import pytest
from VariantValidator import variantValidator as vv

@pytest.mark.skip(reason="Automatically generated tests from LAMP 101")
class TestVariantsAuto2(object):

    @classmethod
    def setup_class(cls):
        vv.my_config()


    def test_variant1(self):
        variant = 'NM_015120.4:c.35T>C'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_015120.4:c.35T>C' in results.keys()
        assert results['NM_015120.4:c.35T>C']['hgvs_lrg_transcript_variant'] == 'LRG_741t1:c.35T>C'
        assert results['NM_015120.4:c.35T>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_015120.4:c.35T>C']['alt_genomic_loci'] == []
        assert results['NM_015120.4:c.35T>C']['gene_symbol'] == 'ALMS1'
        assert results['NM_015120.4:c.35T>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_055935.4(LRG_741p1):p.(Leu12Pro)', 'slr': 'NP_055935.4:p.(L12P)'}
        assert results['NM_015120.4:c.35T>C']['submitted_variant'] == 'NM_015120.4:c.35T>C'
        assert results['NM_015120.4:c.35T>C']['genome_context_intronic_sequence'] == ''
        assert results['NM_015120.4:c.35T>C']['hgvs_lrg_variant'] == 'LRG_741:g.5146T>C'
        assert results['NM_015120.4:c.35T>C']['hgvs_transcript_variant'] == 'NM_015120.4:c.35T>C'
        assert results['NM_015120.4:c.35T>C']['hgvs_refseqgene_variant'] == 'NG_011690.1:g.5146T>C'
        assert results['NM_015120.4:c.35T>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.73613031delinsCGGA', 'vcf': {'chr': 'chr2', 'ref': 'T', 'pos': '73613031', 'alt': 'CGGA'}}
        assert results['NM_015120.4:c.35T>C']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.73385903delinsCGGA', 'vcf': {'chr': 'chr2', 'ref': 'T', 'pos': '73385903', 'alt': 'CGGA'}}
        assert results['NM_015120.4:c.35T>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.73613031delinsCGGA', 'vcf': {'chr': '2', 'ref': 'T', 'pos': '73613031', 'alt': 'CGGA'}}
        assert results['NM_015120.4:c.35T>C']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.73385903delinsCGGA', 'vcf': {'chr': '2', 'ref': 'T', 'pos': '73385903', 'alt': 'CGGA'}}

        assert results['flag'] == 'gene_variant'

    def test_variant2(self):
        variant = 'NM_015120.4:c.39G>C'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_015120.4:c.39G>C' in results.keys()
        assert results['NM_015120.4:c.39G>C']['hgvs_lrg_transcript_variant'] == 'LRG_741t1:c.39G>C'
        assert results['NM_015120.4:c.39G>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_015120.4:c.39G>C']['alt_genomic_loci'] == []
        assert results['NM_015120.4:c.39G>C']['gene_symbol'] == 'ALMS1'
        assert results['NM_015120.4:c.39G>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_055935.4(LRG_741p1):p.(Glu13Asp)', 'slr': 'NP_055935.4:p.(E13D)'}
        assert results['NM_015120.4:c.39G>C']['submitted_variant'] == 'NM_015120.4:c.39G>C'
        assert results['NM_015120.4:c.39G>C']['genome_context_intronic_sequence'] == ''
        assert results['NM_015120.4:c.39G>C']['hgvs_lrg_variant'] == 'LRG_741:g.5150G>C'
        assert results['NM_015120.4:c.39G>C']['hgvs_transcript_variant'] == 'NM_015120.4:c.39G>C'
        assert results['NM_015120.4:c.39G>C']['hgvs_refseqgene_variant'] == 'NG_011690.1:g.5150G>C'
        assert results['NM_015120.4:c.39G>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.73613034_73613035insCGA', 'vcf': {'chr': 'chr2', 'ref': 'G', 'pos': '73613032', 'alt': 'GGAC'}}
        assert results['NM_015120.4:c.39G>C']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.73385906_73385907insCGA', 'vcf': {'chr': 'chr2', 'ref': 'G', 'pos': '73385904', 'alt': 'GGAC'}}
        assert results['NM_015120.4:c.39G>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.73613034_73613035insCGA', 'vcf': {'chr': '2', 'ref': 'G', 'pos': '73613032', 'alt': 'GGAC'}}
        assert results['NM_015120.4:c.39G>C']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.73385906_73385907insCGA', 'vcf': {'chr': '2', 'ref': 'G', 'pos': '73385904', 'alt': 'GGAC'}}


    def test_variant3(self):
        variant = 'NM_015120.4:c.34C>T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_015120.4:c.34C>T' in results.keys()
        assert results['NM_015120.4:c.34C>T']['hgvs_lrg_transcript_variant'] == 'LRG_741t1:c.34C>T'
        assert results['NM_015120.4:c.34C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_015120.4:c.34C>T']['alt_genomic_loci'] == []
        assert results['NM_015120.4:c.34C>T']['gene_symbol'] == 'ALMS1'
        assert results['NM_015120.4:c.34C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_055935.4(LRG_741p1):p.(Leu12=)', 'slr': 'NP_055935.4:p.(L12=)'}
        assert results['NM_015120.4:c.34C>T']['submitted_variant'] == 'NM_015120.4:c.34C>T'
        assert results['NM_015120.4:c.34C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_015120.4:c.34C>T']['hgvs_lrg_variant'] == 'LRG_741:g.5145C>T'
        assert results['NM_015120.4:c.34C>T']['hgvs_transcript_variant'] == 'NM_015120.4:c.34C>T'
        assert results['NM_015120.4:c.34C>T']['hgvs_refseqgene_variant'] == 'NG_011690.1:g.5145C>T'
        assert results['NM_015120.4:c.34C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.73613030C>T', 'vcf': {'chr': 'chr2', 'ref': u'C', 'pos': '73613030', 'alt': u'T'}}
        assert results['NM_015120.4:c.34C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.73385902C>T', 'vcf': {'chr': 'chr2', 'ref': u'C', 'pos': '73385902', 'alt': u'T'}}
        assert results['NM_015120.4:c.34C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.73613030C>T', 'vcf': {'chr': '2', 'ref': u'C', 'pos': '73613030', 'alt': u'T'}}
        assert results['NM_015120.4:c.34C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.73385902C>T', 'vcf': {'chr': '2', 'ref': u'C', 'pos': '73385902', 'alt': u'T'}}


    def test_variant4(self):
        variant = 'NC_000002.11:g.73613030C>T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_015120.4:c.34C>T' in results.keys()
        assert results['NM_015120.4:c.34C>T']['hgvs_lrg_transcript_variant'] == 'LRG_741t1:c.34C>T'
        assert results['NM_015120.4:c.34C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_015120.4:c.34C>T']['alt_genomic_loci'] == []
        assert results['NM_015120.4:c.34C>T']['gene_symbol'] == 'ALMS1'
        assert results['NM_015120.4:c.34C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_055935.4(LRG_741p1):p.(Leu12=)', 'slr': 'NP_055935.4:p.(L12=)'}
        assert results['NM_015120.4:c.34C>T']['submitted_variant'] == 'NC_000002.11:g.73613030C>T'
        assert results['NM_015120.4:c.34C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_015120.4:c.34C>T']['hgvs_lrg_variant'] == 'LRG_741:g.5145C>T'
        assert results['NM_015120.4:c.34C>T']['hgvs_transcript_variant'] == 'NM_015120.4:c.34C>T'
        assert results['NM_015120.4:c.34C>T']['hgvs_refseqgene_variant'] == 'NG_011690.1:g.5145C>T'
        assert results['NM_015120.4:c.34C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.73613030C>T', 'vcf': {'chr': 'chr2', 'ref': u'C', 'pos': '73613030', 'alt': u'T'}}
        assert results['NM_015120.4:c.34C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.73385902C>T', 'vcf': {'chr': 'chr2', 'ref': u'C', 'pos': '73385902', 'alt': u'T'}}
        assert results['NM_015120.4:c.34C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.73613030C>T', 'vcf': {'chr': '2', 'ref': u'C', 'pos': '73613030', 'alt': u'T'}}
        assert results['NM_015120.4:c.34C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.73385902C>T', 'vcf': {'chr': '2', 'ref': u'C', 'pos': '73385902', 'alt': u'T'}}


    def test_variant5(self):
        variant = 'NC_000023.10:g.33229673A>T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000109.3:c.7+127703T>A' in results.keys()
        assert results['NM_000109.3:c.7+127703T>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000109.3:c.7+127703T>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000109.3:c.7+127703T>A']['alt_genomic_loci'] == []
        assert results['NM_000109.3:c.7+127703T>A']['gene_symbol'] == 'DMD'
        assert results['NM_000109.3:c.7+127703T>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000100.2:p.?', 'slr': 'NP_000100.2:p.?'}
        assert results['NM_000109.3:c.7+127703T>A']['submitted_variant'] == 'NC_000023.10:g.33229673A>T'
        assert results['NM_000109.3:c.7+127703T>A']['genome_context_intronic_sequence'] == 'NC_000023.10(NM_000109.3):c.7+127703T>A'
        assert results['NM_000109.3:c.7+127703T>A']['hgvs_lrg_variant'] == ''
        assert results['NM_000109.3:c.7+127703T>A']['hgvs_transcript_variant'] == 'NM_000109.3:c.7+127703T>A'
        assert results['NM_000109.3:c.7+127703T>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_000109.3:c.7+127703T>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.33229673A>T', 'vcf': {'chr': 'chrX', 'ref': u'A', 'pos': '33229673', 'alt': u'T'}}
        assert results['NM_000109.3:c.7+127703T>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.33211556A>T', 'vcf': {'chr': 'chrX', 'ref': u'A', 'pos': '33211556', 'alt': u'T'}}
        assert results['NM_000109.3:c.7+127703T>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.33229673A>T', 'vcf': {'chr': 'X', 'ref': u'A', 'pos': '33229673', 'alt': u'T'}}
        assert results['NM_000109.3:c.7+127703T>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.33211556A>T', 'vcf': {'chr': 'X', 'ref': u'A', 'pos': '33211556', 'alt': u'T'}}

        assert 'NM_004006.2:c.-244T>A' in results.keys()
        assert results['NM_004006.2:c.-244T>A']['hgvs_lrg_transcript_variant'] == 'LRG_199t1:c.-244T>A'
        assert results['NM_004006.2:c.-244T>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_004006.2:c.-244T>A']['alt_genomic_loci'] == []
        assert results['NM_004006.2:c.-244T>A']['gene_symbol'] == 'DMD'
        assert results['NM_004006.2:c.-244T>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003997.1(LRG_199p1):p.?', 'slr': 'NP_003997.1:p.?'}
        assert results['NM_004006.2:c.-244T>A']['submitted_variant'] == 'NC_000023.10:g.33229673A>T'
        assert results['NM_004006.2:c.-244T>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_004006.2:c.-244T>A']['hgvs_lrg_variant'] == 'LRG_199:g.133054T>A'
        assert results['NM_004006.2:c.-244T>A']['hgvs_transcript_variant'] == 'NM_004006.2:c.-244T>A'
        assert results['NM_004006.2:c.-244T>A']['hgvs_refseqgene_variant'] == 'NG_012232.1:g.133054T>A'
        assert results['NM_004006.2:c.-244T>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.33229673A>T', 'vcf': {'chr': 'chrX', 'ref': u'A', 'pos': '33229673', 'alt': u'T'}}
        assert results['NM_004006.2:c.-244T>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.33211556A>T', 'vcf': {'chr': 'chrX', 'ref': u'A', 'pos': '33211556', 'alt': u'T'}}
        assert results['NM_004006.2:c.-244T>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.33229673A>T', 'vcf': {'chr': 'X', 'ref': u'A', 'pos': '33229673', 'alt': u'T'}}
        assert results['NM_004006.2:c.-244T>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.33211556A>T', 'vcf': {'chr': 'X', 'ref': u'A', 'pos': '33211556', 'alt': u'T'}}


    def test_variant6(self):
        variant = 'NM_001145026.1:c.715A>G'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_001145026.1:c.715A>G' in results.keys()
        assert results['NM_001145026.1:c.715A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001145026.1:c.715A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001145026.1:c.715A>G']['alt_genomic_loci'] == []
        assert results['NM_001145026.1:c.715A>G']['gene_symbol'] == 'PTPRQ'
        assert results['NM_001145026.1:c.715A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001138498.1:p.(Arg239Gly)', 'slr': 'NP_001138498.1:p.(R239G)'}
        assert results['NM_001145026.1:c.715A>G']['submitted_variant'] == 'NM_001145026.1:c.715A>G'
        assert results['NM_001145026.1:c.715A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001145026.1:c.715A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001145026.1:c.715A>G']['hgvs_transcript_variant'] == 'NM_001145026.1:c.715A>G'
        assert results['NM_001145026.1:c.715A>G']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['NM_001145026.1:c.715A>G']['primary_assembly_loci'].keys()
        assert results['NM_001145026.1:c.715A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.80460707A>G', 'vcf': {'chr': 'chr12', 'ref': u'A', 'pos': '80460707', 'alt': u'G'}}
        assert 'grch37' not in results['NM_001145026.1:c.715A>G']['primary_assembly_loci'].keys()
        assert results['NM_001145026.1:c.715A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.80460707A>G', 'vcf': {'chr': '12', 'ref': u'A', 'pos': '80460707', 'alt': u'G'}}


    def test_variant7(self):
        variant = 'NC_000016.9:g.2099572TC>T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001077183.2:c.138+821del' in results.keys()
        assert results['NM_001077183.2:c.138+821del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001077183.2:c.138+821del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001077183.2:c.138+821del']['alt_genomic_loci'] == []
        assert results['NM_001077183.2:c.138+821del']['gene_symbol'] == 'TSC2'
        assert results['NM_001077183.2:c.138+821del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001070651.1:p.?', 'slr': 'NP_001070651.1:p.?'}
        assert results['NM_001077183.2:c.138+821del']['submitted_variant'] == 'NC_000016.9:g.2099572TC>T'
        assert results['NM_001077183.2:c.138+821del']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_001077183.2):c.138+821del'
        assert results['NM_001077183.2:c.138+821del']['hgvs_lrg_variant'] == ''
        assert results['NM_001077183.2:c.138+821del']['hgvs_transcript_variant'] == 'NM_001077183.2:c.138+821del'
        assert results['NM_001077183.2:c.138+821del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001077183.2:c.138+821del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.2099575del', 'vcf': {'chr': 'chr16', 'ref': 'TC', 'pos': '2099572', 'alt': 'T'}}
        assert results['NM_001077183.2:c.138+821del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.2049574del', 'vcf': {'chr': 'chr16', 'ref': 'TC', 'pos': '2049571', 'alt': 'T'}}
        assert results['NM_001077183.2:c.138+821del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.2099575del', 'vcf': {'chr': '16', 'ref': 'TC', 'pos': '2099572', 'alt': 'T'}}
        assert results['NM_001077183.2:c.138+821del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.2049574del', 'vcf': {'chr': '16', 'ref': 'TC', 'pos': '2049571', 'alt': 'T'}}

        assert 'NM_001318831.1:c.-89+821del' in results.keys()
        assert results['NM_001318831.1:c.-89+821del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001318831.1:c.-89+821del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001318831.1:c.-89+821del']['alt_genomic_loci'] == []
        assert results['NM_001318831.1:c.-89+821del']['gene_symbol'] == 'TSC2'
        assert results['NM_001318831.1:c.-89+821del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001305760.1:p.?', 'slr': 'NP_001305760.1:p.?'}
        assert results['NM_001318831.1:c.-89+821del']['submitted_variant'] == 'NC_000016.9:g.2099572TC>T'
        assert results['NM_001318831.1:c.-89+821del']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_001318831.1):c.-89+821del'
        assert results['NM_001318831.1:c.-89+821del']['hgvs_lrg_variant'] == ''
        assert results['NM_001318831.1:c.-89+821del']['hgvs_transcript_variant'] == 'NM_001318831.1:c.-89+821del'
        assert results['NM_001318831.1:c.-89+821del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001318831.1:c.-89+821del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.2099575del', 'vcf': {'chr': 'chr16', 'ref': 'TC', 'pos': '2099572', 'alt': 'T'}}
        assert results['NM_001318831.1:c.-89+821del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.2049574del', 'vcf': {'chr': 'chr16', 'ref': 'TC', 'pos': '2049571', 'alt': 'T'}}
        assert results['NM_001318831.1:c.-89+821del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.2099575del', 'vcf': {'chr': '16', 'ref': 'TC', 'pos': '2099572', 'alt': 'T'}}
        assert results['NM_001318831.1:c.-89+821del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.2049574del', 'vcf': {'chr': '16', 'ref': 'TC', 'pos': '2049571', 'alt': 'T'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001318832.1:c.171+821del' in results.keys()
        assert results['NM_001318832.1:c.171+821del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001318832.1:c.171+821del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001318832.1:c.171+821del']['alt_genomic_loci'] == []
        assert results['NM_001318832.1:c.171+821del']['gene_symbol'] == 'TSC2'
        assert results['NM_001318832.1:c.171+821del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001305761.1:p.?', 'slr': 'NP_001305761.1:p.?'}
        assert results['NM_001318832.1:c.171+821del']['submitted_variant'] == 'NC_000016.9:g.2099572TC>T'
        assert results['NM_001318832.1:c.171+821del']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_001318832.1):c.171+821del'
        assert results['NM_001318832.1:c.171+821del']['hgvs_lrg_variant'] == ''
        assert results['NM_001318832.1:c.171+821del']['hgvs_transcript_variant'] == 'NM_001318832.1:c.171+821del'
        assert results['NM_001318832.1:c.171+821del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001318832.1:c.171+821del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.2099575del', 'vcf': {'chr': 'chr16', 'ref': 'TC', 'pos': '2099572', 'alt': 'T'}}
        assert results['NM_001318832.1:c.171+821del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.2049574del', 'vcf': {'chr': 'chr16', 'ref': 'TC', 'pos': '2049571', 'alt': 'T'}}
        assert results['NM_001318832.1:c.171+821del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.2099575del', 'vcf': {'chr': '16', 'ref': 'TC', 'pos': '2099572', 'alt': 'T'}}
        assert results['NM_001318832.1:c.171+821del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.2049574del', 'vcf': {'chr': '16', 'ref': 'TC', 'pos': '2049571', 'alt': 'T'}}

        assert 'NM_001114382.1:c.138+821del' in results.keys()
        assert results['NM_001114382.1:c.138+821del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001114382.1:c.138+821del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001114382.1:c.138+821del']['alt_genomic_loci'] == []
        assert results['NM_001114382.1:c.138+821del']['gene_symbol'] == 'TSC2'
        assert results['NM_001114382.1:c.138+821del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001107854.1:p.?', 'slr': 'NP_001107854.1:p.?'}
        assert results['NM_001114382.1:c.138+821del']['submitted_variant'] == 'NC_000016.9:g.2099572TC>T'
        assert results['NM_001114382.1:c.138+821del']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_001114382.1):c.138+821del'
        assert results['NM_001114382.1:c.138+821del']['hgvs_lrg_variant'] == ''
        assert results['NM_001114382.1:c.138+821del']['hgvs_transcript_variant'] == 'NM_001114382.1:c.138+821del'
        assert results['NM_001114382.1:c.138+821del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001114382.1:c.138+821del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.2099575del', 'vcf': {'chr': 'chr16', 'ref': 'TC', 'pos': '2099572', 'alt': 'T'}}
        assert 'hg38' not in results['NM_001114382.1:c.138+821del']['primary_assembly_loci'].keys()
        assert results['NM_001114382.1:c.138+821del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.2099575del', 'vcf': {'chr': '16', 'ref': 'TC', 'pos': '2099572', 'alt': 'T'}}
        assert 'grch38' not in results['NM_001114382.1:c.138+821del']['primary_assembly_loci'].keys()

        assert 'NM_000548.4:c.138+821del' in results.keys()
        assert results['NM_000548.4:c.138+821del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000548.4:c.138+821del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000548.4:c.138+821del']['alt_genomic_loci'] == []
        assert results['NM_000548.4:c.138+821del']['gene_symbol'] == 'TSC2'
        assert results['NM_000548.4:c.138+821del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000539.2(LRG_487p1):p.?', 'slr': 'NP_000539.2:p.?'}
        assert results['NM_000548.4:c.138+821del']['submitted_variant'] == 'NC_000016.9:g.2099572TC>T'
        assert results['NM_000548.4:c.138+821del']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_000548.4):c.138+821del'
        assert results['NM_000548.4:c.138+821del']['hgvs_lrg_variant'] == ''
        assert results['NM_000548.4:c.138+821del']['hgvs_transcript_variant'] == 'NM_000548.4:c.138+821del'
        assert results['NM_000548.4:c.138+821del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_000548.4:c.138+821del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.2099575del', 'vcf': {'chr': 'chr16', 'ref': 'TC', 'pos': '2099572', 'alt': 'T'}}
        assert results['NM_000548.4:c.138+821del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.2049574del', 'vcf': {'chr': 'chr16', 'ref': 'TC', 'pos': '2049571', 'alt': 'T'}}
        assert results['NM_000548.4:c.138+821del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.2099575del', 'vcf': {'chr': '16', 'ref': 'TC', 'pos': '2099572', 'alt': 'T'}}
        assert results['NM_000548.4:c.138+821del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.2049574del', 'vcf': {'chr': '16', 'ref': 'TC', 'pos': '2049571', 'alt': 'T'}}

        assert 'NM_001363528.1:c.138+821del' in results.keys()
        assert results['NM_001363528.1:c.138+821del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001363528.1:c.138+821del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001363528.1:c.138+821del']['alt_genomic_loci'] == []
        assert results['NM_001363528.1:c.138+821del']['gene_symbol'] == 'TSC2'
        assert results['NM_001363528.1:c.138+821del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001350457.1:p.?', 'slr': 'NP_001350457.1:p.?'}
        assert results['NM_001363528.1:c.138+821del']['submitted_variant'] == 'NC_000016.9:g.2099572TC>T'
        assert results['NM_001363528.1:c.138+821del']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_001363528.1):c.138+821del'
        assert results['NM_001363528.1:c.138+821del']['hgvs_lrg_variant'] == ''
        assert results['NM_001363528.1:c.138+821del']['hgvs_transcript_variant'] == 'NM_001363528.1:c.138+821del'
        assert results['NM_001363528.1:c.138+821del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001363528.1:c.138+821del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.2099575del', 'vcf': {'chr': 'chr16', 'ref': 'TC', 'pos': '2099572', 'alt': 'T'}}
        assert 'hg38' not in results['NM_001363528.1:c.138+821del']['primary_assembly_loci'].keys()
        assert results['NM_001363528.1:c.138+821del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.2099575del', 'vcf': {'chr': '16', 'ref': 'TC', 'pos': '2099572', 'alt': 'T'}}
        assert 'grch38' not in results['NM_001363528.1:c.138+821del']['primary_assembly_loci'].keys()

        assert 'NM_000548.3:c.138+821del' in results.keys()
        assert results['NM_000548.3:c.138+821del']['hgvs_lrg_transcript_variant'] == 'LRG_487t1:c.138+821del'
        assert results['NM_000548.3:c.138+821del']['refseqgene_context_intronic_sequence'] == 'NG_005895.1(NM_000548.3):c.138+821del'
        assert results['NM_000548.3:c.138+821del']['alt_genomic_loci'] == []
        assert results['NM_000548.3:c.138+821del']['gene_symbol'] == 'TSC2'
        assert results['NM_000548.3:c.138+821del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000539.2(LRG_487p1):p.?', 'slr': 'NP_000539.2:p.?'}
        assert results['NM_000548.3:c.138+821del']['submitted_variant'] == 'NC_000016.9:g.2099572TC>T'
        assert results['NM_000548.3:c.138+821del']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_000548.3):c.138+821del'
        assert results['NM_000548.3:c.138+821del']['hgvs_lrg_variant'] == 'LRG_487:g.5269del'
        assert results['NM_000548.3:c.138+821del']['hgvs_transcript_variant'] == 'NM_000548.3:c.138+821del'
        assert results['NM_000548.3:c.138+821del']['hgvs_refseqgene_variant'] == 'NG_005895.1:g.5269del'
        assert results['NM_000548.3:c.138+821del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.2099575del', 'vcf': {'chr': 'chr16', 'ref': 'TC', 'pos': '2099572', 'alt': 'T'}}
        assert 'hg38' not in results['NM_000548.3:c.138+821del']['primary_assembly_loci'].keys()
        assert results['NM_000548.3:c.138+821del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.2099575del', 'vcf': {'chr': '16', 'ref': 'TC', 'pos': '2099572', 'alt': 'T'}}
        assert 'grch38' not in results['NM_000548.3:c.138+821del']['primary_assembly_loci'].keys()

        assert 'NM_021055.2:c.138+821del' in results.keys()
        assert results['NM_021055.2:c.138+821del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_021055.2:c.138+821del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_021055.2:c.138+821del']['alt_genomic_loci'] == []
        assert results['NM_021055.2:c.138+821del']['gene_symbol'] == 'TSC2'
        assert results['NM_021055.2:c.138+821del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_066399.2:p.?', 'slr': 'NP_066399.2:p.?'}
        assert results['NM_021055.2:c.138+821del']['submitted_variant'] == 'NC_000016.9:g.2099572TC>T'
        assert results['NM_021055.2:c.138+821del']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_021055.2):c.138+821del'
        assert results['NM_021055.2:c.138+821del']['hgvs_lrg_variant'] == ''
        assert results['NM_021055.2:c.138+821del']['hgvs_transcript_variant'] == 'NM_021055.2:c.138+821del'
        assert results['NM_021055.2:c.138+821del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_021055.2:c.138+821del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.2099575del', 'vcf': {'chr': 'chr16', 'ref': 'TC', 'pos': '2099572', 'alt': 'T'}}
        assert 'hg38' not in results['NM_021055.2:c.138+821del']['primary_assembly_loci'].keys()
        assert results['NM_021055.2:c.138+821del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.2099575del', 'vcf': {'chr': '16', 'ref': 'TC', 'pos': '2099572', 'alt': 'T'}}
        assert 'grch38' not in results['NM_021055.2:c.138+821del']['primary_assembly_loci'].keys()

        assert 'NM_001077183.1:c.138+821del' in results.keys()
        assert results['NM_001077183.1:c.138+821del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001077183.1:c.138+821del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001077183.1:c.138+821del']['alt_genomic_loci'] == []
        assert results['NM_001077183.1:c.138+821del']['gene_symbol'] == 'TSC2'
        assert results['NM_001077183.1:c.138+821del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001070651.1:p.?', 'slr': 'NP_001070651.1:p.?'}
        assert results['NM_001077183.1:c.138+821del']['submitted_variant'] == 'NC_000016.9:g.2099572TC>T'
        assert results['NM_001077183.1:c.138+821del']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_001077183.1):c.138+821del'
        assert results['NM_001077183.1:c.138+821del']['hgvs_lrg_variant'] == ''
        assert results['NM_001077183.1:c.138+821del']['hgvs_transcript_variant'] == 'NM_001077183.1:c.138+821del'
        assert results['NM_001077183.1:c.138+821del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001077183.1:c.138+821del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.2099575del', 'vcf': {'chr': 'chr16', 'ref': 'TC', 'pos': '2099572', 'alt': 'T'}}
        assert 'hg38' not in results['NM_001077183.1:c.138+821del']['primary_assembly_loci'].keys()
        assert results['NM_001077183.1:c.138+821del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.2099575del', 'vcf': {'chr': '16', 'ref': 'TC', 'pos': '2099572', 'alt': 'T'}}
        assert 'grch38' not in results['NM_001077183.1:c.138+821del']['primary_assembly_loci'].keys()

        assert 'NM_001318827.1:c.138+821del' in results.keys()
        assert results['NM_001318827.1:c.138+821del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001318827.1:c.138+821del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001318827.1:c.138+821del']['alt_genomic_loci'] == []
        assert results['NM_001318827.1:c.138+821del']['gene_symbol'] == 'TSC2'
        assert results['NM_001318827.1:c.138+821del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001305756.1:p.?', 'slr': 'NP_001305756.1:p.?'}
        assert results['NM_001318827.1:c.138+821del']['submitted_variant'] == 'NC_000016.9:g.2099572TC>T'
        assert results['NM_001318827.1:c.138+821del']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_001318827.1):c.138+821del'
        assert results['NM_001318827.1:c.138+821del']['hgvs_lrg_variant'] == ''
        assert results['NM_001318827.1:c.138+821del']['hgvs_transcript_variant'] == 'NM_001318827.1:c.138+821del'
        assert results['NM_001318827.1:c.138+821del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001318827.1:c.138+821del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.2099575del', 'vcf': {'chr': 'chr16', 'ref': 'TC', 'pos': '2099572', 'alt': 'T'}}
        assert results['NM_001318827.1:c.138+821del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.2049574del', 'vcf': {'chr': 'chr16', 'ref': 'TC', 'pos': '2049571', 'alt': 'T'}}
        assert results['NM_001318827.1:c.138+821del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.2099575del', 'vcf': {'chr': '16', 'ref': 'TC', 'pos': '2099572', 'alt': 'T'}}
        assert results['NM_001318827.1:c.138+821del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.2049574del', 'vcf': {'chr': '16', 'ref': 'TC', 'pos': '2049571', 'alt': 'T'}}

        assert 'NM_001114382.2:c.138+821del' in results.keys()
        assert results['NM_001114382.2:c.138+821del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001114382.2:c.138+821del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001114382.2:c.138+821del']['alt_genomic_loci'] == []
        assert results['NM_001114382.2:c.138+821del']['gene_symbol'] == 'TSC2'
        assert results['NM_001114382.2:c.138+821del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001107854.1:p.?', 'slr': 'NP_001107854.1:p.?'}
        assert results['NM_001114382.2:c.138+821del']['submitted_variant'] == 'NC_000016.9:g.2099572TC>T'
        assert results['NM_001114382.2:c.138+821del']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_001114382.2):c.138+821del'
        assert results['NM_001114382.2:c.138+821del']['hgvs_lrg_variant'] == ''
        assert results['NM_001114382.2:c.138+821del']['hgvs_transcript_variant'] == 'NM_001114382.2:c.138+821del'
        assert results['NM_001114382.2:c.138+821del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001114382.2:c.138+821del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.2099575del', 'vcf': {'chr': 'chr16', 'ref': 'TC', 'pos': '2099572', 'alt': 'T'}}
        assert results['NM_001114382.2:c.138+821del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.2049574del', 'vcf': {'chr': 'chr16', 'ref': 'TC', 'pos': '2049571', 'alt': 'T'}}
        assert results['NM_001114382.2:c.138+821del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.2099575del', 'vcf': {'chr': '16', 'ref': 'TC', 'pos': '2099572', 'alt': 'T'}}
        assert results['NM_001114382.2:c.138+821del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.2049574del', 'vcf': {'chr': '16', 'ref': 'TC', 'pos': '2049571', 'alt': 'T'}}

        assert 'NM_001318829.1:c.-9-826del' in results.keys()
        assert results['NM_001318829.1:c.-9-826del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001318829.1:c.-9-826del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001318829.1:c.-9-826del']['alt_genomic_loci'] == []
        assert results['NM_001318829.1:c.-9-826del']['gene_symbol'] == 'TSC2'
        assert results['NM_001318829.1:c.-9-826del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001305758.1:p.?', 'slr': 'NP_001305758.1:p.?'}
        assert results['NM_001318829.1:c.-9-826del']['submitted_variant'] == 'NC_000016.9:g.2099572TC>T'
        assert results['NM_001318829.1:c.-9-826del']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_001318829.1):c.-9-826del'
        assert results['NM_001318829.1:c.-9-826del']['hgvs_lrg_variant'] == ''
        assert results['NM_001318829.1:c.-9-826del']['hgvs_transcript_variant'] == 'NM_001318829.1:c.-9-826del'
        assert results['NM_001318829.1:c.-9-826del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001318829.1:c.-9-826del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.2099575del', 'vcf': {'chr': 'chr16', 'ref': 'TC', 'pos': '2099572', 'alt': 'T'}}
        assert results['NM_001318829.1:c.-9-826del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.2049574del', 'vcf': {'chr': 'chr16', 'ref': 'TC', 'pos': '2049571', 'alt': 'T'}}
        assert results['NM_001318829.1:c.-9-826del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.2099575del', 'vcf': {'chr': '16', 'ref': 'TC', 'pos': '2099572', 'alt': 'T'}}
        assert results['NM_001318829.1:c.-9-826del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.2049574del', 'vcf': {'chr': '16', 'ref': 'TC', 'pos': '2049571', 'alt': 'T'}}


    def test_variant8(self):
        variant = 'NM_000088.3:c.589GG>CT'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000088.3:c.589_590delinsCT' in results.keys()
        assert results['NM_000088.3:c.589_590delinsCT']['hgvs_lrg_transcript_variant'] == 'LRG_1t1:c.589_590delinsCT'
        assert results['NM_000088.3:c.589_590delinsCT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000088.3:c.589_590delinsCT']['alt_genomic_loci'] == []
        assert results['NM_000088.3:c.589_590delinsCT']['gene_symbol'] == 'COL1A1'
        assert results['NM_000088.3:c.589_590delinsCT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000079.2(LRG_1p1):p.(Gly197Leu)', 'slr': 'NP_000079.2:p.(G197L)'}
        assert results['NM_000088.3:c.589_590delinsCT']['submitted_variant'] == 'NM_000088.3:c.589GG>CT'
        assert results['NM_000088.3:c.589_590delinsCT']['genome_context_intronic_sequence'] == ''
        assert results['NM_000088.3:c.589_590delinsCT']['hgvs_lrg_variant'] == 'LRG_1:g.8638_8639delinsCT'
        assert results['NM_000088.3:c.589_590delinsCT']['hgvs_transcript_variant'] == 'NM_000088.3:c.589_590delinsCT'
        assert results['NM_000088.3:c.589_590delinsCT']['hgvs_refseqgene_variant'] == 'NG_007400.1:g.8638_8639delinsCT'
        assert results['NM_000088.3:c.589_590delinsCT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275362_48275363delinsAG', 'vcf': {'chr': 'chr17', 'ref': 'CC', 'pos': '48275362', 'alt': u'AG'}}
        assert results['NM_000088.3:c.589_590delinsCT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198001_50198002delinsAG', 'vcf': {'chr': 'chr17', 'ref': 'CC', 'pos': '50198001', 'alt': u'AG'}}
        assert results['NM_000088.3:c.589_590delinsCT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275362_48275363delinsAG', 'vcf': {'chr': '17', 'ref': 'CC', 'pos': '48275362', 'alt': u'AG'}}
        assert results['NM_000088.3:c.589_590delinsCT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198001_50198002delinsAG', 'vcf': {'chr': '17', 'ref': 'CC', 'pos': '50198001', 'alt': u'AG'}}


    def test_variant9(self):
        variant = 'NM_000094.3:c.6751-2_6751-3del'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'validation_warning_1' in results.keys()
        assert results['validation_warning_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['validation_warning_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['alt_genomic_loci'] == []
        assert results['validation_warning_1']['gene_symbol'] == ''
        assert results['validation_warning_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['validation_warning_1']['submitted_variant'] == 'NM_000094.3:c.6751-2_6751-3del'
        assert results['validation_warning_1']['genome_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['hgvs_lrg_variant'] == ''
        assert results['validation_warning_1']['hgvs_transcript_variant'] == ''
        assert results['validation_warning_1']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['validation_warning_1']['primary_assembly_loci'].keys()

        assert results['flag'] == 'warning'

    def test_variant10(self):
        variant = 'COL5A1:c.5071A>T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'validation_warning_1' in results.keys()
        assert results['validation_warning_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['validation_warning_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['alt_genomic_loci'] == []
        assert results['validation_warning_1']['gene_symbol'] == ''
        assert results['validation_warning_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['validation_warning_1']['submitted_variant'] == 'COL5A1:c.5071A>T'
        assert results['validation_warning_1']['genome_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['hgvs_lrg_variant'] == ''
        assert results['validation_warning_1']['hgvs_transcript_variant'] == ''
        assert results['validation_warning_1']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['validation_warning_1']['primary_assembly_loci'].keys()

        assert results['flag'] == 'warning'

    def test_variant11(self):
        variant = 'NG_007400.1:c.5071A>T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'validation_warning_1' in results.keys()
        assert results['validation_warning_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['validation_warning_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['alt_genomic_loci'] == []
        assert results['validation_warning_1']['gene_symbol'] == ''
        assert results['validation_warning_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['validation_warning_1']['submitted_variant'] == 'NG_007400.1:c.5071A>T'
        assert results['validation_warning_1']['genome_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['hgvs_lrg_variant'] == ''
        assert results['validation_warning_1']['hgvs_transcript_variant'] == ''
        assert results['validation_warning_1']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['validation_warning_1']['primary_assembly_loci'].keys()

        assert results['flag'] == 'warning'

    def test_variant12(self):
        variant = 'chr16:15832508_15832509delinsAC'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001040113.1:c.3055_3056inv' in results.keys()
        assert results['NM_001040113.1:c.3055_3056inv']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001040113.1:c.3055_3056inv']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001040113.1:c.3055_3056inv']['alt_genomic_loci'] == [{'grch38': {'hgvs_genomic_description': 'NT_187607.1:g.1396662_1396663inv', 'vcf': {'chr': 'HSCHR16_1_CTG1', 'ref': 'GT', 'pos': '1396662', 'alt': 'AC'}}}, {'hg38': {'hgvs_genomic_description': 'NT_187607.1:g.1396662_1396663inv', 'vcf': {'chr': 'chr16_KI270853v1_alt', 'ref': 'GT', 'pos': '1396662', 'alt': 'AC'}}}]
        assert results['NM_001040113.1:c.3055_3056inv']['gene_symbol'] == 'MYH11'
        assert results['NM_001040113.1:c.3055_3056inv']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001035202.1:p.(Thr1019Val)', 'slr': 'NP_001035202.1:p.(T1019V)'}
        assert results['NM_001040113.1:c.3055_3056inv']['submitted_variant'] == 'chr16:15832508_15832509delinsAC'
        assert results['NM_001040113.1:c.3055_3056inv']['genome_context_intronic_sequence'] == ''
        assert results['NM_001040113.1:c.3055_3056inv']['hgvs_lrg_variant'] == ''
        assert results['NM_001040113.1:c.3055_3056inv']['hgvs_transcript_variant'] == 'NM_001040113.1:c.3055_3056inv'
        assert results['NM_001040113.1:c.3055_3056inv']['hgvs_refseqgene_variant'] == 'NG_009299.1:g.123379_123380inv'
        assert results['NM_001040113.1:c.3055_3056inv']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.15832508_15832509inv', 'vcf': {'chr': 'chr16', 'ref': 'GT', 'pos': '15832508', 'alt': 'AC'}}
        assert results['NM_001040113.1:c.3055_3056inv']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.15738651_15738652inv', 'vcf': {'chr': 'chr16', 'ref': 'GT', 'pos': '15738651', 'alt': 'AC'}}
        assert results['NM_001040113.1:c.3055_3056inv']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.15832508_15832509inv', 'vcf': {'chr': '16', 'ref': 'GT', 'pos': '15832508', 'alt': 'AC'}}
        assert results['NM_001040113.1:c.3055_3056inv']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.15738651_15738652inv', 'vcf': {'chr': '16', 'ref': 'GT', 'pos': '15738651', 'alt': 'AC'}}

        assert 'NM_001040114.1:c.3055_3056inv' in results.keys()
        assert results['NM_001040114.1:c.3055_3056inv']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001040114.1:c.3055_3056inv']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001040114.1:c.3055_3056inv']['alt_genomic_loci'] == [{'grch38': {'hgvs_genomic_description': 'NT_187607.1:g.1396662_1396663inv', 'vcf': {'chr': 'HSCHR16_1_CTG1', 'ref': 'GT', 'pos': '1396662', 'alt': 'AC'}}}, {'hg38': {'hgvs_genomic_description': 'NT_187607.1:g.1396662_1396663inv', 'vcf': {'chr': 'chr16_KI270853v1_alt', 'ref': 'GT', 'pos': '1396662', 'alt': 'AC'}}}]
        assert results['NM_001040114.1:c.3055_3056inv']['gene_symbol'] == 'MYH11'
        assert results['NM_001040114.1:c.3055_3056inv']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001035203.1:p.(Thr1019Val)', 'slr': 'NP_001035203.1:p.(T1019V)'}
        assert results['NM_001040114.1:c.3055_3056inv']['submitted_variant'] == 'chr16:15832508_15832509delinsAC'
        assert results['NM_001040114.1:c.3055_3056inv']['genome_context_intronic_sequence'] == ''
        assert results['NM_001040114.1:c.3055_3056inv']['hgvs_lrg_variant'] == ''
        assert results['NM_001040114.1:c.3055_3056inv']['hgvs_transcript_variant'] == 'NM_001040114.1:c.3055_3056inv'
        assert results['NM_001040114.1:c.3055_3056inv']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001040114.1:c.3055_3056inv']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.15832508_15832509inv', 'vcf': {'chr': 'chr16', 'ref': 'GT', 'pos': '15832508', 'alt': 'AC'}}
        assert results['NM_001040114.1:c.3055_3056inv']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.15738651_15738652inv', 'vcf': {'chr': 'chr16', 'ref': 'GT', 'pos': '15738651', 'alt': 'AC'}}
        assert results['NM_001040114.1:c.3055_3056inv']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.15832508_15832509inv', 'vcf': {'chr': '16', 'ref': 'GT', 'pos': '15832508', 'alt': 'AC'}}
        assert results['NM_001040114.1:c.3055_3056inv']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.15738651_15738652inv', 'vcf': {'chr': '16', 'ref': 'GT', 'pos': '15738651', 'alt': 'AC'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_002474.2:c.3034_3035inv' in results.keys()
        assert results['NM_002474.2:c.3034_3035inv']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_002474.2:c.3034_3035inv']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_002474.2:c.3034_3035inv']['alt_genomic_loci'] == [{'grch38': {'hgvs_genomic_description': 'NT_187607.1:g.1396662_1396663inv', 'vcf': {'chr': 'HSCHR16_1_CTG1', 'ref': 'GT', 'pos': '1396662', 'alt': 'AC'}}}, {'hg38': {'hgvs_genomic_description': 'NT_187607.1:g.1396662_1396663inv', 'vcf': {'chr': 'chr16_KI270853v1_alt', 'ref': 'GT', 'pos': '1396662', 'alt': 'AC'}}}]
        assert results['NM_002474.2:c.3034_3035inv']['gene_symbol'] == 'MYH11'
        assert results['NM_002474.2:c.3034_3035inv']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_002465.1:p.(Thr1012Val)', 'slr': 'NP_002465.1:p.(T1012V)'}
        assert results['NM_002474.2:c.3034_3035inv']['submitted_variant'] == 'chr16:15832508_15832509delinsAC'
        assert results['NM_002474.2:c.3034_3035inv']['genome_context_intronic_sequence'] == ''
        assert results['NM_002474.2:c.3034_3035inv']['hgvs_lrg_variant'] == ''
        assert results['NM_002474.2:c.3034_3035inv']['hgvs_transcript_variant'] == 'NM_002474.2:c.3034_3035inv'
        assert results['NM_002474.2:c.3034_3035inv']['hgvs_refseqgene_variant'] == ''
        assert results['NM_002474.2:c.3034_3035inv']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.15832508_15832509inv', 'vcf': {'chr': 'chr16', 'ref': 'GT', 'pos': '15832508', 'alt': 'AC'}}
        assert results['NM_002474.2:c.3034_3035inv']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.15738651_15738652inv', 'vcf': {'chr': 'chr16', 'ref': 'GT', 'pos': '15738651', 'alt': 'AC'}}
        assert results['NM_002474.2:c.3034_3035inv']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.15832508_15832509inv', 'vcf': {'chr': '16', 'ref': 'GT', 'pos': '15832508', 'alt': 'AC'}}
        assert results['NM_002474.2:c.3034_3035inv']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.15738651_15738652inv', 'vcf': {'chr': '16', 'ref': 'GT', 'pos': '15738651', 'alt': 'AC'}}

        assert 'NM_022844.2:c.3034_3035inv' in results.keys()
        assert results['NM_022844.2:c.3034_3035inv']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_022844.2:c.3034_3035inv']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_022844.2:c.3034_3035inv']['alt_genomic_loci'] == [{'grch38': {'hgvs_genomic_description': 'NT_187607.1:g.1396662_1396663inv', 'vcf': {'chr': 'HSCHR16_1_CTG1', 'ref': 'GT', 'pos': '1396662', 'alt': 'AC'}}}, {'hg38': {'hgvs_genomic_description': 'NT_187607.1:g.1396662_1396663inv', 'vcf': {'chr': 'chr16_KI270853v1_alt', 'ref': 'GT', 'pos': '1396662', 'alt': 'AC'}}}]
        assert results['NM_022844.2:c.3034_3035inv']['gene_symbol'] == 'MYH11'
        assert results['NM_022844.2:c.3034_3035inv']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_074035.1:p.(Thr1012Val)', 'slr': 'NP_074035.1:p.(T1012V)'}
        assert results['NM_022844.2:c.3034_3035inv']['submitted_variant'] == 'chr16:15832508_15832509delinsAC'
        assert results['NM_022844.2:c.3034_3035inv']['genome_context_intronic_sequence'] == ''
        assert results['NM_022844.2:c.3034_3035inv']['hgvs_lrg_variant'] == ''
        assert results['NM_022844.2:c.3034_3035inv']['hgvs_transcript_variant'] == 'NM_022844.2:c.3034_3035inv'
        assert results['NM_022844.2:c.3034_3035inv']['hgvs_refseqgene_variant'] == ''
        assert results['NM_022844.2:c.3034_3035inv']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.15832508_15832509inv', 'vcf': {'chr': 'chr16', 'ref': 'GT', 'pos': '15832508', 'alt': 'AC'}}
        assert results['NM_022844.2:c.3034_3035inv']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.15738651_15738652inv', 'vcf': {'chr': 'chr16', 'ref': 'GT', 'pos': '15738651', 'alt': 'AC'}}
        assert results['NM_022844.2:c.3034_3035inv']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.15832508_15832509inv', 'vcf': {'chr': '16', 'ref': 'GT', 'pos': '15832508', 'alt': 'AC'}}
        assert results['NM_022844.2:c.3034_3035inv']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.15738651_15738652inv', 'vcf': {'chr': '16', 'ref': 'GT', 'pos': '15738651', 'alt': 'AC'}}


    def test_variant13(self):
        variant = 'NM_000088.3:c.589-1GG>G'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000088.3:c.589-1_589delinsG' in results.keys()
        assert results['NM_000088.3:c.589-1_589delinsG']['hgvs_lrg_transcript_variant'] == 'LRG_1t1:c.590del'
        assert results['NM_000088.3:c.589-1_589delinsG']['refseqgene_context_intronic_sequence'] == 'NG_007400.1(NM_000088.3):c.590del'
        assert results['NM_000088.3:c.589-1_589delinsG']['alt_genomic_loci'] == []
        assert results['NM_000088.3:c.589-1_589delinsG']['gene_symbol'] == 'COL1A1'
        assert results['NM_000088.3:c.589-1_589delinsG']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000079.2(LRG_1p1):p.?', 'slr': 'NP_000079.2:p.?'}
        assert results['NM_000088.3:c.589-1_589delinsG']['submitted_variant'] == 'NM_000088.3:c.589-1GG>G'
        assert results['NM_000088.3:c.589-1_589delinsG']['genome_context_intronic_sequence'] == 'NC_000017.10(NM_000088.3):c.589-1_589delinsG'
        assert results['NM_000088.3:c.589-1_589delinsG']['hgvs_lrg_variant'] == 'LRG_1:g.8639del'
        assert results['NM_000088.3:c.589-1_589delinsG']['hgvs_transcript_variant'] == 'NM_000088.3:c.589-1_589delinsG'
        assert results['NM_000088.3:c.589-1_589delinsG']['hgvs_refseqgene_variant'] == 'NG_007400.1:g.8639del'
        assert results['NM_000088.3:c.589-1_589delinsG']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275363_48275364delinsC', 'vcf': {'chr': 'chr17', 'ref': 'AC', 'pos': '48275361', 'alt': 'A'}}
        assert results['NM_000088.3:c.589-1_589delinsG']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198002_50198003delinsC', 'vcf': {'chr': 'chr17', 'ref': 'AC', 'pos': '50198000', 'alt': 'A'}}
        assert results['NM_000088.3:c.589-1_589delinsG']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275363_48275364delinsC', 'vcf': {'chr': '17', 'ref': 'AC', 'pos': '48275361', 'alt': 'A'}}
        assert results['NM_000088.3:c.589-1_589delinsG']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198002_50198003delinsC', 'vcf': {'chr': '17', 'ref': 'AC', 'pos': '50198000', 'alt': 'A'}}


    def test_variant14(self):
        variant = 'NM_000088.3:c.642+1GT>G'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000088.3:c.642+1_642+2delinsG' in results.keys()
        assert results['NM_000088.3:c.642+1_642+2delinsG']['hgvs_lrg_transcript_variant'] == 'LRG_1t1:c.642+2del'
        assert results['NM_000088.3:c.642+1_642+2delinsG']['refseqgene_context_intronic_sequence'] == 'NG_007400.1(NM_000088.3):c.642+2del'
        assert results['NM_000088.3:c.642+1_642+2delinsG']['alt_genomic_loci'] == []
        assert results['NM_000088.3:c.642+1_642+2delinsG']['gene_symbol'] == 'COL1A1'
        assert results['NM_000088.3:c.642+1_642+2delinsG']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000079.2(LRG_1p1):p.?', 'slr': 'NP_000079.2:p.?'}
        assert results['NM_000088.3:c.642+1_642+2delinsG']['submitted_variant'] == 'NM_000088.3:c.642+1GT>G'
        assert results['NM_000088.3:c.642+1_642+2delinsG']['genome_context_intronic_sequence'] == 'NC_000017.10(NM_000088.3):c.642+1_642+2delinsG'
        assert results['NM_000088.3:c.642+1_642+2delinsG']['hgvs_lrg_variant'] == 'LRG_1:g.8693del'
        assert results['NM_000088.3:c.642+1_642+2delinsG']['hgvs_transcript_variant'] == 'NM_000088.3:c.642+1_642+2delinsG'
        assert results['NM_000088.3:c.642+1_642+2delinsG']['hgvs_refseqgene_variant'] == 'NG_007400.1:g.8693del'
        assert results['NM_000088.3:c.642+1_642+2delinsG']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275308_48275309delinsC', 'vcf': {'chr': 'chr17', 'ref': 'TA', 'pos': '48275307', 'alt': 'T'}}
        assert results['NM_000088.3:c.642+1_642+2delinsG']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50197947_50197948delinsC', 'vcf': {'chr': 'chr17', 'ref': 'TA', 'pos': '50197946', 'alt': 'T'}}
        assert results['NM_000088.3:c.642+1_642+2delinsG']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275308_48275309delinsC', 'vcf': {'chr': '17', 'ref': 'TA', 'pos': '48275307', 'alt': 'T'}}
        assert results['NM_000088.3:c.642+1_642+2delinsG']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50197947_50197948delinsC', 'vcf': {'chr': '17', 'ref': 'TA', 'pos': '50197946', 'alt': 'T'}}


    def test_variant15(self):
        variant = 'NM_000088.3:c.589-2AG>G'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000088.3:c.589-2_589-1delinsG' in results.keys()
        assert results['NM_000088.3:c.589-2_589-1delinsG']['hgvs_lrg_transcript_variant'] == 'LRG_1t1:c.589-2del'
        assert results['NM_000088.3:c.589-2_589-1delinsG']['refseqgene_context_intronic_sequence'] == 'NG_007400.1(NM_000088.3):c.589-2del'
        assert results['NM_000088.3:c.589-2_589-1delinsG']['alt_genomic_loci'] == []
        assert results['NM_000088.3:c.589-2_589-1delinsG']['gene_symbol'] == 'COL1A1'
        assert results['NM_000088.3:c.589-2_589-1delinsG']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000079.2(LRG_1p1):p.?', 'slr': 'NP_000079.2:p.?'}
        assert results['NM_000088.3:c.589-2_589-1delinsG']['submitted_variant'] == 'NM_000088.3:c.589-2AG>G'
        assert results['NM_000088.3:c.589-2_589-1delinsG']['genome_context_intronic_sequence'] == 'NC_000017.10(NM_000088.3):c.589-2_589-1delinsG'
        assert results['NM_000088.3:c.589-2_589-1delinsG']['hgvs_lrg_variant'] == 'LRG_1:g.8636del'
        assert results['NM_000088.3:c.589-2_589-1delinsG']['hgvs_transcript_variant'] == 'NM_000088.3:c.589-2_589-1delinsG'
        assert results['NM_000088.3:c.589-2_589-1delinsG']['hgvs_refseqgene_variant'] == 'NG_007400.1:g.8636del'
        assert results['NM_000088.3:c.589-2_589-1delinsG']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275364_48275365delinsC', 'vcf': {'chr': 'chr17', 'ref': 'CT', 'pos': '48275364', 'alt': 'C'}}
        assert results['NM_000088.3:c.589-2_589-1delinsG']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198003_50198004delinsC', 'vcf': {'chr': 'chr17', 'ref': 'CT', 'pos': '50198003', 'alt': 'C'}}
        assert results['NM_000088.3:c.589-2_589-1delinsG']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275364_48275365delinsC', 'vcf': {'chr': '17', 'ref': 'CT', 'pos': '48275364', 'alt': 'C'}}
        assert results['NM_000088.3:c.589-2_589-1delinsG']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198003_50198004delinsC', 'vcf': {'chr': '17', 'ref': 'CT', 'pos': '50198003', 'alt': 'C'}}


    def test_variant16(self):
        variant = 'NC_000017.10:g.48279242G>T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'intergenic'
        assert 'Intergenic_Variant_1' in results.keys()
        assert results['Intergenic_Variant_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['Intergenic_Variant_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['Intergenic_Variant_1']['alt_genomic_loci'] == []
        assert results['Intergenic_Variant_1']['gene_symbol'] == ''
        assert results['Intergenic_Variant_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['Intergenic_Variant_1']['submitted_variant'] == 'NC_000017.10:g.48279242G>T'
        assert results['Intergenic_Variant_1']['genome_context_intronic_sequence'] == ''
        assert results['Intergenic_Variant_1']['hgvs_lrg_variant'] == 'LRG_1:g.4759C>A'
        assert results['Intergenic_Variant_1']['hgvs_transcript_variant'] == ''
        assert results['Intergenic_Variant_1']['hgvs_refseqgene_variant'] == 'NG_007400.1:g.4759C>A'
        assert results['Intergenic_Variant_1']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48279242G>T', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '48279242', 'alt': 'T'}}
        assert results['Intergenic_Variant_1']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50201881G>T', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '50201881', 'alt': 'T'}}
        assert results['Intergenic_Variant_1']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48279242G>T', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '48279242', 'alt': 'T'}}
        assert results['Intergenic_Variant_1']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50201881G>T', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '50201881', 'alt': 'T'}}


    def test_variant17(self):
        variant = 'NM_000500.7:c.-107-19C>T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'validation_warning_1' in results.keys()
        assert results['validation_warning_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['validation_warning_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['alt_genomic_loci'] == []
        assert results['validation_warning_1']['gene_symbol'] == ''
        assert results['validation_warning_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['validation_warning_1']['submitted_variant'] == 'NM_000500.7:c.-107-19C>T'
        assert results['validation_warning_1']['genome_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['hgvs_lrg_variant'] == ''
        assert results['validation_warning_1']['hgvs_transcript_variant'] == ''
        assert results['validation_warning_1']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['validation_warning_1']['primary_assembly_loci'].keys()

        assert results['flag'] == 'warning'

    def test_variant18(self):
        variant = 'NM_000518.4:c.-130C>T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'validation_warning_1' in results.keys()
        assert results['validation_warning_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['validation_warning_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['alt_genomic_loci'] == []
        assert results['validation_warning_1']['gene_symbol'] == ''
        assert results['validation_warning_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['validation_warning_1']['submitted_variant'] == 'NM_000518.4:c.-130C>T'
        assert results['validation_warning_1']['genome_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['hgvs_lrg_variant'] == ''
        assert results['validation_warning_1']['hgvs_transcript_variant'] == ''
        assert results['validation_warning_1']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['validation_warning_1']['primary_assembly_loci'].keys()

        assert results['flag'] == 'warning'

    def test_variant19(self):
        variant = 'NM_000518.4:c.-50-80C>T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'validation_warning_1' in results.keys()
        assert results['validation_warning_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['validation_warning_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['alt_genomic_loci'] == []
        assert results['validation_warning_1']['gene_symbol'] == ''
        assert results['validation_warning_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['validation_warning_1']['submitted_variant'] == 'NM_000518.4:c.-50-80C>T'
        assert results['validation_warning_1']['genome_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['hgvs_lrg_variant'] == ''
        assert results['validation_warning_1']['hgvs_transcript_variant'] == ''
        assert results['validation_warning_1']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['validation_warning_1']['primary_assembly_loci'].keys()

        assert results['flag'] == 'warning'

    def test_variant20(self):
        variant = 'NM_000518.4:c.316_*342delinsCTACTT'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'validation_warning_1' in results.keys()
        assert results['validation_warning_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['validation_warning_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['alt_genomic_loci'] == []
        assert results['validation_warning_1']['gene_symbol'] == ''
        assert results['validation_warning_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['validation_warning_1']['submitted_variant'] == 'NM_000518.4:c.316_*342delinsCTACTT'
        assert results['validation_warning_1']['genome_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['hgvs_lrg_variant'] == ''
        assert results['validation_warning_1']['hgvs_transcript_variant'] == ''
        assert results['validation_warning_1']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['validation_warning_1']['primary_assembly_loci'].keys()

        assert results['flag'] == 'warning'

    def test_variant21(self):
        variant = 'NM_000518.4:c.316_*100del'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000518.4:c.316_*100del' in results.keys()
        assert results['NM_000518.4:c.316_*100del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000518.4:c.316_*100del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000518.4:c.316_*100del']['alt_genomic_loci'] == []
        assert results['NM_000518.4:c.316_*100del']['gene_symbol'] == 'HBB'
        assert results['NM_000518.4:c.316_*100del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000509.1(LRG_1232p1):p.(Leu106SerfsTer3)', 'slr': 'NP_000509.1:p.(L106Sfs*3)'}
        assert results['NM_000518.4:c.316_*100del']['submitted_variant'] == 'NM_000518.4:c.316_*100del'
        assert results['NM_000518.4:c.316_*100del']['genome_context_intronic_sequence'] == ''
        assert results['NM_000518.4:c.316_*100del']['hgvs_lrg_variant'] == ''
        assert results['NM_000518.4:c.316_*100del']['hgvs_transcript_variant'] == 'NM_000518.4:c.316_*100del'
        assert results['NM_000518.4:c.316_*100del']['hgvs_refseqgene_variant'] == 'NG_000007.3:g.71890_72118del'
        assert results['NM_000518.4:c.316_*100del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.5246728_5246956del', 'vcf': {'chr': 'chr11', 'ref': 'AATCCAGATGCTCAAGGCCCTTCATAATATCCCCCAGTTTAGTAGTTGGACTTAGGGAACAAAGGAACCTTTAATAGAAATTGGACAGCAAGAAAGCGAGCTTAGTGATACTTGTGGGCCAGGGCATTAGCCACACCAGCCACCACTTTCTGATAGGCAGCCTGCACTGGTGGGGTGAATTCTTTGCCAAAGTGATGGGCCAGCACACAGACCAGCACGTTGCCCAGGAG', 'pos': '5246727', 'alt': 'A'}}
        assert results['NM_000518.4:c.316_*100del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.5225498_5225726del', 'vcf': {'chr': 'chr11', 'ref': 'AATCCAGATGCTCAAGGCCCTTCATAATATCCCCCAGTTTAGTAGTTGGACTTAGGGAACAAAGGAACCTTTAATAGAAATTGGACAGCAAGAAAGCGAGCTTAGTGATACTTGTGGGCCAGGGCATTAGCCACACCAGCCACCACTTTCTGATAGGCAGCCTGCACTGGTGGGGTGAATTCTTTGCCAAAGTGATGGGCCAGCACACAGACCAGCACGTTGCCCAGGAG', 'pos': '5225497', 'alt': 'A'}}
        assert results['NM_000518.4:c.316_*100del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.5246728_5246956del', 'vcf': {'chr': '11', 'ref': 'AATCCAGATGCTCAAGGCCCTTCATAATATCCCCCAGTTTAGTAGTTGGACTTAGGGAACAAAGGAACCTTTAATAGAAATTGGACAGCAAGAAAGCGAGCTTAGTGATACTTGTGGGCCAGGGCATTAGCCACACCAGCCACCACTTTCTGATAGGCAGCCTGCACTGGTGGGGTGAATTCTTTGCCAAAGTGATGGGCCAGCACACAGACCAGCACGTTGCCCAGGAG', 'pos': '5246727', 'alt': 'A'}}
        assert results['NM_000518.4:c.316_*100del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.5225498_5225726del', 'vcf': {'chr': '11', 'ref': 'AATCCAGATGCTCAAGGCCCTTCATAATATCCCCCAGTTTAGTAGTTGGACTTAGGGAACAAAGGAACCTTTAATAGAAATTGGACAGCAAGAAAGCGAGCTTAGTGATACTTGTGGGCCAGGGCATTAGCCACACCAGCCACCACTTTCTGATAGGCAGCCTGCACTGGTGGGGTGAATTCTTTGCCAAAGTGATGGGCCAGCACACAGACCAGCACGTTGCCCAGGAG', 'pos': '5225497', 'alt': 'A'}}


    def test_variant22(self):
        variant = 'NM_000518.4:c.*2000C>T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'validation_warning_1' in results.keys()
        assert results['validation_warning_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['validation_warning_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['alt_genomic_loci'] == []
        assert results['validation_warning_1']['gene_symbol'] == ''
        assert results['validation_warning_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['validation_warning_1']['submitted_variant'] == 'NM_000518.4:c.*2000C>T'
        assert results['validation_warning_1']['genome_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['hgvs_lrg_variant'] == ''
        assert results['validation_warning_1']['hgvs_transcript_variant'] == ''
        assert results['validation_warning_1']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['validation_warning_1']['primary_assembly_loci'].keys()

        assert results['flag'] == 'warning'

    def test_variant23(self):
        variant = 'NM_000518.4:c.*132+1868C>T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'validation_warning_1' in results.keys()
        assert results['validation_warning_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['validation_warning_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['alt_genomic_loci'] == []
        assert results['validation_warning_1']['gene_symbol'] == ''
        assert results['validation_warning_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['validation_warning_1']['submitted_variant'] == 'NM_000518.4:c.*132+1868C>T'
        assert results['validation_warning_1']['genome_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['hgvs_lrg_variant'] == ''
        assert results['validation_warning_1']['hgvs_transcript_variant'] == ''
        assert results['validation_warning_1']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['validation_warning_1']['primary_assembly_loci'].keys()

        assert results['flag'] == 'warning'

    def test_variant24(self):
        variant = 'NM_000518.4:c.-130_*2000='
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'validation_warning_1' in results.keys()
        assert results['validation_warning_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['validation_warning_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['alt_genomic_loci'] == []
        assert results['validation_warning_1']['gene_symbol'] == ''
        assert results['validation_warning_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['validation_warning_1']['submitted_variant'] == 'NM_000518.4:c.-130_*2000='
        assert results['validation_warning_1']['genome_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['hgvs_lrg_variant'] == ''
        assert results['validation_warning_1']['hgvs_transcript_variant'] == ''
        assert results['validation_warning_1']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['validation_warning_1']['primary_assembly_loci'].keys()

        assert results['flag'] == 'warning'

    def test_variant25(self):
        variant = 'NM_000518.4:c.-50-80_*132+1868='
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'validation_warning_1' in results.keys()
        assert results['validation_warning_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['validation_warning_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['alt_genomic_loci'] == []
        assert results['validation_warning_1']['gene_symbol'] == ''
        assert results['validation_warning_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['validation_warning_1']['submitted_variant'] == 'NM_000518.4:c.-50-80_*132+1868='
        assert results['validation_warning_1']['genome_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['hgvs_lrg_variant'] == ''
        assert results['validation_warning_1']['hgvs_transcript_variant'] == ''
        assert results['validation_warning_1']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['validation_warning_1']['primary_assembly_loci'].keys()

        assert results['flag'] == 'warning'

    def test_variant26(self):
        variant = 'NR_138595.1:n.-810C>T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'validation_warning_1' in results.keys()
        assert results['validation_warning_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['validation_warning_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['alt_genomic_loci'] == []
        assert results['validation_warning_1']['gene_symbol'] == ''
        assert results['validation_warning_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['validation_warning_1']['submitted_variant'] == 'NR_138595.1:n.-810C>T'
        assert results['validation_warning_1']['genome_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['hgvs_lrg_variant'] == ''
        assert results['validation_warning_1']['hgvs_transcript_variant'] == ''
        assert results['validation_warning_1']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['validation_warning_1']['primary_assembly_loci'].keys()

        assert results['flag'] == 'warning'

    def test_variant27(self):
        variant = 'NR_138595.1:n.1-810C>T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'validation_warning_1' in results.keys()
        assert results['validation_warning_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['validation_warning_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['alt_genomic_loci'] == []
        assert results['validation_warning_1']['gene_symbol'] == ''
        assert results['validation_warning_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['validation_warning_1']['submitted_variant'] == 'NR_138595.1:n.1-810C>T'
        assert results['validation_warning_1']['genome_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['hgvs_lrg_variant'] == ''
        assert results['validation_warning_1']['hgvs_transcript_variant'] == ''
        assert results['validation_warning_1']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['validation_warning_1']['primary_assembly_loci'].keys()

        assert results['flag'] == 'warning'

    def test_variant28(self):
        variant = 'NR_138595.1:n.1071+1A='
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'validation_warning_1' in results.keys()
        assert results['validation_warning_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['validation_warning_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['alt_genomic_loci'] == []
        assert results['validation_warning_1']['gene_symbol'] == ''
        assert results['validation_warning_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['validation_warning_1']['submitted_variant'] == 'NR_138595.1:n.1071+1A='
        assert results['validation_warning_1']['genome_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['hgvs_lrg_variant'] == ''
        assert results['validation_warning_1']['hgvs_transcript_variant'] == ''
        assert results['validation_warning_1']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['validation_warning_1']['primary_assembly_loci'].keys()

        assert results['flag'] == 'warning'

    def test_variant29(self):
        variant = 'NR_138595.1:n.-810_1071+1='
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'validation_warning_1' in results.keys()
        assert results['validation_warning_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['validation_warning_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['alt_genomic_loci'] == []
        assert results['validation_warning_1']['gene_symbol'] == ''
        assert results['validation_warning_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['validation_warning_1']['submitted_variant'] == 'NR_138595.1:n.-810_1071+1='
        assert results['validation_warning_1']['genome_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['hgvs_lrg_variant'] == ''
        assert results['validation_warning_1']['hgvs_transcript_variant'] == ''
        assert results['validation_warning_1']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['validation_warning_1']['primary_assembly_loci'].keys()

        assert results['flag'] == 'warning'

    def test_variant30(self):
        variant = 'NC_000017.10:g.48261457_48261463TTATGTT='
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000088.3:c.*1400_*1406=' in results.keys()
        assert results['NM_000088.3:c.*1400_*1406=']['hgvs_lrg_transcript_variant'] == 'LRG_1t1:c.*1400_*1406='
        assert results['NM_000088.3:c.*1400_*1406=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000088.3:c.*1400_*1406=']['alt_genomic_loci'] == []
        assert results['NM_000088.3:c.*1400_*1406=']['gene_symbol'] == 'COL1A1'
        assert results['NM_000088.3:c.*1400_*1406=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000079.2(LRG_1p1):p.?', 'slr': 'NP_000079.2:p.?'}
        assert results['NM_000088.3:c.*1400_*1406=']['submitted_variant'] == 'NC_000017.10:g.48261457_48261463TTATGTT='
        assert results['NM_000088.3:c.*1400_*1406=']['genome_context_intronic_sequence'] == ''
        assert results['NM_000088.3:c.*1400_*1406=']['hgvs_lrg_variant'] == 'LRG_1:g.22538_22544='
        assert results['NM_000088.3:c.*1400_*1406=']['hgvs_transcript_variant'] == 'NM_000088.3:c.*1400_*1406='
        assert results['NM_000088.3:c.*1400_*1406=']['hgvs_refseqgene_variant'] == 'NG_007400.1:g.22538_22544='
        assert results['NM_000088.3:c.*1400_*1406=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48261457_48261463=', 'vcf': {'chr': 'chr17', 'ref': u'TTATGTT', 'pos': '48261457', 'alt': u'TTATGTT'}}
        assert results['NM_000088.3:c.*1400_*1406=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50184096_50184102=', 'vcf': {'chr': 'chr17', 'ref': u'TTATGTT', 'pos': '50184096', 'alt': u'TTATGTT'}}
        assert results['NM_000088.3:c.*1400_*1406=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48261457_48261463=', 'vcf': {'chr': '17', 'ref': u'TTATGTT', 'pos': '48261457', 'alt': u'TTATGTT'}}
        assert results['NM_000088.3:c.*1400_*1406=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50184096_50184102=', 'vcf': {'chr': '17', 'ref': u'TTATGTT', 'pos': '50184096', 'alt': u'TTATGTT'}}


    def test_variant31(self):
        variant = 'NC_000017.10:g.48275363C>A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000088.3:c.589G>T' in results.keys()
        assert results['NM_000088.3:c.589G>T']['hgvs_lrg_transcript_variant'] == 'LRG_1t1:c.589G>T'
        assert results['NM_000088.3:c.589G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000088.3:c.589G>T']['alt_genomic_loci'] == []
        assert results['NM_000088.3:c.589G>T']['gene_symbol'] == 'COL1A1'
        assert results['NM_000088.3:c.589G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000079.2(LRG_1p1):p.(Gly197Cys)', 'slr': 'NP_000079.2:p.(G197C)'}
        assert results['NM_000088.3:c.589G>T']['submitted_variant'] == 'NC_000017.10:g.48275363C>A'
        assert results['NM_000088.3:c.589G>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_000088.3:c.589G>T']['hgvs_lrg_variant'] == 'LRG_1:g.8638G>T'
        assert results['NM_000088.3:c.589G>T']['hgvs_transcript_variant'] == 'NM_000088.3:c.589G>T'
        assert results['NM_000088.3:c.589G>T']['hgvs_refseqgene_variant'] == 'NG_007400.1:g.8638G>T'
        assert results['NM_000088.3:c.589G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275363C>A', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '48275363', 'alt': u'A'}}
        assert results['NM_000088.3:c.589G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198002C>A', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '50198002', 'alt': u'A'}}
        assert results['NM_000088.3:c.589G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275363C>A', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '48275363', 'alt': u'A'}}
        assert results['NM_000088.3:c.589G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198002C>A', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '50198002', 'alt': u'A'}}


    def test_variant32(self):
        variant = 'NM_000088.3:c.589-1G>T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000088.3:c.589-1G>T' in results.keys()
        assert results['NM_000088.3:c.589-1G>T']['hgvs_lrg_transcript_variant'] == 'LRG_1t1:c.589-1G>T'
        assert results['NM_000088.3:c.589-1G>T']['refseqgene_context_intronic_sequence'] == 'NG_007400.1(NM_000088.3):c.589-1G>T'
        assert results['NM_000088.3:c.589-1G>T']['alt_genomic_loci'] == []
        assert results['NM_000088.3:c.589-1G>T']['gene_symbol'] == 'COL1A1'
        assert results['NM_000088.3:c.589-1G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000079.2(LRG_1p1):p.?', 'slr': 'NP_000079.2:p.?'}
        assert results['NM_000088.3:c.589-1G>T']['submitted_variant'] == 'NM_000088.3:c.589-1G>T'
        assert results['NM_000088.3:c.589-1G>T']['genome_context_intronic_sequence'] == 'NC_000017.10(NM_000088.3):c.589-1G>T'
        assert results['NM_000088.3:c.589-1G>T']['hgvs_lrg_variant'] == 'LRG_1:g.8637G>T'
        assert results['NM_000088.3:c.589-1G>T']['hgvs_transcript_variant'] == 'NM_000088.3:c.589-1G>T'
        assert results['NM_000088.3:c.589-1G>T']['hgvs_refseqgene_variant'] == 'NG_007400.1:g.8637G>T'
        assert results['NM_000088.3:c.589-1G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275364C>A', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '48275364', 'alt': u'A'}}
        assert results['NM_000088.3:c.589-1G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198003C>A', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '50198003', 'alt': u'A'}}
        assert results['NM_000088.3:c.589-1G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275364C>A', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '48275364', 'alt': u'A'}}
        assert results['NM_000088.3:c.589-1G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198003C>A', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '50198003', 'alt': u'A'}}


    def test_variant33(self):
        variant = 'NM_000088.3:c.591_593inv'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_000088.3:c.591_593inv' in results.keys()
        assert results['NM_000088.3:c.591_593inv']['hgvs_lrg_transcript_variant'] == 'LRG_1t1:c.591_593inv'
        assert results['NM_000088.3:c.591_593inv']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000088.3:c.591_593inv']['alt_genomic_loci'] == []
        assert results['NM_000088.3:c.591_593inv']['gene_symbol'] == 'COL1A1'
        assert results['NM_000088.3:c.591_593inv']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000079.2(LRG_1p1):p.(Pro198Asp)', 'slr': 'NP_000079.2:p.(P198D)'}
        assert results['NM_000088.3:c.591_593inv']['submitted_variant'] == 'NM_000088.3:c.591_593inv'
        assert results['NM_000088.3:c.591_593inv']['genome_context_intronic_sequence'] == ''
        assert results['NM_000088.3:c.591_593inv']['hgvs_lrg_variant'] == 'LRG_1:g.8640_8642inv'
        assert results['NM_000088.3:c.591_593inv']['hgvs_transcript_variant'] == 'NM_000088.3:c.591_593inv'
        assert results['NM_000088.3:c.591_593inv']['hgvs_refseqgene_variant'] == 'NG_007400.1:g.8640_8642inv'
        assert results['NM_000088.3:c.591_593inv']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275359_48275361inv', 'vcf': {'chr': 'chr17', 'ref': 'GGA', 'pos': '48275359', 'alt': 'TCC'}}
        assert results['NM_000088.3:c.591_593inv']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50197998_50198000inv', 'vcf': {'chr': 'chr17', 'ref': 'GGA', 'pos': '50197998', 'alt': 'TCC'}}
        assert results['NM_000088.3:c.591_593inv']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275359_48275361inv', 'vcf': {'chr': '17', 'ref': 'GGA', 'pos': '48275359', 'alt': 'TCC'}}
        assert results['NM_000088.3:c.591_593inv']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50197998_50198000inv', 'vcf': {'chr': '17', 'ref': 'GGA', 'pos': '50197998', 'alt': 'TCC'}}

        assert results['flag'] == 'gene_variant'

    def test_variant34(self):
        variant = '11-5248232-T-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000518.5:c.20A>T' in results.keys()
        assert results['NM_000518.5:c.20A>T']['hgvs_lrg_transcript_variant'] == 'LRG_1232t1:c.20A>T'
        assert results['NM_000518.5:c.20A>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000518.5:c.20A>T']['alt_genomic_loci'] == []
        assert results['NM_000518.5:c.20A>T']['gene_symbol'] == 'HBB'
        assert results['NM_000518.5:c.20A>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000509.1(LRG_1232p1):p.(Glu7Val)', 'slr': 'NP_000509.1:p.(E7V)'}
        assert results['NM_000518.5:c.20A>T']['submitted_variant'] == '11-5248232-T-A'
        assert results['NM_000518.5:c.20A>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_000518.5:c.20A>T']['hgvs_lrg_variant'] == ''
        assert results['NM_000518.5:c.20A>T']['hgvs_transcript_variant'] == 'NM_000518.5:c.20A>T'
        assert results['NM_000518.5:c.20A>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_000518.5:c.20A>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.5248232T>A', 'vcf': {'chr': 'chr11', 'ref': u'T', 'pos': '5248232', 'alt': u'A'}}
        assert 'hg38' not in results['NM_000518.5:c.20A>T']['primary_assembly_loci'].keys()
        assert results['NM_000518.5:c.20A>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.5248232T>A', 'vcf': {'chr': '11', 'ref': u'T', 'pos': '5248232', 'alt': u'A'}}
        assert 'grch38' not in results['NM_000518.5:c.20A>T']['primary_assembly_loci'].keys()

        assert 'NM_000518.4:c.20A>T' in results.keys()
        assert results['NM_000518.4:c.20A>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000518.4:c.20A>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000518.4:c.20A>T']['alt_genomic_loci'] == []
        assert results['NM_000518.4:c.20A>T']['gene_symbol'] == 'HBB'
        assert results['NM_000518.4:c.20A>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000509.1(LRG_1232p1):p.(Glu7Val)', 'slr': 'NP_000509.1:p.(E7V)'}
        assert results['NM_000518.4:c.20A>T']['submitted_variant'] == '11-5248232-T-A'
        assert results['NM_000518.4:c.20A>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_000518.4:c.20A>T']['hgvs_lrg_variant'] == ''
        assert results['NM_000518.4:c.20A>T']['hgvs_transcript_variant'] == 'NM_000518.4:c.20A>T'
        assert results['NM_000518.4:c.20A>T']['hgvs_refseqgene_variant'] == 'NG_000007.3:g.70614A>T'
        assert results['NM_000518.4:c.20A>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.5248232T>A', 'vcf': {'chr': 'chr11', 'ref': u'T', 'pos': '5248232', 'alt': u'A'}}
        assert results['NM_000518.4:c.20A>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.5227002T>A', 'vcf': {'chr': 'chr11', 'ref': u'T', 'pos': '5227002', 'alt': u'A'}}
        assert results['NM_000518.4:c.20A>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.5248232T>A', 'vcf': {'chr': '11', 'ref': u'T', 'pos': '5248232', 'alt': u'A'}}
        assert results['NM_000518.4:c.20A>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.5227002T>A', 'vcf': {'chr': '11', 'ref': u'T', 'pos': '5227002', 'alt': u'A'}}


    def test_variant35(self):
        variant = 'NG_007400.1(NM_000088.3):c.589-1G>T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000088.3:c.589-1G>T' in results.keys()
        assert results['NM_000088.3:c.589-1G>T']['hgvs_lrg_transcript_variant'] == 'LRG_1t1:c.589-1G>T'
        assert results['NM_000088.3:c.589-1G>T']['refseqgene_context_intronic_sequence'] == 'NG_007400.1(NM_000088.3):c.589-1G>T'
        assert results['NM_000088.3:c.589-1G>T']['alt_genomic_loci'] == []
        assert results['NM_000088.3:c.589-1G>T']['gene_symbol'] == 'COL1A1'
        assert results['NM_000088.3:c.589-1G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000079.2(LRG_1p1):p.?', 'slr': 'NP_000079.2:p.?'}
        assert results['NM_000088.3:c.589-1G>T']['submitted_variant'] == 'NG_007400.1(NM_000088.3):c.589-1G>T'
        assert results['NM_000088.3:c.589-1G>T']['genome_context_intronic_sequence'] == 'NC_000017.10(NM_000088.3):c.589-1G>T'
        assert results['NM_000088.3:c.589-1G>T']['hgvs_lrg_variant'] == 'LRG_1:g.8637G>T'
        assert results['NM_000088.3:c.589-1G>T']['hgvs_transcript_variant'] == 'NM_000088.3:c.589-1G>T'
        assert results['NM_000088.3:c.589-1G>T']['hgvs_refseqgene_variant'] == 'NG_007400.1:g.8637G>T'
        assert results['NM_000088.3:c.589-1G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275364C>A', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '48275364', 'alt': u'A'}}
        assert results['NM_000088.3:c.589-1G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198003C>A', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '50198003', 'alt': u'A'}}
        assert results['NM_000088.3:c.589-1G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275364C>A', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '48275364', 'alt': u'A'}}
        assert results['NM_000088.3:c.589-1G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198003C>A', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '50198003', 'alt': u'A'}}


    def test_variant36(self):
        variant = '1:150550916G>A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_182763.2:c.688+403C>T' in results.keys()
        assert results['NM_182763.2:c.688+403C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_182763.2:c.688+403C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_182763.2:c.688+403C>T']['alt_genomic_loci'] == []
        assert results['NM_182763.2:c.688+403C>T']['gene_symbol'] == 'MCL1'
        assert results['NM_182763.2:c.688+403C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_877495.1:p.?', 'slr': 'NP_877495.1:p.?'}
        assert results['NM_182763.2:c.688+403C>T']['submitted_variant'] == '1:150550916G>A'
        assert results['NM_182763.2:c.688+403C>T']['genome_context_intronic_sequence'] == 'NC_000001.10(NM_182763.2):c.688+403C>T'
        assert results['NM_182763.2:c.688+403C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_182763.2:c.688+403C>T']['hgvs_transcript_variant'] == 'NM_182763.2:c.688+403C>T'
        assert results['NM_182763.2:c.688+403C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_182763.2:c.688+403C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.150550916G>A', 'vcf': {'chr': 'chr1', 'ref': u'G', 'pos': '150550916', 'alt': u'A'}}
        assert results['NM_182763.2:c.688+403C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.150578440G>A', 'vcf': {'chr': 'chr1', 'ref': u'G', 'pos': '150578440', 'alt': u'A'}}
        assert results['NM_182763.2:c.688+403C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.150550916G>A', 'vcf': {'chr': '1', 'ref': u'G', 'pos': '150550916', 'alt': u'A'}}
        assert results['NM_182763.2:c.688+403C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.150578440G>A', 'vcf': {'chr': '1', 'ref': u'G', 'pos': '150578440', 'alt': u'A'}}

        assert 'NM_001197320.1:c.281C>T' in results.keys()
        assert results['NM_001197320.1:c.281C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001197320.1:c.281C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001197320.1:c.281C>T']['alt_genomic_loci'] == []
        assert results['NM_001197320.1:c.281C>T']['gene_symbol'] == 'MCL1'
        assert results['NM_001197320.1:c.281C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001184249.1:p.(Ser94Phe)', 'slr': 'NP_001184249.1:p.(S94F)'}
        assert results['NM_001197320.1:c.281C>T']['submitted_variant'] == '1:150550916G>A'
        assert results['NM_001197320.1:c.281C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001197320.1:c.281C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001197320.1:c.281C>T']['hgvs_transcript_variant'] == 'NM_001197320.1:c.281C>T'
        assert results['NM_001197320.1:c.281C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001197320.1:c.281C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.150550916G>A', 'vcf': {'chr': 'chr1', 'ref': u'G', 'pos': '150550916', 'alt': u'A'}}
        assert results['NM_001197320.1:c.281C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.150578440G>A', 'vcf': {'chr': 'chr1', 'ref': u'G', 'pos': '150578440', 'alt': u'A'}}
        assert results['NM_001197320.1:c.281C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.150550916G>A', 'vcf': {'chr': '1', 'ref': u'G', 'pos': '150550916', 'alt': u'A'}}
        assert results['NM_001197320.1:c.281C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.150578440G>A', 'vcf': {'chr': '1', 'ref': u'G', 'pos': '150578440', 'alt': u'A'}}

        assert 'NM_021960.4:c.740C>T' in results.keys()
        assert results['NM_021960.4:c.740C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_021960.4:c.740C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_021960.4:c.740C>T']['alt_genomic_loci'] == []
        assert results['NM_021960.4:c.740C>T']['gene_symbol'] == 'MCL1'
        assert results['NM_021960.4:c.740C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_068779.1:p.(Ser247Phe)', 'slr': 'NP_068779.1:p.(S247F)'}
        assert results['NM_021960.4:c.740C>T']['submitted_variant'] == '1:150550916G>A'
        assert results['NM_021960.4:c.740C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_021960.4:c.740C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_021960.4:c.740C>T']['hgvs_transcript_variant'] == 'NM_021960.4:c.740C>T'
        assert results['NM_021960.4:c.740C>T']['hgvs_refseqgene_variant'] == 'NG_029146.1:g.6299C>T'
        assert results['NM_021960.4:c.740C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.150550916G>A', 'vcf': {'chr': 'chr1', 'ref': u'G', 'pos': '150550916', 'alt': u'A'}}
        assert results['NM_021960.4:c.740C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.150578440G>A', 'vcf': {'chr': 'chr1', 'ref': u'G', 'pos': '150578440', 'alt': u'A'}}
        assert results['NM_021960.4:c.740C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.150550916G>A', 'vcf': {'chr': '1', 'ref': u'G', 'pos': '150550916', 'alt': u'A'}}
        assert results['NM_021960.4:c.740C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.150578440G>A', 'vcf': {'chr': '1', 'ref': u'G', 'pos': '150578440', 'alt': u'A'}}


    def test_variant37(self):
        variant = '1-150550916-G-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_182763.2:c.688+403C>T' in results.keys()
        assert results['NM_182763.2:c.688+403C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_182763.2:c.688+403C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_182763.2:c.688+403C>T']['alt_genomic_loci'] == []
        assert results['NM_182763.2:c.688+403C>T']['gene_symbol'] == 'MCL1'
        assert results['NM_182763.2:c.688+403C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_877495.1:p.?', 'slr': 'NP_877495.1:p.?'}
        assert results['NM_182763.2:c.688+403C>T']['submitted_variant'] == '1-150550916-G-A'
        assert results['NM_182763.2:c.688+403C>T']['genome_context_intronic_sequence'] == 'NC_000001.10(NM_182763.2):c.688+403C>T'
        assert results['NM_182763.2:c.688+403C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_182763.2:c.688+403C>T']['hgvs_transcript_variant'] == 'NM_182763.2:c.688+403C>T'
        assert results['NM_182763.2:c.688+403C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_182763.2:c.688+403C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.150550916G>A', 'vcf': {'chr': 'chr1', 'ref': u'G', 'pos': '150550916', 'alt': u'A'}}
        assert results['NM_182763.2:c.688+403C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.150578440G>A', 'vcf': {'chr': 'chr1', 'ref': u'G', 'pos': '150578440', 'alt': u'A'}}
        assert results['NM_182763.2:c.688+403C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.150550916G>A', 'vcf': {'chr': '1', 'ref': u'G', 'pos': '150550916', 'alt': u'A'}}
        assert results['NM_182763.2:c.688+403C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.150578440G>A', 'vcf': {'chr': '1', 'ref': u'G', 'pos': '150578440', 'alt': u'A'}}

        assert 'NM_001197320.1:c.281C>T' in results.keys()
        assert results['NM_001197320.1:c.281C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001197320.1:c.281C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001197320.1:c.281C>T']['alt_genomic_loci'] == []
        assert results['NM_001197320.1:c.281C>T']['gene_symbol'] == 'MCL1'
        assert results['NM_001197320.1:c.281C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001184249.1:p.(Ser94Phe)', 'slr': 'NP_001184249.1:p.(S94F)'}
        assert results['NM_001197320.1:c.281C>T']['submitted_variant'] == '1-150550916-G-A'
        assert results['NM_001197320.1:c.281C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001197320.1:c.281C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001197320.1:c.281C>T']['hgvs_transcript_variant'] == 'NM_001197320.1:c.281C>T'
        assert results['NM_001197320.1:c.281C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001197320.1:c.281C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.150550916G>A', 'vcf': {'chr': 'chr1', 'ref': u'G', 'pos': '150550916', 'alt': u'A'}}
        assert results['NM_001197320.1:c.281C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.150578440G>A', 'vcf': {'chr': 'chr1', 'ref': u'G', 'pos': '150578440', 'alt': u'A'}}
        assert results['NM_001197320.1:c.281C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.150550916G>A', 'vcf': {'chr': '1', 'ref': u'G', 'pos': '150550916', 'alt': u'A'}}
        assert results['NM_001197320.1:c.281C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.150578440G>A', 'vcf': {'chr': '1', 'ref': u'G', 'pos': '150578440', 'alt': u'A'}}

        assert 'NM_021960.4:c.740C>T' in results.keys()
        assert results['NM_021960.4:c.740C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_021960.4:c.740C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_021960.4:c.740C>T']['alt_genomic_loci'] == []
        assert results['NM_021960.4:c.740C>T']['gene_symbol'] == 'MCL1'
        assert results['NM_021960.4:c.740C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_068779.1:p.(Ser247Phe)', 'slr': 'NP_068779.1:p.(S247F)'}
        assert results['NM_021960.4:c.740C>T']['submitted_variant'] == '1-150550916-G-A'
        assert results['NM_021960.4:c.740C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_021960.4:c.740C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_021960.4:c.740C>T']['hgvs_transcript_variant'] == 'NM_021960.4:c.740C>T'
        assert results['NM_021960.4:c.740C>T']['hgvs_refseqgene_variant'] == 'NG_029146.1:g.6299C>T'
        assert results['NM_021960.4:c.740C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.150550916G>A', 'vcf': {'chr': 'chr1', 'ref': u'G', 'pos': '150550916', 'alt': u'A'}}
        assert results['NM_021960.4:c.740C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.150578440G>A', 'vcf': {'chr': 'chr1', 'ref': u'G', 'pos': '150578440', 'alt': u'A'}}
        assert results['NM_021960.4:c.740C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.150550916G>A', 'vcf': {'chr': '1', 'ref': u'G', 'pos': '150550916', 'alt': u'A'}}
        assert results['NM_021960.4:c.740C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.150578440G>A', 'vcf': {'chr': '1', 'ref': u'G', 'pos': '150578440', 'alt': u'A'}}


    def test_variant38(self):
        variant = 'NG_008123.1(LEPRE1_v003):c.2055+18G>A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'validation_warning_1' in results.keys()
        assert results['validation_warning_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['validation_warning_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['alt_genomic_loci'] == []
        assert results['validation_warning_1']['gene_symbol'] == ''
        assert results['validation_warning_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['validation_warning_1']['submitted_variant'] == 'NG_008123.1(LEPRE1_v003):c.2055+18G>A'
        assert results['validation_warning_1']['genome_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['hgvs_lrg_variant'] == ''
        assert results['validation_warning_1']['hgvs_transcript_variant'] == ''
        assert results['validation_warning_1']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['validation_warning_1']['primary_assembly_loci'].keys()

        assert results['flag'] == 'warning'

    def test_variant39(self):
        variant = 'NG_008123.1:c.2055+18G>A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'validation_warning_1' in results.keys()
        assert results['validation_warning_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['validation_warning_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['alt_genomic_loci'] == []
        assert results['validation_warning_1']['gene_symbol'] == ''
        assert results['validation_warning_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['validation_warning_1']['submitted_variant'] == 'NG_008123.1:c.2055+18G>A'
        assert results['validation_warning_1']['genome_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['hgvs_lrg_variant'] == ''
        assert results['validation_warning_1']['hgvs_transcript_variant'] == ''
        assert results['validation_warning_1']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['validation_warning_1']['primary_assembly_loci'].keys()

        assert results['flag'] == 'warning'

    def test_variant40(self):
        variant = 'NG_008123.1(NM_022356.3):c.2055+18G>A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_022356.3:c.2055+18G>A' in results.keys()
        assert results['NM_022356.3:c.2055+18G>A']['hgvs_lrg_transcript_variant'] == 'LRG_5t1:c.2055+18G>A'
        assert results['NM_022356.3:c.2055+18G>A']['refseqgene_context_intronic_sequence'] == 'NG_008123.1(NM_022356.3):c.2055+18G>A'
        assert results['NM_022356.3:c.2055+18G>A']['alt_genomic_loci'] == []
        assert results['NM_022356.3:c.2055+18G>A']['gene_symbol'] == 'P3H1'
        assert results['NM_022356.3:c.2055+18G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_071751.3(LRG_5p1):p.?', 'slr': 'NP_071751.3:p.?'}
        assert results['NM_022356.3:c.2055+18G>A']['submitted_variant'] == 'NG_008123.1(NM_022356.3):c.2055+18G>A'
        assert results['NM_022356.3:c.2055+18G>A']['genome_context_intronic_sequence'] == 'NC_000001.10(NM_022356.3):c.2055+18G>A'
        assert results['NM_022356.3:c.2055+18G>A']['hgvs_lrg_variant'] == 'LRG_5:g.24831G>A'
        assert results['NM_022356.3:c.2055+18G>A']['hgvs_transcript_variant'] == 'NM_022356.3:c.2055+18G>A'
        assert results['NM_022356.3:c.2055+18G>A']['hgvs_refseqgene_variant'] == 'NG_008123.1:g.24831G>A'
        assert results['NM_022356.3:c.2055+18G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.43212925C>T', 'vcf': {'chr': 'chr1', 'ref': u'C', 'pos': '43212925', 'alt': u'T'}}
        assert results['NM_022356.3:c.2055+18G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.42747254C>T', 'vcf': {'chr': 'chr1', 'ref': u'C', 'pos': '42747254', 'alt': u'T'}}
        assert results['NM_022356.3:c.2055+18G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.43212925C>T', 'vcf': {'chr': '1', 'ref': u'C', 'pos': '43212925', 'alt': u'T'}}
        assert results['NM_022356.3:c.2055+18G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.42747254C>T', 'vcf': {'chr': '1', 'ref': u'C', 'pos': '42747254', 'alt': u'T'}}


    def test_variant41(self):
        variant = 'NM_021983.4:c.490G>C'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_021983.4:c.490G>C' in results.keys()
        assert results['NM_021983.4:c.490G>C']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_021983.4:c.490G>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_021983.4:c.490G>C']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NT_167246.1:g.3848158T>G', 'vcf': {'chr': 'HSCHR6_MHC_MANN_CTG1', 'ref': 'T', 'pos': '3848158', 'alt': u'G'}}}, {'hg19': {'hgvs_genomic_description': 'NT_167246.1:g.3848158T>G', 'vcf': {'chr': 'chr6_mann_hap4', 'ref': 'T', 'pos': '3848158', 'alt': u'G'}}}, {'grch38': {'hgvs_genomic_description': 'NT_167246.2:g.3842538T>G', 'vcf': {'chr': 'HSCHR6_MHC_MANN_CTG1', 'ref': 'T', 'pos': '3842538', 'alt': u'G'}}}, {'hg38': {'hgvs_genomic_description': 'NT_167246.2:g.3842538T>G', 'vcf': {'chr': 'chr6_GL000253v2_alt', 'ref': 'T', 'pos': '3842538', 'alt': u'G'}}}, {'grch37': {'hgvs_genomic_description': 'NT_167247.1:g.3884432C>G', 'vcf': {'chr': 'HSCHR6_MHC_MCF_CTG1', 'ref': u'C', 'pos': '3884432', 'alt': u'G'}}}, {'hg19': {'hgvs_genomic_description': 'NT_167247.1:g.3884432C>G', 'vcf': {'chr': 'chr6_mcf_hap5', 'ref': u'C', 'pos': '3884432', 'alt': u'G'}}}, {'grch37': {'hgvs_genomic_description': 'NT_167249.1:g.3852542C>G', 'vcf': {'chr': 'HSCHR6_MHC_SSTO_CTG1', 'ref': u'C', 'pos': '3852542', 'alt': u'G'}}}, {'hg19': {'hgvs_genomic_description': 'NT_167249.1:g.3852542C>G', 'vcf': {'chr': 'chr6_ssto_hap7', 'ref': u'C', 'pos': '3852542', 'alt': u'G'}}}, {'grch38': {'hgvs_genomic_description': 'NT_167249.2:g.3853244C>G', 'vcf': {'chr': 'HSCHR6_MHC_SSTO_CTG1', 'ref': u'C', 'pos': '3853244', 'alt': u'G'}}}, {'hg38': {'hgvs_genomic_description': 'NT_167249.2:g.3853244C>G', 'vcf': {'chr': 'chr6_GL000256v2_alt', 'ref': u'C', 'pos': '3853244', 'alt': u'G'}}}]
        assert results['NM_021983.4:c.490G>C']['gene_symbol'] == 'HLA-DRB4'
        assert results['NM_021983.4:c.490G>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_068818.4:p.(Gly164Arg)', 'slr': 'NP_068818.4:p.(G164R)'}
        assert results['NM_021983.4:c.490G>C']['submitted_variant'] == 'NM_021983.4:c.490G>C'
        assert results['NM_021983.4:c.490G>C']['genome_context_intronic_sequence'] == ''
        assert results['NM_021983.4:c.490G>C']['hgvs_lrg_variant'] == ''
        assert results['NM_021983.4:c.490G>C']['hgvs_transcript_variant'] == 'NM_021983.4:c.490G>C'
        assert results['NM_021983.4:c.490G>C']['hgvs_refseqgene_variant'] == 'NG_002433.1:g.5724C>G'
        assert 'hg19' not in results['NM_021983.4:c.490G>C']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['NM_021983.4:c.490G>C']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['NM_021983.4:c.490G>C']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['NM_021983.4:c.490G>C']['primary_assembly_loci'].keys()


    def test_variant42(self):
        variant = 'NM_032470.3:c.4del'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_032470.3:c.4del' in results.keys()
        assert results['NM_032470.3:c.4del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_032470.3:c.4del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_032470.3:c.4del']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NT_113891.2:g.3483644del', 'vcf': {'chr': 'HSCHR6_MHC_COX_CTG1', 'ref': 'CG', 'pos': '3483643', 'alt': 'C'}}}, {'hg19': {'hgvs_genomic_description': 'NT_113891.2:g.3483644del', 'vcf': {'chr': 'chr6_cox_hap2', 'ref': 'CG', 'pos': '3483643', 'alt': 'C'}}}, {'grch38': {'hgvs_genomic_description': 'NT_113891.3:g.3483538del', 'vcf': {'chr': 'HSCHR6_MHC_COX_CTG1', 'ref': 'CG', 'pos': '3483537', 'alt': 'C'}}}, {'hg38': {'hgvs_genomic_description': 'NT_113891.3:g.3483538del', 'vcf': {'chr': 'chr6_GL000251v2_alt', 'ref': 'CG', 'pos': '3483537', 'alt': 'C'}}}, {'grch37': {'hgvs_genomic_description': 'NT_167245.1:g.3292210del', 'vcf': {'chr': 'HSCHR6_MHC_DBB_CTG1', 'ref': 'CG', 'pos': '3292209', 'alt': 'C'}}}, {'hg19': {'hgvs_genomic_description': 'NT_167245.1:g.3292210del', 'vcf': {'chr': 'chr6_dbb_hap3', 'ref': 'CG', 'pos': '3292209', 'alt': 'C'}}}, {'grch38': {'hgvs_genomic_description': 'NT_167245.2:g.3286625del', 'vcf': {'chr': 'HSCHR6_MHC_DBB_CTG1', 'ref': 'CG', 'pos': '3286624', 'alt': 'C'}}}, {'hg38': {'hgvs_genomic_description': 'NT_167245.2:g.3286625del', 'vcf': {'chr': 'chr6_GL000252v2_alt', 'ref': 'CG', 'pos': '3286624', 'alt': 'C'}}}, {'grch37': {'hgvs_genomic_description': 'NT_167247.1:g.3392834del', 'vcf': {'chr': 'HSCHR6_MHC_MCF_CTG1', 'ref': 'CG', 'pos': '3392833', 'alt': 'C'}}}, {'hg19': {'hgvs_genomic_description': 'NT_167247.1:g.3392834del', 'vcf': {'chr': 'chr6_mcf_hap5', 'ref': 'CG', 'pos': '3392833', 'alt': 'C'}}}, {'grch38': {'hgvs_genomic_description': 'NT_167247.2:g.3387249del', 'vcf': {'chr': 'HSCHR6_MHC_MCF_CTG1', 'ref': 'CG', 'pos': '3387248', 'alt': 'C'}}}, {'hg38': {'hgvs_genomic_description': 'NT_167247.2:g.3387249del', 'vcf': {'chr': 'chr6_GL000254v2_alt', 'ref': 'CG', 'pos': '3387248', 'alt': 'C'}}}, {'grch37': {'hgvs_genomic_description': 'NT_167248.1:g.3274047del', 'vcf': {'chr': 'HSCHR6_MHC_QBL_CTG1', 'ref': 'CG', 'pos': '3274046', 'alt': 'C'}}}, {'hg19': {'hgvs_genomic_description': 'NT_167248.1:g.3274047del', 'vcf': {'chr': 'chr6_qbl_hap6', 'ref': 'CG', 'pos': '3274046', 'alt': 'C'}}}, {'grch38': {'hgvs_genomic_description': 'NT_167248.2:g.3268451del', 'vcf': {'chr': 'HSCHR6_MHC_QBL_CTG1', 'ref': 'CG', 'pos': '3268450', 'alt': 'C'}}}, {'hg38': {'hgvs_genomic_description': 'NT_167248.2:g.3268451del', 'vcf': {'chr': 'chr6_GL000255v2_alt', 'ref': 'CG', 'pos': '3268450', 'alt': 'C'}}}, {'grch37': {'hgvs_genomic_description': 'NT_167249.1:g.3345701del', 'vcf': {'chr': 'HSCHR6_MHC_SSTO_CTG1', 'ref': 'CG', 'pos': '3345700', 'alt': 'C'}}}, {'hg19': {'hgvs_genomic_description': 'NT_167249.1:g.3345701del', 'vcf': {'chr': 'chr6_ssto_hap7', 'ref': 'CG', 'pos': '3345700', 'alt': 'C'}}}, {'grch38': {'hgvs_genomic_description': 'NT_167249.2:g.3346403del', 'vcf': {'chr': 'HSCHR6_MHC_SSTO_CTG1', 'ref': 'CG', 'pos': '3346402', 'alt': 'C'}}}, {'hg38': {'hgvs_genomic_description': 'NT_167249.2:g.3346403del', 'vcf': {'chr': 'chr6_GL000256v2_alt', 'ref': 'CG', 'pos': '3346402', 'alt': 'C'}}}]
        assert results['NM_032470.3:c.4del']['gene_symbol'] == 'TNXB'
        assert results['NM_032470.3:c.4del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_115859.2:p.(Arg2AlafsTer91)', 'slr': 'NP_115859.2:p.(R2Afs*91)'}
        assert results['NM_032470.3:c.4del']['submitted_variant'] == 'NM_032470.3:c.4del'
        assert results['NM_032470.3:c.4del']['genome_context_intronic_sequence'] == ''
        assert results['NM_032470.3:c.4del']['hgvs_lrg_variant'] == ''
        assert results['NM_032470.3:c.4del']['hgvs_transcript_variant'] == 'NM_032470.3:c.4del'
        assert results['NM_032470.3:c.4del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_032470.3:c.4del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000006.11:g.32012993del', 'vcf': {'chr': 'chr6', 'ref': 'CG', 'pos': '32012992', 'alt': 'C'}}
        assert results['NM_032470.3:c.4del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000006.12:g.32045216del', 'vcf': {'chr': 'chr6', 'ref': 'CG', 'pos': '32045215', 'alt': 'C'}}
        assert results['NM_032470.3:c.4del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000006.11:g.32012993del', 'vcf': {'chr': '6', 'ref': 'CG', 'pos': '32012992', 'alt': 'C'}}
        assert results['NM_032470.3:c.4del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000006.12:g.32045216del', 'vcf': {'chr': '6', 'ref': 'CG', 'pos': '32045215', 'alt': 'C'}}


    def test_variant43(self):
        variant = 'NM_001194958.2:c.20C>A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001194958.2:c.20C>A' in results.keys()
        assert results['NM_001194958.2:c.20C>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001194958.2:c.20C>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001194958.2:c.20C>A']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003315950.2:g.355171C>A', 'vcf': {'chr': 'HG987_PATCH', 'ref': 'C', 'pos': '355171', 'alt': 'A'}}}]
        assert results['NM_001194958.2:c.20C>A']['gene_symbol'] == 'KCNJ18'
        assert results['NM_001194958.2:c.20C>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001181887.2:p.(Ala7Asp)', 'slr': 'NP_001181887.2:p.(A7D)'}
        assert results['NM_001194958.2:c.20C>A']['submitted_variant'] == 'NM_001194958.2:c.20C>A'
        assert results['NM_001194958.2:c.20C>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_001194958.2:c.20C>A']['hgvs_lrg_variant'] == ''
        assert results['NM_001194958.2:c.20C>A']['hgvs_transcript_variant'] == 'NM_001194958.2:c.20C>A'
        assert results['NM_001194958.2:c.20C>A']['hgvs_refseqgene_variant'] == 'NG_033093.1:g.15284C>A'
        assert 'hg19' not in results['NM_001194958.2:c.20C>A']['primary_assembly_loci'].keys()
        assert results['NM_001194958.2:c.20C>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.21702806C>A', 'vcf': {'chr': 'chr17', 'ref': 'C', 'pos': '21702806', 'alt': 'A'}}
        assert 'grch37' not in results['NM_001194958.2:c.20C>A']['primary_assembly_loci'].keys()
        assert results['NM_001194958.2:c.20C>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.21702806C>A', 'vcf': {'chr': '17', 'ref': 'C', 'pos': '21702806', 'alt': 'A'}}

        assert results['flag'] == 'gene_variant'

    def test_variant44(self):
        variant = 'NM_000022.2:c.534A>G'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000022.2:c.534A>G' in results.keys()
        assert results['NM_000022.2:c.534A>G']['hgvs_lrg_transcript_variant'] == 'LRG_16t1:c.534A>G'
        assert results['NM_000022.2:c.534A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000022.2:c.534A>G']['alt_genomic_loci'] == []
        assert results['NM_000022.2:c.534A>G']['gene_symbol'] == 'ADA'
        assert results['NM_000022.2:c.534A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000013.2(LRG_16p1):p.(Val178=)', 'slr': 'NP_000013.2:p.(V178=)'}
        assert results['NM_000022.2:c.534A>G']['submitted_variant'] == 'NM_000022.2:c.534A>G'
        assert results['NM_000022.2:c.534A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_000022.2:c.534A>G']['hgvs_lrg_variant'] == 'LRG_16:g.32462A>G'
        assert results['NM_000022.2:c.534A>G']['hgvs_transcript_variant'] == 'NM_000022.2:c.534A>G'
        assert results['NM_000022.2:c.534A>G']['hgvs_refseqgene_variant'] == 'NG_007385.1:g.32462A>G'
        assert results['NM_000022.2:c.534A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000020.10:g.43252915T>C', 'vcf': {'chr': 'chr20', 'ref': u'T', 'pos': '43252915', 'alt': u'C'}}
        assert 'hg38' not in results['NM_000022.2:c.534A>G']['primary_assembly_loci'].keys()
        assert results['NM_000022.2:c.534A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000020.10:g.43252915T>C', 'vcf': {'chr': '20', 'ref': u'T', 'pos': '43252915', 'alt': u'C'}}
        assert 'grch38' not in results['NM_000022.2:c.534A>G']['primary_assembly_loci'].keys()


    def test_variant45(self):
        variant = 'HSCHR6_MHC_SSTO_CTG1-3852542-C-G'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_021983.4:c.490G>C' in results.keys()
        assert results['NM_021983.4:c.490G>C']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_021983.4:c.490G>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_021983.4:c.490G>C']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NT_167246.1:g.3848158T>G', 'vcf': {'chr': 'HSCHR6_MHC_MANN_CTG1', 'ref': 'T', 'pos': '3848158', 'alt': u'G'}}}, {'hg19': {'hgvs_genomic_description': 'NT_167246.1:g.3848158T>G', 'vcf': {'chr': 'chr6_mann_hap4', 'ref': 'T', 'pos': '3848158', 'alt': u'G'}}}, {'grch38': {'hgvs_genomic_description': 'NT_167246.2:g.3842538T>G', 'vcf': {'chr': 'HSCHR6_MHC_MANN_CTG1', 'ref': 'T', 'pos': '3842538', 'alt': u'G'}}}, {'hg38': {'hgvs_genomic_description': 'NT_167246.2:g.3842538T>G', 'vcf': {'chr': 'chr6_GL000253v2_alt', 'ref': 'T', 'pos': '3842538', 'alt': u'G'}}}, {'grch37': {'hgvs_genomic_description': 'NT_167247.1:g.3884432C>G', 'vcf': {'chr': 'HSCHR6_MHC_MCF_CTG1', 'ref': u'C', 'pos': '3884432', 'alt': u'G'}}}, {'hg19': {'hgvs_genomic_description': 'NT_167247.1:g.3884432C>G', 'vcf': {'chr': 'chr6_mcf_hap5', 'ref': u'C', 'pos': '3884432', 'alt': u'G'}}}, {'grch37': {'hgvs_genomic_description': 'NT_167249.1:g.3852542C>G', 'vcf': {'chr': 'HSCHR6_MHC_SSTO_CTG1', 'ref': u'C', 'pos': '3852542', 'alt': u'G'}}}, {'hg19': {'hgvs_genomic_description': 'NT_167249.1:g.3852542C>G', 'vcf': {'chr': 'chr6_ssto_hap7', 'ref': u'C', 'pos': '3852542', 'alt': u'G'}}}, {'grch38': {'hgvs_genomic_description': 'NT_167249.2:g.3853244C>G', 'vcf': {'chr': 'HSCHR6_MHC_SSTO_CTG1', 'ref': u'C', 'pos': '3853244', 'alt': u'G'}}}, {'hg38': {'hgvs_genomic_description': 'NT_167249.2:g.3853244C>G', 'vcf': {'chr': 'chr6_GL000256v2_alt', 'ref': u'C', 'pos': '3853244', 'alt': u'G'}}}]
        assert results['NM_021983.4:c.490G>C']['gene_symbol'] == 'HLA-DRB4'
        assert results['NM_021983.4:c.490G>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_068818.4:p.(Gly164Arg)', 'slr': 'NP_068818.4:p.(G164R)'}
        assert results['NM_021983.4:c.490G>C']['submitted_variant'] == 'HSCHR6_MHC_SSTO_CTG1-3852542-C-G'
        assert results['NM_021983.4:c.490G>C']['genome_context_intronic_sequence'] == ''
        assert results['NM_021983.4:c.490G>C']['hgvs_lrg_variant'] == ''
        assert results['NM_021983.4:c.490G>C']['hgvs_transcript_variant'] == 'NM_021983.4:c.490G>C'
        assert results['NM_021983.4:c.490G>C']['hgvs_refseqgene_variant'] == 'NG_002433.1:g.5724C>G'
        assert 'hg19' not in results['NM_021983.4:c.490G>C']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['NM_021983.4:c.490G>C']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['NM_021983.4:c.490G>C']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['NM_021983.4:c.490G>C']['primary_assembly_loci'].keys()


    def test_variant46(self):
        variant = 'NM_000368.4:c.363+1dupG'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000368.4:c.363+1dup' in results.keys()
        assert results['NM_000368.4:c.363+1dup']['hgvs_lrg_transcript_variant'] == 'LRG_486t1:c.363+1dup'
        assert results['NM_000368.4:c.363+1dup']['refseqgene_context_intronic_sequence'] == 'NG_012386.1(NM_000368.4):c.363+1dup'
        assert results['NM_000368.4:c.363+1dup']['alt_genomic_loci'] == []
        assert results['NM_000368.4:c.363+1dup']['gene_symbol'] == 'TSC1'
        assert results['NM_000368.4:c.363+1dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000359.1(LRG_486p1):p.?', 'slr': 'NP_000359.1:p.?'}
        assert results['NM_000368.4:c.363+1dup']['submitted_variant'] == 'NM_000368.4:c.363+1dupG'
        assert results['NM_000368.4:c.363+1dup']['genome_context_intronic_sequence'] == 'NC_000009.11(NM_000368.4):c.363+1dup'
        assert results['NM_000368.4:c.363+1dup']['hgvs_lrg_variant'] == 'LRG_486:g.24048dup'
        assert results['NM_000368.4:c.363+1dup']['hgvs_transcript_variant'] == 'NM_000368.4:c.363+1dup'
        assert results['NM_000368.4:c.363+1dup']['hgvs_refseqgene_variant'] == 'NG_012386.1:g.24048dup'
        assert results['NM_000368.4:c.363+1dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.135800973dup', 'vcf': {'chr': 'chr9', 'ref': 'C', 'pos': '135800973', 'alt': 'CC'}}
        assert results['NM_000368.4:c.363+1dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.132925586dup', 'vcf': {'chr': 'chr9', 'ref': 'C', 'pos': '132925586', 'alt': 'CC'}}
        assert results['NM_000368.4:c.363+1dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.135800973dup', 'vcf': {'chr': '9', 'ref': 'C', 'pos': '135800973', 'alt': 'CC'}}
        assert results['NM_000368.4:c.363+1dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.132925586dup', 'vcf': {'chr': '9', 'ref': 'C', 'pos': '132925586', 'alt': 'CC'}}


    def test_variant47(self):
        variant = 'NM_000368.4:c.363dupG'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000368.4:c.363+1dup' in results.keys()
        assert results['NM_000368.4:c.363+1dup']['hgvs_lrg_transcript_variant'] == 'LRG_486t1:c.363+1dup'
        assert results['NM_000368.4:c.363+1dup']['refseqgene_context_intronic_sequence'] == 'NG_012386.1(NM_000368.4):c.363+1dup'
        assert results['NM_000368.4:c.363+1dup']['alt_genomic_loci'] == []
        assert results['NM_000368.4:c.363+1dup']['gene_symbol'] == 'TSC1'
        assert results['NM_000368.4:c.363+1dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000359.1(LRG_486p1):p.(Met122AspfsTer4)', 'slr': 'NP_000359.1:p.(M122Dfs*4)'}
        assert results['NM_000368.4:c.363+1dup']['submitted_variant'] == 'NM_000368.4:c.363dupG'
        assert results['NM_000368.4:c.363+1dup']['genome_context_intronic_sequence'] == 'NC_000009.11(NM_000368.4):c.363+1dup'
        assert results['NM_000368.4:c.363+1dup']['hgvs_lrg_variant'] == 'LRG_486:g.24048dup'
        assert results['NM_000368.4:c.363+1dup']['hgvs_transcript_variant'] == 'NM_000368.4:c.363+1dup'
        assert results['NM_000368.4:c.363+1dup']['hgvs_refseqgene_variant'] == 'NG_012386.1:g.24048dup'
        assert results['NM_000368.4:c.363+1dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.135800973dup', 'vcf': {'chr': 'chr9', 'ref': 'C', 'pos': '135800973', 'alt': 'CC'}}
        assert results['NM_000368.4:c.363+1dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.132925586dup', 'vcf': {'chr': 'chr9', 'ref': 'C', 'pos': '132925586', 'alt': 'CC'}}
        assert results['NM_000368.4:c.363+1dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.135800973dup', 'vcf': {'chr': '9', 'ref': 'C', 'pos': '135800973', 'alt': 'CC'}}
        assert results['NM_000368.4:c.363+1dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.132925586dup', 'vcf': {'chr': '9', 'ref': 'C', 'pos': '132925586', 'alt': 'CC'}}


    def test_variant48(self):
        variant = 'NM_000089.3:c.1033_1035delGTT'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000089.3:c.1035_1035+2del' in results.keys()
        assert results['NM_000089.3:c.1035_1035+2del']['hgvs_lrg_transcript_variant'] == 'LRG_2t1:c.1035_1035+2del'
        assert results['NM_000089.3:c.1035_1035+2del']['refseqgene_context_intronic_sequence'] == 'NG_007405.1(NM_000089.3):c.1035_1035+2del'
        assert results['NM_000089.3:c.1035_1035+2del']['alt_genomic_loci'] == []
        assert results['NM_000089.3:c.1035_1035+2del']['gene_symbol'] == 'COL1A2'
        assert results['NM_000089.3:c.1035_1035+2del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000080.2(LRG_2p1):p.(Val345del)', 'slr': 'NP_000080.2:p.(V345del)'}
        assert results['NM_000089.3:c.1035_1035+2del']['submitted_variant'] == 'NM_000089.3:c.1033_1035delGTT'
        assert results['NM_000089.3:c.1035_1035+2del']['genome_context_intronic_sequence'] == 'NC_000007.13(NM_000089.3):c.1035_1035+2del'
        assert results['NM_000089.3:c.1035_1035+2del']['hgvs_lrg_variant'] == 'LRG_2:g.20261_20263del'
        assert results['NM_000089.3:c.1035_1035+2del']['hgvs_transcript_variant'] == 'NM_000089.3:c.1035_1035+2del'
        assert results['NM_000089.3:c.1035_1035+2del']['hgvs_refseqgene_variant'] == 'NG_007405.1:g.20261_20263del'
        assert results['NM_000089.3:c.1035_1035+2del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.94039133_94039135del', 'vcf': {'chr': 'chr7', 'ref': 'CTTG', 'pos': '94039128', 'alt': 'C'}}
        assert results['NM_000089.3:c.1035_1035+2del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.94409821_94409823del', 'vcf': {'chr': 'chr7', 'ref': 'CTTG', 'pos': '94409816', 'alt': 'C'}}
        assert results['NM_000089.3:c.1035_1035+2del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.94039133_94039135del', 'vcf': {'chr': '7', 'ref': 'CTTG', 'pos': '94039128', 'alt': 'C'}}
        assert results['NM_000089.3:c.1035_1035+2del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.94409821_94409823del', 'vcf': {'chr': '7', 'ref': 'CTTG', 'pos': '94409816', 'alt': 'C'}}


    def test_variant49(self):
        variant = 'NM_000089.3:c.1035_1035+2delTGT'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000089.3:c.1035_1035+2del' in results.keys()
        assert results['NM_000089.3:c.1035_1035+2del']['hgvs_lrg_transcript_variant'] == 'LRG_2t1:c.1035_1035+2del'
        assert results['NM_000089.3:c.1035_1035+2del']['refseqgene_context_intronic_sequence'] == 'NG_007405.1(NM_000089.3):c.1035_1035+2del'
        assert results['NM_000089.3:c.1035_1035+2del']['alt_genomic_loci'] == []
        assert results['NM_000089.3:c.1035_1035+2del']['gene_symbol'] == 'COL1A2'
        assert results['NM_000089.3:c.1035_1035+2del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000080.2(LRG_2p1):p.(Val345del)', 'slr': 'NP_000080.2:p.(V345del)'}
        assert results['NM_000089.3:c.1035_1035+2del']['submitted_variant'] == 'NM_000089.3:c.1035_1035+2delTGT'
        assert results['NM_000089.3:c.1035_1035+2del']['genome_context_intronic_sequence'] == 'NC_000007.13(NM_000089.3):c.1035_1035+2del'
        assert results['NM_000089.3:c.1035_1035+2del']['hgvs_lrg_variant'] == 'LRG_2:g.20261_20263del'
        assert results['NM_000089.3:c.1035_1035+2del']['hgvs_transcript_variant'] == 'NM_000089.3:c.1035_1035+2del'
        assert results['NM_000089.3:c.1035_1035+2del']['hgvs_refseqgene_variant'] == 'NG_007405.1:g.20261_20263del'
        assert results['NM_000089.3:c.1035_1035+2del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.94039133_94039135del', 'vcf': {'chr': 'chr7', 'ref': 'CTTG', 'pos': '94039128', 'alt': 'C'}}
        assert results['NM_000089.3:c.1035_1035+2del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.94409821_94409823del', 'vcf': {'chr': 'chr7', 'ref': 'CTTG', 'pos': '94409816', 'alt': 'C'}}
        assert results['NM_000089.3:c.1035_1035+2del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.94039133_94039135del', 'vcf': {'chr': '7', 'ref': 'CTTG', 'pos': '94039128', 'alt': 'C'}}
        assert results['NM_000089.3:c.1035_1035+2del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.94409821_94409823del', 'vcf': {'chr': '7', 'ref': 'CTTG', 'pos': '94409816', 'alt': 'C'}}


    def test_variant50(self):
        variant = 'NM_000088.3:c.2023_2028delGCAAGA'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000088.3:c.2024_2028+1del' in results.keys()
        assert results['NM_000088.3:c.2024_2028+1del']['hgvs_lrg_transcript_variant'] == 'LRG_1t1:c.2024_2028+1del'
        assert results['NM_000088.3:c.2024_2028+1del']['refseqgene_context_intronic_sequence'] == 'NG_007400.1(NM_000088.3):c.2024_2028+1del'
        assert results['NM_000088.3:c.2024_2028+1del']['alt_genomic_loci'] == []
        assert results['NM_000088.3:c.2024_2028+1del']['gene_symbol'] == 'COL1A1'
        assert results['NM_000088.3:c.2024_2028+1del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000079.2(LRG_1p1):p.(Ala675_Arg676del)', 'slr': 'NP_000079.2:p.(A675_R676del)'}
        assert results['NM_000088.3:c.2024_2028+1del']['submitted_variant'] == 'NM_000088.3:c.2023_2028delGCAAGA'
        assert results['NM_000088.3:c.2024_2028+1del']['genome_context_intronic_sequence'] == 'NC_000017.10(NM_000088.3):c.2024_2028+1del'
        assert results['NM_000088.3:c.2024_2028+1del']['hgvs_lrg_variant'] == 'LRG_1:g.14656_14661del'
        assert results['NM_000088.3:c.2024_2028+1del']['hgvs_transcript_variant'] == 'NM_000088.3:c.2024_2028+1del'
        assert results['NM_000088.3:c.2024_2028+1del']['hgvs_refseqgene_variant'] == 'NG_007400.1:g.14656_14661del'
        assert results['NM_000088.3:c.2024_2028+1del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48269340_48269345del', 'vcf': {'chr': 'chr17', 'ref': 'ACTCTTG', 'pos': '48269339', 'alt': 'A'}}
        assert results['NM_000088.3:c.2024_2028+1del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50191979_50191984del', 'vcf': {'chr': 'chr17', 'ref': 'ACTCTTG', 'pos': '50191978', 'alt': 'A'}}
        assert results['NM_000088.3:c.2024_2028+1del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48269340_48269345del', 'vcf': {'chr': '17', 'ref': 'ACTCTTG', 'pos': '48269339', 'alt': 'A'}}
        assert results['NM_000088.3:c.2024_2028+1del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50191979_50191984del', 'vcf': {'chr': '17', 'ref': 'ACTCTTG', 'pos': '50191978', 'alt': 'A'}}


    def test_variant51(self):
        variant = 'NM_000089.3:c.938-1delG'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000089.3:c.938del' in results.keys()
        assert results['NM_000089.3:c.938del']['hgvs_lrg_transcript_variant'] == 'LRG_2t1:c.938del'
        assert results['NM_000089.3:c.938del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000089.3:c.938del']['alt_genomic_loci'] == []
        assert results['NM_000089.3:c.938del']['gene_symbol'] == 'COL1A2'
        assert results['NM_000089.3:c.938del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000080.2(LRG_2p1):p.(Gly313AlafsTer86)', 'slr': 'NP_000080.2:p.(G313Afs*86)'}
        assert results['NM_000089.3:c.938del']['submitted_variant'] == 'NM_000089.3:c.938-1delG'
        assert results['NM_000089.3:c.938del']['genome_context_intronic_sequence'] == ''
        assert results['NM_000089.3:c.938del']['hgvs_lrg_variant'] == 'LRG_2:g.20164del'
        assert results['NM_000089.3:c.938del']['hgvs_transcript_variant'] == 'NM_000089.3:c.938del'
        assert results['NM_000089.3:c.938del']['hgvs_refseqgene_variant'] == 'NG_007405.1:g.20164del'
        assert results['NM_000089.3:c.938del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.94039036del', 'vcf': {'chr': 'chr7', 'ref': 'AG', 'pos': '94039033', 'alt': 'A'}}
        assert results['NM_000089.3:c.938del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.94409724del', 'vcf': {'chr': 'chr7', 'ref': 'AG', 'pos': '94409721', 'alt': 'A'}}
        assert results['NM_000089.3:c.938del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.94039036del', 'vcf': {'chr': '7', 'ref': 'AG', 'pos': '94039033', 'alt': 'A'}}
        assert results['NM_000089.3:c.938del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.94409724del', 'vcf': {'chr': '7', 'ref': 'AG', 'pos': '94409721', 'alt': 'A'}}


    def test_variant52(self):
        variant = 'NM_000088.3:c.589G='
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000088.3:c.589G=' in results.keys()
        assert results['NM_000088.3:c.589G=']['hgvs_lrg_transcript_variant'] == 'LRG_1t1:c.589G='
        assert results['NM_000088.3:c.589G=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000088.3:c.589G=']['alt_genomic_loci'] == []
        assert results['NM_000088.3:c.589G=']['gene_symbol'] == 'COL1A1'
        assert results['NM_000088.3:c.589G=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000079.2(LRG_1p1):p.(Gly197=)', 'slr': 'NP_000079.2:p.(G197=)'}
        assert results['NM_000088.3:c.589G=']['submitted_variant'] == 'NM_000088.3:c.589G='
        assert results['NM_000088.3:c.589G=']['genome_context_intronic_sequence'] == ''
        assert results['NM_000088.3:c.589G=']['hgvs_lrg_variant'] == 'LRG_1:g.8638G='
        assert results['NM_000088.3:c.589G=']['hgvs_transcript_variant'] == 'NM_000088.3:c.589G='
        assert results['NM_000088.3:c.589G=']['hgvs_refseqgene_variant'] == 'NG_007400.1:g.8638G='
        assert results['NM_000088.3:c.589G=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275363C=', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '48275363', 'alt': u'C'}}
        assert results['NM_000088.3:c.589G=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198002C=', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '50198002', 'alt': u'C'}}
        assert results['NM_000088.3:c.589G=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275363C=', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '48275363', 'alt': u'C'}}
        assert results['NM_000088.3:c.589G=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198002C=', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '50198002', 'alt': u'C'}}


    def test_variant53(self):
        variant = 'NM_000088.3:c.642A='
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000088.3:c.642A=' in results.keys()
        assert results['NM_000088.3:c.642A=']['hgvs_lrg_transcript_variant'] == 'LRG_1t1:c.642A='
        assert results['NM_000088.3:c.642A=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000088.3:c.642A=']['alt_genomic_loci'] == []
        assert results['NM_000088.3:c.642A=']['gene_symbol'] == 'COL1A1'
        assert results['NM_000088.3:c.642A=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000079.2(LRG_1p1):p.(Ser214=)', 'slr': 'NP_000079.2:p.(S214=)'}
        assert results['NM_000088.3:c.642A=']['submitted_variant'] == 'NM_000088.3:c.642A='
        assert results['NM_000088.3:c.642A=']['genome_context_intronic_sequence'] == ''
        assert results['NM_000088.3:c.642A=']['hgvs_lrg_variant'] == 'LRG_1:g.8691A='
        assert results['NM_000088.3:c.642A=']['hgvs_transcript_variant'] == 'NM_000088.3:c.642A='
        assert results['NM_000088.3:c.642A=']['hgvs_refseqgene_variant'] == 'NG_007400.1:g.8691A='
        assert results['NM_000088.3:c.642A=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275310T=', 'vcf': {'chr': 'chr17', 'ref': u'T', 'pos': '48275310', 'alt': u'T'}}
        assert results['NM_000088.3:c.642A=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50197949T=', 'vcf': {'chr': 'chr17', 'ref': u'T', 'pos': '50197949', 'alt': u'T'}}
        assert results['NM_000088.3:c.642A=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275310T=', 'vcf': {'chr': '17', 'ref': u'T', 'pos': '48275310', 'alt': u'T'}}
        assert results['NM_000088.3:c.642A=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50197949T=', 'vcf': {'chr': '17', 'ref': u'T', 'pos': '50197949', 'alt': u'T'}}


    def test_variant54(self):
        variant = 'NM_000088.3:c.642+1GG>G'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'validation_warning_1' in results.keys()
        assert results['validation_warning_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['validation_warning_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['alt_genomic_loci'] == []
        assert results['validation_warning_1']['gene_symbol'] == ''
        assert results['validation_warning_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['validation_warning_1']['submitted_variant'] == 'NM_000088.3:c.642+1GG>G'
        assert results['validation_warning_1']['genome_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['hgvs_lrg_variant'] == ''
        assert results['validation_warning_1']['hgvs_transcript_variant'] == ''
        assert results['validation_warning_1']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['validation_warning_1']['primary_assembly_loci'].keys()

        assert results['flag'] == 'warning'

    def test_variant55(self):
        variant = 'NM_000088.3:c.589-2GG>G'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'validation_warning_1' in results.keys()
        assert results['validation_warning_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['validation_warning_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['alt_genomic_loci'] == []
        assert results['validation_warning_1']['gene_symbol'] == ''
        assert results['validation_warning_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['validation_warning_1']['submitted_variant'] == 'NM_000088.3:c.589-2GG>G'
        assert results['validation_warning_1']['genome_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['hgvs_lrg_variant'] == ''
        assert results['validation_warning_1']['hgvs_transcript_variant'] == ''
        assert results['validation_warning_1']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['validation_warning_1']['primary_assembly_loci'].keys()

        assert results['flag'] == 'warning'

    def test_variant56(self):
        variant = 'NM_000088.3:c.589-6_589-5insTTTT'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000088.3:c.589-5_589-4insTTTT' in results.keys()
        assert results['NM_000088.3:c.589-5_589-4insTTTT']['hgvs_lrg_transcript_variant'] == 'LRG_1t1:c.589-5_589-4insTTTT'
        assert results['NM_000088.3:c.589-5_589-4insTTTT']['refseqgene_context_intronic_sequence'] == 'NG_007400.1(NM_000088.3):c.589-5_589-4insTTTT'
        assert results['NM_000088.3:c.589-5_589-4insTTTT']['alt_genomic_loci'] == []
        assert results['NM_000088.3:c.589-5_589-4insTTTT']['gene_symbol'] == 'COL1A1'
        assert results['NM_000088.3:c.589-5_589-4insTTTT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000079.2(LRG_1p1):p.?', 'slr': 'NP_000079.2:p.?'}
        assert results['NM_000088.3:c.589-5_589-4insTTTT']['submitted_variant'] == 'NM_000088.3:c.589-6_589-5insTTTT'
        assert results['NM_000088.3:c.589-5_589-4insTTTT']['genome_context_intronic_sequence'] == 'NC_000017.10(NM_000088.3):c.589-5_589-4insTTTT'
        assert results['NM_000088.3:c.589-5_589-4insTTTT']['hgvs_lrg_variant'] == 'LRG_1:g.8633_8634insTTTT'
        assert results['NM_000088.3:c.589-5_589-4insTTTT']['hgvs_transcript_variant'] == 'NM_000088.3:c.589-5_589-4insTTTT'
        assert results['NM_000088.3:c.589-5_589-4insTTTT']['hgvs_refseqgene_variant'] == 'NG_007400.1:g.8633_8634insTTTT'
        assert results['NM_000088.3:c.589-5_589-4insTTTT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275367_48275368insAAAA', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '48275367', 'alt': u'GAAAA'}}
        assert results['NM_000088.3:c.589-5_589-4insTTTT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198006_50198007insAAAA', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '50198006', 'alt': u'GAAAA'}}
        assert results['NM_000088.3:c.589-5_589-4insTTTT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275367_48275368insAAAA', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '48275367', 'alt': u'GAAAA'}}
        assert results['NM_000088.3:c.589-5_589-4insTTTT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198006_50198007insAAAA', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '50198006', 'alt': u'GAAAA'}}


    def test_variant57(self):
        variant = 'NM_000088.3:c.642+3_642+4insAAAA'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000088.3:c.642+4_642+5insAAAA' in results.keys()
        assert results['NM_000088.3:c.642+4_642+5insAAAA']['hgvs_lrg_transcript_variant'] == 'LRG_1t1:c.642+4_642+5insAAAA'
        assert results['NM_000088.3:c.642+4_642+5insAAAA']['refseqgene_context_intronic_sequence'] == 'NG_007400.1(NM_000088.3):c.642+4_642+5insAAAA'
        assert results['NM_000088.3:c.642+4_642+5insAAAA']['alt_genomic_loci'] == []
        assert results['NM_000088.3:c.642+4_642+5insAAAA']['gene_symbol'] == 'COL1A1'
        assert results['NM_000088.3:c.642+4_642+5insAAAA']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000079.2(LRG_1p1):p.?', 'slr': 'NP_000079.2:p.?'}
        assert results['NM_000088.3:c.642+4_642+5insAAAA']['submitted_variant'] == 'NM_000088.3:c.642+3_642+4insAAAA'
        assert results['NM_000088.3:c.642+4_642+5insAAAA']['genome_context_intronic_sequence'] == 'NC_000017.10(NM_000088.3):c.642+4_642+5insAAAA'
        assert results['NM_000088.3:c.642+4_642+5insAAAA']['hgvs_lrg_variant'] == 'LRG_1:g.8695_8696insAAAA'
        assert results['NM_000088.3:c.642+4_642+5insAAAA']['hgvs_transcript_variant'] == 'NM_000088.3:c.642+4_642+5insAAAA'
        assert results['NM_000088.3:c.642+4_642+5insAAAA']['hgvs_refseqgene_variant'] == 'NG_007400.1:g.8695_8696insAAAA'
        assert results['NM_000088.3:c.642+4_642+5insAAAA']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275305_48275306insTTTT', 'vcf': {'chr': 'chr17', 'ref': 'C', 'pos': '48275305', 'alt': u'CTTTT'}}
        assert results['NM_000088.3:c.642+4_642+5insAAAA']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50197944_50197945insTTTT', 'vcf': {'chr': 'chr17', 'ref': 'C', 'pos': '50197944', 'alt': u'CTTTT'}}
        assert results['NM_000088.3:c.642+4_642+5insAAAA']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275305_48275306insTTTT', 'vcf': {'chr': '17', 'ref': 'C', 'pos': '48275305', 'alt': u'CTTTT'}}
        assert results['NM_000088.3:c.642+4_642+5insAAAA']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50197944_50197945insTTTT', 'vcf': {'chr': '17', 'ref': 'C', 'pos': '50197944', 'alt': u'CTTTT'}}


    def test_variant58(self):
        variant = 'NM_000088.3:c.589-4_589-3insTT'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000088.3:c.589-4_589-3insTT' in results.keys()
        assert results['NM_000088.3:c.589-4_589-3insTT']['hgvs_lrg_transcript_variant'] == 'LRG_1t1:c.589-4_589-3insTT'
        assert results['NM_000088.3:c.589-4_589-3insTT']['refseqgene_context_intronic_sequence'] == 'NG_007400.1(NM_000088.3):c.589-4_589-3insTT'
        assert results['NM_000088.3:c.589-4_589-3insTT']['alt_genomic_loci'] == []
        assert results['NM_000088.3:c.589-4_589-3insTT']['gene_symbol'] == 'COL1A1'
        assert results['NM_000088.3:c.589-4_589-3insTT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000079.2(LRG_1p1):p.?', 'slr': 'NP_000079.2:p.?'}
        assert results['NM_000088.3:c.589-4_589-3insTT']['submitted_variant'] == 'NM_000088.3:c.589-4_589-3insTT'
        assert results['NM_000088.3:c.589-4_589-3insTT']['genome_context_intronic_sequence'] == 'NC_000017.10(NM_000088.3):c.589-4_589-3insTT'
        assert results['NM_000088.3:c.589-4_589-3insTT']['hgvs_lrg_variant'] == 'LRG_1:g.8634_8635insTT'
        assert results['NM_000088.3:c.589-4_589-3insTT']['hgvs_transcript_variant'] == 'NM_000088.3:c.589-4_589-3insTT'
        assert results['NM_000088.3:c.589-4_589-3insTT']['hgvs_refseqgene_variant'] == 'NG_007400.1:g.8634_8635insTT'
        assert results['NM_000088.3:c.589-4_589-3insTT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275366_48275367insAA', 'vcf': {'chr': 'chr17', 'ref': 'T', 'pos': '48275366', 'alt': u'TAA'}}
        assert results['NM_000088.3:c.589-4_589-3insTT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198005_50198006insAA', 'vcf': {'chr': 'chr17', 'ref': 'T', 'pos': '50198005', 'alt': u'TAA'}}
        assert results['NM_000088.3:c.589-4_589-3insTT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275366_48275367insAA', 'vcf': {'chr': '17', 'ref': 'T', 'pos': '48275366', 'alt': u'TAA'}}
        assert results['NM_000088.3:c.589-4_589-3insTT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198005_50198006insAA', 'vcf': {'chr': '17', 'ref': 'T', 'pos': '50198005', 'alt': u'TAA'}}


    def test_variant59(self):
        variant = 'NM_000088.3:c.589-8del'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000088.3:c.589-7del' in results.keys()
        assert results['NM_000088.3:c.589-7del']['hgvs_lrg_transcript_variant'] == 'LRG_1t1:c.589-7del'
        assert results['NM_000088.3:c.589-7del']['refseqgene_context_intronic_sequence'] == 'NG_007400.1(NM_000088.3):c.589-7del'
        assert results['NM_000088.3:c.589-7del']['alt_genomic_loci'] == []
        assert results['NM_000088.3:c.589-7del']['gene_symbol'] == 'COL1A1'
        assert results['NM_000088.3:c.589-7del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000079.2(LRG_1p1):p.?', 'slr': 'NP_000079.2:p.?'}
        assert results['NM_000088.3:c.589-7del']['submitted_variant'] == 'NM_000088.3:c.589-8del'
        assert results['NM_000088.3:c.589-7del']['genome_context_intronic_sequence'] == 'NC_000017.10(NM_000088.3):c.589-7del'
        assert results['NM_000088.3:c.589-7del']['hgvs_lrg_variant'] == 'LRG_1:g.8631del'
        assert results['NM_000088.3:c.589-7del']['hgvs_transcript_variant'] == 'NM_000088.3:c.589-7del'
        assert results['NM_000088.3:c.589-7del']['hgvs_refseqgene_variant'] == 'NG_007400.1:g.8631del'
        assert results['NM_000088.3:c.589-7del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275370del', 'vcf': {'chr': 'chr17', 'ref': 'GA', 'pos': '48275369', 'alt': 'G'}}
        assert results['NM_000088.3:c.589-7del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198009del', 'vcf': {'chr': 'chr17', 'ref': 'GA', 'pos': '50198008', 'alt': 'G'}}
        assert results['NM_000088.3:c.589-7del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275370del', 'vcf': {'chr': '17', 'ref': 'GA', 'pos': '48275369', 'alt': 'G'}}
        assert results['NM_000088.3:c.589-7del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198009del', 'vcf': {'chr': '17', 'ref': 'GA', 'pos': '50198008', 'alt': 'G'}}


    def test_variant60(self):
        variant = 'NM_000527.4:c.-187_-185delCTC'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_000527.4:c.-187_-185del' in results.keys()
        assert results['NM_000527.4:c.-187_-185del']['hgvs_lrg_transcript_variant'] == 'LRG_274t1:c.-187_-185del'
        assert results['NM_000527.4:c.-187_-185del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000527.4:c.-187_-185del']['alt_genomic_loci'] == []
        assert results['NM_000527.4:c.-187_-185del']['gene_symbol'] == 'LDLR'
        assert results['NM_000527.4:c.-187_-185del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000518.1(LRG_274p1):p.?', 'slr': 'NP_000518.1:p.?'}
        assert results['NM_000527.4:c.-187_-185del']['submitted_variant'] == 'NM_000527.4:c.-187_-185delCTC'
        assert results['NM_000527.4:c.-187_-185del']['genome_context_intronic_sequence'] == ''
        assert results['NM_000527.4:c.-187_-185del']['hgvs_lrg_variant'] == 'LRG_274:g.4982_4984del'
        assert results['NM_000527.4:c.-187_-185del']['hgvs_transcript_variant'] == 'NM_000527.4:c.-187_-185del'
        assert results['NM_000527.4:c.-187_-185del']['hgvs_refseqgene_variant'] == 'NG_009060.1:g.4982_4984del'
        assert results['NM_000527.4:c.-187_-185del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.11200038_11200040del', 'vcf': {'chr': 'chr19', 'ref': 'ACTC', 'pos': '11200031', 'alt': 'A'}}
        assert results['NM_000527.4:c.-187_-185del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.11089362_11089364del', 'vcf': {'chr': 'chr19', 'ref': 'ACTC', 'pos': '11089355', 'alt': 'A'}}
        assert results['NM_000527.4:c.-187_-185del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.11200038_11200040del', 'vcf': {'chr': '19', 'ref': 'ACTC', 'pos': '11200031', 'alt': 'A'}}
        assert results['NM_000527.4:c.-187_-185del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.11089362_11089364del', 'vcf': {'chr': '19', 'ref': 'ACTC', 'pos': '11089355', 'alt': 'A'}}

        assert results['flag'] == 'gene_variant'

    def test_variant61(self):
        variant = 'NM_206933.2:c.6317C>G'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_206933.2:c.6317C>G' in results.keys()
        assert results['NM_206933.2:c.6317C>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_206933.2:c.6317C>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_206933.2:c.6317C>G']['alt_genomic_loci'] == []
        assert results['NM_206933.2:c.6317C>G']['gene_symbol'] == 'USH2A'
        assert results['NM_206933.2:c.6317C>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_996816.2:p.(Thr2106Arg)', 'slr': 'NP_996816.2:p.(T2106R)'}
        assert results['NM_206933.2:c.6317C>G']['submitted_variant'] == 'NM_206933.2:c.6317C>G'
        assert results['NM_206933.2:c.6317C>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_206933.2:c.6317C>G']['hgvs_lrg_variant'] == ''
        assert results['NM_206933.2:c.6317C>G']['hgvs_transcript_variant'] == 'NM_206933.2:c.6317C>G'
        assert results['NM_206933.2:c.6317C>G']['hgvs_refseqgene_variant'] == 'NG_009497.1:g.381958C>G'
        assert results['NM_206933.2:c.6317C>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.216219781A>C', 'vcf': {'chr': 'chr1', 'ref': 'A', 'pos': '216219781', 'alt': u'C'}}
        assert results['NM_206933.2:c.6317C>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.216046439A>C', 'vcf': {'chr': 'chr1', 'ref': 'A', 'pos': '216046439', 'alt': u'C'}}
        assert results['NM_206933.2:c.6317C>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.216219781A>C', 'vcf': {'chr': '1', 'ref': 'A', 'pos': '216219781', 'alt': u'C'}}
        assert results['NM_206933.2:c.6317C>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.216046439A>C', 'vcf': {'chr': '1', 'ref': 'A', 'pos': '216046439', 'alt': u'C'}}


    def test_variant62(self):
        variant = 'NC_000013.10:g.32929387T>C'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_000059.3:c.7397C=' in results.keys()
        assert results['NM_000059.3:c.7397C=']['hgvs_lrg_transcript_variant'] == 'LRG_293t1:c.7397C='
        assert results['NM_000059.3:c.7397C=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000059.3:c.7397C=']['alt_genomic_loci'] == []
        assert results['NM_000059.3:c.7397C=']['gene_symbol'] == 'BRCA2'
        assert results['NM_000059.3:c.7397C=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000050.2(LRG_293p1):p.(Ala2466=)', 'slr': 'NP_000050.2:p.(A2466=)'}
        assert results['NM_000059.3:c.7397C=']['submitted_variant'] == 'NC_000013.10:g.32929387T>C'
        assert results['NM_000059.3:c.7397C=']['genome_context_intronic_sequence'] == ''
        assert results['NM_000059.3:c.7397C=']['hgvs_lrg_variant'] == 'LRG_293:g.44771C='
        assert results['NM_000059.3:c.7397C=']['hgvs_transcript_variant'] == 'NM_000059.3:c.7397C='
        assert results['NM_000059.3:c.7397C=']['hgvs_refseqgene_variant'] == 'NG_012772.3:g.44771C='
        assert results['NM_000059.3:c.7397C=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000013.10:g.32929387T>C', 'vcf': {'chr': 'chr13', 'ref': 'T', 'pos': '32929387', 'alt': 'C'}}
        assert results['NM_000059.3:c.7397C=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000013.11:g.32355250T>C', 'vcf': {'chr': 'chr13', 'ref': 'T', 'pos': '32355250', 'alt': 'C'}}
        assert results['NM_000059.3:c.7397C=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000013.10:g.32929387T>C', 'vcf': {'chr': '13', 'ref': 'T', 'pos': '32929387', 'alt': 'C'}}
        assert results['NM_000059.3:c.7397C=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000013.11:g.32355250T>C', 'vcf': {'chr': '13', 'ref': 'T', 'pos': '32355250', 'alt': 'C'}}

        assert results['flag'] == 'gene_variant'

    def test_variant63(self):
        variant = 'NM_015102.3:c.2818-2T>A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_015102.3:c.2818-2T>A' in results.keys()
        assert results['NM_015102.3:c.2818-2T>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_015102.3:c.2818-2T>A']['refseqgene_context_intronic_sequence'] == 'NG_011724.2(NM_015102.3):c.2818-2A='
        assert results['NM_015102.3:c.2818-2T>A']['alt_genomic_loci'] == []
        assert results['NM_015102.3:c.2818-2T>A']['gene_symbol'] == 'NPHP4'
        assert results['NM_015102.3:c.2818-2T>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_055917.1:p.?', 'slr': 'NP_055917.1:p.?'}
        assert results['NM_015102.3:c.2818-2T>A']['submitted_variant'] == 'NM_015102.3:c.2818-2T>A'
        assert results['NM_015102.3:c.2818-2T>A']['genome_context_intronic_sequence'] == 'NC_000001.10(NM_015102.3):c.2818-2T>A'
        assert results['NM_015102.3:c.2818-2T>A']['hgvs_lrg_variant'] == ''
        assert results['NM_015102.3:c.2818-2T>A']['hgvs_transcript_variant'] == 'NM_015102.3:c.2818-2T>A'
        assert results['NM_015102.3:c.2818-2T>A']['hgvs_refseqgene_variant'] == 'NG_011724.2:g.122370A='
        assert results['NM_015102.3:c.2818-2T>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.5935162A>T', 'vcf': {'chr': 'chr1', 'ref': u'A', 'pos': '5935162', 'alt': u'T'}}
        assert 'hg38' not in results['NM_015102.3:c.2818-2T>A']['primary_assembly_loci'].keys()
        assert results['NM_015102.3:c.2818-2T>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.5935162A>T', 'vcf': {'chr': '1', 'ref': u'A', 'pos': '5935162', 'alt': u'T'}}
        assert 'grch38' not in results['NM_015102.3:c.2818-2T>A']['primary_assembly_loci'].keys()

        assert results['flag'] == 'gene_variant'

    def test_variant64(self):
        variant = '19-41123094-G-GG'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_001042544.1:c.3233_3235=' in results.keys()
        assert results['NM_001042544.1:c.3233_3235=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001042544.1:c.3233_3235=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001042544.1:c.3233_3235=']['alt_genomic_loci'] == []
        assert results['NM_001042544.1:c.3233_3235=']['gene_symbol'] == 'LTBP4'
        assert results['NM_001042544.1:c.3233_3235=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001036009.1:p.(Gln1078=)', 'slr': 'NP_001036009.1:p.(Q1078=)'}
        assert results['NM_001042544.1:c.3233_3235=']['submitted_variant'] == '19-41123094-G-GG'
        assert results['NM_001042544.1:c.3233_3235=']['genome_context_intronic_sequence'] == ''
        assert results['NM_001042544.1:c.3233_3235=']['hgvs_lrg_variant'] == ''
        assert results['NM_001042544.1:c.3233_3235=']['hgvs_transcript_variant'] == 'NM_001042544.1:c.3233_3235='
        assert results['NM_001042544.1:c.3233_3235=']['hgvs_refseqgene_variant'] == 'NG_021201.1:g.29022_29024='
        assert results['NM_001042544.1:c.3233_3235=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.41123095dup', 'vcf': {'chr': 'chr19', 'ref': 'G', 'pos': '41123094', 'alt': u'GG'}}
        assert results['NM_001042544.1:c.3233_3235=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.40617187_40617189=', 'vcf': {'chr': 'chr19', 'ref': u'AGG', 'pos': '40617187', 'alt': u'AGG'}}
        assert results['NM_001042544.1:c.3233_3235=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.41123095dup', 'vcf': {'chr': '19', 'ref': 'G', 'pos': '41123094', 'alt': u'GG'}}
        assert results['NM_001042544.1:c.3233_3235=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.40617187_40617189=', 'vcf': {'chr': '19', 'ref': u'AGG', 'pos': '40617187', 'alt': u'AGG'}}

        assert 'NM_001042545.1:c.3032_3034=' in results.keys()
        assert results['NM_001042545.1:c.3032_3034=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001042545.1:c.3032_3034=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001042545.1:c.3032_3034=']['alt_genomic_loci'] == []
        assert results['NM_001042545.1:c.3032_3034=']['gene_symbol'] == 'LTBP4'
        assert results['NM_001042545.1:c.3032_3034=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001036010.1:p.(Gln1011=)', 'slr': 'NP_001036010.1:p.(Q1011=)'}
        assert results['NM_001042545.1:c.3032_3034=']['submitted_variant'] == '19-41123094-G-GG'
        assert results['NM_001042545.1:c.3032_3034=']['genome_context_intronic_sequence'] == ''
        assert results['NM_001042545.1:c.3032_3034=']['hgvs_lrg_variant'] == ''
        assert results['NM_001042545.1:c.3032_3034=']['hgvs_transcript_variant'] == 'NM_001042545.1:c.3032_3034='
        assert results['NM_001042545.1:c.3032_3034=']['hgvs_refseqgene_variant'] == 'NG_021201.1:g.29022_29024='
        assert results['NM_001042545.1:c.3032_3034=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.41123095dup', 'vcf': {'chr': 'chr19', 'ref': 'G', 'pos': '41123094', 'alt': u'GG'}}
        assert results['NM_001042545.1:c.3032_3034=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.40617187_40617189=', 'vcf': {'chr': 'chr19', 'ref': u'AGG', 'pos': '40617187', 'alt': u'AGG'}}
        assert results['NM_001042545.1:c.3032_3034=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.41123095dup', 'vcf': {'chr': '19', 'ref': 'G', 'pos': '41123094', 'alt': u'GG'}}
        assert results['NM_001042545.1:c.3032_3034=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.40617187_40617189=', 'vcf': {'chr': '19', 'ref': u'AGG', 'pos': '40617187', 'alt': u'AGG'}}

        assert 'NM_003573.2:c.3122_3124=' in results.keys()
        assert results['NM_003573.2:c.3122_3124=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003573.2:c.3122_3124=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003573.2:c.3122_3124=']['alt_genomic_loci'] == []
        assert results['NM_003573.2:c.3122_3124=']['gene_symbol'] == 'LTBP4'
        assert results['NM_003573.2:c.3122_3124=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003564.2:p.(Gln1041=)', 'slr': 'NP_003564.2:p.(Q1041=)'}
        assert results['NM_003573.2:c.3122_3124=']['submitted_variant'] == '19-41123094-G-GG'
        assert results['NM_003573.2:c.3122_3124=']['genome_context_intronic_sequence'] == ''
        assert results['NM_003573.2:c.3122_3124=']['hgvs_lrg_variant'] == ''
        assert results['NM_003573.2:c.3122_3124=']['hgvs_transcript_variant'] == 'NM_003573.2:c.3122_3124='
        assert results['NM_003573.2:c.3122_3124=']['hgvs_refseqgene_variant'] == 'NG_021201.1:g.29022_29024='
        assert results['NM_003573.2:c.3122_3124=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.41123095dup', 'vcf': {'chr': 'chr19', 'ref': 'G', 'pos': '41123094', 'alt': u'GG'}}
        assert results['NM_003573.2:c.3122_3124=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.40617187_40617189=', 'vcf': {'chr': 'chr19', 'ref': u'AGG', 'pos': '40617187', 'alt': u'AGG'}}
        assert results['NM_003573.2:c.3122_3124=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.41123095dup', 'vcf': {'chr': '19', 'ref': 'G', 'pos': '41123094', 'alt': u'GG'}}
        assert results['NM_003573.2:c.3122_3124=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.40617187_40617189=', 'vcf': {'chr': '19', 'ref': u'AGG', 'pos': '40617187', 'alt': u'AGG'}}


    def test_variant65(self):
        variant = '15-72105928-AC-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_014249.2:c.946_949=' in results.keys()
        assert results['NM_014249.2:c.946_949=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_014249.2:c.946_949=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_014249.2:c.946_949=']['alt_genomic_loci'] == []
        assert results['NM_014249.2:c.946_949=']['gene_symbol'] == 'NR2E3'
        assert results['NM_014249.2:c.946_949=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_055064.1:p.(Asp316=)', 'slr': 'NP_055064.1:p.(D316=)'}
        assert results['NM_014249.2:c.946_949=']['submitted_variant'] == '15-72105928-AC-A'
        assert results['NM_014249.2:c.946_949=']['genome_context_intronic_sequence'] == ''
        assert results['NM_014249.2:c.946_949=']['hgvs_lrg_variant'] == ''
        assert results['NM_014249.2:c.946_949=']['hgvs_transcript_variant'] == 'NM_014249.2:c.946_949='
        assert results['NM_014249.2:c.946_949=']['hgvs_refseqgene_variant'] == ''
        assert results['NM_014249.2:c.946_949=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105933del', 'vcf': {'chr': 'chr15', 'ref': 'AC', 'pos': '72105928', 'alt': 'A'}}
        assert 'hg38' not in results['NM_014249.2:c.946_949=']['primary_assembly_loci'].keys()
        assert results['NM_014249.2:c.946_949=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105933del', 'vcf': {'chr': '15', 'ref': 'AC', 'pos': '72105928', 'alt': 'A'}}
        assert 'grch38' not in results['NM_014249.2:c.946_949=']['primary_assembly_loci'].keys()

        assert results['flag'] == 'gene_variant'
        assert 'NM_014249.3:c.946_949=' in results.keys()
        assert results['NM_014249.3:c.946_949=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_014249.3:c.946_949=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_014249.3:c.946_949=']['alt_genomic_loci'] == []
        assert results['NM_014249.3:c.946_949=']['gene_symbol'] == 'NR2E3'
        assert results['NM_014249.3:c.946_949=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_055064.1:p.(Asp316=)', 'slr': 'NP_055064.1:p.(D316=)'}
        assert results['NM_014249.3:c.946_949=']['submitted_variant'] == '15-72105928-AC-A'
        assert results['NM_014249.3:c.946_949=']['genome_context_intronic_sequence'] == ''
        assert results['NM_014249.3:c.946_949=']['hgvs_lrg_variant'] == ''
        assert results['NM_014249.3:c.946_949=']['hgvs_transcript_variant'] == 'NM_014249.3:c.946_949='
        assert results['NM_014249.3:c.946_949=']['hgvs_refseqgene_variant'] == 'NG_009113.1:g.8034_8037='
        assert results['NM_014249.3:c.946_949=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105933del', 'vcf': {'chr': 'chr15', 'ref': 'AC', 'pos': '72105928', 'alt': 'A'}}
        assert results['NM_014249.3:c.946_949=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000015.10:g.71813586_71813591=', 'vcf': {'chr': 'chr15', 'ref': u'GGACCC', 'pos': '71813586', 'alt': u'GGACCC'}}
        assert results['NM_014249.3:c.946_949=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105933del', 'vcf': {'chr': '15', 'ref': 'AC', 'pos': '72105928', 'alt': 'A'}}
        assert results['NM_014249.3:c.946_949=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000015.10:g.71813586_71813591=', 'vcf': {'chr': '15', 'ref': u'GGACCC', 'pos': '71813586', 'alt': u'GGACCC'}}

        assert 'NM_016346.2:c.946_949=' in results.keys()
        assert results['NM_016346.2:c.946_949=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_016346.2:c.946_949=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_016346.2:c.946_949=']['alt_genomic_loci'] == []
        assert results['NM_016346.2:c.946_949=']['gene_symbol'] == 'NR2E3'
        assert results['NM_016346.2:c.946_949=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_057430.1:p.(Asp316=)', 'slr': 'NP_057430.1:p.(D316=)'}
        assert results['NM_016346.2:c.946_949=']['submitted_variant'] == '15-72105928-AC-A'
        assert results['NM_016346.2:c.946_949=']['genome_context_intronic_sequence'] == ''
        assert results['NM_016346.2:c.946_949=']['hgvs_lrg_variant'] == ''
        assert results['NM_016346.2:c.946_949=']['hgvs_transcript_variant'] == 'NM_016346.2:c.946_949='
        assert results['NM_016346.2:c.946_949=']['hgvs_refseqgene_variant'] == ''
        assert results['NM_016346.2:c.946_949=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105933del', 'vcf': {'chr': 'chr15', 'ref': 'AC', 'pos': '72105928', 'alt': 'A'}}
        assert 'hg38' not in results['NM_016346.2:c.946_949=']['primary_assembly_loci'].keys()
        assert results['NM_016346.2:c.946_949=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105933del', 'vcf': {'chr': '15', 'ref': 'AC', 'pos': '72105928', 'alt': 'A'}}
        assert 'grch38' not in results['NM_016346.2:c.946_949=']['primary_assembly_loci'].keys()

        assert 'NM_016346.3:c.946_949=' in results.keys()
        assert results['NM_016346.3:c.946_949=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_016346.3:c.946_949=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_016346.3:c.946_949=']['alt_genomic_loci'] == []
        assert results['NM_016346.3:c.946_949=']['gene_symbol'] == 'NR2E3'
        assert results['NM_016346.3:c.946_949=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_057430.1:p.(Asp316=)', 'slr': 'NP_057430.1:p.(D316=)'}
        assert results['NM_016346.3:c.946_949=']['submitted_variant'] == '15-72105928-AC-A'
        assert results['NM_016346.3:c.946_949=']['genome_context_intronic_sequence'] == ''
        assert results['NM_016346.3:c.946_949=']['hgvs_lrg_variant'] == ''
        assert results['NM_016346.3:c.946_949=']['hgvs_transcript_variant'] == 'NM_016346.3:c.946_949='
        assert results['NM_016346.3:c.946_949=']['hgvs_refseqgene_variant'] == ''
        assert results['NM_016346.3:c.946_949=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105933del', 'vcf': {'chr': 'chr15', 'ref': 'AC', 'pos': '72105928', 'alt': 'A'}}
        assert results['NM_016346.3:c.946_949=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000015.10:g.71813587_71813590=', 'vcf': {'chr': 'chr15', 'ref': u'GACC', 'pos': '71813587', 'alt': u'GACC'}}
        assert results['NM_016346.3:c.946_949=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105933del', 'vcf': {'chr': '15', 'ref': 'AC', 'pos': '72105928', 'alt': 'A'}}
        assert results['NM_016346.3:c.946_949=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000015.10:g.71813587_71813590=', 'vcf': {'chr': '15', 'ref': u'GACC', 'pos': '71813587', 'alt': u'GACC'}}


    def test_variant66(self):
        variant = '12-122064773-CCCGCCA-C'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_032790.3:c.126_128=' in results.keys()
        assert results['NM_032790.3:c.126_128=']['hgvs_lrg_transcript_variant'] == 'LRG_93t1:c.126_128='
        assert results['NM_032790.3:c.126_128=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_032790.3:c.126_128=']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_004504303.2:g.302883_302888del', 'vcf': {'chr': 'HG1595_PATCH', 'ref': 'CCCGCCA', 'pos': '302871', 'alt': 'C'}}}]
        assert results['NM_032790.3:c.126_128=']['gene_symbol'] == 'ORAI1'
        assert results['NM_032790.3:c.126_128=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_116179.2(LRG_93p1):p.(Ala42=)', 'slr': 'NP_116179.2:p.(A42=)'}
        assert results['NM_032790.3:c.126_128=']['submitted_variant'] == '12-122064773-CCCGCCA-C'
        assert results['NM_032790.3:c.126_128=']['genome_context_intronic_sequence'] == ''
        assert results['NM_032790.3:c.126_128=']['hgvs_lrg_variant'] == 'LRG_93:g.5299_5301='
        assert results['NM_032790.3:c.126_128=']['hgvs_transcript_variant'] == 'NM_032790.3:c.126_128='
        assert results['NM_032790.3:c.126_128=']['hgvs_refseqgene_variant'] == 'NG_007500.1:g.5299_5301='
        assert results['NM_032790.3:c.126_128=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.122064785_122064790del', 'vcf': {'chr': 'chr12', 'ref': 'CCCGCCA', 'pos': '122064773', 'alt': 'C'}}
        assert results['NM_032790.3:c.126_128=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.121626873_121626875=', 'vcf': {'chr': 'chr12', 'ref': u'CCC', 'pos': '121626873', 'alt': u'CCC'}}
        assert results['NM_032790.3:c.126_128=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.122064785_122064790del', 'vcf': {'chr': '12', 'ref': 'CCCGCCA', 'pos': '122064773', 'alt': 'C'}}
        assert results['NM_032790.3:c.126_128=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.121626873_121626875=', 'vcf': {'chr': '12', 'ref': u'CCC', 'pos': '121626873', 'alt': u'CCC'}}


    def test_variant67(self):
        variant = '12-122064774-CCGCCA-CCGCCA'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_032790.3:c.132_137dup' in results.keys()
        assert results['NM_032790.3:c.132_137dup']['hgvs_lrg_transcript_variant'] == 'LRG_93t1:c.132_137dup'
        assert results['NM_032790.3:c.132_137dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_032790.3:c.132_137dup']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_004504303.2:g.302868_302887=', 'vcf': {'chr': 'HG1595_PATCH', 'ref': 'GGCCCCGCCACCGCCACCGC', 'pos': '302868', 'alt': 'GGCCCCGCCACCGCCACCGC'}}}]
        assert results['NM_032790.3:c.132_137dup']['gene_symbol'] == 'ORAI1'
        assert results['NM_032790.3:c.132_137dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_116179.2(LRG_93p1):p.(Pro46_Pro47dup)', 'slr': 'NP_116179.2:p.(P46_P47dup)'}
        assert results['NM_032790.3:c.132_137dup']['submitted_variant'] == '12-122064774-CCGCCA-CCGCCA'
        assert results['NM_032790.3:c.132_137dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_032790.3:c.132_137dup']['hgvs_lrg_variant'] == 'LRG_93:g.5305_5310dup'
        assert results['NM_032790.3:c.132_137dup']['hgvs_transcript_variant'] == 'NM_032790.3:c.132_137dup'
        assert results['NM_032790.3:c.132_137dup']['hgvs_refseqgene_variant'] == 'NG_007500.1:g.5305_5310dup'
        assert results['NM_032790.3:c.132_137dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.122064770_122064789=', 'vcf': {'chr': 'chr12', 'ref': 'GGCCCCGCCACCGCCACCGC', 'pos': '122064770', 'alt': 'GGCCCCGCCACCGCCACCGC'}}
        assert results['NM_032790.3:c.132_137dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.121626879_121626884dup', 'vcf': {'chr': 'chr12', 'ref': 'CCGCCA', 'pos': '121626874', 'alt': u'CCGCCACCGCCA'}}
        assert results['NM_032790.3:c.132_137dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.122064770_122064789=', 'vcf': {'chr': '12', 'ref': 'GGCCCCGCCACCGCCACCGC', 'pos': '122064770', 'alt': 'GGCCCCGCCACCGCCACCGC'}}
        assert results['NM_032790.3:c.132_137dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.121626879_121626884dup', 'vcf': {'chr': '12', 'ref': 'CCGCCA', 'pos': '121626874', 'alt': u'CCGCCACCGCCA'}}

        assert results['flag'] == 'gene_variant'

    def test_variant68(self):
        variant = '12-122064773-CCCGCCACCGCCACCGC-CCCGCCACCGCCGCCGTC'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_032790.3:c.132_135delinsGCCGT' in results.keys()
        assert results['NM_032790.3:c.132_135delinsGCCGT']['hgvs_lrg_transcript_variant'] == 'LRG_93t1:c.132_135delinsGCCGT'
        assert results['NM_032790.3:c.132_135delinsGCCGT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_032790.3:c.132_135delinsGCCGT']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_004504303.2:g.302883_302886delinsGCCGT', 'vcf': {'chr': 'HG1595_PATCH', 'ref': 'ACCG', 'pos': '302883', 'alt': u'GCCGT'}}}]
        assert results['NM_032790.3:c.132_135delinsGCCGT']['gene_symbol'] == 'ORAI1'
        assert results['NM_032790.3:c.132_135delinsGCCGT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_116179.2(LRG_93p1):p.(Pro46SerfsTer42)', 'slr': 'NP_116179.2:p.(P46Sfs*42)'}
        assert results['NM_032790.3:c.132_135delinsGCCGT']['submitted_variant'] == '12-122064773-CCCGCCACCGCCACCGC-CCCGCCACCGCCGCCGTC'
        assert results['NM_032790.3:c.132_135delinsGCCGT']['genome_context_intronic_sequence'] == ''
        assert results['NM_032790.3:c.132_135delinsGCCGT']['hgvs_lrg_variant'] == 'LRG_93:g.5305_5308delinsGCCGT'
        assert results['NM_032790.3:c.132_135delinsGCCGT']['hgvs_transcript_variant'] == 'NM_032790.3:c.132_135delinsGCCGT'
        assert results['NM_032790.3:c.132_135delinsGCCGT']['hgvs_refseqgene_variant'] == 'NG_007500.1:g.5305_5308delinsGCCGT'
        assert results['NM_032790.3:c.132_135delinsGCCGT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.122064785_122064788delinsGCCGT', 'vcf': {'chr': 'chr12', 'ref': 'ACCG', 'pos': '122064785', 'alt': u'GCCGT'}}
        assert results['NM_032790.3:c.132_135delinsGCCGT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.121626879_121626882delinsGCCGT', 'vcf': {'chr': 'chr12', 'ref': 'ACCG', 'pos': '121626879', 'alt': u'GCCGT'}}
        assert results['NM_032790.3:c.132_135delinsGCCGT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.122064785_122064788delinsGCCGT', 'vcf': {'chr': '12', 'ref': 'ACCG', 'pos': '122064785', 'alt': u'GCCGT'}}
        assert results['NM_032790.3:c.132_135delinsGCCGT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.121626879_121626882delinsGCCGT', 'vcf': {'chr': '12', 'ref': 'ACCG', 'pos': '121626879', 'alt': u'GCCGT'}}


    def test_variant69(self):
        variant = 'NC_000012.11:g.122064777C>A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_032790.3:c.129_130insACACCG' in results.keys()
        assert results['NM_032790.3:c.129_130insACACCG']['hgvs_lrg_transcript_variant'] == 'LRG_93t1:c.129_130insACACCG'
        assert results['NM_032790.3:c.129_130insACACCG']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_032790.3:c.129_130insACACCG']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_004504303.2:g.302875C>A', 'vcf': {'chr': 'HG1595_PATCH', 'ref': 'C', 'pos': '302875', 'alt': 'A'}}}]
        assert results['NM_032790.3:c.129_130insACACCG']['gene_symbol'] == 'ORAI1'
        assert results['NM_032790.3:c.129_130insACACCG']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_116179.2(LRG_93p1):p.(Pro43_Pro44insThrPro)', 'slr': 'NP_116179.2:p.(P43_P44insTP)'}
        assert results['NM_032790.3:c.129_130insACACCG']['submitted_variant'] == 'NC_000012.11:g.122064777C>A'
        assert results['NM_032790.3:c.129_130insACACCG']['genome_context_intronic_sequence'] == ''
        assert results['NM_032790.3:c.129_130insACACCG']['hgvs_lrg_variant'] == 'LRG_93:g.5302_5303insACACCG'
        assert results['NM_032790.3:c.129_130insACACCG']['hgvs_transcript_variant'] == 'NM_032790.3:c.129_130insACACCG'
        assert results['NM_032790.3:c.129_130insACACCG']['hgvs_refseqgene_variant'] == 'NG_007500.1:g.5302_5303insACACCG'
        assert results['NM_032790.3:c.129_130insACACCG']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.122064777C>A', 'vcf': {'chr': 'chr12', 'ref': 'C', 'pos': '122064777', 'alt': 'A'}}
        assert results['NM_032790.3:c.129_130insACACCG']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.121626876_121626877insACACCG', 'vcf': {'chr': 'chr12', 'ref': 'C', 'pos': '121626873', 'alt': u'CCCGACA'}}
        assert results['NM_032790.3:c.129_130insACACCG']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.122064777C>A', 'vcf': {'chr': '12', 'ref': 'C', 'pos': '122064777', 'alt': 'A'}}
        assert results['NM_032790.3:c.129_130insACACCG']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.121626876_121626877insACACCG', 'vcf': {'chr': '12', 'ref': 'C', 'pos': '121626873', 'alt': u'CCCGACA'}}


    def test_variant70(self):
        variant = 'NC_000012.11:g.122064776delG'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_032790.3:c.128_129insCCACC' in results.keys()
        assert results['NM_032790.3:c.128_129insCCACC']['hgvs_lrg_transcript_variant'] == 'LRG_93t1:c.128_129insCCACC'
        assert results['NM_032790.3:c.128_129insCCACC']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_032790.3:c.128_129insCCACC']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_004504303.2:g.302874del', 'vcf': {'chr': 'HG1595_PATCH', 'ref': 'CG', 'pos': '302873', 'alt': 'C'}}}]
        assert results['NM_032790.3:c.128_129insCCACC']['gene_symbol'] == 'ORAI1'
        assert results['NM_032790.3:c.128_129insCCACC']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_116179.2(LRG_93p1):p.(Pro44HisfsTer22)', 'slr': 'NP_116179.2:p.(P44Hfs*22)'}
        assert results['NM_032790.3:c.128_129insCCACC']['submitted_variant'] == 'NC_000012.11:g.122064776delG'
        assert results['NM_032790.3:c.128_129insCCACC']['genome_context_intronic_sequence'] == ''
        assert results['NM_032790.3:c.128_129insCCACC']['hgvs_lrg_variant'] == 'LRG_93:g.5301_5302insCCACC'
        assert results['NM_032790.3:c.128_129insCCACC']['hgvs_transcript_variant'] == 'NM_032790.3:c.128_129insCCACC'
        assert results['NM_032790.3:c.128_129insCCACC']['hgvs_refseqgene_variant'] == 'NG_007500.1:g.5301_5302insCCACC'
        assert results['NM_032790.3:c.128_129insCCACC']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.122064776del', 'vcf': {'chr': 'chr12', 'ref': 'CG', 'pos': '122064775', 'alt': 'C'}}
        assert results['NM_032790.3:c.128_129insCCACC']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.121626875_121626876insCCACC', 'vcf': {'chr': 'chr12', 'ref': 'C', 'pos': '121626873', 'alt': u'CCCCCA'}}
        assert results['NM_032790.3:c.128_129insCCACC']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.122064776del', 'vcf': {'chr': '12', 'ref': 'CG', 'pos': '122064775', 'alt': 'C'}}
        assert results['NM_032790.3:c.128_129insCCACC']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.121626875_121626876insCCACC', 'vcf': {'chr': '12', 'ref': 'C', 'pos': '121626873', 'alt': u'CCCCCA'}}


    def test_variant71(self):
        variant = 'NC_000012.11:g.122064776dupG'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_032790.3:c.129_130insGCCACCG' in results.keys()
        assert results['NM_032790.3:c.129_130insGCCACCG']['hgvs_lrg_transcript_variant'] == 'LRG_93t1:c.129_130insGCCACCG'
        assert results['NM_032790.3:c.129_130insGCCACCG']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_032790.3:c.129_130insGCCACCG']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_004504303.2:g.302874dup', 'vcf': {'chr': 'HG1595_PATCH', 'ref': 'G', 'pos': '302874', 'alt': 'GG'}}}]
        assert results['NM_032790.3:c.129_130insGCCACCG']['gene_symbol'] == 'ORAI1'
        assert results['NM_032790.3:c.129_130insGCCACCG']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_116179.2(LRG_93p1):p.(Pro44AlafsTer46)', 'slr': 'NP_116179.2:p.(P44Afs*46)'}
        assert results['NM_032790.3:c.129_130insGCCACCG']['submitted_variant'] == 'NC_000012.11:g.122064776dupG'
        assert results['NM_032790.3:c.129_130insGCCACCG']['genome_context_intronic_sequence'] == ''
        assert results['NM_032790.3:c.129_130insGCCACCG']['hgvs_lrg_variant'] == 'LRG_93:g.5302_5303insGCCACCG'
        assert results['NM_032790.3:c.129_130insGCCACCG']['hgvs_transcript_variant'] == 'NM_032790.3:c.129_130insGCCACCG'
        assert results['NM_032790.3:c.129_130insGCCACCG']['hgvs_refseqgene_variant'] == 'NG_007500.1:g.5302_5303insGCCACCG'
        assert results['NM_032790.3:c.129_130insGCCACCG']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.122064776dup', 'vcf': {'chr': 'chr12', 'ref': 'G', 'pos': '122064776', 'alt': 'GG'}}
        assert results['NM_032790.3:c.129_130insGCCACCG']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.121626876_121626877insGCCACCG', 'vcf': {'chr': 'chr12', 'ref': 'C', 'pos': '121626873', 'alt': u'CCCGGCCA'}}
        assert results['NM_032790.3:c.129_130insGCCACCG']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.122064776dup', 'vcf': {'chr': '12', 'ref': 'G', 'pos': '122064776', 'alt': 'GG'}}
        assert results['NM_032790.3:c.129_130insGCCACCG']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.121626876_121626877insGCCACCG', 'vcf': {'chr': '12', 'ref': 'C', 'pos': '121626873', 'alt': u'CCCGGCCA'}}


    def test_variant72(self):
        variant = 'NC_000012.11:g.122064776_122064777insTTT'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_032790.3:c.129_130insTTTCCACCG' in results.keys()
        assert results['NM_032790.3:c.129_130insTTTCCACCG']['hgvs_lrg_transcript_variant'] == 'LRG_93t1:c.129_130insTTTCCACCG'
        assert results['NM_032790.3:c.129_130insTTTCCACCG']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_032790.3:c.129_130insTTTCCACCG']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_004504303.2:g.302874_302875insTTT', 'vcf': {'chr': 'HG1595_PATCH', 'ref': 'G', 'pos': '302874', 'alt': 'GTTT'}}}]
        assert results['NM_032790.3:c.129_130insTTTCCACCG']['gene_symbol'] == 'ORAI1'
        assert results['NM_032790.3:c.129_130insTTTCCACCG']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_116179.2(LRG_93p1):p.(Pro43_Pro44insPheProPro)', 'slr': 'NP_116179.2:p.(P43_P44insFPP)'}
        assert results['NM_032790.3:c.129_130insTTTCCACCG']['submitted_variant'] == 'NC_000012.11:g.122064776_122064777insTTT'
        assert results['NM_032790.3:c.129_130insTTTCCACCG']['genome_context_intronic_sequence'] == ''
        assert results['NM_032790.3:c.129_130insTTTCCACCG']['hgvs_lrg_variant'] == 'LRG_93:g.5302_5303insTTTCCACCG'
        assert results['NM_032790.3:c.129_130insTTTCCACCG']['hgvs_transcript_variant'] == 'NM_032790.3:c.129_130insTTTCCACCG'
        assert results['NM_032790.3:c.129_130insTTTCCACCG']['hgvs_refseqgene_variant'] == 'NG_007500.1:g.5302_5303insTTTCCACCG'
        assert results['NM_032790.3:c.129_130insTTTCCACCG']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.122064776_122064777insTTT', 'vcf': {'chr': 'chr12', 'ref': 'G', 'pos': '122064776', 'alt': 'GTTT'}}
        assert results['NM_032790.3:c.129_130insTTTCCACCG']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.121626876_121626877insTTTCCACCG', 'vcf': {'chr': 'chr12', 'ref': 'C', 'pos': '121626873', 'alt': u'CCCGTTTCCA'}}
        assert results['NM_032790.3:c.129_130insTTTCCACCG']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.122064776_122064777insTTT', 'vcf': {'chr': '12', 'ref': 'G', 'pos': '122064776', 'alt': 'GTTT'}}
        assert results['NM_032790.3:c.129_130insTTTCCACCG']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.121626876_121626877insTTTCCACCG', 'vcf': {'chr': '12', 'ref': 'C', 'pos': '121626873', 'alt': u'CCCGTTTCCA'}}


    def test_variant73(self):
        variant = 'NC_000012.11:g.122064772_122064775del'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_032790.3:c.125_126delinsGCCA' in results.keys()
        assert results['NM_032790.3:c.125_126delinsGCCA']['hgvs_lrg_transcript_variant'] == 'LRG_93t1:c.125_126delinsGCCA'
        assert results['NM_032790.3:c.125_126delinsGCCA']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_032790.3:c.125_126delinsGCCA']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_004504303.2:g.302870_302873del', 'vcf': {'chr': 'HG1595_PATCH', 'ref': 'GCCCC', 'pos': '302869', 'alt': 'G'}}}]
        assert results['NM_032790.3:c.125_126delinsGCCA']['gene_symbol'] == 'ORAI1'
        assert results['NM_032790.3:c.125_126delinsGCCA']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_116179.2(LRG_93p1):p.(Ala42GlyfsTer23)', 'slr': 'NP_116179.2:p.(A42Gfs*23)'}
        assert results['NM_032790.3:c.125_126delinsGCCA']['submitted_variant'] == 'NC_000012.11:g.122064772_122064775del'
        assert results['NM_032790.3:c.125_126delinsGCCA']['genome_context_intronic_sequence'] == ''
        assert results['NM_032790.3:c.125_126delinsGCCA']['hgvs_lrg_variant'] == 'LRG_93:g.5298_5299delinsGCCA'
        assert results['NM_032790.3:c.125_126delinsGCCA']['hgvs_transcript_variant'] == 'NM_032790.3:c.125_126delinsGCCA'
        assert results['NM_032790.3:c.125_126delinsGCCA']['hgvs_refseqgene_variant'] == 'NG_007500.1:g.5298_5299delinsGCCA'
        assert results['NM_032790.3:c.125_126delinsGCCA']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.122064772_122064775del', 'vcf': {'chr': 'chr12', 'ref': 'GCCCC', 'pos': '122064771', 'alt': 'G'}}
        assert results['NM_032790.3:c.125_126delinsGCCA']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.121626867_121626873delinsGCCA', 'vcf': {'chr': 'chr12', 'ref': 'CCCCGCC', 'pos': '121626867', 'alt': 'GCCA'}}
        assert results['NM_032790.3:c.125_126delinsGCCA']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.122064772_122064775del', 'vcf': {'chr': '12', 'ref': 'GCCCC', 'pos': '122064771', 'alt': 'G'}}
        assert results['NM_032790.3:c.125_126delinsGCCA']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.121626867_121626873delinsGCCA', 'vcf': {'chr': '12', 'ref': 'CCCCGCC', 'pos': '121626867', 'alt': 'GCCA'}}


    def test_variant74(self):
        variant = 'NC_000012.11:g.122064772_122064775dup'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_032790.3:c.128_129insCCCCGCCACC' in results.keys()
        assert results['NM_032790.3:c.128_129insCCCCGCCACC']['hgvs_lrg_transcript_variant'] == 'LRG_93t1:c.128_129insCCCCGCCACC'
        assert results['NM_032790.3:c.128_129insCCCCGCCACC']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_032790.3:c.128_129insCCCCGCCACC']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_004504303.2:g.302870_302873dup', 'vcf': {'chr': 'HG1595_PATCH', 'ref': 'CCCC', 'pos': '302870', 'alt': 'CCCCCCCC'}}}]
        assert results['NM_032790.3:c.128_129insCCCCGCCACC']['gene_symbol'] == 'ORAI1'
        assert results['NM_032790.3:c.128_129insCCCCGCCACC']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_116179.2(LRG_93p1):p.(Pro45AlafsTer46)', 'slr': 'NP_116179.2:p.(P45Afs*46)'}
        assert results['NM_032790.3:c.128_129insCCCCGCCACC']['submitted_variant'] == 'NC_000012.11:g.122064772_122064775dup'
        assert results['NM_032790.3:c.128_129insCCCCGCCACC']['genome_context_intronic_sequence'] == ''
        assert results['NM_032790.3:c.128_129insCCCCGCCACC']['hgvs_lrg_variant'] == 'LRG_93:g.5301_5302insCCCCGCCACC'
        assert results['NM_032790.3:c.128_129insCCCCGCCACC']['hgvs_transcript_variant'] == 'NM_032790.3:c.128_129insCCCCGCCACC'
        assert results['NM_032790.3:c.128_129insCCCCGCCACC']['hgvs_refseqgene_variant'] == 'NG_007500.1:g.5301_5302insCCCCGCCACC'
        assert results['NM_032790.3:c.128_129insCCCCGCCACC']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.122064772_122064775dup', 'vcf': {'chr': 'chr12', 'ref': 'CCCC', 'pos': '122064772', 'alt': 'CCCCCCCC'}}
        assert results['NM_032790.3:c.128_129insCCCCGCCACC']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.121626875_121626876insCCCCGCCACC', 'vcf': {'chr': 'chr12', 'ref': 'C', 'pos': '121626873', 'alt': u'CCCCCCCGCCA'}}
        assert results['NM_032790.3:c.128_129insCCCCGCCACC']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.122064772_122064775dup', 'vcf': {'chr': '12', 'ref': 'CCCC', 'pos': '122064772', 'alt': 'CCCCCCCC'}}
        assert results['NM_032790.3:c.128_129insCCCCGCCACC']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.121626875_121626876insCCCCGCCACC', 'vcf': {'chr': '12', 'ref': 'C', 'pos': '121626873', 'alt': u'CCCCCCCGCCA'}}


    def test_variant75(self):
        variant = 'NC_000012.11:g.122064773_122064774insTTTT'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_032790.3:c.126_127insTTTTCCGCCA' in results.keys()
        assert results['NM_032790.3:c.126_127insTTTTCCGCCA']['hgvs_lrg_transcript_variant'] == 'LRG_93t1:c.126_127insTTTTCCGCCA'
        assert results['NM_032790.3:c.126_127insTTTTCCGCCA']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_032790.3:c.126_127insTTTTCCGCCA']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_004504303.2:g.302871_302872insTTTT', 'vcf': {'chr': 'HG1595_PATCH', 'ref': 'C', 'pos': '302871', 'alt': 'CTTTT'}}}]
        assert results['NM_032790.3:c.126_127insTTTTCCGCCA']['gene_symbol'] == 'ORAI1'
        assert results['NM_032790.3:c.126_127insTTTTCCGCCA']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_116179.2(LRG_93p1):p.(Pro43PhefsTer48)', 'slr': 'NP_116179.2:p.(P43Ffs*48)'}
        assert results['NM_032790.3:c.126_127insTTTTCCGCCA']['submitted_variant'] == 'NC_000012.11:g.122064773_122064774insTTTT'
        assert results['NM_032790.3:c.126_127insTTTTCCGCCA']['genome_context_intronic_sequence'] == ''
        assert results['NM_032790.3:c.126_127insTTTTCCGCCA']['hgvs_lrg_variant'] == 'LRG_93:g.5299_5300insTTTTCCGCCA'
        assert results['NM_032790.3:c.126_127insTTTTCCGCCA']['hgvs_transcript_variant'] == 'NM_032790.3:c.126_127insTTTTCCGCCA'
        assert results['NM_032790.3:c.126_127insTTTTCCGCCA']['hgvs_refseqgene_variant'] == 'NG_007500.1:g.5299_5300insTTTTCCGCCA'
        assert results['NM_032790.3:c.126_127insTTTTCCGCCA']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.122064773_122064774insTTTT', 'vcf': {'chr': 'chr12', 'ref': 'C', 'pos': '122064773', 'alt': 'CTTTT'}}
        assert results['NM_032790.3:c.126_127insTTTTCCGCCA']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.121626873_121626874insTTTTCCGCCA', 'vcf': {'chr': 'chr12', 'ref': 'C', 'pos': '121626873', 'alt': u'CTTTTCCGCCA'}}
        assert results['NM_032790.3:c.126_127insTTTTCCGCCA']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.122064773_122064774insTTTT', 'vcf': {'chr': '12', 'ref': 'C', 'pos': '122064773', 'alt': 'CTTTT'}}
        assert results['NM_032790.3:c.126_127insTTTTCCGCCA']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.121626873_121626874insTTTTCCGCCA', 'vcf': {'chr': '12', 'ref': 'C', 'pos': '121626873', 'alt': u'CTTTTCCGCCA'}}

        assert results['flag'] == 'gene_variant'

    def test_variant76(self):
        variant = 'NC_000012.11:g.122064772_122064777del'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_032790.3:c.126C>A' in results.keys()
        assert results['NM_032790.3:c.126C>A']['hgvs_lrg_transcript_variant'] == 'LRG_93t1:c.126C>A'
        assert results['NM_032790.3:c.126C>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_032790.3:c.126C>A']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_004504303.2:g.302871_302876del', 'vcf': {'chr': 'HG1595_PATCH', 'ref': 'GCCCCGC', 'pos': '302869', 'alt': 'G'}}}]
        assert results['NM_032790.3:c.126C>A']['gene_symbol'] == 'ORAI1'
        assert results['NM_032790.3:c.126C>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_116179.2(LRG_93p1):p.(Ala42=)', 'slr': 'NP_116179.2:p.(A42=)'}
        assert results['NM_032790.3:c.126C>A']['submitted_variant'] == 'NC_000012.11:g.122064772_122064777del'
        assert results['NM_032790.3:c.126C>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_032790.3:c.126C>A']['hgvs_lrg_variant'] == 'LRG_93:g.5299C>A'
        assert results['NM_032790.3:c.126C>A']['hgvs_transcript_variant'] == 'NM_032790.3:c.126C>A'
        assert results['NM_032790.3:c.126C>A']['hgvs_refseqgene_variant'] == 'NG_007500.1:g.5299C>A'
        assert results['NM_032790.3:c.126C>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.122064773_122064778del', 'vcf': {'chr': 'chr12', 'ref': 'GCCCCGC', 'pos': '122064771', 'alt': 'G'}}
        assert results['NM_032790.3:c.126C>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.121626873C>A', 'vcf': {'chr': 'chr12', 'ref': u'C', 'pos': '121626873', 'alt': u'A'}}
        assert results['NM_032790.3:c.126C>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.122064773_122064778del', 'vcf': {'chr': '12', 'ref': 'GCCCCGC', 'pos': '122064771', 'alt': 'G'}}
        assert results['NM_032790.3:c.126C>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.121626873C>A', 'vcf': {'chr': '12', 'ref': u'C', 'pos': '121626873', 'alt': u'A'}}

        assert results['flag'] == 'gene_variant'

    def test_variant77(self):
        variant = 'NC_000012.11:g.122064772_122064777dup'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_032790.3:c.131_132insCCCGCCACCGCC' in results.keys()
        assert results['NM_032790.3:c.131_132insCCCGCCACCGCC']['hgvs_lrg_transcript_variant'] == 'LRG_93t1:c.131_132insCCCGCCACCGCC'
        assert results['NM_032790.3:c.131_132insCCCGCCACCGCC']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_032790.3:c.131_132insCCCGCCACCGCC']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_004504303.2:g.302871_302876dup', 'vcf': {'chr': 'HG1595_PATCH', 'ref': 'CCCCGC', 'pos': '302870', 'alt': 'CCCCGCCCCCGC'}}}]
        assert results['NM_032790.3:c.131_132insCCCGCCACCGCC']['gene_symbol'] == 'ORAI1'
        assert results['NM_032790.3:c.131_132insCCCGCCACCGCC']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_116179.2(LRG_93p1):p.(Pro44_Pro47dup)', 'slr': 'NP_116179.2:p.(P44_P47dup)'}
        assert results['NM_032790.3:c.131_132insCCCGCCACCGCC']['submitted_variant'] == 'NC_000012.11:g.122064772_122064777dup'
        assert results['NM_032790.3:c.131_132insCCCGCCACCGCC']['genome_context_intronic_sequence'] == ''
        assert results['NM_032790.3:c.131_132insCCCGCCACCGCC']['hgvs_lrg_variant'] == 'LRG_93:g.5304_5305insCCCGCCACCGCC'
        assert results['NM_032790.3:c.131_132insCCCGCCACCGCC']['hgvs_transcript_variant'] == 'NM_032790.3:c.131_132insCCCGCCACCGCC'
        assert results['NM_032790.3:c.131_132insCCCGCCACCGCC']['hgvs_refseqgene_variant'] == 'NG_007500.1:g.5304_5305insCCCGCCACCGCC'
        assert results['NM_032790.3:c.131_132insCCCGCCACCGCC']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.122064773_122064778dup', 'vcf': {'chr': 'chr12', 'ref': 'CCCCGC', 'pos': '122064772', 'alt': 'CCCCGCCCCCGC'}}
        assert results['NM_032790.3:c.131_132insCCCGCCACCGCC']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.121626878_121626879insCCCGCCACCGCC', 'vcf': {'chr': 'chr12', 'ref': 'C', 'pos': '121626873', 'alt': u'CCCGCCCCCGCCA'}}
        assert results['NM_032790.3:c.131_132insCCCGCCACCGCC']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.122064773_122064778dup', 'vcf': {'chr': '12', 'ref': 'CCCCGC', 'pos': '122064772', 'alt': 'CCCCGCCCCCGC'}}
        assert results['NM_032790.3:c.131_132insCCCGCCACCGCC']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.121626878_121626879insCCCGCCACCGCC', 'vcf': {'chr': '12', 'ref': 'C', 'pos': '121626873', 'alt': u'CCCGCCCCCGCCA'}}


    def test_variant78(self):
        variant = 'NC_000012.11:g.122064779_122064782dup'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_032790.3:c.135_136insACCGCCACCG' in results.keys()
        assert results['NM_032790.3:c.135_136insACCGCCACCG']['hgvs_lrg_transcript_variant'] == 'LRG_93t1:c.135_136insACCGCCACCG'
        assert results['NM_032790.3:c.135_136insACCGCCACCG']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_032790.3:c.135_136insACCGCCACCG']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_004504303.2:g.302877_302880dup', 'vcf': {'chr': 'HG1595_PATCH', 'ref': 'ACCG', 'pos': '302877', 'alt': 'ACCGACCG'}}}]
        assert results['NM_032790.3:c.135_136insACCGCCACCG']['gene_symbol'] == 'ORAI1'
        assert results['NM_032790.3:c.135_136insACCGCCACCG']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_116179.2(LRG_93p1):p.(Pro46ThrfsTer45)', 'slr': 'NP_116179.2:p.(P46Tfs*45)'}
        assert results['NM_032790.3:c.135_136insACCGCCACCG']['submitted_variant'] == 'NC_000012.11:g.122064779_122064782dup'
        assert results['NM_032790.3:c.135_136insACCGCCACCG']['genome_context_intronic_sequence'] == ''
        assert results['NM_032790.3:c.135_136insACCGCCACCG']['hgvs_lrg_variant'] == 'LRG_93:g.5308_5309insACCGCCACCG'
        assert results['NM_032790.3:c.135_136insACCGCCACCG']['hgvs_transcript_variant'] == 'NM_032790.3:c.135_136insACCGCCACCG'
        assert results['NM_032790.3:c.135_136insACCGCCACCG']['hgvs_refseqgene_variant'] == 'NG_007500.1:g.5308_5309insACCGCCACCG'
        assert results['NM_032790.3:c.135_136insACCGCCACCG']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.122064779_122064782dup', 'vcf': {'chr': 'chr12', 'ref': 'ACCG', 'pos': '122064779', 'alt': 'ACCGACCG'}}
        assert results['NM_032790.3:c.135_136insACCGCCACCG']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.121626882_121626883insACCGCCACCG', 'vcf': {'chr': 'chr12', 'ref': 'C', 'pos': '121626873', 'alt': u'CCCGCCACCGA'}}
        assert results['NM_032790.3:c.135_136insACCGCCACCG']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.122064779_122064782dup', 'vcf': {'chr': '12', 'ref': 'ACCG', 'pos': '122064779', 'alt': 'ACCGACCG'}}
        assert results['NM_032790.3:c.135_136insACCGCCACCG']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.121626882_121626883insACCGCCACCG', 'vcf': {'chr': '12', 'ref': 'C', 'pos': '121626873', 'alt': u'CCCGCCACCGA'}}

        assert results['flag'] == 'gene_variant'

    def test_variant79(self):
        variant = 'NC_000012.11:g.122064772_122064782del'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_032790.3:c.126_127insA' in results.keys()
        assert results['NM_032790.3:c.126_127insA']['hgvs_lrg_transcript_variant'] == 'LRG_93t1:c.126_127insA'
        assert results['NM_032790.3:c.126_127insA']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_032790.3:c.126_127insA']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_004504303.2:g.302872_302876del', 'vcf': {'chr': 'HG1595_PATCH', 'ref': 'GGCCCC', 'pos': '302868', 'alt': 'G'}}}]
        assert results['NM_032790.3:c.126_127insA']['gene_symbol'] == 'ORAI1'
        assert results['NM_032790.3:c.126_127insA']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_116179.2(LRG_93p1):p.(Pro43ThrfsTer45)', 'slr': 'NP_116179.2:p.(P43Tfs*45)'}
        assert results['NM_032790.3:c.126_127insA']['submitted_variant'] == 'NC_000012.11:g.122064772_122064782del'
        assert results['NM_032790.3:c.126_127insA']['genome_context_intronic_sequence'] == ''
        assert results['NM_032790.3:c.126_127insA']['hgvs_lrg_variant'] == 'LRG_93:g.5299_5300insA'
        assert results['NM_032790.3:c.126_127insA']['hgvs_transcript_variant'] == 'NM_032790.3:c.126_127insA'
        assert results['NM_032790.3:c.126_127insA']['hgvs_refseqgene_variant'] == 'NG_007500.1:g.5299_5300insA'
        assert results['NM_032790.3:c.126_127insA']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.122064774_122064778del', 'vcf': {'chr': 'chr12', 'ref': 'GGCCCC', 'pos': '122064770', 'alt': 'G'}}
        assert results['NM_032790.3:c.126_127insA']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.121626873_121626874insA', 'vcf': {'chr': 'chr12', 'ref': 'C', 'pos': '121626873', 'alt': u'CA'}}
        assert results['NM_032790.3:c.126_127insA']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.122064774_122064778del', 'vcf': {'chr': '12', 'ref': 'GGCCCC', 'pos': '122064770', 'alt': 'G'}}
        assert results['NM_032790.3:c.126_127insA']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.121626873_121626874insA', 'vcf': {'chr': '12', 'ref': 'C', 'pos': '121626873', 'alt': u'CA'}}


    def test_variant80(self):
        variant = 'NC_000002.11:g.95847041_95847043GCG='
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_021088.3:c.471_473dup' in results.keys()
        assert results['NM_021088.3:c.471_473dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_021088.3:c.471_473dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_021088.3:c.471_473dup']['alt_genomic_loci'] == []
        assert results['NM_021088.3:c.471_473dup']['gene_symbol'] == 'ZNF2'
        assert results['NM_021088.3:c.471_473dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_066574.2:p.(Arg159dup)', 'slr': 'NP_066574.2:p.(R159dup)'}
        assert results['NM_021088.3:c.471_473dup']['submitted_variant'] == 'NC_000002.11:g.95847041_95847043GCG='
        assert results['NM_021088.3:c.471_473dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_021088.3:c.471_473dup']['hgvs_lrg_variant'] == ''
        assert results['NM_021088.3:c.471_473dup']['hgvs_transcript_variant'] == 'NM_021088.3:c.471_473dup'
        assert results['NM_021088.3:c.471_473dup']['hgvs_refseqgene_variant'] == 'NG_033798.1:g.20883_20885dup'
        assert results['NM_021088.3:c.471_473dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.95847037_95847050=', 'vcf': {'chr': 'chr2', 'ref': 'GCTTGCGGCGGCGA', 'pos': '95847037', 'alt': 'GCTTGCGGCGGCGA'}}
        assert results['NM_021088.3:c.471_473dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.95181299_95181301dup', 'vcf': {'chr': 'chr2', 'ref': 'GCG', 'pos': '95181296', 'alt': u'GCGGCG'}}
        assert results['NM_021088.3:c.471_473dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.95847037_95847050=', 'vcf': {'chr': '2', 'ref': 'GCTTGCGGCGGCGA', 'pos': '95847037', 'alt': 'GCTTGCGGCGGCGA'}}
        assert results['NM_021088.3:c.471_473dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.95181299_95181301dup', 'vcf': {'chr': '2', 'ref': 'GCG', 'pos': '95181296', 'alt': u'GCGGCG'}}

        assert 'NM_001291605.1:c.510_512dup' in results.keys()
        assert results['NM_001291605.1:c.510_512dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001291605.1:c.510_512dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001291605.1:c.510_512dup']['alt_genomic_loci'] == []
        assert results['NM_001291605.1:c.510_512dup']['gene_symbol'] == 'ZNF2'
        assert results['NM_001291605.1:c.510_512dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001278534.1:p.(Arg172dup)', 'slr': 'NP_001278534.1:p.(R172dup)'}
        assert results['NM_001291605.1:c.510_512dup']['submitted_variant'] == 'NC_000002.11:g.95847041_95847043GCG='
        assert results['NM_001291605.1:c.510_512dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_001291605.1:c.510_512dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001291605.1:c.510_512dup']['hgvs_transcript_variant'] == 'NM_001291605.1:c.510_512dup'
        assert results['NM_001291605.1:c.510_512dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001291605.1:c.510_512dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.95847037_95847050=', 'vcf': {'chr': 'chr2', 'ref': 'GCTTGCGGCGGCGA', 'pos': '95847037', 'alt': 'GCTTGCGGCGGCGA'}}
        assert results['NM_001291605.1:c.510_512dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.95181299_95181301dup', 'vcf': {'chr': 'chr2', 'ref': 'GCG', 'pos': '95181296', 'alt': u'GCGGCG'}}
        assert results['NM_001291605.1:c.510_512dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.95847037_95847050=', 'vcf': {'chr': '2', 'ref': 'GCTTGCGGCGGCGA', 'pos': '95847037', 'alt': 'GCTTGCGGCGGCGA'}}
        assert results['NM_001291605.1:c.510_512dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.95181299_95181301dup', 'vcf': {'chr': '2', 'ref': 'GCG', 'pos': '95181296', 'alt': u'GCGGCG'}}

        assert 'NM_001017396.2:c.345_347dup' in results.keys()
        assert results['NM_001017396.2:c.345_347dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001017396.2:c.345_347dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001017396.2:c.345_347dup']['alt_genomic_loci'] == []
        assert results['NM_001017396.2:c.345_347dup']['gene_symbol'] == 'ZNF2'
        assert results['NM_001017396.2:c.345_347dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001017396.1:p.(Arg117dup)', 'slr': 'NP_001017396.1:p.(R117dup)'}
        assert results['NM_001017396.2:c.345_347dup']['submitted_variant'] == 'NC_000002.11:g.95847041_95847043GCG='
        assert results['NM_001017396.2:c.345_347dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_001017396.2:c.345_347dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001017396.2:c.345_347dup']['hgvs_transcript_variant'] == 'NM_001017396.2:c.345_347dup'
        assert results['NM_001017396.2:c.345_347dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001017396.2:c.345_347dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.95847037_95847050=', 'vcf': {'chr': 'chr2', 'ref': 'GCTTGCGGCGGCGA', 'pos': '95847037', 'alt': 'GCTTGCGGCGGCGA'}}
        assert results['NM_001017396.2:c.345_347dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.95181299_95181301dup', 'vcf': {'chr': 'chr2', 'ref': 'GCG', 'pos': '95181296', 'alt': u'GCGGCG'}}
        assert results['NM_001017396.2:c.345_347dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.95847037_95847050=', 'vcf': {'chr': '2', 'ref': 'GCTTGCGGCGGCGA', 'pos': '95847037', 'alt': 'GCTTGCGGCGGCGA'}}
        assert results['NM_001017396.2:c.345_347dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.95181299_95181301dup', 'vcf': {'chr': '2', 'ref': 'GCG', 'pos': '95181296', 'alt': u'GCGGCG'}}

        assert 'NM_001282398.1:c.357_359dup' in results.keys()
        assert results['NM_001282398.1:c.357_359dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001282398.1:c.357_359dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001282398.1:c.357_359dup']['alt_genomic_loci'] == []
        assert results['NM_001282398.1:c.357_359dup']['gene_symbol'] == 'ZNF2'
        assert results['NM_001282398.1:c.357_359dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001269327.1:p.(Arg121dup)', 'slr': 'NP_001269327.1:p.(R121dup)'}
        assert results['NM_001282398.1:c.357_359dup']['submitted_variant'] == 'NC_000002.11:g.95847041_95847043GCG='
        assert results['NM_001282398.1:c.357_359dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_001282398.1:c.357_359dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001282398.1:c.357_359dup']['hgvs_transcript_variant'] == 'NM_001282398.1:c.357_359dup'
        assert results['NM_001282398.1:c.357_359dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001282398.1:c.357_359dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.95847037_95847050=', 'vcf': {'chr': 'chr2', 'ref': 'GCTTGCGGCGGCGA', 'pos': '95847037', 'alt': 'GCTTGCGGCGGCGA'}}
        assert results['NM_001282398.1:c.357_359dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.95181299_95181301dup', 'vcf': {'chr': 'chr2', 'ref': 'GCG', 'pos': '95181296', 'alt': u'GCGGCG'}}
        assert results['NM_001282398.1:c.357_359dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.95847037_95847050=', 'vcf': {'chr': '2', 'ref': 'GCTTGCGGCGGCGA', 'pos': '95847037', 'alt': 'GCTTGCGGCGGCGA'}}
        assert results['NM_001282398.1:c.357_359dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.95181299_95181301dup', 'vcf': {'chr': '2', 'ref': 'GCG', 'pos': '95181296', 'alt': u'GCGGCG'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001017396.1:c.345_347dup' in results.keys()
        assert results['NM_001017396.1:c.345_347dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001017396.1:c.345_347dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001017396.1:c.345_347dup']['alt_genomic_loci'] == []
        assert results['NM_001017396.1:c.345_347dup']['gene_symbol'] == 'ZNF2'
        assert results['NM_001017396.1:c.345_347dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001017396.1:p.(Arg117dup)', 'slr': 'NP_001017396.1:p.(R117dup)'}
        assert results['NM_001017396.1:c.345_347dup']['submitted_variant'] == 'NC_000002.11:g.95847041_95847043GCG='
        assert results['NM_001017396.1:c.345_347dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_001017396.1:c.345_347dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001017396.1:c.345_347dup']['hgvs_transcript_variant'] == 'NM_001017396.1:c.345_347dup'
        assert results['NM_001017396.1:c.345_347dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001017396.1:c.345_347dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.95847037_95847050=', 'vcf': {'chr': 'chr2', 'ref': 'GCTTGCGGCGGCGA', 'pos': '95847037', 'alt': 'GCTTGCGGCGGCGA'}}
        assert 'hg38' not in results['NM_001017396.1:c.345_347dup']['primary_assembly_loci'].keys()
        assert results['NM_001017396.1:c.345_347dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.95847037_95847050=', 'vcf': {'chr': '2', 'ref': 'GCTTGCGGCGGCGA', 'pos': '95847037', 'alt': 'GCTTGCGGCGGCGA'}}
        assert 'grch38' not in results['NM_001017396.1:c.345_347dup']['primary_assembly_loci'].keys()

        assert 'NM_021088.2:c.471_473dup' in results.keys()
        assert results['NM_021088.2:c.471_473dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_021088.2:c.471_473dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_021088.2:c.471_473dup']['alt_genomic_loci'] == []
        assert results['NM_021088.2:c.471_473dup']['gene_symbol'] == 'ZNF2'
        assert results['NM_021088.2:c.471_473dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_066574.2:p.(Arg159dup)', 'slr': 'NP_066574.2:p.(R159dup)'}
        assert results['NM_021088.2:c.471_473dup']['submitted_variant'] == 'NC_000002.11:g.95847041_95847043GCG='
        assert results['NM_021088.2:c.471_473dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_021088.2:c.471_473dup']['hgvs_lrg_variant'] == ''
        assert results['NM_021088.2:c.471_473dup']['hgvs_transcript_variant'] == 'NM_021088.2:c.471_473dup'
        assert results['NM_021088.2:c.471_473dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_021088.2:c.471_473dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.95847037_95847050=', 'vcf': {'chr': 'chr2', 'ref': 'GCTTGCGGCGGCGA', 'pos': '95847037', 'alt': 'GCTTGCGGCGGCGA'}}
        assert 'hg38' not in results['NM_021088.2:c.471_473dup']['primary_assembly_loci'].keys()
        assert results['NM_021088.2:c.471_473dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.95847037_95847050=', 'vcf': {'chr': '2', 'ref': 'GCTTGCGGCGGCGA', 'pos': '95847037', 'alt': 'GCTTGCGGCGGCGA'}}
        assert 'grch38' not in results['NM_021088.2:c.471_473dup']['primary_assembly_loci'].keys()

        assert 'NM_001291604.1:c.231_233dup' in results.keys()
        assert results['NM_001291604.1:c.231_233dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001291604.1:c.231_233dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001291604.1:c.231_233dup']['alt_genomic_loci'] == []
        assert results['NM_001291604.1:c.231_233dup']['gene_symbol'] == 'ZNF2'
        assert results['NM_001291604.1:c.231_233dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001278533.1:p.(Arg79dup)', 'slr': 'NP_001278533.1:p.(R79dup)'}
        assert results['NM_001291604.1:c.231_233dup']['submitted_variant'] == 'NC_000002.11:g.95847041_95847043GCG='
        assert results['NM_001291604.1:c.231_233dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_001291604.1:c.231_233dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001291604.1:c.231_233dup']['hgvs_transcript_variant'] == 'NM_001291604.1:c.231_233dup'
        assert results['NM_001291604.1:c.231_233dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001291604.1:c.231_233dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.95847037_95847050=', 'vcf': {'chr': 'chr2', 'ref': 'GCTTGCGGCGGCGA', 'pos': '95847037', 'alt': 'GCTTGCGGCGGCGA'}}
        assert results['NM_001291604.1:c.231_233dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.95181299_95181301dup', 'vcf': {'chr': 'chr2', 'ref': 'GCG', 'pos': '95181296', 'alt': u'GCGGCG'}}
        assert results['NM_001291604.1:c.231_233dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.95847037_95847050=', 'vcf': {'chr': '2', 'ref': 'GCTTGCGGCGGCGA', 'pos': '95847037', 'alt': 'GCTTGCGGCGGCGA'}}
        assert results['NM_001291604.1:c.231_233dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.95181299_95181301dup', 'vcf': {'chr': '2', 'ref': 'GCG', 'pos': '95181296', 'alt': u'GCGGCG'}}


    def test_variant81(self):
        variant = 'NC_000017.10:g.5286863_5286889AGTGTTTGGAATTTTCTGTTCATATAG='
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001083585.1:c.*344_*368dup' in results.keys()
        assert results['NM_001083585.1:c.*344_*368dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001083585.1:c.*344_*368dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001083585.1:c.*344_*368dup']['alt_genomic_loci'] == []
        assert results['NM_001083585.1:c.*344_*368dup']['gene_symbol'] == 'RABEP1'
        assert results['NM_001083585.1:c.*344_*368dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001077054.1:p.?', 'slr': 'NP_001077054.1:p.?'}
        assert results['NM_001083585.1:c.*344_*368dup']['submitted_variant'] == 'NC_000017.10:g.5286863_5286889AGTGTTTGGAATTTTCTGTTCATATAG='
        assert results['NM_001083585.1:c.*344_*368dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_001083585.1:c.*344_*368dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001083585.1:c.*344_*368dup']['hgvs_transcript_variant'] == 'NM_001083585.1:c.*344_*368dup'
        assert results['NM_001083585.1:c.*344_*368dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001083585.1:c.*344_*368dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.5286857_5286915=', 'vcf': {'chr': 'chr17', 'ref': 'CCCAGTAGTGTTTGGAATTTTCTGTTCATATAGTGTTTGGAATTTTCTGTTCATAGATA', 'pos': '5286857', 'alt': 'CCCAGTAGTGTTTGGAATTTTCTGTTCATATAGTGTTTGGAATTTTCTGTTCATAGATA'}}
        assert 'hg38' not in results['NM_001083585.1:c.*344_*368dup']['primary_assembly_loci'].keys()
        assert results['NM_001083585.1:c.*344_*368dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.5286857_5286915=', 'vcf': {'chr': '17', 'ref': 'CCCAGTAGTGTTTGGAATTTTCTGTTCATATAGTGTTTGGAATTTTCTGTTCATAGATA', 'pos': '5286857', 'alt': 'CCCAGTAGTGTTTGGAATTTTCTGTTCATATAGTGTTTGGAATTTTCTGTTCATAGATA'}}
        assert 'grch38' not in results['NM_001083585.1:c.*344_*368dup']['primary_assembly_loci'].keys()

        assert 'NM_004703.5:c.*344_*368dup' in results.keys()
        assert results['NM_004703.5:c.*344_*368dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_004703.5:c.*344_*368dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_004703.5:c.*344_*368dup']['alt_genomic_loci'] == []
        assert results['NM_004703.5:c.*344_*368dup']['gene_symbol'] == 'RABEP1'
        assert results['NM_004703.5:c.*344_*368dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_004694.2:p.?', 'slr': 'NP_004694.2:p.?'}
        assert results['NM_004703.5:c.*344_*368dup']['submitted_variant'] == 'NC_000017.10:g.5286863_5286889AGTGTTTGGAATTTTCTGTTCATATAG='
        assert results['NM_004703.5:c.*344_*368dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_004703.5:c.*344_*368dup']['hgvs_lrg_variant'] == ''
        assert results['NM_004703.5:c.*344_*368dup']['hgvs_transcript_variant'] == 'NM_004703.5:c.*344_*368dup'
        assert results['NM_004703.5:c.*344_*368dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_004703.5:c.*344_*368dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.5286857_5286915=', 'vcf': {'chr': 'chr17', 'ref': 'CCCAGTAGTGTTTGGAATTTTCTGTTCATATAGTGTTTGGAATTTTCTGTTCATAGATA', 'pos': '5286857', 'alt': 'CCCAGTAGTGTTTGGAATTTTCTGTTCATATAGTGTTTGGAATTTTCTGTTCATAGATA'}}
        assert results['NM_004703.5:c.*344_*368dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.5383567_5383591dup', 'vcf': {'chr': 'chr17', 'ref': 'TAGTGTTTGGAATTTTCTGTTCATA', 'pos': '5383567', 'alt': u'TAGTGTTTGGAATTTTCTGTTCATATAGTGTTTGGAATTTTCTGTTCATA'}}
        assert results['NM_004703.5:c.*344_*368dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.5286857_5286915=', 'vcf': {'chr': '17', 'ref': 'CCCAGTAGTGTTTGGAATTTTCTGTTCATATAGTGTTTGGAATTTTCTGTTCATAGATA', 'pos': '5286857', 'alt': 'CCCAGTAGTGTTTGGAATTTTCTGTTCATATAGTGTTTGGAATTTTCTGTTCATAGATA'}}
        assert results['NM_004703.5:c.*344_*368dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.5383567_5383591dup', 'vcf': {'chr': '17', 'ref': 'TAGTGTTTGGAATTTTCTGTTCATA', 'pos': '5383567', 'alt': u'TAGTGTTTGGAATTTTCTGTTCATATAGTGTTTGGAATTTTCTGTTCATA'}}

        assert 'NM_004703.4:c.*344_*368dup' in results.keys()
        assert results['NM_004703.4:c.*344_*368dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_004703.4:c.*344_*368dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_004703.4:c.*344_*368dup']['alt_genomic_loci'] == []
        assert results['NM_004703.4:c.*344_*368dup']['gene_symbol'] == 'RABEP1'
        assert results['NM_004703.4:c.*344_*368dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_004694.2:p.?', 'slr': 'NP_004694.2:p.?'}
        assert results['NM_004703.4:c.*344_*368dup']['submitted_variant'] == 'NC_000017.10:g.5286863_5286889AGTGTTTGGAATTTTCTGTTCATATAG='
        assert results['NM_004703.4:c.*344_*368dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_004703.4:c.*344_*368dup']['hgvs_lrg_variant'] == ''
        assert results['NM_004703.4:c.*344_*368dup']['hgvs_transcript_variant'] == 'NM_004703.4:c.*344_*368dup'
        assert results['NM_004703.4:c.*344_*368dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_004703.4:c.*344_*368dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.5286857_5286915=', 'vcf': {'chr': 'chr17', 'ref': 'CCCAGTAGTGTTTGGAATTTTCTGTTCATATAGTGTTTGGAATTTTCTGTTCATAGATA', 'pos': '5286857', 'alt': 'CCCAGTAGTGTTTGGAATTTTCTGTTCATATAGTGTTTGGAATTTTCTGTTCATAGATA'}}
        assert 'hg38' not in results['NM_004703.4:c.*344_*368dup']['primary_assembly_loci'].keys()
        assert results['NM_004703.4:c.*344_*368dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.5286857_5286915=', 'vcf': {'chr': '17', 'ref': 'CCCAGTAGTGTTTGGAATTTTCTGTTCATATAGTGTTTGGAATTTTCTGTTCATAGATA', 'pos': '5286857', 'alt': 'CCCAGTAGTGTTTGGAATTTTCTGTTCATATAGTGTTTGGAATTTTCTGTTCATAGATA'}}
        assert 'grch38' not in results['NM_004703.4:c.*344_*368dup']['primary_assembly_loci'].keys()

        assert results['flag'] == 'gene_variant'
        assert 'NM_001291581.1:c.*344_*368dup' in results.keys()
        assert results['NM_001291581.1:c.*344_*368dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001291581.1:c.*344_*368dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001291581.1:c.*344_*368dup']['alt_genomic_loci'] == []
        assert results['NM_001291581.1:c.*344_*368dup']['gene_symbol'] == 'RABEP1'
        assert results['NM_001291581.1:c.*344_*368dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001278510.1:p.?', 'slr': 'NP_001278510.1:p.?'}
        assert results['NM_001291581.1:c.*344_*368dup']['submitted_variant'] == 'NC_000017.10:g.5286863_5286889AGTGTTTGGAATTTTCTGTTCATATAG='
        assert results['NM_001291581.1:c.*344_*368dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_001291581.1:c.*344_*368dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001291581.1:c.*344_*368dup']['hgvs_transcript_variant'] == 'NM_001291581.1:c.*344_*368dup'
        assert results['NM_001291581.1:c.*344_*368dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001291581.1:c.*344_*368dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.5286857_5286915=', 'vcf': {'chr': 'chr17', 'ref': 'CCCAGTAGTGTTTGGAATTTTCTGTTCATATAGTGTTTGGAATTTTCTGTTCATAGATA', 'pos': '5286857', 'alt': 'CCCAGTAGTGTTTGGAATTTTCTGTTCATATAGTGTTTGGAATTTTCTGTTCATAGATA'}}
        assert results['NM_001291581.1:c.*344_*368dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.5383567_5383591dup', 'vcf': {'chr': 'chr17', 'ref': 'TAGTGTTTGGAATTTTCTGTTCATA', 'pos': '5383567', 'alt': u'TAGTGTTTGGAATTTTCTGTTCATATAGTGTTTGGAATTTTCTGTTCATA'}}
        assert results['NM_001291581.1:c.*344_*368dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.5286857_5286915=', 'vcf': {'chr': '17', 'ref': 'CCCAGTAGTGTTTGGAATTTTCTGTTCATATAGTGTTTGGAATTTTCTGTTCATAGATA', 'pos': '5286857', 'alt': 'CCCAGTAGTGTTTGGAATTTTCTGTTCATATAGTGTTTGGAATTTTCTGTTCATAGATA'}}
        assert results['NM_001291581.1:c.*344_*368dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.5383567_5383591dup', 'vcf': {'chr': '17', 'ref': 'TAGTGTTTGGAATTTTCTGTTCATA', 'pos': '5383567', 'alt': u'TAGTGTTTGGAATTTTCTGTTCATATAGTGTTTGGAATTTTCTGTTCATA'}}

        assert 'NM_001083585.2:c.*344_*368dup' in results.keys()
        assert results['NM_001083585.2:c.*344_*368dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001083585.2:c.*344_*368dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001083585.2:c.*344_*368dup']['alt_genomic_loci'] == []
        assert results['NM_001083585.2:c.*344_*368dup']['gene_symbol'] == 'RABEP1'
        assert results['NM_001083585.2:c.*344_*368dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001077054.1:p.?', 'slr': 'NP_001077054.1:p.?'}
        assert results['NM_001083585.2:c.*344_*368dup']['submitted_variant'] == 'NC_000017.10:g.5286863_5286889AGTGTTTGGAATTTTCTGTTCATATAG='
        assert results['NM_001083585.2:c.*344_*368dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_001083585.2:c.*344_*368dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001083585.2:c.*344_*368dup']['hgvs_transcript_variant'] == 'NM_001083585.2:c.*344_*368dup'
        assert results['NM_001083585.2:c.*344_*368dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001083585.2:c.*344_*368dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.5286857_5286915=', 'vcf': {'chr': 'chr17', 'ref': 'CCCAGTAGTGTTTGGAATTTTCTGTTCATATAGTGTTTGGAATTTTCTGTTCATAGATA', 'pos': '5286857', 'alt': 'CCCAGTAGTGTTTGGAATTTTCTGTTCATATAGTGTTTGGAATTTTCTGTTCATAGATA'}}
        assert results['NM_001083585.2:c.*344_*368dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.5383567_5383591dup', 'vcf': {'chr': 'chr17', 'ref': 'TAGTGTTTGGAATTTTCTGTTCATA', 'pos': '5383567', 'alt': u'TAGTGTTTGGAATTTTCTGTTCATATAGTGTTTGGAATTTTCTGTTCATA'}}
        assert results['NM_001083585.2:c.*344_*368dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.5286857_5286915=', 'vcf': {'chr': '17', 'ref': 'CCCAGTAGTGTTTGGAATTTTCTGTTCATATAGTGTTTGGAATTTTCTGTTCATAGATA', 'pos': '5286857', 'alt': 'CCCAGTAGTGTTTGGAATTTTCTGTTCATATAGTGTTTGGAATTTTCTGTTCATAGATA'}}
        assert results['NM_001083585.2:c.*344_*368dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.5383567_5383591dup', 'vcf': {'chr': '17', 'ref': 'TAGTGTTTGGAATTTTCTGTTCATA', 'pos': '5383567', 'alt': u'TAGTGTTTGGAATTTTCTGTTCATATAGTGTTTGGAATTTTCTGTTCATA'}}


    def test_variant82(self):
        variant = 'NC_000003.11:g.14561629_14561630GC='
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001080423.2:c.1311del' in results.keys()
        assert results['NM_001080423.2:c.1311del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001080423.2:c.1311del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001080423.2:c.1311del']['alt_genomic_loci'] == []
        assert results['NM_001080423.2:c.1311del']['gene_symbol'] == 'GRIP2'
        assert results['NM_001080423.2:c.1311del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001073892.2:p.(Ser438GlnfsTer4)', 'slr': 'NP_001073892.2:p.(S438Qfs*4)'}
        assert results['NM_001080423.2:c.1311del']['submitted_variant'] == 'NC_000003.11:g.14561629_14561630GC='
        assert results['NM_001080423.2:c.1311del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001080423.2:c.1311del']['hgvs_lrg_variant'] == ''
        assert results['NM_001080423.2:c.1311del']['hgvs_transcript_variant'] == 'NM_001080423.2:c.1311del'
        assert results['NM_001080423.2:c.1311del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001080423.2:c.1311del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000003.11:g.14561624_14561630=', 'vcf': {'chr': 'chr3', 'ref': 'CTGAGGC', 'pos': '14561624', 'alt': 'CTGAGGC'}}
        assert 'hg38' not in results['NM_001080423.2:c.1311del']['primary_assembly_loci'].keys()
        assert results['NM_001080423.2:c.1311del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000003.11:g.14561624_14561630=', 'vcf': {'chr': '3', 'ref': 'CTGAGGC', 'pos': '14561624', 'alt': 'CTGAGGC'}}
        assert 'grch38' not in results['NM_001080423.2:c.1311del']['primary_assembly_loci'].keys()

        assert results['flag'] == 'gene_variant'
        assert 'NM_001080423.3:c.1020del' in results.keys()
        assert results['NM_001080423.3:c.1020del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001080423.3:c.1020del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001080423.3:c.1020del']['alt_genomic_loci'] == []
        assert results['NM_001080423.3:c.1020del']['gene_symbol'] == 'GRIP2'
        assert results['NM_001080423.3:c.1020del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001073892.3:p.(Ser341GlnfsTer4)', 'slr': 'NP_001073892.3:p.(S341Qfs*4)'}
        assert results['NM_001080423.3:c.1020del']['submitted_variant'] == 'NC_000003.11:g.14561629_14561630GC='
        assert results['NM_001080423.3:c.1020del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001080423.3:c.1020del']['hgvs_lrg_variant'] == ''
        assert results['NM_001080423.3:c.1020del']['hgvs_transcript_variant'] == 'NM_001080423.3:c.1020del'
        assert results['NM_001080423.3:c.1020del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001080423.3:c.1020del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000003.11:g.14561624_14561630=', 'vcf': {'chr': 'chr3', 'ref': 'CTGAGGC', 'pos': '14561624', 'alt': 'CTGAGGC'}}
        assert results['NM_001080423.3:c.1020del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000003.12:g.14520122del', 'vcf': {'chr': 'chr3', 'ref': 'AG', 'pos': '14520119', 'alt': 'A'}}
        assert results['NM_001080423.3:c.1020del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000003.11:g.14561624_14561630=', 'vcf': {'chr': '3', 'ref': 'CTGAGGC', 'pos': '14561624', 'alt': 'CTGAGGC'}}
        assert results['NM_001080423.3:c.1020del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000003.12:g.14520122del', 'vcf': {'chr': '3', 'ref': 'AG', 'pos': '14520119', 'alt': 'A'}}


    def test_variant83(self):
        variant = 'NC_000003.11:g.14561629_14561630insG'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001080423.3:c.1016_1020=' in results.keys()
        assert results['NM_001080423.3:c.1016_1020=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001080423.3:c.1016_1020=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001080423.3:c.1016_1020=']['alt_genomic_loci'] == []
        assert results['NM_001080423.3:c.1016_1020=']['gene_symbol'] == 'GRIP2'
        assert results['NM_001080423.3:c.1016_1020=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001073892.3:p.(Arg339=)', 'slr': 'NP_001073892.3:p.(R339=)'}
        assert results['NM_001080423.3:c.1016_1020=']['submitted_variant'] == 'NC_000003.11:g.14561629_14561630insG'
        assert results['NM_001080423.3:c.1016_1020=']['genome_context_intronic_sequence'] == ''
        assert results['NM_001080423.3:c.1016_1020=']['hgvs_lrg_variant'] == ''
        assert results['NM_001080423.3:c.1016_1020=']['hgvs_transcript_variant'] == 'NM_001080423.3:c.1016_1020='
        assert results['NM_001080423.3:c.1016_1020=']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001080423.3:c.1016_1020=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000003.11:g.14561629dup', 'vcf': {'chr': 'chr3', 'ref': 'G', 'pos': '14561628', 'alt': u'GG'}}
        assert results['NM_001080423.3:c.1016_1020=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000003.12:g.14520120_14520124=', 'vcf': {'chr': 'chr3', 'ref': u'GGGCC', 'pos': '14520120', 'alt': u'GGGCC'}}
        assert results['NM_001080423.3:c.1016_1020=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000003.11:g.14561629dup', 'vcf': {'chr': '3', 'ref': 'G', 'pos': '14561628', 'alt': u'GG'}}
        assert results['NM_001080423.3:c.1016_1020=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000003.12:g.14520120_14520124=', 'vcf': {'chr': '3', 'ref': u'GGGCC', 'pos': '14520120', 'alt': u'GGGCC'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001080423.2:c.1307_1311=' in results.keys()
        assert results['NM_001080423.2:c.1307_1311=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001080423.2:c.1307_1311=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001080423.2:c.1307_1311=']['alt_genomic_loci'] == []
        assert results['NM_001080423.2:c.1307_1311=']['gene_symbol'] == 'GRIP2'
        assert results['NM_001080423.2:c.1307_1311=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001073892.2:p.(Arg436=)', 'slr': 'NP_001073892.2:p.(R436=)'}
        assert results['NM_001080423.2:c.1307_1311=']['submitted_variant'] == 'NC_000003.11:g.14561629_14561630insG'
        assert results['NM_001080423.2:c.1307_1311=']['genome_context_intronic_sequence'] == ''
        assert results['NM_001080423.2:c.1307_1311=']['hgvs_lrg_variant'] == ''
        assert results['NM_001080423.2:c.1307_1311=']['hgvs_transcript_variant'] == 'NM_001080423.2:c.1307_1311='
        assert results['NM_001080423.2:c.1307_1311=']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001080423.2:c.1307_1311=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000003.11:g.14561629dup', 'vcf': {'chr': 'chr3', 'ref': 'G', 'pos': '14561628', 'alt': u'GG'}}
        assert 'hg38' not in results['NM_001080423.2:c.1307_1311=']['primary_assembly_loci'].keys()
        assert results['NM_001080423.2:c.1307_1311=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000003.11:g.14561629dup', 'vcf': {'chr': '3', 'ref': 'G', 'pos': '14561628', 'alt': u'GG'}}
        assert 'grch38' not in results['NM_001080423.2:c.1307_1311=']['primary_assembly_loci'].keys()


    def test_variant84(self):
        variant = 'NC_000004.11:g.140811111_140811122del'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_018717.5:c.1515_1526del' in results.keys()
        assert results['NM_018717.5:c.1515_1526del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_018717.5:c.1515_1526del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_018717.5:c.1515_1526del']['alt_genomic_loci'] == []
        assert results['NM_018717.5:c.1515_1526del']['gene_symbol'] == 'MAML3'
        assert results['NM_018717.5:c.1515_1526del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_061187.3:p.(Gln507_Gln510del)', 'slr': 'NP_061187.3:p.(Q507_Q510del)'}
        assert results['NM_018717.5:c.1515_1526del']['submitted_variant'] == 'NC_000004.11:g.140811111_140811122del'
        assert results['NM_018717.5:c.1515_1526del']['genome_context_intronic_sequence'] == ''
        assert results['NM_018717.5:c.1515_1526del']['hgvs_lrg_variant'] == ''
        assert results['NM_018717.5:c.1515_1526del']['hgvs_transcript_variant'] == 'NM_018717.5:c.1515_1526del'
        assert results['NM_018717.5:c.1515_1526del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_018717.5:c.1515_1526del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000004.11:g.140811111_140811122del', 'vcf': {'chr': 'chr4', 'ref': 'TTGCTGCTGCTGC', 'pos': '140811063', 'alt': 'T'}}
        assert 'hg38' not in results['NM_018717.5:c.1515_1526del']['primary_assembly_loci'].keys()
        assert results['NM_018717.5:c.1515_1526del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000004.11:g.140811111_140811122del', 'vcf': {'chr': '4', 'ref': 'TTGCTGCTGCTGC', 'pos': '140811063', 'alt': 'T'}}
        assert 'grch38' not in results['NM_018717.5:c.1515_1526del']['primary_assembly_loci'].keys()

        assert 'NM_018717.4:c.1465_1469=' in results.keys()
        assert results['NM_018717.4:c.1465_1469=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_018717.4:c.1465_1469=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_018717.4:c.1465_1469=']['alt_genomic_loci'] == []
        assert results['NM_018717.4:c.1465_1469=']['gene_symbol'] == 'MAML3'
        assert results['NM_018717.4:c.1465_1469=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_061187.2:p.(Gln489=)', 'slr': 'NP_061187.2:p.(Q489=)'}
        assert results['NM_018717.4:c.1465_1469=']['submitted_variant'] == 'NC_000004.11:g.140811111_140811122del'
        assert results['NM_018717.4:c.1465_1469=']['genome_context_intronic_sequence'] == ''
        assert results['NM_018717.4:c.1465_1469=']['hgvs_lrg_variant'] == ''
        assert results['NM_018717.4:c.1465_1469=']['hgvs_transcript_variant'] == 'NM_018717.4:c.1465_1469='
        assert results['NM_018717.4:c.1465_1469=']['hgvs_refseqgene_variant'] == ''
        assert results['NM_018717.4:c.1465_1469=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000004.11:g.140811111_140811122del', 'vcf': {'chr': 'chr4', 'ref': 'TTGCTGCTGCTGC', 'pos': '140811063', 'alt': 'T'}}
        assert results['NM_018717.4:c.1465_1469=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000004.12:g.139889957_139889968del', 'vcf': {'chr': 'chr4', 'ref': 'TTGCTGCTGCTGC', 'pos': '139889909', 'alt': 'T'}}
        assert results['NM_018717.4:c.1465_1469=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000004.11:g.140811111_140811122del', 'vcf': {'chr': '4', 'ref': 'TTGCTGCTGCTGC', 'pos': '140811063', 'alt': 'T'}}
        assert results['NM_018717.4:c.1465_1469=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000004.12:g.139889957_139889968del', 'vcf': {'chr': '4', 'ref': 'TTGCTGCTGCTGC', 'pos': '139889909', 'alt': 'T'}}


    def test_variant85(self):
        variant = 'NC_000004.11:g.140811111_140811122CTGCTGCTGCTG='
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_018717.5:c.1468_1479=' in results.keys()
        assert results['NM_018717.5:c.1468_1479=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_018717.5:c.1468_1479=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_018717.5:c.1468_1479=']['alt_genomic_loci'] == []
        assert results['NM_018717.5:c.1468_1479=']['gene_symbol'] == 'MAML3'
        assert results['NM_018717.5:c.1468_1479=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_061187.3:p.(Gln490=)', 'slr': 'NP_061187.3:p.(Q490=)'}
        assert results['NM_018717.5:c.1468_1479=']['submitted_variant'] == 'NC_000004.11:g.140811111_140811122CTGCTGCTGCTG='
        assert results['NM_018717.5:c.1468_1479=']['genome_context_intronic_sequence'] == ''
        assert results['NM_018717.5:c.1468_1479=']['hgvs_lrg_variant'] == ''
        assert results['NM_018717.5:c.1468_1479=']['hgvs_transcript_variant'] == 'NM_018717.5:c.1468_1479='
        assert results['NM_018717.5:c.1468_1479=']['hgvs_refseqgene_variant'] == ''
        assert results['NM_018717.5:c.1468_1479=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000004.11:g.140811111_140811122=', 'vcf': {'chr': 'chr4', 'ref': u'CTGCTGCTGCTG', 'pos': '140811111', 'alt': u'CTGCTGCTGCTG'}}
        assert 'hg38' not in results['NM_018717.5:c.1468_1479=']['primary_assembly_loci'].keys()
        assert results['NM_018717.5:c.1468_1479=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000004.11:g.140811111_140811122=', 'vcf': {'chr': '4', 'ref': u'CTGCTGCTGCTG', 'pos': '140811111', 'alt': u'CTGCTGCTGCTG'}}
        assert 'grch38' not in results['NM_018717.5:c.1468_1479=']['primary_assembly_loci'].keys()

        assert results['flag'] == 'gene_variant'
        assert 'NM_018717.4:c.1503_1514dup' in results.keys()
        assert results['NM_018717.4:c.1503_1514dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_018717.4:c.1503_1514dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_018717.4:c.1503_1514dup']['alt_genomic_loci'] == []
        assert results['NM_018717.4:c.1503_1514dup']['gene_symbol'] == 'MAML3'
        assert results['NM_018717.4:c.1503_1514dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_061187.2:p.(Gln503_Gln506dup)', 'slr': 'NP_061187.2:p.(Q503_Q506dup)'}
        assert results['NM_018717.4:c.1503_1514dup']['submitted_variant'] == 'NC_000004.11:g.140811111_140811122CTGCTGCTGCTG='
        assert results['NM_018717.4:c.1503_1514dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_018717.4:c.1503_1514dup']['hgvs_lrg_variant'] == ''
        assert results['NM_018717.4:c.1503_1514dup']['hgvs_transcript_variant'] == 'NM_018717.4:c.1503_1514dup'
        assert results['NM_018717.4:c.1503_1514dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_018717.4:c.1503_1514dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000004.11:g.140811095_140811128=', 'vcf': {'chr': 'chr4', 'ref': 'GCTGCTGCTGCTGCTGCTGCTGCTGCTGTTGCTG', 'pos': '140811095', 'alt': 'GCTGCTGCTGCTGCTGCTGCTGCTGCTGTTGCTG'}}
        assert results['NM_018717.4:c.1503_1514dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000004.12:g.139889941_139889974=', 'vcf': {'chr': 'chr4', 'ref': 'GCTGCTGCTGCTGCTGCTGCTGCTGCTGTTGCTG', 'pos': '139889941', 'alt': 'GCTGCTGCTGCTGCTGCTGCTGCTGCTGTTGCTG'}}
        assert results['NM_018717.4:c.1503_1514dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000004.11:g.140811095_140811128=', 'vcf': {'chr': '4', 'ref': 'GCTGCTGCTGCTGCTGCTGCTGCTGCTGTTGCTG', 'pos': '140811095', 'alt': 'GCTGCTGCTGCTGCTGCTGCTGCTGCTGTTGCTG'}}
        assert results['NM_018717.4:c.1503_1514dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000004.12:g.139889941_139889974=', 'vcf': {'chr': '4', 'ref': 'GCTGCTGCTGCTGCTGCTGCTGCTGCTGTTGCTG', 'pos': '139889941', 'alt': 'GCTGCTGCTGCTGCTGCTGCTGCTGCTGTTGCTG'}}


    def test_variant86(self):
        variant = 'NC_000004.11:g.140811117_140811122del'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_018717.5:c.1521_1526del' in results.keys()
        assert results['NM_018717.5:c.1521_1526del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_018717.5:c.1521_1526del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_018717.5:c.1521_1526del']['alt_genomic_loci'] == []
        assert results['NM_018717.5:c.1521_1526del']['gene_symbol'] == 'MAML3'
        assert results['NM_018717.5:c.1521_1526del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_061187.3:p.(Gln509_Gln510del)', 'slr': 'NP_061187.3:p.(Q509_Q510del)'}
        assert results['NM_018717.5:c.1521_1526del']['submitted_variant'] == 'NC_000004.11:g.140811117_140811122del'
        assert results['NM_018717.5:c.1521_1526del']['genome_context_intronic_sequence'] == ''
        assert results['NM_018717.5:c.1521_1526del']['hgvs_lrg_variant'] == ''
        assert results['NM_018717.5:c.1521_1526del']['hgvs_transcript_variant'] == 'NM_018717.5:c.1521_1526del'
        assert results['NM_018717.5:c.1521_1526del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_018717.5:c.1521_1526del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000004.11:g.140811117_140811122del', 'vcf': {'chr': 'chr4', 'ref': 'TTGCTGC', 'pos': '140811063', 'alt': 'T'}}
        assert 'hg38' not in results['NM_018717.5:c.1521_1526del']['primary_assembly_loci'].keys()
        assert results['NM_018717.5:c.1521_1526del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000004.11:g.140811117_140811122del', 'vcf': {'chr': '4', 'ref': 'TTGCTGC', 'pos': '140811063', 'alt': 'T'}}
        assert 'grch38' not in results['NM_018717.5:c.1521_1526del']['primary_assembly_loci'].keys()

        assert 'NM_018717.4:c.1509_1514dup' in results.keys()
        assert results['NM_018717.4:c.1509_1514dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_018717.4:c.1509_1514dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_018717.4:c.1509_1514dup']['alt_genomic_loci'] == []
        assert results['NM_018717.4:c.1509_1514dup']['gene_symbol'] == 'MAML3'
        assert results['NM_018717.4:c.1509_1514dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_061187.2:p.(Gln505_Gln506dup)', 'slr': 'NP_061187.2:p.(Q505_Q506dup)'}
        assert results['NM_018717.4:c.1509_1514dup']['submitted_variant'] == 'NC_000004.11:g.140811117_140811122del'
        assert results['NM_018717.4:c.1509_1514dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_018717.4:c.1509_1514dup']['hgvs_lrg_variant'] == ''
        assert results['NM_018717.4:c.1509_1514dup']['hgvs_transcript_variant'] == 'NM_018717.4:c.1509_1514dup'
        assert results['NM_018717.4:c.1509_1514dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_018717.4:c.1509_1514dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000004.11:g.140811117_140811122del', 'vcf': {'chr': 'chr4', 'ref': 'TTGCTGC', 'pos': '140811063', 'alt': 'T'}}
        assert results['NM_018717.4:c.1509_1514dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000004.12:g.139889963_139889968del', 'vcf': {'chr': 'chr4', 'ref': 'TTGCTGC', 'pos': '139889909', 'alt': 'T'}}
        assert results['NM_018717.4:c.1509_1514dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000004.11:g.140811117_140811122del', 'vcf': {'chr': '4', 'ref': 'TTGCTGC', 'pos': '140811063', 'alt': 'T'}}
        assert results['NM_018717.4:c.1509_1514dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000004.12:g.139889963_139889968del', 'vcf': {'chr': '4', 'ref': 'TTGCTGC', 'pos': '139889909', 'alt': 'T'}}


    def test_variant87(self):
        variant = 'NC_000004.11:g.140811111_140811117del'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_018717.5:c.1473_1479del' in results.keys()
        assert results['NM_018717.5:c.1473_1479del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_018717.5:c.1473_1479del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_018717.5:c.1473_1479del']['alt_genomic_loci'] == []
        assert results['NM_018717.5:c.1473_1479del']['gene_symbol'] == 'MAML3'
        assert results['NM_018717.5:c.1473_1479del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_061187.3:p.(Gln491HisfsTer29)', 'slr': 'NP_061187.3:p.(Q491Hfs*29)'}
        assert results['NM_018717.5:c.1473_1479del']['submitted_variant'] == 'NC_000004.11:g.140811111_140811117del'
        assert results['NM_018717.5:c.1473_1479del']['genome_context_intronic_sequence'] == ''
        assert results['NM_018717.5:c.1473_1479del']['hgvs_lrg_variant'] == ''
        assert results['NM_018717.5:c.1473_1479del']['hgvs_transcript_variant'] == 'NM_018717.5:c.1473_1479del'
        assert results['NM_018717.5:c.1473_1479del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_018717.5:c.1473_1479del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000004.11:g.140811111_140811117del', 'vcf': {'chr': 'chr4', 'ref': 'GCTGCTGC', 'pos': '140811110', 'alt': 'G'}}
        assert 'hg38' not in results['NM_018717.5:c.1473_1479del']['primary_assembly_loci'].keys()
        assert results['NM_018717.5:c.1473_1479del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000004.11:g.140811111_140811117del', 'vcf': {'chr': '4', 'ref': 'GCTGCTGC', 'pos': '140811110', 'alt': 'G'}}
        assert 'grch38' not in results['NM_018717.5:c.1473_1479del']['primary_assembly_loci'].keys()

        assert results['flag'] == 'gene_variant'
        assert 'NM_018717.4:c.1468_1472dup' in results.keys()
        assert results['NM_018717.4:c.1468_1472dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_018717.4:c.1468_1472dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_018717.4:c.1468_1472dup']['alt_genomic_loci'] == []
        assert results['NM_018717.4:c.1468_1472dup']['gene_symbol'] == 'MAML3'
        assert results['NM_018717.4:c.1468_1472dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_061187.2:p.(Gln491HisfsTer29)', 'slr': 'NP_061187.2:p.(Q491Hfs*29)'}
        assert results['NM_018717.4:c.1468_1472dup']['submitted_variant'] == 'NC_000004.11:g.140811111_140811117del'
        assert results['NM_018717.4:c.1468_1472dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_018717.4:c.1468_1472dup']['hgvs_lrg_variant'] == ''
        assert results['NM_018717.4:c.1468_1472dup']['hgvs_transcript_variant'] == 'NM_018717.4:c.1468_1472dup'
        assert results['NM_018717.4:c.1468_1472dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_018717.4:c.1468_1472dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000004.11:g.140811111_140811117del', 'vcf': {'chr': 'chr4', 'ref': 'GCTGCTGC', 'pos': '140811110', 'alt': 'G'}}
        assert results['NM_018717.4:c.1468_1472dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000004.12:g.139889957_139889963del', 'vcf': {'chr': 'chr4', 'ref': 'GCTGCTGC', 'pos': '139889956', 'alt': 'G'}}
        assert results['NM_018717.4:c.1468_1472dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000004.11:g.140811111_140811117del', 'vcf': {'chr': '4', 'ref': 'GCTGCTGC', 'pos': '140811110', 'alt': 'G'}}
        assert results['NM_018717.4:c.1468_1472dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000004.12:g.139889957_139889963del', 'vcf': {'chr': '4', 'ref': 'GCTGCTGC', 'pos': '139889956', 'alt': 'G'}}


    def test_variant88(self):
        variant = 'NC_000004.11:g.140811117C>A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_018717.4:c.1472_1473insTCAGCAGCAGCA' in results.keys()
        assert results['NM_018717.4:c.1472_1473insTCAGCAGCAGCA']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_018717.4:c.1472_1473insTCAGCAGCAGCA']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_018717.4:c.1472_1473insTCAGCAGCAGCA']['alt_genomic_loci'] == []
        assert results['NM_018717.4:c.1472_1473insTCAGCAGCAGCA']['gene_symbol'] == 'MAML3'
        assert results['NM_018717.4:c.1472_1473insTCAGCAGCAGCA']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_061187.2:p.(Gln490_Gln491insHisGlnGlnGln)', 'slr': 'NP_061187.2:p.(Q490_Q491insHQQQ)'}
        assert results['NM_018717.4:c.1472_1473insTCAGCAGCAGCA']['submitted_variant'] == 'NC_000004.11:g.140811117C>A'
        assert results['NM_018717.4:c.1472_1473insTCAGCAGCAGCA']['genome_context_intronic_sequence'] == ''
        assert results['NM_018717.4:c.1472_1473insTCAGCAGCAGCA']['hgvs_lrg_variant'] == ''
        assert results['NM_018717.4:c.1472_1473insTCAGCAGCAGCA']['hgvs_transcript_variant'] == 'NM_018717.4:c.1472_1473insTCAGCAGCAGCA'
        assert results['NM_018717.4:c.1472_1473insTCAGCAGCAGCA']['hgvs_refseqgene_variant'] == ''
        assert results['NM_018717.4:c.1472_1473insTCAGCAGCAGCA']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000004.11:g.140811117C>A', 'vcf': {'chr': 'chr4', 'ref': 'C', 'pos': '140811117', 'alt': u'A'}}
        assert results['NM_018717.4:c.1472_1473insTCAGCAGCAGCA']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000004.12:g.139889963C>A', 'vcf': {'chr': 'chr4', 'ref': 'C', 'pos': '139889963', 'alt': u'A'}}
        assert results['NM_018717.4:c.1472_1473insTCAGCAGCAGCA']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000004.11:g.140811117C>A', 'vcf': {'chr': '4', 'ref': 'C', 'pos': '140811117', 'alt': u'A'}}
        assert results['NM_018717.4:c.1472_1473insTCAGCAGCAGCA']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000004.12:g.139889963C>A', 'vcf': {'chr': '4', 'ref': 'C', 'pos': '139889963', 'alt': u'A'}}

        assert 'NM_018717.5:c.1473G>T' in results.keys()
        assert results['NM_018717.5:c.1473G>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_018717.5:c.1473G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_018717.5:c.1473G>T']['alt_genomic_loci'] == []
        assert results['NM_018717.5:c.1473G>T']['gene_symbol'] == 'MAML3'
        assert results['NM_018717.5:c.1473G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_061187.3:p.(Gln491His)', 'slr': 'NP_061187.3:p.(Q491H)'}
        assert results['NM_018717.5:c.1473G>T']['submitted_variant'] == 'NC_000004.11:g.140811117C>A'
        assert results['NM_018717.5:c.1473G>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_018717.5:c.1473G>T']['hgvs_lrg_variant'] == ''
        assert results['NM_018717.5:c.1473G>T']['hgvs_transcript_variant'] == 'NM_018717.5:c.1473G>T'
        assert results['NM_018717.5:c.1473G>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_018717.5:c.1473G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000004.11:g.140811117C>A', 'vcf': {'chr': 'chr4', 'ref': u'C', 'pos': '140811117', 'alt': u'A'}}
        assert 'hg38' not in results['NM_018717.5:c.1473G>T']['primary_assembly_loci'].keys()
        assert results['NM_018717.5:c.1473G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000004.11:g.140811117C>A', 'vcf': {'chr': '4', 'ref': u'C', 'pos': '140811117', 'alt': u'A'}}
        assert 'grch38' not in results['NM_018717.5:c.1473G>T']['primary_assembly_loci'].keys()


    def test_variant89(self):
        variant = 'NC_000002.11:g.73675227_73675228insCTC'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_015120.4:c.1573_1579=' in results.keys()
        assert results['NM_015120.4:c.1573_1579=']['hgvs_lrg_transcript_variant'] == 'LRG_741t1:c.1573_1579='
        assert results['NM_015120.4:c.1573_1579=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_015120.4:c.1573_1579=']['alt_genomic_loci'] == []
        assert results['NM_015120.4:c.1573_1579=']['gene_symbol'] == 'ALMS1'
        assert results['NM_015120.4:c.1573_1579=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_055935.4(LRG_741p1):p.(Ser525=)', 'slr': 'NP_055935.4:p.(S525=)'}
        assert results['NM_015120.4:c.1573_1579=']['submitted_variant'] == 'NC_000002.11:g.73675227_73675228insCTC'
        assert results['NM_015120.4:c.1573_1579=']['genome_context_intronic_sequence'] == ''
        assert results['NM_015120.4:c.1573_1579=']['hgvs_lrg_variant'] == 'LRG_741:g.67345_67351='
        assert results['NM_015120.4:c.1573_1579=']['hgvs_transcript_variant'] == 'NM_015120.4:c.1573_1579='
        assert results['NM_015120.4:c.1573_1579=']['hgvs_refseqgene_variant'] == 'NG_011690.1:g.67345_67351='
        assert results['NM_015120.4:c.1573_1579=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.73675228_73675230dup', 'vcf': {'chr': 'chr2', 'ref': 'CTC', 'pos': '73675228', 'alt': u'CTCCTC'}}
        assert results['NM_015120.4:c.1573_1579=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.73448097_73448103=', 'vcf': {'chr': 'chr2', 'ref': u'TCTCCTC', 'pos': '73448097', 'alt': u'TCTCCTC'}}
        assert results['NM_015120.4:c.1573_1579=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.73675228_73675230dup', 'vcf': {'chr': '2', 'ref': 'CTC', 'pos': '73675228', 'alt': u'CTCCTC'}}
        assert results['NM_015120.4:c.1573_1579=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.73448097_73448103=', 'vcf': {'chr': '2', 'ref': u'TCTCCTC', 'pos': '73448097', 'alt': u'TCTCCTC'}}

        assert results['flag'] == 'gene_variant'

    def test_variant90(self):
        variant = '9-136132908-T-TC'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_020469.2:c.260_262=' in results.keys()
        assert results['NM_020469.2:c.260_262=']['hgvs_lrg_transcript_variant'] == 'LRG_792t1:c.260_262='
        assert results['NM_020469.2:c.260_262=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_020469.2:c.260_262=']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003315925.1:g.83614_83616=', 'vcf': {'chr': 'HG79_PATCH', 'ref': u'TCA', 'pos': '83614', 'alt': u'TCA'}}}, {'grch38': {'hgvs_genomic_description': 'NW_009646201.1:g.83614_83616=', 'vcf': {'chr': 'HG2030_PATCH', 'ref': u'TCA', 'pos': '83614', 'alt': u'TCA'}}}, {'hg38': {'hgvs_genomic_description': 'NW_009646201.1:g.83614_83616=', 'vcf': {'chr': 'NW_009646201.1', 'ref': u'TCA', 'pos': '83614', 'alt': u'TCA'}}}]
        assert results['NM_020469.2:c.260_262=']['gene_symbol'] == 'ABO'
        assert results['NM_020469.2:c.260_262=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_065202.2(LRG_792p1):p.(Val87=)', 'slr': 'NP_065202.2:p.(V87=)'}
        assert results['NM_020469.2:c.260_262=']['submitted_variant'] == '9-136132908-T-TC'
        assert results['NM_020469.2:c.260_262=']['genome_context_intronic_sequence'] == ''
        assert results['NM_020469.2:c.260_262=']['hgvs_lrg_variant'] == 'LRG_792:g.20145_20147='
        assert results['NM_020469.2:c.260_262=']['hgvs_transcript_variant'] == 'NM_020469.2:c.260_262='
        assert results['NM_020469.2:c.260_262=']['hgvs_refseqgene_variant'] == 'NG_006669.1:g.20145_20147='
        assert results['NM_020469.2:c.260_262=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.136132908_136132909insC', 'vcf': {'chr': 'chr9', 'ref': 'T', 'pos': '136132908', 'alt': u'TC'}}
        assert results['NM_020469.2:c.260_262=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.133257521_133257522insC', 'vcf': {'chr': 'chr9', 'ref': 'T', 'pos': '133257521', 'alt': u'TC'}}
        assert results['NM_020469.2:c.260_262=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.136132908_136132909insC', 'vcf': {'chr': '9', 'ref': 'T', 'pos': '136132908', 'alt': u'TC'}}
        assert results['NM_020469.2:c.260_262=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.133257521_133257522insC', 'vcf': {'chr': '9', 'ref': 'T', 'pos': '133257521', 'alt': u'TC'}}


    def test_variant91(self):
        variant = '9-136132908-TAC-TCA'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_020469.2:c.259del' in results.keys()
        assert results['NM_020469.2:c.259del']['hgvs_lrg_transcript_variant'] == 'LRG_792t1:c.259del'
        assert results['NM_020469.2:c.259del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_020469.2:c.259del']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003315925.1:g.83618del', 'vcf': {'chr': 'HG79_PATCH', 'ref': 'AC', 'pos': '83616', 'alt': 'A'}}}, {'grch38': {'hgvs_genomic_description': 'NW_009646201.1:g.83618del', 'vcf': {'chr': 'HG2030_PATCH', 'ref': 'AC', 'pos': '83616', 'alt': 'A'}}}, {'hg38': {'hgvs_genomic_description': 'NW_009646201.1:g.83618del', 'vcf': {'chr': 'NW_009646201.1', 'ref': 'AC', 'pos': '83616', 'alt': 'A'}}}]
        assert results['NM_020469.2:c.259del']['gene_symbol'] == 'ABO'
        assert results['NM_020469.2:c.259del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_065202.2(LRG_792p1):p.(Val87Ter)', 'slr': 'NP_065202.2:p.(V87*)'}
        assert results['NM_020469.2:c.259del']['submitted_variant'] == '9-136132908-TAC-TCA'
        assert results['NM_020469.2:c.259del']['genome_context_intronic_sequence'] == ''
        assert results['NM_020469.2:c.259del']['hgvs_lrg_variant'] == 'LRG_792:g.20144del'
        assert results['NM_020469.2:c.259del']['hgvs_transcript_variant'] == 'NM_020469.2:c.259del'
        assert results['NM_020469.2:c.259del']['hgvs_refseqgene_variant'] == 'NG_006669.1:g.20144del'
        assert results['NM_020469.2:c.259del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.136132909_136132910delinsCA', 'vcf': {'chr': 'chr9', 'ref': 'AC', 'pos': '136132909', 'alt': u'CA'}}
        assert results['NM_020469.2:c.259del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.133257522_133257523delinsCA', 'vcf': {'chr': 'chr9', 'ref': 'AC', 'pos': '133257522', 'alt': u'CA'}}
        assert results['NM_020469.2:c.259del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.136132909_136132910delinsCA', 'vcf': {'chr': '9', 'ref': 'AC', 'pos': '136132909', 'alt': u'CA'}}
        assert results['NM_020469.2:c.259del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.133257522_133257523delinsCA', 'vcf': {'chr': '9', 'ref': 'AC', 'pos': '133257522', 'alt': u'CA'}}


    def test_variant92(self):
        variant = '9-136132908-TA-TA'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_020469.2:c.261del' in results.keys()
        assert results['NM_020469.2:c.261del']['hgvs_lrg_transcript_variant'] == 'LRG_792t1:c.261del'
        assert results['NM_020469.2:c.261del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_020469.2:c.261del']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003315925.1:g.83615del', 'vcf': {'chr': 'HG79_PATCH', 'ref': 'TC', 'pos': '83614', 'alt': 'T'}}}, {'grch38': {'hgvs_genomic_description': 'NW_009646201.1:g.83615del', 'vcf': {'chr': 'HG2030_PATCH', 'ref': 'TC', 'pos': '83614', 'alt': 'T'}}}, {'hg38': {'hgvs_genomic_description': 'NW_009646201.1:g.83615del', 'vcf': {'chr': 'NW_009646201.1', 'ref': 'TC', 'pos': '83614', 'alt': 'T'}}}]
        assert results['NM_020469.2:c.261del']['gene_symbol'] == 'ABO'
        assert results['NM_020469.2:c.261del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_065202.2(LRG_792p1):p.(Thr88ProfsTer31)', 'slr': 'NP_065202.2:p.(T88Pfs*31)'}
        assert results['NM_020469.2:c.261del']['submitted_variant'] == '9-136132908-TA-TA'
        assert results['NM_020469.2:c.261del']['genome_context_intronic_sequence'] == ''
        assert results['NM_020469.2:c.261del']['hgvs_lrg_variant'] == 'LRG_792:g.20146del'
        assert results['NM_020469.2:c.261del']['hgvs_transcript_variant'] == 'NM_020469.2:c.261del'
        assert results['NM_020469.2:c.261del']['hgvs_refseqgene_variant'] == 'NG_006669.1:g.20146del'
        assert results['NM_020469.2:c.261del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.136132908_136132909=', 'vcf': {'chr': 'chr9', 'ref': 'TA', 'pos': '136132908', 'alt': 'TA'}}
        assert results['NM_020469.2:c.261del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.133257521_133257522=', 'vcf': {'chr': 'chr9', 'ref': 'TA', 'pos': '133257521', 'alt': 'TA'}}
        assert results['NM_020469.2:c.261del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.136132908_136132909=', 'vcf': {'chr': '9', 'ref': 'TA', 'pos': '136132908', 'alt': 'TA'}}
        assert results['NM_020469.2:c.261del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.133257521_133257522=', 'vcf': {'chr': '9', 'ref': 'TA', 'pos': '133257521', 'alt': 'TA'}}

        assert results['flag'] == 'gene_variant'

    def test_variant93(self):
        variant = 'NM_020469.2:c.258delG'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_020469.2:c.259del' in results.keys()
        assert results['NM_020469.2:c.259del']['hgvs_lrg_transcript_variant'] == 'LRG_792t1:c.259del'
        assert results['NM_020469.2:c.259del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_020469.2:c.259del']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003315925.1:g.83618del', 'vcf': {'chr': 'HG79_PATCH', 'ref': 'AC', 'pos': '83616', 'alt': 'A'}}}, {'grch38': {'hgvs_genomic_description': 'NW_009646201.1:g.83618del', 'vcf': {'chr': 'HG2030_PATCH', 'ref': 'AC', 'pos': '83616', 'alt': 'A'}}}, {'hg38': {'hgvs_genomic_description': 'NW_009646201.1:g.83618del', 'vcf': {'chr': 'NW_009646201.1', 'ref': 'AC', 'pos': '83616', 'alt': 'A'}}}]
        assert results['NM_020469.2:c.259del']['gene_symbol'] == 'ABO'
        assert results['NM_020469.2:c.259del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_065202.2(LRG_792p1):p.(Val87Ter)', 'slr': 'NP_065202.2:p.(V87*)'}
        assert results['NM_020469.2:c.259del']['submitted_variant'] == 'NM_020469.2:c.258delG'
        assert results['NM_020469.2:c.259del']['genome_context_intronic_sequence'] == ''
        assert results['NM_020469.2:c.259del']['hgvs_lrg_variant'] == 'LRG_792:g.20144del'
        assert results['NM_020469.2:c.259del']['hgvs_transcript_variant'] == 'NM_020469.2:c.259del'
        assert results['NM_020469.2:c.259del']['hgvs_refseqgene_variant'] == 'NG_006669.1:g.20144del'
        assert results['NM_020469.2:c.259del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.136132909_136132910delinsCA', 'vcf': {'chr': 'chr9', 'ref': 'AC', 'pos': '136132909', 'alt': u'CA'}}
        assert results['NM_020469.2:c.259del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.133257522_133257523delinsCA', 'vcf': {'chr': 'chr9', 'ref': 'AC', 'pos': '133257522', 'alt': u'CA'}}
        assert results['NM_020469.2:c.259del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.136132909_136132910delinsCA', 'vcf': {'chr': '9', 'ref': 'AC', 'pos': '136132909', 'alt': u'CA'}}
        assert results['NM_020469.2:c.259del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.133257522_133257523delinsCA', 'vcf': {'chr': '9', 'ref': 'AC', 'pos': '133257522', 'alt': u'CA'}}


    def test_variant94(self):
        variant = 'NM_020469.2:c.260_262TGA='
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_020469.2:c.260_262=' in results.keys()
        assert results['NM_020469.2:c.260_262=']['hgvs_lrg_transcript_variant'] == 'LRG_792t1:c.260_262='
        assert results['NM_020469.2:c.260_262=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_020469.2:c.260_262=']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003315925.1:g.83614_83616=', 'vcf': {'chr': 'HG79_PATCH', 'ref': u'TCA', 'pos': '83614', 'alt': u'TCA'}}}, {'grch38': {'hgvs_genomic_description': 'NW_009646201.1:g.83614_83616=', 'vcf': {'chr': 'HG2030_PATCH', 'ref': u'TCA', 'pos': '83614', 'alt': u'TCA'}}}, {'hg38': {'hgvs_genomic_description': 'NW_009646201.1:g.83614_83616=', 'vcf': {'chr': 'NW_009646201.1', 'ref': u'TCA', 'pos': '83614', 'alt': u'TCA'}}}]
        assert results['NM_020469.2:c.260_262=']['gene_symbol'] == 'ABO'
        assert results['NM_020469.2:c.260_262=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_065202.2(LRG_792p1):p.(Val87=)', 'slr': 'NP_065202.2:p.(V87=)'}
        assert results['NM_020469.2:c.260_262=']['submitted_variant'] == 'NM_020469.2:c.260_262TGA='
        assert results['NM_020469.2:c.260_262=']['genome_context_intronic_sequence'] == ''
        assert results['NM_020469.2:c.260_262=']['hgvs_lrg_variant'] == 'LRG_792:g.20145_20147='
        assert results['NM_020469.2:c.260_262=']['hgvs_transcript_variant'] == 'NM_020469.2:c.260_262='
        assert results['NM_020469.2:c.260_262=']['hgvs_refseqgene_variant'] == 'NG_006669.1:g.20145_20147='
        assert results['NM_020469.2:c.260_262=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.136132908_136132909insC', 'vcf': {'chr': 'chr9', 'ref': 'T', 'pos': '136132908', 'alt': u'TC'}}
        assert results['NM_020469.2:c.260_262=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.133257521_133257522insC', 'vcf': {'chr': 'chr9', 'ref': 'T', 'pos': '133257521', 'alt': u'TC'}}
        assert results['NM_020469.2:c.260_262=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.136132908_136132909insC', 'vcf': {'chr': '9', 'ref': 'T', 'pos': '136132908', 'alt': u'TC'}}
        assert results['NM_020469.2:c.260_262=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.133257521_133257522insC', 'vcf': {'chr': '9', 'ref': 'T', 'pos': '133257521', 'alt': u'TC'}}


    def test_variant95(self):
        variant = 'NM_020469.2:c.261delG'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_020469.2:c.261del' in results.keys()
        assert results['NM_020469.2:c.261del']['hgvs_lrg_transcript_variant'] == 'LRG_792t1:c.261del'
        assert results['NM_020469.2:c.261del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_020469.2:c.261del']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003315925.1:g.83615del', 'vcf': {'chr': 'HG79_PATCH', 'ref': 'TC', 'pos': '83614', 'alt': 'T'}}}, {'grch38': {'hgvs_genomic_description': 'NW_009646201.1:g.83615del', 'vcf': {'chr': 'HG2030_PATCH', 'ref': 'TC', 'pos': '83614', 'alt': 'T'}}}, {'hg38': {'hgvs_genomic_description': 'NW_009646201.1:g.83615del', 'vcf': {'chr': 'NW_009646201.1', 'ref': 'TC', 'pos': '83614', 'alt': 'T'}}}]
        assert results['NM_020469.2:c.261del']['gene_symbol'] == 'ABO'
        assert results['NM_020469.2:c.261del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_065202.2(LRG_792p1):p.(Thr88ProfsTer31)', 'slr': 'NP_065202.2:p.(T88Pfs*31)'}
        assert results['NM_020469.2:c.261del']['submitted_variant'] == 'NM_020469.2:c.261delG'
        assert results['NM_020469.2:c.261del']['genome_context_intronic_sequence'] == ''
        assert results['NM_020469.2:c.261del']['hgvs_lrg_variant'] == 'LRG_792:g.20146del'
        assert results['NM_020469.2:c.261del']['hgvs_transcript_variant'] == 'NM_020469.2:c.261del'
        assert results['NM_020469.2:c.261del']['hgvs_refseqgene_variant'] == 'NG_006669.1:g.20146del'
        assert results['NM_020469.2:c.261del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.136132908_136132909=', 'vcf': {'chr': 'chr9', 'ref': 'TA', 'pos': '136132908', 'alt': 'TA'}}
        assert results['NM_020469.2:c.261del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.133257521_133257522=', 'vcf': {'chr': 'chr9', 'ref': 'TA', 'pos': '133257521', 'alt': 'TA'}}
        assert results['NM_020469.2:c.261del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.136132908_136132909=', 'vcf': {'chr': '9', 'ref': 'TA', 'pos': '136132908', 'alt': 'TA'}}
        assert results['NM_020469.2:c.261del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.133257521_133257522=', 'vcf': {'chr': '9', 'ref': 'TA', 'pos': '133257521', 'alt': 'TA'}}

        assert results['flag'] == 'gene_variant'

    def test_variant96(self):
        variant = 'NM_020469.2:c.261dupG'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_020469.2:c.261dup' in results.keys()
        assert results['NM_020469.2:c.261dup']['hgvs_lrg_transcript_variant'] == 'LRG_792t1:c.261dup'
        assert results['NM_020469.2:c.261dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_020469.2:c.261dup']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003315925.1:g.83615dup', 'vcf': {'chr': 'HG79_PATCH', 'ref': 'C', 'pos': '83615', 'alt': u'CC'}}}, {'grch38': {'hgvs_genomic_description': 'NW_009646201.1:g.83615dup', 'vcf': {'chr': 'HG2030_PATCH', 'ref': 'C', 'pos': '83615', 'alt': u'CC'}}}, {'hg38': {'hgvs_genomic_description': 'NW_009646201.1:g.83615dup', 'vcf': {'chr': 'NW_009646201.1', 'ref': 'C', 'pos': '83615', 'alt': u'CC'}}}]
        assert results['NM_020469.2:c.261dup']['gene_symbol'] == 'ABO'
        assert results['NM_020469.2:c.261dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_065202.2(LRG_792p1):p.(Thr88AspfsTer107)', 'slr': 'NP_065202.2:p.(T88Dfs*107)'}
        assert results['NM_020469.2:c.261dup']['submitted_variant'] == 'NM_020469.2:c.261dupG'
        assert results['NM_020469.2:c.261dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_020469.2:c.261dup']['hgvs_lrg_variant'] == 'LRG_792:g.20146dup'
        assert results['NM_020469.2:c.261dup']['hgvs_transcript_variant'] == 'NM_020469.2:c.261dup'
        assert results['NM_020469.2:c.261dup']['hgvs_refseqgene_variant'] == 'NG_006669.1:g.20146dup'
        assert results['NM_020469.2:c.261dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.136132908_136132909insCC', 'vcf': {'chr': 'chr9', 'ref': 'T', 'pos': '136132908', 'alt': u'TCC'}}
        assert results['NM_020469.2:c.261dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.133257521_133257522insCC', 'vcf': {'chr': 'chr9', 'ref': 'T', 'pos': '133257521', 'alt': u'TCC'}}
        assert results['NM_020469.2:c.261dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.136132908_136132909insCC', 'vcf': {'chr': '9', 'ref': 'T', 'pos': '136132908', 'alt': u'TCC'}}
        assert results['NM_020469.2:c.261dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.133257521_133257522insCC', 'vcf': {'chr': '9', 'ref': 'T', 'pos': '133257521', 'alt': u'TCC'}}


    def test_variant97(self):
        variant = 'NM_020469.2:c.261_262insTT'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_020469.2:c.261_262insTT' in results.keys()
        assert results['NM_020469.2:c.261_262insTT']['hgvs_lrg_transcript_variant'] == 'LRG_792t1:c.261_262insTT'
        assert results['NM_020469.2:c.261_262insTT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_020469.2:c.261_262insTT']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003315925.1:g.83614_83615insAA', 'vcf': {'chr': 'HG79_PATCH', 'ref': 'T', 'pos': '83614', 'alt': u'TAA'}}}, {'grch38': {'hgvs_genomic_description': 'NW_009646201.1:g.83614_83615insAA', 'vcf': {'chr': 'HG2030_PATCH', 'ref': 'T', 'pos': '83614', 'alt': u'TAA'}}}, {'hg38': {'hgvs_genomic_description': 'NW_009646201.1:g.83614_83615insAA', 'vcf': {'chr': 'NW_009646201.1', 'ref': 'T', 'pos': '83614', 'alt': u'TAA'}}}]
        assert results['NM_020469.2:c.261_262insTT']['gene_symbol'] == 'ABO'
        assert results['NM_020469.2:c.261_262insTT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_065202.2(LRG_792p1):p.(Thr88LeufsTer32)', 'slr': 'NP_065202.2:p.(T88Lfs*32)'}
        assert results['NM_020469.2:c.261_262insTT']['submitted_variant'] == 'NM_020469.2:c.261_262insTT'
        assert results['NM_020469.2:c.261_262insTT']['genome_context_intronic_sequence'] == ''
        assert results['NM_020469.2:c.261_262insTT']['hgvs_lrg_variant'] == 'LRG_792:g.20146_20147insTT'
        assert results['NM_020469.2:c.261_262insTT']['hgvs_transcript_variant'] == 'NM_020469.2:c.261_262insTT'
        assert results['NM_020469.2:c.261_262insTT']['hgvs_refseqgene_variant'] == 'NG_006669.1:g.20146_20147insTT'
        assert results['NM_020469.2:c.261_262insTT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.136132909_136132910insACA', 'vcf': {'chr': 'chr9', 'ref': 'T', 'pos': '136132908', 'alt': u'TAAC'}}
        assert results['NM_020469.2:c.261_262insTT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.133257522_133257523insACA', 'vcf': {'chr': 'chr9', 'ref': 'T', 'pos': '133257521', 'alt': u'TAAC'}}
        assert results['NM_020469.2:c.261_262insTT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.136132909_136132910insACA', 'vcf': {'chr': '9', 'ref': 'T', 'pos': '136132908', 'alt': u'TAAC'}}
        assert results['NM_020469.2:c.261_262insTT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.133257522_133257523insACA', 'vcf': {'chr': '9', 'ref': 'T', 'pos': '133257521', 'alt': u'TAAC'}}


    def test_variant98(self):
        variant = 'NC_000019.10:g.50378563_50378564insTAC'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_007121.5:c.515A>T' in results.keys()
        assert results['NM_007121.5:c.515A>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_007121.5:c.515A>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_007121.5:c.515A>T']['alt_genomic_loci'] == []
        assert results['NM_007121.5:c.515A>T']['gene_symbol'] == 'NR1H2'
        assert results['NM_007121.5:c.515A>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_009052.3:p.(Lys172Ile)', 'slr': 'NP_009052.3:p.(K172I)'}
        assert results['NM_007121.5:c.515A>T']['submitted_variant'] == 'NC_000019.10:g.50378563_50378564insTAC'
        assert results['NM_007121.5:c.515A>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_007121.5:c.515A>T']['hgvs_lrg_variant'] == ''
        assert results['NM_007121.5:c.515A>T']['hgvs_transcript_variant'] == 'NM_007121.5:c.515A>T'
        assert results['NM_007121.5:c.515A>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_007121.5:c.515A>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881820_50881821insTAC', 'vcf': {'chr': 'chr19', 'ref': 'A', 'pos': '50881820', 'alt': 'ATAC'}}
        assert results['NM_007121.5:c.515A>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378563_50378564insTAC', 'vcf': {'chr': 'chr19', 'ref': 'A', 'pos': '50378563', 'alt': 'ATAC'}}
        assert results['NM_007121.5:c.515A>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881820_50881821insTAC', 'vcf': {'chr': '19', 'ref': 'A', 'pos': '50881820', 'alt': 'ATAC'}}
        assert results['NM_007121.5:c.515A>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378563_50378564insTAC', 'vcf': {'chr': '19', 'ref': 'A', 'pos': '50378563', 'alt': 'ATAC'}}

        assert 'NM_001256647.1:c.224A>T' in results.keys()
        assert results['NM_001256647.1:c.224A>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001256647.1:c.224A>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001256647.1:c.224A>T']['alt_genomic_loci'] == []
        assert results['NM_001256647.1:c.224A>T']['gene_symbol'] == 'NR1H2'
        assert results['NM_001256647.1:c.224A>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001243576.1:p.(Lys75Ile)', 'slr': 'NP_001243576.1:p.(K75I)'}
        assert results['NM_001256647.1:c.224A>T']['submitted_variant'] == 'NC_000019.10:g.50378563_50378564insTAC'
        assert results['NM_001256647.1:c.224A>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001256647.1:c.224A>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001256647.1:c.224A>T']['hgvs_transcript_variant'] == 'NM_001256647.1:c.224A>T'
        assert results['NM_001256647.1:c.224A>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001256647.1:c.224A>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881820_50881821insTAC', 'vcf': {'chr': 'chr19', 'ref': 'A', 'pos': '50881820', 'alt': 'ATAC'}}
        assert results['NM_001256647.1:c.224A>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378563_50378564insTAC', 'vcf': {'chr': 'chr19', 'ref': 'A', 'pos': '50378563', 'alt': 'ATAC'}}
        assert results['NM_001256647.1:c.224A>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881820_50881821insTAC', 'vcf': {'chr': '19', 'ref': 'A', 'pos': '50881820', 'alt': 'ATAC'}}
        assert results['NM_001256647.1:c.224A>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378563_50378564insTAC', 'vcf': {'chr': '19', 'ref': 'A', 'pos': '50378563', 'alt': 'ATAC'}}


    def test_variant99(self):
        variant = 'NC_000019.10:g.50378563_50378564insC'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_007121.5:c.515_516del' in results.keys()
        assert results['NM_007121.5:c.515_516del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_007121.5:c.515_516del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_007121.5:c.515_516del']['alt_genomic_loci'] == []
        assert results['NM_007121.5:c.515_516del']['gene_symbol'] == 'NR1H2'
        assert results['NM_007121.5:c.515_516del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_009052.3:p.(Lys172ThrfsTer34)', 'slr': 'NP_009052.3:p.(K172Tfs*34)'}
        assert results['NM_007121.5:c.515_516del']['submitted_variant'] == 'NC_000019.10:g.50378563_50378564insC'
        assert results['NM_007121.5:c.515_516del']['genome_context_intronic_sequence'] == ''
        assert results['NM_007121.5:c.515_516del']['hgvs_lrg_variant'] == ''
        assert results['NM_007121.5:c.515_516del']['hgvs_transcript_variant'] == 'NM_007121.5:c.515_516del'
        assert results['NM_007121.5:c.515_516del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_007121.5:c.515_516del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881820_50881821insC', 'vcf': {'chr': 'chr19', 'ref': 'A', 'pos': '50881820', 'alt': 'AC'}}
        assert results['NM_007121.5:c.515_516del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378563_50378564insC', 'vcf': {'chr': 'chr19', 'ref': 'A', 'pos': '50378563', 'alt': 'AC'}}
        assert results['NM_007121.5:c.515_516del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881820_50881821insC', 'vcf': {'chr': '19', 'ref': 'A', 'pos': '50881820', 'alt': 'AC'}}
        assert results['NM_007121.5:c.515_516del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378563_50378564insC', 'vcf': {'chr': '19', 'ref': 'A', 'pos': '50378563', 'alt': 'AC'}}

        assert 'NM_001256647.1:c.224_225del' in results.keys()
        assert results['NM_001256647.1:c.224_225del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001256647.1:c.224_225del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001256647.1:c.224_225del']['alt_genomic_loci'] == []
        assert results['NM_001256647.1:c.224_225del']['gene_symbol'] == 'NR1H2'
        assert results['NM_001256647.1:c.224_225del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001243576.1:p.(Lys75ThrfsTer34)', 'slr': 'NP_001243576.1:p.(K75Tfs*34)'}
        assert results['NM_001256647.1:c.224_225del']['submitted_variant'] == 'NC_000019.10:g.50378563_50378564insC'
        assert results['NM_001256647.1:c.224_225del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001256647.1:c.224_225del']['hgvs_lrg_variant'] == ''
        assert results['NM_001256647.1:c.224_225del']['hgvs_transcript_variant'] == 'NM_001256647.1:c.224_225del'
        assert results['NM_001256647.1:c.224_225del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001256647.1:c.224_225del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881820_50881821insC', 'vcf': {'chr': 'chr19', 'ref': 'A', 'pos': '50881820', 'alt': 'AC'}}
        assert results['NM_001256647.1:c.224_225del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378563_50378564insC', 'vcf': {'chr': 'chr19', 'ref': 'A', 'pos': '50378563', 'alt': 'AC'}}
        assert results['NM_001256647.1:c.224_225del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881820_50881821insC', 'vcf': {'chr': '19', 'ref': 'A', 'pos': '50881820', 'alt': 'AC'}}
        assert results['NM_001256647.1:c.224_225del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378563_50378564insC', 'vcf': {'chr': '19', 'ref': 'A', 'pos': '50378563', 'alt': 'AC'}}


    def test_variant100(self):
        variant = 'NC_000019.10:g.50378564_50378565insTACA'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_007121.5:c.515_516insT' in results.keys()
        assert results['NM_007121.5:c.515_516insT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_007121.5:c.515_516insT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_007121.5:c.515_516insT']['alt_genomic_loci'] == []
        assert results['NM_007121.5:c.515_516insT']['gene_symbol'] == 'NR1H2'
        assert results['NM_007121.5:c.515_516insT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_009052.3:p.(Lys172AsnfsTer35)', 'slr': 'NP_009052.3:p.(K172Nfs*35)'}
        assert results['NM_007121.5:c.515_516insT']['submitted_variant'] == 'NC_000019.10:g.50378564_50378565insTACA'
        assert results['NM_007121.5:c.515_516insT']['genome_context_intronic_sequence'] == ''
        assert results['NM_007121.5:c.515_516insT']['hgvs_lrg_variant'] == ''
        assert results['NM_007121.5:c.515_516insT']['hgvs_transcript_variant'] == 'NM_007121.5:c.515_516insT'
        assert results['NM_007121.5:c.515_516insT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_007121.5:c.515_516insT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881821_50881822insTACA', 'vcf': {'chr': 'chr19', 'ref': 'A', 'pos': '50881820', 'alt': 'AATAC'}}
        assert results['NM_007121.5:c.515_516insT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378564_50378565insTACA', 'vcf': {'chr': 'chr19', 'ref': 'A', 'pos': '50378563', 'alt': 'AATAC'}}
        assert results['NM_007121.5:c.515_516insT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881821_50881822insTACA', 'vcf': {'chr': '19', 'ref': 'A', 'pos': '50881820', 'alt': 'AATAC'}}
        assert results['NM_007121.5:c.515_516insT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378564_50378565insTACA', 'vcf': {'chr': '19', 'ref': 'A', 'pos': '50378563', 'alt': 'AATAC'}}

        assert 'NM_001256647.1:c.224_225insT' in results.keys()
        assert results['NM_001256647.1:c.224_225insT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001256647.1:c.224_225insT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001256647.1:c.224_225insT']['alt_genomic_loci'] == []
        assert results['NM_001256647.1:c.224_225insT']['gene_symbol'] == 'NR1H2'
        assert results['NM_001256647.1:c.224_225insT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001243576.1:p.(Lys75AsnfsTer35)', 'slr': 'NP_001243576.1:p.(K75Nfs*35)'}
        assert results['NM_001256647.1:c.224_225insT']['submitted_variant'] == 'NC_000019.10:g.50378564_50378565insTACA'
        assert results['NM_001256647.1:c.224_225insT']['genome_context_intronic_sequence'] == ''
        assert results['NM_001256647.1:c.224_225insT']['hgvs_lrg_variant'] == ''
        assert results['NM_001256647.1:c.224_225insT']['hgvs_transcript_variant'] == 'NM_001256647.1:c.224_225insT'
        assert results['NM_001256647.1:c.224_225insT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001256647.1:c.224_225insT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881821_50881822insTACA', 'vcf': {'chr': 'chr19', 'ref': 'A', 'pos': '50881820', 'alt': 'AATAC'}}
        assert results['NM_001256647.1:c.224_225insT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378564_50378565insTACA', 'vcf': {'chr': 'chr19', 'ref': 'A', 'pos': '50378563', 'alt': 'AATAC'}}
        assert results['NM_001256647.1:c.224_225insT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881821_50881822insTACA', 'vcf': {'chr': '19', 'ref': 'A', 'pos': '50881820', 'alt': 'AATAC'}}
        assert results['NM_001256647.1:c.224_225insT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378564_50378565insTACA', 'vcf': {'chr': '19', 'ref': 'A', 'pos': '50378563', 'alt': 'AATAC'}}


    def test_variant101(self):
        variant = 'NC_000019.10:g.50378565_50378567dup'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_007121.5:c.514_520=' in results.keys()
        assert results['NM_007121.5:c.514_520=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_007121.5:c.514_520=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_007121.5:c.514_520=']['alt_genomic_loci'] == []
        assert results['NM_007121.5:c.514_520=']['gene_symbol'] == 'NR1H2'
        assert results['NM_007121.5:c.514_520=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_009052.3:p.(Lys172=)', 'slr': 'NP_009052.3:p.(K172=)'}
        assert results['NM_007121.5:c.514_520=']['submitted_variant'] == 'NC_000019.10:g.50378565_50378567dup'
        assert results['NM_007121.5:c.514_520=']['genome_context_intronic_sequence'] == ''
        assert results['NM_007121.5:c.514_520=']['hgvs_lrg_variant'] == ''
        assert results['NM_007121.5:c.514_520=']['hgvs_transcript_variant'] == 'NM_007121.5:c.514_520='
        assert results['NM_007121.5:c.514_520=']['hgvs_refseqgene_variant'] == ''
        assert results['NM_007121.5:c.514_520=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881822_50881824dup', 'vcf': {'chr': 'chr19', 'ref': 'AAC', 'pos': '50881821', 'alt': u'AACAAC'}}
        assert results['NM_007121.5:c.514_520=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378565_50378567dup', 'vcf': {'chr': 'chr19', 'ref': 'AAC', 'pos': '50378564', 'alt': u'AACAAC'}}
        assert results['NM_007121.5:c.514_520=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881822_50881824dup', 'vcf': {'chr': '19', 'ref': 'AAC', 'pos': '50881821', 'alt': u'AACAAC'}}
        assert results['NM_007121.5:c.514_520=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378565_50378567dup', 'vcf': {'chr': '19', 'ref': 'AAC', 'pos': '50378564', 'alt': u'AACAAC'}}

        assert 'NM_001256647.1:c.223_229=' in results.keys()
        assert results['NM_001256647.1:c.223_229=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001256647.1:c.223_229=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001256647.1:c.223_229=']['alt_genomic_loci'] == []
        assert results['NM_001256647.1:c.223_229=']['gene_symbol'] == 'NR1H2'
        assert results['NM_001256647.1:c.223_229=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001243576.1:p.(Lys75=)', 'slr': 'NP_001243576.1:p.(K75=)'}
        assert results['NM_001256647.1:c.223_229=']['submitted_variant'] == 'NC_000019.10:g.50378565_50378567dup'
        assert results['NM_001256647.1:c.223_229=']['genome_context_intronic_sequence'] == ''
        assert results['NM_001256647.1:c.223_229=']['hgvs_lrg_variant'] == ''
        assert results['NM_001256647.1:c.223_229=']['hgvs_transcript_variant'] == 'NM_001256647.1:c.223_229='
        assert results['NM_001256647.1:c.223_229=']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001256647.1:c.223_229=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881822_50881824dup', 'vcf': {'chr': 'chr19', 'ref': 'AAC', 'pos': '50881821', 'alt': u'AACAAC'}}
        assert results['NM_001256647.1:c.223_229=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378565_50378567dup', 'vcf': {'chr': 'chr19', 'ref': 'AAC', 'pos': '50378564', 'alt': u'AACAAC'}}
        assert results['NM_001256647.1:c.223_229=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881822_50881824dup', 'vcf': {'chr': '19', 'ref': 'AAC', 'pos': '50881821', 'alt': u'AACAAC'}}
        assert results['NM_001256647.1:c.223_229=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378565_50378567dup', 'vcf': {'chr': '19', 'ref': 'AAC', 'pos': '50378564', 'alt': u'AACAAC'}}

        assert results['flag'] == 'gene_variant'

    def test_variant102(self):
        variant = 'NC_000019.10:g.50378563_50378564='
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_007121.5:c.519_521del' in results.keys()
        assert results['NM_007121.5:c.519_521del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_007121.5:c.519_521del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_007121.5:c.519_521del']['alt_genomic_loci'] == []
        assert results['NM_007121.5:c.519_521del']['gene_symbol'] == 'NR1H2'
        assert results['NM_007121.5:c.519_521del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_009052.3:p.(Gln176del)', 'slr': 'NP_009052.3:p.(Q176del)'}
        assert results['NM_007121.5:c.519_521del']['submitted_variant'] == 'NC_000019.10:g.50378563_50378564='
        assert results['NM_007121.5:c.519_521del']['genome_context_intronic_sequence'] == ''
        assert results['NM_007121.5:c.519_521del']['hgvs_lrg_variant'] == ''
        assert results['NM_007121.5:c.519_521del']['hgvs_transcript_variant'] == 'NM_007121.5:c.519_521del'
        assert results['NM_007121.5:c.519_521del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_007121.5:c.519_521del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881821_50881823=', 'vcf': {'chr': 'chr19', 'ref': 'AAC', 'pos': '50881821', 'alt': 'AAC'}}
        assert results['NM_007121.5:c.519_521del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378564_50378566=', 'vcf': {'chr': 'chr19', 'ref': 'AAC', 'pos': '50378564', 'alt': 'AAC'}}
        assert results['NM_007121.5:c.519_521del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881821_50881823=', 'vcf': {'chr': '19', 'ref': 'AAC', 'pos': '50881821', 'alt': 'AAC'}}
        assert results['NM_007121.5:c.519_521del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378564_50378566=', 'vcf': {'chr': '19', 'ref': 'AAC', 'pos': '50378564', 'alt': 'AAC'}}

        assert 'NM_001256647.1:c.228_230del' in results.keys()
        assert results['NM_001256647.1:c.228_230del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001256647.1:c.228_230del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001256647.1:c.228_230del']['alt_genomic_loci'] == []
        assert results['NM_001256647.1:c.228_230del']['gene_symbol'] == 'NR1H2'
        assert results['NM_001256647.1:c.228_230del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001243576.1:p.(Gln79del)', 'slr': 'NP_001243576.1:p.(Q79del)'}
        assert results['NM_001256647.1:c.228_230del']['submitted_variant'] == 'NC_000019.10:g.50378563_50378564='
        assert results['NM_001256647.1:c.228_230del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001256647.1:c.228_230del']['hgvs_lrg_variant'] == ''
        assert results['NM_001256647.1:c.228_230del']['hgvs_transcript_variant'] == 'NM_001256647.1:c.228_230del'
        assert results['NM_001256647.1:c.228_230del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001256647.1:c.228_230del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881821_50881823=', 'vcf': {'chr': 'chr19', 'ref': 'AAC', 'pos': '50881821', 'alt': 'AAC'}}
        assert results['NM_001256647.1:c.228_230del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378564_50378566=', 'vcf': {'chr': 'chr19', 'ref': 'AAC', 'pos': '50378564', 'alt': 'AAC'}}
        assert results['NM_001256647.1:c.228_230del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881821_50881823=', 'vcf': {'chr': '19', 'ref': 'AAC', 'pos': '50881821', 'alt': 'AAC'}}
        assert results['NM_001256647.1:c.228_230del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378564_50378566=', 'vcf': {'chr': '19', 'ref': 'AAC', 'pos': '50378564', 'alt': 'AAC'}}


    def test_variant103(self):
        variant = 'NC_000019.10:g.50378563_50378564insTCGG'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001256647.1:c.224_226delinsTCGG' in results.keys()
        assert results['NM_001256647.1:c.224_226delinsTCGG']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001256647.1:c.224_226delinsTCGG']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001256647.1:c.224_226delinsTCGG']['alt_genomic_loci'] == []
        assert results['NM_001256647.1:c.224_226delinsTCGG']['gene_symbol'] == 'NR1H2'
        assert results['NM_001256647.1:c.224_226delinsTCGG']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001243576.1:p.(Lys75IlefsTer35)', 'slr': 'NP_001243576.1:p.(K75Ifs*35)'}
        assert results['NM_001256647.1:c.224_226delinsTCGG']['submitted_variant'] == 'NC_000019.10:g.50378563_50378564insTCGG'
        assert results['NM_001256647.1:c.224_226delinsTCGG']['genome_context_intronic_sequence'] == ''
        assert results['NM_001256647.1:c.224_226delinsTCGG']['hgvs_lrg_variant'] == ''
        assert results['NM_001256647.1:c.224_226delinsTCGG']['hgvs_transcript_variant'] == 'NM_001256647.1:c.224_226delinsTCGG'
        assert results['NM_001256647.1:c.224_226delinsTCGG']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001256647.1:c.224_226delinsTCGG']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881820_50881821insTCGG', 'vcf': {'chr': 'chr19', 'ref': 'A', 'pos': '50881820', 'alt': 'ATCGG'}}
        assert results['NM_001256647.1:c.224_226delinsTCGG']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378563_50378564insTCGG', 'vcf': {'chr': 'chr19', 'ref': 'A', 'pos': '50378563', 'alt': 'ATCGG'}}
        assert results['NM_001256647.1:c.224_226delinsTCGG']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881820_50881821insTCGG', 'vcf': {'chr': '19', 'ref': 'A', 'pos': '50881820', 'alt': 'ATCGG'}}
        assert results['NM_001256647.1:c.224_226delinsTCGG']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378563_50378564insTCGG', 'vcf': {'chr': '19', 'ref': 'A', 'pos': '50378563', 'alt': 'ATCGG'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_007121.5:c.515_517delinsTCGG' in results.keys()
        assert results['NM_007121.5:c.515_517delinsTCGG']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_007121.5:c.515_517delinsTCGG']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_007121.5:c.515_517delinsTCGG']['alt_genomic_loci'] == []
        assert results['NM_007121.5:c.515_517delinsTCGG']['gene_symbol'] == 'NR1H2'
        assert results['NM_007121.5:c.515_517delinsTCGG']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_009052.3:p.(Lys172IlefsTer35)', 'slr': 'NP_009052.3:p.(K172Ifs*35)'}
        assert results['NM_007121.5:c.515_517delinsTCGG']['submitted_variant'] == 'NC_000019.10:g.50378563_50378564insTCGG'
        assert results['NM_007121.5:c.515_517delinsTCGG']['genome_context_intronic_sequence'] == ''
        assert results['NM_007121.5:c.515_517delinsTCGG']['hgvs_lrg_variant'] == ''
        assert results['NM_007121.5:c.515_517delinsTCGG']['hgvs_transcript_variant'] == 'NM_007121.5:c.515_517delinsTCGG'
        assert results['NM_007121.5:c.515_517delinsTCGG']['hgvs_refseqgene_variant'] == ''
        assert results['NM_007121.5:c.515_517delinsTCGG']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881820_50881821insTCGG', 'vcf': {'chr': 'chr19', 'ref': 'A', 'pos': '50881820', 'alt': 'ATCGG'}}
        assert results['NM_007121.5:c.515_517delinsTCGG']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378563_50378564insTCGG', 'vcf': {'chr': 'chr19', 'ref': 'A', 'pos': '50378563', 'alt': 'ATCGG'}}
        assert results['NM_007121.5:c.515_517delinsTCGG']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881820_50881821insTCGG', 'vcf': {'chr': '19', 'ref': 'A', 'pos': '50881820', 'alt': 'ATCGG'}}
        assert results['NM_007121.5:c.515_517delinsTCGG']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378563_50378564insTCGG', 'vcf': {'chr': '19', 'ref': 'A', 'pos': '50378563', 'alt': 'ATCGG'}}


    def test_variant104(self):
        variant = 'NC_000019.10:g.50378563delinsTTAC'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_007121.5:c.514_515inv' in results.keys()
        assert results['NM_007121.5:c.514_515inv']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_007121.5:c.514_515inv']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_007121.5:c.514_515inv']['alt_genomic_loci'] == []
        assert results['NM_007121.5:c.514_515inv']['gene_symbol'] == 'NR1H2'
        assert results['NM_007121.5:c.514_515inv']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_009052.3:p.(Lys172Leu)', 'slr': 'NP_009052.3:p.(K172L)'}
        assert results['NM_007121.5:c.514_515inv']['submitted_variant'] == 'NC_000019.10:g.50378563delinsTTAC'
        assert results['NM_007121.5:c.514_515inv']['genome_context_intronic_sequence'] == ''
        assert results['NM_007121.5:c.514_515inv']['hgvs_lrg_variant'] == ''
        assert results['NM_007121.5:c.514_515inv']['hgvs_transcript_variant'] == 'NM_007121.5:c.514_515inv'
        assert results['NM_007121.5:c.514_515inv']['hgvs_refseqgene_variant'] == ''
        assert results['NM_007121.5:c.514_515inv']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881820delinsTTAC', 'vcf': {'chr': 'chr19', 'ref': 'A', 'pos': '50881820', 'alt': 'TTAC'}}
        assert results['NM_007121.5:c.514_515inv']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378563delinsTTAC', 'vcf': {'chr': 'chr19', 'ref': 'A', 'pos': '50378563', 'alt': 'TTAC'}}
        assert results['NM_007121.5:c.514_515inv']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881820delinsTTAC', 'vcf': {'chr': '19', 'ref': 'A', 'pos': '50881820', 'alt': 'TTAC'}}
        assert results['NM_007121.5:c.514_515inv']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378563delinsTTAC', 'vcf': {'chr': '19', 'ref': 'A', 'pos': '50378563', 'alt': 'TTAC'}}

        assert 'NM_001256647.1:c.223_224inv' in results.keys()
        assert results['NM_001256647.1:c.223_224inv']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001256647.1:c.223_224inv']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001256647.1:c.223_224inv']['alt_genomic_loci'] == []
        assert results['NM_001256647.1:c.223_224inv']['gene_symbol'] == 'NR1H2'
        assert results['NM_001256647.1:c.223_224inv']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001243576.1:p.(Lys75Leu)', 'slr': 'NP_001243576.1:p.(K75L)'}
        assert results['NM_001256647.1:c.223_224inv']['submitted_variant'] == 'NC_000019.10:g.50378563delinsTTAC'
        assert results['NM_001256647.1:c.223_224inv']['genome_context_intronic_sequence'] == ''
        assert results['NM_001256647.1:c.223_224inv']['hgvs_lrg_variant'] == ''
        assert results['NM_001256647.1:c.223_224inv']['hgvs_transcript_variant'] == 'NM_001256647.1:c.223_224inv'
        assert results['NM_001256647.1:c.223_224inv']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001256647.1:c.223_224inv']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881820delinsTTAC', 'vcf': {'chr': 'chr19', 'ref': 'A', 'pos': '50881820', 'alt': 'TTAC'}}
        assert results['NM_001256647.1:c.223_224inv']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378563delinsTTAC', 'vcf': {'chr': 'chr19', 'ref': 'A', 'pos': '50378563', 'alt': 'TTAC'}}
        assert results['NM_001256647.1:c.223_224inv']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881820delinsTTAC', 'vcf': {'chr': '19', 'ref': 'A', 'pos': '50881820', 'alt': 'TTAC'}}
        assert results['NM_001256647.1:c.223_224inv']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378563delinsTTAC', 'vcf': {'chr': '19', 'ref': 'A', 'pos': '50378563', 'alt': 'TTAC'}}


    def test_variant105(self):
        variant = 'NC_000019.10:g.50378563_50378564insTAAC'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_007121.5:c.514_515insT' in results.keys()
        assert results['NM_007121.5:c.514_515insT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_007121.5:c.514_515insT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_007121.5:c.514_515insT']['alt_genomic_loci'] == []
        assert results['NM_007121.5:c.514_515insT']['gene_symbol'] == 'NR1H2'
        assert results['NM_007121.5:c.514_515insT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_009052.3:p.(Lys172IlefsTer35)', 'slr': 'NP_009052.3:p.(K172Ifs*35)'}
        assert results['NM_007121.5:c.514_515insT']['submitted_variant'] == 'NC_000019.10:g.50378563_50378564insTAAC'
        assert results['NM_007121.5:c.514_515insT']['genome_context_intronic_sequence'] == ''
        assert results['NM_007121.5:c.514_515insT']['hgvs_lrg_variant'] == ''
        assert results['NM_007121.5:c.514_515insT']['hgvs_transcript_variant'] == 'NM_007121.5:c.514_515insT'
        assert results['NM_007121.5:c.514_515insT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_007121.5:c.514_515insT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881820_50881821insTAAC', 'vcf': {'chr': 'chr19', 'ref': 'A', 'pos': '50881820', 'alt': 'ATAAC'}}
        assert results['NM_007121.5:c.514_515insT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378563_50378564insTAAC', 'vcf': {'chr': 'chr19', 'ref': 'A', 'pos': '50378563', 'alt': 'ATAAC'}}
        assert results['NM_007121.5:c.514_515insT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881820_50881821insTAAC', 'vcf': {'chr': '19', 'ref': 'A', 'pos': '50881820', 'alt': 'ATAAC'}}
        assert results['NM_007121.5:c.514_515insT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378563_50378564insTAAC', 'vcf': {'chr': '19', 'ref': 'A', 'pos': '50378563', 'alt': 'ATAAC'}}

        assert 'NM_001256647.1:c.223_224insT' in results.keys()
        assert results['NM_001256647.1:c.223_224insT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001256647.1:c.223_224insT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001256647.1:c.223_224insT']['alt_genomic_loci'] == []
        assert results['NM_001256647.1:c.223_224insT']['gene_symbol'] == 'NR1H2'
        assert results['NM_001256647.1:c.223_224insT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001243576.1:p.(Lys75IlefsTer35)', 'slr': 'NP_001243576.1:p.(K75Ifs*35)'}
        assert results['NM_001256647.1:c.223_224insT']['submitted_variant'] == 'NC_000019.10:g.50378563_50378564insTAAC'
        assert results['NM_001256647.1:c.223_224insT']['genome_context_intronic_sequence'] == ''
        assert results['NM_001256647.1:c.223_224insT']['hgvs_lrg_variant'] == ''
        assert results['NM_001256647.1:c.223_224insT']['hgvs_transcript_variant'] == 'NM_001256647.1:c.223_224insT'
        assert results['NM_001256647.1:c.223_224insT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001256647.1:c.223_224insT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881820_50881821insTAAC', 'vcf': {'chr': 'chr19', 'ref': 'A', 'pos': '50881820', 'alt': 'ATAAC'}}
        assert results['NM_001256647.1:c.223_224insT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378563_50378564insTAAC', 'vcf': {'chr': 'chr19', 'ref': 'A', 'pos': '50378563', 'alt': 'ATAAC'}}
        assert results['NM_001256647.1:c.223_224insT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881820_50881821insTAAC', 'vcf': {'chr': '19', 'ref': 'A', 'pos': '50881820', 'alt': 'ATAAC'}}
        assert results['NM_001256647.1:c.223_224insT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378563_50378564insTAAC', 'vcf': {'chr': '19', 'ref': 'A', 'pos': '50378563', 'alt': 'ATAAC'}}

        assert results['flag'] == 'gene_variant'

    def test_variant106(self):
        variant = 'NC_000019.10:g.50378562_50378565del'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_001256647.1:c.222_228del' in results.keys()
        assert results['NM_001256647.1:c.222_228del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001256647.1:c.222_228del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001256647.1:c.222_228del']['alt_genomic_loci'] == []
        assert results['NM_001256647.1:c.222_228del']['gene_symbol'] == 'NR1H2'
        assert results['NM_001256647.1:c.222_228del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001243576.1:p.(Lys75SerfsTer47)', 'slr': 'NP_001243576.1:p.(K75Sfs*47)'}
        assert results['NM_001256647.1:c.222_228del']['submitted_variant'] == 'NC_000019.10:g.50378562_50378565del'
        assert results['NM_001256647.1:c.222_228del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001256647.1:c.222_228del']['hgvs_lrg_variant'] == ''
        assert results['NM_001256647.1:c.222_228del']['hgvs_transcript_variant'] == 'NM_001256647.1:c.222_228del'
        assert results['NM_001256647.1:c.222_228del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001256647.1:c.222_228del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881819_50881822del', 'vcf': {'chr': 'chr19', 'ref': 'GGAAA', 'pos': '50881818', 'alt': 'G'}}
        assert results['NM_001256647.1:c.222_228del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378562_50378565del', 'vcf': {'chr': 'chr19', 'ref': 'GGAAA', 'pos': '50378561', 'alt': 'G'}}
        assert results['NM_001256647.1:c.222_228del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881819_50881822del', 'vcf': {'chr': '19', 'ref': 'GGAAA', 'pos': '50881818', 'alt': 'G'}}
        assert results['NM_001256647.1:c.222_228del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378562_50378565del', 'vcf': {'chr': '19', 'ref': 'GGAAA', 'pos': '50378561', 'alt': 'G'}}

        assert 'NM_007121.5:c.513_519del' in results.keys()
        assert results['NM_007121.5:c.513_519del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_007121.5:c.513_519del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_007121.5:c.513_519del']['alt_genomic_loci'] == []
        assert results['NM_007121.5:c.513_519del']['gene_symbol'] == 'NR1H2'
        assert results['NM_007121.5:c.513_519del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_009052.3:p.(Lys172SerfsTer47)', 'slr': 'NP_009052.3:p.(K172Sfs*47)'}
        assert results['NM_007121.5:c.513_519del']['submitted_variant'] == 'NC_000019.10:g.50378562_50378565del'
        assert results['NM_007121.5:c.513_519del']['genome_context_intronic_sequence'] == ''
        assert results['NM_007121.5:c.513_519del']['hgvs_lrg_variant'] == ''
        assert results['NM_007121.5:c.513_519del']['hgvs_transcript_variant'] == 'NM_007121.5:c.513_519del'
        assert results['NM_007121.5:c.513_519del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_007121.5:c.513_519del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881819_50881822del', 'vcf': {'chr': 'chr19', 'ref': 'GGAAA', 'pos': '50881818', 'alt': 'G'}}
        assert results['NM_007121.5:c.513_519del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378562_50378565del', 'vcf': {'chr': 'chr19', 'ref': 'GGAAA', 'pos': '50378561', 'alt': 'G'}}
        assert results['NM_007121.5:c.513_519del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881819_50881822del', 'vcf': {'chr': '19', 'ref': 'GGAAA', 'pos': '50881818', 'alt': 'G'}}
        assert results['NM_007121.5:c.513_519del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378562_50378565del', 'vcf': {'chr': '19', 'ref': 'GGAAA', 'pos': '50378561', 'alt': 'G'}}


    def test_variant107(self):
        variant = 'NC_000019.10:g.50378562_50378565delinsTC'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001256647.1:c.222_228delinsTC' in results.keys()
        assert results['NM_001256647.1:c.222_228delinsTC']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001256647.1:c.222_228delinsTC']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001256647.1:c.222_228delinsTC']['alt_genomic_loci'] == []
        assert results['NM_001256647.1:c.222_228delinsTC']['gene_symbol'] == 'NR1H2'
        assert results['NM_001256647.1:c.222_228delinsTC']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001243576.1:p.(Lys75ProfsTer33)', 'slr': 'NP_001243576.1:p.(K75Pfs*33)'}
        assert results['NM_001256647.1:c.222_228delinsTC']['submitted_variant'] == 'NC_000019.10:g.50378562_50378565delinsTC'
        assert results['NM_001256647.1:c.222_228delinsTC']['genome_context_intronic_sequence'] == ''
        assert results['NM_001256647.1:c.222_228delinsTC']['hgvs_lrg_variant'] == ''
        assert results['NM_001256647.1:c.222_228delinsTC']['hgvs_transcript_variant'] == 'NM_001256647.1:c.222_228delinsTC'
        assert results['NM_001256647.1:c.222_228delinsTC']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001256647.1:c.222_228delinsTC']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881819_50881822delinsTC', 'vcf': {'chr': 'chr19', 'ref': 'GAAA', 'pos': '50881819', 'alt': 'TC'}}
        assert results['NM_001256647.1:c.222_228delinsTC']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378562_50378565delinsTC', 'vcf': {'chr': 'chr19', 'ref': 'GAAA', 'pos': '50378562', 'alt': 'TC'}}
        assert results['NM_001256647.1:c.222_228delinsTC']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881819_50881822delinsTC', 'vcf': {'chr': '19', 'ref': 'GAAA', 'pos': '50881819', 'alt': 'TC'}}
        assert results['NM_001256647.1:c.222_228delinsTC']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378562_50378565delinsTC', 'vcf': {'chr': '19', 'ref': 'GAAA', 'pos': '50378562', 'alt': 'TC'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_007121.5:c.513_519delinsTC' in results.keys()
        assert results['NM_007121.5:c.513_519delinsTC']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_007121.5:c.513_519delinsTC']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_007121.5:c.513_519delinsTC']['alt_genomic_loci'] == []
        assert results['NM_007121.5:c.513_519delinsTC']['gene_symbol'] == 'NR1H2'
        assert results['NM_007121.5:c.513_519delinsTC']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_009052.3:p.(Lys172ProfsTer33)', 'slr': 'NP_009052.3:p.(K172Pfs*33)'}
        assert results['NM_007121.5:c.513_519delinsTC']['submitted_variant'] == 'NC_000019.10:g.50378562_50378565delinsTC'
        assert results['NM_007121.5:c.513_519delinsTC']['genome_context_intronic_sequence'] == ''
        assert results['NM_007121.5:c.513_519delinsTC']['hgvs_lrg_variant'] == ''
        assert results['NM_007121.5:c.513_519delinsTC']['hgvs_transcript_variant'] == 'NM_007121.5:c.513_519delinsTC'
        assert results['NM_007121.5:c.513_519delinsTC']['hgvs_refseqgene_variant'] == ''
        assert results['NM_007121.5:c.513_519delinsTC']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881819_50881822delinsTC', 'vcf': {'chr': 'chr19', 'ref': 'GAAA', 'pos': '50881819', 'alt': 'TC'}}
        assert results['NM_007121.5:c.513_519delinsTC']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378562_50378565delinsTC', 'vcf': {'chr': 'chr19', 'ref': 'GAAA', 'pos': '50378562', 'alt': 'TC'}}
        assert results['NM_007121.5:c.513_519delinsTC']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.50881819_50881822delinsTC', 'vcf': {'chr': '19', 'ref': 'GAAA', 'pos': '50881819', 'alt': 'TC'}}
        assert results['NM_007121.5:c.513_519delinsTC']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.50378562_50378565delinsTC', 'vcf': {'chr': '19', 'ref': 'GAAA', 'pos': '50378562', 'alt': 'TC'}}


    def test_variant108(self):
        variant = 'NC_000007.14:g.149779575_149779577delinsT'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_198455.2:c.1115_1116insT' in results.keys()
        assert results['NM_198455.2:c.1115_1116insT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_198455.2:c.1115_1116insT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_198455.2:c.1115_1116insT']['alt_genomic_loci'] == []
        assert results['NM_198455.2:c.1115_1116insT']['gene_symbol'] == 'SSPO'
        assert results['NM_198455.2:c.1115_1116insT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_940857.2:p.(Leu374ProfsTer16)', 'slr': 'NP_940857.2:p.(L374Pfs*16)'}
        assert results['NM_198455.2:c.1115_1116insT']['submitted_variant'] == 'NC_000007.14:g.149779575_149779577delinsT'
        assert results['NM_198455.2:c.1115_1116insT']['genome_context_intronic_sequence'] == ''
        assert results['NM_198455.2:c.1115_1116insT']['hgvs_lrg_variant'] == ''
        assert results['NM_198455.2:c.1115_1116insT']['hgvs_transcript_variant'] == 'NM_198455.2:c.1115_1116insT'
        assert results['NM_198455.2:c.1115_1116insT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_198455.2:c.1115_1116insT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.149476664_149476666delinsTC', 'vcf': {'chr': 'chr7', 'ref': 'CAG', 'pos': '149476664', 'alt': 'TC'}}
        assert results['NM_198455.2:c.1115_1116insT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.149779575_149779577delinsT', 'vcf': {'chr': 'chr7', 'ref': 'CAG', 'pos': '149779575', 'alt': 'T'}}
        assert results['NM_198455.2:c.1115_1116insT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.149476664_149476666delinsTC', 'vcf': {'chr': '7', 'ref': 'CAG', 'pos': '149476664', 'alt': 'TC'}}
        assert results['NM_198455.2:c.1115_1116insT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.149779575_149779577delinsT', 'vcf': {'chr': '7', 'ref': 'CAG', 'pos': '149779575', 'alt': 'T'}}


    def test_variant109(self):
        variant = 'NC_000007.14:g.149779575_149779577='
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_198455.2:c.1116_1118=' in results.keys()
        assert results['NM_198455.2:c.1116_1118=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_198455.2:c.1116_1118=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_198455.2:c.1116_1118=']['alt_genomic_loci'] == []
        assert results['NM_198455.2:c.1116_1118=']['gene_symbol'] == 'SSPO'
        assert results['NM_198455.2:c.1116_1118=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_940857.2:p.(Asp372=)', 'slr': 'NP_940857.2:p.(D372=)'}
        assert results['NM_198455.2:c.1116_1118=']['submitted_variant'] == 'NC_000007.14:g.149779575_149779577='
        assert results['NM_198455.2:c.1116_1118=']['genome_context_intronic_sequence'] == ''
        assert results['NM_198455.2:c.1116_1118=']['hgvs_lrg_variant'] == ''
        assert results['NM_198455.2:c.1116_1118=']['hgvs_transcript_variant'] == 'NM_198455.2:c.1116_1118='
        assert results['NM_198455.2:c.1116_1118=']['hgvs_refseqgene_variant'] == ''
        assert results['NM_198455.2:c.1116_1118=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.149476665_149476666del', 'vcf': {'chr': 'chr7', 'ref': 'CAG', 'pos': '149476664', 'alt': 'C'}}
        assert results['NM_198455.2:c.1116_1118=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.149779576_149779578del', 'vcf': {'chr': 'chr7', 'ref': 'ACAG', 'pos': '149779574', 'alt': 'A'}}
        assert results['NM_198455.2:c.1116_1118=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.149476665_149476666del', 'vcf': {'chr': '7', 'ref': 'CAG', 'pos': '149476664', 'alt': 'C'}}
        assert results['NM_198455.2:c.1116_1118=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.149779576_149779578del', 'vcf': {'chr': '7', 'ref': 'ACAG', 'pos': '149779574', 'alt': 'A'}}


    def test_variant110(self):
        variant = 'NC_000007.14:g.149779576_149779578del'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_198455.2:c.1116_1118=' in results.keys()
        assert results['NM_198455.2:c.1116_1118=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_198455.2:c.1116_1118=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_198455.2:c.1116_1118=']['alt_genomic_loci'] == []
        assert results['NM_198455.2:c.1116_1118=']['gene_symbol'] == 'SSPO'
        assert results['NM_198455.2:c.1116_1118=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_940857.2:p.(Asp372=)', 'slr': 'NP_940857.2:p.(D372=)'}
        assert results['NM_198455.2:c.1116_1118=']['submitted_variant'] == 'NC_000007.14:g.149779576_149779578del'
        assert results['NM_198455.2:c.1116_1118=']['genome_context_intronic_sequence'] == ''
        assert results['NM_198455.2:c.1116_1118=']['hgvs_lrg_variant'] == ''
        assert results['NM_198455.2:c.1116_1118=']['hgvs_transcript_variant'] == 'NM_198455.2:c.1116_1118='
        assert results['NM_198455.2:c.1116_1118=']['hgvs_refseqgene_variant'] == ''
        assert results['NM_198455.2:c.1116_1118=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.149476665_149476666del', 'vcf': {'chr': 'chr7', 'ref': 'CAG', 'pos': '149476664', 'alt': 'C'}}
        assert results['NM_198455.2:c.1116_1118=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.149779576_149779578del', 'vcf': {'chr': 'chr7', 'ref': 'ACAG', 'pos': '149779574', 'alt': 'A'}}
        assert results['NM_198455.2:c.1116_1118=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.149476665_149476666del', 'vcf': {'chr': '7', 'ref': 'CAG', 'pos': '149476664', 'alt': 'C'}}
        assert results['NM_198455.2:c.1116_1118=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.149779576_149779578del', 'vcf': {'chr': '7', 'ref': 'ACAG', 'pos': '149779574', 'alt': 'A'}}


    def test_variant111(self):
        variant = 'NC_000007.14:g.149779577del'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_198455.2:c.1115_1116dup' in results.keys()
        assert results['NM_198455.2:c.1115_1116dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_198455.2:c.1115_1116dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_198455.2:c.1115_1116dup']['alt_genomic_loci'] == []
        assert results['NM_198455.2:c.1115_1116dup']['gene_symbol'] == 'SSPO'
        assert results['NM_198455.2:c.1115_1116dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_940857.2:p.(Pro373ThrfsTer6)', 'slr': 'NP_940857.2:p.(P373Tfs*6)'}
        assert results['NM_198455.2:c.1115_1116dup']['submitted_variant'] == 'NC_000007.14:g.149779577del'
        assert results['NM_198455.2:c.1115_1116dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_198455.2:c.1115_1116dup']['hgvs_lrg_variant'] == ''
        assert results['NM_198455.2:c.1115_1116dup']['hgvs_transcript_variant'] == 'NM_198455.2:c.1115_1116dup'
        assert results['NM_198455.2:c.1115_1116dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_198455.2:c.1115_1116dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.149476666G>C', 'vcf': {'chr': 'chr7', 'ref': 'G', 'pos': '149476666', 'alt': 'C'}}
        assert results['NM_198455.2:c.1115_1116dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.149779577del', 'vcf': {'chr': 'chr7', 'ref': 'AG', 'pos': '149779576', 'alt': 'A'}}
        assert results['NM_198455.2:c.1115_1116dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.149476666G>C', 'vcf': {'chr': '7', 'ref': 'G', 'pos': '149476666', 'alt': 'C'}}
        assert results['NM_198455.2:c.1115_1116dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.149779577del', 'vcf': {'chr': '7', 'ref': 'AG', 'pos': '149779576', 'alt': 'A'}}


    def test_variant112(self):
        variant = 'NC_000007.14:g.149779573_149779579del'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_198455.2:c.1114_1117del' in results.keys()
        assert results['NM_198455.2:c.1114_1117del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_198455.2:c.1114_1117del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_198455.2:c.1114_1117del']['alt_genomic_loci'] == []
        assert results['NM_198455.2:c.1114_1117del']['gene_symbol'] == 'SSPO'
        assert results['NM_198455.2:c.1114_1117del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_940857.2:p.(Asp372ProfsTer5)', 'slr': 'NP_940857.2:p.(D372Pfs*5)'}
        assert results['NM_198455.2:c.1114_1117del']['submitted_variant'] == 'NC_000007.14:g.149779573_149779579del'
        assert results['NM_198455.2:c.1114_1117del']['genome_context_intronic_sequence'] == ''
        assert results['NM_198455.2:c.1114_1117del']['hgvs_lrg_variant'] == ''
        assert results['NM_198455.2:c.1114_1117del']['hgvs_transcript_variant'] == 'NM_198455.2:c.1114_1117del'
        assert results['NM_198455.2:c.1114_1117del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_198455.2:c.1114_1117del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.149476662_149476667del', 'vcf': {'chr': 'chr7', 'ref': 'TGACAGC', 'pos': '149476661', 'alt': 'T'}}
        assert results['NM_198455.2:c.1114_1117del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.149779573_149779579del', 'vcf': {'chr': 'chr7', 'ref': 'TGACAGCC', 'pos': '149779572', 'alt': 'T'}}
        assert results['NM_198455.2:c.1114_1117del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.149476662_149476667del', 'vcf': {'chr': '7', 'ref': 'TGACAGC', 'pos': '149476661', 'alt': 'T'}}
        assert results['NM_198455.2:c.1114_1117del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.149779573_149779579del', 'vcf': {'chr': '7', 'ref': 'TGACAGCC', 'pos': '149779572', 'alt': 'T'}}

        assert results['flag'] == 'gene_variant'

    def test_variant113(self):
        variant = 'NC_000007.14:g.149779573_149779579delinsCA'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_198455.2:c.1114_1117delinsCA' in results.keys()
        assert results['NM_198455.2:c.1114_1117delinsCA']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_198455.2:c.1114_1117delinsCA']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_198455.2:c.1114_1117delinsCA']['alt_genomic_loci'] == []
        assert results['NM_198455.2:c.1114_1117delinsCA']['gene_symbol'] == 'SSPO'
        assert results['NM_198455.2:c.1114_1117delinsCA']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_940857.2:p.(Asp372HisfsTer17)', 'slr': 'NP_940857.2:p.(D372Hfs*17)'}
        assert results['NM_198455.2:c.1114_1117delinsCA']['submitted_variant'] == 'NC_000007.14:g.149779573_149779579delinsCA'
        assert results['NM_198455.2:c.1114_1117delinsCA']['genome_context_intronic_sequence'] == ''
        assert results['NM_198455.2:c.1114_1117delinsCA']['hgvs_lrg_variant'] == ''
        assert results['NM_198455.2:c.1114_1117delinsCA']['hgvs_transcript_variant'] == 'NM_198455.2:c.1114_1117delinsCA'
        assert results['NM_198455.2:c.1114_1117delinsCA']['hgvs_refseqgene_variant'] == ''
        assert results['NM_198455.2:c.1114_1117delinsCA']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.149476662_149476667delinsCA', 'vcf': {'chr': 'chr7', 'ref': 'GACAGC', 'pos': '149476662', 'alt': 'CA'}}
        assert results['NM_198455.2:c.1114_1117delinsCA']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.149779573_149779579delinsCA', 'vcf': {'chr': 'chr7', 'ref': 'GACAGCC', 'pos': '149779573', 'alt': 'CA'}}
        assert results['NM_198455.2:c.1114_1117delinsCA']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.149476662_149476667delinsCA', 'vcf': {'chr': '7', 'ref': 'GACAGC', 'pos': '149476662', 'alt': 'CA'}}
        assert results['NM_198455.2:c.1114_1117delinsCA']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.149779573_149779579delinsCA', 'vcf': {'chr': '7', 'ref': 'GACAGCC', 'pos': '149779573', 'alt': 'CA'}}

        assert results['flag'] == 'gene_variant'

    def test_variant114(self):
        variant = 'NM_000088.3:c.590_591inv'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_000088.3:c.590_591inv' in results.keys()
        assert results['NM_000088.3:c.590_591inv']['hgvs_lrg_transcript_variant'] == 'LRG_1t1:c.590_591inv'
        assert results['NM_000088.3:c.590_591inv']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000088.3:c.590_591inv']['alt_genomic_loci'] == []
        assert results['NM_000088.3:c.590_591inv']['gene_symbol'] == 'COL1A1'
        assert results['NM_000088.3:c.590_591inv']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000079.2(LRG_1p1):p.(Gly197Asp)', 'slr': 'NP_000079.2:p.(G197D)'}
        assert results['NM_000088.3:c.590_591inv']['submitted_variant'] == 'NM_000088.3:c.590_591inv'
        assert results['NM_000088.3:c.590_591inv']['genome_context_intronic_sequence'] == ''
        assert results['NM_000088.3:c.590_591inv']['hgvs_lrg_variant'] == 'LRG_1:g.8639_8640inv'
        assert results['NM_000088.3:c.590_591inv']['hgvs_transcript_variant'] == 'NM_000088.3:c.590_591inv'
        assert results['NM_000088.3:c.590_591inv']['hgvs_refseqgene_variant'] == 'NG_007400.1:g.8639_8640inv'
        assert results['NM_000088.3:c.590_591inv']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275361_48275362inv', 'vcf': {'chr': 'chr17', 'ref': 'AC', 'pos': '48275361', 'alt': 'GT'}}
        assert results['NM_000088.3:c.590_591inv']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198000_50198001inv', 'vcf': {'chr': 'chr17', 'ref': 'AC', 'pos': '50198000', 'alt': 'GT'}}
        assert results['NM_000088.3:c.590_591inv']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275361_48275362inv', 'vcf': {'chr': '17', 'ref': 'AC', 'pos': '48275361', 'alt': 'GT'}}
        assert results['NM_000088.3:c.590_591inv']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198000_50198001inv', 'vcf': {'chr': '17', 'ref': 'AC', 'pos': '50198000', 'alt': 'GT'}}

        assert results['flag'] == 'gene_variant'

    def test_variant115(self):
        variant = 'NM_024989.3:c.1778_1779inv'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_024989.3:c.1778_1779inv' in results.keys()
        assert results['NM_024989.3:c.1778_1779inv']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_024989.3:c.1778_1779inv']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_024989.3:c.1778_1779inv']['alt_genomic_loci'] == []
        assert results['NM_024989.3:c.1778_1779inv']['gene_symbol'] == 'PGAP1'
        assert results['NM_024989.3:c.1778_1779inv']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_079265.2:p.(Phe593Ter)', 'slr': 'NP_079265.2:p.(F593*)'}
        assert results['NM_024989.3:c.1778_1779inv']['submitted_variant'] == 'NM_024989.3:c.1778_1779inv'
        assert results['NM_024989.3:c.1778_1779inv']['genome_context_intronic_sequence'] == ''
        assert results['NM_024989.3:c.1778_1779inv']['hgvs_lrg_variant'] == ''
        assert results['NM_024989.3:c.1778_1779inv']['hgvs_transcript_variant'] == 'NM_024989.3:c.1778_1779inv'
        assert results['NM_024989.3:c.1778_1779inv']['hgvs_refseqgene_variant'] == ''
        assert results['NM_024989.3:c.1778_1779inv']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.197729793_197729794inv', 'vcf': {'chr': 'chr2', 'ref': 'AA', 'pos': '197729793', 'alt': 'TT'}}
        assert results['NM_024989.3:c.1778_1779inv']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.196865069_196865070inv', 'vcf': {'chr': 'chr2', 'ref': 'AA', 'pos': '196865069', 'alt': 'TT'}}
        assert results['NM_024989.3:c.1778_1779inv']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.197729793_197729794inv', 'vcf': {'chr': '2', 'ref': 'AA', 'pos': '197729793', 'alt': 'TT'}}
        assert results['NM_024989.3:c.1778_1779inv']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.196865069_196865070inv', 'vcf': {'chr': '2', 'ref': 'AA', 'pos': '196865069', 'alt': 'TT'}}

        assert results['flag'] == 'gene_variant'

    def test_variant116(self):
        variant = 'NM_032815.3:c.555_556inv'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_032815.3:c.555_556inv' in results.keys()
        assert results['NM_032815.3:c.555_556inv']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_032815.3:c.555_556inv']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_032815.3:c.555_556inv']['alt_genomic_loci'] == []
        assert results['NM_032815.3:c.555_556inv']['gene_symbol'] == 'NFATC2IP'
        assert results['NM_032815.3:c.555_556inv']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_116204.3:p.(Glu185_Glu186delinsAspTer)', 'slr': 'NP_116204.3:p.(E185_E186delinsD*)'}
        assert results['NM_032815.3:c.555_556inv']['submitted_variant'] == 'NM_032815.3:c.555_556inv'
        assert results['NM_032815.3:c.555_556inv']['genome_context_intronic_sequence'] == ''
        assert results['NM_032815.3:c.555_556inv']['hgvs_lrg_variant'] == ''
        assert results['NM_032815.3:c.555_556inv']['hgvs_transcript_variant'] == 'NM_032815.3:c.555_556inv'
        assert results['NM_032815.3:c.555_556inv']['hgvs_refseqgene_variant'] == ''
        assert results['NM_032815.3:c.555_556inv']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.28965980_28965981inv', 'vcf': {'chr': 'chr16', 'ref': 'AG', 'pos': '28965980', 'alt': 'CT'}}
        assert results['NM_032815.3:c.555_556inv']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.28954659_28954660inv', 'vcf': {'chr': 'chr16', 'ref': 'AG', 'pos': '28954659', 'alt': 'CT'}}
        assert results['NM_032815.3:c.555_556inv']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.28965980_28965981inv', 'vcf': {'chr': '16', 'ref': 'AG', 'pos': '28965980', 'alt': 'CT'}}
        assert results['NM_032815.3:c.555_556inv']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.28954659_28954660inv', 'vcf': {'chr': '16', 'ref': 'AG', 'pos': '28954659', 'alt': 'CT'}}


    def test_variant117(self):
        variant = 'NM_006138.4:c.3_4inv'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_006138.4:c.3_4inv' in results.keys()
        assert results['NM_006138.4:c.3_4inv']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_006138.4:c.3_4inv']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_006138.4:c.3_4inv']['alt_genomic_loci'] == []
        assert results['NM_006138.4:c.3_4inv']['gene_symbol'] == 'MS4A3'
        assert results['NM_006138.4:c.3_4inv']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_006129.4:p.(Met1?)', 'slr': 'NP_006129.4:p.(M1?)'}
        assert results['NM_006138.4:c.3_4inv']['submitted_variant'] == 'NM_006138.4:c.3_4inv'
        assert results['NM_006138.4:c.3_4inv']['genome_context_intronic_sequence'] == ''
        assert results['NM_006138.4:c.3_4inv']['hgvs_lrg_variant'] == ''
        assert results['NM_006138.4:c.3_4inv']['hgvs_transcript_variant'] == 'NM_006138.4:c.3_4inv'
        assert results['NM_006138.4:c.3_4inv']['hgvs_refseqgene_variant'] == ''
        assert results['NM_006138.4:c.3_4inv']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.59828636_59828637inv', 'vcf': {'chr': 'chr11', 'ref': 'GG', 'pos': '59828636', 'alt': 'CC'}}
        assert results['NM_006138.4:c.3_4inv']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.60061163_60061164inv', 'vcf': {'chr': 'chr11', 'ref': 'GG', 'pos': '60061163', 'alt': 'CC'}}
        assert results['NM_006138.4:c.3_4inv']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.59828636_59828637inv', 'vcf': {'chr': '11', 'ref': 'GG', 'pos': '59828636', 'alt': 'CC'}}
        assert results['NM_006138.4:c.3_4inv']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.60061163_60061164inv', 'vcf': {'chr': '11', 'ref': 'GG', 'pos': '60061163', 'alt': 'CC'}}

        assert results['flag'] == 'gene_variant'

    def test_variant118(self):
        variant = 'NM_000038.5:c.3927_3928delAAinsTT'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000038.5:c.3927_3928inv' in results.keys()
        assert results['NM_000038.5:c.3927_3928inv']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000038.5:c.3927_3928inv']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000038.5:c.3927_3928inv']['alt_genomic_loci'] == []
        assert results['NM_000038.5:c.3927_3928inv']['gene_symbol'] == 'APC'
        assert results['NM_000038.5:c.3927_3928inv']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000029.2(LRG_130p1):p.(Glu1309_Lys1310delinsAspTer)', 'slr': 'NP_000029.2:p.(E1309_K1310delinsD*)'}
        assert results['NM_000038.5:c.3927_3928inv']['submitted_variant'] == 'NM_000038.5:c.3927_3928delAAinsTT'
        assert results['NM_000038.5:c.3927_3928inv']['genome_context_intronic_sequence'] == ''
        assert results['NM_000038.5:c.3927_3928inv']['hgvs_lrg_variant'] == ''
        assert results['NM_000038.5:c.3927_3928inv']['hgvs_transcript_variant'] == 'NM_000038.5:c.3927_3928inv'
        assert results['NM_000038.5:c.3927_3928inv']['hgvs_refseqgene_variant'] == ''
        assert results['NM_000038.5:c.3927_3928inv']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.112175218_112175219inv', 'vcf': {'chr': 'chr5', 'ref': 'AA', 'pos': '112175218', 'alt': 'TT'}}
        assert results['NM_000038.5:c.3927_3928inv']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000005.10:g.112839521_112839522inv', 'vcf': {'chr': 'chr5', 'ref': 'AA', 'pos': '112839521', 'alt': 'TT'}}
        assert results['NM_000038.5:c.3927_3928inv']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.112175218_112175219inv', 'vcf': {'chr': '5', 'ref': 'AA', 'pos': '112175218', 'alt': 'TT'}}
        assert results['NM_000038.5:c.3927_3928inv']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000005.10:g.112839521_112839522inv', 'vcf': {'chr': '5', 'ref': 'AA', 'pos': '112839521', 'alt': 'TT'}}


    def test_variant119(self):
        variant = 'NM_001034853.1:c.2847_2848delAGinsCT'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_001034853.1:c.2847_2848inv' in results.keys()
        assert results['NM_001034853.1:c.2847_2848inv']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001034853.1:c.2847_2848inv']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001034853.1:c.2847_2848inv']['alt_genomic_loci'] == []
        assert results['NM_001034853.1:c.2847_2848inv']['gene_symbol'] == 'RPGR'
        assert results['NM_001034853.1:c.2847_2848inv']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001030025.1:p.(Glu949_Glu950delinsAspTer)', 'slr': 'NP_001030025.1:p.(E949_E950delinsD*)'}
        assert results['NM_001034853.1:c.2847_2848inv']['submitted_variant'] == 'NM_001034853.1:c.2847_2848delAGinsCT'
        assert results['NM_001034853.1:c.2847_2848inv']['genome_context_intronic_sequence'] == ''
        assert results['NM_001034853.1:c.2847_2848inv']['hgvs_lrg_variant'] == ''
        assert results['NM_001034853.1:c.2847_2848inv']['hgvs_transcript_variant'] == 'NM_001034853.1:c.2847_2848inv'
        assert results['NM_001034853.1:c.2847_2848inv']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001034853.1:c.2847_2848inv']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.38145404_38145405inv', 'vcf': {'chr': 'chrX', 'ref': 'CT', 'pos': '38145404', 'alt': 'AG'}}
        assert results['NM_001034853.1:c.2847_2848inv']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.38286151_38286152inv', 'vcf': {'chr': 'chrX', 'ref': 'CT', 'pos': '38286151', 'alt': 'AG'}}
        assert results['NM_001034853.1:c.2847_2848inv']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.38145404_38145405inv', 'vcf': {'chr': 'X', 'ref': 'CT', 'pos': '38145404', 'alt': 'AG'}}
        assert results['NM_001034853.1:c.2847_2848inv']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.38286151_38286152inv', 'vcf': {'chr': 'X', 'ref': 'CT', 'pos': '38286151', 'alt': 'AG'}}


    def test_variant120(self):
        variant = 'NM_000088.3:c.4392_*2inv'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_000088.3:c.4394_4395inv' in results.keys()
        assert results['NM_000088.3:c.4394_4395inv']['hgvs_lrg_transcript_variant'] == 'LRG_1t1:c.4394_4395inv'
        assert results['NM_000088.3:c.4394_4395inv']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000088.3:c.4394_4395inv']['alt_genomic_loci'] == []
        assert results['NM_000088.3:c.4394_4395inv']['gene_symbol'] == 'COL1A1'
        assert results['NM_000088.3:c.4394_4395inv']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000079.2(LRG_1p1):p.(Ter1465PheextTer27)', 'slr': 'NP_000079.2:p.(*1465Fext*27)'}
        assert results['NM_000088.3:c.4394_4395inv']['submitted_variant'] == 'NM_000088.3:c.4392_*2inv'
        assert results['NM_000088.3:c.4394_4395inv']['genome_context_intronic_sequence'] == ''
        assert results['NM_000088.3:c.4394_4395inv']['hgvs_lrg_variant'] == 'LRG_1:g.21137_21138inv'
        assert results['NM_000088.3:c.4394_4395inv']['hgvs_transcript_variant'] == 'NM_000088.3:c.4394_4395inv'
        assert results['NM_000088.3:c.4394_4395inv']['hgvs_refseqgene_variant'] == 'NG_007400.1:g.21137_21138inv'
        assert results['NM_000088.3:c.4394_4395inv']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48262863_48262864inv', 'vcf': {'chr': 'chr17', 'ref': 'TT', 'pos': '48262863', 'alt': 'AA'}}
        assert results['NM_000088.3:c.4394_4395inv']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50185502_50185503inv', 'vcf': {'chr': 'chr17', 'ref': 'TT', 'pos': '50185502', 'alt': 'AA'}}
        assert results['NM_000088.3:c.4394_4395inv']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48262863_48262864inv', 'vcf': {'chr': '17', 'ref': 'TT', 'pos': '48262863', 'alt': 'AA'}}
        assert results['NM_000088.3:c.4394_4395inv']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50185502_50185503inv', 'vcf': {'chr': '17', 'ref': 'TT', 'pos': '50185502', 'alt': 'AA'}}

        assert results['flag'] == 'gene_variant'

    def test_variant121(self):
        variant = 'NM_000088.3:c.4392_*5inv'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000088.3:c.4392_*5inv' in results.keys()
        assert results['NM_000088.3:c.4392_*5inv']['hgvs_lrg_transcript_variant'] == 'LRG_1t1:c.4393_*4inv'
        assert results['NM_000088.3:c.4392_*5inv']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000088.3:c.4392_*5inv']['alt_genomic_loci'] == []
        assert results['NM_000088.3:c.4392_*5inv']['gene_symbol'] == 'COL1A1'
        assert results['NM_000088.3:c.4392_*5inv']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000079.2(LRG_1p1):p.?', 'slr': 'NP_000079.2:p.?'}
        assert results['NM_000088.3:c.4392_*5inv']['submitted_variant'] == 'NM_000088.3:c.4392_*5inv'
        assert results['NM_000088.3:c.4392_*5inv']['genome_context_intronic_sequence'] == ''
        assert results['NM_000088.3:c.4392_*5inv']['hgvs_lrg_variant'] == 'LRG_1:g.21136_21142inv'
        assert results['NM_000088.3:c.4392_*5inv']['hgvs_transcript_variant'] == 'NM_000088.3:c.4392_*5inv'
        assert results['NM_000088.3:c.4392_*5inv']['hgvs_refseqgene_variant'] == 'NG_007400.1:g.21136_21142inv'
        assert results['NM_000088.3:c.4392_*5inv']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48262858_48262866inv', 'vcf': {'chr': 'chr17', 'ref': 'GAGTTTA', 'pos': '48262859', 'alt': 'TAAACTC'}}
        assert results['NM_000088.3:c.4392_*5inv']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50185497_50185505inv', 'vcf': {'chr': 'chr17', 'ref': 'GAGTTTA', 'pos': '50185498', 'alt': 'TAAACTC'}}
        assert results['NM_000088.3:c.4392_*5inv']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48262858_48262866inv', 'vcf': {'chr': '17', 'ref': 'GAGTTTA', 'pos': '48262859', 'alt': 'TAAACTC'}}
        assert results['NM_000088.3:c.4392_*5inv']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50185497_50185505inv', 'vcf': {'chr': '17', 'ref': 'GAGTTTA', 'pos': '50185498', 'alt': 'TAAACTC'}}


    def test_variant122(self):
        variant = 'NM_000088.3:c.4390_*7inv'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000088.3:c.4390_*7inv' in results.keys()
        assert results['NM_000088.3:c.4390_*7inv']['hgvs_lrg_transcript_variant'] == 'LRG_1t1:c.4390_*7inv'
        assert results['NM_000088.3:c.4390_*7inv']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000088.3:c.4390_*7inv']['alt_genomic_loci'] == []
        assert results['NM_000088.3:c.4390_*7inv']['gene_symbol'] == 'COL1A1'
        assert results['NM_000088.3:c.4390_*7inv']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000079.2(LRG_1p1):p.?', 'slr': 'NP_000079.2:p.?'}
        assert results['NM_000088.3:c.4390_*7inv']['submitted_variant'] == 'NM_000088.3:c.4390_*7inv'
        assert results['NM_000088.3:c.4390_*7inv']['genome_context_intronic_sequence'] == ''
        assert results['NM_000088.3:c.4390_*7inv']['hgvs_lrg_variant'] == 'LRG_1:g.21133_21145inv'
        assert results['NM_000088.3:c.4390_*7inv']['hgvs_transcript_variant'] == 'NM_000088.3:c.4390_*7inv'
        assert results['NM_000088.3:c.4390_*7inv']['hgvs_refseqgene_variant'] == 'NG_007400.1:g.21133_21145inv'
        assert results['NM_000088.3:c.4390_*7inv']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48262856_48262868inv', 'vcf': {'chr': 'chr17', 'ref': 'AGGGAGTTTACAG', 'pos': '48262856', 'alt': 'CTGTAAACTCCCT'}}
        assert results['NM_000088.3:c.4390_*7inv']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50185495_50185507inv', 'vcf': {'chr': 'chr17', 'ref': 'AGGGAGTTTACAG', 'pos': '50185495', 'alt': 'CTGTAAACTCCCT'}}
        assert results['NM_000088.3:c.4390_*7inv']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48262856_48262868inv', 'vcf': {'chr': '17', 'ref': 'AGGGAGTTTACAG', 'pos': '48262856', 'alt': 'CTGTAAACTCCCT'}}
        assert results['NM_000088.3:c.4390_*7inv']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50185495_50185507inv', 'vcf': {'chr': '17', 'ref': 'AGGGAGTTTACAG', 'pos': '50185495', 'alt': 'CTGTAAACTCCCT'}}


    def test_variant123(self):
        variant = 'NM_005732.3:c.2923-5insT'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'validation_warning_1' in results.keys()
        assert results['validation_warning_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['validation_warning_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['alt_genomic_loci'] == []
        assert results['validation_warning_1']['gene_symbol'] == ''
        assert results['validation_warning_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['validation_warning_1']['submitted_variant'] == 'NM_005732.3:c.2923-5insT'
        assert results['validation_warning_1']['genome_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['hgvs_lrg_variant'] == ''
        assert results['validation_warning_1']['hgvs_transcript_variant'] == ''
        assert results['validation_warning_1']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['validation_warning_1']['primary_assembly_loci'].keys()

        assert results['flag'] == 'warning'

    def test_variant124(self):
        variant = 'NM_198283.1(EYS):c.*743120C>T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'validation_warning_1' in results.keys()
        assert results['validation_warning_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['validation_warning_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['alt_genomic_loci'] == []
        assert results['validation_warning_1']['gene_symbol'] == ''
        assert results['validation_warning_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['validation_warning_1']['submitted_variant'] == 'NM_198283.1(EYS):c.*743120C>T'
        assert results['validation_warning_1']['genome_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['hgvs_lrg_variant'] == ''
        assert results['validation_warning_1']['hgvs_transcript_variant'] == ''
        assert results['validation_warning_1']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['validation_warning_1']['primary_assembly_loci'].keys()

        assert results['flag'] == 'warning'

    def test_variant125(self):
        variant = 'NM_133379.4(TTN):c.*265+26591C>T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'validation_warning_1' in results.keys()
        assert results['validation_warning_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['validation_warning_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['alt_genomic_loci'] == []
        assert results['validation_warning_1']['gene_symbol'] == ''
        assert results['validation_warning_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['validation_warning_1']['submitted_variant'] == 'NM_133379.4(TTN):c.*265+26591C>T'
        assert results['validation_warning_1']['genome_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['hgvs_lrg_variant'] == ''
        assert results['validation_warning_1']['hgvs_transcript_variant'] == ''
        assert results['validation_warning_1']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['validation_warning_1']['primary_assembly_loci'].keys()

        assert results['flag'] == 'warning'

    def test_variant126(self):
        variant = 'NM_000088.3:c.589-2_589-1AG>G'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000088.3:c.589-2_589-1delinsG' in results.keys()
        assert results['NM_000088.3:c.589-2_589-1delinsG']['hgvs_lrg_transcript_variant'] == 'LRG_1t1:c.589-2del'
        assert results['NM_000088.3:c.589-2_589-1delinsG']['refseqgene_context_intronic_sequence'] == 'NG_007400.1(NM_000088.3):c.589-2del'
        assert results['NM_000088.3:c.589-2_589-1delinsG']['alt_genomic_loci'] == []
        assert results['NM_000088.3:c.589-2_589-1delinsG']['gene_symbol'] == 'COL1A1'
        assert results['NM_000088.3:c.589-2_589-1delinsG']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000079.2(LRG_1p1):p.?', 'slr': 'NP_000079.2:p.?'}
        assert results['NM_000088.3:c.589-2_589-1delinsG']['submitted_variant'] == 'NM_000088.3:c.589-2_589-1AG>G'
        assert results['NM_000088.3:c.589-2_589-1delinsG']['genome_context_intronic_sequence'] == 'NC_000017.10(NM_000088.3):c.589-2_589-1delinsG'
        assert results['NM_000088.3:c.589-2_589-1delinsG']['hgvs_lrg_variant'] == 'LRG_1:g.8636del'
        assert results['NM_000088.3:c.589-2_589-1delinsG']['hgvs_transcript_variant'] == 'NM_000088.3:c.589-2_589-1delinsG'
        assert results['NM_000088.3:c.589-2_589-1delinsG']['hgvs_refseqgene_variant'] == 'NG_007400.1:g.8636del'
        assert results['NM_000088.3:c.589-2_589-1delinsG']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275364_48275365delinsC', 'vcf': {'chr': 'chr17', 'ref': 'CT', 'pos': '48275364', 'alt': 'C'}}
        assert results['NM_000088.3:c.589-2_589-1delinsG']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198003_50198004delinsC', 'vcf': {'chr': 'chr17', 'ref': 'CT', 'pos': '50198003', 'alt': 'C'}}
        assert results['NM_000088.3:c.589-2_589-1delinsG']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275364_48275365delinsC', 'vcf': {'chr': '17', 'ref': 'CT', 'pos': '48275364', 'alt': 'C'}}
        assert results['NM_000088.3:c.589-2_589-1delinsG']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198003_50198004delinsC', 'vcf': {'chr': '17', 'ref': 'CT', 'pos': '50198003', 'alt': 'C'}}


    def test_variant127(self):
        variant = 'NM_000088.3:c.642+1_642+2delGTinsG'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000088.3:c.642+1_642+2delinsG' in results.keys()
        assert results['NM_000088.3:c.642+1_642+2delinsG']['hgvs_lrg_transcript_variant'] == 'LRG_1t1:c.642+2del'
        assert results['NM_000088.3:c.642+1_642+2delinsG']['refseqgene_context_intronic_sequence'] == 'NG_007400.1(NM_000088.3):c.642+2del'
        assert results['NM_000088.3:c.642+1_642+2delinsG']['alt_genomic_loci'] == []
        assert results['NM_000088.3:c.642+1_642+2delinsG']['gene_symbol'] == 'COL1A1'
        assert results['NM_000088.3:c.642+1_642+2delinsG']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000079.2(LRG_1p1):p.?', 'slr': 'NP_000079.2:p.?'}
        assert results['NM_000088.3:c.642+1_642+2delinsG']['submitted_variant'] == 'NM_000088.3:c.642+1_642+2delGTinsG'
        assert results['NM_000088.3:c.642+1_642+2delinsG']['genome_context_intronic_sequence'] == 'NC_000017.10(NM_000088.3):c.642+1_642+2delinsG'
        assert results['NM_000088.3:c.642+1_642+2delinsG']['hgvs_lrg_variant'] == 'LRG_1:g.8693del'
        assert results['NM_000088.3:c.642+1_642+2delinsG']['hgvs_transcript_variant'] == 'NM_000088.3:c.642+1_642+2delinsG'
        assert results['NM_000088.3:c.642+1_642+2delinsG']['hgvs_refseqgene_variant'] == 'NG_007400.1:g.8693del'
        assert results['NM_000088.3:c.642+1_642+2delinsG']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275308_48275309delinsC', 'vcf': {'chr': 'chr17', 'ref': 'TA', 'pos': '48275307', 'alt': 'T'}}
        assert results['NM_000088.3:c.642+1_642+2delinsG']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50197947_50197948delinsC', 'vcf': {'chr': 'chr17', 'ref': 'TA', 'pos': '50197946', 'alt': 'T'}}
        assert results['NM_000088.3:c.642+1_642+2delinsG']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275308_48275309delinsC', 'vcf': {'chr': '17', 'ref': 'TA', 'pos': '48275307', 'alt': 'T'}}
        assert results['NM_000088.3:c.642+1_642+2delinsG']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50197947_50197948delinsC', 'vcf': {'chr': '17', 'ref': 'TA', 'pos': '50197946', 'alt': 'T'}}


    def test_variant128(self):
        variant = 'NM_004415.3:c.1-1insA'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'validation_warning_1' in results.keys()
        assert results['validation_warning_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['validation_warning_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['alt_genomic_loci'] == []
        assert results['validation_warning_1']['gene_symbol'] == ''
        assert results['validation_warning_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['validation_warning_1']['submitted_variant'] == 'NM_004415.3:c.1-1insA'
        assert results['validation_warning_1']['genome_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['hgvs_lrg_variant'] == ''
        assert results['validation_warning_1']['hgvs_transcript_variant'] == ''
        assert results['validation_warning_1']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['validation_warning_1']['primary_assembly_loci'].keys()

        assert results['flag'] == 'warning'

    def test_variant129(self):
        variant = 'NM_004415.3:c.-1_1insA'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_004415.3:c.-1_1insA' in results.keys()
        assert results['NM_004415.3:c.-1_1insA']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_004415.3:c.-1_1insA']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_004415.3:c.-1_1insA']['alt_genomic_loci'] == []
        assert results['NM_004415.3:c.-1_1insA']['gene_symbol'] == 'DSP'
        assert results['NM_004415.3:c.-1_1insA']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_004406.2(LRG_423p1):p.(Met1?)', 'slr': 'NP_004406.2:p.(M1?)'}
        assert results['NM_004415.3:c.-1_1insA']['submitted_variant'] == 'NM_004415.3:c.-1_1insA'
        assert results['NM_004415.3:c.-1_1insA']['genome_context_intronic_sequence'] == ''
        assert results['NM_004415.3:c.-1_1insA']['hgvs_lrg_variant'] == ''
        assert results['NM_004415.3:c.-1_1insA']['hgvs_transcript_variant'] == 'NM_004415.3:c.-1_1insA'
        assert results['NM_004415.3:c.-1_1insA']['hgvs_refseqgene_variant'] == ''
        assert results['NM_004415.3:c.-1_1insA']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000006.11:g.7542148_7542149insA', 'vcf': {'chr': 'chr6', 'ref': 'A', 'pos': '7542149', 'alt': 'AA'}}
        assert results['NM_004415.3:c.-1_1insA']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000006.12:g.7541915_7541916insA', 'vcf': {'chr': 'chr6', 'ref': 'A', 'pos': '7541916', 'alt': 'AA'}}
        assert results['NM_004415.3:c.-1_1insA']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000006.11:g.7542148_7542149insA', 'vcf': {'chr': '6', 'ref': 'A', 'pos': '7542149', 'alt': 'AA'}}
        assert results['NM_004415.3:c.-1_1insA']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000006.12:g.7541915_7541916insA', 'vcf': {'chr': '6', 'ref': 'A', 'pos': '7541916', 'alt': 'AA'}}


    def test_variant130(self):
        variant = 'NM_000273.2:c.1-5028_253del'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'validation_warning_1' in results.keys()
        assert results['validation_warning_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['validation_warning_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['alt_genomic_loci'] == []
        assert results['validation_warning_1']['gene_symbol'] == ''
        assert results['validation_warning_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['validation_warning_1']['submitted_variant'] == 'NM_000273.2:c.1-5028_253del'
        assert results['validation_warning_1']['genome_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['hgvs_lrg_variant'] == ''
        assert results['validation_warning_1']['hgvs_transcript_variant'] == ''
        assert results['validation_warning_1']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['validation_warning_1']['primary_assembly_loci'].keys()

        assert results['flag'] == 'warning'

    def test_variant131(self):
        variant = 'NM_002929.2:c.1006C>T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_002929.2:c.1006C>T' in results.keys()
        assert results['NM_002929.2:c.1006C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_002929.2:c.1006C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_002929.2:c.1006C>T']['alt_genomic_loci'] == []
        assert results['NM_002929.2:c.1006C>T']['gene_symbol'] == 'GRK1'
        assert results['NM_002929.2:c.1006C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_002920.1:p.(Leu336Phe)', 'slr': 'NP_002920.1:p.(L336F)'}
        assert results['NM_002929.2:c.1006C>T']['submitted_variant'] == 'NM_002929.2:c.1006C>T'
        assert results['NM_002929.2:c.1006C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_002929.2:c.1006C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_002929.2:c.1006C>T']['hgvs_transcript_variant'] == 'NM_002929.2:c.1006C>T'
        assert results['NM_002929.2:c.1006C>T']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['NM_002929.2:c.1006C>T']['primary_assembly_loci'].keys()
        assert results['NM_002929.2:c.1006C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000013.11:g.113723094C>T', 'vcf': {'chr': 'chr13', 'ref': u'C', 'pos': '113723094', 'alt': u'T'}}
        assert 'grch37' not in results['NM_002929.2:c.1006C>T']['primary_assembly_loci'].keys()
        assert results['NM_002929.2:c.1006C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000013.11:g.113723094C>T', 'vcf': {'chr': '13', 'ref': u'C', 'pos': '113723094', 'alt': u'T'}}


    def test_variant132(self):
        variant = 'NR_125367.1:n.167+18165G>A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NR_125367.1:n.167+18165G>A' in results.keys()
        assert results['NR_125367.1:n.167+18165G>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_125367.1:n.167+18165G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_125367.1:n.167+18165G>A']['alt_genomic_loci'] == []
        assert results['NR_125367.1:n.167+18165G>A']['gene_symbol'] == 'MYHAS'
        assert results['NR_125367.1:n.167+18165G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_125367.1:n.167+18165G>A']['submitted_variant'] == 'NR_125367.1:n.167+18165G>A'
        assert results['NR_125367.1:n.167+18165G>A']['genome_context_intronic_sequence'] == 'NC_000017.10(NR_125367.1):c.167+18165G>A'
        assert results['NR_125367.1:n.167+18165G>A']['hgvs_lrg_variant'] == ''
        assert results['NR_125367.1:n.167+18165G>A']['hgvs_transcript_variant'] == 'NR_125367.1:n.167+18165G>A'
        assert results['NR_125367.1:n.167+18165G>A']['hgvs_refseqgene_variant'] == ''
        assert results['NR_125367.1:n.167+18165G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.10327720G>A', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '10327720', 'alt': 'A'}}
        assert results['NR_125367.1:n.167+18165G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.10424403G>A', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '10424403', 'alt': 'A'}}
        assert results['NR_125367.1:n.167+18165G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.10327720G>A', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '10327720', 'alt': 'A'}}
        assert results['NR_125367.1:n.167+18165G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.10424403G>A', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '10424403', 'alt': 'A'}}


    def test_variant133(self):
        variant = 'NM_006005.3:c.3071_3073delinsTTA'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'validation_warning_1' in results.keys()
        assert results['validation_warning_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['validation_warning_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['alt_genomic_loci'] == []
        assert results['validation_warning_1']['gene_symbol'] == ''
        assert results['validation_warning_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['validation_warning_1']['submitted_variant'] == 'NM_006005.3:c.3071_3073delinsTTA'
        assert results['validation_warning_1']['genome_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['hgvs_lrg_variant'] == ''
        assert results['validation_warning_1']['hgvs_transcript_variant'] == ''
        assert results['validation_warning_1']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['validation_warning_1']['primary_assembly_loci'].keys()

        assert results['flag'] == 'warning'

    def test_variant134(self):
        variant = 'NM_000089.3:n.1504_1506del'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'validation_warning_1' in results.keys()
        assert results['validation_warning_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['validation_warning_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['alt_genomic_loci'] == []
        assert results['validation_warning_1']['gene_symbol'] == ''
        assert results['validation_warning_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['validation_warning_1']['submitted_variant'] == 'NM_000089.3:n.1504_1506del'
        assert results['validation_warning_1']['genome_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['hgvs_lrg_variant'] == ''
        assert results['validation_warning_1']['hgvs_transcript_variant'] == ''
        assert results['validation_warning_1']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['validation_warning_1']['primary_assembly_loci'].keys()

        assert results['flag'] == 'warning'

    def test_variant135(self):
        variant = 'NC_012920.1:m.1011C>T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'validation_warning_1' in results.keys()
        assert results['validation_warning_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['validation_warning_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['alt_genomic_loci'] == []
        assert results['validation_warning_1']['gene_symbol'] == ''
        assert results['validation_warning_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['validation_warning_1']['submitted_variant'] == 'NC_012920.1:m.1011C>T'
        assert results['validation_warning_1']['genome_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['hgvs_lrg_variant'] == ''
        assert results['validation_warning_1']['hgvs_transcript_variant'] == ''
        assert results['validation_warning_1']['hgvs_refseqgene_variant'] == ''
        assert results['validation_warning_1']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_012920.1:m.1011C>T', 'vcf': {'chr': 'chrM', 'ref': 'C', 'pos': '1011', 'alt': 'T'}}
        assert results['validation_warning_1']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_012920.1:m.1011C>T', 'vcf': {'chr': 'chrM', 'ref': 'C', 'pos': '1011', 'alt': 'T'}}
        assert results['validation_warning_1']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_012920.1:m.1011C>T', 'vcf': {'chr': 'M', 'ref': 'C', 'pos': '1011', 'alt': 'T'}}
        assert results['validation_warning_1']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_012920.1:m.1011C>T', 'vcf': {'chr': 'M', 'ref': 'C', 'pos': '1011', 'alt': 'T'}}

        assert results['flag'] == 'warning'

    def test_variant136(self):
        variant = 'NC_000006.11:g.90403795G='
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_014611.1:c.9879T>C' in results.keys()
        assert results['NM_014611.1:c.9879T>C']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_014611.1:c.9879T>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_014611.1:c.9879T>C']['alt_genomic_loci'] == []
        assert results['NM_014611.1:c.9879T>C']['gene_symbol'] == 'MDN1'
        assert results['NM_014611.1:c.9879T>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_055426.1:p.(Val3293=)', 'slr': 'NP_055426.1:p.(V3293=)'}
        assert results['NM_014611.1:c.9879T>C']['submitted_variant'] == 'NC_000006.11:g.90403795G='
        assert results['NM_014611.1:c.9879T>C']['genome_context_intronic_sequence'] == ''
        assert results['NM_014611.1:c.9879T>C']['hgvs_lrg_variant'] == ''
        assert results['NM_014611.1:c.9879T>C']['hgvs_transcript_variant'] == 'NM_014611.1:c.9879T>C'
        assert results['NM_014611.1:c.9879T>C']['hgvs_refseqgene_variant'] == ''
        assert results['NM_014611.1:c.9879T>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000006.11:g.90403795G=', 'vcf': {'chr': 'chr6', 'ref': 'G', 'pos': '90403795', 'alt': 'G'}}
        assert 'hg38' not in results['NM_014611.1:c.9879T>C']['primary_assembly_loci'].keys()
        assert results['NM_014611.1:c.9879T>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000006.11:g.90403795G=', 'vcf': {'chr': '6', 'ref': 'G', 'pos': '90403795', 'alt': 'G'}}
        assert 'grch38' not in results['NM_014611.1:c.9879T>C']['primary_assembly_loci'].keys()

        assert 'NM_014611.2:c.9879C=' in results.keys()
        assert results['NM_014611.2:c.9879C=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_014611.2:c.9879C=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_014611.2:c.9879C=']['alt_genomic_loci'] == []
        assert results['NM_014611.2:c.9879C=']['gene_symbol'] == 'MDN1'
        assert results['NM_014611.2:c.9879C=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_055426.1:p.(Val3293=)', 'slr': 'NP_055426.1:p.(V3293=)'}
        assert results['NM_014611.2:c.9879C=']['submitted_variant'] == 'NC_000006.11:g.90403795G='
        assert results['NM_014611.2:c.9879C=']['genome_context_intronic_sequence'] == ''
        assert results['NM_014611.2:c.9879C=']['hgvs_lrg_variant'] == ''
        assert results['NM_014611.2:c.9879C=']['hgvs_transcript_variant'] == 'NM_014611.2:c.9879C='
        assert results['NM_014611.2:c.9879C=']['hgvs_refseqgene_variant'] == ''
        assert results['NM_014611.2:c.9879C=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000006.11:g.90403795G=', 'vcf': {'chr': 'chr6', 'ref': u'G', 'pos': '90403795', 'alt': u'G'}}
        assert results['NM_014611.2:c.9879C=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000006.12:g.89694076G=', 'vcf': {'chr': 'chr6', 'ref': u'G', 'pos': '89694076', 'alt': u'G'}}
        assert results['NM_014611.2:c.9879C=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000006.11:g.90403795G=', 'vcf': {'chr': '6', 'ref': u'G', 'pos': '90403795', 'alt': u'G'}}
        assert results['NM_014611.2:c.9879C=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000006.12:g.89694076G=', 'vcf': {'chr': '6', 'ref': u'G', 'pos': '89694076', 'alt': u'G'}}


    def test_variant137(self):
        variant = '1-169519049-T-.'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_000130.4:c.1602del' in results.keys()
        assert results['NM_000130.4:c.1602del']['hgvs_lrg_transcript_variant'] == 'LRG_553t1:c.1601del'
        assert results['NM_000130.4:c.1602del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000130.4:c.1602del']['alt_genomic_loci'] == []
        assert results['NM_000130.4:c.1602del']['gene_symbol'] == 'F5'
        assert results['NM_000130.4:c.1602del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000121.2(LRG_553p1):p.(Arg534GlnfsTer40)', 'slr': 'NP_000121.2:p.(R534Qfs*40)'}
        assert results['NM_000130.4:c.1602del']['submitted_variant'] == '1-169519049-T-.'
        assert results['NM_000130.4:c.1602del']['genome_context_intronic_sequence'] == ''
        assert results['NM_000130.4:c.1602del']['hgvs_lrg_variant'] == 'LRG_553:g.41721del'
        assert results['NM_000130.4:c.1602del']['hgvs_transcript_variant'] == 'NM_000130.4:c.1602del'
        assert results['NM_000130.4:c.1602del']['hgvs_refseqgene_variant'] == 'NG_011806.1:g.41721del'
        assert results['NM_000130.4:c.1602del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.169519048del', 'vcf': {'chr': 'chr1', 'ref': 'CT', 'pos': '169519047', 'alt': 'C'}}
        assert results['NM_000130.4:c.1602del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.169549810del', 'vcf': {'chr': 'chr1', 'ref': 'CT', 'pos': '169549809', 'alt': 'C'}}
        assert results['NM_000130.4:c.1602del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.169519048del', 'vcf': {'chr': '1', 'ref': 'CT', 'pos': '169519047', 'alt': 'C'}}
        assert results['NM_000130.4:c.1602del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.169549810del', 'vcf': {'chr': '1', 'ref': 'CT', 'pos': '169549809', 'alt': 'C'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_000130.4:c.1601G>A' in results.keys()
        assert results['NM_000130.4:c.1601G>A']['hgvs_lrg_transcript_variant'] == 'LRG_553t1:c.1601G>A'
        assert results['NM_000130.4:c.1601G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000130.4:c.1601G>A']['alt_genomic_loci'] == []
        assert results['NM_000130.4:c.1601G>A']['gene_symbol'] == 'F5'
        assert results['NM_000130.4:c.1601G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000121.2(LRG_553p1):p.(Arg534Gln)', 'slr': 'NP_000121.2:p.(R534Q)'}
        assert results['NM_000130.4:c.1601G>A']['submitted_variant'] == '1-169519049-T-.'
        assert results['NM_000130.4:c.1601G>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_000130.4:c.1601G>A']['hgvs_lrg_variant'] == 'LRG_553:g.41721G>A'
        assert results['NM_000130.4:c.1601G>A']['hgvs_transcript_variant'] == 'NM_000130.4:c.1601G>A'
        assert results['NM_000130.4:c.1601G>A']['hgvs_refseqgene_variant'] == 'NG_011806.1:g.41721G>A'
        assert results['NM_000130.4:c.1601G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.169519049T=', 'vcf': {'chr': 'chr1', 'ref': 'T', 'pos': '169519049', 'alt': 'T'}}
        assert results['NM_000130.4:c.1601G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.169549811C>T', 'vcf': {'chr': 'chr1', 'ref': u'C', 'pos': '169549811', 'alt': u'T'}}
        assert results['NM_000130.4:c.1601G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.169519049T=', 'vcf': {'chr': '1', 'ref': 'T', 'pos': '169519049', 'alt': 'T'}}
        assert results['NM_000130.4:c.1601G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.169549811C>T', 'vcf': {'chr': '1', 'ref': u'C', 'pos': '169549811', 'alt': u'T'}}


    def test_variant138(self):
        variant = 'NC_000005.9:g.35058667_35058668AG='
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001204317.1:c.856-9155_856-9154=' in results.keys()
        assert results['NM_001204317.1:c.856-9155_856-9154=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001204317.1:c.856-9155_856-9154=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001204317.1:c.856-9155_856-9154=']['alt_genomic_loci'] == []
        assert results['NM_001204317.1:c.856-9155_856-9154=']['gene_symbol'] == 'PRLR'
        assert results['NM_001204317.1:c.856-9155_856-9154=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001191246.1:p.?', 'slr': 'NP_001191246.1:p.?'}
        assert results['NM_001204317.1:c.856-9155_856-9154=']['submitted_variant'] == 'NC_000005.9:g.35058667_35058668AG='
        assert results['NM_001204317.1:c.856-9155_856-9154=']['genome_context_intronic_sequence'] == 'NC_000005.9(NM_001204317.1):c.856-9155_856-9154='
        assert results['NM_001204317.1:c.856-9155_856-9154=']['hgvs_lrg_variant'] == ''
        assert results['NM_001204317.1:c.856-9155_856-9154=']['hgvs_transcript_variant'] == 'NM_001204317.1:c.856-9155_856-9154='
        assert results['NM_001204317.1:c.856-9155_856-9154=']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001204317.1:c.856-9155_856-9154=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.35058667_35058668=', 'vcf': {'chr': 'chr5', 'ref': 'AG', 'pos': '35058667', 'alt': 'AG'}}
        assert results['NM_001204317.1:c.856-9155_856-9154=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000005.10:g.35058562_35058563=', 'vcf': {'chr': 'chr5', 'ref': 'AA', 'pos': '35058562', 'alt': 'AA'}}
        assert results['NM_001204317.1:c.856-9155_856-9154=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.35058667_35058668=', 'vcf': {'chr': '5', 'ref': 'AG', 'pos': '35058667', 'alt': 'AG'}}
        assert results['NM_001204317.1:c.856-9155_856-9154=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000005.10:g.35058562_35058563=', 'vcf': {'chr': '5', 'ref': 'AA', 'pos': '35058562', 'alt': 'AA'}}

        assert 'NM_001204316.1:c.1009+7383_1009+7384=' in results.keys()
        assert results['NM_001204316.1:c.1009+7383_1009+7384=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001204316.1:c.1009+7383_1009+7384=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001204316.1:c.1009+7383_1009+7384=']['alt_genomic_loci'] == []
        assert results['NM_001204316.1:c.1009+7383_1009+7384=']['gene_symbol'] == 'PRLR'
        assert results['NM_001204316.1:c.1009+7383_1009+7384=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001191245.1:p.?', 'slr': 'NP_001191245.1:p.?'}
        assert results['NM_001204316.1:c.1009+7383_1009+7384=']['submitted_variant'] == 'NC_000005.9:g.35058667_35058668AG='
        assert results['NM_001204316.1:c.1009+7383_1009+7384=']['genome_context_intronic_sequence'] == 'NC_000005.9(NM_001204316.1):c.1009+7383_1009+7384='
        assert results['NM_001204316.1:c.1009+7383_1009+7384=']['hgvs_lrg_variant'] == ''
        assert results['NM_001204316.1:c.1009+7383_1009+7384=']['hgvs_transcript_variant'] == 'NM_001204316.1:c.1009+7383_1009+7384='
        assert results['NM_001204316.1:c.1009+7383_1009+7384=']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001204316.1:c.1009+7383_1009+7384=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.35058667_35058668=', 'vcf': {'chr': 'chr5', 'ref': 'AG', 'pos': '35058667', 'alt': 'AG'}}
        assert results['NM_001204316.1:c.1009+7383_1009+7384=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000005.10:g.35058565_35058566=', 'vcf': {'chr': 'chr5', 'ref': 'AT', 'pos': '35058565', 'alt': 'AT'}}
        assert results['NM_001204316.1:c.1009+7383_1009+7384=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.35058667_35058668=', 'vcf': {'chr': '5', 'ref': 'AG', 'pos': '35058667', 'alt': 'AG'}}
        assert results['NM_001204316.1:c.1009+7383_1009+7384=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000005.10:g.35058565_35058566=', 'vcf': {'chr': '5', 'ref': 'AT', 'pos': '35058565', 'alt': 'AT'}}

        assert 'NM_001204314.2:c.*6528del' in results.keys()
        assert results['NM_001204314.2:c.*6528del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001204314.2:c.*6528del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001204314.2:c.*6528del']['alt_genomic_loci'] == []
        assert results['NM_001204314.2:c.*6528del']['gene_symbol'] == 'PRLR'
        assert results['NM_001204314.2:c.*6528del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001191243.1:p.?', 'slr': 'NP_001191243.1:p.?'}
        assert results['NM_001204314.2:c.*6528del']['submitted_variant'] == 'NC_000005.9:g.35058667_35058668AG='
        assert results['NM_001204314.2:c.*6528del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001204314.2:c.*6528del']['hgvs_lrg_variant'] == ''
        assert results['NM_001204314.2:c.*6528del']['hgvs_transcript_variant'] == 'NM_001204314.2:c.*6528del'
        assert results['NM_001204314.2:c.*6528del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001204314.2:c.*6528del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.35058662_35058668=', 'vcf': {'chr': 'chr5', 'ref': 'AGACAAG', 'pos': '35058662', 'alt': 'AGACAAG'}}
        assert results['NM_001204314.2:c.*6528del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000005.10:g.35058563del', 'vcf': {'chr': 'chr5', 'ref': 'CA', 'pos': '35058560', 'alt': 'C'}}
        assert results['NM_001204314.2:c.*6528del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.35058662_35058668=', 'vcf': {'chr': '5', 'ref': 'AGACAAG', 'pos': '35058662', 'alt': 'AGACAAG'}}
        assert results['NM_001204314.2:c.*6528del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000005.10:g.35058563del', 'vcf': {'chr': '5', 'ref': 'CA', 'pos': '35058560', 'alt': 'C'}}

        assert 'NM_001204318.1:c.686-9155_686-9154=' in results.keys()
        assert results['NM_001204318.1:c.686-9155_686-9154=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001204318.1:c.686-9155_686-9154=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001204318.1:c.686-9155_686-9154=']['alt_genomic_loci'] == []
        assert results['NM_001204318.1:c.686-9155_686-9154=']['gene_symbol'] == 'PRLR'
        assert results['NM_001204318.1:c.686-9155_686-9154=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001191247.1:p.?', 'slr': 'NP_001191247.1:p.?'}
        assert results['NM_001204318.1:c.686-9155_686-9154=']['submitted_variant'] == 'NC_000005.9:g.35058667_35058668AG='
        assert results['NM_001204318.1:c.686-9155_686-9154=']['genome_context_intronic_sequence'] == 'NC_000005.9(NM_001204318.1):c.686-9155_686-9154='
        assert results['NM_001204318.1:c.686-9155_686-9154=']['hgvs_lrg_variant'] == ''
        assert results['NM_001204318.1:c.686-9155_686-9154=']['hgvs_transcript_variant'] == 'NM_001204318.1:c.686-9155_686-9154='
        assert results['NM_001204318.1:c.686-9155_686-9154=']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001204318.1:c.686-9155_686-9154=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.35058667_35058668=', 'vcf': {'chr': 'chr5', 'ref': 'AG', 'pos': '35058667', 'alt': 'AG'}}
        assert results['NM_001204318.1:c.686-9155_686-9154=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000005.10:g.35058562_35058563=', 'vcf': {'chr': 'chr5', 'ref': 'AA', 'pos': '35058562', 'alt': 'AA'}}
        assert results['NM_001204318.1:c.686-9155_686-9154=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.35058667_35058668=', 'vcf': {'chr': '5', 'ref': 'AG', 'pos': '35058667', 'alt': 'AG'}}
        assert results['NM_001204318.1:c.686-9155_686-9154=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000005.10:g.35058562_35058563=', 'vcf': {'chr': '5', 'ref': 'AA', 'pos': '35058562', 'alt': 'AA'}}

        assert 'NR_037910.1:n.828-9155_828-9154=' in results.keys()
        assert results['NR_037910.1:n.828-9155_828-9154=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_037910.1:n.828-9155_828-9154=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_037910.1:n.828-9155_828-9154=']['alt_genomic_loci'] == []
        assert results['NR_037910.1:n.828-9155_828-9154=']['gene_symbol'] == 'PRLR'
        assert results['NR_037910.1:n.828-9155_828-9154=']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_037910.1:n.828-9155_828-9154=']['submitted_variant'] == 'NC_000005.9:g.35058667_35058668AG='
        assert results['NR_037910.1:n.828-9155_828-9154=']['genome_context_intronic_sequence'] == 'NC_000005.9(NR_037910.1):c.828-9155_828-9154='
        assert results['NR_037910.1:n.828-9155_828-9154=']['hgvs_lrg_variant'] == ''
        assert results['NR_037910.1:n.828-9155_828-9154=']['hgvs_transcript_variant'] == 'NR_037910.1:n.828-9155_828-9154='
        assert results['NR_037910.1:n.828-9155_828-9154=']['hgvs_refseqgene_variant'] == ''
        assert results['NR_037910.1:n.828-9155_828-9154=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.35058667_35058668=', 'vcf': {'chr': 'chr5', 'ref': 'AG', 'pos': '35058667', 'alt': 'AG'}}
        assert results['NR_037910.1:n.828-9155_828-9154=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000005.10:g.35058562_35058563=', 'vcf': {'chr': 'chr5', 'ref': 'AA', 'pos': '35058562', 'alt': 'AA'}}
        assert results['NR_037910.1:n.828-9155_828-9154=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.35058667_35058668=', 'vcf': {'chr': '5', 'ref': 'AG', 'pos': '35058667', 'alt': 'AG'}}
        assert results['NR_037910.1:n.828-9155_828-9154=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000005.10:g.35058562_35058563=', 'vcf': {'chr': '5', 'ref': 'AA', 'pos': '35058562', 'alt': 'AA'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_000949.5:c.*6523_*6524=' in results.keys()
        assert results['NM_000949.5:c.*6523_*6524=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000949.5:c.*6523_*6524=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000949.5:c.*6523_*6524=']['alt_genomic_loci'] == []
        assert results['NM_000949.5:c.*6523_*6524=']['gene_symbol'] == 'PRLR'
        assert results['NM_000949.5:c.*6523_*6524=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000940.1:p.?', 'slr': 'NP_000940.1:p.?'}
        assert results['NM_000949.5:c.*6523_*6524=']['submitted_variant'] == 'NC_000005.9:g.35058667_35058668AG='
        assert results['NM_000949.5:c.*6523_*6524=']['genome_context_intronic_sequence'] == ''
        assert results['NM_000949.5:c.*6523_*6524=']['hgvs_lrg_variant'] == ''
        assert results['NM_000949.5:c.*6523_*6524=']['hgvs_transcript_variant'] == 'NM_000949.5:c.*6523_*6524='
        assert results['NM_000949.5:c.*6523_*6524=']['hgvs_refseqgene_variant'] == 'NG_029042.1:g.177156_177157='
        assert results['NM_000949.5:c.*6523_*6524=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.35058666_35058669=', 'vcf': {'chr': 'chr5', 'ref': u'AAGA', 'pos': '35058666', 'alt': u'AAGA'}}
        assert 'hg38' not in results['NM_000949.5:c.*6523_*6524=']['primary_assembly_loci'].keys()
        assert results['NM_000949.5:c.*6523_*6524=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.35058666_35058669=', 'vcf': {'chr': '5', 'ref': u'AAGA', 'pos': '35058666', 'alt': u'AAGA'}}
        assert 'grch38' not in results['NM_000949.5:c.*6523_*6524=']['primary_assembly_loci'].keys()

        assert 'NM_001204314.1:c.*6523_*6524=' in results.keys()
        assert results['NM_001204314.1:c.*6523_*6524=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001204314.1:c.*6523_*6524=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001204314.1:c.*6523_*6524=']['alt_genomic_loci'] == []
        assert results['NM_001204314.1:c.*6523_*6524=']['gene_symbol'] == 'PRLR'
        assert results['NM_001204314.1:c.*6523_*6524=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001191243.1:p.?', 'slr': 'NP_001191243.1:p.?'}
        assert results['NM_001204314.1:c.*6523_*6524=']['submitted_variant'] == 'NC_000005.9:g.35058667_35058668AG='
        assert results['NM_001204314.1:c.*6523_*6524=']['genome_context_intronic_sequence'] == ''
        assert results['NM_001204314.1:c.*6523_*6524=']['hgvs_lrg_variant'] == ''
        assert results['NM_001204314.1:c.*6523_*6524=']['hgvs_transcript_variant'] == 'NM_001204314.1:c.*6523_*6524='
        assert results['NM_001204314.1:c.*6523_*6524=']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001204314.1:c.*6523_*6524=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.35058667_35058668=', 'vcf': {'chr': 'chr5', 'ref': u'AG', 'pos': '35058667', 'alt': u'AG'}}
        assert 'hg38' not in results['NM_001204314.1:c.*6523_*6524=']['primary_assembly_loci'].keys()
        assert results['NM_001204314.1:c.*6523_*6524=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.35058667_35058668=', 'vcf': {'chr': '5', 'ref': u'AG', 'pos': '35058667', 'alt': u'AG'}}
        assert 'grch38' not in results['NM_001204314.1:c.*6523_*6524=']['primary_assembly_loci'].keys()

        assert 'NM_000949.6:c.*6528del' in results.keys()
        assert results['NM_000949.6:c.*6528del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000949.6:c.*6528del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000949.6:c.*6528del']['alt_genomic_loci'] == []
        assert results['NM_000949.6:c.*6528del']['gene_symbol'] == 'PRLR'
        assert results['NM_000949.6:c.*6528del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000940.1:p.?', 'slr': 'NP_000940.1:p.?'}
        assert results['NM_000949.6:c.*6528del']['submitted_variant'] == 'NC_000005.9:g.35058667_35058668AG='
        assert results['NM_000949.6:c.*6528del']['genome_context_intronic_sequence'] == ''
        assert results['NM_000949.6:c.*6528del']['hgvs_lrg_variant'] == ''
        assert results['NM_000949.6:c.*6528del']['hgvs_transcript_variant'] == 'NM_000949.6:c.*6528del'
        assert results['NM_000949.6:c.*6528del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_000949.6:c.*6528del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.35058662_35058668=', 'vcf': {'chr': 'chr5', 'ref': 'AGACAAG', 'pos': '35058662', 'alt': 'AGACAAG'}}
        assert results['NM_000949.6:c.*6528del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000005.10:g.35058563del', 'vcf': {'chr': 'chr5', 'ref': 'CA', 'pos': '35058560', 'alt': 'C'}}
        assert results['NM_000949.6:c.*6528del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.35058662_35058668=', 'vcf': {'chr': '5', 'ref': 'AGACAAG', 'pos': '35058662', 'alt': 'AGACAAG'}}
        assert results['NM_000949.6:c.*6528del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000005.10:g.35058563del', 'vcf': {'chr': '5', 'ref': 'CA', 'pos': '35058560', 'alt': 'C'}}


    def test_variant139(self):
        variant = 'NM_000251.1:c.1296_1348del'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'validation_warning_1' in results.keys()
        assert results['validation_warning_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['validation_warning_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['alt_genomic_loci'] == []
        assert results['validation_warning_1']['gene_symbol'] == ''
        assert results['validation_warning_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['validation_warning_1']['submitted_variant'] == 'NM_000251.1:c.1296_1348del'
        assert results['validation_warning_1']['genome_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['hgvs_lrg_variant'] == ''
        assert results['validation_warning_1']['hgvs_transcript_variant'] == ''
        assert results['validation_warning_1']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['validation_warning_1']['primary_assembly_loci'].keys()

        assert results['flag'] == 'warning'

    def test_variant140(self):
        variant = 'NM_000088.3:c.2023_2028del'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000088.3:c.2024_2028+1del' in results.keys()
        assert results['NM_000088.3:c.2024_2028+1del']['hgvs_lrg_transcript_variant'] == 'LRG_1t1:c.2024_2028+1del'
        assert results['NM_000088.3:c.2024_2028+1del']['refseqgene_context_intronic_sequence'] == 'NG_007400.1(NM_000088.3):c.2024_2028+1del'
        assert results['NM_000088.3:c.2024_2028+1del']['alt_genomic_loci'] == []
        assert results['NM_000088.3:c.2024_2028+1del']['gene_symbol'] == 'COL1A1'
        assert results['NM_000088.3:c.2024_2028+1del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000079.2(LRG_1p1):p.(Ala675_Arg676del)', 'slr': 'NP_000079.2:p.(A675_R676del)'}
        assert results['NM_000088.3:c.2024_2028+1del']['submitted_variant'] == 'NM_000088.3:c.2023_2028del'
        assert results['NM_000088.3:c.2024_2028+1del']['genome_context_intronic_sequence'] == 'NC_000017.10(NM_000088.3):c.2024_2028+1del'
        assert results['NM_000088.3:c.2024_2028+1del']['hgvs_lrg_variant'] == 'LRG_1:g.14656_14661del'
        assert results['NM_000088.3:c.2024_2028+1del']['hgvs_transcript_variant'] == 'NM_000088.3:c.2024_2028+1del'
        assert results['NM_000088.3:c.2024_2028+1del']['hgvs_refseqgene_variant'] == 'NG_007400.1:g.14656_14661del'
        assert results['NM_000088.3:c.2024_2028+1del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48269340_48269345del', 'vcf': {'chr': 'chr17', 'ref': 'ACTCTTG', 'pos': '48269339', 'alt': 'A'}}
        assert results['NM_000088.3:c.2024_2028+1del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50191979_50191984del', 'vcf': {'chr': 'chr17', 'ref': 'ACTCTTG', 'pos': '50191978', 'alt': 'A'}}
        assert results['NM_000088.3:c.2024_2028+1del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48269340_48269345del', 'vcf': {'chr': '17', 'ref': 'ACTCTTG', 'pos': '48269339', 'alt': 'A'}}
        assert results['NM_000088.3:c.2024_2028+1del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50191979_50191984del', 'vcf': {'chr': '17', 'ref': 'ACTCTTG', 'pos': '50191978', 'alt': 'A'}}


    def test_variant141(self):
        variant = 'NM_000088.3:c.2024_2028+1del'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000088.3:c.2024_2028+1del' in results.keys()
        assert results['NM_000088.3:c.2024_2028+1del']['hgvs_lrg_transcript_variant'] == 'LRG_1t1:c.2024_2028+1del'
        assert results['NM_000088.3:c.2024_2028+1del']['refseqgene_context_intronic_sequence'] == 'NG_007400.1(NM_000088.3):c.2024_2028+1del'
        assert results['NM_000088.3:c.2024_2028+1del']['alt_genomic_loci'] == []
        assert results['NM_000088.3:c.2024_2028+1del']['gene_symbol'] == 'COL1A1'
        assert results['NM_000088.3:c.2024_2028+1del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000079.2(LRG_1p1):p.?', 'slr': 'NP_000079.2:p.?'}
        assert results['NM_000088.3:c.2024_2028+1del']['submitted_variant'] == 'NM_000088.3:c.2024_2028+1del'
        assert results['NM_000088.3:c.2024_2028+1del']['genome_context_intronic_sequence'] == 'NC_000017.10(NM_000088.3):c.2024_2028+1del'
        assert results['NM_000088.3:c.2024_2028+1del']['hgvs_lrg_variant'] == 'LRG_1:g.14656_14661del'
        assert results['NM_000088.3:c.2024_2028+1del']['hgvs_transcript_variant'] == 'NM_000088.3:c.2024_2028+1del'
        assert results['NM_000088.3:c.2024_2028+1del']['hgvs_refseqgene_variant'] == 'NG_007400.1:g.14656_14661del'
        assert results['NM_000088.3:c.2024_2028+1del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48269340_48269345del', 'vcf': {'chr': 'chr17', 'ref': 'ACTCTTG', 'pos': '48269339', 'alt': 'A'}}
        assert results['NM_000088.3:c.2024_2028+1del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50191979_50191984del', 'vcf': {'chr': 'chr17', 'ref': 'ACTCTTG', 'pos': '50191978', 'alt': 'A'}}
        assert results['NM_000088.3:c.2024_2028+1del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48269340_48269345del', 'vcf': {'chr': '17', 'ref': 'ACTCTTG', 'pos': '48269339', 'alt': 'A'}}
        assert results['NM_000088.3:c.2024_2028+1del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50191979_50191984del', 'vcf': {'chr': '17', 'ref': 'ACTCTTG', 'pos': '50191978', 'alt': 'A'}}


    def test_variant142(self):
        variant = 'ENST00000450616.1:n.31+1G>C'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'validation_warning_1' in results.keys()
        assert results['validation_warning_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['validation_warning_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['alt_genomic_loci'] == []
        assert results['validation_warning_1']['gene_symbol'] == ''
        assert results['validation_warning_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['validation_warning_1']['submitted_variant'] == 'ENST00000450616.1:n.31+1G>C'
        assert results['validation_warning_1']['genome_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['hgvs_lrg_variant'] == ''
        assert results['validation_warning_1']['hgvs_transcript_variant'] == ''
        assert results['validation_warning_1']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['validation_warning_1']['primary_assembly_loci'].keys()

        assert results['flag'] == 'warning'

    def test_variant143(self):
        variant = 'ENST00000491747:c.5071A>T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'validation_warning_1' in results.keys()
        assert results['validation_warning_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['validation_warning_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['alt_genomic_loci'] == []
        assert results['validation_warning_1']['gene_symbol'] == ''
        assert results['validation_warning_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['validation_warning_1']['submitted_variant'] == 'ENST00000491747:c.5071A>T'
        assert results['validation_warning_1']['genome_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['hgvs_lrg_variant'] == ''
        assert results['validation_warning_1']['hgvs_transcript_variant'] == ''
        assert results['validation_warning_1']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['validation_warning_1']['primary_assembly_loci'].keys()

        assert results['flag'] == 'warning'

    def test_variant144(self):
        variant = 'NM_000088.3:c.589G>T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000088.3:c.589G>T' in results.keys()
        assert results['NM_000088.3:c.589G>T']['hgvs_lrg_transcript_variant'] == 'LRG_1t1:c.589G>T'
        assert results['NM_000088.3:c.589G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000088.3:c.589G>T']['alt_genomic_loci'] == []
        assert results['NM_000088.3:c.589G>T']['gene_symbol'] == 'COL1A1'
        assert results['NM_000088.3:c.589G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000079.2(LRG_1p1):p.(Gly197Cys)', 'slr': 'NP_000079.2:p.(G197C)'}
        assert results['NM_000088.3:c.589G>T']['submitted_variant'] == 'NM_000088.3:c.589G>T'
        assert results['NM_000088.3:c.589G>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_000088.3:c.589G>T']['hgvs_lrg_variant'] == 'LRG_1:g.8638G>T'
        assert results['NM_000088.3:c.589G>T']['hgvs_transcript_variant'] == 'NM_000088.3:c.589G>T'
        assert results['NM_000088.3:c.589G>T']['hgvs_refseqgene_variant'] == 'NG_007400.1:g.8638G>T'
        assert results['NM_000088.3:c.589G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275363C>A', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '48275363', 'alt': u'A'}}
        assert results['NM_000088.3:c.589G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198002C>A', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '50198002', 'alt': u'A'}}
        assert results['NM_000088.3:c.589G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275363C>A', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '48275363', 'alt': u'A'}}
        assert results['NM_000088.3:c.589G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198002C>A', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '50198002', 'alt': u'A'}}


    def test_variant145(self):
        variant = 'NG_007400.1:g.8638G>T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000088.3:c.589G>T' in results.keys()
        assert results['NM_000088.3:c.589G>T']['hgvs_lrg_transcript_variant'] == 'LRG_1t1:c.589G>T'
        assert results['NM_000088.3:c.589G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000088.3:c.589G>T']['alt_genomic_loci'] == []
        assert results['NM_000088.3:c.589G>T']['gene_symbol'] == 'COL1A1'
        assert results['NM_000088.3:c.589G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000079.2(LRG_1p1):p.(Gly197Cys)', 'slr': 'NP_000079.2:p.(G197C)'}
        assert results['NM_000088.3:c.589G>T']['submitted_variant'] == 'NG_007400.1:g.8638G>T'
        assert results['NM_000088.3:c.589G>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_000088.3:c.589G>T']['hgvs_lrg_variant'] == 'LRG_1:g.8638G>T'
        assert results['NM_000088.3:c.589G>T']['hgvs_transcript_variant'] == 'NM_000088.3:c.589G>T'
        assert results['NM_000088.3:c.589G>T']['hgvs_refseqgene_variant'] == 'NG_007400.1:g.8638G>T'
        assert results['NM_000088.3:c.589G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275363C>A', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '48275363', 'alt': u'A'}}
        assert results['NM_000088.3:c.589G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198002C>A', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '50198002', 'alt': u'A'}}
        assert results['NM_000088.3:c.589G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275363C>A', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '48275363', 'alt': u'A'}}
        assert results['NM_000088.3:c.589G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198002C>A', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '50198002', 'alt': u'A'}}


    def test_variant146(self):
        variant = 'LRG_1:g.8638G>T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000088.3:c.589G>T' in results.keys()
        assert results['NM_000088.3:c.589G>T']['hgvs_lrg_transcript_variant'] == 'LRG_1t1:c.589G>T'
        assert results['NM_000088.3:c.589G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000088.3:c.589G>T']['alt_genomic_loci'] == []
        assert results['NM_000088.3:c.589G>T']['gene_symbol'] == 'COL1A1'
        assert results['NM_000088.3:c.589G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000079.2(LRG_1p1):p.(Gly197Cys)', 'slr': 'NP_000079.2:p.(G197C)'}
        assert results['NM_000088.3:c.589G>T']['submitted_variant'] == 'LRG_1:g.8638G>T'
        assert results['NM_000088.3:c.589G>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_000088.3:c.589G>T']['hgvs_lrg_variant'] == 'LRG_1:g.8638G>T'
        assert results['NM_000088.3:c.589G>T']['hgvs_transcript_variant'] == 'NM_000088.3:c.589G>T'
        assert results['NM_000088.3:c.589G>T']['hgvs_refseqgene_variant'] == 'NG_007400.1:g.8638G>T'
        assert results['NM_000088.3:c.589G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275363C>A', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '48275363', 'alt': u'A'}}
        assert results['NM_000088.3:c.589G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198002C>A', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '50198002', 'alt': u'A'}}
        assert results['NM_000088.3:c.589G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275363C>A', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '48275363', 'alt': u'A'}}
        assert results['NM_000088.3:c.589G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198002C>A', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '50198002', 'alt': u'A'}}


    def test_variant147(self):
        variant = 'LRG_1t1:c.589G>T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'validation_warning_1' in results.keys()
        assert results['validation_warning_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['validation_warning_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['alt_genomic_loci'] == []
        assert results['validation_warning_1']['gene_symbol'] == ''
        assert results['validation_warning_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['validation_warning_1']['submitted_variant'] == 'LRG_1t1:c.589G>T'
        assert results['validation_warning_1']['genome_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['hgvs_lrg_variant'] == ''
        assert results['validation_warning_1']['hgvs_transcript_variant'] == ''
        assert results['validation_warning_1']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['validation_warning_1']['primary_assembly_loci'].keys()

        assert results['flag'] == 'warning'

    def test_variant148(self):
        variant = 'chr16:g.15832508_15832509delinsAC'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001040113.1:c.3055_3056inv' in results.keys()
        assert results['NM_001040113.1:c.3055_3056inv']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001040113.1:c.3055_3056inv']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001040113.1:c.3055_3056inv']['alt_genomic_loci'] == [{'grch38': {'hgvs_genomic_description': 'NT_187607.1:g.1396662_1396663inv', 'vcf': {'chr': 'HSCHR16_1_CTG1', 'ref': 'GT', 'pos': '1396662', 'alt': 'AC'}}}, {'hg38': {'hgvs_genomic_description': 'NT_187607.1:g.1396662_1396663inv', 'vcf': {'chr': 'chr16_KI270853v1_alt', 'ref': 'GT', 'pos': '1396662', 'alt': 'AC'}}}]
        assert results['NM_001040113.1:c.3055_3056inv']['gene_symbol'] == 'MYH11'
        assert results['NM_001040113.1:c.3055_3056inv']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001035202.1:p.(Thr1019Val)', 'slr': 'NP_001035202.1:p.(T1019V)'}
        assert results['NM_001040113.1:c.3055_3056inv']['submitted_variant'] == 'chr16:g.15832508_15832509delinsAC'
        assert results['NM_001040113.1:c.3055_3056inv']['genome_context_intronic_sequence'] == ''
        assert results['NM_001040113.1:c.3055_3056inv']['hgvs_lrg_variant'] == ''
        assert results['NM_001040113.1:c.3055_3056inv']['hgvs_transcript_variant'] == 'NM_001040113.1:c.3055_3056inv'
        assert results['NM_001040113.1:c.3055_3056inv']['hgvs_refseqgene_variant'] == 'NG_009299.1:g.123379_123380inv'
        assert results['NM_001040113.1:c.3055_3056inv']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.15832508_15832509inv', 'vcf': {'chr': 'chr16', 'ref': 'GT', 'pos': '15832508', 'alt': 'AC'}}
        assert results['NM_001040113.1:c.3055_3056inv']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.15738651_15738652inv', 'vcf': {'chr': 'chr16', 'ref': 'GT', 'pos': '15738651', 'alt': 'AC'}}
        assert results['NM_001040113.1:c.3055_3056inv']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.15832508_15832509inv', 'vcf': {'chr': '16', 'ref': 'GT', 'pos': '15832508', 'alt': 'AC'}}
        assert results['NM_001040113.1:c.3055_3056inv']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.15738651_15738652inv', 'vcf': {'chr': '16', 'ref': 'GT', 'pos': '15738651', 'alt': 'AC'}}

        assert 'NM_001040114.1:c.3055_3056inv' in results.keys()
        assert results['NM_001040114.1:c.3055_3056inv']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001040114.1:c.3055_3056inv']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001040114.1:c.3055_3056inv']['alt_genomic_loci'] == [{'grch38': {'hgvs_genomic_description': 'NT_187607.1:g.1396662_1396663inv', 'vcf': {'chr': 'HSCHR16_1_CTG1', 'ref': 'GT', 'pos': '1396662', 'alt': 'AC'}}}, {'hg38': {'hgvs_genomic_description': 'NT_187607.1:g.1396662_1396663inv', 'vcf': {'chr': 'chr16_KI270853v1_alt', 'ref': 'GT', 'pos': '1396662', 'alt': 'AC'}}}]
        assert results['NM_001040114.1:c.3055_3056inv']['gene_symbol'] == 'MYH11'
        assert results['NM_001040114.1:c.3055_3056inv']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001035203.1:p.(Thr1019Val)', 'slr': 'NP_001035203.1:p.(T1019V)'}
        assert results['NM_001040114.1:c.3055_3056inv']['submitted_variant'] == 'chr16:g.15832508_15832509delinsAC'
        assert results['NM_001040114.1:c.3055_3056inv']['genome_context_intronic_sequence'] == ''
        assert results['NM_001040114.1:c.3055_3056inv']['hgvs_lrg_variant'] == ''
        assert results['NM_001040114.1:c.3055_3056inv']['hgvs_transcript_variant'] == 'NM_001040114.1:c.3055_3056inv'
        assert results['NM_001040114.1:c.3055_3056inv']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001040114.1:c.3055_3056inv']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.15832508_15832509inv', 'vcf': {'chr': 'chr16', 'ref': 'GT', 'pos': '15832508', 'alt': 'AC'}}
        assert results['NM_001040114.1:c.3055_3056inv']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.15738651_15738652inv', 'vcf': {'chr': 'chr16', 'ref': 'GT', 'pos': '15738651', 'alt': 'AC'}}
        assert results['NM_001040114.1:c.3055_3056inv']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.15832508_15832509inv', 'vcf': {'chr': '16', 'ref': 'GT', 'pos': '15832508', 'alt': 'AC'}}
        assert results['NM_001040114.1:c.3055_3056inv']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.15738651_15738652inv', 'vcf': {'chr': '16', 'ref': 'GT', 'pos': '15738651', 'alt': 'AC'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_002474.2:c.3034_3035inv' in results.keys()
        assert results['NM_002474.2:c.3034_3035inv']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_002474.2:c.3034_3035inv']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_002474.2:c.3034_3035inv']['alt_genomic_loci'] == [{'grch38': {'hgvs_genomic_description': 'NT_187607.1:g.1396662_1396663inv', 'vcf': {'chr': 'HSCHR16_1_CTG1', 'ref': 'GT', 'pos': '1396662', 'alt': 'AC'}}}, {'hg38': {'hgvs_genomic_description': 'NT_187607.1:g.1396662_1396663inv', 'vcf': {'chr': 'chr16_KI270853v1_alt', 'ref': 'GT', 'pos': '1396662', 'alt': 'AC'}}}]
        assert results['NM_002474.2:c.3034_3035inv']['gene_symbol'] == 'MYH11'
        assert results['NM_002474.2:c.3034_3035inv']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_002465.1:p.(Thr1012Val)', 'slr': 'NP_002465.1:p.(T1012V)'}
        assert results['NM_002474.2:c.3034_3035inv']['submitted_variant'] == 'chr16:g.15832508_15832509delinsAC'
        assert results['NM_002474.2:c.3034_3035inv']['genome_context_intronic_sequence'] == ''
        assert results['NM_002474.2:c.3034_3035inv']['hgvs_lrg_variant'] == ''
        assert results['NM_002474.2:c.3034_3035inv']['hgvs_transcript_variant'] == 'NM_002474.2:c.3034_3035inv'
        assert results['NM_002474.2:c.3034_3035inv']['hgvs_refseqgene_variant'] == ''
        assert results['NM_002474.2:c.3034_3035inv']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.15832508_15832509inv', 'vcf': {'chr': 'chr16', 'ref': 'GT', 'pos': '15832508', 'alt': 'AC'}}
        assert results['NM_002474.2:c.3034_3035inv']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.15738651_15738652inv', 'vcf': {'chr': 'chr16', 'ref': 'GT', 'pos': '15738651', 'alt': 'AC'}}
        assert results['NM_002474.2:c.3034_3035inv']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.15832508_15832509inv', 'vcf': {'chr': '16', 'ref': 'GT', 'pos': '15832508', 'alt': 'AC'}}
        assert results['NM_002474.2:c.3034_3035inv']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.15738651_15738652inv', 'vcf': {'chr': '16', 'ref': 'GT', 'pos': '15738651', 'alt': 'AC'}}

        assert 'NM_022844.2:c.3034_3035inv' in results.keys()
        assert results['NM_022844.2:c.3034_3035inv']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_022844.2:c.3034_3035inv']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_022844.2:c.3034_3035inv']['alt_genomic_loci'] == [{'grch38': {'hgvs_genomic_description': 'NT_187607.1:g.1396662_1396663inv', 'vcf': {'chr': 'HSCHR16_1_CTG1', 'ref': 'GT', 'pos': '1396662', 'alt': 'AC'}}}, {'hg38': {'hgvs_genomic_description': 'NT_187607.1:g.1396662_1396663inv', 'vcf': {'chr': 'chr16_KI270853v1_alt', 'ref': 'GT', 'pos': '1396662', 'alt': 'AC'}}}]
        assert results['NM_022844.2:c.3034_3035inv']['gene_symbol'] == 'MYH11'
        assert results['NM_022844.2:c.3034_3035inv']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_074035.1:p.(Thr1012Val)', 'slr': 'NP_074035.1:p.(T1012V)'}
        assert results['NM_022844.2:c.3034_3035inv']['submitted_variant'] == 'chr16:g.15832508_15832509delinsAC'
        assert results['NM_022844.2:c.3034_3035inv']['genome_context_intronic_sequence'] == ''
        assert results['NM_022844.2:c.3034_3035inv']['hgvs_lrg_variant'] == ''
        assert results['NM_022844.2:c.3034_3035inv']['hgvs_transcript_variant'] == 'NM_022844.2:c.3034_3035inv'
        assert results['NM_022844.2:c.3034_3035inv']['hgvs_refseqgene_variant'] == ''
        assert results['NM_022844.2:c.3034_3035inv']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.15832508_15832509inv', 'vcf': {'chr': 'chr16', 'ref': 'GT', 'pos': '15832508', 'alt': 'AC'}}
        assert results['NM_022844.2:c.3034_3035inv']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.15738651_15738652inv', 'vcf': {'chr': 'chr16', 'ref': 'GT', 'pos': '15738651', 'alt': 'AC'}}
        assert results['NM_022844.2:c.3034_3035inv']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.15832508_15832509inv', 'vcf': {'chr': '16', 'ref': 'GT', 'pos': '15832508', 'alt': 'AC'}}
        assert results['NM_022844.2:c.3034_3035inv']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.15738651_15738652inv', 'vcf': {'chr': '16', 'ref': 'GT', 'pos': '15738651', 'alt': 'AC'}}


    def test_variant149(self):
        variant = 'NG_012386.1:g.24048dupG'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001162426.1:c.363+1dup' in results.keys()
        assert results['NM_001162426.1:c.363+1dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001162426.1:c.363+1dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001162426.1:c.363+1dup']['alt_genomic_loci'] == []
        assert results['NM_001162426.1:c.363+1dup']['gene_symbol'] == 'TSC1'
        assert results['NM_001162426.1:c.363+1dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001155898.1:p.?', 'slr': 'NP_001155898.1:p.?'}
        assert results['NM_001162426.1:c.363+1dup']['submitted_variant'] == 'NG_012386.1:g.24048dupG'
        assert results['NM_001162426.1:c.363+1dup']['genome_context_intronic_sequence'] == 'NC_000009.11(NM_001162426.1):c.363+1dup'
        assert results['NM_001162426.1:c.363+1dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001162426.1:c.363+1dup']['hgvs_transcript_variant'] == 'NM_001162426.1:c.363+1dup'
        assert results['NM_001162426.1:c.363+1dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001162426.1:c.363+1dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.135800973dup', 'vcf': {'chr': 'chr9', 'ref': 'C', 'pos': '135800973', 'alt': 'CC'}}
        assert results['NM_001162426.1:c.363+1dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.132925586dup', 'vcf': {'chr': 'chr9', 'ref': 'C', 'pos': '132925586', 'alt': 'CC'}}
        assert results['NM_001162426.1:c.363+1dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.135800973dup', 'vcf': {'chr': '9', 'ref': 'C', 'pos': '135800973', 'alt': 'CC'}}
        assert results['NM_001162426.1:c.363+1dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.132925586dup', 'vcf': {'chr': '9', 'ref': 'C', 'pos': '132925586', 'alt': 'CC'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001362177.1:c.-1+1dup' in results.keys()
        assert results['NM_001362177.1:c.-1+1dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001362177.1:c.-1+1dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001362177.1:c.-1+1dup']['alt_genomic_loci'] == []
        assert results['NM_001362177.1:c.-1+1dup']['gene_symbol'] == 'TSC1'
        assert results['NM_001362177.1:c.-1+1dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001349106.1:p.?', 'slr': 'NP_001349106.1:p.?'}
        assert results['NM_001362177.1:c.-1+1dup']['submitted_variant'] == 'NG_012386.1:g.24048dupG'
        assert results['NM_001362177.1:c.-1+1dup']['genome_context_intronic_sequence'] == 'NC_000009.11(NM_001362177.1):c.-1+1dup'
        assert results['NM_001362177.1:c.-1+1dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001362177.1:c.-1+1dup']['hgvs_transcript_variant'] == 'NM_001362177.1:c.-1+1dup'
        assert results['NM_001362177.1:c.-1+1dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001362177.1:c.-1+1dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.135800973dup', 'vcf': {'chr': 'chr9', 'ref': 'C', 'pos': '135800973', 'alt': 'CC'}}
        assert results['NM_001362177.1:c.-1+1dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.132925586dup', 'vcf': {'chr': 'chr9', 'ref': 'C', 'pos': '132925586', 'alt': 'CC'}}
        assert results['NM_001362177.1:c.-1+1dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.135800973dup', 'vcf': {'chr': '9', 'ref': 'C', 'pos': '135800973', 'alt': 'CC'}}
        assert results['NM_001362177.1:c.-1+1dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.132925586dup', 'vcf': {'chr': '9', 'ref': 'C', 'pos': '132925586', 'alt': 'CC'}}

        assert 'NM_000368.4:c.363+1dup' in results.keys()
        assert results['NM_000368.4:c.363+1dup']['hgvs_lrg_transcript_variant'] == 'LRG_486t1:c.363+1dup'
        assert results['NM_000368.4:c.363+1dup']['refseqgene_context_intronic_sequence'] == 'NG_012386.1(NM_000368.4):c.363+1dup'
        assert results['NM_000368.4:c.363+1dup']['alt_genomic_loci'] == []
        assert results['NM_000368.4:c.363+1dup']['gene_symbol'] == 'TSC1'
        assert results['NM_000368.4:c.363+1dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000359.1(LRG_486p1):p.?', 'slr': 'NP_000359.1:p.?'}
        assert results['NM_000368.4:c.363+1dup']['submitted_variant'] == 'NG_012386.1:g.24048dupG'
        assert results['NM_000368.4:c.363+1dup']['genome_context_intronic_sequence'] == 'NC_000009.11(NM_000368.4):c.363+1dup'
        assert results['NM_000368.4:c.363+1dup']['hgvs_lrg_variant'] == 'LRG_486:g.24048dup'
        assert results['NM_000368.4:c.363+1dup']['hgvs_transcript_variant'] == 'NM_000368.4:c.363+1dup'
        assert results['NM_000368.4:c.363+1dup']['hgvs_refseqgene_variant'] == 'NG_012386.1:g.24048dup'
        assert results['NM_000368.4:c.363+1dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.135800973dup', 'vcf': {'chr': 'chr9', 'ref': 'C', 'pos': '135800973', 'alt': 'CC'}}
        assert results['NM_000368.4:c.363+1dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.132925586dup', 'vcf': {'chr': 'chr9', 'ref': 'C', 'pos': '132925586', 'alt': 'CC'}}
        assert results['NM_000368.4:c.363+1dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.135800973dup', 'vcf': {'chr': '9', 'ref': 'C', 'pos': '135800973', 'alt': 'CC'}}
        assert results['NM_000368.4:c.363+1dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.132925586dup', 'vcf': {'chr': '9', 'ref': 'C', 'pos': '132925586', 'alt': 'CC'}}

        assert 'NM_001162427.1:c.210+1615dup' in results.keys()
        assert results['NM_001162427.1:c.210+1615dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001162427.1:c.210+1615dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001162427.1:c.210+1615dup']['alt_genomic_loci'] == []
        assert results['NM_001162427.1:c.210+1615dup']['gene_symbol'] == 'TSC1'
        assert results['NM_001162427.1:c.210+1615dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001155899.1:p.?', 'slr': 'NP_001155899.1:p.?'}
        assert results['NM_001162427.1:c.210+1615dup']['submitted_variant'] == 'NG_012386.1:g.24048dupG'
        assert results['NM_001162427.1:c.210+1615dup']['genome_context_intronic_sequence'] == 'NC_000009.11(NM_001162427.1):c.210+1615dup'
        assert results['NM_001162427.1:c.210+1615dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001162427.1:c.210+1615dup']['hgvs_transcript_variant'] == 'NM_001162427.1:c.210+1615dup'
        assert results['NM_001162427.1:c.210+1615dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001162427.1:c.210+1615dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.135800973dup', 'vcf': {'chr': 'chr9', 'ref': 'C', 'pos': '135800973', 'alt': 'CC'}}
        assert results['NM_001162427.1:c.210+1615dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.132925586dup', 'vcf': {'chr': 'chr9', 'ref': 'C', 'pos': '132925586', 'alt': 'CC'}}
        assert results['NM_001162427.1:c.210+1615dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.135800973dup', 'vcf': {'chr': '9', 'ref': 'C', 'pos': '135800973', 'alt': 'CC'}}
        assert results['NM_001162427.1:c.210+1615dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.132925586dup', 'vcf': {'chr': '9', 'ref': 'C', 'pos': '132925586', 'alt': 'CC'}}


    def test_variant150(self):
        variant = 'NM_033517.1:c.1307_1309delCGA'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_033517.1:c.1307_1309del' in results.keys()
        assert results['NM_033517.1:c.1307_1309del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_033517.1:c.1307_1309del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_033517.1:c.1307_1309del']['alt_genomic_loci'] == [{'grch38': {'hgvs_genomic_description': 'NW_015148969.1:g.33721_33723del', 'vcf': {'chr': 'HG1311_PATCH', 'ref': 'CCGA', 'pos': '33720', 'alt': 'C'}}}, {'hg38': {'hgvs_genomic_description': 'NW_015148969.1:g.33721_33723del', 'vcf': {'chr': 'NW_015148969.1', 'ref': 'CCGA', 'pos': '33720', 'alt': 'C'}}}]
        assert results['NM_033517.1:c.1307_1309del']['gene_symbol'] == 'SHANK3'
        assert results['NM_033517.1:c.1307_1309del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_277052.1:p.(Pro436_Ser437delinsArg)', 'slr': 'NP_277052.1:p.(P436_S437delinsR)'}
        assert results['NM_033517.1:c.1307_1309del']['submitted_variant'] == 'NM_033517.1:c.1307_1309delCGA'
        assert results['NM_033517.1:c.1307_1309del']['genome_context_intronic_sequence'] == ''
        assert results['NM_033517.1:c.1307_1309del']['hgvs_lrg_variant'] == ''
        assert results['NM_033517.1:c.1307_1309del']['hgvs_transcript_variant'] == 'NM_033517.1:c.1307_1309del'
        assert results['NM_033517.1:c.1307_1309del']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['NM_033517.1:c.1307_1309del']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['NM_033517.1:c.1307_1309del']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['NM_033517.1:c.1307_1309del']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['NM_033517.1:c.1307_1309del']['primary_assembly_loci'].keys()


    def test_variant151(self):
        variant = 'HG1311_PATCH-33720-CCGA-C'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'validation_warning_1' in results.keys()
        assert results['validation_warning_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['validation_warning_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['alt_genomic_loci'] == []
        assert results['validation_warning_1']['gene_symbol'] == ''
        assert results['validation_warning_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['validation_warning_1']['submitted_variant'] == 'HG1311_PATCH-33720-CCGA-C'
        assert results['validation_warning_1']['genome_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['hgvs_lrg_variant'] == ''
        assert results['validation_warning_1']['hgvs_transcript_variant'] == ''
        assert results['validation_warning_1']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['validation_warning_1']['primary_assembly_loci'].keys()

        assert results['flag'] == 'warning'

    def test_variant152(self):
        variant = '2-73675227-TCTC-TCTCCTC'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_015120.4:c.1573_1579=' in results.keys()
        assert results['NM_015120.4:c.1573_1579=']['hgvs_lrg_transcript_variant'] == 'LRG_741t1:c.1573_1579='
        assert results['NM_015120.4:c.1573_1579=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_015120.4:c.1573_1579=']['alt_genomic_loci'] == []
        assert results['NM_015120.4:c.1573_1579=']['gene_symbol'] == 'ALMS1'
        assert results['NM_015120.4:c.1573_1579=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_055935.4(LRG_741p1):p.(Ser525=)', 'slr': 'NP_055935.4:p.(S525=)'}
        assert results['NM_015120.4:c.1573_1579=']['submitted_variant'] == '2-73675227-TCTC-TCTCCTC'
        assert results['NM_015120.4:c.1573_1579=']['genome_context_intronic_sequence'] == ''
        assert results['NM_015120.4:c.1573_1579=']['hgvs_lrg_variant'] == 'LRG_741:g.67345_67351='
        assert results['NM_015120.4:c.1573_1579=']['hgvs_transcript_variant'] == 'NM_015120.4:c.1573_1579='
        assert results['NM_015120.4:c.1573_1579=']['hgvs_refseqgene_variant'] == 'NG_011690.1:g.67345_67351='
        assert results['NM_015120.4:c.1573_1579=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.73675228_73675230dup', 'vcf': {'chr': 'chr2', 'ref': 'CTC', 'pos': '73675228', 'alt': u'CTCCTC'}}
        assert results['NM_015120.4:c.1573_1579=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.73448097_73448103=', 'vcf': {'chr': 'chr2', 'ref': u'TCTCCTC', 'pos': '73448097', 'alt': u'TCTCCTC'}}
        assert results['NM_015120.4:c.1573_1579=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.73675228_73675230dup', 'vcf': {'chr': '2', 'ref': 'CTC', 'pos': '73675228', 'alt': u'CTCCTC'}}
        assert results['NM_015120.4:c.1573_1579=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.73448097_73448103=', 'vcf': {'chr': '2', 'ref': u'TCTCCTC', 'pos': '73448097', 'alt': u'TCTCCTC'}}

        assert results['flag'] == 'gene_variant'

    def test_variant153(self):
        variant = '2-73675227-TC-TC'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_015120.4:c.1577_1579del' in results.keys()
        assert results['NM_015120.4:c.1577_1579del']['hgvs_lrg_transcript_variant'] == 'LRG_741t1:c.1577_1579del'
        assert results['NM_015120.4:c.1577_1579del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_015120.4:c.1577_1579del']['alt_genomic_loci'] == []
        assert results['NM_015120.4:c.1577_1579del']['gene_symbol'] == 'ALMS1'
        assert results['NM_015120.4:c.1577_1579del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_055935.4(LRG_741p1):p.(Pro526del)', 'slr': 'NP_055935.4:p.(P526del)'}
        assert results['NM_015120.4:c.1577_1579del']['submitted_variant'] == '2-73675227-TC-TC'
        assert results['NM_015120.4:c.1577_1579del']['genome_context_intronic_sequence'] == ''
        assert results['NM_015120.4:c.1577_1579del']['hgvs_lrg_variant'] == 'LRG_741:g.67349_67351del'
        assert results['NM_015120.4:c.1577_1579del']['hgvs_transcript_variant'] == 'NM_015120.4:c.1577_1579del'
        assert results['NM_015120.4:c.1577_1579del']['hgvs_refseqgene_variant'] == 'NG_011690.1:g.67349_67351del'
        assert results['NM_015120.4:c.1577_1579del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.73675227_73675229=', 'vcf': {'chr': 'chr2', 'ref': 'TCT', 'pos': '73675227', 'alt': 'TCT'}}
        assert results['NM_015120.4:c.1577_1579del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.73448101_73448103del', 'vcf': {'chr': 'chr2', 'ref': 'TCTC', 'pos': '73448097', 'alt': 'T'}}
        assert results['NM_015120.4:c.1577_1579del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.73675227_73675229=', 'vcf': {'chr': '2', 'ref': 'TCT', 'pos': '73675227', 'alt': 'TCT'}}
        assert results['NM_015120.4:c.1577_1579del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.73448101_73448103del', 'vcf': {'chr': '2', 'ref': 'TCTC', 'pos': '73448097', 'alt': 'T'}}

        assert results['flag'] == 'gene_variant'

    def test_variant154(self):
        variant = '3-14561627-AG-AGG'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001080423.3:c.1016_1020=' in results.keys()
        assert results['NM_001080423.3:c.1016_1020=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001080423.3:c.1016_1020=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001080423.3:c.1016_1020=']['alt_genomic_loci'] == []
        assert results['NM_001080423.3:c.1016_1020=']['gene_symbol'] == 'GRIP2'
        assert results['NM_001080423.3:c.1016_1020=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001073892.3:p.(Arg339=)', 'slr': 'NP_001073892.3:p.(R339=)'}
        assert results['NM_001080423.3:c.1016_1020=']['submitted_variant'] == '3-14561627-AG-AGG'
        assert results['NM_001080423.3:c.1016_1020=']['genome_context_intronic_sequence'] == ''
        assert results['NM_001080423.3:c.1016_1020=']['hgvs_lrg_variant'] == ''
        assert results['NM_001080423.3:c.1016_1020=']['hgvs_transcript_variant'] == 'NM_001080423.3:c.1016_1020='
        assert results['NM_001080423.3:c.1016_1020=']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001080423.3:c.1016_1020=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000003.11:g.14561629dup', 'vcf': {'chr': 'chr3', 'ref': 'G', 'pos': '14561628', 'alt': u'GG'}}
        assert results['NM_001080423.3:c.1016_1020=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000003.12:g.14520120_14520124=', 'vcf': {'chr': 'chr3', 'ref': u'GGGCC', 'pos': '14520120', 'alt': u'GGGCC'}}
        assert results['NM_001080423.3:c.1016_1020=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000003.11:g.14561629dup', 'vcf': {'chr': '3', 'ref': 'G', 'pos': '14561628', 'alt': u'GG'}}
        assert results['NM_001080423.3:c.1016_1020=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000003.12:g.14520120_14520124=', 'vcf': {'chr': '3', 'ref': u'GGGCC', 'pos': '14520120', 'alt': u'GGGCC'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001080423.2:c.1307_1311=' in results.keys()
        assert results['NM_001080423.2:c.1307_1311=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001080423.2:c.1307_1311=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001080423.2:c.1307_1311=']['alt_genomic_loci'] == []
        assert results['NM_001080423.2:c.1307_1311=']['gene_symbol'] == 'GRIP2'
        assert results['NM_001080423.2:c.1307_1311=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001073892.2:p.(Arg436=)', 'slr': 'NP_001073892.2:p.(R436=)'}
        assert results['NM_001080423.2:c.1307_1311=']['submitted_variant'] == '3-14561627-AG-AGG'
        assert results['NM_001080423.2:c.1307_1311=']['genome_context_intronic_sequence'] == ''
        assert results['NM_001080423.2:c.1307_1311=']['hgvs_lrg_variant'] == ''
        assert results['NM_001080423.2:c.1307_1311=']['hgvs_transcript_variant'] == 'NM_001080423.2:c.1307_1311='
        assert results['NM_001080423.2:c.1307_1311=']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001080423.2:c.1307_1311=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000003.11:g.14561629dup', 'vcf': {'chr': 'chr3', 'ref': 'G', 'pos': '14561628', 'alt': u'GG'}}
        assert 'hg38' not in results['NM_001080423.2:c.1307_1311=']['primary_assembly_loci'].keys()
        assert results['NM_001080423.2:c.1307_1311=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000003.11:g.14561629dup', 'vcf': {'chr': '3', 'ref': 'G', 'pos': '14561628', 'alt': u'GG'}}
        assert 'grch38' not in results['NM_001080423.2:c.1307_1311=']['primary_assembly_loci'].keys()


    def test_variant155(self):
        variant = '3-14561630-CC-CC'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001080423.2:c.1311del' in results.keys()
        assert results['NM_001080423.2:c.1311del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001080423.2:c.1311del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001080423.2:c.1311del']['alt_genomic_loci'] == []
        assert results['NM_001080423.2:c.1311del']['gene_symbol'] == 'GRIP2'
        assert results['NM_001080423.2:c.1311del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001073892.2:p.(Ser438GlnfsTer4)', 'slr': 'NP_001073892.2:p.(S438Qfs*4)'}
        assert results['NM_001080423.2:c.1311del']['submitted_variant'] == '3-14561630-CC-CC'
        assert results['NM_001080423.2:c.1311del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001080423.2:c.1311del']['hgvs_lrg_variant'] == ''
        assert results['NM_001080423.2:c.1311del']['hgvs_transcript_variant'] == 'NM_001080423.2:c.1311del'
        assert results['NM_001080423.2:c.1311del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001080423.2:c.1311del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000003.11:g.14561624_14561630=', 'vcf': {'chr': 'chr3', 'ref': 'CTGAGGC', 'pos': '14561624', 'alt': 'CTGAGGC'}}
        assert 'hg38' not in results['NM_001080423.2:c.1311del']['primary_assembly_loci'].keys()
        assert results['NM_001080423.2:c.1311del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000003.11:g.14561624_14561630=', 'vcf': {'chr': '3', 'ref': 'CTGAGGC', 'pos': '14561624', 'alt': 'CTGAGGC'}}
        assert 'grch38' not in results['NM_001080423.2:c.1311del']['primary_assembly_loci'].keys()

        assert results['flag'] == 'gene_variant'
        assert 'NM_001080423.3:c.1020del' in results.keys()
        assert results['NM_001080423.3:c.1020del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001080423.3:c.1020del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001080423.3:c.1020del']['alt_genomic_loci'] == []
        assert results['NM_001080423.3:c.1020del']['gene_symbol'] == 'GRIP2'
        assert results['NM_001080423.3:c.1020del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001073892.3:p.(Ser341GlnfsTer4)', 'slr': 'NP_001073892.3:p.(S341Qfs*4)'}
        assert results['NM_001080423.3:c.1020del']['submitted_variant'] == '3-14561630-CC-CC'
        assert results['NM_001080423.3:c.1020del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001080423.3:c.1020del']['hgvs_lrg_variant'] == ''
        assert results['NM_001080423.3:c.1020del']['hgvs_transcript_variant'] == 'NM_001080423.3:c.1020del'
        assert results['NM_001080423.3:c.1020del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001080423.3:c.1020del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000003.11:g.14561624_14561630=', 'vcf': {'chr': 'chr3', 'ref': 'CTGAGGC', 'pos': '14561624', 'alt': 'CTGAGGC'}}
        assert results['NM_001080423.3:c.1020del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000003.12:g.14520122del', 'vcf': {'chr': 'chr3', 'ref': 'AG', 'pos': '14520119', 'alt': 'A'}}
        assert results['NM_001080423.3:c.1020del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000003.11:g.14561624_14561630=', 'vcf': {'chr': '3', 'ref': 'CTGAGGC', 'pos': '14561624', 'alt': 'CTGAGGC'}}
        assert results['NM_001080423.3:c.1020del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000003.12:g.14520122del', 'vcf': {'chr': '3', 'ref': 'AG', 'pos': '14520119', 'alt': 'A'}}


    def test_variant156(self):
        variant = '6-90403795-G-G'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_014611.1:c.9879T>C' in results.keys()
        assert results['NM_014611.1:c.9879T>C']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_014611.1:c.9879T>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_014611.1:c.9879T>C']['alt_genomic_loci'] == []
        assert results['NM_014611.1:c.9879T>C']['gene_symbol'] == 'MDN1'
        assert results['NM_014611.1:c.9879T>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_055426.1:p.(Val3293=)', 'slr': 'NP_055426.1:p.(V3293=)'}
        assert results['NM_014611.1:c.9879T>C']['submitted_variant'] == '6-90403795-G-G'
        assert results['NM_014611.1:c.9879T>C']['genome_context_intronic_sequence'] == ''
        assert results['NM_014611.1:c.9879T>C']['hgvs_lrg_variant'] == ''
        assert results['NM_014611.1:c.9879T>C']['hgvs_transcript_variant'] == 'NM_014611.1:c.9879T>C'
        assert results['NM_014611.1:c.9879T>C']['hgvs_refseqgene_variant'] == ''
        assert results['NM_014611.1:c.9879T>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000006.11:g.90403795G=', 'vcf': {'chr': 'chr6', 'ref': 'G', 'pos': '90403795', 'alt': 'G'}}
        assert 'hg38' not in results['NM_014611.1:c.9879T>C']['primary_assembly_loci'].keys()
        assert results['NM_014611.1:c.9879T>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000006.11:g.90403795G=', 'vcf': {'chr': '6', 'ref': 'G', 'pos': '90403795', 'alt': 'G'}}
        assert 'grch38' not in results['NM_014611.1:c.9879T>C']['primary_assembly_loci'].keys()

        assert 'NM_014611.2:c.9879C=' in results.keys()
        assert results['NM_014611.2:c.9879C=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_014611.2:c.9879C=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_014611.2:c.9879C=']['alt_genomic_loci'] == []
        assert results['NM_014611.2:c.9879C=']['gene_symbol'] == 'MDN1'
        assert results['NM_014611.2:c.9879C=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_055426.1:p.(Val3293=)', 'slr': 'NP_055426.1:p.(V3293=)'}
        assert results['NM_014611.2:c.9879C=']['submitted_variant'] == '6-90403795-G-G'
        assert results['NM_014611.2:c.9879C=']['genome_context_intronic_sequence'] == ''
        assert results['NM_014611.2:c.9879C=']['hgvs_lrg_variant'] == ''
        assert results['NM_014611.2:c.9879C=']['hgvs_transcript_variant'] == 'NM_014611.2:c.9879C='
        assert results['NM_014611.2:c.9879C=']['hgvs_refseqgene_variant'] == ''
        assert results['NM_014611.2:c.9879C=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000006.11:g.90403795G=', 'vcf': {'chr': 'chr6', 'ref': u'G', 'pos': '90403795', 'alt': u'G'}}
        assert results['NM_014611.2:c.9879C=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000006.12:g.89694076G=', 'vcf': {'chr': 'chr6', 'ref': u'G', 'pos': '89694076', 'alt': u'G'}}
        assert results['NM_014611.2:c.9879C=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000006.11:g.90403795G=', 'vcf': {'chr': '6', 'ref': u'G', 'pos': '90403795', 'alt': u'G'}}
        assert results['NM_014611.2:c.9879C=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000006.12:g.89694076G=', 'vcf': {'chr': '6', 'ref': u'G', 'pos': '89694076', 'alt': u'G'}}


    def test_variant157(self):
        variant = '6-90403795-G-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_014611.2:c.9879C>T' in results.keys()
        assert results['NM_014611.2:c.9879C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_014611.2:c.9879C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_014611.2:c.9879C>T']['alt_genomic_loci'] == []
        assert results['NM_014611.2:c.9879C>T']['gene_symbol'] == 'MDN1'
        assert results['NM_014611.2:c.9879C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_055426.1:p.(Val3293=)', 'slr': 'NP_055426.1:p.(V3293=)'}
        assert results['NM_014611.2:c.9879C>T']['submitted_variant'] == '6-90403795-G-A'
        assert results['NM_014611.2:c.9879C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_014611.2:c.9879C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_014611.2:c.9879C>T']['hgvs_transcript_variant'] == 'NM_014611.2:c.9879C>T'
        assert results['NM_014611.2:c.9879C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_014611.2:c.9879C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000006.11:g.90403795G>A', 'vcf': {'chr': 'chr6', 'ref': u'G', 'pos': '90403795', 'alt': u'A'}}
        assert results['NM_014611.2:c.9879C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000006.12:g.89694076G>A', 'vcf': {'chr': 'chr6', 'ref': u'G', 'pos': '89694076', 'alt': u'A'}}
        assert results['NM_014611.2:c.9879C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000006.11:g.90403795G>A', 'vcf': {'chr': '6', 'ref': u'G', 'pos': '90403795', 'alt': u'A'}}
        assert results['NM_014611.2:c.9879C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000006.12:g.89694076G>A', 'vcf': {'chr': '6', 'ref': u'G', 'pos': '89694076', 'alt': u'A'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_014611.1:c.9879T=' in results.keys()
        assert results['NM_014611.1:c.9879T=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_014611.1:c.9879T=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_014611.1:c.9879T=']['alt_genomic_loci'] == []
        assert results['NM_014611.1:c.9879T=']['gene_symbol'] == 'MDN1'
        assert results['NM_014611.1:c.9879T=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_055426.1:p.(Val3293=)', 'slr': 'NP_055426.1:p.(V3293=)'}
        assert results['NM_014611.1:c.9879T=']['submitted_variant'] == '6-90403795-G-A'
        assert results['NM_014611.1:c.9879T=']['genome_context_intronic_sequence'] == ''
        assert results['NM_014611.1:c.9879T=']['hgvs_lrg_variant'] == ''
        assert results['NM_014611.1:c.9879T=']['hgvs_transcript_variant'] == 'NM_014611.1:c.9879T='
        assert results['NM_014611.1:c.9879T=']['hgvs_refseqgene_variant'] == ''
        assert results['NM_014611.1:c.9879T=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000006.11:g.90403795G>A', 'vcf': {'chr': 'chr6', 'ref': 'G', 'pos': '90403795', 'alt': u'A'}}
        assert 'hg38' not in results['NM_014611.1:c.9879T=']['primary_assembly_loci'].keys()
        assert results['NM_014611.1:c.9879T=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000006.11:g.90403795G>A', 'vcf': {'chr': '6', 'ref': 'G', 'pos': '90403795', 'alt': u'A'}}
        assert 'grch38' not in results['NM_014611.1:c.9879T=']['primary_assembly_loci'].keys()


    def test_variant158(self):
        variant = '6-32012992-CG-C'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_019105.6:c.10711del' in results.keys()
        assert results['NM_019105.6:c.10711del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_019105.6:c.10711del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_019105.6:c.10711del']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NT_113891.2:g.3483644del', 'vcf': {'chr': 'HSCHR6_MHC_COX_CTG1', 'ref': 'CG', 'pos': '3483643', 'alt': 'C'}}}, {'hg19': {'hgvs_genomic_description': 'NT_113891.2:g.3483644del', 'vcf': {'chr': 'chr6_cox_hap2', 'ref': 'CG', 'pos': '3483643', 'alt': 'C'}}}, {'grch38': {'hgvs_genomic_description': 'NT_113891.3:g.3483538del', 'vcf': {'chr': 'HSCHR6_MHC_COX_CTG1', 'ref': 'CG', 'pos': '3483537', 'alt': 'C'}}}, {'hg38': {'hgvs_genomic_description': 'NT_113891.3:g.3483538del', 'vcf': {'chr': 'chr6_GL000251v2_alt', 'ref': 'CG', 'pos': '3483537', 'alt': 'C'}}}, {'grch37': {'hgvs_genomic_description': 'NT_167245.1:g.3292210del', 'vcf': {'chr': 'HSCHR6_MHC_DBB_CTG1', 'ref': 'CG', 'pos': '3292209', 'alt': 'C'}}}, {'hg19': {'hgvs_genomic_description': 'NT_167245.1:g.3292210del', 'vcf': {'chr': 'chr6_dbb_hap3', 'ref': 'CG', 'pos': '3292209', 'alt': 'C'}}}, {'grch38': {'hgvs_genomic_description': 'NT_167245.2:g.3286625del', 'vcf': {'chr': 'HSCHR6_MHC_DBB_CTG1', 'ref': 'CG', 'pos': '3286624', 'alt': 'C'}}}, {'hg38': {'hgvs_genomic_description': 'NT_167245.2:g.3286625del', 'vcf': {'chr': 'chr6_GL000252v2_alt', 'ref': 'CG', 'pos': '3286624', 'alt': 'C'}}}, {'grch37': {'hgvs_genomic_description': 'NT_167247.1:g.3392834del', 'vcf': {'chr': 'HSCHR6_MHC_MCF_CTG1', 'ref': 'CG', 'pos': '3392833', 'alt': 'C'}}}, {'hg19': {'hgvs_genomic_description': 'NT_167247.1:g.3392834del', 'vcf': {'chr': 'chr6_mcf_hap5', 'ref': 'CG', 'pos': '3392833', 'alt': 'C'}}}, {'grch38': {'hgvs_genomic_description': 'NT_167247.2:g.3387249del', 'vcf': {'chr': 'HSCHR6_MHC_MCF_CTG1', 'ref': 'CG', 'pos': '3387248', 'alt': 'C'}}}, {'hg38': {'hgvs_genomic_description': 'NT_167247.2:g.3387249del', 'vcf': {'chr': 'chr6_GL000254v2_alt', 'ref': 'CG', 'pos': '3387248', 'alt': 'C'}}}, {'grch37': {'hgvs_genomic_description': 'NT_167248.1:g.3271861del', 'vcf': {'chr': 'HSCHR6_MHC_QBL_CTG1', 'ref': 'AG', 'pos': '3271858', 'alt': 'A'}}}, {'hg19': {'hgvs_genomic_description': 'NT_167248.1:g.3271861del', 'vcf': {'chr': 'chr6_qbl_hap6', 'ref': 'AG', 'pos': '3271858', 'alt': 'A'}}}]
        assert results['NM_019105.6:c.10711del']['gene_symbol'] == 'TNXB'
        assert results['NM_019105.6:c.10711del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_061978.6:p.(Arg3571AlafsTer91)', 'slr': 'NP_061978.6:p.(R3571Afs*91)'}
        assert results['NM_019105.6:c.10711del']['submitted_variant'] == '6-32012992-CG-C'
        assert results['NM_019105.6:c.10711del']['genome_context_intronic_sequence'] == ''
        assert results['NM_019105.6:c.10711del']['hgvs_lrg_variant'] == ''
        assert results['NM_019105.6:c.10711del']['hgvs_transcript_variant'] == 'NM_019105.6:c.10711del'
        assert results['NM_019105.6:c.10711del']['hgvs_refseqgene_variant'] == 'NG_008337.2:g.69159del'
        assert results['NM_019105.6:c.10711del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000006.11:g.32012993del', 'vcf': {'chr': 'chr6', 'ref': 'CG', 'pos': '32012992', 'alt': 'C'}}
        assert results['NM_019105.6:c.10711del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000006.12:g.32045216del', 'vcf': {'chr': 'chr6', 'ref': 'CG', 'pos': '32045215', 'alt': 'C'}}
        assert results['NM_019105.6:c.10711del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000006.11:g.32012993del', 'vcf': {'chr': '6', 'ref': 'CG', 'pos': '32012992', 'alt': 'C'}}
        assert results['NM_019105.6:c.10711del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000006.12:g.32045216del', 'vcf': {'chr': '6', 'ref': 'CG', 'pos': '32045215', 'alt': 'C'}}

        assert 'NM_032470.3:c.4del' in results.keys()
        assert results['NM_032470.3:c.4del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_032470.3:c.4del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_032470.3:c.4del']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NT_113891.2:g.3483644del', 'vcf': {'chr': 'HSCHR6_MHC_COX_CTG1', 'ref': 'CG', 'pos': '3483643', 'alt': 'C'}}}, {'hg19': {'hgvs_genomic_description': 'NT_113891.2:g.3483644del', 'vcf': {'chr': 'chr6_cox_hap2', 'ref': 'CG', 'pos': '3483643', 'alt': 'C'}}}, {'grch38': {'hgvs_genomic_description': 'NT_113891.3:g.3483538del', 'vcf': {'chr': 'HSCHR6_MHC_COX_CTG1', 'ref': 'CG', 'pos': '3483537', 'alt': 'C'}}}, {'hg38': {'hgvs_genomic_description': 'NT_113891.3:g.3483538del', 'vcf': {'chr': 'chr6_GL000251v2_alt', 'ref': 'CG', 'pos': '3483537', 'alt': 'C'}}}, {'grch37': {'hgvs_genomic_description': 'NT_167245.1:g.3292210del', 'vcf': {'chr': 'HSCHR6_MHC_DBB_CTG1', 'ref': 'CG', 'pos': '3292209', 'alt': 'C'}}}, {'hg19': {'hgvs_genomic_description': 'NT_167245.1:g.3292210del', 'vcf': {'chr': 'chr6_dbb_hap3', 'ref': 'CG', 'pos': '3292209', 'alt': 'C'}}}, {'grch38': {'hgvs_genomic_description': 'NT_167245.2:g.3286625del', 'vcf': {'chr': 'HSCHR6_MHC_DBB_CTG1', 'ref': 'CG', 'pos': '3286624', 'alt': 'C'}}}, {'hg38': {'hgvs_genomic_description': 'NT_167245.2:g.3286625del', 'vcf': {'chr': 'chr6_GL000252v2_alt', 'ref': 'CG', 'pos': '3286624', 'alt': 'C'}}}, {'grch37': {'hgvs_genomic_description': 'NT_167247.1:g.3392834del', 'vcf': {'chr': 'HSCHR6_MHC_MCF_CTG1', 'ref': 'CG', 'pos': '3392833', 'alt': 'C'}}}, {'hg19': {'hgvs_genomic_description': 'NT_167247.1:g.3392834del', 'vcf': {'chr': 'chr6_mcf_hap5', 'ref': 'CG', 'pos': '3392833', 'alt': 'C'}}}, {'grch38': {'hgvs_genomic_description': 'NT_167247.2:g.3387249del', 'vcf': {'chr': 'HSCHR6_MHC_MCF_CTG1', 'ref': 'CG', 'pos': '3387248', 'alt': 'C'}}}, {'hg38': {'hgvs_genomic_description': 'NT_167247.2:g.3387249del', 'vcf': {'chr': 'chr6_GL000254v2_alt', 'ref': 'CG', 'pos': '3387248', 'alt': 'C'}}}, {'grch37': {'hgvs_genomic_description': 'NT_167248.1:g.3274047del', 'vcf': {'chr': 'HSCHR6_MHC_QBL_CTG1', 'ref': 'CG', 'pos': '3274046', 'alt': 'C'}}}, {'hg19': {'hgvs_genomic_description': 'NT_167248.1:g.3274047del', 'vcf': {'chr': 'chr6_qbl_hap6', 'ref': 'CG', 'pos': '3274046', 'alt': 'C'}}}, {'grch38': {'hgvs_genomic_description': 'NT_167248.2:g.3268451del', 'vcf': {'chr': 'HSCHR6_MHC_QBL_CTG1', 'ref': 'CG', 'pos': '3268450', 'alt': 'C'}}}, {'hg38': {'hgvs_genomic_description': 'NT_167248.2:g.3268451del', 'vcf': {'chr': 'chr6_GL000255v2_alt', 'ref': 'CG', 'pos': '3268450', 'alt': 'C'}}}, {'grch37': {'hgvs_genomic_description': 'NT_167249.1:g.3345701del', 'vcf': {'chr': 'HSCHR6_MHC_SSTO_CTG1', 'ref': 'CG', 'pos': '3345700', 'alt': 'C'}}}, {'hg19': {'hgvs_genomic_description': 'NT_167249.1:g.3345701del', 'vcf': {'chr': 'chr6_ssto_hap7', 'ref': 'CG', 'pos': '3345700', 'alt': 'C'}}}, {'grch38': {'hgvs_genomic_description': 'NT_167249.2:g.3346403del', 'vcf': {'chr': 'HSCHR6_MHC_SSTO_CTG1', 'ref': 'CG', 'pos': '3346402', 'alt': 'C'}}}, {'hg38': {'hgvs_genomic_description': 'NT_167249.2:g.3346403del', 'vcf': {'chr': 'chr6_GL000256v2_alt', 'ref': 'CG', 'pos': '3346402', 'alt': 'C'}}}]
        assert results['NM_032470.3:c.4del']['gene_symbol'] == 'TNXB'
        assert results['NM_032470.3:c.4del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_115859.2:p.(Arg2AlafsTer91)', 'slr': 'NP_115859.2:p.(R2Afs*91)'}
        assert results['NM_032470.3:c.4del']['submitted_variant'] == '6-32012992-CG-C'
        assert results['NM_032470.3:c.4del']['genome_context_intronic_sequence'] == ''
        assert results['NM_032470.3:c.4del']['hgvs_lrg_variant'] == ''
        assert results['NM_032470.3:c.4del']['hgvs_transcript_variant'] == 'NM_032470.3:c.4del'
        assert results['NM_032470.3:c.4del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_032470.3:c.4del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000006.11:g.32012993del', 'vcf': {'chr': 'chr6', 'ref': 'CG', 'pos': '32012992', 'alt': 'C'}}
        assert results['NM_032470.3:c.4del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000006.12:g.32045216del', 'vcf': {'chr': 'chr6', 'ref': 'CG', 'pos': '32045215', 'alt': 'C'}}
        assert results['NM_032470.3:c.4del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000006.11:g.32012993del', 'vcf': {'chr': '6', 'ref': 'CG', 'pos': '32012992', 'alt': 'C'}}
        assert results['NM_032470.3:c.4del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000006.12:g.32045216del', 'vcf': {'chr': '6', 'ref': 'CG', 'pos': '32045215', 'alt': 'C'}}

        assert 'NM_001365276.1:c.10717del' in results.keys()
        assert results['NM_001365276.1:c.10717del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001365276.1:c.10717del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001365276.1:c.10717del']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NT_113891.2:g.3483644del', 'vcf': {'chr': 'HSCHR6_MHC_COX_CTG1', 'ref': 'CG', 'pos': '3483643', 'alt': 'C'}}}, {'hg19': {'hgvs_genomic_description': 'NT_113891.2:g.3483644del', 'vcf': {'chr': 'chr6_cox_hap2', 'ref': 'CG', 'pos': '3483643', 'alt': 'C'}}}, {'grch37': {'hgvs_genomic_description': 'NT_167245.1:g.3292210del', 'vcf': {'chr': 'HSCHR6_MHC_DBB_CTG1', 'ref': 'CG', 'pos': '3292209', 'alt': 'C'}}}, {'hg19': {'hgvs_genomic_description': 'NT_167245.1:g.3292210del', 'vcf': {'chr': 'chr6_dbb_hap3', 'ref': 'CG', 'pos': '3292209', 'alt': 'C'}}}, {'grch37': {'hgvs_genomic_description': 'NT_167247.1:g.3392834del', 'vcf': {'chr': 'HSCHR6_MHC_MCF_CTG1', 'ref': 'CG', 'pos': '3392833', 'alt': 'C'}}}, {'hg19': {'hgvs_genomic_description': 'NT_167247.1:g.3392834del', 'vcf': {'chr': 'chr6_mcf_hap5', 'ref': 'CG', 'pos': '3392833', 'alt': 'C'}}}]
        assert results['NM_001365276.1:c.10717del']['gene_symbol'] == 'TNXB'
        assert results['NM_001365276.1:c.10717del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001352205.1:p.(Arg3573AlafsTer91)', 'slr': 'NP_001352205.1:p.(R3573Afs*91)'}
        assert results['NM_001365276.1:c.10717del']['submitted_variant'] == '6-32012992-CG-C'
        assert results['NM_001365276.1:c.10717del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001365276.1:c.10717del']['hgvs_lrg_variant'] == ''
        assert results['NM_001365276.1:c.10717del']['hgvs_transcript_variant'] == 'NM_001365276.1:c.10717del'
        assert results['NM_001365276.1:c.10717del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001365276.1:c.10717del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000006.11:g.32012993del', 'vcf': {'chr': 'chr6', 'ref': 'CG', 'pos': '32012992', 'alt': 'C'}}
        assert 'hg38' not in results['NM_001365276.1:c.10717del']['primary_assembly_loci'].keys()
        assert results['NM_001365276.1:c.10717del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000006.11:g.32012993del', 'vcf': {'chr': '6', 'ref': 'CG', 'pos': '32012992', 'alt': 'C'}}
        assert 'grch38' not in results['NM_001365276.1:c.10717del']['primary_assembly_loci'].keys()

        assert 'NM_019105.7:c.10711del' in results.keys()
        assert results['NM_019105.7:c.10711del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_019105.7:c.10711del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_019105.7:c.10711del']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NT_113891.2:g.3483644del', 'vcf': {'chr': 'HSCHR6_MHC_COX_CTG1', 'ref': 'CG', 'pos': '3483643', 'alt': 'C'}}}, {'hg19': {'hgvs_genomic_description': 'NT_113891.2:g.3483644del', 'vcf': {'chr': 'chr6_cox_hap2', 'ref': 'CG', 'pos': '3483643', 'alt': 'C'}}}, {'grch37': {'hgvs_genomic_description': 'NT_167245.1:g.3292210del', 'vcf': {'chr': 'HSCHR6_MHC_DBB_CTG1', 'ref': 'CG', 'pos': '3292209', 'alt': 'C'}}}, {'hg19': {'hgvs_genomic_description': 'NT_167245.1:g.3292210del', 'vcf': {'chr': 'chr6_dbb_hap3', 'ref': 'CG', 'pos': '3292209', 'alt': 'C'}}}, {'grch37': {'hgvs_genomic_description': 'NT_167247.1:g.3392834del', 'vcf': {'chr': 'HSCHR6_MHC_MCF_CTG1', 'ref': 'CG', 'pos': '3392833', 'alt': 'C'}}}, {'hg19': {'hgvs_genomic_description': 'NT_167247.1:g.3392834del', 'vcf': {'chr': 'chr6_mcf_hap5', 'ref': 'CG', 'pos': '3392833', 'alt': 'C'}}}]
        assert results['NM_019105.7:c.10711del']['gene_symbol'] == 'TNXB'
        assert results['NM_019105.7:c.10711del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_061978.6:p.(Arg3571AlafsTer91)', 'slr': 'NP_061978.6:p.(R3571Afs*91)'}
        assert results['NM_019105.7:c.10711del']['submitted_variant'] == '6-32012992-CG-C'
        assert results['NM_019105.7:c.10711del']['genome_context_intronic_sequence'] == ''
        assert results['NM_019105.7:c.10711del']['hgvs_lrg_variant'] == ''
        assert results['NM_019105.7:c.10711del']['hgvs_transcript_variant'] == 'NM_019105.7:c.10711del'
        assert results['NM_019105.7:c.10711del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_019105.7:c.10711del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000006.11:g.32012993del', 'vcf': {'chr': 'chr6', 'ref': 'CG', 'pos': '32012992', 'alt': 'C'}}
        assert 'hg38' not in results['NM_019105.7:c.10711del']['primary_assembly_loci'].keys()
        assert results['NM_019105.7:c.10711del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000006.11:g.32012993del', 'vcf': {'chr': '6', 'ref': 'CG', 'pos': '32012992', 'alt': 'C'}}
        assert 'grch38' not in results['NM_019105.7:c.10711del']['primary_assembly_loci'].keys()


    def test_variant159(self):
        variant = '17-48275363-C-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000088.3:c.589G>T' in results.keys()
        assert results['NM_000088.3:c.589G>T']['hgvs_lrg_transcript_variant'] == 'LRG_1t1:c.589G>T'
        assert results['NM_000088.3:c.589G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000088.3:c.589G>T']['alt_genomic_loci'] == []
        assert results['NM_000088.3:c.589G>T']['gene_symbol'] == 'COL1A1'
        assert results['NM_000088.3:c.589G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000079.2(LRG_1p1):p.(Gly197Cys)', 'slr': 'NP_000079.2:p.(G197C)'}
        assert results['NM_000088.3:c.589G>T']['submitted_variant'] == '17-48275363-C-A'
        assert results['NM_000088.3:c.589G>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_000088.3:c.589G>T']['hgvs_lrg_variant'] == 'LRG_1:g.8638G>T'
        assert results['NM_000088.3:c.589G>T']['hgvs_transcript_variant'] == 'NM_000088.3:c.589G>T'
        assert results['NM_000088.3:c.589G>T']['hgvs_refseqgene_variant'] == 'NG_007400.1:g.8638G>T'
        assert results['NM_000088.3:c.589G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275363C>A', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '48275363', 'alt': u'A'}}
        assert results['NM_000088.3:c.589G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198002C>A', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '50198002', 'alt': u'A'}}
        assert results['NM_000088.3:c.589G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275363C>A', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '48275363', 'alt': u'A'}}
        assert results['NM_000088.3:c.589G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198002C>A', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '50198002', 'alt': u'A'}}


    def test_variant160(self):
        variant = '17-48275364-C-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000088.3:c.589-1G>T' in results.keys()
        assert results['NM_000088.3:c.589-1G>T']['hgvs_lrg_transcript_variant'] == 'LRG_1t1:c.589-1G>T'
        assert results['NM_000088.3:c.589-1G>T']['refseqgene_context_intronic_sequence'] == 'NG_007400.1(NM_000088.3):c.589-1G>T'
        assert results['NM_000088.3:c.589-1G>T']['alt_genomic_loci'] == []
        assert results['NM_000088.3:c.589-1G>T']['gene_symbol'] == 'COL1A1'
        assert results['NM_000088.3:c.589-1G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000079.2(LRG_1p1):p.?', 'slr': 'NP_000079.2:p.?'}
        assert results['NM_000088.3:c.589-1G>T']['submitted_variant'] == '17-48275364-C-A'
        assert results['NM_000088.3:c.589-1G>T']['genome_context_intronic_sequence'] == 'NC_000017.10(NM_000088.3):c.589-1G>T'
        assert results['NM_000088.3:c.589-1G>T']['hgvs_lrg_variant'] == 'LRG_1:g.8637G>T'
        assert results['NM_000088.3:c.589-1G>T']['hgvs_transcript_variant'] == 'NM_000088.3:c.589-1G>T'
        assert results['NM_000088.3:c.589-1G>T']['hgvs_refseqgene_variant'] == 'NG_007400.1:g.8637G>T'
        assert results['NM_000088.3:c.589-1G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275364C>A', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '48275364', 'alt': u'A'}}
        assert results['NM_000088.3:c.589-1G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198003C>A', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '50198003', 'alt': u'A'}}
        assert results['NM_000088.3:c.589-1G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275364C>A', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '48275364', 'alt': u'A'}}
        assert results['NM_000088.3:c.589-1G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198003C>A', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '50198003', 'alt': u'A'}}


    def test_variant161(self):
        variant = '17-48275359-GGA-TCC'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_000088.3:c.591_593inv' in results.keys()
        assert results['NM_000088.3:c.591_593inv']['hgvs_lrg_transcript_variant'] == 'LRG_1t1:c.591_593inv'
        assert results['NM_000088.3:c.591_593inv']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000088.3:c.591_593inv']['alt_genomic_loci'] == []
        assert results['NM_000088.3:c.591_593inv']['gene_symbol'] == 'COL1A1'
        assert results['NM_000088.3:c.591_593inv']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000079.2(LRG_1p1):p.(Pro198Asp)', 'slr': 'NP_000079.2:p.(P198D)'}
        assert results['NM_000088.3:c.591_593inv']['submitted_variant'] == '17-48275359-GGA-TCC'
        assert results['NM_000088.3:c.591_593inv']['genome_context_intronic_sequence'] == ''
        assert results['NM_000088.3:c.591_593inv']['hgvs_lrg_variant'] == 'LRG_1:g.8640_8642inv'
        assert results['NM_000088.3:c.591_593inv']['hgvs_transcript_variant'] == 'NM_000088.3:c.591_593inv'
        assert results['NM_000088.3:c.591_593inv']['hgvs_refseqgene_variant'] == 'NG_007400.1:g.8640_8642inv'
        assert results['NM_000088.3:c.591_593inv']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275359_48275361inv', 'vcf': {'chr': 'chr17', 'ref': 'GGA', 'pos': '48275359', 'alt': 'TCC'}}
        assert results['NM_000088.3:c.591_593inv']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50197998_50198000inv', 'vcf': {'chr': 'chr17', 'ref': 'GGA', 'pos': '50197998', 'alt': 'TCC'}}
        assert results['NM_000088.3:c.591_593inv']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275359_48275361inv', 'vcf': {'chr': '17', 'ref': 'GGA', 'pos': '48275359', 'alt': 'TCC'}}
        assert results['NM_000088.3:c.591_593inv']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50197998_50198000inv', 'vcf': {'chr': '17', 'ref': 'GGA', 'pos': '50197998', 'alt': 'TCC'}}

        assert results['flag'] == 'gene_variant'

    def test_variant162(self):
        variant = '7-94039128-CTTG-C'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000089.3:c.1035_1035+2del' in results.keys()
        assert results['NM_000089.3:c.1035_1035+2del']['hgvs_lrg_transcript_variant'] == 'LRG_2t1:c.1035_1035+2del'
        assert results['NM_000089.3:c.1035_1035+2del']['refseqgene_context_intronic_sequence'] == 'NG_007405.1(NM_000089.3):c.1035_1035+2del'
        assert results['NM_000089.3:c.1035_1035+2del']['alt_genomic_loci'] == []
        assert results['NM_000089.3:c.1035_1035+2del']['gene_symbol'] == 'COL1A2'
        assert results['NM_000089.3:c.1035_1035+2del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000080.2(LRG_2p1):p.(Val345del)', 'slr': 'NP_000080.2:p.(V345del)'}
        assert results['NM_000089.3:c.1035_1035+2del']['submitted_variant'] == '7-94039128-CTTG-C'
        assert results['NM_000089.3:c.1035_1035+2del']['genome_context_intronic_sequence'] == 'NC_000007.13(NM_000089.3):c.1035_1035+2del'
        assert results['NM_000089.3:c.1035_1035+2del']['hgvs_lrg_variant'] == 'LRG_2:g.20261_20263del'
        assert results['NM_000089.3:c.1035_1035+2del']['hgvs_transcript_variant'] == 'NM_000089.3:c.1035_1035+2del'
        assert results['NM_000089.3:c.1035_1035+2del']['hgvs_refseqgene_variant'] == 'NG_007405.1:g.20261_20263del'
        assert results['NM_000089.3:c.1035_1035+2del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.94039133_94039135del', 'vcf': {'chr': 'chr7', 'ref': 'CTTG', 'pos': '94039128', 'alt': 'C'}}
        assert results['NM_000089.3:c.1035_1035+2del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.94409821_94409823del', 'vcf': {'chr': 'chr7', 'ref': 'CTTG', 'pos': '94409816', 'alt': 'C'}}
        assert results['NM_000089.3:c.1035_1035+2del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.94039133_94039135del', 'vcf': {'chr': '7', 'ref': 'CTTG', 'pos': '94039128', 'alt': 'C'}}
        assert results['NM_000089.3:c.1035_1035+2del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.94409821_94409823del', 'vcf': {'chr': '7', 'ref': 'CTTG', 'pos': '94409816', 'alt': 'C'}}


    def test_variant163(self):
        variant = '9-135800972-AC-ACC'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001162426.1:c.363+1dup' in results.keys()
        assert results['NM_001162426.1:c.363+1dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001162426.1:c.363+1dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001162426.1:c.363+1dup']['alt_genomic_loci'] == []
        assert results['NM_001162426.1:c.363+1dup']['gene_symbol'] == 'TSC1'
        assert results['NM_001162426.1:c.363+1dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001155898.1:p.(Met122AspfsTer4)', 'slr': 'NP_001155898.1:p.(M122Dfs*4)'}
        assert results['NM_001162426.1:c.363+1dup']['submitted_variant'] == '9-135800972-AC-ACC'
        assert results['NM_001162426.1:c.363+1dup']['genome_context_intronic_sequence'] == 'NC_000009.11(NM_001162426.1):c.363+1dup'
        assert results['NM_001162426.1:c.363+1dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001162426.1:c.363+1dup']['hgvs_transcript_variant'] == 'NM_001162426.1:c.363+1dup'
        assert results['NM_001162426.1:c.363+1dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001162426.1:c.363+1dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.135800973dup', 'vcf': {'chr': 'chr9', 'ref': 'C', 'pos': '135800973', 'alt': 'CC'}}
        assert results['NM_001162426.1:c.363+1dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.132925586dup', 'vcf': {'chr': 'chr9', 'ref': 'C', 'pos': '132925586', 'alt': 'CC'}}
        assert results['NM_001162426.1:c.363+1dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.135800973dup', 'vcf': {'chr': '9', 'ref': 'C', 'pos': '135800973', 'alt': 'CC'}}
        assert results['NM_001162426.1:c.363+1dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.132925586dup', 'vcf': {'chr': '9', 'ref': 'C', 'pos': '132925586', 'alt': 'CC'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001362177.1:c.-1+1dup' in results.keys()
        assert results['NM_001362177.1:c.-1+1dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001362177.1:c.-1+1dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001362177.1:c.-1+1dup']['alt_genomic_loci'] == []
        assert results['NM_001362177.1:c.-1+1dup']['gene_symbol'] == 'TSC1'
        assert results['NM_001362177.1:c.-1+1dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001349106.1:p.?', 'slr': 'NP_001349106.1:p.?'}
        assert results['NM_001362177.1:c.-1+1dup']['submitted_variant'] == '9-135800972-AC-ACC'
        assert results['NM_001362177.1:c.-1+1dup']['genome_context_intronic_sequence'] == 'NC_000009.11(NM_001362177.1):c.-1+1dup'
        assert results['NM_001362177.1:c.-1+1dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001362177.1:c.-1+1dup']['hgvs_transcript_variant'] == 'NM_001362177.1:c.-1+1dup'
        assert results['NM_001362177.1:c.-1+1dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001362177.1:c.-1+1dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.135800973dup', 'vcf': {'chr': 'chr9', 'ref': 'C', 'pos': '135800973', 'alt': 'CC'}}
        assert results['NM_001362177.1:c.-1+1dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.132925586dup', 'vcf': {'chr': 'chr9', 'ref': 'C', 'pos': '132925586', 'alt': 'CC'}}
        assert results['NM_001362177.1:c.-1+1dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.135800973dup', 'vcf': {'chr': '9', 'ref': 'C', 'pos': '135800973', 'alt': 'CC'}}
        assert results['NM_001362177.1:c.-1+1dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.132925586dup', 'vcf': {'chr': '9', 'ref': 'C', 'pos': '132925586', 'alt': 'CC'}}

        assert 'NM_000368.4:c.363+1dup' in results.keys()
        assert results['NM_000368.4:c.363+1dup']['hgvs_lrg_transcript_variant'] == 'LRG_486t1:c.363+1dup'
        assert results['NM_000368.4:c.363+1dup']['refseqgene_context_intronic_sequence'] == 'NG_012386.1(NM_000368.4):c.363+1dup'
        assert results['NM_000368.4:c.363+1dup']['alt_genomic_loci'] == []
        assert results['NM_000368.4:c.363+1dup']['gene_symbol'] == 'TSC1'
        assert results['NM_000368.4:c.363+1dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000359.1(LRG_486p1):p.(Met122AspfsTer4)', 'slr': 'NP_000359.1:p.(M122Dfs*4)'}
        assert results['NM_000368.4:c.363+1dup']['submitted_variant'] == '9-135800972-AC-ACC'
        assert results['NM_000368.4:c.363+1dup']['genome_context_intronic_sequence'] == 'NC_000009.11(NM_000368.4):c.363+1dup'
        assert results['NM_000368.4:c.363+1dup']['hgvs_lrg_variant'] == 'LRG_486:g.24048dup'
        assert results['NM_000368.4:c.363+1dup']['hgvs_transcript_variant'] == 'NM_000368.4:c.363+1dup'
        assert results['NM_000368.4:c.363+1dup']['hgvs_refseqgene_variant'] == 'NG_012386.1:g.24048dup'
        assert results['NM_000368.4:c.363+1dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.135800973dup', 'vcf': {'chr': 'chr9', 'ref': 'C', 'pos': '135800973', 'alt': 'CC'}}
        assert results['NM_000368.4:c.363+1dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.132925586dup', 'vcf': {'chr': 'chr9', 'ref': 'C', 'pos': '132925586', 'alt': 'CC'}}
        assert results['NM_000368.4:c.363+1dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.135800973dup', 'vcf': {'chr': '9', 'ref': 'C', 'pos': '135800973', 'alt': 'CC'}}
        assert results['NM_000368.4:c.363+1dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.132925586dup', 'vcf': {'chr': '9', 'ref': 'C', 'pos': '132925586', 'alt': 'CC'}}

        assert 'NM_001162427.1:c.210+1615dup' in results.keys()
        assert results['NM_001162427.1:c.210+1615dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001162427.1:c.210+1615dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001162427.1:c.210+1615dup']['alt_genomic_loci'] == []
        assert results['NM_001162427.1:c.210+1615dup']['gene_symbol'] == 'TSC1'
        assert results['NM_001162427.1:c.210+1615dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001155899.1:p.?', 'slr': 'NP_001155899.1:p.?'}
        assert results['NM_001162427.1:c.210+1615dup']['submitted_variant'] == '9-135800972-AC-ACC'
        assert results['NM_001162427.1:c.210+1615dup']['genome_context_intronic_sequence'] == 'NC_000009.11(NM_001162427.1):c.210+1615dup'
        assert results['NM_001162427.1:c.210+1615dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001162427.1:c.210+1615dup']['hgvs_transcript_variant'] == 'NM_001162427.1:c.210+1615dup'
        assert results['NM_001162427.1:c.210+1615dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001162427.1:c.210+1615dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.135800973dup', 'vcf': {'chr': 'chr9', 'ref': 'C', 'pos': '135800973', 'alt': 'CC'}}
        assert results['NM_001162427.1:c.210+1615dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.132925586dup', 'vcf': {'chr': 'chr9', 'ref': 'C', 'pos': '132925586', 'alt': 'CC'}}
        assert results['NM_001162427.1:c.210+1615dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.135800973dup', 'vcf': {'chr': '9', 'ref': 'C', 'pos': '135800973', 'alt': 'CC'}}
        assert results['NM_001162427.1:c.210+1615dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.132925586dup', 'vcf': {'chr': '9', 'ref': 'C', 'pos': '132925586', 'alt': 'CC'}}


    def test_variant164(self):
        variant = '1-43212925-C-T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_001243246.1:c.2073G>A' in results.keys()
        assert results['NM_001243246.1:c.2073G>A']['hgvs_lrg_transcript_variant'] == 'LRG_5t3:c.2073G>A'
        assert results['NM_001243246.1:c.2073G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001243246.1:c.2073G>A']['alt_genomic_loci'] == []
        assert results['NM_001243246.1:c.2073G>A']['gene_symbol'] == 'P3H1'
        assert results['NM_001243246.1:c.2073G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001230175.1:p.(Ala691=)', 'slr': 'NP_001230175.1:p.(A691=)'}
        assert results['NM_001243246.1:c.2073G>A']['submitted_variant'] == '1-43212925-C-T'
        assert results['NM_001243246.1:c.2073G>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_001243246.1:c.2073G>A']['hgvs_lrg_variant'] == ''
        assert results['NM_001243246.1:c.2073G>A']['hgvs_transcript_variant'] == 'NM_001243246.1:c.2073G>A'
        assert results['NM_001243246.1:c.2073G>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001243246.1:c.2073G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.43212925C>T', 'vcf': {'chr': 'chr1', 'ref': u'C', 'pos': '43212925', 'alt': u'T'}}
        assert results['NM_001243246.1:c.2073G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.42747254C>T', 'vcf': {'chr': 'chr1', 'ref': u'C', 'pos': '42747254', 'alt': u'T'}}
        assert results['NM_001243246.1:c.2073G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.43212925C>T', 'vcf': {'chr': '1', 'ref': u'C', 'pos': '43212925', 'alt': u'T'}}
        assert results['NM_001243246.1:c.2073G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.42747254C>T', 'vcf': {'chr': '1', 'ref': u'C', 'pos': '42747254', 'alt': u'T'}}

        assert 'NM_001146289.1:c.2073G>A' in results.keys()
        assert results['NM_001146289.1:c.2073G>A']['hgvs_lrg_transcript_variant'] == 'LRG_5t2:c.2073G>A'
        assert results['NM_001146289.1:c.2073G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001146289.1:c.2073G>A']['alt_genomic_loci'] == []
        assert results['NM_001146289.1:c.2073G>A']['gene_symbol'] == 'P3H1'
        assert results['NM_001146289.1:c.2073G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001139761.1:p.(Ala691=)', 'slr': 'NP_001139761.1:p.(A691=)'}
        assert results['NM_001146289.1:c.2073G>A']['submitted_variant'] == '1-43212925-C-T'
        assert results['NM_001146289.1:c.2073G>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_001146289.1:c.2073G>A']['hgvs_lrg_variant'] == 'LRG_5:g.24831G>A'
        assert results['NM_001146289.1:c.2073G>A']['hgvs_transcript_variant'] == 'NM_001146289.1:c.2073G>A'
        assert results['NM_001146289.1:c.2073G>A']['hgvs_refseqgene_variant'] == 'NG_008123.1:g.24831G>A'
        assert results['NM_001146289.1:c.2073G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.43212925C>T', 'vcf': {'chr': 'chr1', 'ref': u'C', 'pos': '43212925', 'alt': u'T'}}
        assert results['NM_001146289.1:c.2073G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.42747254C>T', 'vcf': {'chr': 'chr1', 'ref': u'C', 'pos': '42747254', 'alt': u'T'}}
        assert results['NM_001146289.1:c.2073G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.43212925C>T', 'vcf': {'chr': '1', 'ref': u'C', 'pos': '43212925', 'alt': u'T'}}
        assert results['NM_001146289.1:c.2073G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.42747254C>T', 'vcf': {'chr': '1', 'ref': u'C', 'pos': '42747254', 'alt': u'T'}}

        assert 'NM_022356.3:c.2055+18G>A' in results.keys()
        assert results['NM_022356.3:c.2055+18G>A']['hgvs_lrg_transcript_variant'] == 'LRG_5t1:c.2055+18G>A'
        assert results['NM_022356.3:c.2055+18G>A']['refseqgene_context_intronic_sequence'] == 'NG_008123.1(NM_022356.3):c.2055+18G>A'
        assert results['NM_022356.3:c.2055+18G>A']['alt_genomic_loci'] == []
        assert results['NM_022356.3:c.2055+18G>A']['gene_symbol'] == 'P3H1'
        assert results['NM_022356.3:c.2055+18G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_071751.3(LRG_5p1):p.?', 'slr': 'NP_071751.3:p.?'}
        assert results['NM_022356.3:c.2055+18G>A']['submitted_variant'] == '1-43212925-C-T'
        assert results['NM_022356.3:c.2055+18G>A']['genome_context_intronic_sequence'] == 'NC_000001.10(NM_022356.3):c.2055+18G>A'
        assert results['NM_022356.3:c.2055+18G>A']['hgvs_lrg_variant'] == 'LRG_5:g.24831G>A'
        assert results['NM_022356.3:c.2055+18G>A']['hgvs_transcript_variant'] == 'NM_022356.3:c.2055+18G>A'
        assert results['NM_022356.3:c.2055+18G>A']['hgvs_refseqgene_variant'] == 'NG_008123.1:g.24831G>A'
        assert results['NM_022356.3:c.2055+18G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.43212925C>T', 'vcf': {'chr': 'chr1', 'ref': u'C', 'pos': '43212925', 'alt': u'T'}}
        assert results['NM_022356.3:c.2055+18G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.42747254C>T', 'vcf': {'chr': 'chr1', 'ref': u'C', 'pos': '42747254', 'alt': u'T'}}
        assert results['NM_022356.3:c.2055+18G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.43212925C>T', 'vcf': {'chr': '1', 'ref': u'C', 'pos': '43212925', 'alt': u'T'}}
        assert results['NM_022356.3:c.2055+18G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.42747254C>T', 'vcf': {'chr': '1', 'ref': u'C', 'pos': '42747254', 'alt': u'T'}}


    def test_variant165(self):
        variant = 'HG987_PATCH-355171-C-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001194958.2:c.20C>A' in results.keys()
        assert results['NM_001194958.2:c.20C>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001194958.2:c.20C>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001194958.2:c.20C>A']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003315950.2:g.355171C>A', 'vcf': {'chr': 'HG987_PATCH', 'ref': 'C', 'pos': '355171', 'alt': 'A'}}}]
        assert results['NM_001194958.2:c.20C>A']['gene_symbol'] == 'KCNJ18'
        assert results['NM_001194958.2:c.20C>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001181887.2:p.(Ala7Asp)', 'slr': 'NP_001181887.2:p.(A7D)'}
        assert results['NM_001194958.2:c.20C>A']['submitted_variant'] == 'HG987_PATCH-355171-C-A'
        assert results['NM_001194958.2:c.20C>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_001194958.2:c.20C>A']['hgvs_lrg_variant'] == ''
        assert results['NM_001194958.2:c.20C>A']['hgvs_transcript_variant'] == 'NM_001194958.2:c.20C>A'
        assert results['NM_001194958.2:c.20C>A']['hgvs_refseqgene_variant'] == 'NG_033093.1:g.15284C>A'
        assert 'hg19' not in results['NM_001194958.2:c.20C>A']['primary_assembly_loci'].keys()
        assert results['NM_001194958.2:c.20C>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.21702806C>A', 'vcf': {'chr': 'chr17', 'ref': 'C', 'pos': '21702806', 'alt': 'A'}}
        assert 'grch37' not in results['NM_001194958.2:c.20C>A']['primary_assembly_loci'].keys()
        assert results['NM_001194958.2:c.20C>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.21702806C>A', 'vcf': {'chr': '17', 'ref': 'C', 'pos': '21702806', 'alt': 'A'}}

        assert results['flag'] == 'gene_variant'

    def test_variant166(self):
        variant = '20-43252915-T-C'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_000022.3:c.534A>G' in results.keys()
        assert results['NM_000022.3:c.534A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000022.3:c.534A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000022.3:c.534A>G']['alt_genomic_loci'] == []
        assert results['NM_000022.3:c.534A>G']['gene_symbol'] == 'ADA'
        assert results['NM_000022.3:c.534A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000013.2(LRG_16p1):p.(Val178=)', 'slr': 'NP_000013.2:p.(V178=)'}
        assert results['NM_000022.3:c.534A>G']['submitted_variant'] == '20-43252915-T-C'
        assert results['NM_000022.3:c.534A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_000022.3:c.534A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_000022.3:c.534A>G']['hgvs_transcript_variant'] == 'NM_000022.3:c.534A>G'
        assert results['NM_000022.3:c.534A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_000022.3:c.534A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000020.10:g.43252915T>C', 'vcf': {'chr': 'chr20', 'ref': u'T', 'pos': '43252915', 'alt': u'C'}}
        assert results['NM_000022.3:c.534A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000020.11:g.44624274T>C', 'vcf': {'chr': 'chr20', 'ref': u'T', 'pos': '44624274', 'alt': u'C'}}
        assert results['NM_000022.3:c.534A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000020.10:g.43252915T>C', 'vcf': {'chr': '20', 'ref': u'T', 'pos': '43252915', 'alt': u'C'}}
        assert results['NM_000022.3:c.534A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000020.11:g.44624274T>C', 'vcf': {'chr': '20', 'ref': u'T', 'pos': '44624274', 'alt': u'C'}}

        assert 'NM_001322051.1:c.534A>G' in results.keys()
        assert results['NM_001322051.1:c.534A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001322051.1:c.534A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001322051.1:c.534A>G']['alt_genomic_loci'] == []
        assert results['NM_001322051.1:c.534A>G']['gene_symbol'] == 'ADA'
        assert results['NM_001322051.1:c.534A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001308980.1:p.(Val178=)', 'slr': 'NP_001308980.1:p.(V178=)'}
        assert results['NM_001322051.1:c.534A>G']['submitted_variant'] == '20-43252915-T-C'
        assert results['NM_001322051.1:c.534A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001322051.1:c.534A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001322051.1:c.534A>G']['hgvs_transcript_variant'] == 'NM_001322051.1:c.534A>G'
        assert results['NM_001322051.1:c.534A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001322051.1:c.534A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000020.10:g.43252915T>C', 'vcf': {'chr': 'chr20', 'ref': u'T', 'pos': '43252915', 'alt': u'C'}}
        assert results['NM_001322051.1:c.534A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000020.11:g.44624274T>C', 'vcf': {'chr': 'chr20', 'ref': u'T', 'pos': '44624274', 'alt': u'C'}}
        assert results['NM_001322051.1:c.534A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000020.10:g.43252915T>C', 'vcf': {'chr': '20', 'ref': u'T', 'pos': '43252915', 'alt': u'C'}}
        assert results['NM_001322051.1:c.534A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000020.11:g.44624274T>C', 'vcf': {'chr': '20', 'ref': u'T', 'pos': '44624274', 'alt': u'C'}}

        assert 'NM_000022.2:c.534A>G' in results.keys()
        assert results['NM_000022.2:c.534A>G']['hgvs_lrg_transcript_variant'] == 'LRG_16t1:c.534A>G'
        assert results['NM_000022.2:c.534A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000022.2:c.534A>G']['alt_genomic_loci'] == []
        assert results['NM_000022.2:c.534A>G']['gene_symbol'] == 'ADA'
        assert results['NM_000022.2:c.534A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000013.2(LRG_16p1):p.(Val178=)', 'slr': 'NP_000013.2:p.(V178=)'}
        assert results['NM_000022.2:c.534A>G']['submitted_variant'] == '20-43252915-T-C'
        assert results['NM_000022.2:c.534A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_000022.2:c.534A>G']['hgvs_lrg_variant'] == 'LRG_16:g.32462A>G'
        assert results['NM_000022.2:c.534A>G']['hgvs_transcript_variant'] == 'NM_000022.2:c.534A>G'
        assert results['NM_000022.2:c.534A>G']['hgvs_refseqgene_variant'] == 'NG_007385.1:g.32462A>G'
        assert results['NM_000022.2:c.534A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000020.10:g.43252915T>C', 'vcf': {'chr': 'chr20', 'ref': u'T', 'pos': '43252915', 'alt': u'C'}}
        assert 'hg38' not in results['NM_000022.2:c.534A>G']['primary_assembly_loci'].keys()
        assert results['NM_000022.2:c.534A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000020.10:g.43252915T>C', 'vcf': {'chr': '20', 'ref': u'T', 'pos': '43252915', 'alt': u'C'}}
        assert 'grch38' not in results['NM_000022.2:c.534A>G']['primary_assembly_loci'].keys()

        assert results['flag'] == 'gene_variant'
        assert 'NM_001322050.1:c.129A>G' in results.keys()
        assert results['NM_001322050.1:c.129A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001322050.1:c.129A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001322050.1:c.129A>G']['alt_genomic_loci'] == []
        assert results['NM_001322050.1:c.129A>G']['gene_symbol'] == 'ADA'
        assert results['NM_001322050.1:c.129A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001308979.1:p.(Val43=)', 'slr': 'NP_001308979.1:p.(V43=)'}
        assert results['NM_001322050.1:c.129A>G']['submitted_variant'] == '20-43252915-T-C'
        assert results['NM_001322050.1:c.129A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001322050.1:c.129A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001322050.1:c.129A>G']['hgvs_transcript_variant'] == 'NM_001322050.1:c.129A>G'
        assert results['NM_001322050.1:c.129A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001322050.1:c.129A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000020.10:g.43252915T>C', 'vcf': {'chr': 'chr20', 'ref': u'T', 'pos': '43252915', 'alt': u'C'}}
        assert results['NM_001322050.1:c.129A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000020.11:g.44624274T>C', 'vcf': {'chr': 'chr20', 'ref': u'T', 'pos': '44624274', 'alt': u'C'}}
        assert results['NM_001322050.1:c.129A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000020.10:g.43252915T>C', 'vcf': {'chr': '20', 'ref': u'T', 'pos': '43252915', 'alt': u'C'}}
        assert results['NM_001322050.1:c.129A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000020.11:g.44624274T>C', 'vcf': {'chr': '20', 'ref': u'T', 'pos': '44624274', 'alt': u'C'}}

        assert 'NR_136160.1:n.685A>G' in results.keys()
        assert results['NR_136160.1:n.685A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_136160.1:n.685A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_136160.1:n.685A>G']['alt_genomic_loci'] == []
        assert results['NR_136160.1:n.685A>G']['gene_symbol'] == 'ADA'
        assert results['NR_136160.1:n.685A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_136160.1:n.685A>G']['submitted_variant'] == '20-43252915-T-C'
        assert results['NR_136160.1:n.685A>G']['genome_context_intronic_sequence'] == ''
        assert results['NR_136160.1:n.685A>G']['hgvs_lrg_variant'] == ''
        assert results['NR_136160.1:n.685A>G']['hgvs_transcript_variant'] == 'NR_136160.1:n.685A>G'
        assert results['NR_136160.1:n.685A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NR_136160.1:n.685A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000020.10:g.43252915T>C', 'vcf': {'chr': 'chr20', 'ref': u'T', 'pos': '43252915', 'alt': u'C'}}
        assert results['NR_136160.1:n.685A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000020.11:g.44624274T>C', 'vcf': {'chr': 'chr20', 'ref': u'T', 'pos': '44624274', 'alt': u'C'}}
        assert results['NR_136160.1:n.685A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000020.10:g.43252915T>C', 'vcf': {'chr': '20', 'ref': u'T', 'pos': '43252915', 'alt': u'C'}}
        assert results['NR_136160.1:n.685A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000020.11:g.44624274T>C', 'vcf': {'chr': '20', 'ref': u'T', 'pos': '44624274', 'alt': u'C'}}


    def test_variant167(self):
        variant = '1-216219781-A-C'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_206933.2:c.6317C>G' in results.keys()
        assert results['NM_206933.2:c.6317C>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_206933.2:c.6317C>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_206933.2:c.6317C>G']['alt_genomic_loci'] == []
        assert results['NM_206933.2:c.6317C>G']['gene_symbol'] == 'USH2A'
        assert results['NM_206933.2:c.6317C>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_996816.2:p.(Thr2106Arg)', 'slr': 'NP_996816.2:p.(T2106R)'}
        assert results['NM_206933.2:c.6317C>G']['submitted_variant'] == '1-216219781-A-C'
        assert results['NM_206933.2:c.6317C>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_206933.2:c.6317C>G']['hgvs_lrg_variant'] == ''
        assert results['NM_206933.2:c.6317C>G']['hgvs_transcript_variant'] == 'NM_206933.2:c.6317C>G'
        assert results['NM_206933.2:c.6317C>G']['hgvs_refseqgene_variant'] == 'NG_009497.1:g.381958C>G'
        assert results['NM_206933.2:c.6317C>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.216219781A>C', 'vcf': {'chr': 'chr1', 'ref': 'A', 'pos': '216219781', 'alt': u'C'}}
        assert results['NM_206933.2:c.6317C>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.216046439A>C', 'vcf': {'chr': 'chr1', 'ref': 'A', 'pos': '216046439', 'alt': u'C'}}
        assert results['NM_206933.2:c.6317C>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.216219781A>C', 'vcf': {'chr': '1', 'ref': 'A', 'pos': '216219781', 'alt': u'C'}}
        assert results['NM_206933.2:c.6317C>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.216046439A>C', 'vcf': {'chr': '1', 'ref': 'A', 'pos': '216046439', 'alt': u'C'}}


    def test_variant168(self):
        variant = '2-209113113-G-A,C,T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_005896.3:c.394C>G' in results.keys()
        assert results['NM_005896.3:c.394C>G']['hgvs_lrg_transcript_variant'] == 'LRG_610t1:c.394C>G'
        assert results['NM_005896.3:c.394C>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_005896.3:c.394C>G']['alt_genomic_loci'] == []
        assert results['NM_005896.3:c.394C>G']['gene_symbol'] == 'IDH1'
        assert results['NM_005896.3:c.394C>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_005887.2(LRG_610p1):p.(Arg132Gly)', 'slr': 'NP_005887.2:p.(R132G)'}
        assert results['NM_005896.3:c.394C>G']['submitted_variant'] == '2-209113113-G-A,C,T'
        assert results['NM_005896.3:c.394C>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_005896.3:c.394C>G']['hgvs_lrg_variant'] == ''
        assert results['NM_005896.3:c.394C>G']['hgvs_transcript_variant'] == 'NM_005896.3:c.394C>G'
        assert results['NM_005896.3:c.394C>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_005896.3:c.394C>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.209113113G>C', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '209113113', 'alt': u'C'}}
        assert results['NM_005896.3:c.394C>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.208248389G>C', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '208248389', 'alt': u'C'}}
        assert results['NM_005896.3:c.394C>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.209113113G>C', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '209113113', 'alt': u'C'}}
        assert results['NM_005896.3:c.394C>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.208248389G>C', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '208248389', 'alt': u'C'}}

        assert 'NM_001282387.1:c.394C>G' in results.keys()
        assert results['NM_001282387.1:c.394C>G']['hgvs_lrg_transcript_variant'] == 'LRG_610t2:c.394C>G'
        assert results['NM_001282387.1:c.394C>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001282387.1:c.394C>G']['alt_genomic_loci'] == []
        assert results['NM_001282387.1:c.394C>G']['gene_symbol'] == 'IDH1'
        assert results['NM_001282387.1:c.394C>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001269316.1:p.(Arg132Gly)', 'slr': 'NP_001269316.1:p.(R132G)'}
        assert results['NM_001282387.1:c.394C>G']['submitted_variant'] == '2-209113113-G-A,C,T'
        assert results['NM_001282387.1:c.394C>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001282387.1:c.394C>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001282387.1:c.394C>G']['hgvs_transcript_variant'] == 'NM_001282387.1:c.394C>G'
        assert results['NM_001282387.1:c.394C>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001282387.1:c.394C>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.209113113G>C', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '209113113', 'alt': u'C'}}
        assert results['NM_001282387.1:c.394C>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.208248389G>C', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '208248389', 'alt': u'C'}}
        assert results['NM_001282387.1:c.394C>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.209113113G>C', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '209113113', 'alt': u'C'}}
        assert results['NM_001282387.1:c.394C>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.208248389G>C', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '208248389', 'alt': u'C'}}

        assert 'NM_001282387.1:c.394C>A' in results.keys()
        assert results['NM_001282387.1:c.394C>A']['hgvs_lrg_transcript_variant'] == 'LRG_610t2:c.394C>A'
        assert results['NM_001282387.1:c.394C>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001282387.1:c.394C>A']['alt_genomic_loci'] == []
        assert results['NM_001282387.1:c.394C>A']['gene_symbol'] == 'IDH1'
        assert results['NM_001282387.1:c.394C>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001269316.1:p.(Arg132Ser)', 'slr': 'NP_001269316.1:p.(R132S)'}
        assert results['NM_001282387.1:c.394C>A']['submitted_variant'] == '2-209113113-G-A,C,T'
        assert results['NM_001282387.1:c.394C>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_001282387.1:c.394C>A']['hgvs_lrg_variant'] == ''
        assert results['NM_001282387.1:c.394C>A']['hgvs_transcript_variant'] == 'NM_001282387.1:c.394C>A'
        assert results['NM_001282387.1:c.394C>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001282387.1:c.394C>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.209113113G>T', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '209113113', 'alt': u'T'}}
        assert results['NM_001282387.1:c.394C>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.208248389G>T', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '208248389', 'alt': u'T'}}
        assert results['NM_001282387.1:c.394C>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.209113113G>T', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '209113113', 'alt': u'T'}}
        assert results['NM_001282387.1:c.394C>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.208248389G>T', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '208248389', 'alt': u'T'}}

        assert 'NM_005896.3:c.394C>A' in results.keys()
        assert results['NM_005896.3:c.394C>A']['hgvs_lrg_transcript_variant'] == 'LRG_610t1:c.394C>A'
        assert results['NM_005896.3:c.394C>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_005896.3:c.394C>A']['alt_genomic_loci'] == []
        assert results['NM_005896.3:c.394C>A']['gene_symbol'] == 'IDH1'
        assert results['NM_005896.3:c.394C>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_005887.2(LRG_610p1):p.(Arg132Ser)', 'slr': 'NP_005887.2:p.(R132S)'}
        assert results['NM_005896.3:c.394C>A']['submitted_variant'] == '2-209113113-G-A,C,T'
        assert results['NM_005896.3:c.394C>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_005896.3:c.394C>A']['hgvs_lrg_variant'] == ''
        assert results['NM_005896.3:c.394C>A']['hgvs_transcript_variant'] == 'NM_005896.3:c.394C>A'
        assert results['NM_005896.3:c.394C>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_005896.3:c.394C>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.209113113G>T', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '209113113', 'alt': u'T'}}
        assert results['NM_005896.3:c.394C>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.208248389G>T', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '208248389', 'alt': u'T'}}
        assert results['NM_005896.3:c.394C>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.209113113G>T', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '209113113', 'alt': u'T'}}
        assert results['NM_005896.3:c.394C>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.208248389G>T', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '208248389', 'alt': u'T'}}

        assert 'NM_001282386.1:c.394C>T' in results.keys()
        assert results['NM_001282386.1:c.394C>T']['hgvs_lrg_transcript_variant'] == 'LRG_610t3:c.394C>T'
        assert results['NM_001282386.1:c.394C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001282386.1:c.394C>T']['alt_genomic_loci'] == []
        assert results['NM_001282386.1:c.394C>T']['gene_symbol'] == 'IDH1'
        assert results['NM_001282386.1:c.394C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001269315.1:p.(Arg132Cys)', 'slr': 'NP_001269315.1:p.(R132C)'}
        assert results['NM_001282386.1:c.394C>T']['submitted_variant'] == '2-209113113-G-A,C,T'
        assert results['NM_001282386.1:c.394C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001282386.1:c.394C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001282386.1:c.394C>T']['hgvs_transcript_variant'] == 'NM_001282386.1:c.394C>T'
        assert results['NM_001282386.1:c.394C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001282386.1:c.394C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.209113113G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '209113113', 'alt': u'A'}}
        assert results['NM_001282386.1:c.394C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.208248389G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '208248389', 'alt': u'A'}}
        assert results['NM_001282386.1:c.394C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.209113113G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '209113113', 'alt': u'A'}}
        assert results['NM_001282386.1:c.394C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.208248389G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '208248389', 'alt': u'A'}}

        assert 'NM_005896.2:c.394C>A' in results.keys()
        assert results['NM_005896.2:c.394C>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_005896.2:c.394C>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_005896.2:c.394C>A']['alt_genomic_loci'] == []
        assert results['NM_005896.2:c.394C>A']['gene_symbol'] == 'IDH1'
        assert results['NM_005896.2:c.394C>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_005887.2(LRG_610p1):p.(Arg132Ser)', 'slr': 'NP_005887.2:p.(R132S)'}
        assert results['NM_005896.2:c.394C>A']['submitted_variant'] == '2-209113113-G-A,C,T'
        assert results['NM_005896.2:c.394C>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_005896.2:c.394C>A']['hgvs_lrg_variant'] == 'LRG_610:g.22686C>A'
        assert results['NM_005896.2:c.394C>A']['hgvs_transcript_variant'] == 'NM_005896.2:c.394C>A'
        assert results['NM_005896.2:c.394C>A']['hgvs_refseqgene_variant'] == 'NG_023319.2:g.22686C>A'
        assert results['NM_005896.2:c.394C>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.209113113G>T', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '209113113', 'alt': u'T'}}
        assert 'hg38' not in results['NM_005896.2:c.394C>A']['primary_assembly_loci'].keys()
        assert results['NM_005896.2:c.394C>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.209113113G>T', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '209113113', 'alt': u'T'}}
        assert 'grch38' not in results['NM_005896.2:c.394C>A']['primary_assembly_loci'].keys()

        assert 'NM_005896.2:c.394C>G' in results.keys()
        assert results['NM_005896.2:c.394C>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_005896.2:c.394C>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_005896.2:c.394C>G']['alt_genomic_loci'] == []
        assert results['NM_005896.2:c.394C>G']['gene_symbol'] == 'IDH1'
        assert results['NM_005896.2:c.394C>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_005887.2(LRG_610p1):p.(Arg132Gly)', 'slr': 'NP_005887.2:p.(R132G)'}
        assert results['NM_005896.2:c.394C>G']['submitted_variant'] == '2-209113113-G-A,C,T'
        assert results['NM_005896.2:c.394C>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_005896.2:c.394C>G']['hgvs_lrg_variant'] == 'LRG_610:g.22686C>G'
        assert results['NM_005896.2:c.394C>G']['hgvs_transcript_variant'] == 'NM_005896.2:c.394C>G'
        assert results['NM_005896.2:c.394C>G']['hgvs_refseqgene_variant'] == 'NG_023319.2:g.22686C>G'
        assert results['NM_005896.2:c.394C>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.209113113G>C', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '209113113', 'alt': u'C'}}
        assert 'hg38' not in results['NM_005896.2:c.394C>G']['primary_assembly_loci'].keys()
        assert results['NM_005896.2:c.394C>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.209113113G>C', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '209113113', 'alt': u'C'}}
        assert 'grch38' not in results['NM_005896.2:c.394C>G']['primary_assembly_loci'].keys()

        assert results['flag'] == 'gene_variant'
        assert 'NM_005896.3:c.394C>T' in results.keys()
        assert results['NM_005896.3:c.394C>T']['hgvs_lrg_transcript_variant'] == 'LRG_610t1:c.394C>T'
        assert results['NM_005896.3:c.394C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_005896.3:c.394C>T']['alt_genomic_loci'] == []
        assert results['NM_005896.3:c.394C>T']['gene_symbol'] == 'IDH1'
        assert results['NM_005896.3:c.394C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_005887.2(LRG_610p1):p.(Arg132Cys)', 'slr': 'NP_005887.2:p.(R132C)'}
        assert results['NM_005896.3:c.394C>T']['submitted_variant'] == '2-209113113-G-A,C,T'
        assert results['NM_005896.3:c.394C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_005896.3:c.394C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_005896.3:c.394C>T']['hgvs_transcript_variant'] == 'NM_005896.3:c.394C>T'
        assert results['NM_005896.3:c.394C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_005896.3:c.394C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.209113113G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '209113113', 'alt': u'A'}}
        assert results['NM_005896.3:c.394C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.208248389G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '208248389', 'alt': u'A'}}
        assert results['NM_005896.3:c.394C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.209113113G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '209113113', 'alt': u'A'}}
        assert results['NM_005896.3:c.394C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.208248389G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '208248389', 'alt': u'A'}}

        assert 'NM_001282387.1:c.394C>T' in results.keys()
        assert results['NM_001282387.1:c.394C>T']['hgvs_lrg_transcript_variant'] == 'LRG_610t2:c.394C>T'
        assert results['NM_001282387.1:c.394C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001282387.1:c.394C>T']['alt_genomic_loci'] == []
        assert results['NM_001282387.1:c.394C>T']['gene_symbol'] == 'IDH1'
        assert results['NM_001282387.1:c.394C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001269316.1:p.(Arg132Cys)', 'slr': 'NP_001269316.1:p.(R132C)'}
        assert results['NM_001282387.1:c.394C>T']['submitted_variant'] == '2-209113113-G-A,C,T'
        assert results['NM_001282387.1:c.394C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001282387.1:c.394C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001282387.1:c.394C>T']['hgvs_transcript_variant'] == 'NM_001282387.1:c.394C>T'
        assert results['NM_001282387.1:c.394C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001282387.1:c.394C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.209113113G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '209113113', 'alt': u'A'}}
        assert results['NM_001282387.1:c.394C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.208248389G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '208248389', 'alt': u'A'}}
        assert results['NM_001282387.1:c.394C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.209113113G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '209113113', 'alt': u'A'}}
        assert results['NM_001282387.1:c.394C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.208248389G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '208248389', 'alt': u'A'}}

        assert 'NM_001282386.1:c.394C>G' in results.keys()
        assert results['NM_001282386.1:c.394C>G']['hgvs_lrg_transcript_variant'] == 'LRG_610t3:c.394C>G'
        assert results['NM_001282386.1:c.394C>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001282386.1:c.394C>G']['alt_genomic_loci'] == []
        assert results['NM_001282386.1:c.394C>G']['gene_symbol'] == 'IDH1'
        assert results['NM_001282386.1:c.394C>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001269315.1:p.(Arg132Gly)', 'slr': 'NP_001269315.1:p.(R132G)'}
        assert results['NM_001282386.1:c.394C>G']['submitted_variant'] == '2-209113113-G-A,C,T'
        assert results['NM_001282386.1:c.394C>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001282386.1:c.394C>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001282386.1:c.394C>G']['hgvs_transcript_variant'] == 'NM_001282386.1:c.394C>G'
        assert results['NM_001282386.1:c.394C>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001282386.1:c.394C>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.209113113G>C', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '209113113', 'alt': u'C'}}
        assert results['NM_001282386.1:c.394C>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.208248389G>C', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '208248389', 'alt': u'C'}}
        assert results['NM_001282386.1:c.394C>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.209113113G>C', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '209113113', 'alt': u'C'}}
        assert results['NM_001282386.1:c.394C>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.208248389G>C', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '208248389', 'alt': u'C'}}

        assert 'NM_005896.2:c.394C>T' in results.keys()
        assert results['NM_005896.2:c.394C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_005896.2:c.394C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_005896.2:c.394C>T']['alt_genomic_loci'] == []
        assert results['NM_005896.2:c.394C>T']['gene_symbol'] == 'IDH1'
        assert results['NM_005896.2:c.394C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_005887.2(LRG_610p1):p.(Arg132Cys)', 'slr': 'NP_005887.2:p.(R132C)'}
        assert results['NM_005896.2:c.394C>T']['submitted_variant'] == '2-209113113-G-A,C,T'
        assert results['NM_005896.2:c.394C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_005896.2:c.394C>T']['hgvs_lrg_variant'] == 'LRG_610:g.22686C>T'
        assert results['NM_005896.2:c.394C>T']['hgvs_transcript_variant'] == 'NM_005896.2:c.394C>T'
        assert results['NM_005896.2:c.394C>T']['hgvs_refseqgene_variant'] == 'NG_023319.2:g.22686C>T'
        assert results['NM_005896.2:c.394C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.209113113G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '209113113', 'alt': u'A'}}
        assert 'hg38' not in results['NM_005896.2:c.394C>T']['primary_assembly_loci'].keys()
        assert results['NM_005896.2:c.394C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.209113113G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '209113113', 'alt': u'A'}}
        assert 'grch38' not in results['NM_005896.2:c.394C>T']['primary_assembly_loci'].keys()

        assert 'NM_001282386.1:c.394C>A' in results.keys()
        assert results['NM_001282386.1:c.394C>A']['hgvs_lrg_transcript_variant'] == 'LRG_610t3:c.394C>A'
        assert results['NM_001282386.1:c.394C>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001282386.1:c.394C>A']['alt_genomic_loci'] == []
        assert results['NM_001282386.1:c.394C>A']['gene_symbol'] == 'IDH1'
        assert results['NM_001282386.1:c.394C>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001269315.1:p.(Arg132Ser)', 'slr': 'NP_001269315.1:p.(R132S)'}
        assert results['NM_001282386.1:c.394C>A']['submitted_variant'] == '2-209113113-G-A,C,T'
        assert results['NM_001282386.1:c.394C>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_001282386.1:c.394C>A']['hgvs_lrg_variant'] == ''
        assert results['NM_001282386.1:c.394C>A']['hgvs_transcript_variant'] == 'NM_001282386.1:c.394C>A'
        assert results['NM_001282386.1:c.394C>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001282386.1:c.394C>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.209113113G>T', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '209113113', 'alt': u'T'}}
        assert results['NM_001282386.1:c.394C>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.208248389G>T', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '208248389', 'alt': u'T'}}
        assert results['NM_001282386.1:c.394C>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.209113113G>T', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '209113113', 'alt': u'T'}}
        assert results['NM_001282386.1:c.394C>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.208248389G>T', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '208248389', 'alt': u'T'}}


    def test_variant169(self):
        variant = 'NC_000005.9:g.35058665_35058666CA='
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_000949.5:c.*6525_*6526=' in results.keys()
        assert results['NM_000949.5:c.*6525_*6526=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000949.5:c.*6525_*6526=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000949.5:c.*6525_*6526=']['alt_genomic_loci'] == []
        assert results['NM_000949.5:c.*6525_*6526=']['gene_symbol'] == 'PRLR'
        assert results['NM_000949.5:c.*6525_*6526=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000940.1:p.?', 'slr': 'NP_000940.1:p.?'}
        assert results['NM_000949.5:c.*6525_*6526=']['submitted_variant'] == 'NC_000005.9:g.35058665_35058666CA='
        assert results['NM_000949.5:c.*6525_*6526=']['genome_context_intronic_sequence'] == ''
        assert results['NM_000949.5:c.*6525_*6526=']['hgvs_lrg_variant'] == ''
        assert results['NM_000949.5:c.*6525_*6526=']['hgvs_transcript_variant'] == 'NM_000949.5:c.*6525_*6526='
        assert results['NM_000949.5:c.*6525_*6526=']['hgvs_refseqgene_variant'] == 'NG_029042.1:g.177158_177159='
        assert results['NM_000949.5:c.*6525_*6526=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.35058665_35058666=', 'vcf': {'chr': 'chr5', 'ref': u'CA', 'pos': '35058665', 'alt': u'CA'}}
        assert 'hg38' not in results['NM_000949.5:c.*6525_*6526=']['primary_assembly_loci'].keys()
        assert results['NM_000949.5:c.*6525_*6526=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.35058665_35058666=', 'vcf': {'chr': '5', 'ref': u'CA', 'pos': '35058665', 'alt': u'CA'}}
        assert 'grch38' not in results['NM_000949.5:c.*6525_*6526=']['primary_assembly_loci'].keys()

        assert 'NM_001204314.1:c.*6525_*6526=' in results.keys()
        assert results['NM_001204314.1:c.*6525_*6526=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001204314.1:c.*6525_*6526=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001204314.1:c.*6525_*6526=']['alt_genomic_loci'] == []
        assert results['NM_001204314.1:c.*6525_*6526=']['gene_symbol'] == 'PRLR'
        assert results['NM_001204314.1:c.*6525_*6526=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001191243.1:p.?', 'slr': 'NP_001191243.1:p.?'}
        assert results['NM_001204314.1:c.*6525_*6526=']['submitted_variant'] == 'NC_000005.9:g.35058665_35058666CA='
        assert results['NM_001204314.1:c.*6525_*6526=']['genome_context_intronic_sequence'] == ''
        assert results['NM_001204314.1:c.*6525_*6526=']['hgvs_lrg_variant'] == ''
        assert results['NM_001204314.1:c.*6525_*6526=']['hgvs_transcript_variant'] == 'NM_001204314.1:c.*6525_*6526='
        assert results['NM_001204314.1:c.*6525_*6526=']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001204314.1:c.*6525_*6526=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.35058665_35058666=', 'vcf': {'chr': 'chr5', 'ref': u'CA', 'pos': '35058665', 'alt': u'CA'}}
        assert 'hg38' not in results['NM_001204314.1:c.*6525_*6526=']['primary_assembly_loci'].keys()
        assert results['NM_001204314.1:c.*6525_*6526=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.35058665_35058666=', 'vcf': {'chr': '5', 'ref': u'CA', 'pos': '35058665', 'alt': u'CA'}}
        assert 'grch38' not in results['NM_001204314.1:c.*6525_*6526=']['primary_assembly_loci'].keys()

        assert 'NM_001204314.2:c.*6528del' in results.keys()
        assert results['NM_001204314.2:c.*6528del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001204314.2:c.*6528del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001204314.2:c.*6528del']['alt_genomic_loci'] == []
        assert results['NM_001204314.2:c.*6528del']['gene_symbol'] == 'PRLR'
        assert results['NM_001204314.2:c.*6528del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001191243.1:p.?', 'slr': 'NP_001191243.1:p.?'}
        assert results['NM_001204314.2:c.*6528del']['submitted_variant'] == 'NC_000005.9:g.35058665_35058666CA='
        assert results['NM_001204314.2:c.*6528del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001204314.2:c.*6528del']['hgvs_lrg_variant'] == ''
        assert results['NM_001204314.2:c.*6528del']['hgvs_transcript_variant'] == 'NM_001204314.2:c.*6528del'
        assert results['NM_001204314.2:c.*6528del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001204314.2:c.*6528del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.35058662_35058668=', 'vcf': {'chr': 'chr5', 'ref': 'AGACAAG', 'pos': '35058662', 'alt': 'AGACAAG'}}
        assert results['NM_001204314.2:c.*6528del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000005.10:g.35058563del', 'vcf': {'chr': 'chr5', 'ref': 'CA', 'pos': '35058560', 'alt': 'C'}}
        assert results['NM_001204314.2:c.*6528del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.35058662_35058668=', 'vcf': {'chr': '5', 'ref': 'AGACAAG', 'pos': '35058662', 'alt': 'AGACAAG'}}
        assert results['NM_001204314.2:c.*6528del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000005.10:g.35058563del', 'vcf': {'chr': '5', 'ref': 'CA', 'pos': '35058560', 'alt': 'C'}}

        assert 'NM_001204317.1:c.856-9153_856-9152=' in results.keys()
        assert results['NM_001204317.1:c.856-9153_856-9152=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001204317.1:c.856-9153_856-9152=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001204317.1:c.856-9153_856-9152=']['alt_genomic_loci'] == []
        assert results['NM_001204317.1:c.856-9153_856-9152=']['gene_symbol'] == 'PRLR'
        assert results['NM_001204317.1:c.856-9153_856-9152=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001191246.1:p.?', 'slr': 'NP_001191246.1:p.?'}
        assert results['NM_001204317.1:c.856-9153_856-9152=']['submitted_variant'] == 'NC_000005.9:g.35058665_35058666CA='
        assert results['NM_001204317.1:c.856-9153_856-9152=']['genome_context_intronic_sequence'] == 'NC_000005.9(NM_001204317.1):c.856-9153_856-9152='
        assert results['NM_001204317.1:c.856-9153_856-9152=']['hgvs_lrg_variant'] == ''
        assert results['NM_001204317.1:c.856-9153_856-9152=']['hgvs_transcript_variant'] == 'NM_001204317.1:c.856-9153_856-9152='
        assert results['NM_001204317.1:c.856-9153_856-9152=']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001204317.1:c.856-9153_856-9152=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.35058665_35058666=', 'vcf': {'chr': 'chr5', 'ref': 'CA', 'pos': '35058665', 'alt': 'CA'}}
        assert results['NM_001204317.1:c.856-9153_856-9152=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000005.10:g.35058560_35058561=', 'vcf': {'chr': 'chr5', 'ref': 'CA', 'pos': '35058560', 'alt': 'CA'}}
        assert results['NM_001204317.1:c.856-9153_856-9152=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.35058665_35058666=', 'vcf': {'chr': '5', 'ref': 'CA', 'pos': '35058665', 'alt': 'CA'}}
        assert results['NM_001204317.1:c.856-9153_856-9152=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000005.10:g.35058560_35058561=', 'vcf': {'chr': '5', 'ref': 'CA', 'pos': '35058560', 'alt': 'CA'}}

        assert 'NR_037910.1:n.828-9153_828-9152=' in results.keys()
        assert results['NR_037910.1:n.828-9153_828-9152=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_037910.1:n.828-9153_828-9152=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_037910.1:n.828-9153_828-9152=']['alt_genomic_loci'] == []
        assert results['NR_037910.1:n.828-9153_828-9152=']['gene_symbol'] == 'PRLR'
        assert results['NR_037910.1:n.828-9153_828-9152=']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_037910.1:n.828-9153_828-9152=']['submitted_variant'] == 'NC_000005.9:g.35058665_35058666CA='
        assert results['NR_037910.1:n.828-9153_828-9152=']['genome_context_intronic_sequence'] == 'NC_000005.9(NR_037910.1):c.828-9153_828-9152='
        assert results['NR_037910.1:n.828-9153_828-9152=']['hgvs_lrg_variant'] == ''
        assert results['NR_037910.1:n.828-9153_828-9152=']['hgvs_transcript_variant'] == 'NR_037910.1:n.828-9153_828-9152='
        assert results['NR_037910.1:n.828-9153_828-9152=']['hgvs_refseqgene_variant'] == ''
        assert results['NR_037910.1:n.828-9153_828-9152=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.35058665_35058666=', 'vcf': {'chr': 'chr5', 'ref': 'CA', 'pos': '35058665', 'alt': 'CA'}}
        assert results['NR_037910.1:n.828-9153_828-9152=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000005.10:g.35058560_35058561=', 'vcf': {'chr': 'chr5', 'ref': 'CA', 'pos': '35058560', 'alt': 'CA'}}
        assert results['NR_037910.1:n.828-9153_828-9152=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.35058665_35058666=', 'vcf': {'chr': '5', 'ref': 'CA', 'pos': '35058665', 'alt': 'CA'}}
        assert results['NR_037910.1:n.828-9153_828-9152=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000005.10:g.35058560_35058561=', 'vcf': {'chr': '5', 'ref': 'CA', 'pos': '35058560', 'alt': 'CA'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001204318.1:c.686-9153_686-9152=' in results.keys()
        assert results['NM_001204318.1:c.686-9153_686-9152=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001204318.1:c.686-9153_686-9152=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001204318.1:c.686-9153_686-9152=']['alt_genomic_loci'] == []
        assert results['NM_001204318.1:c.686-9153_686-9152=']['gene_symbol'] == 'PRLR'
        assert results['NM_001204318.1:c.686-9153_686-9152=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001191247.1:p.?', 'slr': 'NP_001191247.1:p.?'}
        assert results['NM_001204318.1:c.686-9153_686-9152=']['submitted_variant'] == 'NC_000005.9:g.35058665_35058666CA='
        assert results['NM_001204318.1:c.686-9153_686-9152=']['genome_context_intronic_sequence'] == 'NC_000005.9(NM_001204318.1):c.686-9153_686-9152='
        assert results['NM_001204318.1:c.686-9153_686-9152=']['hgvs_lrg_variant'] == ''
        assert results['NM_001204318.1:c.686-9153_686-9152=']['hgvs_transcript_variant'] == 'NM_001204318.1:c.686-9153_686-9152='
        assert results['NM_001204318.1:c.686-9153_686-9152=']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001204318.1:c.686-9153_686-9152=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.35058665_35058666=', 'vcf': {'chr': 'chr5', 'ref': 'CA', 'pos': '35058665', 'alt': 'CA'}}
        assert results['NM_001204318.1:c.686-9153_686-9152=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000005.10:g.35058560_35058561=', 'vcf': {'chr': 'chr5', 'ref': 'CA', 'pos': '35058560', 'alt': 'CA'}}
        assert results['NM_001204318.1:c.686-9153_686-9152=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.35058665_35058666=', 'vcf': {'chr': '5', 'ref': 'CA', 'pos': '35058665', 'alt': 'CA'}}
        assert results['NM_001204318.1:c.686-9153_686-9152=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000005.10:g.35058560_35058561=', 'vcf': {'chr': '5', 'ref': 'CA', 'pos': '35058560', 'alt': 'CA'}}

        assert 'NM_001204316.1:c.1009+7385_1009+7386=' in results.keys()
        assert results['NM_001204316.1:c.1009+7385_1009+7386=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001204316.1:c.1009+7385_1009+7386=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001204316.1:c.1009+7385_1009+7386=']['alt_genomic_loci'] == []
        assert results['NM_001204316.1:c.1009+7385_1009+7386=']['gene_symbol'] == 'PRLR'
        assert results['NM_001204316.1:c.1009+7385_1009+7386=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001191245.1:p.?', 'slr': 'NP_001191245.1:p.?'}
        assert results['NM_001204316.1:c.1009+7385_1009+7386=']['submitted_variant'] == 'NC_000005.9:g.35058665_35058666CA='
        assert results['NM_001204316.1:c.1009+7385_1009+7386=']['genome_context_intronic_sequence'] == 'NC_000005.9(NM_001204316.1):c.1009+7385_1009+7386='
        assert results['NM_001204316.1:c.1009+7385_1009+7386=']['hgvs_lrg_variant'] == ''
        assert results['NM_001204316.1:c.1009+7385_1009+7386=']['hgvs_transcript_variant'] == 'NM_001204316.1:c.1009+7385_1009+7386='
        assert results['NM_001204316.1:c.1009+7385_1009+7386=']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001204316.1:c.1009+7385_1009+7386=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.35058665_35058666=', 'vcf': {'chr': 'chr5', 'ref': 'CA', 'pos': '35058665', 'alt': 'CA'}}
        assert results['NM_001204316.1:c.1009+7385_1009+7386=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000005.10:g.35058563_35058564=', 'vcf': {'chr': 'chr5', 'ref': 'AG', 'pos': '35058563', 'alt': 'AG'}}
        assert results['NM_001204316.1:c.1009+7385_1009+7386=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.35058665_35058666=', 'vcf': {'chr': '5', 'ref': 'CA', 'pos': '35058665', 'alt': 'CA'}}
        assert results['NM_001204316.1:c.1009+7385_1009+7386=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000005.10:g.35058563_35058564=', 'vcf': {'chr': '5', 'ref': 'AG', 'pos': '35058563', 'alt': 'AG'}}

        assert 'NM_000949.6:c.*6528del' in results.keys()
        assert results['NM_000949.6:c.*6528del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000949.6:c.*6528del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000949.6:c.*6528del']['alt_genomic_loci'] == []
        assert results['NM_000949.6:c.*6528del']['gene_symbol'] == 'PRLR'
        assert results['NM_000949.6:c.*6528del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000940.1:p.?', 'slr': 'NP_000940.1:p.?'}
        assert results['NM_000949.6:c.*6528del']['submitted_variant'] == 'NC_000005.9:g.35058665_35058666CA='
        assert results['NM_000949.6:c.*6528del']['genome_context_intronic_sequence'] == ''
        assert results['NM_000949.6:c.*6528del']['hgvs_lrg_variant'] == ''
        assert results['NM_000949.6:c.*6528del']['hgvs_transcript_variant'] == 'NM_000949.6:c.*6528del'
        assert results['NM_000949.6:c.*6528del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_000949.6:c.*6528del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.35058662_35058668=', 'vcf': {'chr': 'chr5', 'ref': 'AGACAAG', 'pos': '35058662', 'alt': 'AGACAAG'}}
        assert results['NM_000949.6:c.*6528del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000005.10:g.35058563del', 'vcf': {'chr': 'chr5', 'ref': 'CA', 'pos': '35058560', 'alt': 'C'}}
        assert results['NM_000949.6:c.*6528del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.35058662_35058668=', 'vcf': {'chr': '5', 'ref': 'AGACAAG', 'pos': '35058662', 'alt': 'AGACAAG'}}
        assert results['NM_000949.6:c.*6528del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000005.10:g.35058563del', 'vcf': {'chr': '5', 'ref': 'CA', 'pos': '35058560', 'alt': 'C'}}


    def test_variant170(self):
        variant = 'NC_000002.11:g.73675227_73675229delTCTinsTCTCTC'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_015120.4:c.1580_1581insCCT' in results.keys()
        assert results['NM_015120.4:c.1580_1581insCCT']['hgvs_lrg_transcript_variant'] == 'LRG_741t1:c.1580_1581insCCT'
        assert results['NM_015120.4:c.1580_1581insCCT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_015120.4:c.1580_1581insCCT']['alt_genomic_loci'] == []
        assert results['NM_015120.4:c.1580_1581insCCT']['gene_symbol'] == 'ALMS1'
        assert results['NM_015120.4:c.1580_1581insCCT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_055935.4(LRG_741p1):p.(Leu527dup)', 'slr': 'NP_055935.4:p.(L527dup)'}
        assert results['NM_015120.4:c.1580_1581insCCT']['submitted_variant'] == 'NC_000002.11:g.73675227_73675229delTCTinsTCTCTC'
        assert results['NM_015120.4:c.1580_1581insCCT']['genome_context_intronic_sequence'] == ''
        assert results['NM_015120.4:c.1580_1581insCCT']['hgvs_lrg_variant'] == 'LRG_741:g.67352_67353insCCT'
        assert results['NM_015120.4:c.1580_1581insCCT']['hgvs_transcript_variant'] == 'NM_015120.4:c.1580_1581insCCT'
        assert results['NM_015120.4:c.1580_1581insCCT']['hgvs_refseqgene_variant'] == 'NG_011690.1:g.67352_67353insCCT'
        assert results['NM_015120.4:c.1580_1581insCCT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.73675231_73675232insCCT', 'vcf': {'chr': 'chr2', 'ref': 'T', 'pos': '73675229', 'alt': u'TCTC'}}
        assert results['NM_015120.4:c.1580_1581insCCT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.73448104_73448105insCCT', 'vcf': {'chr': 'chr2', 'ref': 'T', 'pos': '73448102', 'alt': u'TCTC'}}
        assert results['NM_015120.4:c.1580_1581insCCT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.73675231_73675232insCCT', 'vcf': {'chr': '2', 'ref': 'T', 'pos': '73675229', 'alt': u'TCTC'}}
        assert results['NM_015120.4:c.1580_1581insCCT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.73448104_73448105insCCT', 'vcf': {'chr': '2', 'ref': 'T', 'pos': '73448102', 'alt': u'TCTC'}}


    def test_variant173(self):
        variant = 'NM_000828.4:c.-2G>T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000828.4:c.-2G>T' in results.keys()
        assert results['NM_000828.4:c.-2G>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000828.4:c.-2G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000828.4:c.-2G>T']['alt_genomic_loci'] == []
        assert results['NM_000828.4:c.-2G>T']['gene_symbol'] == 'GRIA3'
        assert results['NM_000828.4:c.-2G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000819.3:p.?', 'slr': 'NP_000819.3:p.?'}
        assert results['NM_000828.4:c.-2G>T']['submitted_variant'] == 'NM_000828.4:c.-2G>T'
        assert results['NM_000828.4:c.-2G>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_000828.4:c.-2G>T']['hgvs_lrg_variant'] == ''
        assert results['NM_000828.4:c.-2G>T']['hgvs_transcript_variant'] == 'NM_000828.4:c.-2G>T'
        assert results['NM_000828.4:c.-2G>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_000828.4:c.-2G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.122318386_122318387insT', 'vcf': {'chr': 'chrX', 'ref': 'A', 'pos': '122318386', 'alt': 'AT'}}
        assert results['NM_000828.4:c.-2G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.123184534G>T', 'vcf': {'chr': 'chrX', 'ref': u'G', 'pos': '123184534', 'alt': u'T'}}
        assert results['NM_000828.4:c.-2G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.122318386_122318387insT', 'vcf': {'chr': 'X', 'ref': 'A', 'pos': '122318386', 'alt': 'AT'}}
        assert results['NM_000828.4:c.-2G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.123184534G>T', 'vcf': {'chr': 'X', 'ref': u'G', 'pos': '123184534', 'alt': u'T'}}


    def test_variant174(self):
        variant = 'NM_000828.4:c.-2G='
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_000828.4:c.-2G=' in results.keys()
        assert results['NM_000828.4:c.-2G=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000828.4:c.-2G=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000828.4:c.-2G=']['alt_genomic_loci'] == []
        assert results['NM_000828.4:c.-2G=']['gene_symbol'] == 'GRIA3'
        assert results['NM_000828.4:c.-2G=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000819.3:p.?', 'slr': 'NP_000819.3:p.?'}
        assert results['NM_000828.4:c.-2G=']['submitted_variant'] == 'NM_000828.4:c.-2G='
        assert results['NM_000828.4:c.-2G=']['genome_context_intronic_sequence'] == ''
        assert results['NM_000828.4:c.-2G=']['hgvs_lrg_variant'] == ''
        assert results['NM_000828.4:c.-2G=']['hgvs_transcript_variant'] == 'NM_000828.4:c.-2G='
        assert results['NM_000828.4:c.-2G=']['hgvs_refseqgene_variant'] == ''
        assert results['NM_000828.4:c.-2G=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.122318386_122318387insG', 'vcf': {'chr': 'chrX', 'ref': 'A', 'pos': '122318386', 'alt': u'AG'}}
        assert results['NM_000828.4:c.-2G=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.123184534G=', 'vcf': {'chr': 'chrX', 'ref': u'G', 'pos': '123184534', 'alt': u'G'}}
        assert results['NM_000828.4:c.-2G=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.122318386_122318387insG', 'vcf': {'chr': 'X', 'ref': 'A', 'pos': '122318386', 'alt': u'AG'}}
        assert results['NM_000828.4:c.-2G=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.123184534G=', 'vcf': {'chr': 'X', 'ref': u'G', 'pos': '123184534', 'alt': u'G'}}

        assert results['flag'] == 'gene_variant'

    def test_variant175(self):
        variant = 'X-122318386-A-AT'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000828.4:c.-2G>T' in results.keys()
        assert results['NM_000828.4:c.-2G>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000828.4:c.-2G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000828.4:c.-2G>T']['alt_genomic_loci'] == []
        assert results['NM_000828.4:c.-2G>T']['gene_symbol'] == 'GRIA3'
        assert results['NM_000828.4:c.-2G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000819.3:p.?', 'slr': 'NP_000819.3:p.?'}
        assert results['NM_000828.4:c.-2G>T']['submitted_variant'] == 'X-122318386-A-AT'
        assert results['NM_000828.4:c.-2G>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_000828.4:c.-2G>T']['hgvs_lrg_variant'] == ''
        assert results['NM_000828.4:c.-2G>T']['hgvs_transcript_variant'] == 'NM_000828.4:c.-2G>T'
        assert results['NM_000828.4:c.-2G>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_000828.4:c.-2G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.122318386_122318387insT', 'vcf': {'chr': 'chrX', 'ref': 'A', 'pos': '122318386', 'alt': 'AT'}}
        assert results['NM_000828.4:c.-2G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.123184534G>T', 'vcf': {'chr': 'chrX', 'ref': u'G', 'pos': '123184534', 'alt': u'T'}}
        assert results['NM_000828.4:c.-2G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.122318386_122318387insT', 'vcf': {'chr': 'X', 'ref': 'A', 'pos': '122318386', 'alt': 'AT'}}
        assert results['NM_000828.4:c.-2G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.123184534G>T', 'vcf': {'chr': 'X', 'ref': u'G', 'pos': '123184534', 'alt': u'T'}}

        assert 'NM_001256743.1:c.-2G>T' in results.keys()
        assert results['NM_001256743.1:c.-2G>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001256743.1:c.-2G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001256743.1:c.-2G>T']['alt_genomic_loci'] == []
        assert results['NM_001256743.1:c.-2G>T']['gene_symbol'] == 'GRIA3'
        assert results['NM_001256743.1:c.-2G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001243672.1:p.?', 'slr': 'NP_001243672.1:p.?'}
        assert results['NM_001256743.1:c.-2G>T']['submitted_variant'] == 'X-122318386-A-AT'
        assert results['NM_001256743.1:c.-2G>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001256743.1:c.-2G>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001256743.1:c.-2G>T']['hgvs_transcript_variant'] == 'NM_001256743.1:c.-2G>T'
        assert results['NM_001256743.1:c.-2G>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001256743.1:c.-2G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.122318386_122318387insT', 'vcf': {'chr': 'chrX', 'ref': 'A', 'pos': '122318386', 'alt': 'AT'}}
        assert results['NM_001256743.1:c.-2G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.123184534G>T', 'vcf': {'chr': 'chrX', 'ref': u'G', 'pos': '123184534', 'alt': u'T'}}
        assert results['NM_001256743.1:c.-2G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.122318386_122318387insT', 'vcf': {'chr': 'X', 'ref': 'A', 'pos': '122318386', 'alt': 'AT'}}
        assert results['NM_001256743.1:c.-2G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.123184534G>T', 'vcf': {'chr': 'X', 'ref': u'G', 'pos': '123184534', 'alt': u'T'}}

        assert 'NM_007325.4:c.-2G>T' in results.keys()
        assert results['NM_007325.4:c.-2G>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_007325.4:c.-2G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_007325.4:c.-2G>T']['alt_genomic_loci'] == []
        assert results['NM_007325.4:c.-2G>T']['gene_symbol'] == 'GRIA3'
        assert results['NM_007325.4:c.-2G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_015564.4:p.?', 'slr': 'NP_015564.4:p.?'}
        assert results['NM_007325.4:c.-2G>T']['submitted_variant'] == 'X-122318386-A-AT'
        assert results['NM_007325.4:c.-2G>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_007325.4:c.-2G>T']['hgvs_lrg_variant'] == ''
        assert results['NM_007325.4:c.-2G>T']['hgvs_transcript_variant'] == 'NM_007325.4:c.-2G>T'
        assert results['NM_007325.4:c.-2G>T']['hgvs_refseqgene_variant'] == 'NG_009377.1:g.5292G>T'
        assert results['NM_007325.4:c.-2G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.122318386_122318387insT', 'vcf': {'chr': 'chrX', 'ref': 'A', 'pos': '122318386', 'alt': 'AT'}}
        assert results['NM_007325.4:c.-2G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.123184534G>T', 'vcf': {'chr': 'chrX', 'ref': u'G', 'pos': '123184534', 'alt': u'T'}}
        assert results['NM_007325.4:c.-2G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.122318386_122318387insT', 'vcf': {'chr': 'X', 'ref': 'A', 'pos': '122318386', 'alt': 'AT'}}
        assert results['NM_007325.4:c.-2G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.123184534G>T', 'vcf': {'chr': 'X', 'ref': u'G', 'pos': '123184534', 'alt': u'T'}}


    def test_variant176(self):
        variant = 'NM_000828.4:c.-2_-1insT'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000828.4:c.-2_-1insT' in results.keys()
        assert results['NM_000828.4:c.-2_-1insT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000828.4:c.-2_-1insT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000828.4:c.-2_-1insT']['alt_genomic_loci'] == []
        assert results['NM_000828.4:c.-2_-1insT']['gene_symbol'] == 'GRIA3'
        assert results['NM_000828.4:c.-2_-1insT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000819.3:p.?', 'slr': 'NP_000819.3:p.?'}
        assert results['NM_000828.4:c.-2_-1insT']['submitted_variant'] == 'NM_000828.4:c.-2_-1insT'
        assert results['NM_000828.4:c.-2_-1insT']['genome_context_intronic_sequence'] == ''
        assert results['NM_000828.4:c.-2_-1insT']['hgvs_lrg_variant'] == ''
        assert results['NM_000828.4:c.-2_-1insT']['hgvs_transcript_variant'] == 'NM_000828.4:c.-2_-1insT'
        assert results['NM_000828.4:c.-2_-1insT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_000828.4:c.-2_-1insT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.122318386_122318387insGT', 'vcf': {'chr': 'chrX', 'ref': 'A', 'pos': '122318386', 'alt': 'AGT'}}
        assert results['NM_000828.4:c.-2_-1insT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.123184534_123184535insT', 'vcf': {'chr': 'chrX', 'ref': 'G', 'pos': '123184534', 'alt': u'GT'}}
        assert results['NM_000828.4:c.-2_-1insT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.122318386_122318387insGT', 'vcf': {'chr': 'X', 'ref': 'A', 'pos': '122318386', 'alt': 'AGT'}}
        assert results['NM_000828.4:c.-2_-1insT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.123184534_123184535insT', 'vcf': {'chr': 'X', 'ref': 'G', 'pos': '123184534', 'alt': u'GT'}}


    def test_variant177(self):
        variant = 'NM_000828.4:c.-3_-2insT'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_000828.4:c.-3_-2insT' in results.keys()
        assert results['NM_000828.4:c.-3_-2insT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000828.4:c.-3_-2insT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000828.4:c.-3_-2insT']['alt_genomic_loci'] == []
        assert results['NM_000828.4:c.-3_-2insT']['gene_symbol'] == 'GRIA3'
        assert results['NM_000828.4:c.-3_-2insT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000819.3:p.?', 'slr': 'NP_000819.3:p.?'}
        assert results['NM_000828.4:c.-3_-2insT']['submitted_variant'] == 'NM_000828.4:c.-3_-2insT'
        assert results['NM_000828.4:c.-3_-2insT']['genome_context_intronic_sequence'] == ''
        assert results['NM_000828.4:c.-3_-2insT']['hgvs_lrg_variant'] == ''
        assert results['NM_000828.4:c.-3_-2insT']['hgvs_transcript_variant'] == 'NM_000828.4:c.-3_-2insT'
        assert results['NM_000828.4:c.-3_-2insT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_000828.4:c.-3_-2insT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.122318386_122318387insTG', 'vcf': {'chr': 'chrX', 'ref': 'A', 'pos': '122318386', 'alt': 'ATG'}}
        assert results['NM_000828.4:c.-3_-2insT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.123184533_123184534insT', 'vcf': {'chr': 'chrX', 'ref': 'A', 'pos': '123184533', 'alt': u'AT'}}
        assert results['NM_000828.4:c.-3_-2insT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.122318386_122318387insTG', 'vcf': {'chr': 'X', 'ref': 'A', 'pos': '122318386', 'alt': 'ATG'}}
        assert results['NM_000828.4:c.-3_-2insT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.123184533_123184534insT', 'vcf': {'chr': 'X', 'ref': 'A', 'pos': '123184533', 'alt': u'AT'}}

        assert results['flag'] == 'gene_variant'

    def test_variant178(self):
        variant = 'NM_000828.4:c.-2delGinsTT'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000828.4:c.-2delinsTT' in results.keys()
        assert results['NM_000828.4:c.-2delinsTT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000828.4:c.-2delinsTT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000828.4:c.-2delinsTT']['alt_genomic_loci'] == []
        assert results['NM_000828.4:c.-2delinsTT']['gene_symbol'] == 'GRIA3'
        assert results['NM_000828.4:c.-2delinsTT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000819.3:p.?', 'slr': 'NP_000819.3:p.?'}
        assert results['NM_000828.4:c.-2delinsTT']['submitted_variant'] == 'NM_000828.4:c.-2delGinsTT'
        assert results['NM_000828.4:c.-2delinsTT']['genome_context_intronic_sequence'] == ''
        assert results['NM_000828.4:c.-2delinsTT']['hgvs_lrg_variant'] == ''
        assert results['NM_000828.4:c.-2delinsTT']['hgvs_transcript_variant'] == 'NM_000828.4:c.-2delinsTT'
        assert results['NM_000828.4:c.-2delinsTT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_000828.4:c.-2delinsTT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.122318386_122318387insTT', 'vcf': {'chr': 'chrX', 'ref': 'A', 'pos': '122318386', 'alt': 'ATT'}}
        assert results['NM_000828.4:c.-2delinsTT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.123184534delinsTT', 'vcf': {'chr': 'chrX', 'ref': 'G', 'pos': '123184534', 'alt': u'TT'}}
        assert results['NM_000828.4:c.-2delinsTT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.122318386_122318387insTT', 'vcf': {'chr': 'X', 'ref': 'A', 'pos': '122318386', 'alt': 'ATT'}}
        assert results['NM_000828.4:c.-2delinsTT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.123184534delinsTT', 'vcf': {'chr': 'X', 'ref': 'G', 'pos': '123184534', 'alt': u'TT'}}


    def test_variant179(self):
        variant = 'NM_000828.4:c.-2_-1delGCinsTT'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000828.4:c.-2_-1delinsTT' in results.keys()
        assert results['NM_000828.4:c.-2_-1delinsTT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000828.4:c.-2_-1delinsTT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000828.4:c.-2_-1delinsTT']['alt_genomic_loci'] == []
        assert results['NM_000828.4:c.-2_-1delinsTT']['gene_symbol'] == 'GRIA3'
        assert results['NM_000828.4:c.-2_-1delinsTT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000819.3:p.?', 'slr': 'NP_000819.3:p.?'}
        assert results['NM_000828.4:c.-2_-1delinsTT']['submitted_variant'] == 'NM_000828.4:c.-2_-1delGCinsTT'
        assert results['NM_000828.4:c.-2_-1delinsTT']['genome_context_intronic_sequence'] == ''
        assert results['NM_000828.4:c.-2_-1delinsTT']['hgvs_lrg_variant'] == ''
        assert results['NM_000828.4:c.-2_-1delinsTT']['hgvs_transcript_variant'] == 'NM_000828.4:c.-2_-1delinsTT'
        assert results['NM_000828.4:c.-2_-1delinsTT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_000828.4:c.-2_-1delinsTT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.122318387delinsTT', 'vcf': {'chr': 'chrX', 'ref': 'C', 'pos': '122318387', 'alt': 'TT'}}
        assert results['NM_000828.4:c.-2_-1delinsTT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.123184534_123184535delinsTT', 'vcf': {'chr': 'chrX', 'ref': 'GC', 'pos': '123184534', 'alt': u'TT'}}
        assert results['NM_000828.4:c.-2_-1delinsTT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.122318387delinsTT', 'vcf': {'chr': 'X', 'ref': 'C', 'pos': '122318387', 'alt': 'TT'}}
        assert results['NM_000828.4:c.-2_-1delinsTT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.123184534_123184535delinsTT', 'vcf': {'chr': 'X', 'ref': 'GC', 'pos': '123184534', 'alt': u'TT'}}


    def test_variant180(self):
        variant = 'NM_000828.4:c.-3_-2delAGinsTT'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_000828.4:c.-3_-2delinsTT' in results.keys()
        assert results['NM_000828.4:c.-3_-2delinsTT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000828.4:c.-3_-2delinsTT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000828.4:c.-3_-2delinsTT']['alt_genomic_loci'] == []
        assert results['NM_000828.4:c.-3_-2delinsTT']['gene_symbol'] == 'GRIA3'
        assert results['NM_000828.4:c.-3_-2delinsTT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000819.3:p.?', 'slr': 'NP_000819.3:p.?'}
        assert results['NM_000828.4:c.-3_-2delinsTT']['submitted_variant'] == 'NM_000828.4:c.-3_-2delAGinsTT'
        assert results['NM_000828.4:c.-3_-2delinsTT']['genome_context_intronic_sequence'] == ''
        assert results['NM_000828.4:c.-3_-2delinsTT']['hgvs_lrg_variant'] == ''
        assert results['NM_000828.4:c.-3_-2delinsTT']['hgvs_transcript_variant'] == 'NM_000828.4:c.-3_-2delinsTT'
        assert results['NM_000828.4:c.-3_-2delinsTT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_000828.4:c.-3_-2delinsTT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.122318386delinsTT', 'vcf': {'chr': 'chrX', 'ref': 'A', 'pos': '122318386', 'alt': 'TT'}}
        assert results['NM_000828.4:c.-3_-2delinsTT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.123184533_123184534delinsTT', 'vcf': {'chr': 'chrX', 'ref': 'AG', 'pos': '123184533', 'alt': u'TT'}}
        assert results['NM_000828.4:c.-3_-2delinsTT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.122318386delinsTT', 'vcf': {'chr': 'X', 'ref': 'A', 'pos': '122318386', 'alt': 'TT'}}
        assert results['NM_000828.4:c.-3_-2delinsTT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.123184533_123184534delinsTT', 'vcf': {'chr': 'X', 'ref': 'AG', 'pos': '123184533', 'alt': u'TT'}}

        assert results['flag'] == 'gene_variant'

    def test_variant181(self):
        variant = '15-72105929-C-C'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_014249.3:c.951dup' in results.keys()
        assert results['NM_014249.3:c.951dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_014249.3:c.951dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_014249.3:c.951dup']['alt_genomic_loci'] == []
        assert results['NM_014249.3:c.951dup']['gene_symbol'] == 'NR2E3'
        assert results['NM_014249.3:c.951dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_055064.1:p.(Thr318HisfsTer23)', 'slr': 'NP_055064.1:p.(T318Hfs*23)'}
        assert results['NM_014249.3:c.951dup']['submitted_variant'] == '15-72105929-C-C'
        assert results['NM_014249.3:c.951dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_014249.3:c.951dup']['hgvs_lrg_variant'] == ''
        assert results['NM_014249.3:c.951dup']['hgvs_transcript_variant'] == 'NM_014249.3:c.951dup'
        assert results['NM_014249.3:c.951dup']['hgvs_refseqgene_variant'] == 'NG_009113.1:g.8039dup'
        assert results['NM_014249.3:c.951dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105924_72105934=', 'vcf': {'chr': 'chr15', 'ref': 'GTGGACCCCCA', 'pos': '72105924', 'alt': 'GTGGACCCCCA'}}
        assert results['NM_014249.3:c.951dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000015.10:g.71813592dup', 'vcf': {'chr': 'chr15', 'ref': 'C', 'pos': '71813589', 'alt': u'CC'}}
        assert results['NM_014249.3:c.951dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105924_72105934=', 'vcf': {'chr': '15', 'ref': 'GTGGACCCCCA', 'pos': '72105924', 'alt': 'GTGGACCCCCA'}}
        assert results['NM_014249.3:c.951dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000015.10:g.71813592dup', 'vcf': {'chr': '15', 'ref': 'C', 'pos': '71813589', 'alt': u'CC'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_016346.2:c.951dup' in results.keys()
        assert results['NM_016346.2:c.951dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_016346.2:c.951dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_016346.2:c.951dup']['alt_genomic_loci'] == []
        assert results['NM_016346.2:c.951dup']['gene_symbol'] == 'NR2E3'
        assert results['NM_016346.2:c.951dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_057430.1:p.(Thr318HisfsTer23)', 'slr': 'NP_057430.1:p.(T318Hfs*23)'}
        assert results['NM_016346.2:c.951dup']['submitted_variant'] == '15-72105929-C-C'
        assert results['NM_016346.2:c.951dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_016346.2:c.951dup']['hgvs_lrg_variant'] == ''
        assert results['NM_016346.2:c.951dup']['hgvs_transcript_variant'] == 'NM_016346.2:c.951dup'
        assert results['NM_016346.2:c.951dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_016346.2:c.951dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105924_72105934=', 'vcf': {'chr': 'chr15', 'ref': 'GTGGACCCCCA', 'pos': '72105924', 'alt': 'GTGGACCCCCA'}}
        assert 'hg38' not in results['NM_016346.2:c.951dup']['primary_assembly_loci'].keys()
        assert results['NM_016346.2:c.951dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105924_72105934=', 'vcf': {'chr': '15', 'ref': 'GTGGACCCCCA', 'pos': '72105924', 'alt': 'GTGGACCCCCA'}}
        assert 'grch38' not in results['NM_016346.2:c.951dup']['primary_assembly_loci'].keys()

        assert 'NM_016346.3:c.951dup' in results.keys()
        assert results['NM_016346.3:c.951dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_016346.3:c.951dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_016346.3:c.951dup']['alt_genomic_loci'] == []
        assert results['NM_016346.3:c.951dup']['gene_symbol'] == 'NR2E3'
        assert results['NM_016346.3:c.951dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_057430.1:p.(Thr318HisfsTer23)', 'slr': 'NP_057430.1:p.(T318Hfs*23)'}
        assert results['NM_016346.3:c.951dup']['submitted_variant'] == '15-72105929-C-C'
        assert results['NM_016346.3:c.951dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_016346.3:c.951dup']['hgvs_lrg_variant'] == ''
        assert results['NM_016346.3:c.951dup']['hgvs_transcript_variant'] == 'NM_016346.3:c.951dup'
        assert results['NM_016346.3:c.951dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_016346.3:c.951dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105924_72105934=', 'vcf': {'chr': 'chr15', 'ref': 'GTGGACCCCCA', 'pos': '72105924', 'alt': 'GTGGACCCCCA'}}
        assert results['NM_016346.3:c.951dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000015.10:g.71813592dup', 'vcf': {'chr': 'chr15', 'ref': 'C', 'pos': '71813589', 'alt': u'CC'}}
        assert results['NM_016346.3:c.951dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105924_72105934=', 'vcf': {'chr': '15', 'ref': 'GTGGACCCCCA', 'pos': '72105924', 'alt': 'GTGGACCCCCA'}}
        assert results['NM_016346.3:c.951dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000015.10:g.71813592dup', 'vcf': {'chr': '15', 'ref': 'C', 'pos': '71813589', 'alt': u'CC'}}

        assert 'NM_014249.2:c.951dup' in results.keys()
        assert results['NM_014249.2:c.951dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_014249.2:c.951dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_014249.2:c.951dup']['alt_genomic_loci'] == []
        assert results['NM_014249.2:c.951dup']['gene_symbol'] == 'NR2E3'
        assert results['NM_014249.2:c.951dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_055064.1:p.(Thr318HisfsTer23)', 'slr': 'NP_055064.1:p.(T318Hfs*23)'}
        assert results['NM_014249.2:c.951dup']['submitted_variant'] == '15-72105929-C-C'
        assert results['NM_014249.2:c.951dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_014249.2:c.951dup']['hgvs_lrg_variant'] == ''
        assert results['NM_014249.2:c.951dup']['hgvs_transcript_variant'] == 'NM_014249.2:c.951dup'
        assert results['NM_014249.2:c.951dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_014249.2:c.951dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105924_72105934=', 'vcf': {'chr': 'chr15', 'ref': 'GTGGACCCCCA', 'pos': '72105924', 'alt': 'GTGGACCCCCA'}}
        assert 'hg38' not in results['NM_014249.2:c.951dup']['primary_assembly_loci'].keys()
        assert results['NM_014249.2:c.951dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105924_72105934=', 'vcf': {'chr': '15', 'ref': 'GTGGACCCCCA', 'pos': '72105924', 'alt': 'GTGGACCCCCA'}}
        assert 'grch38' not in results['NM_014249.2:c.951dup']['primary_assembly_loci'].keys()


    def test_variant182(self):
        variant = '15-72105928-AC-ATT'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_014249.2:c.947_948insTT' in results.keys()
        assert results['NM_014249.2:c.947_948insTT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_014249.2:c.947_948insTT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_014249.2:c.947_948insTT']['alt_genomic_loci'] == []
        assert results['NM_014249.2:c.947_948insTT']['gene_symbol'] == 'NR2E3'
        assert results['NM_014249.2:c.947_948insTT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_055064.1:p.(Pro317SerfsTer8)', 'slr': 'NP_055064.1:p.(P317Sfs*8)'}
        assert results['NM_014249.2:c.947_948insTT']['submitted_variant'] == '15-72105928-AC-ATT'
        assert results['NM_014249.2:c.947_948insTT']['genome_context_intronic_sequence'] == ''
        assert results['NM_014249.2:c.947_948insTT']['hgvs_lrg_variant'] == ''
        assert results['NM_014249.2:c.947_948insTT']['hgvs_transcript_variant'] == 'NM_014249.2:c.947_948insTT'
        assert results['NM_014249.2:c.947_948insTT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_014249.2:c.947_948insTT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105929delinsTT', 'vcf': {'chr': 'chr15', 'ref': 'C', 'pos': '72105929', 'alt': 'TT'}}
        assert 'hg38' not in results['NM_014249.2:c.947_948insTT']['primary_assembly_loci'].keys()
        assert results['NM_014249.2:c.947_948insTT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105929delinsTT', 'vcf': {'chr': '15', 'ref': 'C', 'pos': '72105929', 'alt': 'TT'}}
        assert 'grch38' not in results['NM_014249.2:c.947_948insTT']['primary_assembly_loci'].keys()

        assert results['flag'] == 'gene_variant'
        assert 'NM_016346.3:c.947_948insTT' in results.keys()
        assert results['NM_016346.3:c.947_948insTT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_016346.3:c.947_948insTT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_016346.3:c.947_948insTT']['alt_genomic_loci'] == []
        assert results['NM_016346.3:c.947_948insTT']['gene_symbol'] == 'NR2E3'
        assert results['NM_016346.3:c.947_948insTT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_057430.1:p.(Pro317SerfsTer8)', 'slr': 'NP_057430.1:p.(P317Sfs*8)'}
        assert results['NM_016346.3:c.947_948insTT']['submitted_variant'] == '15-72105928-AC-ATT'
        assert results['NM_016346.3:c.947_948insTT']['genome_context_intronic_sequence'] == ''
        assert results['NM_016346.3:c.947_948insTT']['hgvs_lrg_variant'] == ''
        assert results['NM_016346.3:c.947_948insTT']['hgvs_transcript_variant'] == 'NM_016346.3:c.947_948insTT'
        assert results['NM_016346.3:c.947_948insTT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_016346.3:c.947_948insTT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105929delinsTT', 'vcf': {'chr': 'chr15', 'ref': 'C', 'pos': '72105929', 'alt': 'TT'}}
        assert results['NM_016346.3:c.947_948insTT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000015.10:g.71813588_71813589insTT', 'vcf': {'chr': 'chr15', 'ref': 'A', 'pos': '71813588', 'alt': u'ATT'}}
        assert results['NM_016346.3:c.947_948insTT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105929delinsTT', 'vcf': {'chr': '15', 'ref': 'C', 'pos': '72105929', 'alt': 'TT'}}
        assert results['NM_016346.3:c.947_948insTT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000015.10:g.71813588_71813589insTT', 'vcf': {'chr': '15', 'ref': 'A', 'pos': '71813588', 'alt': u'ATT'}}

        assert 'NM_014249.3:c.947_948insTT' in results.keys()
        assert results['NM_014249.3:c.947_948insTT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_014249.3:c.947_948insTT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_014249.3:c.947_948insTT']['alt_genomic_loci'] == []
        assert results['NM_014249.3:c.947_948insTT']['gene_symbol'] == 'NR2E3'
        assert results['NM_014249.3:c.947_948insTT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_055064.1:p.(Pro317SerfsTer8)', 'slr': 'NP_055064.1:p.(P317Sfs*8)'}
        assert results['NM_014249.3:c.947_948insTT']['submitted_variant'] == '15-72105928-AC-ATT'
        assert results['NM_014249.3:c.947_948insTT']['genome_context_intronic_sequence'] == ''
        assert results['NM_014249.3:c.947_948insTT']['hgvs_lrg_variant'] == ''
        assert results['NM_014249.3:c.947_948insTT']['hgvs_transcript_variant'] == 'NM_014249.3:c.947_948insTT'
        assert results['NM_014249.3:c.947_948insTT']['hgvs_refseqgene_variant'] == 'NG_009113.1:g.8035_8036insTT'
        assert results['NM_014249.3:c.947_948insTT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105929delinsTT', 'vcf': {'chr': 'chr15', 'ref': 'C', 'pos': '72105929', 'alt': 'TT'}}
        assert results['NM_014249.3:c.947_948insTT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000015.10:g.71813588_71813589insTT', 'vcf': {'chr': 'chr15', 'ref': 'A', 'pos': '71813588', 'alt': u'ATT'}}
        assert results['NM_014249.3:c.947_948insTT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105929delinsTT', 'vcf': {'chr': '15', 'ref': 'C', 'pos': '72105929', 'alt': 'TT'}}
        assert results['NM_014249.3:c.947_948insTT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000015.10:g.71813588_71813589insTT', 'vcf': {'chr': '15', 'ref': 'A', 'pos': '71813588', 'alt': u'ATT'}}

        assert 'NM_016346.2:c.947_948insTT' in results.keys()
        assert results['NM_016346.2:c.947_948insTT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_016346.2:c.947_948insTT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_016346.2:c.947_948insTT']['alt_genomic_loci'] == []
        assert results['NM_016346.2:c.947_948insTT']['gene_symbol'] == 'NR2E3'
        assert results['NM_016346.2:c.947_948insTT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_057430.1:p.(Pro317SerfsTer8)', 'slr': 'NP_057430.1:p.(P317Sfs*8)'}
        assert results['NM_016346.2:c.947_948insTT']['submitted_variant'] == '15-72105928-AC-ATT'
        assert results['NM_016346.2:c.947_948insTT']['genome_context_intronic_sequence'] == ''
        assert results['NM_016346.2:c.947_948insTT']['hgvs_lrg_variant'] == ''
        assert results['NM_016346.2:c.947_948insTT']['hgvs_transcript_variant'] == 'NM_016346.2:c.947_948insTT'
        assert results['NM_016346.2:c.947_948insTT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_016346.2:c.947_948insTT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105929delinsTT', 'vcf': {'chr': 'chr15', 'ref': 'C', 'pos': '72105929', 'alt': 'TT'}}
        assert 'hg38' not in results['NM_016346.2:c.947_948insTT']['primary_assembly_loci'].keys()
        assert results['NM_016346.2:c.947_948insTT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105929delinsTT', 'vcf': {'chr': '15', 'ref': 'C', 'pos': '72105929', 'alt': 'TT'}}
        assert 'grch38' not in results['NM_016346.2:c.947_948insTT']['primary_assembly_loci'].keys()


    def test_variant183(self):
        variant = '15-72105928-ACC-ATT'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_014249.2:c.947_948insTT' in results.keys()
        assert results['NM_014249.2:c.947_948insTT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_014249.2:c.947_948insTT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_014249.2:c.947_948insTT']['alt_genomic_loci'] == []
        assert results['NM_014249.2:c.947_948insTT']['gene_symbol'] == 'NR2E3'
        assert results['NM_014249.2:c.947_948insTT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_055064.1:p.(Pro317SerfsTer8)', 'slr': 'NP_055064.1:p.(P317Sfs*8)'}
        assert results['NM_014249.2:c.947_948insTT']['submitted_variant'] == '15-72105928-ACC-ATT'
        assert results['NM_014249.2:c.947_948insTT']['genome_context_intronic_sequence'] == ''
        assert results['NM_014249.2:c.947_948insTT']['hgvs_lrg_variant'] == ''
        assert results['NM_014249.2:c.947_948insTT']['hgvs_transcript_variant'] == 'NM_014249.2:c.947_948insTT'
        assert results['NM_014249.2:c.947_948insTT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_014249.2:c.947_948insTT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105929delinsTT', 'vcf': {'chr': 'chr15', 'ref': 'C', 'pos': '72105929', 'alt': 'TT'}}
        assert 'hg38' not in results['NM_014249.2:c.947_948insTT']['primary_assembly_loci'].keys()
        assert results['NM_014249.2:c.947_948insTT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105929delinsTT', 'vcf': {'chr': '15', 'ref': 'C', 'pos': '72105929', 'alt': 'TT'}}
        assert 'grch38' not in results['NM_014249.2:c.947_948insTT']['primary_assembly_loci'].keys()

        assert results['flag'] == 'gene_variant'
        assert 'NM_016346.3:c.947_948insTT' in results.keys()
        assert results['NM_016346.3:c.947_948insTT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_016346.3:c.947_948insTT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_016346.3:c.947_948insTT']['alt_genomic_loci'] == []
        assert results['NM_016346.3:c.947_948insTT']['gene_symbol'] == 'NR2E3'
        assert results['NM_016346.3:c.947_948insTT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_057430.1:p.(Pro317SerfsTer8)', 'slr': 'NP_057430.1:p.(P317Sfs*8)'}
        assert results['NM_016346.3:c.947_948insTT']['submitted_variant'] == '15-72105928-ACC-ATT'
        assert results['NM_016346.3:c.947_948insTT']['genome_context_intronic_sequence'] == ''
        assert results['NM_016346.3:c.947_948insTT']['hgvs_lrg_variant'] == ''
        assert results['NM_016346.3:c.947_948insTT']['hgvs_transcript_variant'] == 'NM_016346.3:c.947_948insTT'
        assert results['NM_016346.3:c.947_948insTT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_016346.3:c.947_948insTT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105929delinsTT', 'vcf': {'chr': 'chr15', 'ref': 'C', 'pos': '72105929', 'alt': 'TT'}}
        assert results['NM_016346.3:c.947_948insTT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000015.10:g.71813588_71813589insTT', 'vcf': {'chr': 'chr15', 'ref': 'A', 'pos': '71813588', 'alt': u'ATT'}}
        assert results['NM_016346.3:c.947_948insTT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105929delinsTT', 'vcf': {'chr': '15', 'ref': 'C', 'pos': '72105929', 'alt': 'TT'}}
        assert results['NM_016346.3:c.947_948insTT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000015.10:g.71813588_71813589insTT', 'vcf': {'chr': '15', 'ref': 'A', 'pos': '71813588', 'alt': u'ATT'}}

        assert 'NM_014249.3:c.947_948insTT' in results.keys()
        assert results['NM_014249.3:c.947_948insTT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_014249.3:c.947_948insTT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_014249.3:c.947_948insTT']['alt_genomic_loci'] == []
        assert results['NM_014249.3:c.947_948insTT']['gene_symbol'] == 'NR2E3'
        assert results['NM_014249.3:c.947_948insTT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_055064.1:p.(Pro317SerfsTer8)', 'slr': 'NP_055064.1:p.(P317Sfs*8)'}
        assert results['NM_014249.3:c.947_948insTT']['submitted_variant'] == '15-72105928-ACC-ATT'
        assert results['NM_014249.3:c.947_948insTT']['genome_context_intronic_sequence'] == ''
        assert results['NM_014249.3:c.947_948insTT']['hgvs_lrg_variant'] == ''
        assert results['NM_014249.3:c.947_948insTT']['hgvs_transcript_variant'] == 'NM_014249.3:c.947_948insTT'
        assert results['NM_014249.3:c.947_948insTT']['hgvs_refseqgene_variant'] == 'NG_009113.1:g.8035_8036insTT'
        assert results['NM_014249.3:c.947_948insTT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105929delinsTT', 'vcf': {'chr': 'chr15', 'ref': 'C', 'pos': '72105929', 'alt': 'TT'}}
        assert results['NM_014249.3:c.947_948insTT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000015.10:g.71813588_71813589insTT', 'vcf': {'chr': 'chr15', 'ref': 'A', 'pos': '71813588', 'alt': u'ATT'}}
        assert results['NM_014249.3:c.947_948insTT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105929delinsTT', 'vcf': {'chr': '15', 'ref': 'C', 'pos': '72105929', 'alt': 'TT'}}
        assert results['NM_014249.3:c.947_948insTT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000015.10:g.71813588_71813589insTT', 'vcf': {'chr': '15', 'ref': 'A', 'pos': '71813588', 'alt': u'ATT'}}

        assert 'NM_016346.2:c.947_948insTT' in results.keys()
        assert results['NM_016346.2:c.947_948insTT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_016346.2:c.947_948insTT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_016346.2:c.947_948insTT']['alt_genomic_loci'] == []
        assert results['NM_016346.2:c.947_948insTT']['gene_symbol'] == 'NR2E3'
        assert results['NM_016346.2:c.947_948insTT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_057430.1:p.(Pro317SerfsTer8)', 'slr': 'NP_057430.1:p.(P317Sfs*8)'}
        assert results['NM_016346.2:c.947_948insTT']['submitted_variant'] == '15-72105928-ACC-ATT'
        assert results['NM_016346.2:c.947_948insTT']['genome_context_intronic_sequence'] == ''
        assert results['NM_016346.2:c.947_948insTT']['hgvs_lrg_variant'] == ''
        assert results['NM_016346.2:c.947_948insTT']['hgvs_transcript_variant'] == 'NM_016346.2:c.947_948insTT'
        assert results['NM_016346.2:c.947_948insTT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_016346.2:c.947_948insTT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105929delinsTT', 'vcf': {'chr': 'chr15', 'ref': 'C', 'pos': '72105929', 'alt': 'TT'}}
        assert 'hg38' not in results['NM_016346.2:c.947_948insTT']['primary_assembly_loci'].keys()
        assert results['NM_016346.2:c.947_948insTT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105929delinsTT', 'vcf': {'chr': '15', 'ref': 'C', 'pos': '72105929', 'alt': 'TT'}}
        assert 'grch38' not in results['NM_016346.2:c.947_948insTT']['primary_assembly_loci'].keys()


    def test_variant184(self):
        variant = '15-72105927-GACC-GTT'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_016346.3:c.947delinsTT' in results.keys()
        assert results['NM_016346.3:c.947delinsTT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_016346.3:c.947delinsTT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_016346.3:c.947delinsTT']['alt_genomic_loci'] == []
        assert results['NM_016346.3:c.947delinsTT']['gene_symbol'] == 'NR2E3'
        assert results['NM_016346.3:c.947delinsTT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_057430.1:p.(Asp316ValfsTer25)', 'slr': 'NP_057430.1:p.(D316Vfs*25)'}
        assert results['NM_016346.3:c.947delinsTT']['submitted_variant'] == '15-72105927-GACC-GTT'
        assert results['NM_016346.3:c.947delinsTT']['genome_context_intronic_sequence'] == ''
        assert results['NM_016346.3:c.947delinsTT']['hgvs_lrg_variant'] == ''
        assert results['NM_016346.3:c.947delinsTT']['hgvs_transcript_variant'] == 'NM_016346.3:c.947delinsTT'
        assert results['NM_016346.3:c.947delinsTT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_016346.3:c.947delinsTT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105928_72105929delinsTT', 'vcf': {'chr': 'chr15', 'ref': 'AC', 'pos': '72105928', 'alt': 'TT'}}
        assert results['NM_016346.3:c.947delinsTT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000015.10:g.71813588delinsTT', 'vcf': {'chr': 'chr15', 'ref': 'A', 'pos': '71813588', 'alt': u'TT'}}
        assert results['NM_016346.3:c.947delinsTT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105928_72105929delinsTT', 'vcf': {'chr': '15', 'ref': 'AC', 'pos': '72105928', 'alt': 'TT'}}
        assert results['NM_016346.3:c.947delinsTT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000015.10:g.71813588delinsTT', 'vcf': {'chr': '15', 'ref': 'A', 'pos': '71813588', 'alt': u'TT'}}

        assert 'NM_014249.2:c.947delinsTT' in results.keys()
        assert results['NM_014249.2:c.947delinsTT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_014249.2:c.947delinsTT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_014249.2:c.947delinsTT']['alt_genomic_loci'] == []
        assert results['NM_014249.2:c.947delinsTT']['gene_symbol'] == 'NR2E3'
        assert results['NM_014249.2:c.947delinsTT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_055064.1:p.(Asp316ValfsTer25)', 'slr': 'NP_055064.1:p.(D316Vfs*25)'}
        assert results['NM_014249.2:c.947delinsTT']['submitted_variant'] == '15-72105927-GACC-GTT'
        assert results['NM_014249.2:c.947delinsTT']['genome_context_intronic_sequence'] == ''
        assert results['NM_014249.2:c.947delinsTT']['hgvs_lrg_variant'] == ''
        assert results['NM_014249.2:c.947delinsTT']['hgvs_transcript_variant'] == 'NM_014249.2:c.947delinsTT'
        assert results['NM_014249.2:c.947delinsTT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_014249.2:c.947delinsTT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105928_72105929delinsTT', 'vcf': {'chr': 'chr15', 'ref': 'AC', 'pos': '72105928', 'alt': 'TT'}}
        assert 'hg38' not in results['NM_014249.2:c.947delinsTT']['primary_assembly_loci'].keys()
        assert results['NM_014249.2:c.947delinsTT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105928_72105929delinsTT', 'vcf': {'chr': '15', 'ref': 'AC', 'pos': '72105928', 'alt': 'TT'}}
        assert 'grch38' not in results['NM_014249.2:c.947delinsTT']['primary_assembly_loci'].keys()

        assert results['flag'] == 'gene_variant'
        assert 'NM_014249.3:c.947delinsTT' in results.keys()
        assert results['NM_014249.3:c.947delinsTT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_014249.3:c.947delinsTT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_014249.3:c.947delinsTT']['alt_genomic_loci'] == []
        assert results['NM_014249.3:c.947delinsTT']['gene_symbol'] == 'NR2E3'
        assert results['NM_014249.3:c.947delinsTT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_055064.1:p.(Asp316ValfsTer25)', 'slr': 'NP_055064.1:p.(D316Vfs*25)'}
        assert results['NM_014249.3:c.947delinsTT']['submitted_variant'] == '15-72105927-GACC-GTT'
        assert results['NM_014249.3:c.947delinsTT']['genome_context_intronic_sequence'] == ''
        assert results['NM_014249.3:c.947delinsTT']['hgvs_lrg_variant'] == ''
        assert results['NM_014249.3:c.947delinsTT']['hgvs_transcript_variant'] == 'NM_014249.3:c.947delinsTT'
        assert results['NM_014249.3:c.947delinsTT']['hgvs_refseqgene_variant'] == 'NG_009113.1:g.8035delinsTT'
        assert results['NM_014249.3:c.947delinsTT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105928_72105929delinsTT', 'vcf': {'chr': 'chr15', 'ref': 'AC', 'pos': '72105928', 'alt': 'TT'}}
        assert results['NM_014249.3:c.947delinsTT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000015.10:g.71813588delinsTT', 'vcf': {'chr': 'chr15', 'ref': 'A', 'pos': '71813588', 'alt': u'TT'}}
        assert results['NM_014249.3:c.947delinsTT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105928_72105929delinsTT', 'vcf': {'chr': '15', 'ref': 'AC', 'pos': '72105928', 'alt': 'TT'}}
        assert results['NM_014249.3:c.947delinsTT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000015.10:g.71813588delinsTT', 'vcf': {'chr': '15', 'ref': 'A', 'pos': '71813588', 'alt': u'TT'}}

        assert 'NM_016346.2:c.947delinsTT' in results.keys()
        assert results['NM_016346.2:c.947delinsTT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_016346.2:c.947delinsTT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_016346.2:c.947delinsTT']['alt_genomic_loci'] == []
        assert results['NM_016346.2:c.947delinsTT']['gene_symbol'] == 'NR2E3'
        assert results['NM_016346.2:c.947delinsTT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_057430.1:p.(Asp316ValfsTer25)', 'slr': 'NP_057430.1:p.(D316Vfs*25)'}
        assert results['NM_016346.2:c.947delinsTT']['submitted_variant'] == '15-72105927-GACC-GTT'
        assert results['NM_016346.2:c.947delinsTT']['genome_context_intronic_sequence'] == ''
        assert results['NM_016346.2:c.947delinsTT']['hgvs_lrg_variant'] == ''
        assert results['NM_016346.2:c.947delinsTT']['hgvs_transcript_variant'] == 'NM_016346.2:c.947delinsTT'
        assert results['NM_016346.2:c.947delinsTT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_016346.2:c.947delinsTT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105928_72105929delinsTT', 'vcf': {'chr': 'chr15', 'ref': 'AC', 'pos': '72105928', 'alt': 'TT'}}
        assert 'hg38' not in results['NM_016346.2:c.947delinsTT']['primary_assembly_loci'].keys()
        assert results['NM_016346.2:c.947delinsTT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105928_72105929delinsTT', 'vcf': {'chr': '15', 'ref': 'AC', 'pos': '72105928', 'alt': 'TT'}}
        assert 'grch38' not in results['NM_016346.2:c.947delinsTT']['primary_assembly_loci'].keys()


    def test_variant185(self):
        variant = '19-41123093-A-AG'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_001042544.1:c.3233_3235=' in results.keys()
        assert results['NM_001042544.1:c.3233_3235=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001042544.1:c.3233_3235=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001042544.1:c.3233_3235=']['alt_genomic_loci'] == []
        assert results['NM_001042544.1:c.3233_3235=']['gene_symbol'] == 'LTBP4'
        assert results['NM_001042544.1:c.3233_3235=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001036009.1:p.(Gln1078=)', 'slr': 'NP_001036009.1:p.(Q1078=)'}
        assert results['NM_001042544.1:c.3233_3235=']['submitted_variant'] == '19-41123093-A-AG'
        assert results['NM_001042544.1:c.3233_3235=']['genome_context_intronic_sequence'] == ''
        assert results['NM_001042544.1:c.3233_3235=']['hgvs_lrg_variant'] == ''
        assert results['NM_001042544.1:c.3233_3235=']['hgvs_transcript_variant'] == 'NM_001042544.1:c.3233_3235='
        assert results['NM_001042544.1:c.3233_3235=']['hgvs_refseqgene_variant'] == 'NG_021201.1:g.29022_29024='
        assert results['NM_001042544.1:c.3233_3235=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.41123095dup', 'vcf': {'chr': 'chr19', 'ref': 'G', 'pos': '41123094', 'alt': u'GG'}}
        assert results['NM_001042544.1:c.3233_3235=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.40617187_40617189=', 'vcf': {'chr': 'chr19', 'ref': u'AGG', 'pos': '40617187', 'alt': u'AGG'}}
        assert results['NM_001042544.1:c.3233_3235=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.41123095dup', 'vcf': {'chr': '19', 'ref': 'G', 'pos': '41123094', 'alt': u'GG'}}
        assert results['NM_001042544.1:c.3233_3235=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.40617187_40617189=', 'vcf': {'chr': '19', 'ref': u'AGG', 'pos': '40617187', 'alt': u'AGG'}}

        assert 'NM_001042545.1:c.3032_3034=' in results.keys()
        assert results['NM_001042545.1:c.3032_3034=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001042545.1:c.3032_3034=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001042545.1:c.3032_3034=']['alt_genomic_loci'] == []
        assert results['NM_001042545.1:c.3032_3034=']['gene_symbol'] == 'LTBP4'
        assert results['NM_001042545.1:c.3032_3034=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001036010.1:p.(Gln1011=)', 'slr': 'NP_001036010.1:p.(Q1011=)'}
        assert results['NM_001042545.1:c.3032_3034=']['submitted_variant'] == '19-41123093-A-AG'
        assert results['NM_001042545.1:c.3032_3034=']['genome_context_intronic_sequence'] == ''
        assert results['NM_001042545.1:c.3032_3034=']['hgvs_lrg_variant'] == ''
        assert results['NM_001042545.1:c.3032_3034=']['hgvs_transcript_variant'] == 'NM_001042545.1:c.3032_3034='
        assert results['NM_001042545.1:c.3032_3034=']['hgvs_refseqgene_variant'] == 'NG_021201.1:g.29022_29024='
        assert results['NM_001042545.1:c.3032_3034=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.41123095dup', 'vcf': {'chr': 'chr19', 'ref': 'G', 'pos': '41123094', 'alt': u'GG'}}
        assert results['NM_001042545.1:c.3032_3034=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.40617187_40617189=', 'vcf': {'chr': 'chr19', 'ref': u'AGG', 'pos': '40617187', 'alt': u'AGG'}}
        assert results['NM_001042545.1:c.3032_3034=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.41123095dup', 'vcf': {'chr': '19', 'ref': 'G', 'pos': '41123094', 'alt': u'GG'}}
        assert results['NM_001042545.1:c.3032_3034=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.40617187_40617189=', 'vcf': {'chr': '19', 'ref': u'AGG', 'pos': '40617187', 'alt': u'AGG'}}

        assert 'NM_003573.2:c.3122_3124=' in results.keys()
        assert results['NM_003573.2:c.3122_3124=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003573.2:c.3122_3124=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003573.2:c.3122_3124=']['alt_genomic_loci'] == []
        assert results['NM_003573.2:c.3122_3124=']['gene_symbol'] == 'LTBP4'
        assert results['NM_003573.2:c.3122_3124=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003564.2:p.(Gln1041=)', 'slr': 'NP_003564.2:p.(Q1041=)'}
        assert results['NM_003573.2:c.3122_3124=']['submitted_variant'] == '19-41123093-A-AG'
        assert results['NM_003573.2:c.3122_3124=']['genome_context_intronic_sequence'] == ''
        assert results['NM_003573.2:c.3122_3124=']['hgvs_lrg_variant'] == ''
        assert results['NM_003573.2:c.3122_3124=']['hgvs_transcript_variant'] == 'NM_003573.2:c.3122_3124='
        assert results['NM_003573.2:c.3122_3124=']['hgvs_refseqgene_variant'] == 'NG_021201.1:g.29022_29024='
        assert results['NM_003573.2:c.3122_3124=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.41123095dup', 'vcf': {'chr': 'chr19', 'ref': 'G', 'pos': '41123094', 'alt': u'GG'}}
        assert results['NM_003573.2:c.3122_3124=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.40617187_40617189=', 'vcf': {'chr': 'chr19', 'ref': u'AGG', 'pos': '40617187', 'alt': u'AGG'}}
        assert results['NM_003573.2:c.3122_3124=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.41123095dup', 'vcf': {'chr': '19', 'ref': 'G', 'pos': '41123094', 'alt': u'GG'}}
        assert results['NM_003573.2:c.3122_3124=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.40617187_40617189=', 'vcf': {'chr': '19', 'ref': u'AGG', 'pos': '40617187', 'alt': u'AGG'}}


    def test_variant186(self):
        variant = '19-41123093-A-AT'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_003573.2:c.3123G>T' in results.keys()
        assert results['NM_003573.2:c.3123G>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003573.2:c.3123G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003573.2:c.3123G>T']['alt_genomic_loci'] == []
        assert results['NM_003573.2:c.3123G>T']['gene_symbol'] == 'LTBP4'
        assert results['NM_003573.2:c.3123G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003564.2:p.(Gln1041His)', 'slr': 'NP_003564.2:p.(Q1041H)'}
        assert results['NM_003573.2:c.3123G>T']['submitted_variant'] == '19-41123093-A-AT'
        assert results['NM_003573.2:c.3123G>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_003573.2:c.3123G>T']['hgvs_lrg_variant'] == ''
        assert results['NM_003573.2:c.3123G>T']['hgvs_transcript_variant'] == 'NM_003573.2:c.3123G>T'
        assert results['NM_003573.2:c.3123G>T']['hgvs_refseqgene_variant'] == 'NG_021201.1:g.29023G>T'
        assert results['NM_003573.2:c.3123G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.41123093_41123094insT', 'vcf': {'chr': 'chr19', 'ref': 'A', 'pos': '41123093', 'alt': 'AT'}}
        assert results['NM_003573.2:c.3123G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.40617188G>T', 'vcf': {'chr': 'chr19', 'ref': u'G', 'pos': '40617188', 'alt': u'T'}}
        assert results['NM_003573.2:c.3123G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.41123093_41123094insT', 'vcf': {'chr': '19', 'ref': 'A', 'pos': '41123093', 'alt': 'AT'}}
        assert results['NM_003573.2:c.3123G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.40617188G>T', 'vcf': {'chr': '19', 'ref': u'G', 'pos': '40617188', 'alt': u'T'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001042545.1:c.3033G>T' in results.keys()
        assert results['NM_001042545.1:c.3033G>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001042545.1:c.3033G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001042545.1:c.3033G>T']['alt_genomic_loci'] == []
        assert results['NM_001042545.1:c.3033G>T']['gene_symbol'] == 'LTBP4'
        assert results['NM_001042545.1:c.3033G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001036010.1:p.(Gln1011His)', 'slr': 'NP_001036010.1:p.(Q1011H)'}
        assert results['NM_001042545.1:c.3033G>T']['submitted_variant'] == '19-41123093-A-AT'
        assert results['NM_001042545.1:c.3033G>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001042545.1:c.3033G>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001042545.1:c.3033G>T']['hgvs_transcript_variant'] == 'NM_001042545.1:c.3033G>T'
        assert results['NM_001042545.1:c.3033G>T']['hgvs_refseqgene_variant'] == 'NG_021201.1:g.29023G>T'
        assert results['NM_001042545.1:c.3033G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.41123093_41123094insT', 'vcf': {'chr': 'chr19', 'ref': 'A', 'pos': '41123093', 'alt': 'AT'}}
        assert results['NM_001042545.1:c.3033G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.40617188G>T', 'vcf': {'chr': 'chr19', 'ref': u'G', 'pos': '40617188', 'alt': u'T'}}
        assert results['NM_001042545.1:c.3033G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.41123093_41123094insT', 'vcf': {'chr': '19', 'ref': 'A', 'pos': '41123093', 'alt': 'AT'}}
        assert results['NM_001042545.1:c.3033G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.40617188G>T', 'vcf': {'chr': '19', 'ref': u'G', 'pos': '40617188', 'alt': u'T'}}

        assert 'NM_001042544.1:c.3234G>T' in results.keys()
        assert results['NM_001042544.1:c.3234G>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001042544.1:c.3234G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001042544.1:c.3234G>T']['alt_genomic_loci'] == []
        assert results['NM_001042544.1:c.3234G>T']['gene_symbol'] == 'LTBP4'
        assert results['NM_001042544.1:c.3234G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001036009.1:p.(Gln1078His)', 'slr': 'NP_001036009.1:p.(Q1078H)'}
        assert results['NM_001042544.1:c.3234G>T']['submitted_variant'] == '19-41123093-A-AT'
        assert results['NM_001042544.1:c.3234G>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001042544.1:c.3234G>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001042544.1:c.3234G>T']['hgvs_transcript_variant'] == 'NM_001042544.1:c.3234G>T'
        assert results['NM_001042544.1:c.3234G>T']['hgvs_refseqgene_variant'] == 'NG_021201.1:g.29023G>T'
        assert results['NM_001042544.1:c.3234G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.41123093_41123094insT', 'vcf': {'chr': 'chr19', 'ref': 'A', 'pos': '41123093', 'alt': 'AT'}}
        assert results['NM_001042544.1:c.3234G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.40617188G>T', 'vcf': {'chr': 'chr19', 'ref': u'G', 'pos': '40617188', 'alt': u'T'}}
        assert results['NM_001042544.1:c.3234G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.41123093_41123094insT', 'vcf': {'chr': '19', 'ref': 'A', 'pos': '41123093', 'alt': 'AT'}}
        assert results['NM_001042544.1:c.3234G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.40617188G>T', 'vcf': {'chr': '19', 'ref': u'G', 'pos': '40617188', 'alt': u'T'}}


    def test_variant187(self):
        variant = '19-41123093-AG-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001042544.1:c.3235_3236del' in results.keys()
        assert results['NM_001042544.1:c.3235_3236del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001042544.1:c.3235_3236del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001042544.1:c.3235_3236del']['alt_genomic_loci'] == []
        assert results['NM_001042544.1:c.3235_3236del']['gene_symbol'] == 'LTBP4'
        assert results['NM_001042544.1:c.3235_3236del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001036009.1:p.(Gly1079LeufsTer17)', 'slr': 'NP_001036009.1:p.(G1079Lfs*17)'}
        assert results['NM_001042544.1:c.3235_3236del']['submitted_variant'] == '19-41123093-AG-A'
        assert results['NM_001042544.1:c.3235_3236del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001042544.1:c.3235_3236del']['hgvs_lrg_variant'] == ''
        assert results['NM_001042544.1:c.3235_3236del']['hgvs_transcript_variant'] == 'NM_001042544.1:c.3235_3236del'
        assert results['NM_001042544.1:c.3235_3236del']['hgvs_refseqgene_variant'] == 'NG_021201.1:g.29024_29025del'
        assert results['NM_001042544.1:c.3235_3236del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.41123095del', 'vcf': {'chr': 'chr19', 'ref': 'AG', 'pos': '41123093', 'alt': 'A'}}
        assert results['NM_001042544.1:c.3235_3236del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.40617189_40617190del', 'vcf': {'chr': 'chr19', 'ref': 'AGG', 'pos': '40617187', 'alt': 'A'}}
        assert results['NM_001042544.1:c.3235_3236del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.41123095del', 'vcf': {'chr': '19', 'ref': 'AG', 'pos': '41123093', 'alt': 'A'}}
        assert results['NM_001042544.1:c.3235_3236del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.40617189_40617190del', 'vcf': {'chr': '19', 'ref': 'AGG', 'pos': '40617187', 'alt': 'A'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001042545.1:c.3034_3035del' in results.keys()
        assert results['NM_001042545.1:c.3034_3035del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001042545.1:c.3034_3035del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001042545.1:c.3034_3035del']['alt_genomic_loci'] == []
        assert results['NM_001042545.1:c.3034_3035del']['gene_symbol'] == 'LTBP4'
        assert results['NM_001042545.1:c.3034_3035del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001036010.1:p.(Gly1012LeufsTer17)', 'slr': 'NP_001036010.1:p.(G1012Lfs*17)'}
        assert results['NM_001042545.1:c.3034_3035del']['submitted_variant'] == '19-41123093-AG-A'
        assert results['NM_001042545.1:c.3034_3035del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001042545.1:c.3034_3035del']['hgvs_lrg_variant'] == ''
        assert results['NM_001042545.1:c.3034_3035del']['hgvs_transcript_variant'] == 'NM_001042545.1:c.3034_3035del'
        assert results['NM_001042545.1:c.3034_3035del']['hgvs_refseqgene_variant'] == 'NG_021201.1:g.29024_29025del'
        assert results['NM_001042545.1:c.3034_3035del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.41123095del', 'vcf': {'chr': 'chr19', 'ref': 'AG', 'pos': '41123093', 'alt': 'A'}}
        assert results['NM_001042545.1:c.3034_3035del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.40617189_40617190del', 'vcf': {'chr': 'chr19', 'ref': 'AGG', 'pos': '40617187', 'alt': 'A'}}
        assert results['NM_001042545.1:c.3034_3035del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.41123095del', 'vcf': {'chr': '19', 'ref': 'AG', 'pos': '41123093', 'alt': 'A'}}
        assert results['NM_001042545.1:c.3034_3035del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.40617189_40617190del', 'vcf': {'chr': '19', 'ref': 'AGG', 'pos': '40617187', 'alt': 'A'}}

        assert 'NM_003573.2:c.3124_3125del' in results.keys()
        assert results['NM_003573.2:c.3124_3125del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003573.2:c.3124_3125del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003573.2:c.3124_3125del']['alt_genomic_loci'] == []
        assert results['NM_003573.2:c.3124_3125del']['gene_symbol'] == 'LTBP4'
        assert results['NM_003573.2:c.3124_3125del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003564.2:p.(Gly1042LeufsTer17)', 'slr': 'NP_003564.2:p.(G1042Lfs*17)'}
        assert results['NM_003573.2:c.3124_3125del']['submitted_variant'] == '19-41123093-AG-A'
        assert results['NM_003573.2:c.3124_3125del']['genome_context_intronic_sequence'] == ''
        assert results['NM_003573.2:c.3124_3125del']['hgvs_lrg_variant'] == ''
        assert results['NM_003573.2:c.3124_3125del']['hgvs_transcript_variant'] == 'NM_003573.2:c.3124_3125del'
        assert results['NM_003573.2:c.3124_3125del']['hgvs_refseqgene_variant'] == 'NG_021201.1:g.29024_29025del'
        assert results['NM_003573.2:c.3124_3125del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.41123095del', 'vcf': {'chr': 'chr19', 'ref': 'AG', 'pos': '41123093', 'alt': 'A'}}
        assert results['NM_003573.2:c.3124_3125del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.40617189_40617190del', 'vcf': {'chr': 'chr19', 'ref': 'AGG', 'pos': '40617187', 'alt': 'A'}}
        assert results['NM_003573.2:c.3124_3125del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.41123095del', 'vcf': {'chr': '19', 'ref': 'AG', 'pos': '41123093', 'alt': 'A'}}
        assert results['NM_003573.2:c.3124_3125del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.40617189_40617190del', 'vcf': {'chr': '19', 'ref': 'AGG', 'pos': '40617187', 'alt': 'A'}}


    def test_variant188(self):
        variant = '19-41123093-AG-AG'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001042545.1:c.3035del' in results.keys()
        assert results['NM_001042545.1:c.3035del']['hgvs_lrg_transcript_variant'] == 'NM_000088.3:c.589G>T'
        assert results['NM_001042545.1:c.3035del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001042545.1:c.3035del']['alt_genomic_loci'] == []
        assert results['NM_001042545.1:c.3035del']['gene_symbol'] == 'LTBP4'
        assert results['NM_001042545.1:c.3035del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001036010.1:p.(Gly1012ValfsTer14)', 'slr': 'NP_001036010.1:p.(G1012Vfs*14)'}
        assert results['NM_001042545.1:c.3035del']['submitted_variant'] == '19-41123093-AG-AG'
        assert results['NM_001042545.1:c.3035del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001042545.1:c.3035del']['hgvs_lrg_variant'] == ''
        assert results['NM_001042545.1:c.3035del']['hgvs_transcript_variant'] == 'NM_001042545.1:c.3035del'
        assert results['NM_001042545.1:c.3035del']['hgvs_refseqgene_variant'] == 'NG_021201.1:g.29025del'
        assert results['NM_001042545.1:c.3035del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.41123094G=', 'vcf': {'chr': 'chr19', 'ref': 'G', 'pos': '41123094', 'alt': 'G'}}
        assert results['NM_001042545.1:c.3035del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.40617190del', 'vcf': {'chr': 'chr19', 'ref': 'AG', 'pos': '40617187', 'alt': 'A'}}
        assert results['NM_001042545.1:c.3035del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.41123094G=', 'vcf': {'chr': '19', 'ref': 'G', 'pos': '41123094', 'alt': 'G'}}
        assert results['NM_001042545.1:c.3035del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.40617190del', 'vcf': {'chr': '19', 'ref': 'AG', 'pos': '40617187', 'alt': 'A'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001042544.1:c.3236del' in results.keys()
        assert results['NM_001042544.1:c.3236del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001042544.1:c.3236del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001042544.1:c.3236del']['alt_genomic_loci'] == []
        assert results['NM_001042544.1:c.3236del']['gene_symbol'] == 'LTBP4'
        assert results['NM_001042544.1:c.3236del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001036009.1:p.(Gly1079ValfsTer14)', 'slr': 'NP_001036009.1:p.(G1079Vfs*14)'}
        assert results['NM_001042544.1:c.3236del']['submitted_variant'] == '19-41123093-AG-AG'
        assert results['NM_001042544.1:c.3236del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001042544.1:c.3236del']['hgvs_lrg_variant'] == ''
        assert results['NM_001042544.1:c.3236del']['hgvs_transcript_variant'] == 'NM_001042544.1:c.3236del'
        assert results['NM_001042544.1:c.3236del']['hgvs_refseqgene_variant'] == 'NG_021201.1:g.29025del'
        assert results['NM_001042544.1:c.3236del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.41123094G=', 'vcf': {'chr': 'chr19', 'ref': 'G', 'pos': '41123094', 'alt': 'G'}}
        assert results['NM_001042544.1:c.3236del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.40617190del', 'vcf': {'chr': 'chr19', 'ref': 'AG', 'pos': '40617187', 'alt': 'A'}}
        assert results['NM_001042544.1:c.3236del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.41123094G=', 'vcf': {'chr': '19', 'ref': 'G', 'pos': '41123094', 'alt': 'G'}}
        assert results['NM_001042544.1:c.3236del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.40617190del', 'vcf': {'chr': '19', 'ref': 'AG', 'pos': '40617187', 'alt': 'A'}}

        assert 'NM_003573.2:c.3125del' in results.keys()
        assert results['NM_003573.2:c.3125del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003573.2:c.3125del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003573.2:c.3125del']['alt_genomic_loci'] == []
        assert results['NM_003573.2:c.3125del']['gene_symbol'] == 'LTBP4'
        assert results['NM_003573.2:c.3125del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003564.2:p.(Gly1042ValfsTer14)', 'slr': 'NP_003564.2:p.(G1042Vfs*14)'}
        assert results['NM_003573.2:c.3125del']['submitted_variant'] == '19-41123093-AG-AG'
        assert results['NM_003573.2:c.3125del']['genome_context_intronic_sequence'] == ''
        assert results['NM_003573.2:c.3125del']['hgvs_lrg_variant'] == ''
        assert results['NM_003573.2:c.3125del']['hgvs_transcript_variant'] == 'NM_003573.2:c.3125del'
        assert results['NM_003573.2:c.3125del']['hgvs_refseqgene_variant'] == 'NG_021201.1:g.29025del'
        assert results['NM_003573.2:c.3125del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.41123094G=', 'vcf': {'chr': 'chr19', 'ref': 'G', 'pos': '41123094', 'alt': 'G'}}
        assert results['NM_003573.2:c.3125del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.40617190del', 'vcf': {'chr': 'chr19', 'ref': 'AG', 'pos': '40617187', 'alt': 'A'}}
        assert results['NM_003573.2:c.3125del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.41123094G=', 'vcf': {'chr': '19', 'ref': 'G', 'pos': '41123094', 'alt': 'G'}}
        assert results['NM_003573.2:c.3125del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.40617190del', 'vcf': {'chr': '19', 'ref': 'AG', 'pos': '40617187', 'alt': 'A'}}


    def test_variant189(self):
        variant = 'NM_012309.4:c.913-5058G>A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_012309.4:c.913-5058G>A' in results.keys()
        assert results['NM_012309.4:c.913-5058G>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_012309.4:c.913-5058G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_012309.4:c.913-5058G>A']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_004070871.1:g.574546C>T', 'vcf': {'chr': 'HG865_PATCH', 'ref': u'C', 'pos': '574546', 'alt': u'T'}}}]
        assert results['NM_012309.4:c.913-5058G>A']['gene_symbol'] == 'SHANK2'
        assert results['NM_012309.4:c.913-5058G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_036441.2:p.?', 'slr': 'NP_036441.2:p.?'}
        assert results['NM_012309.4:c.913-5058G>A']['submitted_variant'] == 'NM_012309.4:c.913-5058G>A'
        assert results['NM_012309.4:c.913-5058G>A']['genome_context_intronic_sequence'] == 'NC_000011.10(NM_012309.4):c.913-5058G>A'
        assert results['NM_012309.4:c.913-5058G>A']['hgvs_lrg_variant'] == ''
        assert results['NM_012309.4:c.913-5058G>A']['hgvs_transcript_variant'] == 'NM_012309.4:c.913-5058G>A'
        assert results['NM_012309.4:c.913-5058G>A']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['NM_012309.4:c.913-5058G>A']['primary_assembly_loci'].keys()
        assert results['NM_012309.4:c.913-5058G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.71080333C>T', 'vcf': {'chr': 'chr11', 'ref': u'C', 'pos': '71080333', 'alt': u'T'}}
        assert 'grch37' not in results['NM_012309.4:c.913-5058G>A']['primary_assembly_loci'].keys()
        assert results['NM_012309.4:c.913-5058G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.71080333C>T', 'vcf': {'chr': '11', 'ref': u'C', 'pos': '71080333', 'alt': u'T'}}


    def test_variant190(self):
        variant = 'LRG_199t1:c.2376[G>C];[G>C]'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_004006.2:c.2376G>C ' in results.keys()
        assert results['NM_004006.2:c.2376G>C ']['hgvs_lrg_transcript_variant'] == 'LRG_199t1:c.2376G>C'
        assert results['NM_004006.2:c.2376G>C ']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_004006.2:c.2376G>C ']['alt_genomic_loci'] == []
        assert results['NM_004006.2:c.2376G>C ']['gene_symbol'] == 'DMD'
        assert results['NM_004006.2:c.2376G>C ']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003997.1(LRG_199p1):p.(Val792=)', 'slr': 'NP_003997.1:p.(V792=)'}
        assert results['NM_004006.2:c.2376G>C ']['submitted_variant'] == 'LRG_199t1:c.2376[G>C];[G>C]'
        assert results['NM_004006.2:c.2376G>C ']['genome_context_intronic_sequence'] == ''
        assert results['NM_004006.2:c.2376G>C ']['hgvs_lrg_variant'] == 'LRG_199:g.842851G>C'
        assert results['NM_004006.2:c.2376G>C ']['hgvs_transcript_variant'] == 'NM_004006.2:c.2376G>C'
        assert results['NM_004006.2:c.2376G>C ']['hgvs_refseqgene_variant'] == 'NG_012232.1:g.842851G>C'
        assert results['NM_004006.2:c.2376G>C ']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.32519876C>G', 'vcf': {'chr': 'chrX', 'ref': u'C', 'pos': '32519876', 'alt': u'G'}}
        assert results['NM_004006.2:c.2376G>C ']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.32501759C>G', 'vcf': {'chr': 'chrX', 'ref': u'C', 'pos': '32501759', 'alt': u'G'}}
        assert results['NM_004006.2:c.2376G>C ']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.32519876C>G', 'vcf': {'chr': 'X', 'ref': u'C', 'pos': '32519876', 'alt': u'G'}}
        assert results['NM_004006.2:c.2376G>C ']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.32501759C>G', 'vcf': {'chr': 'X', 'ref': u'C', 'pos': '32501759', 'alt': u'G'}}

        assert 'NM_004006.2:c.2376G>C' in results.keys()
        assert results['NM_004006.2:c.2376G>C']['hgvs_lrg_transcript_variant'] == 'LRG_199t1:c.2376G>C'
        assert results['NM_004006.2:c.2376G>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_004006.2:c.2376G>C']['alt_genomic_loci'] == []
        assert results['NM_004006.2:c.2376G>C']['gene_symbol'] == 'DMD'
        assert results['NM_004006.2:c.2376G>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003997.1(LRG_199p1):p.(Val792=)', 'slr': 'NP_003997.1:p.(V792=)'}
        assert results['NM_004006.2:c.2376G>C']['submitted_variant'] == 'LRG_199t1:c.2376[G>C];[G>C]'
        assert results['NM_004006.2:c.2376G>C']['genome_context_intronic_sequence'] == ''
        assert results['NM_004006.2:c.2376G>C']['hgvs_lrg_variant'] == 'LRG_199:g.842851G>C'
        assert results['NM_004006.2:c.2376G>C']['hgvs_transcript_variant'] == 'NM_004006.2:c.2376G>C'
        assert results['NM_004006.2:c.2376G>C']['hgvs_refseqgene_variant'] == 'NG_012232.1:g.842851G>C'
        assert results['NM_004006.2:c.2376G>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.32519876C>G', 'vcf': {'chr': 'chrX', 'ref': u'C', 'pos': '32519876', 'alt': u'G'}}
        assert results['NM_004006.2:c.2376G>C']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.32501759C>G', 'vcf': {'chr': 'chrX', 'ref': u'C', 'pos': '32501759', 'alt': u'G'}}
        assert results['NM_004006.2:c.2376G>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.32519876C>G', 'vcf': {'chr': 'X', 'ref': u'C', 'pos': '32519876', 'alt': u'G'}}
        assert results['NM_004006.2:c.2376G>C']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.32501759C>G', 'vcf': {'chr': 'X', 'ref': u'C', 'pos': '32501759', 'alt': u'G'}}


    def test_variant191(self):
        variant = 'LRG_199t1:c.[2376G>C];[3103del]'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_004006.2:c.3103del' in results.keys()
        assert results['NM_004006.2:c.3103del']['hgvs_lrg_transcript_variant'] == 'LRG_199t1:c.3103del'
        assert results['NM_004006.2:c.3103del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_004006.2:c.3103del']['alt_genomic_loci'] == []
        assert results['NM_004006.2:c.3103del']['gene_symbol'] == 'DMD'
        assert results['NM_004006.2:c.3103del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003997.1(LRG_199p1):p.(Gln1035SerfsTer9)', 'slr': 'NP_003997.1:p.(Q1035Sfs*9)'}
        assert results['NM_004006.2:c.3103del']['submitted_variant'] == 'LRG_199t1:c.[2376G>C];[3103del]'
        assert results['NM_004006.2:c.3103del']['genome_context_intronic_sequence'] == ''
        assert results['NM_004006.2:c.3103del']['hgvs_lrg_variant'] == 'LRG_199:g.876053del'
        assert results['NM_004006.2:c.3103del']['hgvs_transcript_variant'] == 'NM_004006.2:c.3103del'
        assert results['NM_004006.2:c.3103del']['hgvs_refseqgene_variant'] == 'NG_012232.1:g.876053del'
        assert results['NM_004006.2:c.3103del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.32486674del', 'vcf': {'chr': 'chrX', 'ref': 'TG', 'pos': '32486673', 'alt': 'T'}}
        assert results['NM_004006.2:c.3103del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.32468557del', 'vcf': {'chr': 'chrX', 'ref': 'TG', 'pos': '32468556', 'alt': 'T'}}
        assert results['NM_004006.2:c.3103del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.32486674del', 'vcf': {'chr': 'X', 'ref': 'TG', 'pos': '32486673', 'alt': 'T'}}
        assert results['NM_004006.2:c.3103del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.32468557del', 'vcf': {'chr': 'X', 'ref': 'TG', 'pos': '32468556', 'alt': 'T'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_004006.2:c.2376G>C' in results.keys()
        assert results['NM_004006.2:c.2376G>C']['hgvs_lrg_transcript_variant'] == 'LRG_199t1:c.2376G>C'
        assert results['NM_004006.2:c.2376G>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_004006.2:c.2376G>C']['alt_genomic_loci'] == []
        assert results['NM_004006.2:c.2376G>C']['gene_symbol'] == 'DMD'
        assert results['NM_004006.2:c.2376G>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003997.1(LRG_199p1):p.(Val792=)', 'slr': 'NP_003997.1:p.(V792=)'}
        assert results['NM_004006.2:c.2376G>C']['submitted_variant'] == 'LRG_199t1:c.[2376G>C];[3103del]'
        assert results['NM_004006.2:c.2376G>C']['genome_context_intronic_sequence'] == ''
        assert results['NM_004006.2:c.2376G>C']['hgvs_lrg_variant'] == 'LRG_199:g.842851G>C'
        assert results['NM_004006.2:c.2376G>C']['hgvs_transcript_variant'] == 'NM_004006.2:c.2376G>C'
        assert results['NM_004006.2:c.2376G>C']['hgvs_refseqgene_variant'] == 'NG_012232.1:g.842851G>C'
        assert results['NM_004006.2:c.2376G>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.32519876C>G', 'vcf': {'chr': 'chrX', 'ref': u'C', 'pos': '32519876', 'alt': u'G'}}
        assert results['NM_004006.2:c.2376G>C']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.32501759C>G', 'vcf': {'chr': 'chrX', 'ref': u'C', 'pos': '32501759', 'alt': u'G'}}
        assert results['NM_004006.2:c.2376G>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.32519876C>G', 'vcf': {'chr': 'X', 'ref': u'C', 'pos': '32519876', 'alt': u'G'}}
        assert results['NM_004006.2:c.2376G>C']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.32501759C>G', 'vcf': {'chr': 'X', 'ref': u'C', 'pos': '32501759', 'alt': u'G'}}


    def test_variant192(self):
        variant = 'LRG_199t1:c.[4358_4359del;4361_4372del]'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_004006.2:c.4358_4372delinsG' in results.keys()
        assert results['NM_004006.2:c.4358_4372delinsG']['hgvs_lrg_transcript_variant'] == 'LRG_199t1:c.4358_4372delinsG'
        assert results['NM_004006.2:c.4358_4372delinsG']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_004006.2:c.4358_4372delinsG']['alt_genomic_loci'] == []
        assert results['NM_004006.2:c.4358_4372delinsG']['gene_symbol'] == 'DMD'
        assert results['NM_004006.2:c.4358_4372delinsG']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003997.1(LRG_199p1):p.(Asp1453GlyfsTer11)', 'slr': 'NP_003997.1:p.(D1453Gfs*11)'}
        assert results['NM_004006.2:c.4358_4372delinsG']['submitted_variant'] == 'LRG_199t1:c.[4358_4359del;4361_4372del]'
        assert results['NM_004006.2:c.4358_4372delinsG']['genome_context_intronic_sequence'] == ''
        assert results['NM_004006.2:c.4358_4372delinsG']['hgvs_lrg_variant'] == 'LRG_199:g.954949_954963delinsG'
        assert results['NM_004006.2:c.4358_4372delinsG']['hgvs_transcript_variant'] == 'NM_004006.2:c.4358_4372delinsG'
        assert results['NM_004006.2:c.4358_4372delinsG']['hgvs_refseqgene_variant'] == 'NG_012232.1:g.954949_954963delinsG'
        assert results['NM_004006.2:c.4358_4372delinsG']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.32407764_32407778delinsC', 'vcf': {'chr': 'chrX', 'ref': 'ACTTCATGGAGACAT', 'pos': '32407764', 'alt': u'C'}}
        assert results['NM_004006.2:c.4358_4372delinsG']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.32389647_32389661delinsC', 'vcf': {'chr': 'chrX', 'ref': 'ACTTCATGGAGACAT', 'pos': '32389647', 'alt': u'C'}}
        assert results['NM_004006.2:c.4358_4372delinsG']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.32407764_32407778delinsC', 'vcf': {'chr': 'X', 'ref': 'ACTTCATGGAGACAT', 'pos': '32407764', 'alt': u'C'}}
        assert results['NM_004006.2:c.4358_4372delinsG']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.32389647_32389661delinsC', 'vcf': {'chr': 'X', 'ref': 'ACTTCATGGAGACAT', 'pos': '32389647', 'alt': u'C'}}


    def test_variant193(self):
        variant = 'LRG_199t1:c.2376G>C(;)3103del'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_004006.2:c.3103del' in results.keys()
        assert results['NM_004006.2:c.3103del']['hgvs_lrg_transcript_variant'] == 'LRG_199t1:c.3103del'
        assert results['NM_004006.2:c.3103del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_004006.2:c.3103del']['alt_genomic_loci'] == []
        assert results['NM_004006.2:c.3103del']['gene_symbol'] == 'DMD'
        assert results['NM_004006.2:c.3103del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003997.1(LRG_199p1):p.(Gln1035SerfsTer9)', 'slr': 'NP_003997.1:p.(Q1035Sfs*9)'}
        assert results['NM_004006.2:c.3103del']['submitted_variant'] == 'LRG_199t1:c.2376G>C(;)3103del'
        assert results['NM_004006.2:c.3103del']['genome_context_intronic_sequence'] == ''
        assert results['NM_004006.2:c.3103del']['hgvs_lrg_variant'] == 'LRG_199:g.876053del'
        assert results['NM_004006.2:c.3103del']['hgvs_transcript_variant'] == 'NM_004006.2:c.3103del'
        assert results['NM_004006.2:c.3103del']['hgvs_refseqgene_variant'] == 'NG_012232.1:g.876053del'
        assert results['NM_004006.2:c.3103del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.32486674del', 'vcf': {'chr': 'chrX', 'ref': 'TG', 'pos': '32486673', 'alt': 'T'}}
        assert results['NM_004006.2:c.3103del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.32468557del', 'vcf': {'chr': 'chrX', 'ref': 'TG', 'pos': '32468556', 'alt': 'T'}}
        assert results['NM_004006.2:c.3103del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.32486674del', 'vcf': {'chr': 'X', 'ref': 'TG', 'pos': '32486673', 'alt': 'T'}}
        assert results['NM_004006.2:c.3103del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.32468557del', 'vcf': {'chr': 'X', 'ref': 'TG', 'pos': '32468556', 'alt': 'T'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_004006.2:c.2376G>C' in results.keys()
        assert results['NM_004006.2:c.2376G>C']['hgvs_lrg_transcript_variant'] == 'LRG_199t1:c.2376G>C'
        assert results['NM_004006.2:c.2376G>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_004006.2:c.2376G>C']['alt_genomic_loci'] == []
        assert results['NM_004006.2:c.2376G>C']['gene_symbol'] == 'DMD'
        assert results['NM_004006.2:c.2376G>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003997.1(LRG_199p1):p.(Val792=)', 'slr': 'NP_003997.1:p.(V792=)'}
        assert results['NM_004006.2:c.2376G>C']['submitted_variant'] == 'LRG_199t1:c.2376G>C(;)3103del'
        assert results['NM_004006.2:c.2376G>C']['genome_context_intronic_sequence'] == ''
        assert results['NM_004006.2:c.2376G>C']['hgvs_lrg_variant'] == 'LRG_199:g.842851G>C'
        assert results['NM_004006.2:c.2376G>C']['hgvs_transcript_variant'] == 'NM_004006.2:c.2376G>C'
        assert results['NM_004006.2:c.2376G>C']['hgvs_refseqgene_variant'] == 'NG_012232.1:g.842851G>C'
        assert results['NM_004006.2:c.2376G>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.32519876C>G', 'vcf': {'chr': 'chrX', 'ref': u'C', 'pos': '32519876', 'alt': u'G'}}
        assert results['NM_004006.2:c.2376G>C']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.32501759C>G', 'vcf': {'chr': 'chrX', 'ref': u'C', 'pos': '32501759', 'alt': u'G'}}
        assert results['NM_004006.2:c.2376G>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.32519876C>G', 'vcf': {'chr': 'X', 'ref': u'C', 'pos': '32519876', 'alt': u'G'}}
        assert results['NM_004006.2:c.2376G>C']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.32501759C>G', 'vcf': {'chr': 'X', 'ref': u'C', 'pos': '32501759', 'alt': u'G'}}


    def test_variant194(self):
        variant = 'LRG_199t1:c.2376[G>C];[(G>C)]'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_004006.2:c.2376G>C' in results.keys()
        assert results['NM_004006.2:c.2376G>C']['hgvs_lrg_transcript_variant'] == 'LRG_199t1:c.2376G>C'
        assert results['NM_004006.2:c.2376G>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_004006.2:c.2376G>C']['alt_genomic_loci'] == []
        assert results['NM_004006.2:c.2376G>C']['gene_symbol'] == 'DMD'
        assert results['NM_004006.2:c.2376G>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003997.1(LRG_199p1):p.(Val792=)', 'slr': 'NP_003997.1:p.(V792=)'}
        assert results['NM_004006.2:c.2376G>C']['submitted_variant'] == 'LRG_199t1:c.2376[G>C];[(G>C)]'
        assert results['NM_004006.2:c.2376G>C']['genome_context_intronic_sequence'] == ''
        assert results['NM_004006.2:c.2376G>C']['hgvs_lrg_variant'] == 'LRG_199:g.842851G>C'
        assert results['NM_004006.2:c.2376G>C']['hgvs_transcript_variant'] == 'NM_004006.2:c.2376G>C'
        assert results['NM_004006.2:c.2376G>C']['hgvs_refseqgene_variant'] == 'NG_012232.1:g.842851G>C'
        assert results['NM_004006.2:c.2376G>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.32519876C>G', 'vcf': {'chr': 'chrX', 'ref': u'C', 'pos': '32519876', 'alt': u'G'}}
        assert results['NM_004006.2:c.2376G>C']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.32501759C>G', 'vcf': {'chr': 'chrX', 'ref': u'C', 'pos': '32501759', 'alt': u'G'}}
        assert results['NM_004006.2:c.2376G>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.32519876C>G', 'vcf': {'chr': 'X', 'ref': u'C', 'pos': '32519876', 'alt': u'G'}}
        assert results['NM_004006.2:c.2376G>C']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.32501759C>G', 'vcf': {'chr': 'X', 'ref': u'C', 'pos': '32501759', 'alt': u'G'}}


    def test_variant195(self):
        variant = 'LRG_199t1:c.[2376G>C];[?]'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_004006.2:c.2376G>C' in results.keys()
        assert results['NM_004006.2:c.2376G>C']['hgvs_lrg_transcript_variant'] == 'LRG_199t1:c.2376G>C'
        assert results['NM_004006.2:c.2376G>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_004006.2:c.2376G>C']['alt_genomic_loci'] == []
        assert results['NM_004006.2:c.2376G>C']['gene_symbol'] == 'DMD'
        assert results['NM_004006.2:c.2376G>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003997.1(LRG_199p1):p.(Val792=)', 'slr': 'NP_003997.1:p.(V792=)'}
        assert results['NM_004006.2:c.2376G>C']['submitted_variant'] == 'LRG_199t1:c.[2376G>C];[?]'
        assert results['NM_004006.2:c.2376G>C']['genome_context_intronic_sequence'] == ''
        assert results['NM_004006.2:c.2376G>C']['hgvs_lrg_variant'] == 'LRG_199:g.842851G>C'
        assert results['NM_004006.2:c.2376G>C']['hgvs_transcript_variant'] == 'NM_004006.2:c.2376G>C'
        assert results['NM_004006.2:c.2376G>C']['hgvs_refseqgene_variant'] == 'NG_012232.1:g.842851G>C'
        assert results['NM_004006.2:c.2376G>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.32519876C>G', 'vcf': {'chr': 'chrX', 'ref': u'C', 'pos': '32519876', 'alt': u'G'}}
        assert results['NM_004006.2:c.2376G>C']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.32501759C>G', 'vcf': {'chr': 'chrX', 'ref': u'C', 'pos': '32501759', 'alt': u'G'}}
        assert results['NM_004006.2:c.2376G>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.32519876C>G', 'vcf': {'chr': 'X', 'ref': u'C', 'pos': '32519876', 'alt': u'G'}}
        assert results['NM_004006.2:c.2376G>C']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.32501759C>G', 'vcf': {'chr': 'X', 'ref': u'C', 'pos': '32501759', 'alt': u'G'}}


    def test_variant196(self):
        variant = 'LRG_199t1:c.[296T>G;476T=];[476T=](;)1083A>C'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_004006.2:c.296_476delinsGCGTAGATGGAAATCATAAACTGACTCTTGGTTTGATTTGGAATATAATCCTCCACTGGCAGGTCAAAAATGTAATGAAAAATATCATGGCTGGATTGCAACAAACCAACAGTGAAAAGATTCTCCTGAGCTGGGTCCGACAATCAACTCGTAATTATCCACAGGTTAATGTAATCAACTT' in results.keys()
        assert results['NM_004006.2:c.296_476delinsGCGTAGATGGAAATCATAAACTGACTCTTGGTTTGATTTGGAATATAATCCTCCACTGGCAGGTCAAAAATGTAATGAAAAATATCATGGCTGGATTGCAACAAACCAACAGTGAAAAGATTCTCCTGAGCTGGGTCCGACAATCAACTCGTAATTATCCACAGGTTAATGTAATCAACTT']['hgvs_lrg_transcript_variant'] == 'LRG_199t1:c.296_358-3delinsGCGTAGATGGAAATCATAAACTGACTCTTGGTTTGATTTGGAATATAATCCTCCACTGGC'
        assert results['NM_004006.2:c.296_476delinsGCGTAGATGGAAATCATAAACTGACTCTTGGTTTGATTTGGAATATAATCCTCCACTGGCAGGTCAAAAATGTAATGAAAAATATCATGGCTGGATTGCAACAAACCAACAGTGAAAAGATTCTCCTGAGCTGGGTCCGACAATCAACTCGTAATTATCCACAGGTTAATGTAATCAACTT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_004006.2:c.296_476delinsGCGTAGATGGAAATCATAAACTGACTCTTGGTTTGATTTGGAATATAATCCTCCACTGGCAGGTCAAAAATGTAATGAAAAATATCATGGCTGGATTGCAACAAACCAACAGTGAAAAGATTCTCCTGAGCTGGGTCCGACAATCAACTCGTAATTATCCACAGGTTAATGTAATCAACTT']['alt_genomic_loci'] == []
        assert results['NM_004006.2:c.296_476delinsGCGTAGATGGAAATCATAAACTGACTCTTGGTTTGATTTGGAATATAATCCTCCACTGGCAGGTCAAAAATGTAATGAAAAATATCATGGCTGGATTGCAACAAACCAACAGTGAAAAGATTCTCCTGAGCTGGGTCCGACAATCAACTCGTAATTATCCACAGGTTAATGTAATCAACTT']['gene_symbol'] == 'DMD'
        assert results['NM_004006.2:c.296_476delinsGCGTAGATGGAAATCATAAACTGACTCTTGGTTTGATTTGGAATATAATCCTCCACTGGCAGGTCAAAAATGTAATGAAAAATATCATGGCTGGATTGCAACAAACCAACAGTGAAAAGATTCTCCTGAGCTGGGTCCGACAATCAACTCGTAATTATCCACAGGTTAATGTAATCAACTT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003997.1(LRG_199p1):p.(Ile99Ser)', 'slr': 'NP_003997.1:p.(I99S)'}
        assert results['NM_004006.2:c.296_476delinsGCGTAGATGGAAATCATAAACTGACTCTTGGTTTGATTTGGAATATAATCCTCCACTGGCAGGTCAAAAATGTAATGAAAAATATCATGGCTGGATTGCAACAAACCAACAGTGAAAAGATTCTCCTGAGCTGGGTCCGACAATCAACTCGTAATTATCCACAGGTTAATGTAATCAACTT']['submitted_variant'] == 'LRG_199t1:c.[296T>G;476T=];[476T=](;)1083A>C'
        assert results['NM_004006.2:c.296_476delinsGCGTAGATGGAAATCATAAACTGACTCTTGGTTTGATTTGGAATATAATCCTCCACTGGCAGGTCAAAAATGTAATGAAAAATATCATGGCTGGATTGCAACAAACCAACAGTGAAAAGATTCTCCTGAGCTGGGTCCGACAATCAACTCGTAATTATCCACAGGTTAATGTAATCAACTT']['genome_context_intronic_sequence'] == ''
        assert results['NM_004006.2:c.296_476delinsGCGTAGATGGAAATCATAAACTGACTCTTGGTTTGATTTGGAATATAATCCTCCACTGGCAGGTCAAAAATGTAATGAAAAATATCATGGCTGGATTGCAACAAACCAACAGTGAAAAGATTCTCCTGAGCTGGGTCCGACAATCAACTCGTAATTATCCACAGGTTAATGTAATCAACTT']['hgvs_lrg_variant'] == 'LRG_199:g.521254_527967delinsGCGTAGATGGAAATCATAAACTGACTCTTGGTTTGATTTGGAATATAATCCTCCACTGGC'
        assert results['NM_004006.2:c.296_476delinsGCGTAGATGGAAATCATAAACTGACTCTTGGTTTGATTTGGAATATAATCCTCCACTGGCAGGTCAAAAATGTAATGAAAAATATCATGGCTGGATTGCAACAAACCAACAGTGAAAAGATTCTCCTGAGCTGGGTCCGACAATCAACTCGTAATTATCCACAGGTTAATGTAATCAACTT']['hgvs_transcript_variant'] == 'NM_004006.2:c.296_476delinsGCGTAGATGGAAATCATAAACTGACTCTTGGTTTGATTTGGAATATAATCCTCCACTGGCAGGTCAAAAATGTAATGAAAAATATCATGGCTGGATTGCAACAAACCAACAGTGAAAAGATTCTCCTGAGCTGGGTCCGACAATCAACTCGTAATTATCCACAGGTTAATGTAATCAACTT'
        assert results['NM_004006.2:c.296_476delinsGCGTAGATGGAAATCATAAACTGACTCTTGGTTTGATTTGGAATATAATCCTCCACTGGCAGGTCAAAAATGTAATGAAAAATATCATGGCTGGATTGCAACAAACCAACAGTGAAAAGATTCTCCTGAGCTGGGTCCGACAATCAACTCGTAATTATCCACAGGTTAATGTAATCAACTT']['hgvs_refseqgene_variant'] == 'NG_012232.1:g.521254_527967delinsGCGTAGATGGAAATCATAAACTGACTCTTGGTTTGATTTGGAATATAATCCTCCACTGGC'
        assert results['NM_004006.2:c.296_476delinsGCGTAGATGGAAATCATAAACTGACTCTTGGTTTGATTTGGAATATAATCCTCCACTGGCAGGTCAAAAATGTAATGAAAAATATCATGGCTGGATTGCAACAAACCAACAGTGAAAAGATTCTCCTGAGCTGGGTCCGACAATCAACTCGTAATTATCCACAGGTTAATGTAATCAACTT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.32834639_32841473delinsAAGTTGATTACATTAACCTGTGGATAATTACGAGTTGATTGTCGGACCCAGCTCAGGAGAATCTTTTCACTGTTGGTTTGTTGCAATCCAGCCATGATATTTTTCATTACATTTTTGACCTGCCAGTGGAGGATTATATTCCAAATCAAACCAAGAGTCAGTTTATGATTTCCATCTACGC', 'vcf': {'chr': 'chrX', 'ref': 'ACATGTGGAAATAAATTTTCATAAGAAAATGCATTCCTTGAGCAAGAACCATGCAAACTTAAATATGAATGTCCTTGATCTTCAGTGATAAATAGAAATTTTAGGGCCAATTAGTAATGAGACATAATAGATTCTACCAGAAGTTAAGTCTATTCTCAAAGGCTAGGAGTCTATTCTGATTCATTGGTATCTATGCCATACCATTTACTGAATGTGTTCACTATTACTCCTGGATTCTGCCATATTAGACCCAACATTTACTGTTTATAGAATTGCTCACCTTCTAGCTTTCAATGAACAAGCTAGAGAAGACAGTGTATAAAACATTAGACTTATTGTTAACCAAGGACTTAAGATATAAAAGCGCTTTTTGGATAGGAGGAATACTATAGATAATTGTTTTTTAAAGTATATGTTCAAATTTTGCCTTTGCTATTAATTGTGGTAATTGGACAAGTTGCTTAACCTCTTCTAGGCCTCATTTCTGTGTATCTTACATGCAAGGTTAGTGTAAGATCAAAGTACAAAATATGGTCAGATAAAACCCCTGATTTCATACCATCATTTAACGCAAGCTATATTTATTGGCTGAACTGAAAAGAAGACAGAAATGTTTGCCTACCTAAAATAAAGAAAAAAAACAACAAAGGAAGGAAATATTTATCCCTAAGAGGTGTACATATGTAGCCTTGGTTCTATTTTCATTACTTGGTTTCCATGTATTTGATTGTCTCATAGGCTTCGTGCATGTGCAGATGTTATTCAACAGATTATCAAGTTTAATTCTTCTGAGTCATCTGATGCAGAAATCCATTCTAAGCATTTTTTCTAAGTTTCATGAATGCAGTGACGCTTCATAGAAAAAACTGTCTACATCTTTTATTAAAGAATTATTATTTATTAAAGACTGCTTTTTAAATATAAGCCCCTTATGAGATGCTGGAAGAAGACATCTATGTAAAAGTGATCAGTCTGTACCACTAAGCACTTTCCTTCTGGGAAGGTTAAGGTTCATATACTATCAACATTTGATCAGATGATGAATATTTATTGTGCATTCTAAATATGACAATTTTATTTTTTCATCCATATAGAAGTTAACAAATGCCTTTGGGGTAAATGCATATTTTAACATCTCCAATGTTAGTCATCTTTTTATTAAAAAGTAAATGCAAATTTTAGGATATACTACTAAATCTCATTTTAACATACTTTTAAAAGATTTGAGATGATGCTTTGTAAACACAGTAACGATTCGGAGATATGAAATCTTTGCATGTATCTGTCACAATCCCTTGATAACTAAGCCTATTATTTAGGAGTGGATTATCCAAGTTTTCCTAATTCCTTCTTGACCTACTTATTTAGTCAGTTACTCAGTGAATACAGGTCAAATAATTACGTAATTTGTCTCATATTCTAAAATTATAAATTGCTCAACCTTTATGTACAGTGACTGCAGGAGGTGTATTTTGTGTCCTGTGCTAAATTACGGTTGGAATCAGCAGATAGTGGCTTAATGGTGGTAATGGATTTGCAATAATCAAGCAATCCTCCAATGATGAATGTGATTCATTTGTGAAATGCTTGGCCAGTCTGCCCTCTATGTGCATAACGTTAAAGGACAACAAAGTATTCAATCCAAAACTCCTTTATAGCATCATTATGTCACTAATTCAGCTTATCAGCCTCAGCCTCGCTGTTATTGCTTCTACTTAGGTGCAAATGTTGCTCAGACAAAAAAAAATAATAAAGGGCAATGTATGATAATCATGTCTCCTTTTCCTACATAAAAGGCAGTTATCAAAAAATCATTGAGCTGGTCCCCAAAGGTGAGCCTGTAAACAGAAAACTGTCCCGTCCTCAGGTCTCACTGTCTCTAAAGGAGGCTCAATTTAAACAGATGTATTTGTGTGTGTGCGCGCATACAACTATAAATCAATTTTTAAAAGGAAAAAAGTTGAAATGTCTTTGAGCAGTGATTTTTAAAAATGACTAGGAGTTAACTGGGCCAAGAAGGCATGATGGACATGGAGGTGCACCACCAAGATTGACCTTCAAGGAAGGGCTTGTTGCTCCAGCTGTGGTCAGGCAGCTTCTAGTTGTTAGTTCTCTCAGGTACCACCTGAGTTGCAGAGTCCTGCCAGCCAATGTCACACCCTCTCTAGGGCAACCCACTACTATTGACTGATGAAGGGGCAGAGTATAAACACTCAGCCATTTTAACCCCATTTGAGACAACTCTGAGGGGTCATCCTAGCTGCAAGTTGTCTATGGGGTTTGAATAAGCGTGTCATAGGGTCTACGTCACAGCTCTAGTTCTTCCTCTTTCCACTCCTGTCACCTCTGCCCTTCTACAGGTGTTTTTTTTTTTTTTTTTTTTTCCAGGGCATGGCTTAATAAACATGCTGCCTGCCAAACTCAACTCCGAAGTCTACTCCCCAGAACCCAGACTGTGATTAAAAGAAGTTGCAGCAGAGGAATAGAAATGAAGAGGGTAGGAATAATCTTATATGGTAAGGATAGTTTCTGTGAAGCTCTGCAGAAGAGCAGAGCAAGAACTGTCTTGCACAGAAGTGACTTGGTAGGCAGGAATTTACAGGGATGGCTTTCAGCAGTGCGAACGTCATCATTCTCCTGTAATTGCAAATTGTATCAATAATATGAACATTCTAAAGTGAATTAGCGTTGCCATTAAGTACATCCCTTGAATTGTCCAGTAAAAGTATGGACCTGACTTCTGATAGCTCACCACTGTTAGCAAGATTTGTTATTACAACCAGCTGGTTGCCAGAGCCAAGTTTAGCACTCCGAAAAAAAAAGAATTGCACTTAAGAAGGAAGAAAAAGATGCCAAACACATGTTATATGCTTTGCAATTTTGTTCAAACCTGCCTATGTTACAAAATATGCGTTACTGTTTTAGGCTAGCAATTTTATTTGAGATGGACAGTAGAAAAACAGTAATAGGAAAACAGAGTTTGCATTCCTGAGTTTTCACAAGTTCAATGAGAAGTCCTCCATGGCACTATTAATAACATAGTATGTTGCATTAGAAAAAGGGGCCAACATGGTTCTATCTGTAAATGCCGCTCGTTATCTGATAGATCTAGAATGTACATTAAGTTGACCAAAATGTCCTTTAATCTTTGCTGAAATTCAAGGGCTTCTACCATTTAATAGGAAGAAAATATGTAATGTTGGTGTAAAAAAAAAAAAAAAAAAGAAAAACACATTTTTAGATTCACTGGATTATGATGAACACTGAGCAAGATAATTTTTCAAAGTGTGCTTAATAGTTTTAGAACAGGGCAGTGTGATATAAGAGAGATGAGATAAGTAAGCAGAGGCTGGGAAGAGTCTACAGTCTTGCAGTCTCAAGGATATGGTGATTCAAGGGCATAAATGAAAAACAACAGGCTATAAGAGAGGTCTCAGACAAAGAAAGATTTAAAAAGCATTCCTTCTGATATTCCTCATTGCAAAAGCGCCTCCACATGCTAATAATGAAAAATGGGCTGGGTGCGGTGGCTCACGCCTGTAATCCAAACACTTTGGGAGGCCGAGACGGGCGGATCACAAGGTCAGGAGATCGAGACCATCCTGGCTAACACGGTGAAACCCCATCTCTACTGAACAAAATACAAAAAATTAGCCGGGCGTGGTGGCGGGCGCCTGTAGTCCCAGCTACTTGGGAGGCTGAGGCAAGAGAATGGCATGAACCCAGGAGGCGGAGCTTGCAGTGAGCAGAGATCGCGCCACTGCACTCCAGCCTGGGTGACAGAGCGAGACTCTGTCTCAAAAATAAAAAATAAAAAAATAATAATAAATGAACAACTTGCAAGCAAGTGAAGAAATTATCAGACATAATAGTCTTGAAAGCTTATCTTGGACTAGTTCAGATGGAGTTGGAAAGGGCTTCAGGATATTTTTAAAAATGGTAAGATCATATAAGAAGTAGTATAAACAATAAATACAGTTCAGAGTGCCTTTTGGAGCTCTGTGCATGCTTGAGTTCTTCTGCTGGCTTTGTCAAAACTTACATCTGTCTTACGCAAAGCGGGATCTGCAAACTGGATTCACTCTAGGAATTATTTGTTACTATGCCATAAGATAAAGTCAACAATTCAGAGTGCATCAAGAAACTTATGTAGCGATTTTACAGAGTACCATTTATGTCTAATGTATTTAATAATAAAGGAGGAAGCATGGGTGTTGGGCATCTTATTTTTGTAACGCTTTGGTTTTATTCTATCTTACATACAATGGATTAGGGGAAAAGTGTTCCTTCCTCAAGATATTTTGAGAAGTACTGAGCAACATATGAAAAGCAGTTTGGGGAGAGATGCGGTATGTTGCTTGCTGGTTCTATTTGACACCATCACTATATGGAACGGGCTGAAAATCGGCCAACTTGGGCTCACTTAAGGCTCCTATGAGCTATTCTTTGTTGCCAGCACATATTAATTCCCGCTCTTTCTCTTCCCCTCTCCCCGCTTACTGTTGTGAAGTAGCATTAAGCCTGTTCAGAGAATTTGGAATAAAAATATATGGGGGCCAATTAGGAGAGCAACATGGCTGCTGAATTTAATAGGTACTCTTTGTGTCTACTCACTATTTGACTCTGAGCAAGAGCAATCTACAATTCAAGTTAAAATTCAAGCCGGGCGCAGTGGCTCATGCCTGCAATCCCAGCACTTTGGGAGGCCGAGGCAGGAGGATCACGAGGTCAGGAGATCGAGACCATCCTGGCTAACACGGTGAAACCCCGTCTCTACTAAAAATACAAAAAAAAATTAGCCGGGCGTGGTGGGGGGGCGCCTGTAGTCCCAGCTACTCGGGAGGCTGAGGCAGGAGAATGGCGTGAATCCGGGAGGCGGAGCTTGCAGTGAGCCGAGATCGCGCCACTGCACTCCAGCCTGGGCGACAGAGCAAGACTCTGTCTCAAAATAAAAATAAAAAAAAATAAAAATAAAAAATAAAAAAATTCAAGTTTACCGAATGAAGCCATGATATCTGCTACAATAGGGCTGGATCTGAACTATTACCCTCATTGCAAGAAGCCAGACACAAAAGAACATGTATTATATTGATATAAAATATCCAAAAAGGCAAATACATATGGACAGAAATAGATTGACAATTGCCTAGGGCTGGGGTGGGAGTGAAAAATGACTGTAAATGGAGAAGAAACTTCTTTTTAAGGTGGTGGAAAAGTGCCAAAATTAGATCGCTATGATGGCTATATAACTCTGTAAATATCCTGCAAATTTTAATGAACGTGGTGGGGGGGAGTTATTCATCCCTCAAAATAGTTTGAGAGGCTACCAAGAACCAGAACTAAACCAGATTGATGCAAATTGAATTTTACACTTAGAATGAGTACATTGATGGTCTAACAATTACATTTCAAAAACGCTGATTTATAAAAAAGTAATCCCGACCAGTAAAAACAGCAGTCATTTCTCAGCCTCCAAATTTTTTAAATGAGTGGAAAACTTTAAGCAATCAATACAAGTCATATGATACACCTCAAATAAAAGTGGTATACTACACTGAGTGCCTTGCAGATATTCTCCCACTTAATCGTGATAAGAAATACATACACATTTATTTTTATTTTTACTCTTTTGATAGTGAAGGGACCATGGCCCAGAACAGGGGTGGCAATATACCAAATGCCACTAAGATGGTAACCGTGGAAGTTTAAATTCAAGTTCATGTTGGTATATATATATACACATACATATTTATACATATGTGTATAACGAGTTACTATTACGTATATATATACAATGAGTGGAAAATATTAGTGGATATAAGTTCATGTATGATTATGTATACATATTATGATTATACTATTATGTATATGTATACAATGAGTGGATAATATTAGTGGATATAAGTTCATGTTGGTCTTATATATACATATATAATGTGTGGATATGAGTGTGTATATATATACATATATAATGTGTGGATATGAGTGTGTGTATATACACACATGCATATTGTATATATGTGTGTATATATAGACACATTATATATATACGCATATATACTACACACACACACACACGGGTGTGTGTCTGTATCTTTTCCACAAATCCTTCAACCCATTTTGCAGAGGTCAAATAGACAGTCGGAAGACCCTATGCTCAGGTGACTTAAAAATAATTTCCAAATCACATTATGGAGTTTGTATGTATTACACACATTTATTGATAGAGATACCCATATTCTACTAATCTTTTATTGGCAATAATTTATGTTAAGAATACCCAAGACTGAGAAAGCCTCATTCCTTTGGTAGTGATTAAAATAAAACATACTAAATTAACTTATAGACAAGTTATAGAACATACATTTGTGAAAAAAATTACTCACCTATGATTGGGACTTTGTATTTTTACCTTATACTTACTCAATGAAATAAAATTTTGAAAAATATTCCTGTAAATGTACCAGAACCTATTTTATACCGTGATGATCCTTAACATTTCAGACGACATGGTAGTGTCAATTTAAAAAGCAGCACTATGGAGCAGGGTTTGTTATTGTTAGAAATACACATTTGTTTCACACGTCAAGGGTAAAAATTAAAAAACAAGATTAATGTTACCCAAAAGGAAACCATTCATCAGGATTCTTACCTGCCAGTGGAGGATTATATTCCAAATCAAACCAAGAGTCAGTTTATGATTTCCATCTACGA', 'pos': '32834760', 'alt': u'GCCAGTGGAGGATTATATTCCAAATCAAACCAAGAGTCAGTTTATGATTTCCATCTACGC'}}
        assert results['NM_004006.2:c.296_476delinsGCGTAGATGGAAATCATAAACTGACTCTTGGTTTGATTTGGAATATAATCCTCCACTGGCAGGTCAAAAATGTAATGAAAAATATCATGGCTGGATTGCAACAAACCAACAGTGAAAAGATTCTCCTGAGCTGGGTCCGACAATCAACTCGTAATTATCCACAGGTTAATGTAATCAACTT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.32816522_32823356delinsAAGTTGATTACATTAACCTGTGGATAATTACGAGTTGATTGTCGGACCCAGCTCAGGAGAATCTTTTCACTGTTGGTTTGTTGCAATCCAGCCATGATATTTTTCATTACATTTTTGACCTGCCAGTGGAGGATTATATTCCAAATCAAACCAAGAGTCAGTTTATGATTTCCATCTACGC', 'vcf': {'chr': 'chrX', 'ref': 'ACATGTGGAAATAAATTTTCATAAGAAAATGCATTCCTTGAGCAAGAACCATGCAAACTTAAATATGAATGTCCTTGATCTTCAGTGATAAATAGAAATTTTAGGGCCAATTAGTAATGAGACATAATAGATTCTACCAGAAGTTAAGTCTATTCTCAAAGGCTAGGAGTCTATTCTGATTCATTGGTATCTATGCCATACCATTTACTGAATGTGTTCACTATTACTCCTGGATTCTGCCATATTAGACCCAACATTTACTGTTTATAGAATTGCTCACCTTCTAGCTTTCAATGAACAAGCTAGAGAAGACAGTGTATAAAACATTAGACTTATTGTTAACCAAGGACTTAAGATATAAAAGCGCTTTTTGGATAGGAGGAATACTATAGATAATTGTTTTTTAAAGTATATGTTCAAATTTTGCCTTTGCTATTAATTGTGGTAATTGGACAAGTTGCTTAACCTCTTCTAGGCCTCATTTCTGTGTATCTTACATGCAAGGTTAGTGTAAGATCAAAGTACAAAATATGGTCAGATAAAACCCCTGATTTCATACCATCATTTAACGCAAGCTATATTTATTGGCTGAACTGAAAAGAAGACAGAAATGTTTGCCTACCTAAAATAAAGAAAAAAAACAACAAAGGAAGGAAATATTTATCCCTAAGAGGTGTACATATGTAGCCTTGGTTCTATTTTCATTACTTGGTTTCCATGTATTTGATTGTCTCATAGGCTTCGTGCATGTGCAGATGTTATTCAACAGATTATCAAGTTTAATTCTTCTGAGTCATCTGATGCAGAAATCCATTCTAAGCATTTTTTCTAAGTTTCATGAATGCAGTGACGCTTCATAGAAAAAACTGTCTACATCTTTTATTAAAGAATTATTATTTATTAAAGACTGCTTTTTAAATATAAGCCCCTTATGAGATGCTGGAAGAAGACATCTATGTAAAAGTGATCAGTCTGTACCACTAAGCACTTTCCTTCTGGGAAGGTTAAGGTTCATATACTATCAACATTTGATCAGATGATGAATATTTATTGTGCATTCTAAATATGACAATTTTATTTTTTCATCCATATAGAAGTTAACAAATGCCTTTGGGGTAAATGCATATTTTAACATCTCCAATGTTAGTCATCTTTTTATTAAAAAGTAAATGCAAATTTTAGGATATACTACTAAATCTCATTTTAACATACTTTTAAAAGATTTGAGATGATGCTTTGTAAACACAGTAACGATTCGGAGATATGAAATCTTTGCATGTATCTGTCACAATCCCTTGATAACTAAGCCTATTATTTAGGAGTGGATTATCCAAGTTTTCCTAATTCCTTCTTGACCTACTTATTTAGTCAGTTACTCAGTGAATACAGGTCAAATAATTACGTAATTTGTCTCATATTCTAAAATTATAAATTGCTCAACCTTTATGTACAGTGACTGCAGGAGGTGTATTTTGTGTCCTGTGCTAAATTACGGTTGGAATCAGCAGATAGTGGCTTAATGGTGGTAATGGATTTGCAATAATCAAGCAATCCTCCAATGATGAATGTGATTCATTTGTGAAATGCTTGGCCAGTCTGCCCTCTATGTGCATAACGTTAAAGGACAACAAAGTATTCAATCCAAAACTCCTTTATAGCATCATTATGTCACTAATTCAGCTTATCAGCCTCAGCCTCGCTGTTATTGCTTCTACTTAGGTGCAAATGTTGCTCAGACAAAAAAAAATAATAAAGGGCAATGTATGATAATCATGTCTCCTTTTCCTACATAAAAGGCAGTTATCAAAAAATCATTGAGCTGGTCCCCAAAGGTGAGCCTGTAAACAGAAAACTGTCCCGTCCTCAGGTCTCACTGTCTCTAAAGGAGGCTCAATTTAAACAGATGTATTTGTGTGTGTGCGCGCATACAACTATAAATCAATTTTTAAAAGGAAAAAAGTTGAAATGTCTTTGAGCAGTGATTTTTAAAAATGACTAGGAGTTAACTGGGCCAAGAAGGCATGATGGACATGGAGGTGCACCACCAAGATTGACCTTCAAGGAAGGGCTTGTTGCTCCAGCTGTGGTCAGGCAGCTTCTAGTTGTTAGTTCTCTCAGGTACCACCTGAGTTGCAGAGTCCTGCCAGCCAATGTCACACCCTCTCTAGGGCAACCCACTACTATTGACTGATGAAGGGGCAGAGTATAAACACTCAGCCATTTTAACCCCATTTGAGACAACTCTGAGGGGTCATCCTAGCTGCAAGTTGTCTATGGGGTTTGAATAAGCGTGTCATAGGGTCTACGTCACAGCTCTAGTTCTTCCTCTTTCCACTCCTGTCACCTCTGCCCTTCTACAGGTGTTTTTTTTTTTTTTTTTTTTTCCAGGGCATGGCTTAATAAACATGCTGCCTGCCAAACTCAACTCCGAAGTCTACTCCCCAGAACCCAGACTGTGATTAAAAGAAGTTGCAGCAGAGGAATAGAAATGAAGAGGGTAGGAATAATCTTATATGGTAAGGATAGTTTCTGTGAAGCTCTGCAGAAGAGCAGAGCAAGAACTGTCTTGCACAGAAGTGACTTGGTAGGCAGGAATTTACAGGGATGGCTTTCAGCAGTGCGAACGTCATCATTCTCCTGTAATTGCAAATTGTATCAATAATATGAACATTCTAAAGTGAATTAGCGTTGCCATTAAGTACATCCCTTGAATTGTCCAGTAAAAGTATGGACCTGACTTCTGATAGCTCACCACTGTTAGCAAGATTTGTTATTACAACCAGCTGGTTGCCAGAGCCAAGTTTAGCACTCCGAAAAAAAAAGAATTGCACTTAAGAAGGAAGAAAAAGATGCCAAACACATGTTATATGCTTTGCAATTTTGTTCAAACCTGCCTATGTTACAAAATATGCGTTACTGTTTTAGGCTAGCAATTTTATTTGAGATGGACAGTAGAAAAACAGTAATAGGAAAACAGAGTTTGCATTCCTGAGTTTTCACAAGTTCAATGAGAAGTCCTCCATGGCACTATTAATAACATAGTATGTTGCATTAGAAAAAGGGGCCAACATGGTTCTATCTGTAAATGCCGCTCGTTATCTGATAGATCTAGAATGTACATTAAGTTGACCAAAATGTCCTTTAATCTTTGCTGAAATTCAAGGGCTTCTACCATTTAATAGGAAGAAAATATGTAATGTTGGTGTAAAAAAAAAAAAAAAAAAGAAAAACACATTTTTAGATTCACTGGATTATGATGAACACTGAGCAAGATAATTTTTCAAAGTGTGCTTAATAGTTTTAGAACAGGGCAGTGTGATATAAGAGAGATGAGATAAGTAAGCAGAGGCTGGGAAGAGTCTACAGTCTTGCAGTCTCAAGGATATGGTGATTCAAGGGCATAAATGAAAAACAACAGGCTATAAGAGAGGTCTCAGACAAAGAAAGATTTAAAAAGCATTCCTTCTGATATTCCTCATTGCAAAAGCGCCTCCACATGCTAATAATGAAAAATGGGCTGGGTGCGGTGGCTCACGCCTGTAATCCAAACACTTTGGGAGGCCGAGACGGGCGGATCACAAGGTCAGGAGATCGAGACCATCCTGGCTAACACGGTGAAACCCCATCTCTACTGAACAAAATACAAAAAATTAGCCGGGCGTGGTGGCGGGCGCCTGTAGTCCCAGCTACTTGGGAGGCTGAGGCAAGAGAATGGCATGAACCCAGGAGGCGGAGCTTGCAGTGAGCAGAGATCGCGCCACTGCACTCCAGCCTGGGTGACAGAGCGAGACTCTGTCTCAAAAATAAAAAATAAAAAAATAATAATAAATGAACAACTTGCAAGCAAGTGAAGAAATTATCAGACATAATAGTCTTGAAAGCTTATCTTGGACTAGTTCAGATGGAGTTGGAAAGGGCTTCAGGATATTTTTAAAAATGGTAAGATCATATAAGAAGTAGTATAAACAATAAATACAGTTCAGAGTGCCTTTTGGAGCTCTGTGCATGCTTGAGTTCTTCTGCTGGCTTTGTCAAAACTTACATCTGTCTTACGCAAAGCGGGATCTGCAAACTGGATTCACTCTAGGAATTATTTGTTACTATGCCATAAGATAAAGTCAACAATTCAGAGTGCATCAAGAAACTTATGTAGCGATTTTACAGAGTACCATTTATGTCTAATGTATTTAATAATAAAGGAGGAAGCATGGGTGTTGGGCATCTTATTTTTGTAACGCTTTGGTTTTATTCTATCTTACATACAATGGATTAGGGGAAAAGTGTTCCTTCCTCAAGATATTTTGAGAAGTACTGAGCAACATATGAAAAGCAGTTTGGGGAGAGATGCGGTATGTTGCTTGCTGGTTCTATTTGACACCATCACTATATGGAACGGGCTGAAAATCGGCCAACTTGGGCTCACTTAAGGCTCCTATGAGCTATTCTTTGTTGCCAGCACATATTAATTCCCGCTCTTTCTCTTCCCCTCTCCCCGCTTACTGTTGTGAAGTAGCATTAAGCCTGTTCAGAGAATTTGGAATAAAAATATATGGGGGCCAATTAGGAGAGCAACATGGCTGCTGAATTTAATAGGTACTCTTTGTGTCTACTCACTATTTGACTCTGAGCAAGAGCAATCTACAATTCAAGTTAAAATTCAAGCCGGGCGCAGTGGCTCATGCCTGCAATCCCAGCACTTTGGGAGGCCGAGGCAGGAGGATCACGAGGTCAGGAGATCGAGACCATCCTGGCTAACACGGTGAAACCCCGTCTCTACTAAAAATACAAAAAAAAATTAGCCGGGCGTGGTGGGGGGGCGCCTGTAGTCCCAGCTACTCGGGAGGCTGAGGCAGGAGAATGGCGTGAATCCGGGAGGCGGAGCTTGCAGTGAGCCGAGATCGCGCCACTGCACTCCAGCCTGGGCGACAGAGCAAGACTCTGTCTCAAAATAAAAATAAAAAAAAATAAAAATAAAAAATAAAAAAATTCAAGTTTACCGAATGAAGCCATGATATCTGCTACAATAGGGCTGGATCTGAACTATTACCCTCATTGCAAGAAGCCAGACACAAAAGAACATGTATTATATTGATATAAAATATCCAAAAAGGCAAATACATATGGACAGAAATAGATTGACAATTGCCTAGGGCTGGGGTGGGAGTGAAAAATGACTGTAAATGGAGAAGAAACTTCTTTTTAAGGTGGTGGAAAAGTGCCAAAATTAGATCGCTATGATGGCTATATAACTCTGTAAATATCCTGCAAATTTTAATGAACGTGGTGGGGGGGAGTTATTCATCCCTCAAAATAGTTTGAGAGGCTACCAAGAACCAGAACTAAACCAGATTGATGCAAATTGAATTTTACACTTAGAATGAGTACATTGATGGTCTAACAATTACATTTCAAAAACGCTGATTTATAAAAAAGTAATCCCGACCAGTAAAAACAGCAGTCATTTCTCAGCCTCCAAATTTTTTAAATGAGTGGAAAACTTTAAGCAATCAATACAAGTCATATGATACACCTCAAATAAAAGTGGTATACTACACTGAGTGCCTTGCAGATATTCTCCCACTTAATCGTGATAAGAAATACATACACATTTATTTTTATTTTTACTCTTTTGATAGTGAAGGGACCATGGCCCAGAACAGGGGTGGCAATATACCAAATGCCACTAAGATGGTAACCGTGGAAGTTTAAATTCAAGTTCATGTTGGTATATATATATACACATACATATTTATACATATGTGTATAACGAGTTACTATTACGTATATATATACAATGAGTGGAAAATATTAGTGGATATAAGTTCATGTATGATTATGTATACATATTATGATTATACTATTATGTATATGTATACAATGAGTGGATAATATTAGTGGATATAAGTTCATGTTGGTCTTATATATACATATATAATGTGTGGATATGAGTGTGTATATATATACATATATAATGTGTGGATATGAGTGTGTGTATATACACACATGCATATTGTATATATGTGTGTATATATAGACACATTATATATATACGCATATATACTACACACACACACACACGGGTGTGTGTCTGTATCTTTTCCACAAATCCTTCAACCCATTTTGCAGAGGTCAAATAGACAGTCGGAAGACCCTATGCTCAGGTGACTTAAAAATAATTTCCAAATCACATTATGGAGTTTGTATGTATTACACACATTTATTGATAGAGATACCCATATTCTACTAATCTTTTATTGGCAATAATTTATGTTAAGAATACCCAAGACTGAGAAAGCCTCATTCCTTTGGTAGTGATTAAAATAAAACATACTAAATTAACTTATAGACAAGTTATAGAACATACATTTGTGAAAAAAATTACTCACCTATGATTGGGACTTTGTATTTTTACCTTATACTTACTCAATGAAATAAAATTTTGAAAAATATTCCTGTAAATGTACCAGAACCTATTTTATACCGTGATGATCCTTAACATTTCAGACGACATGGTAGTGTCAATTTAAAAAGCAGCACTATGGAGCAGGGTTTGTTATTGTTAGAAATACACATTTGTTTCACACGTCAAGGGTAAAAATTAAAAAACAAGATTAATGTTACCCAAAAGGAAACCATTCATCAGGATTCTTACCTGCCAGTGGAGGATTATATTCCAAATCAAACCAAGAGTCAGTTTATGATTTCCATCTACGA', 'pos': '32816643', 'alt': u'GCCAGTGGAGGATTATATTCCAAATCAAACCAAGAGTCAGTTTATGATTTCCATCTACGC'}}
        assert results['NM_004006.2:c.296_476delinsGCGTAGATGGAAATCATAAACTGACTCTTGGTTTGATTTGGAATATAATCCTCCACTGGCAGGTCAAAAATGTAATGAAAAATATCATGGCTGGATTGCAACAAACCAACAGTGAAAAGATTCTCCTGAGCTGGGTCCGACAATCAACTCGTAATTATCCACAGGTTAATGTAATCAACTT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.32834639_32841473delinsAAGTTGATTACATTAACCTGTGGATAATTACGAGTTGATTGTCGGACCCAGCTCAGGAGAATCTTTTCACTGTTGGTTTGTTGCAATCCAGCCATGATATTTTTCATTACATTTTTGACCTGCCAGTGGAGGATTATATTCCAAATCAAACCAAGAGTCAGTTTATGATTTCCATCTACGC', 'vcf': {'chr': 'X', 'ref': 'ACATGTGGAAATAAATTTTCATAAGAAAATGCATTCCTTGAGCAAGAACCATGCAAACTTAAATATGAATGTCCTTGATCTTCAGTGATAAATAGAAATTTTAGGGCCAATTAGTAATGAGACATAATAGATTCTACCAGAAGTTAAGTCTATTCTCAAAGGCTAGGAGTCTATTCTGATTCATTGGTATCTATGCCATACCATTTACTGAATGTGTTCACTATTACTCCTGGATTCTGCCATATTAGACCCAACATTTACTGTTTATAGAATTGCTCACCTTCTAGCTTTCAATGAACAAGCTAGAGAAGACAGTGTATAAAACATTAGACTTATTGTTAACCAAGGACTTAAGATATAAAAGCGCTTTTTGGATAGGAGGAATACTATAGATAATTGTTTTTTAAAGTATATGTTCAAATTTTGCCTTTGCTATTAATTGTGGTAATTGGACAAGTTGCTTAACCTCTTCTAGGCCTCATTTCTGTGTATCTTACATGCAAGGTTAGTGTAAGATCAAAGTACAAAATATGGTCAGATAAAACCCCTGATTTCATACCATCATTTAACGCAAGCTATATTTATTGGCTGAACTGAAAAGAAGACAGAAATGTTTGCCTACCTAAAATAAAGAAAAAAAACAACAAAGGAAGGAAATATTTATCCCTAAGAGGTGTACATATGTAGCCTTGGTTCTATTTTCATTACTTGGTTTCCATGTATTTGATTGTCTCATAGGCTTCGTGCATGTGCAGATGTTATTCAACAGATTATCAAGTTTAATTCTTCTGAGTCATCTGATGCAGAAATCCATTCTAAGCATTTTTTCTAAGTTTCATGAATGCAGTGACGCTTCATAGAAAAAACTGTCTACATCTTTTATTAAAGAATTATTATTTATTAAAGACTGCTTTTTAAATATAAGCCCCTTATGAGATGCTGGAAGAAGACATCTATGTAAAAGTGATCAGTCTGTACCACTAAGCACTTTCCTTCTGGGAAGGTTAAGGTTCATATACTATCAACATTTGATCAGATGATGAATATTTATTGTGCATTCTAAATATGACAATTTTATTTTTTCATCCATATAGAAGTTAACAAATGCCTTTGGGGTAAATGCATATTTTAACATCTCCAATGTTAGTCATCTTTTTATTAAAAAGTAAATGCAAATTTTAGGATATACTACTAAATCTCATTTTAACATACTTTTAAAAGATTTGAGATGATGCTTTGTAAACACAGTAACGATTCGGAGATATGAAATCTTTGCATGTATCTGTCACAATCCCTTGATAACTAAGCCTATTATTTAGGAGTGGATTATCCAAGTTTTCCTAATTCCTTCTTGACCTACTTATTTAGTCAGTTACTCAGTGAATACAGGTCAAATAATTACGTAATTTGTCTCATATTCTAAAATTATAAATTGCTCAACCTTTATGTACAGTGACTGCAGGAGGTGTATTTTGTGTCCTGTGCTAAATTACGGTTGGAATCAGCAGATAGTGGCTTAATGGTGGTAATGGATTTGCAATAATCAAGCAATCCTCCAATGATGAATGTGATTCATTTGTGAAATGCTTGGCCAGTCTGCCCTCTATGTGCATAACGTTAAAGGACAACAAAGTATTCAATCCAAAACTCCTTTATAGCATCATTATGTCACTAATTCAGCTTATCAGCCTCAGCCTCGCTGTTATTGCTTCTACTTAGGTGCAAATGTTGCTCAGACAAAAAAAAATAATAAAGGGCAATGTATGATAATCATGTCTCCTTTTCCTACATAAAAGGCAGTTATCAAAAAATCATTGAGCTGGTCCCCAAAGGTGAGCCTGTAAACAGAAAACTGTCCCGTCCTCAGGTCTCACTGTCTCTAAAGGAGGCTCAATTTAAACAGATGTATTTGTGTGTGTGCGCGCATACAACTATAAATCAATTTTTAAAAGGAAAAAAGTTGAAATGTCTTTGAGCAGTGATTTTTAAAAATGACTAGGAGTTAACTGGGCCAAGAAGGCATGATGGACATGGAGGTGCACCACCAAGATTGACCTTCAAGGAAGGGCTTGTTGCTCCAGCTGTGGTCAGGCAGCTTCTAGTTGTTAGTTCTCTCAGGTACCACCTGAGTTGCAGAGTCCTGCCAGCCAATGTCACACCCTCTCTAGGGCAACCCACTACTATTGACTGATGAAGGGGCAGAGTATAAACACTCAGCCATTTTAACCCCATTTGAGACAACTCTGAGGGGTCATCCTAGCTGCAAGTTGTCTATGGGGTTTGAATAAGCGTGTCATAGGGTCTACGTCACAGCTCTAGTTCTTCCTCTTTCCACTCCTGTCACCTCTGCCCTTCTACAGGTGTTTTTTTTTTTTTTTTTTTTTCCAGGGCATGGCTTAATAAACATGCTGCCTGCCAAACTCAACTCCGAAGTCTACTCCCCAGAACCCAGACTGTGATTAAAAGAAGTTGCAGCAGAGGAATAGAAATGAAGAGGGTAGGAATAATCTTATATGGTAAGGATAGTTTCTGTGAAGCTCTGCAGAAGAGCAGAGCAAGAACTGTCTTGCACAGAAGTGACTTGGTAGGCAGGAATTTACAGGGATGGCTTTCAGCAGTGCGAACGTCATCATTCTCCTGTAATTGCAAATTGTATCAATAATATGAACATTCTAAAGTGAATTAGCGTTGCCATTAAGTACATCCCTTGAATTGTCCAGTAAAAGTATGGACCTGACTTCTGATAGCTCACCACTGTTAGCAAGATTTGTTATTACAACCAGCTGGTTGCCAGAGCCAAGTTTAGCACTCCGAAAAAAAAAGAATTGCACTTAAGAAGGAAGAAAAAGATGCCAAACACATGTTATATGCTTTGCAATTTTGTTCAAACCTGCCTATGTTACAAAATATGCGTTACTGTTTTAGGCTAGCAATTTTATTTGAGATGGACAGTAGAAAAACAGTAATAGGAAAACAGAGTTTGCATTCCTGAGTTTTCACAAGTTCAATGAGAAGTCCTCCATGGCACTATTAATAACATAGTATGTTGCATTAGAAAAAGGGGCCAACATGGTTCTATCTGTAAATGCCGCTCGTTATCTGATAGATCTAGAATGTACATTAAGTTGACCAAAATGTCCTTTAATCTTTGCTGAAATTCAAGGGCTTCTACCATTTAATAGGAAGAAAATATGTAATGTTGGTGTAAAAAAAAAAAAAAAAAAGAAAAACACATTTTTAGATTCACTGGATTATGATGAACACTGAGCAAGATAATTTTTCAAAGTGTGCTTAATAGTTTTAGAACAGGGCAGTGTGATATAAGAGAGATGAGATAAGTAAGCAGAGGCTGGGAAGAGTCTACAGTCTTGCAGTCTCAAGGATATGGTGATTCAAGGGCATAAATGAAAAACAACAGGCTATAAGAGAGGTCTCAGACAAAGAAAGATTTAAAAAGCATTCCTTCTGATATTCCTCATTGCAAAAGCGCCTCCACATGCTAATAATGAAAAATGGGCTGGGTGCGGTGGCTCACGCCTGTAATCCAAACACTTTGGGAGGCCGAGACGGGCGGATCACAAGGTCAGGAGATCGAGACCATCCTGGCTAACACGGTGAAACCCCATCTCTACTGAACAAAATACAAAAAATTAGCCGGGCGTGGTGGCGGGCGCCTGTAGTCCCAGCTACTTGGGAGGCTGAGGCAAGAGAATGGCATGAACCCAGGAGGCGGAGCTTGCAGTGAGCAGAGATCGCGCCACTGCACTCCAGCCTGGGTGACAGAGCGAGACTCTGTCTCAAAAATAAAAAATAAAAAAATAATAATAAATGAACAACTTGCAAGCAAGTGAAGAAATTATCAGACATAATAGTCTTGAAAGCTTATCTTGGACTAGTTCAGATGGAGTTGGAAAGGGCTTCAGGATATTTTTAAAAATGGTAAGATCATATAAGAAGTAGTATAAACAATAAATACAGTTCAGAGTGCCTTTTGGAGCTCTGTGCATGCTTGAGTTCTTCTGCTGGCTTTGTCAAAACTTACATCTGTCTTACGCAAAGCGGGATCTGCAAACTGGATTCACTCTAGGAATTATTTGTTACTATGCCATAAGATAAAGTCAACAATTCAGAGTGCATCAAGAAACTTATGTAGCGATTTTACAGAGTACCATTTATGTCTAATGTATTTAATAATAAAGGAGGAAGCATGGGTGTTGGGCATCTTATTTTTGTAACGCTTTGGTTTTATTCTATCTTACATACAATGGATTAGGGGAAAAGTGTTCCTTCCTCAAGATATTTTGAGAAGTACTGAGCAACATATGAAAAGCAGTTTGGGGAGAGATGCGGTATGTTGCTTGCTGGTTCTATTTGACACCATCACTATATGGAACGGGCTGAAAATCGGCCAACTTGGGCTCACTTAAGGCTCCTATGAGCTATTCTTTGTTGCCAGCACATATTAATTCCCGCTCTTTCTCTTCCCCTCTCCCCGCTTACTGTTGTGAAGTAGCATTAAGCCTGTTCAGAGAATTTGGAATAAAAATATATGGGGGCCAATTAGGAGAGCAACATGGCTGCTGAATTTAATAGGTACTCTTTGTGTCTACTCACTATTTGACTCTGAGCAAGAGCAATCTACAATTCAAGTTAAAATTCAAGCCGGGCGCAGTGGCTCATGCCTGCAATCCCAGCACTTTGGGAGGCCGAGGCAGGAGGATCACGAGGTCAGGAGATCGAGACCATCCTGGCTAACACGGTGAAACCCCGTCTCTACTAAAAATACAAAAAAAAATTAGCCGGGCGTGGTGGGGGGGCGCCTGTAGTCCCAGCTACTCGGGAGGCTGAGGCAGGAGAATGGCGTGAATCCGGGAGGCGGAGCTTGCAGTGAGCCGAGATCGCGCCACTGCACTCCAGCCTGGGCGACAGAGCAAGACTCTGTCTCAAAATAAAAATAAAAAAAAATAAAAATAAAAAATAAAAAAATTCAAGTTTACCGAATGAAGCCATGATATCTGCTACAATAGGGCTGGATCTGAACTATTACCCTCATTGCAAGAAGCCAGACACAAAAGAACATGTATTATATTGATATAAAATATCCAAAAAGGCAAATACATATGGACAGAAATAGATTGACAATTGCCTAGGGCTGGGGTGGGAGTGAAAAATGACTGTAAATGGAGAAGAAACTTCTTTTTAAGGTGGTGGAAAAGTGCCAAAATTAGATCGCTATGATGGCTATATAACTCTGTAAATATCCTGCAAATTTTAATGAACGTGGTGGGGGGGAGTTATTCATCCCTCAAAATAGTTTGAGAGGCTACCAAGAACCAGAACTAAACCAGATTGATGCAAATTGAATTTTACACTTAGAATGAGTACATTGATGGTCTAACAATTACATTTCAAAAACGCTGATTTATAAAAAAGTAATCCCGACCAGTAAAAACAGCAGTCATTTCTCAGCCTCCAAATTTTTTAAATGAGTGGAAAACTTTAAGCAATCAATACAAGTCATATGATACACCTCAAATAAAAGTGGTATACTACACTGAGTGCCTTGCAGATATTCTCCCACTTAATCGTGATAAGAAATACATACACATTTATTTTTATTTTTACTCTTTTGATAGTGAAGGGACCATGGCCCAGAACAGGGGTGGCAATATACCAAATGCCACTAAGATGGTAACCGTGGAAGTTTAAATTCAAGTTCATGTTGGTATATATATATACACATACATATTTATACATATGTGTATAACGAGTTACTATTACGTATATATATACAATGAGTGGAAAATATTAGTGGATATAAGTTCATGTATGATTATGTATACATATTATGATTATACTATTATGTATATGTATACAATGAGTGGATAATATTAGTGGATATAAGTTCATGTTGGTCTTATATATACATATATAATGTGTGGATATGAGTGTGTATATATATACATATATAATGTGTGGATATGAGTGTGTGTATATACACACATGCATATTGTATATATGTGTGTATATATAGACACATTATATATATACGCATATATACTACACACACACACACACGGGTGTGTGTCTGTATCTTTTCCACAAATCCTTCAACCCATTTTGCAGAGGTCAAATAGACAGTCGGAAGACCCTATGCTCAGGTGACTTAAAAATAATTTCCAAATCACATTATGGAGTTTGTATGTATTACACACATTTATTGATAGAGATACCCATATTCTACTAATCTTTTATTGGCAATAATTTATGTTAAGAATACCCAAGACTGAGAAAGCCTCATTCCTTTGGTAGTGATTAAAATAAAACATACTAAATTAACTTATAGACAAGTTATAGAACATACATTTGTGAAAAAAATTACTCACCTATGATTGGGACTTTGTATTTTTACCTTATACTTACTCAATGAAATAAAATTTTGAAAAATATTCCTGTAAATGTACCAGAACCTATTTTATACCGTGATGATCCTTAACATTTCAGACGACATGGTAGTGTCAATTTAAAAAGCAGCACTATGGAGCAGGGTTTGTTATTGTTAGAAATACACATTTGTTTCACACGTCAAGGGTAAAAATTAAAAAACAAGATTAATGTTACCCAAAAGGAAACCATTCATCAGGATTCTTACCTGCCAGTGGAGGATTATATTCCAAATCAAACCAAGAGTCAGTTTATGATTTCCATCTACGA', 'pos': '32834760', 'alt': u'GCCAGTGGAGGATTATATTCCAAATCAAACCAAGAGTCAGTTTATGATTTCCATCTACGC'}}
        assert results['NM_004006.2:c.296_476delinsGCGTAGATGGAAATCATAAACTGACTCTTGGTTTGATTTGGAATATAATCCTCCACTGGCAGGTCAAAAATGTAATGAAAAATATCATGGCTGGATTGCAACAAACCAACAGTGAAAAGATTCTCCTGAGCTGGGTCCGACAATCAACTCGTAATTATCCACAGGTTAATGTAATCAACTT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.32816522_32823356delinsAAGTTGATTACATTAACCTGTGGATAATTACGAGTTGATTGTCGGACCCAGCTCAGGAGAATCTTTTCACTGTTGGTTTGTTGCAATCCAGCCATGATATTTTTCATTACATTTTTGACCTGCCAGTGGAGGATTATATTCCAAATCAAACCAAGAGTCAGTTTATGATTTCCATCTACGC', 'vcf': {'chr': 'X', 'ref': 'ACATGTGGAAATAAATTTTCATAAGAAAATGCATTCCTTGAGCAAGAACCATGCAAACTTAAATATGAATGTCCTTGATCTTCAGTGATAAATAGAAATTTTAGGGCCAATTAGTAATGAGACATAATAGATTCTACCAGAAGTTAAGTCTATTCTCAAAGGCTAGGAGTCTATTCTGATTCATTGGTATCTATGCCATACCATTTACTGAATGTGTTCACTATTACTCCTGGATTCTGCCATATTAGACCCAACATTTACTGTTTATAGAATTGCTCACCTTCTAGCTTTCAATGAACAAGCTAGAGAAGACAGTGTATAAAACATTAGACTTATTGTTAACCAAGGACTTAAGATATAAAAGCGCTTTTTGGATAGGAGGAATACTATAGATAATTGTTTTTTAAAGTATATGTTCAAATTTTGCCTTTGCTATTAATTGTGGTAATTGGACAAGTTGCTTAACCTCTTCTAGGCCTCATTTCTGTGTATCTTACATGCAAGGTTAGTGTAAGATCAAAGTACAAAATATGGTCAGATAAAACCCCTGATTTCATACCATCATTTAACGCAAGCTATATTTATTGGCTGAACTGAAAAGAAGACAGAAATGTTTGCCTACCTAAAATAAAGAAAAAAAACAACAAAGGAAGGAAATATTTATCCCTAAGAGGTGTACATATGTAGCCTTGGTTCTATTTTCATTACTTGGTTTCCATGTATTTGATTGTCTCATAGGCTTCGTGCATGTGCAGATGTTATTCAACAGATTATCAAGTTTAATTCTTCTGAGTCATCTGATGCAGAAATCCATTCTAAGCATTTTTTCTAAGTTTCATGAATGCAGTGACGCTTCATAGAAAAAACTGTCTACATCTTTTATTAAAGAATTATTATTTATTAAAGACTGCTTTTTAAATATAAGCCCCTTATGAGATGCTGGAAGAAGACATCTATGTAAAAGTGATCAGTCTGTACCACTAAGCACTTTCCTTCTGGGAAGGTTAAGGTTCATATACTATCAACATTTGATCAGATGATGAATATTTATTGTGCATTCTAAATATGACAATTTTATTTTTTCATCCATATAGAAGTTAACAAATGCCTTTGGGGTAAATGCATATTTTAACATCTCCAATGTTAGTCATCTTTTTATTAAAAAGTAAATGCAAATTTTAGGATATACTACTAAATCTCATTTTAACATACTTTTAAAAGATTTGAGATGATGCTTTGTAAACACAGTAACGATTCGGAGATATGAAATCTTTGCATGTATCTGTCACAATCCCTTGATAACTAAGCCTATTATTTAGGAGTGGATTATCCAAGTTTTCCTAATTCCTTCTTGACCTACTTATTTAGTCAGTTACTCAGTGAATACAGGTCAAATAATTACGTAATTTGTCTCATATTCTAAAATTATAAATTGCTCAACCTTTATGTACAGTGACTGCAGGAGGTGTATTTTGTGTCCTGTGCTAAATTACGGTTGGAATCAGCAGATAGTGGCTTAATGGTGGTAATGGATTTGCAATAATCAAGCAATCCTCCAATGATGAATGTGATTCATTTGTGAAATGCTTGGCCAGTCTGCCCTCTATGTGCATAACGTTAAAGGACAACAAAGTATTCAATCCAAAACTCCTTTATAGCATCATTATGTCACTAATTCAGCTTATCAGCCTCAGCCTCGCTGTTATTGCTTCTACTTAGGTGCAAATGTTGCTCAGACAAAAAAAAATAATAAAGGGCAATGTATGATAATCATGTCTCCTTTTCCTACATAAAAGGCAGTTATCAAAAAATCATTGAGCTGGTCCCCAAAGGTGAGCCTGTAAACAGAAAACTGTCCCGTCCTCAGGTCTCACTGTCTCTAAAGGAGGCTCAATTTAAACAGATGTATTTGTGTGTGTGCGCGCATACAACTATAAATCAATTTTTAAAAGGAAAAAAGTTGAAATGTCTTTGAGCAGTGATTTTTAAAAATGACTAGGAGTTAACTGGGCCAAGAAGGCATGATGGACATGGAGGTGCACCACCAAGATTGACCTTCAAGGAAGGGCTTGTTGCTCCAGCTGTGGTCAGGCAGCTTCTAGTTGTTAGTTCTCTCAGGTACCACCTGAGTTGCAGAGTCCTGCCAGCCAATGTCACACCCTCTCTAGGGCAACCCACTACTATTGACTGATGAAGGGGCAGAGTATAAACACTCAGCCATTTTAACCCCATTTGAGACAACTCTGAGGGGTCATCCTAGCTGCAAGTTGTCTATGGGGTTTGAATAAGCGTGTCATAGGGTCTACGTCACAGCTCTAGTTCTTCCTCTTTCCACTCCTGTCACCTCTGCCCTTCTACAGGTGTTTTTTTTTTTTTTTTTTTTTCCAGGGCATGGCTTAATAAACATGCTGCCTGCCAAACTCAACTCCGAAGTCTACTCCCCAGAACCCAGACTGTGATTAAAAGAAGTTGCAGCAGAGGAATAGAAATGAAGAGGGTAGGAATAATCTTATATGGTAAGGATAGTTTCTGTGAAGCTCTGCAGAAGAGCAGAGCAAGAACTGTCTTGCACAGAAGTGACTTGGTAGGCAGGAATTTACAGGGATGGCTTTCAGCAGTGCGAACGTCATCATTCTCCTGTAATTGCAAATTGTATCAATAATATGAACATTCTAAAGTGAATTAGCGTTGCCATTAAGTACATCCCTTGAATTGTCCAGTAAAAGTATGGACCTGACTTCTGATAGCTCACCACTGTTAGCAAGATTTGTTATTACAACCAGCTGGTTGCCAGAGCCAAGTTTAGCACTCCGAAAAAAAAAGAATTGCACTTAAGAAGGAAGAAAAAGATGCCAAACACATGTTATATGCTTTGCAATTTTGTTCAAACCTGCCTATGTTACAAAATATGCGTTACTGTTTTAGGCTAGCAATTTTATTTGAGATGGACAGTAGAAAAACAGTAATAGGAAAACAGAGTTTGCATTCCTGAGTTTTCACAAGTTCAATGAGAAGTCCTCCATGGCACTATTAATAACATAGTATGTTGCATTAGAAAAAGGGGCCAACATGGTTCTATCTGTAAATGCCGCTCGTTATCTGATAGATCTAGAATGTACATTAAGTTGACCAAAATGTCCTTTAATCTTTGCTGAAATTCAAGGGCTTCTACCATTTAATAGGAAGAAAATATGTAATGTTGGTGTAAAAAAAAAAAAAAAAAAGAAAAACACATTTTTAGATTCACTGGATTATGATGAACACTGAGCAAGATAATTTTTCAAAGTGTGCTTAATAGTTTTAGAACAGGGCAGTGTGATATAAGAGAGATGAGATAAGTAAGCAGAGGCTGGGAAGAGTCTACAGTCTTGCAGTCTCAAGGATATGGTGATTCAAGGGCATAAATGAAAAACAACAGGCTATAAGAGAGGTCTCAGACAAAGAAAGATTTAAAAAGCATTCCTTCTGATATTCCTCATTGCAAAAGCGCCTCCACATGCTAATAATGAAAAATGGGCTGGGTGCGGTGGCTCACGCCTGTAATCCAAACACTTTGGGAGGCCGAGACGGGCGGATCACAAGGTCAGGAGATCGAGACCATCCTGGCTAACACGGTGAAACCCCATCTCTACTGAACAAAATACAAAAAATTAGCCGGGCGTGGTGGCGGGCGCCTGTAGTCCCAGCTACTTGGGAGGCTGAGGCAAGAGAATGGCATGAACCCAGGAGGCGGAGCTTGCAGTGAGCAGAGATCGCGCCACTGCACTCCAGCCTGGGTGACAGAGCGAGACTCTGTCTCAAAAATAAAAAATAAAAAAATAATAATAAATGAACAACTTGCAAGCAAGTGAAGAAATTATCAGACATAATAGTCTTGAAAGCTTATCTTGGACTAGTTCAGATGGAGTTGGAAAGGGCTTCAGGATATTTTTAAAAATGGTAAGATCATATAAGAAGTAGTATAAACAATAAATACAGTTCAGAGTGCCTTTTGGAGCTCTGTGCATGCTTGAGTTCTTCTGCTGGCTTTGTCAAAACTTACATCTGTCTTACGCAAAGCGGGATCTGCAAACTGGATTCACTCTAGGAATTATTTGTTACTATGCCATAAGATAAAGTCAACAATTCAGAGTGCATCAAGAAACTTATGTAGCGATTTTACAGAGTACCATTTATGTCTAATGTATTTAATAATAAAGGAGGAAGCATGGGTGTTGGGCATCTTATTTTTGTAACGCTTTGGTTTTATTCTATCTTACATACAATGGATTAGGGGAAAAGTGTTCCTTCCTCAAGATATTTTGAGAAGTACTGAGCAACATATGAAAAGCAGTTTGGGGAGAGATGCGGTATGTTGCTTGCTGGTTCTATTTGACACCATCACTATATGGAACGGGCTGAAAATCGGCCAACTTGGGCTCACTTAAGGCTCCTATGAGCTATTCTTTGTTGCCAGCACATATTAATTCCCGCTCTTTCTCTTCCCCTCTCCCCGCTTACTGTTGTGAAGTAGCATTAAGCCTGTTCAGAGAATTTGGAATAAAAATATATGGGGGCCAATTAGGAGAGCAACATGGCTGCTGAATTTAATAGGTACTCTTTGTGTCTACTCACTATTTGACTCTGAGCAAGAGCAATCTACAATTCAAGTTAAAATTCAAGCCGGGCGCAGTGGCTCATGCCTGCAATCCCAGCACTTTGGGAGGCCGAGGCAGGAGGATCACGAGGTCAGGAGATCGAGACCATCCTGGCTAACACGGTGAAACCCCGTCTCTACTAAAAATACAAAAAAAAATTAGCCGGGCGTGGTGGGGGGGCGCCTGTAGTCCCAGCTACTCGGGAGGCTGAGGCAGGAGAATGGCGTGAATCCGGGAGGCGGAGCTTGCAGTGAGCCGAGATCGCGCCACTGCACTCCAGCCTGGGCGACAGAGCAAGACTCTGTCTCAAAATAAAAATAAAAAAAAATAAAAATAAAAAATAAAAAAATTCAAGTTTACCGAATGAAGCCATGATATCTGCTACAATAGGGCTGGATCTGAACTATTACCCTCATTGCAAGAAGCCAGACACAAAAGAACATGTATTATATTGATATAAAATATCCAAAAAGGCAAATACATATGGACAGAAATAGATTGACAATTGCCTAGGGCTGGGGTGGGAGTGAAAAATGACTGTAAATGGAGAAGAAACTTCTTTTTAAGGTGGTGGAAAAGTGCCAAAATTAGATCGCTATGATGGCTATATAACTCTGTAAATATCCTGCAAATTTTAATGAACGTGGTGGGGGGGAGTTATTCATCCCTCAAAATAGTTTGAGAGGCTACCAAGAACCAGAACTAAACCAGATTGATGCAAATTGAATTTTACACTTAGAATGAGTACATTGATGGTCTAACAATTACATTTCAAAAACGCTGATTTATAAAAAAGTAATCCCGACCAGTAAAAACAGCAGTCATTTCTCAGCCTCCAAATTTTTTAAATGAGTGGAAAACTTTAAGCAATCAATACAAGTCATATGATACACCTCAAATAAAAGTGGTATACTACACTGAGTGCCTTGCAGATATTCTCCCACTTAATCGTGATAAGAAATACATACACATTTATTTTTATTTTTACTCTTTTGATAGTGAAGGGACCATGGCCCAGAACAGGGGTGGCAATATACCAAATGCCACTAAGATGGTAACCGTGGAAGTTTAAATTCAAGTTCATGTTGGTATATATATATACACATACATATTTATACATATGTGTATAACGAGTTACTATTACGTATATATATACAATGAGTGGAAAATATTAGTGGATATAAGTTCATGTATGATTATGTATACATATTATGATTATACTATTATGTATATGTATACAATGAGTGGATAATATTAGTGGATATAAGTTCATGTTGGTCTTATATATACATATATAATGTGTGGATATGAGTGTGTATATATATACATATATAATGTGTGGATATGAGTGTGTGTATATACACACATGCATATTGTATATATGTGTGTATATATAGACACATTATATATATACGCATATATACTACACACACACACACACGGGTGTGTGTCTGTATCTTTTCCACAAATCCTTCAACCCATTTTGCAGAGGTCAAATAGACAGTCGGAAGACCCTATGCTCAGGTGACTTAAAAATAATTTCCAAATCACATTATGGAGTTTGTATGTATTACACACATTTATTGATAGAGATACCCATATTCTACTAATCTTTTATTGGCAATAATTTATGTTAAGAATACCCAAGACTGAGAAAGCCTCATTCCTTTGGTAGTGATTAAAATAAAACATACTAAATTAACTTATAGACAAGTTATAGAACATACATTTGTGAAAAAAATTACTCACCTATGATTGGGACTTTGTATTTTTACCTTATACTTACTCAATGAAATAAAATTTTGAAAAATATTCCTGTAAATGTACCAGAACCTATTTTATACCGTGATGATCCTTAACATTTCAGACGACATGGTAGTGTCAATTTAAAAAGCAGCACTATGGAGCAGGGTTTGTTATTGTTAGAAATACACATTTGTTTCACACGTCAAGGGTAAAAATTAAAAAACAAGATTAATGTTACCCAAAAGGAAACCATTCATCAGGATTCTTACCTGCCAGTGGAGGATTATATTCCAAATCAAACCAAGAGTCAGTTTATGATTTCCATCTACGA', 'pos': '32816643', 'alt': u'GCCAGTGGAGGATTATATTCCAAATCAAACCAAGAGTCAGTTTATGATTTCCATCTACGC'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_004006.2:c.476T=' in results.keys()
        assert results['NM_004006.2:c.476T=']['hgvs_lrg_transcript_variant'] == 'LRG_199t1:c.476T='
        assert results['NM_004006.2:c.476T=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_004006.2:c.476T=']['alt_genomic_loci'] == []
        assert results['NM_004006.2:c.476T=']['gene_symbol'] == 'DMD'
        assert results['NM_004006.2:c.476T=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003997.1(LRG_199p1):p.(Phe159=)', 'slr': 'NP_003997.1:p.(F159=)'}
        assert results['NM_004006.2:c.476T=']['submitted_variant'] == 'LRG_199t1:c.[296T>G;476T=];[476T=](;)1083A>C'
        assert results['NM_004006.2:c.476T=']['genome_context_intronic_sequence'] == ''
        assert results['NM_004006.2:c.476T=']['hgvs_lrg_variant'] == 'LRG_199:g.528088T='
        assert results['NM_004006.2:c.476T=']['hgvs_transcript_variant'] == 'NM_004006.2:c.476T='
        assert results['NM_004006.2:c.476T=']['hgvs_refseqgene_variant'] == 'NG_012232.1:g.528088T='
        assert results['NM_004006.2:c.476T=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.32834639A=', 'vcf': {'chr': 'chrX', 'ref': u'A', 'pos': '32834639', 'alt': u'A'}}
        assert results['NM_004006.2:c.476T=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.32816522A=', 'vcf': {'chr': 'chrX', 'ref': u'A', 'pos': '32816522', 'alt': u'A'}}
        assert results['NM_004006.2:c.476T=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.32834639A=', 'vcf': {'chr': 'X', 'ref': u'A', 'pos': '32834639', 'alt': u'A'}}
        assert results['NM_004006.2:c.476T=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.32816522A=', 'vcf': {'chr': 'X', 'ref': u'A', 'pos': '32816522', 'alt': u'A'}}

        assert 'NM_004006.2:c.1083A>C' in results.keys()
        assert results['NM_004006.2:c.1083A>C']['hgvs_lrg_transcript_variant'] == 'LRG_199t1:c.1083A>C'
        assert results['NM_004006.2:c.1083A>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_004006.2:c.1083A>C']['alt_genomic_loci'] == []
        assert results['NM_004006.2:c.1083A>C']['gene_symbol'] == 'DMD'
        assert results['NM_004006.2:c.1083A>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003997.1(LRG_199p1):p.(Thr361=)', 'slr': 'NP_003997.1:p.(T361=)'}
        assert results['NM_004006.2:c.1083A>C']['submitted_variant'] == 'LRG_199t1:c.[296T>G;476T=];[476T=](;)1083A>C'
        assert results['NM_004006.2:c.1083A>C']['genome_context_intronic_sequence'] == ''
        assert results['NM_004006.2:c.1083A>C']['hgvs_lrg_variant'] == 'LRG_199:g.699580A>C'
        assert results['NM_004006.2:c.1083A>C']['hgvs_transcript_variant'] == 'NM_004006.2:c.1083A>C'
        assert results['NM_004006.2:c.1083A>C']['hgvs_refseqgene_variant'] == 'NG_012232.1:g.699580A>C'
        assert results['NM_004006.2:c.1083A>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.32663147T>G', 'vcf': {'chr': 'chrX', 'ref': u'T', 'pos': '32663147', 'alt': u'G'}}
        assert results['NM_004006.2:c.1083A>C']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.32645030T>G', 'vcf': {'chr': 'chrX', 'ref': u'T', 'pos': '32645030', 'alt': u'G'}}
        assert results['NM_004006.2:c.1083A>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.32663147T>G', 'vcf': {'chr': 'X', 'ref': u'T', 'pos': '32663147', 'alt': u'G'}}
        assert results['NM_004006.2:c.1083A>C']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.32645030T>G', 'vcf': {'chr': 'X', 'ref': u'T', 'pos': '32645030', 'alt': u'G'}}


    def test_variant197(self):
        variant = 'LRG_199t1:c.[296T>G];[476T>C](;)1083A>C(;)1406del'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_004006.2:c.1408del' in results.keys()
        assert results['NM_004006.2:c.1408del']['hgvs_lrg_transcript_variant'] == 'LRG_199t1:c.1408del'
        assert results['NM_004006.2:c.1408del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_004006.2:c.1408del']['alt_genomic_loci'] == []
        assert results['NM_004006.2:c.1408del']['gene_symbol'] == 'DMD'
        assert results['NM_004006.2:c.1408del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003997.1(LRG_199p1):p.(Arg470GlufsTer17)', 'slr': 'NP_003997.1:p.(R470Efs*17)'}
        assert results['NM_004006.2:c.1408del']['submitted_variant'] == 'LRG_199t1:c.[296T>G];[476T>C](;)1083A>C(;)1406del'
        assert results['NM_004006.2:c.1408del']['genome_context_intronic_sequence'] == ''
        assert results['NM_004006.2:c.1408del']['hgvs_lrg_variant'] == 'LRG_199:g.730233del'
        assert results['NM_004006.2:c.1408del']['hgvs_transcript_variant'] == 'NM_004006.2:c.1408del'
        assert results['NM_004006.2:c.1408del']['hgvs_refseqgene_variant'] == 'NG_012232.1:g.730233del'
        assert results['NM_004006.2:c.1408del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.32632494del', 'vcf': {'chr': 'chrX', 'ref': 'CT', 'pos': '32632493', 'alt': 'C'}}
        assert results['NM_004006.2:c.1408del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.32614377del', 'vcf': {'chr': 'chrX', 'ref': 'CT', 'pos': '32614376', 'alt': 'C'}}
        assert results['NM_004006.2:c.1408del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.32632494del', 'vcf': {'chr': 'X', 'ref': 'CT', 'pos': '32632493', 'alt': 'C'}}
        assert results['NM_004006.2:c.1408del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.32614377del', 'vcf': {'chr': 'X', 'ref': 'CT', 'pos': '32614376', 'alt': 'C'}}

        assert 'NM_004006.2:c.476T>C' in results.keys()
        assert results['NM_004006.2:c.476T>C']['hgvs_lrg_transcript_variant'] == 'LRG_199t1:c.476T>C'
        assert results['NM_004006.2:c.476T>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_004006.2:c.476T>C']['alt_genomic_loci'] == []
        assert results['NM_004006.2:c.476T>C']['gene_symbol'] == 'DMD'
        assert results['NM_004006.2:c.476T>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003997.1(LRG_199p1):p.(Phe159Ser)', 'slr': 'NP_003997.1:p.(F159S)'}
        assert results['NM_004006.2:c.476T>C']['submitted_variant'] == 'LRG_199t1:c.[296T>G];[476T>C](;)1083A>C(;)1406del'
        assert results['NM_004006.2:c.476T>C']['genome_context_intronic_sequence'] == ''
        assert results['NM_004006.2:c.476T>C']['hgvs_lrg_variant'] == 'LRG_199:g.528088T>C'
        assert results['NM_004006.2:c.476T>C']['hgvs_transcript_variant'] == 'NM_004006.2:c.476T>C'
        assert results['NM_004006.2:c.476T>C']['hgvs_refseqgene_variant'] == 'NG_012232.1:g.528088T>C'
        assert results['NM_004006.2:c.476T>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.32834639A>G', 'vcf': {'chr': 'chrX', 'ref': u'A', 'pos': '32834639', 'alt': u'G'}}
        assert results['NM_004006.2:c.476T>C']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.32816522A>G', 'vcf': {'chr': 'chrX', 'ref': u'A', 'pos': '32816522', 'alt': u'G'}}
        assert results['NM_004006.2:c.476T>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.32834639A>G', 'vcf': {'chr': 'X', 'ref': u'A', 'pos': '32834639', 'alt': u'G'}}
        assert results['NM_004006.2:c.476T>C']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.32816522A>G', 'vcf': {'chr': 'X', 'ref': u'A', 'pos': '32816522', 'alt': u'G'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_004006.2:c.1083A>C' in results.keys()
        assert results['NM_004006.2:c.1083A>C']['hgvs_lrg_transcript_variant'] == 'LRG_199t1:c.1083A>C'
        assert results['NM_004006.2:c.1083A>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_004006.2:c.1083A>C']['alt_genomic_loci'] == []
        assert results['NM_004006.2:c.1083A>C']['gene_symbol'] == 'DMD'
        assert results['NM_004006.2:c.1083A>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003997.1(LRG_199p1):p.(Thr361=)', 'slr': 'NP_003997.1:p.(T361=)'}
        assert results['NM_004006.2:c.1083A>C']['submitted_variant'] == 'LRG_199t1:c.[296T>G];[476T>C](;)1083A>C(;)1406del'
        assert results['NM_004006.2:c.1083A>C']['genome_context_intronic_sequence'] == ''
        assert results['NM_004006.2:c.1083A>C']['hgvs_lrg_variant'] == 'LRG_199:g.699580A>C'
        assert results['NM_004006.2:c.1083A>C']['hgvs_transcript_variant'] == 'NM_004006.2:c.1083A>C'
        assert results['NM_004006.2:c.1083A>C']['hgvs_refseqgene_variant'] == 'NG_012232.1:g.699580A>C'
        assert results['NM_004006.2:c.1083A>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.32663147T>G', 'vcf': {'chr': 'chrX', 'ref': u'T', 'pos': '32663147', 'alt': u'G'}}
        assert results['NM_004006.2:c.1083A>C']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.32645030T>G', 'vcf': {'chr': 'chrX', 'ref': u'T', 'pos': '32645030', 'alt': u'G'}}
        assert results['NM_004006.2:c.1083A>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.32663147T>G', 'vcf': {'chr': 'X', 'ref': u'T', 'pos': '32663147', 'alt': u'G'}}
        assert results['NM_004006.2:c.1083A>C']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.32645030T>G', 'vcf': {'chr': 'X', 'ref': u'T', 'pos': '32645030', 'alt': u'G'}}

        assert 'NM_004006.2:c.296T>G' in results.keys()
        assert results['NM_004006.2:c.296T>G']['hgvs_lrg_transcript_variant'] == 'LRG_199t1:c.296T>G'
        assert results['NM_004006.2:c.296T>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_004006.2:c.296T>G']['alt_genomic_loci'] == []
        assert results['NM_004006.2:c.296T>G']['gene_symbol'] == 'DMD'
        assert results['NM_004006.2:c.296T>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003997.1(LRG_199p1):p.(Ile99Ser)', 'slr': 'NP_003997.1:p.(I99S)'}
        assert results['NM_004006.2:c.296T>G']['submitted_variant'] == 'LRG_199t1:c.[296T>G];[476T>C](;)1083A>C(;)1406del'
        assert results['NM_004006.2:c.296T>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_004006.2:c.296T>G']['hgvs_lrg_variant'] == 'LRG_199:g.521254T>G'
        assert results['NM_004006.2:c.296T>G']['hgvs_transcript_variant'] == 'NM_004006.2:c.296T>G'
        assert results['NM_004006.2:c.296T>G']['hgvs_refseqgene_variant'] == 'NG_012232.1:g.521254T>G'
        assert results['NM_004006.2:c.296T>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.32841473A>C', 'vcf': {'chr': 'chrX', 'ref': u'A', 'pos': '32841473', 'alt': u'C'}}
        assert results['NM_004006.2:c.296T>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.32823356A>C', 'vcf': {'chr': 'chrX', 'ref': u'A', 'pos': '32823356', 'alt': u'C'}}
        assert results['NM_004006.2:c.296T>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.32841473A>C', 'vcf': {'chr': 'X', 'ref': u'A', 'pos': '32841473', 'alt': u'C'}}
        assert results['NM_004006.2:c.296T>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.32823356A>C', 'vcf': {'chr': 'X', 'ref': u'A', 'pos': '32823356', 'alt': u'C'}}


    def test_variant198(self):
        variant = 'LRG_199t1:c.[976-20T>A;976-17_976-1dup]'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'validation_warning_1' in results.keys()
        assert results['validation_warning_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['validation_warning_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['alt_genomic_loci'] == []
        assert results['validation_warning_1']['gene_symbol'] == ''
        assert results['validation_warning_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['validation_warning_1']['submitted_variant'] == 'LRG_199t1:c.[976-20T>A;976-17_976-1dup]'
        assert results['validation_warning_1']['genome_context_intronic_sequence'] == ''
        assert results['validation_warning_1']['hgvs_lrg_variant'] == ''
        assert results['validation_warning_1']['hgvs_transcript_variant'] == ''
        assert results['validation_warning_1']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['validation_warning_1']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['validation_warning_1']['primary_assembly_loci'].keys()

        assert results['flag'] == 'warning'

    def test_variant199(self):
        variant = '1-5935162-A-T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_015102.3:c.2818-2T>A' in results.keys()
        assert results['NM_015102.3:c.2818-2T>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_015102.3:c.2818-2T>A']['refseqgene_context_intronic_sequence'] == 'NG_011724.2(NM_015102.3):c.2818-2A='
        assert results['NM_015102.3:c.2818-2T>A']['alt_genomic_loci'] == []
        assert results['NM_015102.3:c.2818-2T>A']['gene_symbol'] == 'NPHP4'
        assert results['NM_015102.3:c.2818-2T>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_055917.1:p.?', 'slr': 'NP_055917.1:p.?'}
        assert results['NM_015102.3:c.2818-2T>A']['submitted_variant'] == '1-5935162-A-T'
        assert results['NM_015102.3:c.2818-2T>A']['genome_context_intronic_sequence'] == 'NC_000001.10(NM_015102.3):c.2818-2T>A'
        assert results['NM_015102.3:c.2818-2T>A']['hgvs_lrg_variant'] == ''
        assert results['NM_015102.3:c.2818-2T>A']['hgvs_transcript_variant'] == 'NM_015102.3:c.2818-2T>A'
        assert results['NM_015102.3:c.2818-2T>A']['hgvs_refseqgene_variant'] == 'NG_011724.2:g.122370A='
        assert results['NM_015102.3:c.2818-2T>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.5935162A>T', 'vcf': {'chr': 'chr1', 'ref': u'A', 'pos': '5935162', 'alt': u'T'}}
        assert 'hg38' not in results['NM_015102.3:c.2818-2T>A']['primary_assembly_loci'].keys()
        assert results['NM_015102.3:c.2818-2T>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.5935162A>T', 'vcf': {'chr': '1', 'ref': u'A', 'pos': '5935162', 'alt': u'T'}}
        assert 'grch38' not in results['NM_015102.3:c.2818-2T>A']['primary_assembly_loci'].keys()

        assert 'NM_001291593.1:c.1279-2T>A' in results.keys()
        assert results['NM_001291593.1:c.1279-2T>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001291593.1:c.1279-2T>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001291593.1:c.1279-2T>A']['alt_genomic_loci'] == []
        assert results['NM_001291593.1:c.1279-2T>A']['gene_symbol'] == 'NPHP4'
        assert results['NM_001291593.1:c.1279-2T>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001278522.1:p.?', 'slr': 'NP_001278522.1:p.?'}
        assert results['NM_001291593.1:c.1279-2T>A']['submitted_variant'] == '1-5935162-A-T'
        assert results['NM_001291593.1:c.1279-2T>A']['genome_context_intronic_sequence'] == 'NC_000001.10(NM_001291593.1):c.1279-2T>A'
        assert results['NM_001291593.1:c.1279-2T>A']['hgvs_lrg_variant'] == ''
        assert results['NM_001291593.1:c.1279-2T>A']['hgvs_transcript_variant'] == 'NM_001291593.1:c.1279-2T>A'
        assert results['NM_001291593.1:c.1279-2T>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001291593.1:c.1279-2T>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.5935162A>T', 'vcf': {'chr': 'chr1', 'ref': u'A', 'pos': '5935162', 'alt': u'T'}}
        assert results['NM_001291593.1:c.1279-2T>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.5875102T=', 'vcf': {'chr': 'chr1', 'ref': 'T', 'pos': '5875102', 'alt': 'T'}}
        assert results['NM_001291593.1:c.1279-2T>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.5935162A>T', 'vcf': {'chr': '1', 'ref': u'A', 'pos': '5935162', 'alt': u'T'}}
        assert results['NM_001291593.1:c.1279-2T>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.5875102T=', 'vcf': {'chr': '1', 'ref': 'T', 'pos': '5875102', 'alt': 'T'}}

        assert 'NM_015102.4:c.2818-2T>A' in results.keys()
        assert results['NM_015102.4:c.2818-2T>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_015102.4:c.2818-2T>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_015102.4:c.2818-2T>A']['alt_genomic_loci'] == []
        assert results['NM_015102.4:c.2818-2T>A']['gene_symbol'] == 'NPHP4'
        assert results['NM_015102.4:c.2818-2T>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_055917.1:p.?', 'slr': 'NP_055917.1:p.?'}
        assert results['NM_015102.4:c.2818-2T>A']['submitted_variant'] == '1-5935162-A-T'
        assert results['NM_015102.4:c.2818-2T>A']['genome_context_intronic_sequence'] == 'NC_000001.10(NM_015102.4):c.2818-2T>A'
        assert results['NM_015102.4:c.2818-2T>A']['hgvs_lrg_variant'] == ''
        assert results['NM_015102.4:c.2818-2T>A']['hgvs_transcript_variant'] == 'NM_015102.4:c.2818-2T>A'
        assert results['NM_015102.4:c.2818-2T>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_015102.4:c.2818-2T>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.5935162A>T', 'vcf': {'chr': 'chr1', 'ref': u'A', 'pos': '5935162', 'alt': u'T'}}
        assert results['NM_015102.4:c.2818-2T>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.5875102T=', 'vcf': {'chr': 'chr1', 'ref': 'T', 'pos': '5875102', 'alt': 'T'}}
        assert results['NM_015102.4:c.2818-2T>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.5935162A>T', 'vcf': {'chr': '1', 'ref': u'A', 'pos': '5935162', 'alt': u'T'}}
        assert results['NM_015102.4:c.2818-2T>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.5875102T=', 'vcf': {'chr': '1', 'ref': 'T', 'pos': '5875102', 'alt': 'T'}}

        assert 'NM_001291594.1:c.1282-2T>A' in results.keys()
        assert results['NM_001291594.1:c.1282-2T>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001291594.1:c.1282-2T>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001291594.1:c.1282-2T>A']['alt_genomic_loci'] == []
        assert results['NM_001291594.1:c.1282-2T>A']['gene_symbol'] == 'NPHP4'
        assert results['NM_001291594.1:c.1282-2T>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001278523.1:p.?', 'slr': 'NP_001278523.1:p.?'}
        assert results['NM_001291594.1:c.1282-2T>A']['submitted_variant'] == '1-5935162-A-T'
        assert results['NM_001291594.1:c.1282-2T>A']['genome_context_intronic_sequence'] == 'NC_000001.10(NM_001291594.1):c.1282-2T>A'
        assert results['NM_001291594.1:c.1282-2T>A']['hgvs_lrg_variant'] == ''
        assert results['NM_001291594.1:c.1282-2T>A']['hgvs_transcript_variant'] == 'NM_001291594.1:c.1282-2T>A'
        assert results['NM_001291594.1:c.1282-2T>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001291594.1:c.1282-2T>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.5935162A>T', 'vcf': {'chr': 'chr1', 'ref': u'A', 'pos': '5935162', 'alt': u'T'}}
        assert results['NM_001291594.1:c.1282-2T>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.5875102T=', 'vcf': {'chr': 'chr1', 'ref': 'T', 'pos': '5875102', 'alt': 'T'}}
        assert results['NM_001291594.1:c.1282-2T>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.5935162A>T', 'vcf': {'chr': '1', 'ref': u'A', 'pos': '5935162', 'alt': u'T'}}
        assert results['NM_001291594.1:c.1282-2T>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.5875102T=', 'vcf': {'chr': '1', 'ref': 'T', 'pos': '5875102', 'alt': 'T'}}

        assert results['flag'] == 'gene_variant'
        assert 'NR_111987.1:n.3633-2T>A' in results.keys()
        assert results['NR_111987.1:n.3633-2T>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_111987.1:n.3633-2T>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_111987.1:n.3633-2T>A']['alt_genomic_loci'] == []
        assert results['NR_111987.1:n.3633-2T>A']['gene_symbol'] == 'NPHP4'
        assert results['NR_111987.1:n.3633-2T>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_111987.1:n.3633-2T>A']['submitted_variant'] == '1-5935162-A-T'
        assert results['NR_111987.1:n.3633-2T>A']['genome_context_intronic_sequence'] == 'NC_000001.10(NR_111987.1):c.3633-2T>A'
        assert results['NR_111987.1:n.3633-2T>A']['hgvs_lrg_variant'] == ''
        assert results['NR_111987.1:n.3633-2T>A']['hgvs_transcript_variant'] == 'NR_111987.1:n.3633-2T>A'
        assert results['NR_111987.1:n.3633-2T>A']['hgvs_refseqgene_variant'] == ''
        assert results['NR_111987.1:n.3633-2T>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.5935162A>T', 'vcf': {'chr': 'chr1', 'ref': u'A', 'pos': '5935162', 'alt': u'T'}}
        assert results['NR_111987.1:n.3633-2T>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.5875102T=', 'vcf': {'chr': 'chr1', 'ref': 'T', 'pos': '5875102', 'alt': 'T'}}
        assert results['NR_111987.1:n.3633-2T>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.5935162A>T', 'vcf': {'chr': '1', 'ref': u'A', 'pos': '5935162', 'alt': u'T'}}
        assert results['NR_111987.1:n.3633-2T>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.5875102T=', 'vcf': {'chr': '1', 'ref': 'T', 'pos': '5875102', 'alt': 'T'}}


    def test_variant200(self):
        variant = '1-12065948-C-T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_001127660.1:c.1676C>T' in results.keys()
        assert results['NM_001127660.1:c.1676C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001127660.1:c.1676C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001127660.1:c.1676C>T']['alt_genomic_loci'] == []
        assert results['NM_001127660.1:c.1676C>T']['gene_symbol'] == 'MFN2'
        assert results['NM_001127660.1:c.1676C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001121132.1:p.(Pro559Leu)', 'slr': 'NP_001121132.1:p.(P559L)'}
        assert results['NM_001127660.1:c.1676C>T']['submitted_variant'] == '1-12065948-C-T'
        assert results['NM_001127660.1:c.1676C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001127660.1:c.1676C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001127660.1:c.1676C>T']['hgvs_transcript_variant'] == 'NM_001127660.1:c.1676C>T'
        assert results['NM_001127660.1:c.1676C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001127660.1:c.1676C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.12065948C>T', 'vcf': {'chr': 'chr1', 'ref': 'C', 'pos': '12065948', 'alt': 'T'}}
        assert results['NM_001127660.1:c.1676C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.12005891C>T', 'vcf': {'chr': 'chr1', 'ref': 'C', 'pos': '12005891', 'alt': 'T'}}
        assert results['NM_001127660.1:c.1676C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.12065948C>T', 'vcf': {'chr': '1', 'ref': 'C', 'pos': '12065948', 'alt': 'T'}}
        assert results['NM_001127660.1:c.1676C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.12005891C>T', 'vcf': {'chr': '1', 'ref': 'C', 'pos': '12005891', 'alt': 'T'}}

        assert 'NM_014874.3:c.1676C>T' in results.keys()
        assert results['NM_014874.3:c.1676C>T']['hgvs_lrg_transcript_variant'] == 'LRG_255t1:c.1676C>T'
        assert results['NM_014874.3:c.1676C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_014874.3:c.1676C>T']['alt_genomic_loci'] == []
        assert results['NM_014874.3:c.1676C>T']['gene_symbol'] == 'MFN2'
        assert results['NM_014874.3:c.1676C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_055689.1(LRG_255p1):p.(Pro559Leu)', 'slr': 'NP_055689.1:p.(P559L)'}
        assert results['NM_014874.3:c.1676C>T']['submitted_variant'] == '1-12065948-C-T'
        assert results['NM_014874.3:c.1676C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_014874.3:c.1676C>T']['hgvs_lrg_variant'] == 'LRG_255:g.30711C>T'
        assert results['NM_014874.3:c.1676C>T']['hgvs_transcript_variant'] == 'NM_014874.3:c.1676C>T'
        assert results['NM_014874.3:c.1676C>T']['hgvs_refseqgene_variant'] == 'NG_007945.1:g.30711C>T'
        assert results['NM_014874.3:c.1676C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.12065948C>T', 'vcf': {'chr': 'chr1', 'ref': 'C', 'pos': '12065948', 'alt': 'T'}}
        assert results['NM_014874.3:c.1676C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.12005891C>T', 'vcf': {'chr': 'chr1', 'ref': 'C', 'pos': '12005891', 'alt': 'T'}}
        assert results['NM_014874.3:c.1676C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.12065948C>T', 'vcf': {'chr': '1', 'ref': 'C', 'pos': '12065948', 'alt': 'T'}}
        assert results['NM_014874.3:c.1676C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.12005891C>T', 'vcf': {'chr': '1', 'ref': 'C', 'pos': '12005891', 'alt': 'T'}}


    def test_variant201(self):
        variant = '1-46655125-CTCAC-C'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_017739.3:c.1895+5_1895+8del' in results.keys()
        assert results['NM_017739.3:c.1895+5_1895+8del']['hgvs_lrg_transcript_variant'] == 'LRG_701t2:c.1895+5_1895+8del'
        assert results['NM_017739.3:c.1895+5_1895+8del']['refseqgene_context_intronic_sequence'] == 'NG_009205.2(NM_017739.3):c.1895+5_1895+8del'
        assert results['NM_017739.3:c.1895+5_1895+8del']['alt_genomic_loci'] == []
        assert results['NM_017739.3:c.1895+5_1895+8del']['gene_symbol'] == 'POMGNT1'
        assert results['NM_017739.3:c.1895+5_1895+8del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_060209.3(LRG_701p2):p.?', 'slr': 'NP_060209.3:p.?'}
        assert results['NM_017739.3:c.1895+5_1895+8del']['submitted_variant'] == '1-46655125-CTCAC-C'
        assert results['NM_017739.3:c.1895+5_1895+8del']['genome_context_intronic_sequence'] == 'NC_000001.10(NM_017739.3):c.1895+5_1895+8del'
        assert results['NM_017739.3:c.1895+5_1895+8del']['hgvs_lrg_variant'] == 'LRG_701:g.35853_35856del'
        assert results['NM_017739.3:c.1895+5_1895+8del']['hgvs_transcript_variant'] == 'NM_017739.3:c.1895+5_1895+8del'
        assert results['NM_017739.3:c.1895+5_1895+8del']['hgvs_refseqgene_variant'] == 'NG_009205.2:g.35853_35856del'
        assert results['NM_017739.3:c.1895+5_1895+8del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.46655122_46655125del', 'vcf': {'chr': 'chr1', 'ref': 'GTCAC', 'pos': '46655121', 'alt': 'G'}}
        assert results['NM_017739.3:c.1895+5_1895+8del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.46189450_46189453del', 'vcf': {'chr': 'chr1', 'ref': 'GTCAC', 'pos': '46189449', 'alt': 'G'}}
        assert results['NM_017739.3:c.1895+5_1895+8del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.46655122_46655125del', 'vcf': {'chr': '1', 'ref': 'GTCAC', 'pos': '46655121', 'alt': 'G'}}
        assert results['NM_017739.3:c.1895+5_1895+8del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.46189450_46189453del', 'vcf': {'chr': '1', 'ref': 'GTCAC', 'pos': '46189449', 'alt': 'G'}}

        assert 'NM_001243766.1:c.1869+31_1869+34del' in results.keys()
        assert results['NM_001243766.1:c.1869+31_1869+34del']['hgvs_lrg_transcript_variant'] == 'LRG_701t1:c.1869+31_1869+34del'
        assert results['NM_001243766.1:c.1869+31_1869+34del']['refseqgene_context_intronic_sequence'] == 'NG_009205.2(NM_001243766.1):c.1869+31_1869+34del'
        assert results['NM_001243766.1:c.1869+31_1869+34del']['alt_genomic_loci'] == []
        assert results['NM_001243766.1:c.1869+31_1869+34del']['gene_symbol'] == 'POMGNT1'
        assert results['NM_001243766.1:c.1869+31_1869+34del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001230695.1:p.?', 'slr': 'NP_001230695.1:p.?'}
        assert results['NM_001243766.1:c.1869+31_1869+34del']['submitted_variant'] == '1-46655125-CTCAC-C'
        assert results['NM_001243766.1:c.1869+31_1869+34del']['genome_context_intronic_sequence'] == 'NC_000001.10(NM_001243766.1):c.1869+31_1869+34del'
        assert results['NM_001243766.1:c.1869+31_1869+34del']['hgvs_lrg_variant'] == 'LRG_701:g.35853_35856del'
        assert results['NM_001243766.1:c.1869+31_1869+34del']['hgvs_transcript_variant'] == 'NM_001243766.1:c.1869+31_1869+34del'
        assert results['NM_001243766.1:c.1869+31_1869+34del']['hgvs_refseqgene_variant'] == 'NG_009205.2:g.35853_35856del'
        assert results['NM_001243766.1:c.1869+31_1869+34del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.46655122_46655125del', 'vcf': {'chr': 'chr1', 'ref': 'GTCAC', 'pos': '46655121', 'alt': 'G'}}
        assert results['NM_001243766.1:c.1869+31_1869+34del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.46189450_46189453del', 'vcf': {'chr': 'chr1', 'ref': 'GTCAC', 'pos': '46189449', 'alt': 'G'}}
        assert results['NM_001243766.1:c.1869+31_1869+34del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.46655122_46655125del', 'vcf': {'chr': '1', 'ref': 'GTCAC', 'pos': '46655121', 'alt': 'G'}}
        assert results['NM_001243766.1:c.1869+31_1869+34del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.46189450_46189453del', 'vcf': {'chr': '1', 'ref': 'GTCAC', 'pos': '46189449', 'alt': 'G'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001290129.1:c.1829+5_1829+8del' in results.keys()
        assert results['NM_001290129.1:c.1829+5_1829+8del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001290129.1:c.1829+5_1829+8del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001290129.1:c.1829+5_1829+8del']['alt_genomic_loci'] == []
        assert results['NM_001290129.1:c.1829+5_1829+8del']['gene_symbol'] == 'POMGNT1'
        assert results['NM_001290129.1:c.1829+5_1829+8del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001277058.1:p.?', 'slr': 'NP_001277058.1:p.?'}
        assert results['NM_001290129.1:c.1829+5_1829+8del']['submitted_variant'] == '1-46655125-CTCAC-C'
        assert results['NM_001290129.1:c.1829+5_1829+8del']['genome_context_intronic_sequence'] == 'NC_000001.10(NM_001290129.1):c.1829+5_1829+8del'
        assert results['NM_001290129.1:c.1829+5_1829+8del']['hgvs_lrg_variant'] == ''
        assert results['NM_001290129.1:c.1829+5_1829+8del']['hgvs_transcript_variant'] == 'NM_001290129.1:c.1829+5_1829+8del'
        assert results['NM_001290129.1:c.1829+5_1829+8del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001290129.1:c.1829+5_1829+8del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.46655122_46655125del', 'vcf': {'chr': 'chr1', 'ref': 'GTCAC', 'pos': '46655121', 'alt': 'G'}}
        assert results['NM_001290129.1:c.1829+5_1829+8del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.46189450_46189453del', 'vcf': {'chr': 'chr1', 'ref': 'GTCAC', 'pos': '46189449', 'alt': 'G'}}
        assert results['NM_001290129.1:c.1829+5_1829+8del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.46655122_46655125del', 'vcf': {'chr': '1', 'ref': 'GTCAC', 'pos': '46655121', 'alt': 'G'}}
        assert results['NM_001290129.1:c.1829+5_1829+8del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.46189450_46189453del', 'vcf': {'chr': '1', 'ref': 'GTCAC', 'pos': '46189449', 'alt': 'G'}}

        assert 'NM_001290130.1:c.1466+5_1466+8del' in results.keys()
        assert results['NM_001290130.1:c.1466+5_1466+8del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001290130.1:c.1466+5_1466+8del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001290130.1:c.1466+5_1466+8del']['alt_genomic_loci'] == []
        assert results['NM_001290130.1:c.1466+5_1466+8del']['gene_symbol'] == 'POMGNT1'
        assert results['NM_001290130.1:c.1466+5_1466+8del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001277059.1:p.?', 'slr': 'NP_001277059.1:p.?'}
        assert results['NM_001290130.1:c.1466+5_1466+8del']['submitted_variant'] == '1-46655125-CTCAC-C'
        assert results['NM_001290130.1:c.1466+5_1466+8del']['genome_context_intronic_sequence'] == 'NC_000001.10(NM_001290130.1):c.1466+5_1466+8del'
        assert results['NM_001290130.1:c.1466+5_1466+8del']['hgvs_lrg_variant'] == ''
        assert results['NM_001290130.1:c.1466+5_1466+8del']['hgvs_transcript_variant'] == 'NM_001290130.1:c.1466+5_1466+8del'
        assert results['NM_001290130.1:c.1466+5_1466+8del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001290130.1:c.1466+5_1466+8del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.46655122_46655125del', 'vcf': {'chr': 'chr1', 'ref': 'GTCAC', 'pos': '46655121', 'alt': 'G'}}
        assert results['NM_001290130.1:c.1466+5_1466+8del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.46189450_46189453del', 'vcf': {'chr': 'chr1', 'ref': 'GTCAC', 'pos': '46189449', 'alt': 'G'}}
        assert results['NM_001290130.1:c.1466+5_1466+8del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.46655122_46655125del', 'vcf': {'chr': '1', 'ref': 'GTCAC', 'pos': '46655121', 'alt': 'G'}}
        assert results['NM_001290130.1:c.1466+5_1466+8del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.46189450_46189453del', 'vcf': {'chr': '1', 'ref': 'GTCAC', 'pos': '46189449', 'alt': 'G'}}


    def test_variant202(self):
        variant = '1-68912523-TGAGCCAGAG-T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_000329.2:c.106_114del' in results.keys()
        assert results['NM_000329.2:c.106_114del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000329.2:c.106_114del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000329.2:c.106_114del']['alt_genomic_loci'] == []
        assert results['NM_000329.2:c.106_114del']['gene_symbol'] == 'RPE65'
        assert results['NM_000329.2:c.106_114del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000320.1:p.(Leu36_Leu38del)', 'slr': 'NP_000320.1:p.(L36_L38del)'}
        assert results['NM_000329.2:c.106_114del']['submitted_variant'] == '1-68912523-TGAGCCAGAG-T'
        assert results['NM_000329.2:c.106_114del']['genome_context_intronic_sequence'] == ''
        assert results['NM_000329.2:c.106_114del']['hgvs_lrg_variant'] == ''
        assert results['NM_000329.2:c.106_114del']['hgvs_transcript_variant'] == 'NM_000329.2:c.106_114del'
        assert results['NM_000329.2:c.106_114del']['hgvs_refseqgene_variant'] == 'NG_008472.1:g.8111_8119del'
        assert results['NM_000329.2:c.106_114del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.68912524_68912532del', 'vcf': {'chr': 'chr1', 'ref': 'TGAGCCAGAG', 'pos': '68912523', 'alt': 'T'}}
        assert results['NM_000329.2:c.106_114del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.68446841_68446849del', 'vcf': {'chr': 'chr1', 'ref': 'TGAGCCAGAG', 'pos': '68446840', 'alt': 'T'}}
        assert results['NM_000329.2:c.106_114del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.68912524_68912532del', 'vcf': {'chr': '1', 'ref': 'TGAGCCAGAG', 'pos': '68912523', 'alt': 'T'}}
        assert results['NM_000329.2:c.106_114del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.68446841_68446849del', 'vcf': {'chr': '1', 'ref': 'TGAGCCAGAG', 'pos': '68446840', 'alt': 'T'}}

        assert results['flag'] == 'gene_variant'

    def test_variant203(self):
        variant = '1-68912526-GCCAGAG-G'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_000329.2:c.109_114del' in results.keys()
        assert results['NM_000329.2:c.109_114del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000329.2:c.109_114del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000329.2:c.109_114del']['alt_genomic_loci'] == []
        assert results['NM_000329.2:c.109_114del']['gene_symbol'] == 'RPE65'
        assert results['NM_000329.2:c.109_114del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000320.1:p.(Trp37_Leu38del)', 'slr': 'NP_000320.1:p.(W37_L38del)'}
        assert results['NM_000329.2:c.109_114del']['submitted_variant'] == '1-68912526-GCCAGAG-G'
        assert results['NM_000329.2:c.109_114del']['genome_context_intronic_sequence'] == ''
        assert results['NM_000329.2:c.109_114del']['hgvs_lrg_variant'] == ''
        assert results['NM_000329.2:c.109_114del']['hgvs_transcript_variant'] == 'NM_000329.2:c.109_114del'
        assert results['NM_000329.2:c.109_114del']['hgvs_refseqgene_variant'] == 'NG_008472.1:g.8114_8119del'
        assert results['NM_000329.2:c.109_114del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.68912524_68912529del', 'vcf': {'chr': 'chr1', 'ref': 'TGAGCCA', 'pos': '68912523', 'alt': 'T'}}
        assert results['NM_000329.2:c.109_114del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.68446841_68446846del', 'vcf': {'chr': 'chr1', 'ref': 'TGAGCCA', 'pos': '68446840', 'alt': 'T'}}
        assert results['NM_000329.2:c.109_114del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.68912524_68912529del', 'vcf': {'chr': '1', 'ref': 'TGAGCCA', 'pos': '68912523', 'alt': 'T'}}
        assert results['NM_000329.2:c.109_114del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.68446841_68446846del', 'vcf': {'chr': '1', 'ref': 'TGAGCCA', 'pos': '68446840', 'alt': 'T'}}

        assert results['flag'] == 'gene_variant'

    def test_variant204(self):
        variant = '1-109817590-G-T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_001408.2:c.*919G>T' in results.keys()
        assert results['NM_001408.2:c.*919G>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001408.2:c.*919G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001408.2:c.*919G>T']['alt_genomic_loci'] == []
        assert results['NM_001408.2:c.*919G>T']['gene_symbol'] == 'CELSR2'
        assert results['NM_001408.2:c.*919G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001399.1:p.?', 'slr': 'NP_001399.1:p.?'}
        assert results['NM_001408.2:c.*919G>T']['submitted_variant'] == '1-109817590-G-T'
        assert results['NM_001408.2:c.*919G>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001408.2:c.*919G>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001408.2:c.*919G>T']['hgvs_transcript_variant'] == 'NM_001408.2:c.*919G>T'
        assert results['NM_001408.2:c.*919G>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001408.2:c.*919G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.109817590G>T', 'vcf': {'chr': 'chr1', 'ref': 'G', 'pos': '109817590', 'alt': 'T'}}
        assert results['NM_001408.2:c.*919G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.109274968G>T', 'vcf': {'chr': 'chr1', 'ref': 'G', 'pos': '109274968', 'alt': 'T'}}
        assert results['NM_001408.2:c.*919G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.109817590G>T', 'vcf': {'chr': '1', 'ref': 'G', 'pos': '109817590', 'alt': 'T'}}
        assert results['NM_001408.2:c.*919G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.109274968G>T', 'vcf': {'chr': '1', 'ref': 'G', 'pos': '109274968', 'alt': 'T'}}


    def test_variant205(self):
        variant = '1-145597475-GAAGT-G'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_006468.6:c.1070+35_1070+38del' in results.keys()
        assert results['NM_006468.6:c.1070+35_1070+38del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_006468.6:c.1070+35_1070+38del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_006468.6:c.1070+35_1070+38del']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003871055.3:g.2653044_2653047del', 'vcf': {'chr': 'HG1287_PATCH', 'ref': 'ATACT', 'pos': '2653042', 'alt': 'A'}}}]
        assert results['NM_006468.6:c.1070+35_1070+38del']['gene_symbol'] == 'POLR3C'
        assert results['NM_006468.6:c.1070+35_1070+38del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_006459.3:p.?', 'slr': 'NP_006459.3:p.?'}
        assert results['NM_006468.6:c.1070+35_1070+38del']['submitted_variant'] == '1-145597475-GAAGT-G'
        assert results['NM_006468.6:c.1070+35_1070+38del']['genome_context_intronic_sequence'] == 'NC_000001.10(NM_006468.6):c.1070+35_1070+38del'
        assert results['NM_006468.6:c.1070+35_1070+38del']['hgvs_lrg_variant'] == ''
        assert results['NM_006468.6:c.1070+35_1070+38del']['hgvs_transcript_variant'] == 'NM_006468.6:c.1070+35_1070+38del'
        assert results['NM_006468.6:c.1070+35_1070+38del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_006468.6:c.1070+35_1070+38del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.145597477_145597480del', 'vcf': {'chr': 'chr1', 'ref': 'GAAGT', 'pos': '145597475', 'alt': 'G'}}
        assert 'hg38' not in results['NM_006468.6:c.1070+35_1070+38del']['primary_assembly_loci'].keys()
        assert results['NM_006468.6:c.1070+35_1070+38del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.145597477_145597480del', 'vcf': {'chr': '1', 'ref': 'GAAGT', 'pos': '145597475', 'alt': 'G'}}
        assert 'grch38' not in results['NM_006468.6:c.1070+35_1070+38del']['primary_assembly_loci'].keys()

        assert 'NM_001303456.1:c.1109+35_1109+38del' in results.keys()
        assert results['NM_001303456.1:c.1109+35_1109+38del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001303456.1:c.1109+35_1109+38del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001303456.1:c.1109+35_1109+38del']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003871055.3:g.2653044_2653047del', 'vcf': {'chr': 'HG1287_PATCH', 'ref': 'ATACT', 'pos': '2653042', 'alt': 'A'}}}]
        assert results['NM_001303456.1:c.1109+35_1109+38del']['gene_symbol'] == 'POLR3C'
        assert results['NM_001303456.1:c.1109+35_1109+38del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001290385.1:p.?', 'slr': 'NP_001290385.1:p.?'}
        assert results['NM_001303456.1:c.1109+35_1109+38del']['submitted_variant'] == '1-145597475-GAAGT-G'
        assert results['NM_001303456.1:c.1109+35_1109+38del']['genome_context_intronic_sequence'] == 'NC_000001.10(NM_001303456.1):c.1109+35_1109+38del'
        assert results['NM_001303456.1:c.1109+35_1109+38del']['hgvs_lrg_variant'] == ''
        assert results['NM_001303456.1:c.1109+35_1109+38del']['hgvs_transcript_variant'] == 'NM_001303456.1:c.1109+35_1109+38del'
        assert results['NM_001303456.1:c.1109+35_1109+38del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001303456.1:c.1109+35_1109+38del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.145597477_145597480del', 'vcf': {'chr': 'chr1', 'ref': 'GAAGT', 'pos': '145597475', 'alt': 'G'}}
        assert results['NM_001303456.1:c.1109+35_1109+38del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.145837631_145837634del', 'vcf': {'chr': 'chr1', 'ref': 'ATACT', 'pos': '145837629', 'alt': 'A'}}
        assert results['NM_001303456.1:c.1109+35_1109+38del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.145597477_145597480del', 'vcf': {'chr': '1', 'ref': 'GAAGT', 'pos': '145597475', 'alt': 'G'}}
        assert results['NM_001303456.1:c.1109+35_1109+38del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.145837631_145837634del', 'vcf': {'chr': '1', 'ref': 'ATACT', 'pos': '145837629', 'alt': 'A'}}

        assert 'NM_006468.7:c.1070+35_1070+38del' in results.keys()
        assert results['NM_006468.7:c.1070+35_1070+38del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_006468.7:c.1070+35_1070+38del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_006468.7:c.1070+35_1070+38del']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003871055.3:g.2653044_2653047del', 'vcf': {'chr': 'HG1287_PATCH', 'ref': 'ATACT', 'pos': '2653042', 'alt': 'A'}}}]
        assert results['NM_006468.7:c.1070+35_1070+38del']['gene_symbol'] == 'POLR3C'
        assert results['NM_006468.7:c.1070+35_1070+38del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_006459.3:p.?', 'slr': 'NP_006459.3:p.?'}
        assert results['NM_006468.7:c.1070+35_1070+38del']['submitted_variant'] == '1-145597475-GAAGT-G'
        assert results['NM_006468.7:c.1070+35_1070+38del']['genome_context_intronic_sequence'] == 'NC_000001.10(NM_006468.7):c.1070+35_1070+38del'
        assert results['NM_006468.7:c.1070+35_1070+38del']['hgvs_lrg_variant'] == ''
        assert results['NM_006468.7:c.1070+35_1070+38del']['hgvs_transcript_variant'] == 'NM_006468.7:c.1070+35_1070+38del'
        assert results['NM_006468.7:c.1070+35_1070+38del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_006468.7:c.1070+35_1070+38del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.145597477_145597480del', 'vcf': {'chr': 'chr1', 'ref': 'GAAGT', 'pos': '145597475', 'alt': 'G'}}
        assert results['NM_006468.7:c.1070+35_1070+38del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.145837631_145837634del', 'vcf': {'chr': 'chr1', 'ref': 'ATACT', 'pos': '145837629', 'alt': 'A'}}
        assert results['NM_006468.7:c.1070+35_1070+38del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.145597477_145597480del', 'vcf': {'chr': '1', 'ref': 'GAAGT', 'pos': '145597475', 'alt': 'G'}}
        assert results['NM_006468.7:c.1070+35_1070+38del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.145837631_145837634del', 'vcf': {'chr': '1', 'ref': 'ATACT', 'pos': '145837629', 'alt': 'A'}}


    def test_variant206(self):
        variant = '1-153791300-CTG-C'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_020699.2:c.562_563del' in results.keys()
        assert results['NM_020699.2:c.562_563del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_020699.2:c.562_563del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_020699.2:c.562_563del']['alt_genomic_loci'] == []
        assert results['NM_020699.2:c.562_563del']['gene_symbol'] == 'GATAD2B'
        assert results['NM_020699.2:c.562_563del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_065750.1:p.(Gln188GlufsTer36)', 'slr': 'NP_065750.1:p.(Q188Efs*36)'}
        assert results['NM_020699.2:c.562_563del']['submitted_variant'] == '1-153791300-CTG-C'
        assert results['NM_020699.2:c.562_563del']['genome_context_intronic_sequence'] == ''
        assert results['NM_020699.2:c.562_563del']['hgvs_lrg_variant'] == ''
        assert results['NM_020699.2:c.562_563del']['hgvs_transcript_variant'] == 'NM_020699.2:c.562_563del'
        assert results['NM_020699.2:c.562_563del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_020699.2:c.562_563del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.153791301_153791302del', 'vcf': {'chr': 'chr1', 'ref': 'CTG', 'pos': '153791300', 'alt': 'C'}}
        assert results['NM_020699.2:c.562_563del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.153818825_153818826del', 'vcf': {'chr': 'chr1', 'ref': 'CTG', 'pos': '153818824', 'alt': 'C'}}
        assert results['NM_020699.2:c.562_563del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.153791301_153791302del', 'vcf': {'chr': '1', 'ref': 'CTG', 'pos': '153791300', 'alt': 'C'}}
        assert results['NM_020699.2:c.562_563del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.153818825_153818826del', 'vcf': {'chr': '1', 'ref': 'CTG', 'pos': '153818824', 'alt': 'C'}}

        assert 'NM_020699.3:c.562_563del' in results.keys()
        assert results['NM_020699.3:c.562_563del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_020699.3:c.562_563del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_020699.3:c.562_563del']['alt_genomic_loci'] == []
        assert results['NM_020699.3:c.562_563del']['gene_symbol'] == 'GATAD2B'
        assert results['NM_020699.3:c.562_563del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_065750.1:p.(Gln188GlufsTer36)', 'slr': 'NP_065750.1:p.(Q188Efs*36)'}
        assert results['NM_020699.3:c.562_563del']['submitted_variant'] == '1-153791300-CTG-C'
        assert results['NM_020699.3:c.562_563del']['genome_context_intronic_sequence'] == ''
        assert results['NM_020699.3:c.562_563del']['hgvs_lrg_variant'] == ''
        assert results['NM_020699.3:c.562_563del']['hgvs_transcript_variant'] == 'NM_020699.3:c.562_563del'
        assert results['NM_020699.3:c.562_563del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_020699.3:c.562_563del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.153791301_153791302del', 'vcf': {'chr': 'chr1', 'ref': 'CTG', 'pos': '153791300', 'alt': 'C'}}
        assert results['NM_020699.3:c.562_563del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.153818825_153818826del', 'vcf': {'chr': 'chr1', 'ref': 'CTG', 'pos': '153818824', 'alt': 'C'}}
        assert results['NM_020699.3:c.562_563del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.153791301_153791302del', 'vcf': {'chr': '1', 'ref': 'CTG', 'pos': '153791300', 'alt': 'C'}}
        assert results['NM_020699.3:c.562_563del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.153818825_153818826del', 'vcf': {'chr': '1', 'ref': 'CTG', 'pos': '153818824', 'alt': 'C'}}


    def test_variant207(self):
        variant = '1-156104666-TTGAGAGCCGGCTGGCGGATGCGCT-TCCCC'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_005572.3:c.711_734delinsCCCC' in results.keys()
        assert results['NM_005572.3:c.711_734delinsCCCC']['hgvs_lrg_transcript_variant'] == 'LRG_254t1:c.711_734delinsCCCC'
        assert results['NM_005572.3:c.711_734delinsCCCC']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_005572.3:c.711_734delinsCCCC']['alt_genomic_loci'] == []
        assert results['NM_005572.3:c.711_734delinsCCCC']['gene_symbol'] == 'LMNA'
        assert results['NM_005572.3:c.711_734delinsCCCC']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_005563.1(LRG_254p1):p.(Glu238ProfsTer9)', 'slr': 'NP_005563.1:p.(E238Pfs*9)'}
        assert results['NM_005572.3:c.711_734delinsCCCC']['submitted_variant'] == '1-156104666-TTGAGAGCCGGCTGGCGGATGCGCT-TCCCC'
        assert results['NM_005572.3:c.711_734delinsCCCC']['genome_context_intronic_sequence'] == ''
        assert results['NM_005572.3:c.711_734delinsCCCC']['hgvs_lrg_variant'] == 'LRG_254:g.57304_57327delinsCCCC'
        assert results['NM_005572.3:c.711_734delinsCCCC']['hgvs_transcript_variant'] == 'NM_005572.3:c.711_734delinsCCCC'
        assert results['NM_005572.3:c.711_734delinsCCCC']['hgvs_refseqgene_variant'] == 'NG_008692.2:g.57304_57327delinsCCCC'
        assert results['NM_005572.3:c.711_734delinsCCCC']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.156104667_156104690delinsCCCC', 'vcf': {'chr': 'chr1', 'ref': 'TGAGAGCCGGCTGGCGGATGCGCT', 'pos': '156104667', 'alt': 'CCCC'}}
        assert results['NM_005572.3:c.711_734delinsCCCC']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.156134876_156134899delinsCCCC', 'vcf': {'chr': 'chr1', 'ref': 'TGAGAGCCGGCTGGCGGATGCGCT', 'pos': '156134876', 'alt': 'CCCC'}}
        assert results['NM_005572.3:c.711_734delinsCCCC']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.156104667_156104690delinsCCCC', 'vcf': {'chr': '1', 'ref': 'TGAGAGCCGGCTGGCGGATGCGCT', 'pos': '156104667', 'alt': 'CCCC'}}
        assert results['NM_005572.3:c.711_734delinsCCCC']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.156134876_156134899delinsCCCC', 'vcf': {'chr': '1', 'ref': 'TGAGAGCCGGCTGGCGGATGCGCT', 'pos': '156134876', 'alt': 'CCCC'}}

        assert 'NM_001257374.1:c.375_398delinsCCCC' in results.keys()
        assert results['NM_001257374.1:c.375_398delinsCCCC']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001257374.1:c.375_398delinsCCCC']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001257374.1:c.375_398delinsCCCC']['alt_genomic_loci'] == []
        assert results['NM_001257374.1:c.375_398delinsCCCC']['gene_symbol'] == 'LMNA'
        assert results['NM_001257374.1:c.375_398delinsCCCC']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001244303.1:p.(Glu126ProfsTer9)', 'slr': 'NP_001244303.1:p.(E126Pfs*9)'}
        assert results['NM_001257374.1:c.375_398delinsCCCC']['submitted_variant'] == '1-156104666-TTGAGAGCCGGCTGGCGGATGCGCT-TCCCC'
        assert results['NM_001257374.1:c.375_398delinsCCCC']['genome_context_intronic_sequence'] == ''
        assert results['NM_001257374.1:c.375_398delinsCCCC']['hgvs_lrg_variant'] == ''
        assert results['NM_001257374.1:c.375_398delinsCCCC']['hgvs_transcript_variant'] == 'NM_001257374.1:c.375_398delinsCCCC'
        assert results['NM_001257374.1:c.375_398delinsCCCC']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001257374.1:c.375_398delinsCCCC']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.156104667_156104690delinsCCCC', 'vcf': {'chr': 'chr1', 'ref': 'TGAGAGCCGGCTGGCGGATGCGCT', 'pos': '156104667', 'alt': 'CCCC'}}
        assert 'hg38' not in results['NM_001257374.1:c.375_398delinsCCCC']['primary_assembly_loci'].keys()
        assert results['NM_001257374.1:c.375_398delinsCCCC']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.156104667_156104690delinsCCCC', 'vcf': {'chr': '1', 'ref': 'TGAGAGCCGGCTGGCGGATGCGCT', 'pos': '156104667', 'alt': 'CCCC'}}
        assert 'grch38' not in results['NM_001257374.1:c.375_398delinsCCCC']['primary_assembly_loci'].keys()

        assert 'NM_001257374.2:c.375_398delinsCCCC' in results.keys()
        assert results['NM_001257374.2:c.375_398delinsCCCC']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001257374.2:c.375_398delinsCCCC']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001257374.2:c.375_398delinsCCCC']['alt_genomic_loci'] == []
        assert results['NM_001257374.2:c.375_398delinsCCCC']['gene_symbol'] == 'LMNA'
        assert results['NM_001257374.2:c.375_398delinsCCCC']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001244303.1:p.(Glu126ProfsTer9)', 'slr': 'NP_001244303.1:p.(E126Pfs*9)'}
        assert results['NM_001257374.2:c.375_398delinsCCCC']['submitted_variant'] == '1-156104666-TTGAGAGCCGGCTGGCGGATGCGCT-TCCCC'
        assert results['NM_001257374.2:c.375_398delinsCCCC']['genome_context_intronic_sequence'] == ''
        assert results['NM_001257374.2:c.375_398delinsCCCC']['hgvs_lrg_variant'] == ''
        assert results['NM_001257374.2:c.375_398delinsCCCC']['hgvs_transcript_variant'] == 'NM_001257374.2:c.375_398delinsCCCC'
        assert results['NM_001257374.2:c.375_398delinsCCCC']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001257374.2:c.375_398delinsCCCC']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.156104667_156104690delinsCCCC', 'vcf': {'chr': 'chr1', 'ref': 'TGAGAGCCGGCTGGCGGATGCGCT', 'pos': '156104667', 'alt': 'CCCC'}}
        assert results['NM_001257374.2:c.375_398delinsCCCC']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.156134876_156134899delinsCCCC', 'vcf': {'chr': 'chr1', 'ref': 'TGAGAGCCGGCTGGCGGATGCGCT', 'pos': '156134876', 'alt': 'CCCC'}}
        assert results['NM_001257374.2:c.375_398delinsCCCC']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.156104667_156104690delinsCCCC', 'vcf': {'chr': '1', 'ref': 'TGAGAGCCGGCTGGCGGATGCGCT', 'pos': '156104667', 'alt': 'CCCC'}}
        assert results['NM_001257374.2:c.375_398delinsCCCC']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.156134876_156134899delinsCCCC', 'vcf': {'chr': '1', 'ref': 'TGAGAGCCGGCTGGCGGATGCGCT', 'pos': '156134876', 'alt': 'CCCC'}}

        assert 'NM_001282624.1:c.468_491delinsCCCC' in results.keys()
        assert results['NM_001282624.1:c.468_491delinsCCCC']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001282624.1:c.468_491delinsCCCC']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001282624.1:c.468_491delinsCCCC']['alt_genomic_loci'] == []
        assert results['NM_001282624.1:c.468_491delinsCCCC']['gene_symbol'] == 'LMNA'
        assert results['NM_001282624.1:c.468_491delinsCCCC']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001269553.1:p.(Glu157ProfsTer9)', 'slr': 'NP_001269553.1:p.(E157Pfs*9)'}
        assert results['NM_001282624.1:c.468_491delinsCCCC']['submitted_variant'] == '1-156104666-TTGAGAGCCGGCTGGCGGATGCGCT-TCCCC'
        assert results['NM_001282624.1:c.468_491delinsCCCC']['genome_context_intronic_sequence'] == ''
        assert results['NM_001282624.1:c.468_491delinsCCCC']['hgvs_lrg_variant'] == ''
        assert results['NM_001282624.1:c.468_491delinsCCCC']['hgvs_transcript_variant'] == 'NM_001282624.1:c.468_491delinsCCCC'
        assert results['NM_001282624.1:c.468_491delinsCCCC']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001282624.1:c.468_491delinsCCCC']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.156104667_156104690delinsCCCC', 'vcf': {'chr': 'chr1', 'ref': 'TGAGAGCCGGCTGGCGGATGCGCT', 'pos': '156104667', 'alt': 'CCCC'}}
        assert results['NM_001282624.1:c.468_491delinsCCCC']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.156134876_156134899delinsCCCC', 'vcf': {'chr': 'chr1', 'ref': 'TGAGAGCCGGCTGGCGGATGCGCT', 'pos': '156134876', 'alt': 'CCCC'}}
        assert results['NM_001282624.1:c.468_491delinsCCCC']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.156104667_156104690delinsCCCC', 'vcf': {'chr': '1', 'ref': 'TGAGAGCCGGCTGGCGGATGCGCT', 'pos': '156104667', 'alt': 'CCCC'}}
        assert results['NM_001282624.1:c.468_491delinsCCCC']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.156134876_156134899delinsCCCC', 'vcf': {'chr': '1', 'ref': 'TGAGAGCCGGCTGGCGGATGCGCT', 'pos': '156134876', 'alt': 'CCCC'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_170708.3:c.711_734delinsCCCC' in results.keys()
        assert results['NM_170708.3:c.711_734delinsCCCC']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_170708.3:c.711_734delinsCCCC']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_170708.3:c.711_734delinsCCCC']['alt_genomic_loci'] == []
        assert results['NM_170708.3:c.711_734delinsCCCC']['gene_symbol'] == 'LMNA'
        assert results['NM_170708.3:c.711_734delinsCCCC']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_733822.1(LRG_254p3):p.(Glu238ProfsTer9)', 'slr': 'NP_733822.1:p.(E238Pfs*9)'}
        assert results['NM_170708.3:c.711_734delinsCCCC']['submitted_variant'] == '1-156104666-TTGAGAGCCGGCTGGCGGATGCGCT-TCCCC'
        assert results['NM_170708.3:c.711_734delinsCCCC']['genome_context_intronic_sequence'] == ''
        assert results['NM_170708.3:c.711_734delinsCCCC']['hgvs_lrg_variant'] == ''
        assert results['NM_170708.3:c.711_734delinsCCCC']['hgvs_transcript_variant'] == 'NM_170708.3:c.711_734delinsCCCC'
        assert results['NM_170708.3:c.711_734delinsCCCC']['hgvs_refseqgene_variant'] == ''
        assert results['NM_170708.3:c.711_734delinsCCCC']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.156104667_156104690delinsCCCC', 'vcf': {'chr': 'chr1', 'ref': 'TGAGAGCCGGCTGGCGGATGCGCT', 'pos': '156104667', 'alt': 'CCCC'}}
        assert results['NM_170708.3:c.711_734delinsCCCC']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.156134876_156134899delinsCCCC', 'vcf': {'chr': 'chr1', 'ref': 'TGAGAGCCGGCTGGCGGATGCGCT', 'pos': '156134876', 'alt': 'CCCC'}}
        assert results['NM_170708.3:c.711_734delinsCCCC']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.156104667_156104690delinsCCCC', 'vcf': {'chr': '1', 'ref': 'TGAGAGCCGGCTGGCGGATGCGCT', 'pos': '156104667', 'alt': 'CCCC'}}
        assert results['NM_170708.3:c.711_734delinsCCCC']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.156134876_156134899delinsCCCC', 'vcf': {'chr': '1', 'ref': 'TGAGAGCCGGCTGGCGGATGCGCT', 'pos': '156134876', 'alt': 'CCCC'}}

        assert 'NM_170707.3:c.711_734delinsCCCC' in results.keys()
        assert results['NM_170707.3:c.711_734delinsCCCC']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_170707.3:c.711_734delinsCCCC']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_170707.3:c.711_734delinsCCCC']['alt_genomic_loci'] == []
        assert results['NM_170707.3:c.711_734delinsCCCC']['gene_symbol'] == 'LMNA'
        assert results['NM_170707.3:c.711_734delinsCCCC']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_733821.1(LRG_254p2):p.(Glu238ProfsTer9)', 'slr': 'NP_733821.1:p.(E238Pfs*9)'}
        assert results['NM_170707.3:c.711_734delinsCCCC']['submitted_variant'] == '1-156104666-TTGAGAGCCGGCTGGCGGATGCGCT-TCCCC'
        assert results['NM_170707.3:c.711_734delinsCCCC']['genome_context_intronic_sequence'] == ''
        assert results['NM_170707.3:c.711_734delinsCCCC']['hgvs_lrg_variant'] == ''
        assert results['NM_170707.3:c.711_734delinsCCCC']['hgvs_transcript_variant'] == 'NM_170707.3:c.711_734delinsCCCC'
        assert results['NM_170707.3:c.711_734delinsCCCC']['hgvs_refseqgene_variant'] == ''
        assert results['NM_170707.3:c.711_734delinsCCCC']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.156104667_156104690delinsCCCC', 'vcf': {'chr': 'chr1', 'ref': 'TGAGAGCCGGCTGGCGGATGCGCT', 'pos': '156104667', 'alt': 'CCCC'}}
        assert results['NM_170707.3:c.711_734delinsCCCC']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.156134876_156134899delinsCCCC', 'vcf': {'chr': 'chr1', 'ref': 'TGAGAGCCGGCTGGCGGATGCGCT', 'pos': '156134876', 'alt': 'CCCC'}}
        assert results['NM_170707.3:c.711_734delinsCCCC']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.156104667_156104690delinsCCCC', 'vcf': {'chr': '1', 'ref': 'TGAGAGCCGGCTGGCGGATGCGCT', 'pos': '156104667', 'alt': 'CCCC'}}
        assert results['NM_170707.3:c.711_734delinsCCCC']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.156134876_156134899delinsCCCC', 'vcf': {'chr': '1', 'ref': 'TGAGAGCCGGCTGGCGGATGCGCT', 'pos': '156134876', 'alt': 'CCCC'}}

        assert 'NM_001282626.1:c.711_734delinsCCCC' in results.keys()
        assert results['NM_001282626.1:c.711_734delinsCCCC']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001282626.1:c.711_734delinsCCCC']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001282626.1:c.711_734delinsCCCC']['alt_genomic_loci'] == []
        assert results['NM_001282626.1:c.711_734delinsCCCC']['gene_symbol'] == 'LMNA'
        assert results['NM_001282626.1:c.711_734delinsCCCC']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001269555.1:p.(Glu238ProfsTer9)', 'slr': 'NP_001269555.1:p.(E238Pfs*9)'}
        assert results['NM_001282626.1:c.711_734delinsCCCC']['submitted_variant'] == '1-156104666-TTGAGAGCCGGCTGGCGGATGCGCT-TCCCC'
        assert results['NM_001282626.1:c.711_734delinsCCCC']['genome_context_intronic_sequence'] == ''
        assert results['NM_001282626.1:c.711_734delinsCCCC']['hgvs_lrg_variant'] == ''
        assert results['NM_001282626.1:c.711_734delinsCCCC']['hgvs_transcript_variant'] == 'NM_001282626.1:c.711_734delinsCCCC'
        assert results['NM_001282626.1:c.711_734delinsCCCC']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001282626.1:c.711_734delinsCCCC']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.156104667_156104690delinsCCCC', 'vcf': {'chr': 'chr1', 'ref': 'TGAGAGCCGGCTGGCGGATGCGCT', 'pos': '156104667', 'alt': 'CCCC'}}
        assert results['NM_001282626.1:c.711_734delinsCCCC']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.156134876_156134899delinsCCCC', 'vcf': {'chr': 'chr1', 'ref': 'TGAGAGCCGGCTGGCGGATGCGCT', 'pos': '156134876', 'alt': 'CCCC'}}
        assert results['NM_001282626.1:c.711_734delinsCCCC']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.156104667_156104690delinsCCCC', 'vcf': {'chr': '1', 'ref': 'TGAGAGCCGGCTGGCGGATGCGCT', 'pos': '156104667', 'alt': 'CCCC'}}
        assert results['NM_001282626.1:c.711_734delinsCCCC']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.156134876_156134899delinsCCCC', 'vcf': {'chr': '1', 'ref': 'TGAGAGCCGGCTGGCGGATGCGCT', 'pos': '156134876', 'alt': 'CCCC'}}

        assert 'NM_001282625.1:c.711_734delinsCCCC' in results.keys()
        assert results['NM_001282625.1:c.711_734delinsCCCC']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001282625.1:c.711_734delinsCCCC']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001282625.1:c.711_734delinsCCCC']['alt_genomic_loci'] == []
        assert results['NM_001282625.1:c.711_734delinsCCCC']['gene_symbol'] == 'LMNA'
        assert results['NM_001282625.1:c.711_734delinsCCCC']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001269554.1:p.(Glu238ProfsTer9)', 'slr': 'NP_001269554.1:p.(E238Pfs*9)'}
        assert results['NM_001282625.1:c.711_734delinsCCCC']['submitted_variant'] == '1-156104666-TTGAGAGCCGGCTGGCGGATGCGCT-TCCCC'
        assert results['NM_001282625.1:c.711_734delinsCCCC']['genome_context_intronic_sequence'] == ''
        assert results['NM_001282625.1:c.711_734delinsCCCC']['hgvs_lrg_variant'] == ''
        assert results['NM_001282625.1:c.711_734delinsCCCC']['hgvs_transcript_variant'] == 'NM_001282625.1:c.711_734delinsCCCC'
        assert results['NM_001282625.1:c.711_734delinsCCCC']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001282625.1:c.711_734delinsCCCC']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.156104667_156104690delinsCCCC', 'vcf': {'chr': 'chr1', 'ref': 'TGAGAGCCGGCTGGCGGATGCGCT', 'pos': '156104667', 'alt': 'CCCC'}}
        assert results['NM_001282625.1:c.711_734delinsCCCC']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.156134876_156134899delinsCCCC', 'vcf': {'chr': 'chr1', 'ref': 'TGAGAGCCGGCTGGCGGATGCGCT', 'pos': '156134876', 'alt': 'CCCC'}}
        assert results['NM_001282625.1:c.711_734delinsCCCC']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.156104667_156104690delinsCCCC', 'vcf': {'chr': '1', 'ref': 'TGAGAGCCGGCTGGCGGATGCGCT', 'pos': '156104667', 'alt': 'CCCC'}}
        assert results['NM_001282625.1:c.711_734delinsCCCC']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.156134876_156134899delinsCCCC', 'vcf': {'chr': '1', 'ref': 'TGAGAGCCGGCTGGCGGATGCGCT', 'pos': '156134876', 'alt': 'CCCC'}}


    def test_variant208(self):
        variant = '1-156108541-G-GG'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_170707.3:c.1961dup' in results.keys()
        assert results['NM_170707.3:c.1961dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_170707.3:c.1961dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_170707.3:c.1961dup']['alt_genomic_loci'] == []
        assert results['NM_170707.3:c.1961dup']['gene_symbol'] == 'LMNA'
        assert results['NM_170707.3:c.1961dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_733821.1(LRG_254p2):p.(Thr655AsnfsTer49)', 'slr': 'NP_733821.1:p.(T655Nfs*49)'}
        assert results['NM_170707.3:c.1961dup']['submitted_variant'] == '1-156108541-G-GG'
        assert results['NM_170707.3:c.1961dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_170707.3:c.1961dup']['hgvs_lrg_variant'] == ''
        assert results['NM_170707.3:c.1961dup']['hgvs_transcript_variant'] == 'NM_170707.3:c.1961dup'
        assert results['NM_170707.3:c.1961dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_170707.3:c.1961dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.156108541dup', 'vcf': {'chr': 'chr1', 'ref': 'G', 'pos': '156108541', 'alt': 'GG'}}
        assert results['NM_170707.3:c.1961dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.156138750dup', 'vcf': {'chr': 'chr1', 'ref': 'G', 'pos': '156138750', 'alt': 'GG'}}
        assert results['NM_170707.3:c.1961dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.156108541dup', 'vcf': {'chr': '1', 'ref': 'G', 'pos': '156108541', 'alt': 'GG'}}
        assert results['NM_170707.3:c.1961dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.156138750dup', 'vcf': {'chr': '1', 'ref': 'G', 'pos': '156138750', 'alt': 'GG'}}

        assert 'NM_001282626.1:c.1818+143dup' in results.keys()
        assert results['NM_001282626.1:c.1818+143dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001282626.1:c.1818+143dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001282626.1:c.1818+143dup']['alt_genomic_loci'] == []
        assert results['NM_001282626.1:c.1818+143dup']['gene_symbol'] == 'LMNA'
        assert results['NM_001282626.1:c.1818+143dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001269555.1:p.?', 'slr': 'NP_001269555.1:p.?'}
        assert results['NM_001282626.1:c.1818+143dup']['submitted_variant'] == '1-156108541-G-GG'
        assert results['NM_001282626.1:c.1818+143dup']['genome_context_intronic_sequence'] == 'NC_000001.10(NM_001282626.1):c.1818+143dup'
        assert results['NM_001282626.1:c.1818+143dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001282626.1:c.1818+143dup']['hgvs_transcript_variant'] == 'NM_001282626.1:c.1818+143dup'
        assert results['NM_001282626.1:c.1818+143dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001282626.1:c.1818+143dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.156108541dup', 'vcf': {'chr': 'chr1', 'ref': 'G', 'pos': '156108541', 'alt': 'GG'}}
        assert results['NM_001282626.1:c.1818+143dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.156138750dup', 'vcf': {'chr': 'chr1', 'ref': 'G', 'pos': '156138750', 'alt': 'GG'}}
        assert results['NM_001282626.1:c.1818+143dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.156108541dup', 'vcf': {'chr': '1', 'ref': 'G', 'pos': '156108541', 'alt': 'GG'}}
        assert results['NM_001282626.1:c.1818+143dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.156138750dup', 'vcf': {'chr': '1', 'ref': 'G', 'pos': '156138750', 'alt': 'GG'}}

        assert 'NM_001257374.1:c.1625dup' in results.keys()
        assert results['NM_001257374.1:c.1625dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001257374.1:c.1625dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001257374.1:c.1625dup']['alt_genomic_loci'] == []
        assert results['NM_001257374.1:c.1625dup']['gene_symbol'] == 'LMNA'
        assert results['NM_001257374.1:c.1625dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001244303.1:p.(Thr543AsnfsTer90)', 'slr': 'NP_001244303.1:p.(T543Nfs*90)'}
        assert results['NM_001257374.1:c.1625dup']['submitted_variant'] == '1-156108541-G-GG'
        assert results['NM_001257374.1:c.1625dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_001257374.1:c.1625dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001257374.1:c.1625dup']['hgvs_transcript_variant'] == 'NM_001257374.1:c.1625dup'
        assert results['NM_001257374.1:c.1625dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001257374.1:c.1625dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.156108541dup', 'vcf': {'chr': 'chr1', 'ref': 'G', 'pos': '156108541', 'alt': 'GG'}}
        assert 'hg38' not in results['NM_001257374.1:c.1625dup']['primary_assembly_loci'].keys()
        assert results['NM_001257374.1:c.1625dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.156108541dup', 'vcf': {'chr': '1', 'ref': 'G', 'pos': '156108541', 'alt': 'GG'}}
        assert 'grch38' not in results['NM_001257374.1:c.1625dup']['primary_assembly_loci'].keys()

        assert 'NM_001257374.2:c.1625dup' in results.keys()
        assert results['NM_001257374.2:c.1625dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001257374.2:c.1625dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001257374.2:c.1625dup']['alt_genomic_loci'] == []
        assert results['NM_001257374.2:c.1625dup']['gene_symbol'] == 'LMNA'
        assert results['NM_001257374.2:c.1625dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001244303.1:p.(Thr543AsnfsTer90)', 'slr': 'NP_001244303.1:p.(T543Nfs*90)'}
        assert results['NM_001257374.2:c.1625dup']['submitted_variant'] == '1-156108541-G-GG'
        assert results['NM_001257374.2:c.1625dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_001257374.2:c.1625dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001257374.2:c.1625dup']['hgvs_transcript_variant'] == 'NM_001257374.2:c.1625dup'
        assert results['NM_001257374.2:c.1625dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001257374.2:c.1625dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.156108541dup', 'vcf': {'chr': 'chr1', 'ref': 'G', 'pos': '156108541', 'alt': 'GG'}}
        assert results['NM_001257374.2:c.1625dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.156138750dup', 'vcf': {'chr': 'chr1', 'ref': 'G', 'pos': '156138750', 'alt': 'GG'}}
        assert results['NM_001257374.2:c.1625dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.156108541dup', 'vcf': {'chr': '1', 'ref': 'G', 'pos': '156108541', 'alt': 'GG'}}
        assert results['NM_001257374.2:c.1625dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.156138750dup', 'vcf': {'chr': '1', 'ref': 'G', 'pos': '156138750', 'alt': 'GG'}}

        assert 'NM_170708.3:c.1871dup' in results.keys()
        assert results['NM_170708.3:c.1871dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_170708.3:c.1871dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_170708.3:c.1871dup']['alt_genomic_loci'] == []
        assert results['NM_170708.3:c.1871dup']['gene_symbol'] == 'LMNA'
        assert results['NM_170708.3:c.1871dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_733822.1(LRG_254p3):p.(Thr625AsnfsTer49)', 'slr': 'NP_733822.1:p.(T625Nfs*49)'}
        assert results['NM_170708.3:c.1871dup']['submitted_variant'] == '1-156108541-G-GG'
        assert results['NM_170708.3:c.1871dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_170708.3:c.1871dup']['hgvs_lrg_variant'] == ''
        assert results['NM_170708.3:c.1871dup']['hgvs_transcript_variant'] == 'NM_170708.3:c.1871dup'
        assert results['NM_170708.3:c.1871dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_170708.3:c.1871dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.156108541dup', 'vcf': {'chr': 'chr1', 'ref': 'G', 'pos': '156108541', 'alt': 'GG'}}
        assert results['NM_170708.3:c.1871dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.156138750dup', 'vcf': {'chr': 'chr1', 'ref': 'G', 'pos': '156138750', 'alt': 'GG'}}
        assert results['NM_170708.3:c.1871dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.156108541dup', 'vcf': {'chr': '1', 'ref': 'G', 'pos': '156108541', 'alt': 'GG'}}
        assert results['NM_170708.3:c.1871dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.156138750dup', 'vcf': {'chr': '1', 'ref': 'G', 'pos': '156138750', 'alt': 'GG'}}

        assert results['flag'] == 'gene_variant'

    def test_variant209(self):
        variant = '1-161279695-T-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_001315491.1:c.1A>T' in results.keys()
        assert results['NM_001315491.1:c.1A>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001315491.1:c.1A>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001315491.1:c.1A>T']['alt_genomic_loci'] == []
        assert results['NM_001315491.1:c.1A>T']['gene_symbol'] == 'MPZ'
        assert results['NM_001315491.1:c.1A>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001302420.1:p.(Met1?)', 'slr': 'NP_001302420.1:p.(M1?)'}
        assert results['NM_001315491.1:c.1A>T']['submitted_variant'] == '1-161279695-T-A'
        assert results['NM_001315491.1:c.1A>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001315491.1:c.1A>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001315491.1:c.1A>T']['hgvs_transcript_variant'] == 'NM_001315491.1:c.1A>T'
        assert results['NM_001315491.1:c.1A>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001315491.1:c.1A>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.161279695T>A', 'vcf': {'chr': 'chr1', 'ref': u'T', 'pos': '161279695', 'alt': u'A'}}
        assert results['NM_001315491.1:c.1A>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.161309905T>A', 'vcf': {'chr': 'chr1', 'ref': u'T', 'pos': '161309905', 'alt': u'A'}}
        assert results['NM_001315491.1:c.1A>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.161279695T>A', 'vcf': {'chr': '1', 'ref': u'T', 'pos': '161279695', 'alt': u'A'}}
        assert results['NM_001315491.1:c.1A>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.161309905T>A', 'vcf': {'chr': '1', 'ref': u'T', 'pos': '161309905', 'alt': u'A'}}

        assert 'NM_000530.7:c.1A>T' in results.keys()
        assert results['NM_000530.7:c.1A>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000530.7:c.1A>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000530.7:c.1A>T']['alt_genomic_loci'] == []
        assert results['NM_000530.7:c.1A>T']['gene_symbol'] == 'MPZ'
        assert results['NM_000530.7:c.1A>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000521.2(LRG_256p1):p.(Met1?)', 'slr': 'NP_000521.2:p.(M1?)'}
        assert results['NM_000530.7:c.1A>T']['submitted_variant'] == '1-161279695-T-A'
        assert results['NM_000530.7:c.1A>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_000530.7:c.1A>T']['hgvs_lrg_variant'] == ''
        assert results['NM_000530.7:c.1A>T']['hgvs_transcript_variant'] == 'NM_000530.7:c.1A>T'
        assert results['NM_000530.7:c.1A>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_000530.7:c.1A>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.161279695T>A', 'vcf': {'chr': 'chr1', 'ref': u'T', 'pos': '161279695', 'alt': u'A'}}
        assert results['NM_000530.7:c.1A>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.161309905T>A', 'vcf': {'chr': 'chr1', 'ref': u'T', 'pos': '161309905', 'alt': u'A'}}
        assert results['NM_000530.7:c.1A>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.161279695T>A', 'vcf': {'chr': '1', 'ref': u'T', 'pos': '161279695', 'alt': u'A'}}
        assert results['NM_000530.7:c.1A>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.161309905T>A', 'vcf': {'chr': '1', 'ref': u'T', 'pos': '161309905', 'alt': u'A'}}

        assert 'NM_000530.6:c.1A>T' in results.keys()
        assert results['NM_000530.6:c.1A>T']['hgvs_lrg_transcript_variant'] == 'LRG_256t1:c.1A>T'
        assert results['NM_000530.6:c.1A>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000530.6:c.1A>T']['alt_genomic_loci'] == []
        assert results['NM_000530.6:c.1A>T']['gene_symbol'] == 'MPZ'
        assert results['NM_000530.6:c.1A>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000521.2(LRG_256p1):p.(Met1?)', 'slr': 'NP_000521.2:p.(M1?)'}
        assert results['NM_000530.6:c.1A>T']['submitted_variant'] == '1-161279695-T-A'
        assert results['NM_000530.6:c.1A>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_000530.6:c.1A>T']['hgvs_lrg_variant'] == 'LRG_256:g.5068A>T'
        assert results['NM_000530.6:c.1A>T']['hgvs_transcript_variant'] == 'NM_000530.6:c.1A>T'
        assert results['NM_000530.6:c.1A>T']['hgvs_refseqgene_variant'] == 'NG_008055.1:g.5068A>T'
        assert results['NM_000530.6:c.1A>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.161279695T>A', 'vcf': {'chr': 'chr1', 'ref': u'T', 'pos': '161279695', 'alt': u'A'}}
        assert 'hg38' not in results['NM_000530.6:c.1A>T']['primary_assembly_loci'].keys()
        assert results['NM_000530.6:c.1A>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.161279695T>A', 'vcf': {'chr': '1', 'ref': u'T', 'pos': '161279695', 'alt': u'A'}}
        assert 'grch38' not in results['NM_000530.6:c.1A>T']['primary_assembly_loci'].keys()


    def test_variant210(self):
        variant = '1-169519049-T-T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000130.4:c.1601G>A' in results.keys()
        assert results['NM_000130.4:c.1601G>A']['hgvs_lrg_transcript_variant'] == 'LRG_553t1:c.1601G>A'
        assert results['NM_000130.4:c.1601G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000130.4:c.1601G>A']['alt_genomic_loci'] == []
        assert results['NM_000130.4:c.1601G>A']['gene_symbol'] == 'F5'
        assert results['NM_000130.4:c.1601G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000121.2(LRG_553p1):p.(Arg534Gln)', 'slr': 'NP_000121.2:p.(R534Q)'}
        assert results['NM_000130.4:c.1601G>A']['submitted_variant'] == '1-169519049-T-T'
        assert results['NM_000130.4:c.1601G>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_000130.4:c.1601G>A']['hgvs_lrg_variant'] == 'LRG_553:g.41721G>A'
        assert results['NM_000130.4:c.1601G>A']['hgvs_transcript_variant'] == 'NM_000130.4:c.1601G>A'
        assert results['NM_000130.4:c.1601G>A']['hgvs_refseqgene_variant'] == 'NG_011806.1:g.41721G>A'
        assert results['NM_000130.4:c.1601G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.169519049T=', 'vcf': {'chr': 'chr1', 'ref': 'T', 'pos': '169519049', 'alt': 'T'}}
        assert results['NM_000130.4:c.1601G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.169549811C>T', 'vcf': {'chr': 'chr1', 'ref': u'C', 'pos': '169549811', 'alt': u'T'}}
        assert results['NM_000130.4:c.1601G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.169519049T=', 'vcf': {'chr': '1', 'ref': 'T', 'pos': '169519049', 'alt': 'T'}}
        assert results['NM_000130.4:c.1601G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.169549811C>T', 'vcf': {'chr': '1', 'ref': u'C', 'pos': '169549811', 'alt': u'T'}}


    def test_variant211(self):
        variant = '1-226125468-G-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_001172425.1:c.672C>T' in results.keys()
        assert results['NM_001172425.1:c.672C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001172425.1:c.672C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001172425.1:c.672C>T']['alt_genomic_loci'] == []
        assert results['NM_001172425.1:c.672C>T']['gene_symbol'] == 'LEFTY2'
        assert results['NM_001172425.1:c.672C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001165896.1:p.(Thr224=)', 'slr': 'NP_001165896.1:p.(T224=)'}
        assert results['NM_001172425.1:c.672C>T']['submitted_variant'] == '1-226125468-G-A'
        assert results['NM_001172425.1:c.672C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001172425.1:c.672C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001172425.1:c.672C>T']['hgvs_transcript_variant'] == 'NM_001172425.1:c.672C>T'
        assert results['NM_001172425.1:c.672C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001172425.1:c.672C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.226125468G>A', 'vcf': {'chr': 'chr1', 'ref': u'G', 'pos': '226125468', 'alt': u'A'}}
        assert results['NM_001172425.1:c.672C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.225937768G>A', 'vcf': {'chr': 'chr1', 'ref': u'G', 'pos': '225937768', 'alt': u'A'}}
        assert results['NM_001172425.1:c.672C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.226125468G>A', 'vcf': {'chr': '1', 'ref': u'G', 'pos': '226125468', 'alt': u'A'}}
        assert results['NM_001172425.1:c.672C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.225937768G>A', 'vcf': {'chr': '1', 'ref': u'G', 'pos': '225937768', 'alt': u'A'}}

        assert 'NM_003240.4:c.774C>T' in results.keys()
        assert results['NM_003240.4:c.774C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003240.4:c.774C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003240.4:c.774C>T']['alt_genomic_loci'] == []
        assert results['NM_003240.4:c.774C>T']['gene_symbol'] == 'LEFTY2'
        assert results['NM_003240.4:c.774C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003231.2:p.(Thr258=)', 'slr': 'NP_003231.2:p.(T258=)'}
        assert results['NM_003240.4:c.774C>T']['submitted_variant'] == '1-226125468-G-A'
        assert results['NM_003240.4:c.774C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_003240.4:c.774C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_003240.4:c.774C>T']['hgvs_transcript_variant'] == 'NM_003240.4:c.774C>T'
        assert results['NM_003240.4:c.774C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_003240.4:c.774C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.226125468G>A', 'vcf': {'chr': 'chr1', 'ref': u'G', 'pos': '226125468', 'alt': u'A'}}
        assert results['NM_003240.4:c.774C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.225937768G>A', 'vcf': {'chr': 'chr1', 'ref': u'G', 'pos': '225937768', 'alt': u'A'}}
        assert results['NM_003240.4:c.774C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.226125468G>A', 'vcf': {'chr': '1', 'ref': u'G', 'pos': '226125468', 'alt': u'A'}}
        assert results['NM_003240.4:c.774C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.225937768G>A', 'vcf': {'chr': '1', 'ref': u'G', 'pos': '225937768', 'alt': u'A'}}

        assert 'NM_003240.3:c.774C>T' in results.keys()
        assert results['NM_003240.3:c.774C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003240.3:c.774C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003240.3:c.774C>T']['alt_genomic_loci'] == []
        assert results['NM_003240.3:c.774C>T']['gene_symbol'] == 'LEFTY2'
        assert results['NM_003240.3:c.774C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003231.2:p.(Thr258=)', 'slr': 'NP_003231.2:p.(T258=)'}
        assert results['NM_003240.3:c.774C>T']['submitted_variant'] == '1-226125468-G-A'
        assert results['NM_003240.3:c.774C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_003240.3:c.774C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_003240.3:c.774C>T']['hgvs_transcript_variant'] == 'NM_003240.3:c.774C>T'
        assert results['NM_003240.3:c.774C>T']['hgvs_refseqgene_variant'] == 'NG_008118.1:g.8453C>T'
        assert results['NM_003240.3:c.774C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.226125468G>A', 'vcf': {'chr': 'chr1', 'ref': u'G', 'pos': '226125468', 'alt': u'A'}}
        assert results['NM_003240.3:c.774C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.225937768G>A', 'vcf': {'chr': 'chr1', 'ref': u'G', 'pos': '225937768', 'alt': u'A'}}
        assert results['NM_003240.3:c.774C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.226125468G>A', 'vcf': {'chr': '1', 'ref': u'G', 'pos': '226125468', 'alt': u'A'}}
        assert results['NM_003240.3:c.774C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.225937768G>A', 'vcf': {'chr': '1', 'ref': u'G', 'pos': '225937768', 'alt': u'A'}}

        assert 'NM_001172425.2:c.672C>T' in results.keys()
        assert results['NM_001172425.2:c.672C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001172425.2:c.672C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001172425.2:c.672C>T']['alt_genomic_loci'] == []
        assert results['NM_001172425.2:c.672C>T']['gene_symbol'] == 'LEFTY2'
        assert results['NM_001172425.2:c.672C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001165896.1:p.(Thr224=)', 'slr': 'NP_001165896.1:p.(T224=)'}
        assert results['NM_001172425.2:c.672C>T']['submitted_variant'] == '1-226125468-G-A'
        assert results['NM_001172425.2:c.672C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001172425.2:c.672C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001172425.2:c.672C>T']['hgvs_transcript_variant'] == 'NM_001172425.2:c.672C>T'
        assert results['NM_001172425.2:c.672C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001172425.2:c.672C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.226125468G>A', 'vcf': {'chr': 'chr1', 'ref': u'G', 'pos': '226125468', 'alt': u'A'}}
        assert results['NM_001172425.2:c.672C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.225937768G>A', 'vcf': {'chr': 'chr1', 'ref': u'G', 'pos': '225937768', 'alt': u'A'}}
        assert results['NM_001172425.2:c.672C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.226125468G>A', 'vcf': {'chr': '1', 'ref': u'G', 'pos': '226125468', 'alt': u'A'}}
        assert results['NM_001172425.2:c.672C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.225937768G>A', 'vcf': {'chr': '1', 'ref': u'G', 'pos': '225937768', 'alt': u'A'}}


    def test_variant212(self):
        variant = '10-89623035-CGCA-C'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_001126049.1:c.-794_-792del' in results.keys()
        assert results['NM_001126049.1:c.-794_-792del']['hgvs_lrg_transcript_variant'] == 'LRG_1087t1:c.-794_-792del'
        assert results['NM_001126049.1:c.-794_-792del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001126049.1:c.-794_-792del']['alt_genomic_loci'] == [{'grch38': {'hgvs_genomic_description': 'NW_013171807.1:g.79106_79108del', 'vcf': {'chr': 'HG2334_PATCH', 'ref': 'CGCA', 'pos': '79102', 'alt': 'C'}}}, {'hg38': {'hgvs_genomic_description': 'NW_013171807.1:g.79106_79108del', 'vcf': {'chr': 'NW_013171807.1', 'ref': 'CGCA', 'pos': '79102', 'alt': 'C'}}}]
        assert results['NM_001126049.1:c.-794_-792del']['gene_symbol'] == 'KLLN'
        assert results['NM_001126049.1:c.-794_-792del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001119521.1:p.?', 'slr': 'NP_001119521.1:p.?'}
        assert results['NM_001126049.1:c.-794_-792del']['submitted_variant'] == '10-89623035-CGCA-C'
        assert results['NM_001126049.1:c.-794_-792del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001126049.1:c.-794_-792del']['hgvs_lrg_variant'] == 'LRG_1087:g.5157_5159del'
        assert results['NM_001126049.1:c.-794_-792del']['hgvs_transcript_variant'] == 'NM_001126049.1:c.-794_-792del'
        assert results['NM_001126049.1:c.-794_-792del']['hgvs_refseqgene_variant'] == 'NG_033079.1:g.5157_5159del'
        assert results['NM_001126049.1:c.-794_-792del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000010.10:g.89623039_89623041del', 'vcf': {'chr': 'chr10', 'ref': 'CGCA', 'pos': '89623035', 'alt': 'C'}}
        assert results['NM_001126049.1:c.-794_-792del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000010.11:g.87863282_87863284del', 'vcf': {'chr': 'chr10', 'ref': 'CGCA', 'pos': '87863278', 'alt': 'C'}}
        assert results['NM_001126049.1:c.-794_-792del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000010.10:g.89623039_89623041del', 'vcf': {'chr': '10', 'ref': 'CGCA', 'pos': '89623035', 'alt': 'C'}}
        assert results['NM_001126049.1:c.-794_-792del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000010.11:g.87863282_87863284del', 'vcf': {'chr': '10', 'ref': 'CGCA', 'pos': '87863278', 'alt': 'C'}}


    def test_variant213(self):
        variant = '11-62457852-C-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NR_037946.1:n.3896G>T' in results.keys()
        assert results['NR_037946.1:n.3896G>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_037946.1:n.3896G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_037946.1:n.3896G>T']['alt_genomic_loci'] == []
        assert results['NR_037946.1:n.3896G>T']['gene_symbol'] == 'HNRNPUL2-BSCL2'
        assert results['NR_037946.1:n.3896G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_037946.1:n.3896G>T']['submitted_variant'] == '11-62457852-C-A'
        assert results['NR_037946.1:n.3896G>T']['genome_context_intronic_sequence'] == ''
        assert results['NR_037946.1:n.3896G>T']['hgvs_lrg_variant'] == ''
        assert results['NR_037946.1:n.3896G>T']['hgvs_transcript_variant'] == 'NR_037946.1:n.3896G>T'
        assert results['NR_037946.1:n.3896G>T']['hgvs_refseqgene_variant'] == ''
        assert results['NR_037946.1:n.3896G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.62457852C>A', 'vcf': {'chr': 'chr11', 'ref': u'C', 'pos': '62457852', 'alt': u'A'}}
        assert results['NR_037946.1:n.3896G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.62690380C>A', 'vcf': {'chr': 'chr11', 'ref': u'C', 'pos': '62690380', 'alt': u'A'}}
        assert results['NR_037946.1:n.3896G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.62457852C>A', 'vcf': {'chr': '11', 'ref': u'C', 'pos': '62457852', 'alt': u'A'}}
        assert results['NR_037946.1:n.3896G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.62690380C>A', 'vcf': {'chr': '11', 'ref': u'C', 'pos': '62690380', 'alt': u'A'}}

        assert 'NM_032667.6:c.1184G>T' in results.keys()
        assert results['NM_032667.6:c.1184G>T']['hgvs_lrg_transcript_variant'] == 'LRG_235t2:c.1184G>T'
        assert results['NM_032667.6:c.1184G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_032667.6:c.1184G>T']['alt_genomic_loci'] == []
        assert results['NM_032667.6:c.1184G>T']['gene_symbol'] == 'BSCL2'
        assert results['NM_032667.6:c.1184G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_116056.3(LRG_235p2):p.(Cys395Phe)', 'slr': 'NP_116056.3:p.(C395F)'}
        assert results['NM_032667.6:c.1184G>T']['submitted_variant'] == '11-62457852-C-A'
        assert results['NM_032667.6:c.1184G>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_032667.6:c.1184G>T']['hgvs_lrg_variant'] == ''
        assert results['NM_032667.6:c.1184G>T']['hgvs_transcript_variant'] == 'NM_032667.6:c.1184G>T'
        assert results['NM_032667.6:c.1184G>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_032667.6:c.1184G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.62457852C>A', 'vcf': {'chr': 'chr11', 'ref': u'C', 'pos': '62457852', 'alt': u'A'}}
        assert results['NM_032667.6:c.1184G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.62690380C>A', 'vcf': {'chr': 'chr11', 'ref': u'C', 'pos': '62690380', 'alt': u'A'}}
        assert results['NM_032667.6:c.1184G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.62457852C>A', 'vcf': {'chr': '11', 'ref': u'C', 'pos': '62457852', 'alt': u'A'}}
        assert results['NM_032667.6:c.1184G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.62690380C>A', 'vcf': {'chr': '11', 'ref': u'C', 'pos': '62690380', 'alt': u'A'}}

        assert results['flag'] == 'gene_variant'
        assert 'NR_037948.1:n.1978G>T' in results.keys()
        assert results['NR_037948.1:n.1978G>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_037948.1:n.1978G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_037948.1:n.1978G>T']['alt_genomic_loci'] == []
        assert results['NR_037948.1:n.1978G>T']['gene_symbol'] == 'BSCL2'
        assert results['NR_037948.1:n.1978G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_037948.1:n.1978G>T']['submitted_variant'] == '11-62457852-C-A'
        assert results['NR_037948.1:n.1978G>T']['genome_context_intronic_sequence'] == ''
        assert results['NR_037948.1:n.1978G>T']['hgvs_lrg_variant'] == ''
        assert results['NR_037948.1:n.1978G>T']['hgvs_transcript_variant'] == 'NR_037948.1:n.1978G>T'
        assert results['NR_037948.1:n.1978G>T']['hgvs_refseqgene_variant'] == ''
        assert results['NR_037948.1:n.1978G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.62457852C>A', 'vcf': {'chr': 'chr11', 'ref': u'C', 'pos': '62457852', 'alt': u'A'}}
        assert results['NR_037948.1:n.1978G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.62690380C>A', 'vcf': {'chr': 'chr11', 'ref': u'C', 'pos': '62690380', 'alt': u'A'}}
        assert results['NR_037948.1:n.1978G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.62457852C>A', 'vcf': {'chr': '11', 'ref': u'C', 'pos': '62457852', 'alt': u'A'}}
        assert results['NR_037948.1:n.1978G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.62690380C>A', 'vcf': {'chr': '11', 'ref': u'C', 'pos': '62690380', 'alt': u'A'}}

        assert 'NM_001122955.3:c.1376G>T' in results.keys()
        assert results['NM_001122955.3:c.1376G>T']['hgvs_lrg_transcript_variant'] == 'LRG_235t1:c.1376G>T'
        assert results['NM_001122955.3:c.1376G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001122955.3:c.1376G>T']['alt_genomic_loci'] == []
        assert results['NM_001122955.3:c.1376G>T']['gene_symbol'] == 'BSCL2'
        assert results['NM_001122955.3:c.1376G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001116427.1:p.(Cys459Phe)', 'slr': 'NP_001116427.1:p.(C459F)'}
        assert results['NM_001122955.3:c.1376G>T']['submitted_variant'] == '11-62457852-C-A'
        assert results['NM_001122955.3:c.1376G>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001122955.3:c.1376G>T']['hgvs_lrg_variant'] == 'LRG_235:g.24195G>T'
        assert results['NM_001122955.3:c.1376G>T']['hgvs_transcript_variant'] == 'NM_001122955.3:c.1376G>T'
        assert results['NM_001122955.3:c.1376G>T']['hgvs_refseqgene_variant'] == 'NG_008461.1:g.24195G>T'
        assert results['NM_001122955.3:c.1376G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.62457852C>A', 'vcf': {'chr': 'chr11', 'ref': u'C', 'pos': '62457852', 'alt': u'A'}}
        assert results['NM_001122955.3:c.1376G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.62690380C>A', 'vcf': {'chr': 'chr11', 'ref': u'C', 'pos': '62690380', 'alt': u'A'}}
        assert results['NM_001122955.3:c.1376G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.62457852C>A', 'vcf': {'chr': '11', 'ref': u'C', 'pos': '62457852', 'alt': u'A'}}
        assert results['NM_001122955.3:c.1376G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.62690380C>A', 'vcf': {'chr': '11', 'ref': u'C', 'pos': '62690380', 'alt': u'A'}}

        assert 'NR_037949.1:n.1984G>T' in results.keys()
        assert results['NR_037949.1:n.1984G>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_037949.1:n.1984G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_037949.1:n.1984G>T']['alt_genomic_loci'] == []
        assert results['NR_037949.1:n.1984G>T']['gene_symbol'] == 'BSCL2'
        assert results['NR_037949.1:n.1984G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_037949.1:n.1984G>T']['submitted_variant'] == '11-62457852-C-A'
        assert results['NR_037949.1:n.1984G>T']['genome_context_intronic_sequence'] == ''
        assert results['NR_037949.1:n.1984G>T']['hgvs_lrg_variant'] == ''
        assert results['NR_037949.1:n.1984G>T']['hgvs_transcript_variant'] == 'NR_037949.1:n.1984G>T'
        assert results['NR_037949.1:n.1984G>T']['hgvs_refseqgene_variant'] == ''
        assert results['NR_037949.1:n.1984G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.62457852C>A', 'vcf': {'chr': 'chr11', 'ref': u'C', 'pos': '62457852', 'alt': u'A'}}
        assert results['NR_037949.1:n.1984G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.62690380C>A', 'vcf': {'chr': 'chr11', 'ref': u'C', 'pos': '62690380', 'alt': u'A'}}
        assert results['NR_037949.1:n.1984G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.62457852C>A', 'vcf': {'chr': '11', 'ref': u'C', 'pos': '62457852', 'alt': u'A'}}
        assert results['NR_037949.1:n.1984G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.62690380C>A', 'vcf': {'chr': '11', 'ref': u'C', 'pos': '62690380', 'alt': u'A'}}

        assert 'NM_001130702.2:c.*178G>T' in results.keys()
        assert results['NM_001130702.2:c.*178G>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001130702.2:c.*178G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001130702.2:c.*178G>T']['alt_genomic_loci'] == []
        assert results['NM_001130702.2:c.*178G>T']['gene_symbol'] == 'BSCL2'
        assert results['NM_001130702.2:c.*178G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001124174.2:p.?', 'slr': 'NP_001124174.2:p.?'}
        assert results['NM_001130702.2:c.*178G>T']['submitted_variant'] == '11-62457852-C-A'
        assert results['NM_001130702.2:c.*178G>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001130702.2:c.*178G>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001130702.2:c.*178G>T']['hgvs_transcript_variant'] == 'NM_001130702.2:c.*178G>T'
        assert results['NM_001130702.2:c.*178G>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001130702.2:c.*178G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.62457852C>A', 'vcf': {'chr': 'chr11', 'ref': u'C', 'pos': '62457852', 'alt': u'A'}}
        assert results['NM_001130702.2:c.*178G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.62690380C>A', 'vcf': {'chr': 'chr11', 'ref': u'C', 'pos': '62690380', 'alt': u'A'}}
        assert results['NM_001130702.2:c.*178G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.62457852C>A', 'vcf': {'chr': '11', 'ref': u'C', 'pos': '62457852', 'alt': u'A'}}
        assert results['NM_001130702.2:c.*178G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.62690380C>A', 'vcf': {'chr': '11', 'ref': u'C', 'pos': '62690380', 'alt': u'A'}}


    def test_variant214(self):
        variant = '11-108178710-A-AT'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001351834.1:c.5761_5762insT' in results.keys()
        assert results['NM_001351834.1:c.5761_5762insT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001351834.1:c.5761_5762insT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001351834.1:c.5761_5762insT']['alt_genomic_loci'] == []
        assert results['NM_001351834.1:c.5761_5762insT']['gene_symbol'] == 'ATM'
        assert results['NM_001351834.1:c.5761_5762insT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001338763.1:p.(Arg1921MetfsTer9)', 'slr': 'NP_001338763.1:p.(R1921Mfs*9)'}
        assert results['NM_001351834.1:c.5761_5762insT']['submitted_variant'] == '11-108178710-A-AT'
        assert results['NM_001351834.1:c.5761_5762insT']['genome_context_intronic_sequence'] == ''
        assert results['NM_001351834.1:c.5761_5762insT']['hgvs_lrg_variant'] == ''
        assert results['NM_001351834.1:c.5761_5762insT']['hgvs_transcript_variant'] == 'NM_001351834.1:c.5761_5762insT'
        assert results['NM_001351834.1:c.5761_5762insT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001351834.1:c.5761_5762insT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.108178710_108178711insT', 'vcf': {'chr': 'chr11', 'ref': 'A', 'pos': '108178710', 'alt': 'AT'}}
        assert results['NM_001351834.1:c.5761_5762insT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.108307983_108307984insT', 'vcf': {'chr': 'chr11', 'ref': 'A', 'pos': '108307983', 'alt': 'AT'}}
        assert results['NM_001351834.1:c.5761_5762insT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.108178710_108178711insT', 'vcf': {'chr': '11', 'ref': 'A', 'pos': '108178710', 'alt': 'AT'}}
        assert results['NM_001351834.1:c.5761_5762insT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.108307983_108307984insT', 'vcf': {'chr': '11', 'ref': 'A', 'pos': '108307983', 'alt': 'AT'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_000051.3:c.5761_5762insT' in results.keys()
        assert results['NM_000051.3:c.5761_5762insT']['hgvs_lrg_transcript_variant'] == 'LRG_135t1:c.5761_5762insT'
        assert results['NM_000051.3:c.5761_5762insT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000051.3:c.5761_5762insT']['alt_genomic_loci'] == []
        assert results['NM_000051.3:c.5761_5762insT']['gene_symbol'] == 'ATM'
        assert results['NM_000051.3:c.5761_5762insT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000042.3(LRG_135p1):p.(Arg1921MetfsTer9)', 'slr': 'NP_000042.3:p.(R1921Mfs*9)'}
        assert results['NM_000051.3:c.5761_5762insT']['submitted_variant'] == '11-108178710-A-AT'
        assert results['NM_000051.3:c.5761_5762insT']['genome_context_intronic_sequence'] == ''
        assert results['NM_000051.3:c.5761_5762insT']['hgvs_lrg_variant'] == 'LRG_135:g.90152_90153insT'
        assert results['NM_000051.3:c.5761_5762insT']['hgvs_transcript_variant'] == 'NM_000051.3:c.5761_5762insT'
        assert results['NM_000051.3:c.5761_5762insT']['hgvs_refseqgene_variant'] == 'NG_009830.1:g.90152_90153insT'
        assert results['NM_000051.3:c.5761_5762insT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.108178710_108178711insT', 'vcf': {'chr': 'chr11', 'ref': 'A', 'pos': '108178710', 'alt': 'AT'}}
        assert results['NM_000051.3:c.5761_5762insT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.108307983_108307984insT', 'vcf': {'chr': 'chr11', 'ref': 'A', 'pos': '108307983', 'alt': 'AT'}}
        assert results['NM_000051.3:c.5761_5762insT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.108178710_108178711insT', 'vcf': {'chr': '11', 'ref': 'A', 'pos': '108178710', 'alt': 'AT'}}
        assert results['NM_000051.3:c.5761_5762insT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.108307983_108307984insT', 'vcf': {'chr': '11', 'ref': 'A', 'pos': '108307983', 'alt': 'AT'}}


    def test_variant215(self):
        variant = '11-111735981-G-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001352413.1:c.-108-7C>T' in results.keys()
        assert results['NM_001352413.1:c.-108-7C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001352413.1:c.-108-7C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001352413.1:c.-108-7C>T']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003871080.1:g.117249G>A', 'vcf': {'chr': 'HG388_HG400_PATCH', 'ref': u'G', 'pos': '117249', 'alt': u'A'}}}]
        assert results['NM_001352413.1:c.-108-7C>T']['gene_symbol'] == 'ALG9'
        assert results['NM_001352413.1:c.-108-7C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001339342.1:p.?', 'slr': 'NP_001339342.1:p.?'}
        assert results['NM_001352413.1:c.-108-7C>T']['submitted_variant'] == '11-111735981-G-A'
        assert results['NM_001352413.1:c.-108-7C>T']['genome_context_intronic_sequence'] == 'NC_000011.9(NM_001352413.1):c.-108-7C>T'
        assert results['NM_001352413.1:c.-108-7C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001352413.1:c.-108-7C>T']['hgvs_transcript_variant'] == 'NM_001352413.1:c.-108-7C>T'
        assert results['NM_001352413.1:c.-108-7C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001352413.1:c.-108-7C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001352413.1:c.-108-7C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}
        assert results['NM_001352413.1:c.-108-7C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001352413.1:c.-108-7C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}

        assert 'NM_001352415.1:c.-108-7C>T' in results.keys()
        assert results['NM_001352415.1:c.-108-7C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001352415.1:c.-108-7C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001352415.1:c.-108-7C>T']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003871080.1:g.117249G>A', 'vcf': {'chr': 'HG388_HG400_PATCH', 'ref': u'G', 'pos': '117249', 'alt': u'A'}}}]
        assert results['NM_001352415.1:c.-108-7C>T']['gene_symbol'] == 'ALG9'
        assert results['NM_001352415.1:c.-108-7C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001339344.1:p.?', 'slr': 'NP_001339344.1:p.?'}
        assert results['NM_001352415.1:c.-108-7C>T']['submitted_variant'] == '11-111735981-G-A'
        assert results['NM_001352415.1:c.-108-7C>T']['genome_context_intronic_sequence'] == 'NC_000011.9(NM_001352415.1):c.-108-7C>T'
        assert results['NM_001352415.1:c.-108-7C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001352415.1:c.-108-7C>T']['hgvs_transcript_variant'] == 'NM_001352415.1:c.-108-7C>T'
        assert results['NM_001352415.1:c.-108-7C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001352415.1:c.-108-7C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001352415.1:c.-108-7C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}
        assert results['NM_001352415.1:c.-108-7C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001352415.1:c.-108-7C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}

        assert 'NM_001352410.1:c.-108-7C>T' in results.keys()
        assert results['NM_001352410.1:c.-108-7C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001352410.1:c.-108-7C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001352410.1:c.-108-7C>T']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003871080.1:g.117249G>A', 'vcf': {'chr': 'HG388_HG400_PATCH', 'ref': u'G', 'pos': '117249', 'alt': u'A'}}}]
        assert results['NM_001352410.1:c.-108-7C>T']['gene_symbol'] == 'ALG9'
        assert results['NM_001352410.1:c.-108-7C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001339339.1:p.?', 'slr': 'NP_001339339.1:p.?'}
        assert results['NM_001352410.1:c.-108-7C>T']['submitted_variant'] == '11-111735981-G-A'
        assert results['NM_001352410.1:c.-108-7C>T']['genome_context_intronic_sequence'] == 'NC_000011.9(NM_001352410.1):c.-108-7C>T'
        assert results['NM_001352410.1:c.-108-7C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001352410.1:c.-108-7C>T']['hgvs_transcript_variant'] == 'NM_001352410.1:c.-108-7C>T'
        assert results['NM_001352410.1:c.-108-7C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001352410.1:c.-108-7C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001352410.1:c.-108-7C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}
        assert results['NM_001352410.1:c.-108-7C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001352410.1:c.-108-7C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}

        assert 'NM_024740.2:c.406-7C>T' in results.keys()
        assert results['NM_024740.2:c.406-7C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_024740.2:c.406-7C>T']['refseqgene_context_intronic_sequence'] == 'NG_009210.1(NM_024740.2):c.406-7C>T'
        assert results['NM_024740.2:c.406-7C>T']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003871080.1:g.117249G>A', 'vcf': {'chr': 'HG388_HG400_PATCH', 'ref': u'G', 'pos': '117249', 'alt': u'A'}}}]
        assert results['NM_024740.2:c.406-7C>T']['gene_symbol'] == 'ALG9'
        assert results['NM_024740.2:c.406-7C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_079016.2:p.?', 'slr': 'NP_079016.2:p.?'}
        assert results['NM_024740.2:c.406-7C>T']['submitted_variant'] == '11-111735981-G-A'
        assert results['NM_024740.2:c.406-7C>T']['genome_context_intronic_sequence'] == 'NC_000011.9(NM_024740.2):c.406-7C>T'
        assert results['NM_024740.2:c.406-7C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_024740.2:c.406-7C>T']['hgvs_transcript_variant'] == 'NM_024740.2:c.406-7C>T'
        assert results['NM_024740.2:c.406-7C>T']['hgvs_refseqgene_variant'] == 'NG_009210.1:g.11324C>T'
        assert results['NM_024740.2:c.406-7C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_024740.2:c.406-7C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}
        assert results['NM_024740.2:c.406-7C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_024740.2:c.406-7C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}

        assert 'NM_001352419.1:c.-108-7C>T' in results.keys()
        assert results['NM_001352419.1:c.-108-7C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001352419.1:c.-108-7C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001352419.1:c.-108-7C>T']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003871080.1:g.117249G>A', 'vcf': {'chr': 'HG388_HG400_PATCH', 'ref': u'G', 'pos': '117249', 'alt': u'A'}}}]
        assert results['NM_001352419.1:c.-108-7C>T']['gene_symbol'] == 'ALG9'
        assert results['NM_001352419.1:c.-108-7C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001339348.1:p.?', 'slr': 'NP_001339348.1:p.?'}
        assert results['NM_001352419.1:c.-108-7C>T']['submitted_variant'] == '11-111735981-G-A'
        assert results['NM_001352419.1:c.-108-7C>T']['genome_context_intronic_sequence'] == 'NC_000011.9(NM_001352419.1):c.-108-7C>T'
        assert results['NM_001352419.1:c.-108-7C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001352419.1:c.-108-7C>T']['hgvs_transcript_variant'] == 'NM_001352419.1:c.-108-7C>T'
        assert results['NM_001352419.1:c.-108-7C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001352419.1:c.-108-7C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001352419.1:c.-108-7C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}
        assert results['NM_001352419.1:c.-108-7C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001352419.1:c.-108-7C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}

        assert 'NM_001352421.1:c.-108-7C>T' in results.keys()
        assert results['NM_001352421.1:c.-108-7C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001352421.1:c.-108-7C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001352421.1:c.-108-7C>T']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003871080.1:g.117249G>A', 'vcf': {'chr': 'HG388_HG400_PATCH', 'ref': u'G', 'pos': '117249', 'alt': u'A'}}}]
        assert results['NM_001352421.1:c.-108-7C>T']['gene_symbol'] == 'ALG9'
        assert results['NM_001352421.1:c.-108-7C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001339350.1:p.?', 'slr': 'NP_001339350.1:p.?'}
        assert results['NM_001352421.1:c.-108-7C>T']['submitted_variant'] == '11-111735981-G-A'
        assert results['NM_001352421.1:c.-108-7C>T']['genome_context_intronic_sequence'] == 'NC_000011.9(NM_001352421.1):c.-108-7C>T'
        assert results['NM_001352421.1:c.-108-7C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001352421.1:c.-108-7C>T']['hgvs_transcript_variant'] == 'NM_001352421.1:c.-108-7C>T'
        assert results['NM_001352421.1:c.-108-7C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001352421.1:c.-108-7C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001352421.1:c.-108-7C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}
        assert results['NM_001352421.1:c.-108-7C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001352421.1:c.-108-7C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}

        assert 'NM_001352417.1:c.406-7C>T' in results.keys()
        assert results['NM_001352417.1:c.406-7C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001352417.1:c.406-7C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001352417.1:c.406-7C>T']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003871080.1:g.117249G>A', 'vcf': {'chr': 'HG388_HG400_PATCH', 'ref': u'G', 'pos': '117249', 'alt': u'A'}}}]
        assert results['NM_001352417.1:c.406-7C>T']['gene_symbol'] == 'ALG9'
        assert results['NM_001352417.1:c.406-7C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001339346.1:p.?', 'slr': 'NP_001339346.1:p.?'}
        assert results['NM_001352417.1:c.406-7C>T']['submitted_variant'] == '11-111735981-G-A'
        assert results['NM_001352417.1:c.406-7C>T']['genome_context_intronic_sequence'] == 'NC_000011.9(NM_001352417.1):c.406-7C>T'
        assert results['NM_001352417.1:c.406-7C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001352417.1:c.406-7C>T']['hgvs_transcript_variant'] == 'NM_001352417.1:c.406-7C>T'
        assert results['NM_001352417.1:c.406-7C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001352417.1:c.406-7C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001352417.1:c.406-7C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}
        assert results['NM_001352417.1:c.406-7C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001352417.1:c.406-7C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}

        assert 'NM_001352414.1:c.-108-7C>T' in results.keys()
        assert results['NM_001352414.1:c.-108-7C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001352414.1:c.-108-7C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001352414.1:c.-108-7C>T']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003871080.1:g.117249G>A', 'vcf': {'chr': 'HG388_HG400_PATCH', 'ref': u'G', 'pos': '117249', 'alt': u'A'}}}]
        assert results['NM_001352414.1:c.-108-7C>T']['gene_symbol'] == 'ALG9'
        assert results['NM_001352414.1:c.-108-7C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001339343.1:p.?', 'slr': 'NP_001339343.1:p.?'}
        assert results['NM_001352414.1:c.-108-7C>T']['submitted_variant'] == '11-111735981-G-A'
        assert results['NM_001352414.1:c.-108-7C>T']['genome_context_intronic_sequence'] == 'NC_000011.9(NM_001352414.1):c.-108-7C>T'
        assert results['NM_001352414.1:c.-108-7C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001352414.1:c.-108-7C>T']['hgvs_transcript_variant'] == 'NM_001352414.1:c.-108-7C>T'
        assert results['NM_001352414.1:c.-108-7C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001352414.1:c.-108-7C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001352414.1:c.-108-7C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}
        assert results['NM_001352414.1:c.-108-7C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001352414.1:c.-108-7C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}

        assert 'NM_001352420.1:c.-108-7C>T' in results.keys()
        assert results['NM_001352420.1:c.-108-7C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001352420.1:c.-108-7C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001352420.1:c.-108-7C>T']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003871080.1:g.117249G>A', 'vcf': {'chr': 'HG388_HG400_PATCH', 'ref': u'G', 'pos': '117249', 'alt': u'A'}}}]
        assert results['NM_001352420.1:c.-108-7C>T']['gene_symbol'] == 'ALG9'
        assert results['NM_001352420.1:c.-108-7C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001339349.1:p.?', 'slr': 'NP_001339349.1:p.?'}
        assert results['NM_001352420.1:c.-108-7C>T']['submitted_variant'] == '11-111735981-G-A'
        assert results['NM_001352420.1:c.-108-7C>T']['genome_context_intronic_sequence'] == 'NC_000011.9(NM_001352420.1):c.-108-7C>T'
        assert results['NM_001352420.1:c.-108-7C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001352420.1:c.-108-7C>T']['hgvs_transcript_variant'] == 'NM_001352420.1:c.-108-7C>T'
        assert results['NM_001352420.1:c.-108-7C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001352420.1:c.-108-7C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001352420.1:c.-108-7C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}
        assert results['NM_001352420.1:c.-108-7C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001352420.1:c.-108-7C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}

        assert 'NM_001352411.1:c.-108-7C>T' in results.keys()
        assert results['NM_001352411.1:c.-108-7C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001352411.1:c.-108-7C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001352411.1:c.-108-7C>T']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003871080.1:g.117249G>A', 'vcf': {'chr': 'HG388_HG400_PATCH', 'ref': u'G', 'pos': '117249', 'alt': u'A'}}}]
        assert results['NM_001352411.1:c.-108-7C>T']['gene_symbol'] == 'ALG9'
        assert results['NM_001352411.1:c.-108-7C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001339340.1:p.?', 'slr': 'NP_001339340.1:p.?'}
        assert results['NM_001352411.1:c.-108-7C>T']['submitted_variant'] == '11-111735981-G-A'
        assert results['NM_001352411.1:c.-108-7C>T']['genome_context_intronic_sequence'] == 'NC_000011.9(NM_001352411.1):c.-108-7C>T'
        assert results['NM_001352411.1:c.-108-7C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001352411.1:c.-108-7C>T']['hgvs_transcript_variant'] == 'NM_001352411.1:c.-108-7C>T'
        assert results['NM_001352411.1:c.-108-7C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001352411.1:c.-108-7C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001352411.1:c.-108-7C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}
        assert results['NM_001352411.1:c.-108-7C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001352411.1:c.-108-7C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}

        assert 'NM_001077692.1:c.-108-7C>T' in results.keys()
        assert results['NM_001077692.1:c.-108-7C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001077692.1:c.-108-7C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001077692.1:c.-108-7C>T']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003871080.1:g.117249G>A', 'vcf': {'chr': 'HG388_HG400_PATCH', 'ref': u'G', 'pos': '117249', 'alt': u'A'}}}]
        assert results['NM_001077692.1:c.-108-7C>T']['gene_symbol'] == 'ALG9'
        assert results['NM_001077692.1:c.-108-7C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001071160.1:p.?', 'slr': 'NP_001071160.1:p.?'}
        assert results['NM_001077692.1:c.-108-7C>T']['submitted_variant'] == '11-111735981-G-A'
        assert results['NM_001077692.1:c.-108-7C>T']['genome_context_intronic_sequence'] == 'NC_000011.9(NM_001077692.1):c.-108-7C>T'
        assert results['NM_001077692.1:c.-108-7C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001077692.1:c.-108-7C>T']['hgvs_transcript_variant'] == 'NM_001077692.1:c.-108-7C>T'
        assert results['NM_001077692.1:c.-108-7C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001077692.1:c.-108-7C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001077692.1:c.-108-7C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}
        assert results['NM_001077692.1:c.-108-7C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001077692.1:c.-108-7C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}

        assert results['flag'] == 'gene_variant'
        assert 'NR_147984.1:n.782-7C>T' in results.keys()
        assert results['NR_147984.1:n.782-7C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_147984.1:n.782-7C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_147984.1:n.782-7C>T']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003871080.1:g.117249G>A', 'vcf': {'chr': 'HG388_HG400_PATCH', 'ref': u'G', 'pos': '117249', 'alt': u'A'}}}]
        assert results['NR_147984.1:n.782-7C>T']['gene_symbol'] == 'ALG9'
        assert results['NR_147984.1:n.782-7C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_147984.1:n.782-7C>T']['submitted_variant'] == '11-111735981-G-A'
        assert results['NR_147984.1:n.782-7C>T']['genome_context_intronic_sequence'] == 'NC_000011.9(NR_147984.1):c.782-7C>T'
        assert results['NR_147984.1:n.782-7C>T']['hgvs_lrg_variant'] == ''
        assert results['NR_147984.1:n.782-7C>T']['hgvs_transcript_variant'] == 'NR_147984.1:n.782-7C>T'
        assert results['NR_147984.1:n.782-7C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NR_147984.1:n.782-7C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NR_147984.1:n.782-7C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}
        assert results['NR_147984.1:n.782-7C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NR_147984.1:n.782-7C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}

        assert 'NM_001352409.1:c.-108-7C>T' in results.keys()
        assert results['NM_001352409.1:c.-108-7C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001352409.1:c.-108-7C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001352409.1:c.-108-7C>T']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003871080.1:g.117249G>A', 'vcf': {'chr': 'HG388_HG400_PATCH', 'ref': u'G', 'pos': '117249', 'alt': u'A'}}}]
        assert results['NM_001352409.1:c.-108-7C>T']['gene_symbol'] == 'ALG9'
        assert results['NM_001352409.1:c.-108-7C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001339338.1:p.?', 'slr': 'NP_001339338.1:p.?'}
        assert results['NM_001352409.1:c.-108-7C>T']['submitted_variant'] == '11-111735981-G-A'
        assert results['NM_001352409.1:c.-108-7C>T']['genome_context_intronic_sequence'] == 'NC_000011.9(NM_001352409.1):c.-108-7C>T'
        assert results['NM_001352409.1:c.-108-7C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001352409.1:c.-108-7C>T']['hgvs_transcript_variant'] == 'NM_001352409.1:c.-108-7C>T'
        assert results['NM_001352409.1:c.-108-7C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001352409.1:c.-108-7C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001352409.1:c.-108-7C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}
        assert results['NM_001352409.1:c.-108-7C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001352409.1:c.-108-7C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}

        assert 'NM_001352412.1:c.-108-7C>T' in results.keys()
        assert results['NM_001352412.1:c.-108-7C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001352412.1:c.-108-7C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001352412.1:c.-108-7C>T']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003871080.1:g.117249G>A', 'vcf': {'chr': 'HG388_HG400_PATCH', 'ref': u'G', 'pos': '117249', 'alt': u'A'}}}]
        assert results['NM_001352412.1:c.-108-7C>T']['gene_symbol'] == 'ALG9'
        assert results['NM_001352412.1:c.-108-7C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001339341.1:p.?', 'slr': 'NP_001339341.1:p.?'}
        assert results['NM_001352412.1:c.-108-7C>T']['submitted_variant'] == '11-111735981-G-A'
        assert results['NM_001352412.1:c.-108-7C>T']['genome_context_intronic_sequence'] == 'NC_000011.9(NM_001352412.1):c.-108-7C>T'
        assert results['NM_001352412.1:c.-108-7C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001352412.1:c.-108-7C>T']['hgvs_transcript_variant'] == 'NM_001352412.1:c.-108-7C>T'
        assert results['NM_001352412.1:c.-108-7C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001352412.1:c.-108-7C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001352412.1:c.-108-7C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}
        assert results['NM_001352412.1:c.-108-7C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001352412.1:c.-108-7C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}

        assert 'NM_001077691.1:c.-108-7C>T' in results.keys()
        assert results['NM_001077691.1:c.-108-7C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001077691.1:c.-108-7C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001077691.1:c.-108-7C>T']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003871080.1:g.117249G>A', 'vcf': {'chr': 'HG388_HG400_PATCH', 'ref': u'G', 'pos': '117249', 'alt': u'A'}}}]
        assert results['NM_001077691.1:c.-108-7C>T']['gene_symbol'] == 'ALG9'
        assert results['NM_001077691.1:c.-108-7C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001071159.1:p.?', 'slr': 'NP_001071159.1:p.?'}
        assert results['NM_001077691.1:c.-108-7C>T']['submitted_variant'] == '11-111735981-G-A'
        assert results['NM_001077691.1:c.-108-7C>T']['genome_context_intronic_sequence'] == 'NC_000011.9(NM_001077691.1):c.-108-7C>T'
        assert results['NM_001077691.1:c.-108-7C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001077691.1:c.-108-7C>T']['hgvs_transcript_variant'] == 'NM_001077691.1:c.-108-7C>T'
        assert results['NM_001077691.1:c.-108-7C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001077691.1:c.-108-7C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001077691.1:c.-108-7C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}
        assert results['NM_001077691.1:c.-108-7C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001077691.1:c.-108-7C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}

        assert 'NM_001352422.1:c.-326-7C>T' in results.keys()
        assert results['NM_001352422.1:c.-326-7C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001352422.1:c.-326-7C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001352422.1:c.-326-7C>T']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003871080.1:g.117249G>A', 'vcf': {'chr': 'HG388_HG400_PATCH', 'ref': u'G', 'pos': '117249', 'alt': u'A'}}}]
        assert results['NM_001352422.1:c.-326-7C>T']['gene_symbol'] == 'ALG9'
        assert results['NM_001352422.1:c.-326-7C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001339351.1:p.?', 'slr': 'NP_001339351.1:p.?'}
        assert results['NM_001352422.1:c.-326-7C>T']['submitted_variant'] == '11-111735981-G-A'
        assert results['NM_001352422.1:c.-326-7C>T']['genome_context_intronic_sequence'] == 'NC_000011.9(NM_001352422.1):c.-326-7C>T'
        assert results['NM_001352422.1:c.-326-7C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001352422.1:c.-326-7C>T']['hgvs_transcript_variant'] == 'NM_001352422.1:c.-326-7C>T'
        assert results['NM_001352422.1:c.-326-7C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001352422.1:c.-326-7C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001352422.1:c.-326-7C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}
        assert results['NM_001352422.1:c.-326-7C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001352422.1:c.-326-7C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}

        assert 'NM_001077690.1:c.406-7C>T' in results.keys()
        assert results['NM_001077690.1:c.406-7C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001077690.1:c.406-7C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001077690.1:c.406-7C>T']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003871080.1:g.117249G>A', 'vcf': {'chr': 'HG388_HG400_PATCH', 'ref': u'G', 'pos': '117249', 'alt': u'A'}}}]
        assert results['NM_001077690.1:c.406-7C>T']['gene_symbol'] == 'ALG9'
        assert results['NM_001077690.1:c.406-7C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001071158.1:p.?', 'slr': 'NP_001071158.1:p.?'}
        assert results['NM_001077690.1:c.406-7C>T']['submitted_variant'] == '11-111735981-G-A'
        assert results['NM_001077690.1:c.406-7C>T']['genome_context_intronic_sequence'] == 'NC_000011.9(NM_001077690.1):c.406-7C>T'
        assert results['NM_001077690.1:c.406-7C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001077690.1:c.406-7C>T']['hgvs_transcript_variant'] == 'NM_001077690.1:c.406-7C>T'
        assert results['NM_001077690.1:c.406-7C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001077690.1:c.406-7C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001077690.1:c.406-7C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}
        assert results['NM_001077690.1:c.406-7C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001077690.1:c.406-7C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}

        assert 'NM_001352418.1:c.406-7C>T' in results.keys()
        assert results['NM_001352418.1:c.406-7C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001352418.1:c.406-7C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001352418.1:c.406-7C>T']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003871080.1:g.117249G>A', 'vcf': {'chr': 'HG388_HG400_PATCH', 'ref': u'G', 'pos': '117249', 'alt': u'A'}}}]
        assert results['NM_001352418.1:c.406-7C>T']['gene_symbol'] == 'ALG9'
        assert results['NM_001352418.1:c.406-7C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001339347.1:p.?', 'slr': 'NP_001339347.1:p.?'}
        assert results['NM_001352418.1:c.406-7C>T']['submitted_variant'] == '11-111735981-G-A'
        assert results['NM_001352418.1:c.406-7C>T']['genome_context_intronic_sequence'] == 'NC_000011.9(NM_001352418.1):c.406-7C>T'
        assert results['NM_001352418.1:c.406-7C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001352418.1:c.406-7C>T']['hgvs_transcript_variant'] == 'NM_001352418.1:c.406-7C>T'
        assert results['NM_001352418.1:c.406-7C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001352418.1:c.406-7C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001352418.1:c.406-7C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}
        assert results['NM_001352418.1:c.406-7C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001352418.1:c.406-7C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}

        assert 'NM_001352423.1:c.-108-7C>T' in results.keys()
        assert results['NM_001352423.1:c.-108-7C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001352423.1:c.-108-7C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001352423.1:c.-108-7C>T']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003871080.1:g.117249G>A', 'vcf': {'chr': 'HG388_HG400_PATCH', 'ref': u'G', 'pos': '117249', 'alt': u'A'}}}]
        assert results['NM_001352423.1:c.-108-7C>T']['gene_symbol'] == 'ALG9'
        assert results['NM_001352423.1:c.-108-7C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001339352.1:p.?', 'slr': 'NP_001339352.1:p.?'}
        assert results['NM_001352423.1:c.-108-7C>T']['submitted_variant'] == '11-111735981-G-A'
        assert results['NM_001352423.1:c.-108-7C>T']['genome_context_intronic_sequence'] == 'NC_000011.9(NM_001352423.1):c.-108-7C>T'
        assert results['NM_001352423.1:c.-108-7C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001352423.1:c.-108-7C>T']['hgvs_transcript_variant'] == 'NM_001352423.1:c.-108-7C>T'
        assert results['NM_001352423.1:c.-108-7C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001352423.1:c.-108-7C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001352423.1:c.-108-7C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}
        assert results['NM_001352423.1:c.-108-7C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001352423.1:c.-108-7C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}

        assert 'NM_001352416.1:c.-108-7C>T' in results.keys()
        assert results['NM_001352416.1:c.-108-7C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001352416.1:c.-108-7C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001352416.1:c.-108-7C>T']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003871080.1:g.117249G>A', 'vcf': {'chr': 'HG388_HG400_PATCH', 'ref': u'G', 'pos': '117249', 'alt': u'A'}}}]
        assert results['NM_001352416.1:c.-108-7C>T']['gene_symbol'] == 'ALG9'
        assert results['NM_001352416.1:c.-108-7C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001339345.1:p.?', 'slr': 'NP_001339345.1:p.?'}
        assert results['NM_001352416.1:c.-108-7C>T']['submitted_variant'] == '11-111735981-G-A'
        assert results['NM_001352416.1:c.-108-7C>T']['genome_context_intronic_sequence'] == 'NC_000011.9(NM_001352416.1):c.-108-7C>T'
        assert results['NM_001352416.1:c.-108-7C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001352416.1:c.-108-7C>T']['hgvs_transcript_variant'] == 'NM_001352416.1:c.-108-7C>T'
        assert results['NM_001352416.1:c.-108-7C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001352416.1:c.-108-7C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001352416.1:c.-108-7C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}
        assert results['NM_001352416.1:c.-108-7C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.111735981G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111735981', 'alt': u'A'}}
        assert results['NM_001352416.1:c.-108-7C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.111865258G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '111865258', 'alt': u'A'}}


    def test_variant216(self):
        variant = '12-11023080-C-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NR_037918.2:n.1184+11736G>T' in results.keys()
        assert results['NR_037918.2:n.1184+11736G>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_037918.2:n.1184+11736G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_037918.2:n.1184+11736G>T']['alt_genomic_loci'] == [{'grch38': {'hgvs_genomic_description': 'NT_187658.1:g.69187C>A', 'vcf': {'chr': 'HSCHR12_3_CTG2', 'ref': u'C', 'pos': '69187', 'alt': u'A'}}}, {'hg38': {'hgvs_genomic_description': 'NT_187658.1:g.69187C>A', 'vcf': {'chr': 'chr12_KI270904v1_alt', 'ref': u'C', 'pos': '69187', 'alt': u'A'}}}, {'grch37': {'hgvs_genomic_description': 'NW_003571047.1:g.69187C>A', 'vcf': {'chr': 'HG1133_PATCH', 'ref': u'C', 'pos': '69187', 'alt': u'A'}}}, {'grch37': {'hgvs_genomic_description': 'NW_003571050.1:g.69187C>A', 'vcf': {'chr': 'HSCHR12_2_CTG2', 'ref': u'C', 'pos': '69187', 'alt': u'A'}}}, {'grch38': {'hgvs_genomic_description': 'NW_003571050.1:g.69187C>A', 'vcf': {'chr': 'HSCHR12_2_CTG2', 'ref': u'C', 'pos': '69187', 'alt': u'A'}}}, {'hg38': {'hgvs_genomic_description': 'NW_003571050.1:g.69187C>A', 'vcf': {'chr': 'chr12_GL877876v1_alt', 'ref': u'C', 'pos': '69187', 'alt': u'A'}}}]
        assert results['NR_037918.2:n.1184+11736G>T']['gene_symbol'] == 'PRH1-PRR4'
        assert results['NR_037918.2:n.1184+11736G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_037918.2:n.1184+11736G>T']['submitted_variant'] == '12-11023080-C-A'
        assert results['NR_037918.2:n.1184+11736G>T']['genome_context_intronic_sequence'] == 'NC_000012.11(NR_037918.2):c.1184+11736G>T'
        assert results['NR_037918.2:n.1184+11736G>T']['hgvs_lrg_variant'] == ''
        assert results['NR_037918.2:n.1184+11736G>T']['hgvs_transcript_variant'] == 'NR_037918.2:n.1184+11736G>T'
        assert results['NR_037918.2:n.1184+11736G>T']['hgvs_refseqgene_variant'] == ''
        assert results['NR_037918.2:n.1184+11736G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.11023080C>A', 'vcf': {'chr': 'chr12', 'ref': u'C', 'pos': '11023080', 'alt': u'A'}}
        assert results['NR_037918.2:n.1184+11736G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.10870481C>A', 'vcf': {'chr': 'chr12', 'ref': u'C', 'pos': '10870481', 'alt': u'A'}}
        assert results['NR_037918.2:n.1184+11736G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.11023080C>A', 'vcf': {'chr': '12', 'ref': u'C', 'pos': '11023080', 'alt': u'A'}}
        assert results['NR_037918.2:n.1184+11736G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.10870481C>A', 'vcf': {'chr': '12', 'ref': u'C', 'pos': '10870481', 'alt': u'A'}}


    def test_variant217(self):
        variant = '12-22018712-TC-T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_020297.3:c.2199-1302del' in results.keys()
        assert results['NM_020297.3:c.2199-1302del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_020297.3:c.2199-1302del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_020297.3:c.2199-1302del']['alt_genomic_loci'] == []
        assert results['NM_020297.3:c.2199-1302del']['gene_symbol'] == 'ABCC9'
        assert results['NM_020297.3:c.2199-1302del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_064693.2(LRG_377p1):p.?', 'slr': 'NP_064693.2:p.?'}
        assert results['NM_020297.3:c.2199-1302del']['submitted_variant'] == '12-22018712-TC-T'
        assert results['NM_020297.3:c.2199-1302del']['genome_context_intronic_sequence'] == 'NC_000012.11(NM_020297.3):c.2199-1302del'
        assert results['NM_020297.3:c.2199-1302del']['hgvs_lrg_variant'] == ''
        assert results['NM_020297.3:c.2199-1302del']['hgvs_transcript_variant'] == 'NM_020297.3:c.2199-1302del'
        assert results['NM_020297.3:c.2199-1302del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_020297.3:c.2199-1302del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.22018713del', 'vcf': {'chr': 'chr12', 'ref': 'TC', 'pos': '22018712', 'alt': 'T'}}
        assert results['NM_020297.3:c.2199-1302del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.21865779del', 'vcf': {'chr': 'chr12', 'ref': 'TC', 'pos': '21865778', 'alt': 'T'}}
        assert results['NM_020297.3:c.2199-1302del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.22018713del', 'vcf': {'chr': '12', 'ref': 'TC', 'pos': '22018712', 'alt': 'T'}}
        assert results['NM_020297.3:c.2199-1302del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.21865779del', 'vcf': {'chr': '12', 'ref': 'TC', 'pos': '21865778', 'alt': 'T'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_005691.3:c.2199-1302del' in results.keys()
        assert results['NM_005691.3:c.2199-1302del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_005691.3:c.2199-1302del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_005691.3:c.2199-1302del']['alt_genomic_loci'] == []
        assert results['NM_005691.3:c.2199-1302del']['gene_symbol'] == 'ABCC9'
        assert results['NM_005691.3:c.2199-1302del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_005682.2(LRG_377p2):p.?', 'slr': 'NP_005682.2:p.?'}
        assert results['NM_005691.3:c.2199-1302del']['submitted_variant'] == '12-22018712-TC-T'
        assert results['NM_005691.3:c.2199-1302del']['genome_context_intronic_sequence'] == 'NC_000012.11(NM_005691.3):c.2199-1302del'
        assert results['NM_005691.3:c.2199-1302del']['hgvs_lrg_variant'] == ''
        assert results['NM_005691.3:c.2199-1302del']['hgvs_transcript_variant'] == 'NM_005691.3:c.2199-1302del'
        assert results['NM_005691.3:c.2199-1302del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_005691.3:c.2199-1302del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.22018713del', 'vcf': {'chr': 'chr12', 'ref': 'TC', 'pos': '22018712', 'alt': 'T'}}
        assert results['NM_005691.3:c.2199-1302del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.21865779del', 'vcf': {'chr': 'chr12', 'ref': 'TC', 'pos': '21865778', 'alt': 'T'}}
        assert results['NM_005691.3:c.2199-1302del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.22018713del', 'vcf': {'chr': '12', 'ref': 'TC', 'pos': '22018712', 'alt': 'T'}}
        assert results['NM_005691.3:c.2199-1302del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.21865779del', 'vcf': {'chr': '12', 'ref': 'TC', 'pos': '21865778', 'alt': 'T'}}

        assert 'NM_020297.2:c.2199-1302del' in results.keys()
        assert results['NM_020297.2:c.2199-1302del']['hgvs_lrg_transcript_variant'] == 'LRG_377t1:c.2199-1302del'
        assert results['NM_020297.2:c.2199-1302del']['refseqgene_context_intronic_sequence'] == 'NG_012819.1(NM_020297.2):c.2199-1302del'
        assert results['NM_020297.2:c.2199-1302del']['alt_genomic_loci'] == []
        assert results['NM_020297.2:c.2199-1302del']['gene_symbol'] == 'ABCC9'
        assert results['NM_020297.2:c.2199-1302del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_064693.2(LRG_377p1):p.?', 'slr': 'NP_064693.2:p.?'}
        assert results['NM_020297.2:c.2199-1302del']['submitted_variant'] == '12-22018712-TC-T'
        assert results['NM_020297.2:c.2199-1302del']['genome_context_intronic_sequence'] == 'NC_000012.11(NM_020297.2):c.2199-1302del'
        assert results['NM_020297.2:c.2199-1302del']['hgvs_lrg_variant'] == 'LRG_377:g.75916del'
        assert results['NM_020297.2:c.2199-1302del']['hgvs_transcript_variant'] == 'NM_020297.2:c.2199-1302del'
        assert results['NM_020297.2:c.2199-1302del']['hgvs_refseqgene_variant'] == 'NG_012819.1:g.75916del'
        assert results['NM_020297.2:c.2199-1302del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.22018713del', 'vcf': {'chr': 'chr12', 'ref': 'TC', 'pos': '22018712', 'alt': 'T'}}
        assert 'hg38' not in results['NM_020297.2:c.2199-1302del']['primary_assembly_loci'].keys()
        assert results['NM_020297.2:c.2199-1302del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.22018713del', 'vcf': {'chr': '12', 'ref': 'TC', 'pos': '22018712', 'alt': 'T'}}
        assert 'grch38' not in results['NM_020297.2:c.2199-1302del']['primary_assembly_loci'].keys()

        assert 'NM_005691.2:c.2199-1302del' in results.keys()
        assert results['NM_005691.2:c.2199-1302del']['hgvs_lrg_transcript_variant'] == 'LRG_377t2:c.2199-1302del'
        assert results['NM_005691.2:c.2199-1302del']['refseqgene_context_intronic_sequence'] == 'NG_012819.1(NM_005691.2):c.2199-1302del'
        assert results['NM_005691.2:c.2199-1302del']['alt_genomic_loci'] == []
        assert results['NM_005691.2:c.2199-1302del']['gene_symbol'] == 'ABCC9'
        assert results['NM_005691.2:c.2199-1302del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_005682.2(LRG_377p2):p.?', 'slr': 'NP_005682.2:p.?'}
        assert results['NM_005691.2:c.2199-1302del']['submitted_variant'] == '12-22018712-TC-T'
        assert results['NM_005691.2:c.2199-1302del']['genome_context_intronic_sequence'] == 'NC_000012.11(NM_005691.2):c.2199-1302del'
        assert results['NM_005691.2:c.2199-1302del']['hgvs_lrg_variant'] == 'LRG_377:g.75916del'
        assert results['NM_005691.2:c.2199-1302del']['hgvs_transcript_variant'] == 'NM_005691.2:c.2199-1302del'
        assert results['NM_005691.2:c.2199-1302del']['hgvs_refseqgene_variant'] == 'NG_012819.1:g.75916del'
        assert results['NM_005691.2:c.2199-1302del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.22018713del', 'vcf': {'chr': 'chr12', 'ref': 'TC', 'pos': '22018712', 'alt': 'T'}}
        assert 'hg38' not in results['NM_005691.2:c.2199-1302del']['primary_assembly_loci'].keys()
        assert results['NM_005691.2:c.2199-1302del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.22018713del', 'vcf': {'chr': '12', 'ref': 'TC', 'pos': '22018712', 'alt': 'T'}}
        assert 'grch38' not in results['NM_005691.2:c.2199-1302del']['primary_assembly_loci'].keys()


    def test_variant218(self):
        variant = '12-52912946-T-C'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000424.3:c.556-2A>G' in results.keys()
        assert results['NM_000424.3:c.556-2A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000424.3:c.556-2A>G']['refseqgene_context_intronic_sequence'] == 'NG_008297.1(NM_000424.3):c.556-2A>G'
        assert results['NM_000424.3:c.556-2A>G']['alt_genomic_loci'] == []
        assert results['NM_000424.3:c.556-2A>G']['gene_symbol'] == 'KRT5'
        assert results['NM_000424.3:c.556-2A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000415.2:p.?', 'slr': 'NP_000415.2:p.?'}
        assert results['NM_000424.3:c.556-2A>G']['submitted_variant'] == '12-52912946-T-C'
        assert results['NM_000424.3:c.556-2A>G']['genome_context_intronic_sequence'] == 'NC_000012.11(NM_000424.3):c.556-2A>G'
        assert results['NM_000424.3:c.556-2A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_000424.3:c.556-2A>G']['hgvs_transcript_variant'] == 'NM_000424.3:c.556-2A>G'
        assert results['NM_000424.3:c.556-2A>G']['hgvs_refseqgene_variant'] == 'NG_008297.1:g.6298A>G'
        assert results['NM_000424.3:c.556-2A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.52912946T>C', 'vcf': {'chr': 'chr12', 'ref': u'T', 'pos': '52912946', 'alt': u'C'}}
        assert results['NM_000424.3:c.556-2A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.52519162T>C', 'vcf': {'chr': 'chr12', 'ref': u'T', 'pos': '52519162', 'alt': u'C'}}
        assert results['NM_000424.3:c.556-2A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.52912946T>C', 'vcf': {'chr': '12', 'ref': u'T', 'pos': '52912946', 'alt': u'C'}}
        assert results['NM_000424.3:c.556-2A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.52519162T>C', 'vcf': {'chr': '12', 'ref': u'T', 'pos': '52519162', 'alt': u'C'}}


    def test_variant219(self):
        variant = '12-103234292-TC-T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001354304.1:c.1200del' in results.keys()
        assert results['NM_001354304.1:c.1200del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001354304.1:c.1200del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001354304.1:c.1200del']['alt_genomic_loci'] == []
        assert results['NM_001354304.1:c.1200del']['gene_symbol'] == 'PAH'
        assert results['NM_001354304.1:c.1200del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001341233.1:p.(Asn401ThrfsTer51)', 'slr': 'NP_001341233.1:p.(N401Tfs*51)'}
        assert results['NM_001354304.1:c.1200del']['submitted_variant'] == '12-103234292-TC-T'
        assert results['NM_001354304.1:c.1200del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001354304.1:c.1200del']['hgvs_lrg_variant'] == ''
        assert results['NM_001354304.1:c.1200del']['hgvs_transcript_variant'] == 'NM_001354304.1:c.1200del'
        assert results['NM_001354304.1:c.1200del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001354304.1:c.1200del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.103234294del', 'vcf': {'chr': 'chr12', 'ref': 'TC', 'pos': '103234292', 'alt': 'T'}}
        assert results['NM_001354304.1:c.1200del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.102840516del', 'vcf': {'chr': 'chr12', 'ref': 'TC', 'pos': '102840514', 'alt': 'T'}}
        assert results['NM_001354304.1:c.1200del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.103234294del', 'vcf': {'chr': '12', 'ref': 'TC', 'pos': '103234292', 'alt': 'T'}}
        assert results['NM_001354304.1:c.1200del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.102840516del', 'vcf': {'chr': '12', 'ref': 'TC', 'pos': '102840514', 'alt': 'T'}}

        assert 'NM_000277.2:c.1200del' in results.keys()
        assert results['NM_000277.2:c.1200del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000277.2:c.1200del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000277.2:c.1200del']['alt_genomic_loci'] == []
        assert results['NM_000277.2:c.1200del']['gene_symbol'] == 'PAH'
        assert results['NM_000277.2:c.1200del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000268.1:p.(Asn401ThrfsTer51)', 'slr': 'NP_000268.1:p.(N401Tfs*51)'}
        assert results['NM_000277.2:c.1200del']['submitted_variant'] == '12-103234292-TC-T'
        assert results['NM_000277.2:c.1200del']['genome_context_intronic_sequence'] == ''
        assert results['NM_000277.2:c.1200del']['hgvs_lrg_variant'] == ''
        assert results['NM_000277.2:c.1200del']['hgvs_transcript_variant'] == 'NM_000277.2:c.1200del'
        assert results['NM_000277.2:c.1200del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_000277.2:c.1200del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.103234294del', 'vcf': {'chr': 'chr12', 'ref': 'TC', 'pos': '103234292', 'alt': 'T'}}
        assert results['NM_000277.2:c.1200del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.102840516del', 'vcf': {'chr': 'chr12', 'ref': 'TC', 'pos': '102840514', 'alt': 'T'}}
        assert results['NM_000277.2:c.1200del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.103234294del', 'vcf': {'chr': '12', 'ref': 'TC', 'pos': '103234292', 'alt': 'T'}}
        assert results['NM_000277.2:c.1200del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.102840516del', 'vcf': {'chr': '12', 'ref': 'TC', 'pos': '102840514', 'alt': 'T'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_000277.1:c.1200del' in results.keys()
        assert results['NM_000277.1:c.1200del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000277.1:c.1200del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000277.1:c.1200del']['alt_genomic_loci'] == []
        assert results['NM_000277.1:c.1200del']['gene_symbol'] == 'PAH'
        assert results['NM_000277.1:c.1200del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000268.1:p.(Asn401ThrfsTer51)', 'slr': 'NP_000268.1:p.(N401Tfs*51)'}
        assert results['NM_000277.1:c.1200del']['submitted_variant'] == '12-103234292-TC-T'
        assert results['NM_000277.1:c.1200del']['genome_context_intronic_sequence'] == ''
        assert results['NM_000277.1:c.1200del']['hgvs_lrg_variant'] == ''
        assert results['NM_000277.1:c.1200del']['hgvs_transcript_variant'] == 'NM_000277.1:c.1200del'
        assert results['NM_000277.1:c.1200del']['hgvs_refseqgene_variant'] == 'NG_008690.1:g.82088del'
        assert results['NM_000277.1:c.1200del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.103234294del', 'vcf': {'chr': 'chr12', 'ref': 'TC', 'pos': '103234292', 'alt': 'T'}}
        assert results['NM_000277.1:c.1200del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.102840516del', 'vcf': {'chr': 'chr12', 'ref': 'TC', 'pos': '102840514', 'alt': 'T'}}
        assert results['NM_000277.1:c.1200del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.103234294del', 'vcf': {'chr': '12', 'ref': 'TC', 'pos': '103234292', 'alt': 'T'}}
        assert results['NM_000277.1:c.1200del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.102840516del', 'vcf': {'chr': '12', 'ref': 'TC', 'pos': '102840514', 'alt': 'T'}}


    def test_variant220(self):
        variant = '12-103311124-T-C'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001354304.1:c.-95-121A>G' in results.keys()
        assert results['NM_001354304.1:c.-95-121A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001354304.1:c.-95-121A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001354304.1:c.-95-121A>G']['alt_genomic_loci'] == []
        assert results['NM_001354304.1:c.-95-121A>G']['gene_symbol'] == 'PAH'
        assert results['NM_001354304.1:c.-95-121A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001341233.1:p.?', 'slr': 'NP_001341233.1:p.?'}
        assert results['NM_001354304.1:c.-95-121A>G']['submitted_variant'] == '12-103311124-T-C'
        assert results['NM_001354304.1:c.-95-121A>G']['genome_context_intronic_sequence'] == 'NC_000012.11(NM_001354304.1):c.-95-121A>G'
        assert results['NM_001354304.1:c.-95-121A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001354304.1:c.-95-121A>G']['hgvs_transcript_variant'] == 'NM_001354304.1:c.-95-121A>G'
        assert results['NM_001354304.1:c.-95-121A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001354304.1:c.-95-121A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.103311124T>C', 'vcf': {'chr': 'chr12', 'ref': u'T', 'pos': '103311124', 'alt': u'C'}}
        assert results['NM_001354304.1:c.-95-121A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.102917346T>C', 'vcf': {'chr': 'chr12', 'ref': u'T', 'pos': '102917346', 'alt': u'C'}}
        assert results['NM_001354304.1:c.-95-121A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.103311124T>C', 'vcf': {'chr': '12', 'ref': u'T', 'pos': '103311124', 'alt': u'C'}}
        assert results['NM_001354304.1:c.-95-121A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.102917346T>C', 'vcf': {'chr': '12', 'ref': u'T', 'pos': '102917346', 'alt': u'C'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_000277.2:c.-216A>G' in results.keys()
        assert results['NM_000277.2:c.-216A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000277.2:c.-216A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000277.2:c.-216A>G']['alt_genomic_loci'] == []
        assert results['NM_000277.2:c.-216A>G']['gene_symbol'] == 'PAH'
        assert results['NM_000277.2:c.-216A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000268.1:p.?', 'slr': 'NP_000268.1:p.?'}
        assert results['NM_000277.2:c.-216A>G']['submitted_variant'] == '12-103311124-T-C'
        assert results['NM_000277.2:c.-216A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_000277.2:c.-216A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_000277.2:c.-216A>G']['hgvs_transcript_variant'] == 'NM_000277.2:c.-216A>G'
        assert results['NM_000277.2:c.-216A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_000277.2:c.-216A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.103311124T>C', 'vcf': {'chr': 'chr12', 'ref': u'T', 'pos': '103311124', 'alt': u'C'}}
        assert results['NM_000277.2:c.-216A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.102917346T>C', 'vcf': {'chr': 'chr12', 'ref': u'T', 'pos': '102917346', 'alt': u'C'}}
        assert results['NM_000277.2:c.-216A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.103311124T>C', 'vcf': {'chr': '12', 'ref': u'T', 'pos': '103311124', 'alt': u'C'}}
        assert results['NM_000277.2:c.-216A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.102917346T>C', 'vcf': {'chr': '12', 'ref': u'T', 'pos': '102917346', 'alt': u'C'}}

        assert 'NM_000277.1:c.-215A>G' in results.keys()
        assert results['NM_000277.1:c.-215A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000277.1:c.-215A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000277.1:c.-215A>G']['alt_genomic_loci'] == []
        assert results['NM_000277.1:c.-215A>G']['gene_symbol'] == 'PAH'
        assert results['NM_000277.1:c.-215A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000268.1:p.?', 'slr': 'NP_000268.1:p.?'}
        assert results['NM_000277.1:c.-215A>G']['submitted_variant'] == '12-103311124-T-C'
        assert results['NM_000277.1:c.-215A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_000277.1:c.-215A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_000277.1:c.-215A>G']['hgvs_transcript_variant'] == 'NM_000277.1:c.-215A>G'
        assert results['NM_000277.1:c.-215A>G']['hgvs_refseqgene_variant'] == 'NG_008690.1:g.5258A>G'
        assert results['NM_000277.1:c.-215A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.103311124T>C', 'vcf': {'chr': 'chr12', 'ref': u'T', 'pos': '103311124', 'alt': u'C'}}
        assert results['NM_000277.1:c.-215A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.102917346T>C', 'vcf': {'chr': 'chr12', 'ref': u'T', 'pos': '102917346', 'alt': u'C'}}
        assert results['NM_000277.1:c.-215A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.103311124T>C', 'vcf': {'chr': '12', 'ref': u'T', 'pos': '103311124', 'alt': u'C'}}
        assert results['NM_000277.1:c.-215A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.102917346T>C', 'vcf': {'chr': '12', 'ref': u'T', 'pos': '102917346', 'alt': u'C'}}


    def test_variant221(self):
        variant = '12-111064166-G-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001319681.1:c.-366-1G>A' in results.keys()
        assert results['NM_001319681.1:c.-366-1G>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001319681.1:c.-366-1G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001319681.1:c.-366-1G>A']['alt_genomic_loci'] == []
        assert results['NM_001319681.1:c.-366-1G>A']['gene_symbol'] == 'TCTN1'
        assert results['NM_001319681.1:c.-366-1G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001306610.1:p.?', 'slr': 'NP_001306610.1:p.?'}
        assert results['NM_001319681.1:c.-366-1G>A']['submitted_variant'] == '12-111064166-G-A'
        assert results['NM_001319681.1:c.-366-1G>A']['genome_context_intronic_sequence'] == 'NC_000012.11(NM_001319681.1):c.-366-1G>A'
        assert results['NM_001319681.1:c.-366-1G>A']['hgvs_lrg_variant'] == ''
        assert results['NM_001319681.1:c.-366-1G>A']['hgvs_transcript_variant'] == 'NM_001319681.1:c.-366-1G>A'
        assert results['NM_001319681.1:c.-366-1G>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001319681.1:c.-366-1G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.111064166G>A', 'vcf': {'chr': 'chr12', 'ref': 'G', 'pos': '111064166', 'alt': 'A'}}
        assert results['NM_001319681.1:c.-366-1G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.110626361G>A', 'vcf': {'chr': 'chr12', 'ref': 'G', 'pos': '110626361', 'alt': 'A'}}
        assert results['NM_001319681.1:c.-366-1G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.111064166G>A', 'vcf': {'chr': '12', 'ref': 'G', 'pos': '111064166', 'alt': 'A'}}
        assert results['NM_001319681.1:c.-366-1G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.110626361G>A', 'vcf': {'chr': '12', 'ref': 'G', 'pos': '110626361', 'alt': 'A'}}

        assert 'NM_001319680.1:c.342-1G>A' in results.keys()
        assert results['NM_001319680.1:c.342-1G>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001319680.1:c.342-1G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001319680.1:c.342-1G>A']['alt_genomic_loci'] == []
        assert results['NM_001319680.1:c.342-1G>A']['gene_symbol'] == 'TCTN1'
        assert results['NM_001319680.1:c.342-1G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001306609.1:p.?', 'slr': 'NP_001306609.1:p.?'}
        assert results['NM_001319680.1:c.342-1G>A']['submitted_variant'] == '12-111064166-G-A'
        assert results['NM_001319680.1:c.342-1G>A']['genome_context_intronic_sequence'] == 'NC_000012.11(NM_001319680.1):c.342-1G>A'
        assert results['NM_001319680.1:c.342-1G>A']['hgvs_lrg_variant'] == ''
        assert results['NM_001319680.1:c.342-1G>A']['hgvs_transcript_variant'] == 'NM_001319680.1:c.342-1G>A'
        assert results['NM_001319680.1:c.342-1G>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001319680.1:c.342-1G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.111064166G>A', 'vcf': {'chr': 'chr12', 'ref': 'G', 'pos': '111064166', 'alt': 'A'}}
        assert results['NM_001319680.1:c.342-1G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.110626361G>A', 'vcf': {'chr': 'chr12', 'ref': 'G', 'pos': '110626361', 'alt': 'A'}}
        assert results['NM_001319680.1:c.342-1G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.111064166G>A', 'vcf': {'chr': '12', 'ref': 'G', 'pos': '111064166', 'alt': 'A'}}
        assert results['NM_001319680.1:c.342-1G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.110626361G>A', 'vcf': {'chr': '12', 'ref': 'G', 'pos': '110626361', 'alt': 'A'}}

        assert 'NM_001082538.2:c.342-1G>A' in results.keys()
        assert results['NM_001082538.2:c.342-1G>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001082538.2:c.342-1G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001082538.2:c.342-1G>A']['alt_genomic_loci'] == []
        assert results['NM_001082538.2:c.342-1G>A']['gene_symbol'] == 'TCTN1'
        assert results['NM_001082538.2:c.342-1G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001076007.1:p.?', 'slr': 'NP_001076007.1:p.?'}
        assert results['NM_001082538.2:c.342-1G>A']['submitted_variant'] == '12-111064166-G-A'
        assert results['NM_001082538.2:c.342-1G>A']['genome_context_intronic_sequence'] == 'NC_000012.11(NM_001082538.2):c.342-1G>A'
        assert results['NM_001082538.2:c.342-1G>A']['hgvs_lrg_variant'] == ''
        assert results['NM_001082538.2:c.342-1G>A']['hgvs_transcript_variant'] == 'NM_001082538.2:c.342-1G>A'
        assert results['NM_001082538.2:c.342-1G>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001082538.2:c.342-1G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.111064166G>A', 'vcf': {'chr': 'chr12', 'ref': 'G', 'pos': '111064166', 'alt': 'A'}}
        assert results['NM_001082538.2:c.342-1G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.110626361G>A', 'vcf': {'chr': 'chr12', 'ref': 'G', 'pos': '110626361', 'alt': 'A'}}
        assert results['NM_001082538.2:c.342-1G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.111064166G>A', 'vcf': {'chr': '12', 'ref': 'G', 'pos': '111064166', 'alt': 'A'}}
        assert results['NM_001082538.2:c.342-1G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.110626361G>A', 'vcf': {'chr': '12', 'ref': 'G', 'pos': '110626361', 'alt': 'A'}}

        assert 'NM_001173976.1:c.162-1G>A' in results.keys()
        assert results['NM_001173976.1:c.162-1G>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001173976.1:c.162-1G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001173976.1:c.162-1G>A']['alt_genomic_loci'] == []
        assert results['NM_001173976.1:c.162-1G>A']['gene_symbol'] == 'TCTN1'
        assert results['NM_001173976.1:c.162-1G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001167447.1:p.?', 'slr': 'NP_001167447.1:p.?'}
        assert results['NM_001173976.1:c.162-1G>A']['submitted_variant'] == '12-111064166-G-A'
        assert results['NM_001173976.1:c.162-1G>A']['genome_context_intronic_sequence'] == 'NC_000012.11(NM_001173976.1):c.162-1G>A'
        assert results['NM_001173976.1:c.162-1G>A']['hgvs_lrg_variant'] == ''
        assert results['NM_001173976.1:c.162-1G>A']['hgvs_transcript_variant'] == 'NM_001173976.1:c.162-1G>A'
        assert results['NM_001173976.1:c.162-1G>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001173976.1:c.162-1G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.111064166G>A', 'vcf': {'chr': 'chr12', 'ref': 'G', 'pos': '111064166', 'alt': 'A'}}
        assert results['NM_001173976.1:c.162-1G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.110626361G>A', 'vcf': {'chr': 'chr12', 'ref': 'G', 'pos': '110626361', 'alt': 'A'}}
        assert results['NM_001173976.1:c.162-1G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.111064166G>A', 'vcf': {'chr': '12', 'ref': 'G', 'pos': '111064166', 'alt': 'A'}}
        assert results['NM_001173976.1:c.162-1G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.110626361G>A', 'vcf': {'chr': '12', 'ref': 'G', 'pos': '110626361', 'alt': 'A'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001082537.2:c.342-1G>A' in results.keys()
        assert results['NM_001082537.2:c.342-1G>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001082537.2:c.342-1G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001082537.2:c.342-1G>A']['alt_genomic_loci'] == []
        assert results['NM_001082537.2:c.342-1G>A']['gene_symbol'] == 'TCTN1'
        assert results['NM_001082537.2:c.342-1G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001076006.1:p.?', 'slr': 'NP_001076006.1:p.?'}
        assert results['NM_001082537.2:c.342-1G>A']['submitted_variant'] == '12-111064166-G-A'
        assert results['NM_001082537.2:c.342-1G>A']['genome_context_intronic_sequence'] == 'NC_000012.11(NM_001082537.2):c.342-1G>A'
        assert results['NM_001082537.2:c.342-1G>A']['hgvs_lrg_variant'] == ''
        assert results['NM_001082537.2:c.342-1G>A']['hgvs_transcript_variant'] == 'NM_001082537.2:c.342-1G>A'
        assert results['NM_001082537.2:c.342-1G>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001082537.2:c.342-1G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.111064166G>A', 'vcf': {'chr': 'chr12', 'ref': 'G', 'pos': '111064166', 'alt': 'A'}}
        assert results['NM_001082537.2:c.342-1G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.110626361G>A', 'vcf': {'chr': 'chr12', 'ref': 'G', 'pos': '110626361', 'alt': 'A'}}
        assert results['NM_001082537.2:c.342-1G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.111064166G>A', 'vcf': {'chr': '12', 'ref': 'G', 'pos': '111064166', 'alt': 'A'}}
        assert results['NM_001082537.2:c.342-1G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.110626361G>A', 'vcf': {'chr': '12', 'ref': 'G', 'pos': '110626361', 'alt': 'A'}}

        assert 'NR_135088.1:n.559-1G>A' in results.keys()
        assert results['NR_135088.1:n.559-1G>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_135088.1:n.559-1G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_135088.1:n.559-1G>A']['alt_genomic_loci'] == []
        assert results['NR_135088.1:n.559-1G>A']['gene_symbol'] == 'TCTN1'
        assert results['NR_135088.1:n.559-1G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_135088.1:n.559-1G>A']['submitted_variant'] == '12-111064166-G-A'
        assert results['NR_135088.1:n.559-1G>A']['genome_context_intronic_sequence'] == 'NC_000012.11(NR_135088.1):c.559-1G>A'
        assert results['NR_135088.1:n.559-1G>A']['hgvs_lrg_variant'] == ''
        assert results['NR_135088.1:n.559-1G>A']['hgvs_transcript_variant'] == 'NR_135088.1:n.559-1G>A'
        assert results['NR_135088.1:n.559-1G>A']['hgvs_refseqgene_variant'] == ''
        assert results['NR_135088.1:n.559-1G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.111064166G>A', 'vcf': {'chr': 'chr12', 'ref': 'G', 'pos': '111064166', 'alt': 'A'}}
        assert results['NR_135088.1:n.559-1G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.110626361G>A', 'vcf': {'chr': 'chr12', 'ref': 'G', 'pos': '110626361', 'alt': 'A'}}
        assert results['NR_135088.1:n.559-1G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.111064166G>A', 'vcf': {'chr': '12', 'ref': 'G', 'pos': '111064166', 'alt': 'A'}}
        assert results['NR_135088.1:n.559-1G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.110626361G>A', 'vcf': {'chr': '12', 'ref': 'G', 'pos': '110626361', 'alt': 'A'}}

        assert 'NM_024549.5:c.342-1G>A' in results.keys()
        assert results['NM_024549.5:c.342-1G>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_024549.5:c.342-1G>A']['refseqgene_context_intronic_sequence'] == 'NG_030381.1(NM_024549.5):c.342-1G>A'
        assert results['NM_024549.5:c.342-1G>A']['alt_genomic_loci'] == []
        assert results['NM_024549.5:c.342-1G>A']['gene_symbol'] == 'TCTN1'
        assert results['NM_024549.5:c.342-1G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_078825.2:p.?', 'slr': 'NP_078825.2:p.?'}
        assert results['NM_024549.5:c.342-1G>A']['submitted_variant'] == '12-111064166-G-A'
        assert results['NM_024549.5:c.342-1G>A']['genome_context_intronic_sequence'] == 'NC_000012.11(NM_024549.5):c.342-1G>A'
        assert results['NM_024549.5:c.342-1G>A']['hgvs_lrg_variant'] == ''
        assert results['NM_024549.5:c.342-1G>A']['hgvs_transcript_variant'] == 'NM_024549.5:c.342-1G>A'
        assert results['NM_024549.5:c.342-1G>A']['hgvs_refseqgene_variant'] == 'NG_030381.1:g.17335G>A'
        assert results['NM_024549.5:c.342-1G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.111064166G>A', 'vcf': {'chr': 'chr12', 'ref': 'G', 'pos': '111064166', 'alt': 'A'}}
        assert results['NM_024549.5:c.342-1G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.110626361G>A', 'vcf': {'chr': 'chr12', 'ref': 'G', 'pos': '110626361', 'alt': 'A'}}
        assert results['NM_024549.5:c.342-1G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.111064166G>A', 'vcf': {'chr': '12', 'ref': 'G', 'pos': '111064166', 'alt': 'A'}}
        assert results['NM_024549.5:c.342-1G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.110626361G>A', 'vcf': {'chr': '12', 'ref': 'G', 'pos': '110626361', 'alt': 'A'}}

        assert 'NM_001319682.1:c.174-1G>A' in results.keys()
        assert results['NM_001319682.1:c.174-1G>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001319682.1:c.174-1G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001319682.1:c.174-1G>A']['alt_genomic_loci'] == []
        assert results['NM_001319682.1:c.174-1G>A']['gene_symbol'] == 'TCTN1'
        assert results['NM_001319682.1:c.174-1G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001306611.1:p.?', 'slr': 'NP_001306611.1:p.?'}
        assert results['NM_001319682.1:c.174-1G>A']['submitted_variant'] == '12-111064166-G-A'
        assert results['NM_001319682.1:c.174-1G>A']['genome_context_intronic_sequence'] == 'NC_000012.11(NM_001319682.1):c.174-1G>A'
        assert results['NM_001319682.1:c.174-1G>A']['hgvs_lrg_variant'] == ''
        assert results['NM_001319682.1:c.174-1G>A']['hgvs_transcript_variant'] == 'NM_001319682.1:c.174-1G>A'
        assert results['NM_001319682.1:c.174-1G>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001319682.1:c.174-1G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.111064166G>A', 'vcf': {'chr': 'chr12', 'ref': 'G', 'pos': '111064166', 'alt': 'A'}}
        assert results['NM_001319682.1:c.174-1G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.110626361G>A', 'vcf': {'chr': 'chr12', 'ref': 'G', 'pos': '110626361', 'alt': 'A'}}
        assert results['NM_001319682.1:c.174-1G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.111064166G>A', 'vcf': {'chr': '12', 'ref': 'G', 'pos': '111064166', 'alt': 'A'}}
        assert results['NM_001319682.1:c.174-1G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.110626361G>A', 'vcf': {'chr': '12', 'ref': 'G', 'pos': '110626361', 'alt': 'A'}}

        assert 'NM_001173975.1:c.174-1G>A' in results.keys()
        assert results['NM_001173975.1:c.174-1G>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001173975.1:c.174-1G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001173975.1:c.174-1G>A']['alt_genomic_loci'] == []
        assert results['NM_001173975.1:c.174-1G>A']['gene_symbol'] == 'TCTN1'
        assert results['NM_001173975.1:c.174-1G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001167446.1:p.?', 'slr': 'NP_001167446.1:p.?'}
        assert results['NM_001173975.1:c.174-1G>A']['submitted_variant'] == '12-111064166-G-A'
        assert results['NM_001173975.1:c.174-1G>A']['genome_context_intronic_sequence'] == 'NC_000012.11(NM_001173975.1):c.174-1G>A'
        assert results['NM_001173975.1:c.174-1G>A']['hgvs_lrg_variant'] == ''
        assert results['NM_001173975.1:c.174-1G>A']['hgvs_transcript_variant'] == 'NM_001173975.1:c.174-1G>A'
        assert results['NM_001173975.1:c.174-1G>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001173975.1:c.174-1G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.111064166G>A', 'vcf': {'chr': 'chr12', 'ref': 'G', 'pos': '111064166', 'alt': 'A'}}
        assert 'hg38' not in results['NM_001173975.1:c.174-1G>A']['primary_assembly_loci'].keys()
        assert results['NM_001173975.1:c.174-1G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.111064166G>A', 'vcf': {'chr': '12', 'ref': 'G', 'pos': '111064166', 'alt': 'A'}}
        assert 'grch38' not in results['NM_001173975.1:c.174-1G>A']['primary_assembly_loci'].keys()

        assert 'NM_001173975.2:c.174-1G>A' in results.keys()
        assert results['NM_001173975.2:c.174-1G>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001173975.2:c.174-1G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001173975.2:c.174-1G>A']['alt_genomic_loci'] == []
        assert results['NM_001173975.2:c.174-1G>A']['gene_symbol'] == 'TCTN1'
        assert results['NM_001173975.2:c.174-1G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001167446.1:p.?', 'slr': 'NP_001167446.1:p.?'}
        assert results['NM_001173975.2:c.174-1G>A']['submitted_variant'] == '12-111064166-G-A'
        assert results['NM_001173975.2:c.174-1G>A']['genome_context_intronic_sequence'] == 'NC_000012.11(NM_001173975.2):c.174-1G>A'
        assert results['NM_001173975.2:c.174-1G>A']['hgvs_lrg_variant'] == ''
        assert results['NM_001173975.2:c.174-1G>A']['hgvs_transcript_variant'] == 'NM_001173975.2:c.174-1G>A'
        assert results['NM_001173975.2:c.174-1G>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001173975.2:c.174-1G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.111064166G>A', 'vcf': {'chr': 'chr12', 'ref': 'G', 'pos': '111064166', 'alt': 'A'}}
        assert results['NM_001173975.2:c.174-1G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.110626361G>A', 'vcf': {'chr': 'chr12', 'ref': 'G', 'pos': '110626361', 'alt': 'A'}}
        assert results['NM_001173975.2:c.174-1G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.111064166G>A', 'vcf': {'chr': '12', 'ref': 'G', 'pos': '111064166', 'alt': 'A'}}
        assert results['NM_001173975.2:c.174-1G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.110626361G>A', 'vcf': {'chr': '12', 'ref': 'G', 'pos': '110626361', 'alt': 'A'}}


    def test_variant222(self):
        variant = '12-123738430-CA-C'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001194995.1:c.210del' in results.keys()
        assert results['NM_001194995.1:c.210del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001194995.1:c.210del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001194995.1:c.210del']['alt_genomic_loci'] == []
        assert results['NM_001194995.1:c.210del']['gene_symbol'] == 'C12orf65'
        assert results['NM_001194995.1:c.210del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001181924.1:p.(Gly72AlafsTer13)', 'slr': 'NP_001181924.1:p.(G72Afs*13)'}
        assert results['NM_001194995.1:c.210del']['submitted_variant'] == '12-123738430-CA-C'
        assert results['NM_001194995.1:c.210del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001194995.1:c.210del']['hgvs_lrg_variant'] == ''
        assert results['NM_001194995.1:c.210del']['hgvs_transcript_variant'] == 'NM_001194995.1:c.210del'
        assert results['NM_001194995.1:c.210del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001194995.1:c.210del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.123738431del', 'vcf': {'chr': 'chr12', 'ref': 'CA', 'pos': '123738430', 'alt': 'C'}}
        assert results['NM_001194995.1:c.210del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.123253884del', 'vcf': {'chr': 'chr12', 'ref': 'CA', 'pos': '123253883', 'alt': 'C'}}
        assert results['NM_001194995.1:c.210del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.123738431del', 'vcf': {'chr': '12', 'ref': 'CA', 'pos': '123738430', 'alt': 'C'}}
        assert results['NM_001194995.1:c.210del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.123253884del', 'vcf': {'chr': '12', 'ref': 'CA', 'pos': '123253883', 'alt': 'C'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_152269.4:c.210del' in results.keys()
        assert results['NM_152269.4:c.210del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_152269.4:c.210del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_152269.4:c.210del']['alt_genomic_loci'] == []
        assert results['NM_152269.4:c.210del']['gene_symbol'] == 'C12orf65'
        assert results['NM_152269.4:c.210del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_689482.1:p.(Gly72AlafsTer13)', 'slr': 'NP_689482.1:p.(G72Afs*13)'}
        assert results['NM_152269.4:c.210del']['submitted_variant'] == '12-123738430-CA-C'
        assert results['NM_152269.4:c.210del']['genome_context_intronic_sequence'] == ''
        assert results['NM_152269.4:c.210del']['hgvs_lrg_variant'] == ''
        assert results['NM_152269.4:c.210del']['hgvs_transcript_variant'] == 'NM_152269.4:c.210del'
        assert results['NM_152269.4:c.210del']['hgvs_refseqgene_variant'] == 'NG_027517.1:g.25588del'
        assert results['NM_152269.4:c.210del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.123738431del', 'vcf': {'chr': 'chr12', 'ref': 'CA', 'pos': '123738430', 'alt': 'C'}}
        assert results['NM_152269.4:c.210del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.123253884del', 'vcf': {'chr': 'chr12', 'ref': 'CA', 'pos': '123253883', 'alt': 'C'}}
        assert results['NM_152269.4:c.210del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.123738431del', 'vcf': {'chr': '12', 'ref': 'CA', 'pos': '123738430', 'alt': 'C'}}
        assert results['NM_152269.4:c.210del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.123253884del', 'vcf': {'chr': '12', 'ref': 'CA', 'pos': '123253883', 'alt': 'C'}}

        assert 'NM_001143905.2:c.210del' in results.keys()
        assert results['NM_001143905.2:c.210del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001143905.2:c.210del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001143905.2:c.210del']['alt_genomic_loci'] == []
        assert results['NM_001143905.2:c.210del']['gene_symbol'] == 'C12orf65'
        assert results['NM_001143905.2:c.210del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001137377.1:p.(Gly72AlafsTer13)', 'slr': 'NP_001137377.1:p.(G72Afs*13)'}
        assert results['NM_001143905.2:c.210del']['submitted_variant'] == '12-123738430-CA-C'
        assert results['NM_001143905.2:c.210del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001143905.2:c.210del']['hgvs_lrg_variant'] == ''
        assert results['NM_001143905.2:c.210del']['hgvs_transcript_variant'] == 'NM_001143905.2:c.210del'
        assert results['NM_001143905.2:c.210del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001143905.2:c.210del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000012.11:g.123738431del', 'vcf': {'chr': 'chr12', 'ref': 'CA', 'pos': '123738430', 'alt': 'C'}}
        assert results['NM_001143905.2:c.210del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000012.12:g.123253884del', 'vcf': {'chr': 'chr12', 'ref': 'CA', 'pos': '123253883', 'alt': 'C'}}
        assert results['NM_001143905.2:c.210del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000012.11:g.123738431del', 'vcf': {'chr': '12', 'ref': 'CA', 'pos': '123738430', 'alt': 'C'}}
        assert results['NM_001143905.2:c.210del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000012.12:g.123253884del', 'vcf': {'chr': '12', 'ref': 'CA', 'pos': '123253883', 'alt': 'C'}}


    def test_variant223(self):
        variant = '13-31789169-CT-C'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_194318.3:c.71-5del' in results.keys()
        assert results['NM_194318.3:c.71-5del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_194318.3:c.71-5del']['refseqgene_context_intronic_sequence'] == 'NG_011732.1(NM_194318.3):c.71-5del'
        assert results['NM_194318.3:c.71-5del']['alt_genomic_loci'] == []
        assert results['NM_194318.3:c.71-5del']['gene_symbol'] == 'B3GLCT'
        assert results['NM_194318.3:c.71-5del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_919299.3:p.?', 'slr': 'NP_919299.3:p.?'}
        assert results['NM_194318.3:c.71-5del']['submitted_variant'] == '13-31789169-CT-C'
        assert results['NM_194318.3:c.71-5del']['genome_context_intronic_sequence'] == 'NC_000013.10(NM_194318.3):c.71-5del'
        assert results['NM_194318.3:c.71-5del']['hgvs_lrg_variant'] == ''
        assert results['NM_194318.3:c.71-5del']['hgvs_transcript_variant'] == 'NM_194318.3:c.71-5del'
        assert results['NM_194318.3:c.71-5del']['hgvs_refseqgene_variant'] == 'NG_011732.1:g.20072del'
        assert results['NM_194318.3:c.71-5del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000013.10:g.31789183del', 'vcf': {'chr': 'chr13', 'ref': 'CT', 'pos': '31789169', 'alt': 'C'}}
        assert results['NM_194318.3:c.71-5del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000013.11:g.31215046del', 'vcf': {'chr': 'chr13', 'ref': 'CT', 'pos': '31215032', 'alt': 'C'}}
        assert results['NM_194318.3:c.71-5del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000013.10:g.31789183del', 'vcf': {'chr': '13', 'ref': 'CT', 'pos': '31789169', 'alt': 'C'}}
        assert results['NM_194318.3:c.71-5del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000013.11:g.31215046del', 'vcf': {'chr': '13', 'ref': 'CT', 'pos': '31215032', 'alt': 'C'}}


    def test_variant224(self):
        variant = '14-62187287-G-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001530.3:c.223G>A' in results.keys()
        assert results['NM_001530.3:c.223G>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001530.3:c.223G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001530.3:c.223G>A']['alt_genomic_loci'] == []
        assert results['NM_001530.3:c.223G>A']['gene_symbol'] == 'HIF1A'
        assert results['NM_001530.3:c.223G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001521.1:p.(Ala75Thr)', 'slr': 'NP_001521.1:p.(A75T)'}
        assert results['NM_001530.3:c.223G>A']['submitted_variant'] == '14-62187287-G-A'
        assert results['NM_001530.3:c.223G>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_001530.3:c.223G>A']['hgvs_lrg_variant'] == ''
        assert results['NM_001530.3:c.223G>A']['hgvs_transcript_variant'] == 'NM_001530.3:c.223G>A'
        assert results['NM_001530.3:c.223G>A']['hgvs_refseqgene_variant'] == 'NG_029606.1:g.30169G>A'
        assert results['NM_001530.3:c.223G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000014.8:g.62187287G>A', 'vcf': {'chr': 'chr14', 'ref': 'G', 'pos': '62187287', 'alt': 'A'}}
        assert results['NM_001530.3:c.223G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000014.9:g.61720569G>A', 'vcf': {'chr': 'chr14', 'ref': 'G', 'pos': '61720569', 'alt': 'A'}}
        assert results['NM_001530.3:c.223G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000014.8:g.62187287G>A', 'vcf': {'chr': '14', 'ref': 'G', 'pos': '62187287', 'alt': 'A'}}
        assert results['NM_001530.3:c.223G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000014.9:g.61720569G>A', 'vcf': {'chr': '14', 'ref': 'G', 'pos': '61720569', 'alt': 'A'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001243084.1:c.295G>A' in results.keys()
        assert results['NM_001243084.1:c.295G>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001243084.1:c.295G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001243084.1:c.295G>A']['alt_genomic_loci'] == []
        assert results['NM_001243084.1:c.295G>A']['gene_symbol'] == 'HIF1A'
        assert results['NM_001243084.1:c.295G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001230013.1:p.(Ala99Thr)', 'slr': 'NP_001230013.1:p.(A99T)'}
        assert results['NM_001243084.1:c.295G>A']['submitted_variant'] == '14-62187287-G-A'
        assert results['NM_001243084.1:c.295G>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_001243084.1:c.295G>A']['hgvs_lrg_variant'] == ''
        assert results['NM_001243084.1:c.295G>A']['hgvs_transcript_variant'] == 'NM_001243084.1:c.295G>A'
        assert results['NM_001243084.1:c.295G>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001243084.1:c.295G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000014.8:g.62187287G>A', 'vcf': {'chr': 'chr14', 'ref': 'G', 'pos': '62187287', 'alt': 'A'}}
        assert results['NM_001243084.1:c.295G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000014.9:g.61720569G>A', 'vcf': {'chr': 'chr14', 'ref': 'G', 'pos': '61720569', 'alt': 'A'}}
        assert results['NM_001243084.1:c.295G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000014.8:g.62187287G>A', 'vcf': {'chr': '14', 'ref': 'G', 'pos': '62187287', 'alt': 'A'}}
        assert results['NM_001243084.1:c.295G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000014.9:g.61720569G>A', 'vcf': {'chr': '14', 'ref': 'G', 'pos': '61720569', 'alt': 'A'}}

        assert 'NR_144368.1:n.214-3552C>T' in results.keys()
        assert results['NR_144368.1:n.214-3552C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_144368.1:n.214-3552C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_144368.1:n.214-3552C>T']['alt_genomic_loci'] == []
        assert results['NR_144368.1:n.214-3552C>T']['gene_symbol'] == 'LOC105370526'
        assert results['NR_144368.1:n.214-3552C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_144368.1:n.214-3552C>T']['submitted_variant'] == '14-62187287-G-A'
        assert results['NR_144368.1:n.214-3552C>T']['genome_context_intronic_sequence'] == 'NC_000014.8(NR_144368.1):c.214-3552C>T'
        assert results['NR_144368.1:n.214-3552C>T']['hgvs_lrg_variant'] == ''
        assert results['NR_144368.1:n.214-3552C>T']['hgvs_transcript_variant'] == 'NR_144368.1:n.214-3552C>T'
        assert results['NR_144368.1:n.214-3552C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NR_144368.1:n.214-3552C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000014.8:g.62187287G>A', 'vcf': {'chr': 'chr14', 'ref': u'G', 'pos': '62187287', 'alt': u'A'}}
        assert results['NR_144368.1:n.214-3552C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000014.9:g.61720569G>A', 'vcf': {'chr': 'chr14', 'ref': u'G', 'pos': '61720569', 'alt': u'A'}}
        assert results['NR_144368.1:n.214-3552C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000014.8:g.62187287G>A', 'vcf': {'chr': '14', 'ref': u'G', 'pos': '62187287', 'alt': u'A'}}
        assert results['NR_144368.1:n.214-3552C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000014.9:g.61720569G>A', 'vcf': {'chr': '14', 'ref': u'G', 'pos': '61720569', 'alt': u'A'}}

        assert 'NM_181054.2:c.223G>A' in results.keys()
        assert results['NM_181054.2:c.223G>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_181054.2:c.223G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_181054.2:c.223G>A']['alt_genomic_loci'] == []
        assert results['NM_181054.2:c.223G>A']['gene_symbol'] == 'HIF1A'
        assert results['NM_181054.2:c.223G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_851397.1:p.(Ala75Thr)', 'slr': 'NP_851397.1:p.(A75T)'}
        assert results['NM_181054.2:c.223G>A']['submitted_variant'] == '14-62187287-G-A'
        assert results['NM_181054.2:c.223G>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_181054.2:c.223G>A']['hgvs_lrg_variant'] == ''
        assert results['NM_181054.2:c.223G>A']['hgvs_transcript_variant'] == 'NM_181054.2:c.223G>A'
        assert results['NM_181054.2:c.223G>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_181054.2:c.223G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000014.8:g.62187287G>A', 'vcf': {'chr': 'chr14', 'ref': 'G', 'pos': '62187287', 'alt': 'A'}}
        assert results['NM_181054.2:c.223G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000014.9:g.61720569G>A', 'vcf': {'chr': 'chr14', 'ref': 'G', 'pos': '61720569', 'alt': 'A'}}
        assert results['NM_181054.2:c.223G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000014.8:g.62187287G>A', 'vcf': {'chr': '14', 'ref': 'G', 'pos': '62187287', 'alt': 'A'}}
        assert results['NM_181054.2:c.223G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000014.9:g.61720569G>A', 'vcf': {'chr': '14', 'ref': 'G', 'pos': '61720569', 'alt': 'A'}}


    def test_variant225(self):
        variant = '14-62188231-TT-GA'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_001243084.1:c.303_304delinsGA' in results.keys()
        assert results['NM_001243084.1:c.303_304delinsGA']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001243084.1:c.303_304delinsGA']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001243084.1:c.303_304delinsGA']['alt_genomic_loci'] == []
        assert results['NM_001243084.1:c.303_304delinsGA']['gene_symbol'] == 'HIF1A'
        assert results['NM_001243084.1:c.303_304delinsGA']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001230013.1:p.(Asp101_Leu102delinsGluMet)', 'slr': 'NP_001230013.1:p.(D101_L102delinsEM)'}
        assert results['NM_001243084.1:c.303_304delinsGA']['submitted_variant'] == '14-62188231-TT-GA'
        assert results['NM_001243084.1:c.303_304delinsGA']['genome_context_intronic_sequence'] == ''
        assert results['NM_001243084.1:c.303_304delinsGA']['hgvs_lrg_variant'] == ''
        assert results['NM_001243084.1:c.303_304delinsGA']['hgvs_transcript_variant'] == 'NM_001243084.1:c.303_304delinsGA'
        assert results['NM_001243084.1:c.303_304delinsGA']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001243084.1:c.303_304delinsGA']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000014.8:g.62188231_62188232delinsGA', 'vcf': {'chr': 'chr14', 'ref': 'TT', 'pos': '62188231', 'alt': 'GA'}}
        assert results['NM_001243084.1:c.303_304delinsGA']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000014.9:g.61721513_61721514delinsGA', 'vcf': {'chr': 'chr14', 'ref': 'TT', 'pos': '61721513', 'alt': 'GA'}}
        assert results['NM_001243084.1:c.303_304delinsGA']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000014.8:g.62188231_62188232delinsGA', 'vcf': {'chr': '14', 'ref': 'TT', 'pos': '62188231', 'alt': 'GA'}}
        assert results['NM_001243084.1:c.303_304delinsGA']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000014.9:g.61721513_61721514delinsGA', 'vcf': {'chr': '14', 'ref': 'TT', 'pos': '61721513', 'alt': 'GA'}}

        assert 'NM_181054.2:c.231_232delinsGA' in results.keys()
        assert results['NM_181054.2:c.231_232delinsGA']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_181054.2:c.231_232delinsGA']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_181054.2:c.231_232delinsGA']['alt_genomic_loci'] == []
        assert results['NM_181054.2:c.231_232delinsGA']['gene_symbol'] == 'HIF1A'
        assert results['NM_181054.2:c.231_232delinsGA']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_851397.1:p.(Asp77_Leu78delinsGluMet)', 'slr': 'NP_851397.1:p.(D77_L78delinsEM)'}
        assert results['NM_181054.2:c.231_232delinsGA']['submitted_variant'] == '14-62188231-TT-GA'
        assert results['NM_181054.2:c.231_232delinsGA']['genome_context_intronic_sequence'] == ''
        assert results['NM_181054.2:c.231_232delinsGA']['hgvs_lrg_variant'] == ''
        assert results['NM_181054.2:c.231_232delinsGA']['hgvs_transcript_variant'] == 'NM_181054.2:c.231_232delinsGA'
        assert results['NM_181054.2:c.231_232delinsGA']['hgvs_refseqgene_variant'] == ''
        assert results['NM_181054.2:c.231_232delinsGA']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000014.8:g.62188231_62188232delinsGA', 'vcf': {'chr': 'chr14', 'ref': 'TT', 'pos': '62188231', 'alt': 'GA'}}
        assert results['NM_181054.2:c.231_232delinsGA']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000014.9:g.61721513_61721514delinsGA', 'vcf': {'chr': 'chr14', 'ref': 'TT', 'pos': '61721513', 'alt': 'GA'}}
        assert results['NM_181054.2:c.231_232delinsGA']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000014.8:g.62188231_62188232delinsGA', 'vcf': {'chr': '14', 'ref': 'TT', 'pos': '62188231', 'alt': 'GA'}}
        assert results['NM_181054.2:c.231_232delinsGA']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000014.9:g.61721513_61721514delinsGA', 'vcf': {'chr': '14', 'ref': 'TT', 'pos': '61721513', 'alt': 'GA'}}

        assert 'NR_144368.1:n.214-4497_214-4496delinsTC' in results.keys()
        assert results['NR_144368.1:n.214-4497_214-4496delinsTC']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_144368.1:n.214-4497_214-4496delinsTC']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_144368.1:n.214-4497_214-4496delinsTC']['alt_genomic_loci'] == []
        assert results['NR_144368.1:n.214-4497_214-4496delinsTC']['gene_symbol'] == 'LOC105370526'
        assert results['NR_144368.1:n.214-4497_214-4496delinsTC']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_144368.1:n.214-4497_214-4496delinsTC']['submitted_variant'] == '14-62188231-TT-GA'
        assert results['NR_144368.1:n.214-4497_214-4496delinsTC']['genome_context_intronic_sequence'] == 'NC_000014.8(NR_144368.1):c.214-4497_214-4496delinsTC'
        assert results['NR_144368.1:n.214-4497_214-4496delinsTC']['hgvs_lrg_variant'] == ''
        assert results['NR_144368.1:n.214-4497_214-4496delinsTC']['hgvs_transcript_variant'] == 'NR_144368.1:n.214-4497_214-4496delinsTC'
        assert results['NR_144368.1:n.214-4497_214-4496delinsTC']['hgvs_refseqgene_variant'] == ''
        assert results['NR_144368.1:n.214-4497_214-4496delinsTC']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000014.8:g.62188231_62188232delinsGA', 'vcf': {'chr': 'chr14', 'ref': 'TT', 'pos': '62188231', 'alt': u'GA'}}
        assert results['NR_144368.1:n.214-4497_214-4496delinsTC']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000014.9:g.61721513_61721514delinsGA', 'vcf': {'chr': 'chr14', 'ref': 'TT', 'pos': '61721513', 'alt': u'GA'}}
        assert results['NR_144368.1:n.214-4497_214-4496delinsTC']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000014.8:g.62188231_62188232delinsGA', 'vcf': {'chr': '14', 'ref': 'TT', 'pos': '62188231', 'alt': u'GA'}}
        assert results['NR_144368.1:n.214-4497_214-4496delinsTC']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000014.9:g.61721513_61721514delinsGA', 'vcf': {'chr': '14', 'ref': 'TT', 'pos': '61721513', 'alt': u'GA'}}

        assert 'NM_001530.3:c.231_232delinsGA' in results.keys()
        assert results['NM_001530.3:c.231_232delinsGA']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001530.3:c.231_232delinsGA']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001530.3:c.231_232delinsGA']['alt_genomic_loci'] == []
        assert results['NM_001530.3:c.231_232delinsGA']['gene_symbol'] == 'HIF1A'
        assert results['NM_001530.3:c.231_232delinsGA']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001521.1:p.(Asp77_Leu78delinsGluMet)', 'slr': 'NP_001521.1:p.(D77_L78delinsEM)'}
        assert results['NM_001530.3:c.231_232delinsGA']['submitted_variant'] == '14-62188231-TT-GA'
        assert results['NM_001530.3:c.231_232delinsGA']['genome_context_intronic_sequence'] == ''
        assert results['NM_001530.3:c.231_232delinsGA']['hgvs_lrg_variant'] == ''
        assert results['NM_001530.3:c.231_232delinsGA']['hgvs_transcript_variant'] == 'NM_001530.3:c.231_232delinsGA'
        assert results['NM_001530.3:c.231_232delinsGA']['hgvs_refseqgene_variant'] == 'NG_029606.1:g.31113_31114delinsGA'
        assert results['NM_001530.3:c.231_232delinsGA']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000014.8:g.62188231_62188232delinsGA', 'vcf': {'chr': 'chr14', 'ref': 'TT', 'pos': '62188231', 'alt': 'GA'}}
        assert results['NM_001530.3:c.231_232delinsGA']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000014.9:g.61721513_61721514delinsGA', 'vcf': {'chr': 'chr14', 'ref': 'TT', 'pos': '61721513', 'alt': 'GA'}}
        assert results['NM_001530.3:c.231_232delinsGA']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000014.8:g.62188231_62188232delinsGA', 'vcf': {'chr': '14', 'ref': 'TT', 'pos': '62188231', 'alt': 'GA'}}
        assert results['NM_001530.3:c.231_232delinsGA']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000014.9:g.61721513_61721514delinsGA', 'vcf': {'chr': '14', 'ref': 'TT', 'pos': '61721513', 'alt': 'GA'}}


    def test_variant226(self):
        variant = '14-63174827-C-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_172375.1:c.*333G>T' in results.keys()
        assert results['NM_172375.1:c.*333G>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_172375.1:c.*333G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_172375.1:c.*333G>T']['alt_genomic_loci'] == []
        assert results['NM_172375.1:c.*333G>T']['gene_symbol'] == 'KCNH5'
        assert results['NM_172375.1:c.*333G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_758963.1:p.?', 'slr': 'NP_758963.1:p.?'}
        assert results['NM_172375.1:c.*333G>T']['submitted_variant'] == '14-63174827-C-A'
        assert results['NM_172375.1:c.*333G>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_172375.1:c.*333G>T']['hgvs_lrg_variant'] == ''
        assert results['NM_172375.1:c.*333G>T']['hgvs_transcript_variant'] == 'NM_172375.1:c.*333G>T'
        assert results['NM_172375.1:c.*333G>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_172375.1:c.*333G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000014.8:g.63174827C>A', 'vcf': {'chr': 'chr14', 'ref': u'C', 'pos': '63174827', 'alt': u'A'}}
        assert 'hg38' not in results['NM_172375.1:c.*333G>T']['primary_assembly_loci'].keys()
        assert results['NM_172375.1:c.*333G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000014.8:g.63174827C>A', 'vcf': {'chr': '14', 'ref': u'C', 'pos': '63174827', 'alt': u'A'}}
        assert 'grch38' not in results['NM_172375.1:c.*333G>T']['primary_assembly_loci'].keys()

        assert 'NM_139318.4:c.2366G>T' in results.keys()
        assert results['NM_139318.4:c.2366G>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_139318.4:c.2366G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_139318.4:c.2366G>T']['alt_genomic_loci'] == []
        assert results['NM_139318.4:c.2366G>T']['gene_symbol'] == 'KCNH5'
        assert results['NM_139318.4:c.2366G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_647479.2:p.(Gly789Val)', 'slr': 'NP_647479.2:p.(G789V)'}
        assert results['NM_139318.4:c.2366G>T']['submitted_variant'] == '14-63174827-C-A'
        assert results['NM_139318.4:c.2366G>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_139318.4:c.2366G>T']['hgvs_lrg_variant'] == ''
        assert results['NM_139318.4:c.2366G>T']['hgvs_transcript_variant'] == 'NM_139318.4:c.2366G>T'
        assert results['NM_139318.4:c.2366G>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_139318.4:c.2366G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000014.8:g.63174827C>A', 'vcf': {'chr': 'chr14', 'ref': u'C', 'pos': '63174827', 'alt': u'A'}}
        assert results['NM_139318.4:c.2366G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000014.9:g.62708109C>A', 'vcf': {'chr': 'chr14', 'ref': u'C', 'pos': '62708109', 'alt': u'A'}}
        assert results['NM_139318.4:c.2366G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000014.8:g.63174827C>A', 'vcf': {'chr': '14', 'ref': u'C', 'pos': '63174827', 'alt': u'A'}}
        assert results['NM_139318.4:c.2366G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000014.9:g.62708109C>A', 'vcf': {'chr': '14', 'ref': u'C', 'pos': '62708109', 'alt': u'A'}}

        assert 'NM_139318.3:c.2366G>T' in results.keys()
        assert results['NM_139318.3:c.2366G>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_139318.3:c.2366G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_139318.3:c.2366G>T']['alt_genomic_loci'] == []
        assert results['NM_139318.3:c.2366G>T']['gene_symbol'] == 'KCNH5'
        assert results['NM_139318.3:c.2366G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_647479.2:p.(Gly789Val)', 'slr': 'NP_647479.2:p.(G789V)'}
        assert results['NM_139318.3:c.2366G>T']['submitted_variant'] == '14-63174827-C-A'
        assert results['NM_139318.3:c.2366G>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_139318.3:c.2366G>T']['hgvs_lrg_variant'] == ''
        assert results['NM_139318.3:c.2366G>T']['hgvs_transcript_variant'] == 'NM_139318.3:c.2366G>T'
        assert results['NM_139318.3:c.2366G>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_139318.3:c.2366G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000014.8:g.63174827C>A', 'vcf': {'chr': 'chr14', 'ref': u'C', 'pos': '63174827', 'alt': u'A'}}
        assert 'hg38' not in results['NM_139318.3:c.2366G>T']['primary_assembly_loci'].keys()
        assert results['NM_139318.3:c.2366G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000014.8:g.63174827C>A', 'vcf': {'chr': '14', 'ref': u'C', 'pos': '63174827', 'alt': u'A'}}
        assert 'grch38' not in results['NM_139318.3:c.2366G>T']['primary_assembly_loci'].keys()

        assert results['flag'] == 'gene_variant'
        assert 'NM_172375.2:c.*333G>T' in results.keys()
        assert results['NM_172375.2:c.*333G>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_172375.2:c.*333G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_172375.2:c.*333G>T']['alt_genomic_loci'] == []
        assert results['NM_172375.2:c.*333G>T']['gene_symbol'] == 'KCNH5'
        assert results['NM_172375.2:c.*333G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_758963.1:p.?', 'slr': 'NP_758963.1:p.?'}
        assert results['NM_172375.2:c.*333G>T']['submitted_variant'] == '14-63174827-C-A'
        assert results['NM_172375.2:c.*333G>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_172375.2:c.*333G>T']['hgvs_lrg_variant'] == ''
        assert results['NM_172375.2:c.*333G>T']['hgvs_transcript_variant'] == 'NM_172375.2:c.*333G>T'
        assert results['NM_172375.2:c.*333G>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_172375.2:c.*333G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000014.8:g.63174827C>A', 'vcf': {'chr': 'chr14', 'ref': u'C', 'pos': '63174827', 'alt': u'A'}}
        assert results['NM_172375.2:c.*333G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000014.9:g.62708109C>A', 'vcf': {'chr': 'chr14', 'ref': u'C', 'pos': '62708109', 'alt': u'A'}}
        assert results['NM_172375.2:c.*333G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000014.8:g.63174827C>A', 'vcf': {'chr': '14', 'ref': u'C', 'pos': '63174827', 'alt': u'A'}}
        assert results['NM_172375.2:c.*333G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000014.9:g.62708109C>A', 'vcf': {'chr': '14', 'ref': u'C', 'pos': '62708109', 'alt': u'A'}}


    def test_variant227(self):
        variant = '15-42680000-CA-C'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_000070.2:c.550del' in results.keys()
        assert results['NM_000070.2:c.550del']['hgvs_lrg_transcript_variant'] == 'LRG_849t1:c.550del'
        assert results['NM_000070.2:c.550del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000070.2:c.550del']['alt_genomic_loci'] == []
        assert results['NM_000070.2:c.550del']['gene_symbol'] == 'CAPN3'
        assert results['NM_000070.2:c.550del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000061.1(LRG_849p1):p.(Thr184ArgfsTer36)', 'slr': 'NP_000061.1:p.(T184Rfs*36)'}
        assert results['NM_000070.2:c.550del']['submitted_variant'] == '15-42680000-CA-C'
        assert results['NM_000070.2:c.550del']['genome_context_intronic_sequence'] == ''
        assert results['NM_000070.2:c.550del']['hgvs_lrg_variant'] == 'LRG_849:g.44702del'
        assert results['NM_000070.2:c.550del']['hgvs_transcript_variant'] == 'NM_000070.2:c.550del'
        assert results['NM_000070.2:c.550del']['hgvs_refseqgene_variant'] == 'NG_008660.1:g.44702del'
        assert results['NM_000070.2:c.550del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.42680002del', 'vcf': {'chr': 'chr15', 'ref': 'CA', 'pos': '42680000', 'alt': 'C'}}
        assert results['NM_000070.2:c.550del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42387804del', 'vcf': {'chr': 'chr15', 'ref': 'CA', 'pos': '42387802', 'alt': 'C'}}
        assert results['NM_000070.2:c.550del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.42680002del', 'vcf': {'chr': '15', 'ref': 'CA', 'pos': '42680000', 'alt': 'C'}}
        assert results['NM_000070.2:c.550del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42387804del', 'vcf': {'chr': '15', 'ref': 'CA', 'pos': '42387802', 'alt': 'C'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_024344.1:c.550del' in results.keys()
        assert results['NM_024344.1:c.550del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_024344.1:c.550del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_024344.1:c.550del']['alt_genomic_loci'] == []
        assert results['NM_024344.1:c.550del']['gene_symbol'] == 'CAPN3'
        assert results['NM_024344.1:c.550del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_077320.1:p.(Thr184ArgfsTer36)', 'slr': 'NP_077320.1:p.(T184Rfs*36)'}
        assert results['NM_024344.1:c.550del']['submitted_variant'] == '15-42680000-CA-C'
        assert results['NM_024344.1:c.550del']['genome_context_intronic_sequence'] == ''
        assert results['NM_024344.1:c.550del']['hgvs_lrg_variant'] == ''
        assert results['NM_024344.1:c.550del']['hgvs_transcript_variant'] == 'NM_024344.1:c.550del'
        assert results['NM_024344.1:c.550del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_024344.1:c.550del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.42680002del', 'vcf': {'chr': 'chr15', 'ref': 'CA', 'pos': '42680000', 'alt': 'C'}}
        assert results['NM_024344.1:c.550del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42387804del', 'vcf': {'chr': 'chr15', 'ref': 'CA', 'pos': '42387802', 'alt': 'C'}}
        assert results['NM_024344.1:c.550del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.42680002del', 'vcf': {'chr': '15', 'ref': 'CA', 'pos': '42680000', 'alt': 'C'}}
        assert results['NM_024344.1:c.550del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42387804del', 'vcf': {'chr': '15', 'ref': 'CA', 'pos': '42387802', 'alt': 'C'}}

        assert 'NM_173087.1:c.550del' in results.keys()
        assert results['NM_173087.1:c.550del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_173087.1:c.550del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_173087.1:c.550del']['alt_genomic_loci'] == []
        assert results['NM_173087.1:c.550del']['gene_symbol'] == 'CAPN3'
        assert results['NM_173087.1:c.550del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_775110.1:p.(Thr184ArgfsTer36)', 'slr': 'NP_775110.1:p.(T184Rfs*36)'}
        assert results['NM_173087.1:c.550del']['submitted_variant'] == '15-42680000-CA-C'
        assert results['NM_173087.1:c.550del']['genome_context_intronic_sequence'] == ''
        assert results['NM_173087.1:c.550del']['hgvs_lrg_variant'] == ''
        assert results['NM_173087.1:c.550del']['hgvs_transcript_variant'] == 'NM_173087.1:c.550del'
        assert results['NM_173087.1:c.550del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_173087.1:c.550del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.42680002del', 'vcf': {'chr': 'chr15', 'ref': 'CA', 'pos': '42680000', 'alt': 'C'}}
        assert results['NM_173087.1:c.550del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42387804del', 'vcf': {'chr': 'chr15', 'ref': 'CA', 'pos': '42387802', 'alt': 'C'}}
        assert results['NM_173087.1:c.550del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.42680002del', 'vcf': {'chr': '15', 'ref': 'CA', 'pos': '42680000', 'alt': 'C'}}
        assert results['NM_173087.1:c.550del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42387804del', 'vcf': {'chr': '15', 'ref': 'CA', 'pos': '42387802', 'alt': 'C'}}


    def test_variant228(self):
        variant = '15-42680000-CA-CAA'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_024344.1:c.550dup' in results.keys()
        assert results['NM_024344.1:c.550dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_024344.1:c.550dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_024344.1:c.550dup']['alt_genomic_loci'] == []
        assert results['NM_024344.1:c.550dup']['gene_symbol'] == 'CAPN3'
        assert results['NM_024344.1:c.550dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_077320.1:p.(Thr184AsnfsTer16)', 'slr': 'NP_077320.1:p.(T184Nfs*16)'}
        assert results['NM_024344.1:c.550dup']['submitted_variant'] == '15-42680000-CA-CAA'
        assert results['NM_024344.1:c.550dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_024344.1:c.550dup']['hgvs_lrg_variant'] == ''
        assert results['NM_024344.1:c.550dup']['hgvs_transcript_variant'] == 'NM_024344.1:c.550dup'
        assert results['NM_024344.1:c.550dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_024344.1:c.550dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.42680002dup', 'vcf': {'chr': 'chr15', 'ref': 'A', 'pos': '42680001', 'alt': 'AA'}}
        assert results['NM_024344.1:c.550dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42387804dup', 'vcf': {'chr': 'chr15', 'ref': 'A', 'pos': '42387803', 'alt': 'AA'}}
        assert results['NM_024344.1:c.550dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.42680002dup', 'vcf': {'chr': '15', 'ref': 'A', 'pos': '42680001', 'alt': 'AA'}}
        assert results['NM_024344.1:c.550dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42387804dup', 'vcf': {'chr': '15', 'ref': 'A', 'pos': '42387803', 'alt': 'AA'}}

        assert 'NM_173087.1:c.550dup' in results.keys()
        assert results['NM_173087.1:c.550dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_173087.1:c.550dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_173087.1:c.550dup']['alt_genomic_loci'] == []
        assert results['NM_173087.1:c.550dup']['gene_symbol'] == 'CAPN3'
        assert results['NM_173087.1:c.550dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_775110.1:p.(Thr184AsnfsTer16)', 'slr': 'NP_775110.1:p.(T184Nfs*16)'}
        assert results['NM_173087.1:c.550dup']['submitted_variant'] == '15-42680000-CA-CAA'
        assert results['NM_173087.1:c.550dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_173087.1:c.550dup']['hgvs_lrg_variant'] == ''
        assert results['NM_173087.1:c.550dup']['hgvs_transcript_variant'] == 'NM_173087.1:c.550dup'
        assert results['NM_173087.1:c.550dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_173087.1:c.550dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.42680002dup', 'vcf': {'chr': 'chr15', 'ref': 'A', 'pos': '42680001', 'alt': 'AA'}}
        assert results['NM_173087.1:c.550dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42387804dup', 'vcf': {'chr': 'chr15', 'ref': 'A', 'pos': '42387803', 'alt': 'AA'}}
        assert results['NM_173087.1:c.550dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.42680002dup', 'vcf': {'chr': '15', 'ref': 'A', 'pos': '42680001', 'alt': 'AA'}}
        assert results['NM_173087.1:c.550dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42387804dup', 'vcf': {'chr': '15', 'ref': 'A', 'pos': '42387803', 'alt': 'AA'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_000070.2:c.550dup' in results.keys()
        assert results['NM_000070.2:c.550dup']['hgvs_lrg_transcript_variant'] == 'LRG_849t1:c.550dup'
        assert results['NM_000070.2:c.550dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000070.2:c.550dup']['alt_genomic_loci'] == []
        assert results['NM_000070.2:c.550dup']['gene_symbol'] == 'CAPN3'
        assert results['NM_000070.2:c.550dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000061.1(LRG_849p1):p.(Thr184AsnfsTer16)', 'slr': 'NP_000061.1:p.(T184Nfs*16)'}
        assert results['NM_000070.2:c.550dup']['submitted_variant'] == '15-42680000-CA-CAA'
        assert results['NM_000070.2:c.550dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_000070.2:c.550dup']['hgvs_lrg_variant'] == 'LRG_849:g.44702dup'
        assert results['NM_000070.2:c.550dup']['hgvs_transcript_variant'] == 'NM_000070.2:c.550dup'
        assert results['NM_000070.2:c.550dup']['hgvs_refseqgene_variant'] == 'NG_008660.1:g.44702dup'
        assert results['NM_000070.2:c.550dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.42680002dup', 'vcf': {'chr': 'chr15', 'ref': 'A', 'pos': '42680001', 'alt': 'AA'}}
        assert results['NM_000070.2:c.550dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42387804dup', 'vcf': {'chr': 'chr15', 'ref': 'A', 'pos': '42387803', 'alt': 'AA'}}
        assert results['NM_000070.2:c.550dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.42680002dup', 'vcf': {'chr': '15', 'ref': 'A', 'pos': '42680001', 'alt': 'AA'}}
        assert results['NM_000070.2:c.550dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42387804dup', 'vcf': {'chr': '15', 'ref': 'A', 'pos': '42387803', 'alt': 'AA'}}


    def test_variant229(self):
        variant = '15-42703179-T-TTCA'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_173088.1:c.825_826insTCA' in results.keys()
        assert results['NM_173088.1:c.825_826insTCA']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_173088.1:c.825_826insTCA']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_173088.1:c.825_826insTCA']['alt_genomic_loci'] == []
        assert results['NM_173088.1:c.825_826insTCA']['gene_symbol'] == 'CAPN3'
        assert results['NM_173088.1:c.825_826insTCA']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_775111.1:p.(Val275_Arg276insSer)', 'slr': 'NP_775111.1:p.(V275_R276insS)'}
        assert results['NM_173088.1:c.825_826insTCA']['submitted_variant'] == '15-42703179-T-TTCA'
        assert results['NM_173088.1:c.825_826insTCA']['genome_context_intronic_sequence'] == ''
        assert results['NM_173088.1:c.825_826insTCA']['hgvs_lrg_variant'] == ''
        assert results['NM_173088.1:c.825_826insTCA']['hgvs_transcript_variant'] == 'NM_173088.1:c.825_826insTCA'
        assert results['NM_173088.1:c.825_826insTCA']['hgvs_refseqgene_variant'] == ''
        assert results['NM_173088.1:c.825_826insTCA']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.42703179_42703180insTCA', 'vcf': {'chr': 'chr15', 'ref': 'T', 'pos': '42703179', 'alt': 'TTCA'}}
        assert results['NM_173088.1:c.825_826insTCA']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42410981_42410982insTCA', 'vcf': {'chr': 'chr15', 'ref': 'T', 'pos': '42410981', 'alt': 'TTCA'}}
        assert results['NM_173088.1:c.825_826insTCA']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.42703179_42703180insTCA', 'vcf': {'chr': '15', 'ref': 'T', 'pos': '42703179', 'alt': 'TTCA'}}
        assert results['NM_173088.1:c.825_826insTCA']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42410981_42410982insTCA', 'vcf': {'chr': '15', 'ref': 'T', 'pos': '42410981', 'alt': 'TTCA'}}

        assert 'NM_173090.1:c.366_367insTCA' in results.keys()
        assert results['NM_173090.1:c.366_367insTCA']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_173090.1:c.366_367insTCA']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_173090.1:c.366_367insTCA']['alt_genomic_loci'] == []
        assert results['NM_173090.1:c.366_367insTCA']['gene_symbol'] == 'CAPN3'
        assert results['NM_173090.1:c.366_367insTCA']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_775113.1:p.(Val122_Arg123insSer)', 'slr': 'NP_775113.1:p.(V122_R123insS)'}
        assert results['NM_173090.1:c.366_367insTCA']['submitted_variant'] == '15-42703179-T-TTCA'
        assert results['NM_173090.1:c.366_367insTCA']['genome_context_intronic_sequence'] == ''
        assert results['NM_173090.1:c.366_367insTCA']['hgvs_lrg_variant'] == ''
        assert results['NM_173090.1:c.366_367insTCA']['hgvs_transcript_variant'] == 'NM_173090.1:c.366_367insTCA'
        assert results['NM_173090.1:c.366_367insTCA']['hgvs_refseqgene_variant'] == ''
        assert results['NM_173090.1:c.366_367insTCA']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.42703179_42703180insTCA', 'vcf': {'chr': 'chr15', 'ref': 'T', 'pos': '42703179', 'alt': 'TTCA'}}
        assert results['NM_173090.1:c.366_367insTCA']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42410981_42410982insTCA', 'vcf': {'chr': 'chr15', 'ref': 'T', 'pos': '42410981', 'alt': 'TTCA'}}
        assert results['NM_173090.1:c.366_367insTCA']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.42703179_42703180insTCA', 'vcf': {'chr': '15', 'ref': 'T', 'pos': '42703179', 'alt': 'TTCA'}}
        assert results['NM_173090.1:c.366_367insTCA']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42410981_42410982insTCA', 'vcf': {'chr': '15', 'ref': 'T', 'pos': '42410981', 'alt': 'TTCA'}}

        assert 'NM_173089.1:c.366_367insTCA' in results.keys()
        assert results['NM_173089.1:c.366_367insTCA']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_173089.1:c.366_367insTCA']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_173089.1:c.366_367insTCA']['alt_genomic_loci'] == []
        assert results['NM_173089.1:c.366_367insTCA']['gene_symbol'] == 'CAPN3'
        assert results['NM_173089.1:c.366_367insTCA']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_775112.1:p.(Val122_Arg123insSer)', 'slr': 'NP_775112.1:p.(V122_R123insS)'}
        assert results['NM_173089.1:c.366_367insTCA']['submitted_variant'] == '15-42703179-T-TTCA'
        assert results['NM_173089.1:c.366_367insTCA']['genome_context_intronic_sequence'] == ''
        assert results['NM_173089.1:c.366_367insTCA']['hgvs_lrg_variant'] == ''
        assert results['NM_173089.1:c.366_367insTCA']['hgvs_transcript_variant'] == 'NM_173089.1:c.366_367insTCA'
        assert results['NM_173089.1:c.366_367insTCA']['hgvs_refseqgene_variant'] == ''
        assert results['NM_173089.1:c.366_367insTCA']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.42703179_42703180insTCA', 'vcf': {'chr': 'chr15', 'ref': 'T', 'pos': '42703179', 'alt': 'TTCA'}}
        assert results['NM_173089.1:c.366_367insTCA']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42410981_42410982insTCA', 'vcf': {'chr': 'chr15', 'ref': 'T', 'pos': '42410981', 'alt': 'TTCA'}}
        assert results['NM_173089.1:c.366_367insTCA']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.42703179_42703180insTCA', 'vcf': {'chr': '15', 'ref': 'T', 'pos': '42703179', 'alt': 'TTCA'}}
        assert results['NM_173089.1:c.366_367insTCA']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42410981_42410982insTCA', 'vcf': {'chr': '15', 'ref': 'T', 'pos': '42410981', 'alt': 'TTCA'}}

        assert 'NM_173087.1:c.2085_2086insTCA' in results.keys()
        assert results['NM_173087.1:c.2085_2086insTCA']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_173087.1:c.2085_2086insTCA']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_173087.1:c.2085_2086insTCA']['alt_genomic_loci'] == []
        assert results['NM_173087.1:c.2085_2086insTCA']['gene_symbol'] == 'CAPN3'
        assert results['NM_173087.1:c.2085_2086insTCA']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_775110.1:p.(Val695_Arg696insSer)', 'slr': 'NP_775110.1:p.(V695_R696insS)'}
        assert results['NM_173087.1:c.2085_2086insTCA']['submitted_variant'] == '15-42703179-T-TTCA'
        assert results['NM_173087.1:c.2085_2086insTCA']['genome_context_intronic_sequence'] == ''
        assert results['NM_173087.1:c.2085_2086insTCA']['hgvs_lrg_variant'] == ''
        assert results['NM_173087.1:c.2085_2086insTCA']['hgvs_transcript_variant'] == 'NM_173087.1:c.2085_2086insTCA'
        assert results['NM_173087.1:c.2085_2086insTCA']['hgvs_refseqgene_variant'] == ''
        assert results['NM_173087.1:c.2085_2086insTCA']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.42703179_42703180insTCA', 'vcf': {'chr': 'chr15', 'ref': 'T', 'pos': '42703179', 'alt': 'TTCA'}}
        assert results['NM_173087.1:c.2085_2086insTCA']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42410981_42410982insTCA', 'vcf': {'chr': 'chr15', 'ref': 'T', 'pos': '42410981', 'alt': 'TTCA'}}
        assert results['NM_173087.1:c.2085_2086insTCA']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.42703179_42703180insTCA', 'vcf': {'chr': '15', 'ref': 'T', 'pos': '42703179', 'alt': 'TTCA'}}
        assert results['NM_173087.1:c.2085_2086insTCA']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42410981_42410982insTCA', 'vcf': {'chr': '15', 'ref': 'T', 'pos': '42410981', 'alt': 'TTCA'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_000070.2:c.2361_2362insTCA' in results.keys()
        assert results['NM_000070.2:c.2361_2362insTCA']['hgvs_lrg_transcript_variant'] == 'LRG_849t1:c.2361_2362insTCA'
        assert results['NM_000070.2:c.2361_2362insTCA']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000070.2:c.2361_2362insTCA']['alt_genomic_loci'] == []
        assert results['NM_000070.2:c.2361_2362insTCA']['gene_symbol'] == 'CAPN3'
        assert results['NM_000070.2:c.2361_2362insTCA']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000061.1(LRG_849p1):p.(Val787_Arg788insSer)', 'slr': 'NP_000061.1:p.(V787_R788insS)'}
        assert results['NM_000070.2:c.2361_2362insTCA']['submitted_variant'] == '15-42703179-T-TTCA'
        assert results['NM_000070.2:c.2361_2362insTCA']['genome_context_intronic_sequence'] == ''
        assert results['NM_000070.2:c.2361_2362insTCA']['hgvs_lrg_variant'] == 'LRG_849:g.67879_67880insTCA'
        assert results['NM_000070.2:c.2361_2362insTCA']['hgvs_transcript_variant'] == 'NM_000070.2:c.2361_2362insTCA'
        assert results['NM_000070.2:c.2361_2362insTCA']['hgvs_refseqgene_variant'] == 'NG_008660.1:g.67879_67880insTCA'
        assert results['NM_000070.2:c.2361_2362insTCA']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.42703179_42703180insTCA', 'vcf': {'chr': 'chr15', 'ref': 'T', 'pos': '42703179', 'alt': 'TTCA'}}
        assert results['NM_000070.2:c.2361_2362insTCA']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42410981_42410982insTCA', 'vcf': {'chr': 'chr15', 'ref': 'T', 'pos': '42410981', 'alt': 'TTCA'}}
        assert results['NM_000070.2:c.2361_2362insTCA']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.42703179_42703180insTCA', 'vcf': {'chr': '15', 'ref': 'T', 'pos': '42703179', 'alt': 'TTCA'}}
        assert results['NM_000070.2:c.2361_2362insTCA']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42410981_42410982insTCA', 'vcf': {'chr': '15', 'ref': 'T', 'pos': '42410981', 'alt': 'TTCA'}}

        assert 'NM_024344.1:c.2343_2344insTCA' in results.keys()
        assert results['NM_024344.1:c.2343_2344insTCA']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_024344.1:c.2343_2344insTCA']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_024344.1:c.2343_2344insTCA']['alt_genomic_loci'] == []
        assert results['NM_024344.1:c.2343_2344insTCA']['gene_symbol'] == 'CAPN3'
        assert results['NM_024344.1:c.2343_2344insTCA']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_077320.1:p.(Val781_Arg782insSer)', 'slr': 'NP_077320.1:p.(V781_R782insS)'}
        assert results['NM_024344.1:c.2343_2344insTCA']['submitted_variant'] == '15-42703179-T-TTCA'
        assert results['NM_024344.1:c.2343_2344insTCA']['genome_context_intronic_sequence'] == ''
        assert results['NM_024344.1:c.2343_2344insTCA']['hgvs_lrg_variant'] == ''
        assert results['NM_024344.1:c.2343_2344insTCA']['hgvs_transcript_variant'] == 'NM_024344.1:c.2343_2344insTCA'
        assert results['NM_024344.1:c.2343_2344insTCA']['hgvs_refseqgene_variant'] == ''
        assert results['NM_024344.1:c.2343_2344insTCA']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.42703179_42703180insTCA', 'vcf': {'chr': 'chr15', 'ref': 'T', 'pos': '42703179', 'alt': 'TTCA'}}
        assert results['NM_024344.1:c.2343_2344insTCA']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42410981_42410982insTCA', 'vcf': {'chr': 'chr15', 'ref': 'T', 'pos': '42410981', 'alt': 'TTCA'}}
        assert results['NM_024344.1:c.2343_2344insTCA']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.42703179_42703180insTCA', 'vcf': {'chr': '15', 'ref': 'T', 'pos': '42703179', 'alt': 'TTCA'}}
        assert results['NM_024344.1:c.2343_2344insTCA']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42410981_42410982insTCA', 'vcf': {'chr': '15', 'ref': 'T', 'pos': '42410981', 'alt': 'TTCA'}}


    def test_variant230(self):
        variant = '15-42703179-TAG-TTCATCT'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_024344.1:c.2344_2345delinsTCATCT' in results.keys()
        assert results['NM_024344.1:c.2344_2345delinsTCATCT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_024344.1:c.2344_2345delinsTCATCT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_024344.1:c.2344_2345delinsTCATCT']['alt_genomic_loci'] == []
        assert results['NM_024344.1:c.2344_2345delinsTCATCT']['gene_symbol'] == 'CAPN3'
        assert results['NM_024344.1:c.2344_2345delinsTCATCT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_077320.1:p.(Arg782SerfsTer14)', 'slr': 'NP_077320.1:p.(R782Sfs*14)'}
        assert results['NM_024344.1:c.2344_2345delinsTCATCT']['submitted_variant'] == '15-42703179-TAG-TTCATCT'
        assert results['NM_024344.1:c.2344_2345delinsTCATCT']['genome_context_intronic_sequence'] == ''
        assert results['NM_024344.1:c.2344_2345delinsTCATCT']['hgvs_lrg_variant'] == ''
        assert results['NM_024344.1:c.2344_2345delinsTCATCT']['hgvs_transcript_variant'] == 'NM_024344.1:c.2344_2345delinsTCATCT'
        assert results['NM_024344.1:c.2344_2345delinsTCATCT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_024344.1:c.2344_2345delinsTCATCT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.42703180_42703181delinsTCATCT', 'vcf': {'chr': 'chr15', 'ref': 'AG', 'pos': '42703180', 'alt': 'TCATCT'}}
        assert results['NM_024344.1:c.2344_2345delinsTCATCT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42410982_42410983delinsTCATCT', 'vcf': {'chr': 'chr15', 'ref': 'AG', 'pos': '42410982', 'alt': 'TCATCT'}}
        assert results['NM_024344.1:c.2344_2345delinsTCATCT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.42703180_42703181delinsTCATCT', 'vcf': {'chr': '15', 'ref': 'AG', 'pos': '42703180', 'alt': 'TCATCT'}}
        assert results['NM_024344.1:c.2344_2345delinsTCATCT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42410982_42410983delinsTCATCT', 'vcf': {'chr': '15', 'ref': 'AG', 'pos': '42410982', 'alt': 'TCATCT'}}

        assert 'NM_173090.1:c.367_368delinsTCATCT' in results.keys()
        assert results['NM_173090.1:c.367_368delinsTCATCT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_173090.1:c.367_368delinsTCATCT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_173090.1:c.367_368delinsTCATCT']['alt_genomic_loci'] == []
        assert results['NM_173090.1:c.367_368delinsTCATCT']['gene_symbol'] == 'CAPN3'
        assert results['NM_173090.1:c.367_368delinsTCATCT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_775113.1:p.(Arg123SerfsTer14)', 'slr': 'NP_775113.1:p.(R123Sfs*14)'}
        assert results['NM_173090.1:c.367_368delinsTCATCT']['submitted_variant'] == '15-42703179-TAG-TTCATCT'
        assert results['NM_173090.1:c.367_368delinsTCATCT']['genome_context_intronic_sequence'] == ''
        assert results['NM_173090.1:c.367_368delinsTCATCT']['hgvs_lrg_variant'] == ''
        assert results['NM_173090.1:c.367_368delinsTCATCT']['hgvs_transcript_variant'] == 'NM_173090.1:c.367_368delinsTCATCT'
        assert results['NM_173090.1:c.367_368delinsTCATCT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_173090.1:c.367_368delinsTCATCT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.42703180_42703181delinsTCATCT', 'vcf': {'chr': 'chr15', 'ref': 'AG', 'pos': '42703180', 'alt': 'TCATCT'}}
        assert results['NM_173090.1:c.367_368delinsTCATCT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42410982_42410983delinsTCATCT', 'vcf': {'chr': 'chr15', 'ref': 'AG', 'pos': '42410982', 'alt': 'TCATCT'}}
        assert results['NM_173090.1:c.367_368delinsTCATCT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.42703180_42703181delinsTCATCT', 'vcf': {'chr': '15', 'ref': 'AG', 'pos': '42703180', 'alt': 'TCATCT'}}
        assert results['NM_173090.1:c.367_368delinsTCATCT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42410982_42410983delinsTCATCT', 'vcf': {'chr': '15', 'ref': 'AG', 'pos': '42410982', 'alt': 'TCATCT'}}

        assert 'NM_173088.1:c.826_827delinsTCATCT' in results.keys()
        assert results['NM_173088.1:c.826_827delinsTCATCT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_173088.1:c.826_827delinsTCATCT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_173088.1:c.826_827delinsTCATCT']['alt_genomic_loci'] == []
        assert results['NM_173088.1:c.826_827delinsTCATCT']['gene_symbol'] == 'CAPN3'
        assert results['NM_173088.1:c.826_827delinsTCATCT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_775111.1:p.(Arg276SerfsTer14)', 'slr': 'NP_775111.1:p.(R276Sfs*14)'}
        assert results['NM_173088.1:c.826_827delinsTCATCT']['submitted_variant'] == '15-42703179-TAG-TTCATCT'
        assert results['NM_173088.1:c.826_827delinsTCATCT']['genome_context_intronic_sequence'] == ''
        assert results['NM_173088.1:c.826_827delinsTCATCT']['hgvs_lrg_variant'] == ''
        assert results['NM_173088.1:c.826_827delinsTCATCT']['hgvs_transcript_variant'] == 'NM_173088.1:c.826_827delinsTCATCT'
        assert results['NM_173088.1:c.826_827delinsTCATCT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_173088.1:c.826_827delinsTCATCT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.42703180_42703181delinsTCATCT', 'vcf': {'chr': 'chr15', 'ref': 'AG', 'pos': '42703180', 'alt': 'TCATCT'}}
        assert results['NM_173088.1:c.826_827delinsTCATCT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42410982_42410983delinsTCATCT', 'vcf': {'chr': 'chr15', 'ref': 'AG', 'pos': '42410982', 'alt': 'TCATCT'}}
        assert results['NM_173088.1:c.826_827delinsTCATCT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.42703180_42703181delinsTCATCT', 'vcf': {'chr': '15', 'ref': 'AG', 'pos': '42703180', 'alt': 'TCATCT'}}
        assert results['NM_173088.1:c.826_827delinsTCATCT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42410982_42410983delinsTCATCT', 'vcf': {'chr': '15', 'ref': 'AG', 'pos': '42410982', 'alt': 'TCATCT'}}

        assert 'NM_000070.2:c.2362_2363delinsTCATCT' in results.keys()
        assert results['NM_000070.2:c.2362_2363delinsTCATCT']['hgvs_lrg_transcript_variant'] == 'LRG_849t1:c.2362_2363delinsTCATCT'
        assert results['NM_000070.2:c.2362_2363delinsTCATCT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000070.2:c.2362_2363delinsTCATCT']['alt_genomic_loci'] == []
        assert results['NM_000070.2:c.2362_2363delinsTCATCT']['gene_symbol'] == 'CAPN3'
        assert results['NM_000070.2:c.2362_2363delinsTCATCT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000061.1(LRG_849p1):p.(Arg788SerfsTer14)', 'slr': 'NP_000061.1:p.(R788Sfs*14)'}
        assert results['NM_000070.2:c.2362_2363delinsTCATCT']['submitted_variant'] == '15-42703179-TAG-TTCATCT'
        assert results['NM_000070.2:c.2362_2363delinsTCATCT']['genome_context_intronic_sequence'] == ''
        assert results['NM_000070.2:c.2362_2363delinsTCATCT']['hgvs_lrg_variant'] == 'LRG_849:g.67880_67881delinsTCATCT'
        assert results['NM_000070.2:c.2362_2363delinsTCATCT']['hgvs_transcript_variant'] == 'NM_000070.2:c.2362_2363delinsTCATCT'
        assert results['NM_000070.2:c.2362_2363delinsTCATCT']['hgvs_refseqgene_variant'] == 'NG_008660.1:g.67880_67881delinsTCATCT'
        assert results['NM_000070.2:c.2362_2363delinsTCATCT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.42703180_42703181delinsTCATCT', 'vcf': {'chr': 'chr15', 'ref': 'AG', 'pos': '42703180', 'alt': 'TCATCT'}}
        assert results['NM_000070.2:c.2362_2363delinsTCATCT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42410982_42410983delinsTCATCT', 'vcf': {'chr': 'chr15', 'ref': 'AG', 'pos': '42410982', 'alt': 'TCATCT'}}
        assert results['NM_000070.2:c.2362_2363delinsTCATCT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.42703180_42703181delinsTCATCT', 'vcf': {'chr': '15', 'ref': 'AG', 'pos': '42703180', 'alt': 'TCATCT'}}
        assert results['NM_000070.2:c.2362_2363delinsTCATCT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42410982_42410983delinsTCATCT', 'vcf': {'chr': '15', 'ref': 'AG', 'pos': '42410982', 'alt': 'TCATCT'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_173089.1:c.367_368delinsTCATCT' in results.keys()
        assert results['NM_173089.1:c.367_368delinsTCATCT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_173089.1:c.367_368delinsTCATCT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_173089.1:c.367_368delinsTCATCT']['alt_genomic_loci'] == []
        assert results['NM_173089.1:c.367_368delinsTCATCT']['gene_symbol'] == 'CAPN3'
        assert results['NM_173089.1:c.367_368delinsTCATCT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_775112.1:p.(Arg123SerfsTer14)', 'slr': 'NP_775112.1:p.(R123Sfs*14)'}
        assert results['NM_173089.1:c.367_368delinsTCATCT']['submitted_variant'] == '15-42703179-TAG-TTCATCT'
        assert results['NM_173089.1:c.367_368delinsTCATCT']['genome_context_intronic_sequence'] == ''
        assert results['NM_173089.1:c.367_368delinsTCATCT']['hgvs_lrg_variant'] == ''
        assert results['NM_173089.1:c.367_368delinsTCATCT']['hgvs_transcript_variant'] == 'NM_173089.1:c.367_368delinsTCATCT'
        assert results['NM_173089.1:c.367_368delinsTCATCT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_173089.1:c.367_368delinsTCATCT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.42703180_42703181delinsTCATCT', 'vcf': {'chr': 'chr15', 'ref': 'AG', 'pos': '42703180', 'alt': 'TCATCT'}}
        assert results['NM_173089.1:c.367_368delinsTCATCT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42410982_42410983delinsTCATCT', 'vcf': {'chr': 'chr15', 'ref': 'AG', 'pos': '42410982', 'alt': 'TCATCT'}}
        assert results['NM_173089.1:c.367_368delinsTCATCT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.42703180_42703181delinsTCATCT', 'vcf': {'chr': '15', 'ref': 'AG', 'pos': '42703180', 'alt': 'TCATCT'}}
        assert results['NM_173089.1:c.367_368delinsTCATCT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42410982_42410983delinsTCATCT', 'vcf': {'chr': '15', 'ref': 'AG', 'pos': '42410982', 'alt': 'TCATCT'}}

        assert 'NM_173087.1:c.2086_2087delinsTCATCT' in results.keys()
        assert results['NM_173087.1:c.2086_2087delinsTCATCT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_173087.1:c.2086_2087delinsTCATCT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_173087.1:c.2086_2087delinsTCATCT']['alt_genomic_loci'] == []
        assert results['NM_173087.1:c.2086_2087delinsTCATCT']['gene_symbol'] == 'CAPN3'
        assert results['NM_173087.1:c.2086_2087delinsTCATCT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_775110.1:p.(Arg696SerfsTer14)', 'slr': 'NP_775110.1:p.(R696Sfs*14)'}
        assert results['NM_173087.1:c.2086_2087delinsTCATCT']['submitted_variant'] == '15-42703179-TAG-TTCATCT'
        assert results['NM_173087.1:c.2086_2087delinsTCATCT']['genome_context_intronic_sequence'] == ''
        assert results['NM_173087.1:c.2086_2087delinsTCATCT']['hgvs_lrg_variant'] == ''
        assert results['NM_173087.1:c.2086_2087delinsTCATCT']['hgvs_transcript_variant'] == 'NM_173087.1:c.2086_2087delinsTCATCT'
        assert results['NM_173087.1:c.2086_2087delinsTCATCT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_173087.1:c.2086_2087delinsTCATCT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.42703180_42703181delinsTCATCT', 'vcf': {'chr': 'chr15', 'ref': 'AG', 'pos': '42703180', 'alt': 'TCATCT'}}
        assert results['NM_173087.1:c.2086_2087delinsTCATCT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42410982_42410983delinsTCATCT', 'vcf': {'chr': 'chr15', 'ref': 'AG', 'pos': '42410982', 'alt': 'TCATCT'}}
        assert results['NM_173087.1:c.2086_2087delinsTCATCT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.42703180_42703181delinsTCATCT', 'vcf': {'chr': '15', 'ref': 'AG', 'pos': '42703180', 'alt': 'TCATCT'}}
        assert results['NM_173087.1:c.2086_2087delinsTCATCT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000015.10:g.42410982_42410983delinsTCATCT', 'vcf': {'chr': '15', 'ref': 'AG', 'pos': '42410982', 'alt': 'TCATCT'}}


    def test_variant231(self):
        variant = '15-48782203-C-T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_000138.4:c.2927G>A' in results.keys()
        assert results['NM_000138.4:c.2927G>A']['hgvs_lrg_transcript_variant'] == 'LRG_778t1:c.2927G>A'
        assert results['NM_000138.4:c.2927G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000138.4:c.2927G>A']['alt_genomic_loci'] == []
        assert results['NM_000138.4:c.2927G>A']['gene_symbol'] == 'FBN1'
        assert results['NM_000138.4:c.2927G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000129.3(LRG_778p1):p.(Arg976His)', 'slr': 'NP_000129.3:p.(R976H)'}
        assert results['NM_000138.4:c.2927G>A']['submitted_variant'] == '15-48782203-C-T'
        assert results['NM_000138.4:c.2927G>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_000138.4:c.2927G>A']['hgvs_lrg_variant'] == 'LRG_778:g.160783G>A'
        assert results['NM_000138.4:c.2927G>A']['hgvs_transcript_variant'] == 'NM_000138.4:c.2927G>A'
        assert results['NM_000138.4:c.2927G>A']['hgvs_refseqgene_variant'] == 'NG_008805.2:g.160783G>A'
        assert results['NM_000138.4:c.2927G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.48782203C>T', 'vcf': {'chr': 'chr15', 'ref': u'C', 'pos': '48782203', 'alt': u'T'}}
        assert results['NM_000138.4:c.2927G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000015.10:g.48490006C>T', 'vcf': {'chr': 'chr15', 'ref': u'C', 'pos': '48490006', 'alt': u'T'}}
        assert results['NM_000138.4:c.2927G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.48782203C>T', 'vcf': {'chr': '15', 'ref': u'C', 'pos': '48782203', 'alt': u'T'}}
        assert results['NM_000138.4:c.2927G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000015.10:g.48490006C>T', 'vcf': {'chr': '15', 'ref': u'C', 'pos': '48490006', 'alt': u'T'}}

        assert results['flag'] == 'gene_variant'

    def test_variant232(self):
        variant = '15-72105929-CC-C'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_014249.2:c.946_949=' in results.keys()
        assert results['NM_014249.2:c.946_949=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_014249.2:c.946_949=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_014249.2:c.946_949=']['alt_genomic_loci'] == []
        assert results['NM_014249.2:c.946_949=']['gene_symbol'] == 'NR2E3'
        assert results['NM_014249.2:c.946_949=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_055064.1:p.(Asp316=)', 'slr': 'NP_055064.1:p.(D316=)'}
        assert results['NM_014249.2:c.946_949=']['submitted_variant'] == '15-72105929-CC-C'
        assert results['NM_014249.2:c.946_949=']['genome_context_intronic_sequence'] == ''
        assert results['NM_014249.2:c.946_949=']['hgvs_lrg_variant'] == ''
        assert results['NM_014249.2:c.946_949=']['hgvs_transcript_variant'] == 'NM_014249.2:c.946_949='
        assert results['NM_014249.2:c.946_949=']['hgvs_refseqgene_variant'] == ''
        assert results['NM_014249.2:c.946_949=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105933del', 'vcf': {'chr': 'chr15', 'ref': 'AC', 'pos': '72105928', 'alt': 'A'}}
        assert 'hg38' not in results['NM_014249.2:c.946_949=']['primary_assembly_loci'].keys()
        assert results['NM_014249.2:c.946_949=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105933del', 'vcf': {'chr': '15', 'ref': 'AC', 'pos': '72105928', 'alt': 'A'}}
        assert 'grch38' not in results['NM_014249.2:c.946_949=']['primary_assembly_loci'].keys()

        assert results['flag'] == 'gene_variant'
        assert 'NM_014249.3:c.946_949=' in results.keys()
        assert results['NM_014249.3:c.946_949=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_014249.3:c.946_949=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_014249.3:c.946_949=']['alt_genomic_loci'] == []
        assert results['NM_014249.3:c.946_949=']['gene_symbol'] == 'NR2E3'
        assert results['NM_014249.3:c.946_949=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_055064.1:p.(Asp316=)', 'slr': 'NP_055064.1:p.(D316=)'}
        assert results['NM_014249.3:c.946_949=']['submitted_variant'] == '15-72105929-CC-C'
        assert results['NM_014249.3:c.946_949=']['genome_context_intronic_sequence'] == ''
        assert results['NM_014249.3:c.946_949=']['hgvs_lrg_variant'] == ''
        assert results['NM_014249.3:c.946_949=']['hgvs_transcript_variant'] == 'NM_014249.3:c.946_949='
        assert results['NM_014249.3:c.946_949=']['hgvs_refseqgene_variant'] == 'NG_009113.1:g.8034_8037='
        assert results['NM_014249.3:c.946_949=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105933del', 'vcf': {'chr': 'chr15', 'ref': 'AC', 'pos': '72105928', 'alt': 'A'}}
        assert results['NM_014249.3:c.946_949=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000015.10:g.71813586_71813591=', 'vcf': {'chr': 'chr15', 'ref': u'GGACCC', 'pos': '71813586', 'alt': u'GGACCC'}}
        assert results['NM_014249.3:c.946_949=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105933del', 'vcf': {'chr': '15', 'ref': 'AC', 'pos': '72105928', 'alt': 'A'}}
        assert results['NM_014249.3:c.946_949=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000015.10:g.71813586_71813591=', 'vcf': {'chr': '15', 'ref': u'GGACCC', 'pos': '71813586', 'alt': u'GGACCC'}}

        assert 'NM_016346.2:c.946_949=' in results.keys()
        assert results['NM_016346.2:c.946_949=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_016346.2:c.946_949=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_016346.2:c.946_949=']['alt_genomic_loci'] == []
        assert results['NM_016346.2:c.946_949=']['gene_symbol'] == 'NR2E3'
        assert results['NM_016346.2:c.946_949=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_057430.1:p.(Asp316=)', 'slr': 'NP_057430.1:p.(D316=)'}
        assert results['NM_016346.2:c.946_949=']['submitted_variant'] == '15-72105929-CC-C'
        assert results['NM_016346.2:c.946_949=']['genome_context_intronic_sequence'] == ''
        assert results['NM_016346.2:c.946_949=']['hgvs_lrg_variant'] == ''
        assert results['NM_016346.2:c.946_949=']['hgvs_transcript_variant'] == 'NM_016346.2:c.946_949='
        assert results['NM_016346.2:c.946_949=']['hgvs_refseqgene_variant'] == ''
        assert results['NM_016346.2:c.946_949=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105933del', 'vcf': {'chr': 'chr15', 'ref': 'AC', 'pos': '72105928', 'alt': 'A'}}
        assert 'hg38' not in results['NM_016346.2:c.946_949=']['primary_assembly_loci'].keys()
        assert results['NM_016346.2:c.946_949=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105933del', 'vcf': {'chr': '15', 'ref': 'AC', 'pos': '72105928', 'alt': 'A'}}
        assert 'grch38' not in results['NM_016346.2:c.946_949=']['primary_assembly_loci'].keys()

        assert 'NM_016346.3:c.946_949=' in results.keys()
        assert results['NM_016346.3:c.946_949=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_016346.3:c.946_949=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_016346.3:c.946_949=']['alt_genomic_loci'] == []
        assert results['NM_016346.3:c.946_949=']['gene_symbol'] == 'NR2E3'
        assert results['NM_016346.3:c.946_949=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_057430.1:p.(Asp316=)', 'slr': 'NP_057430.1:p.(D316=)'}
        assert results['NM_016346.3:c.946_949=']['submitted_variant'] == '15-72105929-CC-C'
        assert results['NM_016346.3:c.946_949=']['genome_context_intronic_sequence'] == ''
        assert results['NM_016346.3:c.946_949=']['hgvs_lrg_variant'] == ''
        assert results['NM_016346.3:c.946_949=']['hgvs_transcript_variant'] == 'NM_016346.3:c.946_949='
        assert results['NM_016346.3:c.946_949=']['hgvs_refseqgene_variant'] == ''
        assert results['NM_016346.3:c.946_949=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105933del', 'vcf': {'chr': 'chr15', 'ref': 'AC', 'pos': '72105928', 'alt': 'A'}}
        assert results['NM_016346.3:c.946_949=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000015.10:g.71813587_71813590=', 'vcf': {'chr': 'chr15', 'ref': u'GACC', 'pos': '71813587', 'alt': u'GACC'}}
        assert results['NM_016346.3:c.946_949=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.72105933del', 'vcf': {'chr': '15', 'ref': 'AC', 'pos': '72105928', 'alt': 'A'}}
        assert results['NM_016346.3:c.946_949=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000015.10:g.71813587_71813590=', 'vcf': {'chr': '15', 'ref': u'GACC', 'pos': '71813587', 'alt': u'GACC'}}


    def test_variant233(self):
        variant = '15-89873415-G-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_002693.2:c.752C>T' in results.keys()
        assert results['NM_002693.2:c.752C>T']['hgvs_lrg_transcript_variant'] == 'LRG_765t1:c.752C>T'
        assert results['NM_002693.2:c.752C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_002693.2:c.752C>T']['alt_genomic_loci'] == []
        assert results['NM_002693.2:c.752C>T']['gene_symbol'] == 'POLG'
        assert results['NM_002693.2:c.752C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_002684.1(LRG_765p1):p.(Thr251Ile)', 'slr': 'NP_002684.1:p.(T251I)'}
        assert results['NM_002693.2:c.752C>T']['submitted_variant'] == '15-89873415-G-A'
        assert results['NM_002693.2:c.752C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_002693.2:c.752C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_002693.2:c.752C>T']['hgvs_transcript_variant'] == 'NM_002693.2:c.752C>T'
        assert results['NM_002693.2:c.752C>T']['hgvs_refseqgene_variant'] == 'NG_008218.1:g.9612C>T'
        assert results['NM_002693.2:c.752C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.89873415G>A', 'vcf': {'chr': 'chr15', 'ref': u'G', 'pos': '89873415', 'alt': u'A'}}
        assert results['NM_002693.2:c.752C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000015.10:g.89330184G>A', 'vcf': {'chr': 'chr15', 'ref': u'G', 'pos': '89330184', 'alt': u'A'}}
        assert results['NM_002693.2:c.752C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.89873415G>A', 'vcf': {'chr': '15', 'ref': u'G', 'pos': '89873415', 'alt': u'A'}}
        assert results['NM_002693.2:c.752C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000015.10:g.89330184G>A', 'vcf': {'chr': '15', 'ref': u'G', 'pos': '89330184', 'alt': u'A'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001126131.1:c.752C>T' in results.keys()
        assert results['NM_001126131.1:c.752C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001126131.1:c.752C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001126131.1:c.752C>T']['alt_genomic_loci'] == []
        assert results['NM_001126131.1:c.752C>T']['gene_symbol'] == 'POLG'
        assert results['NM_001126131.1:c.752C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001119603.1:p.(Thr251Ile)', 'slr': 'NP_001119603.1:p.(T251I)'}
        assert results['NM_001126131.1:c.752C>T']['submitted_variant'] == '15-89873415-G-A'
        assert results['NM_001126131.1:c.752C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001126131.1:c.752C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001126131.1:c.752C>T']['hgvs_transcript_variant'] == 'NM_001126131.1:c.752C>T'
        assert results['NM_001126131.1:c.752C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001126131.1:c.752C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.89873415G>A', 'vcf': {'chr': 'chr15', 'ref': u'G', 'pos': '89873415', 'alt': u'A'}}
        assert results['NM_001126131.1:c.752C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000015.10:g.89330184G>A', 'vcf': {'chr': 'chr15', 'ref': u'G', 'pos': '89330184', 'alt': u'A'}}
        assert results['NM_001126131.1:c.752C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.89873415G>A', 'vcf': {'chr': '15', 'ref': u'G', 'pos': '89873415', 'alt': u'A'}}
        assert results['NM_001126131.1:c.752C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000015.10:g.89330184G>A', 'vcf': {'chr': '15', 'ref': u'G', 'pos': '89330184', 'alt': u'A'}}


    def test_variant234(self):
        variant = '16-2103394-C-T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_000548.3:c.277C>T' in results.keys()
        assert results['NM_000548.3:c.277C>T']['hgvs_lrg_transcript_variant'] == 'LRG_487t1:c.277C>T'
        assert results['NM_000548.3:c.277C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000548.3:c.277C>T']['alt_genomic_loci'] == []
        assert results['NM_000548.3:c.277C>T']['gene_symbol'] == 'TSC2'
        assert results['NM_000548.3:c.277C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000539.2(LRG_487p1):p.(Arg93Trp)', 'slr': 'NP_000539.2:p.(R93W)'}
        assert results['NM_000548.3:c.277C>T']['submitted_variant'] == '16-2103394-C-T'
        assert results['NM_000548.3:c.277C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_000548.3:c.277C>T']['hgvs_lrg_variant'] == 'LRG_487:g.9088C>T'
        assert results['NM_000548.3:c.277C>T']['hgvs_transcript_variant'] == 'NM_000548.3:c.277C>T'
        assert results['NM_000548.3:c.277C>T']['hgvs_refseqgene_variant'] == 'NG_005895.1:g.9088C>T'
        assert results['NM_000548.3:c.277C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.2103394C>T', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '2103394', 'alt': 'T'}}
        assert 'hg38' not in results['NM_000548.3:c.277C>T']['primary_assembly_loci'].keys()
        assert results['NM_000548.3:c.277C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.2103394C>T', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '2103394', 'alt': 'T'}}
        assert 'grch38' not in results['NM_000548.3:c.277C>T']['primary_assembly_loci'].keys()

        assert 'NM_001318832.1:c.310C>T' in results.keys()
        assert results['NM_001318832.1:c.310C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001318832.1:c.310C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001318832.1:c.310C>T']['alt_genomic_loci'] == []
        assert results['NM_001318832.1:c.310C>T']['gene_symbol'] == 'TSC2'
        assert results['NM_001318832.1:c.310C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001305761.1:p.(Arg104Trp)', 'slr': 'NP_001305761.1:p.(R104W)'}
        assert results['NM_001318832.1:c.310C>T']['submitted_variant'] == '16-2103394-C-T'
        assert results['NM_001318832.1:c.310C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001318832.1:c.310C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001318832.1:c.310C>T']['hgvs_transcript_variant'] == 'NM_001318832.1:c.310C>T'
        assert results['NM_001318832.1:c.310C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001318832.1:c.310C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.2103394C>T', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '2103394', 'alt': 'T'}}
        assert results['NM_001318832.1:c.310C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.2053393C>T', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '2053393', 'alt': 'T'}}
        assert results['NM_001318832.1:c.310C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.2103394C>T', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '2103394', 'alt': 'T'}}
        assert results['NM_001318832.1:c.310C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.2053393C>T', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '2053393', 'alt': 'T'}}

        assert 'NM_001318829.1:c.130C>T' in results.keys()
        assert results['NM_001318829.1:c.130C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001318829.1:c.130C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001318829.1:c.130C>T']['alt_genomic_loci'] == []
        assert results['NM_001318829.1:c.130C>T']['gene_symbol'] == 'TSC2'
        assert results['NM_001318829.1:c.130C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001305758.1:p.(Arg44Trp)', 'slr': 'NP_001305758.1:p.(R44W)'}
        assert results['NM_001318829.1:c.130C>T']['submitted_variant'] == '16-2103394-C-T'
        assert results['NM_001318829.1:c.130C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001318829.1:c.130C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001318829.1:c.130C>T']['hgvs_transcript_variant'] == 'NM_001318829.1:c.130C>T'
        assert results['NM_001318829.1:c.130C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001318829.1:c.130C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.2103394C>T', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '2103394', 'alt': 'T'}}
        assert results['NM_001318829.1:c.130C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.2053393C>T', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '2053393', 'alt': 'T'}}
        assert results['NM_001318829.1:c.130C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.2103394C>T', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '2103394', 'alt': 'T'}}
        assert results['NM_001318829.1:c.130C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.2053393C>T', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '2053393', 'alt': 'T'}}

        assert 'NM_000548.4:c.277C>T' in results.keys()
        assert results['NM_000548.4:c.277C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000548.4:c.277C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000548.4:c.277C>T']['alt_genomic_loci'] == []
        assert results['NM_000548.4:c.277C>T']['gene_symbol'] == 'TSC2'
        assert results['NM_000548.4:c.277C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000539.2(LRG_487p1):p.(Arg93Trp)', 'slr': 'NP_000539.2:p.(R93W)'}
        assert results['NM_000548.4:c.277C>T']['submitted_variant'] == '16-2103394-C-T'
        assert results['NM_000548.4:c.277C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_000548.4:c.277C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_000548.4:c.277C>T']['hgvs_transcript_variant'] == 'NM_000548.4:c.277C>T'
        assert results['NM_000548.4:c.277C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_000548.4:c.277C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.2103394C>T', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '2103394', 'alt': 'T'}}
        assert results['NM_000548.4:c.277C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.2053393C>T', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '2053393', 'alt': 'T'}}
        assert results['NM_000548.4:c.277C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.2103394C>T', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '2103394', 'alt': 'T'}}
        assert results['NM_000548.4:c.277C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.2053393C>T', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '2053393', 'alt': 'T'}}

        assert 'NM_001114382.1:c.277C>T' in results.keys()
        assert results['NM_001114382.1:c.277C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001114382.1:c.277C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001114382.1:c.277C>T']['alt_genomic_loci'] == []
        assert results['NM_001114382.1:c.277C>T']['gene_symbol'] == 'TSC2'
        assert results['NM_001114382.1:c.277C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001107854.1:p.(Arg93Trp)', 'slr': 'NP_001107854.1:p.(R93W)'}
        assert results['NM_001114382.1:c.277C>T']['submitted_variant'] == '16-2103394-C-T'
        assert results['NM_001114382.1:c.277C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001114382.1:c.277C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001114382.1:c.277C>T']['hgvs_transcript_variant'] == 'NM_001114382.1:c.277C>T'
        assert results['NM_001114382.1:c.277C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001114382.1:c.277C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.2103394C>T', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '2103394', 'alt': 'T'}}
        assert 'hg38' not in results['NM_001114382.1:c.277C>T']['primary_assembly_loci'].keys()
        assert results['NM_001114382.1:c.277C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.2103394C>T', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '2103394', 'alt': 'T'}}
        assert 'grch38' not in results['NM_001114382.1:c.277C>T']['primary_assembly_loci'].keys()

        assert 'NM_001077183.1:c.277C>T' in results.keys()
        assert results['NM_001077183.1:c.277C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001077183.1:c.277C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001077183.1:c.277C>T']['alt_genomic_loci'] == []
        assert results['NM_001077183.1:c.277C>T']['gene_symbol'] == 'TSC2'
        assert results['NM_001077183.1:c.277C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001070651.1:p.(Arg93Trp)', 'slr': 'NP_001070651.1:p.(R93W)'}
        assert results['NM_001077183.1:c.277C>T']['submitted_variant'] == '16-2103394-C-T'
        assert results['NM_001077183.1:c.277C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001077183.1:c.277C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001077183.1:c.277C>T']['hgvs_transcript_variant'] == 'NM_001077183.1:c.277C>T'
        assert results['NM_001077183.1:c.277C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001077183.1:c.277C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.2103394C>T', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '2103394', 'alt': 'T'}}
        assert 'hg38' not in results['NM_001077183.1:c.277C>T']['primary_assembly_loci'].keys()
        assert results['NM_001077183.1:c.277C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.2103394C>T', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '2103394', 'alt': 'T'}}
        assert 'grch38' not in results['NM_001077183.1:c.277C>T']['primary_assembly_loci'].keys()

        assert 'NM_001077183.2:c.277C>T' in results.keys()
        assert results['NM_001077183.2:c.277C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001077183.2:c.277C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001077183.2:c.277C>T']['alt_genomic_loci'] == []
        assert results['NM_001077183.2:c.277C>T']['gene_symbol'] == 'TSC2'
        assert results['NM_001077183.2:c.277C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001070651.1:p.(Arg93Trp)', 'slr': 'NP_001070651.1:p.(R93W)'}
        assert results['NM_001077183.2:c.277C>T']['submitted_variant'] == '16-2103394-C-T'
        assert results['NM_001077183.2:c.277C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001077183.2:c.277C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001077183.2:c.277C>T']['hgvs_transcript_variant'] == 'NM_001077183.2:c.277C>T'
        assert results['NM_001077183.2:c.277C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001077183.2:c.277C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.2103394C>T', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '2103394', 'alt': 'T'}}
        assert results['NM_001077183.2:c.277C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.2053393C>T', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '2053393', 'alt': 'T'}}
        assert results['NM_001077183.2:c.277C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.2103394C>T', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '2103394', 'alt': 'T'}}
        assert results['NM_001077183.2:c.277C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.2053393C>T', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '2053393', 'alt': 'T'}}

        assert 'NM_001318827.1:c.226-903C>T' in results.keys()
        assert results['NM_001318827.1:c.226-903C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001318827.1:c.226-903C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001318827.1:c.226-903C>T']['alt_genomic_loci'] == []
        assert results['NM_001318827.1:c.226-903C>T']['gene_symbol'] == 'TSC2'
        assert results['NM_001318827.1:c.226-903C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001305756.1:p.?', 'slr': 'NP_001305756.1:p.?'}
        assert results['NM_001318827.1:c.226-903C>T']['submitted_variant'] == '16-2103394-C-T'
        assert results['NM_001318827.1:c.226-903C>T']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_001318827.1):c.226-903C>T'
        assert results['NM_001318827.1:c.226-903C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001318827.1:c.226-903C>T']['hgvs_transcript_variant'] == 'NM_001318827.1:c.226-903C>T'
        assert results['NM_001318827.1:c.226-903C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001318827.1:c.226-903C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.2103394C>T', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '2103394', 'alt': 'T'}}
        assert results['NM_001318827.1:c.226-903C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.2053393C>T', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '2053393', 'alt': 'T'}}
        assert results['NM_001318827.1:c.226-903C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.2103394C>T', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '2103394', 'alt': 'T'}}
        assert results['NM_001318827.1:c.226-903C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.2053393C>T', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '2053393', 'alt': 'T'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001114382.2:c.277C>T' in results.keys()
        assert results['NM_001114382.2:c.277C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001114382.2:c.277C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001114382.2:c.277C>T']['alt_genomic_loci'] == []
        assert results['NM_001114382.2:c.277C>T']['gene_symbol'] == 'TSC2'
        assert results['NM_001114382.2:c.277C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001107854.1:p.(Arg93Trp)', 'slr': 'NP_001107854.1:p.(R93W)'}
        assert results['NM_001114382.2:c.277C>T']['submitted_variant'] == '16-2103394-C-T'
        assert results['NM_001114382.2:c.277C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001114382.2:c.277C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001114382.2:c.277C>T']['hgvs_transcript_variant'] == 'NM_001114382.2:c.277C>T'
        assert results['NM_001114382.2:c.277C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001114382.2:c.277C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.2103394C>T', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '2103394', 'alt': 'T'}}
        assert results['NM_001114382.2:c.277C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.2053393C>T', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '2053393', 'alt': 'T'}}
        assert results['NM_001114382.2:c.277C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.2103394C>T', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '2103394', 'alt': 'T'}}
        assert results['NM_001114382.2:c.277C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.2053393C>T', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '2053393', 'alt': 'T'}}

        assert 'NM_001363528.1:c.277C>T' in results.keys()
        assert results['NM_001363528.1:c.277C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001363528.1:c.277C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001363528.1:c.277C>T']['alt_genomic_loci'] == []
        assert results['NM_001363528.1:c.277C>T']['gene_symbol'] == 'TSC2'
        assert results['NM_001363528.1:c.277C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001350457.1:p.(Arg93Trp)', 'slr': 'NP_001350457.1:p.(R93W)'}
        assert results['NM_001363528.1:c.277C>T']['submitted_variant'] == '16-2103394-C-T'
        assert results['NM_001363528.1:c.277C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001363528.1:c.277C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001363528.1:c.277C>T']['hgvs_transcript_variant'] == 'NM_001363528.1:c.277C>T'
        assert results['NM_001363528.1:c.277C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001363528.1:c.277C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.2103394C>T', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '2103394', 'alt': 'T'}}
        assert 'hg38' not in results['NM_001363528.1:c.277C>T']['primary_assembly_loci'].keys()
        assert results['NM_001363528.1:c.277C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.2103394C>T', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '2103394', 'alt': 'T'}}
        assert 'grch38' not in results['NM_001363528.1:c.277C>T']['primary_assembly_loci'].keys()

        assert 'NM_021055.2:c.277C>T' in results.keys()
        assert results['NM_021055.2:c.277C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_021055.2:c.277C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_021055.2:c.277C>T']['alt_genomic_loci'] == []
        assert results['NM_021055.2:c.277C>T']['gene_symbol'] == 'TSC2'
        assert results['NM_021055.2:c.277C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_066399.2:p.(Arg93Trp)', 'slr': 'NP_066399.2:p.(R93W)'}
        assert results['NM_021055.2:c.277C>T']['submitted_variant'] == '16-2103394-C-T'
        assert results['NM_021055.2:c.277C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_021055.2:c.277C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_021055.2:c.277C>T']['hgvs_transcript_variant'] == 'NM_021055.2:c.277C>T'
        assert results['NM_021055.2:c.277C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_021055.2:c.277C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.2103394C>T', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '2103394', 'alt': 'T'}}
        assert 'hg38' not in results['NM_021055.2:c.277C>T']['primary_assembly_loci'].keys()
        assert results['NM_021055.2:c.277C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.2103394C>T', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '2103394', 'alt': 'T'}}
        assert 'grch38' not in results['NM_021055.2:c.277C>T']['primary_assembly_loci'].keys()

        assert 'NM_001318831.1:c.-1-2803C>T' in results.keys()
        assert results['NM_001318831.1:c.-1-2803C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001318831.1:c.-1-2803C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001318831.1:c.-1-2803C>T']['alt_genomic_loci'] == []
        assert results['NM_001318831.1:c.-1-2803C>T']['gene_symbol'] == 'TSC2'
        assert results['NM_001318831.1:c.-1-2803C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001305760.1:p.?', 'slr': 'NP_001305760.1:p.?'}
        assert results['NM_001318831.1:c.-1-2803C>T']['submitted_variant'] == '16-2103394-C-T'
        assert results['NM_001318831.1:c.-1-2803C>T']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_001318831.1):c.-1-2803C>T'
        assert results['NM_001318831.1:c.-1-2803C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001318831.1:c.-1-2803C>T']['hgvs_transcript_variant'] == 'NM_001318831.1:c.-1-2803C>T'
        assert results['NM_001318831.1:c.-1-2803C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001318831.1:c.-1-2803C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.2103394C>T', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '2103394', 'alt': 'T'}}
        assert results['NM_001318831.1:c.-1-2803C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.2053393C>T', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '2053393', 'alt': 'T'}}
        assert results['NM_001318831.1:c.-1-2803C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.2103394C>T', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '2103394', 'alt': 'T'}}
        assert results['NM_001318831.1:c.-1-2803C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.2053393C>T', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '2053393', 'alt': 'T'}}


    def test_variant235(self):
        variant = '16-3779300-C-G'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_001079846.1:c.5634G>C' in results.keys()
        assert results['NM_001079846.1:c.5634G>C']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001079846.1:c.5634G>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001079846.1:c.5634G>C']['alt_genomic_loci'] == []
        assert results['NM_001079846.1:c.5634G>C']['gene_symbol'] == 'CREBBP'
        assert results['NM_001079846.1:c.5634G>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001073315.1:p.(Met1878Ile)', 'slr': 'NP_001073315.1:p.(M1878I)'}
        assert results['NM_001079846.1:c.5634G>C']['submitted_variant'] == '16-3779300-C-G'
        assert results['NM_001079846.1:c.5634G>C']['genome_context_intronic_sequence'] == ''
        assert results['NM_001079846.1:c.5634G>C']['hgvs_lrg_variant'] == ''
        assert results['NM_001079846.1:c.5634G>C']['hgvs_transcript_variant'] == 'NM_001079846.1:c.5634G>C'
        assert results['NM_001079846.1:c.5634G>C']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001079846.1:c.5634G>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.3779300C>G', 'vcf': {'chr': 'chr16', 'ref': u'C', 'pos': '3779300', 'alt': u'G'}}
        assert results['NM_001079846.1:c.5634G>C']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.3729299C>G', 'vcf': {'chr': 'chr16', 'ref': u'C', 'pos': '3729299', 'alt': u'G'}}
        assert results['NM_001079846.1:c.5634G>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.3779300C>G', 'vcf': {'chr': '16', 'ref': u'C', 'pos': '3779300', 'alt': u'G'}}
        assert results['NM_001079846.1:c.5634G>C']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.3729299C>G', 'vcf': {'chr': '16', 'ref': u'C', 'pos': '3729299', 'alt': u'G'}}

        assert 'NM_004380.2:c.5748G>C' in results.keys()
        assert results['NM_004380.2:c.5748G>C']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_004380.2:c.5748G>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_004380.2:c.5748G>C']['alt_genomic_loci'] == []
        assert results['NM_004380.2:c.5748G>C']['gene_symbol'] == 'CREBBP'
        assert results['NM_004380.2:c.5748G>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_004371.2:p.(Met1916Ile)', 'slr': 'NP_004371.2:p.(M1916I)'}
        assert results['NM_004380.2:c.5748G>C']['submitted_variant'] == '16-3779300-C-G'
        assert results['NM_004380.2:c.5748G>C']['genome_context_intronic_sequence'] == ''
        assert results['NM_004380.2:c.5748G>C']['hgvs_lrg_variant'] == ''
        assert results['NM_004380.2:c.5748G>C']['hgvs_transcript_variant'] == 'NM_004380.2:c.5748G>C'
        assert results['NM_004380.2:c.5748G>C']['hgvs_refseqgene_variant'] == 'NG_009873.1:g.155822G>C'
        assert results['NM_004380.2:c.5748G>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.3779300C>G', 'vcf': {'chr': 'chr16', 'ref': u'C', 'pos': '3779300', 'alt': u'G'}}
        assert results['NM_004380.2:c.5748G>C']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.3729299C>G', 'vcf': {'chr': 'chr16', 'ref': u'C', 'pos': '3729299', 'alt': u'G'}}
        assert results['NM_004380.2:c.5748G>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.3779300C>G', 'vcf': {'chr': '16', 'ref': u'C', 'pos': '3779300', 'alt': u'G'}}
        assert results['NM_004380.2:c.5748G>C']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.3729299C>G', 'vcf': {'chr': '16', 'ref': u'C', 'pos': '3729299', 'alt': u'G'}}


    def test_variant236(self):
        variant = '16-5128843-C-G'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001330504.1:c.493C>G' in results.keys()
        assert results['NM_001330504.1:c.493C>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001330504.1:c.493C>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001330504.1:c.493C>G']['alt_genomic_loci'] == []
        assert results['NM_001330504.1:c.493C>G']['gene_symbol'] == 'ALG1'
        assert results['NM_001330504.1:c.493C>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001317433.1:p.(Arg165Gly)', 'slr': 'NP_001317433.1:p.(R165G)'}
        assert results['NM_001330504.1:c.493C>G']['submitted_variant'] == '16-5128843-C-G'
        assert results['NM_001330504.1:c.493C>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001330504.1:c.493C>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001330504.1:c.493C>G']['hgvs_transcript_variant'] == 'NM_001330504.1:c.493C>G'
        assert results['NM_001330504.1:c.493C>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001330504.1:c.493C>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.5128843C>G', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '5128843', 'alt': 'G'}}
        assert results['NM_001330504.1:c.493C>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.5078842C>G', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '5078842', 'alt': 'G'}}
        assert results['NM_001330504.1:c.493C>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.5128843C>G', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '5128843', 'alt': 'G'}}
        assert results['NM_001330504.1:c.493C>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.5078842C>G', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '5078842', 'alt': 'G'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_019109.4:c.826C>G' in results.keys()
        assert results['NM_019109.4:c.826C>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_019109.4:c.826C>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_019109.4:c.826C>G']['alt_genomic_loci'] == []
        assert results['NM_019109.4:c.826C>G']['gene_symbol'] == 'ALG1'
        assert results['NM_019109.4:c.826C>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_061982.3:p.(Arg276Gly)', 'slr': 'NP_061982.3:p.(R276G)'}
        assert results['NM_019109.4:c.826C>G']['submitted_variant'] == '16-5128843-C-G'
        assert results['NM_019109.4:c.826C>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_019109.4:c.826C>G']['hgvs_lrg_variant'] == ''
        assert results['NM_019109.4:c.826C>G']['hgvs_transcript_variant'] == 'NM_019109.4:c.826C>G'
        assert results['NM_019109.4:c.826C>G']['hgvs_refseqgene_variant'] == 'NG_009202.1:g.12034C>G'
        assert results['NM_019109.4:c.826C>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.5128843C>G', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '5128843', 'alt': 'G'}}
        assert results['NM_019109.4:c.826C>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.5078842C>G', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '5078842', 'alt': 'G'}}
        assert results['NM_019109.4:c.826C>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.5128843C>G', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '5128843', 'alt': 'G'}}
        assert results['NM_019109.4:c.826C>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.5078842C>G', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '5078842', 'alt': 'G'}}


    def test_variant237(self):
        variant = '16-74808559-C-T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_024306.4:c.95G>A' in results.keys()
        assert results['NM_024306.4:c.95G>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_024306.4:c.95G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_024306.4:c.95G>A']['alt_genomic_loci'] == []
        assert results['NM_024306.4:c.95G>A']['gene_symbol'] == 'FA2H'
        assert results['NM_024306.4:c.95G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_077282.3:p.(Arg32His)', 'slr': 'NP_077282.3:p.(R32H)'}
        assert results['NM_024306.4:c.95G>A']['submitted_variant'] == '16-74808559-C-T'
        assert results['NM_024306.4:c.95G>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_024306.4:c.95G>A']['hgvs_lrg_variant'] == ''
        assert results['NM_024306.4:c.95G>A']['hgvs_transcript_variant'] == 'NM_024306.4:c.95G>A'
        assert results['NM_024306.4:c.95G>A']['hgvs_refseqgene_variant'] == 'NG_017070.1:g.5171G>A'
        assert results['NM_024306.4:c.95G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.74808559C>T', 'vcf': {'chr': 'chr16', 'ref': u'C', 'pos': '74808559', 'alt': u'T'}}
        assert results['NM_024306.4:c.95G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.74774661C>T', 'vcf': {'chr': 'chr16', 'ref': u'C', 'pos': '74774661', 'alt': u'T'}}
        assert results['NM_024306.4:c.95G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.74808559C>T', 'vcf': {'chr': '16', 'ref': u'C', 'pos': '74808559', 'alt': u'T'}}
        assert results['NM_024306.4:c.95G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.74774661C>T', 'vcf': {'chr': '16', 'ref': u'C', 'pos': '74774661', 'alt': u'T'}}

        assert results['flag'] == 'gene_variant'

    def test_variant238(self):
        variant = '16-89574804-C-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_003119.3:c.-22C>A' in results.keys()
        assert results['NM_003119.3:c.-22C>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003119.3:c.-22C>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003119.3:c.-22C>A']['alt_genomic_loci'] == []
        assert results['NM_003119.3:c.-22C>A']['gene_symbol'] == 'SPG7'
        assert results['NM_003119.3:c.-22C>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003110.1:p.?', 'slr': 'NP_003110.1:p.?'}
        assert results['NM_003119.3:c.-22C>A']['submitted_variant'] == '16-89574804-C-A'
        assert results['NM_003119.3:c.-22C>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_003119.3:c.-22C>A']['hgvs_lrg_variant'] == ''
        assert results['NM_003119.3:c.-22C>A']['hgvs_transcript_variant'] == 'NM_003119.3:c.-22C>A'
        assert results['NM_003119.3:c.-22C>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_003119.3:c.-22C>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574804C>A', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '89574804', 'alt': 'A'}}
        assert results['NM_003119.3:c.-22C>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89508396C>A', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '89508396', 'alt': 'A'}}
        assert results['NM_003119.3:c.-22C>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574804C>A', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '89574804', 'alt': 'A'}}
        assert results['NM_003119.3:c.-22C>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89508396C>A', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '89508396', 'alt': 'A'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_199367.2:c.-22C>A' in results.keys()
        assert results['NM_199367.2:c.-22C>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_199367.2:c.-22C>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_199367.2:c.-22C>A']['alt_genomic_loci'] == []
        assert results['NM_199367.2:c.-22C>A']['gene_symbol'] == 'SPG7'
        assert results['NM_199367.2:c.-22C>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_955399.1:p.?', 'slr': 'NP_955399.1:p.?'}
        assert results['NM_199367.2:c.-22C>A']['submitted_variant'] == '16-89574804-C-A'
        assert results['NM_199367.2:c.-22C>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_199367.2:c.-22C>A']['hgvs_lrg_variant'] == ''
        assert results['NM_199367.2:c.-22C>A']['hgvs_transcript_variant'] == 'NM_199367.2:c.-22C>A'
        assert results['NM_199367.2:c.-22C>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_199367.2:c.-22C>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574804C>A', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '89574804', 'alt': 'A'}}
        assert results['NM_199367.2:c.-22C>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89508396C>A', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '89508396', 'alt': 'A'}}
        assert results['NM_199367.2:c.-22C>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574804C>A', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '89574804', 'alt': 'A'}}
        assert results['NM_199367.2:c.-22C>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89508396C>A', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '89508396', 'alt': 'A'}}

        assert 'NM_001363850.1:c.-22C>A' in results.keys()
        assert results['NM_001363850.1:c.-22C>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001363850.1:c.-22C>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001363850.1:c.-22C>A']['alt_genomic_loci'] == []
        assert results['NM_001363850.1:c.-22C>A']['gene_symbol'] == 'SPG7'
        assert results['NM_001363850.1:c.-22C>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001350779.1:p.?', 'slr': 'NP_001350779.1:p.?'}
        assert results['NM_001363850.1:c.-22C>A']['submitted_variant'] == '16-89574804-C-A'
        assert results['NM_001363850.1:c.-22C>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_001363850.1:c.-22C>A']['hgvs_lrg_variant'] == ''
        assert results['NM_001363850.1:c.-22C>A']['hgvs_transcript_variant'] == 'NM_001363850.1:c.-22C>A'
        assert results['NM_001363850.1:c.-22C>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001363850.1:c.-22C>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574804C>A', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '89574804', 'alt': 'A'}}
        assert 'hg38' not in results['NM_001363850.1:c.-22C>A']['primary_assembly_loci'].keys()
        assert results['NM_001363850.1:c.-22C>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574804C>A', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '89574804', 'alt': 'A'}}
        assert 'grch38' not in results['NM_001363850.1:c.-22C>A']['primary_assembly_loci'].keys()


    def test_variant239(self):
        variant = '16-89574826-A-C'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_003119.2:c.1A>C' in results.keys()
        assert results['NM_003119.2:c.1A>C']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003119.2:c.1A>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003119.2:c.1A>C']['alt_genomic_loci'] == []
        assert results['NM_003119.2:c.1A>C']['gene_symbol'] == 'SPG7'
        assert results['NM_003119.2:c.1A>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003110.1:p.(Met1?)', 'slr': 'NP_003110.1:p.(M1?)'}
        assert results['NM_003119.2:c.1A>C']['submitted_variant'] == '16-89574826-A-C'
        assert results['NM_003119.2:c.1A>C']['genome_context_intronic_sequence'] == ''
        assert results['NM_003119.2:c.1A>C']['hgvs_lrg_variant'] == ''
        assert results['NM_003119.2:c.1A>C']['hgvs_transcript_variant'] == 'NM_003119.2:c.1A>C'
        assert results['NM_003119.2:c.1A>C']['hgvs_refseqgene_variant'] == 'NG_008082.1:g.5022A>C'
        assert results['NM_003119.2:c.1A>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574826A>C', 'vcf': {'chr': 'chr16', 'ref': 'A', 'pos': '89574826', 'alt': 'C'}}
        assert 'hg38' not in results['NM_003119.2:c.1A>C']['primary_assembly_loci'].keys()
        assert results['NM_003119.2:c.1A>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574826A>C', 'vcf': {'chr': '16', 'ref': 'A', 'pos': '89574826', 'alt': 'C'}}
        assert 'grch38' not in results['NM_003119.2:c.1A>C']['primary_assembly_loci'].keys()

        assert 'NM_199367.1:c.1A>C' in results.keys()
        assert results['NM_199367.1:c.1A>C']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_199367.1:c.1A>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_199367.1:c.1A>C']['alt_genomic_loci'] == []
        assert results['NM_199367.1:c.1A>C']['gene_symbol'] == 'SPG7'
        assert results['NM_199367.1:c.1A>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_955399.1:p.(Met1?)', 'slr': 'NP_955399.1:p.(M1?)'}
        assert results['NM_199367.1:c.1A>C']['submitted_variant'] == '16-89574826-A-C'
        assert results['NM_199367.1:c.1A>C']['genome_context_intronic_sequence'] == ''
        assert results['NM_199367.1:c.1A>C']['hgvs_lrg_variant'] == ''
        assert results['NM_199367.1:c.1A>C']['hgvs_transcript_variant'] == 'NM_199367.1:c.1A>C'
        assert results['NM_199367.1:c.1A>C']['hgvs_refseqgene_variant'] == ''
        assert results['NM_199367.1:c.1A>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574826A>C', 'vcf': {'chr': 'chr16', 'ref': 'A', 'pos': '89574826', 'alt': 'C'}}
        assert 'hg38' not in results['NM_199367.1:c.1A>C']['primary_assembly_loci'].keys()
        assert results['NM_199367.1:c.1A>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574826A>C', 'vcf': {'chr': '16', 'ref': 'A', 'pos': '89574826', 'alt': 'C'}}
        assert 'grch38' not in results['NM_199367.1:c.1A>C']['primary_assembly_loci'].keys()

        assert 'NM_001363850.1:c.1A>C' in results.keys()
        assert results['NM_001363850.1:c.1A>C']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001363850.1:c.1A>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001363850.1:c.1A>C']['alt_genomic_loci'] == []
        assert results['NM_001363850.1:c.1A>C']['gene_symbol'] == 'SPG7'
        assert results['NM_001363850.1:c.1A>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001350779.1:p.(Met1?)', 'slr': 'NP_001350779.1:p.(M1?)'}
        assert results['NM_001363850.1:c.1A>C']['submitted_variant'] == '16-89574826-A-C'
        assert results['NM_001363850.1:c.1A>C']['genome_context_intronic_sequence'] == ''
        assert results['NM_001363850.1:c.1A>C']['hgvs_lrg_variant'] == ''
        assert results['NM_001363850.1:c.1A>C']['hgvs_transcript_variant'] == 'NM_001363850.1:c.1A>C'
        assert results['NM_001363850.1:c.1A>C']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001363850.1:c.1A>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574826A>C', 'vcf': {'chr': 'chr16', 'ref': 'A', 'pos': '89574826', 'alt': 'C'}}
        assert 'hg38' not in results['NM_001363850.1:c.1A>C']['primary_assembly_loci'].keys()
        assert results['NM_001363850.1:c.1A>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574826A>C', 'vcf': {'chr': '16', 'ref': 'A', 'pos': '89574826', 'alt': 'C'}}
        assert 'grch38' not in results['NM_001363850.1:c.1A>C']['primary_assembly_loci'].keys()

        assert 'NM_199367.2:c.1A>C' in results.keys()
        assert results['NM_199367.2:c.1A>C']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_199367.2:c.1A>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_199367.2:c.1A>C']['alt_genomic_loci'] == []
        assert results['NM_199367.2:c.1A>C']['gene_symbol'] == 'SPG7'
        assert results['NM_199367.2:c.1A>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_955399.1:p.(Met1?)', 'slr': 'NP_955399.1:p.(M1?)'}
        assert results['NM_199367.2:c.1A>C']['submitted_variant'] == '16-89574826-A-C'
        assert results['NM_199367.2:c.1A>C']['genome_context_intronic_sequence'] == ''
        assert results['NM_199367.2:c.1A>C']['hgvs_lrg_variant'] == ''
        assert results['NM_199367.2:c.1A>C']['hgvs_transcript_variant'] == 'NM_199367.2:c.1A>C'
        assert results['NM_199367.2:c.1A>C']['hgvs_refseqgene_variant'] == ''
        assert results['NM_199367.2:c.1A>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574826A>C', 'vcf': {'chr': 'chr16', 'ref': 'A', 'pos': '89574826', 'alt': 'C'}}
        assert results['NM_199367.2:c.1A>C']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89508418A>C', 'vcf': {'chr': 'chr16', 'ref': 'A', 'pos': '89508418', 'alt': 'C'}}
        assert results['NM_199367.2:c.1A>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574826A>C', 'vcf': {'chr': '16', 'ref': 'A', 'pos': '89574826', 'alt': 'C'}}
        assert results['NM_199367.2:c.1A>C']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89508418A>C', 'vcf': {'chr': '16', 'ref': 'A', 'pos': '89508418', 'alt': 'C'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_003119.3:c.1A>C' in results.keys()
        assert results['NM_003119.3:c.1A>C']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003119.3:c.1A>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003119.3:c.1A>C']['alt_genomic_loci'] == []
        assert results['NM_003119.3:c.1A>C']['gene_symbol'] == 'SPG7'
        assert results['NM_003119.3:c.1A>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003110.1:p.(Met1?)', 'slr': 'NP_003110.1:p.(M1?)'}
        assert results['NM_003119.3:c.1A>C']['submitted_variant'] == '16-89574826-A-C'
        assert results['NM_003119.3:c.1A>C']['genome_context_intronic_sequence'] == ''
        assert results['NM_003119.3:c.1A>C']['hgvs_lrg_variant'] == ''
        assert results['NM_003119.3:c.1A>C']['hgvs_transcript_variant'] == 'NM_003119.3:c.1A>C'
        assert results['NM_003119.3:c.1A>C']['hgvs_refseqgene_variant'] == ''
        assert results['NM_003119.3:c.1A>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574826A>C', 'vcf': {'chr': 'chr16', 'ref': 'A', 'pos': '89574826', 'alt': 'C'}}
        assert results['NM_003119.3:c.1A>C']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89508418A>C', 'vcf': {'chr': 'chr16', 'ref': 'A', 'pos': '89508418', 'alt': 'C'}}
        assert results['NM_003119.3:c.1A>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574826A>C', 'vcf': {'chr': '16', 'ref': 'A', 'pos': '89574826', 'alt': 'C'}}
        assert results['NM_003119.3:c.1A>C']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89508418A>C', 'vcf': {'chr': '16', 'ref': 'A', 'pos': '89508418', 'alt': 'C'}}


    def test_variant240(self):
        variant = '16-89574914-G-GT'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001363850.1:c.90dup' in results.keys()
        assert results['NM_001363850.1:c.90dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001363850.1:c.90dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001363850.1:c.90dup']['alt_genomic_loci'] == []
        assert results['NM_001363850.1:c.90dup']['gene_symbol'] == 'SPG7'
        assert results['NM_001363850.1:c.90dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001350779.1:p.(Pro31SerfsTer43)', 'slr': 'NP_001350779.1:p.(P31Sfs*43)'}
        assert results['NM_001363850.1:c.90dup']['submitted_variant'] == '16-89574914-G-GT'
        assert results['NM_001363850.1:c.90dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_001363850.1:c.90dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001363850.1:c.90dup']['hgvs_transcript_variant'] == 'NM_001363850.1:c.90dup'
        assert results['NM_001363850.1:c.90dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001363850.1:c.90dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574915dup', 'vcf': {'chr': 'chr16', 'ref': 'T', 'pos': '89574915', 'alt': 'TT'}}
        assert 'hg38' not in results['NM_001363850.1:c.90dup']['primary_assembly_loci'].keys()
        assert results['NM_001363850.1:c.90dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574915dup', 'vcf': {'chr': '16', 'ref': 'T', 'pos': '89574915', 'alt': 'TT'}}
        assert 'grch38' not in results['NM_001363850.1:c.90dup']['primary_assembly_loci'].keys()

        assert 'NM_199367.1:c.90dup' in results.keys()
        assert results['NM_199367.1:c.90dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_199367.1:c.90dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_199367.1:c.90dup']['alt_genomic_loci'] == []
        assert results['NM_199367.1:c.90dup']['gene_symbol'] == 'SPG7'
        assert results['NM_199367.1:c.90dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_955399.1:p.(Pro31SerfsTer43)', 'slr': 'NP_955399.1:p.(P31Sfs*43)'}
        assert results['NM_199367.1:c.90dup']['submitted_variant'] == '16-89574914-G-GT'
        assert results['NM_199367.1:c.90dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_199367.1:c.90dup']['hgvs_lrg_variant'] == ''
        assert results['NM_199367.1:c.90dup']['hgvs_transcript_variant'] == 'NM_199367.1:c.90dup'
        assert results['NM_199367.1:c.90dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_199367.1:c.90dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574915dup', 'vcf': {'chr': 'chr16', 'ref': 'T', 'pos': '89574915', 'alt': 'TT'}}
        assert 'hg38' not in results['NM_199367.1:c.90dup']['primary_assembly_loci'].keys()
        assert results['NM_199367.1:c.90dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574915dup', 'vcf': {'chr': '16', 'ref': 'T', 'pos': '89574915', 'alt': 'TT'}}
        assert 'grch38' not in results['NM_199367.1:c.90dup']['primary_assembly_loci'].keys()

        assert 'NM_003119.2:c.90dup' in results.keys()
        assert results['NM_003119.2:c.90dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003119.2:c.90dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003119.2:c.90dup']['alt_genomic_loci'] == []
        assert results['NM_003119.2:c.90dup']['gene_symbol'] == 'SPG7'
        assert results['NM_003119.2:c.90dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003110.1:p.(Pro31SerfsTer43)', 'slr': 'NP_003110.1:p.(P31Sfs*43)'}
        assert results['NM_003119.2:c.90dup']['submitted_variant'] == '16-89574914-G-GT'
        assert results['NM_003119.2:c.90dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_003119.2:c.90dup']['hgvs_lrg_variant'] == ''
        assert results['NM_003119.2:c.90dup']['hgvs_transcript_variant'] == 'NM_003119.2:c.90dup'
        assert results['NM_003119.2:c.90dup']['hgvs_refseqgene_variant'] == 'NG_008082.1:g.5111dup'
        assert results['NM_003119.2:c.90dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574915dup', 'vcf': {'chr': 'chr16', 'ref': 'T', 'pos': '89574915', 'alt': 'TT'}}
        assert 'hg38' not in results['NM_003119.2:c.90dup']['primary_assembly_loci'].keys()
        assert results['NM_003119.2:c.90dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574915dup', 'vcf': {'chr': '16', 'ref': 'T', 'pos': '89574915', 'alt': 'TT'}}
        assert 'grch38' not in results['NM_003119.2:c.90dup']['primary_assembly_loci'].keys()

        assert 'NM_199367.2:c.90dup' in results.keys()
        assert results['NM_199367.2:c.90dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_199367.2:c.90dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_199367.2:c.90dup']['alt_genomic_loci'] == []
        assert results['NM_199367.2:c.90dup']['gene_symbol'] == 'SPG7'
        assert results['NM_199367.2:c.90dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_955399.1:p.(Pro31SerfsTer43)', 'slr': 'NP_955399.1:p.(P31Sfs*43)'}
        assert results['NM_199367.2:c.90dup']['submitted_variant'] == '16-89574914-G-GT'
        assert results['NM_199367.2:c.90dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_199367.2:c.90dup']['hgvs_lrg_variant'] == ''
        assert results['NM_199367.2:c.90dup']['hgvs_transcript_variant'] == 'NM_199367.2:c.90dup'
        assert results['NM_199367.2:c.90dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_199367.2:c.90dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574915dup', 'vcf': {'chr': 'chr16', 'ref': 'T', 'pos': '89574915', 'alt': 'TT'}}
        assert results['NM_199367.2:c.90dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89508507dup', 'vcf': {'chr': 'chr16', 'ref': 'T', 'pos': '89508507', 'alt': 'TT'}}
        assert results['NM_199367.2:c.90dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574915dup', 'vcf': {'chr': '16', 'ref': 'T', 'pos': '89574915', 'alt': 'TT'}}
        assert results['NM_199367.2:c.90dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89508507dup', 'vcf': {'chr': '16', 'ref': 'T', 'pos': '89508507', 'alt': 'TT'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_003119.3:c.90dup' in results.keys()
        assert results['NM_003119.3:c.90dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003119.3:c.90dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003119.3:c.90dup']['alt_genomic_loci'] == []
        assert results['NM_003119.3:c.90dup']['gene_symbol'] == 'SPG7'
        assert results['NM_003119.3:c.90dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003110.1:p.(Pro31SerfsTer43)', 'slr': 'NP_003110.1:p.(P31Sfs*43)'}
        assert results['NM_003119.3:c.90dup']['submitted_variant'] == '16-89574914-G-GT'
        assert results['NM_003119.3:c.90dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_003119.3:c.90dup']['hgvs_lrg_variant'] == ''
        assert results['NM_003119.3:c.90dup']['hgvs_transcript_variant'] == 'NM_003119.3:c.90dup'
        assert results['NM_003119.3:c.90dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_003119.3:c.90dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574915dup', 'vcf': {'chr': 'chr16', 'ref': 'T', 'pos': '89574915', 'alt': 'TT'}}
        assert results['NM_003119.3:c.90dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89508507dup', 'vcf': {'chr': 'chr16', 'ref': 'T', 'pos': '89508507', 'alt': 'TT'}}
        assert results['NM_003119.3:c.90dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574915dup', 'vcf': {'chr': '16', 'ref': 'T', 'pos': '89574915', 'alt': 'TT'}}
        assert results['NM_003119.3:c.90dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89508507dup', 'vcf': {'chr': '16', 'ref': 'T', 'pos': '89508507', 'alt': 'TT'}}


    def test_variant241(self):
        variant = '16-89574916-C-CGTC'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_199367.2:c.89_91dup' in results.keys()
        assert results['NM_199367.2:c.89_91dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_199367.2:c.89_91dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_199367.2:c.89_91dup']['alt_genomic_loci'] == []
        assert results['NM_199367.2:c.89_91dup']['gene_symbol'] == 'SPG7'
        assert results['NM_199367.2:c.89_91dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_955399.1:p.(Ser30_Pro31insArg)', 'slr': 'NP_955399.1:p.(S30_P31insR)'}
        assert results['NM_199367.2:c.89_91dup']['submitted_variant'] == '16-89574916-C-CGTC'
        assert results['NM_199367.2:c.89_91dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_199367.2:c.89_91dup']['hgvs_lrg_variant'] == ''
        assert results['NM_199367.2:c.89_91dup']['hgvs_transcript_variant'] == 'NM_199367.2:c.89_91dup'
        assert results['NM_199367.2:c.89_91dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_199367.2:c.89_91dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574914_89574916dup', 'vcf': {'chr': 'chr16', 'ref': 'GTC', 'pos': '89574914', 'alt': 'GTCGTC'}}
        assert results['NM_199367.2:c.89_91dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89508506_89508508dup', 'vcf': {'chr': 'chr16', 'ref': 'GTC', 'pos': '89508506', 'alt': 'GTCGTC'}}
        assert results['NM_199367.2:c.89_91dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574914_89574916dup', 'vcf': {'chr': '16', 'ref': 'GTC', 'pos': '89574914', 'alt': 'GTCGTC'}}
        assert results['NM_199367.2:c.89_91dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89508506_89508508dup', 'vcf': {'chr': '16', 'ref': 'GTC', 'pos': '89508506', 'alt': 'GTCGTC'}}

        assert 'NM_003119.3:c.89_91dup' in results.keys()
        assert results['NM_003119.3:c.89_91dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003119.3:c.89_91dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003119.3:c.89_91dup']['alt_genomic_loci'] == []
        assert results['NM_003119.3:c.89_91dup']['gene_symbol'] == 'SPG7'
        assert results['NM_003119.3:c.89_91dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003110.1:p.(Ser30_Pro31insArg)', 'slr': 'NP_003110.1:p.(S30_P31insR)'}
        assert results['NM_003119.3:c.89_91dup']['submitted_variant'] == '16-89574916-C-CGTC'
        assert results['NM_003119.3:c.89_91dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_003119.3:c.89_91dup']['hgvs_lrg_variant'] == ''
        assert results['NM_003119.3:c.89_91dup']['hgvs_transcript_variant'] == 'NM_003119.3:c.89_91dup'
        assert results['NM_003119.3:c.89_91dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_003119.3:c.89_91dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574914_89574916dup', 'vcf': {'chr': 'chr16', 'ref': 'GTC', 'pos': '89574914', 'alt': 'GTCGTC'}}
        assert results['NM_003119.3:c.89_91dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89508506_89508508dup', 'vcf': {'chr': 'chr16', 'ref': 'GTC', 'pos': '89508506', 'alt': 'GTCGTC'}}
        assert results['NM_003119.3:c.89_91dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574914_89574916dup', 'vcf': {'chr': '16', 'ref': 'GTC', 'pos': '89574914', 'alt': 'GTCGTC'}}
        assert results['NM_003119.3:c.89_91dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89508506_89508508dup', 'vcf': {'chr': '16', 'ref': 'GTC', 'pos': '89508506', 'alt': 'GTCGTC'}}

        assert 'NM_001363850.1:c.89_91dup' in results.keys()
        assert results['NM_001363850.1:c.89_91dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001363850.1:c.89_91dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001363850.1:c.89_91dup']['alt_genomic_loci'] == []
        assert results['NM_001363850.1:c.89_91dup']['gene_symbol'] == 'SPG7'
        assert results['NM_001363850.1:c.89_91dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001350779.1:p.(Ser30_Pro31insArg)', 'slr': 'NP_001350779.1:p.(S30_P31insR)'}
        assert results['NM_001363850.1:c.89_91dup']['submitted_variant'] == '16-89574916-C-CGTC'
        assert results['NM_001363850.1:c.89_91dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_001363850.1:c.89_91dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001363850.1:c.89_91dup']['hgvs_transcript_variant'] == 'NM_001363850.1:c.89_91dup'
        assert results['NM_001363850.1:c.89_91dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001363850.1:c.89_91dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574914_89574916dup', 'vcf': {'chr': 'chr16', 'ref': 'GTC', 'pos': '89574914', 'alt': 'GTCGTC'}}
        assert 'hg38' not in results['NM_001363850.1:c.89_91dup']['primary_assembly_loci'].keys()
        assert results['NM_001363850.1:c.89_91dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574914_89574916dup', 'vcf': {'chr': '16', 'ref': 'GTC', 'pos': '89574914', 'alt': 'GTCGTC'}}
        assert 'grch38' not in results['NM_001363850.1:c.89_91dup']['primary_assembly_loci'].keys()

        assert results['flag'] == 'gene_variant'
        assert 'NM_199367.1:c.89_91dup' in results.keys()
        assert results['NM_199367.1:c.89_91dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_199367.1:c.89_91dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_199367.1:c.89_91dup']['alt_genomic_loci'] == []
        assert results['NM_199367.1:c.89_91dup']['gene_symbol'] == 'SPG7'
        assert results['NM_199367.1:c.89_91dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_955399.1:p.(Ser30_Pro31insArg)', 'slr': 'NP_955399.1:p.(S30_P31insR)'}
        assert results['NM_199367.1:c.89_91dup']['submitted_variant'] == '16-89574916-C-CGTC'
        assert results['NM_199367.1:c.89_91dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_199367.1:c.89_91dup']['hgvs_lrg_variant'] == ''
        assert results['NM_199367.1:c.89_91dup']['hgvs_transcript_variant'] == 'NM_199367.1:c.89_91dup'
        assert results['NM_199367.1:c.89_91dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_199367.1:c.89_91dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574914_89574916dup', 'vcf': {'chr': 'chr16', 'ref': 'GTC', 'pos': '89574914', 'alt': 'GTCGTC'}}
        assert 'hg38' not in results['NM_199367.1:c.89_91dup']['primary_assembly_loci'].keys()
        assert results['NM_199367.1:c.89_91dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574914_89574916dup', 'vcf': {'chr': '16', 'ref': 'GTC', 'pos': '89574914', 'alt': 'GTCGTC'}}
        assert 'grch38' not in results['NM_199367.1:c.89_91dup']['primary_assembly_loci'].keys()

        assert 'NM_003119.2:c.89_91dup' in results.keys()
        assert results['NM_003119.2:c.89_91dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003119.2:c.89_91dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003119.2:c.89_91dup']['alt_genomic_loci'] == []
        assert results['NM_003119.2:c.89_91dup']['gene_symbol'] == 'SPG7'
        assert results['NM_003119.2:c.89_91dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003110.1:p.(Ser30_Pro31insArg)', 'slr': 'NP_003110.1:p.(S30_P31insR)'}
        assert results['NM_003119.2:c.89_91dup']['submitted_variant'] == '16-89574916-C-CGTC'
        assert results['NM_003119.2:c.89_91dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_003119.2:c.89_91dup']['hgvs_lrg_variant'] == ''
        assert results['NM_003119.2:c.89_91dup']['hgvs_transcript_variant'] == 'NM_003119.2:c.89_91dup'
        assert results['NM_003119.2:c.89_91dup']['hgvs_refseqgene_variant'] == 'NG_008082.1:g.5110_5112dup'
        assert results['NM_003119.2:c.89_91dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574914_89574916dup', 'vcf': {'chr': 'chr16', 'ref': 'GTC', 'pos': '89574914', 'alt': 'GTCGTC'}}
        assert 'hg38' not in results['NM_003119.2:c.89_91dup']['primary_assembly_loci'].keys()
        assert results['NM_003119.2:c.89_91dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89574914_89574916dup', 'vcf': {'chr': '16', 'ref': 'GTC', 'pos': '89574914', 'alt': 'GTCGTC'}}
        assert 'grch38' not in results['NM_003119.2:c.89_91dup']['primary_assembly_loci'].keys()


    def test_variant242(self):
        variant = '16-89575009-G-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_003119.2:c.183+1G>A' in results.keys()
        assert results['NM_003119.2:c.183+1G>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003119.2:c.183+1G>A']['refseqgene_context_intronic_sequence'] == 'NG_008082.1(NM_003119.2):c.183+1G>A'
        assert results['NM_003119.2:c.183+1G>A']['alt_genomic_loci'] == []
        assert results['NM_003119.2:c.183+1G>A']['gene_symbol'] == 'SPG7'
        assert results['NM_003119.2:c.183+1G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003110.1:p.?', 'slr': 'NP_003110.1:p.?'}
        assert results['NM_003119.2:c.183+1G>A']['submitted_variant'] == '16-89575009-G-A'
        assert results['NM_003119.2:c.183+1G>A']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_003119.2):c.183+1G>A'
        assert results['NM_003119.2:c.183+1G>A']['hgvs_lrg_variant'] == ''
        assert results['NM_003119.2:c.183+1G>A']['hgvs_transcript_variant'] == 'NM_003119.2:c.183+1G>A'
        assert results['NM_003119.2:c.183+1G>A']['hgvs_refseqgene_variant'] == 'NG_008082.1:g.5205G>A'
        assert results['NM_003119.2:c.183+1G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89575009G>A', 'vcf': {'chr': 'chr16', 'ref': 'G', 'pos': '89575009', 'alt': 'A'}}
        assert 'hg38' not in results['NM_003119.2:c.183+1G>A']['primary_assembly_loci'].keys()
        assert results['NM_003119.2:c.183+1G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89575009G>A', 'vcf': {'chr': '16', 'ref': 'G', 'pos': '89575009', 'alt': 'A'}}
        assert 'grch38' not in results['NM_003119.2:c.183+1G>A']['primary_assembly_loci'].keys()

        assert 'NM_199367.2:c.183+1G>A' in results.keys()
        assert results['NM_199367.2:c.183+1G>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_199367.2:c.183+1G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_199367.2:c.183+1G>A']['alt_genomic_loci'] == []
        assert results['NM_199367.2:c.183+1G>A']['gene_symbol'] == 'SPG7'
        assert results['NM_199367.2:c.183+1G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_955399.1:p.?', 'slr': 'NP_955399.1:p.?'}
        assert results['NM_199367.2:c.183+1G>A']['submitted_variant'] == '16-89575009-G-A'
        assert results['NM_199367.2:c.183+1G>A']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_199367.2):c.183+1G>A'
        assert results['NM_199367.2:c.183+1G>A']['hgvs_lrg_variant'] == ''
        assert results['NM_199367.2:c.183+1G>A']['hgvs_transcript_variant'] == 'NM_199367.2:c.183+1G>A'
        assert results['NM_199367.2:c.183+1G>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_199367.2:c.183+1G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89575009G>A', 'vcf': {'chr': 'chr16', 'ref': 'G', 'pos': '89575009', 'alt': 'A'}}
        assert results['NM_199367.2:c.183+1G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89508601G>A', 'vcf': {'chr': 'chr16', 'ref': 'G', 'pos': '89508601', 'alt': 'A'}}
        assert results['NM_199367.2:c.183+1G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89575009G>A', 'vcf': {'chr': '16', 'ref': 'G', 'pos': '89575009', 'alt': 'A'}}
        assert results['NM_199367.2:c.183+1G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89508601G>A', 'vcf': {'chr': '16', 'ref': 'G', 'pos': '89508601', 'alt': 'A'}}

        assert 'NM_199367.1:c.183+1G>A' in results.keys()
        assert results['NM_199367.1:c.183+1G>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_199367.1:c.183+1G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_199367.1:c.183+1G>A']['alt_genomic_loci'] == []
        assert results['NM_199367.1:c.183+1G>A']['gene_symbol'] == 'SPG7'
        assert results['NM_199367.1:c.183+1G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_955399.1:p.?', 'slr': 'NP_955399.1:p.?'}
        assert results['NM_199367.1:c.183+1G>A']['submitted_variant'] == '16-89575009-G-A'
        assert results['NM_199367.1:c.183+1G>A']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_199367.1):c.183+1G>A'
        assert results['NM_199367.1:c.183+1G>A']['hgvs_lrg_variant'] == ''
        assert results['NM_199367.1:c.183+1G>A']['hgvs_transcript_variant'] == 'NM_199367.1:c.183+1G>A'
        assert results['NM_199367.1:c.183+1G>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_199367.1:c.183+1G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89575009G>A', 'vcf': {'chr': 'chr16', 'ref': 'G', 'pos': '89575009', 'alt': 'A'}}
        assert 'hg38' not in results['NM_199367.1:c.183+1G>A']['primary_assembly_loci'].keys()
        assert results['NM_199367.1:c.183+1G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89575009G>A', 'vcf': {'chr': '16', 'ref': 'G', 'pos': '89575009', 'alt': 'A'}}
        assert 'grch38' not in results['NM_199367.1:c.183+1G>A']['primary_assembly_loci'].keys()

        assert 'NM_001363850.1:c.183+1G>A' in results.keys()
        assert results['NM_001363850.1:c.183+1G>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001363850.1:c.183+1G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001363850.1:c.183+1G>A']['alt_genomic_loci'] == []
        assert results['NM_001363850.1:c.183+1G>A']['gene_symbol'] == 'SPG7'
        assert results['NM_001363850.1:c.183+1G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001350779.1:p.?', 'slr': 'NP_001350779.1:p.?'}
        assert results['NM_001363850.1:c.183+1G>A']['submitted_variant'] == '16-89575009-G-A'
        assert results['NM_001363850.1:c.183+1G>A']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_001363850.1):c.183+1G>A'
        assert results['NM_001363850.1:c.183+1G>A']['hgvs_lrg_variant'] == ''
        assert results['NM_001363850.1:c.183+1G>A']['hgvs_transcript_variant'] == 'NM_001363850.1:c.183+1G>A'
        assert results['NM_001363850.1:c.183+1G>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001363850.1:c.183+1G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89575009G>A', 'vcf': {'chr': 'chr16', 'ref': 'G', 'pos': '89575009', 'alt': 'A'}}
        assert 'hg38' not in results['NM_001363850.1:c.183+1G>A']['primary_assembly_loci'].keys()
        assert results['NM_001363850.1:c.183+1G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89575009G>A', 'vcf': {'chr': '16', 'ref': 'G', 'pos': '89575009', 'alt': 'A'}}
        assert 'grch38' not in results['NM_001363850.1:c.183+1G>A']['primary_assembly_loci'].keys()

        assert 'NM_003119.3:c.183+1G>A' in results.keys()
        assert results['NM_003119.3:c.183+1G>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003119.3:c.183+1G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003119.3:c.183+1G>A']['alt_genomic_loci'] == []
        assert results['NM_003119.3:c.183+1G>A']['gene_symbol'] == 'SPG7'
        assert results['NM_003119.3:c.183+1G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003110.1:p.?', 'slr': 'NP_003110.1:p.?'}
        assert results['NM_003119.3:c.183+1G>A']['submitted_variant'] == '16-89575009-G-A'
        assert results['NM_003119.3:c.183+1G>A']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_003119.3):c.183+1G>A'
        assert results['NM_003119.3:c.183+1G>A']['hgvs_lrg_variant'] == ''
        assert results['NM_003119.3:c.183+1G>A']['hgvs_transcript_variant'] == 'NM_003119.3:c.183+1G>A'
        assert results['NM_003119.3:c.183+1G>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_003119.3:c.183+1G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89575009G>A', 'vcf': {'chr': 'chr16', 'ref': 'G', 'pos': '89575009', 'alt': 'A'}}
        assert results['NM_003119.3:c.183+1G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89508601G>A', 'vcf': {'chr': 'chr16', 'ref': 'G', 'pos': '89508601', 'alt': 'A'}}
        assert results['NM_003119.3:c.183+1G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89575009G>A', 'vcf': {'chr': '16', 'ref': 'G', 'pos': '89575009', 'alt': 'A'}}
        assert results['NM_003119.3:c.183+1G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89508601G>A', 'vcf': {'chr': '16', 'ref': 'G', 'pos': '89508601', 'alt': 'A'}}


    def test_variant243(self):
        variant = '16-89575040-C-A,CA'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_199367.1:c.183+32_183+33insA' in results.keys()
        assert results['NM_199367.1:c.183+32_183+33insA']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_199367.1:c.183+32_183+33insA']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_199367.1:c.183+32_183+33insA']['alt_genomic_loci'] == []
        assert results['NM_199367.1:c.183+32_183+33insA']['gene_symbol'] == 'SPG7'
        assert results['NM_199367.1:c.183+32_183+33insA']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_955399.1:p.?', 'slr': 'NP_955399.1:p.?'}
        assert results['NM_199367.1:c.183+32_183+33insA']['submitted_variant'] == '16-89575040-C-A,CA'
        assert results['NM_199367.1:c.183+32_183+33insA']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_199367.1):c.183+32_183+33insA'
        assert results['NM_199367.1:c.183+32_183+33insA']['hgvs_lrg_variant'] == ''
        assert results['NM_199367.1:c.183+32_183+33insA']['hgvs_transcript_variant'] == 'NM_199367.1:c.183+32_183+33insA'
        assert results['NM_199367.1:c.183+32_183+33insA']['hgvs_refseqgene_variant'] == ''
        assert results['NM_199367.1:c.183+32_183+33insA']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89575040_89575041insA', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '89575040', 'alt': 'CA'}}
        assert 'hg38' not in results['NM_199367.1:c.183+32_183+33insA']['primary_assembly_loci'].keys()
        assert results['NM_199367.1:c.183+32_183+33insA']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89575040_89575041insA', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '89575040', 'alt': 'CA'}}
        assert 'grch38' not in results['NM_199367.1:c.183+32_183+33insA']['primary_assembly_loci'].keys()

        assert 'NM_001363850.1:c.183+32C>A' in results.keys()
        assert results['NM_001363850.1:c.183+32C>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001363850.1:c.183+32C>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001363850.1:c.183+32C>A']['alt_genomic_loci'] == []
        assert results['NM_001363850.1:c.183+32C>A']['gene_symbol'] == 'SPG7'
        assert results['NM_001363850.1:c.183+32C>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001350779.1:p.?', 'slr': 'NP_001350779.1:p.?'}
        assert results['NM_001363850.1:c.183+32C>A']['submitted_variant'] == '16-89575040-C-A,CA'
        assert results['NM_001363850.1:c.183+32C>A']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_001363850.1):c.183+32C>A'
        assert results['NM_001363850.1:c.183+32C>A']['hgvs_lrg_variant'] == ''
        assert results['NM_001363850.1:c.183+32C>A']['hgvs_transcript_variant'] == 'NM_001363850.1:c.183+32C>A'
        assert results['NM_001363850.1:c.183+32C>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001363850.1:c.183+32C>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89575040C>A', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '89575040', 'alt': 'A'}}
        assert 'hg38' not in results['NM_001363850.1:c.183+32C>A']['primary_assembly_loci'].keys()
        assert results['NM_001363850.1:c.183+32C>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89575040C>A', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '89575040', 'alt': 'A'}}
        assert 'grch38' not in results['NM_001363850.1:c.183+32C>A']['primary_assembly_loci'].keys()

        assert 'NM_001363850.1:c.183+32_183+33insA' in results.keys()
        assert results['NM_001363850.1:c.183+32_183+33insA']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001363850.1:c.183+32_183+33insA']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001363850.1:c.183+32_183+33insA']['alt_genomic_loci'] == []
        assert results['NM_001363850.1:c.183+32_183+33insA']['gene_symbol'] == 'SPG7'
        assert results['NM_001363850.1:c.183+32_183+33insA']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001350779.1:p.?', 'slr': 'NP_001350779.1:p.?'}
        assert results['NM_001363850.1:c.183+32_183+33insA']['submitted_variant'] == '16-89575040-C-A,CA'
        assert results['NM_001363850.1:c.183+32_183+33insA']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_001363850.1):c.183+32_183+33insA'
        assert results['NM_001363850.1:c.183+32_183+33insA']['hgvs_lrg_variant'] == ''
        assert results['NM_001363850.1:c.183+32_183+33insA']['hgvs_transcript_variant'] == 'NM_001363850.1:c.183+32_183+33insA'
        assert results['NM_001363850.1:c.183+32_183+33insA']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001363850.1:c.183+32_183+33insA']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89575040_89575041insA', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '89575040', 'alt': 'CA'}}
        assert 'hg38' not in results['NM_001363850.1:c.183+32_183+33insA']['primary_assembly_loci'].keys()
        assert results['NM_001363850.1:c.183+32_183+33insA']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89575040_89575041insA', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '89575040', 'alt': 'CA'}}
        assert 'grch38' not in results['NM_001363850.1:c.183+32_183+33insA']['primary_assembly_loci'].keys()

        assert 'NM_199367.2:c.183+32C>A' in results.keys()
        assert results['NM_199367.2:c.183+32C>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_199367.2:c.183+32C>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_199367.2:c.183+32C>A']['alt_genomic_loci'] == []
        assert results['NM_199367.2:c.183+32C>A']['gene_symbol'] == 'SPG7'
        assert results['NM_199367.2:c.183+32C>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_955399.1:p.?', 'slr': 'NP_955399.1:p.?'}
        assert results['NM_199367.2:c.183+32C>A']['submitted_variant'] == '16-89575040-C-A,CA'
        assert results['NM_199367.2:c.183+32C>A']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_199367.2):c.183+32C>A'
        assert results['NM_199367.2:c.183+32C>A']['hgvs_lrg_variant'] == ''
        assert results['NM_199367.2:c.183+32C>A']['hgvs_transcript_variant'] == 'NM_199367.2:c.183+32C>A'
        assert results['NM_199367.2:c.183+32C>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_199367.2:c.183+32C>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89575040C>A', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '89575040', 'alt': 'A'}}
        assert results['NM_199367.2:c.183+32C>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89508632C>A', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '89508632', 'alt': 'A'}}
        assert results['NM_199367.2:c.183+32C>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89575040C>A', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '89575040', 'alt': 'A'}}
        assert results['NM_199367.2:c.183+32C>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89508632C>A', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '89508632', 'alt': 'A'}}

        assert 'NM_003119.3:c.183+32_183+33insA' in results.keys()
        assert results['NM_003119.3:c.183+32_183+33insA']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003119.3:c.183+32_183+33insA']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003119.3:c.183+32_183+33insA']['alt_genomic_loci'] == []
        assert results['NM_003119.3:c.183+32_183+33insA']['gene_symbol'] == 'SPG7'
        assert results['NM_003119.3:c.183+32_183+33insA']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003110.1:p.?', 'slr': 'NP_003110.1:p.?'}
        assert results['NM_003119.3:c.183+32_183+33insA']['submitted_variant'] == '16-89575040-C-A,CA'
        assert results['NM_003119.3:c.183+32_183+33insA']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_003119.3):c.183+32_183+33insA'
        assert results['NM_003119.3:c.183+32_183+33insA']['hgvs_lrg_variant'] == ''
        assert results['NM_003119.3:c.183+32_183+33insA']['hgvs_transcript_variant'] == 'NM_003119.3:c.183+32_183+33insA'
        assert results['NM_003119.3:c.183+32_183+33insA']['hgvs_refseqgene_variant'] == ''
        assert results['NM_003119.3:c.183+32_183+33insA']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89575040_89575041insA', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '89575040', 'alt': 'CA'}}
        assert results['NM_003119.3:c.183+32_183+33insA']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89508632_89508633insA', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '89508632', 'alt': 'CA'}}
        assert results['NM_003119.3:c.183+32_183+33insA']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89575040_89575041insA', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '89575040', 'alt': 'CA'}}
        assert results['NM_003119.3:c.183+32_183+33insA']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89508632_89508633insA', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '89508632', 'alt': 'CA'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_003119.2:c.183+32_183+33insA' in results.keys()
        assert results['NM_003119.2:c.183+32_183+33insA']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003119.2:c.183+32_183+33insA']['refseqgene_context_intronic_sequence'] == 'NG_008082.1(NM_003119.2):c.183+32_183+33insA'
        assert results['NM_003119.2:c.183+32_183+33insA']['alt_genomic_loci'] == []
        assert results['NM_003119.2:c.183+32_183+33insA']['gene_symbol'] == 'SPG7'
        assert results['NM_003119.2:c.183+32_183+33insA']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003110.1:p.?', 'slr': 'NP_003110.1:p.?'}
        assert results['NM_003119.2:c.183+32_183+33insA']['submitted_variant'] == '16-89575040-C-A,CA'
        assert results['NM_003119.2:c.183+32_183+33insA']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_003119.2):c.183+32_183+33insA'
        assert results['NM_003119.2:c.183+32_183+33insA']['hgvs_lrg_variant'] == ''
        assert results['NM_003119.2:c.183+32_183+33insA']['hgvs_transcript_variant'] == 'NM_003119.2:c.183+32_183+33insA'
        assert results['NM_003119.2:c.183+32_183+33insA']['hgvs_refseqgene_variant'] == 'NG_008082.1:g.5236_5237insA'
        assert results['NM_003119.2:c.183+32_183+33insA']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89575040_89575041insA', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '89575040', 'alt': 'CA'}}
        assert 'hg38' not in results['NM_003119.2:c.183+32_183+33insA']['primary_assembly_loci'].keys()
        assert results['NM_003119.2:c.183+32_183+33insA']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89575040_89575041insA', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '89575040', 'alt': 'CA'}}
        assert 'grch38' not in results['NM_003119.2:c.183+32_183+33insA']['primary_assembly_loci'].keys()

        assert 'NM_199367.1:c.183+32C>A' in results.keys()
        assert results['NM_199367.1:c.183+32C>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_199367.1:c.183+32C>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_199367.1:c.183+32C>A']['alt_genomic_loci'] == []
        assert results['NM_199367.1:c.183+32C>A']['gene_symbol'] == 'SPG7'
        assert results['NM_199367.1:c.183+32C>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_955399.1:p.?', 'slr': 'NP_955399.1:p.?'}
        assert results['NM_199367.1:c.183+32C>A']['submitted_variant'] == '16-89575040-C-A,CA'
        assert results['NM_199367.1:c.183+32C>A']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_199367.1):c.183+32C>A'
        assert results['NM_199367.1:c.183+32C>A']['hgvs_lrg_variant'] == ''
        assert results['NM_199367.1:c.183+32C>A']['hgvs_transcript_variant'] == 'NM_199367.1:c.183+32C>A'
        assert results['NM_199367.1:c.183+32C>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_199367.1:c.183+32C>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89575040C>A', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '89575040', 'alt': 'A'}}
        assert 'hg38' not in results['NM_199367.1:c.183+32C>A']['primary_assembly_loci'].keys()
        assert results['NM_199367.1:c.183+32C>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89575040C>A', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '89575040', 'alt': 'A'}}
        assert 'grch38' not in results['NM_199367.1:c.183+32C>A']['primary_assembly_loci'].keys()

        assert 'NM_003119.3:c.183+32C>A' in results.keys()
        assert results['NM_003119.3:c.183+32C>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003119.3:c.183+32C>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003119.3:c.183+32C>A']['alt_genomic_loci'] == []
        assert results['NM_003119.3:c.183+32C>A']['gene_symbol'] == 'SPG7'
        assert results['NM_003119.3:c.183+32C>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003110.1:p.?', 'slr': 'NP_003110.1:p.?'}
        assert results['NM_003119.3:c.183+32C>A']['submitted_variant'] == '16-89575040-C-A,CA'
        assert results['NM_003119.3:c.183+32C>A']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_003119.3):c.183+32C>A'
        assert results['NM_003119.3:c.183+32C>A']['hgvs_lrg_variant'] == ''
        assert results['NM_003119.3:c.183+32C>A']['hgvs_transcript_variant'] == 'NM_003119.3:c.183+32C>A'
        assert results['NM_003119.3:c.183+32C>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_003119.3:c.183+32C>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89575040C>A', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '89575040', 'alt': 'A'}}
        assert results['NM_003119.3:c.183+32C>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89508632C>A', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '89508632', 'alt': 'A'}}
        assert results['NM_003119.3:c.183+32C>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89575040C>A', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '89575040', 'alt': 'A'}}
        assert results['NM_003119.3:c.183+32C>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89508632C>A', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '89508632', 'alt': 'A'}}

        assert 'NM_199367.2:c.183+32_183+33insA' in results.keys()
        assert results['NM_199367.2:c.183+32_183+33insA']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_199367.2:c.183+32_183+33insA']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_199367.2:c.183+32_183+33insA']['alt_genomic_loci'] == []
        assert results['NM_199367.2:c.183+32_183+33insA']['gene_symbol'] == 'SPG7'
        assert results['NM_199367.2:c.183+32_183+33insA']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_955399.1:p.?', 'slr': 'NP_955399.1:p.?'}
        assert results['NM_199367.2:c.183+32_183+33insA']['submitted_variant'] == '16-89575040-C-A,CA'
        assert results['NM_199367.2:c.183+32_183+33insA']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_199367.2):c.183+32_183+33insA'
        assert results['NM_199367.2:c.183+32_183+33insA']['hgvs_lrg_variant'] == ''
        assert results['NM_199367.2:c.183+32_183+33insA']['hgvs_transcript_variant'] == 'NM_199367.2:c.183+32_183+33insA'
        assert results['NM_199367.2:c.183+32_183+33insA']['hgvs_refseqgene_variant'] == ''
        assert results['NM_199367.2:c.183+32_183+33insA']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89575040_89575041insA', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '89575040', 'alt': 'CA'}}
        assert results['NM_199367.2:c.183+32_183+33insA']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89508632_89508633insA', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '89508632', 'alt': 'CA'}}
        assert results['NM_199367.2:c.183+32_183+33insA']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89575040_89575041insA', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '89575040', 'alt': 'CA'}}
        assert results['NM_199367.2:c.183+32_183+33insA']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89508632_89508633insA', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '89508632', 'alt': 'CA'}}

        assert 'NM_003119.2:c.183+32C>A' in results.keys()
        assert results['NM_003119.2:c.183+32C>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003119.2:c.183+32C>A']['refseqgene_context_intronic_sequence'] == 'NG_008082.1(NM_003119.2):c.183+32C>A'
        assert results['NM_003119.2:c.183+32C>A']['alt_genomic_loci'] == []
        assert results['NM_003119.2:c.183+32C>A']['gene_symbol'] == 'SPG7'
        assert results['NM_003119.2:c.183+32C>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003110.1:p.?', 'slr': 'NP_003110.1:p.?'}
        assert results['NM_003119.2:c.183+32C>A']['submitted_variant'] == '16-89575040-C-A,CA'
        assert results['NM_003119.2:c.183+32C>A']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_003119.2):c.183+32C>A'
        assert results['NM_003119.2:c.183+32C>A']['hgvs_lrg_variant'] == ''
        assert results['NM_003119.2:c.183+32C>A']['hgvs_transcript_variant'] == 'NM_003119.2:c.183+32C>A'
        assert results['NM_003119.2:c.183+32C>A']['hgvs_refseqgene_variant'] == 'NG_008082.1:g.5236C>A'
        assert results['NM_003119.2:c.183+32C>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89575040C>A', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '89575040', 'alt': 'A'}}
        assert 'hg38' not in results['NM_003119.2:c.183+32C>A']['primary_assembly_loci'].keys()
        assert results['NM_003119.2:c.183+32C>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89575040C>A', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '89575040', 'alt': 'A'}}
        assert 'grch38' not in results['NM_003119.2:c.183+32C>A']['primary_assembly_loci'].keys()


    def test_variant244(self):
        variant = '16-89576896-A-C'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_199367.2:c.184-2A>C' in results.keys()
        assert results['NM_199367.2:c.184-2A>C']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_199367.2:c.184-2A>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_199367.2:c.184-2A>C']['alt_genomic_loci'] == []
        assert results['NM_199367.2:c.184-2A>C']['gene_symbol'] == 'SPG7'
        assert results['NM_199367.2:c.184-2A>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_955399.1:p.?', 'slr': 'NP_955399.1:p.?'}
        assert results['NM_199367.2:c.184-2A>C']['submitted_variant'] == '16-89576896-A-C'
        assert results['NM_199367.2:c.184-2A>C']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_199367.2):c.184-2A>C'
        assert results['NM_199367.2:c.184-2A>C']['hgvs_lrg_variant'] == ''
        assert results['NM_199367.2:c.184-2A>C']['hgvs_transcript_variant'] == 'NM_199367.2:c.184-2A>C'
        assert results['NM_199367.2:c.184-2A>C']['hgvs_refseqgene_variant'] == ''
        assert results['NM_199367.2:c.184-2A>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576896A>C', 'vcf': {'chr': 'chr16', 'ref': 'A', 'pos': '89576896', 'alt': 'C'}}
        assert results['NM_199367.2:c.184-2A>C']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89510488A>C', 'vcf': {'chr': 'chr16', 'ref': 'A', 'pos': '89510488', 'alt': 'C'}}
        assert results['NM_199367.2:c.184-2A>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576896A>C', 'vcf': {'chr': '16', 'ref': 'A', 'pos': '89576896', 'alt': 'C'}}
        assert results['NM_199367.2:c.184-2A>C']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89510488A>C', 'vcf': {'chr': '16', 'ref': 'A', 'pos': '89510488', 'alt': 'C'}}

        assert 'NM_003119.2:c.184-2A>C' in results.keys()
        assert results['NM_003119.2:c.184-2A>C']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003119.2:c.184-2A>C']['refseqgene_context_intronic_sequence'] == 'NG_008082.1(NM_003119.2):c.184-2A>C'
        assert results['NM_003119.2:c.184-2A>C']['alt_genomic_loci'] == []
        assert results['NM_003119.2:c.184-2A>C']['gene_symbol'] == 'SPG7'
        assert results['NM_003119.2:c.184-2A>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003110.1:p.?', 'slr': 'NP_003110.1:p.?'}
        assert results['NM_003119.2:c.184-2A>C']['submitted_variant'] == '16-89576896-A-C'
        assert results['NM_003119.2:c.184-2A>C']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_003119.2):c.184-2A>C'
        assert results['NM_003119.2:c.184-2A>C']['hgvs_lrg_variant'] == ''
        assert results['NM_003119.2:c.184-2A>C']['hgvs_transcript_variant'] == 'NM_003119.2:c.184-2A>C'
        assert results['NM_003119.2:c.184-2A>C']['hgvs_refseqgene_variant'] == 'NG_008082.1:g.7092A>C'
        assert results['NM_003119.2:c.184-2A>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576896A>C', 'vcf': {'chr': 'chr16', 'ref': 'A', 'pos': '89576896', 'alt': 'C'}}
        assert 'hg38' not in results['NM_003119.2:c.184-2A>C']['primary_assembly_loci'].keys()
        assert results['NM_003119.2:c.184-2A>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576896A>C', 'vcf': {'chr': '16', 'ref': 'A', 'pos': '89576896', 'alt': 'C'}}
        assert 'grch38' not in results['NM_003119.2:c.184-2A>C']['primary_assembly_loci'].keys()

        assert 'NM_003119.3:c.184-2A>C' in results.keys()
        assert results['NM_003119.3:c.184-2A>C']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003119.3:c.184-2A>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003119.3:c.184-2A>C']['alt_genomic_loci'] == []
        assert results['NM_003119.3:c.184-2A>C']['gene_symbol'] == 'SPG7'
        assert results['NM_003119.3:c.184-2A>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003110.1:p.?', 'slr': 'NP_003110.1:p.?'}
        assert results['NM_003119.3:c.184-2A>C']['submitted_variant'] == '16-89576896-A-C'
        assert results['NM_003119.3:c.184-2A>C']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_003119.3):c.184-2A>C'
        assert results['NM_003119.3:c.184-2A>C']['hgvs_lrg_variant'] == ''
        assert results['NM_003119.3:c.184-2A>C']['hgvs_transcript_variant'] == 'NM_003119.3:c.184-2A>C'
        assert results['NM_003119.3:c.184-2A>C']['hgvs_refseqgene_variant'] == ''
        assert results['NM_003119.3:c.184-2A>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576896A>C', 'vcf': {'chr': 'chr16', 'ref': 'A', 'pos': '89576896', 'alt': 'C'}}
        assert results['NM_003119.3:c.184-2A>C']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89510488A>C', 'vcf': {'chr': 'chr16', 'ref': 'A', 'pos': '89510488', 'alt': 'C'}}
        assert results['NM_003119.3:c.184-2A>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576896A>C', 'vcf': {'chr': '16', 'ref': 'A', 'pos': '89576896', 'alt': 'C'}}
        assert results['NM_003119.3:c.184-2A>C']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89510488A>C', 'vcf': {'chr': '16', 'ref': 'A', 'pos': '89510488', 'alt': 'C'}}

        assert 'NM_001363850.1:c.184-2A>C' in results.keys()
        assert results['NM_001363850.1:c.184-2A>C']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001363850.1:c.184-2A>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001363850.1:c.184-2A>C']['alt_genomic_loci'] == []
        assert results['NM_001363850.1:c.184-2A>C']['gene_symbol'] == 'SPG7'
        assert results['NM_001363850.1:c.184-2A>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001350779.1:p.?', 'slr': 'NP_001350779.1:p.?'}
        assert results['NM_001363850.1:c.184-2A>C']['submitted_variant'] == '16-89576896-A-C'
        assert results['NM_001363850.1:c.184-2A>C']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_001363850.1):c.184-2A>C'
        assert results['NM_001363850.1:c.184-2A>C']['hgvs_lrg_variant'] == ''
        assert results['NM_001363850.1:c.184-2A>C']['hgvs_transcript_variant'] == 'NM_001363850.1:c.184-2A>C'
        assert results['NM_001363850.1:c.184-2A>C']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001363850.1:c.184-2A>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576896A>C', 'vcf': {'chr': 'chr16', 'ref': 'A', 'pos': '89576896', 'alt': 'C'}}
        assert 'hg38' not in results['NM_001363850.1:c.184-2A>C']['primary_assembly_loci'].keys()
        assert results['NM_001363850.1:c.184-2A>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576896A>C', 'vcf': {'chr': '16', 'ref': 'A', 'pos': '89576896', 'alt': 'C'}}
        assert 'grch38' not in results['NM_001363850.1:c.184-2A>C']['primary_assembly_loci'].keys()

        assert results['flag'] == 'gene_variant'
        assert 'NM_199367.1:c.184-2A>C' in results.keys()
        assert results['NM_199367.1:c.184-2A>C']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_199367.1:c.184-2A>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_199367.1:c.184-2A>C']['alt_genomic_loci'] == []
        assert results['NM_199367.1:c.184-2A>C']['gene_symbol'] == 'SPG7'
        assert results['NM_199367.1:c.184-2A>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_955399.1:p.?', 'slr': 'NP_955399.1:p.?'}
        assert results['NM_199367.1:c.184-2A>C']['submitted_variant'] == '16-89576896-A-C'
        assert results['NM_199367.1:c.184-2A>C']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_199367.1):c.184-2A>C'
        assert results['NM_199367.1:c.184-2A>C']['hgvs_lrg_variant'] == ''
        assert results['NM_199367.1:c.184-2A>C']['hgvs_transcript_variant'] == 'NM_199367.1:c.184-2A>C'
        assert results['NM_199367.1:c.184-2A>C']['hgvs_refseqgene_variant'] == ''
        assert results['NM_199367.1:c.184-2A>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576896A>C', 'vcf': {'chr': 'chr16', 'ref': 'A', 'pos': '89576896', 'alt': 'C'}}
        assert 'hg38' not in results['NM_199367.1:c.184-2A>C']['primary_assembly_loci'].keys()
        assert results['NM_199367.1:c.184-2A>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576896A>C', 'vcf': {'chr': '16', 'ref': 'A', 'pos': '89576896', 'alt': 'C'}}
        assert 'grch38' not in results['NM_199367.1:c.184-2A>C']['primary_assembly_loci'].keys()


    def test_variant245(self):
        variant = '16-89576930-T-TA,TT'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_003119.3:c.216dup' in results.keys()
        assert results['NM_003119.3:c.216dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003119.3:c.216dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003119.3:c.216dup']['alt_genomic_loci'] == []
        assert results['NM_003119.3:c.216dup']['gene_symbol'] == 'SPG7'
        assert results['NM_003119.3:c.216dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003110.1:p.(Glu73Ter)', 'slr': 'NP_003110.1:p.(E73*)'}
        assert results['NM_003119.3:c.216dup']['submitted_variant'] == '16-89576930-T-TA,TT'
        assert results['NM_003119.3:c.216dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_003119.3:c.216dup']['hgvs_lrg_variant'] == ''
        assert results['NM_003119.3:c.216dup']['hgvs_transcript_variant'] == 'NM_003119.3:c.216dup'
        assert results['NM_003119.3:c.216dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_003119.3:c.216dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576930dup', 'vcf': {'chr': 'chr16', 'ref': 'T', 'pos': '89576928', 'alt': 'TT'}}
        assert results['NM_003119.3:c.216dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89510522dup', 'vcf': {'chr': 'chr16', 'ref': 'T', 'pos': '89510520', 'alt': 'TT'}}
        assert results['NM_003119.3:c.216dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576930dup', 'vcf': {'chr': '16', 'ref': 'T', 'pos': '89576928', 'alt': 'TT'}}
        assert results['NM_003119.3:c.216dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89510522dup', 'vcf': {'chr': '16', 'ref': 'T', 'pos': '89510520', 'alt': 'TT'}}

        assert 'NM_003119.2:c.216_217insA' in results.keys()
        assert results['NM_003119.2:c.216_217insA']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003119.2:c.216_217insA']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003119.2:c.216_217insA']['alt_genomic_loci'] == []
        assert results['NM_003119.2:c.216_217insA']['gene_symbol'] == 'SPG7'
        assert results['NM_003119.2:c.216_217insA']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003110.1:p.(Glu73ArgfsTer30)', 'slr': 'NP_003110.1:p.(E73Rfs*30)'}
        assert results['NM_003119.2:c.216_217insA']['submitted_variant'] == '16-89576930-T-TA,TT'
        assert results['NM_003119.2:c.216_217insA']['genome_context_intronic_sequence'] == ''
        assert results['NM_003119.2:c.216_217insA']['hgvs_lrg_variant'] == ''
        assert results['NM_003119.2:c.216_217insA']['hgvs_transcript_variant'] == 'NM_003119.2:c.216_217insA'
        assert results['NM_003119.2:c.216_217insA']['hgvs_refseqgene_variant'] == 'NG_008082.1:g.7126_7127insA'
        assert results['NM_003119.2:c.216_217insA']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576930_89576931insA', 'vcf': {'chr': 'chr16', 'ref': 'T', 'pos': '89576930', 'alt': 'TA'}}
        assert 'hg38' not in results['NM_003119.2:c.216_217insA']['primary_assembly_loci'].keys()
        assert results['NM_003119.2:c.216_217insA']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576930_89576931insA', 'vcf': {'chr': '16', 'ref': 'T', 'pos': '89576930', 'alt': 'TA'}}
        assert 'grch38' not in results['NM_003119.2:c.216_217insA']['primary_assembly_loci'].keys()

        assert 'NM_199367.2:c.216dup' in results.keys()
        assert results['NM_199367.2:c.216dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_199367.2:c.216dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_199367.2:c.216dup']['alt_genomic_loci'] == []
        assert results['NM_199367.2:c.216dup']['gene_symbol'] == 'SPG7'
        assert results['NM_199367.2:c.216dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_955399.1:p.(Glu73Ter)', 'slr': 'NP_955399.1:p.(E73*)'}
        assert results['NM_199367.2:c.216dup']['submitted_variant'] == '16-89576930-T-TA,TT'
        assert results['NM_199367.2:c.216dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_199367.2:c.216dup']['hgvs_lrg_variant'] == ''
        assert results['NM_199367.2:c.216dup']['hgvs_transcript_variant'] == 'NM_199367.2:c.216dup'
        assert results['NM_199367.2:c.216dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_199367.2:c.216dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576930dup', 'vcf': {'chr': 'chr16', 'ref': 'T', 'pos': '89576928', 'alt': 'TT'}}
        assert results['NM_199367.2:c.216dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89510522dup', 'vcf': {'chr': 'chr16', 'ref': 'T', 'pos': '89510520', 'alt': 'TT'}}
        assert results['NM_199367.2:c.216dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576930dup', 'vcf': {'chr': '16', 'ref': 'T', 'pos': '89576928', 'alt': 'TT'}}
        assert results['NM_199367.2:c.216dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89510522dup', 'vcf': {'chr': '16', 'ref': 'T', 'pos': '89510520', 'alt': 'TT'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001363850.1:c.216dup' in results.keys()
        assert results['NM_001363850.1:c.216dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001363850.1:c.216dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001363850.1:c.216dup']['alt_genomic_loci'] == []
        assert results['NM_001363850.1:c.216dup']['gene_symbol'] == 'SPG7'
        assert results['NM_001363850.1:c.216dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001350779.1:p.(Glu73Ter)', 'slr': 'NP_001350779.1:p.(E73*)'}
        assert results['NM_001363850.1:c.216dup']['submitted_variant'] == '16-89576930-T-TA,TT'
        assert results['NM_001363850.1:c.216dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_001363850.1:c.216dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001363850.1:c.216dup']['hgvs_transcript_variant'] == 'NM_001363850.1:c.216dup'
        assert results['NM_001363850.1:c.216dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001363850.1:c.216dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576930dup', 'vcf': {'chr': 'chr16', 'ref': 'T', 'pos': '89576928', 'alt': 'TT'}}
        assert 'hg38' not in results['NM_001363850.1:c.216dup']['primary_assembly_loci'].keys()
        assert results['NM_001363850.1:c.216dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576930dup', 'vcf': {'chr': '16', 'ref': 'T', 'pos': '89576928', 'alt': 'TT'}}
        assert 'grch38' not in results['NM_001363850.1:c.216dup']['primary_assembly_loci'].keys()

        assert 'NM_199367.2:c.216_217insA' in results.keys()
        assert results['NM_199367.2:c.216_217insA']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_199367.2:c.216_217insA']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_199367.2:c.216_217insA']['alt_genomic_loci'] == []
        assert results['NM_199367.2:c.216_217insA']['gene_symbol'] == 'SPG7'
        assert results['NM_199367.2:c.216_217insA']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_955399.1:p.(Glu73ArgfsTer30)', 'slr': 'NP_955399.1:p.(E73Rfs*30)'}
        assert results['NM_199367.2:c.216_217insA']['submitted_variant'] == '16-89576930-T-TA,TT'
        assert results['NM_199367.2:c.216_217insA']['genome_context_intronic_sequence'] == ''
        assert results['NM_199367.2:c.216_217insA']['hgvs_lrg_variant'] == ''
        assert results['NM_199367.2:c.216_217insA']['hgvs_transcript_variant'] == 'NM_199367.2:c.216_217insA'
        assert results['NM_199367.2:c.216_217insA']['hgvs_refseqgene_variant'] == ''
        assert results['NM_199367.2:c.216_217insA']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576930_89576931insA', 'vcf': {'chr': 'chr16', 'ref': 'T', 'pos': '89576930', 'alt': 'TA'}}
        assert results['NM_199367.2:c.216_217insA']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89510522_89510523insA', 'vcf': {'chr': 'chr16', 'ref': 'T', 'pos': '89510522', 'alt': 'TA'}}
        assert results['NM_199367.2:c.216_217insA']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576930_89576931insA', 'vcf': {'chr': '16', 'ref': 'T', 'pos': '89576930', 'alt': 'TA'}}
        assert results['NM_199367.2:c.216_217insA']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89510522_89510523insA', 'vcf': {'chr': '16', 'ref': 'T', 'pos': '89510522', 'alt': 'TA'}}

        assert 'NM_001363850.1:c.216_217insA' in results.keys()
        assert results['NM_001363850.1:c.216_217insA']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001363850.1:c.216_217insA']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001363850.1:c.216_217insA']['alt_genomic_loci'] == []
        assert results['NM_001363850.1:c.216_217insA']['gene_symbol'] == 'SPG7'
        assert results['NM_001363850.1:c.216_217insA']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001350779.1:p.(Glu73ArgfsTer30)', 'slr': 'NP_001350779.1:p.(E73Rfs*30)'}
        assert results['NM_001363850.1:c.216_217insA']['submitted_variant'] == '16-89576930-T-TA,TT'
        assert results['NM_001363850.1:c.216_217insA']['genome_context_intronic_sequence'] == ''
        assert results['NM_001363850.1:c.216_217insA']['hgvs_lrg_variant'] == ''
        assert results['NM_001363850.1:c.216_217insA']['hgvs_transcript_variant'] == 'NM_001363850.1:c.216_217insA'
        assert results['NM_001363850.1:c.216_217insA']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001363850.1:c.216_217insA']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576930_89576931insA', 'vcf': {'chr': 'chr16', 'ref': 'T', 'pos': '89576930', 'alt': 'TA'}}
        assert 'hg38' not in results['NM_001363850.1:c.216_217insA']['primary_assembly_loci'].keys()
        assert results['NM_001363850.1:c.216_217insA']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576930_89576931insA', 'vcf': {'chr': '16', 'ref': 'T', 'pos': '89576930', 'alt': 'TA'}}
        assert 'grch38' not in results['NM_001363850.1:c.216_217insA']['primary_assembly_loci'].keys()

        assert 'NM_199367.1:c.216_217insA' in results.keys()
        assert results['NM_199367.1:c.216_217insA']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_199367.1:c.216_217insA']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_199367.1:c.216_217insA']['alt_genomic_loci'] == []
        assert results['NM_199367.1:c.216_217insA']['gene_symbol'] == 'SPG7'
        assert results['NM_199367.1:c.216_217insA']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_955399.1:p.(Glu73ArgfsTer30)', 'slr': 'NP_955399.1:p.(E73Rfs*30)'}
        assert results['NM_199367.1:c.216_217insA']['submitted_variant'] == '16-89576930-T-TA,TT'
        assert results['NM_199367.1:c.216_217insA']['genome_context_intronic_sequence'] == ''
        assert results['NM_199367.1:c.216_217insA']['hgvs_lrg_variant'] == ''
        assert results['NM_199367.1:c.216_217insA']['hgvs_transcript_variant'] == 'NM_199367.1:c.216_217insA'
        assert results['NM_199367.1:c.216_217insA']['hgvs_refseqgene_variant'] == ''
        assert results['NM_199367.1:c.216_217insA']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576930_89576931insA', 'vcf': {'chr': 'chr16', 'ref': 'T', 'pos': '89576930', 'alt': 'TA'}}
        assert 'hg38' not in results['NM_199367.1:c.216_217insA']['primary_assembly_loci'].keys()
        assert results['NM_199367.1:c.216_217insA']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576930_89576931insA', 'vcf': {'chr': '16', 'ref': 'T', 'pos': '89576930', 'alt': 'TA'}}
        assert 'grch38' not in results['NM_199367.1:c.216_217insA']['primary_assembly_loci'].keys()

        assert 'NM_199367.1:c.216dup' in results.keys()
        assert results['NM_199367.1:c.216dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_199367.1:c.216dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_199367.1:c.216dup']['alt_genomic_loci'] == []
        assert results['NM_199367.1:c.216dup']['gene_symbol'] == 'SPG7'
        assert results['NM_199367.1:c.216dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_955399.1:p.(Glu73Ter)', 'slr': 'NP_955399.1:p.(E73*)'}
        assert results['NM_199367.1:c.216dup']['submitted_variant'] == '16-89576930-T-TA,TT'
        assert results['NM_199367.1:c.216dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_199367.1:c.216dup']['hgvs_lrg_variant'] == ''
        assert results['NM_199367.1:c.216dup']['hgvs_transcript_variant'] == 'NM_199367.1:c.216dup'
        assert results['NM_199367.1:c.216dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_199367.1:c.216dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576930dup', 'vcf': {'chr': 'chr16', 'ref': 'T', 'pos': '89576928', 'alt': 'TT'}}
        assert 'hg38' not in results['NM_199367.1:c.216dup']['primary_assembly_loci'].keys()
        assert results['NM_199367.1:c.216dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576930dup', 'vcf': {'chr': '16', 'ref': 'T', 'pos': '89576928', 'alt': 'TT'}}
        assert 'grch38' not in results['NM_199367.1:c.216dup']['primary_assembly_loci'].keys()

        assert 'NM_003119.3:c.216_217insA' in results.keys()
        assert results['NM_003119.3:c.216_217insA']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003119.3:c.216_217insA']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003119.3:c.216_217insA']['alt_genomic_loci'] == []
        assert results['NM_003119.3:c.216_217insA']['gene_symbol'] == 'SPG7'
        assert results['NM_003119.3:c.216_217insA']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003110.1:p.(Glu73ArgfsTer30)', 'slr': 'NP_003110.1:p.(E73Rfs*30)'}
        assert results['NM_003119.3:c.216_217insA']['submitted_variant'] == '16-89576930-T-TA,TT'
        assert results['NM_003119.3:c.216_217insA']['genome_context_intronic_sequence'] == ''
        assert results['NM_003119.3:c.216_217insA']['hgvs_lrg_variant'] == ''
        assert results['NM_003119.3:c.216_217insA']['hgvs_transcript_variant'] == 'NM_003119.3:c.216_217insA'
        assert results['NM_003119.3:c.216_217insA']['hgvs_refseqgene_variant'] == ''
        assert results['NM_003119.3:c.216_217insA']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576930_89576931insA', 'vcf': {'chr': 'chr16', 'ref': 'T', 'pos': '89576930', 'alt': 'TA'}}
        assert results['NM_003119.3:c.216_217insA']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89510522_89510523insA', 'vcf': {'chr': 'chr16', 'ref': 'T', 'pos': '89510522', 'alt': 'TA'}}
        assert results['NM_003119.3:c.216_217insA']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576930_89576931insA', 'vcf': {'chr': '16', 'ref': 'T', 'pos': '89576930', 'alt': 'TA'}}
        assert results['NM_003119.3:c.216_217insA']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89510522_89510523insA', 'vcf': {'chr': '16', 'ref': 'T', 'pos': '89510522', 'alt': 'TA'}}

        assert 'NM_003119.2:c.216dup' in results.keys()
        assert results['NM_003119.2:c.216dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003119.2:c.216dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003119.2:c.216dup']['alt_genomic_loci'] == []
        assert results['NM_003119.2:c.216dup']['gene_symbol'] == 'SPG7'
        assert results['NM_003119.2:c.216dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003110.1:p.(Glu73Ter)', 'slr': 'NP_003110.1:p.(E73*)'}
        assert results['NM_003119.2:c.216dup']['submitted_variant'] == '16-89576930-T-TA,TT'
        assert results['NM_003119.2:c.216dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_003119.2:c.216dup']['hgvs_lrg_variant'] == ''
        assert results['NM_003119.2:c.216dup']['hgvs_transcript_variant'] == 'NM_003119.2:c.216dup'
        assert results['NM_003119.2:c.216dup']['hgvs_refseqgene_variant'] == 'NG_008082.1:g.7126dup'
        assert results['NM_003119.2:c.216dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576930dup', 'vcf': {'chr': 'chr16', 'ref': 'T', 'pos': '89576928', 'alt': 'TT'}}
        assert 'hg38' not in results['NM_003119.2:c.216dup']['primary_assembly_loci'].keys()
        assert results['NM_003119.2:c.216dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576930dup', 'vcf': {'chr': '16', 'ref': 'T', 'pos': '89576928', 'alt': 'TT'}}
        assert 'grch38' not in results['NM_003119.2:c.216dup']['primary_assembly_loci'].keys()


    def test_variant246(self):
        variant = '16-89576931-G-GTG'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_199367.1:c.216_217dup' in results.keys()
        assert results['NM_199367.1:c.216_217dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_199367.1:c.216_217dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_199367.1:c.216_217dup']['alt_genomic_loci'] == []
        assert results['NM_199367.1:c.216_217dup']['gene_symbol'] == 'SPG7'
        assert results['NM_199367.1:c.216_217dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_955399.1:p.(Glu73ValfsTer9)', 'slr': 'NP_955399.1:p.(E73Vfs*9)'}
        assert results['NM_199367.1:c.216_217dup']['submitted_variant'] == '16-89576931-G-GTG'
        assert results['NM_199367.1:c.216_217dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_199367.1:c.216_217dup']['hgvs_lrg_variant'] == ''
        assert results['NM_199367.1:c.216_217dup']['hgvs_transcript_variant'] == 'NM_199367.1:c.216_217dup'
        assert results['NM_199367.1:c.216_217dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_199367.1:c.216_217dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576930_89576931dup', 'vcf': {'chr': 'chr16', 'ref': 'TG', 'pos': '89576930', 'alt': 'TGTG'}}
        assert 'hg38' not in results['NM_199367.1:c.216_217dup']['primary_assembly_loci'].keys()
        assert results['NM_199367.1:c.216_217dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576930_89576931dup', 'vcf': {'chr': '16', 'ref': 'TG', 'pos': '89576930', 'alt': 'TGTG'}}
        assert 'grch38' not in results['NM_199367.1:c.216_217dup']['primary_assembly_loci'].keys()

        assert 'NM_003119.3:c.216_217dup' in results.keys()
        assert results['NM_003119.3:c.216_217dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003119.3:c.216_217dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003119.3:c.216_217dup']['alt_genomic_loci'] == []
        assert results['NM_003119.3:c.216_217dup']['gene_symbol'] == 'SPG7'
        assert results['NM_003119.3:c.216_217dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003110.1:p.(Glu73ValfsTer9)', 'slr': 'NP_003110.1:p.(E73Vfs*9)'}
        assert results['NM_003119.3:c.216_217dup']['submitted_variant'] == '16-89576931-G-GTG'
        assert results['NM_003119.3:c.216_217dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_003119.3:c.216_217dup']['hgvs_lrg_variant'] == ''
        assert results['NM_003119.3:c.216_217dup']['hgvs_transcript_variant'] == 'NM_003119.3:c.216_217dup'
        assert results['NM_003119.3:c.216_217dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_003119.3:c.216_217dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576930_89576931dup', 'vcf': {'chr': 'chr16', 'ref': 'TG', 'pos': '89576930', 'alt': 'TGTG'}}
        assert results['NM_003119.3:c.216_217dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89510522_89510523dup', 'vcf': {'chr': 'chr16', 'ref': 'TG', 'pos': '89510522', 'alt': 'TGTG'}}
        assert results['NM_003119.3:c.216_217dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576930_89576931dup', 'vcf': {'chr': '16', 'ref': 'TG', 'pos': '89576930', 'alt': 'TGTG'}}
        assert results['NM_003119.3:c.216_217dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89510522_89510523dup', 'vcf': {'chr': '16', 'ref': 'TG', 'pos': '89510522', 'alt': 'TGTG'}}

        assert 'NM_199367.2:c.216_217dup' in results.keys()
        assert results['NM_199367.2:c.216_217dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_199367.2:c.216_217dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_199367.2:c.216_217dup']['alt_genomic_loci'] == []
        assert results['NM_199367.2:c.216_217dup']['gene_symbol'] == 'SPG7'
        assert results['NM_199367.2:c.216_217dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_955399.1:p.(Glu73ValfsTer9)', 'slr': 'NP_955399.1:p.(E73Vfs*9)'}
        assert results['NM_199367.2:c.216_217dup']['submitted_variant'] == '16-89576931-G-GTG'
        assert results['NM_199367.2:c.216_217dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_199367.2:c.216_217dup']['hgvs_lrg_variant'] == ''
        assert results['NM_199367.2:c.216_217dup']['hgvs_transcript_variant'] == 'NM_199367.2:c.216_217dup'
        assert results['NM_199367.2:c.216_217dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_199367.2:c.216_217dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576930_89576931dup', 'vcf': {'chr': 'chr16', 'ref': 'TG', 'pos': '89576930', 'alt': 'TGTG'}}
        assert results['NM_199367.2:c.216_217dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89510522_89510523dup', 'vcf': {'chr': 'chr16', 'ref': 'TG', 'pos': '89510522', 'alt': 'TGTG'}}
        assert results['NM_199367.2:c.216_217dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576930_89576931dup', 'vcf': {'chr': '16', 'ref': 'TG', 'pos': '89576930', 'alt': 'TGTG'}}
        assert results['NM_199367.2:c.216_217dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89510522_89510523dup', 'vcf': {'chr': '16', 'ref': 'TG', 'pos': '89510522', 'alt': 'TGTG'}}

        assert 'NM_001363850.1:c.216_217dup' in results.keys()
        assert results['NM_001363850.1:c.216_217dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001363850.1:c.216_217dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001363850.1:c.216_217dup']['alt_genomic_loci'] == []
        assert results['NM_001363850.1:c.216_217dup']['gene_symbol'] == 'SPG7'
        assert results['NM_001363850.1:c.216_217dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001350779.1:p.(Glu73ValfsTer9)', 'slr': 'NP_001350779.1:p.(E73Vfs*9)'}
        assert results['NM_001363850.1:c.216_217dup']['submitted_variant'] == '16-89576931-G-GTG'
        assert results['NM_001363850.1:c.216_217dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_001363850.1:c.216_217dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001363850.1:c.216_217dup']['hgvs_transcript_variant'] == 'NM_001363850.1:c.216_217dup'
        assert results['NM_001363850.1:c.216_217dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001363850.1:c.216_217dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576930_89576931dup', 'vcf': {'chr': 'chr16', 'ref': 'TG', 'pos': '89576930', 'alt': 'TGTG'}}
        assert 'hg38' not in results['NM_001363850.1:c.216_217dup']['primary_assembly_loci'].keys()
        assert results['NM_001363850.1:c.216_217dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576930_89576931dup', 'vcf': {'chr': '16', 'ref': 'TG', 'pos': '89576930', 'alt': 'TGTG'}}
        assert 'grch38' not in results['NM_001363850.1:c.216_217dup']['primary_assembly_loci'].keys()

        assert results['flag'] == 'gene_variant'
        assert 'NM_003119.2:c.216_217dup' in results.keys()
        assert results['NM_003119.2:c.216_217dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003119.2:c.216_217dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003119.2:c.216_217dup']['alt_genomic_loci'] == []
        assert results['NM_003119.2:c.216_217dup']['gene_symbol'] == 'SPG7'
        assert results['NM_003119.2:c.216_217dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003110.1:p.(Glu73ValfsTer9)', 'slr': 'NP_003110.1:p.(E73Vfs*9)'}
        assert results['NM_003119.2:c.216_217dup']['submitted_variant'] == '16-89576931-G-GTG'
        assert results['NM_003119.2:c.216_217dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_003119.2:c.216_217dup']['hgvs_lrg_variant'] == ''
        assert results['NM_003119.2:c.216_217dup']['hgvs_transcript_variant'] == 'NM_003119.2:c.216_217dup'
        assert results['NM_003119.2:c.216_217dup']['hgvs_refseqgene_variant'] == 'NG_008082.1:g.7126_7127dup'
        assert results['NM_003119.2:c.216_217dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576930_89576931dup', 'vcf': {'chr': 'chr16', 'ref': 'TG', 'pos': '89576930', 'alt': 'TGTG'}}
        assert 'hg38' not in results['NM_003119.2:c.216_217dup']['primary_assembly_loci'].keys()
        assert results['NM_003119.2:c.216_217dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89576930_89576931dup', 'vcf': {'chr': '16', 'ref': 'TG', 'pos': '89576930', 'alt': 'TGTG'}}
        assert 'grch38' not in results['NM_003119.2:c.216_217dup']['primary_assembly_loci'].keys()


    def test_variant247(self):
        variant = '16-89598368-CGGCCCCCCCGGCTGTGGGAAGACGCT-C'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_199367.1:c.1046_1071del' in results.keys()
        assert results['NM_199367.1:c.1046_1071del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_199367.1:c.1046_1071del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_199367.1:c.1046_1071del']['alt_genomic_loci'] == []
        assert results['NM_199367.1:c.1046_1071del']['gene_symbol'] == 'SPG7'
        assert results['NM_199367.1:c.1046_1071del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_955399.1:p.(Gly349AlafsTer38)', 'slr': 'NP_955399.1:p.(G349Afs*38)'}
        assert results['NM_199367.1:c.1046_1071del']['submitted_variant'] == '16-89598368-CGGCCCCCCCGGCTGTGGGAAGACGCT-C'
        assert results['NM_199367.1:c.1046_1071del']['genome_context_intronic_sequence'] == ''
        assert results['NM_199367.1:c.1046_1071del']['hgvs_lrg_variant'] == ''
        assert results['NM_199367.1:c.1046_1071del']['hgvs_transcript_variant'] == 'NM_199367.1:c.1046_1071del'
        assert results['NM_199367.1:c.1046_1071del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_199367.1:c.1046_1071del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89598370_89598395del', 'vcf': {'chr': 'chr16', 'ref': 'CGGCCCCCCCGGCTGTGGGAAGACGCT', 'pos': '89598368', 'alt': 'C'}}
        assert 'hg38' not in results['NM_199367.1:c.1046_1071del']['primary_assembly_loci'].keys()
        assert results['NM_199367.1:c.1046_1071del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89598370_89598395del', 'vcf': {'chr': '16', 'ref': 'CGGCCCCCCCGGCTGTGGGAAGACGCT', 'pos': '89598368', 'alt': 'C'}}
        assert 'grch38' not in results['NM_199367.1:c.1046_1071del']['primary_assembly_loci'].keys()

        assert 'NM_003119.2:c.1046_1071del' in results.keys()
        assert results['NM_003119.2:c.1046_1071del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003119.2:c.1046_1071del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003119.2:c.1046_1071del']['alt_genomic_loci'] == []
        assert results['NM_003119.2:c.1046_1071del']['gene_symbol'] == 'SPG7'
        assert results['NM_003119.2:c.1046_1071del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003110.1:p.(Gly349AlafsTer38)', 'slr': 'NP_003110.1:p.(G349Afs*38)'}
        assert results['NM_003119.2:c.1046_1071del']['submitted_variant'] == '16-89598368-CGGCCCCCCCGGCTGTGGGAAGACGCT-C'
        assert results['NM_003119.2:c.1046_1071del']['genome_context_intronic_sequence'] == ''
        assert results['NM_003119.2:c.1046_1071del']['hgvs_lrg_variant'] == ''
        assert results['NM_003119.2:c.1046_1071del']['hgvs_transcript_variant'] == 'NM_003119.2:c.1046_1071del'
        assert results['NM_003119.2:c.1046_1071del']['hgvs_refseqgene_variant'] == 'NG_008082.1:g.28566_28591del'
        assert results['NM_003119.2:c.1046_1071del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89598370_89598395del', 'vcf': {'chr': 'chr16', 'ref': 'CGGCCCCCCCGGCTGTGGGAAGACGCT', 'pos': '89598368', 'alt': 'C'}}
        assert 'hg38' not in results['NM_003119.2:c.1046_1071del']['primary_assembly_loci'].keys()
        assert results['NM_003119.2:c.1046_1071del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89598370_89598395del', 'vcf': {'chr': '16', 'ref': 'CGGCCCCCCCGGCTGTGGGAAGACGCT', 'pos': '89598368', 'alt': 'C'}}
        assert 'grch38' not in results['NM_003119.2:c.1046_1071del']['primary_assembly_loci'].keys()

        assert 'NM_199367.2:c.1046_1071del' in results.keys()
        assert results['NM_199367.2:c.1046_1071del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_199367.2:c.1046_1071del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_199367.2:c.1046_1071del']['alt_genomic_loci'] == []
        assert results['NM_199367.2:c.1046_1071del']['gene_symbol'] == 'SPG7'
        assert results['NM_199367.2:c.1046_1071del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_955399.1:p.(Gly349AlafsTer38)', 'slr': 'NP_955399.1:p.(G349Afs*38)'}
        assert results['NM_199367.2:c.1046_1071del']['submitted_variant'] == '16-89598368-CGGCCCCCCCGGCTGTGGGAAGACGCT-C'
        assert results['NM_199367.2:c.1046_1071del']['genome_context_intronic_sequence'] == ''
        assert results['NM_199367.2:c.1046_1071del']['hgvs_lrg_variant'] == ''
        assert results['NM_199367.2:c.1046_1071del']['hgvs_transcript_variant'] == 'NM_199367.2:c.1046_1071del'
        assert results['NM_199367.2:c.1046_1071del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_199367.2:c.1046_1071del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89598370_89598395del', 'vcf': {'chr': 'chr16', 'ref': 'CGGCCCCCCCGGCTGTGGGAAGACGCT', 'pos': '89598368', 'alt': 'C'}}
        assert results['NM_199367.2:c.1046_1071del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89531962_89531987del', 'vcf': {'chr': 'chr16', 'ref': 'CGGCCCCCCCGGCTGTGGGAAGACGCT', 'pos': '89531960', 'alt': 'C'}}
        assert results['NM_199367.2:c.1046_1071del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89598370_89598395del', 'vcf': {'chr': '16', 'ref': 'CGGCCCCCCCGGCTGTGGGAAGACGCT', 'pos': '89598368', 'alt': 'C'}}
        assert results['NM_199367.2:c.1046_1071del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89531962_89531987del', 'vcf': {'chr': '16', 'ref': 'CGGCCCCCCCGGCTGTGGGAAGACGCT', 'pos': '89531960', 'alt': 'C'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001363850.1:c.1046_1071del' in results.keys()
        assert results['NM_001363850.1:c.1046_1071del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001363850.1:c.1046_1071del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001363850.1:c.1046_1071del']['alt_genomic_loci'] == []
        assert results['NM_001363850.1:c.1046_1071del']['gene_symbol'] == 'SPG7'
        assert results['NM_001363850.1:c.1046_1071del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001350779.1:p.(Gly349AlafsTer38)', 'slr': 'NP_001350779.1:p.(G349Afs*38)'}
        assert results['NM_001363850.1:c.1046_1071del']['submitted_variant'] == '16-89598368-CGGCCCCCCCGGCTGTGGGAAGACGCT-C'
        assert results['NM_001363850.1:c.1046_1071del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001363850.1:c.1046_1071del']['hgvs_lrg_variant'] == ''
        assert results['NM_001363850.1:c.1046_1071del']['hgvs_transcript_variant'] == 'NM_001363850.1:c.1046_1071del'
        assert results['NM_001363850.1:c.1046_1071del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001363850.1:c.1046_1071del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89598370_89598395del', 'vcf': {'chr': 'chr16', 'ref': 'CGGCCCCCCCGGCTGTGGGAAGACGCT', 'pos': '89598368', 'alt': 'C'}}
        assert 'hg38' not in results['NM_001363850.1:c.1046_1071del']['primary_assembly_loci'].keys()
        assert results['NM_001363850.1:c.1046_1071del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89598370_89598395del', 'vcf': {'chr': '16', 'ref': 'CGGCCCCCCCGGCTGTGGGAAGACGCT', 'pos': '89598368', 'alt': 'C'}}
        assert 'grch38' not in results['NM_001363850.1:c.1046_1071del']['primary_assembly_loci'].keys()

        assert 'NM_003119.3:c.1046_1071del' in results.keys()
        assert results['NM_003119.3:c.1046_1071del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003119.3:c.1046_1071del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003119.3:c.1046_1071del']['alt_genomic_loci'] == []
        assert results['NM_003119.3:c.1046_1071del']['gene_symbol'] == 'SPG7'
        assert results['NM_003119.3:c.1046_1071del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003110.1:p.(Gly349AlafsTer38)', 'slr': 'NP_003110.1:p.(G349Afs*38)'}
        assert results['NM_003119.3:c.1046_1071del']['submitted_variant'] == '16-89598368-CGGCCCCCCCGGCTGTGGGAAGACGCT-C'
        assert results['NM_003119.3:c.1046_1071del']['genome_context_intronic_sequence'] == ''
        assert results['NM_003119.3:c.1046_1071del']['hgvs_lrg_variant'] == ''
        assert results['NM_003119.3:c.1046_1071del']['hgvs_transcript_variant'] == 'NM_003119.3:c.1046_1071del'
        assert results['NM_003119.3:c.1046_1071del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_003119.3:c.1046_1071del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89598370_89598395del', 'vcf': {'chr': 'chr16', 'ref': 'CGGCCCCCCCGGCTGTGGGAAGACGCT', 'pos': '89598368', 'alt': 'C'}}
        assert results['NM_003119.3:c.1046_1071del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89531962_89531987del', 'vcf': {'chr': 'chr16', 'ref': 'CGGCCCCCCCGGCTGTGGGAAGACGCT', 'pos': '89531960', 'alt': 'C'}}
        assert results['NM_003119.3:c.1046_1071del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89598370_89598395del', 'vcf': {'chr': '16', 'ref': 'CGGCCCCCCCGGCTGTGGGAAGACGCT', 'pos': '89598368', 'alt': 'C'}}
        assert results['NM_003119.3:c.1046_1071del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89531962_89531987del', 'vcf': {'chr': '16', 'ref': 'CGGCCCCCCCGGCTGTGGGAAGACGCT', 'pos': '89531960', 'alt': 'C'}}


    def test_variant248(self):
        variant = '16-89613064-AGGAGAGGCG-AT'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001363850.1:c.1450-1_1457delinsT' in results.keys()
        assert results['NM_001363850.1:c.1450-1_1457delinsT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001363850.1:c.1450-1_1457delinsT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001363850.1:c.1450-1_1457delinsT']['alt_genomic_loci'] == []
        assert results['NM_001363850.1:c.1450-1_1457delinsT']['gene_symbol'] == 'SPG7'
        assert results['NM_001363850.1:c.1450-1_1457delinsT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001350779.1:p.?', 'slr': 'NP_001350779.1:p.?'}
        assert results['NM_001363850.1:c.1450-1_1457delinsT']['submitted_variant'] == '16-89613064-AGGAGAGGCG-AT'
        assert results['NM_001363850.1:c.1450-1_1457delinsT']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_001363850.1):c.1450-1_1457delinsT'
        assert results['NM_001363850.1:c.1450-1_1457delinsT']['hgvs_lrg_variant'] == ''
        assert results['NM_001363850.1:c.1450-1_1457delinsT']['hgvs_transcript_variant'] == 'NM_001363850.1:c.1450-1_1457delinsT'
        assert results['NM_001363850.1:c.1450-1_1457delinsT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001363850.1:c.1450-1_1457delinsT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89613065_89613073delinsT', 'vcf': {'chr': 'chr16', 'ref': 'GGAGAGGCG', 'pos': '89613065', 'alt': 'T'}}
        assert 'hg38' not in results['NM_001363850.1:c.1450-1_1457delinsT']['primary_assembly_loci'].keys()
        assert results['NM_001363850.1:c.1450-1_1457delinsT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89613065_89613073delinsT', 'vcf': {'chr': '16', 'ref': 'GGAGAGGCG', 'pos': '89613065', 'alt': 'T'}}
        assert 'grch38' not in results['NM_001363850.1:c.1450-1_1457delinsT']['primary_assembly_loci'].keys()

        assert results['flag'] == 'gene_variant'
        assert 'NM_003119.2:c.1450-1_1457delinsT' in results.keys()
        assert results['NM_003119.2:c.1450-1_1457delinsT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003119.2:c.1450-1_1457delinsT']['refseqgene_context_intronic_sequence'] == 'NG_008082.1(NM_003119.2):c.1450-1_1457delinsT'
        assert results['NM_003119.2:c.1450-1_1457delinsT']['alt_genomic_loci'] == []
        assert results['NM_003119.2:c.1450-1_1457delinsT']['gene_symbol'] == 'SPG7'
        assert results['NM_003119.2:c.1450-1_1457delinsT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003110.1:p.?', 'slr': 'NP_003110.1:p.?'}
        assert results['NM_003119.2:c.1450-1_1457delinsT']['submitted_variant'] == '16-89613064-AGGAGAGGCG-AT'
        assert results['NM_003119.2:c.1450-1_1457delinsT']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_003119.2):c.1450-1_1457delinsT'
        assert results['NM_003119.2:c.1450-1_1457delinsT']['hgvs_lrg_variant'] == ''
        assert results['NM_003119.2:c.1450-1_1457delinsT']['hgvs_transcript_variant'] == 'NM_003119.2:c.1450-1_1457delinsT'
        assert results['NM_003119.2:c.1450-1_1457delinsT']['hgvs_refseqgene_variant'] == 'NG_008082.1:g.43261_43269delinsT'
        assert results['NM_003119.2:c.1450-1_1457delinsT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89613065_89613073delinsT', 'vcf': {'chr': 'chr16', 'ref': 'GGAGAGGCG', 'pos': '89613065', 'alt': 'T'}}
        assert 'hg38' not in results['NM_003119.2:c.1450-1_1457delinsT']['primary_assembly_loci'].keys()
        assert results['NM_003119.2:c.1450-1_1457delinsT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89613065_89613073delinsT', 'vcf': {'chr': '16', 'ref': 'GGAGAGGCG', 'pos': '89613065', 'alt': 'T'}}
        assert 'grch38' not in results['NM_003119.2:c.1450-1_1457delinsT']['primary_assembly_loci'].keys()

        assert 'NM_003119.3:c.1450-1_1457delinsT' in results.keys()
        assert results['NM_003119.3:c.1450-1_1457delinsT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003119.3:c.1450-1_1457delinsT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003119.3:c.1450-1_1457delinsT']['alt_genomic_loci'] == []
        assert results['NM_003119.3:c.1450-1_1457delinsT']['gene_symbol'] == 'SPG7'
        assert results['NM_003119.3:c.1450-1_1457delinsT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003110.1:p.?', 'slr': 'NP_003110.1:p.?'}
        assert results['NM_003119.3:c.1450-1_1457delinsT']['submitted_variant'] == '16-89613064-AGGAGAGGCG-AT'
        assert results['NM_003119.3:c.1450-1_1457delinsT']['genome_context_intronic_sequence'] == 'NC_000016.9(NM_003119.3):c.1450-1_1457delinsT'
        assert results['NM_003119.3:c.1450-1_1457delinsT']['hgvs_lrg_variant'] == ''
        assert results['NM_003119.3:c.1450-1_1457delinsT']['hgvs_transcript_variant'] == 'NM_003119.3:c.1450-1_1457delinsT'
        assert results['NM_003119.3:c.1450-1_1457delinsT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_003119.3:c.1450-1_1457delinsT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89613065_89613073delinsT', 'vcf': {'chr': 'chr16', 'ref': 'GGAGAGGCG', 'pos': '89613065', 'alt': 'T'}}
        assert results['NM_003119.3:c.1450-1_1457delinsT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89546657_89546665delinsT', 'vcf': {'chr': 'chr16', 'ref': 'GGAGAGGCG', 'pos': '89546657', 'alt': 'T'}}
        assert results['NM_003119.3:c.1450-1_1457delinsT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89613065_89613073delinsT', 'vcf': {'chr': '16', 'ref': 'GGAGAGGCG', 'pos': '89613065', 'alt': 'T'}}
        assert results['NM_003119.3:c.1450-1_1457delinsT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89546657_89546665delinsT', 'vcf': {'chr': '16', 'ref': 'GGAGAGGCG', 'pos': '89546657', 'alt': 'T'}}


    def test_variant249(self):
        variant = '16-89613069-AGGCGGGAGA-AT'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001363850.1:c.1454_1462delinsT' in results.keys()
        assert results['NM_001363850.1:c.1454_1462delinsT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001363850.1:c.1454_1462delinsT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001363850.1:c.1454_1462delinsT']['alt_genomic_loci'] == []
        assert results['NM_001363850.1:c.1454_1462delinsT']['gene_symbol'] == 'SPG7'
        assert results['NM_001363850.1:c.1454_1462delinsT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001350779.1:p.(Arg485IlefsTer3)', 'slr': 'NP_001350779.1:p.(R485Ifs*3)'}
        assert results['NM_001363850.1:c.1454_1462delinsT']['submitted_variant'] == '16-89613069-AGGCGGGAGA-AT'
        assert results['NM_001363850.1:c.1454_1462delinsT']['genome_context_intronic_sequence'] == ''
        assert results['NM_001363850.1:c.1454_1462delinsT']['hgvs_lrg_variant'] == ''
        assert results['NM_001363850.1:c.1454_1462delinsT']['hgvs_transcript_variant'] == 'NM_001363850.1:c.1454_1462delinsT'
        assert results['NM_001363850.1:c.1454_1462delinsT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001363850.1:c.1454_1462delinsT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89613070_89613078delinsT', 'vcf': {'chr': 'chr16', 'ref': 'GGCGGGAGA', 'pos': '89613070', 'alt': 'T'}}
        assert 'hg38' not in results['NM_001363850.1:c.1454_1462delinsT']['primary_assembly_loci'].keys()
        assert results['NM_001363850.1:c.1454_1462delinsT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89613070_89613078delinsT', 'vcf': {'chr': '16', 'ref': 'GGCGGGAGA', 'pos': '89613070', 'alt': 'T'}}
        assert 'grch38' not in results['NM_001363850.1:c.1454_1462delinsT']['primary_assembly_loci'].keys()

        assert results['flag'] == 'gene_variant'
        assert 'NM_003119.2:c.1454_1462delinsT' in results.keys()
        assert results['NM_003119.2:c.1454_1462delinsT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003119.2:c.1454_1462delinsT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003119.2:c.1454_1462delinsT']['alt_genomic_loci'] == []
        assert results['NM_003119.2:c.1454_1462delinsT']['gene_symbol'] == 'SPG7'
        assert results['NM_003119.2:c.1454_1462delinsT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003110.1:p.(Arg485IlefsTer3)', 'slr': 'NP_003110.1:p.(R485Ifs*3)'}
        assert results['NM_003119.2:c.1454_1462delinsT']['submitted_variant'] == '16-89613069-AGGCGGGAGA-AT'
        assert results['NM_003119.2:c.1454_1462delinsT']['genome_context_intronic_sequence'] == ''
        assert results['NM_003119.2:c.1454_1462delinsT']['hgvs_lrg_variant'] == ''
        assert results['NM_003119.2:c.1454_1462delinsT']['hgvs_transcript_variant'] == 'NM_003119.2:c.1454_1462delinsT'
        assert results['NM_003119.2:c.1454_1462delinsT']['hgvs_refseqgene_variant'] == 'NG_008082.1:g.43266_43274delinsT'
        assert results['NM_003119.2:c.1454_1462delinsT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89613070_89613078delinsT', 'vcf': {'chr': 'chr16', 'ref': 'GGCGGGAGA', 'pos': '89613070', 'alt': 'T'}}
        assert 'hg38' not in results['NM_003119.2:c.1454_1462delinsT']['primary_assembly_loci'].keys()
        assert results['NM_003119.2:c.1454_1462delinsT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89613070_89613078delinsT', 'vcf': {'chr': '16', 'ref': 'GGCGGGAGA', 'pos': '89613070', 'alt': 'T'}}
        assert 'grch38' not in results['NM_003119.2:c.1454_1462delinsT']['primary_assembly_loci'].keys()

        assert 'NM_003119.3:c.1454_1462delinsT' in results.keys()
        assert results['NM_003119.3:c.1454_1462delinsT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003119.3:c.1454_1462delinsT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003119.3:c.1454_1462delinsT']['alt_genomic_loci'] == []
        assert results['NM_003119.3:c.1454_1462delinsT']['gene_symbol'] == 'SPG7'
        assert results['NM_003119.3:c.1454_1462delinsT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003110.1:p.(Arg485IlefsTer3)', 'slr': 'NP_003110.1:p.(R485Ifs*3)'}
        assert results['NM_003119.3:c.1454_1462delinsT']['submitted_variant'] == '16-89613069-AGGCGGGAGA-AT'
        assert results['NM_003119.3:c.1454_1462delinsT']['genome_context_intronic_sequence'] == ''
        assert results['NM_003119.3:c.1454_1462delinsT']['hgvs_lrg_variant'] == ''
        assert results['NM_003119.3:c.1454_1462delinsT']['hgvs_transcript_variant'] == 'NM_003119.3:c.1454_1462delinsT'
        assert results['NM_003119.3:c.1454_1462delinsT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_003119.3:c.1454_1462delinsT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89613070_89613078delinsT', 'vcf': {'chr': 'chr16', 'ref': 'GGCGGGAGA', 'pos': '89613070', 'alt': 'T'}}
        assert results['NM_003119.3:c.1454_1462delinsT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89546662_89546670delinsT', 'vcf': {'chr': 'chr16', 'ref': 'GGCGGGAGA', 'pos': '89546662', 'alt': 'T'}}
        assert results['NM_003119.3:c.1454_1462delinsT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89613070_89613078delinsT', 'vcf': {'chr': '16', 'ref': 'GGCGGGAGA', 'pos': '89613070', 'alt': 'T'}}
        assert results['NM_003119.3:c.1454_1462delinsT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89546662_89546670delinsT', 'vcf': {'chr': '16', 'ref': 'GGCGGGAGA', 'pos': '89546662', 'alt': 'T'}}


    def test_variant250(self):
        variant = '16-89613145-C-T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001363850.1:c.1529C>T' in results.keys()
        assert results['NM_001363850.1:c.1529C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001363850.1:c.1529C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001363850.1:c.1529C>T']['alt_genomic_loci'] == []
        assert results['NM_001363850.1:c.1529C>T']['gene_symbol'] == 'SPG7'
        assert results['NM_001363850.1:c.1529C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001350779.1:p.(Ala510Val)', 'slr': 'NP_001350779.1:p.(A510V)'}
        assert results['NM_001363850.1:c.1529C>T']['submitted_variant'] == '16-89613145-C-T'
        assert results['NM_001363850.1:c.1529C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001363850.1:c.1529C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001363850.1:c.1529C>T']['hgvs_transcript_variant'] == 'NM_001363850.1:c.1529C>T'
        assert results['NM_001363850.1:c.1529C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001363850.1:c.1529C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89613145C>T', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '89613145', 'alt': 'T'}}
        assert 'hg38' not in results['NM_001363850.1:c.1529C>T']['primary_assembly_loci'].keys()
        assert results['NM_001363850.1:c.1529C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89613145C>T', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '89613145', 'alt': 'T'}}
        assert 'grch38' not in results['NM_001363850.1:c.1529C>T']['primary_assembly_loci'].keys()

        assert 'NM_003119.3:c.1529C>T' in results.keys()
        assert results['NM_003119.3:c.1529C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003119.3:c.1529C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003119.3:c.1529C>T']['alt_genomic_loci'] == []
        assert results['NM_003119.3:c.1529C>T']['gene_symbol'] == 'SPG7'
        assert results['NM_003119.3:c.1529C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003110.1:p.(Ala510Val)', 'slr': 'NP_003110.1:p.(A510V)'}
        assert results['NM_003119.3:c.1529C>T']['submitted_variant'] == '16-89613145-C-T'
        assert results['NM_003119.3:c.1529C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_003119.3:c.1529C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_003119.3:c.1529C>T']['hgvs_transcript_variant'] == 'NM_003119.3:c.1529C>T'
        assert results['NM_003119.3:c.1529C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_003119.3:c.1529C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89613145C>T', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '89613145', 'alt': 'T'}}
        assert results['NM_003119.3:c.1529C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89546737C>T', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '89546737', 'alt': 'T'}}
        assert results['NM_003119.3:c.1529C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89613145C>T', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '89613145', 'alt': 'T'}}
        assert results['NM_003119.3:c.1529C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000016.10:g.89546737C>T', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '89546737', 'alt': 'T'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_003119.2:c.1529C>T' in results.keys()
        assert results['NM_003119.2:c.1529C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003119.2:c.1529C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003119.2:c.1529C>T']['alt_genomic_loci'] == []
        assert results['NM_003119.2:c.1529C>T']['gene_symbol'] == 'SPG7'
        assert results['NM_003119.2:c.1529C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003110.1:p.(Ala510Val)', 'slr': 'NP_003110.1:p.(A510V)'}
        assert results['NM_003119.2:c.1529C>T']['submitted_variant'] == '16-89613145-C-T'
        assert results['NM_003119.2:c.1529C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_003119.2:c.1529C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_003119.2:c.1529C>T']['hgvs_transcript_variant'] == 'NM_003119.2:c.1529C>T'
        assert results['NM_003119.2:c.1529C>T']['hgvs_refseqgene_variant'] == 'NG_008082.1:g.43341C>T'
        assert results['NM_003119.2:c.1529C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000016.9:g.89613145C>T', 'vcf': {'chr': 'chr16', 'ref': 'C', 'pos': '89613145', 'alt': 'T'}}
        assert 'hg38' not in results['NM_003119.2:c.1529C>T']['primary_assembly_loci'].keys()
        assert results['NM_003119.2:c.1529C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000016.9:g.89613145C>T', 'vcf': {'chr': '16', 'ref': 'C', 'pos': '89613145', 'alt': 'T'}}
        assert 'grch38' not in results['NM_003119.2:c.1529C>T']['primary_assembly_loci'].keys()


    def test_variant251(self):
        variant = '17-7578194-GCAC-G'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001276695.1:c.535_537del' in results.keys()
        assert results['NM_001276695.1:c.535_537del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001276695.1:c.535_537del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001276695.1:c.535_537del']['alt_genomic_loci'] == []
        assert results['NM_001276695.1:c.535_537del']['gene_symbol'] == 'TP53'
        assert results['NM_001276695.1:c.535_537del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001263624.1:p.(Val179del)', 'slr': 'NP_001263624.1:p.(V179del)'}
        assert results['NM_001276695.1:c.535_537del']['submitted_variant'] == '17-7578194-GCAC-G'
        assert results['NM_001276695.1:c.535_537del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001276695.1:c.535_537del']['hgvs_lrg_variant'] == ''
        assert results['NM_001276695.1:c.535_537del']['hgvs_transcript_variant'] == 'NM_001276695.1:c.535_537del'
        assert results['NM_001276695.1:c.535_537del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001276695.1:c.535_537del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578195_7578197del', 'vcf': {'chr': 'chr17', 'ref': 'GCAC', 'pos': '7578194', 'alt': 'G'}}
        assert results['NM_001276695.1:c.535_537del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7674877_7674879del', 'vcf': {'chr': 'chr17', 'ref': 'GCAC', 'pos': '7674876', 'alt': 'G'}}
        assert results['NM_001276695.1:c.535_537del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578195_7578197del', 'vcf': {'chr': '17', 'ref': 'GCAC', 'pos': '7578194', 'alt': 'G'}}
        assert results['NM_001276695.1:c.535_537del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7674877_7674879del', 'vcf': {'chr': '17', 'ref': 'GCAC', 'pos': '7674876', 'alt': 'G'}}

        assert 'NM_001126118.1:c.535_537del' in results.keys()
        assert results['NM_001126118.1:c.535_537del']['hgvs_lrg_transcript_variant'] == 'LRG_321t8:c.535_537del'
        assert results['NM_001126118.1:c.535_537del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001126118.1:c.535_537del']['alt_genomic_loci'] == []
        assert results['NM_001126118.1:c.535_537del']['gene_symbol'] == 'TP53'
        assert results['NM_001126118.1:c.535_537del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001119590.1:p.(Val179del)', 'slr': 'NP_001119590.1:p.(V179del)'}
        assert results['NM_001126118.1:c.535_537del']['submitted_variant'] == '17-7578194-GCAC-G'
        assert results['NM_001126118.1:c.535_537del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001126118.1:c.535_537del']['hgvs_lrg_variant'] == 'LRG_321:g.17672_17674del'
        assert results['NM_001126118.1:c.535_537del']['hgvs_transcript_variant'] == 'NM_001126118.1:c.535_537del'
        assert results['NM_001126118.1:c.535_537del']['hgvs_refseqgene_variant'] == 'NG_017013.2:g.17672_17674del'
        assert results['NM_001126118.1:c.535_537del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578195_7578197del', 'vcf': {'chr': 'chr17', 'ref': 'GCAC', 'pos': '7578194', 'alt': 'G'}}
        assert results['NM_001126118.1:c.535_537del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7674877_7674879del', 'vcf': {'chr': 'chr17', 'ref': 'GCAC', 'pos': '7674876', 'alt': 'G'}}
        assert results['NM_001126118.1:c.535_537del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578195_7578197del', 'vcf': {'chr': '17', 'ref': 'GCAC', 'pos': '7578194', 'alt': 'G'}}
        assert results['NM_001126118.1:c.535_537del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7674877_7674879del', 'vcf': {'chr': '17', 'ref': 'GCAC', 'pos': '7674876', 'alt': 'G'}}

        assert 'NM_001126113.2:c.652_654del' in results.keys()
        assert results['NM_001126113.2:c.652_654del']['hgvs_lrg_transcript_variant'] == 'LRG_321t4:c.652_654del'
        assert results['NM_001126113.2:c.652_654del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001126113.2:c.652_654del']['alt_genomic_loci'] == []
        assert results['NM_001126113.2:c.652_654del']['gene_symbol'] == 'TP53'
        assert results['NM_001126113.2:c.652_654del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001119585.1:p.(Val218del)', 'slr': 'NP_001119585.1:p.(V218del)'}
        assert results['NM_001126113.2:c.652_654del']['submitted_variant'] == '17-7578194-GCAC-G'
        assert results['NM_001126113.2:c.652_654del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001126113.2:c.652_654del']['hgvs_lrg_variant'] == 'LRG_321:g.17672_17674del'
        assert results['NM_001126113.2:c.652_654del']['hgvs_transcript_variant'] == 'NM_001126113.2:c.652_654del'
        assert results['NM_001126113.2:c.652_654del']['hgvs_refseqgene_variant'] == 'NG_017013.2:g.17672_17674del'
        assert results['NM_001126113.2:c.652_654del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578195_7578197del', 'vcf': {'chr': 'chr17', 'ref': 'GCAC', 'pos': '7578194', 'alt': 'G'}}
        assert results['NM_001126113.2:c.652_654del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7674877_7674879del', 'vcf': {'chr': 'chr17', 'ref': 'GCAC', 'pos': '7674876', 'alt': 'G'}}
        assert results['NM_001126113.2:c.652_654del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578195_7578197del', 'vcf': {'chr': '17', 'ref': 'GCAC', 'pos': '7578194', 'alt': 'G'}}
        assert results['NM_001126113.2:c.652_654del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7674877_7674879del', 'vcf': {'chr': '17', 'ref': 'GCAC', 'pos': '7674876', 'alt': 'G'}}

        assert 'NM_001276761.1:c.535_537del' in results.keys()
        assert results['NM_001276761.1:c.535_537del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001276761.1:c.535_537del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001276761.1:c.535_537del']['alt_genomic_loci'] == []
        assert results['NM_001276761.1:c.535_537del']['gene_symbol'] == 'TP53'
        assert results['NM_001276761.1:c.535_537del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001263690.1:p.(Val179del)', 'slr': 'NP_001263690.1:p.(V179del)'}
        assert results['NM_001276761.1:c.535_537del']['submitted_variant'] == '17-7578194-GCAC-G'
        assert results['NM_001276761.1:c.535_537del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001276761.1:c.535_537del']['hgvs_lrg_variant'] == ''
        assert results['NM_001276761.1:c.535_537del']['hgvs_transcript_variant'] == 'NM_001276761.1:c.535_537del'
        assert results['NM_001276761.1:c.535_537del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001276761.1:c.535_537del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578195_7578197del', 'vcf': {'chr': 'chr17', 'ref': 'GCAC', 'pos': '7578194', 'alt': 'G'}}
        assert results['NM_001276761.1:c.535_537del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7674877_7674879del', 'vcf': {'chr': 'chr17', 'ref': 'GCAC', 'pos': '7674876', 'alt': 'G'}}
        assert results['NM_001276761.1:c.535_537del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578195_7578197del', 'vcf': {'chr': '17', 'ref': 'GCAC', 'pos': '7578194', 'alt': 'G'}}
        assert results['NM_001276761.1:c.535_537del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7674877_7674879del', 'vcf': {'chr': '17', 'ref': 'GCAC', 'pos': '7674876', 'alt': 'G'}}

        assert 'NM_001126117.1:c.256_258del' in results.keys()
        assert results['NM_001126117.1:c.256_258del']['hgvs_lrg_transcript_variant'] == 'LRG_321t7:c.256_258del'
        assert results['NM_001126117.1:c.256_258del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001126117.1:c.256_258del']['alt_genomic_loci'] == []
        assert results['NM_001126117.1:c.256_258del']['gene_symbol'] == 'TP53'
        assert results['NM_001126117.1:c.256_258del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001119589.1:p.(Val86del)', 'slr': 'NP_001119589.1:p.(V86del)'}
        assert results['NM_001126117.1:c.256_258del']['submitted_variant'] == '17-7578194-GCAC-G'
        assert results['NM_001126117.1:c.256_258del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001126117.1:c.256_258del']['hgvs_lrg_variant'] == 'LRG_321:g.17672_17674del'
        assert results['NM_001126117.1:c.256_258del']['hgvs_transcript_variant'] == 'NM_001126117.1:c.256_258del'
        assert results['NM_001126117.1:c.256_258del']['hgvs_refseqgene_variant'] == 'NG_017013.2:g.17672_17674del'
        assert results['NM_001126117.1:c.256_258del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578195_7578197del', 'vcf': {'chr': 'chr17', 'ref': 'GCAC', 'pos': '7578194', 'alt': 'G'}}
        assert results['NM_001126117.1:c.256_258del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7674877_7674879del', 'vcf': {'chr': 'chr17', 'ref': 'GCAC', 'pos': '7674876', 'alt': 'G'}}
        assert results['NM_001126117.1:c.256_258del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578195_7578197del', 'vcf': {'chr': '17', 'ref': 'GCAC', 'pos': '7578194', 'alt': 'G'}}
        assert results['NM_001126117.1:c.256_258del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7674877_7674879del', 'vcf': {'chr': '17', 'ref': 'GCAC', 'pos': '7674876', 'alt': 'G'}}

        assert 'NM_001126116.1:c.256_258del' in results.keys()
        assert results['NM_001126116.1:c.256_258del']['hgvs_lrg_transcript_variant'] == 'LRG_321t6:c.256_258del'
        assert results['NM_001126116.1:c.256_258del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001126116.1:c.256_258del']['alt_genomic_loci'] == []
        assert results['NM_001126116.1:c.256_258del']['gene_symbol'] == 'TP53'
        assert results['NM_001126116.1:c.256_258del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001119588.1:p.(Val86del)', 'slr': 'NP_001119588.1:p.(V86del)'}
        assert results['NM_001126116.1:c.256_258del']['submitted_variant'] == '17-7578194-GCAC-G'
        assert results['NM_001126116.1:c.256_258del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001126116.1:c.256_258del']['hgvs_lrg_variant'] == 'LRG_321:g.17672_17674del'
        assert results['NM_001126116.1:c.256_258del']['hgvs_transcript_variant'] == 'NM_001126116.1:c.256_258del'
        assert results['NM_001126116.1:c.256_258del']['hgvs_refseqgene_variant'] == 'NG_017013.2:g.17672_17674del'
        assert results['NM_001126116.1:c.256_258del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578195_7578197del', 'vcf': {'chr': 'chr17', 'ref': 'GCAC', 'pos': '7578194', 'alt': 'G'}}
        assert results['NM_001126116.1:c.256_258del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7674877_7674879del', 'vcf': {'chr': 'chr17', 'ref': 'GCAC', 'pos': '7674876', 'alt': 'G'}}
        assert results['NM_001126116.1:c.256_258del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578195_7578197del', 'vcf': {'chr': '17', 'ref': 'GCAC', 'pos': '7578194', 'alt': 'G'}}
        assert results['NM_001126116.1:c.256_258del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7674877_7674879del', 'vcf': {'chr': '17', 'ref': 'GCAC', 'pos': '7674876', 'alt': 'G'}}

        assert 'NM_001126112.2:c.652_654del' in results.keys()
        assert results['NM_001126112.2:c.652_654del']['hgvs_lrg_transcript_variant'] == 'LRG_321t2:c.652_654del'
        assert results['NM_001126112.2:c.652_654del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001126112.2:c.652_654del']['alt_genomic_loci'] == []
        assert results['NM_001126112.2:c.652_654del']['gene_symbol'] == 'TP53'
        assert results['NM_001126112.2:c.652_654del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001119584.1:p.(Val218del)', 'slr': 'NP_001119584.1:p.(V218del)'}
        assert results['NM_001126112.2:c.652_654del']['submitted_variant'] == '17-7578194-GCAC-G'
        assert results['NM_001126112.2:c.652_654del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001126112.2:c.652_654del']['hgvs_lrg_variant'] == 'LRG_321:g.17672_17674del'
        assert results['NM_001126112.2:c.652_654del']['hgvs_transcript_variant'] == 'NM_001126112.2:c.652_654del'
        assert results['NM_001126112.2:c.652_654del']['hgvs_refseqgene_variant'] == 'NG_017013.2:g.17672_17674del'
        assert results['NM_001126112.2:c.652_654del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578195_7578197del', 'vcf': {'chr': 'chr17', 'ref': 'GCAC', 'pos': '7578194', 'alt': 'G'}}
        assert results['NM_001126112.2:c.652_654del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7674877_7674879del', 'vcf': {'chr': 'chr17', 'ref': 'GCAC', 'pos': '7674876', 'alt': 'G'}}
        assert results['NM_001126112.2:c.652_654del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578195_7578197del', 'vcf': {'chr': '17', 'ref': 'GCAC', 'pos': '7578194', 'alt': 'G'}}
        assert results['NM_001126112.2:c.652_654del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7674877_7674879del', 'vcf': {'chr': '17', 'ref': 'GCAC', 'pos': '7674876', 'alt': 'G'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001276697.1:c.175_177del' in results.keys()
        assert results['NM_001276697.1:c.175_177del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001276697.1:c.175_177del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001276697.1:c.175_177del']['alt_genomic_loci'] == []
        assert results['NM_001276697.1:c.175_177del']['gene_symbol'] == 'TP53'
        assert results['NM_001276697.1:c.175_177del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001263626.1:p.(Val59del)', 'slr': 'NP_001263626.1:p.(V59del)'}
        assert results['NM_001276697.1:c.175_177del']['submitted_variant'] == '17-7578194-GCAC-G'
        assert results['NM_001276697.1:c.175_177del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001276697.1:c.175_177del']['hgvs_lrg_variant'] == ''
        assert results['NM_001276697.1:c.175_177del']['hgvs_transcript_variant'] == 'NM_001276697.1:c.175_177del'
        assert results['NM_001276697.1:c.175_177del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001276697.1:c.175_177del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578195_7578197del', 'vcf': {'chr': 'chr17', 'ref': 'GCAC', 'pos': '7578194', 'alt': 'G'}}
        assert results['NM_001276697.1:c.175_177del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7674877_7674879del', 'vcf': {'chr': 'chr17', 'ref': 'GCAC', 'pos': '7674876', 'alt': 'G'}}
        assert results['NM_001276697.1:c.175_177del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578195_7578197del', 'vcf': {'chr': '17', 'ref': 'GCAC', 'pos': '7578194', 'alt': 'G'}}
        assert results['NM_001276697.1:c.175_177del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7674877_7674879del', 'vcf': {'chr': '17', 'ref': 'GCAC', 'pos': '7674876', 'alt': 'G'}}

        assert 'NM_001276696.1:c.535_537del' in results.keys()
        assert results['NM_001276696.1:c.535_537del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001276696.1:c.535_537del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001276696.1:c.535_537del']['alt_genomic_loci'] == []
        assert results['NM_001276696.1:c.535_537del']['gene_symbol'] == 'TP53'
        assert results['NM_001276696.1:c.535_537del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001263625.1:p.(Val179del)', 'slr': 'NP_001263625.1:p.(V179del)'}
        assert results['NM_001276696.1:c.535_537del']['submitted_variant'] == '17-7578194-GCAC-G'
        assert results['NM_001276696.1:c.535_537del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001276696.1:c.535_537del']['hgvs_lrg_variant'] == ''
        assert results['NM_001276696.1:c.535_537del']['hgvs_transcript_variant'] == 'NM_001276696.1:c.535_537del'
        assert results['NM_001276696.1:c.535_537del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001276696.1:c.535_537del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578195_7578197del', 'vcf': {'chr': 'chr17', 'ref': 'GCAC', 'pos': '7578194', 'alt': 'G'}}
        assert results['NM_001276696.1:c.535_537del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7674877_7674879del', 'vcf': {'chr': 'chr17', 'ref': 'GCAC', 'pos': '7674876', 'alt': 'G'}}
        assert results['NM_001276696.1:c.535_537del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578195_7578197del', 'vcf': {'chr': '17', 'ref': 'GCAC', 'pos': '7578194', 'alt': 'G'}}
        assert results['NM_001276696.1:c.535_537del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7674877_7674879del', 'vcf': {'chr': '17', 'ref': 'GCAC', 'pos': '7674876', 'alt': 'G'}}

        assert 'NM_001276698.1:c.175_177del' in results.keys()
        assert results['NM_001276698.1:c.175_177del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001276698.1:c.175_177del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001276698.1:c.175_177del']['alt_genomic_loci'] == []
        assert results['NM_001276698.1:c.175_177del']['gene_symbol'] == 'TP53'
        assert results['NM_001276698.1:c.175_177del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001263627.1:p.(Val59del)', 'slr': 'NP_001263627.1:p.(V59del)'}
        assert results['NM_001276698.1:c.175_177del']['submitted_variant'] == '17-7578194-GCAC-G'
        assert results['NM_001276698.1:c.175_177del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001276698.1:c.175_177del']['hgvs_lrg_variant'] == ''
        assert results['NM_001276698.1:c.175_177del']['hgvs_transcript_variant'] == 'NM_001276698.1:c.175_177del'
        assert results['NM_001276698.1:c.175_177del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001276698.1:c.175_177del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578195_7578197del', 'vcf': {'chr': 'chr17', 'ref': 'GCAC', 'pos': '7578194', 'alt': 'G'}}
        assert results['NM_001276698.1:c.175_177del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7674877_7674879del', 'vcf': {'chr': 'chr17', 'ref': 'GCAC', 'pos': '7674876', 'alt': 'G'}}
        assert results['NM_001276698.1:c.175_177del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578195_7578197del', 'vcf': {'chr': '17', 'ref': 'GCAC', 'pos': '7578194', 'alt': 'G'}}
        assert results['NM_001276698.1:c.175_177del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7674877_7674879del', 'vcf': {'chr': '17', 'ref': 'GCAC', 'pos': '7674876', 'alt': 'G'}}

        assert 'NM_001126115.1:c.256_258del' in results.keys()
        assert results['NM_001126115.1:c.256_258del']['hgvs_lrg_transcript_variant'] == 'LRG_321t5:c.256_258del'
        assert results['NM_001126115.1:c.256_258del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001126115.1:c.256_258del']['alt_genomic_loci'] == []
        assert results['NM_001126115.1:c.256_258del']['gene_symbol'] == 'TP53'
        assert results['NM_001126115.1:c.256_258del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001119587.1:p.(Val86del)', 'slr': 'NP_001119587.1:p.(V86del)'}
        assert results['NM_001126115.1:c.256_258del']['submitted_variant'] == '17-7578194-GCAC-G'
        assert results['NM_001126115.1:c.256_258del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001126115.1:c.256_258del']['hgvs_lrg_variant'] == 'LRG_321:g.17672_17674del'
        assert results['NM_001126115.1:c.256_258del']['hgvs_transcript_variant'] == 'NM_001126115.1:c.256_258del'
        assert results['NM_001126115.1:c.256_258del']['hgvs_refseqgene_variant'] == 'NG_017013.2:g.17672_17674del'
        assert results['NM_001126115.1:c.256_258del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578195_7578197del', 'vcf': {'chr': 'chr17', 'ref': 'GCAC', 'pos': '7578194', 'alt': 'G'}}
        assert results['NM_001126115.1:c.256_258del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7674877_7674879del', 'vcf': {'chr': 'chr17', 'ref': 'GCAC', 'pos': '7674876', 'alt': 'G'}}
        assert results['NM_001126115.1:c.256_258del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578195_7578197del', 'vcf': {'chr': '17', 'ref': 'GCAC', 'pos': '7578194', 'alt': 'G'}}
        assert results['NM_001126115.1:c.256_258del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7674877_7674879del', 'vcf': {'chr': '17', 'ref': 'GCAC', 'pos': '7674876', 'alt': 'G'}}

        assert 'NM_001126114.2:c.652_654del' in results.keys()
        assert results['NM_001126114.2:c.652_654del']['hgvs_lrg_transcript_variant'] == 'LRG_321t3:c.652_654del'
        assert results['NM_001126114.2:c.652_654del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001126114.2:c.652_654del']['alt_genomic_loci'] == []
        assert results['NM_001126114.2:c.652_654del']['gene_symbol'] == 'TP53'
        assert results['NM_001126114.2:c.652_654del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001119586.1:p.(Val218del)', 'slr': 'NP_001119586.1:p.(V218del)'}
        assert results['NM_001126114.2:c.652_654del']['submitted_variant'] == '17-7578194-GCAC-G'
        assert results['NM_001126114.2:c.652_654del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001126114.2:c.652_654del']['hgvs_lrg_variant'] == 'LRG_321:g.17672_17674del'
        assert results['NM_001126114.2:c.652_654del']['hgvs_transcript_variant'] == 'NM_001126114.2:c.652_654del'
        assert results['NM_001126114.2:c.652_654del']['hgvs_refseqgene_variant'] == 'NG_017013.2:g.17672_17674del'
        assert results['NM_001126114.2:c.652_654del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578195_7578197del', 'vcf': {'chr': 'chr17', 'ref': 'GCAC', 'pos': '7578194', 'alt': 'G'}}
        assert results['NM_001126114.2:c.652_654del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7674877_7674879del', 'vcf': {'chr': 'chr17', 'ref': 'GCAC', 'pos': '7674876', 'alt': 'G'}}
        assert results['NM_001126114.2:c.652_654del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578195_7578197del', 'vcf': {'chr': '17', 'ref': 'GCAC', 'pos': '7578194', 'alt': 'G'}}
        assert results['NM_001126114.2:c.652_654del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7674877_7674879del', 'vcf': {'chr': '17', 'ref': 'GCAC', 'pos': '7674876', 'alt': 'G'}}

        assert 'NM_001276699.1:c.175_177del' in results.keys()
        assert results['NM_001276699.1:c.175_177del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001276699.1:c.175_177del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001276699.1:c.175_177del']['alt_genomic_loci'] == []
        assert results['NM_001276699.1:c.175_177del']['gene_symbol'] == 'TP53'
        assert results['NM_001276699.1:c.175_177del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001263628.1:p.(Val59del)', 'slr': 'NP_001263628.1:p.(V59del)'}
        assert results['NM_001276699.1:c.175_177del']['submitted_variant'] == '17-7578194-GCAC-G'
        assert results['NM_001276699.1:c.175_177del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001276699.1:c.175_177del']['hgvs_lrg_variant'] == ''
        assert results['NM_001276699.1:c.175_177del']['hgvs_transcript_variant'] == 'NM_001276699.1:c.175_177del'
        assert results['NM_001276699.1:c.175_177del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001276699.1:c.175_177del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578195_7578197del', 'vcf': {'chr': 'chr17', 'ref': 'GCAC', 'pos': '7578194', 'alt': 'G'}}
        assert results['NM_001276699.1:c.175_177del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7674877_7674879del', 'vcf': {'chr': 'chr17', 'ref': 'GCAC', 'pos': '7674876', 'alt': 'G'}}
        assert results['NM_001276699.1:c.175_177del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578195_7578197del', 'vcf': {'chr': '17', 'ref': 'GCAC', 'pos': '7578194', 'alt': 'G'}}
        assert results['NM_001276699.1:c.175_177del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7674877_7674879del', 'vcf': {'chr': '17', 'ref': 'GCAC', 'pos': '7674876', 'alt': 'G'}}

        assert 'NM_001276760.1:c.535_537del' in results.keys()
        assert results['NM_001276760.1:c.535_537del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001276760.1:c.535_537del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001276760.1:c.535_537del']['alt_genomic_loci'] == []
        assert results['NM_001276760.1:c.535_537del']['gene_symbol'] == 'TP53'
        assert results['NM_001276760.1:c.535_537del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001263689.1:p.(Val179del)', 'slr': 'NP_001263689.1:p.(V179del)'}
        assert results['NM_001276760.1:c.535_537del']['submitted_variant'] == '17-7578194-GCAC-G'
        assert results['NM_001276760.1:c.535_537del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001276760.1:c.535_537del']['hgvs_lrg_variant'] == ''
        assert results['NM_001276760.1:c.535_537del']['hgvs_transcript_variant'] == 'NM_001276760.1:c.535_537del'
        assert results['NM_001276760.1:c.535_537del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001276760.1:c.535_537del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578195_7578197del', 'vcf': {'chr': 'chr17', 'ref': 'GCAC', 'pos': '7578194', 'alt': 'G'}}
        assert results['NM_001276760.1:c.535_537del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7674877_7674879del', 'vcf': {'chr': 'chr17', 'ref': 'GCAC', 'pos': '7674876', 'alt': 'G'}}
        assert results['NM_001276760.1:c.535_537del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578195_7578197del', 'vcf': {'chr': '17', 'ref': 'GCAC', 'pos': '7578194', 'alt': 'G'}}
        assert results['NM_001276760.1:c.535_537del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7674877_7674879del', 'vcf': {'chr': '17', 'ref': 'GCAC', 'pos': '7674876', 'alt': 'G'}}

        assert 'NM_000546.5:c.652_654del' in results.keys()
        assert results['NM_000546.5:c.652_654del']['hgvs_lrg_transcript_variant'] == 'LRG_321t1:c.652_654del'
        assert results['NM_000546.5:c.652_654del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000546.5:c.652_654del']['alt_genomic_loci'] == []
        assert results['NM_000546.5:c.652_654del']['gene_symbol'] == 'TP53'
        assert results['NM_000546.5:c.652_654del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000537.3(LRG_321p1):p.(Val218del)', 'slr': 'NP_000537.3:p.(V218del)'}
        assert results['NM_000546.5:c.652_654del']['submitted_variant'] == '17-7578194-GCAC-G'
        assert results['NM_000546.5:c.652_654del']['genome_context_intronic_sequence'] == ''
        assert results['NM_000546.5:c.652_654del']['hgvs_lrg_variant'] == 'LRG_321:g.17672_17674del'
        assert results['NM_000546.5:c.652_654del']['hgvs_transcript_variant'] == 'NM_000546.5:c.652_654del'
        assert results['NM_000546.5:c.652_654del']['hgvs_refseqgene_variant'] == 'NG_017013.2:g.17672_17674del'
        assert results['NM_000546.5:c.652_654del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578195_7578197del', 'vcf': {'chr': 'chr17', 'ref': 'GCAC', 'pos': '7578194', 'alt': 'G'}}
        assert results['NM_000546.5:c.652_654del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7674877_7674879del', 'vcf': {'chr': 'chr17', 'ref': 'GCAC', 'pos': '7674876', 'alt': 'G'}}
        assert results['NM_000546.5:c.652_654del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578195_7578197del', 'vcf': {'chr': '17', 'ref': 'GCAC', 'pos': '7578194', 'alt': 'G'}}
        assert results['NM_000546.5:c.652_654del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7674877_7674879del', 'vcf': {'chr': '17', 'ref': 'GCAC', 'pos': '7674876', 'alt': 'G'}}


    def test_variant252(self):
        variant = '17-7578523-T-TG'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001276760.1:c.289dup' in results.keys()
        assert results['NM_001276760.1:c.289dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001276760.1:c.289dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001276760.1:c.289dup']['alt_genomic_loci'] == []
        assert results['NM_001276760.1:c.289dup']['gene_symbol'] == 'TP53'
        assert results['NM_001276760.1:c.289dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001263689.1:p.(Gln97ProfsTer13)', 'slr': 'NP_001263689.1:p.(Q97Pfs*13)'}
        assert results['NM_001276760.1:c.289dup']['submitted_variant'] == '17-7578523-T-TG'
        assert results['NM_001276760.1:c.289dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_001276760.1:c.289dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001276760.1:c.289dup']['hgvs_transcript_variant'] == 'NM_001276760.1:c.289dup'
        assert results['NM_001276760.1:c.289dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001276760.1:c.289dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578524dup', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '7578524', 'alt': u'GG'}}
        assert results['NM_001276760.1:c.289dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7675206dup', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '7675206', 'alt': u'GG'}}
        assert results['NM_001276760.1:c.289dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578524dup', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '7578524', 'alt': u'GG'}}
        assert results['NM_001276760.1:c.289dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7675206dup', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '7675206', 'alt': u'GG'}}

        assert 'NM_001276695.1:c.289dup' in results.keys()
        assert results['NM_001276695.1:c.289dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001276695.1:c.289dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001276695.1:c.289dup']['alt_genomic_loci'] == []
        assert results['NM_001276695.1:c.289dup']['gene_symbol'] == 'TP53'
        assert results['NM_001276695.1:c.289dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001263624.1:p.(Gln97ProfsTer13)', 'slr': 'NP_001263624.1:p.(Q97Pfs*13)'}
        assert results['NM_001276695.1:c.289dup']['submitted_variant'] == '17-7578523-T-TG'
        assert results['NM_001276695.1:c.289dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_001276695.1:c.289dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001276695.1:c.289dup']['hgvs_transcript_variant'] == 'NM_001276695.1:c.289dup'
        assert results['NM_001276695.1:c.289dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001276695.1:c.289dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578524dup', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '7578524', 'alt': u'GG'}}
        assert results['NM_001276695.1:c.289dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7675206dup', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '7675206', 'alt': u'GG'}}
        assert results['NM_001276695.1:c.289dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578524dup', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '7578524', 'alt': u'GG'}}
        assert results['NM_001276695.1:c.289dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7675206dup', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '7675206', 'alt': u'GG'}}

        assert 'NM_001276699.1:c.-72dup' in results.keys()
        assert results['NM_001276699.1:c.-72dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001276699.1:c.-72dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001276699.1:c.-72dup']['alt_genomic_loci'] == []
        assert results['NM_001276699.1:c.-72dup']['gene_symbol'] == 'TP53'
        assert results['NM_001276699.1:c.-72dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001263628.1:p.?', 'slr': 'NP_001263628.1:p.?'}
        assert results['NM_001276699.1:c.-72dup']['submitted_variant'] == '17-7578523-T-TG'
        assert results['NM_001276699.1:c.-72dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_001276699.1:c.-72dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001276699.1:c.-72dup']['hgvs_transcript_variant'] == 'NM_001276699.1:c.-72dup'
        assert results['NM_001276699.1:c.-72dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001276699.1:c.-72dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578524dup', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '7578524', 'alt': u'GG'}}
        assert results['NM_001276699.1:c.-72dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7675206dup', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '7675206', 'alt': u'GG'}}
        assert results['NM_001276699.1:c.-72dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578524dup', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '7578524', 'alt': u'GG'}}
        assert results['NM_001276699.1:c.-72dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7675206dup', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '7675206', 'alt': u'GG'}}

        assert 'NM_001126115.1:c.10dup' in results.keys()
        assert results['NM_001126115.1:c.10dup']['hgvs_lrg_transcript_variant'] == 'LRG_321t5:c.10dup'
        assert results['NM_001126115.1:c.10dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001126115.1:c.10dup']['alt_genomic_loci'] == []
        assert results['NM_001126115.1:c.10dup']['gene_symbol'] == 'TP53'
        assert results['NM_001126115.1:c.10dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001119587.1:p.(Gln4ProfsTer13)', 'slr': 'NP_001119587.1:p.(Q4Pfs*13)'}
        assert results['NM_001126115.1:c.10dup']['submitted_variant'] == '17-7578523-T-TG'
        assert results['NM_001126115.1:c.10dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_001126115.1:c.10dup']['hgvs_lrg_variant'] == 'LRG_321:g.17345dup'
        assert results['NM_001126115.1:c.10dup']['hgvs_transcript_variant'] == 'NM_001126115.1:c.10dup'
        assert results['NM_001126115.1:c.10dup']['hgvs_refseqgene_variant'] == 'NG_017013.2:g.17345dup'
        assert results['NM_001126115.1:c.10dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578524dup', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '7578524', 'alt': u'GG'}}
        assert results['NM_001126115.1:c.10dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7675206dup', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '7675206', 'alt': u'GG'}}
        assert results['NM_001126115.1:c.10dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578524dup', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '7578524', 'alt': u'GG'}}
        assert results['NM_001126115.1:c.10dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7675206dup', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '7675206', 'alt': u'GG'}}

        assert 'NM_001276697.1:c.-72dup' in results.keys()
        assert results['NM_001276697.1:c.-72dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001276697.1:c.-72dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001276697.1:c.-72dup']['alt_genomic_loci'] == []
        assert results['NM_001276697.1:c.-72dup']['gene_symbol'] == 'TP53'
        assert results['NM_001276697.1:c.-72dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001263626.1:p.?', 'slr': 'NP_001263626.1:p.?'}
        assert results['NM_001276697.1:c.-72dup']['submitted_variant'] == '17-7578523-T-TG'
        assert results['NM_001276697.1:c.-72dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_001276697.1:c.-72dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001276697.1:c.-72dup']['hgvs_transcript_variant'] == 'NM_001276697.1:c.-72dup'
        assert results['NM_001276697.1:c.-72dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001276697.1:c.-72dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578524dup', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '7578524', 'alt': u'GG'}}
        assert results['NM_001276697.1:c.-72dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7675206dup', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '7675206', 'alt': u'GG'}}
        assert results['NM_001276697.1:c.-72dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578524dup', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '7578524', 'alt': u'GG'}}
        assert results['NM_001276697.1:c.-72dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7675206dup', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '7675206', 'alt': u'GG'}}

        assert 'NM_001126117.1:c.10dup' in results.keys()
        assert results['NM_001126117.1:c.10dup']['hgvs_lrg_transcript_variant'] == 'LRG_321t7:c.10dup'
        assert results['NM_001126117.1:c.10dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001126117.1:c.10dup']['alt_genomic_loci'] == []
        assert results['NM_001126117.1:c.10dup']['gene_symbol'] == 'TP53'
        assert results['NM_001126117.1:c.10dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001119589.1:p.(Gln4ProfsTer13)', 'slr': 'NP_001119589.1:p.(Q4Pfs*13)'}
        assert results['NM_001126117.1:c.10dup']['submitted_variant'] == '17-7578523-T-TG'
        assert results['NM_001126117.1:c.10dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_001126117.1:c.10dup']['hgvs_lrg_variant'] == 'LRG_321:g.17345dup'
        assert results['NM_001126117.1:c.10dup']['hgvs_transcript_variant'] == 'NM_001126117.1:c.10dup'
        assert results['NM_001126117.1:c.10dup']['hgvs_refseqgene_variant'] == 'NG_017013.2:g.17345dup'
        assert results['NM_001126117.1:c.10dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578524dup', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '7578524', 'alt': u'GG'}}
        assert results['NM_001126117.1:c.10dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7675206dup', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '7675206', 'alt': u'GG'}}
        assert results['NM_001126117.1:c.10dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578524dup', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '7578524', 'alt': u'GG'}}
        assert results['NM_001126117.1:c.10dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7675206dup', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '7675206', 'alt': u'GG'}}

        assert 'NM_000546.5:c.406dup' in results.keys()
        assert results['NM_000546.5:c.406dup']['hgvs_lrg_transcript_variant'] == 'LRG_321t1:c.406dup'
        assert results['NM_000546.5:c.406dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000546.5:c.406dup']['alt_genomic_loci'] == []
        assert results['NM_000546.5:c.406dup']['gene_symbol'] == 'TP53'
        assert results['NM_000546.5:c.406dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000537.3(LRG_321p1):p.(Gln136ProfsTer13)', 'slr': 'NP_000537.3:p.(Q136Pfs*13)'}
        assert results['NM_000546.5:c.406dup']['submitted_variant'] == '17-7578523-T-TG'
        assert results['NM_000546.5:c.406dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_000546.5:c.406dup']['hgvs_lrg_variant'] == 'LRG_321:g.17345dup'
        assert results['NM_000546.5:c.406dup']['hgvs_transcript_variant'] == 'NM_000546.5:c.406dup'
        assert results['NM_000546.5:c.406dup']['hgvs_refseqgene_variant'] == 'NG_017013.2:g.17345dup'
        assert results['NM_000546.5:c.406dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578524dup', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '7578524', 'alt': u'GG'}}
        assert results['NM_000546.5:c.406dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7675206dup', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '7675206', 'alt': u'GG'}}
        assert results['NM_000546.5:c.406dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578524dup', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '7578524', 'alt': u'GG'}}
        assert results['NM_000546.5:c.406dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7675206dup', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '7675206', 'alt': u'GG'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001276696.1:c.289dup' in results.keys()
        assert results['NM_001276696.1:c.289dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001276696.1:c.289dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001276696.1:c.289dup']['alt_genomic_loci'] == []
        assert results['NM_001276696.1:c.289dup']['gene_symbol'] == 'TP53'
        assert results['NM_001276696.1:c.289dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001263625.1:p.(Gln97ProfsTer13)', 'slr': 'NP_001263625.1:p.(Q97Pfs*13)'}
        assert results['NM_001276696.1:c.289dup']['submitted_variant'] == '17-7578523-T-TG'
        assert results['NM_001276696.1:c.289dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_001276696.1:c.289dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001276696.1:c.289dup']['hgvs_transcript_variant'] == 'NM_001276696.1:c.289dup'
        assert results['NM_001276696.1:c.289dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001276696.1:c.289dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578524dup', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '7578524', 'alt': u'GG'}}
        assert results['NM_001276696.1:c.289dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7675206dup', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '7675206', 'alt': u'GG'}}
        assert results['NM_001276696.1:c.289dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578524dup', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '7578524', 'alt': u'GG'}}
        assert results['NM_001276696.1:c.289dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7675206dup', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '7675206', 'alt': u'GG'}}

        assert 'NM_001276698.1:c.-72dup' in results.keys()
        assert results['NM_001276698.1:c.-72dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001276698.1:c.-72dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001276698.1:c.-72dup']['alt_genomic_loci'] == []
        assert results['NM_001276698.1:c.-72dup']['gene_symbol'] == 'TP53'
        assert results['NM_001276698.1:c.-72dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001263627.1:p.?', 'slr': 'NP_001263627.1:p.?'}
        assert results['NM_001276698.1:c.-72dup']['submitted_variant'] == '17-7578523-T-TG'
        assert results['NM_001276698.1:c.-72dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_001276698.1:c.-72dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001276698.1:c.-72dup']['hgvs_transcript_variant'] == 'NM_001276698.1:c.-72dup'
        assert results['NM_001276698.1:c.-72dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001276698.1:c.-72dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578524dup', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '7578524', 'alt': u'GG'}}
        assert results['NM_001276698.1:c.-72dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7675206dup', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '7675206', 'alt': u'GG'}}
        assert results['NM_001276698.1:c.-72dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578524dup', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '7578524', 'alt': u'GG'}}
        assert results['NM_001276698.1:c.-72dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7675206dup', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '7675206', 'alt': u'GG'}}

        assert 'NM_001126114.2:c.406dup' in results.keys()
        assert results['NM_001126114.2:c.406dup']['hgvs_lrg_transcript_variant'] == 'LRG_321t3:c.406dup'
        assert results['NM_001126114.2:c.406dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001126114.2:c.406dup']['alt_genomic_loci'] == []
        assert results['NM_001126114.2:c.406dup']['gene_symbol'] == 'TP53'
        assert results['NM_001126114.2:c.406dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001119586.1:p.(Gln136ProfsTer13)', 'slr': 'NP_001119586.1:p.(Q136Pfs*13)'}
        assert results['NM_001126114.2:c.406dup']['submitted_variant'] == '17-7578523-T-TG'
        assert results['NM_001126114.2:c.406dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_001126114.2:c.406dup']['hgvs_lrg_variant'] == 'LRG_321:g.17345dup'
        assert results['NM_001126114.2:c.406dup']['hgvs_transcript_variant'] == 'NM_001126114.2:c.406dup'
        assert results['NM_001126114.2:c.406dup']['hgvs_refseqgene_variant'] == 'NG_017013.2:g.17345dup'
        assert results['NM_001126114.2:c.406dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578524dup', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '7578524', 'alt': u'GG'}}
        assert results['NM_001126114.2:c.406dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7675206dup', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '7675206', 'alt': u'GG'}}
        assert results['NM_001126114.2:c.406dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578524dup', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '7578524', 'alt': u'GG'}}
        assert results['NM_001126114.2:c.406dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7675206dup', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '7675206', 'alt': u'GG'}}

        assert 'NM_001126118.1:c.289dup' in results.keys()
        assert results['NM_001126118.1:c.289dup']['hgvs_lrg_transcript_variant'] == 'LRG_321t8:c.289dup'
        assert results['NM_001126118.1:c.289dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001126118.1:c.289dup']['alt_genomic_loci'] == []
        assert results['NM_001126118.1:c.289dup']['gene_symbol'] == 'TP53'
        assert results['NM_001126118.1:c.289dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001119590.1:p.(Gln97ProfsTer13)', 'slr': 'NP_001119590.1:p.(Q97Pfs*13)'}
        assert results['NM_001126118.1:c.289dup']['submitted_variant'] == '17-7578523-T-TG'
        assert results['NM_001126118.1:c.289dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_001126118.1:c.289dup']['hgvs_lrg_variant'] == 'LRG_321:g.17345dup'
        assert results['NM_001126118.1:c.289dup']['hgvs_transcript_variant'] == 'NM_001126118.1:c.289dup'
        assert results['NM_001126118.1:c.289dup']['hgvs_refseqgene_variant'] == 'NG_017013.2:g.17345dup'
        assert results['NM_001126118.1:c.289dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578524dup', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '7578524', 'alt': u'GG'}}
        assert results['NM_001126118.1:c.289dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7675206dup', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '7675206', 'alt': u'GG'}}
        assert results['NM_001126118.1:c.289dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578524dup', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '7578524', 'alt': u'GG'}}
        assert results['NM_001126118.1:c.289dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7675206dup', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '7675206', 'alt': u'GG'}}

        assert 'NM_001126113.2:c.406dup' in results.keys()
        assert results['NM_001126113.2:c.406dup']['hgvs_lrg_transcript_variant'] == 'LRG_321t4:c.406dup'
        assert results['NM_001126113.2:c.406dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001126113.2:c.406dup']['alt_genomic_loci'] == []
        assert results['NM_001126113.2:c.406dup']['gene_symbol'] == 'TP53'
        assert results['NM_001126113.2:c.406dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001119585.1:p.(Gln136ProfsTer13)', 'slr': 'NP_001119585.1:p.(Q136Pfs*13)'}
        assert results['NM_001126113.2:c.406dup']['submitted_variant'] == '17-7578523-T-TG'
        assert results['NM_001126113.2:c.406dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_001126113.2:c.406dup']['hgvs_lrg_variant'] == 'LRG_321:g.17345dup'
        assert results['NM_001126113.2:c.406dup']['hgvs_transcript_variant'] == 'NM_001126113.2:c.406dup'
        assert results['NM_001126113.2:c.406dup']['hgvs_refseqgene_variant'] == 'NG_017013.2:g.17345dup'
        assert results['NM_001126113.2:c.406dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578524dup', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '7578524', 'alt': u'GG'}}
        assert results['NM_001126113.2:c.406dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7675206dup', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '7675206', 'alt': u'GG'}}
        assert results['NM_001126113.2:c.406dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578524dup', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '7578524', 'alt': u'GG'}}
        assert results['NM_001126113.2:c.406dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7675206dup', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '7675206', 'alt': u'GG'}}

        assert 'NM_001276761.1:c.289dup' in results.keys()
        assert results['NM_001276761.1:c.289dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001276761.1:c.289dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001276761.1:c.289dup']['alt_genomic_loci'] == []
        assert results['NM_001276761.1:c.289dup']['gene_symbol'] == 'TP53'
        assert results['NM_001276761.1:c.289dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001263690.1:p.(Gln97ProfsTer13)', 'slr': 'NP_001263690.1:p.(Q97Pfs*13)'}
        assert results['NM_001276761.1:c.289dup']['submitted_variant'] == '17-7578523-T-TG'
        assert results['NM_001276761.1:c.289dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_001276761.1:c.289dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001276761.1:c.289dup']['hgvs_transcript_variant'] == 'NM_001276761.1:c.289dup'
        assert results['NM_001276761.1:c.289dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001276761.1:c.289dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578524dup', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '7578524', 'alt': u'GG'}}
        assert results['NM_001276761.1:c.289dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7675206dup', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '7675206', 'alt': u'GG'}}
        assert results['NM_001276761.1:c.289dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578524dup', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '7578524', 'alt': u'GG'}}
        assert results['NM_001276761.1:c.289dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7675206dup', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '7675206', 'alt': u'GG'}}

        assert 'NM_001126116.1:c.10dup' in results.keys()
        assert results['NM_001126116.1:c.10dup']['hgvs_lrg_transcript_variant'] == 'LRG_321t6:c.10dup'
        assert results['NM_001126116.1:c.10dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001126116.1:c.10dup']['alt_genomic_loci'] == []
        assert results['NM_001126116.1:c.10dup']['gene_symbol'] == 'TP53'
        assert results['NM_001126116.1:c.10dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001119588.1:p.(Gln4ProfsTer13)', 'slr': 'NP_001119588.1:p.(Q4Pfs*13)'}
        assert results['NM_001126116.1:c.10dup']['submitted_variant'] == '17-7578523-T-TG'
        assert results['NM_001126116.1:c.10dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_001126116.1:c.10dup']['hgvs_lrg_variant'] == 'LRG_321:g.17345dup'
        assert results['NM_001126116.1:c.10dup']['hgvs_transcript_variant'] == 'NM_001126116.1:c.10dup'
        assert results['NM_001126116.1:c.10dup']['hgvs_refseqgene_variant'] == 'NG_017013.2:g.17345dup'
        assert results['NM_001126116.1:c.10dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578524dup', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '7578524', 'alt': u'GG'}}
        assert results['NM_001126116.1:c.10dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7675206dup', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '7675206', 'alt': u'GG'}}
        assert results['NM_001126116.1:c.10dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578524dup', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '7578524', 'alt': u'GG'}}
        assert results['NM_001126116.1:c.10dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7675206dup', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '7675206', 'alt': u'GG'}}

        assert 'NM_001126112.2:c.406dup' in results.keys()
        assert results['NM_001126112.2:c.406dup']['hgvs_lrg_transcript_variant'] == 'LRG_321t2:c.406dup'
        assert results['NM_001126112.2:c.406dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001126112.2:c.406dup']['alt_genomic_loci'] == []
        assert results['NM_001126112.2:c.406dup']['gene_symbol'] == 'TP53'
        assert results['NM_001126112.2:c.406dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001119584.1:p.(Gln136ProfsTer13)', 'slr': 'NP_001119584.1:p.(Q136Pfs*13)'}
        assert results['NM_001126112.2:c.406dup']['submitted_variant'] == '17-7578523-T-TG'
        assert results['NM_001126112.2:c.406dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_001126112.2:c.406dup']['hgvs_lrg_variant'] == 'LRG_321:g.17345dup'
        assert results['NM_001126112.2:c.406dup']['hgvs_transcript_variant'] == 'NM_001126112.2:c.406dup'
        assert results['NM_001126112.2:c.406dup']['hgvs_refseqgene_variant'] == 'NG_017013.2:g.17345dup'
        assert results['NM_001126112.2:c.406dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578524dup', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '7578524', 'alt': u'GG'}}
        assert results['NM_001126112.2:c.406dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7675206dup', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '7675206', 'alt': u'GG'}}
        assert results['NM_001126112.2:c.406dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.7578524dup', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '7578524', 'alt': u'GG'}}
        assert results['NM_001126112.2:c.406dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.7675206dup', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '7675206', 'alt': u'GG'}}


    def test_variant253(self):
        variant = '17-17119692-A-C'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_144997.6:c.1300+2T>G' in results.keys()
        assert results['NM_144997.6:c.1300+2T>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_144997.6:c.1300+2T>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_144997.6:c.1300+2T>G']['alt_genomic_loci'] == []
        assert results['NM_144997.6:c.1300+2T>G']['gene_symbol'] == 'FLCN'
        assert results['NM_144997.6:c.1300+2T>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_659434.2(LRG_325p1):p.?', 'slr': 'NP_659434.2:p.?'}
        assert results['NM_144997.6:c.1300+2T>G']['submitted_variant'] == '17-17119692-A-C'
        assert results['NM_144997.6:c.1300+2T>G']['genome_context_intronic_sequence'] == 'NC_000017.10(NM_144997.6):c.1300+2T>G'
        assert results['NM_144997.6:c.1300+2T>G']['hgvs_lrg_variant'] == ''
        assert results['NM_144997.6:c.1300+2T>G']['hgvs_transcript_variant'] == 'NM_144997.6:c.1300+2T>G'
        assert results['NM_144997.6:c.1300+2T>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_144997.6:c.1300+2T>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.17119692A>C', 'vcf': {'chr': 'chr17', 'ref': u'A', 'pos': '17119692', 'alt': u'C'}}
        assert results['NM_144997.6:c.1300+2T>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.17216378A>C', 'vcf': {'chr': 'chr17', 'ref': u'A', 'pos': '17216378', 'alt': u'C'}}
        assert results['NM_144997.6:c.1300+2T>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.17119692A>C', 'vcf': {'chr': '17', 'ref': u'A', 'pos': '17119692', 'alt': u'C'}}
        assert results['NM_144997.6:c.1300+2T>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.17216378A>C', 'vcf': {'chr': '17', 'ref': u'A', 'pos': '17216378', 'alt': u'C'}}

        assert 'NM_001353230.1:c.1300+2T>G' in results.keys()
        assert results['NM_001353230.1:c.1300+2T>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001353230.1:c.1300+2T>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001353230.1:c.1300+2T>G']['alt_genomic_loci'] == []
        assert results['NM_001353230.1:c.1300+2T>G']['gene_symbol'] == 'FLCN'
        assert results['NM_001353230.1:c.1300+2T>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001340159.1:p.?', 'slr': 'NP_001340159.1:p.?'}
        assert results['NM_001353230.1:c.1300+2T>G']['submitted_variant'] == '17-17119692-A-C'
        assert results['NM_001353230.1:c.1300+2T>G']['genome_context_intronic_sequence'] == 'NC_000017.10(NM_001353230.1):c.1300+2T>G'
        assert results['NM_001353230.1:c.1300+2T>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001353230.1:c.1300+2T>G']['hgvs_transcript_variant'] == 'NM_001353230.1:c.1300+2T>G'
        assert results['NM_001353230.1:c.1300+2T>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001353230.1:c.1300+2T>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.17119692A>C', 'vcf': {'chr': 'chr17', 'ref': u'A', 'pos': '17119692', 'alt': u'C'}}
        assert results['NM_001353230.1:c.1300+2T>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.17216378A>C', 'vcf': {'chr': 'chr17', 'ref': u'A', 'pos': '17216378', 'alt': u'C'}}
        assert results['NM_001353230.1:c.1300+2T>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.17119692A>C', 'vcf': {'chr': '17', 'ref': u'A', 'pos': '17119692', 'alt': u'C'}}
        assert results['NM_001353230.1:c.1300+2T>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.17216378A>C', 'vcf': {'chr': '17', 'ref': u'A', 'pos': '17216378', 'alt': u'C'}}

        assert 'NM_001353229.1:c.1354+2T>G' in results.keys()
        assert results['NM_001353229.1:c.1354+2T>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001353229.1:c.1354+2T>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001353229.1:c.1354+2T>G']['alt_genomic_loci'] == []
        assert results['NM_001353229.1:c.1354+2T>G']['gene_symbol'] == 'FLCN'
        assert results['NM_001353229.1:c.1354+2T>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001340158.1:p.?', 'slr': 'NP_001340158.1:p.?'}
        assert results['NM_001353229.1:c.1354+2T>G']['submitted_variant'] == '17-17119692-A-C'
        assert results['NM_001353229.1:c.1354+2T>G']['genome_context_intronic_sequence'] == 'NC_000017.10(NM_001353229.1):c.1354+2T>G'
        assert results['NM_001353229.1:c.1354+2T>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001353229.1:c.1354+2T>G']['hgvs_transcript_variant'] == 'NM_001353229.1:c.1354+2T>G'
        assert results['NM_001353229.1:c.1354+2T>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001353229.1:c.1354+2T>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.17119692A>C', 'vcf': {'chr': 'chr17', 'ref': u'A', 'pos': '17119692', 'alt': u'C'}}
        assert results['NM_001353229.1:c.1354+2T>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.17216378A>C', 'vcf': {'chr': 'chr17', 'ref': u'A', 'pos': '17216378', 'alt': u'C'}}
        assert results['NM_001353229.1:c.1354+2T>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.17119692A>C', 'vcf': {'chr': '17', 'ref': u'A', 'pos': '17119692', 'alt': u'C'}}
        assert results['NM_001353229.1:c.1354+2T>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.17216378A>C', 'vcf': {'chr': '17', 'ref': u'A', 'pos': '17216378', 'alt': u'C'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_144997.5:c.1300+2T>G' in results.keys()
        assert results['NM_144997.5:c.1300+2T>G']['hgvs_lrg_transcript_variant'] == 'LRG_325t1:c.1300+2T>G'
        assert results['NM_144997.5:c.1300+2T>G']['refseqgene_context_intronic_sequence'] == 'NG_008001.2(NM_144997.5):c.1300+2T>G'
        assert results['NM_144997.5:c.1300+2T>G']['alt_genomic_loci'] == []
        assert results['NM_144997.5:c.1300+2T>G']['gene_symbol'] == 'FLCN'
        assert results['NM_144997.5:c.1300+2T>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_659434.2(LRG_325p1):p.?', 'slr': 'NP_659434.2:p.?'}
        assert results['NM_144997.5:c.1300+2T>G']['submitted_variant'] == '17-17119692-A-C'
        assert results['NM_144997.5:c.1300+2T>G']['genome_context_intronic_sequence'] == 'NC_000017.10(NM_144997.5):c.1300+2T>G'
        assert results['NM_144997.5:c.1300+2T>G']['hgvs_lrg_variant'] == 'LRG_325:g.25811T>G'
        assert results['NM_144997.5:c.1300+2T>G']['hgvs_transcript_variant'] == 'NM_144997.5:c.1300+2T>G'
        assert results['NM_144997.5:c.1300+2T>G']['hgvs_refseqgene_variant'] == 'NG_008001.2:g.25811T>G'
        assert results['NM_144997.5:c.1300+2T>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.17119692A>C', 'vcf': {'chr': 'chr17', 'ref': u'A', 'pos': '17119692', 'alt': u'C'}}
        assert results['NM_144997.5:c.1300+2T>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.17216378A>C', 'vcf': {'chr': 'chr17', 'ref': u'A', 'pos': '17216378', 'alt': u'C'}}
        assert results['NM_144997.5:c.1300+2T>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.17119692A>C', 'vcf': {'chr': '17', 'ref': u'A', 'pos': '17119692', 'alt': u'C'}}
        assert results['NM_144997.5:c.1300+2T>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.17216378A>C', 'vcf': {'chr': '17', 'ref': u'A', 'pos': '17216378', 'alt': u'C'}}

        assert 'NM_001353231.1:c.1300+2T>G' in results.keys()
        assert results['NM_001353231.1:c.1300+2T>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001353231.1:c.1300+2T>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001353231.1:c.1300+2T>G']['alt_genomic_loci'] == []
        assert results['NM_001353231.1:c.1300+2T>G']['gene_symbol'] == 'FLCN'
        assert results['NM_001353231.1:c.1300+2T>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001340160.1:p.?', 'slr': 'NP_001340160.1:p.?'}
        assert results['NM_001353231.1:c.1300+2T>G']['submitted_variant'] == '17-17119692-A-C'
        assert results['NM_001353231.1:c.1300+2T>G']['genome_context_intronic_sequence'] == 'NC_000017.10(NM_001353231.1):c.1300+2T>G'
        assert results['NM_001353231.1:c.1300+2T>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001353231.1:c.1300+2T>G']['hgvs_transcript_variant'] == 'NM_001353231.1:c.1300+2T>G'
        assert results['NM_001353231.1:c.1300+2T>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001353231.1:c.1300+2T>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.17119692A>C', 'vcf': {'chr': 'chr17', 'ref': u'A', 'pos': '17119692', 'alt': u'C'}}
        assert results['NM_001353231.1:c.1300+2T>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.17216378A>C', 'vcf': {'chr': 'chr17', 'ref': u'A', 'pos': '17216378', 'alt': u'C'}}
        assert results['NM_001353231.1:c.1300+2T>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.17119692A>C', 'vcf': {'chr': '17', 'ref': u'A', 'pos': '17119692', 'alt': u'C'}}
        assert results['NM_001353231.1:c.1300+2T>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.17216378A>C', 'vcf': {'chr': '17', 'ref': u'A', 'pos': '17216378', 'alt': u'C'}}


    def test_variant254(self):
        variant = '17-41197588-GGACA-G'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_007294.3:c.*103_*106del' in results.keys()
        assert results['NM_007294.3:c.*103_*106del']['hgvs_lrg_transcript_variant'] == 'LRG_292t1:c.*103_*106del'
        assert results['NM_007294.3:c.*103_*106del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_007294.3:c.*103_*106del']['alt_genomic_loci'] == []
        assert results['NM_007294.3:c.*103_*106del']['gene_symbol'] == 'BRCA1'
        assert results['NM_007294.3:c.*103_*106del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_009225.1(LRG_292p1):p.?', 'slr': 'NP_009225.1:p.?'}
        assert results['NM_007294.3:c.*103_*106del']['submitted_variant'] == '17-41197588-GGACA-G'
        assert results['NM_007294.3:c.*103_*106del']['genome_context_intronic_sequence'] == ''
        assert results['NM_007294.3:c.*103_*106del']['hgvs_lrg_variant'] == 'LRG_292:g.172409_172412del'
        assert results['NM_007294.3:c.*103_*106del']['hgvs_transcript_variant'] == 'NM_007294.3:c.*103_*106del'
        assert results['NM_007294.3:c.*103_*106del']['hgvs_refseqgene_variant'] == 'NG_005905.2:g.172409_172412del'
        assert results['NM_007294.3:c.*103_*106del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.41197589_41197592del', 'vcf': {'chr': 'chr17', 'ref': 'GGACA', 'pos': '41197588', 'alt': 'G'}}
        assert results['NM_007294.3:c.*103_*106del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.43045572_43045575del', 'vcf': {'chr': 'chr17', 'ref': 'GGACA', 'pos': '43045571', 'alt': 'G'}}
        assert results['NM_007294.3:c.*103_*106del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.41197589_41197592del', 'vcf': {'chr': '17', 'ref': 'GGACA', 'pos': '41197588', 'alt': 'G'}}
        assert results['NM_007294.3:c.*103_*106del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.43045572_43045575del', 'vcf': {'chr': '17', 'ref': 'GGACA', 'pos': '43045571', 'alt': 'G'}}

        assert 'NM_007297.3:c.*103_*106del' in results.keys()
        assert results['NM_007297.3:c.*103_*106del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_007297.3:c.*103_*106del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_007297.3:c.*103_*106del']['alt_genomic_loci'] == []
        assert results['NM_007297.3:c.*103_*106del']['gene_symbol'] == 'BRCA1'
        assert results['NM_007297.3:c.*103_*106del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_009228.2:p.?', 'slr': 'NP_009228.2:p.?'}
        assert results['NM_007297.3:c.*103_*106del']['submitted_variant'] == '17-41197588-GGACA-G'
        assert results['NM_007297.3:c.*103_*106del']['genome_context_intronic_sequence'] == ''
        assert results['NM_007297.3:c.*103_*106del']['hgvs_lrg_variant'] == ''
        assert results['NM_007297.3:c.*103_*106del']['hgvs_transcript_variant'] == 'NM_007297.3:c.*103_*106del'
        assert results['NM_007297.3:c.*103_*106del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_007297.3:c.*103_*106del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.41197589_41197592del', 'vcf': {'chr': 'chr17', 'ref': 'GGACA', 'pos': '41197588', 'alt': 'G'}}
        assert results['NM_007297.3:c.*103_*106del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.43045572_43045575del', 'vcf': {'chr': 'chr17', 'ref': 'GGACA', 'pos': '43045571', 'alt': 'G'}}
        assert results['NM_007297.3:c.*103_*106del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.41197589_41197592del', 'vcf': {'chr': '17', 'ref': 'GGACA', 'pos': '41197588', 'alt': 'G'}}
        assert results['NM_007297.3:c.*103_*106del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.43045572_43045575del', 'vcf': {'chr': '17', 'ref': 'GGACA', 'pos': '43045571', 'alt': 'G'}}

        assert 'NM_007300.3:c.*103_*106del' in results.keys()
        assert results['NM_007300.3:c.*103_*106del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_007300.3:c.*103_*106del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_007300.3:c.*103_*106del']['alt_genomic_loci'] == []
        assert results['NM_007300.3:c.*103_*106del']['gene_symbol'] == 'BRCA1'
        assert results['NM_007300.3:c.*103_*106del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_009231.2:p.?', 'slr': 'NP_009231.2:p.?'}
        assert results['NM_007300.3:c.*103_*106del']['submitted_variant'] == '17-41197588-GGACA-G'
        assert results['NM_007300.3:c.*103_*106del']['genome_context_intronic_sequence'] == ''
        assert results['NM_007300.3:c.*103_*106del']['hgvs_lrg_variant'] == ''
        assert results['NM_007300.3:c.*103_*106del']['hgvs_transcript_variant'] == 'NM_007300.3:c.*103_*106del'
        assert results['NM_007300.3:c.*103_*106del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_007300.3:c.*103_*106del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.41197589_41197592del', 'vcf': {'chr': 'chr17', 'ref': 'GGACA', 'pos': '41197588', 'alt': 'G'}}
        assert results['NM_007300.3:c.*103_*106del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.43045572_43045575del', 'vcf': {'chr': 'chr17', 'ref': 'GGACA', 'pos': '43045571', 'alt': 'G'}}
        assert results['NM_007300.3:c.*103_*106del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.41197589_41197592del', 'vcf': {'chr': '17', 'ref': 'GGACA', 'pos': '41197588', 'alt': 'G'}}
        assert results['NM_007300.3:c.*103_*106del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.43045572_43045575del', 'vcf': {'chr': '17', 'ref': 'GGACA', 'pos': '43045571', 'alt': 'G'}}

        assert 'NR_027676.1:n.5831_5834del' in results.keys()
        assert results['NR_027676.1:n.5831_5834del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_027676.1:n.5831_5834del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_027676.1:n.5831_5834del']['alt_genomic_loci'] == []
        assert results['NR_027676.1:n.5831_5834del']['gene_symbol'] == 'BRCA1'
        assert results['NR_027676.1:n.5831_5834del']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_027676.1:n.5831_5834del']['submitted_variant'] == '17-41197588-GGACA-G'
        assert results['NR_027676.1:n.5831_5834del']['genome_context_intronic_sequence'] == ''
        assert results['NR_027676.1:n.5831_5834del']['hgvs_lrg_variant'] == ''
        assert results['NR_027676.1:n.5831_5834del']['hgvs_transcript_variant'] == 'NR_027676.1:n.5831_5834del'
        assert results['NR_027676.1:n.5831_5834del']['hgvs_refseqgene_variant'] == ''
        assert results['NR_027676.1:n.5831_5834del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.41197589_41197592del', 'vcf': {'chr': 'chr17', 'ref': 'GGACA', 'pos': '41197588', 'alt': 'G'}}
        assert results['NR_027676.1:n.5831_5834del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.43045572_43045575del', 'vcf': {'chr': 'chr17', 'ref': 'GGACA', 'pos': '43045571', 'alt': 'G'}}
        assert results['NR_027676.1:n.5831_5834del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.41197589_41197592del', 'vcf': {'chr': '17', 'ref': 'GGACA', 'pos': '41197588', 'alt': 'G'}}
        assert results['NR_027676.1:n.5831_5834del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.43045572_43045575del', 'vcf': {'chr': '17', 'ref': 'GGACA', 'pos': '43045571', 'alt': 'G'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_007299.3:c.*209_*212del' in results.keys()
        assert results['NM_007299.3:c.*209_*212del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_007299.3:c.*209_*212del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_007299.3:c.*209_*212del']['alt_genomic_loci'] == []
        assert results['NM_007299.3:c.*209_*212del']['gene_symbol'] == 'BRCA1'
        assert results['NM_007299.3:c.*209_*212del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_009230.2:p.?', 'slr': 'NP_009230.2:p.?'}
        assert results['NM_007299.3:c.*209_*212del']['submitted_variant'] == '17-41197588-GGACA-G'
        assert results['NM_007299.3:c.*209_*212del']['genome_context_intronic_sequence'] == ''
        assert results['NM_007299.3:c.*209_*212del']['hgvs_lrg_variant'] == ''
        assert results['NM_007299.3:c.*209_*212del']['hgvs_transcript_variant'] == 'NM_007299.3:c.*209_*212del'
        assert results['NM_007299.3:c.*209_*212del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_007299.3:c.*209_*212del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.41197589_41197592del', 'vcf': {'chr': 'chr17', 'ref': 'GGACA', 'pos': '41197588', 'alt': 'G'}}
        assert results['NM_007299.3:c.*209_*212del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.43045572_43045575del', 'vcf': {'chr': 'chr17', 'ref': 'GGACA', 'pos': '43045571', 'alt': 'G'}}
        assert results['NM_007299.3:c.*209_*212del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.41197589_41197592del', 'vcf': {'chr': '17', 'ref': 'GGACA', 'pos': '41197588', 'alt': 'G'}}
        assert results['NM_007299.3:c.*209_*212del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.43045572_43045575del', 'vcf': {'chr': '17', 'ref': 'GGACA', 'pos': '43045571', 'alt': 'G'}}

        assert 'NM_007298.3:c.*103_*106del' in results.keys()
        assert results['NM_007298.3:c.*103_*106del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_007298.3:c.*103_*106del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_007298.3:c.*103_*106del']['alt_genomic_loci'] == []
        assert results['NM_007298.3:c.*103_*106del']['gene_symbol'] == 'BRCA1'
        assert results['NM_007298.3:c.*103_*106del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_009229.2:p.?', 'slr': 'NP_009229.2:p.?'}
        assert results['NM_007298.3:c.*103_*106del']['submitted_variant'] == '17-41197588-GGACA-G'
        assert results['NM_007298.3:c.*103_*106del']['genome_context_intronic_sequence'] == ''
        assert results['NM_007298.3:c.*103_*106del']['hgvs_lrg_variant'] == ''
        assert results['NM_007298.3:c.*103_*106del']['hgvs_transcript_variant'] == 'NM_007298.3:c.*103_*106del'
        assert results['NM_007298.3:c.*103_*106del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_007298.3:c.*103_*106del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.41197589_41197592del', 'vcf': {'chr': 'chr17', 'ref': 'GGACA', 'pos': '41197588', 'alt': 'G'}}
        assert results['NM_007298.3:c.*103_*106del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.43045572_43045575del', 'vcf': {'chr': 'chr17', 'ref': 'GGACA', 'pos': '43045571', 'alt': 'G'}}
        assert results['NM_007298.3:c.*103_*106del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.41197589_41197592del', 'vcf': {'chr': '17', 'ref': 'GGACA', 'pos': '41197588', 'alt': 'G'}}
        assert results['NM_007298.3:c.*103_*106del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.43045572_43045575del', 'vcf': {'chr': '17', 'ref': 'GGACA', 'pos': '43045571', 'alt': 'G'}}


    def test_variant255(self):
        variant = '17-41256884-C-G'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_007299.3:c.301+1G>C' in results.keys()
        assert results['NM_007299.3:c.301+1G>C']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_007299.3:c.301+1G>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_007299.3:c.301+1G>C']['alt_genomic_loci'] == []
        assert results['NM_007299.3:c.301+1G>C']['gene_symbol'] == 'BRCA1'
        assert results['NM_007299.3:c.301+1G>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_009230.2:p.?', 'slr': 'NP_009230.2:p.?'}
        assert results['NM_007299.3:c.301+1G>C']['submitted_variant'] == '17-41256884-C-G'
        assert results['NM_007299.3:c.301+1G>C']['genome_context_intronic_sequence'] == 'NC_000017.10(NM_007299.3):c.301+1G>C'
        assert results['NM_007299.3:c.301+1G>C']['hgvs_lrg_variant'] == ''
        assert results['NM_007299.3:c.301+1G>C']['hgvs_transcript_variant'] == 'NM_007299.3:c.301+1G>C'
        assert results['NM_007299.3:c.301+1G>C']['hgvs_refseqgene_variant'] == ''
        assert results['NM_007299.3:c.301+1G>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.41256884C>G', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '41256884', 'alt': u'G'}}
        assert results['NM_007299.3:c.301+1G>C']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.43104867C>G', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '43104867', 'alt': u'G'}}
        assert results['NM_007299.3:c.301+1G>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.41256884C>G', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '41256884', 'alt': u'G'}}
        assert results['NM_007299.3:c.301+1G>C']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.43104867C>G', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '43104867', 'alt': u'G'}}

        assert 'NR_027676.1:n.440+1G>C' in results.keys()
        assert results['NR_027676.1:n.440+1G>C']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_027676.1:n.440+1G>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_027676.1:n.440+1G>C']['alt_genomic_loci'] == []
        assert results['NR_027676.1:n.440+1G>C']['gene_symbol'] == 'BRCA1'
        assert results['NR_027676.1:n.440+1G>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_027676.1:n.440+1G>C']['submitted_variant'] == '17-41256884-C-G'
        assert results['NR_027676.1:n.440+1G>C']['genome_context_intronic_sequence'] == 'NC_000017.10(NR_027676.1):c.440+1G>C'
        assert results['NR_027676.1:n.440+1G>C']['hgvs_lrg_variant'] == ''
        assert results['NR_027676.1:n.440+1G>C']['hgvs_transcript_variant'] == 'NR_027676.1:n.440+1G>C'
        assert results['NR_027676.1:n.440+1G>C']['hgvs_refseqgene_variant'] == ''
        assert results['NR_027676.1:n.440+1G>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.41256884C>G', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '41256884', 'alt': u'G'}}
        assert results['NR_027676.1:n.440+1G>C']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.43104867C>G', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '43104867', 'alt': u'G'}}
        assert results['NR_027676.1:n.440+1G>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.41256884C>G', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '41256884', 'alt': u'G'}}
        assert results['NR_027676.1:n.440+1G>C']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.43104867C>G', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '43104867', 'alt': u'G'}}

        assert 'NM_007300.3:c.301+1G>C' in results.keys()
        assert results['NM_007300.3:c.301+1G>C']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_007300.3:c.301+1G>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_007300.3:c.301+1G>C']['alt_genomic_loci'] == []
        assert results['NM_007300.3:c.301+1G>C']['gene_symbol'] == 'BRCA1'
        assert results['NM_007300.3:c.301+1G>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_009231.2:p.?', 'slr': 'NP_009231.2:p.?'}
        assert results['NM_007300.3:c.301+1G>C']['submitted_variant'] == '17-41256884-C-G'
        assert results['NM_007300.3:c.301+1G>C']['genome_context_intronic_sequence'] == 'NC_000017.10(NM_007300.3):c.301+1G>C'
        assert results['NM_007300.3:c.301+1G>C']['hgvs_lrg_variant'] == ''
        assert results['NM_007300.3:c.301+1G>C']['hgvs_transcript_variant'] == 'NM_007300.3:c.301+1G>C'
        assert results['NM_007300.3:c.301+1G>C']['hgvs_refseqgene_variant'] == ''
        assert results['NM_007300.3:c.301+1G>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.41256884C>G', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '41256884', 'alt': u'G'}}
        assert results['NM_007300.3:c.301+1G>C']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.43104867C>G', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '43104867', 'alt': u'G'}}
        assert results['NM_007300.3:c.301+1G>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.41256884C>G', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '41256884', 'alt': u'G'}}
        assert results['NM_007300.3:c.301+1G>C']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.43104867C>G', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '43104867', 'alt': u'G'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_007297.3:c.160+1G>C' in results.keys()
        assert results['NM_007297.3:c.160+1G>C']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_007297.3:c.160+1G>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_007297.3:c.160+1G>C']['alt_genomic_loci'] == []
        assert results['NM_007297.3:c.160+1G>C']['gene_symbol'] == 'BRCA1'
        assert results['NM_007297.3:c.160+1G>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_009228.2:p.?', 'slr': 'NP_009228.2:p.?'}
        assert results['NM_007297.3:c.160+1G>C']['submitted_variant'] == '17-41256884-C-G'
        assert results['NM_007297.3:c.160+1G>C']['genome_context_intronic_sequence'] == 'NC_000017.10(NM_007297.3):c.160+1G>C'
        assert results['NM_007297.3:c.160+1G>C']['hgvs_lrg_variant'] == ''
        assert results['NM_007297.3:c.160+1G>C']['hgvs_transcript_variant'] == 'NM_007297.3:c.160+1G>C'
        assert results['NM_007297.3:c.160+1G>C']['hgvs_refseqgene_variant'] == ''
        assert results['NM_007297.3:c.160+1G>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.41256884C>G', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '41256884', 'alt': u'G'}}
        assert results['NM_007297.3:c.160+1G>C']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.43104867C>G', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '43104867', 'alt': u'G'}}
        assert results['NM_007297.3:c.160+1G>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.41256884C>G', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '41256884', 'alt': u'G'}}
        assert results['NM_007297.3:c.160+1G>C']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.43104867C>G', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '43104867', 'alt': u'G'}}

        assert 'NM_007298.3:c.301+1G>C' in results.keys()
        assert results['NM_007298.3:c.301+1G>C']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_007298.3:c.301+1G>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_007298.3:c.301+1G>C']['alt_genomic_loci'] == []
        assert results['NM_007298.3:c.301+1G>C']['gene_symbol'] == 'BRCA1'
        assert results['NM_007298.3:c.301+1G>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_009229.2:p.?', 'slr': 'NP_009229.2:p.?'}
        assert results['NM_007298.3:c.301+1G>C']['submitted_variant'] == '17-41256884-C-G'
        assert results['NM_007298.3:c.301+1G>C']['genome_context_intronic_sequence'] == 'NC_000017.10(NM_007298.3):c.301+1G>C'
        assert results['NM_007298.3:c.301+1G>C']['hgvs_lrg_variant'] == ''
        assert results['NM_007298.3:c.301+1G>C']['hgvs_transcript_variant'] == 'NM_007298.3:c.301+1G>C'
        assert results['NM_007298.3:c.301+1G>C']['hgvs_refseqgene_variant'] == ''
        assert results['NM_007298.3:c.301+1G>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.41256884C>G', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '41256884', 'alt': u'G'}}
        assert results['NM_007298.3:c.301+1G>C']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.43104867C>G', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '43104867', 'alt': u'G'}}
        assert results['NM_007298.3:c.301+1G>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.41256884C>G', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '41256884', 'alt': u'G'}}
        assert results['NM_007298.3:c.301+1G>C']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.43104867C>G', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '43104867', 'alt': u'G'}}

        assert 'NM_007294.3:c.301+1G>C' in results.keys()
        assert results['NM_007294.3:c.301+1G>C']['hgvs_lrg_transcript_variant'] == 'LRG_292t1:c.301+1G>C'
        assert results['NM_007294.3:c.301+1G>C']['refseqgene_context_intronic_sequence'] == 'NG_005905.2(NM_007294.3):c.301+1G>C'
        assert results['NM_007294.3:c.301+1G>C']['alt_genomic_loci'] == []
        assert results['NM_007294.3:c.301+1G>C']['gene_symbol'] == 'BRCA1'
        assert results['NM_007294.3:c.301+1G>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_009225.1(LRG_292p1):p.?', 'slr': 'NP_009225.1:p.?'}
        assert results['NM_007294.3:c.301+1G>C']['submitted_variant'] == '17-41256884-C-G'
        assert results['NM_007294.3:c.301+1G>C']['genome_context_intronic_sequence'] == 'NC_000017.10(NM_007294.3):c.301+1G>C'
        assert results['NM_007294.3:c.301+1G>C']['hgvs_lrg_variant'] == 'LRG_292:g.113117G>C'
        assert results['NM_007294.3:c.301+1G>C']['hgvs_transcript_variant'] == 'NM_007294.3:c.301+1G>C'
        assert results['NM_007294.3:c.301+1G>C']['hgvs_refseqgene_variant'] == 'NG_005905.2:g.113117G>C'
        assert results['NM_007294.3:c.301+1G>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.41256884C>G', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '41256884', 'alt': u'G'}}
        assert results['NM_007294.3:c.301+1G>C']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.43104867C>G', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '43104867', 'alt': u'G'}}
        assert results['NM_007294.3:c.301+1G>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.41256884C>G', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '41256884', 'alt': u'G'}}
        assert results['NM_007294.3:c.301+1G>C']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.43104867C>G', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '43104867', 'alt': u'G'}}


    def test_variant256(self):
        variant = '17-42991428-C-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001131019.2:c.490G>T' in results.keys()
        assert results['NM_001131019.2:c.490G>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001131019.2:c.490G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001131019.2:c.490G>T']['alt_genomic_loci'] == []
        assert results['NM_001131019.2:c.490G>T']['gene_symbol'] == 'GFAP'
        assert results['NM_001131019.2:c.490G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001124491.1:p.(Glu164Ter)', 'slr': 'NP_001124491.1:p.(E164*)'}
        assert results['NM_001131019.2:c.490G>T']['submitted_variant'] == '17-42991428-C-A'
        assert results['NM_001131019.2:c.490G>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001131019.2:c.490G>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001131019.2:c.490G>T']['hgvs_transcript_variant'] == 'NM_001131019.2:c.490G>T'
        assert results['NM_001131019.2:c.490G>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001131019.2:c.490G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.42991428C>A', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '42991428', 'alt': u'A'}}
        assert results['NM_001131019.2:c.490G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.44914060C>A', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '44914060', 'alt': u'A'}}
        assert results['NM_001131019.2:c.490G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.42991428C>A', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '42991428', 'alt': u'A'}}
        assert results['NM_001131019.2:c.490G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.44914060C>A', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '44914060', 'alt': u'A'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001242376.1:c.490G>T' in results.keys()
        assert results['NM_001242376.1:c.490G>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001242376.1:c.490G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001242376.1:c.490G>T']['alt_genomic_loci'] == []
        assert results['NM_001242376.1:c.490G>T']['gene_symbol'] == 'GFAP'
        assert results['NM_001242376.1:c.490G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001229305.1:p.(Glu164Ter)', 'slr': 'NP_001229305.1:p.(E164*)'}
        assert results['NM_001242376.1:c.490G>T']['submitted_variant'] == '17-42991428-C-A'
        assert results['NM_001242376.1:c.490G>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001242376.1:c.490G>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001242376.1:c.490G>T']['hgvs_transcript_variant'] == 'NM_001242376.1:c.490G>T'
        assert results['NM_001242376.1:c.490G>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001242376.1:c.490G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.42991428C>A', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '42991428', 'alt': u'A'}}
        assert results['NM_001242376.1:c.490G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.44914060C>A', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '44914060', 'alt': u'A'}}
        assert results['NM_001242376.1:c.490G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.42991428C>A', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '42991428', 'alt': u'A'}}
        assert results['NM_001242376.1:c.490G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.44914060C>A', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '44914060', 'alt': u'A'}}

        assert 'NM_001363846.1:c.490G>T' in results.keys()
        assert results['NM_001363846.1:c.490G>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001363846.1:c.490G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001363846.1:c.490G>T']['alt_genomic_loci'] == []
        assert results['NM_001363846.1:c.490G>T']['gene_symbol'] == 'GFAP'
        assert results['NM_001363846.1:c.490G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001350775.1:p.(Glu164Ter)', 'slr': 'NP_001350775.1:p.(E164*)'}
        assert results['NM_001363846.1:c.490G>T']['submitted_variant'] == '17-42991428-C-A'
        assert results['NM_001363846.1:c.490G>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001363846.1:c.490G>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001363846.1:c.490G>T']['hgvs_transcript_variant'] == 'NM_001363846.1:c.490G>T'
        assert results['NM_001363846.1:c.490G>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001363846.1:c.490G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.42991428C>A', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '42991428', 'alt': u'A'}}
        assert 'hg38' not in results['NM_001363846.1:c.490G>T']['primary_assembly_loci'].keys()
        assert results['NM_001363846.1:c.490G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.42991428C>A', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '42991428', 'alt': u'A'}}
        assert 'grch38' not in results['NM_001363846.1:c.490G>T']['primary_assembly_loci'].keys()

        assert 'NM_002055.4:c.490G>T' in results.keys()
        assert results['NM_002055.4:c.490G>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_002055.4:c.490G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_002055.4:c.490G>T']['alt_genomic_loci'] == []
        assert results['NM_002055.4:c.490G>T']['gene_symbol'] == 'GFAP'
        assert results['NM_002055.4:c.490G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_002046.1:p.(Glu164Ter)', 'slr': 'NP_002046.1:p.(E164*)'}
        assert results['NM_002055.4:c.490G>T']['submitted_variant'] == '17-42991428-C-A'
        assert results['NM_002055.4:c.490G>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_002055.4:c.490G>T']['hgvs_lrg_variant'] == ''
        assert results['NM_002055.4:c.490G>T']['hgvs_transcript_variant'] == 'NM_002055.4:c.490G>T'
        assert results['NM_002055.4:c.490G>T']['hgvs_refseqgene_variant'] == 'NG_008401.1:g.6487G>T'
        assert results['NM_002055.4:c.490G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.42991428C>A', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '42991428', 'alt': u'A'}}
        assert results['NM_002055.4:c.490G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.44914060C>A', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '44914060', 'alt': u'A'}}
        assert results['NM_002055.4:c.490G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.42991428C>A', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '42991428', 'alt': u'A'}}
        assert results['NM_002055.4:c.490G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.44914060C>A', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '44914060', 'alt': u'A'}}


    def test_variant257(self):
        variant = '17-48252809-A-T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NR_135553.1:n.1022A>T' in results.keys()
        assert results['NR_135553.1:n.1022A>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_135553.1:n.1022A>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_135553.1:n.1022A>T']['alt_genomic_loci'] == []
        assert results['NR_135553.1:n.1022A>T']['gene_symbol'] == 'SGCA'
        assert results['NR_135553.1:n.1022A>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_135553.1:n.1022A>T']['submitted_variant'] == '17-48252809-A-T'
        assert results['NR_135553.1:n.1022A>T']['genome_context_intronic_sequence'] == ''
        assert results['NR_135553.1:n.1022A>T']['hgvs_lrg_variant'] == ''
        assert results['NR_135553.1:n.1022A>T']['hgvs_transcript_variant'] == 'NR_135553.1:n.1022A>T'
        assert results['NR_135553.1:n.1022A>T']['hgvs_refseqgene_variant'] == ''
        assert results['NR_135553.1:n.1022A>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48252809A>T', 'vcf': {'chr': 'chr17', 'ref': 'A', 'pos': '48252809', 'alt': 'T'}}
        assert results['NR_135553.1:n.1022A>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50175448A>T', 'vcf': {'chr': 'chr17', 'ref': 'A', 'pos': '50175448', 'alt': 'T'}}
        assert results['NR_135553.1:n.1022A>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48252809A>T', 'vcf': {'chr': '17', 'ref': 'A', 'pos': '48252809', 'alt': 'T'}}
        assert results['NR_135553.1:n.1022A>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50175448A>T', 'vcf': {'chr': '17', 'ref': 'A', 'pos': '50175448', 'alt': 'T'}}

        assert 'NM_001135697.1:c.*11A>T' in results.keys()
        assert results['NM_001135697.1:c.*11A>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001135697.1:c.*11A>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001135697.1:c.*11A>T']['alt_genomic_loci'] == []
        assert results['NM_001135697.1:c.*11A>T']['gene_symbol'] == 'SGCA'
        assert results['NM_001135697.1:c.*11A>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001129169.1:p.?', 'slr': 'NP_001129169.1:p.?'}
        assert results['NM_001135697.1:c.*11A>T']['submitted_variant'] == '17-48252809-A-T'
        assert results['NM_001135697.1:c.*11A>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001135697.1:c.*11A>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001135697.1:c.*11A>T']['hgvs_transcript_variant'] == 'NM_001135697.1:c.*11A>T'
        assert results['NM_001135697.1:c.*11A>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001135697.1:c.*11A>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48252809A>T', 'vcf': {'chr': 'chr17', 'ref': 'A', 'pos': '48252809', 'alt': 'T'}}
        assert 'hg38' not in results['NM_001135697.1:c.*11A>T']['primary_assembly_loci'].keys()
        assert results['NM_001135697.1:c.*11A>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48252809A>T', 'vcf': {'chr': '17', 'ref': 'A', 'pos': '48252809', 'alt': 'T'}}
        assert 'grch38' not in results['NM_001135697.1:c.*11A>T']['primary_assembly_loci'].keys()

        assert results['flag'] == 'gene_variant'
        assert 'NM_000023.2:c.*11A>T' in results.keys()
        assert results['NM_000023.2:c.*11A>T']['hgvs_lrg_transcript_variant'] == 'LRG_203t1:c.*11A>T'
        assert results['NM_000023.2:c.*11A>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000023.2:c.*11A>T']['alt_genomic_loci'] == []
        assert results['NM_000023.2:c.*11A>T']['gene_symbol'] == 'SGCA'
        assert results['NM_000023.2:c.*11A>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000014.1(LRG_203p1):p.?', 'slr': 'NP_000014.1:p.?'}
        assert results['NM_000023.2:c.*11A>T']['submitted_variant'] == '17-48252809-A-T'
        assert results['NM_000023.2:c.*11A>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_000023.2:c.*11A>T']['hgvs_lrg_variant'] == 'LRG_203:g.14444A>T'
        assert results['NM_000023.2:c.*11A>T']['hgvs_transcript_variant'] == 'NM_000023.2:c.*11A>T'
        assert results['NM_000023.2:c.*11A>T']['hgvs_refseqgene_variant'] == 'NG_008889.1:g.14444A>T'
        assert results['NM_000023.2:c.*11A>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48252809A>T', 'vcf': {'chr': 'chr17', 'ref': 'A', 'pos': '48252809', 'alt': 'T'}}
        assert 'hg38' not in results['NM_000023.2:c.*11A>T']['primary_assembly_loci'].keys()
        assert results['NM_000023.2:c.*11A>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48252809A>T', 'vcf': {'chr': '17', 'ref': 'A', 'pos': '48252809', 'alt': 'T'}}
        assert 'grch38' not in results['NM_000023.2:c.*11A>T']['primary_assembly_loci'].keys()

        assert 'NM_001135697.2:c.*11A>T' in results.keys()
        assert results['NM_001135697.2:c.*11A>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001135697.2:c.*11A>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001135697.2:c.*11A>T']['alt_genomic_loci'] == []
        assert results['NM_001135697.2:c.*11A>T']['gene_symbol'] == 'SGCA'
        assert results['NM_001135697.2:c.*11A>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001129169.1:p.?', 'slr': 'NP_001129169.1:p.?'}
        assert results['NM_001135697.2:c.*11A>T']['submitted_variant'] == '17-48252809-A-T'
        assert results['NM_001135697.2:c.*11A>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001135697.2:c.*11A>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001135697.2:c.*11A>T']['hgvs_transcript_variant'] == 'NM_001135697.2:c.*11A>T'
        assert results['NM_001135697.2:c.*11A>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001135697.2:c.*11A>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48252809A>T', 'vcf': {'chr': 'chr17', 'ref': 'A', 'pos': '48252809', 'alt': 'T'}}
        assert results['NM_001135697.2:c.*11A>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50175448A>T', 'vcf': {'chr': 'chr17', 'ref': 'A', 'pos': '50175448', 'alt': 'T'}}
        assert results['NM_001135697.2:c.*11A>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48252809A>T', 'vcf': {'chr': '17', 'ref': 'A', 'pos': '48252809', 'alt': 'T'}}
        assert results['NM_001135697.2:c.*11A>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50175448A>T', 'vcf': {'chr': '17', 'ref': 'A', 'pos': '50175448', 'alt': 'T'}}

        assert 'NM_000023.3:c.*11A>T' in results.keys()
        assert results['NM_000023.3:c.*11A>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000023.3:c.*11A>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000023.3:c.*11A>T']['alt_genomic_loci'] == []
        assert results['NM_000023.3:c.*11A>T']['gene_symbol'] == 'SGCA'
        assert results['NM_000023.3:c.*11A>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000014.1(LRG_203p1):p.?', 'slr': 'NP_000014.1:p.?'}
        assert results['NM_000023.3:c.*11A>T']['submitted_variant'] == '17-48252809-A-T'
        assert results['NM_000023.3:c.*11A>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_000023.3:c.*11A>T']['hgvs_lrg_variant'] == ''
        assert results['NM_000023.3:c.*11A>T']['hgvs_transcript_variant'] == 'NM_000023.3:c.*11A>T'
        assert results['NM_000023.3:c.*11A>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_000023.3:c.*11A>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48252809A>T', 'vcf': {'chr': 'chr17', 'ref': 'A', 'pos': '48252809', 'alt': 'T'}}
        assert results['NM_000023.3:c.*11A>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50175448A>T', 'vcf': {'chr': 'chr17', 'ref': 'A', 'pos': '50175448', 'alt': 'T'}}
        assert results['NM_000023.3:c.*11A>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48252809A>T', 'vcf': {'chr': '17', 'ref': 'A', 'pos': '48252809', 'alt': 'T'}}
        assert results['NM_000023.3:c.*11A>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50175448A>T', 'vcf': {'chr': '17', 'ref': 'A', 'pos': '50175448', 'alt': 'T'}}


    def test_variant258(self):
        variant = '17-62022709-G-GTC'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000334.4:c.3720+9_3720+10dup' in results.keys()
        assert results['NM_000334.4:c.3720+9_3720+10dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000334.4:c.3720+9_3720+10dup']['refseqgene_context_intronic_sequence'] == 'NG_011699.1(NM_000334.4):c.3720+9_3720+10dup'
        assert results['NM_000334.4:c.3720+9_3720+10dup']['alt_genomic_loci'] == []
        assert results['NM_000334.4:c.3720+9_3720+10dup']['gene_symbol'] == 'SCN4A'
        assert results['NM_000334.4:c.3720+9_3720+10dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000325.4:p.?', 'slr': 'NP_000325.4:p.?'}
        assert results['NM_000334.4:c.3720+9_3720+10dup']['submitted_variant'] == '17-62022709-G-GTC'
        assert results['NM_000334.4:c.3720+9_3720+10dup']['genome_context_intronic_sequence'] == 'NC_000017.10(NM_000334.4):c.3720+9_3720+10dup'
        assert results['NM_000334.4:c.3720+9_3720+10dup']['hgvs_lrg_variant'] == ''
        assert results['NM_000334.4:c.3720+9_3720+10dup']['hgvs_transcript_variant'] == 'NM_000334.4:c.3720+9_3720+10dup'
        assert results['NM_000334.4:c.3720+9_3720+10dup']['hgvs_refseqgene_variant'] == 'NG_011699.1:g.32568_32569dup'
        assert results['NM_000334.4:c.3720+9_3720+10dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.62022710_62022711dup', 'vcf': {'chr': 'chr17', 'ref': 'TC', 'pos': '62022710', 'alt': 'TCTC'}}
        assert results['NM_000334.4:c.3720+9_3720+10dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.63945350_63945351dup', 'vcf': {'chr': 'chr17', 'ref': 'TC', 'pos': '63945350', 'alt': 'TCTC'}}
        assert results['NM_000334.4:c.3720+9_3720+10dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.62022710_62022711dup', 'vcf': {'chr': '17', 'ref': 'TC', 'pos': '62022710', 'alt': 'TCTC'}}
        assert results['NM_000334.4:c.3720+9_3720+10dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.63945350_63945351dup', 'vcf': {'chr': '17', 'ref': 'TC', 'pos': '63945350', 'alt': 'TCTC'}}


    def test_variant259(self):
        variant = '17-62022711-C-CT'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000334.4:c.3720+8_3720+9insA' in results.keys()
        assert results['NM_000334.4:c.3720+8_3720+9insA']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000334.4:c.3720+8_3720+9insA']['refseqgene_context_intronic_sequence'] == 'NG_011699.1(NM_000334.4):c.3720+8_3720+9insA'
        assert results['NM_000334.4:c.3720+8_3720+9insA']['alt_genomic_loci'] == []
        assert results['NM_000334.4:c.3720+8_3720+9insA']['gene_symbol'] == 'SCN4A'
        assert results['NM_000334.4:c.3720+8_3720+9insA']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000325.4:p.?', 'slr': 'NP_000325.4:p.?'}
        assert results['NM_000334.4:c.3720+8_3720+9insA']['submitted_variant'] == '17-62022711-C-CT'
        assert results['NM_000334.4:c.3720+8_3720+9insA']['genome_context_intronic_sequence'] == 'NC_000017.10(NM_000334.4):c.3720+8_3720+9insA'
        assert results['NM_000334.4:c.3720+8_3720+9insA']['hgvs_lrg_variant'] == ''
        assert results['NM_000334.4:c.3720+8_3720+9insA']['hgvs_transcript_variant'] == 'NM_000334.4:c.3720+8_3720+9insA'
        assert results['NM_000334.4:c.3720+8_3720+9insA']['hgvs_refseqgene_variant'] == 'NG_011699.1:g.32567_32568insA'
        assert results['NM_000334.4:c.3720+8_3720+9insA']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.62022711_62022712insT', 'vcf': {'chr': 'chr17', 'ref': 'C', 'pos': '62022711', 'alt': u'CT'}}
        assert results['NM_000334.4:c.3720+8_3720+9insA']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.63945351_63945352insT', 'vcf': {'chr': 'chr17', 'ref': 'C', 'pos': '63945351', 'alt': u'CT'}}
        assert results['NM_000334.4:c.3720+8_3720+9insA']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.62022711_62022712insT', 'vcf': {'chr': '17', 'ref': 'C', 'pos': '62022711', 'alt': u'CT'}}
        assert results['NM_000334.4:c.3720+8_3720+9insA']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.63945351_63945352insT', 'vcf': {'chr': '17', 'ref': 'C', 'pos': '63945351', 'alt': u'CT'}}


    def test_variant260(self):
        variant = '17-62023005-G-GGC'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000334.4:c.3442-8_3442-7insGC' in results.keys()
        assert results['NM_000334.4:c.3442-8_3442-7insGC']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000334.4:c.3442-8_3442-7insGC']['refseqgene_context_intronic_sequence'] == 'NG_011699.1(NM_000334.4):c.3442-8_3442-7insGC'
        assert results['NM_000334.4:c.3442-8_3442-7insGC']['alt_genomic_loci'] == []
        assert results['NM_000334.4:c.3442-8_3442-7insGC']['gene_symbol'] == 'SCN4A'
        assert results['NM_000334.4:c.3442-8_3442-7insGC']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000325.4:p.?', 'slr': 'NP_000325.4:p.?'}
        assert results['NM_000334.4:c.3442-8_3442-7insGC']['submitted_variant'] == '17-62023005-G-GGC'
        assert results['NM_000334.4:c.3442-8_3442-7insGC']['genome_context_intronic_sequence'] == 'NC_000017.10(NM_000334.4):c.3442-8_3442-7insGC'
        assert results['NM_000334.4:c.3442-8_3442-7insGC']['hgvs_lrg_variant'] == ''
        assert results['NM_000334.4:c.3442-8_3442-7insGC']['hgvs_transcript_variant'] == 'NM_000334.4:c.3442-8_3442-7insGC'
        assert results['NM_000334.4:c.3442-8_3442-7insGC']['hgvs_refseqgene_variant'] == 'NG_011699.1:g.32273_32274insGC'
        assert results['NM_000334.4:c.3442-8_3442-7insGC']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.62023005_62023006insGC', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '62023005', 'alt': u'GGC'}}
        assert results['NM_000334.4:c.3442-8_3442-7insGC']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.63945645_63945646insGC', 'vcf': {'chr': 'chr17', 'ref': 'G', 'pos': '63945645', 'alt': u'GGC'}}
        assert results['NM_000334.4:c.3442-8_3442-7insGC']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.62023005_62023006insGC', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '62023005', 'alt': u'GGC'}}
        assert results['NM_000334.4:c.3442-8_3442-7insGC']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.63945645_63945646insGC', 'vcf': {'chr': '17', 'ref': 'G', 'pos': '63945645', 'alt': u'GGC'}}


    def test_variant261(self):
        variant = '17-62023006-C-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000334.4:c.3442-8G>T' in results.keys()
        assert results['NM_000334.4:c.3442-8G>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000334.4:c.3442-8G>T']['refseqgene_context_intronic_sequence'] == 'NG_011699.1(NM_000334.4):c.3442-8G>T'
        assert results['NM_000334.4:c.3442-8G>T']['alt_genomic_loci'] == []
        assert results['NM_000334.4:c.3442-8G>T']['gene_symbol'] == 'SCN4A'
        assert results['NM_000334.4:c.3442-8G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000325.4:p.?', 'slr': 'NP_000325.4:p.?'}
        assert results['NM_000334.4:c.3442-8G>T']['submitted_variant'] == '17-62023006-C-A'
        assert results['NM_000334.4:c.3442-8G>T']['genome_context_intronic_sequence'] == 'NC_000017.10(NM_000334.4):c.3442-8G>T'
        assert results['NM_000334.4:c.3442-8G>T']['hgvs_lrg_variant'] == ''
        assert results['NM_000334.4:c.3442-8G>T']['hgvs_transcript_variant'] == 'NM_000334.4:c.3442-8G>T'
        assert results['NM_000334.4:c.3442-8G>T']['hgvs_refseqgene_variant'] == 'NG_011699.1:g.32273G>T'
        assert results['NM_000334.4:c.3442-8G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.62023006C>A', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '62023006', 'alt': u'A'}}
        assert results['NM_000334.4:c.3442-8G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.63945646C>A', 'vcf': {'chr': 'chr17', 'ref': u'C', 'pos': '63945646', 'alt': u'A'}}
        assert results['NM_000334.4:c.3442-8G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.62023006C>A', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '62023006', 'alt': u'A'}}
        assert results['NM_000334.4:c.3442-8G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.63945646C>A', 'vcf': {'chr': '17', 'ref': u'C', 'pos': '63945646', 'alt': u'A'}}


    def test_variant262(self):
        variant = '17-62034787-G-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000334.4:c.2111C>T' in results.keys()
        assert results['NM_000334.4:c.2111C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000334.4:c.2111C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000334.4:c.2111C>T']['alt_genomic_loci'] == []
        assert results['NM_000334.4:c.2111C>T']['gene_symbol'] == 'SCN4A'
        assert results['NM_000334.4:c.2111C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000325.4:p.(Thr704Met)', 'slr': 'NP_000325.4:p.(T704M)'}
        assert results['NM_000334.4:c.2111C>T']['submitted_variant'] == '17-62034787-G-A'
        assert results['NM_000334.4:c.2111C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_000334.4:c.2111C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_000334.4:c.2111C>T']['hgvs_transcript_variant'] == 'NM_000334.4:c.2111C>T'
        assert results['NM_000334.4:c.2111C>T']['hgvs_refseqgene_variant'] == 'NG_011699.1:g.20492C>T'
        assert results['NM_000334.4:c.2111C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.62034787G>A', 'vcf': {'chr': 'chr17', 'ref': u'G', 'pos': '62034787', 'alt': u'A'}}
        assert results['NM_000334.4:c.2111C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.63957427G>A', 'vcf': {'chr': 'chr17', 'ref': u'G', 'pos': '63957427', 'alt': u'A'}}
        assert results['NM_000334.4:c.2111C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.62034787G>A', 'vcf': {'chr': '17', 'ref': u'G', 'pos': '62034787', 'alt': u'A'}}
        assert results['NM_000334.4:c.2111C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.63957427G>A', 'vcf': {'chr': '17', 'ref': u'G', 'pos': '63957427', 'alt': u'A'}}


    def test_variant263(self):
        variant = '18-24128261-GTCCTCC-G'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001351443.1:c.-16+941_-16+946del' in results.keys()
        assert results['NM_001351443.1:c.-16+941_-16+946del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001351443.1:c.-16+941_-16+946del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001351443.1:c.-16+941_-16+946del']['alt_genomic_loci'] == []
        assert results['NM_001351443.1:c.-16+941_-16+946del']['gene_symbol'] == 'KCTD1'
        assert results['NM_001351443.1:c.-16+941_-16+946del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001338372.1:p.?', 'slr': 'NP_001338372.1:p.?'}
        assert results['NM_001351443.1:c.-16+941_-16+946del']['submitted_variant'] == '18-24128261-GTCCTCC-G'
        assert results['NM_001351443.1:c.-16+941_-16+946del']['genome_context_intronic_sequence'] == 'NC_000018.9(NM_001351443.1):c.-16+941_-16+946del'
        assert results['NM_001351443.1:c.-16+941_-16+946del']['hgvs_lrg_variant'] == ''
        assert results['NM_001351443.1:c.-16+941_-16+946del']['hgvs_transcript_variant'] == 'NM_001351443.1:c.-16+941_-16+946del'
        assert results['NM_001351443.1:c.-16+941_-16+946del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001351443.1:c.-16+941_-16+946del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000018.9:g.24128262_24128267del', 'vcf': {'chr': 'chr18', 'ref': 'GTCCTCC', 'pos': '24128261', 'alt': 'G'}}
        assert results['NM_001351443.1:c.-16+941_-16+946del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000018.10:g.26548298_26548303del', 'vcf': {'chr': 'chr18', 'ref': 'GTCCTCC', 'pos': '26548297', 'alt': 'G'}}
        assert results['NM_001351443.1:c.-16+941_-16+946del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000018.9:g.24128262_24128267del', 'vcf': {'chr': '18', 'ref': 'GTCCTCC', 'pos': '24128261', 'alt': 'G'}}
        assert results['NM_001351443.1:c.-16+941_-16+946del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000018.10:g.26548298_26548303del', 'vcf': {'chr': '18', 'ref': 'GTCCTCC', 'pos': '26548297', 'alt': 'G'}}

        assert 'NM_001258221.1:c.-16+1426_-16+1431del' in results.keys()
        assert results['NM_001258221.1:c.-16+1426_-16+1431del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001258221.1:c.-16+1426_-16+1431del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001258221.1:c.-16+1426_-16+1431del']['alt_genomic_loci'] == []
        assert results['NM_001258221.1:c.-16+1426_-16+1431del']['gene_symbol'] == 'KCTD1'
        assert results['NM_001258221.1:c.-16+1426_-16+1431del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001245150.1:p.?', 'slr': 'NP_001245150.1:p.?'}
        assert results['NM_001258221.1:c.-16+1426_-16+1431del']['submitted_variant'] == '18-24128261-GTCCTCC-G'
        assert results['NM_001258221.1:c.-16+1426_-16+1431del']['genome_context_intronic_sequence'] == 'NC_000018.9(NM_001258221.1):c.-16+1426_-16+1431del'
        assert results['NM_001258221.1:c.-16+1426_-16+1431del']['hgvs_lrg_variant'] == ''
        assert results['NM_001258221.1:c.-16+1426_-16+1431del']['hgvs_transcript_variant'] == 'NM_001258221.1:c.-16+1426_-16+1431del'
        assert results['NM_001258221.1:c.-16+1426_-16+1431del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001258221.1:c.-16+1426_-16+1431del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000018.9:g.24128262_24128267del', 'vcf': {'chr': 'chr18', 'ref': 'GTCCTCC', 'pos': '24128261', 'alt': 'G'}}
        assert results['NM_001258221.1:c.-16+1426_-16+1431del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000018.10:g.26548298_26548303del', 'vcf': {'chr': 'chr18', 'ref': 'GTCCTCC', 'pos': '26548297', 'alt': 'G'}}
        assert results['NM_001258221.1:c.-16+1426_-16+1431del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000018.9:g.24128262_24128267del', 'vcf': {'chr': '18', 'ref': 'GTCCTCC', 'pos': '24128261', 'alt': 'G'}}
        assert results['NM_001258221.1:c.-16+1426_-16+1431del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000018.10:g.26548298_26548303del', 'vcf': {'chr': '18', 'ref': 'GTCCTCC', 'pos': '26548297', 'alt': 'G'}}

        assert 'NM_001258222.2:c.10-47053_10-47048del' in results.keys()
        assert results['NM_001258222.2:c.10-47053_10-47048del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001258222.2:c.10-47053_10-47048del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001258222.2:c.10-47053_10-47048del']['alt_genomic_loci'] == []
        assert results['NM_001258222.2:c.10-47053_10-47048del']['gene_symbol'] == 'KCTD1'
        assert results['NM_001258222.2:c.10-47053_10-47048del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001245151.1:p.?', 'slr': 'NP_001245151.1:p.?'}
        assert results['NM_001258222.2:c.10-47053_10-47048del']['submitted_variant'] == '18-24128261-GTCCTCC-G'
        assert results['NM_001258222.2:c.10-47053_10-47048del']['genome_context_intronic_sequence'] == 'NC_000018.9(NM_001258222.2):c.10-47053_10-47048del'
        assert results['NM_001258222.2:c.10-47053_10-47048del']['hgvs_lrg_variant'] == ''
        assert results['NM_001258222.2:c.10-47053_10-47048del']['hgvs_transcript_variant'] == 'NM_001258222.2:c.10-47053_10-47048del'
        assert results['NM_001258222.2:c.10-47053_10-47048del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001258222.2:c.10-47053_10-47048del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000018.9:g.24128262_24128267del', 'vcf': {'chr': 'chr18', 'ref': 'GTCCTCC', 'pos': '24128261', 'alt': 'G'}}
        assert results['NM_001258222.2:c.10-47053_10-47048del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000018.10:g.26548298_26548303del', 'vcf': {'chr': 'chr18', 'ref': 'GTCCTCC', 'pos': '26548297', 'alt': 'G'}}
        assert results['NM_001258222.2:c.10-47053_10-47048del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000018.9:g.24128262_24128267del', 'vcf': {'chr': '18', 'ref': 'GTCCTCC', 'pos': '24128261', 'alt': 'G'}}
        assert results['NM_001258222.2:c.10-47053_10-47048del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000018.10:g.26548298_26548303del', 'vcf': {'chr': '18', 'ref': 'GTCCTCC', 'pos': '26548297', 'alt': 'G'}}

        assert 'NM_001142730.2:c.234_239del' in results.keys()
        assert results['NM_001142730.2:c.234_239del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001142730.2:c.234_239del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001142730.2:c.234_239del']['alt_genomic_loci'] == []
        assert results['NM_001142730.2:c.234_239del']['gene_symbol'] == 'KCTD1'
        assert results['NM_001142730.2:c.234_239del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001136202.1:p.(Glu78_Glu79del)', 'slr': 'NP_001136202.1:p.(E78_E79del)'}
        assert results['NM_001142730.2:c.234_239del']['submitted_variant'] == '18-24128261-GTCCTCC-G'
        assert results['NM_001142730.2:c.234_239del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001142730.2:c.234_239del']['hgvs_lrg_variant'] == ''
        assert results['NM_001142730.2:c.234_239del']['hgvs_transcript_variant'] == 'NM_001142730.2:c.234_239del'
        assert results['NM_001142730.2:c.234_239del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001142730.2:c.234_239del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000018.9:g.24128262_24128267del', 'vcf': {'chr': 'chr18', 'ref': 'GTCCTCC', 'pos': '24128261', 'alt': 'G'}}
        assert results['NM_001142730.2:c.234_239del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000018.10:g.26548298_26548303del', 'vcf': {'chr': 'chr18', 'ref': 'GTCCTCC', 'pos': '26548297', 'alt': 'G'}}
        assert results['NM_001142730.2:c.234_239del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000018.9:g.24128262_24128267del', 'vcf': {'chr': '18', 'ref': 'GTCCTCC', 'pos': '24128261', 'alt': 'G'}}
        assert results['NM_001142730.2:c.234_239del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000018.10:g.26548298_26548303del', 'vcf': {'chr': '18', 'ref': 'GTCCTCC', 'pos': '26548297', 'alt': 'G'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001136205.2:c.-16+588_-16+593del' in results.keys()
        assert results['NM_001136205.2:c.-16+588_-16+593del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001136205.2:c.-16+588_-16+593del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001136205.2:c.-16+588_-16+593del']['alt_genomic_loci'] == []
        assert results['NM_001136205.2:c.-16+588_-16+593del']['gene_symbol'] == 'KCTD1'
        assert results['NM_001136205.2:c.-16+588_-16+593del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001129677.1:p.?', 'slr': 'NP_001129677.1:p.?'}
        assert results['NM_001136205.2:c.-16+588_-16+593del']['submitted_variant'] == '18-24128261-GTCCTCC-G'
        assert results['NM_001136205.2:c.-16+588_-16+593del']['genome_context_intronic_sequence'] == 'NC_000018.9(NM_001136205.2):c.-16+588_-16+593del'
        assert results['NM_001136205.2:c.-16+588_-16+593del']['hgvs_lrg_variant'] == ''
        assert results['NM_001136205.2:c.-16+588_-16+593del']['hgvs_transcript_variant'] == 'NM_001136205.2:c.-16+588_-16+593del'
        assert results['NM_001136205.2:c.-16+588_-16+593del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001136205.2:c.-16+588_-16+593del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000018.9:g.24128262_24128267del', 'vcf': {'chr': 'chr18', 'ref': 'GTCCTCC', 'pos': '24128261', 'alt': 'G'}}
        assert results['NM_001136205.2:c.-16+588_-16+593del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000018.10:g.26548298_26548303del', 'vcf': {'chr': 'chr18', 'ref': 'GTCCTCC', 'pos': '26548297', 'alt': 'G'}}
        assert results['NM_001136205.2:c.-16+588_-16+593del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000018.9:g.24128262_24128267del', 'vcf': {'chr': '18', 'ref': 'GTCCTCC', 'pos': '24128261', 'alt': 'G'}}
        assert results['NM_001136205.2:c.-16+588_-16+593del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000018.10:g.26548298_26548303del', 'vcf': {'chr': '18', 'ref': 'GTCCTCC', 'pos': '26548297', 'alt': 'G'}}

        assert 'NM_001258222.1:c.10-47053_10-47048del' in results.keys()
        assert results['NM_001258222.1:c.10-47053_10-47048del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001258222.1:c.10-47053_10-47048del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001258222.1:c.10-47053_10-47048del']['alt_genomic_loci'] == []
        assert results['NM_001258222.1:c.10-47053_10-47048del']['gene_symbol'] == 'KCTD1'
        assert results['NM_001258222.1:c.10-47053_10-47048del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001245151.1:p.?', 'slr': 'NP_001245151.1:p.?'}
        assert results['NM_001258222.1:c.10-47053_10-47048del']['submitted_variant'] == '18-24128261-GTCCTCC-G'
        assert results['NM_001258222.1:c.10-47053_10-47048del']['genome_context_intronic_sequence'] == 'NC_000018.9(NM_001258222.1):c.10-47053_10-47048del'
        assert results['NM_001258222.1:c.10-47053_10-47048del']['hgvs_lrg_variant'] == ''
        assert results['NM_001258222.1:c.10-47053_10-47048del']['hgvs_transcript_variant'] == 'NM_001258222.1:c.10-47053_10-47048del'
        assert results['NM_001258222.1:c.10-47053_10-47048del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001258222.1:c.10-47053_10-47048del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000018.9:g.24128262_24128267del', 'vcf': {'chr': 'chr18', 'ref': 'GTCCTCC', 'pos': '24128261', 'alt': 'G'}}
        assert results['NM_001258222.1:c.10-47053_10-47048del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000018.10:g.26548298_26548303del', 'vcf': {'chr': 'chr18', 'ref': 'GTCCTCC', 'pos': '26548297', 'alt': 'G'}}
        assert results['NM_001258222.1:c.10-47053_10-47048del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000018.9:g.24128262_24128267del', 'vcf': {'chr': '18', 'ref': 'GTCCTCC', 'pos': '24128261', 'alt': 'G'}}
        assert results['NM_001258222.1:c.10-47053_10-47048del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000018.10:g.26548298_26548303del', 'vcf': {'chr': '18', 'ref': 'GTCCTCC', 'pos': '26548297', 'alt': 'G'}}

        assert 'NM_198991.3:c.-15-47053_-15-47048del' in results.keys()
        assert results['NM_198991.3:c.-15-47053_-15-47048del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_198991.3:c.-15-47053_-15-47048del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_198991.3:c.-15-47053_-15-47048del']['alt_genomic_loci'] == []
        assert results['NM_198991.3:c.-15-47053_-15-47048del']['gene_symbol'] == 'KCTD1'
        assert results['NM_198991.3:c.-15-47053_-15-47048del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_945342.1:p.?', 'slr': 'NP_945342.1:p.?'}
        assert results['NM_198991.3:c.-15-47053_-15-47048del']['submitted_variant'] == '18-24128261-GTCCTCC-G'
        assert results['NM_198991.3:c.-15-47053_-15-47048del']['genome_context_intronic_sequence'] == 'NC_000018.9(NM_198991.3):c.-15-47053_-15-47048del'
        assert results['NM_198991.3:c.-15-47053_-15-47048del']['hgvs_lrg_variant'] == ''
        assert results['NM_198991.3:c.-15-47053_-15-47048del']['hgvs_transcript_variant'] == 'NM_198991.3:c.-15-47053_-15-47048del'
        assert results['NM_198991.3:c.-15-47053_-15-47048del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_198991.3:c.-15-47053_-15-47048del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000018.9:g.24128262_24128267del', 'vcf': {'chr': 'chr18', 'ref': 'GTCCTCC', 'pos': '24128261', 'alt': 'G'}}
        assert results['NM_198991.3:c.-15-47053_-15-47048del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000018.10:g.26548298_26548303del', 'vcf': {'chr': 'chr18', 'ref': 'GTCCTCC', 'pos': '26548297', 'alt': 'G'}}
        assert results['NM_198991.3:c.-15-47053_-15-47048del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000018.9:g.24128262_24128267del', 'vcf': {'chr': '18', 'ref': 'GTCCTCC', 'pos': '24128261', 'alt': 'G'}}
        assert results['NM_198991.3:c.-15-47053_-15-47048del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000018.10:g.26548298_26548303del', 'vcf': {'chr': '18', 'ref': 'GTCCTCC', 'pos': '26548297', 'alt': 'G'}}


    def test_variant264(self):
        variant = '19-15291774-G-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000435.2:c.2992C>T' in results.keys()
        assert results['NM_000435.2:c.2992C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000435.2:c.2992C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000435.2:c.2992C>T']['alt_genomic_loci'] == []
        assert results['NM_000435.2:c.2992C>T']['gene_symbol'] == 'NOTCH3'
        assert results['NM_000435.2:c.2992C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000426.2:p.(Gln998Ter)', 'slr': 'NP_000426.2:p.(Q998*)'}
        assert results['NM_000435.2:c.2992C>T']['submitted_variant'] == '19-15291774-G-A'
        assert results['NM_000435.2:c.2992C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_000435.2:c.2992C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_000435.2:c.2992C>T']['hgvs_transcript_variant'] == 'NM_000435.2:c.2992C>T'
        assert results['NM_000435.2:c.2992C>T']['hgvs_refseqgene_variant'] == 'NG_009819.1:g.25019C>T'
        assert results['NM_000435.2:c.2992C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.15291774G>A', 'vcf': {'chr': 'chr19', 'ref': u'G', 'pos': '15291774', 'alt': u'A'}}
        assert results['NM_000435.2:c.2992C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.15180963G>A', 'vcf': {'chr': 'chr19', 'ref': u'G', 'pos': '15180963', 'alt': u'A'}}
        assert results['NM_000435.2:c.2992C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.15291774G>A', 'vcf': {'chr': '19', 'ref': u'G', 'pos': '15291774', 'alt': u'A'}}
        assert results['NM_000435.2:c.2992C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.15180963G>A', 'vcf': {'chr': '19', 'ref': u'G', 'pos': '15180963', 'alt': u'A'}}


    def test_variant265(self):
        variant = '19-15311794-A-G'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'intergenic'
        assert 'Intergenic_Variant_1' in results.keys()
        assert results['Intergenic_Variant_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['Intergenic_Variant_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['Intergenic_Variant_1']['alt_genomic_loci'] == []
        assert results['Intergenic_Variant_1']['gene_symbol'] == ''
        assert results['Intergenic_Variant_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['Intergenic_Variant_1']['submitted_variant'] == '19-15311794-A-G'
        assert results['Intergenic_Variant_1']['genome_context_intronic_sequence'] == ''
        assert results['Intergenic_Variant_1']['hgvs_lrg_variant'] == ''
        assert results['Intergenic_Variant_1']['hgvs_transcript_variant'] == ''
        assert results['Intergenic_Variant_1']['hgvs_refseqgene_variant'] == 'NG_009819.1:g.4999T>C'
        assert results['Intergenic_Variant_1']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.15311794A>G', 'vcf': {'chr': 'chr19', 'ref': 'A', 'pos': '15311794', 'alt': 'G'}}
        assert results['Intergenic_Variant_1']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.15200983A>G', 'vcf': {'chr': 'chr19', 'ref': 'A', 'pos': '15200983', 'alt': 'G'}}
        assert results['Intergenic_Variant_1']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.15311794A>G', 'vcf': {'chr': '19', 'ref': 'A', 'pos': '15311794', 'alt': 'G'}}
        assert results['Intergenic_Variant_1']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.15200983A>G', 'vcf': {'chr': '19', 'ref': 'A', 'pos': '15200983', 'alt': 'G'}}


    def test_variant266(self):
        variant = '19-39076592-G-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000540.2:c.14818G>A' in results.keys()
        assert results['NM_000540.2:c.14818G>A']['hgvs_lrg_transcript_variant'] == 'LRG_766t1:c.14818G>A'
        assert results['NM_000540.2:c.14818G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000540.2:c.14818G>A']['alt_genomic_loci'] == []
        assert results['NM_000540.2:c.14818G>A']['gene_symbol'] == 'RYR1'
        assert results['NM_000540.2:c.14818G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000531.2(LRG_766p1):p.(Ala4940Thr)', 'slr': 'NP_000531.2:p.(A4940T)'}
        assert results['NM_000540.2:c.14818G>A']['submitted_variant'] == '19-39076592-G-A'
        assert results['NM_000540.2:c.14818G>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_000540.2:c.14818G>A']['hgvs_lrg_variant'] == 'LRG_766:g.157253G>A'
        assert results['NM_000540.2:c.14818G>A']['hgvs_transcript_variant'] == 'NM_000540.2:c.14818G>A'
        assert results['NM_000540.2:c.14818G>A']['hgvs_refseqgene_variant'] == 'NG_008866.1:g.157253G>A'
        assert results['NM_000540.2:c.14818G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.39076592G>A', 'vcf': {'chr': 'chr19', 'ref': 'G', 'pos': '39076592', 'alt': 'A'}}
        assert results['NM_000540.2:c.14818G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.38585952G>A', 'vcf': {'chr': 'chr19', 'ref': 'G', 'pos': '38585952', 'alt': 'A'}}
        assert results['NM_000540.2:c.14818G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.39076592G>A', 'vcf': {'chr': '19', 'ref': 'G', 'pos': '39076592', 'alt': 'A'}}
        assert results['NM_000540.2:c.14818G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.38585952G>A', 'vcf': {'chr': '19', 'ref': 'G', 'pos': '38585952', 'alt': 'A'}}

        assert 'NM_001042723.1:c.14803G>A' in results.keys()
        assert results['NM_001042723.1:c.14803G>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001042723.1:c.14803G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001042723.1:c.14803G>A']['alt_genomic_loci'] == []
        assert results['NM_001042723.1:c.14803G>A']['gene_symbol'] == 'RYR1'
        assert results['NM_001042723.1:c.14803G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001036188.1:p.(Ala4935Thr)', 'slr': 'NP_001036188.1:p.(A4935T)'}
        assert results['NM_001042723.1:c.14803G>A']['submitted_variant'] == '19-39076592-G-A'
        assert results['NM_001042723.1:c.14803G>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_001042723.1:c.14803G>A']['hgvs_lrg_variant'] == ''
        assert results['NM_001042723.1:c.14803G>A']['hgvs_transcript_variant'] == 'NM_001042723.1:c.14803G>A'
        assert results['NM_001042723.1:c.14803G>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001042723.1:c.14803G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000019.9:g.39076592G>A', 'vcf': {'chr': 'chr19', 'ref': 'G', 'pos': '39076592', 'alt': 'A'}}
        assert results['NM_001042723.1:c.14803G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000019.10:g.38585952G>A', 'vcf': {'chr': 'chr19', 'ref': 'G', 'pos': '38585952', 'alt': 'A'}}
        assert results['NM_001042723.1:c.14803G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000019.9:g.39076592G>A', 'vcf': {'chr': '19', 'ref': 'G', 'pos': '39076592', 'alt': 'A'}}
        assert results['NM_001042723.1:c.14803G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000019.10:g.38585952G>A', 'vcf': {'chr': '19', 'ref': 'G', 'pos': '38585952', 'alt': 'A'}}


    def test_variant267(self):
        variant = '2-50149352-T-C'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001330086.1:c.4245A>G' in results.keys()
        assert results['NM_001330086.1:c.4245A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001330086.1:c.4245A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001330086.1:c.4245A>G']['alt_genomic_loci'] == []
        assert results['NM_001330086.1:c.4245A>G']['gene_symbol'] == 'NRXN1'
        assert results['NM_001330086.1:c.4245A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001317015.1:p.(Pro1415=)', 'slr': 'NP_001317015.1:p.(P1415=)'}
        assert results['NM_001330086.1:c.4245A>G']['submitted_variant'] == '2-50149352-T-C'
        assert results['NM_001330086.1:c.4245A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001330086.1:c.4245A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001330086.1:c.4245A>G']['hgvs_transcript_variant'] == 'NM_001330086.1:c.4245A>G'
        assert results['NM_001330086.1:c.4245A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001330086.1:c.4245A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001330086.1:c.4245A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}
        assert results['NM_001330086.1:c.4245A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001330086.1:c.4245A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}

        assert 'NM_001330083.1:c.4089A>G' in results.keys()
        assert results['NM_001330083.1:c.4089A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001330083.1:c.4089A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001330083.1:c.4089A>G']['alt_genomic_loci'] == []
        assert results['NM_001330083.1:c.4089A>G']['gene_symbol'] == 'NRXN1'
        assert results['NM_001330083.1:c.4089A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001317012.1:p.(Pro1363=)', 'slr': 'NP_001317012.1:p.(P1363=)'}
        assert results['NM_001330083.1:c.4089A>G']['submitted_variant'] == '2-50149352-T-C'
        assert results['NM_001330083.1:c.4089A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001330083.1:c.4089A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001330083.1:c.4089A>G']['hgvs_transcript_variant'] == 'NM_001330083.1:c.4089A>G'
        assert results['NM_001330083.1:c.4089A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001330083.1:c.4089A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001330083.1:c.4089A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}
        assert results['NM_001330083.1:c.4089A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001330083.1:c.4089A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}

        assert 'NM_001330095.1:c.4113A>G' in results.keys()
        assert results['NM_001330095.1:c.4113A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001330095.1:c.4113A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001330095.1:c.4113A>G']['alt_genomic_loci'] == []
        assert results['NM_001330095.1:c.4113A>G']['gene_symbol'] == 'NRXN1'
        assert results['NM_001330095.1:c.4113A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001317024.1:p.(Pro1371=)', 'slr': 'NP_001317024.1:p.(P1371=)'}
        assert results['NM_001330095.1:c.4113A>G']['submitted_variant'] == '2-50149352-T-C'
        assert results['NM_001330095.1:c.4113A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001330095.1:c.4113A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001330095.1:c.4113A>G']['hgvs_transcript_variant'] == 'NM_001330095.1:c.4113A>G'
        assert results['NM_001330095.1:c.4113A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001330095.1:c.4113A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001330095.1:c.4113A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}
        assert results['NM_001330095.1:c.4113A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001330095.1:c.4113A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}

        assert 'NM_138735.2:c.1059A>G' in results.keys()
        assert results['NM_138735.2:c.1059A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_138735.2:c.1059A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_138735.2:c.1059A>G']['alt_genomic_loci'] == []
        assert results['NM_138735.2:c.1059A>G']['gene_symbol'] == 'NRXN1'
        assert results['NM_138735.2:c.1059A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_620072.1:p.(Pro353=)', 'slr': 'NP_620072.1:p.(P353=)'}
        assert results['NM_138735.2:c.1059A>G']['submitted_variant'] == '2-50149352-T-C'
        assert results['NM_138735.2:c.1059A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_138735.2:c.1059A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_138735.2:c.1059A>G']['hgvs_transcript_variant'] == 'NM_138735.2:c.1059A>G'
        assert results['NM_138735.2:c.1059A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_138735.2:c.1059A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_138735.2:c.1059A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}
        assert results['NM_138735.2:c.1059A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_138735.2:c.1059A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}

        assert 'NM_001330078.1:c.4254A>G' in results.keys()
        assert results['NM_001330078.1:c.4254A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001330078.1:c.4254A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001330078.1:c.4254A>G']['alt_genomic_loci'] == []
        assert results['NM_001330078.1:c.4254A>G']['gene_symbol'] == 'NRXN1'
        assert results['NM_001330078.1:c.4254A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001317007.1:p.(Pro1418=)', 'slr': 'NP_001317007.1:p.(P1418=)'}
        assert results['NM_001330078.1:c.4254A>G']['submitted_variant'] == '2-50149352-T-C'
        assert results['NM_001330078.1:c.4254A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001330078.1:c.4254A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001330078.1:c.4254A>G']['hgvs_transcript_variant'] == 'NM_001330078.1:c.4254A>G'
        assert results['NM_001330078.1:c.4254A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001330078.1:c.4254A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001330078.1:c.4254A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}
        assert results['NM_001330078.1:c.4254A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001330078.1:c.4254A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}

        assert 'NM_001330094.1:c.4233A>G' in results.keys()
        assert results['NM_001330094.1:c.4233A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001330094.1:c.4233A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001330094.1:c.4233A>G']['alt_genomic_loci'] == []
        assert results['NM_001330094.1:c.4233A>G']['gene_symbol'] == 'NRXN1'
        assert results['NM_001330094.1:c.4233A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001317023.1:p.(Pro1411=)', 'slr': 'NP_001317023.1:p.(P1411=)'}
        assert results['NM_001330094.1:c.4233A>G']['submitted_variant'] == '2-50149352-T-C'
        assert results['NM_001330094.1:c.4233A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001330094.1:c.4233A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001330094.1:c.4233A>G']['hgvs_transcript_variant'] == 'NM_001330094.1:c.4233A>G'
        assert results['NM_001330094.1:c.4233A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001330094.1:c.4233A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001330094.1:c.4233A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}
        assert results['NM_001330094.1:c.4233A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001330094.1:c.4233A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}

        assert 'NM_001320156.3:c.159A>G' in results.keys()
        assert results['NM_001320156.3:c.159A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001320156.3:c.159A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001320156.3:c.159A>G']['alt_genomic_loci'] == []
        assert results['NM_001320156.3:c.159A>G']['gene_symbol'] == 'NRXN1'
        assert results['NM_001320156.3:c.159A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001307085.1:p.(Pro53=)', 'slr': 'NP_001307085.1:p.(P53=)'}
        assert results['NM_001320156.3:c.159A>G']['submitted_variant'] == '2-50149352-T-C'
        assert results['NM_001320156.3:c.159A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001320156.3:c.159A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001320156.3:c.159A>G']['hgvs_transcript_variant'] == 'NM_001320156.3:c.159A>G'
        assert results['NM_001320156.3:c.159A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001320156.3:c.159A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001320156.3:c.159A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}
        assert results['NM_001320156.3:c.159A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001320156.3:c.159A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}

        assert 'NM_001330088.1:c.4074A>G' in results.keys()
        assert results['NM_001330088.1:c.4074A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001330088.1:c.4074A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001330088.1:c.4074A>G']['alt_genomic_loci'] == []
        assert results['NM_001330088.1:c.4074A>G']['gene_symbol'] == 'NRXN1'
        assert results['NM_001330088.1:c.4074A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001317017.1:p.(Pro1358=)', 'slr': 'NP_001317017.1:p.(P1358=)'}
        assert results['NM_001330088.1:c.4074A>G']['submitted_variant'] == '2-50149352-T-C'
        assert results['NM_001330088.1:c.4074A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001330088.1:c.4074A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001330088.1:c.4074A>G']['hgvs_transcript_variant'] == 'NM_001330088.1:c.4074A>G'
        assert results['NM_001330088.1:c.4074A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001330088.1:c.4074A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001330088.1:c.4074A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}
        assert results['NM_001330088.1:c.4074A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001330088.1:c.4074A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}

        assert 'NM_001330092.1:c.1149A>G' in results.keys()
        assert results['NM_001330092.1:c.1149A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001330092.1:c.1149A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001330092.1:c.1149A>G']['alt_genomic_loci'] == []
        assert results['NM_001330092.1:c.1149A>G']['gene_symbol'] == 'NRXN1'
        assert results['NM_001330092.1:c.1149A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001317021.1:p.(Pro383=)', 'slr': 'NP_001317021.1:p.(P383=)'}
        assert results['NM_001330092.1:c.1149A>G']['submitted_variant'] == '2-50149352-T-C'
        assert results['NM_001330092.1:c.1149A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001330092.1:c.1149A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001330092.1:c.1149A>G']['hgvs_transcript_variant'] == 'NM_001330092.1:c.1149A>G'
        assert results['NM_001330092.1:c.1149A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001330092.1:c.1149A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001330092.1:c.1149A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}
        assert results['NM_001330092.1:c.1149A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001330092.1:c.1149A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}

        assert 'NM_138735.4:c.1059A>G' in results.keys()
        assert results['NM_138735.4:c.1059A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_138735.4:c.1059A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_138735.4:c.1059A>G']['alt_genomic_loci'] == []
        assert results['NM_138735.4:c.1059A>G']['gene_symbol'] == 'NRXN1'
        assert results['NM_138735.4:c.1059A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_620072.1:p.(Pro353=)', 'slr': 'NP_620072.1:p.(P353=)'}
        assert results['NM_138735.4:c.1059A>G']['submitted_variant'] == '2-50149352-T-C'
        assert results['NM_138735.4:c.1059A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_138735.4:c.1059A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_138735.4:c.1059A>G']['hgvs_transcript_variant'] == 'NM_138735.4:c.1059A>G'
        assert results['NM_138735.4:c.1059A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_138735.4:c.1059A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_138735.4:c.1059A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}
        assert results['NM_138735.4:c.1059A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_138735.4:c.1059A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}

        assert 'NM_001330096.1:c.4044A>G' in results.keys()
        assert results['NM_001330096.1:c.4044A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001330096.1:c.4044A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001330096.1:c.4044A>G']['alt_genomic_loci'] == []
        assert results['NM_001330096.1:c.4044A>G']['gene_symbol'] == 'NRXN1'
        assert results['NM_001330096.1:c.4044A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001317025.1:p.(Pro1348=)', 'slr': 'NP_001317025.1:p.(P1348=)'}
        assert results['NM_001330096.1:c.4044A>G']['submitted_variant'] == '2-50149352-T-C'
        assert results['NM_001330096.1:c.4044A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001330096.1:c.4044A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001330096.1:c.4044A>G']['hgvs_transcript_variant'] == 'NM_001330096.1:c.4044A>G'
        assert results['NM_001330096.1:c.4044A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001330096.1:c.4044A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001330096.1:c.4044A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}
        assert results['NM_001330096.1:c.4044A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001330096.1:c.4044A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}

        assert 'NM_001135659.2:c.4374A>G' in results.keys()
        assert results['NM_001135659.2:c.4374A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001135659.2:c.4374A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001135659.2:c.4374A>G']['alt_genomic_loci'] == []
        assert results['NM_001135659.2:c.4374A>G']['gene_symbol'] == 'NRXN1'
        assert results['NM_001135659.2:c.4374A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001129131.1:p.(Pro1458=)', 'slr': 'NP_001129131.1:p.(P1458=)'}
        assert results['NM_001135659.2:c.4374A>G']['submitted_variant'] == '2-50149352-T-C'
        assert results['NM_001135659.2:c.4374A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001135659.2:c.4374A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001135659.2:c.4374A>G']['hgvs_transcript_variant'] == 'NM_001135659.2:c.4374A>G'
        assert results['NM_001135659.2:c.4374A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001135659.2:c.4374A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001135659.2:c.4374A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}
        assert results['NM_001135659.2:c.4374A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001135659.2:c.4374A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}

        assert 'NM_001330085.1:c.4227A>G' in results.keys()
        assert results['NM_001330085.1:c.4227A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001330085.1:c.4227A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001330085.1:c.4227A>G']['alt_genomic_loci'] == []
        assert results['NM_001330085.1:c.4227A>G']['gene_symbol'] == 'NRXN1'
        assert results['NM_001330085.1:c.4227A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001317014.1:p.(Pro1409=)', 'slr': 'NP_001317014.1:p.(P1409=)'}
        assert results['NM_001330085.1:c.4227A>G']['submitted_variant'] == '2-50149352-T-C'
        assert results['NM_001330085.1:c.4227A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001330085.1:c.4227A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001330085.1:c.4227A>G']['hgvs_transcript_variant'] == 'NM_001330085.1:c.4227A>G'
        assert results['NM_001330085.1:c.4227A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001330085.1:c.4227A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001330085.1:c.4227A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}
        assert results['NM_001330085.1:c.4227A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001330085.1:c.4227A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}

        assert 'NM_001320156.1:c.159A>G' in results.keys()
        assert results['NM_001320156.1:c.159A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001320156.1:c.159A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001320156.1:c.159A>G']['alt_genomic_loci'] == []
        assert results['NM_001320156.1:c.159A>G']['gene_symbol'] == 'NRXN1'
        assert results['NM_001320156.1:c.159A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001307085.1:p.(Pro53=)', 'slr': 'NP_001307085.1:p.(P53=)'}
        assert results['NM_001320156.1:c.159A>G']['submitted_variant'] == '2-50149352-T-C'
        assert results['NM_001320156.1:c.159A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001320156.1:c.159A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001320156.1:c.159A>G']['hgvs_transcript_variant'] == 'NM_001320156.1:c.159A>G'
        assert results['NM_001320156.1:c.159A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001320156.1:c.159A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001320156.1:c.159A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}
        assert results['NM_001320156.1:c.159A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001320156.1:c.159A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}

        assert 'NM_001330077.1:c.4230A>G' in results.keys()
        assert results['NM_001330077.1:c.4230A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001330077.1:c.4230A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001330077.1:c.4230A>G']['alt_genomic_loci'] == []
        assert results['NM_001330077.1:c.4230A>G']['gene_symbol'] == 'NRXN1'
        assert results['NM_001330077.1:c.4230A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001317006.1:p.(Pro1410=)', 'slr': 'NP_001317006.1:p.(P1410=)'}
        assert results['NM_001330077.1:c.4230A>G']['submitted_variant'] == '2-50149352-T-C'
        assert results['NM_001330077.1:c.4230A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001330077.1:c.4230A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001330077.1:c.4230A>G']['hgvs_transcript_variant'] == 'NM_001330077.1:c.4230A>G'
        assert results['NM_001330077.1:c.4230A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001330077.1:c.4230A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001330077.1:c.4230A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}
        assert results['NM_001330077.1:c.4230A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001330077.1:c.4230A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}

        assert 'NM_001330093.1:c.4251A>G' in results.keys()
        assert results['NM_001330093.1:c.4251A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001330093.1:c.4251A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001330093.1:c.4251A>G']['alt_genomic_loci'] == []
        assert results['NM_001330093.1:c.4251A>G']['gene_symbol'] == 'NRXN1'
        assert results['NM_001330093.1:c.4251A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001317022.1:p.(Pro1417=)', 'slr': 'NP_001317022.1:p.(P1417=)'}
        assert results['NM_001330093.1:c.4251A>G']['submitted_variant'] == '2-50149352-T-C'
        assert results['NM_001330093.1:c.4251A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001330093.1:c.4251A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001330093.1:c.4251A>G']['hgvs_transcript_variant'] == 'NM_001330093.1:c.4251A>G'
        assert results['NM_001330093.1:c.4251A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001330093.1:c.4251A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001330093.1:c.4251A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}
        assert results['NM_001330093.1:c.4251A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001330093.1:c.4251A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}

        assert 'NM_001135659.1:c.4374A>G' in results.keys()
        assert results['NM_001135659.1:c.4374A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001135659.1:c.4374A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001135659.1:c.4374A>G']['alt_genomic_loci'] == []
        assert results['NM_001135659.1:c.4374A>G']['gene_symbol'] == 'NRXN1'
        assert results['NM_001135659.1:c.4374A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001129131.1:p.(Pro1458=)', 'slr': 'NP_001129131.1:p.(P1458=)'}
        assert results['NM_001135659.1:c.4374A>G']['submitted_variant'] == '2-50149352-T-C'
        assert results['NM_001135659.1:c.4374A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001135659.1:c.4374A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001135659.1:c.4374A>G']['hgvs_transcript_variant'] == 'NM_001135659.1:c.4374A>G'
        assert results['NM_001135659.1:c.4374A>G']['hgvs_refseqgene_variant'] == 'NG_011878.1:g.1115323A>G'
        assert results['NM_001135659.1:c.4374A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001135659.1:c.4374A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}
        assert results['NM_001135659.1:c.4374A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001135659.1:c.4374A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}

        assert 'NM_001320157.1:c.150A>G' in results.keys()
        assert results['NM_001320157.1:c.150A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001320157.1:c.150A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001320157.1:c.150A>G']['alt_genomic_loci'] == []
        assert results['NM_001320157.1:c.150A>G']['gene_symbol'] == 'NRXN1'
        assert results['NM_001320157.1:c.150A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001307086.1:p.(Pro50=)', 'slr': 'NP_001307086.1:p.(P50=)'}
        assert results['NM_001320157.1:c.150A>G']['submitted_variant'] == '2-50149352-T-C'
        assert results['NM_001320157.1:c.150A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001320157.1:c.150A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001320157.1:c.150A>G']['hgvs_transcript_variant'] == 'NM_001320157.1:c.150A>G'
        assert results['NM_001320157.1:c.150A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001320157.1:c.150A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001320157.1:c.150A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}
        assert results['NM_001320157.1:c.150A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001320157.1:c.150A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}

        assert 'NM_001330084.1:c.4188A>G' in results.keys()
        assert results['NM_001330084.1:c.4188A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001330084.1:c.4188A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001330084.1:c.4188A>G']['alt_genomic_loci'] == []
        assert results['NM_001330084.1:c.4188A>G']['gene_symbol'] == 'NRXN1'
        assert results['NM_001330084.1:c.4188A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001317013.1:p.(Pro1396=)', 'slr': 'NP_001317013.1:p.(P1396=)'}
        assert results['NM_001330084.1:c.4188A>G']['submitted_variant'] == '2-50149352-T-C'
        assert results['NM_001330084.1:c.4188A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001330084.1:c.4188A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001330084.1:c.4188A>G']['hgvs_transcript_variant'] == 'NM_001330084.1:c.4188A>G'
        assert results['NM_001330084.1:c.4188A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001330084.1:c.4188A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001330084.1:c.4188A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}
        assert results['NM_001330084.1:c.4188A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001330084.1:c.4188A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}

        assert 'NM_004801.4:c.4164A>G' in results.keys()
        assert results['NM_004801.4:c.4164A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_004801.4:c.4164A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_004801.4:c.4164A>G']['alt_genomic_loci'] == []
        assert results['NM_004801.4:c.4164A>G']['gene_symbol'] == 'NRXN1'
        assert results['NM_004801.4:c.4164A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_004792.1:p.(Pro1388=)', 'slr': 'NP_004792.1:p.(P1388=)'}
        assert results['NM_004801.4:c.4164A>G']['submitted_variant'] == '2-50149352-T-C'
        assert results['NM_004801.4:c.4164A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_004801.4:c.4164A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_004801.4:c.4164A>G']['hgvs_transcript_variant'] == 'NM_004801.4:c.4164A>G'
        assert results['NM_004801.4:c.4164A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_004801.4:c.4164A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_004801.4:c.4164A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}
        assert results['NM_004801.4:c.4164A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_004801.4:c.4164A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}

        assert 'NM_001330082.1:c.4221A>G' in results.keys()
        assert results['NM_001330082.1:c.4221A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001330082.1:c.4221A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001330082.1:c.4221A>G']['alt_genomic_loci'] == []
        assert results['NM_001330082.1:c.4221A>G']['gene_symbol'] == 'NRXN1'
        assert results['NM_001330082.1:c.4221A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001317011.1:p.(Pro1407=)', 'slr': 'NP_001317011.1:p.(P1407=)'}
        assert results['NM_001330082.1:c.4221A>G']['submitted_variant'] == '2-50149352-T-C'
        assert results['NM_001330082.1:c.4221A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001330082.1:c.4221A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001330082.1:c.4221A>G']['hgvs_transcript_variant'] == 'NM_001330082.1:c.4221A>G'
        assert results['NM_001330082.1:c.4221A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001330082.1:c.4221A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001330082.1:c.4221A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}
        assert results['NM_001330082.1:c.4221A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001330082.1:c.4221A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001330097.1:c.1050A>G' in results.keys()
        assert results['NM_001330097.1:c.1050A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001330097.1:c.1050A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001330097.1:c.1050A>G']['alt_genomic_loci'] == []
        assert results['NM_001330097.1:c.1050A>G']['gene_symbol'] == 'NRXN1'
        assert results['NM_001330097.1:c.1050A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001317026.1:p.(Pro350=)', 'slr': 'NP_001317026.1:p.(P350=)'}
        assert results['NM_001330097.1:c.1050A>G']['submitted_variant'] == '2-50149352-T-C'
        assert results['NM_001330097.1:c.1050A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001330097.1:c.1050A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001330097.1:c.1050A>G']['hgvs_transcript_variant'] == 'NM_001330097.1:c.1050A>G'
        assert results['NM_001330097.1:c.1050A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001330097.1:c.1050A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001330097.1:c.1050A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}
        assert results['NM_001330097.1:c.1050A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001330097.1:c.1050A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}

        assert 'NM_001320157.3:c.150A>G' in results.keys()
        assert results['NM_001320157.3:c.150A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001320157.3:c.150A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001320157.3:c.150A>G']['alt_genomic_loci'] == []
        assert results['NM_001320157.3:c.150A>G']['gene_symbol'] == 'NRXN1'
        assert results['NM_001320157.3:c.150A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001307086.1:p.(Pro50=)', 'slr': 'NP_001307086.1:p.(P50=)'}
        assert results['NM_001320157.3:c.150A>G']['submitted_variant'] == '2-50149352-T-C'
        assert results['NM_001320157.3:c.150A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001320157.3:c.150A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001320157.3:c.150A>G']['hgvs_transcript_variant'] == 'NM_001320157.3:c.150A>G'
        assert results['NM_001320157.3:c.150A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001320157.3:c.150A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001320157.3:c.150A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}
        assert results['NM_001320157.3:c.150A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001320157.3:c.150A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}

        assert 'NM_001330087.1:c.4053A>G' in results.keys()
        assert results['NM_001330087.1:c.4053A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001330087.1:c.4053A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001330087.1:c.4053A>G']['alt_genomic_loci'] == []
        assert results['NM_001330087.1:c.4053A>G']['gene_symbol'] == 'NRXN1'
        assert results['NM_001330087.1:c.4053A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001317016.1:p.(Pro1351=)', 'slr': 'NP_001317016.1:p.(P1351=)'}
        assert results['NM_001330087.1:c.4053A>G']['submitted_variant'] == '2-50149352-T-C'
        assert results['NM_001330087.1:c.4053A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001330087.1:c.4053A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001330087.1:c.4053A>G']['hgvs_transcript_variant'] == 'NM_001330087.1:c.4053A>G'
        assert results['NM_001330087.1:c.4053A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001330087.1:c.4053A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001330087.1:c.4053A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}
        assert results['NM_001330087.1:c.4053A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001330087.1:c.4053A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}

        assert 'NM_001330091.1:c.1140A>G' in results.keys()
        assert results['NM_001330091.1:c.1140A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001330091.1:c.1140A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001330091.1:c.1140A>G']['alt_genomic_loci'] == []
        assert results['NM_001330091.1:c.1140A>G']['gene_symbol'] == 'NRXN1'
        assert results['NM_001330091.1:c.1140A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001317020.1:p.(Pro380=)', 'slr': 'NP_001317020.1:p.(P380=)'}
        assert results['NM_001330091.1:c.1140A>G']['submitted_variant'] == '2-50149352-T-C'
        assert results['NM_001330091.1:c.1140A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001330091.1:c.1140A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001330091.1:c.1140A>G']['hgvs_transcript_variant'] == 'NM_001330091.1:c.1140A>G'
        assert results['NM_001330091.1:c.1140A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001330091.1:c.1140A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001330091.1:c.1140A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}
        assert results['NM_001330091.1:c.1140A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_001330091.1:c.1140A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}

        assert 'NM_004801.5:c.4164A>G' in results.keys()
        assert results['NM_004801.5:c.4164A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_004801.5:c.4164A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_004801.5:c.4164A>G']['alt_genomic_loci'] == []
        assert results['NM_004801.5:c.4164A>G']['gene_symbol'] == 'NRXN1'
        assert results['NM_004801.5:c.4164A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_004792.1:p.(Pro1388=)', 'slr': 'NP_004792.1:p.(P1388=)'}
        assert results['NM_004801.5:c.4164A>G']['submitted_variant'] == '2-50149352-T-C'
        assert results['NM_004801.5:c.4164A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_004801.5:c.4164A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_004801.5:c.4164A>G']['hgvs_transcript_variant'] == 'NM_004801.5:c.4164A>G'
        assert results['NM_004801.5:c.4164A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_004801.5:c.4164A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_004801.5:c.4164A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': 'chr2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}
        assert results['NM_004801.5:c.4164A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50149352T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '50149352', 'alt': u'C'}}
        assert results['NM_004801.5:c.4164A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.49922214T>C', 'vcf': {'chr': '2', 'ref': u'T', 'pos': '49922214', 'alt': u'C'}}


    def test_variant268(self):
        variant = '2-50847195-G-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001330096.1:c.1201C>T' in results.keys()
        assert results['NM_001330096.1:c.1201C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001330096.1:c.1201C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001330096.1:c.1201C>T']['alt_genomic_loci'] == []
        assert results['NM_001330096.1:c.1201C>T']['gene_symbol'] == 'NRXN1'
        assert results['NM_001330096.1:c.1201C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001317025.1:p.(Pro401Ser)', 'slr': 'NP_001317025.1:p.(P401S)'}
        assert results['NM_001330096.1:c.1201C>T']['submitted_variant'] == '2-50847195-G-A'
        assert results['NM_001330096.1:c.1201C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001330096.1:c.1201C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001330096.1:c.1201C>T']['hgvs_transcript_variant'] == 'NM_001330096.1:c.1201C>T'
        assert results['NM_001330096.1:c.1201C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001330096.1:c.1201C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50847195G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '50847195', 'alt': u'A'}}
        assert results['NM_001330096.1:c.1201C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.50620057G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '50620057', 'alt': u'A'}}
        assert results['NM_001330096.1:c.1201C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50847195G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '50847195', 'alt': u'A'}}
        assert results['NM_001330096.1:c.1201C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.50620057G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '50620057', 'alt': u'A'}}

        assert 'NM_001135659.1:c.1405C>T' in results.keys()
        assert results['NM_001135659.1:c.1405C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001135659.1:c.1405C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001135659.1:c.1405C>T']['alt_genomic_loci'] == []
        assert results['NM_001135659.1:c.1405C>T']['gene_symbol'] == 'NRXN1'
        assert results['NM_001135659.1:c.1405C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001129131.1:p.(Pro469Ser)', 'slr': 'NP_001129131.1:p.(P469S)'}
        assert results['NM_001135659.1:c.1405C>T']['submitted_variant'] == '2-50847195-G-A'
        assert results['NM_001135659.1:c.1405C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001135659.1:c.1405C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001135659.1:c.1405C>T']['hgvs_transcript_variant'] == 'NM_001135659.1:c.1405C>T'
        assert results['NM_001135659.1:c.1405C>T']['hgvs_refseqgene_variant'] == 'NG_011878.1:g.417480C>T'
        assert results['NM_001135659.1:c.1405C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50847195G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '50847195', 'alt': u'A'}}
        assert results['NM_001135659.1:c.1405C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.50620057G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '50620057', 'alt': u'A'}}
        assert results['NM_001135659.1:c.1405C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50847195G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '50847195', 'alt': u'A'}}
        assert results['NM_001135659.1:c.1405C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.50620057G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '50620057', 'alt': u'A'}}

        assert 'NM_001330077.1:c.1261C>T' in results.keys()
        assert results['NM_001330077.1:c.1261C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001330077.1:c.1261C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001330077.1:c.1261C>T']['alt_genomic_loci'] == []
        assert results['NM_001330077.1:c.1261C>T']['gene_symbol'] == 'NRXN1'
        assert results['NM_001330077.1:c.1261C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001317006.1:p.(Pro421Ser)', 'slr': 'NP_001317006.1:p.(P421S)'}
        assert results['NM_001330077.1:c.1261C>T']['submitted_variant'] == '2-50847195-G-A'
        assert results['NM_001330077.1:c.1261C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001330077.1:c.1261C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001330077.1:c.1261C>T']['hgvs_transcript_variant'] == 'NM_001330077.1:c.1261C>T'
        assert results['NM_001330077.1:c.1261C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001330077.1:c.1261C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50847195G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '50847195', 'alt': u'A'}}
        assert results['NM_001330077.1:c.1261C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.50620057G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '50620057', 'alt': u'A'}}
        assert results['NM_001330077.1:c.1261C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50847195G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '50847195', 'alt': u'A'}}
        assert results['NM_001330077.1:c.1261C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.50620057G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '50620057', 'alt': u'A'}}

        assert 'NM_001330094.1:c.1273C>T' in results.keys()
        assert results['NM_001330094.1:c.1273C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001330094.1:c.1273C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001330094.1:c.1273C>T']['alt_genomic_loci'] == []
        assert results['NM_001330094.1:c.1273C>T']['gene_symbol'] == 'NRXN1'
        assert results['NM_001330094.1:c.1273C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001317023.1:p.(Pro425Ser)', 'slr': 'NP_001317023.1:p.(P425S)'}
        assert results['NM_001330094.1:c.1273C>T']['submitted_variant'] == '2-50847195-G-A'
        assert results['NM_001330094.1:c.1273C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001330094.1:c.1273C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001330094.1:c.1273C>T']['hgvs_transcript_variant'] == 'NM_001330094.1:c.1273C>T'
        assert results['NM_001330094.1:c.1273C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001330094.1:c.1273C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50847195G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '50847195', 'alt': u'A'}}
        assert results['NM_001330094.1:c.1273C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.50620057G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '50620057', 'alt': u'A'}}
        assert results['NM_001330094.1:c.1273C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50847195G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '50847195', 'alt': u'A'}}
        assert results['NM_001330094.1:c.1273C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.50620057G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '50620057', 'alt': u'A'}}

        assert 'NM_001330088.1:c.1231C>T' in results.keys()
        assert results['NM_001330088.1:c.1231C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001330088.1:c.1231C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001330088.1:c.1231C>T']['alt_genomic_loci'] == []
        assert results['NM_001330088.1:c.1231C>T']['gene_symbol'] == 'NRXN1'
        assert results['NM_001330088.1:c.1231C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001317017.1:p.(Pro411Ser)', 'slr': 'NP_001317017.1:p.(P411S)'}
        assert results['NM_001330088.1:c.1231C>T']['submitted_variant'] == '2-50847195-G-A'
        assert results['NM_001330088.1:c.1231C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001330088.1:c.1231C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001330088.1:c.1231C>T']['hgvs_transcript_variant'] == 'NM_001330088.1:c.1231C>T'
        assert results['NM_001330088.1:c.1231C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001330088.1:c.1231C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50847195G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '50847195', 'alt': u'A'}}
        assert results['NM_001330088.1:c.1231C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.50620057G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '50620057', 'alt': u'A'}}
        assert results['NM_001330088.1:c.1231C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50847195G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '50847195', 'alt': u'A'}}
        assert results['NM_001330088.1:c.1231C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.50620057G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '50620057', 'alt': u'A'}}

        assert 'NM_001330093.1:c.1282C>T' in results.keys()
        assert results['NM_001330093.1:c.1282C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001330093.1:c.1282C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001330093.1:c.1282C>T']['alt_genomic_loci'] == []
        assert results['NM_001330093.1:c.1282C>T']['gene_symbol'] == 'NRXN1'
        assert results['NM_001330093.1:c.1282C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001317022.1:p.(Pro428Ser)', 'slr': 'NP_001317022.1:p.(P428S)'}
        assert results['NM_001330093.1:c.1282C>T']['submitted_variant'] == '2-50847195-G-A'
        assert results['NM_001330093.1:c.1282C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001330093.1:c.1282C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001330093.1:c.1282C>T']['hgvs_transcript_variant'] == 'NM_001330093.1:c.1282C>T'
        assert results['NM_001330093.1:c.1282C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001330093.1:c.1282C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50847195G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '50847195', 'alt': u'A'}}
        assert results['NM_001330093.1:c.1282C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.50620057G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '50620057', 'alt': u'A'}}
        assert results['NM_001330093.1:c.1282C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50847195G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '50847195', 'alt': u'A'}}
        assert results['NM_001330093.1:c.1282C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.50620057G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '50620057', 'alt': u'A'}}

        assert 'NM_001330087.1:c.1201C>T' in results.keys()
        assert results['NM_001330087.1:c.1201C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001330087.1:c.1201C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001330087.1:c.1201C>T']['alt_genomic_loci'] == []
        assert results['NM_001330087.1:c.1201C>T']['gene_symbol'] == 'NRXN1'
        assert results['NM_001330087.1:c.1201C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001317016.1:p.(Pro401Ser)', 'slr': 'NP_001317016.1:p.(P401S)'}
        assert results['NM_001330087.1:c.1201C>T']['submitted_variant'] == '2-50847195-G-A'
        assert results['NM_001330087.1:c.1201C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001330087.1:c.1201C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001330087.1:c.1201C>T']['hgvs_transcript_variant'] == 'NM_001330087.1:c.1201C>T'
        assert results['NM_001330087.1:c.1201C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001330087.1:c.1201C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50847195G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '50847195', 'alt': u'A'}}
        assert results['NM_001330087.1:c.1201C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.50620057G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '50620057', 'alt': u'A'}}
        assert results['NM_001330087.1:c.1201C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50847195G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '50847195', 'alt': u'A'}}
        assert results['NM_001330087.1:c.1201C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.50620057G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '50620057', 'alt': u'A'}}

        assert 'NM_001330082.1:c.1261C>T' in results.keys()
        assert results['NM_001330082.1:c.1261C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001330082.1:c.1261C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001330082.1:c.1261C>T']['alt_genomic_loci'] == []
        assert results['NM_001330082.1:c.1261C>T']['gene_symbol'] == 'NRXN1'
        assert results['NM_001330082.1:c.1261C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001317011.1:p.(Pro421Ser)', 'slr': 'NP_001317011.1:p.(P421S)'}
        assert results['NM_001330082.1:c.1261C>T']['submitted_variant'] == '2-50847195-G-A'
        assert results['NM_001330082.1:c.1261C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001330082.1:c.1261C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001330082.1:c.1261C>T']['hgvs_transcript_variant'] == 'NM_001330082.1:c.1261C>T'
        assert results['NM_001330082.1:c.1261C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001330082.1:c.1261C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50847195G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '50847195', 'alt': u'A'}}
        assert results['NM_001330082.1:c.1261C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.50620057G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '50620057', 'alt': u'A'}}
        assert results['NM_001330082.1:c.1261C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50847195G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '50847195', 'alt': u'A'}}
        assert results['NM_001330082.1:c.1261C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.50620057G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '50620057', 'alt': u'A'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001330083.1:c.1246C>T' in results.keys()
        assert results['NM_001330083.1:c.1246C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001330083.1:c.1246C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001330083.1:c.1246C>T']['alt_genomic_loci'] == []
        assert results['NM_001330083.1:c.1246C>T']['gene_symbol'] == 'NRXN1'
        assert results['NM_001330083.1:c.1246C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001317012.1:p.(Pro416Ser)', 'slr': 'NP_001317012.1:p.(P416S)'}
        assert results['NM_001330083.1:c.1246C>T']['submitted_variant'] == '2-50847195-G-A'
        assert results['NM_001330083.1:c.1246C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001330083.1:c.1246C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001330083.1:c.1246C>T']['hgvs_transcript_variant'] == 'NM_001330083.1:c.1246C>T'
        assert results['NM_001330083.1:c.1246C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001330083.1:c.1246C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50847195G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '50847195', 'alt': u'A'}}
        assert results['NM_001330083.1:c.1246C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.50620057G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '50620057', 'alt': u'A'}}
        assert results['NM_001330083.1:c.1246C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50847195G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '50847195', 'alt': u'A'}}
        assert results['NM_001330083.1:c.1246C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.50620057G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '50620057', 'alt': u'A'}}

        assert 'NM_001135659.2:c.1405C>T' in results.keys()
        assert results['NM_001135659.2:c.1405C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001135659.2:c.1405C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001135659.2:c.1405C>T']['alt_genomic_loci'] == []
        assert results['NM_001135659.2:c.1405C>T']['gene_symbol'] == 'NRXN1'
        assert results['NM_001135659.2:c.1405C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001129131.1:p.(Pro469Ser)', 'slr': 'NP_001129131.1:p.(P469S)'}
        assert results['NM_001135659.2:c.1405C>T']['submitted_variant'] == '2-50847195-G-A'
        assert results['NM_001135659.2:c.1405C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001135659.2:c.1405C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001135659.2:c.1405C>T']['hgvs_transcript_variant'] == 'NM_001135659.2:c.1405C>T'
        assert results['NM_001135659.2:c.1405C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001135659.2:c.1405C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50847195G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '50847195', 'alt': u'A'}}
        assert results['NM_001135659.2:c.1405C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.50620057G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '50620057', 'alt': u'A'}}
        assert results['NM_001135659.2:c.1405C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50847195G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '50847195', 'alt': u'A'}}
        assert results['NM_001135659.2:c.1405C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.50620057G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '50620057', 'alt': u'A'}}

        assert 'NM_004801.5:c.1285C>T' in results.keys()
        assert results['NM_004801.5:c.1285C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_004801.5:c.1285C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_004801.5:c.1285C>T']['alt_genomic_loci'] == []
        assert results['NM_004801.5:c.1285C>T']['gene_symbol'] == 'NRXN1'
        assert results['NM_004801.5:c.1285C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_004792.1:p.(Pro429Ser)', 'slr': 'NP_004792.1:p.(P429S)'}
        assert results['NM_004801.5:c.1285C>T']['submitted_variant'] == '2-50847195-G-A'
        assert results['NM_004801.5:c.1285C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_004801.5:c.1285C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_004801.5:c.1285C>T']['hgvs_transcript_variant'] == 'NM_004801.5:c.1285C>T'
        assert results['NM_004801.5:c.1285C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_004801.5:c.1285C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50847195G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '50847195', 'alt': u'A'}}
        assert results['NM_004801.5:c.1285C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.50620057G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '50620057', 'alt': u'A'}}
        assert results['NM_004801.5:c.1285C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50847195G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '50847195', 'alt': u'A'}}
        assert results['NM_004801.5:c.1285C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.50620057G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '50620057', 'alt': u'A'}}

        assert 'NM_001330086.1:c.1285C>T' in results.keys()
        assert results['NM_001330086.1:c.1285C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001330086.1:c.1285C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001330086.1:c.1285C>T']['alt_genomic_loci'] == []
        assert results['NM_001330086.1:c.1285C>T']['gene_symbol'] == 'NRXN1'
        assert results['NM_001330086.1:c.1285C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001317015.1:p.(Pro429Ser)', 'slr': 'NP_001317015.1:p.(P429S)'}
        assert results['NM_001330086.1:c.1285C>T']['submitted_variant'] == '2-50847195-G-A'
        assert results['NM_001330086.1:c.1285C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001330086.1:c.1285C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001330086.1:c.1285C>T']['hgvs_transcript_variant'] == 'NM_001330086.1:c.1285C>T'
        assert results['NM_001330086.1:c.1285C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001330086.1:c.1285C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50847195G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '50847195', 'alt': u'A'}}
        assert results['NM_001330086.1:c.1285C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.50620057G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '50620057', 'alt': u'A'}}
        assert results['NM_001330086.1:c.1285C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50847195G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '50847195', 'alt': u'A'}}
        assert results['NM_001330086.1:c.1285C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.50620057G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '50620057', 'alt': u'A'}}

        assert 'NM_001330078.1:c.1285C>T' in results.keys()
        assert results['NM_001330078.1:c.1285C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001330078.1:c.1285C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001330078.1:c.1285C>T']['alt_genomic_loci'] == []
        assert results['NM_001330078.1:c.1285C>T']['gene_symbol'] == 'NRXN1'
        assert results['NM_001330078.1:c.1285C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001317007.1:p.(Pro429Ser)', 'slr': 'NP_001317007.1:p.(P429S)'}
        assert results['NM_001330078.1:c.1285C>T']['submitted_variant'] == '2-50847195-G-A'
        assert results['NM_001330078.1:c.1285C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001330078.1:c.1285C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001330078.1:c.1285C>T']['hgvs_transcript_variant'] == 'NM_001330078.1:c.1285C>T'
        assert results['NM_001330078.1:c.1285C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001330078.1:c.1285C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50847195G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '50847195', 'alt': u'A'}}
        assert results['NM_001330078.1:c.1285C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.50620057G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '50620057', 'alt': u'A'}}
        assert results['NM_001330078.1:c.1285C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50847195G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '50847195', 'alt': u'A'}}
        assert results['NM_001330078.1:c.1285C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.50620057G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '50620057', 'alt': u'A'}}

        assert 'NM_001330085.1:c.1285C>T' in results.keys()
        assert results['NM_001330085.1:c.1285C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001330085.1:c.1285C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001330085.1:c.1285C>T']['alt_genomic_loci'] == []
        assert results['NM_001330085.1:c.1285C>T']['gene_symbol'] == 'NRXN1'
        assert results['NM_001330085.1:c.1285C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001317014.1:p.(Pro429Ser)', 'slr': 'NP_001317014.1:p.(P429S)'}
        assert results['NM_001330085.1:c.1285C>T']['submitted_variant'] == '2-50847195-G-A'
        assert results['NM_001330085.1:c.1285C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001330085.1:c.1285C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001330085.1:c.1285C>T']['hgvs_transcript_variant'] == 'NM_001330085.1:c.1285C>T'
        assert results['NM_001330085.1:c.1285C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001330085.1:c.1285C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50847195G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '50847195', 'alt': u'A'}}
        assert results['NM_001330085.1:c.1285C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.50620057G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '50620057', 'alt': u'A'}}
        assert results['NM_001330085.1:c.1285C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50847195G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '50847195', 'alt': u'A'}}
        assert results['NM_001330085.1:c.1285C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.50620057G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '50620057', 'alt': u'A'}}

        assert 'NM_001330084.1:c.1246C>T' in results.keys()
        assert results['NM_001330084.1:c.1246C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001330084.1:c.1246C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001330084.1:c.1246C>T']['alt_genomic_loci'] == []
        assert results['NM_001330084.1:c.1246C>T']['gene_symbol'] == 'NRXN1'
        assert results['NM_001330084.1:c.1246C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001317013.1:p.(Pro416Ser)', 'slr': 'NP_001317013.1:p.(P416S)'}
        assert results['NM_001330084.1:c.1246C>T']['submitted_variant'] == '2-50847195-G-A'
        assert results['NM_001330084.1:c.1246C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001330084.1:c.1246C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001330084.1:c.1246C>T']['hgvs_transcript_variant'] == 'NM_001330084.1:c.1246C>T'
        assert results['NM_001330084.1:c.1246C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001330084.1:c.1246C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50847195G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '50847195', 'alt': u'A'}}
        assert results['NM_001330084.1:c.1246C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.50620057G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '50620057', 'alt': u'A'}}
        assert results['NM_001330084.1:c.1246C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50847195G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '50847195', 'alt': u'A'}}
        assert results['NM_001330084.1:c.1246C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.50620057G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '50620057', 'alt': u'A'}}

        assert 'NM_004801.4:c.1285C>T' in results.keys()
        assert results['NM_004801.4:c.1285C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_004801.4:c.1285C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_004801.4:c.1285C>T']['alt_genomic_loci'] == []
        assert results['NM_004801.4:c.1285C>T']['gene_symbol'] == 'NRXN1'
        assert results['NM_004801.4:c.1285C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_004792.1:p.(Pro429Ser)', 'slr': 'NP_004792.1:p.(P429S)'}
        assert results['NM_004801.4:c.1285C>T']['submitted_variant'] == '2-50847195-G-A'
        assert results['NM_004801.4:c.1285C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_004801.4:c.1285C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_004801.4:c.1285C>T']['hgvs_transcript_variant'] == 'NM_004801.4:c.1285C>T'
        assert results['NM_004801.4:c.1285C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_004801.4:c.1285C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50847195G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '50847195', 'alt': u'A'}}
        assert results['NM_004801.4:c.1285C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.50620057G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '50620057', 'alt': u'A'}}
        assert results['NM_004801.4:c.1285C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50847195G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '50847195', 'alt': u'A'}}
        assert results['NM_004801.4:c.1285C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.50620057G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '50620057', 'alt': u'A'}}

        assert 'NM_001330095.1:c.1261C>T' in results.keys()
        assert results['NM_001330095.1:c.1261C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001330095.1:c.1261C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001330095.1:c.1261C>T']['alt_genomic_loci'] == []
        assert results['NM_001330095.1:c.1261C>T']['gene_symbol'] == 'NRXN1'
        assert results['NM_001330095.1:c.1261C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001317024.1:p.(Pro421Ser)', 'slr': 'NP_001317024.1:p.(P421S)'}
        assert results['NM_001330095.1:c.1261C>T']['submitted_variant'] == '2-50847195-G-A'
        assert results['NM_001330095.1:c.1261C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001330095.1:c.1261C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001330095.1:c.1261C>T']['hgvs_transcript_variant'] == 'NM_001330095.1:c.1261C>T'
        assert results['NM_001330095.1:c.1261C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001330095.1:c.1261C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.50847195G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '50847195', 'alt': u'A'}}
        assert results['NM_001330095.1:c.1261C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.50620057G>A', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '50620057', 'alt': u'A'}}
        assert results['NM_001330095.1:c.1261C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.50847195G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '50847195', 'alt': u'A'}}
        assert results['NM_001330095.1:c.1261C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.50620057G>A', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '50620057', 'alt': u'A'}}


    def test_variant269(self):
        variant = '2-71825797-C-G'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001130976.1:c.3582C>G' in results.keys()
        assert results['NM_001130976.1:c.3582C>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001130976.1:c.3582C>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001130976.1:c.3582C>G']['alt_genomic_loci'] == []
        assert results['NM_001130976.1:c.3582C>G']['gene_symbol'] == 'DYSF'
        assert results['NM_001130976.1:c.3582C>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001124448.1:p.(Ile1194Met)', 'slr': 'NP_001124448.1:p.(I1194M)'}
        assert results['NM_001130976.1:c.3582C>G']['submitted_variant'] == '2-71825797-C-G'
        assert results['NM_001130976.1:c.3582C>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001130976.1:c.3582C>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001130976.1:c.3582C>G']['hgvs_transcript_variant'] == 'NM_001130976.1:c.3582C>G'
        assert results['NM_001130976.1:c.3582C>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001130976.1:c.3582C>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.71825797C>G', 'vcf': {'chr': 'chr2', 'ref': 'C', 'pos': '71825797', 'alt': 'G'}}
        assert results['NM_001130976.1:c.3582C>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.71598667C>G', 'vcf': {'chr': 'chr2', 'ref': 'C', 'pos': '71598667', 'alt': 'G'}}
        assert results['NM_001130976.1:c.3582C>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.71825797C>G', 'vcf': {'chr': '2', 'ref': 'C', 'pos': '71825797', 'alt': 'G'}}
        assert results['NM_001130976.1:c.3582C>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.71598667C>G', 'vcf': {'chr': '2', 'ref': 'C', 'pos': '71598667', 'alt': 'G'}}

        assert 'NM_001130981.1:c.3675C>G' in results.keys()
        assert results['NM_001130981.1:c.3675C>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001130981.1:c.3675C>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001130981.1:c.3675C>G']['alt_genomic_loci'] == []
        assert results['NM_001130981.1:c.3675C>G']['gene_symbol'] == 'DYSF'
        assert results['NM_001130981.1:c.3675C>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001124453.1:p.(Ile1225Met)', 'slr': 'NP_001124453.1:p.(I1225M)'}
        assert results['NM_001130981.1:c.3675C>G']['submitted_variant'] == '2-71825797-C-G'
        assert results['NM_001130981.1:c.3675C>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001130981.1:c.3675C>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001130981.1:c.3675C>G']['hgvs_transcript_variant'] == 'NM_001130981.1:c.3675C>G'
        assert results['NM_001130981.1:c.3675C>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001130981.1:c.3675C>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.71825797C>G', 'vcf': {'chr': 'chr2', 'ref': 'C', 'pos': '71825797', 'alt': 'G'}}
        assert results['NM_001130981.1:c.3675C>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.71598667C>G', 'vcf': {'chr': 'chr2', 'ref': 'C', 'pos': '71598667', 'alt': 'G'}}
        assert results['NM_001130981.1:c.3675C>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.71825797C>G', 'vcf': {'chr': '2', 'ref': 'C', 'pos': '71825797', 'alt': 'G'}}
        assert results['NM_001130981.1:c.3675C>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.71598667C>G', 'vcf': {'chr': '2', 'ref': 'C', 'pos': '71598667', 'alt': 'G'}}

        assert 'NM_003494.3:c.3624C>G' in results.keys()
        assert results['NM_003494.3:c.3624C>G']['hgvs_lrg_transcript_variant'] == 'LRG_845t1:c.3624C>G'
        assert results['NM_003494.3:c.3624C>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003494.3:c.3624C>G']['alt_genomic_loci'] == []
        assert results['NM_003494.3:c.3624C>G']['gene_symbol'] == 'DYSF'
        assert results['NM_003494.3:c.3624C>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003485.1(LRG_845p1):p.(Ile1208Met)', 'slr': 'NP_003485.1:p.(I1208M)'}
        assert results['NM_003494.3:c.3624C>G']['submitted_variant'] == '2-71825797-C-G'
        assert results['NM_003494.3:c.3624C>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_003494.3:c.3624C>G']['hgvs_lrg_variant'] == ''
        assert results['NM_003494.3:c.3624C>G']['hgvs_transcript_variant'] == 'NM_003494.3:c.3624C>G'
        assert results['NM_003494.3:c.3624C>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_003494.3:c.3624C>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.71825797C>G', 'vcf': {'chr': 'chr2', 'ref': 'C', 'pos': '71825797', 'alt': 'G'}}
        assert results['NM_003494.3:c.3624C>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.71598667C>G', 'vcf': {'chr': 'chr2', 'ref': 'C', 'pos': '71598667', 'alt': 'G'}}
        assert results['NM_003494.3:c.3624C>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.71825797C>G', 'vcf': {'chr': '2', 'ref': 'C', 'pos': '71825797', 'alt': 'G'}}
        assert results['NM_003494.3:c.3624C>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.71598667C>G', 'vcf': {'chr': '2', 'ref': 'C', 'pos': '71598667', 'alt': 'G'}}

        assert 'NM_001130985.1:c.3678C>G' in results.keys()
        assert results['NM_001130985.1:c.3678C>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001130985.1:c.3678C>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001130985.1:c.3678C>G']['alt_genomic_loci'] == []
        assert results['NM_001130985.1:c.3678C>G']['gene_symbol'] == 'DYSF'
        assert results['NM_001130985.1:c.3678C>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001124457.1:p.(Ile1226Met)', 'slr': 'NP_001124457.1:p.(I1226M)'}
        assert results['NM_001130985.1:c.3678C>G']['submitted_variant'] == '2-71825797-C-G'
        assert results['NM_001130985.1:c.3678C>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001130985.1:c.3678C>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001130985.1:c.3678C>G']['hgvs_transcript_variant'] == 'NM_001130985.1:c.3678C>G'
        assert results['NM_001130985.1:c.3678C>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001130985.1:c.3678C>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.71825797C>G', 'vcf': {'chr': 'chr2', 'ref': 'C', 'pos': '71825797', 'alt': 'G'}}
        assert results['NM_001130985.1:c.3678C>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.71598667C>G', 'vcf': {'chr': 'chr2', 'ref': 'C', 'pos': '71598667', 'alt': 'G'}}
        assert results['NM_001130985.1:c.3678C>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.71825797C>G', 'vcf': {'chr': '2', 'ref': 'C', 'pos': '71825797', 'alt': 'G'}}
        assert results['NM_001130985.1:c.3678C>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.71598667C>G', 'vcf': {'chr': '2', 'ref': 'C', 'pos': '71598667', 'alt': 'G'}}

        assert 'NM_001130983.1:c.3627C>G' in results.keys()
        assert results['NM_001130983.1:c.3627C>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001130983.1:c.3627C>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001130983.1:c.3627C>G']['alt_genomic_loci'] == []
        assert results['NM_001130983.1:c.3627C>G']['gene_symbol'] == 'DYSF'
        assert results['NM_001130983.1:c.3627C>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001124455.1:p.(Ile1209Met)', 'slr': 'NP_001124455.1:p.(I1209M)'}
        assert results['NM_001130983.1:c.3627C>G']['submitted_variant'] == '2-71825797-C-G'
        assert results['NM_001130983.1:c.3627C>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001130983.1:c.3627C>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001130983.1:c.3627C>G']['hgvs_transcript_variant'] == 'NM_001130983.1:c.3627C>G'
        assert results['NM_001130983.1:c.3627C>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001130983.1:c.3627C>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.71825797C>G', 'vcf': {'chr': 'chr2', 'ref': 'C', 'pos': '71825797', 'alt': 'G'}}
        assert results['NM_001130983.1:c.3627C>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.71598667C>G', 'vcf': {'chr': 'chr2', 'ref': 'C', 'pos': '71598667', 'alt': 'G'}}
        assert results['NM_001130983.1:c.3627C>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.71825797C>G', 'vcf': {'chr': '2', 'ref': 'C', 'pos': '71825797', 'alt': 'G'}}
        assert results['NM_001130983.1:c.3627C>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.71598667C>G', 'vcf': {'chr': '2', 'ref': 'C', 'pos': '71598667', 'alt': 'G'}}

        assert 'NM_001130987.1:c.3678C>G' in results.keys()
        assert results['NM_001130987.1:c.3678C>G']['hgvs_lrg_transcript_variant'] == 'LRG_845t2:c.3678C>G'
        assert results['NM_001130987.1:c.3678C>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001130987.1:c.3678C>G']['alt_genomic_loci'] == []
        assert results['NM_001130987.1:c.3678C>G']['gene_symbol'] == 'DYSF'
        assert results['NM_001130987.1:c.3678C>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001124459.1:p.(Ile1226Met)', 'slr': 'NP_001124459.1:p.(I1226M)'}
        assert results['NM_001130987.1:c.3678C>G']['submitted_variant'] == '2-71825797-C-G'
        assert results['NM_001130987.1:c.3678C>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001130987.1:c.3678C>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001130987.1:c.3678C>G']['hgvs_transcript_variant'] == 'NM_001130987.1:c.3678C>G'
        assert results['NM_001130987.1:c.3678C>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001130987.1:c.3678C>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.71825797C>G', 'vcf': {'chr': 'chr2', 'ref': 'C', 'pos': '71825797', 'alt': 'G'}}
        assert results['NM_001130987.1:c.3678C>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.71598667C>G', 'vcf': {'chr': 'chr2', 'ref': 'C', 'pos': '71598667', 'alt': 'G'}}
        assert results['NM_001130987.1:c.3678C>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.71825797C>G', 'vcf': {'chr': '2', 'ref': 'C', 'pos': '71825797', 'alt': 'G'}}
        assert results['NM_001130987.1:c.3678C>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.71598667C>G', 'vcf': {'chr': '2', 'ref': 'C', 'pos': '71598667', 'alt': 'G'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001130982.1:c.3720C>G' in results.keys()
        assert results['NM_001130982.1:c.3720C>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001130982.1:c.3720C>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001130982.1:c.3720C>G']['alt_genomic_loci'] == []
        assert results['NM_001130982.1:c.3720C>G']['gene_symbol'] == 'DYSF'
        assert results['NM_001130982.1:c.3720C>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001124454.1:p.(Ile1240Met)', 'slr': 'NP_001124454.1:p.(I1240M)'}
        assert results['NM_001130982.1:c.3720C>G']['submitted_variant'] == '2-71825797-C-G'
        assert results['NM_001130982.1:c.3720C>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001130982.1:c.3720C>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001130982.1:c.3720C>G']['hgvs_transcript_variant'] == 'NM_001130982.1:c.3720C>G'
        assert results['NM_001130982.1:c.3720C>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001130982.1:c.3720C>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.71825797C>G', 'vcf': {'chr': 'chr2', 'ref': 'C', 'pos': '71825797', 'alt': 'G'}}
        assert results['NM_001130982.1:c.3720C>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.71598667C>G', 'vcf': {'chr': 'chr2', 'ref': 'C', 'pos': '71598667', 'alt': 'G'}}
        assert results['NM_001130982.1:c.3720C>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.71825797C>G', 'vcf': {'chr': '2', 'ref': 'C', 'pos': '71825797', 'alt': 'G'}}
        assert results['NM_001130982.1:c.3720C>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.71598667C>G', 'vcf': {'chr': '2', 'ref': 'C', 'pos': '71598667', 'alt': 'G'}}

        assert 'NM_001130980.1:c.3675C>G' in results.keys()
        assert results['NM_001130980.1:c.3675C>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001130980.1:c.3675C>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001130980.1:c.3675C>G']['alt_genomic_loci'] == []
        assert results['NM_001130980.1:c.3675C>G']['gene_symbol'] == 'DYSF'
        assert results['NM_001130980.1:c.3675C>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001124452.1:p.(Ile1225Met)', 'slr': 'NP_001124452.1:p.(I1225M)'}
        assert results['NM_001130980.1:c.3675C>G']['submitted_variant'] == '2-71825797-C-G'
        assert results['NM_001130980.1:c.3675C>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001130980.1:c.3675C>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001130980.1:c.3675C>G']['hgvs_transcript_variant'] == 'NM_001130980.1:c.3675C>G'
        assert results['NM_001130980.1:c.3675C>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001130980.1:c.3675C>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.71825797C>G', 'vcf': {'chr': 'chr2', 'ref': 'C', 'pos': '71825797', 'alt': 'G'}}
        assert results['NM_001130980.1:c.3675C>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.71598667C>G', 'vcf': {'chr': 'chr2', 'ref': 'C', 'pos': '71598667', 'alt': 'G'}}
        assert results['NM_001130980.1:c.3675C>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.71825797C>G', 'vcf': {'chr': '2', 'ref': 'C', 'pos': '71825797', 'alt': 'G'}}
        assert results['NM_001130980.1:c.3675C>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.71598667C>G', 'vcf': {'chr': '2', 'ref': 'C', 'pos': '71598667', 'alt': 'G'}}

        assert 'NM_001130979.1:c.3717C>G' in results.keys()
        assert results['NM_001130979.1:c.3717C>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001130979.1:c.3717C>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001130979.1:c.3717C>G']['alt_genomic_loci'] == []
        assert results['NM_001130979.1:c.3717C>G']['gene_symbol'] == 'DYSF'
        assert results['NM_001130979.1:c.3717C>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001124451.1:p.(Ile1239Met)', 'slr': 'NP_001124451.1:p.(I1239M)'}
        assert results['NM_001130979.1:c.3717C>G']['submitted_variant'] == '2-71825797-C-G'
        assert results['NM_001130979.1:c.3717C>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001130979.1:c.3717C>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001130979.1:c.3717C>G']['hgvs_transcript_variant'] == 'NM_001130979.1:c.3717C>G'
        assert results['NM_001130979.1:c.3717C>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001130979.1:c.3717C>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.71825797C>G', 'vcf': {'chr': 'chr2', 'ref': 'C', 'pos': '71825797', 'alt': 'G'}}
        assert results['NM_001130979.1:c.3717C>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.71598667C>G', 'vcf': {'chr': 'chr2', 'ref': 'C', 'pos': '71598667', 'alt': 'G'}}
        assert results['NM_001130979.1:c.3717C>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.71825797C>G', 'vcf': {'chr': '2', 'ref': 'C', 'pos': '71825797', 'alt': 'G'}}
        assert results['NM_001130979.1:c.3717C>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.71598667C>G', 'vcf': {'chr': '2', 'ref': 'C', 'pos': '71598667', 'alt': 'G'}}

        assert 'NM_001130984.1:c.3585C>G' in results.keys()
        assert results['NM_001130984.1:c.3585C>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001130984.1:c.3585C>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001130984.1:c.3585C>G']['alt_genomic_loci'] == []
        assert results['NM_001130984.1:c.3585C>G']['gene_symbol'] == 'DYSF'
        assert results['NM_001130984.1:c.3585C>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001124456.1:p.(Ile1195Met)', 'slr': 'NP_001124456.1:p.(I1195M)'}
        assert results['NM_001130984.1:c.3585C>G']['submitted_variant'] == '2-71825797-C-G'
        assert results['NM_001130984.1:c.3585C>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001130984.1:c.3585C>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001130984.1:c.3585C>G']['hgvs_transcript_variant'] == 'NM_001130984.1:c.3585C>G'
        assert results['NM_001130984.1:c.3585C>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001130984.1:c.3585C>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.71825797C>G', 'vcf': {'chr': 'chr2', 'ref': 'C', 'pos': '71825797', 'alt': 'G'}}
        assert results['NM_001130984.1:c.3585C>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.71598667C>G', 'vcf': {'chr': 'chr2', 'ref': 'C', 'pos': '71598667', 'alt': 'G'}}
        assert results['NM_001130984.1:c.3585C>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.71825797C>G', 'vcf': {'chr': '2', 'ref': 'C', 'pos': '71825797', 'alt': 'G'}}
        assert results['NM_001130984.1:c.3585C>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.71598667C>G', 'vcf': {'chr': '2', 'ref': 'C', 'pos': '71598667', 'alt': 'G'}}

        assert 'NM_001130977.1:c.3582C>G' in results.keys()
        assert results['NM_001130977.1:c.3582C>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001130977.1:c.3582C>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001130977.1:c.3582C>G']['alt_genomic_loci'] == []
        assert results['NM_001130977.1:c.3582C>G']['gene_symbol'] == 'DYSF'
        assert results['NM_001130977.1:c.3582C>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001124449.1:p.(Ile1194Met)', 'slr': 'NP_001124449.1:p.(I1194M)'}
        assert results['NM_001130977.1:c.3582C>G']['submitted_variant'] == '2-71825797-C-G'
        assert results['NM_001130977.1:c.3582C>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001130977.1:c.3582C>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001130977.1:c.3582C>G']['hgvs_transcript_variant'] == 'NM_001130977.1:c.3582C>G'
        assert results['NM_001130977.1:c.3582C>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001130977.1:c.3582C>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.71825797C>G', 'vcf': {'chr': 'chr2', 'ref': 'C', 'pos': '71825797', 'alt': 'G'}}
        assert results['NM_001130977.1:c.3582C>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.71598667C>G', 'vcf': {'chr': 'chr2', 'ref': 'C', 'pos': '71598667', 'alt': 'G'}}
        assert results['NM_001130977.1:c.3582C>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.71825797C>G', 'vcf': {'chr': '2', 'ref': 'C', 'pos': '71825797', 'alt': 'G'}}
        assert results['NM_001130977.1:c.3582C>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.71598667C>G', 'vcf': {'chr': '2', 'ref': 'C', 'pos': '71598667', 'alt': 'G'}}

        assert 'NM_001130455.1:c.3627C>G' in results.keys()
        assert results['NM_001130455.1:c.3627C>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001130455.1:c.3627C>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001130455.1:c.3627C>G']['alt_genomic_loci'] == []
        assert results['NM_001130455.1:c.3627C>G']['gene_symbol'] == 'DYSF'
        assert results['NM_001130455.1:c.3627C>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001123927.1:p.(Ile1209Met)', 'slr': 'NP_001123927.1:p.(I1209M)'}
        assert results['NM_001130455.1:c.3627C>G']['submitted_variant'] == '2-71825797-C-G'
        assert results['NM_001130455.1:c.3627C>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001130455.1:c.3627C>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001130455.1:c.3627C>G']['hgvs_transcript_variant'] == 'NM_001130455.1:c.3627C>G'
        assert results['NM_001130455.1:c.3627C>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001130455.1:c.3627C>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.71825797C>G', 'vcf': {'chr': 'chr2', 'ref': 'C', 'pos': '71825797', 'alt': 'G'}}
        assert results['NM_001130455.1:c.3627C>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.71598667C>G', 'vcf': {'chr': 'chr2', 'ref': 'C', 'pos': '71598667', 'alt': 'G'}}
        assert results['NM_001130455.1:c.3627C>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.71825797C>G', 'vcf': {'chr': '2', 'ref': 'C', 'pos': '71825797', 'alt': 'G'}}
        assert results['NM_001130455.1:c.3627C>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.71598667C>G', 'vcf': {'chr': '2', 'ref': 'C', 'pos': '71598667', 'alt': 'G'}}

        assert 'NM_001130978.1:c.3624C>G' in results.keys()
        assert results['NM_001130978.1:c.3624C>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001130978.1:c.3624C>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001130978.1:c.3624C>G']['alt_genomic_loci'] == []
        assert results['NM_001130978.1:c.3624C>G']['gene_symbol'] == 'DYSF'
        assert results['NM_001130978.1:c.3624C>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001124450.1:p.(Ile1208Met)', 'slr': 'NP_001124450.1:p.(I1208M)'}
        assert results['NM_001130978.1:c.3624C>G']['submitted_variant'] == '2-71825797-C-G'
        assert results['NM_001130978.1:c.3624C>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001130978.1:c.3624C>G']['hgvs_lrg_variant'] == 'LRG_845:g.150045C>G'
        assert results['NM_001130978.1:c.3624C>G']['hgvs_transcript_variant'] == 'NM_001130978.1:c.3624C>G'
        assert results['NM_001130978.1:c.3624C>G']['hgvs_refseqgene_variant'] == 'NG_008694.1:g.150045C>G'
        assert results['NM_001130978.1:c.3624C>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.71825797C>G', 'vcf': {'chr': 'chr2', 'ref': 'C', 'pos': '71825797', 'alt': 'G'}}
        assert results['NM_001130978.1:c.3624C>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.71598667C>G', 'vcf': {'chr': 'chr2', 'ref': 'C', 'pos': '71598667', 'alt': 'G'}}
        assert results['NM_001130978.1:c.3624C>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.71825797C>G', 'vcf': {'chr': '2', 'ref': 'C', 'pos': '71825797', 'alt': 'G'}}
        assert results['NM_001130978.1:c.3624C>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.71598667C>G', 'vcf': {'chr': '2', 'ref': 'C', 'pos': '71598667', 'alt': 'G'}}

        assert 'NM_001130986.1:c.3585C>G' in results.keys()
        assert results['NM_001130986.1:c.3585C>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001130986.1:c.3585C>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001130986.1:c.3585C>G']['alt_genomic_loci'] == []
        assert results['NM_001130986.1:c.3585C>G']['gene_symbol'] == 'DYSF'
        assert results['NM_001130986.1:c.3585C>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001124458.1:p.(Ile1195Met)', 'slr': 'NP_001124458.1:p.(I1195M)'}
        assert results['NM_001130986.1:c.3585C>G']['submitted_variant'] == '2-71825797-C-G'
        assert results['NM_001130986.1:c.3585C>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001130986.1:c.3585C>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001130986.1:c.3585C>G']['hgvs_transcript_variant'] == 'NM_001130986.1:c.3585C>G'
        assert results['NM_001130986.1:c.3585C>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001130986.1:c.3585C>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.71825797C>G', 'vcf': {'chr': 'chr2', 'ref': 'C', 'pos': '71825797', 'alt': 'G'}}
        assert results['NM_001130986.1:c.3585C>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.71598667C>G', 'vcf': {'chr': 'chr2', 'ref': 'C', 'pos': '71598667', 'alt': 'G'}}
        assert results['NM_001130986.1:c.3585C>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.71825797C>G', 'vcf': {'chr': '2', 'ref': 'C', 'pos': '71825797', 'alt': 'G'}}
        assert results['NM_001130986.1:c.3585C>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.71598667C>G', 'vcf': {'chr': '2', 'ref': 'C', 'pos': '71598667', 'alt': 'G'}}


    def test_variant270(self):
        variant = '2-166179712-G-C'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_021007.2:c.1718G>C' in results.keys()
        assert results['NM_021007.2:c.1718G>C']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_021007.2:c.1718G>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_021007.2:c.1718G>C']['alt_genomic_loci'] == []
        assert results['NM_021007.2:c.1718G>C']['gene_symbol'] == 'SCN2A'
        assert results['NM_021007.2:c.1718G>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_066287.2:p.(Ser573Thr)', 'slr': 'NP_066287.2:p.(S573T)'}
        assert results['NM_021007.2:c.1718G>C']['submitted_variant'] == '2-166179712-G-C'
        assert results['NM_021007.2:c.1718G>C']['genome_context_intronic_sequence'] == ''
        assert results['NM_021007.2:c.1718G>C']['hgvs_lrg_variant'] == ''
        assert results['NM_021007.2:c.1718G>C']['hgvs_transcript_variant'] == 'NM_021007.2:c.1718G>C'
        assert results['NM_021007.2:c.1718G>C']['hgvs_refseqgene_variant'] == ''
        assert results['NM_021007.2:c.1718G>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166179712G>C', 'vcf': {'chr': 'chr2', 'ref': 'G', 'pos': '166179712', 'alt': 'C'}}
        assert results['NM_021007.2:c.1718G>C']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.165323202G>C', 'vcf': {'chr': 'chr2', 'ref': 'G', 'pos': '165323202', 'alt': 'C'}}
        assert results['NM_021007.2:c.1718G>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166179712G>C', 'vcf': {'chr': '2', 'ref': 'G', 'pos': '166179712', 'alt': 'C'}}
        assert results['NM_021007.2:c.1718G>C']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.165323202G>C', 'vcf': {'chr': '2', 'ref': 'G', 'pos': '165323202', 'alt': 'C'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001040143.1:c.1718G>C' in results.keys()
        assert results['NM_001040143.1:c.1718G>C']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001040143.1:c.1718G>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001040143.1:c.1718G>C']['alt_genomic_loci'] == []
        assert results['NM_001040143.1:c.1718G>C']['gene_symbol'] == 'SCN2A'
        assert results['NM_001040143.1:c.1718G>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001035233.1:p.(Ser573Thr)', 'slr': 'NP_001035233.1:p.(S573T)'}
        assert results['NM_001040143.1:c.1718G>C']['submitted_variant'] == '2-166179712-G-C'
        assert results['NM_001040143.1:c.1718G>C']['genome_context_intronic_sequence'] == ''
        assert results['NM_001040143.1:c.1718G>C']['hgvs_lrg_variant'] == ''
        assert results['NM_001040143.1:c.1718G>C']['hgvs_transcript_variant'] == 'NM_001040143.1:c.1718G>C'
        assert results['NM_001040143.1:c.1718G>C']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001040143.1:c.1718G>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166179712G>C', 'vcf': {'chr': 'chr2', 'ref': 'G', 'pos': '166179712', 'alt': 'C'}}
        assert results['NM_001040143.1:c.1718G>C']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.165323202G>C', 'vcf': {'chr': 'chr2', 'ref': 'G', 'pos': '165323202', 'alt': 'C'}}
        assert results['NM_001040143.1:c.1718G>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166179712G>C', 'vcf': {'chr': '2', 'ref': 'G', 'pos': '166179712', 'alt': 'C'}}
        assert results['NM_001040143.1:c.1718G>C']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.165323202G>C', 'vcf': {'chr': '2', 'ref': 'G', 'pos': '165323202', 'alt': 'C'}}

        assert 'NM_001040142.1:c.1718G>C' in results.keys()
        assert results['NM_001040142.1:c.1718G>C']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001040142.1:c.1718G>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001040142.1:c.1718G>C']['alt_genomic_loci'] == []
        assert results['NM_001040142.1:c.1718G>C']['gene_symbol'] == 'SCN2A'
        assert results['NM_001040142.1:c.1718G>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001035232.1:p.(Ser573Thr)', 'slr': 'NP_001035232.1:p.(S573T)'}
        assert results['NM_001040142.1:c.1718G>C']['submitted_variant'] == '2-166179712-G-C'
        assert results['NM_001040142.1:c.1718G>C']['genome_context_intronic_sequence'] == ''
        assert results['NM_001040142.1:c.1718G>C']['hgvs_lrg_variant'] == ''
        assert results['NM_001040142.1:c.1718G>C']['hgvs_transcript_variant'] == 'NM_001040142.1:c.1718G>C'
        assert results['NM_001040142.1:c.1718G>C']['hgvs_refseqgene_variant'] == 'NG_008143.1:g.88801G>C'
        assert results['NM_001040142.1:c.1718G>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166179712G>C', 'vcf': {'chr': 'chr2', 'ref': 'G', 'pos': '166179712', 'alt': 'C'}}
        assert results['NM_001040142.1:c.1718G>C']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.165323202G>C', 'vcf': {'chr': 'chr2', 'ref': 'G', 'pos': '165323202', 'alt': 'C'}}
        assert results['NM_001040142.1:c.1718G>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166179712G>C', 'vcf': {'chr': '2', 'ref': 'G', 'pos': '166179712', 'alt': 'C'}}
        assert results['NM_001040142.1:c.1718G>C']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.165323202G>C', 'vcf': {'chr': '2', 'ref': 'G', 'pos': '165323202', 'alt': 'C'}}


    def test_variant271(self):
        variant = '2-166183371-A-G'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_021007.2:c.2026A>G' in results.keys()
        assert results['NM_021007.2:c.2026A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_021007.2:c.2026A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_021007.2:c.2026A>G']['alt_genomic_loci'] == []
        assert results['NM_021007.2:c.2026A>G']['gene_symbol'] == 'SCN2A'
        assert results['NM_021007.2:c.2026A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_066287.2:p.(Thr676Ala)', 'slr': 'NP_066287.2:p.(T676A)'}
        assert results['NM_021007.2:c.2026A>G']['submitted_variant'] == '2-166183371-A-G'
        assert results['NM_021007.2:c.2026A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_021007.2:c.2026A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_021007.2:c.2026A>G']['hgvs_transcript_variant'] == 'NM_021007.2:c.2026A>G'
        assert results['NM_021007.2:c.2026A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_021007.2:c.2026A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166183371A>G', 'vcf': {'chr': 'chr2', 'ref': 'A', 'pos': '166183371', 'alt': 'G'}}
        assert results['NM_021007.2:c.2026A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.165326861A>G', 'vcf': {'chr': 'chr2', 'ref': 'A', 'pos': '165326861', 'alt': 'G'}}
        assert results['NM_021007.2:c.2026A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166183371A>G', 'vcf': {'chr': '2', 'ref': 'A', 'pos': '166183371', 'alt': 'G'}}
        assert results['NM_021007.2:c.2026A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.165326861A>G', 'vcf': {'chr': '2', 'ref': 'A', 'pos': '165326861', 'alt': 'G'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001040143.1:c.2026A>G' in results.keys()
        assert results['NM_001040143.1:c.2026A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001040143.1:c.2026A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001040143.1:c.2026A>G']['alt_genomic_loci'] == []
        assert results['NM_001040143.1:c.2026A>G']['gene_symbol'] == 'SCN2A'
        assert results['NM_001040143.1:c.2026A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001035233.1:p.(Thr676Ala)', 'slr': 'NP_001035233.1:p.(T676A)'}
        assert results['NM_001040143.1:c.2026A>G']['submitted_variant'] == '2-166183371-A-G'
        assert results['NM_001040143.1:c.2026A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001040143.1:c.2026A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001040143.1:c.2026A>G']['hgvs_transcript_variant'] == 'NM_001040143.1:c.2026A>G'
        assert results['NM_001040143.1:c.2026A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001040143.1:c.2026A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166183371A>G', 'vcf': {'chr': 'chr2', 'ref': 'A', 'pos': '166183371', 'alt': 'G'}}
        assert results['NM_001040143.1:c.2026A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.165326861A>G', 'vcf': {'chr': 'chr2', 'ref': 'A', 'pos': '165326861', 'alt': 'G'}}
        assert results['NM_001040143.1:c.2026A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166183371A>G', 'vcf': {'chr': '2', 'ref': 'A', 'pos': '166183371', 'alt': 'G'}}
        assert results['NM_001040143.1:c.2026A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.165326861A>G', 'vcf': {'chr': '2', 'ref': 'A', 'pos': '165326861', 'alt': 'G'}}

        assert 'NM_001040142.1:c.2026A>G' in results.keys()
        assert results['NM_001040142.1:c.2026A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001040142.1:c.2026A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001040142.1:c.2026A>G']['alt_genomic_loci'] == []
        assert results['NM_001040142.1:c.2026A>G']['gene_symbol'] == 'SCN2A'
        assert results['NM_001040142.1:c.2026A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001035232.1:p.(Thr676Ala)', 'slr': 'NP_001035232.1:p.(T676A)'}
        assert results['NM_001040142.1:c.2026A>G']['submitted_variant'] == '2-166183371-A-G'
        assert results['NM_001040142.1:c.2026A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001040142.1:c.2026A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001040142.1:c.2026A>G']['hgvs_transcript_variant'] == 'NM_001040142.1:c.2026A>G'
        assert results['NM_001040142.1:c.2026A>G']['hgvs_refseqgene_variant'] == 'NG_008143.1:g.92460A>G'
        assert results['NM_001040142.1:c.2026A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166183371A>G', 'vcf': {'chr': 'chr2', 'ref': 'A', 'pos': '166183371', 'alt': 'G'}}
        assert results['NM_001040142.1:c.2026A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.165326861A>G', 'vcf': {'chr': 'chr2', 'ref': 'A', 'pos': '165326861', 'alt': 'G'}}
        assert results['NM_001040142.1:c.2026A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166183371A>G', 'vcf': {'chr': '2', 'ref': 'A', 'pos': '166183371', 'alt': 'G'}}
        assert results['NM_001040142.1:c.2026A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.165326861A>G', 'vcf': {'chr': '2', 'ref': 'A', 'pos': '165326861', 'alt': 'G'}}


    def test_variant272(self):
        variant = '2-166929889-GTCCAGGTCCT-GAC'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001353951.1:c.233_242delinsGT' in results.keys()
        assert results['NM_001353951.1:c.233_242delinsGT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001353951.1:c.233_242delinsGT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001353951.1:c.233_242delinsGT']['alt_genomic_loci'] == []
        assert results['NM_001353951.1:c.233_242delinsGT']['gene_symbol'] == 'SCN1A'
        assert results['NM_001353951.1:c.233_242delinsGT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001340880.1:p.(Glu78GlyfsTer7)', 'slr': 'NP_001340880.1:p.(E78Gfs*7)'}
        assert results['NM_001353951.1:c.233_242delinsGT']['submitted_variant'] == '2-166929889-GTCCAGGTCCT-GAC'
        assert results['NM_001353951.1:c.233_242delinsGT']['genome_context_intronic_sequence'] == ''
        assert results['NM_001353951.1:c.233_242delinsGT']['hgvs_lrg_variant'] == ''
        assert results['NM_001353951.1:c.233_242delinsGT']['hgvs_transcript_variant'] == 'NM_001353951.1:c.233_242delinsGT'
        assert results['NM_001353951.1:c.233_242delinsGT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001353951.1:c.233_242delinsGT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_001353951.1:c.233_242delinsGT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}
        assert results['NM_001353951.1:c.233_242delinsGT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_001353951.1:c.233_242delinsGT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}

        assert 'NM_001353958.1:c.233_242delinsGT' in results.keys()
        assert results['NM_001353958.1:c.233_242delinsGT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001353958.1:c.233_242delinsGT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001353958.1:c.233_242delinsGT']['alt_genomic_loci'] == []
        assert results['NM_001353958.1:c.233_242delinsGT']['gene_symbol'] == 'SCN1A'
        assert results['NM_001353958.1:c.233_242delinsGT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001340887.1:p.(Glu78GlyfsTer7)', 'slr': 'NP_001340887.1:p.(E78Gfs*7)'}
        assert results['NM_001353958.1:c.233_242delinsGT']['submitted_variant'] == '2-166929889-GTCCAGGTCCT-GAC'
        assert results['NM_001353958.1:c.233_242delinsGT']['genome_context_intronic_sequence'] == ''
        assert results['NM_001353958.1:c.233_242delinsGT']['hgvs_lrg_variant'] == ''
        assert results['NM_001353958.1:c.233_242delinsGT']['hgvs_transcript_variant'] == 'NM_001353958.1:c.233_242delinsGT'
        assert results['NM_001353958.1:c.233_242delinsGT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001353958.1:c.233_242delinsGT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_001353958.1:c.233_242delinsGT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}
        assert results['NM_001353958.1:c.233_242delinsGT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_001353958.1:c.233_242delinsGT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}

        assert 'NM_001202435.1:c.233_242delinsGT' in results.keys()
        assert results['NM_001202435.1:c.233_242delinsGT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001202435.1:c.233_242delinsGT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001202435.1:c.233_242delinsGT']['alt_genomic_loci'] == []
        assert results['NM_001202435.1:c.233_242delinsGT']['gene_symbol'] == 'SCN1A'
        assert results['NM_001202435.1:c.233_242delinsGT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001189364.1:p.(Glu78GlyfsTer7)', 'slr': 'NP_001189364.1:p.(E78Gfs*7)'}
        assert results['NM_001202435.1:c.233_242delinsGT']['submitted_variant'] == '2-166929889-GTCCAGGTCCT-GAC'
        assert results['NM_001202435.1:c.233_242delinsGT']['genome_context_intronic_sequence'] == ''
        assert results['NM_001202435.1:c.233_242delinsGT']['hgvs_lrg_variant'] == ''
        assert results['NM_001202435.1:c.233_242delinsGT']['hgvs_transcript_variant'] == 'NM_001202435.1:c.233_242delinsGT'
        assert results['NM_001202435.1:c.233_242delinsGT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001202435.1:c.233_242delinsGT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_001202435.1:c.233_242delinsGT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}
        assert results['NM_001202435.1:c.233_242delinsGT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_001202435.1:c.233_242delinsGT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}

        assert 'NM_001165963.2:c.233_242delinsGT' in results.keys()
        assert results['NM_001165963.2:c.233_242delinsGT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001165963.2:c.233_242delinsGT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001165963.2:c.233_242delinsGT']['alt_genomic_loci'] == []
        assert results['NM_001165963.2:c.233_242delinsGT']['gene_symbol'] == 'SCN1A'
        assert results['NM_001165963.2:c.233_242delinsGT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001159435.1:p.(Glu78GlyfsTer7)', 'slr': 'NP_001159435.1:p.(E78Gfs*7)'}
        assert results['NM_001165963.2:c.233_242delinsGT']['submitted_variant'] == '2-166929889-GTCCAGGTCCT-GAC'
        assert results['NM_001165963.2:c.233_242delinsGT']['genome_context_intronic_sequence'] == ''
        assert results['NM_001165963.2:c.233_242delinsGT']['hgvs_lrg_variant'] == ''
        assert results['NM_001165963.2:c.233_242delinsGT']['hgvs_transcript_variant'] == 'NM_001165963.2:c.233_242delinsGT'
        assert results['NM_001165963.2:c.233_242delinsGT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001165963.2:c.233_242delinsGT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_001165963.2:c.233_242delinsGT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}
        assert results['NM_001165963.2:c.233_242delinsGT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_001165963.2:c.233_242delinsGT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}

        assert 'NM_001353950.1:c.233_242delinsGT' in results.keys()
        assert results['NM_001353950.1:c.233_242delinsGT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001353950.1:c.233_242delinsGT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001353950.1:c.233_242delinsGT']['alt_genomic_loci'] == []
        assert results['NM_001353950.1:c.233_242delinsGT']['gene_symbol'] == 'SCN1A'
        assert results['NM_001353950.1:c.233_242delinsGT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001340879.1:p.(Glu78GlyfsTer7)', 'slr': 'NP_001340879.1:p.(E78Gfs*7)'}
        assert results['NM_001353950.1:c.233_242delinsGT']['submitted_variant'] == '2-166929889-GTCCAGGTCCT-GAC'
        assert results['NM_001353950.1:c.233_242delinsGT']['genome_context_intronic_sequence'] == ''
        assert results['NM_001353950.1:c.233_242delinsGT']['hgvs_lrg_variant'] == ''
        assert results['NM_001353950.1:c.233_242delinsGT']['hgvs_transcript_variant'] == 'NM_001353950.1:c.233_242delinsGT'
        assert results['NM_001353950.1:c.233_242delinsGT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001353950.1:c.233_242delinsGT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_001353950.1:c.233_242delinsGT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}
        assert results['NM_001353950.1:c.233_242delinsGT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_001353950.1:c.233_242delinsGT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}

        assert 'NM_001353952.1:c.233_242delinsGT' in results.keys()
        assert results['NM_001353952.1:c.233_242delinsGT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001353952.1:c.233_242delinsGT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001353952.1:c.233_242delinsGT']['alt_genomic_loci'] == []
        assert results['NM_001353952.1:c.233_242delinsGT']['gene_symbol'] == 'SCN1A'
        assert results['NM_001353952.1:c.233_242delinsGT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001340881.1:p.(Glu78GlyfsTer7)', 'slr': 'NP_001340881.1:p.(E78Gfs*7)'}
        assert results['NM_001353952.1:c.233_242delinsGT']['submitted_variant'] == '2-166929889-GTCCAGGTCCT-GAC'
        assert results['NM_001353952.1:c.233_242delinsGT']['genome_context_intronic_sequence'] == ''
        assert results['NM_001353952.1:c.233_242delinsGT']['hgvs_lrg_variant'] == ''
        assert results['NM_001353952.1:c.233_242delinsGT']['hgvs_transcript_variant'] == 'NM_001353952.1:c.233_242delinsGT'
        assert results['NM_001353952.1:c.233_242delinsGT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001353952.1:c.233_242delinsGT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_001353952.1:c.233_242delinsGT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}
        assert results['NM_001353952.1:c.233_242delinsGT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_001353952.1:c.233_242delinsGT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}

        assert 'NM_001165963.1:c.233_242delinsGT' in results.keys()
        assert results['NM_001165963.1:c.233_242delinsGT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001165963.1:c.233_242delinsGT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001165963.1:c.233_242delinsGT']['alt_genomic_loci'] == []
        assert results['NM_001165963.1:c.233_242delinsGT']['gene_symbol'] == 'SCN1A'
        assert results['NM_001165963.1:c.233_242delinsGT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001159435.1:p.(Glu78GlyfsTer7)', 'slr': 'NP_001159435.1:p.(E78Gfs*7)'}
        assert results['NM_001165963.1:c.233_242delinsGT']['submitted_variant'] == '2-166929889-GTCCAGGTCCT-GAC'
        assert results['NM_001165963.1:c.233_242delinsGT']['genome_context_intronic_sequence'] == ''
        assert results['NM_001165963.1:c.233_242delinsGT']['hgvs_lrg_variant'] == ''
        assert results['NM_001165963.1:c.233_242delinsGT']['hgvs_transcript_variant'] == 'NM_001165963.1:c.233_242delinsGT'
        assert results['NM_001165963.1:c.233_242delinsGT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001165963.1:c.233_242delinsGT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_001165963.1:c.233_242delinsGT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}
        assert results['NM_001165963.1:c.233_242delinsGT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_001165963.1:c.233_242delinsGT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}

        assert 'NM_001353957.1:c.233_242delinsGT' in results.keys()
        assert results['NM_001353957.1:c.233_242delinsGT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001353957.1:c.233_242delinsGT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001353957.1:c.233_242delinsGT']['alt_genomic_loci'] == []
        assert results['NM_001353957.1:c.233_242delinsGT']['gene_symbol'] == 'SCN1A'
        assert results['NM_001353957.1:c.233_242delinsGT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001340886.1:p.(Glu78GlyfsTer7)', 'slr': 'NP_001340886.1:p.(E78Gfs*7)'}
        assert results['NM_001353957.1:c.233_242delinsGT']['submitted_variant'] == '2-166929889-GTCCAGGTCCT-GAC'
        assert results['NM_001353957.1:c.233_242delinsGT']['genome_context_intronic_sequence'] == ''
        assert results['NM_001353957.1:c.233_242delinsGT']['hgvs_lrg_variant'] == ''
        assert results['NM_001353957.1:c.233_242delinsGT']['hgvs_transcript_variant'] == 'NM_001353957.1:c.233_242delinsGT'
        assert results['NM_001353957.1:c.233_242delinsGT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001353957.1:c.233_242delinsGT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_001353957.1:c.233_242delinsGT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}
        assert results['NM_001353957.1:c.233_242delinsGT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_001353957.1:c.233_242delinsGT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}

        assert 'NM_001165964.2:c.233_242delinsGT' in results.keys()
        assert results['NM_001165964.2:c.233_242delinsGT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001165964.2:c.233_242delinsGT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001165964.2:c.233_242delinsGT']['alt_genomic_loci'] == []
        assert results['NM_001165964.2:c.233_242delinsGT']['gene_symbol'] == 'SCN1A'
        assert results['NM_001165964.2:c.233_242delinsGT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001159436.1:p.(Glu78GlyfsTer7)', 'slr': 'NP_001159436.1:p.(E78Gfs*7)'}
        assert results['NM_001165964.2:c.233_242delinsGT']['submitted_variant'] == '2-166929889-GTCCAGGTCCT-GAC'
        assert results['NM_001165964.2:c.233_242delinsGT']['genome_context_intronic_sequence'] == ''
        assert results['NM_001165964.2:c.233_242delinsGT']['hgvs_lrg_variant'] == ''
        assert results['NM_001165964.2:c.233_242delinsGT']['hgvs_transcript_variant'] == 'NM_001165964.2:c.233_242delinsGT'
        assert results['NM_001165964.2:c.233_242delinsGT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001165964.2:c.233_242delinsGT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_001165964.2:c.233_242delinsGT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}
        assert results['NM_001165964.2:c.233_242delinsGT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_001165964.2:c.233_242delinsGT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}

        assert results['flag'] == 'gene_variant'
        assert 'NR_148667.1:n.638_647delinsGT' in results.keys()
        assert results['NR_148667.1:n.638_647delinsGT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_148667.1:n.638_647delinsGT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_148667.1:n.638_647delinsGT']['alt_genomic_loci'] == []
        assert results['NR_148667.1:n.638_647delinsGT']['gene_symbol'] == 'SCN1A'
        assert results['NR_148667.1:n.638_647delinsGT']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_148667.1:n.638_647delinsGT']['submitted_variant'] == '2-166929889-GTCCAGGTCCT-GAC'
        assert results['NR_148667.1:n.638_647delinsGT']['genome_context_intronic_sequence'] == ''
        assert results['NR_148667.1:n.638_647delinsGT']['hgvs_lrg_variant'] == ''
        assert results['NR_148667.1:n.638_647delinsGT']['hgvs_transcript_variant'] == 'NR_148667.1:n.638_647delinsGT'
        assert results['NR_148667.1:n.638_647delinsGT']['hgvs_refseqgene_variant'] == ''
        assert results['NR_148667.1:n.638_647delinsGT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NR_148667.1:n.638_647delinsGT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}
        assert results['NR_148667.1:n.638_647delinsGT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NR_148667.1:n.638_647delinsGT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}

        assert 'NM_001353954.1:c.233_242delinsGT' in results.keys()
        assert results['NM_001353954.1:c.233_242delinsGT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001353954.1:c.233_242delinsGT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001353954.1:c.233_242delinsGT']['alt_genomic_loci'] == []
        assert results['NM_001353954.1:c.233_242delinsGT']['gene_symbol'] == 'SCN1A'
        assert results['NM_001353954.1:c.233_242delinsGT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001340883.1:p.(Glu78GlyfsTer7)', 'slr': 'NP_001340883.1:p.(E78Gfs*7)'}
        assert results['NM_001353954.1:c.233_242delinsGT']['submitted_variant'] == '2-166929889-GTCCAGGTCCT-GAC'
        assert results['NM_001353954.1:c.233_242delinsGT']['genome_context_intronic_sequence'] == ''
        assert results['NM_001353954.1:c.233_242delinsGT']['hgvs_lrg_variant'] == ''
        assert results['NM_001353954.1:c.233_242delinsGT']['hgvs_transcript_variant'] == 'NM_001353954.1:c.233_242delinsGT'
        assert results['NM_001353954.1:c.233_242delinsGT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001353954.1:c.233_242delinsGT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_001353954.1:c.233_242delinsGT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}
        assert results['NM_001353954.1:c.233_242delinsGT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_001353954.1:c.233_242delinsGT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}

        assert 'NM_006920.4:c.233_242delinsGT' in results.keys()
        assert results['NM_006920.4:c.233_242delinsGT']['hgvs_lrg_transcript_variant'] == 'LRG_8t1:c.233_242delinsGT'
        assert results['NM_006920.4:c.233_242delinsGT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_006920.4:c.233_242delinsGT']['alt_genomic_loci'] == []
        assert results['NM_006920.4:c.233_242delinsGT']['gene_symbol'] == 'SCN1A'
        assert results['NM_006920.4:c.233_242delinsGT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_008851.3(LRG_8p1):p.(Glu78GlyfsTer7)', 'slr': 'NP_008851.3:p.(E78Gfs*7)'}
        assert results['NM_006920.4:c.233_242delinsGT']['submitted_variant'] == '2-166929889-GTCCAGGTCCT-GAC'
        assert results['NM_006920.4:c.233_242delinsGT']['genome_context_intronic_sequence'] == ''
        assert results['NM_006920.4:c.233_242delinsGT']['hgvs_lrg_variant'] == 'LRG_8:g.5251_5260delinsGT'
        assert results['NM_006920.4:c.233_242delinsGT']['hgvs_transcript_variant'] == 'NM_006920.4:c.233_242delinsGT'
        assert results['NM_006920.4:c.233_242delinsGT']['hgvs_refseqgene_variant'] == 'NG_011906.1:g.5251_5260delinsGT'
        assert results['NM_006920.4:c.233_242delinsGT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_006920.4:c.233_242delinsGT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}
        assert results['NM_006920.4:c.233_242delinsGT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_006920.4:c.233_242delinsGT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}

        assert 'NM_001165964.1:c.233_242delinsGT' in results.keys()
        assert results['NM_001165964.1:c.233_242delinsGT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001165964.1:c.233_242delinsGT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001165964.1:c.233_242delinsGT']['alt_genomic_loci'] == []
        assert results['NM_001165964.1:c.233_242delinsGT']['gene_symbol'] == 'SCN1A'
        assert results['NM_001165964.1:c.233_242delinsGT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001159436.1:p.(Glu78GlyfsTer7)', 'slr': 'NP_001159436.1:p.(E78Gfs*7)'}
        assert results['NM_001165964.1:c.233_242delinsGT']['submitted_variant'] == '2-166929889-GTCCAGGTCCT-GAC'
        assert results['NM_001165964.1:c.233_242delinsGT']['genome_context_intronic_sequence'] == ''
        assert results['NM_001165964.1:c.233_242delinsGT']['hgvs_lrg_variant'] == ''
        assert results['NM_001165964.1:c.233_242delinsGT']['hgvs_transcript_variant'] == 'NM_001165964.1:c.233_242delinsGT'
        assert results['NM_001165964.1:c.233_242delinsGT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001165964.1:c.233_242delinsGT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_001165964.1:c.233_242delinsGT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}
        assert results['NM_001165964.1:c.233_242delinsGT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_001165964.1:c.233_242delinsGT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}

        assert 'NM_001353960.1:c.233_242delinsGT' in results.keys()
        assert results['NM_001353960.1:c.233_242delinsGT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001353960.1:c.233_242delinsGT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001353960.1:c.233_242delinsGT']['alt_genomic_loci'] == []
        assert results['NM_001353960.1:c.233_242delinsGT']['gene_symbol'] == 'SCN1A'
        assert results['NM_001353960.1:c.233_242delinsGT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001340889.1:p.(Glu78GlyfsTer7)', 'slr': 'NP_001340889.1:p.(E78Gfs*7)'}
        assert results['NM_001353960.1:c.233_242delinsGT']['submitted_variant'] == '2-166929889-GTCCAGGTCCT-GAC'
        assert results['NM_001353960.1:c.233_242delinsGT']['genome_context_intronic_sequence'] == ''
        assert results['NM_001353960.1:c.233_242delinsGT']['hgvs_lrg_variant'] == ''
        assert results['NM_001353960.1:c.233_242delinsGT']['hgvs_transcript_variant'] == 'NM_001353960.1:c.233_242delinsGT'
        assert results['NM_001353960.1:c.233_242delinsGT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001353960.1:c.233_242delinsGT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_001353960.1:c.233_242delinsGT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}
        assert results['NM_001353960.1:c.233_242delinsGT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_001353960.1:c.233_242delinsGT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}

        assert 'NM_001202435.2:c.233_242delinsGT' in results.keys()
        assert results['NM_001202435.2:c.233_242delinsGT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001202435.2:c.233_242delinsGT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001202435.2:c.233_242delinsGT']['alt_genomic_loci'] == []
        assert results['NM_001202435.2:c.233_242delinsGT']['gene_symbol'] == 'SCN1A'
        assert results['NM_001202435.2:c.233_242delinsGT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001189364.1:p.(Glu78GlyfsTer7)', 'slr': 'NP_001189364.1:p.(E78Gfs*7)'}
        assert results['NM_001202435.2:c.233_242delinsGT']['submitted_variant'] == '2-166929889-GTCCAGGTCCT-GAC'
        assert results['NM_001202435.2:c.233_242delinsGT']['genome_context_intronic_sequence'] == ''
        assert results['NM_001202435.2:c.233_242delinsGT']['hgvs_lrg_variant'] == ''
        assert results['NM_001202435.2:c.233_242delinsGT']['hgvs_transcript_variant'] == 'NM_001202435.2:c.233_242delinsGT'
        assert results['NM_001202435.2:c.233_242delinsGT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001202435.2:c.233_242delinsGT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_001202435.2:c.233_242delinsGT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}
        assert results['NM_001202435.2:c.233_242delinsGT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_001202435.2:c.233_242delinsGT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}

        assert 'NM_006920.5:c.233_242delinsGT' in results.keys()
        assert results['NM_006920.5:c.233_242delinsGT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_006920.5:c.233_242delinsGT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_006920.5:c.233_242delinsGT']['alt_genomic_loci'] == []
        assert results['NM_006920.5:c.233_242delinsGT']['gene_symbol'] == 'SCN1A'
        assert results['NM_006920.5:c.233_242delinsGT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_008851.3(LRG_8p1):p.(Glu78GlyfsTer7)', 'slr': 'NP_008851.3:p.(E78Gfs*7)'}
        assert results['NM_006920.5:c.233_242delinsGT']['submitted_variant'] == '2-166929889-GTCCAGGTCCT-GAC'
        assert results['NM_006920.5:c.233_242delinsGT']['genome_context_intronic_sequence'] == ''
        assert results['NM_006920.5:c.233_242delinsGT']['hgvs_lrg_variant'] == ''
        assert results['NM_006920.5:c.233_242delinsGT']['hgvs_transcript_variant'] == 'NM_006920.5:c.233_242delinsGT'
        assert results['NM_006920.5:c.233_242delinsGT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_006920.5:c.233_242delinsGT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_006920.5:c.233_242delinsGT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}
        assert results['NM_006920.5:c.233_242delinsGT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_006920.5:c.233_242delinsGT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}

        assert 'NM_001353955.1:c.233_242delinsGT' in results.keys()
        assert results['NM_001353955.1:c.233_242delinsGT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001353955.1:c.233_242delinsGT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001353955.1:c.233_242delinsGT']['alt_genomic_loci'] == []
        assert results['NM_001353955.1:c.233_242delinsGT']['gene_symbol'] == 'SCN1A'
        assert results['NM_001353955.1:c.233_242delinsGT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001340884.1:p.(Glu78GlyfsTer7)', 'slr': 'NP_001340884.1:p.(E78Gfs*7)'}
        assert results['NM_001353955.1:c.233_242delinsGT']['submitted_variant'] == '2-166929889-GTCCAGGTCCT-GAC'
        assert results['NM_001353955.1:c.233_242delinsGT']['genome_context_intronic_sequence'] == ''
        assert results['NM_001353955.1:c.233_242delinsGT']['hgvs_lrg_variant'] == ''
        assert results['NM_001353955.1:c.233_242delinsGT']['hgvs_transcript_variant'] == 'NM_001353955.1:c.233_242delinsGT'
        assert results['NM_001353955.1:c.233_242delinsGT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001353955.1:c.233_242delinsGT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_001353955.1:c.233_242delinsGT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}
        assert results['NM_001353955.1:c.233_242delinsGT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_001353955.1:c.233_242delinsGT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}

        assert 'NM_001353961.1:c.-2193_-2184delinsGT' in results.keys()
        assert results['NM_001353961.1:c.-2193_-2184delinsGT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001353961.1:c.-2193_-2184delinsGT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001353961.1:c.-2193_-2184delinsGT']['alt_genomic_loci'] == []
        assert results['NM_001353961.1:c.-2193_-2184delinsGT']['gene_symbol'] == 'SCN1A'
        assert results['NM_001353961.1:c.-2193_-2184delinsGT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001340890.1:p.?', 'slr': 'NP_001340890.1:p.?'}
        assert results['NM_001353961.1:c.-2193_-2184delinsGT']['submitted_variant'] == '2-166929889-GTCCAGGTCCT-GAC'
        assert results['NM_001353961.1:c.-2193_-2184delinsGT']['genome_context_intronic_sequence'] == ''
        assert results['NM_001353961.1:c.-2193_-2184delinsGT']['hgvs_lrg_variant'] == ''
        assert results['NM_001353961.1:c.-2193_-2184delinsGT']['hgvs_transcript_variant'] == 'NM_001353961.1:c.-2193_-2184delinsGT'
        assert results['NM_001353961.1:c.-2193_-2184delinsGT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001353961.1:c.-2193_-2184delinsGT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_001353961.1:c.-2193_-2184delinsGT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}
        assert results['NM_001353961.1:c.-2193_-2184delinsGT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_001353961.1:c.-2193_-2184delinsGT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}

        assert 'NM_001353948.1:c.233_242delinsGT' in results.keys()
        assert results['NM_001353948.1:c.233_242delinsGT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001353948.1:c.233_242delinsGT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001353948.1:c.233_242delinsGT']['alt_genomic_loci'] == []
        assert results['NM_001353948.1:c.233_242delinsGT']['gene_symbol'] == 'SCN1A'
        assert results['NM_001353948.1:c.233_242delinsGT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001340877.1:p.(Glu78GlyfsTer7)', 'slr': 'NP_001340877.1:p.(E78Gfs*7)'}
        assert results['NM_001353948.1:c.233_242delinsGT']['submitted_variant'] == '2-166929889-GTCCAGGTCCT-GAC'
        assert results['NM_001353948.1:c.233_242delinsGT']['genome_context_intronic_sequence'] == ''
        assert results['NM_001353948.1:c.233_242delinsGT']['hgvs_lrg_variant'] == ''
        assert results['NM_001353948.1:c.233_242delinsGT']['hgvs_transcript_variant'] == 'NM_001353948.1:c.233_242delinsGT'
        assert results['NM_001353948.1:c.233_242delinsGT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001353948.1:c.233_242delinsGT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_001353948.1:c.233_242delinsGT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}
        assert results['NM_001353948.1:c.233_242delinsGT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_001353948.1:c.233_242delinsGT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}

        assert 'NM_001353949.1:c.233_242delinsGT' in results.keys()
        assert results['NM_001353949.1:c.233_242delinsGT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001353949.1:c.233_242delinsGT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001353949.1:c.233_242delinsGT']['alt_genomic_loci'] == []
        assert results['NM_001353949.1:c.233_242delinsGT']['gene_symbol'] == 'SCN1A'
        assert results['NM_001353949.1:c.233_242delinsGT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001340878.1:p.(Glu78GlyfsTer7)', 'slr': 'NP_001340878.1:p.(E78Gfs*7)'}
        assert results['NM_001353949.1:c.233_242delinsGT']['submitted_variant'] == '2-166929889-GTCCAGGTCCT-GAC'
        assert results['NM_001353949.1:c.233_242delinsGT']['genome_context_intronic_sequence'] == ''
        assert results['NM_001353949.1:c.233_242delinsGT']['hgvs_lrg_variant'] == ''
        assert results['NM_001353949.1:c.233_242delinsGT']['hgvs_transcript_variant'] == 'NM_001353949.1:c.233_242delinsGT'
        assert results['NM_001353949.1:c.233_242delinsGT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001353949.1:c.233_242delinsGT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_001353949.1:c.233_242delinsGT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': 'chr2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}
        assert results['NM_001353949.1:c.233_242delinsGT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929890_166929899delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166929890', 'alt': u'AC'}}
        assert results['NM_001353949.1:c.233_242delinsGT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073380_166073389delinsAC', 'vcf': {'chr': '2', 'ref': 'TCCAGGTCCT', 'pos': '166073380', 'alt': u'AC'}}


    def test_variant273(self):
        variant = '2-166929891-CCAGGTCCT-C'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001353950.1:c.233_240del' in results.keys()
        assert results['NM_001353950.1:c.233_240del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001353950.1:c.233_240del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001353950.1:c.233_240del']['alt_genomic_loci'] == []
        assert results['NM_001353950.1:c.233_240del']['gene_symbol'] == 'SCN1A'
        assert results['NM_001353950.1:c.233_240del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001340879.1:p.(Glu78GlyfsTer7)', 'slr': 'NP_001340879.1:p.(E78Gfs*7)'}
        assert results['NM_001353950.1:c.233_240del']['submitted_variant'] == '2-166929891-CCAGGTCCT-C'
        assert results['NM_001353950.1:c.233_240del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001353950.1:c.233_240del']['hgvs_lrg_variant'] == ''
        assert results['NM_001353950.1:c.233_240del']['hgvs_transcript_variant'] == 'NM_001353950.1:c.233_240del'
        assert results['NM_001353950.1:c.233_240del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001353950.1:c.233_240del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_001353950.1:c.233_240del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}
        assert results['NM_001353950.1:c.233_240del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_001353950.1:c.233_240del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}

        assert 'NM_001353960.1:c.233_240del' in results.keys()
        assert results['NM_001353960.1:c.233_240del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001353960.1:c.233_240del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001353960.1:c.233_240del']['alt_genomic_loci'] == []
        assert results['NM_001353960.1:c.233_240del']['gene_symbol'] == 'SCN1A'
        assert results['NM_001353960.1:c.233_240del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001340889.1:p.(Glu78GlyfsTer7)', 'slr': 'NP_001340889.1:p.(E78Gfs*7)'}
        assert results['NM_001353960.1:c.233_240del']['submitted_variant'] == '2-166929891-CCAGGTCCT-C'
        assert results['NM_001353960.1:c.233_240del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001353960.1:c.233_240del']['hgvs_lrg_variant'] == ''
        assert results['NM_001353960.1:c.233_240del']['hgvs_transcript_variant'] == 'NM_001353960.1:c.233_240del'
        assert results['NM_001353960.1:c.233_240del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001353960.1:c.233_240del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_001353960.1:c.233_240del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}
        assert results['NM_001353960.1:c.233_240del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_001353960.1:c.233_240del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}

        assert 'NM_006920.5:c.233_240del' in results.keys()
        assert results['NM_006920.5:c.233_240del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_006920.5:c.233_240del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_006920.5:c.233_240del']['alt_genomic_loci'] == []
        assert results['NM_006920.5:c.233_240del']['gene_symbol'] == 'SCN1A'
        assert results['NM_006920.5:c.233_240del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_008851.3(LRG_8p1):p.(Glu78GlyfsTer7)', 'slr': 'NP_008851.3:p.(E78Gfs*7)'}
        assert results['NM_006920.5:c.233_240del']['submitted_variant'] == '2-166929891-CCAGGTCCT-C'
        assert results['NM_006920.5:c.233_240del']['genome_context_intronic_sequence'] == ''
        assert results['NM_006920.5:c.233_240del']['hgvs_lrg_variant'] == ''
        assert results['NM_006920.5:c.233_240del']['hgvs_transcript_variant'] == 'NM_006920.5:c.233_240del'
        assert results['NM_006920.5:c.233_240del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_006920.5:c.233_240del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_006920.5:c.233_240del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}
        assert results['NM_006920.5:c.233_240del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_006920.5:c.233_240del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}

        assert 'NM_001353955.1:c.233_240del' in results.keys()
        assert results['NM_001353955.1:c.233_240del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001353955.1:c.233_240del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001353955.1:c.233_240del']['alt_genomic_loci'] == []
        assert results['NM_001353955.1:c.233_240del']['gene_symbol'] == 'SCN1A'
        assert results['NM_001353955.1:c.233_240del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001340884.1:p.(Glu78GlyfsTer7)', 'slr': 'NP_001340884.1:p.(E78Gfs*7)'}
        assert results['NM_001353955.1:c.233_240del']['submitted_variant'] == '2-166929891-CCAGGTCCT-C'
        assert results['NM_001353955.1:c.233_240del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001353955.1:c.233_240del']['hgvs_lrg_variant'] == ''
        assert results['NM_001353955.1:c.233_240del']['hgvs_transcript_variant'] == 'NM_001353955.1:c.233_240del'
        assert results['NM_001353955.1:c.233_240del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001353955.1:c.233_240del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_001353955.1:c.233_240del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}
        assert results['NM_001353955.1:c.233_240del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_001353955.1:c.233_240del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}

        assert 'NM_001165964.2:c.233_240del' in results.keys()
        assert results['NM_001165964.2:c.233_240del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001165964.2:c.233_240del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001165964.2:c.233_240del']['alt_genomic_loci'] == []
        assert results['NM_001165964.2:c.233_240del']['gene_symbol'] == 'SCN1A'
        assert results['NM_001165964.2:c.233_240del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001159436.1:p.(Glu78GlyfsTer7)', 'slr': 'NP_001159436.1:p.(E78Gfs*7)'}
        assert results['NM_001165964.2:c.233_240del']['submitted_variant'] == '2-166929891-CCAGGTCCT-C'
        assert results['NM_001165964.2:c.233_240del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001165964.2:c.233_240del']['hgvs_lrg_variant'] == ''
        assert results['NM_001165964.2:c.233_240del']['hgvs_transcript_variant'] == 'NM_001165964.2:c.233_240del'
        assert results['NM_001165964.2:c.233_240del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001165964.2:c.233_240del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_001165964.2:c.233_240del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}
        assert results['NM_001165964.2:c.233_240del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_001165964.2:c.233_240del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}

        assert 'NM_001165963.2:c.233_240del' in results.keys()
        assert results['NM_001165963.2:c.233_240del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001165963.2:c.233_240del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001165963.2:c.233_240del']['alt_genomic_loci'] == []
        assert results['NM_001165963.2:c.233_240del']['gene_symbol'] == 'SCN1A'
        assert results['NM_001165963.2:c.233_240del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001159435.1:p.(Glu78GlyfsTer7)', 'slr': 'NP_001159435.1:p.(E78Gfs*7)'}
        assert results['NM_001165963.2:c.233_240del']['submitted_variant'] == '2-166929891-CCAGGTCCT-C'
        assert results['NM_001165963.2:c.233_240del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001165963.2:c.233_240del']['hgvs_lrg_variant'] == ''
        assert results['NM_001165963.2:c.233_240del']['hgvs_transcript_variant'] == 'NM_001165963.2:c.233_240del'
        assert results['NM_001165963.2:c.233_240del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001165963.2:c.233_240del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_001165963.2:c.233_240del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}
        assert results['NM_001165963.2:c.233_240del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_001165963.2:c.233_240del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}

        assert 'NM_001202435.1:c.233_240del' in results.keys()
        assert results['NM_001202435.1:c.233_240del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001202435.1:c.233_240del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001202435.1:c.233_240del']['alt_genomic_loci'] == []
        assert results['NM_001202435.1:c.233_240del']['gene_symbol'] == 'SCN1A'
        assert results['NM_001202435.1:c.233_240del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001189364.1:p.(Glu78GlyfsTer7)', 'slr': 'NP_001189364.1:p.(E78Gfs*7)'}
        assert results['NM_001202435.1:c.233_240del']['submitted_variant'] == '2-166929891-CCAGGTCCT-C'
        assert results['NM_001202435.1:c.233_240del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001202435.1:c.233_240del']['hgvs_lrg_variant'] == ''
        assert results['NM_001202435.1:c.233_240del']['hgvs_transcript_variant'] == 'NM_001202435.1:c.233_240del'
        assert results['NM_001202435.1:c.233_240del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001202435.1:c.233_240del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_001202435.1:c.233_240del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}
        assert results['NM_001202435.1:c.233_240del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_001202435.1:c.233_240del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}

        assert 'NM_001353951.1:c.233_240del' in results.keys()
        assert results['NM_001353951.1:c.233_240del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001353951.1:c.233_240del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001353951.1:c.233_240del']['alt_genomic_loci'] == []
        assert results['NM_001353951.1:c.233_240del']['gene_symbol'] == 'SCN1A'
        assert results['NM_001353951.1:c.233_240del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001340880.1:p.(Glu78GlyfsTer7)', 'slr': 'NP_001340880.1:p.(E78Gfs*7)'}
        assert results['NM_001353951.1:c.233_240del']['submitted_variant'] == '2-166929891-CCAGGTCCT-C'
        assert results['NM_001353951.1:c.233_240del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001353951.1:c.233_240del']['hgvs_lrg_variant'] == ''
        assert results['NM_001353951.1:c.233_240del']['hgvs_transcript_variant'] == 'NM_001353951.1:c.233_240del'
        assert results['NM_001353951.1:c.233_240del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001353951.1:c.233_240del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_001353951.1:c.233_240del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}
        assert results['NM_001353951.1:c.233_240del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_001353951.1:c.233_240del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}

        assert 'NM_001353952.1:c.233_240del' in results.keys()
        assert results['NM_001353952.1:c.233_240del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001353952.1:c.233_240del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001353952.1:c.233_240del']['alt_genomic_loci'] == []
        assert results['NM_001353952.1:c.233_240del']['gene_symbol'] == 'SCN1A'
        assert results['NM_001353952.1:c.233_240del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001340881.1:p.(Glu78GlyfsTer7)', 'slr': 'NP_001340881.1:p.(E78Gfs*7)'}
        assert results['NM_001353952.1:c.233_240del']['submitted_variant'] == '2-166929891-CCAGGTCCT-C'
        assert results['NM_001353952.1:c.233_240del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001353952.1:c.233_240del']['hgvs_lrg_variant'] == ''
        assert results['NM_001353952.1:c.233_240del']['hgvs_transcript_variant'] == 'NM_001353952.1:c.233_240del'
        assert results['NM_001353952.1:c.233_240del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001353952.1:c.233_240del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_001353952.1:c.233_240del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}
        assert results['NM_001353952.1:c.233_240del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_001353952.1:c.233_240del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}

        assert 'NM_001353957.1:c.233_240del' in results.keys()
        assert results['NM_001353957.1:c.233_240del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001353957.1:c.233_240del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001353957.1:c.233_240del']['alt_genomic_loci'] == []
        assert results['NM_001353957.1:c.233_240del']['gene_symbol'] == 'SCN1A'
        assert results['NM_001353957.1:c.233_240del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001340886.1:p.(Glu78GlyfsTer7)', 'slr': 'NP_001340886.1:p.(E78Gfs*7)'}
        assert results['NM_001353957.1:c.233_240del']['submitted_variant'] == '2-166929891-CCAGGTCCT-C'
        assert results['NM_001353957.1:c.233_240del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001353957.1:c.233_240del']['hgvs_lrg_variant'] == ''
        assert results['NM_001353957.1:c.233_240del']['hgvs_transcript_variant'] == 'NM_001353957.1:c.233_240del'
        assert results['NM_001353957.1:c.233_240del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001353957.1:c.233_240del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_001353957.1:c.233_240del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}
        assert results['NM_001353957.1:c.233_240del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_001353957.1:c.233_240del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}

        assert 'NM_001353954.1:c.233_240del' in results.keys()
        assert results['NM_001353954.1:c.233_240del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001353954.1:c.233_240del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001353954.1:c.233_240del']['alt_genomic_loci'] == []
        assert results['NM_001353954.1:c.233_240del']['gene_symbol'] == 'SCN1A'
        assert results['NM_001353954.1:c.233_240del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001340883.1:p.(Glu78GlyfsTer7)', 'slr': 'NP_001340883.1:p.(E78Gfs*7)'}
        assert results['NM_001353954.1:c.233_240del']['submitted_variant'] == '2-166929891-CCAGGTCCT-C'
        assert results['NM_001353954.1:c.233_240del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001353954.1:c.233_240del']['hgvs_lrg_variant'] == ''
        assert results['NM_001353954.1:c.233_240del']['hgvs_transcript_variant'] == 'NM_001353954.1:c.233_240del'
        assert results['NM_001353954.1:c.233_240del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001353954.1:c.233_240del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_001353954.1:c.233_240del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}
        assert results['NM_001353954.1:c.233_240del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_001353954.1:c.233_240del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}

        assert 'NM_001353948.1:c.233_240del' in results.keys()
        assert results['NM_001353948.1:c.233_240del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001353948.1:c.233_240del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001353948.1:c.233_240del']['alt_genomic_loci'] == []
        assert results['NM_001353948.1:c.233_240del']['gene_symbol'] == 'SCN1A'
        assert results['NM_001353948.1:c.233_240del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001340877.1:p.(Glu78GlyfsTer7)', 'slr': 'NP_001340877.1:p.(E78Gfs*7)'}
        assert results['NM_001353948.1:c.233_240del']['submitted_variant'] == '2-166929891-CCAGGTCCT-C'
        assert results['NM_001353948.1:c.233_240del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001353948.1:c.233_240del']['hgvs_lrg_variant'] == ''
        assert results['NM_001353948.1:c.233_240del']['hgvs_transcript_variant'] == 'NM_001353948.1:c.233_240del'
        assert results['NM_001353948.1:c.233_240del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001353948.1:c.233_240del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_001353948.1:c.233_240del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}
        assert results['NM_001353948.1:c.233_240del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_001353948.1:c.233_240del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}

        assert 'NM_001202435.2:c.233_240del' in results.keys()
        assert results['NM_001202435.2:c.233_240del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001202435.2:c.233_240del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001202435.2:c.233_240del']['alt_genomic_loci'] == []
        assert results['NM_001202435.2:c.233_240del']['gene_symbol'] == 'SCN1A'
        assert results['NM_001202435.2:c.233_240del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001189364.1:p.(Glu78GlyfsTer7)', 'slr': 'NP_001189364.1:p.(E78Gfs*7)'}
        assert results['NM_001202435.2:c.233_240del']['submitted_variant'] == '2-166929891-CCAGGTCCT-C'
        assert results['NM_001202435.2:c.233_240del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001202435.2:c.233_240del']['hgvs_lrg_variant'] == ''
        assert results['NM_001202435.2:c.233_240del']['hgvs_transcript_variant'] == 'NM_001202435.2:c.233_240del'
        assert results['NM_001202435.2:c.233_240del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001202435.2:c.233_240del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_001202435.2:c.233_240del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}
        assert results['NM_001202435.2:c.233_240del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_001202435.2:c.233_240del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001353961.1:c.-2193_-2186del' in results.keys()
        assert results['NM_001353961.1:c.-2193_-2186del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001353961.1:c.-2193_-2186del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001353961.1:c.-2193_-2186del']['alt_genomic_loci'] == []
        assert results['NM_001353961.1:c.-2193_-2186del']['gene_symbol'] == 'SCN1A'
        assert results['NM_001353961.1:c.-2193_-2186del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001340890.1:p.?', 'slr': 'NP_001340890.1:p.?'}
        assert results['NM_001353961.1:c.-2193_-2186del']['submitted_variant'] == '2-166929891-CCAGGTCCT-C'
        assert results['NM_001353961.1:c.-2193_-2186del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001353961.1:c.-2193_-2186del']['hgvs_lrg_variant'] == ''
        assert results['NM_001353961.1:c.-2193_-2186del']['hgvs_transcript_variant'] == 'NM_001353961.1:c.-2193_-2186del'
        assert results['NM_001353961.1:c.-2193_-2186del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001353961.1:c.-2193_-2186del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_001353961.1:c.-2193_-2186del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}
        assert results['NM_001353961.1:c.-2193_-2186del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_001353961.1:c.-2193_-2186del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}

        assert 'NM_001165963.1:c.233_240del' in results.keys()
        assert results['NM_001165963.1:c.233_240del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001165963.1:c.233_240del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001165963.1:c.233_240del']['alt_genomic_loci'] == []
        assert results['NM_001165963.1:c.233_240del']['gene_symbol'] == 'SCN1A'
        assert results['NM_001165963.1:c.233_240del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001159435.1:p.(Glu78GlyfsTer7)', 'slr': 'NP_001159435.1:p.(E78Gfs*7)'}
        assert results['NM_001165963.1:c.233_240del']['submitted_variant'] == '2-166929891-CCAGGTCCT-C'
        assert results['NM_001165963.1:c.233_240del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001165963.1:c.233_240del']['hgvs_lrg_variant'] == ''
        assert results['NM_001165963.1:c.233_240del']['hgvs_transcript_variant'] == 'NM_001165963.1:c.233_240del'
        assert results['NM_001165963.1:c.233_240del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001165963.1:c.233_240del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_001165963.1:c.233_240del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}
        assert results['NM_001165963.1:c.233_240del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_001165963.1:c.233_240del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}

        assert 'NM_001165964.1:c.233_240del' in results.keys()
        assert results['NM_001165964.1:c.233_240del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001165964.1:c.233_240del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001165964.1:c.233_240del']['alt_genomic_loci'] == []
        assert results['NM_001165964.1:c.233_240del']['gene_symbol'] == 'SCN1A'
        assert results['NM_001165964.1:c.233_240del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001159436.1:p.(Glu78GlyfsTer7)', 'slr': 'NP_001159436.1:p.(E78Gfs*7)'}
        assert results['NM_001165964.1:c.233_240del']['submitted_variant'] == '2-166929891-CCAGGTCCT-C'
        assert results['NM_001165964.1:c.233_240del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001165964.1:c.233_240del']['hgvs_lrg_variant'] == ''
        assert results['NM_001165964.1:c.233_240del']['hgvs_transcript_variant'] == 'NM_001165964.1:c.233_240del'
        assert results['NM_001165964.1:c.233_240del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001165964.1:c.233_240del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_001165964.1:c.233_240del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}
        assert results['NM_001165964.1:c.233_240del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_001165964.1:c.233_240del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}

        assert 'NM_006920.4:c.233_240del' in results.keys()
        assert results['NM_006920.4:c.233_240del']['hgvs_lrg_transcript_variant'] == 'LRG_8t1:c.233_240del'
        assert results['NM_006920.4:c.233_240del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_006920.4:c.233_240del']['alt_genomic_loci'] == []
        assert results['NM_006920.4:c.233_240del']['gene_symbol'] == 'SCN1A'
        assert results['NM_006920.4:c.233_240del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_008851.3(LRG_8p1):p.(Glu78GlyfsTer7)', 'slr': 'NP_008851.3:p.(E78Gfs*7)'}
        assert results['NM_006920.4:c.233_240del']['submitted_variant'] == '2-166929891-CCAGGTCCT-C'
        assert results['NM_006920.4:c.233_240del']['genome_context_intronic_sequence'] == ''
        assert results['NM_006920.4:c.233_240del']['hgvs_lrg_variant'] == 'LRG_8:g.5251_5258del'
        assert results['NM_006920.4:c.233_240del']['hgvs_transcript_variant'] == 'NM_006920.4:c.233_240del'
        assert results['NM_006920.4:c.233_240del']['hgvs_refseqgene_variant'] == 'NG_011906.1:g.5251_5258del'
        assert results['NM_006920.4:c.233_240del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_006920.4:c.233_240del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}
        assert results['NM_006920.4:c.233_240del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_006920.4:c.233_240del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}

        assert 'NM_001353958.1:c.233_240del' in results.keys()
        assert results['NM_001353958.1:c.233_240del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001353958.1:c.233_240del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001353958.1:c.233_240del']['alt_genomic_loci'] == []
        assert results['NM_001353958.1:c.233_240del']['gene_symbol'] == 'SCN1A'
        assert results['NM_001353958.1:c.233_240del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001340887.1:p.(Glu78GlyfsTer7)', 'slr': 'NP_001340887.1:p.(E78Gfs*7)'}
        assert results['NM_001353958.1:c.233_240del']['submitted_variant'] == '2-166929891-CCAGGTCCT-C'
        assert results['NM_001353958.1:c.233_240del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001353958.1:c.233_240del']['hgvs_lrg_variant'] == ''
        assert results['NM_001353958.1:c.233_240del']['hgvs_transcript_variant'] == 'NM_001353958.1:c.233_240del'
        assert results['NM_001353958.1:c.233_240del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001353958.1:c.233_240del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_001353958.1:c.233_240del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}
        assert results['NM_001353958.1:c.233_240del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_001353958.1:c.233_240del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}

        assert 'NR_148667.1:n.638_645del' in results.keys()
        assert results['NR_148667.1:n.638_645del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_148667.1:n.638_645del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_148667.1:n.638_645del']['alt_genomic_loci'] == []
        assert results['NR_148667.1:n.638_645del']['gene_symbol'] == 'SCN1A'
        assert results['NR_148667.1:n.638_645del']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_148667.1:n.638_645del']['submitted_variant'] == '2-166929891-CCAGGTCCT-C'
        assert results['NR_148667.1:n.638_645del']['genome_context_intronic_sequence'] == ''
        assert results['NR_148667.1:n.638_645del']['hgvs_lrg_variant'] == ''
        assert results['NR_148667.1:n.638_645del']['hgvs_transcript_variant'] == 'NR_148667.1:n.638_645del'
        assert results['NR_148667.1:n.638_645del']['hgvs_refseqgene_variant'] == ''
        assert results['NR_148667.1:n.638_645del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NR_148667.1:n.638_645del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}
        assert results['NR_148667.1:n.638_645del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NR_148667.1:n.638_645del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}

        assert 'NM_001353949.1:c.233_240del' in results.keys()
        assert results['NM_001353949.1:c.233_240del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001353949.1:c.233_240del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001353949.1:c.233_240del']['alt_genomic_loci'] == []
        assert results['NM_001353949.1:c.233_240del']['gene_symbol'] == 'SCN1A'
        assert results['NM_001353949.1:c.233_240del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001340878.1:p.(Glu78GlyfsTer7)', 'slr': 'NP_001340878.1:p.(E78Gfs*7)'}
        assert results['NM_001353949.1:c.233_240del']['submitted_variant'] == '2-166929891-CCAGGTCCT-C'
        assert results['NM_001353949.1:c.233_240del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001353949.1:c.233_240del']['hgvs_lrg_variant'] == ''
        assert results['NM_001353949.1:c.233_240del']['hgvs_transcript_variant'] == 'NM_001353949.1:c.233_240del'
        assert results['NM_001353949.1:c.233_240del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001353949.1:c.233_240del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_001353949.1:c.233_240del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': 'chr2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}
        assert results['NM_001353949.1:c.233_240del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.166929892_166929899del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166929891', 'alt': 'C'}}
        assert results['NM_001353949.1:c.233_240del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.166073382_166073389del', 'vcf': {'chr': '2', 'ref': 'CCAGGTCCT', 'pos': '166073381', 'alt': 'C'}}


    def test_variant274(self):
        variant = '2-179393504-G-T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001256850.1:c.102051C>A' in results.keys()
        assert results['NM_001256850.1:c.102051C>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001256850.1:c.102051C>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001256850.1:c.102051C>A']['alt_genomic_loci'] == []
        assert results['NM_001256850.1:c.102051C>A']['gene_symbol'] == 'TTN'
        assert results['NM_001256850.1:c.102051C>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001243779.1:p.(Ser34017Arg)', 'slr': 'NP_001243779.1:p.(S34017R)'}
        assert results['NM_001256850.1:c.102051C>A']['submitted_variant'] == '2-179393504-G-T'
        assert results['NM_001256850.1:c.102051C>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_001256850.1:c.102051C>A']['hgvs_lrg_variant'] == ''
        assert results['NM_001256850.1:c.102051C>A']['hgvs_transcript_variant'] == 'NM_001256850.1:c.102051C>A'
        assert results['NM_001256850.1:c.102051C>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001256850.1:c.102051C>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.179393504G>T', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '179393504', 'alt': u'T'}}
        assert results['NM_001256850.1:c.102051C>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.178528777G>T', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '178528777', 'alt': u'T'}}
        assert results['NM_001256850.1:c.102051C>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.179393504G>T', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '179393504', 'alt': u'T'}}
        assert results['NM_001256850.1:c.102051C>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.178528777G>T', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '178528777', 'alt': u'T'}}

        assert 'NR_038271.1:n.446+5141G>T' in results.keys()
        assert results['NR_038271.1:n.446+5141G>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_038271.1:n.446+5141G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_038271.1:n.446+5141G>T']['alt_genomic_loci'] == []
        assert results['NR_038271.1:n.446+5141G>T']['gene_symbol'] == 'TTN-AS1'
        assert results['NR_038271.1:n.446+5141G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_038271.1:n.446+5141G>T']['submitted_variant'] == '2-179393504-G-T'
        assert results['NR_038271.1:n.446+5141G>T']['genome_context_intronic_sequence'] == 'NC_000002.11(NR_038271.1):c.446+5141G>T'
        assert results['NR_038271.1:n.446+5141G>T']['hgvs_lrg_variant'] == ''
        assert results['NR_038271.1:n.446+5141G>T']['hgvs_transcript_variant'] == 'NR_038271.1:n.446+5141G>T'
        assert results['NR_038271.1:n.446+5141G>T']['hgvs_refseqgene_variant'] == ''
        assert results['NR_038271.1:n.446+5141G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.179393504G>T', 'vcf': {'chr': 'chr2', 'ref': 'G', 'pos': '179393504', 'alt': 'T'}}
        assert results['NR_038271.1:n.446+5141G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.178528777G>T', 'vcf': {'chr': 'chr2', 'ref': 'G', 'pos': '178528777', 'alt': 'T'}}
        assert results['NR_038271.1:n.446+5141G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.179393504G>T', 'vcf': {'chr': '2', 'ref': 'G', 'pos': '179393504', 'alt': 'T'}}
        assert results['NR_038271.1:n.446+5141G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.178528777G>T', 'vcf': {'chr': '2', 'ref': 'G', 'pos': '178528777', 'alt': 'T'}}

        assert 'NM_133432.3:c.80154C>A' in results.keys()
        assert results['NM_133432.3:c.80154C>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_133432.3:c.80154C>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_133432.3:c.80154C>A']['alt_genomic_loci'] == []
        assert results['NM_133432.3:c.80154C>A']['gene_symbol'] == 'TTN'
        assert results['NM_133432.3:c.80154C>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_597676.3:p.(Ser26718Arg)', 'slr': 'NP_597676.3:p.(S26718R)'}
        assert results['NM_133432.3:c.80154C>A']['submitted_variant'] == '2-179393504-G-T'
        assert results['NM_133432.3:c.80154C>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_133432.3:c.80154C>A']['hgvs_lrg_variant'] == ''
        assert results['NM_133432.3:c.80154C>A']['hgvs_transcript_variant'] == 'NM_133432.3:c.80154C>A'
        assert results['NM_133432.3:c.80154C>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_133432.3:c.80154C>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.179393504G>T', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '179393504', 'alt': u'T'}}
        assert results['NM_133432.3:c.80154C>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.178528777G>T', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '178528777', 'alt': u'T'}}
        assert results['NM_133432.3:c.80154C>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.179393504G>T', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '179393504', 'alt': u'T'}}
        assert results['NM_133432.3:c.80154C>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.178528777G>T', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '178528777', 'alt': u'T'}}

        assert 'NM_001267550.1:c.106974C>A' in results.keys()
        assert results['NM_001267550.1:c.106974C>A']['hgvs_lrg_transcript_variant'] == 'LRG_391t1:c.106974C>A'
        assert results['NM_001267550.1:c.106974C>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001267550.1:c.106974C>A']['alt_genomic_loci'] == []
        assert results['NM_001267550.1:c.106974C>A']['gene_symbol'] == 'TTN'
        assert results['NM_001267550.1:c.106974C>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001254479.1:p.(Ser35658Arg)', 'slr': 'NP_001254479.1:p.(S35658R)'}
        assert results['NM_001267550.1:c.106974C>A']['submitted_variant'] == '2-179393504-G-T'
        assert results['NM_001267550.1:c.106974C>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_001267550.1:c.106974C>A']['hgvs_lrg_variant'] == 'LRG_391:g.307026C>A'
        assert results['NM_001267550.1:c.106974C>A']['hgvs_transcript_variant'] == 'NM_001267550.1:c.106974C>A'
        assert results['NM_001267550.1:c.106974C>A']['hgvs_refseqgene_variant'] == 'NG_011618.3:g.307026C>A'
        assert results['NM_001267550.1:c.106974C>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.179393504G>T', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '179393504', 'alt': u'T'}}
        assert 'hg38' not in results['NM_001267550.1:c.106974C>A']['primary_assembly_loci'].keys()
        assert results['NM_001267550.1:c.106974C>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.179393504G>T', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '179393504', 'alt': u'T'}}
        assert 'grch38' not in results['NM_001267550.1:c.106974C>A']['primary_assembly_loci'].keys()

        assert 'NR_038272.1:n.219+5141G>T' in results.keys()
        assert results['NR_038272.1:n.219+5141G>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_038272.1:n.219+5141G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_038272.1:n.219+5141G>T']['alt_genomic_loci'] == []
        assert results['NR_038272.1:n.219+5141G>T']['gene_symbol'] == 'TTN-AS1'
        assert results['NR_038272.1:n.219+5141G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_038272.1:n.219+5141G>T']['submitted_variant'] == '2-179393504-G-T'
        assert results['NR_038272.1:n.219+5141G>T']['genome_context_intronic_sequence'] == 'NC_000002.11(NR_038272.1):c.219+5141G>T'
        assert results['NR_038272.1:n.219+5141G>T']['hgvs_lrg_variant'] == ''
        assert results['NR_038272.1:n.219+5141G>T']['hgvs_transcript_variant'] == 'NR_038272.1:n.219+5141G>T'
        assert results['NR_038272.1:n.219+5141G>T']['hgvs_refseqgene_variant'] == ''
        assert results['NR_038272.1:n.219+5141G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.179393504G>T', 'vcf': {'chr': 'chr2', 'ref': 'G', 'pos': '179393504', 'alt': 'T'}}
        assert results['NR_038272.1:n.219+5141G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.178528777G>T', 'vcf': {'chr': 'chr2', 'ref': 'G', 'pos': '178528777', 'alt': 'T'}}
        assert results['NR_038272.1:n.219+5141G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.179393504G>T', 'vcf': {'chr': '2', 'ref': 'G', 'pos': '179393504', 'alt': 'T'}}
        assert results['NR_038272.1:n.219+5141G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.178528777G>T', 'vcf': {'chr': '2', 'ref': 'G', 'pos': '178528777', 'alt': 'T'}}

        assert 'NM_133437.4:c.80355C>A' in results.keys()
        assert results['NM_133437.4:c.80355C>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_133437.4:c.80355C>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_133437.4:c.80355C>A']['alt_genomic_loci'] == []
        assert results['NM_133437.4:c.80355C>A']['gene_symbol'] == 'TTN'
        assert results['NM_133437.4:c.80355C>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_597681.4:p.(Ser26785Arg)', 'slr': 'NP_597681.4:p.(S26785R)'}
        assert results['NM_133437.4:c.80355C>A']['submitted_variant'] == '2-179393504-G-T'
        assert results['NM_133437.4:c.80355C>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_133437.4:c.80355C>A']['hgvs_lrg_variant'] == ''
        assert results['NM_133437.4:c.80355C>A']['hgvs_transcript_variant'] == 'NM_133437.4:c.80355C>A'
        assert results['NM_133437.4:c.80355C>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_133437.4:c.80355C>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.179393504G>T', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '179393504', 'alt': u'T'}}
        assert results['NM_133437.4:c.80355C>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.178528777G>T', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '178528777', 'alt': u'T'}}
        assert results['NM_133437.4:c.80355C>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.179393504G>T', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '179393504', 'alt': u'T'}}
        assert results['NM_133437.4:c.80355C>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.178528777G>T', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '178528777', 'alt': u'T'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_133378.4:c.99270C>A' in results.keys()
        assert results['NM_133378.4:c.99270C>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_133378.4:c.99270C>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_133378.4:c.99270C>A']['alt_genomic_loci'] == []
        assert results['NM_133378.4:c.99270C>A']['gene_symbol'] == 'TTN'
        assert results['NM_133378.4:c.99270C>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_596869.4:p.(Ser33090Arg)', 'slr': 'NP_596869.4:p.(S33090R)'}
        assert results['NM_133378.4:c.99270C>A']['submitted_variant'] == '2-179393504-G-T'
        assert results['NM_133378.4:c.99270C>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_133378.4:c.99270C>A']['hgvs_lrg_variant'] == ''
        assert results['NM_133378.4:c.99270C>A']['hgvs_transcript_variant'] == 'NM_133378.4:c.99270C>A'
        assert results['NM_133378.4:c.99270C>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_133378.4:c.99270C>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.179393504G>T', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '179393504', 'alt': u'T'}}
        assert results['NM_133378.4:c.99270C>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.178528777G>T', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '178528777', 'alt': u'T'}}
        assert results['NM_133378.4:c.99270C>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.179393504G>T', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '179393504', 'alt': u'T'}}
        assert results['NM_133378.4:c.99270C>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.178528777G>T', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '178528777', 'alt': u'T'}}

        assert 'NM_001267550.2:c.106974C>A' in results.keys()
        assert results['NM_001267550.2:c.106974C>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001267550.2:c.106974C>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001267550.2:c.106974C>A']['alt_genomic_loci'] == []
        assert results['NM_001267550.2:c.106974C>A']['gene_symbol'] == 'TTN'
        assert results['NM_001267550.2:c.106974C>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001254479.2:p.(Ser35658Arg)', 'slr': 'NP_001254479.2:p.(S35658R)'}
        assert results['NM_001267550.2:c.106974C>A']['submitted_variant'] == '2-179393504-G-T'
        assert results['NM_001267550.2:c.106974C>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_001267550.2:c.106974C>A']['hgvs_lrg_variant'] == ''
        assert results['NM_001267550.2:c.106974C>A']['hgvs_transcript_variant'] == 'NM_001267550.2:c.106974C>A'
        assert results['NM_001267550.2:c.106974C>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001267550.2:c.106974C>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.179393504G>T', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '179393504', 'alt': u'T'}}
        assert results['NM_001267550.2:c.106974C>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.178528777G>T', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '178528777', 'alt': u'T'}}
        assert results['NM_001267550.2:c.106974C>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.179393504G>T', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '179393504', 'alt': u'T'}}
        assert results['NM_001267550.2:c.106974C>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.178528777G>T', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '178528777', 'alt': u'T'}}

        assert 'NM_133437.3:c.80355C>A' in results.keys()
        assert results['NM_133437.3:c.80355C>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_133437.3:c.80355C>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_133437.3:c.80355C>A']['alt_genomic_loci'] == []
        assert results['NM_133437.3:c.80355C>A']['gene_symbol'] == 'TTN'
        assert results['NM_133437.3:c.80355C>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_597681.3:p.(Ser26785Arg)', 'slr': 'NP_597681.3:p.(S26785R)'}
        assert results['NM_133437.3:c.80355C>A']['submitted_variant'] == '2-179393504-G-T'
        assert results['NM_133437.3:c.80355C>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_133437.3:c.80355C>A']['hgvs_lrg_variant'] == ''
        assert results['NM_133437.3:c.80355C>A']['hgvs_transcript_variant'] == 'NM_133437.3:c.80355C>A'
        assert results['NM_133437.3:c.80355C>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_133437.3:c.80355C>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.179393504G>T', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '179393504', 'alt': u'T'}}
        assert 'hg38' not in results['NM_133437.3:c.80355C>A']['primary_assembly_loci'].keys()
        assert results['NM_133437.3:c.80355C>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.179393504G>T', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '179393504', 'alt': u'T'}}
        assert 'grch38' not in results['NM_133437.3:c.80355C>A']['primary_assembly_loci'].keys()

        assert 'NM_003319.4:c.79779C>A' in results.keys()
        assert results['NM_003319.4:c.79779C>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003319.4:c.79779C>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003319.4:c.79779C>A']['alt_genomic_loci'] == []
        assert results['NM_003319.4:c.79779C>A']['gene_symbol'] == 'TTN'
        assert results['NM_003319.4:c.79779C>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003310.4:p.(Ser26593Arg)', 'slr': 'NP_003310.4:p.(S26593R)'}
        assert results['NM_003319.4:c.79779C>A']['submitted_variant'] == '2-179393504-G-T'
        assert results['NM_003319.4:c.79779C>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_003319.4:c.79779C>A']['hgvs_lrg_variant'] == ''
        assert results['NM_003319.4:c.79779C>A']['hgvs_transcript_variant'] == 'NM_003319.4:c.79779C>A'
        assert results['NM_003319.4:c.79779C>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_003319.4:c.79779C>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.179393504G>T', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '179393504', 'alt': u'T'}}
        assert results['NM_003319.4:c.79779C>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.178528777G>T', 'vcf': {'chr': 'chr2', 'ref': u'G', 'pos': '178528777', 'alt': u'T'}}
        assert results['NM_003319.4:c.79779C>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.179393504G>T', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '179393504', 'alt': u'T'}}
        assert results['NM_003319.4:c.79779C>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.178528777G>T', 'vcf': {'chr': '2', 'ref': u'G', 'pos': '178528777', 'alt': u'T'}}


    def test_variant275(self):
        variant = '2-185803444-TGCAGCTGCTGCAGCTGCAGCTGCA-T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_194250.1:c.3324_3347del' in results.keys()
        assert results['NM_194250.1:c.3324_3347del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_194250.1:c.3324_3347del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_194250.1:c.3324_3347del']['alt_genomic_loci'] == []
        assert results['NM_194250.1:c.3324_3347del']['gene_symbol'] == 'ZNF804A'
        assert results['NM_194250.1:c.3324_3347del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_919226.1:p.(Ala1112_Ala1119del)', 'slr': 'NP_919226.1:p.(A1112_A1119del)'}
        assert results['NM_194250.1:c.3324_3347del']['submitted_variant'] == '2-185803444-TGCAGCTGCTGCAGCTGCAGCTGCA-T'
        assert results['NM_194250.1:c.3324_3347del']['genome_context_intronic_sequence'] == ''
        assert results['NM_194250.1:c.3324_3347del']['hgvs_lrg_variant'] == ''
        assert results['NM_194250.1:c.3324_3347del']['hgvs_transcript_variant'] == 'NM_194250.1:c.3324_3347del'
        assert results['NM_194250.1:c.3324_3347del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_194250.1:c.3324_3347del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.185803447_185803470del', 'vcf': {'chr': 'chr2', 'ref': 'TGCAGCTGCTGCAGCTGCAGCTGCA', 'pos': '185803444', 'alt': 'T'}}
        assert results['NM_194250.1:c.3324_3347del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.184938720_184938743del', 'vcf': {'chr': 'chr2', 'ref': 'TGCAGCTGCTGCAGCTGCAGCTGCA', 'pos': '184938717', 'alt': 'T'}}
        assert results['NM_194250.1:c.3324_3347del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.185803447_185803470del', 'vcf': {'chr': '2', 'ref': 'TGCAGCTGCTGCAGCTGCAGCTGCA', 'pos': '185803444', 'alt': 'T'}}
        assert results['NM_194250.1:c.3324_3347del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.184938720_184938743del', 'vcf': {'chr': '2', 'ref': 'TGCAGCTGCTGCAGCTGCAGCTGCA', 'pos': '184938717', 'alt': 'T'}}


    def test_variant276(self):
        variant = '2-201950249-G-T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_002491.2:c.208G>T' in results.keys()
        assert results['NM_002491.2:c.208G>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_002491.2:c.208G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_002491.2:c.208G>T']['alt_genomic_loci'] == []
        assert results['NM_002491.2:c.208G>T']['gene_symbol'] == 'NDUFB3'
        assert results['NM_002491.2:c.208G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_002482.1:p.(Gly70Ter)', 'slr': 'NP_002482.1:p.(G70*)'}
        assert results['NM_002491.2:c.208G>T']['submitted_variant'] == '2-201950249-G-T'
        assert results['NM_002491.2:c.208G>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_002491.2:c.208G>T']['hgvs_lrg_variant'] == ''
        assert results['NM_002491.2:c.208G>T']['hgvs_transcript_variant'] == 'NM_002491.2:c.208G>T'
        assert results['NM_002491.2:c.208G>T']['hgvs_refseqgene_variant'] == 'NG_032156.1:g.18788G>T'
        assert results['NM_002491.2:c.208G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.201950249G>T', 'vcf': {'chr': 'chr2', 'ref': 'G', 'pos': '201950249', 'alt': 'T'}}
        assert results['NM_002491.2:c.208G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.201085526G>T', 'vcf': {'chr': 'chr2', 'ref': 'G', 'pos': '201085526', 'alt': 'T'}}
        assert results['NM_002491.2:c.208G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.201950249G>T', 'vcf': {'chr': '2', 'ref': 'G', 'pos': '201950249', 'alt': 'T'}}
        assert results['NM_002491.2:c.208G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.201085526G>T', 'vcf': {'chr': '2', 'ref': 'G', 'pos': '201085526', 'alt': 'T'}}

        assert 'NM_001257102.1:c.208G>T' in results.keys()
        assert results['NM_001257102.1:c.208G>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001257102.1:c.208G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001257102.1:c.208G>T']['alt_genomic_loci'] == []
        assert results['NM_001257102.1:c.208G>T']['gene_symbol'] == 'NDUFB3'
        assert results['NM_001257102.1:c.208G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001244031.1:p.(Gly70Ter)', 'slr': 'NP_001244031.1:p.(G70*)'}
        assert results['NM_001257102.1:c.208G>T']['submitted_variant'] == '2-201950249-G-T'
        assert results['NM_001257102.1:c.208G>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001257102.1:c.208G>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001257102.1:c.208G>T']['hgvs_transcript_variant'] == 'NM_001257102.1:c.208G>T'
        assert results['NM_001257102.1:c.208G>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001257102.1:c.208G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.201950249G>T', 'vcf': {'chr': 'chr2', 'ref': 'G', 'pos': '201950249', 'alt': 'T'}}
        assert results['NM_001257102.1:c.208G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.201085526G>T', 'vcf': {'chr': 'chr2', 'ref': 'G', 'pos': '201085526', 'alt': 'T'}}
        assert results['NM_001257102.1:c.208G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.201950249G>T', 'vcf': {'chr': '2', 'ref': 'G', 'pos': '201950249', 'alt': 'T'}}
        assert results['NM_001257102.1:c.208G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.201085526G>T', 'vcf': {'chr': '2', 'ref': 'G', 'pos': '201085526', 'alt': 'T'}}


    def test_variant277(self):
        variant = '2-238268730-C-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_004369.3:c.6282+1G>T' in results.keys()
        assert results['NM_004369.3:c.6282+1G>T']['hgvs_lrg_transcript_variant'] == 'LRG_473t1:c.6282+1G>T'
        assert results['NM_004369.3:c.6282+1G>T']['refseqgene_context_intronic_sequence'] == 'NG_008676.1(NM_004369.3):c.6282+1G>T'
        assert results['NM_004369.3:c.6282+1G>T']['alt_genomic_loci'] == []
        assert results['NM_004369.3:c.6282+1G>T']['gene_symbol'] == 'COL6A3'
        assert results['NM_004369.3:c.6282+1G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_004360.2(LRG_473p1):p.?', 'slr': 'NP_004360.2:p.?'}
        assert results['NM_004369.3:c.6282+1G>T']['submitted_variant'] == '2-238268730-C-A'
        assert results['NM_004369.3:c.6282+1G>T']['genome_context_intronic_sequence'] == 'NC_000002.11(NM_004369.3):c.6282+1G>T'
        assert results['NM_004369.3:c.6282+1G>T']['hgvs_lrg_variant'] == 'LRG_473:g.59121G>T'
        assert results['NM_004369.3:c.6282+1G>T']['hgvs_transcript_variant'] == 'NM_004369.3:c.6282+1G>T'
        assert results['NM_004369.3:c.6282+1G>T']['hgvs_refseqgene_variant'] == 'NG_008676.1:g.59121G>T'
        assert results['NM_004369.3:c.6282+1G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.238268730C>A', 'vcf': {'chr': 'chr2', 'ref': u'C', 'pos': '238268730', 'alt': u'A'}}
        assert results['NM_004369.3:c.6282+1G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.237360087C>A', 'vcf': {'chr': 'chr2', 'ref': u'C', 'pos': '237360087', 'alt': u'A'}}
        assert results['NM_004369.3:c.6282+1G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.238268730C>A', 'vcf': {'chr': '2', 'ref': u'C', 'pos': '238268730', 'alt': u'A'}}
        assert results['NM_004369.3:c.6282+1G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.237360087C>A', 'vcf': {'chr': '2', 'ref': u'C', 'pos': '237360087', 'alt': u'A'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_057166.4:c.4461+1G>T' in results.keys()
        assert results['NM_057166.4:c.4461+1G>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_057166.4:c.4461+1G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_057166.4:c.4461+1G>T']['alt_genomic_loci'] == []
        assert results['NM_057166.4:c.4461+1G>T']['gene_symbol'] == 'COL6A3'
        assert results['NM_057166.4:c.4461+1G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_476507.3:p.?', 'slr': 'NP_476507.3:p.?'}
        assert results['NM_057166.4:c.4461+1G>T']['submitted_variant'] == '2-238268730-C-A'
        assert results['NM_057166.4:c.4461+1G>T']['genome_context_intronic_sequence'] == 'NC_000002.11(NM_057166.4):c.4461+1G>T'
        assert results['NM_057166.4:c.4461+1G>T']['hgvs_lrg_variant'] == ''
        assert results['NM_057166.4:c.4461+1G>T']['hgvs_transcript_variant'] == 'NM_057166.4:c.4461+1G>T'
        assert results['NM_057166.4:c.4461+1G>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_057166.4:c.4461+1G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.238268730C>A', 'vcf': {'chr': 'chr2', 'ref': u'C', 'pos': '238268730', 'alt': u'A'}}
        assert results['NM_057166.4:c.4461+1G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.237360087C>A', 'vcf': {'chr': 'chr2', 'ref': u'C', 'pos': '237360087', 'alt': u'A'}}
        assert results['NM_057166.4:c.4461+1G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.238268730C>A', 'vcf': {'chr': '2', 'ref': u'C', 'pos': '238268730', 'alt': u'A'}}
        assert results['NM_057166.4:c.4461+1G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.237360087C>A', 'vcf': {'chr': '2', 'ref': u'C', 'pos': '237360087', 'alt': u'A'}}

        assert 'NM_057167.3:c.5664+1G>T' in results.keys()
        assert results['NM_057167.3:c.5664+1G>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_057167.3:c.5664+1G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_057167.3:c.5664+1G>T']['alt_genomic_loci'] == []
        assert results['NM_057167.3:c.5664+1G>T']['gene_symbol'] == 'COL6A3'
        assert results['NM_057167.3:c.5664+1G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_476508.2:p.?', 'slr': 'NP_476508.2:p.?'}
        assert results['NM_057167.3:c.5664+1G>T']['submitted_variant'] == '2-238268730-C-A'
        assert results['NM_057167.3:c.5664+1G>T']['genome_context_intronic_sequence'] == 'NC_000002.11(NM_057167.3):c.5664+1G>T'
        assert results['NM_057167.3:c.5664+1G>T']['hgvs_lrg_variant'] == ''
        assert results['NM_057167.3:c.5664+1G>T']['hgvs_transcript_variant'] == 'NM_057167.3:c.5664+1G>T'
        assert results['NM_057167.3:c.5664+1G>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_057167.3:c.5664+1G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.238268730C>A', 'vcf': {'chr': 'chr2', 'ref': u'C', 'pos': '238268730', 'alt': u'A'}}
        assert results['NM_057167.3:c.5664+1G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.237360087C>A', 'vcf': {'chr': 'chr2', 'ref': u'C', 'pos': '237360087', 'alt': u'A'}}
        assert results['NM_057167.3:c.5664+1G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.238268730C>A', 'vcf': {'chr': '2', 'ref': u'C', 'pos': '238268730', 'alt': u'A'}}
        assert results['NM_057167.3:c.5664+1G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.237360087C>A', 'vcf': {'chr': '2', 'ref': u'C', 'pos': '237360087', 'alt': u'A'}}


    def test_variant278(self):
        variant = '21-43897396-C-T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_080860.2:c.727+5G>A' in results.keys()
        assert results['NM_080860.2:c.727+5G>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_080860.2:c.727+5G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_080860.2:c.727+5G>A']['alt_genomic_loci'] == []
        assert results['NM_080860.2:c.727+5G>A']['gene_symbol'] == 'RSPH1'
        assert results['NM_080860.2:c.727+5G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_543136.1:p.?', 'slr': 'NP_543136.1:p.?'}
        assert results['NM_080860.2:c.727+5G>A']['submitted_variant'] == '21-43897396-C-T'
        assert results['NM_080860.2:c.727+5G>A']['genome_context_intronic_sequence'] == 'NC_000021.8(NM_080860.2):c.727+5G>A'
        assert results['NM_080860.2:c.727+5G>A']['hgvs_lrg_variant'] == ''
        assert results['NM_080860.2:c.727+5G>A']['hgvs_transcript_variant'] == 'NM_080860.2:c.727+5G>A'
        assert results['NM_080860.2:c.727+5G>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_080860.2:c.727+5G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000021.8:g.43897396C>T', 'vcf': {'chr': 'chr21', 'ref': u'C', 'pos': '43897396', 'alt': u'T'}}
        assert 'hg38' not in results['NM_080860.2:c.727+5G>A']['primary_assembly_loci'].keys()
        assert results['NM_080860.2:c.727+5G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000021.8:g.43897396C>T', 'vcf': {'chr': '21', 'ref': u'C', 'pos': '43897396', 'alt': u'T'}}
        assert 'grch38' not in results['NM_080860.2:c.727+5G>A']['primary_assembly_loci'].keys()

        assert results['flag'] == 'gene_variant'
        assert 'NM_080860.3:c.727+5G>A' in results.keys()
        assert results['NM_080860.3:c.727+5G>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_080860.3:c.727+5G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_080860.3:c.727+5G>A']['alt_genomic_loci'] == []
        assert results['NM_080860.3:c.727+5G>A']['gene_symbol'] == 'RSPH1'
        assert results['NM_080860.3:c.727+5G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_543136.1:p.?', 'slr': 'NP_543136.1:p.?'}
        assert results['NM_080860.3:c.727+5G>A']['submitted_variant'] == '21-43897396-C-T'
        assert results['NM_080860.3:c.727+5G>A']['genome_context_intronic_sequence'] == 'NC_000021.8(NM_080860.3):c.727+5G>A'
        assert results['NM_080860.3:c.727+5G>A']['hgvs_lrg_variant'] == ''
        assert results['NM_080860.3:c.727+5G>A']['hgvs_transcript_variant'] == 'NM_080860.3:c.727+5G>A'
        assert results['NM_080860.3:c.727+5G>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_080860.3:c.727+5G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000021.8:g.43897396C>T', 'vcf': {'chr': 'chr21', 'ref': u'C', 'pos': '43897396', 'alt': u'T'}}
        assert results['NM_080860.3:c.727+5G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000021.9:g.42477286C>T', 'vcf': {'chr': 'chr21', 'ref': u'C', 'pos': '42477286', 'alt': u'T'}}
        assert results['NM_080860.3:c.727+5G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000021.8:g.43897396C>T', 'vcf': {'chr': '21', 'ref': u'C', 'pos': '43897396', 'alt': u'T'}}
        assert results['NM_080860.3:c.727+5G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000021.9:g.42477286C>T', 'vcf': {'chr': '21', 'ref': u'C', 'pos': '42477286', 'alt': u'T'}}

        assert 'NM_001286506.1:c.613+5G>A' in results.keys()
        assert results['NM_001286506.1:c.613+5G>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001286506.1:c.613+5G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001286506.1:c.613+5G>A']['alt_genomic_loci'] == []
        assert results['NM_001286506.1:c.613+5G>A']['gene_symbol'] == 'RSPH1'
        assert results['NM_001286506.1:c.613+5G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001273435.1:p.?', 'slr': 'NP_001273435.1:p.?'}
        assert results['NM_001286506.1:c.613+5G>A']['submitted_variant'] == '21-43897396-C-T'
        assert results['NM_001286506.1:c.613+5G>A']['genome_context_intronic_sequence'] == 'NC_000021.8(NM_001286506.1):c.613+5G>A'
        assert results['NM_001286506.1:c.613+5G>A']['hgvs_lrg_variant'] == ''
        assert results['NM_001286506.1:c.613+5G>A']['hgvs_transcript_variant'] == 'NM_001286506.1:c.613+5G>A'
        assert results['NM_001286506.1:c.613+5G>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001286506.1:c.613+5G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000021.8:g.43897396C>T', 'vcf': {'chr': 'chr21', 'ref': u'C', 'pos': '43897396', 'alt': u'T'}}
        assert results['NM_001286506.1:c.613+5G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000021.9:g.42477286C>T', 'vcf': {'chr': 'chr21', 'ref': u'C', 'pos': '42477286', 'alt': u'T'}}
        assert results['NM_001286506.1:c.613+5G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000021.8:g.43897396C>T', 'vcf': {'chr': '21', 'ref': u'C', 'pos': '43897396', 'alt': u'T'}}
        assert results['NM_001286506.1:c.613+5G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000021.9:g.42477286C>T', 'vcf': {'chr': '21', 'ref': u'C', 'pos': '42477286', 'alt': u'T'}}


    def test_variant279(self):
        variant = '22-30064360-G-GCGACGC'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_000268.3:c.924_925insCGACGC' in results.keys()
        assert results['NM_000268.3:c.924_925insCGACGC']['hgvs_lrg_transcript_variant'] == 'LRG_511t1:c.924_925insCGACGC'
        assert results['NM_000268.3:c.924_925insCGACGC']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000268.3:c.924_925insCGACGC']['alt_genomic_loci'] == []
        assert results['NM_000268.3:c.924_925insCGACGC']['gene_symbol'] == 'NF2'
        assert results['NM_000268.3:c.924_925insCGACGC']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000259.1(LRG_511p1):p.(Arg310_Arg311dup)', 'slr': 'NP_000259.1:p.(R310_R311dup)'}
        assert results['NM_000268.3:c.924_925insCGACGC']['submitted_variant'] == '22-30064360-G-GCGACGC'
        assert results['NM_000268.3:c.924_925insCGACGC']['genome_context_intronic_sequence'] == ''
        assert results['NM_000268.3:c.924_925insCGACGC']['hgvs_lrg_variant'] == 'LRG_511:g.69816_69817insCGACGC'
        assert results['NM_000268.3:c.924_925insCGACGC']['hgvs_transcript_variant'] == 'NM_000268.3:c.924_925insCGACGC'
        assert results['NM_000268.3:c.924_925insCGACGC']['hgvs_refseqgene_variant'] == 'NG_009057.1:g.69816_69817insCGACGC'
        assert results['NM_000268.3:c.924_925insCGACGC']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000022.10:g.30064360_30064361insCGACGC', 'vcf': {'chr': 'chr22', 'ref': 'G', 'pos': '30064360', 'alt': 'GCGACGC'}}
        assert results['NM_000268.3:c.924_925insCGACGC']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000022.11:g.29668371_29668372insCGACGC', 'vcf': {'chr': 'chr22', 'ref': 'G', 'pos': '29668371', 'alt': 'GCGACGC'}}
        assert results['NM_000268.3:c.924_925insCGACGC']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000022.10:g.30064360_30064361insCGACGC', 'vcf': {'chr': '22', 'ref': 'G', 'pos': '30064360', 'alt': 'GCGACGC'}}
        assert results['NM_000268.3:c.924_925insCGACGC']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000022.11:g.29668371_29668372insCGACGC', 'vcf': {'chr': '22', 'ref': 'G', 'pos': '29668371', 'alt': 'GCGACGC'}}

        assert 'NM_181828.2:c.798_799insCGACGC' in results.keys()
        assert results['NM_181828.2:c.798_799insCGACGC']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_181828.2:c.798_799insCGACGC']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_181828.2:c.798_799insCGACGC']['alt_genomic_loci'] == []
        assert results['NM_181828.2:c.798_799insCGACGC']['gene_symbol'] == 'NF2'
        assert results['NM_181828.2:c.798_799insCGACGC']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_861966.1:p.(Arg268_Arg269dup)', 'slr': 'NP_861966.1:p.(R268_R269dup)'}
        assert results['NM_181828.2:c.798_799insCGACGC']['submitted_variant'] == '22-30064360-G-GCGACGC'
        assert results['NM_181828.2:c.798_799insCGACGC']['genome_context_intronic_sequence'] == ''
        assert results['NM_181828.2:c.798_799insCGACGC']['hgvs_lrg_variant'] == ''
        assert results['NM_181828.2:c.798_799insCGACGC']['hgvs_transcript_variant'] == 'NM_181828.2:c.798_799insCGACGC'
        assert results['NM_181828.2:c.798_799insCGACGC']['hgvs_refseqgene_variant'] == ''
        assert results['NM_181828.2:c.798_799insCGACGC']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000022.10:g.30064360_30064361insCGACGC', 'vcf': {'chr': 'chr22', 'ref': 'G', 'pos': '30064360', 'alt': 'GCGACGC'}}
        assert results['NM_181828.2:c.798_799insCGACGC']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000022.11:g.29668371_29668372insCGACGC', 'vcf': {'chr': 'chr22', 'ref': 'G', 'pos': '29668371', 'alt': 'GCGACGC'}}
        assert results['NM_181828.2:c.798_799insCGACGC']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000022.10:g.30064360_30064361insCGACGC', 'vcf': {'chr': '22', 'ref': 'G', 'pos': '30064360', 'alt': 'GCGACGC'}}
        assert results['NM_181828.2:c.798_799insCGACGC']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000022.11:g.29668371_29668372insCGACGC', 'vcf': {'chr': '22', 'ref': 'G', 'pos': '29668371', 'alt': 'GCGACGC'}}

        assert 'NM_181830.2:c.675_676insCGACGC' in results.keys()
        assert results['NM_181830.2:c.675_676insCGACGC']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_181830.2:c.675_676insCGACGC']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_181830.2:c.675_676insCGACGC']['alt_genomic_loci'] == []
        assert results['NM_181830.2:c.675_676insCGACGC']['gene_symbol'] == 'NF2'
        assert results['NM_181830.2:c.675_676insCGACGC']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_861968.1:p.(Arg227_Arg228dup)', 'slr': 'NP_861968.1:p.(R227_R228dup)'}
        assert results['NM_181830.2:c.675_676insCGACGC']['submitted_variant'] == '22-30064360-G-GCGACGC'
        assert results['NM_181830.2:c.675_676insCGACGC']['genome_context_intronic_sequence'] == ''
        assert results['NM_181830.2:c.675_676insCGACGC']['hgvs_lrg_variant'] == ''
        assert results['NM_181830.2:c.675_676insCGACGC']['hgvs_transcript_variant'] == 'NM_181830.2:c.675_676insCGACGC'
        assert results['NM_181830.2:c.675_676insCGACGC']['hgvs_refseqgene_variant'] == ''
        assert results['NM_181830.2:c.675_676insCGACGC']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000022.10:g.30064360_30064361insCGACGC', 'vcf': {'chr': 'chr22', 'ref': 'G', 'pos': '30064360', 'alt': 'GCGACGC'}}
        assert results['NM_181830.2:c.675_676insCGACGC']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000022.11:g.29668371_29668372insCGACGC', 'vcf': {'chr': 'chr22', 'ref': 'G', 'pos': '29668371', 'alt': 'GCGACGC'}}
        assert results['NM_181830.2:c.675_676insCGACGC']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000022.10:g.30064360_30064361insCGACGC', 'vcf': {'chr': '22', 'ref': 'G', 'pos': '30064360', 'alt': 'GCGACGC'}}
        assert results['NM_181830.2:c.675_676insCGACGC']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000022.11:g.29668371_29668372insCGACGC', 'vcf': {'chr': '22', 'ref': 'G', 'pos': '29668371', 'alt': 'GCGACGC'}}

        assert 'NM_181825.2:c.924_925insCGACGC' in results.keys()
        assert results['NM_181825.2:c.924_925insCGACGC']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_181825.2:c.924_925insCGACGC']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_181825.2:c.924_925insCGACGC']['alt_genomic_loci'] == []
        assert results['NM_181825.2:c.924_925insCGACGC']['gene_symbol'] == 'NF2'
        assert results['NM_181825.2:c.924_925insCGACGC']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_861546.1:p.(Arg310_Arg311dup)', 'slr': 'NP_861546.1:p.(R310_R311dup)'}
        assert results['NM_181825.2:c.924_925insCGACGC']['submitted_variant'] == '22-30064360-G-GCGACGC'
        assert results['NM_181825.2:c.924_925insCGACGC']['genome_context_intronic_sequence'] == ''
        assert results['NM_181825.2:c.924_925insCGACGC']['hgvs_lrg_variant'] == ''
        assert results['NM_181825.2:c.924_925insCGACGC']['hgvs_transcript_variant'] == 'NM_181825.2:c.924_925insCGACGC'
        assert results['NM_181825.2:c.924_925insCGACGC']['hgvs_refseqgene_variant'] == ''
        assert results['NM_181825.2:c.924_925insCGACGC']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000022.10:g.30064360_30064361insCGACGC', 'vcf': {'chr': 'chr22', 'ref': 'G', 'pos': '30064360', 'alt': 'GCGACGC'}}
        assert results['NM_181825.2:c.924_925insCGACGC']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000022.11:g.29668371_29668372insCGACGC', 'vcf': {'chr': 'chr22', 'ref': 'G', 'pos': '29668371', 'alt': 'GCGACGC'}}
        assert results['NM_181825.2:c.924_925insCGACGC']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000022.10:g.30064360_30064361insCGACGC', 'vcf': {'chr': '22', 'ref': 'G', 'pos': '30064360', 'alt': 'GCGACGC'}}
        assert results['NM_181825.2:c.924_925insCGACGC']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000022.11:g.29668371_29668372insCGACGC', 'vcf': {'chr': '22', 'ref': 'G', 'pos': '29668371', 'alt': 'GCGACGC'}}

        assert 'NM_016418.5:c.924_925insCGACGC' in results.keys()
        assert results['NM_016418.5:c.924_925insCGACGC']['hgvs_lrg_transcript_variant'] == 'LRG_511t2:c.924_925insCGACGC'
        assert results['NM_016418.5:c.924_925insCGACGC']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_016418.5:c.924_925insCGACGC']['alt_genomic_loci'] == []
        assert results['NM_016418.5:c.924_925insCGACGC']['gene_symbol'] == 'NF2'
        assert results['NM_016418.5:c.924_925insCGACGC']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_057502.2(LRG_511p2):p.(Arg310_Arg311dup)', 'slr': 'NP_057502.2:p.(R310_R311dup)'}
        assert results['NM_016418.5:c.924_925insCGACGC']['submitted_variant'] == '22-30064360-G-GCGACGC'
        assert results['NM_016418.5:c.924_925insCGACGC']['genome_context_intronic_sequence'] == ''
        assert results['NM_016418.5:c.924_925insCGACGC']['hgvs_lrg_variant'] == 'LRG_511:g.69816_69817insCGACGC'
        assert results['NM_016418.5:c.924_925insCGACGC']['hgvs_transcript_variant'] == 'NM_016418.5:c.924_925insCGACGC'
        assert results['NM_016418.5:c.924_925insCGACGC']['hgvs_refseqgene_variant'] == 'NG_009057.1:g.69816_69817insCGACGC'
        assert results['NM_016418.5:c.924_925insCGACGC']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000022.10:g.30064360_30064361insCGACGC', 'vcf': {'chr': 'chr22', 'ref': 'G', 'pos': '30064360', 'alt': 'GCGACGC'}}
        assert results['NM_016418.5:c.924_925insCGACGC']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000022.11:g.29668371_29668372insCGACGC', 'vcf': {'chr': 'chr22', 'ref': 'G', 'pos': '29668371', 'alt': 'GCGACGC'}}
        assert results['NM_016418.5:c.924_925insCGACGC']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000022.10:g.30064360_30064361insCGACGC', 'vcf': {'chr': '22', 'ref': 'G', 'pos': '30064360', 'alt': 'GCGACGC'}}
        assert results['NM_016418.5:c.924_925insCGACGC']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000022.11:g.29668371_29668372insCGACGC', 'vcf': {'chr': '22', 'ref': 'G', 'pos': '29668371', 'alt': 'GCGACGC'}}

        assert 'NR_156186.1:n.1483_1484insCGACGC' in results.keys()
        assert results['NR_156186.1:n.1483_1484insCGACGC']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_156186.1:n.1483_1484insCGACGC']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_156186.1:n.1483_1484insCGACGC']['alt_genomic_loci'] == []
        assert results['NR_156186.1:n.1483_1484insCGACGC']['gene_symbol'] == 'NF2'
        assert results['NR_156186.1:n.1483_1484insCGACGC']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_156186.1:n.1483_1484insCGACGC']['submitted_variant'] == '22-30064360-G-GCGACGC'
        assert results['NR_156186.1:n.1483_1484insCGACGC']['genome_context_intronic_sequence'] == ''
        assert results['NR_156186.1:n.1483_1484insCGACGC']['hgvs_lrg_variant'] == ''
        assert results['NR_156186.1:n.1483_1484insCGACGC']['hgvs_transcript_variant'] == 'NR_156186.1:n.1483_1484insCGACGC'
        assert results['NR_156186.1:n.1483_1484insCGACGC']['hgvs_refseqgene_variant'] == ''
        assert results['NR_156186.1:n.1483_1484insCGACGC']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000022.10:g.30064360_30064361insCGACGC', 'vcf': {'chr': 'chr22', 'ref': 'G', 'pos': '30064360', 'alt': 'GCGACGC'}}
        assert 'hg38' not in results['NR_156186.1:n.1483_1484insCGACGC']['primary_assembly_loci'].keys()
        assert results['NR_156186.1:n.1483_1484insCGACGC']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000022.10:g.30064360_30064361insCGACGC', 'vcf': {'chr': '22', 'ref': 'G', 'pos': '30064360', 'alt': 'GCGACGC'}}
        assert 'grch38' not in results['NR_156186.1:n.1483_1484insCGACGC']['primary_assembly_loci'].keys()

        assert 'NM_181832.2:c.924_925insCGACGC' in results.keys()
        assert results['NM_181832.2:c.924_925insCGACGC']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_181832.2:c.924_925insCGACGC']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_181832.2:c.924_925insCGACGC']['alt_genomic_loci'] == []
        assert results['NM_181832.2:c.924_925insCGACGC']['gene_symbol'] == 'NF2'
        assert results['NM_181832.2:c.924_925insCGACGC']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_861970.1:p.(Arg310_Arg311dup)', 'slr': 'NP_861970.1:p.(R310_R311dup)'}
        assert results['NM_181832.2:c.924_925insCGACGC']['submitted_variant'] == '22-30064360-G-GCGACGC'
        assert results['NM_181832.2:c.924_925insCGACGC']['genome_context_intronic_sequence'] == ''
        assert results['NM_181832.2:c.924_925insCGACGC']['hgvs_lrg_variant'] == ''
        assert results['NM_181832.2:c.924_925insCGACGC']['hgvs_transcript_variant'] == 'NM_181832.2:c.924_925insCGACGC'
        assert results['NM_181832.2:c.924_925insCGACGC']['hgvs_refseqgene_variant'] == ''
        assert results['NM_181832.2:c.924_925insCGACGC']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000022.10:g.30064360_30064361insCGACGC', 'vcf': {'chr': 'chr22', 'ref': 'G', 'pos': '30064360', 'alt': 'GCGACGC'}}
        assert results['NM_181832.2:c.924_925insCGACGC']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000022.11:g.29668371_29668372insCGACGC', 'vcf': {'chr': 'chr22', 'ref': 'G', 'pos': '29668371', 'alt': 'GCGACGC'}}
        assert results['NM_181832.2:c.924_925insCGACGC']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000022.10:g.30064360_30064361insCGACGC', 'vcf': {'chr': '22', 'ref': 'G', 'pos': '30064360', 'alt': 'GCGACGC'}}
        assert results['NM_181832.2:c.924_925insCGACGC']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000022.11:g.29668371_29668372insCGACGC', 'vcf': {'chr': '22', 'ref': 'G', 'pos': '29668371', 'alt': 'GCGACGC'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_181829.2:c.801_802insCGACGC' in results.keys()
        assert results['NM_181829.2:c.801_802insCGACGC']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_181829.2:c.801_802insCGACGC']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_181829.2:c.801_802insCGACGC']['alt_genomic_loci'] == []
        assert results['NM_181829.2:c.801_802insCGACGC']['gene_symbol'] == 'NF2'
        assert results['NM_181829.2:c.801_802insCGACGC']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_861967.1:p.(Arg269_Arg270dup)', 'slr': 'NP_861967.1:p.(R269_R270dup)'}
        assert results['NM_181829.2:c.801_802insCGACGC']['submitted_variant'] == '22-30064360-G-GCGACGC'
        assert results['NM_181829.2:c.801_802insCGACGC']['genome_context_intronic_sequence'] == ''
        assert results['NM_181829.2:c.801_802insCGACGC']['hgvs_lrg_variant'] == ''
        assert results['NM_181829.2:c.801_802insCGACGC']['hgvs_transcript_variant'] == 'NM_181829.2:c.801_802insCGACGC'
        assert results['NM_181829.2:c.801_802insCGACGC']['hgvs_refseqgene_variant'] == ''
        assert results['NM_181829.2:c.801_802insCGACGC']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000022.10:g.30064360_30064361insCGACGC', 'vcf': {'chr': 'chr22', 'ref': 'G', 'pos': '30064360', 'alt': 'GCGACGC'}}
        assert results['NM_181829.2:c.801_802insCGACGC']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000022.11:g.29668371_29668372insCGACGC', 'vcf': {'chr': 'chr22', 'ref': 'G', 'pos': '29668371', 'alt': 'GCGACGC'}}
        assert results['NM_181829.2:c.801_802insCGACGC']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000022.10:g.30064360_30064361insCGACGC', 'vcf': {'chr': '22', 'ref': 'G', 'pos': '30064360', 'alt': 'GCGACGC'}}
        assert results['NM_181829.2:c.801_802insCGACGC']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000022.11:g.29668371_29668372insCGACGC', 'vcf': {'chr': '22', 'ref': 'G', 'pos': '29668371', 'alt': 'GCGACGC'}}

        assert 'NM_181833.2:c.447+26086_447+26087insCGACGC' in results.keys()
        assert results['NM_181833.2:c.447+26086_447+26087insCGACGC']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_181833.2:c.447+26086_447+26087insCGACGC']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_181833.2:c.447+26086_447+26087insCGACGC']['alt_genomic_loci'] == []
        assert results['NM_181833.2:c.447+26086_447+26087insCGACGC']['gene_symbol'] == 'NF2'
        assert results['NM_181833.2:c.447+26086_447+26087insCGACGC']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_861971.1:p.?', 'slr': 'NP_861971.1:p.?'}
        assert results['NM_181833.2:c.447+26086_447+26087insCGACGC']['submitted_variant'] == '22-30064360-G-GCGACGC'
        assert results['NM_181833.2:c.447+26086_447+26087insCGACGC']['genome_context_intronic_sequence'] == 'NC_000022.10(NM_181833.2):c.447+26086_447+26087insCGACGC'
        assert results['NM_181833.2:c.447+26086_447+26087insCGACGC']['hgvs_lrg_variant'] == ''
        assert results['NM_181833.2:c.447+26086_447+26087insCGACGC']['hgvs_transcript_variant'] == 'NM_181833.2:c.447+26086_447+26087insCGACGC'
        assert results['NM_181833.2:c.447+26086_447+26087insCGACGC']['hgvs_refseqgene_variant'] == ''
        assert results['NM_181833.2:c.447+26086_447+26087insCGACGC']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000022.10:g.30064360_30064361insCGACGC', 'vcf': {'chr': 'chr22', 'ref': 'G', 'pos': '30064360', 'alt': 'GCGACGC'}}
        assert results['NM_181833.2:c.447+26086_447+26087insCGACGC']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000022.11:g.29668371_29668372insCGACGC', 'vcf': {'chr': 'chr22', 'ref': 'G', 'pos': '29668371', 'alt': 'GCGACGC'}}
        assert results['NM_181833.2:c.447+26086_447+26087insCGACGC']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000022.10:g.30064360_30064361insCGACGC', 'vcf': {'chr': '22', 'ref': 'G', 'pos': '30064360', 'alt': 'GCGACGC'}}
        assert results['NM_181833.2:c.447+26086_447+26087insCGACGC']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000022.11:g.29668371_29668372insCGACGC', 'vcf': {'chr': '22', 'ref': 'G', 'pos': '29668371', 'alt': 'GCGACGC'}}

        assert 'NM_181831.2:c.675_676insCGACGC' in results.keys()
        assert results['NM_181831.2:c.675_676insCGACGC']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_181831.2:c.675_676insCGACGC']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_181831.2:c.675_676insCGACGC']['alt_genomic_loci'] == []
        assert results['NM_181831.2:c.675_676insCGACGC']['gene_symbol'] == 'NF2'
        assert results['NM_181831.2:c.675_676insCGACGC']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_861969.1:p.(Arg227_Arg228dup)', 'slr': 'NP_861969.1:p.(R227_R228dup)'}
        assert results['NM_181831.2:c.675_676insCGACGC']['submitted_variant'] == '22-30064360-G-GCGACGC'
        assert results['NM_181831.2:c.675_676insCGACGC']['genome_context_intronic_sequence'] == ''
        assert results['NM_181831.2:c.675_676insCGACGC']['hgvs_lrg_variant'] == ''
        assert results['NM_181831.2:c.675_676insCGACGC']['hgvs_transcript_variant'] == 'NM_181831.2:c.675_676insCGACGC'
        assert results['NM_181831.2:c.675_676insCGACGC']['hgvs_refseqgene_variant'] == ''
        assert results['NM_181831.2:c.675_676insCGACGC']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000022.10:g.30064360_30064361insCGACGC', 'vcf': {'chr': 'chr22', 'ref': 'G', 'pos': '30064360', 'alt': 'GCGACGC'}}
        assert results['NM_181831.2:c.675_676insCGACGC']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000022.11:g.29668371_29668372insCGACGC', 'vcf': {'chr': 'chr22', 'ref': 'G', 'pos': '29668371', 'alt': 'GCGACGC'}}
        assert results['NM_181831.2:c.675_676insCGACGC']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000022.10:g.30064360_30064361insCGACGC', 'vcf': {'chr': '22', 'ref': 'G', 'pos': '30064360', 'alt': 'GCGACGC'}}
        assert results['NM_181831.2:c.675_676insCGACGC']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000022.11:g.29668371_29668372insCGACGC', 'vcf': {'chr': '22', 'ref': 'G', 'pos': '29668371', 'alt': 'GCGACGC'}}


    def test_variant280(self):
        variant = '3-10188187-TGTCCCGATAG-T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_198156.2:c.341-3280_341-3271del' in results.keys()
        assert results['NM_198156.2:c.341-3280_341-3271del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_198156.2:c.341-3280_341-3271del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_198156.2:c.341-3280_341-3271del']['alt_genomic_loci'] == []
        assert results['NM_198156.2:c.341-3280_341-3271del']['gene_symbol'] == 'VHL'
        assert results['NM_198156.2:c.341-3280_341-3271del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_937799.1:p.?', 'slr': 'NP_937799.1:p.?'}
        assert results['NM_198156.2:c.341-3280_341-3271del']['submitted_variant'] == '3-10188187-TGTCCCGATAG-T'
        assert results['NM_198156.2:c.341-3280_341-3271del']['genome_context_intronic_sequence'] == 'NC_000003.11(NM_198156.2):c.341-3280_341-3271del'
        assert results['NM_198156.2:c.341-3280_341-3271del']['hgvs_lrg_variant'] == ''
        assert results['NM_198156.2:c.341-3280_341-3271del']['hgvs_transcript_variant'] == 'NM_198156.2:c.341-3280_341-3271del'
        assert results['NM_198156.2:c.341-3280_341-3271del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_198156.2:c.341-3280_341-3271del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000003.11:g.10188191_10188200del', 'vcf': {'chr': 'chr3', 'ref': 'TGTCCCGATAG', 'pos': '10188187', 'alt': 'T'}}
        assert results['NM_198156.2:c.341-3280_341-3271del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000003.12:g.10146507_10146516del', 'vcf': {'chr': 'chr3', 'ref': 'TGTCCCGATAG', 'pos': '10146503', 'alt': 'T'}}
        assert results['NM_198156.2:c.341-3280_341-3271del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000003.11:g.10188191_10188200del', 'vcf': {'chr': '3', 'ref': 'TGTCCCGATAG', 'pos': '10188187', 'alt': 'T'}}
        assert results['NM_198156.2:c.341-3280_341-3271del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000003.12:g.10146507_10146516del', 'vcf': {'chr': '3', 'ref': 'TGTCCCGATAG', 'pos': '10146503', 'alt': 'T'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001354723.1:c.*18-3280_*18-3271del' in results.keys()
        assert results['NM_001354723.1:c.*18-3280_*18-3271del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001354723.1:c.*18-3280_*18-3271del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001354723.1:c.*18-3280_*18-3271del']['alt_genomic_loci'] == []
        assert results['NM_001354723.1:c.*18-3280_*18-3271del']['gene_symbol'] == 'VHL'
        assert results['NM_001354723.1:c.*18-3280_*18-3271del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001341652.1:p.?', 'slr': 'NP_001341652.1:p.?'}
        assert results['NM_001354723.1:c.*18-3280_*18-3271del']['submitted_variant'] == '3-10188187-TGTCCCGATAG-T'
        assert results['NM_001354723.1:c.*18-3280_*18-3271del']['genome_context_intronic_sequence'] == 'NC_000003.11(NM_001354723.1):c.*18-3280_*18-3271del'
        assert results['NM_001354723.1:c.*18-3280_*18-3271del']['hgvs_lrg_variant'] == ''
        assert results['NM_001354723.1:c.*18-3280_*18-3271del']['hgvs_transcript_variant'] == 'NM_001354723.1:c.*18-3280_*18-3271del'
        assert results['NM_001354723.1:c.*18-3280_*18-3271del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001354723.1:c.*18-3280_*18-3271del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000003.11:g.10188191_10188200del', 'vcf': {'chr': 'chr3', 'ref': 'TGTCCCGATAG', 'pos': '10188187', 'alt': 'T'}}
        assert results['NM_001354723.1:c.*18-3280_*18-3271del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000003.12:g.10146507_10146516del', 'vcf': {'chr': 'chr3', 'ref': 'TGTCCCGATAG', 'pos': '10146503', 'alt': 'T'}}
        assert results['NM_001354723.1:c.*18-3280_*18-3271del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000003.11:g.10188191_10188200del', 'vcf': {'chr': '3', 'ref': 'TGTCCCGATAG', 'pos': '10188187', 'alt': 'T'}}
        assert results['NM_001354723.1:c.*18-3280_*18-3271del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000003.12:g.10146507_10146516del', 'vcf': {'chr': '3', 'ref': 'TGTCCCGATAG', 'pos': '10146503', 'alt': 'T'}}

        assert 'NM_000551.3:c.341-7_343del' in results.keys()
        assert results['NM_000551.3:c.341-7_343del']['hgvs_lrg_transcript_variant'] == 'LRG_322t1:c.341-7_343del'
        assert results['NM_000551.3:c.341-7_343del']['refseqgene_context_intronic_sequence'] == 'NG_008212.3(NM_000551.3):c.341-7_343del'
        assert results['NM_000551.3:c.341-7_343del']['alt_genomic_loci'] == []
        assert results['NM_000551.3:c.341-7_343del']['gene_symbol'] == 'VHL'
        assert results['NM_000551.3:c.341-7_343del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000542.1(LRG_322p1):p.?', 'slr': 'NP_000542.1:p.?'}
        assert results['NM_000551.3:c.341-7_343del']['submitted_variant'] == '3-10188187-TGTCCCGATAG-T'
        assert results['NM_000551.3:c.341-7_343del']['genome_context_intronic_sequence'] == 'NC_000003.11(NM_000551.3):c.341-7_343del'
        assert results['NM_000551.3:c.341-7_343del']['hgvs_lrg_variant'] == 'LRG_322:g.9873_9882del'
        assert results['NM_000551.3:c.341-7_343del']['hgvs_transcript_variant'] == 'NM_000551.3:c.341-7_343del'
        assert results['NM_000551.3:c.341-7_343del']['hgvs_refseqgene_variant'] == 'NG_008212.3:g.9873_9882del'
        assert results['NM_000551.3:c.341-7_343del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000003.11:g.10188191_10188200del', 'vcf': {'chr': 'chr3', 'ref': 'TGTCCCGATAG', 'pos': '10188187', 'alt': 'T'}}
        assert results['NM_000551.3:c.341-7_343del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000003.12:g.10146507_10146516del', 'vcf': {'chr': 'chr3', 'ref': 'TGTCCCGATAG', 'pos': '10146503', 'alt': 'T'}}
        assert results['NM_000551.3:c.341-7_343del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000003.11:g.10188191_10188200del', 'vcf': {'chr': '3', 'ref': 'TGTCCCGATAG', 'pos': '10188187', 'alt': 'T'}}
        assert results['NM_000551.3:c.341-7_343del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000003.12:g.10146507_10146516del', 'vcf': {'chr': '3', 'ref': 'TGTCCCGATAG', 'pos': '10146503', 'alt': 'T'}}


    def test_variant281(self):
        variant = '3-50402127-T-G'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001005505.2:c.3408A>C' in results.keys()
        assert results['NM_001005505.2:c.3408A>C']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001005505.2:c.3408A>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001005505.2:c.3408A>C']['alt_genomic_loci'] == []
        assert results['NM_001005505.2:c.3408A>C']['gene_symbol'] == 'CACNA2D2'
        assert results['NM_001005505.2:c.3408A>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001005505.1:p.(Gln1136His)', 'slr': 'NP_001005505.1:p.(Q1136H)'}
        assert results['NM_001005505.2:c.3408A>C']['submitted_variant'] == '3-50402127-T-G'
        assert results['NM_001005505.2:c.3408A>C']['genome_context_intronic_sequence'] == ''
        assert results['NM_001005505.2:c.3408A>C']['hgvs_lrg_variant'] == ''
        assert results['NM_001005505.2:c.3408A>C']['hgvs_transcript_variant'] == 'NM_001005505.2:c.3408A>C'
        assert results['NM_001005505.2:c.3408A>C']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001005505.2:c.3408A>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402127T>G', 'vcf': {'chr': 'chr3', 'ref': u'T', 'pos': '50402127', 'alt': u'G'}}
        assert results['NM_001005505.2:c.3408A>C']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000003.12:g.50364696T>G', 'vcf': {'chr': 'chr3', 'ref': u'T', 'pos': '50364696', 'alt': u'G'}}
        assert results['NM_001005505.2:c.3408A>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402127T>G', 'vcf': {'chr': '3', 'ref': u'T', 'pos': '50402127', 'alt': u'G'}}
        assert results['NM_001005505.2:c.3408A>C']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000003.12:g.50364696T>G', 'vcf': {'chr': '3', 'ref': u'T', 'pos': '50364696', 'alt': u'G'}}

        assert 'NM_006030.2:c.3402A>C' in results.keys()
        assert results['NM_006030.2:c.3402A>C']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_006030.2:c.3402A>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_006030.2:c.3402A>C']['alt_genomic_loci'] == []
        assert results['NM_006030.2:c.3402A>C']['gene_symbol'] == 'CACNA2D2'
        assert results['NM_006030.2:c.3402A>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_006021.2:p.(Gln1134His)', 'slr': 'NP_006021.2:p.(Q1134H)'}
        assert results['NM_006030.2:c.3402A>C']['submitted_variant'] == '3-50402127-T-G'
        assert results['NM_006030.2:c.3402A>C']['genome_context_intronic_sequence'] == ''
        assert results['NM_006030.2:c.3402A>C']['hgvs_lrg_variant'] == ''
        assert results['NM_006030.2:c.3402A>C']['hgvs_transcript_variant'] == 'NM_006030.2:c.3402A>C'
        assert results['NM_006030.2:c.3402A>C']['hgvs_refseqgene_variant'] == ''
        assert results['NM_006030.2:c.3402A>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402127T>G', 'vcf': {'chr': 'chr3', 'ref': u'T', 'pos': '50402127', 'alt': u'G'}}
        assert 'hg38' not in results['NM_006030.2:c.3402A>C']['primary_assembly_loci'].keys()
        assert results['NM_006030.2:c.3402A>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402127T>G', 'vcf': {'chr': '3', 'ref': u'T', 'pos': '50402127', 'alt': u'G'}}
        assert 'grch38' not in results['NM_006030.2:c.3402A>C']['primary_assembly_loci'].keys()

        assert 'NM_001291101.1:c.3201A>C' in results.keys()
        assert results['NM_001291101.1:c.3201A>C']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001291101.1:c.3201A>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001291101.1:c.3201A>C']['alt_genomic_loci'] == []
        assert results['NM_001291101.1:c.3201A>C']['gene_symbol'] == 'CACNA2D2'
        assert results['NM_001291101.1:c.3201A>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001278030.1:p.(Gln1067His)', 'slr': 'NP_001278030.1:p.(Q1067H)'}
        assert results['NM_001291101.1:c.3201A>C']['submitted_variant'] == '3-50402127-T-G'
        assert results['NM_001291101.1:c.3201A>C']['genome_context_intronic_sequence'] == ''
        assert results['NM_001291101.1:c.3201A>C']['hgvs_lrg_variant'] == ''
        assert results['NM_001291101.1:c.3201A>C']['hgvs_transcript_variant'] == 'NM_001291101.1:c.3201A>C'
        assert results['NM_001291101.1:c.3201A>C']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001291101.1:c.3201A>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402127T>G', 'vcf': {'chr': 'chr3', 'ref': u'T', 'pos': '50402127', 'alt': u'G'}}
        assert results['NM_001291101.1:c.3201A>C']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000003.12:g.50364696T>G', 'vcf': {'chr': 'chr3', 'ref': u'T', 'pos': '50364696', 'alt': u'G'}}
        assert results['NM_001291101.1:c.3201A>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402127T>G', 'vcf': {'chr': '3', 'ref': u'T', 'pos': '50402127', 'alt': u'G'}}
        assert results['NM_001291101.1:c.3201A>C']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000003.12:g.50364696T>G', 'vcf': {'chr': '3', 'ref': u'T', 'pos': '50364696', 'alt': u'G'}}

        assert 'NM_001174051.2:c.3423A>C' in results.keys()
        assert results['NM_001174051.2:c.3423A>C']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001174051.2:c.3423A>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001174051.2:c.3423A>C']['alt_genomic_loci'] == []
        assert results['NM_001174051.2:c.3423A>C']['gene_symbol'] == 'CACNA2D2'
        assert results['NM_001174051.2:c.3423A>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001167522.1:p.(Gln1141His)', 'slr': 'NP_001167522.1:p.(Q1141H)'}
        assert results['NM_001174051.2:c.3423A>C']['submitted_variant'] == '3-50402127-T-G'
        assert results['NM_001174051.2:c.3423A>C']['genome_context_intronic_sequence'] == ''
        assert results['NM_001174051.2:c.3423A>C']['hgvs_lrg_variant'] == ''
        assert results['NM_001174051.2:c.3423A>C']['hgvs_transcript_variant'] == 'NM_001174051.2:c.3423A>C'
        assert results['NM_001174051.2:c.3423A>C']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001174051.2:c.3423A>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402127T>G', 'vcf': {'chr': 'chr3', 'ref': u'T', 'pos': '50402127', 'alt': u'G'}}
        assert results['NM_001174051.2:c.3423A>C']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000003.12:g.50364696T>G', 'vcf': {'chr': 'chr3', 'ref': u'T', 'pos': '50364696', 'alt': u'G'}}
        assert results['NM_001174051.2:c.3423A>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402127T>G', 'vcf': {'chr': '3', 'ref': u'T', 'pos': '50402127', 'alt': u'G'}}
        assert results['NM_001174051.2:c.3423A>C']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000003.12:g.50364696T>G', 'vcf': {'chr': '3', 'ref': u'T', 'pos': '50364696', 'alt': u'G'}}

        assert 'NM_006030.3:c.3402A>C' in results.keys()
        assert results['NM_006030.3:c.3402A>C']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_006030.3:c.3402A>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_006030.3:c.3402A>C']['alt_genomic_loci'] == []
        assert results['NM_006030.3:c.3402A>C']['gene_symbol'] == 'CACNA2D2'
        assert results['NM_006030.3:c.3402A>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_006021.2:p.(Gln1134His)', 'slr': 'NP_006021.2:p.(Q1134H)'}
        assert results['NM_006030.3:c.3402A>C']['submitted_variant'] == '3-50402127-T-G'
        assert results['NM_006030.3:c.3402A>C']['genome_context_intronic_sequence'] == ''
        assert results['NM_006030.3:c.3402A>C']['hgvs_lrg_variant'] == ''
        assert results['NM_006030.3:c.3402A>C']['hgvs_transcript_variant'] == 'NM_006030.3:c.3402A>C'
        assert results['NM_006030.3:c.3402A>C']['hgvs_refseqgene_variant'] == ''
        assert results['NM_006030.3:c.3402A>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402127T>G', 'vcf': {'chr': 'chr3', 'ref': u'T', 'pos': '50402127', 'alt': u'G'}}
        assert results['NM_006030.3:c.3402A>C']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000003.12:g.50364696T>G', 'vcf': {'chr': 'chr3', 'ref': u'T', 'pos': '50364696', 'alt': u'G'}}
        assert results['NM_006030.3:c.3402A>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402127T>G', 'vcf': {'chr': '3', 'ref': u'T', 'pos': '50402127', 'alt': u'G'}}
        assert results['NM_006030.3:c.3402A>C']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000003.12:g.50364696T>G', 'vcf': {'chr': '3', 'ref': u'T', 'pos': '50364696', 'alt': u'G'}}

        assert 'NM_001174051.1:c.3423A>C' in results.keys()
        assert results['NM_001174051.1:c.3423A>C']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001174051.1:c.3423A>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001174051.1:c.3423A>C']['alt_genomic_loci'] == []
        assert results['NM_001174051.1:c.3423A>C']['gene_symbol'] == 'CACNA2D2'
        assert results['NM_001174051.1:c.3423A>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001167522.1:p.(Gln1141His)', 'slr': 'NP_001167522.1:p.(Q1141H)'}
        assert results['NM_001174051.1:c.3423A>C']['submitted_variant'] == '3-50402127-T-G'
        assert results['NM_001174051.1:c.3423A>C']['genome_context_intronic_sequence'] == ''
        assert results['NM_001174051.1:c.3423A>C']['hgvs_lrg_variant'] == ''
        assert results['NM_001174051.1:c.3423A>C']['hgvs_transcript_variant'] == 'NM_001174051.1:c.3423A>C'
        assert results['NM_001174051.1:c.3423A>C']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001174051.1:c.3423A>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402127T>G', 'vcf': {'chr': 'chr3', 'ref': u'T', 'pos': '50402127', 'alt': u'G'}}
        assert 'hg38' not in results['NM_001174051.1:c.3423A>C']['primary_assembly_loci'].keys()
        assert results['NM_001174051.1:c.3423A>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402127T>G', 'vcf': {'chr': '3', 'ref': u'T', 'pos': '50402127', 'alt': u'G'}}
        assert 'grch38' not in results['NM_001174051.1:c.3423A>C']['primary_assembly_loci'].keys()

        assert results['flag'] == 'gene_variant'
        assert 'NR_111912.1:n.443-1601T>G' in results.keys()
        assert results['NR_111912.1:n.443-1601T>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_111912.1:n.443-1601T>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_111912.1:n.443-1601T>G']['alt_genomic_loci'] == []
        assert results['NR_111912.1:n.443-1601T>G']['gene_symbol'] == 'CYB561D2'
        assert results['NR_111912.1:n.443-1601T>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_111912.1:n.443-1601T>G']['submitted_variant'] == '3-50402127-T-G'
        assert results['NR_111912.1:n.443-1601T>G']['genome_context_intronic_sequence'] == 'NC_000003.11(NR_111912.1):c.443-1601T>G'
        assert results['NR_111912.1:n.443-1601T>G']['hgvs_lrg_variant'] == ''
        assert results['NR_111912.1:n.443-1601T>G']['hgvs_transcript_variant'] == 'NR_111912.1:n.443-1601T>G'
        assert results['NR_111912.1:n.443-1601T>G']['hgvs_refseqgene_variant'] == ''
        assert results['NR_111912.1:n.443-1601T>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402127T>G', 'vcf': {'chr': 'chr3', 'ref': 'T', 'pos': '50402127', 'alt': 'G'}}
        assert results['NR_111912.1:n.443-1601T>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000003.12:g.50364696T>G', 'vcf': {'chr': 'chr3', 'ref': 'T', 'pos': '50364696', 'alt': 'G'}}
        assert results['NR_111912.1:n.443-1601T>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402127T>G', 'vcf': {'chr': '3', 'ref': 'T', 'pos': '50402127', 'alt': 'G'}}
        assert results['NR_111912.1:n.443-1601T>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000003.12:g.50364696T>G', 'vcf': {'chr': '3', 'ref': 'T', 'pos': '50364696', 'alt': 'G'}}

        assert 'NM_001005505.1:c.3408A>C' in results.keys()
        assert results['NM_001005505.1:c.3408A>C']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001005505.1:c.3408A>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001005505.1:c.3408A>C']['alt_genomic_loci'] == []
        assert results['NM_001005505.1:c.3408A>C']['gene_symbol'] == 'CACNA2D2'
        assert results['NM_001005505.1:c.3408A>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001005505.1:p.(Gln1136His)', 'slr': 'NP_001005505.1:p.(Q1136H)'}
        assert results['NM_001005505.1:c.3408A>C']['submitted_variant'] == '3-50402127-T-G'
        assert results['NM_001005505.1:c.3408A>C']['genome_context_intronic_sequence'] == ''
        assert results['NM_001005505.1:c.3408A>C']['hgvs_lrg_variant'] == ''
        assert results['NM_001005505.1:c.3408A>C']['hgvs_transcript_variant'] == 'NM_001005505.1:c.3408A>C'
        assert results['NM_001005505.1:c.3408A>C']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001005505.1:c.3408A>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402127T>G', 'vcf': {'chr': 'chr3', 'ref': u'T', 'pos': '50402127', 'alt': u'G'}}
        assert 'hg38' not in results['NM_001005505.1:c.3408A>C']['primary_assembly_loci'].keys()
        assert results['NM_001005505.1:c.3408A>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402127T>G', 'vcf': {'chr': '3', 'ref': u'T', 'pos': '50402127', 'alt': u'G'}}
        assert 'grch38' not in results['NM_001005505.1:c.3408A>C']['primary_assembly_loci'].keys()


    def test_variant282(self):
        variant = '3-50402890-G-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NR_111913.1:n.126G>A' in results.keys()
        assert results['NR_111913.1:n.126G>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_111913.1:n.126G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_111913.1:n.126G>A']['alt_genomic_loci'] == []
        assert results['NR_111913.1:n.126G>A']['gene_symbol'] == 'CYB561D2'
        assert results['NR_111913.1:n.126G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_111913.1:n.126G>A']['submitted_variant'] == '3-50402890-G-A'
        assert results['NR_111913.1:n.126G>A']['genome_context_intronic_sequence'] == ''
        assert results['NR_111913.1:n.126G>A']['hgvs_lrg_variant'] == ''
        assert results['NR_111913.1:n.126G>A']['hgvs_transcript_variant'] == 'NR_111913.1:n.126G>A'
        assert results['NR_111913.1:n.126G>A']['hgvs_refseqgene_variant'] == ''
        assert results['NR_111913.1:n.126G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402890G>A', 'vcf': {'chr': 'chr3', 'ref': 'G', 'pos': '50402890', 'alt': 'A'}}
        assert results['NR_111913.1:n.126G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000003.12:g.50365459G>A', 'vcf': {'chr': 'chr3', 'ref': 'G', 'pos': '50365459', 'alt': 'A'}}
        assert results['NR_111913.1:n.126G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402890G>A', 'vcf': {'chr': '3', 'ref': 'G', 'pos': '50402890', 'alt': 'A'}}
        assert results['NR_111913.1:n.126G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000003.12:g.50365459G>A', 'vcf': {'chr': '3', 'ref': 'G', 'pos': '50365459', 'alt': 'A'}}

        assert 'NR_111912.1:n.443-838G>A' in results.keys()
        assert results['NR_111912.1:n.443-838G>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_111912.1:n.443-838G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_111912.1:n.443-838G>A']['alt_genomic_loci'] == []
        assert results['NR_111912.1:n.443-838G>A']['gene_symbol'] == 'CYB561D2'
        assert results['NR_111912.1:n.443-838G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_111912.1:n.443-838G>A']['submitted_variant'] == '3-50402890-G-A'
        assert results['NR_111912.1:n.443-838G>A']['genome_context_intronic_sequence'] == 'NC_000003.11(NR_111912.1):c.443-838G>A'
        assert results['NR_111912.1:n.443-838G>A']['hgvs_lrg_variant'] == ''
        assert results['NR_111912.1:n.443-838G>A']['hgvs_transcript_variant'] == 'NR_111912.1:n.443-838G>A'
        assert results['NR_111912.1:n.443-838G>A']['hgvs_refseqgene_variant'] == ''
        assert results['NR_111912.1:n.443-838G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402890G>A', 'vcf': {'chr': 'chr3', 'ref': 'G', 'pos': '50402890', 'alt': 'A'}}
        assert results['NR_111912.1:n.443-838G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000003.12:g.50365459G>A', 'vcf': {'chr': 'chr3', 'ref': 'G', 'pos': '50365459', 'alt': 'A'}}
        assert results['NR_111912.1:n.443-838G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402890G>A', 'vcf': {'chr': '3', 'ref': 'G', 'pos': '50402890', 'alt': 'A'}}
        assert results['NR_111912.1:n.443-838G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000003.12:g.50365459G>A', 'vcf': {'chr': '3', 'ref': 'G', 'pos': '50365459', 'alt': 'A'}}

        assert 'NM_001291101.1:c.2788C>T' in results.keys()
        assert results['NM_001291101.1:c.2788C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001291101.1:c.2788C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001291101.1:c.2788C>T']['alt_genomic_loci'] == []
        assert results['NM_001291101.1:c.2788C>T']['gene_symbol'] == 'CACNA2D2'
        assert results['NM_001291101.1:c.2788C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001278030.1:p.(Pro930Ser)', 'slr': 'NP_001278030.1:p.(P930S)'}
        assert results['NM_001291101.1:c.2788C>T']['submitted_variant'] == '3-50402890-G-A'
        assert results['NM_001291101.1:c.2788C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001291101.1:c.2788C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001291101.1:c.2788C>T']['hgvs_transcript_variant'] == 'NM_001291101.1:c.2788C>T'
        assert results['NM_001291101.1:c.2788C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001291101.1:c.2788C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402890G>A', 'vcf': {'chr': 'chr3', 'ref': u'G', 'pos': '50402890', 'alt': u'A'}}
        assert results['NM_001291101.1:c.2788C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000003.12:g.50365459G>A', 'vcf': {'chr': 'chr3', 'ref': u'G', 'pos': '50365459', 'alt': u'A'}}
        assert results['NM_001291101.1:c.2788C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402890G>A', 'vcf': {'chr': '3', 'ref': u'G', 'pos': '50402890', 'alt': u'A'}}
        assert results['NM_001291101.1:c.2788C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000003.12:g.50365459G>A', 'vcf': {'chr': '3', 'ref': u'G', 'pos': '50365459', 'alt': u'A'}}

        assert 'NM_006030.2:c.2995C>T' in results.keys()
        assert results['NM_006030.2:c.2995C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_006030.2:c.2995C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_006030.2:c.2995C>T']['alt_genomic_loci'] == []
        assert results['NM_006030.2:c.2995C>T']['gene_symbol'] == 'CACNA2D2'
        assert results['NM_006030.2:c.2995C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_006021.2:p.(Pro999Ser)', 'slr': 'NP_006021.2:p.(P999S)'}
        assert results['NM_006030.2:c.2995C>T']['submitted_variant'] == '3-50402890-G-A'
        assert results['NM_006030.2:c.2995C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_006030.2:c.2995C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_006030.2:c.2995C>T']['hgvs_transcript_variant'] == 'NM_006030.2:c.2995C>T'
        assert results['NM_006030.2:c.2995C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_006030.2:c.2995C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402890G>A', 'vcf': {'chr': 'chr3', 'ref': u'G', 'pos': '50402890', 'alt': u'A'}}
        assert 'hg38' not in results['NM_006030.2:c.2995C>T']['primary_assembly_loci'].keys()
        assert results['NM_006030.2:c.2995C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402890G>A', 'vcf': {'chr': '3', 'ref': u'G', 'pos': '50402890', 'alt': u'A'}}
        assert 'grch38' not in results['NM_006030.2:c.2995C>T']['primary_assembly_loci'].keys()

        assert 'NR_111914.1:n.126G>A' in results.keys()
        assert results['NR_111914.1:n.126G>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_111914.1:n.126G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_111914.1:n.126G>A']['alt_genomic_loci'] == []
        assert results['NR_111914.1:n.126G>A']['gene_symbol'] == 'CYB561D2'
        assert results['NR_111914.1:n.126G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_111914.1:n.126G>A']['submitted_variant'] == '3-50402890-G-A'
        assert results['NR_111914.1:n.126G>A']['genome_context_intronic_sequence'] == ''
        assert results['NR_111914.1:n.126G>A']['hgvs_lrg_variant'] == ''
        assert results['NR_111914.1:n.126G>A']['hgvs_transcript_variant'] == 'NR_111914.1:n.126G>A'
        assert results['NR_111914.1:n.126G>A']['hgvs_refseqgene_variant'] == ''
        assert results['NR_111914.1:n.126G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402890G>A', 'vcf': {'chr': 'chr3', 'ref': 'G', 'pos': '50402890', 'alt': 'A'}}
        assert results['NR_111914.1:n.126G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000003.12:g.50365459G>A', 'vcf': {'chr': 'chr3', 'ref': 'G', 'pos': '50365459', 'alt': 'A'}}
        assert results['NR_111914.1:n.126G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402890G>A', 'vcf': {'chr': '3', 'ref': 'G', 'pos': '50402890', 'alt': 'A'}}
        assert results['NR_111914.1:n.126G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000003.12:g.50365459G>A', 'vcf': {'chr': '3', 'ref': 'G', 'pos': '50365459', 'alt': 'A'}}

        assert 'NM_001005505.2:c.2995C>T' in results.keys()
        assert results['NM_001005505.2:c.2995C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001005505.2:c.2995C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001005505.2:c.2995C>T']['alt_genomic_loci'] == []
        assert results['NM_001005505.2:c.2995C>T']['gene_symbol'] == 'CACNA2D2'
        assert results['NM_001005505.2:c.2995C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001005505.1:p.(Pro999Ser)', 'slr': 'NP_001005505.1:p.(P999S)'}
        assert results['NM_001005505.2:c.2995C>T']['submitted_variant'] == '3-50402890-G-A'
        assert results['NM_001005505.2:c.2995C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001005505.2:c.2995C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001005505.2:c.2995C>T']['hgvs_transcript_variant'] == 'NM_001005505.2:c.2995C>T'
        assert results['NM_001005505.2:c.2995C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001005505.2:c.2995C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402890G>A', 'vcf': {'chr': 'chr3', 'ref': u'G', 'pos': '50402890', 'alt': u'A'}}
        assert results['NM_001005505.2:c.2995C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000003.12:g.50365459G>A', 'vcf': {'chr': 'chr3', 'ref': u'G', 'pos': '50365459', 'alt': u'A'}}
        assert results['NM_001005505.2:c.2995C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402890G>A', 'vcf': {'chr': '3', 'ref': u'G', 'pos': '50402890', 'alt': u'A'}}
        assert results['NM_001005505.2:c.2995C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000003.12:g.50365459G>A', 'vcf': {'chr': '3', 'ref': u'G', 'pos': '50365459', 'alt': u'A'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001174051.1:c.3016C>T' in results.keys()
        assert results['NM_001174051.1:c.3016C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001174051.1:c.3016C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001174051.1:c.3016C>T']['alt_genomic_loci'] == []
        assert results['NM_001174051.1:c.3016C>T']['gene_symbol'] == 'CACNA2D2'
        assert results['NM_001174051.1:c.3016C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001167522.1:p.(Pro1006Ser)', 'slr': 'NP_001167522.1:p.(P1006S)'}
        assert results['NM_001174051.1:c.3016C>T']['submitted_variant'] == '3-50402890-G-A'
        assert results['NM_001174051.1:c.3016C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001174051.1:c.3016C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001174051.1:c.3016C>T']['hgvs_transcript_variant'] == 'NM_001174051.1:c.3016C>T'
        assert results['NM_001174051.1:c.3016C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001174051.1:c.3016C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402890G>A', 'vcf': {'chr': 'chr3', 'ref': u'G', 'pos': '50402890', 'alt': u'A'}}
        assert 'hg38' not in results['NM_001174051.1:c.3016C>T']['primary_assembly_loci'].keys()
        assert results['NM_001174051.1:c.3016C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402890G>A', 'vcf': {'chr': '3', 'ref': u'G', 'pos': '50402890', 'alt': u'A'}}
        assert 'grch38' not in results['NM_001174051.1:c.3016C>T']['primary_assembly_loci'].keys()

        assert 'NM_001174051.2:c.3016C>T' in results.keys()
        assert results['NM_001174051.2:c.3016C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001174051.2:c.3016C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001174051.2:c.3016C>T']['alt_genomic_loci'] == []
        assert results['NM_001174051.2:c.3016C>T']['gene_symbol'] == 'CACNA2D2'
        assert results['NM_001174051.2:c.3016C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001167522.1:p.(Pro1006Ser)', 'slr': 'NP_001167522.1:p.(P1006S)'}
        assert results['NM_001174051.2:c.3016C>T']['submitted_variant'] == '3-50402890-G-A'
        assert results['NM_001174051.2:c.3016C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001174051.2:c.3016C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001174051.2:c.3016C>T']['hgvs_transcript_variant'] == 'NM_001174051.2:c.3016C>T'
        assert results['NM_001174051.2:c.3016C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001174051.2:c.3016C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402890G>A', 'vcf': {'chr': 'chr3', 'ref': u'G', 'pos': '50402890', 'alt': u'A'}}
        assert results['NM_001174051.2:c.3016C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000003.12:g.50365459G>A', 'vcf': {'chr': 'chr3', 'ref': u'G', 'pos': '50365459', 'alt': u'A'}}
        assert results['NM_001174051.2:c.3016C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402890G>A', 'vcf': {'chr': '3', 'ref': u'G', 'pos': '50402890', 'alt': u'A'}}
        assert results['NM_001174051.2:c.3016C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000003.12:g.50365459G>A', 'vcf': {'chr': '3', 'ref': u'G', 'pos': '50365459', 'alt': u'A'}}

        assert 'NM_006030.3:c.2995C>T' in results.keys()
        assert results['NM_006030.3:c.2995C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_006030.3:c.2995C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_006030.3:c.2995C>T']['alt_genomic_loci'] == []
        assert results['NM_006030.3:c.2995C>T']['gene_symbol'] == 'CACNA2D2'
        assert results['NM_006030.3:c.2995C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_006021.2:p.(Pro999Ser)', 'slr': 'NP_006021.2:p.(P999S)'}
        assert results['NM_006030.3:c.2995C>T']['submitted_variant'] == '3-50402890-G-A'
        assert results['NM_006030.3:c.2995C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_006030.3:c.2995C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_006030.3:c.2995C>T']['hgvs_transcript_variant'] == 'NM_006030.3:c.2995C>T'
        assert results['NM_006030.3:c.2995C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_006030.3:c.2995C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402890G>A', 'vcf': {'chr': 'chr3', 'ref': u'G', 'pos': '50402890', 'alt': u'A'}}
        assert results['NM_006030.3:c.2995C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000003.12:g.50365459G>A', 'vcf': {'chr': 'chr3', 'ref': u'G', 'pos': '50365459', 'alt': u'A'}}
        assert results['NM_006030.3:c.2995C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402890G>A', 'vcf': {'chr': '3', 'ref': u'G', 'pos': '50402890', 'alt': u'A'}}
        assert results['NM_006030.3:c.2995C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000003.12:g.50365459G>A', 'vcf': {'chr': '3', 'ref': u'G', 'pos': '50365459', 'alt': u'A'}}

        assert 'NM_001005505.1:c.2995C>T' in results.keys()
        assert results['NM_001005505.1:c.2995C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001005505.1:c.2995C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001005505.1:c.2995C>T']['alt_genomic_loci'] == []
        assert results['NM_001005505.1:c.2995C>T']['gene_symbol'] == 'CACNA2D2'
        assert results['NM_001005505.1:c.2995C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001005505.1:p.(Pro999Ser)', 'slr': 'NP_001005505.1:p.(P999S)'}
        assert results['NM_001005505.1:c.2995C>T']['submitted_variant'] == '3-50402890-G-A'
        assert results['NM_001005505.1:c.2995C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001005505.1:c.2995C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001005505.1:c.2995C>T']['hgvs_transcript_variant'] == 'NM_001005505.1:c.2995C>T'
        assert results['NM_001005505.1:c.2995C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001005505.1:c.2995C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402890G>A', 'vcf': {'chr': 'chr3', 'ref': u'G', 'pos': '50402890', 'alt': u'A'}}
        assert 'hg38' not in results['NM_001005505.1:c.2995C>T']['primary_assembly_loci'].keys()
        assert results['NM_001005505.1:c.2995C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000003.11:g.50402890G>A', 'vcf': {'chr': '3', 'ref': u'G', 'pos': '50402890', 'alt': u'A'}}
        assert 'grch38' not in results['NM_001005505.1:c.2995C>T']['primary_assembly_loci'].keys()


    def test_variant283(self):
        variant = '3-57851007-AG-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert '' in results.keys()
        assert results['']['hgvs_lrg_transcript_variant'] == ''
        assert results['']['refseqgene_context_intronic_sequence'] == ''
        assert results['']['alt_genomic_loci'] == []
        assert results['']['gene_symbol'] == ''
        assert results['']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['']['submitted_variant'] == '3-57851007-AG-A'
        assert results['']['genome_context_intronic_sequence'] == ''
        assert results['']['hgvs_lrg_variant'] == ''
        assert results['']['hgvs_transcript_variant'] == ''
        assert results['']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['']['primary_assembly_loci'].keys()

        assert 'NM_001304420.2:c.1186+424del' in results.keys()
        assert results['NM_001304420.2:c.1186+424del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001304420.2:c.1186+424del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001304420.2:c.1186+424del']['alt_genomic_loci'] == []
        assert results['NM_001304420.2:c.1186+424del']['gene_symbol'] == 'SLMAP'
        assert results['NM_001304420.2:c.1186+424del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001291349.1:p.?', 'slr': 'NP_001291349.1:p.?'}
        assert results['NM_001304420.2:c.1186+424del']['submitted_variant'] == '3-57851007-AG-A'
        assert results['NM_001304420.2:c.1186+424del']['genome_context_intronic_sequence'] == 'NC_000003.11(NM_001304420.2):c.1186+424del'
        assert results['NM_001304420.2:c.1186+424del']['hgvs_lrg_variant'] == ''
        assert results['NM_001304420.2:c.1186+424del']['hgvs_transcript_variant'] == 'NM_001304420.2:c.1186+424del'
        assert results['NM_001304420.2:c.1186+424del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001304420.2:c.1186+424del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000003.11:g.57851008del', 'vcf': {'chr': 'chr3', 'ref': 'AG', 'pos': '57851007', 'alt': 'A'}}
        assert results['NM_001304420.2:c.1186+424del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000003.12:g.57865281del', 'vcf': {'chr': 'chr3', 'ref': 'AG', 'pos': '57865280', 'alt': 'A'}}
        assert results['NM_001304420.2:c.1186+424del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000003.11:g.57851008del', 'vcf': {'chr': '3', 'ref': 'AG', 'pos': '57851007', 'alt': 'A'}}
        assert results['NM_001304420.2:c.1186+424del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000003.12:g.57865281del', 'vcf': {'chr': '3', 'ref': 'AG', 'pos': '57865280', 'alt': 'A'}}

        assert '  ' in results.keys()
        assert results['  ']['hgvs_lrg_transcript_variant'] == ''
        assert results['  ']['refseqgene_context_intronic_sequence'] == ''
        assert results['  ']['alt_genomic_loci'] == []
        assert results['  ']['gene_symbol'] == ''
        assert results['  ']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['  ']['submitted_variant'] == '3-57851007-AG-A'
        assert results['  ']['genome_context_intronic_sequence'] == ''
        assert results['  ']['hgvs_lrg_variant'] == ''
        assert results['  ']['hgvs_transcript_variant'] == ''
        assert results['  ']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['  ']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['  ']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['  ']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['  ']['primary_assembly_loci'].keys()

        assert ' ' in results.keys()
        assert results[' ']['hgvs_lrg_transcript_variant'] == ''
        assert results[' ']['refseqgene_context_intronic_sequence'] == ''
        assert results[' ']['alt_genomic_loci'] == []
        assert results[' ']['gene_symbol'] == ''
        assert results[' ']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results[' ']['submitted_variant'] == '3-57851007-AG-A'
        assert results[' ']['genome_context_intronic_sequence'] == ''
        assert results[' ']['hgvs_lrg_variant'] == ''
        assert results[' ']['hgvs_transcript_variant'] == ''
        assert results[' ']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results[' ']['primary_assembly_loci'].keys()
        assert 'hg38' not in results[' ']['primary_assembly_loci'].keys()
        assert 'grch37' not in results[' ']['primary_assembly_loci'].keys()
        assert 'grch38' not in results[' ']['primary_assembly_loci'].keys()

        assert 'NM_007159.4:c.1135+565del' in results.keys()
        assert results['NM_007159.4:c.1135+565del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_007159.4:c.1135+565del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_007159.4:c.1135+565del']['alt_genomic_loci'] == []
        assert results['NM_007159.4:c.1135+565del']['gene_symbol'] == 'SLMAP'
        assert results['NM_007159.4:c.1135+565del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_009090.2:p.?', 'slr': 'NP_009090.2:p.?'}
        assert results['NM_007159.4:c.1135+565del']['submitted_variant'] == '3-57851007-AG-A'
        assert results['NM_007159.4:c.1135+565del']['genome_context_intronic_sequence'] == 'NC_000003.11(NM_007159.4):c.1135+565del'
        assert results['NM_007159.4:c.1135+565del']['hgvs_lrg_variant'] == ''
        assert results['NM_007159.4:c.1135+565del']['hgvs_transcript_variant'] == 'NM_007159.4:c.1135+565del'
        assert results['NM_007159.4:c.1135+565del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_007159.4:c.1135+565del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000003.11:g.57851008del', 'vcf': {'chr': 'chr3', 'ref': 'AG', 'pos': '57851007', 'alt': 'A'}}
        assert results['NM_007159.4:c.1135+565del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000003.12:g.57865281del', 'vcf': {'chr': 'chr3', 'ref': 'AG', 'pos': '57865280', 'alt': 'A'}}
        assert results['NM_007159.4:c.1135+565del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000003.11:g.57851008del', 'vcf': {'chr': '3', 'ref': 'AG', 'pos': '57851007', 'alt': 'A'}}
        assert results['NM_007159.4:c.1135+565del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000003.12:g.57865281del', 'vcf': {'chr': '3', 'ref': 'AG', 'pos': '57865280', 'alt': 'A'}}

        assert 'NM_001304421.2:c.1135+565del' in results.keys()
        assert results['NM_001304421.2:c.1135+565del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001304421.2:c.1135+565del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001304421.2:c.1135+565del']['alt_genomic_loci'] == []
        assert results['NM_001304421.2:c.1135+565del']['gene_symbol'] == 'SLMAP'
        assert results['NM_001304421.2:c.1135+565del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001291350.1:p.?', 'slr': 'NP_001291350.1:p.?'}
        assert results['NM_001304421.2:c.1135+565del']['submitted_variant'] == '3-57851007-AG-A'
        assert results['NM_001304421.2:c.1135+565del']['genome_context_intronic_sequence'] == 'NC_000003.11(NM_001304421.2):c.1135+565del'
        assert results['NM_001304421.2:c.1135+565del']['hgvs_lrg_variant'] == ''
        assert results['NM_001304421.2:c.1135+565del']['hgvs_transcript_variant'] == 'NM_001304421.2:c.1135+565del'
        assert results['NM_001304421.2:c.1135+565del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001304421.2:c.1135+565del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000003.11:g.57851008del', 'vcf': {'chr': 'chr3', 'ref': 'AG', 'pos': '57851007', 'alt': 'A'}}
        assert results['NM_001304421.2:c.1135+565del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000003.12:g.57865281del', 'vcf': {'chr': 'chr3', 'ref': 'AG', 'pos': '57865280', 'alt': 'A'}}
        assert results['NM_001304421.2:c.1135+565del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000003.11:g.57851008del', 'vcf': {'chr': '3', 'ref': 'AG', 'pos': '57851007', 'alt': 'A'}}
        assert results['NM_001304421.2:c.1135+565del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000003.12:g.57865281del', 'vcf': {'chr': '3', 'ref': 'AG', 'pos': '57865280', 'alt': 'A'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_007159.2:c.1135+565del' in results.keys()
        assert results['NM_007159.2:c.1135+565del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_007159.2:c.1135+565del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_007159.2:c.1135+565del']['alt_genomic_loci'] == []
        assert results['NM_007159.2:c.1135+565del']['gene_symbol'] == 'SLMAP'
        assert results['NM_007159.2:c.1135+565del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_009090.2:p.?', 'slr': 'NP_009090.2:p.?'}
        assert results['NM_007159.2:c.1135+565del']['submitted_variant'] == '3-57851007-AG-A'
        assert results['NM_007159.2:c.1135+565del']['genome_context_intronic_sequence'] == 'NC_000003.11(NM_007159.2):c.1135+565del'
        assert results['NM_007159.2:c.1135+565del']['hgvs_lrg_variant'] == ''
        assert results['NM_007159.2:c.1135+565del']['hgvs_transcript_variant'] == 'NM_007159.2:c.1135+565del'
        assert results['NM_007159.2:c.1135+565del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_007159.2:c.1135+565del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000003.11:g.57851008del', 'vcf': {'chr': 'chr3', 'ref': 'AG', 'pos': '57851007', 'alt': 'A'}}
        assert 'hg38' not in results['NM_007159.2:c.1135+565del']['primary_assembly_loci'].keys()
        assert results['NM_007159.2:c.1135+565del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000003.11:g.57851008del', 'vcf': {'chr': '3', 'ref': 'AG', 'pos': '57851007', 'alt': 'A'}}
        assert 'grch38' not in results['NM_007159.2:c.1135+565del']['primary_assembly_loci'].keys()


    def test_variant284(self):
        variant = '3-122003832-G-C'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001178065.1:c.3061C=' in results.keys()
        assert results['NM_001178065.1:c.3061C=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001178065.1:c.3061C=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001178065.1:c.3061C=']['alt_genomic_loci'] == []
        assert results['NM_001178065.1:c.3061C=']['gene_symbol'] == 'CASR'
        assert results['NM_001178065.1:c.3061C=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001171536.1:p.(Gln1021=)', 'slr': 'NP_001171536.1:p.(Q1021=)'}
        assert results['NM_001178065.1:c.3061C=']['submitted_variant'] == '3-122003832-G-C'
        assert results['NM_001178065.1:c.3061C=']['genome_context_intronic_sequence'] == ''
        assert results['NM_001178065.1:c.3061C=']['hgvs_lrg_variant'] == ''
        assert results['NM_001178065.1:c.3061C=']['hgvs_transcript_variant'] == 'NM_001178065.1:c.3061C='
        assert results['NM_001178065.1:c.3061C=']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001178065.1:c.3061C=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000003.11:g.122003832G>C', 'vcf': {'chr': 'chr3', 'ref': 'G', 'pos': '122003832', 'alt': 'C'}}
        assert results['NM_001178065.1:c.3061C=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000003.12:g.122284985G>C', 'vcf': {'chr': 'chr3', 'ref': 'G', 'pos': '122284985', 'alt': 'C'}}
        assert results['NM_001178065.1:c.3061C=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000003.11:g.122003832G>C', 'vcf': {'chr': '3', 'ref': 'G', 'pos': '122003832', 'alt': 'C'}}
        assert results['NM_001178065.1:c.3061C=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000003.12:g.122284985G>C', 'vcf': {'chr': '3', 'ref': 'G', 'pos': '122284985', 'alt': 'C'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_000388.3:c.3031C=' in results.keys()
        assert results['NM_000388.3:c.3031C=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000388.3:c.3031C=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000388.3:c.3031C=']['alt_genomic_loci'] == []
        assert results['NM_000388.3:c.3031C=']['gene_symbol'] == 'CASR'
        assert results['NM_000388.3:c.3031C=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000379.2:p.(Gln1011=)', 'slr': 'NP_000379.2:p.(Q1011=)'}
        assert results['NM_000388.3:c.3031C=']['submitted_variant'] == '3-122003832-G-C'
        assert results['NM_000388.3:c.3031C=']['genome_context_intronic_sequence'] == ''
        assert results['NM_000388.3:c.3031C=']['hgvs_lrg_variant'] == ''
        assert results['NM_000388.3:c.3031C=']['hgvs_transcript_variant'] == 'NM_000388.3:c.3031C='
        assert results['NM_000388.3:c.3031C=']['hgvs_refseqgene_variant'] == 'NG_009058.1:g.106303C='
        assert results['NM_000388.3:c.3031C=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000003.11:g.122003832G>C', 'vcf': {'chr': 'chr3', 'ref': 'G', 'pos': '122003832', 'alt': 'C'}}
        assert results['NM_000388.3:c.3031C=']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000003.12:g.122284985G>C', 'vcf': {'chr': 'chr3', 'ref': 'G', 'pos': '122284985', 'alt': 'C'}}
        assert results['NM_000388.3:c.3031C=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000003.11:g.122003832G>C', 'vcf': {'chr': '3', 'ref': 'G', 'pos': '122003832', 'alt': 'C'}}
        assert results['NM_000388.3:c.3031C=']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000003.12:g.122284985G>C', 'vcf': {'chr': '3', 'ref': 'G', 'pos': '122284985', 'alt': 'C'}}


    def test_variant285(self):
        variant = '4-153332910-C-CAGG'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_033632.3:c.45_46insCCT' in results.keys()
        assert results['NM_033632.3:c.45_46insCCT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_033632.3:c.45_46insCCT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_033632.3:c.45_46insCCT']['alt_genomic_loci'] == []
        assert results['NM_033632.3:c.45_46insCCT']['gene_symbol'] == 'FBXW7'
        assert results['NM_033632.3:c.45_46insCCT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_361014.1:p.(Thr15_Gly16insPro)', 'slr': 'NP_361014.1:p.(T15_G16insP)'}
        assert results['NM_033632.3:c.45_46insCCT']['submitted_variant'] == '4-153332910-C-CAGG'
        assert results['NM_033632.3:c.45_46insCCT']['genome_context_intronic_sequence'] == ''
        assert results['NM_033632.3:c.45_46insCCT']['hgvs_lrg_variant'] == ''
        assert results['NM_033632.3:c.45_46insCCT']['hgvs_transcript_variant'] == 'NM_033632.3:c.45_46insCCT'
        assert results['NM_033632.3:c.45_46insCCT']['hgvs_refseqgene_variant'] == 'NG_029466.1:g.128262_128263insCCT'
        assert results['NM_033632.3:c.45_46insCCT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000004.11:g.153332910_153332911insAGG', 'vcf': {'chr': 'chr4', 'ref': 'C', 'pos': '153332910', 'alt': u'CAGG'}}
        assert results['NM_033632.3:c.45_46insCCT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000004.12:g.152411758_152411759insAGG', 'vcf': {'chr': 'chr4', 'ref': 'C', 'pos': '152411758', 'alt': u'CAGG'}}
        assert results['NM_033632.3:c.45_46insCCT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000004.11:g.153332910_153332911insAGG', 'vcf': {'chr': '4', 'ref': 'C', 'pos': '153332910', 'alt': u'CAGG'}}
        assert results['NM_033632.3:c.45_46insCCT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000004.12:g.152411758_152411759insAGG', 'vcf': {'chr': '4', 'ref': 'C', 'pos': '152411758', 'alt': u'CAGG'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001257069.1:c.45_46insCCT' in results.keys()
        assert results['NM_001257069.1:c.45_46insCCT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001257069.1:c.45_46insCCT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001257069.1:c.45_46insCCT']['alt_genomic_loci'] == []
        assert results['NM_001257069.1:c.45_46insCCT']['gene_symbol'] == 'FBXW7'
        assert results['NM_001257069.1:c.45_46insCCT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001243998.1:p.(Thr15_Gly16insPro)', 'slr': 'NP_001243998.1:p.(T15_G16insP)'}
        assert results['NM_001257069.1:c.45_46insCCT']['submitted_variant'] == '4-153332910-C-CAGG'
        assert results['NM_001257069.1:c.45_46insCCT']['genome_context_intronic_sequence'] == ''
        assert results['NM_001257069.1:c.45_46insCCT']['hgvs_lrg_variant'] == ''
        assert results['NM_001257069.1:c.45_46insCCT']['hgvs_transcript_variant'] == 'NM_001257069.1:c.45_46insCCT'
        assert results['NM_001257069.1:c.45_46insCCT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001257069.1:c.45_46insCCT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000004.11:g.153332910_153332911insAGG', 'vcf': {'chr': 'chr4', 'ref': 'C', 'pos': '153332910', 'alt': u'CAGG'}}
        assert results['NM_001257069.1:c.45_46insCCT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000004.12:g.152411758_152411759insAGG', 'vcf': {'chr': 'chr4', 'ref': 'C', 'pos': '152411758', 'alt': u'CAGG'}}
        assert results['NM_001257069.1:c.45_46insCCT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000004.11:g.153332910_153332911insAGG', 'vcf': {'chr': '4', 'ref': 'C', 'pos': '153332910', 'alt': u'CAGG'}}
        assert results['NM_001257069.1:c.45_46insCCT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000004.12:g.152411758_152411759insAGG', 'vcf': {'chr': '4', 'ref': 'C', 'pos': '152411758', 'alt': u'CAGG'}}

        assert 'NM_001349798.2:c.45_46insCCT' in results.keys()
        assert results['NM_001349798.2:c.45_46insCCT']['hgvs_lrg_transcript_variant'] == 'LRG_1141t1:c.45_46insCCT'
        assert results['NM_001349798.2:c.45_46insCCT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001349798.2:c.45_46insCCT']['alt_genomic_loci'] == []
        assert results['NM_001349798.2:c.45_46insCCT']['gene_symbol'] == 'FBXW7'
        assert results['NM_001349798.2:c.45_46insCCT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001336727.1:p.(Thr15_Gly16insPro)', 'slr': 'NP_001336727.1:p.(T15_G16insP)'}
        assert results['NM_001349798.2:c.45_46insCCT']['submitted_variant'] == '4-153332910-C-CAGG'
        assert results['NM_001349798.2:c.45_46insCCT']['genome_context_intronic_sequence'] == ''
        assert results['NM_001349798.2:c.45_46insCCT']['hgvs_lrg_variant'] == ''
        assert results['NM_001349798.2:c.45_46insCCT']['hgvs_transcript_variant'] == 'NM_001349798.2:c.45_46insCCT'
        assert results['NM_001349798.2:c.45_46insCCT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001349798.2:c.45_46insCCT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000004.11:g.153332910_153332911insAGG', 'vcf': {'chr': 'chr4', 'ref': 'C', 'pos': '153332910', 'alt': u'CAGG'}}
        assert 'hg38' not in results['NM_001349798.2:c.45_46insCCT']['primary_assembly_loci'].keys()
        assert results['NM_001349798.2:c.45_46insCCT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000004.11:g.153332910_153332911insAGG', 'vcf': {'chr': '4', 'ref': 'C', 'pos': '153332910', 'alt': u'CAGG'}}
        assert 'grch38' not in results['NM_001349798.2:c.45_46insCCT']['primary_assembly_loci'].keys()

        assert 'NM_001349798.1:c.45_46insCCT' in results.keys()
        assert results['NM_001349798.1:c.45_46insCCT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001349798.1:c.45_46insCCT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001349798.1:c.45_46insCCT']['alt_genomic_loci'] == []
        assert results['NM_001349798.1:c.45_46insCCT']['gene_symbol'] == 'FBXW7'
        assert results['NM_001349798.1:c.45_46insCCT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_361014.1:p.(Thr15_Gly16insPro)', 'slr': 'NP_361014.1:p.(T15_G16insP)'}
        assert results['NM_001349798.1:c.45_46insCCT']['submitted_variant'] == '4-153332910-C-CAGG'
        assert results['NM_001349798.1:c.45_46insCCT']['genome_context_intronic_sequence'] == ''
        assert results['NM_001349798.1:c.45_46insCCT']['hgvs_lrg_variant'] == ''
        assert results['NM_001349798.1:c.45_46insCCT']['hgvs_transcript_variant'] == 'NM_001349798.1:c.45_46insCCT'
        assert results['NM_001349798.1:c.45_46insCCT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001349798.1:c.45_46insCCT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000004.11:g.153332910_153332911insAGG', 'vcf': {'chr': 'chr4', 'ref': 'C', 'pos': '153332910', 'alt': u'CAGG'}}
        assert results['NM_001349798.1:c.45_46insCCT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000004.12:g.152411758_152411759insAGG', 'vcf': {'chr': 'chr4', 'ref': 'C', 'pos': '152411758', 'alt': u'CAGG'}}
        assert results['NM_001349798.1:c.45_46insCCT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000004.11:g.153332910_153332911insAGG', 'vcf': {'chr': '4', 'ref': 'C', 'pos': '153332910', 'alt': u'CAGG'}}
        assert results['NM_001349798.1:c.45_46insCCT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000004.12:g.152411758_152411759insAGG', 'vcf': {'chr': '4', 'ref': 'C', 'pos': '152411758', 'alt': u'CAGG'}}


    def test_variant286(self):
        variant = '5-1295183-G-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'intergenic'
        assert 'Intergenic_Variant_1' in results.keys()
        assert results['Intergenic_Variant_1']['hgvs_lrg_transcript_variant'] == ''
        assert results['Intergenic_Variant_1']['refseqgene_context_intronic_sequence'] == ''
        assert results['Intergenic_Variant_1']['alt_genomic_loci'] == []
        assert results['Intergenic_Variant_1']['gene_symbol'] == ''
        assert results['Intergenic_Variant_1']['hgvs_predicted_protein_consequence'] == {'tlr': '', 'slr': ''}
        assert results['Intergenic_Variant_1']['submitted_variant'] == '5-1295183-G-A'
        assert results['Intergenic_Variant_1']['genome_context_intronic_sequence'] == ''
        assert results['Intergenic_Variant_1']['hgvs_lrg_variant'] == 'LRG_343:g.4980C>T'
        assert results['Intergenic_Variant_1']['hgvs_transcript_variant'] == ''
        assert results['Intergenic_Variant_1']['hgvs_refseqgene_variant'] == 'NG_009265.1:g.4980C>T'
        assert results['Intergenic_Variant_1']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.1295183G>A', 'vcf': {'chr': 'chr5', 'ref': 'G', 'pos': '1295183', 'alt': 'A'}}
        assert results['Intergenic_Variant_1']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000005.10:g.1295068G>A', 'vcf': {'chr': 'chr5', 'ref': 'G', 'pos': '1295068', 'alt': 'A'}}
        assert results['Intergenic_Variant_1']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.1295183G>A', 'vcf': {'chr': '5', 'ref': 'G', 'pos': '1295183', 'alt': 'A'}}
        assert results['Intergenic_Variant_1']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000005.10:g.1295068G>A', 'vcf': {'chr': '5', 'ref': 'G', 'pos': '1295068', 'alt': 'A'}}


    def test_variant287(self):
        variant = '5-77396835-TTTC-T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_003664.4:c.2409_2411del' in results.keys()
        assert results['NM_003664.4:c.2409_2411del']['hgvs_lrg_transcript_variant'] == 'LRG_170t1:c.2409_2411del'
        assert results['NM_003664.4:c.2409_2411del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003664.4:c.2409_2411del']['alt_genomic_loci'] == []
        assert results['NM_003664.4:c.2409_2411del']['gene_symbol'] == 'AP3B1'
        assert results['NM_003664.4:c.2409_2411del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003655.3(LRG_170p1):p.(Lys804del)', 'slr': 'NP_003655.3:p.(K804del)'}
        assert results['NM_003664.4:c.2409_2411del']['submitted_variant'] == '5-77396835-TTTC-T'
        assert results['NM_003664.4:c.2409_2411del']['genome_context_intronic_sequence'] == ''
        assert results['NM_003664.4:c.2409_2411del']['hgvs_lrg_variant'] == 'LRG_170:g.198691_198693del'
        assert results['NM_003664.4:c.2409_2411del']['hgvs_transcript_variant'] == 'NM_003664.4:c.2409_2411del'
        assert results['NM_003664.4:c.2409_2411del']['hgvs_refseqgene_variant'] == 'NG_007268.1:g.198691_198693del'
        assert results['NM_003664.4:c.2409_2411del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.77396836_77396838del', 'vcf': {'chr': 'chr5', 'ref': 'TTTC', 'pos': '77396835', 'alt': 'T'}}
        assert results['NM_003664.4:c.2409_2411del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000005.10:g.78101012_78101014del', 'vcf': {'chr': 'chr5', 'ref': 'TTTC', 'pos': '78101011', 'alt': 'T'}}
        assert results['NM_003664.4:c.2409_2411del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.77396836_77396838del', 'vcf': {'chr': '5', 'ref': 'TTTC', 'pos': '77396835', 'alt': 'T'}}
        assert results['NM_003664.4:c.2409_2411del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000005.10:g.78101012_78101014del', 'vcf': {'chr': '5', 'ref': 'TTTC', 'pos': '78101011', 'alt': 'T'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_003664.3:c.2409_2411del' in results.keys()
        assert results['NM_003664.3:c.2409_2411del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003664.3:c.2409_2411del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003664.3:c.2409_2411del']['alt_genomic_loci'] == []
        assert results['NM_003664.3:c.2409_2411del']['gene_symbol'] == 'AP3B1'
        assert results['NM_003664.3:c.2409_2411del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003655.3(LRG_170p1):p.(Lys804del)', 'slr': 'NP_003655.3:p.(K804del)'}
        assert results['NM_003664.3:c.2409_2411del']['submitted_variant'] == '5-77396835-TTTC-T'
        assert results['NM_003664.3:c.2409_2411del']['genome_context_intronic_sequence'] == ''
        assert results['NM_003664.3:c.2409_2411del']['hgvs_lrg_variant'] == ''
        assert results['NM_003664.3:c.2409_2411del']['hgvs_transcript_variant'] == 'NM_003664.3:c.2409_2411del'
        assert results['NM_003664.3:c.2409_2411del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_003664.3:c.2409_2411del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.77396836_77396838del', 'vcf': {'chr': 'chr5', 'ref': 'TTTC', 'pos': '77396835', 'alt': 'T'}}
        assert 'hg38' not in results['NM_003664.3:c.2409_2411del']['primary_assembly_loci'].keys()
        assert results['NM_003664.3:c.2409_2411del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.77396836_77396838del', 'vcf': {'chr': '5', 'ref': 'TTTC', 'pos': '77396835', 'alt': 'T'}}
        assert 'grch38' not in results['NM_003664.3:c.2409_2411del']['primary_assembly_loci'].keys()

        assert 'NM_001271769.1:c.2262_2264del' in results.keys()
        assert results['NM_001271769.1:c.2262_2264del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001271769.1:c.2262_2264del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001271769.1:c.2262_2264del']['alt_genomic_loci'] == []
        assert results['NM_001271769.1:c.2262_2264del']['gene_symbol'] == 'AP3B1'
        assert results['NM_001271769.1:c.2262_2264del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001258698.1:p.(Lys755del)', 'slr': 'NP_001258698.1:p.(K755del)'}
        assert results['NM_001271769.1:c.2262_2264del']['submitted_variant'] == '5-77396835-TTTC-T'
        assert results['NM_001271769.1:c.2262_2264del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001271769.1:c.2262_2264del']['hgvs_lrg_variant'] == ''
        assert results['NM_001271769.1:c.2262_2264del']['hgvs_transcript_variant'] == 'NM_001271769.1:c.2262_2264del'
        assert results['NM_001271769.1:c.2262_2264del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001271769.1:c.2262_2264del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.77396836_77396838del', 'vcf': {'chr': 'chr5', 'ref': 'TTTC', 'pos': '77396835', 'alt': 'T'}}
        assert results['NM_001271769.1:c.2262_2264del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000005.10:g.78101012_78101014del', 'vcf': {'chr': 'chr5', 'ref': 'TTTC', 'pos': '78101011', 'alt': 'T'}}
        assert results['NM_001271769.1:c.2262_2264del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.77396836_77396838del', 'vcf': {'chr': '5', 'ref': 'TTTC', 'pos': '77396835', 'alt': 'T'}}
        assert results['NM_001271769.1:c.2262_2264del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000005.10:g.78101012_78101014del', 'vcf': {'chr': '5', 'ref': 'TTTC', 'pos': '78101011', 'alt': 'T'}}


    def test_variant288(self):
        variant = '5-118811422-GGTGA-G'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_000414.3:c.302+3_302+6del' in results.keys()
        assert results['NM_000414.3:c.302+3_302+6del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000414.3:c.302+3_302+6del']['refseqgene_context_intronic_sequence'] == 'NG_008182.1(NM_000414.3):c.302+3_302+6del'
        assert results['NM_000414.3:c.302+3_302+6del']['alt_genomic_loci'] == []
        assert results['NM_000414.3:c.302+3_302+6del']['gene_symbol'] == 'HSD17B4'
        assert results['NM_000414.3:c.302+3_302+6del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000405.1:p.?', 'slr': 'NP_000405.1:p.?'}
        assert results['NM_000414.3:c.302+3_302+6del']['submitted_variant'] == '5-118811422-GGTGA-G'
        assert results['NM_000414.3:c.302+3_302+6del']['genome_context_intronic_sequence'] == 'NC_000005.9(NM_000414.3):c.302+3_302+6del'
        assert results['NM_000414.3:c.302+3_302+6del']['hgvs_lrg_variant'] == ''
        assert results['NM_000414.3:c.302+3_302+6del']['hgvs_transcript_variant'] == 'NM_000414.3:c.302+3_302+6del'
        assert results['NM_000414.3:c.302+3_302+6del']['hgvs_refseqgene_variant'] == 'NG_008182.1:g.28278_28281del'
        assert results['NM_000414.3:c.302+3_302+6del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.118811425_118811428del', 'vcf': {'chr': 'chr5', 'ref': 'GGTGA', 'pos': '118811422', 'alt': 'G'}}
        assert results['NM_000414.3:c.302+3_302+6del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000005.10:g.119475730_119475733del', 'vcf': {'chr': 'chr5', 'ref': 'GGTGA', 'pos': '119475727', 'alt': 'G'}}
        assert results['NM_000414.3:c.302+3_302+6del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.118811425_118811428del', 'vcf': {'chr': '5', 'ref': 'GGTGA', 'pos': '118811422', 'alt': 'G'}}
        assert results['NM_000414.3:c.302+3_302+6del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000005.10:g.119475730_119475733del', 'vcf': {'chr': '5', 'ref': 'GGTGA', 'pos': '119475727', 'alt': 'G'}}

        assert 'NM_001292028.1:c.-110+3_-110+6del' in results.keys()
        assert results['NM_001292028.1:c.-110+3_-110+6del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001292028.1:c.-110+3_-110+6del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001292028.1:c.-110+3_-110+6del']['alt_genomic_loci'] == []
        assert results['NM_001292028.1:c.-110+3_-110+6del']['gene_symbol'] == 'HSD17B4'
        assert results['NM_001292028.1:c.-110+3_-110+6del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001278957.1:p.?', 'slr': 'NP_001278957.1:p.?'}
        assert results['NM_001292028.1:c.-110+3_-110+6del']['submitted_variant'] == '5-118811422-GGTGA-G'
        assert results['NM_001292028.1:c.-110+3_-110+6del']['genome_context_intronic_sequence'] == 'NC_000005.9(NM_001292028.1):c.-110+3_-110+6del'
        assert results['NM_001292028.1:c.-110+3_-110+6del']['hgvs_lrg_variant'] == ''
        assert results['NM_001292028.1:c.-110+3_-110+6del']['hgvs_transcript_variant'] == 'NM_001292028.1:c.-110+3_-110+6del'
        assert results['NM_001292028.1:c.-110+3_-110+6del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001292028.1:c.-110+3_-110+6del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.118811425_118811428del', 'vcf': {'chr': 'chr5', 'ref': 'GGTGA', 'pos': '118811422', 'alt': 'G'}}
        assert results['NM_001292028.1:c.-110+3_-110+6del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000005.10:g.119475730_119475733del', 'vcf': {'chr': 'chr5', 'ref': 'GGTGA', 'pos': '119475727', 'alt': 'G'}}
        assert results['NM_001292028.1:c.-110+3_-110+6del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.118811425_118811428del', 'vcf': {'chr': '5', 'ref': 'GGTGA', 'pos': '118811422', 'alt': 'G'}}
        assert results['NM_001292028.1:c.-110+3_-110+6del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000005.10:g.119475730_119475733del', 'vcf': {'chr': '5', 'ref': 'GGTGA', 'pos': '119475727', 'alt': 'G'}}

        assert 'NM_001199291.2:c.377+3_377+6del' in results.keys()
        assert results['NM_001199291.2:c.377+3_377+6del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001199291.2:c.377+3_377+6del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001199291.2:c.377+3_377+6del']['alt_genomic_loci'] == []
        assert results['NM_001199291.2:c.377+3_377+6del']['gene_symbol'] == 'HSD17B4'
        assert results['NM_001199291.2:c.377+3_377+6del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001186220.1:p.?', 'slr': 'NP_001186220.1:p.?'}
        assert results['NM_001199291.2:c.377+3_377+6del']['submitted_variant'] == '5-118811422-GGTGA-G'
        assert results['NM_001199291.2:c.377+3_377+6del']['genome_context_intronic_sequence'] == 'NC_000005.9(NM_001199291.2):c.377+3_377+6del'
        assert results['NM_001199291.2:c.377+3_377+6del']['hgvs_lrg_variant'] == ''
        assert results['NM_001199291.2:c.377+3_377+6del']['hgvs_transcript_variant'] == 'NM_001199291.2:c.377+3_377+6del'
        assert results['NM_001199291.2:c.377+3_377+6del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001199291.2:c.377+3_377+6del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.118811425_118811428del', 'vcf': {'chr': 'chr5', 'ref': 'GGTGA', 'pos': '118811422', 'alt': 'G'}}
        assert results['NM_001199291.2:c.377+3_377+6del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000005.10:g.119475730_119475733del', 'vcf': {'chr': 'chr5', 'ref': 'GGTGA', 'pos': '119475727', 'alt': 'G'}}
        assert results['NM_001199291.2:c.377+3_377+6del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.118811425_118811428del', 'vcf': {'chr': '5', 'ref': 'GGTGA', 'pos': '118811422', 'alt': 'G'}}
        assert results['NM_001199291.2:c.377+3_377+6del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000005.10:g.119475730_119475733del', 'vcf': {'chr': '5', 'ref': 'GGTGA', 'pos': '119475727', 'alt': 'G'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001292027.1:c.230+3_230+6del' in results.keys()
        assert results['NM_001292027.1:c.230+3_230+6del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001292027.1:c.230+3_230+6del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001292027.1:c.230+3_230+6del']['alt_genomic_loci'] == []
        assert results['NM_001292027.1:c.230+3_230+6del']['gene_symbol'] == 'HSD17B4'
        assert results['NM_001292027.1:c.230+3_230+6del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001278956.1:p.?', 'slr': 'NP_001278956.1:p.?'}
        assert results['NM_001292027.1:c.230+3_230+6del']['submitted_variant'] == '5-118811422-GGTGA-G'
        assert results['NM_001292027.1:c.230+3_230+6del']['genome_context_intronic_sequence'] == 'NC_000005.9(NM_001292027.1):c.230+3_230+6del'
        assert results['NM_001292027.1:c.230+3_230+6del']['hgvs_lrg_variant'] == ''
        assert results['NM_001292027.1:c.230+3_230+6del']['hgvs_transcript_variant'] == 'NM_001292027.1:c.230+3_230+6del'
        assert results['NM_001292027.1:c.230+3_230+6del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001292027.1:c.230+3_230+6del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.118811425_118811428del', 'vcf': {'chr': 'chr5', 'ref': 'GGTGA', 'pos': '118811422', 'alt': 'G'}}
        assert results['NM_001292027.1:c.230+3_230+6del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000005.10:g.119475730_119475733del', 'vcf': {'chr': 'chr5', 'ref': 'GGTGA', 'pos': '119475727', 'alt': 'G'}}
        assert results['NM_001292027.1:c.230+3_230+6del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.118811425_118811428del', 'vcf': {'chr': '5', 'ref': 'GGTGA', 'pos': '118811422', 'alt': 'G'}}
        assert results['NM_001292027.1:c.230+3_230+6del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000005.10:g.119475730_119475733del', 'vcf': {'chr': '5', 'ref': 'GGTGA', 'pos': '119475727', 'alt': 'G'}}

        assert 'NM_001199291.1:c.377+3_377+6del' in results.keys()
        assert results['NM_001199291.1:c.377+3_377+6del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001199291.1:c.377+3_377+6del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001199291.1:c.377+3_377+6del']['alt_genomic_loci'] == []
        assert results['NM_001199291.1:c.377+3_377+6del']['gene_symbol'] == 'HSD17B4'
        assert results['NM_001199291.1:c.377+3_377+6del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001186220.1:p.?', 'slr': 'NP_001186220.1:p.?'}
        assert results['NM_001199291.1:c.377+3_377+6del']['submitted_variant'] == '5-118811422-GGTGA-G'
        assert results['NM_001199291.1:c.377+3_377+6del']['genome_context_intronic_sequence'] == 'NC_000005.9(NM_001199291.1):c.377+3_377+6del'
        assert results['NM_001199291.1:c.377+3_377+6del']['hgvs_lrg_variant'] == ''
        assert results['NM_001199291.1:c.377+3_377+6del']['hgvs_transcript_variant'] == 'NM_001199291.1:c.377+3_377+6del'
        assert results['NM_001199291.1:c.377+3_377+6del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001199291.1:c.377+3_377+6del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.118811425_118811428del', 'vcf': {'chr': 'chr5', 'ref': 'GGTGA', 'pos': '118811422', 'alt': 'G'}}
        assert 'hg38' not in results['NM_001199291.1:c.377+3_377+6del']['primary_assembly_loci'].keys()
        assert results['NM_001199291.1:c.377+3_377+6del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.118811425_118811428del', 'vcf': {'chr': '5', 'ref': 'GGTGA', 'pos': '118811422', 'alt': 'G'}}
        assert 'grch38' not in results['NM_001199291.1:c.377+3_377+6del']['primary_assembly_loci'].keys()

        assert 'NM_001199292.1:c.248+3_248+6del' in results.keys()
        assert results['NM_001199292.1:c.248+3_248+6del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001199292.1:c.248+3_248+6del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001199292.1:c.248+3_248+6del']['alt_genomic_loci'] == []
        assert results['NM_001199292.1:c.248+3_248+6del']['gene_symbol'] == 'HSD17B4'
        assert results['NM_001199292.1:c.248+3_248+6del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001186221.1:p.?', 'slr': 'NP_001186221.1:p.?'}
        assert results['NM_001199292.1:c.248+3_248+6del']['submitted_variant'] == '5-118811422-GGTGA-G'
        assert results['NM_001199292.1:c.248+3_248+6del']['genome_context_intronic_sequence'] == 'NC_000005.9(NM_001199292.1):c.248+3_248+6del'
        assert results['NM_001199292.1:c.248+3_248+6del']['hgvs_lrg_variant'] == ''
        assert results['NM_001199292.1:c.248+3_248+6del']['hgvs_transcript_variant'] == 'NM_001199292.1:c.248+3_248+6del'
        assert results['NM_001199292.1:c.248+3_248+6del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001199292.1:c.248+3_248+6del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.118811425_118811428del', 'vcf': {'chr': 'chr5', 'ref': 'GGTGA', 'pos': '118811422', 'alt': 'G'}}
        assert results['NM_001199292.1:c.248+3_248+6del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000005.10:g.119475730_119475733del', 'vcf': {'chr': 'chr5', 'ref': 'GGTGA', 'pos': '119475727', 'alt': 'G'}}
        assert results['NM_001199292.1:c.248+3_248+6del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.118811425_118811428del', 'vcf': {'chr': '5', 'ref': 'GGTGA', 'pos': '118811422', 'alt': 'G'}}
        assert results['NM_001199292.1:c.248+3_248+6del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000005.10:g.119475730_119475733del', 'vcf': {'chr': '5', 'ref': 'GGTGA', 'pos': '119475727', 'alt': 'G'}}


    def test_variant289(self):
        variant = '5-118811422-GGTGAG-G'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001292028.1:c.-110+1_-110+5del' in results.keys()
        assert results['NM_001292028.1:c.-110+1_-110+5del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001292028.1:c.-110+1_-110+5del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001292028.1:c.-110+1_-110+5del']['alt_genomic_loci'] == []
        assert results['NM_001292028.1:c.-110+1_-110+5del']['gene_symbol'] == 'HSD17B4'
        assert results['NM_001292028.1:c.-110+1_-110+5del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001278957.1:p.?', 'slr': 'NP_001278957.1:p.?'}
        assert results['NM_001292028.1:c.-110+1_-110+5del']['submitted_variant'] == '5-118811422-GGTGAG-G'
        assert results['NM_001292028.1:c.-110+1_-110+5del']['genome_context_intronic_sequence'] == 'NC_000005.9(NM_001292028.1):c.-110+1_-110+5del'
        assert results['NM_001292028.1:c.-110+1_-110+5del']['hgvs_lrg_variant'] == ''
        assert results['NM_001292028.1:c.-110+1_-110+5del']['hgvs_transcript_variant'] == 'NM_001292028.1:c.-110+1_-110+5del'
        assert results['NM_001292028.1:c.-110+1_-110+5del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001292028.1:c.-110+1_-110+5del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.118811423_118811427del', 'vcf': {'chr': 'chr5', 'ref': 'GGGTGA', 'pos': '118811421', 'alt': 'G'}}
        assert results['NM_001292028.1:c.-110+1_-110+5del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000005.10:g.119475728_119475732del', 'vcf': {'chr': 'chr5', 'ref': 'GGGTGA', 'pos': '119475726', 'alt': 'G'}}
        assert results['NM_001292028.1:c.-110+1_-110+5del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.118811423_118811427del', 'vcf': {'chr': '5', 'ref': 'GGGTGA', 'pos': '118811421', 'alt': 'G'}}
        assert results['NM_001292028.1:c.-110+1_-110+5del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000005.10:g.119475728_119475732del', 'vcf': {'chr': '5', 'ref': 'GGGTGA', 'pos': '119475726', 'alt': 'G'}}

        assert 'NM_000414.3:c.302+1_302+5del' in results.keys()
        assert results['NM_000414.3:c.302+1_302+5del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000414.3:c.302+1_302+5del']['refseqgene_context_intronic_sequence'] == 'NG_008182.1(NM_000414.3):c.302+1_302+5del'
        assert results['NM_000414.3:c.302+1_302+5del']['alt_genomic_loci'] == []
        assert results['NM_000414.3:c.302+1_302+5del']['gene_symbol'] == 'HSD17B4'
        assert results['NM_000414.3:c.302+1_302+5del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000405.1:p.?', 'slr': 'NP_000405.1:p.?'}
        assert results['NM_000414.3:c.302+1_302+5del']['submitted_variant'] == '5-118811422-GGTGAG-G'
        assert results['NM_000414.3:c.302+1_302+5del']['genome_context_intronic_sequence'] == 'NC_000005.9(NM_000414.3):c.302+1_302+5del'
        assert results['NM_000414.3:c.302+1_302+5del']['hgvs_lrg_variant'] == ''
        assert results['NM_000414.3:c.302+1_302+5del']['hgvs_transcript_variant'] == 'NM_000414.3:c.302+1_302+5del'
        assert results['NM_000414.3:c.302+1_302+5del']['hgvs_refseqgene_variant'] == 'NG_008182.1:g.28276_28280del'
        assert results['NM_000414.3:c.302+1_302+5del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.118811423_118811427del', 'vcf': {'chr': 'chr5', 'ref': 'GGGTGA', 'pos': '118811421', 'alt': 'G'}}
        assert results['NM_000414.3:c.302+1_302+5del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000005.10:g.119475728_119475732del', 'vcf': {'chr': 'chr5', 'ref': 'GGGTGA', 'pos': '119475726', 'alt': 'G'}}
        assert results['NM_000414.3:c.302+1_302+5del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.118811423_118811427del', 'vcf': {'chr': '5', 'ref': 'GGGTGA', 'pos': '118811421', 'alt': 'G'}}
        assert results['NM_000414.3:c.302+1_302+5del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000005.10:g.119475728_119475732del', 'vcf': {'chr': '5', 'ref': 'GGGTGA', 'pos': '119475726', 'alt': 'G'}}

        assert 'NM_001199291.2:c.377+1_377+5del' in results.keys()
        assert results['NM_001199291.2:c.377+1_377+5del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001199291.2:c.377+1_377+5del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001199291.2:c.377+1_377+5del']['alt_genomic_loci'] == []
        assert results['NM_001199291.2:c.377+1_377+5del']['gene_symbol'] == 'HSD17B4'
        assert results['NM_001199291.2:c.377+1_377+5del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001186220.1:p.?', 'slr': 'NP_001186220.1:p.?'}
        assert results['NM_001199291.2:c.377+1_377+5del']['submitted_variant'] == '5-118811422-GGTGAG-G'
        assert results['NM_001199291.2:c.377+1_377+5del']['genome_context_intronic_sequence'] == 'NC_000005.9(NM_001199291.2):c.377+1_377+5del'
        assert results['NM_001199291.2:c.377+1_377+5del']['hgvs_lrg_variant'] == ''
        assert results['NM_001199291.2:c.377+1_377+5del']['hgvs_transcript_variant'] == 'NM_001199291.2:c.377+1_377+5del'
        assert results['NM_001199291.2:c.377+1_377+5del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001199291.2:c.377+1_377+5del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.118811423_118811427del', 'vcf': {'chr': 'chr5', 'ref': 'GGGTGA', 'pos': '118811421', 'alt': 'G'}}
        assert results['NM_001199291.2:c.377+1_377+5del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000005.10:g.119475728_119475732del', 'vcf': {'chr': 'chr5', 'ref': 'GGGTGA', 'pos': '119475726', 'alt': 'G'}}
        assert results['NM_001199291.2:c.377+1_377+5del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.118811423_118811427del', 'vcf': {'chr': '5', 'ref': 'GGGTGA', 'pos': '118811421', 'alt': 'G'}}
        assert results['NM_001199291.2:c.377+1_377+5del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000005.10:g.119475728_119475732del', 'vcf': {'chr': '5', 'ref': 'GGGTGA', 'pos': '119475726', 'alt': 'G'}}

        assert 'NM_001199292.1:c.248+1_248+5del' in results.keys()
        assert results['NM_001199292.1:c.248+1_248+5del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001199292.1:c.248+1_248+5del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001199292.1:c.248+1_248+5del']['alt_genomic_loci'] == []
        assert results['NM_001199292.1:c.248+1_248+5del']['gene_symbol'] == 'HSD17B4'
        assert results['NM_001199292.1:c.248+1_248+5del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001186221.1:p.?', 'slr': 'NP_001186221.1:p.?'}
        assert results['NM_001199292.1:c.248+1_248+5del']['submitted_variant'] == '5-118811422-GGTGAG-G'
        assert results['NM_001199292.1:c.248+1_248+5del']['genome_context_intronic_sequence'] == 'NC_000005.9(NM_001199292.1):c.248+1_248+5del'
        assert results['NM_001199292.1:c.248+1_248+5del']['hgvs_lrg_variant'] == ''
        assert results['NM_001199292.1:c.248+1_248+5del']['hgvs_transcript_variant'] == 'NM_001199292.1:c.248+1_248+5del'
        assert results['NM_001199292.1:c.248+1_248+5del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001199292.1:c.248+1_248+5del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.118811423_118811427del', 'vcf': {'chr': 'chr5', 'ref': 'GGGTGA', 'pos': '118811421', 'alt': 'G'}}
        assert results['NM_001199292.1:c.248+1_248+5del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000005.10:g.119475728_119475732del', 'vcf': {'chr': 'chr5', 'ref': 'GGGTGA', 'pos': '119475726', 'alt': 'G'}}
        assert results['NM_001199292.1:c.248+1_248+5del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.118811423_118811427del', 'vcf': {'chr': '5', 'ref': 'GGGTGA', 'pos': '118811421', 'alt': 'G'}}
        assert results['NM_001199292.1:c.248+1_248+5del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000005.10:g.119475728_119475732del', 'vcf': {'chr': '5', 'ref': 'GGGTGA', 'pos': '119475726', 'alt': 'G'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001199291.1:c.377+1_377+5del' in results.keys()
        assert results['NM_001199291.1:c.377+1_377+5del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001199291.1:c.377+1_377+5del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001199291.1:c.377+1_377+5del']['alt_genomic_loci'] == []
        assert results['NM_001199291.1:c.377+1_377+5del']['gene_symbol'] == 'HSD17B4'
        assert results['NM_001199291.1:c.377+1_377+5del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001186220.1:p.?', 'slr': 'NP_001186220.1:p.?'}
        assert results['NM_001199291.1:c.377+1_377+5del']['submitted_variant'] == '5-118811422-GGTGAG-G'
        assert results['NM_001199291.1:c.377+1_377+5del']['genome_context_intronic_sequence'] == 'NC_000005.9(NM_001199291.1):c.377+1_377+5del'
        assert results['NM_001199291.1:c.377+1_377+5del']['hgvs_lrg_variant'] == ''
        assert results['NM_001199291.1:c.377+1_377+5del']['hgvs_transcript_variant'] == 'NM_001199291.1:c.377+1_377+5del'
        assert results['NM_001199291.1:c.377+1_377+5del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001199291.1:c.377+1_377+5del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.118811423_118811427del', 'vcf': {'chr': 'chr5', 'ref': 'GGGTGA', 'pos': '118811421', 'alt': 'G'}}
        assert 'hg38' not in results['NM_001199291.1:c.377+1_377+5del']['primary_assembly_loci'].keys()
        assert results['NM_001199291.1:c.377+1_377+5del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.118811423_118811427del', 'vcf': {'chr': '5', 'ref': 'GGGTGA', 'pos': '118811421', 'alt': 'G'}}
        assert 'grch38' not in results['NM_001199291.1:c.377+1_377+5del']['primary_assembly_loci'].keys()

        assert 'NM_001292027.1:c.230+1_230+5del' in results.keys()
        assert results['NM_001292027.1:c.230+1_230+5del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001292027.1:c.230+1_230+5del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001292027.1:c.230+1_230+5del']['alt_genomic_loci'] == []
        assert results['NM_001292027.1:c.230+1_230+5del']['gene_symbol'] == 'HSD17B4'
        assert results['NM_001292027.1:c.230+1_230+5del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001278956.1:p.?', 'slr': 'NP_001278956.1:p.?'}
        assert results['NM_001292027.1:c.230+1_230+5del']['submitted_variant'] == '5-118811422-GGTGAG-G'
        assert results['NM_001292027.1:c.230+1_230+5del']['genome_context_intronic_sequence'] == 'NC_000005.9(NM_001292027.1):c.230+1_230+5del'
        assert results['NM_001292027.1:c.230+1_230+5del']['hgvs_lrg_variant'] == ''
        assert results['NM_001292027.1:c.230+1_230+5del']['hgvs_transcript_variant'] == 'NM_001292027.1:c.230+1_230+5del'
        assert results['NM_001292027.1:c.230+1_230+5del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001292027.1:c.230+1_230+5del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.118811423_118811427del', 'vcf': {'chr': 'chr5', 'ref': 'GGGTGA', 'pos': '118811421', 'alt': 'G'}}
        assert results['NM_001292027.1:c.230+1_230+5del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000005.10:g.119475728_119475732del', 'vcf': {'chr': 'chr5', 'ref': 'GGGTGA', 'pos': '119475726', 'alt': 'G'}}
        assert results['NM_001292027.1:c.230+1_230+5del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.118811423_118811427del', 'vcf': {'chr': '5', 'ref': 'GGGTGA', 'pos': '118811421', 'alt': 'G'}}
        assert results['NM_001292027.1:c.230+1_230+5del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000005.10:g.119475728_119475732del', 'vcf': {'chr': '5', 'ref': 'GGGTGA', 'pos': '119475726', 'alt': 'G'}}


    def test_variant290(self):
        variant = '5-131705587-CG-C'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NR_110997.1:n.21del' in results.keys()
        assert results['NR_110997.1:n.21del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_110997.1:n.21del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_110997.1:n.21del']['alt_genomic_loci'] == []
        assert results['NR_110997.1:n.21del']['gene_symbol'] == 'MIR3936HG'
        assert results['NR_110997.1:n.21del']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_110997.1:n.21del']['submitted_variant'] == '5-131705587-CG-C'
        assert results['NR_110997.1:n.21del']['genome_context_intronic_sequence'] == ''
        assert results['NR_110997.1:n.21del']['hgvs_lrg_variant'] == ''
        assert results['NR_110997.1:n.21del']['hgvs_transcript_variant'] == 'NR_110997.1:n.21del'
        assert results['NR_110997.1:n.21del']['hgvs_refseqgene_variant'] == ''
        assert results['NR_110997.1:n.21del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.131705590del', 'vcf': {'chr': 'chr5', 'ref': 'CG', 'pos': '131705587', 'alt': 'C'}}
        assert results['NR_110997.1:n.21del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000005.10:g.132369898del', 'vcf': {'chr': 'chr5', 'ref': 'CG', 'pos': '132369895', 'alt': 'C'}}
        assert results['NR_110997.1:n.21del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.131705590del', 'vcf': {'chr': '5', 'ref': 'CG', 'pos': '131705587', 'alt': 'C'}}
        assert results['NR_110997.1:n.21del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000005.10:g.132369898del', 'vcf': {'chr': '5', 'ref': 'CG', 'pos': '132369895', 'alt': 'C'}}

        assert 'NM_003060.3:c.-75del' in results.keys()
        assert results['NM_003060.3:c.-75del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003060.3:c.-75del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003060.3:c.-75del']['alt_genomic_loci'] == []
        assert results['NM_003060.3:c.-75del']['gene_symbol'] == 'SLC22A5'
        assert results['NM_003060.3:c.-75del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003051.1:p.?', 'slr': 'NP_003051.1:p.?'}
        assert results['NM_003060.3:c.-75del']['submitted_variant'] == '5-131705587-CG-C'
        assert results['NM_003060.3:c.-75del']['genome_context_intronic_sequence'] == ''
        assert results['NM_003060.3:c.-75del']['hgvs_lrg_variant'] == ''
        assert results['NM_003060.3:c.-75del']['hgvs_transcript_variant'] == 'NM_003060.3:c.-75del'
        assert results['NM_003060.3:c.-75del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_003060.3:c.-75del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.131705590del', 'vcf': {'chr': 'chr5', 'ref': 'CG', 'pos': '131705587', 'alt': 'C'}}
        assert results['NM_003060.3:c.-75del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000005.10:g.132369898del', 'vcf': {'chr': 'chr5', 'ref': 'CG', 'pos': '132369895', 'alt': 'C'}}
        assert results['NM_003060.3:c.-75del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.131705590del', 'vcf': {'chr': '5', 'ref': 'CG', 'pos': '131705587', 'alt': 'C'}}
        assert results['NM_003060.3:c.-75del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000005.10:g.132369898del', 'vcf': {'chr': '5', 'ref': 'CG', 'pos': '132369895', 'alt': 'C'}}

        assert 'NM_001308122.1:c.-75del' in results.keys()
        assert results['NM_001308122.1:c.-75del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001308122.1:c.-75del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001308122.1:c.-75del']['alt_genomic_loci'] == []
        assert results['NM_001308122.1:c.-75del']['gene_symbol'] == 'SLC22A5'
        assert results['NM_001308122.1:c.-75del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001295051.1:p.?', 'slr': 'NP_001295051.1:p.?'}
        assert results['NM_001308122.1:c.-75del']['submitted_variant'] == '5-131705587-CG-C'
        assert results['NM_001308122.1:c.-75del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001308122.1:c.-75del']['hgvs_lrg_variant'] == ''
        assert results['NM_001308122.1:c.-75del']['hgvs_transcript_variant'] == 'NM_001308122.1:c.-75del'
        assert results['NM_001308122.1:c.-75del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001308122.1:c.-75del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.131705590del', 'vcf': {'chr': 'chr5', 'ref': 'CG', 'pos': '131705587', 'alt': 'C'}}
        assert results['NM_001308122.1:c.-75del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000005.10:g.132369898del', 'vcf': {'chr': 'chr5', 'ref': 'CG', 'pos': '132369895', 'alt': 'C'}}
        assert results['NM_001308122.1:c.-75del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.131705590del', 'vcf': {'chr': '5', 'ref': 'CG', 'pos': '131705587', 'alt': 'C'}}
        assert results['NM_001308122.1:c.-75del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000005.10:g.132369898del', 'vcf': {'chr': '5', 'ref': 'CG', 'pos': '132369895', 'alt': 'C'}}


    def test_variant291(self):
        variant = '5-148406482-T-C'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_024577.3:c.2813A>G' in results.keys()
        assert results['NM_024577.3:c.2813A>G']['hgvs_lrg_transcript_variant'] == 'LRG_269t1:c.2813A>G'
        assert results['NM_024577.3:c.2813A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_024577.3:c.2813A>G']['alt_genomic_loci'] == []
        assert results['NM_024577.3:c.2813A>G']['gene_symbol'] == 'SH3TC2'
        assert results['NM_024577.3:c.2813A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_078853.2(LRG_269p1):p.(His938Arg)', 'slr': 'NP_078853.2:p.(H938R)'}
        assert results['NM_024577.3:c.2813A>G']['submitted_variant'] == '5-148406482-T-C'
        assert results['NM_024577.3:c.2813A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_024577.3:c.2813A>G']['hgvs_lrg_variant'] == 'LRG_269:g.41256A>G'
        assert results['NM_024577.3:c.2813A>G']['hgvs_transcript_variant'] == 'NM_024577.3:c.2813A>G'
        assert results['NM_024577.3:c.2813A>G']['hgvs_refseqgene_variant'] == 'NG_007947.2:g.41256A>G'
        assert results['NM_024577.3:c.2813A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000005.9:g.148406482T>C', 'vcf': {'chr': 'chr5', 'ref': u'T', 'pos': '148406482', 'alt': u'C'}}
        assert results['NM_024577.3:c.2813A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000005.10:g.149026919T>C', 'vcf': {'chr': 'chr5', 'ref': u'T', 'pos': '149026919', 'alt': u'C'}}
        assert results['NM_024577.3:c.2813A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000005.9:g.148406482T>C', 'vcf': {'chr': '5', 'ref': u'T', 'pos': '148406482', 'alt': u'C'}}
        assert results['NM_024577.3:c.2813A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000005.10:g.149026919T>C', 'vcf': {'chr': '5', 'ref': u'T', 'pos': '149026919', 'alt': u'C'}}


    def test_variant292(self):
        variant = '6-110036337-T-TCAG'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_014845.5:c.123_124insCAG' in results.keys()
        assert results['NM_014845.5:c.123_124insCAG']['hgvs_lrg_transcript_variant'] == 'LRG_241t1:c.123_124insCAG'
        assert results['NM_014845.5:c.123_124insCAG']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_014845.5:c.123_124insCAG']['alt_genomic_loci'] == []
        assert results['NM_014845.5:c.123_124insCAG']['gene_symbol'] == 'FIG4'
        assert results['NM_014845.5:c.123_124insCAG']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_055660.1(LRG_241p1):p.(Ile41_Asp42insGln)', 'slr': 'NP_055660.1:p.(I41_D42insQ)'}
        assert results['NM_014845.5:c.123_124insCAG']['submitted_variant'] == '6-110036337-T-TCAG'
        assert results['NM_014845.5:c.123_124insCAG']['genome_context_intronic_sequence'] == ''
        assert results['NM_014845.5:c.123_124insCAG']['hgvs_lrg_variant'] == 'LRG_241:g.28914_28915insCAG'
        assert results['NM_014845.5:c.123_124insCAG']['hgvs_transcript_variant'] == 'NM_014845.5:c.123_124insCAG'
        assert results['NM_014845.5:c.123_124insCAG']['hgvs_refseqgene_variant'] == 'NG_007977.1:g.28914_28915insCAG'
        assert results['NM_014845.5:c.123_124insCAG']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000006.11:g.110036337_110036338insCAG', 'vcf': {'chr': 'chr6', 'ref': 'T', 'pos': '110036337', 'alt': 'TCAG'}}
        assert results['NM_014845.5:c.123_124insCAG']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000006.12:g.109715134_109715135insCAG', 'vcf': {'chr': 'chr6', 'ref': 'T', 'pos': '109715134', 'alt': 'TCAG'}}
        assert results['NM_014845.5:c.123_124insCAG']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000006.11:g.110036337_110036338insCAG', 'vcf': {'chr': '6', 'ref': 'T', 'pos': '110036337', 'alt': 'TCAG'}}
        assert results['NM_014845.5:c.123_124insCAG']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000006.12:g.109715134_109715135insCAG', 'vcf': {'chr': '6', 'ref': 'T', 'pos': '109715134', 'alt': 'TCAG'}}

        assert results['flag'] == 'gene_variant'

    def test_variant293(self):
        variant = '6-110036337-TGAT-T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_014845.5:c.124_126del' in results.keys()
        assert results['NM_014845.5:c.124_126del']['hgvs_lrg_transcript_variant'] == 'LRG_241t1:c.124_126del'
        assert results['NM_014845.5:c.124_126del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_014845.5:c.124_126del']['alt_genomic_loci'] == []
        assert results['NM_014845.5:c.124_126del']['gene_symbol'] == 'FIG4'
        assert results['NM_014845.5:c.124_126del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_055660.1(LRG_241p1):p.(Asp42del)', 'slr': 'NP_055660.1:p.(D42del)'}
        assert results['NM_014845.5:c.124_126del']['submitted_variant'] == '6-110036337-TGAT-T'
        assert results['NM_014845.5:c.124_126del']['genome_context_intronic_sequence'] == ''
        assert results['NM_014845.5:c.124_126del']['hgvs_lrg_variant'] == 'LRG_241:g.28915_28917del'
        assert results['NM_014845.5:c.124_126del']['hgvs_transcript_variant'] == 'NM_014845.5:c.124_126del'
        assert results['NM_014845.5:c.124_126del']['hgvs_refseqgene_variant'] == 'NG_007977.1:g.28915_28917del'
        assert results['NM_014845.5:c.124_126del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000006.11:g.110036338_110036340del', 'vcf': {'chr': 'chr6', 'ref': 'TTGA', 'pos': '110036336', 'alt': 'T'}}
        assert results['NM_014845.5:c.124_126del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000006.12:g.109715135_109715137del', 'vcf': {'chr': 'chr6', 'ref': 'TTGA', 'pos': '109715133', 'alt': 'T'}}
        assert results['NM_014845.5:c.124_126del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000006.11:g.110036338_110036340del', 'vcf': {'chr': '6', 'ref': 'TTGA', 'pos': '110036336', 'alt': 'T'}}
        assert results['NM_014845.5:c.124_126del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000006.12:g.109715135_109715137del', 'vcf': {'chr': '6', 'ref': 'TTGA', 'pos': '109715133', 'alt': 'T'}}

        assert results['flag'] == 'gene_variant'

    def test_variant294(self):
        variant = '6-152651802-C-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_182961.3:c.14018G>T' in results.keys()
        assert results['NM_182961.3:c.14018G>T']['hgvs_lrg_transcript_variant'] == 'LRG_427t1:c.14018G>T'
        assert results['NM_182961.3:c.14018G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_182961.3:c.14018G>T']['alt_genomic_loci'] == []
        assert results['NM_182961.3:c.14018G>T']['gene_symbol'] == 'SYNE1'
        assert results['NM_182961.3:c.14018G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_892006.3(LRG_427p1):p.(Arg4673Leu)', 'slr': 'NP_892006.3:p.(R4673L)'}
        assert results['NM_182961.3:c.14018G>T']['submitted_variant'] == '6-152651802-C-A'
        assert results['NM_182961.3:c.14018G>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_182961.3:c.14018G>T']['hgvs_lrg_variant'] == 'LRG_427:g.311733G>T'
        assert results['NM_182961.3:c.14018G>T']['hgvs_transcript_variant'] == 'NM_182961.3:c.14018G>T'
        assert results['NM_182961.3:c.14018G>T']['hgvs_refseqgene_variant'] == 'NG_012855.1:g.311733G>T'
        assert results['NM_182961.3:c.14018G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000006.11:g.152651802C>A', 'vcf': {'chr': 'chr6', 'ref': u'C', 'pos': '152651802', 'alt': u'A'}}
        assert results['NM_182961.3:c.14018G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000006.12:g.152330667C>A', 'vcf': {'chr': 'chr6', 'ref': u'C', 'pos': '152330667', 'alt': u'A'}}
        assert results['NM_182961.3:c.14018G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000006.11:g.152651802C>A', 'vcf': {'chr': '6', 'ref': u'C', 'pos': '152651802', 'alt': u'A'}}
        assert results['NM_182961.3:c.14018G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000006.12:g.152330667C>A', 'vcf': {'chr': '6', 'ref': u'C', 'pos': '152330667', 'alt': u'A'}}

        assert 'NM_033071.3:c.13805G>T' in results.keys()
        assert results['NM_033071.3:c.13805G>T']['hgvs_lrg_transcript_variant'] == 'LRG_427t2:c.13805G>T'
        assert results['NM_033071.3:c.13805G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_033071.3:c.13805G>T']['alt_genomic_loci'] == []
        assert results['NM_033071.3:c.13805G>T']['gene_symbol'] == 'SYNE1'
        assert results['NM_033071.3:c.13805G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_149062.1(LRG_427p2):p.(Arg4602Leu)', 'slr': 'NP_149062.1:p.(R4602L)'}
        assert results['NM_033071.3:c.13805G>T']['submitted_variant'] == '6-152651802-C-A'
        assert results['NM_033071.3:c.13805G>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_033071.3:c.13805G>T']['hgvs_lrg_variant'] == 'LRG_427:g.311733G>T'
        assert results['NM_033071.3:c.13805G>T']['hgvs_transcript_variant'] == 'NM_033071.3:c.13805G>T'
        assert results['NM_033071.3:c.13805G>T']['hgvs_refseqgene_variant'] == 'NG_012855.1:g.311733G>T'
        assert results['NM_033071.3:c.13805G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000006.11:g.152651802C>A', 'vcf': {'chr': 'chr6', 'ref': u'C', 'pos': '152651802', 'alt': u'A'}}
        assert results['NM_033071.3:c.13805G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000006.12:g.152330667C>A', 'vcf': {'chr': 'chr6', 'ref': u'C', 'pos': '152330667', 'alt': u'A'}}
        assert results['NM_033071.3:c.13805G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000006.11:g.152651802C>A', 'vcf': {'chr': '6', 'ref': u'C', 'pos': '152651802', 'alt': u'A'}}
        assert results['NM_033071.3:c.13805G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000006.12:g.152330667C>A', 'vcf': {'chr': '6', 'ref': u'C', 'pos': '152330667', 'alt': u'A'}}


    def test_variant295(self):
        variant = '6-152737643-C-G'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_033071.3:c.5950G>C' in results.keys()
        assert results['NM_033071.3:c.5950G>C']['hgvs_lrg_transcript_variant'] == 'LRG_427t2:c.5950G>C'
        assert results['NM_033071.3:c.5950G>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_033071.3:c.5950G>C']['alt_genomic_loci'] == []
        assert results['NM_033071.3:c.5950G>C']['gene_symbol'] == 'SYNE1'
        assert results['NM_033071.3:c.5950G>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_149062.1(LRG_427p2):p.(Ala1984Pro)', 'slr': 'NP_149062.1:p.(A1984P)'}
        assert results['NM_033071.3:c.5950G>C']['submitted_variant'] == '6-152737643-C-G'
        assert results['NM_033071.3:c.5950G>C']['genome_context_intronic_sequence'] == ''
        assert results['NM_033071.3:c.5950G>C']['hgvs_lrg_variant'] == 'LRG_427:g.225892G>C'
        assert results['NM_033071.3:c.5950G>C']['hgvs_transcript_variant'] == 'NM_033071.3:c.5950G>C'
        assert results['NM_033071.3:c.5950G>C']['hgvs_refseqgene_variant'] == 'NG_012855.1:g.225892G>C'
        assert results['NM_033071.3:c.5950G>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000006.11:g.152737643C>G', 'vcf': {'chr': 'chr6', 'ref': u'C', 'pos': '152737643', 'alt': u'G'}}
        assert results['NM_033071.3:c.5950G>C']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000006.12:g.152416508C>G', 'vcf': {'chr': 'chr6', 'ref': u'C', 'pos': '152416508', 'alt': u'G'}}
        assert results['NM_033071.3:c.5950G>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000006.11:g.152737643C>G', 'vcf': {'chr': '6', 'ref': u'C', 'pos': '152737643', 'alt': u'G'}}
        assert results['NM_033071.3:c.5950G>C']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000006.12:g.152416508C>G', 'vcf': {'chr': '6', 'ref': u'C', 'pos': '152416508', 'alt': u'G'}}

        assert 'NM_182961.3:c.5929G>C' in results.keys()
        assert results['NM_182961.3:c.5929G>C']['hgvs_lrg_transcript_variant'] == 'LRG_427t1:c.5929G>C'
        assert results['NM_182961.3:c.5929G>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_182961.3:c.5929G>C']['alt_genomic_loci'] == []
        assert results['NM_182961.3:c.5929G>C']['gene_symbol'] == 'SYNE1'
        assert results['NM_182961.3:c.5929G>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_892006.3(LRG_427p1):p.(Ala1977Pro)', 'slr': 'NP_892006.3:p.(A1977P)'}
        assert results['NM_182961.3:c.5929G>C']['submitted_variant'] == '6-152737643-C-G'
        assert results['NM_182961.3:c.5929G>C']['genome_context_intronic_sequence'] == ''
        assert results['NM_182961.3:c.5929G>C']['hgvs_lrg_variant'] == 'LRG_427:g.225892G>C'
        assert results['NM_182961.3:c.5929G>C']['hgvs_transcript_variant'] == 'NM_182961.3:c.5929G>C'
        assert results['NM_182961.3:c.5929G>C']['hgvs_refseqgene_variant'] == 'NG_012855.1:g.225892G>C'
        assert results['NM_182961.3:c.5929G>C']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000006.11:g.152737643C>G', 'vcf': {'chr': 'chr6', 'ref': u'C', 'pos': '152737643', 'alt': u'G'}}
        assert results['NM_182961.3:c.5929G>C']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000006.12:g.152416508C>G', 'vcf': {'chr': 'chr6', 'ref': u'C', 'pos': '152416508', 'alt': u'G'}}
        assert results['NM_182961.3:c.5929G>C']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000006.11:g.152737643C>G', 'vcf': {'chr': '6', 'ref': u'C', 'pos': '152737643', 'alt': u'G'}}
        assert results['NM_182961.3:c.5929G>C']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000006.12:g.152416508C>G', 'vcf': {'chr': '6', 'ref': u'C', 'pos': '152416508', 'alt': u'G'}}


    def test_variant296(self):
        variant = '7-6026775-T-C'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001322012.1:c.688A>G' in results.keys()
        assert results['NM_001322012.1:c.688A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001322012.1:c.688A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001322012.1:c.688A>G']['alt_genomic_loci'] == []
        assert results['NM_001322012.1:c.688A>G']['gene_symbol'] == 'PMS2'
        assert results['NM_001322012.1:c.688A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001308941.1:p.(Lys230Glu)', 'slr': 'NP_001308941.1:p.(K230E)'}
        assert results['NM_001322012.1:c.688A>G']['submitted_variant'] == '7-6026775-T-C'
        assert results['NM_001322012.1:c.688A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001322012.1:c.688A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001322012.1:c.688A>G']['hgvs_transcript_variant'] == 'NM_001322012.1:c.688A>G'
        assert results['NM_001322012.1:c.688A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001322012.1:c.688A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.6026775T>C', 'vcf': {'chr': 'chr7', 'ref': u'T', 'pos': '6026775', 'alt': u'C'}}
        assert results['NM_001322012.1:c.688A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.5987144T>C', 'vcf': {'chr': 'chr7', 'ref': u'T', 'pos': '5987144', 'alt': u'C'}}
        assert results['NM_001322012.1:c.688A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.6026775T>C', 'vcf': {'chr': '7', 'ref': u'T', 'pos': '6026775', 'alt': u'C'}}
        assert results['NM_001322012.1:c.688A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.5987144T>C', 'vcf': {'chr': '7', 'ref': u'T', 'pos': '5987144', 'alt': u'C'}}

        assert 'NM_001322010.1:c.1060A>G' in results.keys()
        assert results['NM_001322010.1:c.1060A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001322010.1:c.1060A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001322010.1:c.1060A>G']['alt_genomic_loci'] == []
        assert results['NM_001322010.1:c.1060A>G']['gene_symbol'] == 'PMS2'
        assert results['NM_001322010.1:c.1060A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001308939.1:p.(Lys354Glu)', 'slr': 'NP_001308939.1:p.(K354E)'}
        assert results['NM_001322010.1:c.1060A>G']['submitted_variant'] == '7-6026775-T-C'
        assert results['NM_001322010.1:c.1060A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001322010.1:c.1060A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001322010.1:c.1060A>G']['hgvs_transcript_variant'] == 'NM_001322010.1:c.1060A>G'
        assert results['NM_001322010.1:c.1060A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001322010.1:c.1060A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.6026775T>C', 'vcf': {'chr': 'chr7', 'ref': u'T', 'pos': '6026775', 'alt': u'C'}}
        assert results['NM_001322010.1:c.1060A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.5987144T>C', 'vcf': {'chr': 'chr7', 'ref': u'T', 'pos': '5987144', 'alt': u'C'}}
        assert results['NM_001322010.1:c.1060A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.6026775T>C', 'vcf': {'chr': '7', 'ref': u'T', 'pos': '6026775', 'alt': u'C'}}
        assert results['NM_001322010.1:c.1060A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.5987144T>C', 'vcf': {'chr': '7', 'ref': u'T', 'pos': '5987144', 'alt': u'C'}}

        assert 'NM_001322015.1:c.1312A>G' in results.keys()
        assert results['NM_001322015.1:c.1312A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001322015.1:c.1312A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001322015.1:c.1312A>G']['alt_genomic_loci'] == []
        assert results['NM_001322015.1:c.1312A>G']['gene_symbol'] == 'PMS2'
        assert results['NM_001322015.1:c.1312A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001308944.1:p.(Lys438Glu)', 'slr': 'NP_001308944.1:p.(K438E)'}
        assert results['NM_001322015.1:c.1312A>G']['submitted_variant'] == '7-6026775-T-C'
        assert results['NM_001322015.1:c.1312A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001322015.1:c.1312A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001322015.1:c.1312A>G']['hgvs_transcript_variant'] == 'NM_001322015.1:c.1312A>G'
        assert results['NM_001322015.1:c.1312A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001322015.1:c.1312A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.6026775T>C', 'vcf': {'chr': 'chr7', 'ref': u'T', 'pos': '6026775', 'alt': u'C'}}
        assert results['NM_001322015.1:c.1312A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.5987144T>C', 'vcf': {'chr': 'chr7', 'ref': u'T', 'pos': '5987144', 'alt': u'C'}}
        assert results['NM_001322015.1:c.1312A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.6026775T>C', 'vcf': {'chr': '7', 'ref': u'T', 'pos': '6026775', 'alt': u'C'}}
        assert results['NM_001322015.1:c.1312A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.5987144T>C', 'vcf': {'chr': '7', 'ref': u'T', 'pos': '5987144', 'alt': u'C'}}

        assert 'NM_001322003.1:c.1216A>G' in results.keys()
        assert results['NM_001322003.1:c.1216A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001322003.1:c.1216A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001322003.1:c.1216A>G']['alt_genomic_loci'] == []
        assert results['NM_001322003.1:c.1216A>G']['gene_symbol'] == 'PMS2'
        assert results['NM_001322003.1:c.1216A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001308932.1:p.(Lys406Glu)', 'slr': 'NP_001308932.1:p.(K406E)'}
        assert results['NM_001322003.1:c.1216A>G']['submitted_variant'] == '7-6026775-T-C'
        assert results['NM_001322003.1:c.1216A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001322003.1:c.1216A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001322003.1:c.1216A>G']['hgvs_transcript_variant'] == 'NM_001322003.1:c.1216A>G'
        assert results['NM_001322003.1:c.1216A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001322003.1:c.1216A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.6026775T>C', 'vcf': {'chr': 'chr7', 'ref': u'T', 'pos': '6026775', 'alt': u'C'}}
        assert results['NM_001322003.1:c.1216A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.5987144T>C', 'vcf': {'chr': 'chr7', 'ref': u'T', 'pos': '5987144', 'alt': u'C'}}
        assert results['NM_001322003.1:c.1216A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.6026775T>C', 'vcf': {'chr': '7', 'ref': u'T', 'pos': '6026775', 'alt': u'C'}}
        assert results['NM_001322003.1:c.1216A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.5987144T>C', 'vcf': {'chr': '7', 'ref': u'T', 'pos': '5987144', 'alt': u'C'}}

        assert 'NM_001322014.1:c.1621A>G' in results.keys()
        assert results['NM_001322014.1:c.1621A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001322014.1:c.1621A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001322014.1:c.1621A>G']['alt_genomic_loci'] == []
        assert results['NM_001322014.1:c.1621A>G']['gene_symbol'] == 'PMS2'
        assert results['NM_001322014.1:c.1621A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001308943.1:p.(Lys541Glu)', 'slr': 'NP_001308943.1:p.(K541E)'}
        assert results['NM_001322014.1:c.1621A>G']['submitted_variant'] == '7-6026775-T-C'
        assert results['NM_001322014.1:c.1621A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001322014.1:c.1621A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001322014.1:c.1621A>G']['hgvs_transcript_variant'] == 'NM_001322014.1:c.1621A>G'
        assert results['NM_001322014.1:c.1621A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001322014.1:c.1621A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.6026775T>C', 'vcf': {'chr': 'chr7', 'ref': u'T', 'pos': '6026775', 'alt': u'C'}}
        assert results['NM_001322014.1:c.1621A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.5987144T>C', 'vcf': {'chr': 'chr7', 'ref': u'T', 'pos': '5987144', 'alt': u'C'}}
        assert results['NM_001322014.1:c.1621A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.6026775T>C', 'vcf': {'chr': '7', 'ref': u'T', 'pos': '6026775', 'alt': u'C'}}
        assert results['NM_001322014.1:c.1621A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.5987144T>C', 'vcf': {'chr': '7', 'ref': u'T', 'pos': '5987144', 'alt': u'C'}}

        assert 'NM_001322004.1:c.1216A>G' in results.keys()
        assert results['NM_001322004.1:c.1216A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001322004.1:c.1216A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001322004.1:c.1216A>G']['alt_genomic_loci'] == []
        assert results['NM_001322004.1:c.1216A>G']['gene_symbol'] == 'PMS2'
        assert results['NM_001322004.1:c.1216A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001308933.1:p.(Lys406Glu)', 'slr': 'NP_001308933.1:p.(K406E)'}
        assert results['NM_001322004.1:c.1216A>G']['submitted_variant'] == '7-6026775-T-C'
        assert results['NM_001322004.1:c.1216A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001322004.1:c.1216A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001322004.1:c.1216A>G']['hgvs_transcript_variant'] == 'NM_001322004.1:c.1216A>G'
        assert results['NM_001322004.1:c.1216A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001322004.1:c.1216A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.6026775T>C', 'vcf': {'chr': 'chr7', 'ref': u'T', 'pos': '6026775', 'alt': u'C'}}
        assert results['NM_001322004.1:c.1216A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.5987144T>C', 'vcf': {'chr': 'chr7', 'ref': u'T', 'pos': '5987144', 'alt': u'C'}}
        assert results['NM_001322004.1:c.1216A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.6026775T>C', 'vcf': {'chr': '7', 'ref': u'T', 'pos': '6026775', 'alt': u'C'}}
        assert results['NM_001322004.1:c.1216A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.5987144T>C', 'vcf': {'chr': '7', 'ref': u'T', 'pos': '5987144', 'alt': u'C'}}

        assert 'NM_001322008.1:c.1303A>G' in results.keys()
        assert results['NM_001322008.1:c.1303A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001322008.1:c.1303A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001322008.1:c.1303A>G']['alt_genomic_loci'] == []
        assert results['NM_001322008.1:c.1303A>G']['gene_symbol'] == 'PMS2'
        assert results['NM_001322008.1:c.1303A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001308937.1:p.(Lys435Glu)', 'slr': 'NP_001308937.1:p.(K435E)'}
        assert results['NM_001322008.1:c.1303A>G']['submitted_variant'] == '7-6026775-T-C'
        assert results['NM_001322008.1:c.1303A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001322008.1:c.1303A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001322008.1:c.1303A>G']['hgvs_transcript_variant'] == 'NM_001322008.1:c.1303A>G'
        assert results['NM_001322008.1:c.1303A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001322008.1:c.1303A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.6026775T>C', 'vcf': {'chr': 'chr7', 'ref': u'T', 'pos': '6026775', 'alt': u'C'}}
        assert results['NM_001322008.1:c.1303A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.5987144T>C', 'vcf': {'chr': 'chr7', 'ref': u'T', 'pos': '5987144', 'alt': u'C'}}
        assert results['NM_001322008.1:c.1303A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.6026775T>C', 'vcf': {'chr': '7', 'ref': u'T', 'pos': '6026775', 'alt': u'C'}}
        assert results['NM_001322008.1:c.1303A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.5987144T>C', 'vcf': {'chr': '7', 'ref': u'T', 'pos': '5987144', 'alt': u'C'}}

        assert 'NM_001322006.1:c.1465A>G' in results.keys()
        assert results['NM_001322006.1:c.1465A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001322006.1:c.1465A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001322006.1:c.1465A>G']['alt_genomic_loci'] == []
        assert results['NM_001322006.1:c.1465A>G']['gene_symbol'] == 'PMS2'
        assert results['NM_001322006.1:c.1465A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001308935.1:p.(Lys489Glu)', 'slr': 'NP_001308935.1:p.(K489E)'}
        assert results['NM_001322006.1:c.1465A>G']['submitted_variant'] == '7-6026775-T-C'
        assert results['NM_001322006.1:c.1465A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001322006.1:c.1465A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001322006.1:c.1465A>G']['hgvs_transcript_variant'] == 'NM_001322006.1:c.1465A>G'
        assert results['NM_001322006.1:c.1465A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001322006.1:c.1465A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.6026775T>C', 'vcf': {'chr': 'chr7', 'ref': u'T', 'pos': '6026775', 'alt': u'C'}}
        assert results['NM_001322006.1:c.1465A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.5987144T>C', 'vcf': {'chr': 'chr7', 'ref': u'T', 'pos': '5987144', 'alt': u'C'}}
        assert results['NM_001322006.1:c.1465A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.6026775T>C', 'vcf': {'chr': '7', 'ref': u'T', 'pos': '6026775', 'alt': u'C'}}
        assert results['NM_001322006.1:c.1465A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.5987144T>C', 'vcf': {'chr': '7', 'ref': u'T', 'pos': '5987144', 'alt': u'C'}}

        assert 'NM_001322009.1:c.1216A>G' in results.keys()
        assert results['NM_001322009.1:c.1216A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001322009.1:c.1216A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001322009.1:c.1216A>G']['alt_genomic_loci'] == []
        assert results['NM_001322009.1:c.1216A>G']['gene_symbol'] == 'PMS2'
        assert results['NM_001322009.1:c.1216A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001308938.1:p.(Lys406Glu)', 'slr': 'NP_001308938.1:p.(K406E)'}
        assert results['NM_001322009.1:c.1216A>G']['submitted_variant'] == '7-6026775-T-C'
        assert results['NM_001322009.1:c.1216A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001322009.1:c.1216A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001322009.1:c.1216A>G']['hgvs_transcript_variant'] == 'NM_001322009.1:c.1216A>G'
        assert results['NM_001322009.1:c.1216A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001322009.1:c.1216A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.6026775T>C', 'vcf': {'chr': 'chr7', 'ref': u'T', 'pos': '6026775', 'alt': u'C'}}
        assert results['NM_001322009.1:c.1216A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.5987144T>C', 'vcf': {'chr': 'chr7', 'ref': u'T', 'pos': '5987144', 'alt': u'C'}}
        assert results['NM_001322009.1:c.1216A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.6026775T>C', 'vcf': {'chr': '7', 'ref': u'T', 'pos': '6026775', 'alt': u'C'}}
        assert results['NM_001322009.1:c.1216A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.5987144T>C', 'vcf': {'chr': '7', 'ref': u'T', 'pos': '5987144', 'alt': u'C'}}

        assert 'NR_003085.2:n.1703G=' in results.keys()
        assert results['NR_003085.2:n.1703G=']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_003085.2:n.1703G=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_003085.2:n.1703G=']['alt_genomic_loci'] == []
        assert results['NR_003085.2:n.1703G=']['gene_symbol'] == 'PMS2'
        assert results['NR_003085.2:n.1703G=']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_003085.2:n.1703G=']['submitted_variant'] == '7-6026775-T-C'
        assert results['NR_003085.2:n.1703G=']['genome_context_intronic_sequence'] == ''
        assert results['NR_003085.2:n.1703G=']['hgvs_lrg_variant'] == ''
        assert results['NR_003085.2:n.1703G=']['hgvs_transcript_variant'] == 'NR_003085.2:n.1703G='
        assert results['NR_003085.2:n.1703G=']['hgvs_refseqgene_variant'] == ''
        assert results['NR_003085.2:n.1703G=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.6026775T>C', 'vcf': {'chr': 'chr7', 'ref': 'T', 'pos': '6026775', 'alt': u'C'}}
        assert 'hg38' not in results['NR_003085.2:n.1703G=']['primary_assembly_loci'].keys()
        assert results['NR_003085.2:n.1703G=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.6026775T>C', 'vcf': {'chr': '7', 'ref': 'T', 'pos': '6026775', 'alt': u'C'}}
        assert 'grch38' not in results['NR_003085.2:n.1703G=']['primary_assembly_loci'].keys()

        assert results['flag'] == 'gene_variant'
        assert 'NM_001322005.1:c.1216A>G' in results.keys()
        assert results['NM_001322005.1:c.1216A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001322005.1:c.1216A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001322005.1:c.1216A>G']['alt_genomic_loci'] == []
        assert results['NM_001322005.1:c.1216A>G']['gene_symbol'] == 'PMS2'
        assert results['NM_001322005.1:c.1216A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001308934.1:p.(Lys406Glu)', 'slr': 'NP_001308934.1:p.(K406E)'}
        assert results['NM_001322005.1:c.1216A>G']['submitted_variant'] == '7-6026775-T-C'
        assert results['NM_001322005.1:c.1216A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001322005.1:c.1216A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001322005.1:c.1216A>G']['hgvs_transcript_variant'] == 'NM_001322005.1:c.1216A>G'
        assert results['NM_001322005.1:c.1216A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001322005.1:c.1216A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.6026775T>C', 'vcf': {'chr': 'chr7', 'ref': u'T', 'pos': '6026775', 'alt': u'C'}}
        assert results['NM_001322005.1:c.1216A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.5987144T>C', 'vcf': {'chr': 'chr7', 'ref': u'T', 'pos': '5987144', 'alt': u'C'}}
        assert results['NM_001322005.1:c.1216A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.6026775T>C', 'vcf': {'chr': '7', 'ref': u'T', 'pos': '6026775', 'alt': u'C'}}
        assert results['NM_001322005.1:c.1216A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.5987144T>C', 'vcf': {'chr': '7', 'ref': u'T', 'pos': '5987144', 'alt': u'C'}}

        assert 'NM_001322007.1:c.1303A>G' in results.keys()
        assert results['NM_001322007.1:c.1303A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001322007.1:c.1303A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001322007.1:c.1303A>G']['alt_genomic_loci'] == []
        assert results['NM_001322007.1:c.1303A>G']['gene_symbol'] == 'PMS2'
        assert results['NM_001322007.1:c.1303A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001308936.1:p.(Lys435Glu)', 'slr': 'NP_001308936.1:p.(K435E)'}
        assert results['NM_001322007.1:c.1303A>G']['submitted_variant'] == '7-6026775-T-C'
        assert results['NM_001322007.1:c.1303A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001322007.1:c.1303A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001322007.1:c.1303A>G']['hgvs_transcript_variant'] == 'NM_001322007.1:c.1303A>G'
        assert results['NM_001322007.1:c.1303A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001322007.1:c.1303A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.6026775T>C', 'vcf': {'chr': 'chr7', 'ref': u'T', 'pos': '6026775', 'alt': u'C'}}
        assert results['NM_001322007.1:c.1303A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.5987144T>C', 'vcf': {'chr': 'chr7', 'ref': u'T', 'pos': '5987144', 'alt': u'C'}}
        assert results['NM_001322007.1:c.1303A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.6026775T>C', 'vcf': {'chr': '7', 'ref': u'T', 'pos': '6026775', 'alt': u'C'}}
        assert results['NM_001322007.1:c.1303A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.5987144T>C', 'vcf': {'chr': '7', 'ref': u'T', 'pos': '5987144', 'alt': u'C'}}

        assert 'NM_000535.5:c.1621G=' in results.keys()
        assert results['NM_000535.5:c.1621G=']['hgvs_lrg_transcript_variant'] == 'LRG_161t1:c.1621G='
        assert results['NM_000535.5:c.1621G=']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000535.5:c.1621G=']['alt_genomic_loci'] == []
        assert results['NM_000535.5:c.1621G=']['gene_symbol'] == 'PMS2'
        assert results['NM_000535.5:c.1621G=']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000526.1(LRG_161p1):p.(Glu541=)', 'slr': 'NP_000526.1:p.(E541=)'}
        assert results['NM_000535.5:c.1621G=']['submitted_variant'] == '7-6026775-T-C'
        assert results['NM_000535.5:c.1621G=']['genome_context_intronic_sequence'] == ''
        assert results['NM_000535.5:c.1621G=']['hgvs_lrg_variant'] == 'LRG_161:g.26963G='
        assert results['NM_000535.5:c.1621G=']['hgvs_transcript_variant'] == 'NM_000535.5:c.1621G='
        assert results['NM_000535.5:c.1621G=']['hgvs_refseqgene_variant'] == 'NG_008466.1:g.26963G='
        assert results['NM_000535.5:c.1621G=']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.6026775T>C', 'vcf': {'chr': 'chr7', 'ref': 'T', 'pos': '6026775', 'alt': u'C'}}
        assert 'hg38' not in results['NM_000535.5:c.1621G=']['primary_assembly_loci'].keys()
        assert results['NM_000535.5:c.1621G=']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.6026775T>C', 'vcf': {'chr': '7', 'ref': 'T', 'pos': '6026775', 'alt': u'C'}}
        assert 'grch38' not in results['NM_000535.5:c.1621G=']['primary_assembly_loci'].keys()

        assert 'NM_001322013.1:c.1048A>G' in results.keys()
        assert results['NM_001322013.1:c.1048A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001322013.1:c.1048A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001322013.1:c.1048A>G']['alt_genomic_loci'] == []
        assert results['NM_001322013.1:c.1048A>G']['gene_symbol'] == 'PMS2'
        assert results['NM_001322013.1:c.1048A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001308942.1:p.(Lys350Glu)', 'slr': 'NP_001308942.1:p.(K350E)'}
        assert results['NM_001322013.1:c.1048A>G']['submitted_variant'] == '7-6026775-T-C'
        assert results['NM_001322013.1:c.1048A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001322013.1:c.1048A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001322013.1:c.1048A>G']['hgvs_transcript_variant'] == 'NM_001322013.1:c.1048A>G'
        assert results['NM_001322013.1:c.1048A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001322013.1:c.1048A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.6026775T>C', 'vcf': {'chr': 'chr7', 'ref': u'T', 'pos': '6026775', 'alt': u'C'}}
        assert results['NM_001322013.1:c.1048A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.5987144T>C', 'vcf': {'chr': 'chr7', 'ref': u'T', 'pos': '5987144', 'alt': u'C'}}
        assert results['NM_001322013.1:c.1048A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.6026775T>C', 'vcf': {'chr': '7', 'ref': u'T', 'pos': '6026775', 'alt': u'C'}}
        assert results['NM_001322013.1:c.1048A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.5987144T>C', 'vcf': {'chr': '7', 'ref': u'T', 'pos': '5987144', 'alt': u'C'}}

        assert 'NR_136154.1:n.1708A>G' in results.keys()
        assert results['NR_136154.1:n.1708A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_136154.1:n.1708A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_136154.1:n.1708A>G']['alt_genomic_loci'] == []
        assert results['NR_136154.1:n.1708A>G']['gene_symbol'] == 'PMS2'
        assert results['NR_136154.1:n.1708A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_136154.1:n.1708A>G']['submitted_variant'] == '7-6026775-T-C'
        assert results['NR_136154.1:n.1708A>G']['genome_context_intronic_sequence'] == ''
        assert results['NR_136154.1:n.1708A>G']['hgvs_lrg_variant'] == ''
        assert results['NR_136154.1:n.1708A>G']['hgvs_transcript_variant'] == 'NR_136154.1:n.1708A>G'
        assert results['NR_136154.1:n.1708A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NR_136154.1:n.1708A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.6026775T>C', 'vcf': {'chr': 'chr7', 'ref': u'T', 'pos': '6026775', 'alt': u'C'}}
        assert results['NR_136154.1:n.1708A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.5987144T>C', 'vcf': {'chr': 'chr7', 'ref': u'T', 'pos': '5987144', 'alt': u'C'}}
        assert results['NR_136154.1:n.1708A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.6026775T>C', 'vcf': {'chr': '7', 'ref': u'T', 'pos': '6026775', 'alt': u'C'}}
        assert results['NR_136154.1:n.1708A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.5987144T>C', 'vcf': {'chr': '7', 'ref': u'T', 'pos': '5987144', 'alt': u'C'}}

        assert 'NM_001322011.1:c.688A>G' in results.keys()
        assert results['NM_001322011.1:c.688A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001322011.1:c.688A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001322011.1:c.688A>G']['alt_genomic_loci'] == []
        assert results['NM_001322011.1:c.688A>G']['gene_symbol'] == 'PMS2'
        assert results['NM_001322011.1:c.688A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001308940.1:p.(Lys230Glu)', 'slr': 'NP_001308940.1:p.(K230E)'}
        assert results['NM_001322011.1:c.688A>G']['submitted_variant'] == '7-6026775-T-C'
        assert results['NM_001322011.1:c.688A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_001322011.1:c.688A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_001322011.1:c.688A>G']['hgvs_transcript_variant'] == 'NM_001322011.1:c.688A>G'
        assert results['NM_001322011.1:c.688A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001322011.1:c.688A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.6026775T>C', 'vcf': {'chr': 'chr7', 'ref': u'T', 'pos': '6026775', 'alt': u'C'}}
        assert results['NM_001322011.1:c.688A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.5987144T>C', 'vcf': {'chr': 'chr7', 'ref': u'T', 'pos': '5987144', 'alt': u'C'}}
        assert results['NM_001322011.1:c.688A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.6026775T>C', 'vcf': {'chr': '7', 'ref': u'T', 'pos': '6026775', 'alt': u'C'}}
        assert results['NM_001322011.1:c.688A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.5987144T>C', 'vcf': {'chr': '7', 'ref': u'T', 'pos': '5987144', 'alt': u'C'}}

        assert 'NM_000535.6:c.1621A>G' in results.keys()
        assert results['NM_000535.6:c.1621A>G']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000535.6:c.1621A>G']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000535.6:c.1621A>G']['alt_genomic_loci'] == []
        assert results['NM_000535.6:c.1621A>G']['gene_symbol'] == 'PMS2'
        assert results['NM_000535.6:c.1621A>G']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000526.2:p.(Lys541Glu)', 'slr': 'NP_000526.2:p.(K541E)'}
        assert results['NM_000535.6:c.1621A>G']['submitted_variant'] == '7-6026775-T-C'
        assert results['NM_000535.6:c.1621A>G']['genome_context_intronic_sequence'] == ''
        assert results['NM_000535.6:c.1621A>G']['hgvs_lrg_variant'] == ''
        assert results['NM_000535.6:c.1621A>G']['hgvs_transcript_variant'] == 'NM_000535.6:c.1621A>G'
        assert results['NM_000535.6:c.1621A>G']['hgvs_refseqgene_variant'] == ''
        assert results['NM_000535.6:c.1621A>G']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.6026775T>C', 'vcf': {'chr': 'chr7', 'ref': u'T', 'pos': '6026775', 'alt': u'C'}}
        assert results['NM_000535.6:c.1621A>G']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.5987144T>C', 'vcf': {'chr': 'chr7', 'ref': u'T', 'pos': '5987144', 'alt': u'C'}}
        assert results['NM_000535.6:c.1621A>G']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.6026775T>C', 'vcf': {'chr': '7', 'ref': u'T', 'pos': '6026775', 'alt': u'C'}}
        assert results['NM_000535.6:c.1621A>G']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.5987144T>C', 'vcf': {'chr': '7', 'ref': u'T', 'pos': '5987144', 'alt': u'C'}}


    def test_variant297(self):
        variant = '7-55242465-GGAATTAAGAGAAGCA-G'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001346900.1:c.2077_2091del' in results.keys()
        assert results['NM_001346900.1:c.2077_2091del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001346900.1:c.2077_2091del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001346900.1:c.2077_2091del']['alt_genomic_loci'] == []
        assert results['NM_001346900.1:c.2077_2091del']['gene_symbol'] == 'EGFR'
        assert results['NM_001346900.1:c.2077_2091del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001333829.1:p.(Glu693_Ala697del)', 'slr': 'NP_001333829.1:p.(E693_A697del)'}
        assert results['NM_001346900.1:c.2077_2091del']['submitted_variant'] == '7-55242465-GGAATTAAGAGAAGCA-G'
        assert results['NM_001346900.1:c.2077_2091del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001346900.1:c.2077_2091del']['hgvs_lrg_variant'] == ''
        assert results['NM_001346900.1:c.2077_2091del']['hgvs_transcript_variant'] == 'NM_001346900.1:c.2077_2091del'
        assert results['NM_001346900.1:c.2077_2091del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001346900.1:c.2077_2091del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.55242466_55242480del', 'vcf': {'chr': 'chr7', 'ref': 'GGAATTAAGAGAAGCA', 'pos': '55242465', 'alt': 'G'}}
        assert results['NM_001346900.1:c.2077_2091del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.55174773_55174787del', 'vcf': {'chr': 'chr7', 'ref': 'GGAATTAAGAGAAGCA', 'pos': '55174772', 'alt': 'G'}}
        assert results['NM_001346900.1:c.2077_2091del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.55242466_55242480del', 'vcf': {'chr': '7', 'ref': 'GGAATTAAGAGAAGCA', 'pos': '55242465', 'alt': 'G'}}
        assert results['NM_001346900.1:c.2077_2091del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.55174773_55174787del', 'vcf': {'chr': '7', 'ref': 'GGAATTAAGAGAAGCA', 'pos': '55174772', 'alt': 'G'}}

        assert 'NM_001346898.1:c.2236_2250del' in results.keys()
        assert results['NM_001346898.1:c.2236_2250del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001346898.1:c.2236_2250del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001346898.1:c.2236_2250del']['alt_genomic_loci'] == []
        assert results['NM_001346898.1:c.2236_2250del']['gene_symbol'] == 'EGFR'
        assert results['NM_001346898.1:c.2236_2250del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001333827.1:p.(Glu746_Ala750del)', 'slr': 'NP_001333827.1:p.(E746_A750del)'}
        assert results['NM_001346898.1:c.2236_2250del']['submitted_variant'] == '7-55242465-GGAATTAAGAGAAGCA-G'
        assert results['NM_001346898.1:c.2236_2250del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001346898.1:c.2236_2250del']['hgvs_lrg_variant'] == ''
        assert results['NM_001346898.1:c.2236_2250del']['hgvs_transcript_variant'] == 'NM_001346898.1:c.2236_2250del'
        assert results['NM_001346898.1:c.2236_2250del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001346898.1:c.2236_2250del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.55242466_55242480del', 'vcf': {'chr': 'chr7', 'ref': 'GGAATTAAGAGAAGCA', 'pos': '55242465', 'alt': 'G'}}
        assert results['NM_001346898.1:c.2236_2250del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.55174773_55174787del', 'vcf': {'chr': 'chr7', 'ref': 'GGAATTAAGAGAAGCA', 'pos': '55174772', 'alt': 'G'}}
        assert results['NM_001346898.1:c.2236_2250del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.55242466_55242480del', 'vcf': {'chr': '7', 'ref': 'GGAATTAAGAGAAGCA', 'pos': '55242465', 'alt': 'G'}}
        assert results['NM_001346898.1:c.2236_2250del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.55174773_55174787del', 'vcf': {'chr': '7', 'ref': 'GGAATTAAGAGAAGCA', 'pos': '55174772', 'alt': 'G'}}

        assert 'NM_001346941.1:c.1435_1449del' in results.keys()
        assert results['NM_001346941.1:c.1435_1449del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001346941.1:c.1435_1449del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001346941.1:c.1435_1449del']['alt_genomic_loci'] == []
        assert results['NM_001346941.1:c.1435_1449del']['gene_symbol'] == 'EGFR'
        assert results['NM_001346941.1:c.1435_1449del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001333870.1:p.(Glu479_Ala483del)', 'slr': 'NP_001333870.1:p.(E479_A483del)'}
        assert results['NM_001346941.1:c.1435_1449del']['submitted_variant'] == '7-55242465-GGAATTAAGAGAAGCA-G'
        assert results['NM_001346941.1:c.1435_1449del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001346941.1:c.1435_1449del']['hgvs_lrg_variant'] == ''
        assert results['NM_001346941.1:c.1435_1449del']['hgvs_transcript_variant'] == 'NM_001346941.1:c.1435_1449del'
        assert results['NM_001346941.1:c.1435_1449del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001346941.1:c.1435_1449del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.55242466_55242480del', 'vcf': {'chr': 'chr7', 'ref': 'GGAATTAAGAGAAGCA', 'pos': '55242465', 'alt': 'G'}}
        assert results['NM_001346941.1:c.1435_1449del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.55174773_55174787del', 'vcf': {'chr': 'chr7', 'ref': 'GGAATTAAGAGAAGCA', 'pos': '55174772', 'alt': 'G'}}
        assert results['NM_001346941.1:c.1435_1449del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.55242466_55242480del', 'vcf': {'chr': '7', 'ref': 'GGAATTAAGAGAAGCA', 'pos': '55242465', 'alt': 'G'}}
        assert results['NM_001346941.1:c.1435_1449del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.55174773_55174787del', 'vcf': {'chr': '7', 'ref': 'GGAATTAAGAGAAGCA', 'pos': '55174772', 'alt': 'G'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001346899.1:c.2101_2115del' in results.keys()
        assert results['NM_001346899.1:c.2101_2115del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001346899.1:c.2101_2115del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001346899.1:c.2101_2115del']['alt_genomic_loci'] == []
        assert results['NM_001346899.1:c.2101_2115del']['gene_symbol'] == 'EGFR'
        assert results['NM_001346899.1:c.2101_2115del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001333828.1:p.(Glu701_Ala705del)', 'slr': 'NP_001333828.1:p.(E701_A705del)'}
        assert results['NM_001346899.1:c.2101_2115del']['submitted_variant'] == '7-55242465-GGAATTAAGAGAAGCA-G'
        assert results['NM_001346899.1:c.2101_2115del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001346899.1:c.2101_2115del']['hgvs_lrg_variant'] == ''
        assert results['NM_001346899.1:c.2101_2115del']['hgvs_transcript_variant'] == 'NM_001346899.1:c.2101_2115del'
        assert results['NM_001346899.1:c.2101_2115del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001346899.1:c.2101_2115del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.55242466_55242480del', 'vcf': {'chr': 'chr7', 'ref': 'GGAATTAAGAGAAGCA', 'pos': '55242465', 'alt': 'G'}}
        assert results['NM_001346899.1:c.2101_2115del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.55174773_55174787del', 'vcf': {'chr': 'chr7', 'ref': 'GGAATTAAGAGAAGCA', 'pos': '55174772', 'alt': 'G'}}
        assert results['NM_001346899.1:c.2101_2115del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.55242466_55242480del', 'vcf': {'chr': '7', 'ref': 'GGAATTAAGAGAAGCA', 'pos': '55242465', 'alt': 'G'}}
        assert results['NM_001346899.1:c.2101_2115del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.55174773_55174787del', 'vcf': {'chr': '7', 'ref': 'GGAATTAAGAGAAGCA', 'pos': '55174772', 'alt': 'G'}}

        assert 'NM_001346897.1:c.2101_2115del' in results.keys()
        assert results['NM_001346897.1:c.2101_2115del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001346897.1:c.2101_2115del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001346897.1:c.2101_2115del']['alt_genomic_loci'] == []
        assert results['NM_001346897.1:c.2101_2115del']['gene_symbol'] == 'EGFR'
        assert results['NM_001346897.1:c.2101_2115del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001333826.1:p.(Glu701_Ala705del)', 'slr': 'NP_001333826.1:p.(E701_A705del)'}
        assert results['NM_001346897.1:c.2101_2115del']['submitted_variant'] == '7-55242465-GGAATTAAGAGAAGCA-G'
        assert results['NM_001346897.1:c.2101_2115del']['genome_context_intronic_sequence'] == ''
        assert results['NM_001346897.1:c.2101_2115del']['hgvs_lrg_variant'] == ''
        assert results['NM_001346897.1:c.2101_2115del']['hgvs_transcript_variant'] == 'NM_001346897.1:c.2101_2115del'
        assert results['NM_001346897.1:c.2101_2115del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001346897.1:c.2101_2115del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.55242466_55242480del', 'vcf': {'chr': 'chr7', 'ref': 'GGAATTAAGAGAAGCA', 'pos': '55242465', 'alt': 'G'}}
        assert results['NM_001346897.1:c.2101_2115del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.55174773_55174787del', 'vcf': {'chr': 'chr7', 'ref': 'GGAATTAAGAGAAGCA', 'pos': '55174772', 'alt': 'G'}}
        assert results['NM_001346897.1:c.2101_2115del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.55242466_55242480del', 'vcf': {'chr': '7', 'ref': 'GGAATTAAGAGAAGCA', 'pos': '55242465', 'alt': 'G'}}
        assert results['NM_001346897.1:c.2101_2115del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.55174773_55174787del', 'vcf': {'chr': '7', 'ref': 'GGAATTAAGAGAAGCA', 'pos': '55174772', 'alt': 'G'}}

        assert 'NM_005228.3:c.2236_2250del' in results.keys()
        assert results['NM_005228.3:c.2236_2250del']['hgvs_lrg_transcript_variant'] == 'LRG_304t1:c.2236_2250del'
        assert results['NM_005228.3:c.2236_2250del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_005228.3:c.2236_2250del']['alt_genomic_loci'] == []
        assert results['NM_005228.3:c.2236_2250del']['gene_symbol'] == 'EGFR'
        assert results['NM_005228.3:c.2236_2250del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_005219.2(LRG_304p1):p.(Glu746_Ala750del)', 'slr': 'NP_005219.2:p.(E746_A750del)'}
        assert results['NM_005228.3:c.2236_2250del']['submitted_variant'] == '7-55242465-GGAATTAAGAGAAGCA-G'
        assert results['NM_005228.3:c.2236_2250del']['genome_context_intronic_sequence'] == ''
        assert results['NM_005228.3:c.2236_2250del']['hgvs_lrg_variant'] == 'LRG_304:g.160742_160756del'
        assert results['NM_005228.3:c.2236_2250del']['hgvs_transcript_variant'] == 'NM_005228.3:c.2236_2250del'
        assert results['NM_005228.3:c.2236_2250del']['hgvs_refseqgene_variant'] == 'NG_007726.3:g.160742_160756del'
        assert results['NM_005228.3:c.2236_2250del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.55242466_55242480del', 'vcf': {'chr': 'chr7', 'ref': 'GGAATTAAGAGAAGCA', 'pos': '55242465', 'alt': 'G'}}
        assert results['NM_005228.3:c.2236_2250del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.55174773_55174787del', 'vcf': {'chr': 'chr7', 'ref': 'GGAATTAAGAGAAGCA', 'pos': '55174772', 'alt': 'G'}}
        assert results['NM_005228.3:c.2236_2250del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.55242466_55242480del', 'vcf': {'chr': '7', 'ref': 'GGAATTAAGAGAAGCA', 'pos': '55242465', 'alt': 'G'}}
        assert results['NM_005228.3:c.2236_2250del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.55174773_55174787del', 'vcf': {'chr': '7', 'ref': 'GGAATTAAGAGAAGCA', 'pos': '55174772', 'alt': 'G'}}

        assert 'NM_005228.4:c.2236_2250del' in results.keys()
        assert results['NM_005228.4:c.2236_2250del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_005228.4:c.2236_2250del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_005228.4:c.2236_2250del']['alt_genomic_loci'] == []
        assert results['NM_005228.4:c.2236_2250del']['gene_symbol'] == 'EGFR'
        assert results['NM_005228.4:c.2236_2250del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_005219.2(LRG_304p1):p.(Glu746_Ala750del)', 'slr': 'NP_005219.2:p.(E746_A750del)'}
        assert results['NM_005228.4:c.2236_2250del']['submitted_variant'] == '7-55242465-GGAATTAAGAGAAGCA-G'
        assert results['NM_005228.4:c.2236_2250del']['genome_context_intronic_sequence'] == ''
        assert results['NM_005228.4:c.2236_2250del']['hgvs_lrg_variant'] == ''
        assert results['NM_005228.4:c.2236_2250del']['hgvs_transcript_variant'] == 'NM_005228.4:c.2236_2250del'
        assert results['NM_005228.4:c.2236_2250del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_005228.4:c.2236_2250del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.55242466_55242480del', 'vcf': {'chr': 'chr7', 'ref': 'GGAATTAAGAGAAGCA', 'pos': '55242465', 'alt': 'G'}}
        assert results['NM_005228.4:c.2236_2250del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.55174773_55174787del', 'vcf': {'chr': 'chr7', 'ref': 'GGAATTAAGAGAAGCA', 'pos': '55174772', 'alt': 'G'}}
        assert results['NM_005228.4:c.2236_2250del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.55242466_55242480del', 'vcf': {'chr': '7', 'ref': 'GGAATTAAGAGAAGCA', 'pos': '55242465', 'alt': 'G'}}
        assert results['NM_005228.4:c.2236_2250del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.55174773_55174787del', 'vcf': {'chr': '7', 'ref': 'GGAATTAAGAGAAGCA', 'pos': '55174772', 'alt': 'G'}}


    def test_variant298(self):
        variant = '7-55248992-T-TTCCAGGAAGCCT'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_005228.3:c.2284-5_2290dup' in results.keys()
        assert results['NM_005228.3:c.2284-5_2290dup']['hgvs_lrg_transcript_variant'] == 'LRG_304t1:c.2284-5_2290dup'
        assert results['NM_005228.3:c.2284-5_2290dup']['refseqgene_context_intronic_sequence'] == 'NG_007726.3(NM_005228.3):c.2284-5_2290dup'
        assert results['NM_005228.3:c.2284-5_2290dup']['alt_genomic_loci'] == []
        assert results['NM_005228.3:c.2284-5_2290dup']['gene_symbol'] == 'EGFR'
        assert results['NM_005228.3:c.2284-5_2290dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_005219.2(LRG_304p1):p.?', 'slr': 'NP_005219.2:p.?'}
        assert results['NM_005228.3:c.2284-5_2290dup']['submitted_variant'] == '7-55248992-T-TTCCAGGAAGCCT'
        assert results['NM_005228.3:c.2284-5_2290dup']['genome_context_intronic_sequence'] == 'NC_000007.13(NM_005228.3):c.2284-5_2290dup'
        assert results['NM_005228.3:c.2284-5_2290dup']['hgvs_lrg_variant'] == 'LRG_304:g.167257_167268dup'
        assert results['NM_005228.3:c.2284-5_2290dup']['hgvs_transcript_variant'] == 'NM_005228.3:c.2284-5_2290dup'
        assert results['NM_005228.3:c.2284-5_2290dup']['hgvs_refseqgene_variant'] == 'NG_007726.3:g.167257_167268dup'
        assert results['NM_005228.3:c.2284-5_2290dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.55248981_55248992dup', 'vcf': {'chr': 'chr7', 'ref': 'TCCAGGAAGCCT', 'pos': '55248981', 'alt': 'TCCAGGAAGCCTTCCAGGAAGCCT'}}
        assert results['NM_005228.3:c.2284-5_2290dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.55181288_55181299dup', 'vcf': {'chr': 'chr7', 'ref': 'TCCAGGAAGCCT', 'pos': '55181288', 'alt': 'TCCAGGAAGCCTTCCAGGAAGCCT'}}
        assert results['NM_005228.3:c.2284-5_2290dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.55248981_55248992dup', 'vcf': {'chr': '7', 'ref': 'TCCAGGAAGCCT', 'pos': '55248981', 'alt': 'TCCAGGAAGCCTTCCAGGAAGCCT'}}
        assert results['NM_005228.3:c.2284-5_2290dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.55181288_55181299dup', 'vcf': {'chr': '7', 'ref': 'TCCAGGAAGCCT', 'pos': '55181288', 'alt': 'TCCAGGAAGCCTTCCAGGAAGCCT'}}

        assert 'NM_001346899.1:c.2149-5_2155dup' in results.keys()
        assert results['NM_001346899.1:c.2149-5_2155dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001346899.1:c.2149-5_2155dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001346899.1:c.2149-5_2155dup']['alt_genomic_loci'] == []
        assert results['NM_001346899.1:c.2149-5_2155dup']['gene_symbol'] == 'EGFR'
        assert results['NM_001346899.1:c.2149-5_2155dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001333828.1:p.?', 'slr': 'NP_001333828.1:p.?'}
        assert results['NM_001346899.1:c.2149-5_2155dup']['submitted_variant'] == '7-55248992-T-TTCCAGGAAGCCT'
        assert results['NM_001346899.1:c.2149-5_2155dup']['genome_context_intronic_sequence'] == 'NC_000007.13(NM_001346899.1):c.2149-5_2155dup'
        assert results['NM_001346899.1:c.2149-5_2155dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001346899.1:c.2149-5_2155dup']['hgvs_transcript_variant'] == 'NM_001346899.1:c.2149-5_2155dup'
        assert results['NM_001346899.1:c.2149-5_2155dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001346899.1:c.2149-5_2155dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.55248981_55248992dup', 'vcf': {'chr': 'chr7', 'ref': 'TCCAGGAAGCCT', 'pos': '55248981', 'alt': 'TCCAGGAAGCCTTCCAGGAAGCCT'}}
        assert results['NM_001346899.1:c.2149-5_2155dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.55181288_55181299dup', 'vcf': {'chr': 'chr7', 'ref': 'TCCAGGAAGCCT', 'pos': '55181288', 'alt': 'TCCAGGAAGCCTTCCAGGAAGCCT'}}
        assert results['NM_001346899.1:c.2149-5_2155dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.55248981_55248992dup', 'vcf': {'chr': '7', 'ref': 'TCCAGGAAGCCT', 'pos': '55248981', 'alt': 'TCCAGGAAGCCTTCCAGGAAGCCT'}}
        assert results['NM_001346899.1:c.2149-5_2155dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.55181288_55181299dup', 'vcf': {'chr': '7', 'ref': 'TCCAGGAAGCCT', 'pos': '55181288', 'alt': 'TCCAGGAAGCCTTCCAGGAAGCCT'}}

        assert 'NM_005228.4:c.2284-5_2290dup' in results.keys()
        assert results['NM_005228.4:c.2284-5_2290dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_005228.4:c.2284-5_2290dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_005228.4:c.2284-5_2290dup']['alt_genomic_loci'] == []
        assert results['NM_005228.4:c.2284-5_2290dup']['gene_symbol'] == 'EGFR'
        assert results['NM_005228.4:c.2284-5_2290dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_005219.2(LRG_304p1):p.?', 'slr': 'NP_005219.2:p.?'}
        assert results['NM_005228.4:c.2284-5_2290dup']['submitted_variant'] == '7-55248992-T-TTCCAGGAAGCCT'
        assert results['NM_005228.4:c.2284-5_2290dup']['genome_context_intronic_sequence'] == 'NC_000007.13(NM_005228.4):c.2284-5_2290dup'
        assert results['NM_005228.4:c.2284-5_2290dup']['hgvs_lrg_variant'] == ''
        assert results['NM_005228.4:c.2284-5_2290dup']['hgvs_transcript_variant'] == 'NM_005228.4:c.2284-5_2290dup'
        assert results['NM_005228.4:c.2284-5_2290dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_005228.4:c.2284-5_2290dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.55248981_55248992dup', 'vcf': {'chr': 'chr7', 'ref': 'TCCAGGAAGCCT', 'pos': '55248981', 'alt': 'TCCAGGAAGCCTTCCAGGAAGCCT'}}
        assert results['NM_005228.4:c.2284-5_2290dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.55181288_55181299dup', 'vcf': {'chr': 'chr7', 'ref': 'TCCAGGAAGCCT', 'pos': '55181288', 'alt': 'TCCAGGAAGCCTTCCAGGAAGCCT'}}
        assert results['NM_005228.4:c.2284-5_2290dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.55248981_55248992dup', 'vcf': {'chr': '7', 'ref': 'TCCAGGAAGCCT', 'pos': '55248981', 'alt': 'TCCAGGAAGCCTTCCAGGAAGCCT'}}
        assert results['NM_005228.4:c.2284-5_2290dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.55181288_55181299dup', 'vcf': {'chr': '7', 'ref': 'TCCAGGAAGCCT', 'pos': '55181288', 'alt': 'TCCAGGAAGCCTTCCAGGAAGCCT'}}

        assert 'NM_001346898.1:c.2284-5_2290dup' in results.keys()
        assert results['NM_001346898.1:c.2284-5_2290dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001346898.1:c.2284-5_2290dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001346898.1:c.2284-5_2290dup']['alt_genomic_loci'] == []
        assert results['NM_001346898.1:c.2284-5_2290dup']['gene_symbol'] == 'EGFR'
        assert results['NM_001346898.1:c.2284-5_2290dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001333827.1:p.?', 'slr': 'NP_001333827.1:p.?'}
        assert results['NM_001346898.1:c.2284-5_2290dup']['submitted_variant'] == '7-55248992-T-TTCCAGGAAGCCT'
        assert results['NM_001346898.1:c.2284-5_2290dup']['genome_context_intronic_sequence'] == 'NC_000007.13(NM_001346898.1):c.2284-5_2290dup'
        assert results['NM_001346898.1:c.2284-5_2290dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001346898.1:c.2284-5_2290dup']['hgvs_transcript_variant'] == 'NM_001346898.1:c.2284-5_2290dup'
        assert results['NM_001346898.1:c.2284-5_2290dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001346898.1:c.2284-5_2290dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.55248981_55248992dup', 'vcf': {'chr': 'chr7', 'ref': 'TCCAGGAAGCCT', 'pos': '55248981', 'alt': 'TCCAGGAAGCCTTCCAGGAAGCCT'}}
        assert results['NM_001346898.1:c.2284-5_2290dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.55181288_55181299dup', 'vcf': {'chr': 'chr7', 'ref': 'TCCAGGAAGCCT', 'pos': '55181288', 'alt': 'TCCAGGAAGCCTTCCAGGAAGCCT'}}
        assert results['NM_001346898.1:c.2284-5_2290dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.55248981_55248992dup', 'vcf': {'chr': '7', 'ref': 'TCCAGGAAGCCT', 'pos': '55248981', 'alt': 'TCCAGGAAGCCTTCCAGGAAGCCT'}}
        assert results['NM_001346898.1:c.2284-5_2290dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.55181288_55181299dup', 'vcf': {'chr': '7', 'ref': 'TCCAGGAAGCCT', 'pos': '55181288', 'alt': 'TCCAGGAAGCCTTCCAGGAAGCCT'}}

        assert 'NM_001346941.1:c.1483-5_1489dup' in results.keys()
        assert results['NM_001346941.1:c.1483-5_1489dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001346941.1:c.1483-5_1489dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001346941.1:c.1483-5_1489dup']['alt_genomic_loci'] == []
        assert results['NM_001346941.1:c.1483-5_1489dup']['gene_symbol'] == 'EGFR'
        assert results['NM_001346941.1:c.1483-5_1489dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001333870.1:p.?', 'slr': 'NP_001333870.1:p.?'}
        assert results['NM_001346941.1:c.1483-5_1489dup']['submitted_variant'] == '7-55248992-T-TTCCAGGAAGCCT'
        assert results['NM_001346941.1:c.1483-5_1489dup']['genome_context_intronic_sequence'] == 'NC_000007.13(NM_001346941.1):c.1483-5_1489dup'
        assert results['NM_001346941.1:c.1483-5_1489dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001346941.1:c.1483-5_1489dup']['hgvs_transcript_variant'] == 'NM_001346941.1:c.1483-5_1489dup'
        assert results['NM_001346941.1:c.1483-5_1489dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001346941.1:c.1483-5_1489dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.55248981_55248992dup', 'vcf': {'chr': 'chr7', 'ref': 'TCCAGGAAGCCT', 'pos': '55248981', 'alt': 'TCCAGGAAGCCTTCCAGGAAGCCT'}}
        assert results['NM_001346941.1:c.1483-5_1489dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.55181288_55181299dup', 'vcf': {'chr': 'chr7', 'ref': 'TCCAGGAAGCCT', 'pos': '55181288', 'alt': 'TCCAGGAAGCCTTCCAGGAAGCCT'}}
        assert results['NM_001346941.1:c.1483-5_1489dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.55248981_55248992dup', 'vcf': {'chr': '7', 'ref': 'TCCAGGAAGCCT', 'pos': '55248981', 'alt': 'TCCAGGAAGCCTTCCAGGAAGCCT'}}
        assert results['NM_001346941.1:c.1483-5_1489dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.55181288_55181299dup', 'vcf': {'chr': '7', 'ref': 'TCCAGGAAGCCT', 'pos': '55181288', 'alt': 'TCCAGGAAGCCTTCCAGGAAGCCT'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001346900.1:c.2125-5_2131dup' in results.keys()
        assert results['NM_001346900.1:c.2125-5_2131dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001346900.1:c.2125-5_2131dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001346900.1:c.2125-5_2131dup']['alt_genomic_loci'] == []
        assert results['NM_001346900.1:c.2125-5_2131dup']['gene_symbol'] == 'EGFR'
        assert results['NM_001346900.1:c.2125-5_2131dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001333829.1:p.?', 'slr': 'NP_001333829.1:p.?'}
        assert results['NM_001346900.1:c.2125-5_2131dup']['submitted_variant'] == '7-55248992-T-TTCCAGGAAGCCT'
        assert results['NM_001346900.1:c.2125-5_2131dup']['genome_context_intronic_sequence'] == 'NC_000007.13(NM_001346900.1):c.2125-5_2131dup'
        assert results['NM_001346900.1:c.2125-5_2131dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001346900.1:c.2125-5_2131dup']['hgvs_transcript_variant'] == 'NM_001346900.1:c.2125-5_2131dup'
        assert results['NM_001346900.1:c.2125-5_2131dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001346900.1:c.2125-5_2131dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.55248981_55248992dup', 'vcf': {'chr': 'chr7', 'ref': 'TCCAGGAAGCCT', 'pos': '55248981', 'alt': 'TCCAGGAAGCCTTCCAGGAAGCCT'}}
        assert results['NM_001346900.1:c.2125-5_2131dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.55181288_55181299dup', 'vcf': {'chr': 'chr7', 'ref': 'TCCAGGAAGCCT', 'pos': '55181288', 'alt': 'TCCAGGAAGCCTTCCAGGAAGCCT'}}
        assert results['NM_001346900.1:c.2125-5_2131dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.55248981_55248992dup', 'vcf': {'chr': '7', 'ref': 'TCCAGGAAGCCT', 'pos': '55248981', 'alt': 'TCCAGGAAGCCTTCCAGGAAGCCT'}}
        assert results['NM_001346900.1:c.2125-5_2131dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.55181288_55181299dup', 'vcf': {'chr': '7', 'ref': 'TCCAGGAAGCCT', 'pos': '55181288', 'alt': 'TCCAGGAAGCCTTCCAGGAAGCCT'}}

        assert 'NR_047551.1:n.1272_1283dup' in results.keys()
        assert results['NR_047551.1:n.1272_1283dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_047551.1:n.1272_1283dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_047551.1:n.1272_1283dup']['alt_genomic_loci'] == []
        assert results['NR_047551.1:n.1272_1283dup']['gene_symbol'] == 'EGFR-AS1'
        assert results['NR_047551.1:n.1272_1283dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_047551.1:n.1272_1283dup']['submitted_variant'] == '7-55248992-T-TTCCAGGAAGCCT'
        assert results['NR_047551.1:n.1272_1283dup']['genome_context_intronic_sequence'] == ''
        assert results['NR_047551.1:n.1272_1283dup']['hgvs_lrg_variant'] == ''
        assert results['NR_047551.1:n.1272_1283dup']['hgvs_transcript_variant'] == 'NR_047551.1:n.1272_1283dup'
        assert results['NR_047551.1:n.1272_1283dup']['hgvs_refseqgene_variant'] == ''
        assert results['NR_047551.1:n.1272_1283dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.55248981_55248992dup', 'vcf': {'chr': 'chr7', 'ref': 'TCCAGGAAGCCT', 'pos': '55248981', 'alt': u'TCCAGGAAGCCTTCCAGGAAGCCT'}}
        assert results['NR_047551.1:n.1272_1283dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.55181288_55181299dup', 'vcf': {'chr': 'chr7', 'ref': 'TCCAGGAAGCCT', 'pos': '55181288', 'alt': u'TCCAGGAAGCCTTCCAGGAAGCCT'}}
        assert results['NR_047551.1:n.1272_1283dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.55248981_55248992dup', 'vcf': {'chr': '7', 'ref': 'TCCAGGAAGCCT', 'pos': '55248981', 'alt': u'TCCAGGAAGCCTTCCAGGAAGCCT'}}
        assert results['NR_047551.1:n.1272_1283dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.55181288_55181299dup', 'vcf': {'chr': '7', 'ref': 'TCCAGGAAGCCT', 'pos': '55181288', 'alt': u'TCCAGGAAGCCTTCCAGGAAGCCT'}}

        assert 'NM_001346897.1:c.2149-5_2155dup' in results.keys()
        assert results['NM_001346897.1:c.2149-5_2155dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001346897.1:c.2149-5_2155dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001346897.1:c.2149-5_2155dup']['alt_genomic_loci'] == []
        assert results['NM_001346897.1:c.2149-5_2155dup']['gene_symbol'] == 'EGFR'
        assert results['NM_001346897.1:c.2149-5_2155dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001333826.1:p.?', 'slr': 'NP_001333826.1:p.?'}
        assert results['NM_001346897.1:c.2149-5_2155dup']['submitted_variant'] == '7-55248992-T-TTCCAGGAAGCCT'
        assert results['NM_001346897.1:c.2149-5_2155dup']['genome_context_intronic_sequence'] == 'NC_000007.13(NM_001346897.1):c.2149-5_2155dup'
        assert results['NM_001346897.1:c.2149-5_2155dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001346897.1:c.2149-5_2155dup']['hgvs_transcript_variant'] == 'NM_001346897.1:c.2149-5_2155dup'
        assert results['NM_001346897.1:c.2149-5_2155dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001346897.1:c.2149-5_2155dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.55248981_55248992dup', 'vcf': {'chr': 'chr7', 'ref': 'TCCAGGAAGCCT', 'pos': '55248981', 'alt': 'TCCAGGAAGCCTTCCAGGAAGCCT'}}
        assert results['NM_001346897.1:c.2149-5_2155dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.55181288_55181299dup', 'vcf': {'chr': 'chr7', 'ref': 'TCCAGGAAGCCT', 'pos': '55181288', 'alt': 'TCCAGGAAGCCTTCCAGGAAGCCT'}}
        assert results['NM_001346897.1:c.2149-5_2155dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.55248981_55248992dup', 'vcf': {'chr': '7', 'ref': 'TCCAGGAAGCCT', 'pos': '55248981', 'alt': 'TCCAGGAAGCCTTCCAGGAAGCCT'}}
        assert results['NM_001346897.1:c.2149-5_2155dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.55181288_55181299dup', 'vcf': {'chr': '7', 'ref': 'TCCAGGAAGCCT', 'pos': '55181288', 'alt': 'TCCAGGAAGCCTTCCAGGAAGCCT'}}


    def test_variant299(self):
        variant = '7-75932111-C-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001540.4:c.82C>A' in results.keys()
        assert results['NM_001540.4:c.82C>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001540.4:c.82C>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001540.4:c.82C>A']['alt_genomic_loci'] == []
        assert results['NM_001540.4:c.82C>A']['gene_symbol'] == 'HSPB1'
        assert results['NM_001540.4:c.82C>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001531.1(LRG_248p1):p.(Leu28Ile)', 'slr': 'NP_001531.1:p.(L28I)'}
        assert results['NM_001540.4:c.82C>A']['submitted_variant'] == '7-75932111-C-A'
        assert results['NM_001540.4:c.82C>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_001540.4:c.82C>A']['hgvs_lrg_variant'] == ''
        assert results['NM_001540.4:c.82C>A']['hgvs_transcript_variant'] == 'NM_001540.4:c.82C>A'
        assert results['NM_001540.4:c.82C>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001540.4:c.82C>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.75932111C>A', 'vcf': {'chr': 'chr7', 'ref': 'C', 'pos': '75932111', 'alt': 'A'}}
        assert results['NM_001540.4:c.82C>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.76302794C>A', 'vcf': {'chr': 'chr7', 'ref': 'C', 'pos': '76302794', 'alt': 'A'}}
        assert results['NM_001540.4:c.82C>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.75932111C>A', 'vcf': {'chr': '7', 'ref': 'C', 'pos': '75932111', 'alt': 'A'}}
        assert results['NM_001540.4:c.82C>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.76302794C>A', 'vcf': {'chr': '7', 'ref': 'C', 'pos': '76302794', 'alt': 'A'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001540.3:c.82C>A' in results.keys()
        assert results['NM_001540.3:c.82C>A']['hgvs_lrg_transcript_variant'] == 'LRG_248t1:c.82C>A'
        assert results['NM_001540.3:c.82C>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001540.3:c.82C>A']['alt_genomic_loci'] == []
        assert results['NM_001540.3:c.82C>A']['gene_symbol'] == 'HSPB1'
        assert results['NM_001540.3:c.82C>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001531.1(LRG_248p1):p.(Leu28Ile)', 'slr': 'NP_001531.1:p.(L28I)'}
        assert results['NM_001540.3:c.82C>A']['submitted_variant'] == '7-75932111-C-A'
        assert results['NM_001540.3:c.82C>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_001540.3:c.82C>A']['hgvs_lrg_variant'] == 'LRG_248:g.5237C>A'
        assert results['NM_001540.3:c.82C>A']['hgvs_transcript_variant'] == 'NM_001540.3:c.82C>A'
        assert results['NM_001540.3:c.82C>A']['hgvs_refseqgene_variant'] == 'NG_008995.1:g.5237C>A'
        assert results['NM_001540.3:c.82C>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.75932111C>A', 'vcf': {'chr': 'chr7', 'ref': 'C', 'pos': '75932111', 'alt': 'A'}}
        assert results['NM_001540.3:c.82C>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.76302794C>A', 'vcf': {'chr': 'chr7', 'ref': 'C', 'pos': '76302794', 'alt': 'A'}}
        assert results['NM_001540.3:c.82C>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.75932111C>A', 'vcf': {'chr': '7', 'ref': 'C', 'pos': '75932111', 'alt': 'A'}}
        assert results['NM_001540.3:c.82C>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.76302794C>A', 'vcf': {'chr': '7', 'ref': 'C', 'pos': '76302794', 'alt': 'A'}}


    def test_variant300(self):
        variant = '7-91652178-A-AAAC'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_005751.4:c.4004_4006dup' in results.keys()
        assert results['NM_005751.4:c.4004_4006dup']['hgvs_lrg_transcript_variant'] == 'LRG_331t1:c.4004_4006dup'
        assert results['NM_005751.4:c.4004_4006dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_005751.4:c.4004_4006dup']['alt_genomic_loci'] == []
        assert results['NM_005751.4:c.4004_4006dup']['gene_symbol'] == 'AKAP9'
        assert results['NM_005751.4:c.4004_4006dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_005742.4(LRG_331p1):p.(Lys1335_Leu1336insGln)', 'slr': 'NP_005742.4:p.(K1335_L1336insQ)'}
        assert results['NM_005751.4:c.4004_4006dup']['submitted_variant'] == '7-91652178-A-AAAC'
        assert results['NM_005751.4:c.4004_4006dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_005751.4:c.4004_4006dup']['hgvs_lrg_variant'] == 'LRG_331:g.86991_86993dup'
        assert results['NM_005751.4:c.4004_4006dup']['hgvs_transcript_variant'] == 'NM_005751.4:c.4004_4006dup'
        assert results['NM_005751.4:c.4004_4006dup']['hgvs_refseqgene_variant'] == 'NG_011623.1:g.86991_86993dup'
        assert results['NM_005751.4:c.4004_4006dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.91652179_91652181dup', 'vcf': {'chr': 'chr7', 'ref': 'AAC', 'pos': '91652179', 'alt': 'AACAAC'}}
        assert results['NM_005751.4:c.4004_4006dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.92022865_92022867dup', 'vcf': {'chr': 'chr7', 'ref': 'AAC', 'pos': '92022865', 'alt': 'AACAAC'}}
        assert results['NM_005751.4:c.4004_4006dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.91652179_91652181dup', 'vcf': {'chr': '7', 'ref': 'AAC', 'pos': '91652179', 'alt': 'AACAAC'}}
        assert results['NM_005751.4:c.4004_4006dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.92022865_92022867dup', 'vcf': {'chr': '7', 'ref': 'AAC', 'pos': '92022865', 'alt': 'AACAAC'}}

        assert 'NM_147185.2:c.4004_4006dup' in results.keys()
        assert results['NM_147185.2:c.4004_4006dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_147185.2:c.4004_4006dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_147185.2:c.4004_4006dup']['alt_genomic_loci'] == []
        assert results['NM_147185.2:c.4004_4006dup']['gene_symbol'] == 'AKAP9'
        assert results['NM_147185.2:c.4004_4006dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_671714.1:p.(Lys1335_Leu1336insGln)', 'slr': 'NP_671714.1:p.(K1335_L1336insQ)'}
        assert results['NM_147185.2:c.4004_4006dup']['submitted_variant'] == '7-91652178-A-AAAC'
        assert results['NM_147185.2:c.4004_4006dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_147185.2:c.4004_4006dup']['hgvs_lrg_variant'] == ''
        assert results['NM_147185.2:c.4004_4006dup']['hgvs_transcript_variant'] == 'NM_147185.2:c.4004_4006dup'
        assert results['NM_147185.2:c.4004_4006dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_147185.2:c.4004_4006dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.91652179_91652181dup', 'vcf': {'chr': 'chr7', 'ref': 'AAC', 'pos': '91652179', 'alt': 'AACAAC'}}
        assert results['NM_147185.2:c.4004_4006dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.92022865_92022867dup', 'vcf': {'chr': 'chr7', 'ref': 'AAC', 'pos': '92022865', 'alt': 'AACAAC'}}
        assert results['NM_147185.2:c.4004_4006dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.91652179_91652181dup', 'vcf': {'chr': '7', 'ref': 'AAC', 'pos': '91652179', 'alt': 'AACAAC'}}
        assert results['NM_147185.2:c.4004_4006dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.92022865_92022867dup', 'vcf': {'chr': '7', 'ref': 'AAC', 'pos': '92022865', 'alt': 'AACAAC'}}


    def test_variant301(self):
        variant = '7-117199644-ATCT-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NR_149084.1:n.221+1140_221+1142del' in results.keys()
        assert results['NR_149084.1:n.221+1140_221+1142del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_149084.1:n.221+1140_221+1142del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_149084.1:n.221+1140_221+1142del']['alt_genomic_loci'] == []
        assert results['NR_149084.1:n.221+1140_221+1142del']['gene_symbol'] == 'CFTR-AS1'
        assert results['NR_149084.1:n.221+1140_221+1142del']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_149084.1:n.221+1140_221+1142del']['submitted_variant'] == '7-117199644-ATCT-A'
        assert results['NR_149084.1:n.221+1140_221+1142del']['genome_context_intronic_sequence'] == 'NC_000007.13(NR_149084.1):c.221+1140_221+1142del'
        assert results['NR_149084.1:n.221+1140_221+1142del']['hgvs_lrg_variant'] == ''
        assert results['NR_149084.1:n.221+1140_221+1142del']['hgvs_transcript_variant'] == 'NR_149084.1:n.221+1140_221+1142del'
        assert results['NR_149084.1:n.221+1140_221+1142del']['hgvs_refseqgene_variant'] == ''
        assert results['NR_149084.1:n.221+1140_221+1142del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.117199645_117199647del', 'vcf': {'chr': 'chr7', 'ref': 'ATCT', 'pos': '117199644', 'alt': 'A'}}
        assert results['NR_149084.1:n.221+1140_221+1142del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.117559591_117559593del', 'vcf': {'chr': 'chr7', 'ref': 'ATCT', 'pos': '117559590', 'alt': 'A'}}
        assert results['NR_149084.1:n.221+1140_221+1142del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.117199645_117199647del', 'vcf': {'chr': '7', 'ref': 'ATCT', 'pos': '117199644', 'alt': 'A'}}
        assert results['NR_149084.1:n.221+1140_221+1142del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.117559591_117559593del', 'vcf': {'chr': '7', 'ref': 'ATCT', 'pos': '117559590', 'alt': 'A'}}

        assert 'NM_000492.3:c.1521_1523del' in results.keys()
        assert results['NM_000492.3:c.1521_1523del']['hgvs_lrg_transcript_variant'] == 'LRG_663t1:c.1521_1523del'
        assert results['NM_000492.3:c.1521_1523del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000492.3:c.1521_1523del']['alt_genomic_loci'] == []
        assert results['NM_000492.3:c.1521_1523del']['gene_symbol'] == 'CFTR'
        assert results['NM_000492.3:c.1521_1523del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000483.3(LRG_663p1):p.(Phe508del)', 'slr': 'NP_000483.3:p.(F508del)'}
        assert results['NM_000492.3:c.1521_1523del']['submitted_variant'] == '7-117199644-ATCT-A'
        assert results['NM_000492.3:c.1521_1523del']['genome_context_intronic_sequence'] == ''
        assert results['NM_000492.3:c.1521_1523del']['hgvs_lrg_variant'] == ''
        assert results['NM_000492.3:c.1521_1523del']['hgvs_transcript_variant'] == 'NM_000492.3:c.1521_1523del'
        assert results['NM_000492.3:c.1521_1523del']['hgvs_refseqgene_variant'] == 'NG_016465.3:g.98809_98811del'
        assert results['NM_000492.3:c.1521_1523del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.117199646_117199648del', 'vcf': {'chr': 'chr7', 'ref': 'ATCT', 'pos': '117199644', 'alt': 'A'}}
        assert results['NM_000492.3:c.1521_1523del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.117559592_117559594del', 'vcf': {'chr': 'chr7', 'ref': 'ATCT', 'pos': '117559590', 'alt': 'A'}}
        assert results['NM_000492.3:c.1521_1523del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.117199646_117199648del', 'vcf': {'chr': '7', 'ref': 'ATCT', 'pos': '117199644', 'alt': 'A'}}
        assert results['NM_000492.3:c.1521_1523del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.117559592_117559594del', 'vcf': {'chr': '7', 'ref': 'ATCT', 'pos': '117559590', 'alt': 'A'}}


    def test_variant302(self):
        variant = '7-140453136-AC-CT'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_004333.4:c.1798_1799delinsAG' in results.keys()
        assert results['NM_004333.4:c.1798_1799delinsAG']['hgvs_lrg_transcript_variant'] == 'LRG_299t1:c.1798_1799delinsAG'
        assert results['NM_004333.4:c.1798_1799delinsAG']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_004333.4:c.1798_1799delinsAG']['alt_genomic_loci'] == []
        assert results['NM_004333.4:c.1798_1799delinsAG']['gene_symbol'] == 'BRAF'
        assert results['NM_004333.4:c.1798_1799delinsAG']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_004324.2(LRG_299p1):p.(Val600Arg)', 'slr': 'NP_004324.2:p.(V600R)'}
        assert results['NM_004333.4:c.1798_1799delinsAG']['submitted_variant'] == '7-140453136-AC-CT'
        assert results['NM_004333.4:c.1798_1799delinsAG']['genome_context_intronic_sequence'] == ''
        assert results['NM_004333.4:c.1798_1799delinsAG']['hgvs_lrg_variant'] == ''
        assert results['NM_004333.4:c.1798_1799delinsAG']['hgvs_transcript_variant'] == 'NM_004333.4:c.1798_1799delinsAG'
        assert results['NM_004333.4:c.1798_1799delinsAG']['hgvs_refseqgene_variant'] == 'NG_007873.2:g.176428_176429delinsAG'
        assert results['NM_004333.4:c.1798_1799delinsAG']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.140453136_140453137delinsCT', 'vcf': {'chr': 'chr7', 'ref': 'AC', 'pos': '140453136', 'alt': u'CT'}}
        assert results['NM_004333.4:c.1798_1799delinsAG']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.140753336_140753337delinsCT', 'vcf': {'chr': 'chr7', 'ref': 'AC', 'pos': '140753336', 'alt': u'CT'}}
        assert results['NM_004333.4:c.1798_1799delinsAG']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.140453136_140453137delinsCT', 'vcf': {'chr': '7', 'ref': 'AC', 'pos': '140453136', 'alt': u'CT'}}
        assert results['NM_004333.4:c.1798_1799delinsAG']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.140753336_140753337delinsCT', 'vcf': {'chr': '7', 'ref': 'AC', 'pos': '140753336', 'alt': u'CT'}}

        assert 'NM_004333.5:c.1798_1799delinsAG' in results.keys()
        assert results['NM_004333.5:c.1798_1799delinsAG']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_004333.5:c.1798_1799delinsAG']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_004333.5:c.1798_1799delinsAG']['alt_genomic_loci'] == []
        assert results['NM_004333.5:c.1798_1799delinsAG']['gene_symbol'] == 'BRAF'
        assert results['NM_004333.5:c.1798_1799delinsAG']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_004324.2(LRG_299p1):p.(Val600Arg)', 'slr': 'NP_004324.2:p.(V600R)'}
        assert results['NM_004333.5:c.1798_1799delinsAG']['submitted_variant'] == '7-140453136-AC-CT'
        assert results['NM_004333.5:c.1798_1799delinsAG']['genome_context_intronic_sequence'] == ''
        assert results['NM_004333.5:c.1798_1799delinsAG']['hgvs_lrg_variant'] == ''
        assert results['NM_004333.5:c.1798_1799delinsAG']['hgvs_transcript_variant'] == 'NM_004333.5:c.1798_1799delinsAG'
        assert results['NM_004333.5:c.1798_1799delinsAG']['hgvs_refseqgene_variant'] == ''
        assert results['NM_004333.5:c.1798_1799delinsAG']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.140453136_140453137delinsCT', 'vcf': {'chr': 'chr7', 'ref': 'AC', 'pos': '140453136', 'alt': u'CT'}}
        assert results['NM_004333.5:c.1798_1799delinsAG']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.140753336_140753337delinsCT', 'vcf': {'chr': 'chr7', 'ref': 'AC', 'pos': '140753336', 'alt': u'CT'}}
        assert results['NM_004333.5:c.1798_1799delinsAG']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.140453136_140453137delinsCT', 'vcf': {'chr': '7', 'ref': 'AC', 'pos': '140453136', 'alt': u'CT'}}
        assert results['NM_004333.5:c.1798_1799delinsAG']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.140753336_140753337delinsCT', 'vcf': {'chr': '7', 'ref': 'AC', 'pos': '140753336', 'alt': u'CT'}}

        assert results['flag'] == 'gene_variant'
        assert 'NR_148928.1:n.2896_2897delinsAG' in results.keys()
        assert results['NR_148928.1:n.2896_2897delinsAG']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_148928.1:n.2896_2897delinsAG']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_148928.1:n.2896_2897delinsAG']['alt_genomic_loci'] == []
        assert results['NR_148928.1:n.2896_2897delinsAG']['gene_symbol'] == 'BRAF'
        assert results['NR_148928.1:n.2896_2897delinsAG']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_148928.1:n.2896_2897delinsAG']['submitted_variant'] == '7-140453136-AC-CT'
        assert results['NR_148928.1:n.2896_2897delinsAG']['genome_context_intronic_sequence'] == ''
        assert results['NR_148928.1:n.2896_2897delinsAG']['hgvs_lrg_variant'] == ''
        assert results['NR_148928.1:n.2896_2897delinsAG']['hgvs_transcript_variant'] == 'NR_148928.1:n.2896_2897delinsAG'
        assert results['NR_148928.1:n.2896_2897delinsAG']['hgvs_refseqgene_variant'] == ''
        assert results['NR_148928.1:n.2896_2897delinsAG']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.140453136_140453137delinsCT', 'vcf': {'chr': 'chr7', 'ref': 'AC', 'pos': '140453136', 'alt': u'CT'}}
        assert results['NR_148928.1:n.2896_2897delinsAG']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.140753336_140753337delinsCT', 'vcf': {'chr': 'chr7', 'ref': 'AC', 'pos': '140753336', 'alt': u'CT'}}
        assert results['NR_148928.1:n.2896_2897delinsAG']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.140453136_140453137delinsCT', 'vcf': {'chr': '7', 'ref': 'AC', 'pos': '140453136', 'alt': u'CT'}}
        assert results['NR_148928.1:n.2896_2897delinsAG']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.140753336_140753337delinsCT', 'vcf': {'chr': '7', 'ref': 'AC', 'pos': '140753336', 'alt': u'CT'}}

        assert 'NM_001354609.1:c.1798_1799delinsAG' in results.keys()
        assert results['NM_001354609.1:c.1798_1799delinsAG']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001354609.1:c.1798_1799delinsAG']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001354609.1:c.1798_1799delinsAG']['alt_genomic_loci'] == []
        assert results['NM_001354609.1:c.1798_1799delinsAG']['gene_symbol'] == 'BRAF'
        assert results['NM_001354609.1:c.1798_1799delinsAG']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001341538.1:p.(Val600Arg)', 'slr': 'NP_001341538.1:p.(V600R)'}
        assert results['NM_001354609.1:c.1798_1799delinsAG']['submitted_variant'] == '7-140453136-AC-CT'
        assert results['NM_001354609.1:c.1798_1799delinsAG']['genome_context_intronic_sequence'] == ''
        assert results['NM_001354609.1:c.1798_1799delinsAG']['hgvs_lrg_variant'] == ''
        assert results['NM_001354609.1:c.1798_1799delinsAG']['hgvs_transcript_variant'] == 'NM_001354609.1:c.1798_1799delinsAG'
        assert results['NM_001354609.1:c.1798_1799delinsAG']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001354609.1:c.1798_1799delinsAG']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.140453136_140453137delinsCT', 'vcf': {'chr': 'chr7', 'ref': 'AC', 'pos': '140453136', 'alt': u'CT'}}
        assert results['NM_001354609.1:c.1798_1799delinsAG']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.140753336_140753337delinsCT', 'vcf': {'chr': 'chr7', 'ref': 'AC', 'pos': '140753336', 'alt': u'CT'}}
        assert results['NM_001354609.1:c.1798_1799delinsAG']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.140453136_140453137delinsCT', 'vcf': {'chr': '7', 'ref': 'AC', 'pos': '140453136', 'alt': u'CT'}}
        assert results['NM_001354609.1:c.1798_1799delinsAG']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.140753336_140753337delinsCT', 'vcf': {'chr': '7', 'ref': 'AC', 'pos': '140753336', 'alt': u'CT'}}


    def test_variant303(self):
        variant = '7-140453136-A-T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001354609.1:c.1799T>A' in results.keys()
        assert results['NM_001354609.1:c.1799T>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001354609.1:c.1799T>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001354609.1:c.1799T>A']['alt_genomic_loci'] == []
        assert results['NM_001354609.1:c.1799T>A']['gene_symbol'] == 'BRAF'
        assert results['NM_001354609.1:c.1799T>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001341538.1:p.(Val600Glu)', 'slr': 'NP_001341538.1:p.(V600E)'}
        assert results['NM_001354609.1:c.1799T>A']['submitted_variant'] == '7-140453136-A-T'
        assert results['NM_001354609.1:c.1799T>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_001354609.1:c.1799T>A']['hgvs_lrg_variant'] == ''
        assert results['NM_001354609.1:c.1799T>A']['hgvs_transcript_variant'] == 'NM_001354609.1:c.1799T>A'
        assert results['NM_001354609.1:c.1799T>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001354609.1:c.1799T>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.140453136A>T', 'vcf': {'chr': 'chr7', 'ref': u'A', 'pos': '140453136', 'alt': u'T'}}
        assert results['NM_001354609.1:c.1799T>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.140753336A>T', 'vcf': {'chr': 'chr7', 'ref': u'A', 'pos': '140753336', 'alt': u'T'}}
        assert results['NM_001354609.1:c.1799T>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.140453136A>T', 'vcf': {'chr': '7', 'ref': u'A', 'pos': '140453136', 'alt': u'T'}}
        assert results['NM_001354609.1:c.1799T>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.140753336A>T', 'vcf': {'chr': '7', 'ref': u'A', 'pos': '140753336', 'alt': u'T'}}

        assert results['flag'] == 'gene_variant'
        assert 'NR_148928.1:n.2897T>A' in results.keys()
        assert results['NR_148928.1:n.2897T>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_148928.1:n.2897T>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_148928.1:n.2897T>A']['alt_genomic_loci'] == []
        assert results['NR_148928.1:n.2897T>A']['gene_symbol'] == 'BRAF'
        assert results['NR_148928.1:n.2897T>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_148928.1:n.2897T>A']['submitted_variant'] == '7-140453136-A-T'
        assert results['NR_148928.1:n.2897T>A']['genome_context_intronic_sequence'] == ''
        assert results['NR_148928.1:n.2897T>A']['hgvs_lrg_variant'] == ''
        assert results['NR_148928.1:n.2897T>A']['hgvs_transcript_variant'] == 'NR_148928.1:n.2897T>A'
        assert results['NR_148928.1:n.2897T>A']['hgvs_refseqgene_variant'] == ''
        assert results['NR_148928.1:n.2897T>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.140453136A>T', 'vcf': {'chr': 'chr7', 'ref': u'A', 'pos': '140453136', 'alt': u'T'}}
        assert results['NR_148928.1:n.2897T>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.140753336A>T', 'vcf': {'chr': 'chr7', 'ref': u'A', 'pos': '140753336', 'alt': u'T'}}
        assert results['NR_148928.1:n.2897T>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.140453136A>T', 'vcf': {'chr': '7', 'ref': u'A', 'pos': '140453136', 'alt': u'T'}}
        assert results['NR_148928.1:n.2897T>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.140753336A>T', 'vcf': {'chr': '7', 'ref': u'A', 'pos': '140753336', 'alt': u'T'}}

        assert 'NM_004333.4:c.1799T>A' in results.keys()
        assert results['NM_004333.4:c.1799T>A']['hgvs_lrg_transcript_variant'] == 'LRG_299t1:c.1799T>A'
        assert results['NM_004333.4:c.1799T>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_004333.4:c.1799T>A']['alt_genomic_loci'] == []
        assert results['NM_004333.4:c.1799T>A']['gene_symbol'] == 'BRAF'
        assert results['NM_004333.4:c.1799T>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_004324.2(LRG_299p1):p.(Val600Glu)', 'slr': 'NP_004324.2:p.(V600E)'}
        assert results['NM_004333.4:c.1799T>A']['submitted_variant'] == '7-140453136-A-T'
        assert results['NM_004333.4:c.1799T>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_004333.4:c.1799T>A']['hgvs_lrg_variant'] == ''
        assert results['NM_004333.4:c.1799T>A']['hgvs_transcript_variant'] == 'NM_004333.4:c.1799T>A'
        assert results['NM_004333.4:c.1799T>A']['hgvs_refseqgene_variant'] == 'NG_007873.2:g.176429T>A'
        assert results['NM_004333.4:c.1799T>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.140453136A>T', 'vcf': {'chr': 'chr7', 'ref': u'A', 'pos': '140453136', 'alt': u'T'}}
        assert results['NM_004333.4:c.1799T>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.140753336A>T', 'vcf': {'chr': 'chr7', 'ref': u'A', 'pos': '140753336', 'alt': u'T'}}
        assert results['NM_004333.4:c.1799T>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.140453136A>T', 'vcf': {'chr': '7', 'ref': u'A', 'pos': '140453136', 'alt': u'T'}}
        assert results['NM_004333.4:c.1799T>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.140753336A>T', 'vcf': {'chr': '7', 'ref': u'A', 'pos': '140753336', 'alt': u'T'}}

        assert 'NM_004333.5:c.1799T>A' in results.keys()
        assert results['NM_004333.5:c.1799T>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_004333.5:c.1799T>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_004333.5:c.1799T>A']['alt_genomic_loci'] == []
        assert results['NM_004333.5:c.1799T>A']['gene_symbol'] == 'BRAF'
        assert results['NM_004333.5:c.1799T>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_004324.2(LRG_299p1):p.(Val600Glu)', 'slr': 'NP_004324.2:p.(V600E)'}
        assert results['NM_004333.5:c.1799T>A']['submitted_variant'] == '7-140453136-A-T'
        assert results['NM_004333.5:c.1799T>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_004333.5:c.1799T>A']['hgvs_lrg_variant'] == ''
        assert results['NM_004333.5:c.1799T>A']['hgvs_transcript_variant'] == 'NM_004333.5:c.1799T>A'
        assert results['NM_004333.5:c.1799T>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_004333.5:c.1799T>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.140453136A>T', 'vcf': {'chr': 'chr7', 'ref': u'A', 'pos': '140453136', 'alt': u'T'}}
        assert results['NM_004333.5:c.1799T>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.140753336A>T', 'vcf': {'chr': 'chr7', 'ref': u'A', 'pos': '140753336', 'alt': u'T'}}
        assert results['NM_004333.5:c.1799T>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.140453136A>T', 'vcf': {'chr': '7', 'ref': u'A', 'pos': '140453136', 'alt': u'T'}}
        assert results['NM_004333.5:c.1799T>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.140753336A>T', 'vcf': {'chr': '7', 'ref': u'A', 'pos': '140753336', 'alt': u'T'}}


    def test_variant304(self):
        variant = '7-140453137-C-T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_004333.5:c.1798G>A' in results.keys()
        assert results['NM_004333.5:c.1798G>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_004333.5:c.1798G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_004333.5:c.1798G>A']['alt_genomic_loci'] == []
        assert results['NM_004333.5:c.1798G>A']['gene_symbol'] == 'BRAF'
        assert results['NM_004333.5:c.1798G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_004324.2(LRG_299p1):p.(Val600Met)', 'slr': 'NP_004324.2:p.(V600M)'}
        assert results['NM_004333.5:c.1798G>A']['submitted_variant'] == '7-140453137-C-T'
        assert results['NM_004333.5:c.1798G>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_004333.5:c.1798G>A']['hgvs_lrg_variant'] == ''
        assert results['NM_004333.5:c.1798G>A']['hgvs_transcript_variant'] == 'NM_004333.5:c.1798G>A'
        assert results['NM_004333.5:c.1798G>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_004333.5:c.1798G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.140453137C>T', 'vcf': {'chr': 'chr7', 'ref': u'C', 'pos': '140453137', 'alt': u'T'}}
        assert results['NM_004333.5:c.1798G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.140753337C>T', 'vcf': {'chr': 'chr7', 'ref': u'C', 'pos': '140753337', 'alt': u'T'}}
        assert results['NM_004333.5:c.1798G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.140453137C>T', 'vcf': {'chr': '7', 'ref': u'C', 'pos': '140453137', 'alt': u'T'}}
        assert results['NM_004333.5:c.1798G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.140753337C>T', 'vcf': {'chr': '7', 'ref': u'C', 'pos': '140753337', 'alt': u'T'}}

        assert 'NR_148928.1:n.2896G>A' in results.keys()
        assert results['NR_148928.1:n.2896G>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_148928.1:n.2896G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_148928.1:n.2896G>A']['alt_genomic_loci'] == []
        assert results['NR_148928.1:n.2896G>A']['gene_symbol'] == 'BRAF'
        assert results['NR_148928.1:n.2896G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_148928.1:n.2896G>A']['submitted_variant'] == '7-140453137-C-T'
        assert results['NR_148928.1:n.2896G>A']['genome_context_intronic_sequence'] == ''
        assert results['NR_148928.1:n.2896G>A']['hgvs_lrg_variant'] == ''
        assert results['NR_148928.1:n.2896G>A']['hgvs_transcript_variant'] == 'NR_148928.1:n.2896G>A'
        assert results['NR_148928.1:n.2896G>A']['hgvs_refseqgene_variant'] == ''
        assert results['NR_148928.1:n.2896G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.140453137C>T', 'vcf': {'chr': 'chr7', 'ref': u'C', 'pos': '140453137', 'alt': u'T'}}
        assert results['NR_148928.1:n.2896G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.140753337C>T', 'vcf': {'chr': 'chr7', 'ref': u'C', 'pos': '140753337', 'alt': u'T'}}
        assert results['NR_148928.1:n.2896G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.140453137C>T', 'vcf': {'chr': '7', 'ref': u'C', 'pos': '140453137', 'alt': u'T'}}
        assert results['NR_148928.1:n.2896G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.140753337C>T', 'vcf': {'chr': '7', 'ref': u'C', 'pos': '140753337', 'alt': u'T'}}

        assert 'NM_004333.4:c.1798G>A' in results.keys()
        assert results['NM_004333.4:c.1798G>A']['hgvs_lrg_transcript_variant'] == 'LRG_299t1:c.1798G>A'
        assert results['NM_004333.4:c.1798G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_004333.4:c.1798G>A']['alt_genomic_loci'] == []
        assert results['NM_004333.4:c.1798G>A']['gene_symbol'] == 'BRAF'
        assert results['NM_004333.4:c.1798G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_004324.2(LRG_299p1):p.(Val600Met)', 'slr': 'NP_004324.2:p.(V600M)'}
        assert results['NM_004333.4:c.1798G>A']['submitted_variant'] == '7-140453137-C-T'
        assert results['NM_004333.4:c.1798G>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_004333.4:c.1798G>A']['hgvs_lrg_variant'] == ''
        assert results['NM_004333.4:c.1798G>A']['hgvs_transcript_variant'] == 'NM_004333.4:c.1798G>A'
        assert results['NM_004333.4:c.1798G>A']['hgvs_refseqgene_variant'] == 'NG_007873.2:g.176428G>A'
        assert results['NM_004333.4:c.1798G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.140453137C>T', 'vcf': {'chr': 'chr7', 'ref': u'C', 'pos': '140453137', 'alt': u'T'}}
        assert results['NM_004333.4:c.1798G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.140753337C>T', 'vcf': {'chr': 'chr7', 'ref': u'C', 'pos': '140753337', 'alt': u'T'}}
        assert results['NM_004333.4:c.1798G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.140453137C>T', 'vcf': {'chr': '7', 'ref': u'C', 'pos': '140453137', 'alt': u'T'}}
        assert results['NM_004333.4:c.1798G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.140753337C>T', 'vcf': {'chr': '7', 'ref': u'C', 'pos': '140753337', 'alt': u'T'}}

        assert 'NM_001354609.1:c.1798G>A' in results.keys()
        assert results['NM_001354609.1:c.1798G>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001354609.1:c.1798G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001354609.1:c.1798G>A']['alt_genomic_loci'] == []
        assert results['NM_001354609.1:c.1798G>A']['gene_symbol'] == 'BRAF'
        assert results['NM_001354609.1:c.1798G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001341538.1:p.(Val600Met)', 'slr': 'NP_001341538.1:p.(V600M)'}
        assert results['NM_001354609.1:c.1798G>A']['submitted_variant'] == '7-140453137-C-T'
        assert results['NM_001354609.1:c.1798G>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_001354609.1:c.1798G>A']['hgvs_lrg_variant'] == ''
        assert results['NM_001354609.1:c.1798G>A']['hgvs_transcript_variant'] == 'NM_001354609.1:c.1798G>A'
        assert results['NM_001354609.1:c.1798G>A']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001354609.1:c.1798G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.140453137C>T', 'vcf': {'chr': 'chr7', 'ref': u'C', 'pos': '140453137', 'alt': u'T'}}
        assert results['NM_001354609.1:c.1798G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.140753337C>T', 'vcf': {'chr': 'chr7', 'ref': u'C', 'pos': '140753337', 'alt': u'T'}}
        assert results['NM_001354609.1:c.1798G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.140453137C>T', 'vcf': {'chr': '7', 'ref': u'C', 'pos': '140453137', 'alt': u'T'}}
        assert results['NM_001354609.1:c.1798G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.140753337C>T', 'vcf': {'chr': '7', 'ref': u'C', 'pos': '140753337', 'alt': u'T'}}


    def test_variant305(self):
        variant = '7-143013488-A-T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000083.2:c.180+3A>T' in results.keys()
        assert results['NM_000083.2:c.180+3A>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000083.2:c.180+3A>T']['refseqgene_context_intronic_sequence'] == 'NG_009815.1(NM_000083.2):c.180+3A>T'
        assert results['NM_000083.2:c.180+3A>T']['alt_genomic_loci'] == []
        assert results['NM_000083.2:c.180+3A>T']['gene_symbol'] == 'CLCN1'
        assert results['NM_000083.2:c.180+3A>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000074.2:p.?', 'slr': 'NP_000074.2:p.?'}
        assert results['NM_000083.2:c.180+3A>T']['submitted_variant'] == '7-143013488-A-T'
        assert results['NM_000083.2:c.180+3A>T']['genome_context_intronic_sequence'] == 'NC_000007.13(NM_000083.2):c.180+3A>T'
        assert results['NM_000083.2:c.180+3A>T']['hgvs_lrg_variant'] == ''
        assert results['NM_000083.2:c.180+3A>T']['hgvs_transcript_variant'] == 'NM_000083.2:c.180+3A>T'
        assert results['NM_000083.2:c.180+3A>T']['hgvs_refseqgene_variant'] == 'NG_009815.1:g.5270A>T'
        assert results['NM_000083.2:c.180+3A>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.143013488A>T', 'vcf': {'chr': 'chr7', 'ref': 'A', 'pos': '143013488', 'alt': 'T'}}
        assert results['NM_000083.2:c.180+3A>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.143316395A>T', 'vcf': {'chr': 'chr7', 'ref': 'A', 'pos': '143316395', 'alt': 'T'}}
        assert results['NM_000083.2:c.180+3A>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.143013488A>T', 'vcf': {'chr': '7', 'ref': 'A', 'pos': '143013488', 'alt': 'T'}}
        assert results['NM_000083.2:c.180+3A>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.143316395A>T', 'vcf': {'chr': '7', 'ref': 'A', 'pos': '143316395', 'alt': 'T'}}

        assert 'NR_046453.1:n.267+3A>T' in results.keys()
        assert results['NR_046453.1:n.267+3A>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_046453.1:n.267+3A>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_046453.1:n.267+3A>T']['alt_genomic_loci'] == []
        assert results['NR_046453.1:n.267+3A>T']['gene_symbol'] == 'CLCN1'
        assert results['NR_046453.1:n.267+3A>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_046453.1:n.267+3A>T']['submitted_variant'] == '7-143013488-A-T'
        assert results['NR_046453.1:n.267+3A>T']['genome_context_intronic_sequence'] == 'NC_000007.13(NR_046453.1):c.267+3A>T'
        assert results['NR_046453.1:n.267+3A>T']['hgvs_lrg_variant'] == ''
        assert results['NR_046453.1:n.267+3A>T']['hgvs_transcript_variant'] == 'NR_046453.1:n.267+3A>T'
        assert results['NR_046453.1:n.267+3A>T']['hgvs_refseqgene_variant'] == ''
        assert results['NR_046453.1:n.267+3A>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.143013488A>T', 'vcf': {'chr': 'chr7', 'ref': 'A', 'pos': '143013488', 'alt': 'T'}}
        assert results['NR_046453.1:n.267+3A>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.143316395A>T', 'vcf': {'chr': 'chr7', 'ref': 'A', 'pos': '143316395', 'alt': 'T'}}
        assert results['NR_046453.1:n.267+3A>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.143013488A>T', 'vcf': {'chr': '7', 'ref': 'A', 'pos': '143013488', 'alt': 'T'}}
        assert results['NR_046453.1:n.267+3A>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.143316395A>T', 'vcf': {'chr': '7', 'ref': 'A', 'pos': '143316395', 'alt': 'T'}}


    def test_variant306(self):
        variant = '7-143018934-G-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NR_046453.1:n.776G>A' in results.keys()
        assert results['NR_046453.1:n.776G>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_046453.1:n.776G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_046453.1:n.776G>A']['alt_genomic_loci'] == []
        assert results['NR_046453.1:n.776G>A']['gene_symbol'] == 'CLCN1'
        assert results['NR_046453.1:n.776G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_046453.1:n.776G>A']['submitted_variant'] == '7-143018934-G-A'
        assert results['NR_046453.1:n.776G>A']['genome_context_intronic_sequence'] == ''
        assert results['NR_046453.1:n.776G>A']['hgvs_lrg_variant'] == ''
        assert results['NR_046453.1:n.776G>A']['hgvs_transcript_variant'] == 'NR_046453.1:n.776G>A'
        assert results['NR_046453.1:n.776G>A']['hgvs_refseqgene_variant'] == ''
        assert results['NR_046453.1:n.776G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.143018934G>A', 'vcf': {'chr': 'chr7', 'ref': 'G', 'pos': '143018934', 'alt': 'A'}}
        assert results['NR_046453.1:n.776G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.143321841G>A', 'vcf': {'chr': 'chr7', 'ref': 'G', 'pos': '143321841', 'alt': 'A'}}
        assert results['NR_046453.1:n.776G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.143018934G>A', 'vcf': {'chr': '7', 'ref': 'G', 'pos': '143018934', 'alt': 'A'}}
        assert results['NR_046453.1:n.776G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.143321841G>A', 'vcf': {'chr': '7', 'ref': 'G', 'pos': '143321841', 'alt': 'A'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_000083.2:c.689G>A' in results.keys()
        assert results['NM_000083.2:c.689G>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000083.2:c.689G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000083.2:c.689G>A']['alt_genomic_loci'] == []
        assert results['NM_000083.2:c.689G>A']['gene_symbol'] == 'CLCN1'
        assert results['NM_000083.2:c.689G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000074.2:p.(Gly230Glu)', 'slr': 'NP_000074.2:p.(G230E)'}
        assert results['NM_000083.2:c.689G>A']['submitted_variant'] == '7-143018934-G-A'
        assert results['NM_000083.2:c.689G>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_000083.2:c.689G>A']['hgvs_lrg_variant'] == ''
        assert results['NM_000083.2:c.689G>A']['hgvs_transcript_variant'] == 'NM_000083.2:c.689G>A'
        assert results['NM_000083.2:c.689G>A']['hgvs_refseqgene_variant'] == 'NG_009815.1:g.10716G>A'
        assert results['NM_000083.2:c.689G>A']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.143018934G>A', 'vcf': {'chr': 'chr7', 'ref': 'G', 'pos': '143018934', 'alt': 'A'}}
        assert results['NM_000083.2:c.689G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.143321841G>A', 'vcf': {'chr': 'chr7', 'ref': 'G', 'pos': '143321841', 'alt': 'A'}}
        assert results['NM_000083.2:c.689G>A']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.143018934G>A', 'vcf': {'chr': '7', 'ref': 'G', 'pos': '143018934', 'alt': 'A'}}
        assert results['NM_000083.2:c.689G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.143321841G>A', 'vcf': {'chr': '7', 'ref': 'G', 'pos': '143321841', 'alt': 'A'}}


    def test_variant307(self):
        variant = '7-143048771-C-T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NR_046453.1:n.2620C>T' in results.keys()
        assert results['NR_046453.1:n.2620C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_046453.1:n.2620C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_046453.1:n.2620C>T']['alt_genomic_loci'] == []
        assert results['NR_046453.1:n.2620C>T']['gene_symbol'] == 'CLCN1'
        assert results['NR_046453.1:n.2620C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_046453.1:n.2620C>T']['submitted_variant'] == '7-143048771-C-T'
        assert results['NR_046453.1:n.2620C>T']['genome_context_intronic_sequence'] == ''
        assert results['NR_046453.1:n.2620C>T']['hgvs_lrg_variant'] == ''
        assert results['NR_046453.1:n.2620C>T']['hgvs_transcript_variant'] == 'NR_046453.1:n.2620C>T'
        assert results['NR_046453.1:n.2620C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NR_046453.1:n.2620C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.143048771C>T', 'vcf': {'chr': 'chr7', 'ref': 'C', 'pos': '143048771', 'alt': 'T'}}
        assert results['NR_046453.1:n.2620C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.143351678C>T', 'vcf': {'chr': 'chr7', 'ref': 'C', 'pos': '143351678', 'alt': 'T'}}
        assert results['NR_046453.1:n.2620C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.143048771C>T', 'vcf': {'chr': '7', 'ref': 'C', 'pos': '143048771', 'alt': 'T'}}
        assert results['NR_046453.1:n.2620C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.143351678C>T', 'vcf': {'chr': '7', 'ref': 'C', 'pos': '143351678', 'alt': 'T'}}

        assert 'NM_000083.2:c.2680C>T' in results.keys()
        assert results['NM_000083.2:c.2680C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000083.2:c.2680C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000083.2:c.2680C>T']['alt_genomic_loci'] == []
        assert results['NM_000083.2:c.2680C>T']['gene_symbol'] == 'CLCN1'
        assert results['NM_000083.2:c.2680C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000074.2:p.(Arg894Ter)', 'slr': 'NP_000074.2:p.(R894*)'}
        assert results['NM_000083.2:c.2680C>T']['submitted_variant'] == '7-143048771-C-T'
        assert results['NM_000083.2:c.2680C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_000083.2:c.2680C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_000083.2:c.2680C>T']['hgvs_transcript_variant'] == 'NM_000083.2:c.2680C>T'
        assert results['NM_000083.2:c.2680C>T']['hgvs_refseqgene_variant'] == 'NG_009815.1:g.40553C>T'
        assert results['NM_000083.2:c.2680C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000007.13:g.143048771C>T', 'vcf': {'chr': 'chr7', 'ref': 'C', 'pos': '143048771', 'alt': 'T'}}
        assert results['NM_000083.2:c.2680C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000007.14:g.143351678C>T', 'vcf': {'chr': 'chr7', 'ref': 'C', 'pos': '143351678', 'alt': 'T'}}
        assert results['NM_000083.2:c.2680C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000007.13:g.143048771C>T', 'vcf': {'chr': '7', 'ref': 'C', 'pos': '143048771', 'alt': 'T'}}
        assert results['NM_000083.2:c.2680C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000007.14:g.143351678C>T', 'vcf': {'chr': '7', 'ref': 'C', 'pos': '143351678', 'alt': 'T'}}


    def test_variant308(self):
        variant = '8-1871951-C-T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_014629.2:c.2399C>T' in results.keys()
        assert results['NM_014629.2:c.2399C>T']['hgvs_lrg_transcript_variant'] == 'LRG_234t1:c.2399C>T'
        assert results['NM_014629.2:c.2399C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_014629.2:c.2399C>T']['alt_genomic_loci'] == []
        assert results['NM_014629.2:c.2399C>T']['gene_symbol'] == 'ARHGEF10'
        assert results['NM_014629.2:c.2399C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_055444.2(LRG_234p1):p.(Pro800Leu)', 'slr': 'NP_055444.2:p.(P800L)'}
        assert results['NM_014629.2:c.2399C>T']['submitted_variant'] == '8-1871951-C-T'
        assert results['NM_014629.2:c.2399C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_014629.2:c.2399C>T']['hgvs_lrg_variant'] == 'LRG_234:g.104803C>T'
        assert results['NM_014629.2:c.2399C>T']['hgvs_transcript_variant'] == 'NM_014629.2:c.2399C>T'
        assert results['NM_014629.2:c.2399C>T']['hgvs_refseqgene_variant'] == 'NG_008480.1:g.104803C>T'
        assert results['NM_014629.2:c.2399C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000008.10:g.1871951C>T', 'vcf': {'chr': 'chr8', 'ref': 'C', 'pos': '1871951', 'alt': 'T'}}
        assert 'hg38' not in results['NM_014629.2:c.2399C>T']['primary_assembly_loci'].keys()
        assert results['NM_014629.2:c.2399C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000008.10:g.1871951C>T', 'vcf': {'chr': '8', 'ref': 'C', 'pos': '1871951', 'alt': 'T'}}
        assert 'grch38' not in results['NM_014629.2:c.2399C>T']['primary_assembly_loci'].keys()

        assert 'NM_014629.3:c.2399C>T' in results.keys()
        assert results['NM_014629.3:c.2399C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_014629.3:c.2399C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_014629.3:c.2399C>T']['alt_genomic_loci'] == [{'grch38': {'hgvs_genomic_description': 'NT_187576.1:g.107161C>T', 'vcf': {'chr': 'HSCHR8_8_CTG1', 'ref': 'C', 'pos': '107161', 'alt': 'T'}}}, {'hg38': {'hgvs_genomic_description': 'NT_187576.1:g.107161C>T', 'vcf': {'chr': 'chr8_KI270821v1_alt', 'ref': 'C', 'pos': '107161', 'alt': 'T'}}}]
        assert results['NM_014629.3:c.2399C>T']['gene_symbol'] == 'ARHGEF10'
        assert results['NM_014629.3:c.2399C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_055444.2(LRG_234p1):p.(Pro800Leu)', 'slr': 'NP_055444.2:p.(P800L)'}
        assert results['NM_014629.3:c.2399C>T']['submitted_variant'] == '8-1871951-C-T'
        assert results['NM_014629.3:c.2399C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_014629.3:c.2399C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_014629.3:c.2399C>T']['hgvs_transcript_variant'] == 'NM_014629.3:c.2399C>T'
        assert results['NM_014629.3:c.2399C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_014629.3:c.2399C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000008.10:g.1871951C>T', 'vcf': {'chr': 'chr8', 'ref': 'C', 'pos': '1871951', 'alt': 'T'}}
        assert results['NM_014629.3:c.2399C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000008.11:g.1923785C>T', 'vcf': {'chr': 'chr8', 'ref': 'C', 'pos': '1923785', 'alt': 'T'}}
        assert results['NM_014629.3:c.2399C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000008.10:g.1871951C>T', 'vcf': {'chr': '8', 'ref': 'C', 'pos': '1871951', 'alt': 'T'}}
        assert results['NM_014629.3:c.2399C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000008.11:g.1923785C>T', 'vcf': {'chr': '8', 'ref': 'C', 'pos': '1923785', 'alt': 'T'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001308153.1:c.2471C>T' in results.keys()
        assert results['NM_001308153.1:c.2471C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001308153.1:c.2471C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001308153.1:c.2471C>T']['alt_genomic_loci'] == [{'grch38': {'hgvs_genomic_description': 'NT_187576.1:g.107161C>T', 'vcf': {'chr': 'HSCHR8_8_CTG1', 'ref': 'C', 'pos': '107161', 'alt': 'T'}}}, {'hg38': {'hgvs_genomic_description': 'NT_187576.1:g.107161C>T', 'vcf': {'chr': 'chr8_KI270821v1_alt', 'ref': 'C', 'pos': '107161', 'alt': 'T'}}}]
        assert results['NM_001308153.1:c.2471C>T']['gene_symbol'] == 'ARHGEF10'
        assert results['NM_001308153.1:c.2471C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001295082.1:p.(Pro824Leu)', 'slr': 'NP_001295082.1:p.(P824L)'}
        assert results['NM_001308153.1:c.2471C>T']['submitted_variant'] == '8-1871951-C-T'
        assert results['NM_001308153.1:c.2471C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001308153.1:c.2471C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001308153.1:c.2471C>T']['hgvs_transcript_variant'] == 'NM_001308153.1:c.2471C>T'
        assert results['NM_001308153.1:c.2471C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001308153.1:c.2471C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000008.10:g.1871951C>T', 'vcf': {'chr': 'chr8', 'ref': 'C', 'pos': '1871951', 'alt': 'T'}}
        assert results['NM_001308153.1:c.2471C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000008.11:g.1923785C>T', 'vcf': {'chr': 'chr8', 'ref': 'C', 'pos': '1923785', 'alt': 'T'}}
        assert results['NM_001308153.1:c.2471C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000008.10:g.1871951C>T', 'vcf': {'chr': '8', 'ref': 'C', 'pos': '1871951', 'alt': 'T'}}
        assert results['NM_001308153.1:c.2471C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000008.11:g.1923785C>T', 'vcf': {'chr': '8', 'ref': 'C', 'pos': '1923785', 'alt': 'T'}}

        assert 'NM_001308152.1:c.2285C>T' in results.keys()
        assert results['NM_001308152.1:c.2285C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001308152.1:c.2285C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001308152.1:c.2285C>T']['alt_genomic_loci'] == [{'grch38': {'hgvs_genomic_description': 'NT_187576.1:g.107161C>T', 'vcf': {'chr': 'HSCHR8_8_CTG1', 'ref': 'C', 'pos': '107161', 'alt': 'T'}}}, {'hg38': {'hgvs_genomic_description': 'NT_187576.1:g.107161C>T', 'vcf': {'chr': 'chr8_KI270821v1_alt', 'ref': 'C', 'pos': '107161', 'alt': 'T'}}}]
        assert results['NM_001308152.1:c.2285C>T']['gene_symbol'] == 'ARHGEF10'
        assert results['NM_001308152.1:c.2285C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001295081.1:p.(Pro762Leu)', 'slr': 'NP_001295081.1:p.(P762L)'}
        assert results['NM_001308152.1:c.2285C>T']['submitted_variant'] == '8-1871951-C-T'
        assert results['NM_001308152.1:c.2285C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001308152.1:c.2285C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001308152.1:c.2285C>T']['hgvs_transcript_variant'] == 'NM_001308152.1:c.2285C>T'
        assert results['NM_001308152.1:c.2285C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001308152.1:c.2285C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000008.10:g.1871951C>T', 'vcf': {'chr': 'chr8', 'ref': 'C', 'pos': '1871951', 'alt': 'T'}}
        assert results['NM_001308152.1:c.2285C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000008.11:g.1923785C>T', 'vcf': {'chr': 'chr8', 'ref': 'C', 'pos': '1923785', 'alt': 'T'}}
        assert results['NM_001308152.1:c.2285C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000008.10:g.1871951C>T', 'vcf': {'chr': '8', 'ref': 'C', 'pos': '1871951', 'alt': 'T'}}
        assert results['NM_001308152.1:c.2285C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000008.11:g.1923785C>T', 'vcf': {'chr': '8', 'ref': 'C', 'pos': '1923785', 'alt': 'T'}}


    def test_variant309(self):
        variant = '9-13112056-T-TG'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_001261406.1:c.5591dup' in results.keys()
        assert results['NM_001261406.1:c.5591dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001261406.1:c.5591dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001261406.1:c.5591dup']['alt_genomic_loci'] == []
        assert results['NM_001261406.1:c.5591dup']['gene_symbol'] == 'MPDZ'
        assert results['NM_001261406.1:c.5591dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001248335.1:p.(Thr1865AsnfsTer15)', 'slr': 'NP_001248335.1:p.(T1865Nfs*15)'}
        assert results['NM_001261406.1:c.5591dup']['submitted_variant'] == '9-13112056-T-TG'
        assert results['NM_001261406.1:c.5591dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_001261406.1:c.5591dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001261406.1:c.5591dup']['hgvs_transcript_variant'] == 'NM_001261406.1:c.5591dup'
        assert results['NM_001261406.1:c.5591dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001261406.1:c.5591dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.13112057dup', 'vcf': {'chr': 'chr9', 'ref': 'G', 'pos': '13112057', 'alt': u'GG'}}
        assert results['NM_001261406.1:c.5591dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.13112058dup', 'vcf': {'chr': 'chr9', 'ref': 'G', 'pos': '13112058', 'alt': u'GG'}}
        assert results['NM_001261406.1:c.5591dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.13112057dup', 'vcf': {'chr': '9', 'ref': 'G', 'pos': '13112057', 'alt': u'GG'}}
        assert results['NM_001261406.1:c.5591dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.13112058dup', 'vcf': {'chr': '9', 'ref': 'G', 'pos': '13112058', 'alt': u'GG'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001330637.1:c.5690dup' in results.keys()
        assert results['NM_001330637.1:c.5690dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001330637.1:c.5690dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001330637.1:c.5690dup']['alt_genomic_loci'] == []
        assert results['NM_001330637.1:c.5690dup']['gene_symbol'] == 'MPDZ'
        assert results['NM_001330637.1:c.5690dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001317566.1:p.(Thr1898AsnfsTer15)', 'slr': 'NP_001317566.1:p.(T1898Nfs*15)'}
        assert results['NM_001330637.1:c.5690dup']['submitted_variant'] == '9-13112056-T-TG'
        assert results['NM_001330637.1:c.5690dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_001330637.1:c.5690dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001330637.1:c.5690dup']['hgvs_transcript_variant'] == 'NM_001330637.1:c.5690dup'
        assert results['NM_001330637.1:c.5690dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001330637.1:c.5690dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.13112057dup', 'vcf': {'chr': 'chr9', 'ref': 'G', 'pos': '13112057', 'alt': u'GG'}}
        assert results['NM_001330637.1:c.5690dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.13112058dup', 'vcf': {'chr': 'chr9', 'ref': 'G', 'pos': '13112058', 'alt': u'GG'}}
        assert results['NM_001330637.1:c.5690dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.13112057dup', 'vcf': {'chr': '9', 'ref': 'G', 'pos': '13112057', 'alt': u'GG'}}
        assert results['NM_001330637.1:c.5690dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.13112058dup', 'vcf': {'chr': '9', 'ref': 'G', 'pos': '13112058', 'alt': u'GG'}}

        assert 'NM_001261407.1:c.5504dup' in results.keys()
        assert results['NM_001261407.1:c.5504dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001261407.1:c.5504dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001261407.1:c.5504dup']['alt_genomic_loci'] == []
        assert results['NM_001261407.1:c.5504dup']['gene_symbol'] == 'MPDZ'
        assert results['NM_001261407.1:c.5504dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001248336.1:p.(Thr1836AsnfsTer15)', 'slr': 'NP_001248336.1:p.(T1836Nfs*15)'}
        assert results['NM_001261407.1:c.5504dup']['submitted_variant'] == '9-13112056-T-TG'
        assert results['NM_001261407.1:c.5504dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_001261407.1:c.5504dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001261407.1:c.5504dup']['hgvs_transcript_variant'] == 'NM_001261407.1:c.5504dup'
        assert results['NM_001261407.1:c.5504dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001261407.1:c.5504dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.13112057dup', 'vcf': {'chr': 'chr9', 'ref': 'G', 'pos': '13112057', 'alt': u'GG'}}
        assert results['NM_001261407.1:c.5504dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.13112058dup', 'vcf': {'chr': 'chr9', 'ref': 'G', 'pos': '13112058', 'alt': u'GG'}}
        assert results['NM_001261407.1:c.5504dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.13112057dup', 'vcf': {'chr': '9', 'ref': 'G', 'pos': '13112057', 'alt': u'GG'}}
        assert results['NM_001261407.1:c.5504dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.13112058dup', 'vcf': {'chr': '9', 'ref': 'G', 'pos': '13112058', 'alt': u'GG'}}

        assert 'NM_003829.4:c.5603dup' in results.keys()
        assert results['NM_003829.4:c.5603dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_003829.4:c.5603dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_003829.4:c.5603dup']['alt_genomic_loci'] == []
        assert results['NM_003829.4:c.5603dup']['gene_symbol'] == 'MPDZ'
        assert results['NM_003829.4:c.5603dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003820.2:p.(Thr1869AsnfsTer15)', 'slr': 'NP_003820.2:p.(T1869Nfs*15)'}
        assert results['NM_003829.4:c.5603dup']['submitted_variant'] == '9-13112056-T-TG'
        assert results['NM_003829.4:c.5603dup']['genome_context_intronic_sequence'] == ''
        assert results['NM_003829.4:c.5603dup']['hgvs_lrg_variant'] == ''
        assert results['NM_003829.4:c.5603dup']['hgvs_transcript_variant'] == 'NM_003829.4:c.5603dup'
        assert results['NM_003829.4:c.5603dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_003829.4:c.5603dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.13112057dup', 'vcf': {'chr': 'chr9', 'ref': 'G', 'pos': '13112057', 'alt': u'GG'}}
        assert results['NM_003829.4:c.5603dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.13112058dup', 'vcf': {'chr': 'chr9', 'ref': 'G', 'pos': '13112058', 'alt': u'GG'}}
        assert results['NM_003829.4:c.5603dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.13112057dup', 'vcf': {'chr': '9', 'ref': 'G', 'pos': '13112057', 'alt': u'GG'}}
        assert results['NM_003829.4:c.5603dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.13112058dup', 'vcf': {'chr': '9', 'ref': 'G', 'pos': '13112058', 'alt': u'GG'}}


    def test_variant310(self):
        variant = '9-21971208-C-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_058197.4:c.*74-1G>T' in results.keys()
        assert results['NM_058197.4:c.*74-1G>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_058197.4:c.*74-1G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_058197.4:c.*74-1G>T']['alt_genomic_loci'] == []
        assert results['NM_058197.4:c.*74-1G>T']['gene_symbol'] == 'CDKN2A'
        assert results['NM_058197.4:c.*74-1G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_478104.2:p.?', 'slr': 'NP_478104.2:p.?'}
        assert results['NM_058197.4:c.*74-1G>T']['submitted_variant'] == '9-21971208-C-A'
        assert results['NM_058197.4:c.*74-1G>T']['genome_context_intronic_sequence'] == 'NC_000009.11(NM_058197.4):c.*74-1G>T'
        assert results['NM_058197.4:c.*74-1G>T']['hgvs_lrg_variant'] == ''
        assert results['NM_058197.4:c.*74-1G>T']['hgvs_transcript_variant'] == 'NM_058197.4:c.*74-1G>T'
        assert results['NM_058197.4:c.*74-1G>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_058197.4:c.*74-1G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.21971208C>A', 'vcf': {'chr': 'chr9', 'ref': u'C', 'pos': '21971208', 'alt': u'A'}}
        assert results['NM_058197.4:c.*74-1G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.21971209C>A', 'vcf': {'chr': 'chr9', 'ref': u'C', 'pos': '21971209', 'alt': u'A'}}
        assert results['NM_058197.4:c.*74-1G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.21971208C>A', 'vcf': {'chr': '9', 'ref': u'C', 'pos': '21971208', 'alt': u'A'}}
        assert results['NM_058197.4:c.*74-1G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.21971209C>A', 'vcf': {'chr': '9', 'ref': u'C', 'pos': '21971209', 'alt': u'A'}}

        assert 'NM_000077.4:c.151-1G>T' in results.keys()
        assert results['NM_000077.4:c.151-1G>T']['hgvs_lrg_transcript_variant'] == 'LRG_11t1:c.151-1G>T'
        assert results['NM_000077.4:c.151-1G>T']['refseqgene_context_intronic_sequence'] == 'NG_007485.1(NM_000077.4):c.151-1G>T'
        assert results['NM_000077.4:c.151-1G>T']['alt_genomic_loci'] == []
        assert results['NM_000077.4:c.151-1G>T']['gene_symbol'] == 'CDKN2A'
        assert results['NM_000077.4:c.151-1G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000068.1(LRG_11p1):p.?', 'slr': 'NP_000068.1:p.?'}
        assert results['NM_000077.4:c.151-1G>T']['submitted_variant'] == '9-21971208-C-A'
        assert results['NM_000077.4:c.151-1G>T']['genome_context_intronic_sequence'] == 'NC_000009.11(NM_000077.4):c.151-1G>T'
        assert results['NM_000077.4:c.151-1G>T']['hgvs_lrg_variant'] == 'LRG_11:g.28283G>T'
        assert results['NM_000077.4:c.151-1G>T']['hgvs_transcript_variant'] == 'NM_000077.4:c.151-1G>T'
        assert results['NM_000077.4:c.151-1G>T']['hgvs_refseqgene_variant'] == 'NG_007485.1:g.28283G>T'
        assert results['NM_000077.4:c.151-1G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.21971208C>A', 'vcf': {'chr': 'chr9', 'ref': u'C', 'pos': '21971208', 'alt': u'A'}}
        assert results['NM_000077.4:c.151-1G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.21971209C>A', 'vcf': {'chr': 'chr9', 'ref': u'C', 'pos': '21971209', 'alt': u'A'}}
        assert results['NM_000077.4:c.151-1G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.21971208C>A', 'vcf': {'chr': '9', 'ref': u'C', 'pos': '21971208', 'alt': u'A'}}
        assert results['NM_000077.4:c.151-1G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.21971209C>A', 'vcf': {'chr': '9', 'ref': u'C', 'pos': '21971209', 'alt': u'A'}}

        assert 'NM_001363763.1:c.-3-1G>T' in results.keys()
        assert results['NM_001363763.1:c.-3-1G>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001363763.1:c.-3-1G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001363763.1:c.-3-1G>T']['alt_genomic_loci'] == []
        assert results['NM_001363763.1:c.-3-1G>T']['gene_symbol'] == 'CDKN2A'
        assert results['NM_001363763.1:c.-3-1G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001350692.1:p.?', 'slr': 'NP_001350692.1:p.?'}
        assert results['NM_001363763.1:c.-3-1G>T']['submitted_variant'] == '9-21971208-C-A'
        assert results['NM_001363763.1:c.-3-1G>T']['genome_context_intronic_sequence'] == 'NC_000009.11(NM_001363763.1):c.-3-1G>T'
        assert results['NM_001363763.1:c.-3-1G>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001363763.1:c.-3-1G>T']['hgvs_transcript_variant'] == 'NM_001363763.1:c.-3-1G>T'
        assert results['NM_001363763.1:c.-3-1G>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001363763.1:c.-3-1G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.21971208C>A', 'vcf': {'chr': 'chr9', 'ref': u'C', 'pos': '21971208', 'alt': u'A'}}
        assert 'hg38' not in results['NM_001363763.1:c.-3-1G>T']['primary_assembly_loci'].keys()
        assert results['NM_001363763.1:c.-3-1G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.21971208C>A', 'vcf': {'chr': '9', 'ref': u'C', 'pos': '21971208', 'alt': u'A'}}
        assert 'grch38' not in results['NM_001363763.1:c.-3-1G>T']['primary_assembly_loci'].keys()

        assert 'NM_001195132.1:c.151-1G>T' in results.keys()
        assert results['NM_001195132.1:c.151-1G>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001195132.1:c.151-1G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001195132.1:c.151-1G>T']['alt_genomic_loci'] == []
        assert results['NM_001195132.1:c.151-1G>T']['gene_symbol'] == 'CDKN2A'
        assert results['NM_001195132.1:c.151-1G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001182061.1:p.?', 'slr': 'NP_001182061.1:p.?'}
        assert results['NM_001195132.1:c.151-1G>T']['submitted_variant'] == '9-21971208-C-A'
        assert results['NM_001195132.1:c.151-1G>T']['genome_context_intronic_sequence'] == 'NC_000009.11(NM_001195132.1):c.151-1G>T'
        assert results['NM_001195132.1:c.151-1G>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001195132.1:c.151-1G>T']['hgvs_transcript_variant'] == 'NM_001195132.1:c.151-1G>T'
        assert results['NM_001195132.1:c.151-1G>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001195132.1:c.151-1G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.21971208C>A', 'vcf': {'chr': 'chr9', 'ref': u'C', 'pos': '21971208', 'alt': u'A'}}
        assert results['NM_001195132.1:c.151-1G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.21971209C>A', 'vcf': {'chr': 'chr9', 'ref': u'C', 'pos': '21971209', 'alt': u'A'}}
        assert results['NM_001195132.1:c.151-1G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.21971208C>A', 'vcf': {'chr': '9', 'ref': u'C', 'pos': '21971208', 'alt': u'A'}}
        assert results['NM_001195132.1:c.151-1G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.21971209C>A', 'vcf': {'chr': '9', 'ref': u'C', 'pos': '21971209', 'alt': u'A'}}

        assert 'NM_058195.3:c.194-1G>T' in results.keys()
        assert results['NM_058195.3:c.194-1G>T']['hgvs_lrg_transcript_variant'] == 'LRG_11t2:c.194-1G>T'
        assert results['NM_058195.3:c.194-1G>T']['refseqgene_context_intronic_sequence'] == 'NG_007485.1(NM_058195.3):c.194-1G>T'
        assert results['NM_058195.3:c.194-1G>T']['alt_genomic_loci'] == []
        assert results['NM_058195.3:c.194-1G>T']['gene_symbol'] == 'CDKN2A'
        assert results['NM_058195.3:c.194-1G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_478102.2(LRG_11p2):p.?', 'slr': 'NP_478102.2:p.?'}
        assert results['NM_058195.3:c.194-1G>T']['submitted_variant'] == '9-21971208-C-A'
        assert results['NM_058195.3:c.194-1G>T']['genome_context_intronic_sequence'] == 'NC_000009.11(NM_058195.3):c.194-1G>T'
        assert results['NM_058195.3:c.194-1G>T']['hgvs_lrg_variant'] == 'LRG_11:g.28283G>T'
        assert results['NM_058195.3:c.194-1G>T']['hgvs_transcript_variant'] == 'NM_058195.3:c.194-1G>T'
        assert results['NM_058195.3:c.194-1G>T']['hgvs_refseqgene_variant'] == 'NG_007485.1:g.28283G>T'
        assert results['NM_058195.3:c.194-1G>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.21971208C>A', 'vcf': {'chr': 'chr9', 'ref': u'C', 'pos': '21971208', 'alt': u'A'}}
        assert results['NM_058195.3:c.194-1G>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.21971209C>A', 'vcf': {'chr': 'chr9', 'ref': u'C', 'pos': '21971209', 'alt': u'A'}}
        assert results['NM_058195.3:c.194-1G>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.21971208C>A', 'vcf': {'chr': '9', 'ref': u'C', 'pos': '21971208', 'alt': u'A'}}
        assert results['NM_058195.3:c.194-1G>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.21971209C>A', 'vcf': {'chr': '9', 'ref': u'C', 'pos': '21971209', 'alt': u'A'}}

        assert results['flag'] == 'gene_variant'

    def test_variant311(self):
        variant = '9-35683240-T-TG'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_003289.3:c.773-3dup' in results.keys()
        assert results['NM_003289.3:c.773-3dup']['hgvs_lrg_transcript_variant'] == 'LRG_680t2:c.773-3dup'
        assert results['NM_003289.3:c.773-3dup']['refseqgene_context_intronic_sequence'] == 'NG_011620.1(NM_003289.3):c.773-3dup'
        assert results['NM_003289.3:c.773-3dup']['alt_genomic_loci'] == []
        assert results['NM_003289.3:c.773-3dup']['gene_symbol'] == 'TPM2'
        assert results['NM_003289.3:c.773-3dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_003280.2(LRG_680p2):p.?', 'slr': 'NP_003280.2:p.?'}
        assert results['NM_003289.3:c.773-3dup']['submitted_variant'] == '9-35683240-T-TG'
        assert results['NM_003289.3:c.773-3dup']['genome_context_intronic_sequence'] == 'NC_000009.11(NM_003289.3):c.773-3dup'
        assert results['NM_003289.3:c.773-3dup']['hgvs_lrg_variant'] == 'LRG_680:g.11814dup'
        assert results['NM_003289.3:c.773-3dup']['hgvs_transcript_variant'] == 'NM_003289.3:c.773-3dup'
        assert results['NM_003289.3:c.773-3dup']['hgvs_refseqgene_variant'] == 'NG_011620.1:g.11814dup'
        assert results['NM_003289.3:c.773-3dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.35683241dup', 'vcf': {'chr': 'chr9', 'ref': 'G', 'pos': '35683241', 'alt': 'GG'}}
        assert results['NM_003289.3:c.773-3dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.35683244dup', 'vcf': {'chr': 'chr9', 'ref': 'G', 'pos': '35683244', 'alt': 'GG'}}
        assert results['NM_003289.3:c.773-3dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.35683241dup', 'vcf': {'chr': '9', 'ref': 'G', 'pos': '35683241', 'alt': 'GG'}}
        assert results['NM_003289.3:c.773-3dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.35683244dup', 'vcf': {'chr': '9', 'ref': 'G', 'pos': '35683244', 'alt': 'GG'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_213674.1:c.772+1002dup' in results.keys()
        assert results['NM_213674.1:c.772+1002dup']['hgvs_lrg_transcript_variant'] == 'LRG_680t1:c.772+1002dup'
        assert results['NM_213674.1:c.772+1002dup']['refseqgene_context_intronic_sequence'] == 'NG_011620.1(NM_213674.1):c.772+1002dup'
        assert results['NM_213674.1:c.772+1002dup']['alt_genomic_loci'] == []
        assert results['NM_213674.1:c.772+1002dup']['gene_symbol'] == 'TPM2'
        assert results['NM_213674.1:c.772+1002dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_998839.1(LRG_680p1):p.?', 'slr': 'NP_998839.1:p.?'}
        assert results['NM_213674.1:c.772+1002dup']['submitted_variant'] == '9-35683240-T-TG'
        assert results['NM_213674.1:c.772+1002dup']['genome_context_intronic_sequence'] == 'NC_000009.11(NM_213674.1):c.772+1002dup'
        assert results['NM_213674.1:c.772+1002dup']['hgvs_lrg_variant'] == 'LRG_680:g.11814dup'
        assert results['NM_213674.1:c.772+1002dup']['hgvs_transcript_variant'] == 'NM_213674.1:c.772+1002dup'
        assert results['NM_213674.1:c.772+1002dup']['hgvs_refseqgene_variant'] == 'NG_011620.1:g.11814dup'
        assert results['NM_213674.1:c.772+1002dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.35683241dup', 'vcf': {'chr': 'chr9', 'ref': 'G', 'pos': '35683241', 'alt': 'GG'}}
        assert results['NM_213674.1:c.772+1002dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.35683244dup', 'vcf': {'chr': 'chr9', 'ref': 'G', 'pos': '35683244', 'alt': 'GG'}}
        assert results['NM_213674.1:c.772+1002dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.35683241dup', 'vcf': {'chr': '9', 'ref': 'G', 'pos': '35683241', 'alt': 'GG'}}
        assert results['NM_213674.1:c.772+1002dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.35683244dup', 'vcf': {'chr': '9', 'ref': 'G', 'pos': '35683244', 'alt': 'GG'}}

        assert 'NM_001301226.1:c.772+1002dup' in results.keys()
        assert results['NM_001301226.1:c.772+1002dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001301226.1:c.772+1002dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001301226.1:c.772+1002dup']['alt_genomic_loci'] == []
        assert results['NM_001301226.1:c.772+1002dup']['gene_symbol'] == 'TPM2'
        assert results['NM_001301226.1:c.772+1002dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001288155.1:p.?', 'slr': 'NP_001288155.1:p.?'}
        assert results['NM_001301226.1:c.772+1002dup']['submitted_variant'] == '9-35683240-T-TG'
        assert results['NM_001301226.1:c.772+1002dup']['genome_context_intronic_sequence'] == 'NC_000009.11(NM_001301226.1):c.772+1002dup'
        assert results['NM_001301226.1:c.772+1002dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001301226.1:c.772+1002dup']['hgvs_transcript_variant'] == 'NM_001301226.1:c.772+1002dup'
        assert results['NM_001301226.1:c.772+1002dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001301226.1:c.772+1002dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.35683241dup', 'vcf': {'chr': 'chr9', 'ref': 'G', 'pos': '35683241', 'alt': 'GG'}}
        assert results['NM_001301226.1:c.772+1002dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.35683244dup', 'vcf': {'chr': 'chr9', 'ref': 'G', 'pos': '35683244', 'alt': 'GG'}}
        assert results['NM_001301226.1:c.772+1002dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.35683241dup', 'vcf': {'chr': '9', 'ref': 'G', 'pos': '35683241', 'alt': 'GG'}}
        assert results['NM_001301226.1:c.772+1002dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.35683244dup', 'vcf': {'chr': '9', 'ref': 'G', 'pos': '35683244', 'alt': 'GG'}}

        assert 'NM_001301227.1:c.773-3dup' in results.keys()
        assert results['NM_001301227.1:c.773-3dup']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001301227.1:c.773-3dup']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001301227.1:c.773-3dup']['alt_genomic_loci'] == []
        assert results['NM_001301227.1:c.773-3dup']['gene_symbol'] == 'TPM2'
        assert results['NM_001301227.1:c.773-3dup']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001288156.1:p.?', 'slr': 'NP_001288156.1:p.?'}
        assert results['NM_001301227.1:c.773-3dup']['submitted_variant'] == '9-35683240-T-TG'
        assert results['NM_001301227.1:c.773-3dup']['genome_context_intronic_sequence'] == 'NC_000009.11(NM_001301227.1):c.773-3dup'
        assert results['NM_001301227.1:c.773-3dup']['hgvs_lrg_variant'] == ''
        assert results['NM_001301227.1:c.773-3dup']['hgvs_transcript_variant'] == 'NM_001301227.1:c.773-3dup'
        assert results['NM_001301227.1:c.773-3dup']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001301227.1:c.773-3dup']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.35683241dup', 'vcf': {'chr': 'chr9', 'ref': 'G', 'pos': '35683241', 'alt': 'GG'}}
        assert results['NM_001301227.1:c.773-3dup']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.35683244dup', 'vcf': {'chr': 'chr9', 'ref': 'G', 'pos': '35683244', 'alt': 'GG'}}
        assert results['NM_001301227.1:c.773-3dup']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.35683241dup', 'vcf': {'chr': '9', 'ref': 'G', 'pos': '35683241', 'alt': 'GG'}}
        assert results['NM_001301227.1:c.773-3dup']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.35683244dup', 'vcf': {'chr': '9', 'ref': 'G', 'pos': '35683244', 'alt': 'GG'}}


    def test_variant312(self):
        variant = '9-135796754-G-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_000368.4:c.733C>T' in results.keys()
        assert results['NM_000368.4:c.733C>T']['hgvs_lrg_transcript_variant'] == 'LRG_486t1:c.733C>T'
        assert results['NM_000368.4:c.733C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000368.4:c.733C>T']['alt_genomic_loci'] == []
        assert results['NM_000368.4:c.733C>T']['gene_symbol'] == 'TSC1'
        assert results['NM_000368.4:c.733C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000359.1(LRG_486p1):p.(Arg245Ter)', 'slr': 'NP_000359.1:p.(R245*)'}
        assert results['NM_000368.4:c.733C>T']['submitted_variant'] == '9-135796754-G-A'
        assert results['NM_000368.4:c.733C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_000368.4:c.733C>T']['hgvs_lrg_variant'] == 'LRG_486:g.28267C>T'
        assert results['NM_000368.4:c.733C>T']['hgvs_transcript_variant'] == 'NM_000368.4:c.733C>T'
        assert results['NM_000368.4:c.733C>T']['hgvs_refseqgene_variant'] == 'NG_012386.1:g.28267C>T'
        assert results['NM_000368.4:c.733C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.135796754G>A', 'vcf': {'chr': 'chr9', 'ref': u'G', 'pos': '135796754', 'alt': u'A'}}
        assert results['NM_000368.4:c.733C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.132921367G>A', 'vcf': {'chr': 'chr9', 'ref': u'G', 'pos': '132921367', 'alt': u'A'}}
        assert results['NM_000368.4:c.733C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.135796754G>A', 'vcf': {'chr': '9', 'ref': u'G', 'pos': '135796754', 'alt': u'A'}}
        assert results['NM_000368.4:c.733C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.132921367G>A', 'vcf': {'chr': '9', 'ref': u'G', 'pos': '132921367', 'alt': u'A'}}

        assert 'NM_001162426.1:c.733C>T' in results.keys()
        assert results['NM_001162426.1:c.733C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001162426.1:c.733C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001162426.1:c.733C>T']['alt_genomic_loci'] == []
        assert results['NM_001162426.1:c.733C>T']['gene_symbol'] == 'TSC1'
        assert results['NM_001162426.1:c.733C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001155898.1:p.(Arg245Ter)', 'slr': 'NP_001155898.1:p.(R245*)'}
        assert results['NM_001162426.1:c.733C>T']['submitted_variant'] == '9-135796754-G-A'
        assert results['NM_001162426.1:c.733C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001162426.1:c.733C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001162426.1:c.733C>T']['hgvs_transcript_variant'] == 'NM_001162426.1:c.733C>T'
        assert results['NM_001162426.1:c.733C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001162426.1:c.733C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.135796754G>A', 'vcf': {'chr': 'chr9', 'ref': u'G', 'pos': '135796754', 'alt': u'A'}}
        assert results['NM_001162426.1:c.733C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.132921367G>A', 'vcf': {'chr': 'chr9', 'ref': u'G', 'pos': '132921367', 'alt': u'A'}}
        assert results['NM_001162426.1:c.733C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.135796754G>A', 'vcf': {'chr': '9', 'ref': u'G', 'pos': '135796754', 'alt': u'A'}}
        assert results['NM_001162426.1:c.733C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.132921367G>A', 'vcf': {'chr': '9', 'ref': u'G', 'pos': '132921367', 'alt': u'A'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001362177.1:c.370C>T' in results.keys()
        assert results['NM_001362177.1:c.370C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001362177.1:c.370C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001362177.1:c.370C>T']['alt_genomic_loci'] == []
        assert results['NM_001362177.1:c.370C>T']['gene_symbol'] == 'TSC1'
        assert results['NM_001362177.1:c.370C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001349106.1:p.(Arg124Ter)', 'slr': 'NP_001349106.1:p.(R124*)'}
        assert results['NM_001362177.1:c.370C>T']['submitted_variant'] == '9-135796754-G-A'
        assert results['NM_001362177.1:c.370C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001362177.1:c.370C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001362177.1:c.370C>T']['hgvs_transcript_variant'] == 'NM_001362177.1:c.370C>T'
        assert results['NM_001362177.1:c.370C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001362177.1:c.370C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.135796754G>A', 'vcf': {'chr': 'chr9', 'ref': u'G', 'pos': '135796754', 'alt': u'A'}}
        assert results['NM_001362177.1:c.370C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.132921367G>A', 'vcf': {'chr': 'chr9', 'ref': u'G', 'pos': '132921367', 'alt': u'A'}}
        assert results['NM_001362177.1:c.370C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.135796754G>A', 'vcf': {'chr': '9', 'ref': u'G', 'pos': '135796754', 'alt': u'A'}}
        assert results['NM_001362177.1:c.370C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.132921367G>A', 'vcf': {'chr': '9', 'ref': u'G', 'pos': '132921367', 'alt': u'A'}}

        assert 'NM_001162427.1:c.580C>T' in results.keys()
        assert results['NM_001162427.1:c.580C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001162427.1:c.580C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001162427.1:c.580C>T']['alt_genomic_loci'] == []
        assert results['NM_001162427.1:c.580C>T']['gene_symbol'] == 'TSC1'
        assert results['NM_001162427.1:c.580C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001155899.1:p.(Arg194Ter)', 'slr': 'NP_001155899.1:p.(R194*)'}
        assert results['NM_001162427.1:c.580C>T']['submitted_variant'] == '9-135796754-G-A'
        assert results['NM_001162427.1:c.580C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001162427.1:c.580C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001162427.1:c.580C>T']['hgvs_transcript_variant'] == 'NM_001162427.1:c.580C>T'
        assert results['NM_001162427.1:c.580C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001162427.1:c.580C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.135796754G>A', 'vcf': {'chr': 'chr9', 'ref': u'G', 'pos': '135796754', 'alt': u'A'}}
        assert results['NM_001162427.1:c.580C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.132921367G>A', 'vcf': {'chr': 'chr9', 'ref': u'G', 'pos': '132921367', 'alt': u'A'}}
        assert results['NM_001162427.1:c.580C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.135796754G>A', 'vcf': {'chr': '9', 'ref': u'G', 'pos': '135796754', 'alt': u'A'}}
        assert results['NM_001162427.1:c.580C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.132921367G>A', 'vcf': {'chr': '9', 'ref': u'G', 'pos': '132921367', 'alt': u'A'}}


    def test_variant313(self):
        variant = 'HG536_PATCH-10391-AC-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_005247.2:c.616del' in results.keys()
        assert results['NM_005247.2:c.616del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_005247.2:c.616del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_005247.2:c.616del']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003571046.1:g.10392del', 'vcf': {'chr': 'HG536_PATCH', 'ref': 'AC', 'pos': '10391', 'alt': 'A'}}}]
        assert results['NM_005247.2:c.616del']['gene_symbol'] == 'FGF3'
        assert results['NM_005247.2:c.616del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_005238.1:p.(Val206SerfsTer117)', 'slr': 'NP_005238.1:p.(V206Sfs*117)'}
        assert results['NM_005247.2:c.616del']['submitted_variant'] == 'HG536_PATCH-10391-AC-A'
        assert results['NM_005247.2:c.616del']['genome_context_intronic_sequence'] == ''
        assert results['NM_005247.2:c.616del']['hgvs_lrg_variant'] == ''
        assert results['NM_005247.2:c.616del']['hgvs_transcript_variant'] == 'NM_005247.2:c.616del'
        assert results['NM_005247.2:c.616del']['hgvs_refseqgene_variant'] == 'NG_009016.1:g.14016del'
        assert results['NM_005247.2:c.616del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.69625177del', 'vcf': {'chr': 'chr11', 'ref': 'AC', 'pos': '69625176', 'alt': 'A'}}
        assert results['NM_005247.2:c.616del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.69810409del', 'vcf': {'chr': 'chr11', 'ref': 'AC', 'pos': '69810408', 'alt': 'A'}}
        assert results['NM_005247.2:c.616del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.69625177del', 'vcf': {'chr': '11', 'ref': 'AC', 'pos': '69625176', 'alt': 'A'}}
        assert results['NM_005247.2:c.616del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.69810409del', 'vcf': {'chr': '11', 'ref': 'AC', 'pos': '69810408', 'alt': 'A'}}


    def test_variant314(self):
        variant = 'HG865_PATCH-33547-G-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_133266.4:c.802C>T' in results.keys()
        assert results['NM_133266.4:c.802C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_133266.4:c.802C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_133266.4:c.802C>T']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_004070871.1:g.33547G>A', 'vcf': {'chr': 'HG865_PATCH', 'ref': u'G', 'pos': '33547', 'alt': u'A'}}}]
        assert results['NM_133266.4:c.802C>T']['gene_symbol'] == 'SHANK2'
        assert results['NM_133266.4:c.802C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_573573.2:p.(Leu268=)', 'slr': 'NP_573573.2:p.(L268=)'}
        assert results['NM_133266.4:c.802C>T']['submitted_variant'] == 'HG865_PATCH-33547-G-A'
        assert results['NM_133266.4:c.802C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_133266.4:c.802C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_133266.4:c.802C>T']['hgvs_transcript_variant'] == 'NM_133266.4:c.802C>T'
        assert results['NM_133266.4:c.802C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_133266.4:c.802C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.70335439G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '70335439', 'alt': u'A'}}
        assert results['NM_133266.4:c.802C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.70489334G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '70489334', 'alt': u'A'}}
        assert results['NM_133266.4:c.802C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.70335439G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '70335439', 'alt': u'A'}}
        assert results['NM_133266.4:c.802C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.70489334G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '70489334', 'alt': u'A'}}

        assert 'NR_110766.1:n.833+969C>T' in results.keys()
        assert results['NR_110766.1:n.833+969C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NR_110766.1:n.833+969C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NR_110766.1:n.833+969C>T']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_004070871.1:g.33547G>A', 'vcf': {'chr': 'HG865_PATCH', 'ref': u'G', 'pos': '33547', 'alt': u'A'}}}]
        assert results['NR_110766.1:n.833+969C>T']['gene_symbol'] == 'SHANK2'
        assert results['NR_110766.1:n.833+969C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'Non-coding :n.', 'slr': ''}
        assert results['NR_110766.1:n.833+969C>T']['submitted_variant'] == 'HG865_PATCH-33547-G-A'
        assert results['NR_110766.1:n.833+969C>T']['genome_context_intronic_sequence'] == 'NC_000011.9(NR_110766.1):c.833+969C>T'
        assert results['NR_110766.1:n.833+969C>T']['hgvs_lrg_variant'] == ''
        assert results['NR_110766.1:n.833+969C>T']['hgvs_transcript_variant'] == 'NR_110766.1:n.833+969C>T'
        assert results['NR_110766.1:n.833+969C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NR_110766.1:n.833+969C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.70335439G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '70335439', 'alt': u'A'}}
        assert results['NR_110766.1:n.833+969C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.70489334G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '70489334', 'alt': u'A'}}
        assert results['NR_110766.1:n.833+969C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.70335439G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '70335439', 'alt': u'A'}}
        assert results['NR_110766.1:n.833+969C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.70489334G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '70489334', 'alt': u'A'}}

        assert 'NM_133266.3:c.802C>T' in results.keys()
        assert results['NM_133266.3:c.802C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_133266.3:c.802C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_133266.3:c.802C>T']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_004070871.1:g.33547G>A', 'vcf': {'chr': 'HG865_PATCH', 'ref': u'G', 'pos': '33547', 'alt': u'A'}}}]
        assert results['NM_133266.3:c.802C>T']['gene_symbol'] == 'SHANK2'
        assert results['NM_133266.3:c.802C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_573573.2:p.(Leu268=)', 'slr': 'NP_573573.2:p.(L268=)'}
        assert results['NM_133266.3:c.802C>T']['submitted_variant'] == 'HG865_PATCH-33547-G-A'
        assert results['NM_133266.3:c.802C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_133266.3:c.802C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_133266.3:c.802C>T']['hgvs_transcript_variant'] == 'NM_133266.3:c.802C>T'
        assert results['NM_133266.3:c.802C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_133266.3:c.802C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.70335439G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '70335439', 'alt': u'A'}}
        assert 'hg38' not in results['NM_133266.3:c.802C>T']['primary_assembly_loci'].keys()
        assert results['NM_133266.3:c.802C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.70335439G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '70335439', 'alt': u'A'}}
        assert 'grch38' not in results['NM_133266.3:c.802C>T']['primary_assembly_loci'].keys()

        assert 'NM_012309.4:c.2566C>T' in results.keys()
        assert results['NM_012309.4:c.2566C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_012309.4:c.2566C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_012309.4:c.2566C>T']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_004070871.1:g.33547G>A', 'vcf': {'chr': 'HG865_PATCH', 'ref': u'G', 'pos': '33547', 'alt': u'A'}}}]
        assert results['NM_012309.4:c.2566C>T']['gene_symbol'] == 'SHANK2'
        assert results['NM_012309.4:c.2566C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_036441.2:p.(Leu856=)', 'slr': 'NP_036441.2:p.(L856=)'}
        assert results['NM_012309.4:c.2566C>T']['submitted_variant'] == 'HG865_PATCH-33547-G-A'
        assert results['NM_012309.4:c.2566C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_012309.4:c.2566C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_012309.4:c.2566C>T']['hgvs_transcript_variant'] == 'NM_012309.4:c.2566C>T'
        assert results['NM_012309.4:c.2566C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_012309.4:c.2566C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000011.9:g.70336423G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '70336423', 'alt': u'A'}}
        assert results['NM_012309.4:c.2566C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.70489334G>A', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '70489334', 'alt': u'A'}}
        assert results['NM_012309.4:c.2566C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000011.9:g.70336423G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '70336423', 'alt': u'A'}}
        assert results['NM_012309.4:c.2566C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.70489334G>A', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '70489334', 'alt': u'A'}}


    def test_variant315(self):
        variant = 'HG865_PATCH-569441-G-T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_012309.4:c.960C>A' in results.keys()
        assert results['NM_012309.4:c.960C>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_012309.4:c.960C>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_012309.4:c.960C>A']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_004070871.1:g.569441G>T', 'vcf': {'chr': 'HG865_PATCH', 'ref': u'G', 'pos': '569441', 'alt': u'T'}}}]
        assert results['NM_012309.4:c.960C>A']['gene_symbol'] == 'SHANK2'
        assert results['NM_012309.4:c.960C>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_036441.2:p.(Tyr320Ter)', 'slr': 'NP_036441.2:p.(Y320*)'}
        assert results['NM_012309.4:c.960C>A']['submitted_variant'] == 'HG865_PATCH-569441-G-T'
        assert results['NM_012309.4:c.960C>A']['genome_context_intronic_sequence'] == ''
        assert results['NM_012309.4:c.960C>A']['hgvs_lrg_variant'] == ''
        assert results['NM_012309.4:c.960C>A']['hgvs_transcript_variant'] == 'NM_012309.4:c.960C>A'
        assert results['NM_012309.4:c.960C>A']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['NM_012309.4:c.960C>A']['primary_assembly_loci'].keys()
        assert results['NM_012309.4:c.960C>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.71075228G>T', 'vcf': {'chr': 'chr11', 'ref': u'G', 'pos': '71075228', 'alt': u'T'}}
        assert 'grch37' not in results['NM_012309.4:c.960C>A']['primary_assembly_loci'].keys()
        assert results['NM_012309.4:c.960C>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.71075228G>T', 'vcf': {'chr': '11', 'ref': u'G', 'pos': '71075228', 'alt': u'T'}}


    def test_variant316(self):
        variant = 'HG865_PATCH-574546-C-T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_012309.4:c.913-5058G>A' in results.keys()
        assert results['NM_012309.4:c.913-5058G>A']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_012309.4:c.913-5058G>A']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_012309.4:c.913-5058G>A']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_004070871.1:g.574546C>T', 'vcf': {'chr': 'HG865_PATCH', 'ref': u'C', 'pos': '574546', 'alt': u'T'}}}]
        assert results['NM_012309.4:c.913-5058G>A']['gene_symbol'] == 'SHANK2'
        assert results['NM_012309.4:c.913-5058G>A']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_036441.2:p.?', 'slr': 'NP_036441.2:p.?'}
        assert results['NM_012309.4:c.913-5058G>A']['submitted_variant'] == 'HG865_PATCH-574546-C-T'
        assert results['NM_012309.4:c.913-5058G>A']['genome_context_intronic_sequence'] == 'NC_000011.10(NM_012309.4):c.913-5058G>A'
        assert results['NM_012309.4:c.913-5058G>A']['hgvs_lrg_variant'] == ''
        assert results['NM_012309.4:c.913-5058G>A']['hgvs_transcript_variant'] == 'NM_012309.4:c.913-5058G>A'
        assert results['NM_012309.4:c.913-5058G>A']['hgvs_refseqgene_variant'] == ''
        assert 'hg19' not in results['NM_012309.4:c.913-5058G>A']['primary_assembly_loci'].keys()
        assert results['NM_012309.4:c.913-5058G>A']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000011.10:g.71080333C>T', 'vcf': {'chr': 'chr11', 'ref': u'C', 'pos': '71080333', 'alt': u'T'}}
        assert 'grch37' not in results['NM_012309.4:c.913-5058G>A']['primary_assembly_loci'].keys()
        assert results['NM_012309.4:c.913-5058G>A']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000011.10:g.71080333C>T', 'vcf': {'chr': '11', 'ref': u'C', 'pos': '71080333', 'alt': u'T'}}


    def test_variant317(self):
        variant = 'HSCHR1_1_CTG31-133178-TAG-T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_020699.2:c.802_803insTT' in results.keys()
        assert results['NM_020699.2:c.802_803insTT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_020699.2:c.802_803insTT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_020699.2:c.802_803insTT']['alt_genomic_loci'] == []
        assert results['NM_020699.2:c.802_803insTT']['gene_symbol'] == 'GATAD2B'
        assert results['NM_020699.2:c.802_803insTT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_065750.1:p.(Pro268LeufsTer26)', 'slr': 'NP_065750.1:p.(P268Lfs*26)'}
        assert results['NM_020699.2:c.802_803insTT']['submitted_variant'] == 'HSCHR1_1_CTG31-133178-TAG-T'
        assert results['NM_020699.2:c.802_803insTT']['genome_context_intronic_sequence'] == ''
        assert results['NM_020699.2:c.802_803insTT']['hgvs_lrg_variant'] == ''
        assert results['NM_020699.2:c.802_803insTT']['hgvs_transcript_variant'] == 'NM_020699.2:c.802_803insTT'
        assert results['NM_020699.2:c.802_803insTT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_020699.2:c.802_803insTT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000001.10:g.153789945_153789946delinsGAAG', 'vcf': {'chr': 'chr1', 'ref': 'G', 'pos': '153789945', 'alt': u'GAA'}}
        assert results['NM_020699.2:c.802_803insTT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000001.11:g.153817469_153817470insAA', 'vcf': {'chr': 'chr1', 'ref': 'G', 'pos': '153817469', 'alt': u'GAA'}}
        assert results['NM_020699.2:c.802_803insTT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000001.10:g.153789945_153789946delinsGAAG', 'vcf': {'chr': '1', 'ref': 'G', 'pos': '153789945', 'alt': u'GAA'}}
        assert results['NM_020699.2:c.802_803insTT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000001.11:g.153817469_153817470insAA', 'vcf': {'chr': '1', 'ref': 'G', 'pos': '153817469', 'alt': u'GAA'}}


    def test_variant318(self):
        variant = 'HSCHR6_MHC_MANN_CTG1-3848158-T-G'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_021983.4:c.490G>C' in results.keys()
        assert results['NM_021983.4:c.490G>C']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_021983.4:c.490G>C']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_021983.4:c.490G>C']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NT_167246.1:g.3848158T>G', 'vcf': {'chr': 'HSCHR6_MHC_MANN_CTG1', 'ref': 'T', 'pos': '3848158', 'alt': u'G'}}}, {'hg19': {'hgvs_genomic_description': 'NT_167246.1:g.3848158T>G', 'vcf': {'chr': 'chr6_mann_hap4', 'ref': 'T', 'pos': '3848158', 'alt': u'G'}}}, {'grch38': {'hgvs_genomic_description': 'NT_167246.2:g.3842538T>G', 'vcf': {'chr': 'HSCHR6_MHC_MANN_CTG1', 'ref': 'T', 'pos': '3842538', 'alt': u'G'}}}, {'hg38': {'hgvs_genomic_description': 'NT_167246.2:g.3842538T>G', 'vcf': {'chr': 'chr6_GL000253v2_alt', 'ref': 'T', 'pos': '3842538', 'alt': u'G'}}}, {'grch37': {'hgvs_genomic_description': 'NT_167247.1:g.3884432C>G', 'vcf': {'chr': 'HSCHR6_MHC_MCF_CTG1', 'ref': u'C', 'pos': '3884432', 'alt': u'G'}}}, {'hg19': {'hgvs_genomic_description': 'NT_167247.1:g.3884432C>G', 'vcf': {'chr': 'chr6_mcf_hap5', 'ref': u'C', 'pos': '3884432', 'alt': u'G'}}}, {'grch37': {'hgvs_genomic_description': 'NT_167249.1:g.3852542C>G', 'vcf': {'chr': 'HSCHR6_MHC_SSTO_CTG1', 'ref': u'C', 'pos': '3852542', 'alt': u'G'}}}, {'hg19': {'hgvs_genomic_description': 'NT_167249.1:g.3852542C>G', 'vcf': {'chr': 'chr6_ssto_hap7', 'ref': u'C', 'pos': '3852542', 'alt': u'G'}}}, {'grch38': {'hgvs_genomic_description': 'NT_167249.2:g.3853244C>G', 'vcf': {'chr': 'HSCHR6_MHC_SSTO_CTG1', 'ref': u'C', 'pos': '3853244', 'alt': u'G'}}}, {'hg38': {'hgvs_genomic_description': 'NT_167249.2:g.3853244C>G', 'vcf': {'chr': 'chr6_GL000256v2_alt', 'ref': u'C', 'pos': '3853244', 'alt': u'G'}}}]
        assert results['NM_021983.4:c.490G>C']['gene_symbol'] == 'HLA-DRB4'
        assert results['NM_021983.4:c.490G>C']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_068818.4:p.(Gly164Arg)', 'slr': 'NP_068818.4:p.(G164R)'}
        assert results['NM_021983.4:c.490G>C']['submitted_variant'] == 'HSCHR6_MHC_MANN_CTG1-3848158-T-G'
        assert results['NM_021983.4:c.490G>C']['genome_context_intronic_sequence'] == ''
        assert results['NM_021983.4:c.490G>C']['hgvs_lrg_variant'] == ''
        assert results['NM_021983.4:c.490G>C']['hgvs_transcript_variant'] == 'NM_021983.4:c.490G>C'
        assert results['NM_021983.4:c.490G>C']['hgvs_refseqgene_variant'] == 'NG_002433.1:g.5724C>G'
        assert 'hg19' not in results['NM_021983.4:c.490G>C']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['NM_021983.4:c.490G>C']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['NM_021983.4:c.490G>C']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['NM_021983.4:c.490G>C']['primary_assembly_loci'].keys()


    def test_variant319(self):
        variant = 'HSCHR6_MHC_MANN_CTG1-3851043-C-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_021983.4:c.346G>T' in results.keys()
        assert results['NM_021983.4:c.346G>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_021983.4:c.346G>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_021983.4:c.346G>T']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NT_167246.1:g.3851043C>A', 'vcf': {'chr': 'HSCHR6_MHC_MANN_CTG1', 'ref': u'C', 'pos': '3851043', 'alt': u'A'}}}, {'hg19': {'hgvs_genomic_description': 'NT_167246.1:g.3851043C>A', 'vcf': {'chr': 'chr6_mann_hap4', 'ref': u'C', 'pos': '3851043', 'alt': u'A'}}}, {'grch38': {'hgvs_genomic_description': 'NT_167246.2:g.3845423C>A', 'vcf': {'chr': 'HSCHR6_MHC_MANN_CTG1', 'ref': u'C', 'pos': '3845423', 'alt': u'A'}}}, {'hg38': {'hgvs_genomic_description': 'NT_167246.2:g.3845423C>A', 'vcf': {'chr': 'chr6_GL000253v2_alt', 'ref': u'C', 'pos': '3845423', 'alt': u'A'}}}, {'grch37': {'hgvs_genomic_description': 'NT_167247.1:g.3887313C>A', 'vcf': {'chr': 'HSCHR6_MHC_MCF_CTG1', 'ref': u'C', 'pos': '3887313', 'alt': u'A'}}}, {'hg19': {'hgvs_genomic_description': 'NT_167247.1:g.3887313C>A', 'vcf': {'chr': 'chr6_mcf_hap5', 'ref': u'C', 'pos': '3887313', 'alt': u'A'}}}, {'grch37': {'hgvs_genomic_description': 'NT_167249.1:g.3855423C>A', 'vcf': {'chr': 'HSCHR6_MHC_SSTO_CTG1', 'ref': u'C', 'pos': '3855423', 'alt': u'A'}}}, {'hg19': {'hgvs_genomic_description': 'NT_167249.1:g.3855423C>A', 'vcf': {'chr': 'chr6_ssto_hap7', 'ref': u'C', 'pos': '3855423', 'alt': u'A'}}}, {'grch38': {'hgvs_genomic_description': 'NT_167249.2:g.3856125C>A', 'vcf': {'chr': 'HSCHR6_MHC_SSTO_CTG1', 'ref': u'C', 'pos': '3856125', 'alt': u'A'}}}, {'hg38': {'hgvs_genomic_description': 'NT_167249.2:g.3856125C>A', 'vcf': {'chr': 'chr6_GL000256v2_alt', 'ref': u'C', 'pos': '3856125', 'alt': u'A'}}}]
        assert results['NM_021983.4:c.346G>T']['gene_symbol'] == 'HLA-DRB4'
        assert results['NM_021983.4:c.346G>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_068818.4:p.(Glu116Ter)', 'slr': 'NP_068818.4:p.(E116*)'}
        assert results['NM_021983.4:c.346G>T']['submitted_variant'] == 'HSCHR6_MHC_MANN_CTG1-3851043-C-A'
        assert results['NM_021983.4:c.346G>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_021983.4:c.346G>T']['hgvs_lrg_variant'] == ''
        assert results['NM_021983.4:c.346G>T']['hgvs_transcript_variant'] == 'NM_021983.4:c.346G>T'
        assert results['NM_021983.4:c.346G>T']['hgvs_refseqgene_variant'] == 'NG_002433.1:g.8605C>A'
        assert 'hg19' not in results['NM_021983.4:c.346G>T']['primary_assembly_loci'].keys()
        assert 'hg38' not in results['NM_021983.4:c.346G>T']['primary_assembly_loci'].keys()
        assert 'grch37' not in results['NM_021983.4:c.346G>T']['primary_assembly_loci'].keys()
        assert 'grch38' not in results['NM_021983.4:c.346G>T']['primary_assembly_loci'].keys()


    def test_variant320(self):
        variant = 'X-70443101-C-T'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_001097642.2:c.-16-441C>T' in results.keys()
        assert results['NM_001097642.2:c.-16-441C>T']['hgvs_lrg_transcript_variant'] == 'LRG_245t1:c.-16-441C>T'
        assert results['NM_001097642.2:c.-16-441C>T']['refseqgene_context_intronic_sequence'] == 'NG_008357.1(NM_001097642.2):c.-16-441C>T'
        assert results['NM_001097642.2:c.-16-441C>T']['alt_genomic_loci'] == []
        assert results['NM_001097642.2:c.-16-441C>T']['gene_symbol'] == 'GJB1'
        assert results['NM_001097642.2:c.-16-441C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001091111.1:p.?', 'slr': 'NP_001091111.1:p.?'}
        assert results['NM_001097642.2:c.-16-441C>T']['submitted_variant'] == 'X-70443101-C-T'
        assert results['NM_001097642.2:c.-16-441C>T']['genome_context_intronic_sequence'] == 'NC_000023.10(NM_001097642.2):c.-16-441C>T'
        assert results['NM_001097642.2:c.-16-441C>T']['hgvs_lrg_variant'] == 'LRG_245:g.13040C>T'
        assert results['NM_001097642.2:c.-16-441C>T']['hgvs_transcript_variant'] == 'NM_001097642.2:c.-16-441C>T'
        assert results['NM_001097642.2:c.-16-441C>T']['hgvs_refseqgene_variant'] == 'NG_008357.1:g.13040C>T'
        assert results['NM_001097642.2:c.-16-441C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.70443101C>T', 'vcf': {'chr': 'chrX', 'ref': 'C', 'pos': '70443101', 'alt': 'T'}}
        assert results['NM_001097642.2:c.-16-441C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.71223251C>T', 'vcf': {'chr': 'chrX', 'ref': 'C', 'pos': '71223251', 'alt': 'T'}}
        assert results['NM_001097642.2:c.-16-441C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.70443101C>T', 'vcf': {'chr': 'X', 'ref': 'C', 'pos': '70443101', 'alt': 'T'}}
        assert results['NM_001097642.2:c.-16-441C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.71223251C>T', 'vcf': {'chr': 'X', 'ref': 'C', 'pos': '71223251', 'alt': 'T'}}

        assert 'NM_000166.5:c.-101C>T' in results.keys()
        assert results['NM_000166.5:c.-101C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000166.5:c.-101C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000166.5:c.-101C>T']['alt_genomic_loci'] == []
        assert results['NM_000166.5:c.-101C>T']['gene_symbol'] == 'GJB1'
        assert results['NM_000166.5:c.-101C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000157.1:p.?', 'slr': 'NP_000157.1:p.?'}
        assert results['NM_000166.5:c.-101C>T']['submitted_variant'] == 'X-70443101-C-T'
        assert results['NM_000166.5:c.-101C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_000166.5:c.-101C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_000166.5:c.-101C>T']['hgvs_transcript_variant'] == 'NM_000166.5:c.-101C>T'
        assert results['NM_000166.5:c.-101C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_000166.5:c.-101C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.70443101C>T', 'vcf': {'chr': 'chrX', 'ref': 'C', 'pos': '70443101', 'alt': 'T'}}
        assert results['NM_000166.5:c.-101C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.71223251C>T', 'vcf': {'chr': 'chrX', 'ref': 'C', 'pos': '71223251', 'alt': 'T'}}
        assert results['NM_000166.5:c.-101C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.70443101C>T', 'vcf': {'chr': 'X', 'ref': 'C', 'pos': '70443101', 'alt': 'T'}}
        assert results['NM_000166.5:c.-101C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.71223251C>T', 'vcf': {'chr': 'X', 'ref': 'C', 'pos': '71223251', 'alt': 'T'}}


    def test_variant321(self):
        variant = 'X-107845202-GACCACC-GACC,G'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_000495.4:c.2130_2135del' in results.keys()
        assert results['NM_000495.4:c.2130_2135del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000495.4:c.2130_2135del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000495.4:c.2130_2135del']['alt_genomic_loci'] == []
        assert results['NM_000495.4:c.2130_2135del']['gene_symbol'] == 'COL4A5'
        assert results['NM_000495.4:c.2130_2135del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000486.1(LRG_232p1):p.(Pro711_Pro712del)', 'slr': 'NP_000486.1:p.(P711_P712del)'}
        assert results['NM_000495.4:c.2130_2135del']['submitted_variant'] == 'X-107845202-GACCACC-GACC,G'
        assert results['NM_000495.4:c.2130_2135del']['genome_context_intronic_sequence'] == ''
        assert results['NM_000495.4:c.2130_2135del']['hgvs_lrg_variant'] == ''
        assert results['NM_000495.4:c.2130_2135del']['hgvs_transcript_variant'] == 'NM_000495.4:c.2130_2135del'
        assert results['NM_000495.4:c.2130_2135del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_000495.4:c.2130_2135del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.107845203_107845208del', 'vcf': {'chr': 'chrX', 'ref': 'GACCACC', 'pos': '107845202', 'alt': 'G'}}
        assert results['NM_000495.4:c.2130_2135del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.108601973_108601978del', 'vcf': {'chr': 'chrX', 'ref': 'GACCACC', 'pos': '108601972', 'alt': 'G'}}
        assert results['NM_000495.4:c.2130_2135del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.107845203_107845208del', 'vcf': {'chr': 'X', 'ref': 'GACCACC', 'pos': '107845202', 'alt': 'G'}}
        assert results['NM_000495.4:c.2130_2135del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.108601973_108601978del', 'vcf': {'chr': 'X', 'ref': 'GACCACC', 'pos': '108601972', 'alt': 'G'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_033380.2:c.2130_2135del' in results.keys()
        assert results['NM_033380.2:c.2130_2135del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_033380.2:c.2130_2135del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_033380.2:c.2130_2135del']['alt_genomic_loci'] == []
        assert results['NM_033380.2:c.2130_2135del']['gene_symbol'] == 'COL4A5'
        assert results['NM_033380.2:c.2130_2135del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_203699.1(LRG_232p2):p.(Pro711_Pro712del)', 'slr': 'NP_203699.1:p.(P711_P712del)'}
        assert results['NM_033380.2:c.2130_2135del']['submitted_variant'] == 'X-107845202-GACCACC-GACC,G'
        assert results['NM_033380.2:c.2130_2135del']['genome_context_intronic_sequence'] == ''
        assert results['NM_033380.2:c.2130_2135del']['hgvs_lrg_variant'] == ''
        assert results['NM_033380.2:c.2130_2135del']['hgvs_transcript_variant'] == 'NM_033380.2:c.2130_2135del'
        assert results['NM_033380.2:c.2130_2135del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_033380.2:c.2130_2135del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.107845203_107845208del', 'vcf': {'chr': 'chrX', 'ref': 'GACCACC', 'pos': '107845202', 'alt': 'G'}}
        assert results['NM_033380.2:c.2130_2135del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.108601973_108601978del', 'vcf': {'chr': 'chrX', 'ref': 'GACCACC', 'pos': '108601972', 'alt': 'G'}}
        assert results['NM_033380.2:c.2130_2135del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.107845203_107845208del', 'vcf': {'chr': 'X', 'ref': 'GACCACC', 'pos': '107845202', 'alt': 'G'}}
        assert results['NM_033380.2:c.2130_2135del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.108601973_108601978del', 'vcf': {'chr': 'X', 'ref': 'GACCACC', 'pos': '108601972', 'alt': 'G'}}

        assert 'NM_033380.2:c.2133_2135del' in results.keys()
        assert results['NM_033380.2:c.2133_2135del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_033380.2:c.2133_2135del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_033380.2:c.2133_2135del']['alt_genomic_loci'] == []
        assert results['NM_033380.2:c.2133_2135del']['gene_symbol'] == 'COL4A5'
        assert results['NM_033380.2:c.2133_2135del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_203699.1(LRG_232p2):p.(Pro712del)', 'slr': 'NP_203699.1:p.(P712del)'}
        assert results['NM_033380.2:c.2133_2135del']['submitted_variant'] == 'X-107845202-GACCACC-GACC,G'
        assert results['NM_033380.2:c.2133_2135del']['genome_context_intronic_sequence'] == ''
        assert results['NM_033380.2:c.2133_2135del']['hgvs_lrg_variant'] == ''
        assert results['NM_033380.2:c.2133_2135del']['hgvs_transcript_variant'] == 'NM_033380.2:c.2133_2135del'
        assert results['NM_033380.2:c.2133_2135del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_033380.2:c.2133_2135del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.107845206_107845208del', 'vcf': {'chr': 'chrX', 'ref': 'GACC', 'pos': '107845202', 'alt': 'G'}}
        assert results['NM_033380.2:c.2133_2135del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.108601976_108601978del', 'vcf': {'chr': 'chrX', 'ref': 'GACC', 'pos': '108601972', 'alt': 'G'}}
        assert results['NM_033380.2:c.2133_2135del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.107845206_107845208del', 'vcf': {'chr': 'X', 'ref': 'GACC', 'pos': '107845202', 'alt': 'G'}}
        assert results['NM_033380.2:c.2133_2135del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.108601976_108601978del', 'vcf': {'chr': 'X', 'ref': 'GACC', 'pos': '108601972', 'alt': 'G'}}

        assert 'NM_000495.4:c.2133_2135del' in results.keys()
        assert results['NM_000495.4:c.2133_2135del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000495.4:c.2133_2135del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000495.4:c.2133_2135del']['alt_genomic_loci'] == []
        assert results['NM_000495.4:c.2133_2135del']['gene_symbol'] == 'COL4A5'
        assert results['NM_000495.4:c.2133_2135del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000486.1(LRG_232p1):p.(Pro712del)', 'slr': 'NP_000486.1:p.(P712del)'}
        assert results['NM_000495.4:c.2133_2135del']['submitted_variant'] == 'X-107845202-GACCACC-GACC,G'
        assert results['NM_000495.4:c.2133_2135del']['genome_context_intronic_sequence'] == ''
        assert results['NM_000495.4:c.2133_2135del']['hgvs_lrg_variant'] == ''
        assert results['NM_000495.4:c.2133_2135del']['hgvs_transcript_variant'] == 'NM_000495.4:c.2133_2135del'
        assert results['NM_000495.4:c.2133_2135del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_000495.4:c.2133_2135del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.107845206_107845208del', 'vcf': {'chr': 'chrX', 'ref': 'GACC', 'pos': '107845202', 'alt': 'G'}}
        assert results['NM_000495.4:c.2133_2135del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.108601976_108601978del', 'vcf': {'chr': 'chrX', 'ref': 'GACC', 'pos': '108601972', 'alt': 'G'}}
        assert results['NM_000495.4:c.2133_2135del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.107845206_107845208del', 'vcf': {'chr': 'X', 'ref': 'GACC', 'pos': '107845202', 'alt': 'G'}}
        assert results['NM_000495.4:c.2133_2135del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.108601976_108601978del', 'vcf': {'chr': 'X', 'ref': 'GACC', 'pos': '108601972', 'alt': 'G'}}


    def test_variant322(self):
        variant = 'X-153296777-G-A'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_004992.3:c.502C>T' in results.keys()
        assert results['NM_004992.3:c.502C>T']['hgvs_lrg_transcript_variant'] == 'LRG_764t2:c.502C>T'
        assert results['NM_004992.3:c.502C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_004992.3:c.502C>T']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003871103.3:g.1465305G>A', 'vcf': {'chr': 'HG1497_PATCH', 'ref': u'G', 'pos': '1465305', 'alt': u'A'}}}]
        assert results['NM_004992.3:c.502C>T']['gene_symbol'] == 'MECP2'
        assert results['NM_004992.3:c.502C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_004983.1(LRG_764p2):p.(Arg168Ter)', 'slr': 'NP_004983.1:p.(R168*)'}
        assert results['NM_004992.3:c.502C>T']['submitted_variant'] == 'X-153296777-G-A'
        assert results['NM_004992.3:c.502C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_004992.3:c.502C>T']['hgvs_lrg_variant'] == 'LRG_764:g.110802C>T'
        assert results['NM_004992.3:c.502C>T']['hgvs_transcript_variant'] == 'NM_004992.3:c.502C>T'
        assert results['NM_004992.3:c.502C>T']['hgvs_refseqgene_variant'] == 'NG_007107.2:g.110802C>T'
        assert results['NM_004992.3:c.502C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.153296777G>A', 'vcf': {'chr': 'chrX', 'ref': u'G', 'pos': '153296777', 'alt': u'A'}}
        assert results['NM_004992.3:c.502C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.154031326G>A', 'vcf': {'chr': 'chrX', 'ref': u'G', 'pos': '154031326', 'alt': u'A'}}
        assert results['NM_004992.3:c.502C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.153296777G>A', 'vcf': {'chr': 'X', 'ref': u'G', 'pos': '153296777', 'alt': u'A'}}
        assert results['NM_004992.3:c.502C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.154031326G>A', 'vcf': {'chr': 'X', 'ref': u'G', 'pos': '154031326', 'alt': u'A'}}

        assert results['flag'] == 'gene_variant'
        assert 'NM_001316337.1:c.223C>T' in results.keys()
        assert results['NM_001316337.1:c.223C>T']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001316337.1:c.223C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001316337.1:c.223C>T']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003871103.3:g.1465305G>A', 'vcf': {'chr': 'HG1497_PATCH', 'ref': u'G', 'pos': '1465305', 'alt': u'A'}}}]
        assert results['NM_001316337.1:c.223C>T']['gene_symbol'] == 'MECP2'
        assert results['NM_001316337.1:c.223C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001303266.1:p.(Arg75Ter)', 'slr': 'NP_001303266.1:p.(R75*)'}
        assert results['NM_001316337.1:c.223C>T']['submitted_variant'] == 'X-153296777-G-A'
        assert results['NM_001316337.1:c.223C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001316337.1:c.223C>T']['hgvs_lrg_variant'] == ''
        assert results['NM_001316337.1:c.223C>T']['hgvs_transcript_variant'] == 'NM_001316337.1:c.223C>T'
        assert results['NM_001316337.1:c.223C>T']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001316337.1:c.223C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.153296777G>A', 'vcf': {'chr': 'chrX', 'ref': u'G', 'pos': '153296777', 'alt': u'A'}}
        assert results['NM_001316337.1:c.223C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.154031326G>A', 'vcf': {'chr': 'chrX', 'ref': u'G', 'pos': '154031326', 'alt': u'A'}}
        assert results['NM_001316337.1:c.223C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.153296777G>A', 'vcf': {'chr': 'X', 'ref': u'G', 'pos': '153296777', 'alt': u'A'}}
        assert results['NM_001316337.1:c.223C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.154031326G>A', 'vcf': {'chr': 'X', 'ref': u'G', 'pos': '154031326', 'alt': u'A'}}

        assert 'NM_001110792.1:c.538C>T' in results.keys()
        assert results['NM_001110792.1:c.538C>T']['hgvs_lrg_transcript_variant'] == 'LRG_764t1:c.538C>T'
        assert results['NM_001110792.1:c.538C>T']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001110792.1:c.538C>T']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_003871103.3:g.1465305G>A', 'vcf': {'chr': 'HG1497_PATCH', 'ref': u'G', 'pos': '1465305', 'alt': u'A'}}}]
        assert results['NM_001110792.1:c.538C>T']['gene_symbol'] == 'MECP2'
        assert results['NM_001110792.1:c.538C>T']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001104262.1:p.(Arg180Ter)', 'slr': 'NP_001104262.1:p.(R180*)'}
        assert results['NM_001110792.1:c.538C>T']['submitted_variant'] == 'X-153296777-G-A'
        assert results['NM_001110792.1:c.538C>T']['genome_context_intronic_sequence'] == ''
        assert results['NM_001110792.1:c.538C>T']['hgvs_lrg_variant'] == 'LRG_764:g.110802C>T'
        assert results['NM_001110792.1:c.538C>T']['hgvs_transcript_variant'] == 'NM_001110792.1:c.538C>T'
        assert results['NM_001110792.1:c.538C>T']['hgvs_refseqgene_variant'] == 'NG_007107.2:g.110802C>T'
        assert results['NM_001110792.1:c.538C>T']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.153296777G>A', 'vcf': {'chr': 'chrX', 'ref': u'G', 'pos': '153296777', 'alt': u'A'}}
        assert results['NM_001110792.1:c.538C>T']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.154031326G>A', 'vcf': {'chr': 'chrX', 'ref': u'G', 'pos': '154031326', 'alt': u'A'}}
        assert results['NM_001110792.1:c.538C>T']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.153296777G>A', 'vcf': {'chr': 'X', 'ref': u'G', 'pos': '153296777', 'alt': u'A'}}
        assert results['NM_001110792.1:c.538C>T']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.154031326G>A', 'vcf': {'chr': 'X', 'ref': u'G', 'pos': '154031326', 'alt': u'A'}}


    def test_variant323(self):
        variant = 'NM_198180.2:c.408_410delGTG'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_198180.2:c.408_410del' in results.keys()
        assert results['NM_198180.2:c.408_410del']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_198180.2:c.408_410del']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_198180.2:c.408_410del']['alt_genomic_loci'] == []
        assert results['NM_198180.2:c.408_410del']['gene_symbol'] == 'QRFP'
        assert results['NM_198180.2:c.408_410del']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_937823.1:p.?', 'slr': 'NP_937823.1:p.?'}
        assert results['NM_198180.2:c.408_410del']['submitted_variant'] == 'NM_198180.2:c.408_410delGTG'
        assert results['NM_198180.2:c.408_410del']['genome_context_intronic_sequence'] == ''
        assert results['NM_198180.2:c.408_410del']['hgvs_lrg_variant'] == ''
        assert results['NM_198180.2:c.408_410del']['hgvs_transcript_variant'] == 'NM_198180.2:c.408_410del'
        assert results['NM_198180.2:c.408_410del']['hgvs_refseqgene_variant'] == ''
        assert results['NM_198180.2:c.408_410del']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.133768816_133768818del', 'vcf': {'chr': 'chr9', 'ref': 'TCAC', 'pos': '133768815', 'alt': 'T'}}
        assert results['NM_198180.2:c.408_410del']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.130893429_130893431del', 'vcf': {'chr': 'chr9', 'ref': 'TCAC', 'pos': '130893428', 'alt': 'T'}}
        assert results['NM_198180.2:c.408_410del']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.133768816_133768818del', 'vcf': {'chr': '9', 'ref': 'TCAC', 'pos': '133768815', 'alt': 'T'}}
        assert results['NM_198180.2:c.408_410del']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.130893429_130893431del', 'vcf': {'chr': '9', 'ref': 'TCAC', 'pos': '130893428', 'alt': 'T'}}


    def test_variant324(self):
        variant = 'NM_080877.2:c.1733_1735delinsTTT'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_080877.2:c.1733_1735delinsTTT' in results.keys()
        assert results['NM_080877.2:c.1733_1735delinsTTT']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_080877.2:c.1733_1735delinsTTT']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_080877.2:c.1733_1735delinsTTT']['alt_genomic_loci'] == []
        assert results['NM_080877.2:c.1733_1735delinsTTT']['gene_symbol'] == 'SLC34A3'
        assert results['NM_080877.2:c.1733_1735delinsTTT']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_543153.1:p.(Pro578_Lys579delinsLeuTer)', 'slr': 'NP_543153.1:p.(P578_K579delinsL*)'}
        assert results['NM_080877.2:c.1733_1735delinsTTT']['submitted_variant'] == 'NM_080877.2:c.1733_1735delinsTTT'
        assert results['NM_080877.2:c.1733_1735delinsTTT']['genome_context_intronic_sequence'] == ''
        assert results['NM_080877.2:c.1733_1735delinsTTT']['hgvs_lrg_variant'] == ''
        assert results['NM_080877.2:c.1733_1735delinsTTT']['hgvs_transcript_variant'] == 'NM_080877.2:c.1733_1735delinsTTT'
        assert results['NM_080877.2:c.1733_1735delinsTTT']['hgvs_refseqgene_variant'] == ''
        assert results['NM_080877.2:c.1733_1735delinsTTT']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.140130801_140130803delinsTTT', 'vcf': {'chr': 'chr9', 'ref': 'CGA', 'pos': '140130801', 'alt': 'TTT'}}
        assert results['NM_080877.2:c.1733_1735delinsTTT']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.137236349_137236351delinsTTT', 'vcf': {'chr': 'chr9', 'ref': 'CGA', 'pos': '137236349', 'alt': 'TTT'}}
        assert results['NM_080877.2:c.1733_1735delinsTTT']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.140130801_140130803delinsTTT', 'vcf': {'chr': '9', 'ref': 'CGA', 'pos': '140130801', 'alt': 'TTT'}}
        assert results['NM_080877.2:c.1733_1735delinsTTT']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.137236349_137236351delinsTTT', 'vcf': {'chr': '9', 'ref': 'CGA', 'pos': '137236349', 'alt': 'TTT'}}

        assert results['flag'] == 'gene_variant'

    def test_variant325(self):
        variant = 'NM_080877.2:c.1735_1737delinsTGA'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_080877.2:c.1735_1737delinsTGA' in results.keys()
        assert results['NM_080877.2:c.1735_1737delinsTGA']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_080877.2:c.1735_1737delinsTGA']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_080877.2:c.1735_1737delinsTGA']['alt_genomic_loci'] == []
        assert results['NM_080877.2:c.1735_1737delinsTGA']['gene_symbol'] == 'SLC34A3'
        assert results['NM_080877.2:c.1735_1737delinsTGA']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_543153.1:p.(Lys579Ter)', 'slr': 'NP_543153.1:p.(K579*)'}
        assert results['NM_080877.2:c.1735_1737delinsTGA']['submitted_variant'] == 'NM_080877.2:c.1735_1737delinsTGA'
        assert results['NM_080877.2:c.1735_1737delinsTGA']['genome_context_intronic_sequence'] == ''
        assert results['NM_080877.2:c.1735_1737delinsTGA']['hgvs_lrg_variant'] == ''
        assert results['NM_080877.2:c.1735_1737delinsTGA']['hgvs_transcript_variant'] == 'NM_080877.2:c.1735_1737delinsTGA'
        assert results['NM_080877.2:c.1735_1737delinsTGA']['hgvs_refseqgene_variant'] == ''
        assert results['NM_080877.2:c.1735_1737delinsTGA']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.140130803_140130805delinsTGA', 'vcf': {'chr': 'chr9', 'ref': 'AAG', 'pos': '140130803', 'alt': 'TGA'}}
        assert results['NM_080877.2:c.1735_1737delinsTGA']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.137236351_137236353delinsTGA', 'vcf': {'chr': 'chr9', 'ref': 'AAG', 'pos': '137236351', 'alt': 'TGA'}}
        assert results['NM_080877.2:c.1735_1737delinsTGA']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.140130803_140130805delinsTGA', 'vcf': {'chr': '9', 'ref': 'AAG', 'pos': '140130803', 'alt': 'TGA'}}
        assert results['NM_080877.2:c.1735_1737delinsTGA']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.137236351_137236353delinsTGA', 'vcf': {'chr': '9', 'ref': 'AAG', 'pos': '137236351', 'alt': 'TGA'}}

        assert results['flag'] == 'gene_variant'

    def test_variant326(self):
        variant = 'NM_080877.2:c.1735_1737delinsTAATTGTTC'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_080877.2:c.1735_1737delinsTAATTGTTC' in results.keys()
        assert results['NM_080877.2:c.1735_1737delinsTAATTGTTC']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_080877.2:c.1735_1737delinsTAATTGTTC']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_080877.2:c.1735_1737delinsTAATTGTTC']['alt_genomic_loci'] == []
        assert results['NM_080877.2:c.1735_1737delinsTAATTGTTC']['gene_symbol'] == 'SLC34A3'
        assert results['NM_080877.2:c.1735_1737delinsTAATTGTTC']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_543153.1:p.(Lys579Ter)', 'slr': 'NP_543153.1:p.(K579*)'}
        assert results['NM_080877.2:c.1735_1737delinsTAATTGTTC']['submitted_variant'] == 'NM_080877.2:c.1735_1737delinsTAATTGTTC'
        assert results['NM_080877.2:c.1735_1737delinsTAATTGTTC']['genome_context_intronic_sequence'] == ''
        assert results['NM_080877.2:c.1735_1737delinsTAATTGTTC']['hgvs_lrg_variant'] == ''
        assert results['NM_080877.2:c.1735_1737delinsTAATTGTTC']['hgvs_transcript_variant'] == 'NM_080877.2:c.1735_1737delinsTAATTGTTC'
        assert results['NM_080877.2:c.1735_1737delinsTAATTGTTC']['hgvs_refseqgene_variant'] == ''
        assert results['NM_080877.2:c.1735_1737delinsTAATTGTTC']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.140130803_140130805delinsTAATTGTTC', 'vcf': {'chr': 'chr9', 'ref': 'AAG', 'pos': '140130803', 'alt': 'TAATTGTTC'}}
        assert results['NM_080877.2:c.1735_1737delinsTAATTGTTC']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.137236351_137236353delinsTAATTGTTC', 'vcf': {'chr': 'chr9', 'ref': 'AAG', 'pos': '137236351', 'alt': 'TAATTGTTC'}}
        assert results['NM_080877.2:c.1735_1737delinsTAATTGTTC']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.140130803_140130805delinsTAATTGTTC', 'vcf': {'chr': '9', 'ref': 'AAG', 'pos': '140130803', 'alt': 'TAATTGTTC'}}
        assert results['NM_080877.2:c.1735_1737delinsTAATTGTTC']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.137236351_137236353delinsTAATTGTTC', 'vcf': {'chr': '9', 'ref': 'AAG', 'pos': '137236351', 'alt': 'TAATTGTTC'}}


    def test_variant327(self):
        variant = 'NM_080877.2:c.1737delinsATTGTTC'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_080877.2:c.1737delinsATTGTTC' in results.keys()
        assert results['NM_080877.2:c.1737delinsATTGTTC']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_080877.2:c.1737delinsATTGTTC']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_080877.2:c.1737delinsATTGTTC']['alt_genomic_loci'] == []
        assert results['NM_080877.2:c.1737delinsATTGTTC']['gene_symbol'] == 'SLC34A3'
        assert results['NM_080877.2:c.1737delinsATTGTTC']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_543153.1:p.(Lys579_Ala580insLeuPhe)', 'slr': 'NP_543153.1:p.(K579_A580insLF)'}
        assert results['NM_080877.2:c.1737delinsATTGTTC']['submitted_variant'] == 'NM_080877.2:c.1737delinsATTGTTC'
        assert results['NM_080877.2:c.1737delinsATTGTTC']['genome_context_intronic_sequence'] == ''
        assert results['NM_080877.2:c.1737delinsATTGTTC']['hgvs_lrg_variant'] == ''
        assert results['NM_080877.2:c.1737delinsATTGTTC']['hgvs_transcript_variant'] == 'NM_080877.2:c.1737delinsATTGTTC'
        assert results['NM_080877.2:c.1737delinsATTGTTC']['hgvs_refseqgene_variant'] == ''
        assert results['NM_080877.2:c.1737delinsATTGTTC']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000009.11:g.140130805delinsATTGTTC', 'vcf': {'chr': 'chr9', 'ref': 'G', 'pos': '140130805', 'alt': 'ATTGTTC'}}
        assert results['NM_080877.2:c.1737delinsATTGTTC']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000009.12:g.137236353delinsATTGTTC', 'vcf': {'chr': 'chr9', 'ref': 'G', 'pos': '137236353', 'alt': 'ATTGTTC'}}
        assert results['NM_080877.2:c.1737delinsATTGTTC']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000009.11:g.140130805delinsATTGTTC', 'vcf': {'chr': '9', 'ref': 'G', 'pos': '140130805', 'alt': 'ATTGTTC'}}
        assert results['NM_080877.2:c.1737delinsATTGTTC']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000009.12:g.137236353delinsATTGTTC', 'vcf': {'chr': '9', 'ref': 'G', 'pos': '137236353', 'alt': 'ATTGTTC'}}


    def test_variant328(self):
        variant = 'NM_000088.3:c.4392_*2delinsAGAG'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000088.3:c.4392_*2delinsAGAG' in results.keys()
        assert results['NM_000088.3:c.4392_*2delinsAGAG']['hgvs_lrg_transcript_variant'] == 'LRG_1t1:c.4392_*2delinsAGAG'
        assert results['NM_000088.3:c.4392_*2delinsAGAG']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000088.3:c.4392_*2delinsAGAG']['alt_genomic_loci'] == []
        assert results['NM_000088.3:c.4392_*2delinsAGAG']['gene_symbol'] == 'COL1A1'
        assert results['NM_000088.3:c.4392_*2delinsAGAG']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000079.2(LRG_1p1):p.(Ter1465GluextTer84)', 'slr': 'NP_000079.2:p.(*1465Eext*84)'}
        assert results['NM_000088.3:c.4392_*2delinsAGAG']['submitted_variant'] == 'NM_000088.3:c.4392_*2delinsAGAG'
        assert results['NM_000088.3:c.4392_*2delinsAGAG']['genome_context_intronic_sequence'] == ''
        assert results['NM_000088.3:c.4392_*2delinsAGAG']['hgvs_lrg_variant'] == 'LRG_1:g.21135_21140delinsAGAG'
        assert results['NM_000088.3:c.4392_*2delinsAGAG']['hgvs_transcript_variant'] == 'NM_000088.3:c.4392_*2delinsAGAG'
        assert results['NM_000088.3:c.4392_*2delinsAGAG']['hgvs_refseqgene_variant'] == 'NG_007400.1:g.21135_21140delinsAGAG'
        assert results['NM_000088.3:c.4392_*2delinsAGAG']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48262861_48262866delinsCTCT', 'vcf': {'chr': 'chr17', 'ref': 'GTTTAC', 'pos': '48262861', 'alt': u'CTCT'}}
        assert results['NM_000088.3:c.4392_*2delinsAGAG']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50185500_50185505delinsCTCT', 'vcf': {'chr': 'chr17', 'ref': 'GTTTAC', 'pos': '50185500', 'alt': u'CTCT'}}
        assert results['NM_000088.3:c.4392_*2delinsAGAG']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48262861_48262866delinsCTCT', 'vcf': {'chr': '17', 'ref': 'GTTTAC', 'pos': '48262861', 'alt': u'CTCT'}}
        assert results['NM_000088.3:c.4392_*2delinsAGAG']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50185500_50185505delinsCTCT', 'vcf': {'chr': '17', 'ref': 'GTTTAC', 'pos': '50185500', 'alt': u'CTCT'}}


    def test_variant329(self):
        variant = 'NM_000088.3:c.589_591delinsAGAAGC'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000088.3:c.589_591delinsAGAAGC' in results.keys()
        assert results['NM_000088.3:c.589_591delinsAGAAGC']['hgvs_lrg_transcript_variant'] == 'LRG_1t1:c.589_591delinsAGAAGC'
        assert results['NM_000088.3:c.589_591delinsAGAAGC']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000088.3:c.589_591delinsAGAAGC']['alt_genomic_loci'] == []
        assert results['NM_000088.3:c.589_591delinsAGAAGC']['gene_symbol'] == 'COL1A1'
        assert results['NM_000088.3:c.589_591delinsAGAAGC']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000079.2(LRG_1p1):p.(Gly197delinsArgSer)', 'slr': 'NP_000079.2:p.(G197delinsRS)'}
        assert results['NM_000088.3:c.589_591delinsAGAAGC']['submitted_variant'] == 'NM_000088.3:c.589_591delinsAGAAGC'
        assert results['NM_000088.3:c.589_591delinsAGAAGC']['genome_context_intronic_sequence'] == ''
        assert results['NM_000088.3:c.589_591delinsAGAAGC']['hgvs_lrg_variant'] == 'LRG_1:g.8638_8640delinsAGAAGC'
        assert results['NM_000088.3:c.589_591delinsAGAAGC']['hgvs_transcript_variant'] == 'NM_000088.3:c.589_591delinsAGAAGC'
        assert results['NM_000088.3:c.589_591delinsAGAAGC']['hgvs_refseqgene_variant'] == 'NG_007400.1:g.8638_8640delinsAGAAGC'
        assert results['NM_000088.3:c.589_591delinsAGAAGC']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275361_48275363delinsGCTTCT', 'vcf': {'chr': 'chr17', 'ref': 'ACC', 'pos': '48275361', 'alt': u'GCTTCT'}}
        assert results['NM_000088.3:c.589_591delinsAGAAGC']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198000_50198002delinsGCTTCT', 'vcf': {'chr': 'chr17', 'ref': 'ACC', 'pos': '50198000', 'alt': u'GCTTCT'}}
        assert results['NM_000088.3:c.589_591delinsAGAAGC']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000017.10:g.48275361_48275363delinsGCTTCT', 'vcf': {'chr': '17', 'ref': 'ACC', 'pos': '48275361', 'alt': u'GCTTCT'}}
        assert results['NM_000088.3:c.589_591delinsAGAAGC']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000017.11:g.50198000_50198002delinsGCTTCT', 'vcf': {'chr': '17', 'ref': 'ACC', 'pos': '50198000', 'alt': u'GCTTCT'}}


    def test_variant330(self):
        variant = 'NM_000885.5:c.*2536delinsAGAAAAATCA'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_000885.5:c.*2536delinsAGAAAAATCA' in results.keys()
        assert results['NM_000885.5:c.*2536delinsAGAAAAATCA']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_000885.5:c.*2536delinsAGAAAAATCA']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000885.5:c.*2536delinsAGAAAAATCA']['alt_genomic_loci'] == []
        assert results['NM_000885.5:c.*2536delinsAGAAAAATCA']['gene_symbol'] == 'ITGA4'
        assert results['NM_000885.5:c.*2536delinsAGAAAAATCA']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000876.3:p.?', 'slr': 'NP_000876.3:p.?'}
        assert results['NM_000885.5:c.*2536delinsAGAAAAATCA']['submitted_variant'] == 'NM_000885.5:c.*2536delinsAGAAAAATCA'
        assert results['NM_000885.5:c.*2536delinsAGAAAAATCA']['genome_context_intronic_sequence'] == ''
        assert results['NM_000885.5:c.*2536delinsAGAAAAATCA']['hgvs_lrg_variant'] == ''
        assert results['NM_000885.5:c.*2536delinsAGAAAAATCA']['hgvs_transcript_variant'] == 'NM_000885.5:c.*2536delinsAGAAAAATCA'
        assert results['NM_000885.5:c.*2536delinsAGAAAAATCA']['hgvs_refseqgene_variant'] == ''
        assert results['NM_000885.5:c.*2536delinsAGAAAAATCA']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000002.11:g.182402790delinsAGAAAAATCA', 'vcf': {'chr': 'chr2', 'ref': 'G', 'pos': '182402790', 'alt': 'AGAAAAATCA'}}
        assert results['NM_000885.5:c.*2536delinsAGAAAAATCA']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000002.12:g.181538063delinsAGAAAAATCA', 'vcf': {'chr': 'chr2', 'ref': 'G', 'pos': '181538063', 'alt': 'AGAAAAATCA'}}
        assert results['NM_000885.5:c.*2536delinsAGAAAAATCA']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000002.11:g.182402790delinsAGAAAAATCA', 'vcf': {'chr': '2', 'ref': 'G', 'pos': '182402790', 'alt': 'AGAAAAATCA'}}
        assert results['NM_000885.5:c.*2536delinsAGAAAAATCA']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000002.12:g.181538063delinsAGAAAAATCA', 'vcf': {'chr': '2', 'ref': 'G', 'pos': '181538063', 'alt': 'AGAAAAATCA'}}

        assert results['flag'] == 'gene_variant'

    def test_variant331(self):
        variant = 'NM_002693.2:c.-186_-185delinsCC'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert 'NM_002693.2:c.-186_-185delinsCC' in results.keys()
        assert results['NM_002693.2:c.-186_-185delinsCC']['hgvs_lrg_transcript_variant'] == 'LRG_765t1:c.-186_-185delinsCC'
        assert results['NM_002693.2:c.-186_-185delinsCC']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_002693.2:c.-186_-185delinsCC']['alt_genomic_loci'] == []
        assert results['NM_002693.2:c.-186_-185delinsCC']['gene_symbol'] == 'POLG'
        assert results['NM_002693.2:c.-186_-185delinsCC']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_002684.1(LRG_765p1):p.?', 'slr': 'NP_002684.1:p.?'}
        assert results['NM_002693.2:c.-186_-185delinsCC']['submitted_variant'] == 'NM_002693.2:c.-186_-185delinsCC'
        assert results['NM_002693.2:c.-186_-185delinsCC']['genome_context_intronic_sequence'] == ''
        assert results['NM_002693.2:c.-186_-185delinsCC']['hgvs_lrg_variant'] == ''
        assert results['NM_002693.2:c.-186_-185delinsCC']['hgvs_transcript_variant'] == 'NM_002693.2:c.-186_-185delinsCC'
        assert results['NM_002693.2:c.-186_-185delinsCC']['hgvs_refseqgene_variant'] == 'NG_008218.1:g.5097_5098delinsCC'
        assert results['NM_002693.2:c.-186_-185delinsCC']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000015.9:g.89877929_89877930delinsGG', 'vcf': {'chr': 'chr15', 'ref': 'CT', 'pos': '89877929', 'alt': u'GG'}}
        assert results['NM_002693.2:c.-186_-185delinsCC']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000015.10:g.89334698_89334699delinsGG', 'vcf': {'chr': 'chr15', 'ref': 'CT', 'pos': '89334698', 'alt': u'GG'}}
        assert results['NM_002693.2:c.-186_-185delinsCC']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000015.9:g.89877929_89877930delinsGG', 'vcf': {'chr': '15', 'ref': 'CT', 'pos': '89877929', 'alt': u'GG'}}
        assert results['NM_002693.2:c.-186_-185delinsCC']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000015.10:g.89334698_89334699delinsGG', 'vcf': {'chr': '15', 'ref': 'CT', 'pos': '89334698', 'alt': u'GG'}}

        assert results['flag'] == 'gene_variant'

    def test_variant332(self):
        variant = 'NG_009616.1:g.29052_29053insCTACATAG'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_001287344.1:c.690_690+1insCTACATAG' in results.keys()
        assert results['NM_001287344.1:c.690_690+1insCTACATAG']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001287344.1:c.690_690+1insCTACATAG']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001287344.1:c.690_690+1insCTACATAG']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_004070883.1:g.43848_43849insCTATGTAG', 'vcf': {'chr': 'HG1439_PATCH', 'ref': 'C', 'pos': '43848', 'alt': u'CCTATGTAG'}}}]
        assert results['NM_001287344.1:c.690_690+1insCTACATAG']['gene_symbol'] == 'BTK'
        assert results['NM_001287344.1:c.690_690+1insCTACATAG']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001274273.1:p.?', 'slr': 'NP_001274273.1:p.?'}
        assert results['NM_001287344.1:c.690_690+1insCTACATAG']['submitted_variant'] == 'NG_009616.1:g.29052_29053insCTACATAG'
        assert results['NM_001287344.1:c.690_690+1insCTACATAG']['genome_context_intronic_sequence'] == 'NC_000023.10(NM_001287344.1):c.690_690+1insCTACATAG'
        assert results['NM_001287344.1:c.690_690+1insCTACATAG']['hgvs_lrg_variant'] == ''
        assert results['NM_001287344.1:c.690_690+1insCTACATAG']['hgvs_transcript_variant'] == 'NM_001287344.1:c.690_690+1insCTACATAG'
        assert results['NM_001287344.1:c.690_690+1insCTACATAG']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001287344.1:c.690_690+1insCTACATAG']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.100617160_100617161insCTATGTAG', 'vcf': {'chr': 'chrX', 'ref': 'C', 'pos': '100617160', 'alt': u'CCTATGTAG'}}
        assert results['NM_001287344.1:c.690_690+1insCTACATAG']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.101362172_101362173insCTATGTAG', 'vcf': {'chr': 'chrX', 'ref': 'C', 'pos': '101362172', 'alt': u'CCTATGTAG'}}
        assert results['NM_001287344.1:c.690_690+1insCTACATAG']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.100617160_100617161insCTATGTAG', 'vcf': {'chr': 'X', 'ref': 'C', 'pos': '100617160', 'alt': u'CCTATGTAG'}}
        assert results['NM_001287344.1:c.690_690+1insCTACATAG']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.101362172_101362173insCTATGTAG', 'vcf': {'chr': 'X', 'ref': 'C', 'pos': '101362172', 'alt': u'CCTATGTAG'}}

        assert 'NM_001287345.1:c.588_588+1insCTACATAG' in results.keys()
        assert results['NM_001287345.1:c.588_588+1insCTACATAG']['hgvs_lrg_transcript_variant'] == ''
        assert results['NM_001287345.1:c.588_588+1insCTACATAG']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_001287345.1:c.588_588+1insCTACATAG']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_004070883.1:g.43848_43849insCTATGTAG', 'vcf': {'chr': 'HG1439_PATCH', 'ref': 'C', 'pos': '43848', 'alt': u'CCTATGTAG'}}}]
        assert results['NM_001287345.1:c.588_588+1insCTACATAG']['gene_symbol'] == 'BTK'
        assert results['NM_001287345.1:c.588_588+1insCTACATAG']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_001274274.1:p.?', 'slr': 'NP_001274274.1:p.?'}
        assert results['NM_001287345.1:c.588_588+1insCTACATAG']['submitted_variant'] == 'NG_009616.1:g.29052_29053insCTACATAG'
        assert results['NM_001287345.1:c.588_588+1insCTACATAG']['genome_context_intronic_sequence'] == 'NC_000023.10(NM_001287345.1):c.588_588+1insCTACATAG'
        assert results['NM_001287345.1:c.588_588+1insCTACATAG']['hgvs_lrg_variant'] == ''
        assert results['NM_001287345.1:c.588_588+1insCTACATAG']['hgvs_transcript_variant'] == 'NM_001287345.1:c.588_588+1insCTACATAG'
        assert results['NM_001287345.1:c.588_588+1insCTACATAG']['hgvs_refseqgene_variant'] == ''
        assert results['NM_001287345.1:c.588_588+1insCTACATAG']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.100617160_100617161insCTATGTAG', 'vcf': {'chr': 'chrX', 'ref': 'C', 'pos': '100617160', 'alt': u'CCTATGTAG'}}
        assert results['NM_001287345.1:c.588_588+1insCTACATAG']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.101362172_101362173insCTATGTAG', 'vcf': {'chr': 'chrX', 'ref': 'C', 'pos': '101362172', 'alt': u'CCTATGTAG'}}
        assert results['NM_001287345.1:c.588_588+1insCTACATAG']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.100617160_100617161insCTATGTAG', 'vcf': {'chr': 'X', 'ref': 'C', 'pos': '100617160', 'alt': u'CCTATGTAG'}}
        assert results['NM_001287345.1:c.588_588+1insCTACATAG']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.101362172_101362173insCTATGTAG', 'vcf': {'chr': 'X', 'ref': 'C', 'pos': '101362172', 'alt': u'CCTATGTAG'}}

        assert 'NM_000061.2:c.588_588+1insCTACATAG' in results.keys()
        assert results['NM_000061.2:c.588_588+1insCTACATAG']['hgvs_lrg_transcript_variant'] == 'LRG_128t1:c.588_588+1insCTACATAG'
        assert results['NM_000061.2:c.588_588+1insCTACATAG']['refseqgene_context_intronic_sequence'] == 'NG_009616.1(NM_000061.2):c.588_588+1insCTACATAG'
        assert results['NM_000061.2:c.588_588+1insCTACATAG']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_004070883.1:g.43848_43849insCTATGTAG', 'vcf': {'chr': 'HG1439_PATCH', 'ref': 'C', 'pos': '43848', 'alt': u'CCTATGTAG'}}}]
        assert results['NM_000061.2:c.588_588+1insCTACATAG']['gene_symbol'] == 'BTK'
        assert results['NM_000061.2:c.588_588+1insCTACATAG']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000052.1(LRG_128p1):p.?', 'slr': 'NP_000052.1:p.?'}
        assert results['NM_000061.2:c.588_588+1insCTACATAG']['submitted_variant'] == 'NG_009616.1:g.29052_29053insCTACATAG'
        assert results['NM_000061.2:c.588_588+1insCTACATAG']['genome_context_intronic_sequence'] == 'NC_000023.10(NM_000061.2):c.588_588+1insCTACATAG'
        assert results['NM_000061.2:c.588_588+1insCTACATAG']['hgvs_lrg_variant'] == 'LRG_128:g.29052_29053insCTACATAG'
        assert results['NM_000061.2:c.588_588+1insCTACATAG']['hgvs_transcript_variant'] == 'NM_000061.2:c.588_588+1insCTACATAG'
        assert results['NM_000061.2:c.588_588+1insCTACATAG']['hgvs_refseqgene_variant'] == 'NG_009616.1:g.29052_29053insCTACATAG'
        assert results['NM_000061.2:c.588_588+1insCTACATAG']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.100617160_100617161insCTATGTAG', 'vcf': {'chr': 'chrX', 'ref': 'C', 'pos': '100617160', 'alt': u'CCTATGTAG'}}
        assert results['NM_000061.2:c.588_588+1insCTACATAG']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.101362172_101362173insCTATGTAG', 'vcf': {'chr': 'chrX', 'ref': 'C', 'pos': '101362172', 'alt': u'CCTATGTAG'}}
        assert results['NM_000061.2:c.588_588+1insCTACATAG']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.100617160_100617161insCTATGTAG', 'vcf': {'chr': 'X', 'ref': 'C', 'pos': '100617160', 'alt': u'CCTATGTAG'}}
        assert results['NM_000061.2:c.588_588+1insCTACATAG']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.101362172_101362173insCTATGTAG', 'vcf': {'chr': 'X', 'ref': 'C', 'pos': '101362172', 'alt': u'CCTATGTAG'}}


    def test_variant333(self):
        variant = 'NM_000061.2:c.588_588+1insCTACATAG'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000061.2:c.588_588+1insCTACATAG' in results.keys()
        assert results['NM_000061.2:c.588_588+1insCTACATAG']['hgvs_lrg_transcript_variant'] == 'LRG_128t1:c.588_588+1insCTACATAG'
        assert results['NM_000061.2:c.588_588+1insCTACATAG']['refseqgene_context_intronic_sequence'] == 'NG_009616.1(NM_000061.2):c.588_588+1insCTACATAG'
        assert results['NM_000061.2:c.588_588+1insCTACATAG']['alt_genomic_loci'] == [{'grch37': {'hgvs_genomic_description': 'NW_004070883.1:g.43848_43849insCTATGTAG', 'vcf': {'chr': 'HG1439_PATCH', 'ref': 'C', 'pos': '43848', 'alt': u'CCTATGTAG'}}}]
        assert results['NM_000061.2:c.588_588+1insCTACATAG']['gene_symbol'] == 'BTK'
        assert results['NM_000061.2:c.588_588+1insCTACATAG']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000052.1(LRG_128p1):p.?', 'slr': 'NP_000052.1:p.?'}
        assert results['NM_000061.2:c.588_588+1insCTACATAG']['submitted_variant'] == 'NM_000061.2:c.588_588+1insCTACATAG'
        assert results['NM_000061.2:c.588_588+1insCTACATAG']['genome_context_intronic_sequence'] == 'NC_000023.10(NM_000061.2):c.588_588+1insCTACATAG'
        assert results['NM_000061.2:c.588_588+1insCTACATAG']['hgvs_lrg_variant'] == 'LRG_128:g.29052_29053insCTACATAG'
        assert results['NM_000061.2:c.588_588+1insCTACATAG']['hgvs_transcript_variant'] == 'NM_000061.2:c.588_588+1insCTACATAG'
        assert results['NM_000061.2:c.588_588+1insCTACATAG']['hgvs_refseqgene_variant'] == 'NG_009616.1:g.29052_29053insCTACATAG'
        assert results['NM_000061.2:c.588_588+1insCTACATAG']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.100617160_100617161insCTATGTAG', 'vcf': {'chr': 'chrX', 'ref': 'C', 'pos': '100617160', 'alt': u'CCTATGTAG'}}
        assert results['NM_000061.2:c.588_588+1insCTACATAG']['primary_assembly_loci']['hg38'] == {'hgvs_genomic_description': 'NC_000023.11:g.101362172_101362173insCTATGTAG', 'vcf': {'chr': 'chrX', 'ref': 'C', 'pos': '101362172', 'alt': u'CCTATGTAG'}}
        assert results['NM_000061.2:c.588_588+1insCTACATAG']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.100617160_100617161insCTATGTAG', 'vcf': {'chr': 'X', 'ref': 'C', 'pos': '100617160', 'alt': u'CCTATGTAG'}}
        assert results['NM_000061.2:c.588_588+1insCTACATAG']['primary_assembly_loci']['grch38'] == {'hgvs_genomic_description': 'NC_000023.11:g.101362172_101362173insCTATGTAG', 'vcf': {'chr': 'X', 'ref': 'C', 'pos': '101362172', 'alt': u'CCTATGTAG'}}


    def test_variant334(self):
        variant = 'NM_000061.2:c.588_589insCTACATAG'
        results = vv.validator(variant, 'GRCh37', 'all')
        print results

        assert results['flag'] == 'gene_variant'
        assert 'NM_000061.2:c.588_589insCTACATAG' in results.keys()
        assert results['NM_000061.2:c.588_589insCTACATAG']['hgvs_lrg_transcript_variant'] == 'LRG_128t1:c.588_589insCTACATAG'
        assert results['NM_000061.2:c.588_589insCTACATAG']['refseqgene_context_intronic_sequence'] == ''
        assert results['NM_000061.2:c.588_589insCTACATAG']['alt_genomic_loci'] == []
        assert results['NM_000061.2:c.588_589insCTACATAG']['gene_symbol'] == 'BTK'
        assert results['NM_000061.2:c.588_589insCTACATAG']['hgvs_predicted_protein_consequence'] == {'tlr': 'NP_000052.1(LRG_128p1):p.(Ile197LeufsTer5)', 'slr': 'NP_000052.1:p.(I197Lfs*5)'}
        assert results['NM_000061.2:c.588_589insCTACATAG']['submitted_variant'] == 'NM_000061.2:c.588_589insCTACATAG'
        assert results['NM_000061.2:c.588_589insCTACATAG']['genome_context_intronic_sequence'] == ''
        assert results['NM_000061.2:c.588_589insCTACATAG']['hgvs_lrg_variant'] == ''
        assert results['NM_000061.2:c.588_589insCTACATAG']['hgvs_transcript_variant'] == 'NM_000061.2:c.588_589insCTACATAG'
        assert results['NM_000061.2:c.588_589insCTACATAG']['hgvs_refseqgene_variant'] == ''
        assert results['NM_000061.2:c.588_589insCTACATAG']['primary_assembly_loci']['hg19'] == {'hgvs_genomic_description': 'NC_000023.10:g.100615743_100617161delinsTCTATGTAGC', 'vcf': {'chr': 'chrX', 'ref': 'GTTAGGAGAAAAGGTAGGAGGGTTTGTCAAGATACCAAGCACTCTTCTCTTCTCTCCCAACTCTCTGGCTTACTCAAGACACCCAAATCAGGCATACTAAAATATTACTCAGCAGTCATTCAACAACCATTTTTAAGCACCAGTGCAGGAGTTCTCAGCCTTGCACACATATAAAGACCATGTATGGAACTTTTAAATTCCAATGTACTTTCGGAGGCCAAGGCGGGCGGATCAGTTGAGGCCAGGAGTTCGAGACCAGCCTGGCCAACGTGACGAAACCCCATCTCTACTAAAAATGCAAAAATCAGCTGGGCATGGTAGTGTGTGCATATAGCCCCAGCTACTCAGGAGGCTGAGGCAGGAGAATAACTTGACCCCAGGAGGCGGAGGTTGTAGTGAGCCAAGACCACGCCACTGCACTCCAGCCTGGGTGACAGAGTGAGACTGTCTCAAAAAAAAACCCACAAAAAACAAAAAACCAATTCCAATACCTAGTCAGTTTCCTCACAGACCAATTACATCAAAATCAAACTCTCAGGAATGGGACCCAAACATTACTATTTTTAAAGCTCACTAGACAAAAACCATTTATAGCTAAGGTCAGGAAACCGGCTTGGCACTAAACTTGTACGTGAATCTACTAAGTGGCTCAGAACCTTGGTTTCCTTCTTTGTAAAATGAGTATAATAATACCTGCTCTACTTACTTACAATATGTGAGAAAGGGCTTTCTAGCCCTCAAGAAGGAACCAAAAAAAAAAAAAAAACTTCTGAAGTGTTAGTGATAGGTGGTGTTAGTGCTAAGTGTTGAGTATGTTGGTATTAAGTGTTAAATTCTTCTAACTTTACTGTATGTTTGAAAATACTTTCCAGCTGGGCACGGTGGCTCACGCCTGTAATCCCAGCACTTTAGGAGGCCGAGGCGGGTTCGAGACCAGCCTGAACAACATGGTGACACCCCCATCTCTACTAAAAATACAAAAATTAGCTGGGCGTGGTGGTGCATGCCTGTAATCCCAGCTACTCAGGAGGCTGAGACAGGAGAATCACTTGAATCTGGGAGGCGGAGGTTGCAGTGAGCCGAGATCATGCCATTACACTCGAACCTGGGCGACAGAGCGAGACTCCGTCTCAAAACAAAAACAAAAGCAAAAACAAAACAACAACAAAAACTTTACATTAAAAAAATCAGGTTTTGTTCTAAACAGGTGATTGGATTACATGGTTGCTGAGAGCCTTCTATCTTTCCATCGAGGAGGAAATCCTAATTAGAAGAACAAATCCCCCATCTTAGCAAGAATACCAATTAACACTGCCAAGTCCCAGGGTAATTCTAAGACTCTAGTGTGTTCTTAGGGCTTGACTATAAGTTTCCATTTAAGCAGTGGCAGCACCCAGTTTCCCTGTATAC', 'pos': '100615751', 'alt': 'G'}}
        assert 'hg38' not in results['NM_000061.2:c.588_589insCTACATAG']['primary_assembly_loci'].keys()
        assert results['NM_000061.2:c.588_589insCTACATAG']['primary_assembly_loci']['grch37'] == {'hgvs_genomic_description': 'NC_000023.10:g.100615743_100617161delinsTCTATGTAGC', 'vcf': {'chr': 'X', 'ref': 'GTTAGGAGAAAAGGTAGGAGGGTTTGTCAAGATACCAAGCACTCTTCTCTTCTCTCCCAACTCTCTGGCTTACTCAAGACACCCAAATCAGGCATACTAAAATATTACTCAGCAGTCATTCAACAACCATTTTTAAGCACCAGTGCAGGAGTTCTCAGCCTTGCACACATATAAAGACCATGTATGGAACTTTTAAATTCCAATGTACTTTCGGAGGCCAAGGCGGGCGGATCAGTTGAGGCCAGGAGTTCGAGACCAGCCTGGCCAACGTGACGAAACCCCATCTCTACTAAAAATGCAAAAATCAGCTGGGCATGGTAGTGTGTGCATATAGCCCCAGCTACTCAGGAGGCTGAGGCAGGAGAATAACTTGACCCCAGGAGGCGGAGGTTGTAGTGAGCCAAGACCACGCCACTGCACTCCAGCCTGGGTGACAGAGTGAGACTGTCTCAAAAAAAAACCCACAAAAAACAAAAAACCAATTCCAATACCTAGTCAGTTTCCTCACAGACCAATTACATCAAAATCAAACTCTCAGGAATGGGACCCAAACATTACTATTTTTAAAGCTCACTAGACAAAAACCATTTATAGCTAAGGTCAGGAAACCGGCTTGGCACTAAACTTGTACGTGAATCTACTAAGTGGCTCAGAACCTTGGTTTCCTTCTTTGTAAAATGAGTATAATAATACCTGCTCTACTTACTTACAATATGTGAGAAAGGGCTTTCTAGCCCTCAAGAAGGAACCAAAAAAAAAAAAAAAACTTCTGAAGTGTTAGTGATAGGTGGTGTTAGTGCTAAGTGTTGAGTATGTTGGTATTAAGTGTTAAATTCTTCTAACTTTACTGTATGTTTGAAAATACTTTCCAGCTGGGCACGGTGGCTCACGCCTGTAATCCCAGCACTTTAGGAGGCCGAGGCGGGTTCGAGACCAGCCTGAACAACATGGTGACACCCCCATCTCTACTAAAAATACAAAAATTAGCTGGGCGTGGTGGTGCATGCCTGTAATCCCAGCTACTCAGGAGGCTGAGACAGGAGAATCACTTGAATCTGGGAGGCGGAGGTTGCAGTGAGCCGAGATCATGCCATTACACTCGAACCTGGGCGACAGAGCGAGACTCCGTCTCAAAACAAAAACAAAAGCAAAAACAAAACAACAACAAAAACTTTACATTAAAAAAATCAGGTTTTGTTCTAAACAGGTGATTGGATTACATGGTTGCTGAGAGCCTTCTATCTTTCCATCGAGGAGGAAATCCTAATTAGAAGAACAAATCCCCCATCTTAGCAAGAATACCAATTAACACTGCCAAGTCCCAGGGTAATTCTAAGACTCTAGTGTGTTCTTAGGGCTTGACTATAAGTTTCCATTTAAGCAGTGGCAGCACCCAGTTTCCCTGTATAC', 'pos': '100615751', 'alt': 'G'}}
        assert 'grch38' not in results['NM_000061.2:c.588_589insCTACATAG']['primary_assembly_loci'].keys()
